// @cdktf/provider-aws/lib/data-aws-location-tracker-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLocationTrackerAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/location_tracker_association aws_location_tracker_association}
*/
class DataAwsLocationTrackerAssociation extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsLocationTrackerAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsLocationTrackerAssociation to import
    * @param importFromId The id of the existing DataAwsLocationTrackerAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/location_tracker_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsLocationTrackerAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_location_tracker_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/location_tracker_association aws_location_tracker_association} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLocationTrackerAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_location_tracker_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._consumerArn = config.consumerArn;
        this._id = config.id;
        this._trackerName = config.trackerName;
    }
    get consumerArn() {
        return this.getStringAttribute('consumer_arn');
    }
    set consumerArn(value) {
        this._consumerArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get consumerArnInput() {
        return this._consumerArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get trackerName() {
        return this.getStringAttribute('tracker_name');
    }
    set trackerName(value) {
        this._trackerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get trackerNameInput() {
        return this._trackerName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            consumer_arn: cdktf.stringToTerraform(this._consumerArn),
            id: cdktf.stringToTerraform(this._id),
            tracker_name: cdktf.stringToTerraform(this._trackerName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            consumer_arn: {
                value: cdktf.stringToHclTerraform(this._consumerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tracker_name: {
                value: cdktf.stringToHclTerraform(this._trackerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsLocationTrackerAssociation = DataAwsLocationTrackerAssociation;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLocationTrackerAssociation[_a] = { fqn: "@cdktf/provider-aws.dataAwsLocationTrackerAssociation.DataAwsLocationTrackerAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLocationTrackerAssociation.tfResourceType = "aws_location_tracker_association";
