// @cdktf/provider-aws/lib/auditmanager-control/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditmanagerControl = exports.AuditmanagerControlControlMappingSourcesList = exports.AuditmanagerControlControlMappingSourcesOutputReference = exports.auditmanagerControlControlMappingSourcesToHclTerraform = exports.auditmanagerControlControlMappingSourcesToTerraform = exports.AuditmanagerControlControlMappingSourcesSourceKeywordList = exports.AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference = exports.auditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform = exports.auditmanagerControlControlMappingSourcesSourceKeywordToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function auditmanagerControlControlMappingSourcesSourceKeywordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        keyword_input_type: cdktf.stringToTerraform(struct.keywordInputType),
        keyword_value: cdktf.stringToTerraform(struct.keywordValue),
    };
}
exports.auditmanagerControlControlMappingSourcesSourceKeywordToTerraform = auditmanagerControlControlMappingSourcesSourceKeywordToTerraform;
function auditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        keyword_input_type: {
            value: cdktf.stringToHclTerraform(struct.keywordInputType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        keyword_value: {
            value: cdktf.stringToHclTerraform(struct.keywordValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform = auditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform;
class AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keywordInputType !== undefined) {
            hasAnyValues = true;
            internalValueResult.keywordInputType = this._keywordInputType;
        }
        if (this._keywordValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.keywordValue = this._keywordValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._keywordInputType = undefined;
            this._keywordValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._keywordInputType = value.keywordInputType;
            this._keywordValue = value.keywordValue;
        }
    }
    get keywordInputType() {
        return this.getStringAttribute('keyword_input_type');
    }
    set keywordInputType(value) {
        this._keywordInputType = value;
    }
    // Temporarily expose input value. Use with caution.
    get keywordInputTypeInput() {
        return this._keywordInputType;
    }
    get keywordValue() {
        return this.getStringAttribute('keyword_value');
    }
    set keywordValue(value) {
        this._keywordValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get keywordValueInput() {
        return this._keywordValue;
    }
}
exports.AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference = AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference[_a] = { fqn: "@cdktf/provider-aws.auditmanagerControl.AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference", version: "19.15.0" };
class AuditmanagerControlControlMappingSourcesSourceKeywordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerControlControlMappingSourcesSourceKeywordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerControlControlMappingSourcesSourceKeywordList = AuditmanagerControlControlMappingSourcesSourceKeywordList;
_b = JSII_RTTI_SYMBOL_1;
AuditmanagerControlControlMappingSourcesSourceKeywordList[_b] = { fqn: "@cdktf/provider-aws.auditmanagerControl.AuditmanagerControlControlMappingSourcesSourceKeywordList", version: "19.15.0" };
function auditmanagerControlControlMappingSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_description: cdktf.stringToTerraform(struct.sourceDescription),
        source_frequency: cdktf.stringToTerraform(struct.sourceFrequency),
        source_name: cdktf.stringToTerraform(struct.sourceName),
        source_set_up_option: cdktf.stringToTerraform(struct.sourceSetUpOption),
        source_type: cdktf.stringToTerraform(struct.sourceType),
        troubleshooting_text: cdktf.stringToTerraform(struct.troubleshootingText),
        source_keyword: cdktf.listMapper(auditmanagerControlControlMappingSourcesSourceKeywordToTerraform, true)(struct.sourceKeyword),
    };
}
exports.auditmanagerControlControlMappingSourcesToTerraform = auditmanagerControlControlMappingSourcesToTerraform;
function auditmanagerControlControlMappingSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_description: {
            value: cdktf.stringToHclTerraform(struct.sourceDescription),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_frequency: {
            value: cdktf.stringToHclTerraform(struct.sourceFrequency),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_name: {
            value: cdktf.stringToHclTerraform(struct.sourceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_set_up_option: {
            value: cdktf.stringToHclTerraform(struct.sourceSetUpOption),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_type: {
            value: cdktf.stringToHclTerraform(struct.sourceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        troubleshooting_text: {
            value: cdktf.stringToHclTerraform(struct.troubleshootingText),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_keyword: {
            value: cdktf.listMapperHcl(auditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform, true)(struct.sourceKeyword),
            isBlock: true,
            type: "list",
            storageClassType: "AuditmanagerControlControlMappingSourcesSourceKeywordList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerControlControlMappingSourcesToHclTerraform = auditmanagerControlControlMappingSourcesToHclTerraform;
class AuditmanagerControlControlMappingSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // source_keyword - computed: false, optional: true, required: false
        this._sourceKeyword = new AuditmanagerControlControlMappingSourcesSourceKeywordList(this, "source_keyword", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceDescription !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceDescription = this._sourceDescription;
        }
        if (this._sourceFrequency !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFrequency = this._sourceFrequency;
        }
        if (this._sourceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceName = this._sourceName;
        }
        if (this._sourceSetUpOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceSetUpOption = this._sourceSetUpOption;
        }
        if (this._sourceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceType = this._sourceType;
        }
        if (this._troubleshootingText !== undefined) {
            hasAnyValues = true;
            internalValueResult.troubleshootingText = this._troubleshootingText;
        }
        if (this._sourceKeyword?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceKeyword = this._sourceKeyword?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._sourceDescription = undefined;
            this._sourceFrequency = undefined;
            this._sourceName = undefined;
            this._sourceSetUpOption = undefined;
            this._sourceType = undefined;
            this._troubleshootingText = undefined;
            this._sourceKeyword.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._sourceDescription = value.sourceDescription;
            this._sourceFrequency = value.sourceFrequency;
            this._sourceName = value.sourceName;
            this._sourceSetUpOption = value.sourceSetUpOption;
            this._sourceType = value.sourceType;
            this._troubleshootingText = value.troubleshootingText;
            this._sourceKeyword.internalValue = value.sourceKeyword;
        }
    }
    get sourceDescription() {
        return this.getStringAttribute('source_description');
    }
    set sourceDescription(value) {
        this._sourceDescription = value;
    }
    resetSourceDescription() {
        this._sourceDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDescriptionInput() {
        return this._sourceDescription;
    }
    get sourceFrequency() {
        return this.getStringAttribute('source_frequency');
    }
    set sourceFrequency(value) {
        this._sourceFrequency = value;
    }
    resetSourceFrequency() {
        this._sourceFrequency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFrequencyInput() {
        return this._sourceFrequency;
    }
    // source_id - computed: true, optional: false, required: false
    get sourceId() {
        return this.getStringAttribute('source_id');
    }
    get sourceName() {
        return this.getStringAttribute('source_name');
    }
    set sourceName(value) {
        this._sourceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceNameInput() {
        return this._sourceName;
    }
    get sourceSetUpOption() {
        return this.getStringAttribute('source_set_up_option');
    }
    set sourceSetUpOption(value) {
        this._sourceSetUpOption = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceSetUpOptionInput() {
        return this._sourceSetUpOption;
    }
    get sourceType() {
        return this.getStringAttribute('source_type');
    }
    set sourceType(value) {
        this._sourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceTypeInput() {
        return this._sourceType;
    }
    get troubleshootingText() {
        return this.getStringAttribute('troubleshooting_text');
    }
    set troubleshootingText(value) {
        this._troubleshootingText = value;
    }
    resetTroubleshootingText() {
        this._troubleshootingText = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get troubleshootingTextInput() {
        return this._troubleshootingText;
    }
    get sourceKeyword() {
        return this._sourceKeyword;
    }
    putSourceKeyword(value) {
        this._sourceKeyword.internalValue = value;
    }
    resetSourceKeyword() {
        this._sourceKeyword.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceKeywordInput() {
        return this._sourceKeyword.internalValue;
    }
}
exports.AuditmanagerControlControlMappingSourcesOutputReference = AuditmanagerControlControlMappingSourcesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AuditmanagerControlControlMappingSourcesOutputReference[_c] = { fqn: "@cdktf/provider-aws.auditmanagerControl.AuditmanagerControlControlMappingSourcesOutputReference", version: "19.15.0" };
class AuditmanagerControlControlMappingSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerControlControlMappingSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerControlControlMappingSourcesList = AuditmanagerControlControlMappingSourcesList;
_d = JSII_RTTI_SYMBOL_1;
AuditmanagerControlControlMappingSourcesList[_d] = { fqn: "@cdktf/provider-aws.auditmanagerControl.AuditmanagerControlControlMappingSourcesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_control aws_auditmanager_control}
*/
class AuditmanagerControl extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AuditmanagerControl resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AuditmanagerControl to import
    * @param importFromId The id of the existing AuditmanagerControl that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_control#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AuditmanagerControl to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_auditmanager_control", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_control aws_auditmanager_control} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AuditmanagerControlConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_auditmanager_control',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // control_mapping_sources - computed: false, optional: true, required: false
        this._controlMappingSources = new AuditmanagerControlControlMappingSourcesList(this, "control_mapping_sources", true);
        this._actionPlanInstructions = config.actionPlanInstructions;
        this._actionPlanTitle = config.actionPlanTitle;
        this._description = config.description;
        this._name = config.name;
        this._tags = config.tags;
        this._testingInformation = config.testingInformation;
        this._controlMappingSources.internalValue = config.controlMappingSources;
    }
    get actionPlanInstructions() {
        return this.getStringAttribute('action_plan_instructions');
    }
    set actionPlanInstructions(value) {
        this._actionPlanInstructions = value;
    }
    resetActionPlanInstructions() {
        this._actionPlanInstructions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionPlanInstructionsInput() {
        return this._actionPlanInstructions;
    }
    get actionPlanTitle() {
        return this.getStringAttribute('action_plan_title');
    }
    set actionPlanTitle(value) {
        this._actionPlanTitle = value;
    }
    resetActionPlanTitle() {
        this._actionPlanTitle = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionPlanTitleInput() {
        return this._actionPlanTitle;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get testingInformation() {
        return this.getStringAttribute('testing_information');
    }
    set testingInformation(value) {
        this._testingInformation = value;
    }
    resetTestingInformation() {
        this._testingInformation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get testingInformationInput() {
        return this._testingInformation;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    get controlMappingSources() {
        return this._controlMappingSources;
    }
    putControlMappingSources(value) {
        this._controlMappingSources.internalValue = value;
    }
    resetControlMappingSources() {
        this._controlMappingSources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlMappingSourcesInput() {
        return this._controlMappingSources.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action_plan_instructions: cdktf.stringToTerraform(this._actionPlanInstructions),
            action_plan_title: cdktf.stringToTerraform(this._actionPlanTitle),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            testing_information: cdktf.stringToTerraform(this._testingInformation),
            control_mapping_sources: cdktf.listMapper(auditmanagerControlControlMappingSourcesToTerraform, true)(this._controlMappingSources.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action_plan_instructions: {
                value: cdktf.stringToHclTerraform(this._actionPlanInstructions),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            action_plan_title: {
                value: cdktf.stringToHclTerraform(this._actionPlanTitle),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            testing_information: {
                value: cdktf.stringToHclTerraform(this._testingInformation),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            control_mapping_sources: {
                value: cdktf.listMapperHcl(auditmanagerControlControlMappingSourcesToHclTerraform, true)(this._controlMappingSources.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "AuditmanagerControlControlMappingSourcesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AuditmanagerControl = AuditmanagerControl;
_e = JSII_RTTI_SYMBOL_1;
AuditmanagerControl[_e] = { fqn: "@cdktf/provider-aws.auditmanagerControl.AuditmanagerControl", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AuditmanagerControl.tfResourceType = "aws_auditmanager_control";
