// @cdktf/provider-aws/lib/data-aws-mq-broker-engine-types/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsMqBrokerEngineTypes = exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesList = exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference = exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesToHclTerraform = exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesToTerraform = exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList = exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference = exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToHclTerraform = exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToTerraform = dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToTerraform;
function dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToHclTerraform = dataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsToHclTerraform;
class DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference = DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsMqBrokerEngineTypes.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference", version: "19.15.0" };
class DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList = DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsMqBrokerEngineTypes.DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList", version: "19.15.0" };
function dataAwsMqBrokerEngineTypesBrokerEngineTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesToTerraform = dataAwsMqBrokerEngineTypesBrokerEngineTypesToTerraform;
function dataAwsMqBrokerEngineTypesBrokerEngineTypesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMqBrokerEngineTypesBrokerEngineTypesToHclTerraform = dataAwsMqBrokerEngineTypesBrokerEngineTypesToHclTerraform;
class DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // engine_versions - computed: true, optional: false, required: false
        this._engineVersions = new DataAwsMqBrokerEngineTypesBrokerEngineTypesEngineVersionsList(this, "engine_versions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // engine_type - computed: true, optional: false, required: false
    get engineType() {
        return this.getStringAttribute('engine_type');
    }
    get engineVersions() {
        return this._engineVersions;
    }
}
exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference = DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsMqBrokerEngineTypes.DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference", version: "19.15.0" };
class DataAwsMqBrokerEngineTypesBrokerEngineTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMqBrokerEngineTypesBrokerEngineTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMqBrokerEngineTypesBrokerEngineTypesList = DataAwsMqBrokerEngineTypesBrokerEngineTypesList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsMqBrokerEngineTypesBrokerEngineTypesList[_d] = { fqn: "@cdktf/provider-aws.dataAwsMqBrokerEngineTypes.DataAwsMqBrokerEngineTypesBrokerEngineTypesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/mq_broker_engine_types aws_mq_broker_engine_types}
*/
class DataAwsMqBrokerEngineTypes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsMqBrokerEngineTypes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsMqBrokerEngineTypes to import
    * @param importFromId The id of the existing DataAwsMqBrokerEngineTypes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/mq_broker_engine_types#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsMqBrokerEngineTypes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_mq_broker_engine_types", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/mq_broker_engine_types aws_mq_broker_engine_types} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsMqBrokerEngineTypesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_mq_broker_engine_types',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // broker_engine_types - computed: true, optional: false, required: false
        this._brokerEngineTypes = new DataAwsMqBrokerEngineTypesBrokerEngineTypesList(this, "broker_engine_types", false);
        this._engineType = config.engineType;
        this._id = config.id;
    }
    get brokerEngineTypes() {
        return this._brokerEngineTypes;
    }
    get engineType() {
        return this.getStringAttribute('engine_type');
    }
    set engineType(value) {
        this._engineType = value;
    }
    resetEngineType() {
        this._engineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engineTypeInput() {
        return this._engineType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            engine_type: cdktf.stringToTerraform(this._engineType),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            engine_type: {
                value: cdktf.stringToHclTerraform(this._engineType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsMqBrokerEngineTypes = DataAwsMqBrokerEngineTypes;
_e = JSII_RTTI_SYMBOL_1;
DataAwsMqBrokerEngineTypes[_e] = { fqn: "@cdktf/provider-aws.dataAwsMqBrokerEngineTypes.DataAwsMqBrokerEngineTypes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsMqBrokerEngineTypes.tfResourceType = "aws_mq_broker_engine_types";
