// @cdktf/provider-aws/lib/comprehend-entity-recognizer/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComprehendEntityRecognizer = exports.ComprehendEntityRecognizerVpcConfigOutputReference = exports.comprehendEntityRecognizerVpcConfigToHclTerraform = exports.comprehendEntityRecognizerVpcConfigToTerraform = exports.ComprehendEntityRecognizerTimeoutsOutputReference = exports.comprehendEntityRecognizerTimeoutsToHclTerraform = exports.comprehendEntityRecognizerTimeoutsToTerraform = exports.ComprehendEntityRecognizerInputDataConfigOutputReference = exports.comprehendEntityRecognizerInputDataConfigToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigToTerraform = exports.ComprehendEntityRecognizerInputDataConfigEntityTypesList = exports.ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference = exports.comprehendEntityRecognizerInputDataConfigEntityTypesToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigEntityTypesToTerraform = exports.ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference = exports.comprehendEntityRecognizerInputDataConfigEntityListStructToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigEntityListStructToTerraform = exports.ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference = exports.comprehendEntityRecognizerInputDataConfigDocumentsToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigDocumentsToTerraform = exports.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList = exports.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference = exports.comprehendEntityRecognizerInputDataConfigAugmentedManifestsToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigAugmentedManifestsToTerraform = exports.ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference = exports.comprehendEntityRecognizerInputDataConfigAnnotationsToHclTerraform = exports.comprehendEntityRecognizerInputDataConfigAnnotationsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function comprehendEntityRecognizerInputDataConfigAnnotationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
        test_s3_uri: cdktf.stringToTerraform(struct.testS3Uri),
    };
}
exports.comprehendEntityRecognizerInputDataConfigAnnotationsToTerraform = comprehendEntityRecognizerInputDataConfigAnnotationsToTerraform;
function comprehendEntityRecognizerInputDataConfigAnnotationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        test_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.testS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigAnnotationsToHclTerraform = comprehendEntityRecognizerInputDataConfigAnnotationsToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        if (this._testS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.testS3Uri = this._testS3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3Uri = undefined;
            this._testS3Uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3Uri = value.s3Uri;
            this._testS3Uri = value.testS3Uri;
        }
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
    get testS3Uri() {
        return this.getStringAttribute('test_s3_uri');
    }
    set testS3Uri(value) {
        this._testS3Uri = value;
    }
    resetTestS3Uri() {
        this._testS3Uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get testS3UriInput() {
        return this._testS3Uri;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference = ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference[_a] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference", version: "19.15.0" };
function comprehendEntityRecognizerInputDataConfigAugmentedManifestsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        annotation_data_s3_uri: cdktf.stringToTerraform(struct.annotationDataS3Uri),
        attribute_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.attributeNames),
        document_type: cdktf.stringToTerraform(struct.documentType),
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
        source_documents_s3_uri: cdktf.stringToTerraform(struct.sourceDocumentsS3Uri),
        split: cdktf.stringToTerraform(struct.split),
    };
}
exports.comprehendEntityRecognizerInputDataConfigAugmentedManifestsToTerraform = comprehendEntityRecognizerInputDataConfigAugmentedManifestsToTerraform;
function comprehendEntityRecognizerInputDataConfigAugmentedManifestsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        annotation_data_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.annotationDataS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        attribute_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.attributeNames),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        document_type: {
            value: cdktf.stringToHclTerraform(struct.documentType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_documents_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.sourceDocumentsS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        split: {
            value: cdktf.stringToHclTerraform(struct.split),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigAugmentedManifestsToHclTerraform = comprehendEntityRecognizerInputDataConfigAugmentedManifestsToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._annotationDataS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.annotationDataS3Uri = this._annotationDataS3Uri;
        }
        if (this._attributeNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.attributeNames = this._attributeNames;
        }
        if (this._documentType !== undefined) {
            hasAnyValues = true;
            internalValueResult.documentType = this._documentType;
        }
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        if (this._sourceDocumentsS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceDocumentsS3Uri = this._sourceDocumentsS3Uri;
        }
        if (this._split !== undefined) {
            hasAnyValues = true;
            internalValueResult.split = this._split;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._annotationDataS3Uri = undefined;
            this._attributeNames = undefined;
            this._documentType = undefined;
            this._s3Uri = undefined;
            this._sourceDocumentsS3Uri = undefined;
            this._split = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._annotationDataS3Uri = value.annotationDataS3Uri;
            this._attributeNames = value.attributeNames;
            this._documentType = value.documentType;
            this._s3Uri = value.s3Uri;
            this._sourceDocumentsS3Uri = value.sourceDocumentsS3Uri;
            this._split = value.split;
        }
    }
    get annotationDataS3Uri() {
        return this.getStringAttribute('annotation_data_s3_uri');
    }
    set annotationDataS3Uri(value) {
        this._annotationDataS3Uri = value;
    }
    resetAnnotationDataS3Uri() {
        this._annotationDataS3Uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationDataS3UriInput() {
        return this._annotationDataS3Uri;
    }
    get attributeNames() {
        return this.getListAttribute('attribute_names');
    }
    set attributeNames(value) {
        this._attributeNames = value;
    }
    // Temporarily expose input value. Use with caution.
    get attributeNamesInput() {
        return this._attributeNames;
    }
    get documentType() {
        return this.getStringAttribute('document_type');
    }
    set documentType(value) {
        this._documentType = value;
    }
    resetDocumentType() {
        this._documentType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentTypeInput() {
        return this._documentType;
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
    get sourceDocumentsS3Uri() {
        return this.getStringAttribute('source_documents_s3_uri');
    }
    set sourceDocumentsS3Uri(value) {
        this._sourceDocumentsS3Uri = value;
    }
    resetSourceDocumentsS3Uri() {
        this._sourceDocumentsS3Uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDocumentsS3UriInput() {
        return this._sourceDocumentsS3Uri;
    }
    get split() {
        return this.getStringAttribute('split');
    }
    set split(value) {
        this._split = value;
    }
    resetSplit() {
        this._split = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get splitInput() {
        return this._split;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference = ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference[_b] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference", version: "19.15.0" };
class ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComprehendEntityRecognizerInputDataConfigAugmentedManifestsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList = ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList;
_c = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList[_c] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList", version: "19.15.0" };
function comprehendEntityRecognizerInputDataConfigDocumentsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_format: cdktf.stringToTerraform(struct.inputFormat),
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
        test_s3_uri: cdktf.stringToTerraform(struct.testS3Uri),
    };
}
exports.comprehendEntityRecognizerInputDataConfigDocumentsToTerraform = comprehendEntityRecognizerInputDataConfigDocumentsToTerraform;
function comprehendEntityRecognizerInputDataConfigDocumentsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        input_format: {
            value: cdktf.stringToHclTerraform(struct.inputFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        test_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.testS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigDocumentsToHclTerraform = comprehendEntityRecognizerInputDataConfigDocumentsToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputFormat = this._inputFormat;
        }
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        if (this._testS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.testS3Uri = this._testS3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._inputFormat = undefined;
            this._s3Uri = undefined;
            this._testS3Uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._inputFormat = value.inputFormat;
            this._s3Uri = value.s3Uri;
            this._testS3Uri = value.testS3Uri;
        }
    }
    get inputFormat() {
        return this.getStringAttribute('input_format');
    }
    set inputFormat(value) {
        this._inputFormat = value;
    }
    resetInputFormat() {
        this._inputFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputFormatInput() {
        return this._inputFormat;
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
    get testS3Uri() {
        return this.getStringAttribute('test_s3_uri');
    }
    set testS3Uri(value) {
        this._testS3Uri = value;
    }
    resetTestS3Uri() {
        this._testS3Uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get testS3UriInput() {
        return this._testS3Uri;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference = ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference[_d] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference", version: "19.15.0" };
function comprehendEntityRecognizerInputDataConfigEntityListStructToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
    };
}
exports.comprehendEntityRecognizerInputDataConfigEntityListStructToTerraform = comprehendEntityRecognizerInputDataConfigEntityListStructToTerraform;
function comprehendEntityRecognizerInputDataConfigEntityListStructToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigEntityListStructToHclTerraform = comprehendEntityRecognizerInputDataConfigEntityListStructToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3Uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3Uri = value.s3Uri;
        }
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference = ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference[_e] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference", version: "19.15.0" };
function comprehendEntityRecognizerInputDataConfigEntityTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.comprehendEntityRecognizerInputDataConfigEntityTypesToTerraform = comprehendEntityRecognizerInputDataConfigEntityTypesToTerraform;
function comprehendEntityRecognizerInputDataConfigEntityTypesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigEntityTypesToHclTerraform = comprehendEntityRecognizerInputDataConfigEntityTypesToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference = ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference[_f] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference", version: "19.15.0" };
class ComprehendEntityRecognizerInputDataConfigEntityTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComprehendEntityRecognizerInputDataConfigEntityTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComprehendEntityRecognizerInputDataConfigEntityTypesList = ComprehendEntityRecognizerInputDataConfigEntityTypesList;
_g = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigEntityTypesList[_g] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigEntityTypesList", version: "19.15.0" };
function comprehendEntityRecognizerInputDataConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_format: cdktf.stringToTerraform(struct.dataFormat),
        annotations: comprehendEntityRecognizerInputDataConfigAnnotationsToTerraform(struct.annotations),
        augmented_manifests: cdktf.listMapper(comprehendEntityRecognizerInputDataConfigAugmentedManifestsToTerraform, true)(struct.augmentedManifests),
        documents: comprehendEntityRecognizerInputDataConfigDocumentsToTerraform(struct.documents),
        entity_list: comprehendEntityRecognizerInputDataConfigEntityListStructToTerraform(struct.entityList),
        entity_types: cdktf.listMapper(comprehendEntityRecognizerInputDataConfigEntityTypesToTerraform, true)(struct.entityTypes),
    };
}
exports.comprehendEntityRecognizerInputDataConfigToTerraform = comprehendEntityRecognizerInputDataConfigToTerraform;
function comprehendEntityRecognizerInputDataConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_format: {
            value: cdktf.stringToHclTerraform(struct.dataFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        annotations: {
            value: comprehendEntityRecognizerInputDataConfigAnnotationsToHclTerraform(struct.annotations),
            isBlock: true,
            type: "list",
            storageClassType: "ComprehendEntityRecognizerInputDataConfigAnnotationsList",
        },
        augmented_manifests: {
            value: cdktf.listMapperHcl(comprehendEntityRecognizerInputDataConfigAugmentedManifestsToHclTerraform, true)(struct.augmentedManifests),
            isBlock: true,
            type: "set",
            storageClassType: "ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList",
        },
        documents: {
            value: comprehendEntityRecognizerInputDataConfigDocumentsToHclTerraform(struct.documents),
            isBlock: true,
            type: "list",
            storageClassType: "ComprehendEntityRecognizerInputDataConfigDocumentsList",
        },
        entity_list: {
            value: comprehendEntityRecognizerInputDataConfigEntityListStructToHclTerraform(struct.entityList),
            isBlock: true,
            type: "list",
            storageClassType: "ComprehendEntityRecognizerInputDataConfigEntityListStructList",
        },
        entity_types: {
            value: cdktf.listMapperHcl(comprehendEntityRecognizerInputDataConfigEntityTypesToHclTerraform, true)(struct.entityTypes),
            isBlock: true,
            type: "set",
            storageClassType: "ComprehendEntityRecognizerInputDataConfigEntityTypesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerInputDataConfigToHclTerraform = comprehendEntityRecognizerInputDataConfigToHclTerraform;
class ComprehendEntityRecognizerInputDataConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // annotations - computed: false, optional: true, required: false
        this._annotations = new ComprehendEntityRecognizerInputDataConfigAnnotationsOutputReference(this, "annotations");
        // augmented_manifests - computed: false, optional: true, required: false
        this._augmentedManifests = new ComprehendEntityRecognizerInputDataConfigAugmentedManifestsList(this, "augmented_manifests", true);
        // documents - computed: false, optional: true, required: false
        this._documents = new ComprehendEntityRecognizerInputDataConfigDocumentsOutputReference(this, "documents");
        // entity_list - computed: false, optional: true, required: false
        this._entityList = new ComprehendEntityRecognizerInputDataConfigEntityListStructOutputReference(this, "entity_list");
        // entity_types - computed: false, optional: false, required: true
        this._entityTypes = new ComprehendEntityRecognizerInputDataConfigEntityTypesList(this, "entity_types", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataFormat = this._dataFormat;
        }
        if (this._annotations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.annotations = this._annotations?.internalValue;
        }
        if (this._augmentedManifests?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.augmentedManifests = this._augmentedManifests?.internalValue;
        }
        if (this._documents?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.documents = this._documents?.internalValue;
        }
        if (this._entityList?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityList = this._entityList?.internalValue;
        }
        if (this._entityTypes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityTypes = this._entityTypes?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataFormat = undefined;
            this._annotations.internalValue = undefined;
            this._augmentedManifests.internalValue = undefined;
            this._documents.internalValue = undefined;
            this._entityList.internalValue = undefined;
            this._entityTypes.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataFormat = value.dataFormat;
            this._annotations.internalValue = value.annotations;
            this._augmentedManifests.internalValue = value.augmentedManifests;
            this._documents.internalValue = value.documents;
            this._entityList.internalValue = value.entityList;
            this._entityTypes.internalValue = value.entityTypes;
        }
    }
    get dataFormat() {
        return this.getStringAttribute('data_format');
    }
    set dataFormat(value) {
        this._dataFormat = value;
    }
    resetDataFormat() {
        this._dataFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataFormatInput() {
        return this._dataFormat;
    }
    get annotations() {
        return this._annotations;
    }
    putAnnotations(value) {
        this._annotations.internalValue = value;
    }
    resetAnnotations() {
        this._annotations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations.internalValue;
    }
    get augmentedManifests() {
        return this._augmentedManifests;
    }
    putAugmentedManifests(value) {
        this._augmentedManifests.internalValue = value;
    }
    resetAugmentedManifests() {
        this._augmentedManifests.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get augmentedManifestsInput() {
        return this._augmentedManifests.internalValue;
    }
    get documents() {
        return this._documents;
    }
    putDocuments(value) {
        this._documents.internalValue = value;
    }
    resetDocuments() {
        this._documents.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentsInput() {
        return this._documents.internalValue;
    }
    get entityList() {
        return this._entityList;
    }
    putEntityList(value) {
        this._entityList.internalValue = value;
    }
    resetEntityList() {
        this._entityList.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entityListInput() {
        return this._entityList.internalValue;
    }
    get entityTypes() {
        return this._entityTypes;
    }
    putEntityTypes(value) {
        this._entityTypes.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get entityTypesInput() {
        return this._entityTypes.internalValue;
    }
}
exports.ComprehendEntityRecognizerInputDataConfigOutputReference = ComprehendEntityRecognizerInputDataConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerInputDataConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerInputDataConfigOutputReference", version: "19.15.0" };
function comprehendEntityRecognizerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.comprehendEntityRecognizerTimeoutsToTerraform = comprehendEntityRecognizerTimeoutsToTerraform;
function comprehendEntityRecognizerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerTimeoutsToHclTerraform = comprehendEntityRecognizerTimeoutsToHclTerraform;
class ComprehendEntityRecognizerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComprehendEntityRecognizerTimeoutsOutputReference = ComprehendEntityRecognizerTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerTimeoutsOutputReference", version: "19.15.0" };
function comprehendEntityRecognizerVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.comprehendEntityRecognizerVpcConfigToTerraform = comprehendEntityRecognizerVpcConfigToTerraform;
function comprehendEntityRecognizerVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.comprehendEntityRecognizerVpcConfigToHclTerraform = comprehendEntityRecognizerVpcConfigToHclTerraform;
class ComprehendEntityRecognizerVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnets = value.subnets;
        }
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.ComprehendEntityRecognizerVpcConfigOutputReference = ComprehendEntityRecognizerVpcConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizerVpcConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizerVpcConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/comprehend_entity_recognizer aws_comprehend_entity_recognizer}
*/
class ComprehendEntityRecognizer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComprehendEntityRecognizer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComprehendEntityRecognizer to import
    * @param importFromId The id of the existing ComprehendEntityRecognizer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/comprehend_entity_recognizer#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComprehendEntityRecognizer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_comprehend_entity_recognizer", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/comprehend_entity_recognizer aws_comprehend_entity_recognizer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComprehendEntityRecognizerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_comprehend_entity_recognizer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // input_data_config - computed: false, optional: false, required: true
        this._inputDataConfig = new ComprehendEntityRecognizerInputDataConfigOutputReference(this, "input_data_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComprehendEntityRecognizerTimeoutsOutputReference(this, "timeouts");
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new ComprehendEntityRecognizerVpcConfigOutputReference(this, "vpc_config");
        this._dataAccessRoleArn = config.dataAccessRoleArn;
        this._id = config.id;
        this._languageCode = config.languageCode;
        this._modelKmsKeyId = config.modelKmsKeyId;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._versionName = config.versionName;
        this._versionNamePrefix = config.versionNamePrefix;
        this._volumeKmsKeyId = config.volumeKmsKeyId;
        this._inputDataConfig.internalValue = config.inputDataConfig;
        this._timeouts.internalValue = config.timeouts;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get dataAccessRoleArn() {
        return this.getStringAttribute('data_access_role_arn');
    }
    set dataAccessRoleArn(value) {
        this._dataAccessRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataAccessRoleArnInput() {
        return this._dataAccessRoleArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get languageCode() {
        return this.getStringAttribute('language_code');
    }
    set languageCode(value) {
        this._languageCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get languageCodeInput() {
        return this._languageCode;
    }
    get modelKmsKeyId() {
        return this.getStringAttribute('model_kms_key_id');
    }
    set modelKmsKeyId(value) {
        this._modelKmsKeyId = value;
    }
    resetModelKmsKeyId() {
        this._modelKmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelKmsKeyIdInput() {
        return this._modelKmsKeyId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get versionName() {
        return this.getStringAttribute('version_name');
    }
    set versionName(value) {
        this._versionName = value;
    }
    resetVersionName() {
        this._versionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionNameInput() {
        return this._versionName;
    }
    get versionNamePrefix() {
        return this.getStringAttribute('version_name_prefix');
    }
    set versionNamePrefix(value) {
        this._versionNamePrefix = value;
    }
    resetVersionNamePrefix() {
        this._versionNamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionNamePrefixInput() {
        return this._versionNamePrefix;
    }
    get volumeKmsKeyId() {
        return this.getStringAttribute('volume_kms_key_id');
    }
    set volumeKmsKeyId(value) {
        this._volumeKmsKeyId = value;
    }
    resetVolumeKmsKeyId() {
        this._volumeKmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeKmsKeyIdInput() {
        return this._volumeKmsKeyId;
    }
    get inputDataConfig() {
        return this._inputDataConfig;
    }
    putInputDataConfig(value) {
        this._inputDataConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputDataConfigInput() {
        return this._inputDataConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            data_access_role_arn: cdktf.stringToTerraform(this._dataAccessRoleArn),
            id: cdktf.stringToTerraform(this._id),
            language_code: cdktf.stringToTerraform(this._languageCode),
            model_kms_key_id: cdktf.stringToTerraform(this._modelKmsKeyId),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            version_name: cdktf.stringToTerraform(this._versionName),
            version_name_prefix: cdktf.stringToTerraform(this._versionNamePrefix),
            volume_kms_key_id: cdktf.stringToTerraform(this._volumeKmsKeyId),
            input_data_config: comprehendEntityRecognizerInputDataConfigToTerraform(this._inputDataConfig.internalValue),
            timeouts: comprehendEntityRecognizerTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_config: comprehendEntityRecognizerVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            data_access_role_arn: {
                value: cdktf.stringToHclTerraform(this._dataAccessRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            language_code: {
                value: cdktf.stringToHclTerraform(this._languageCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            model_kms_key_id: {
                value: cdktf.stringToHclTerraform(this._modelKmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            version_name: {
                value: cdktf.stringToHclTerraform(this._versionName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version_name_prefix: {
                value: cdktf.stringToHclTerraform(this._versionNamePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            volume_kms_key_id: {
                value: cdktf.stringToHclTerraform(this._volumeKmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            input_data_config: {
                value: comprehendEntityRecognizerInputDataConfigToHclTerraform(this._inputDataConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComprehendEntityRecognizerInputDataConfigList",
            },
            timeouts: {
                value: comprehendEntityRecognizerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComprehendEntityRecognizerTimeouts",
            },
            vpc_config: {
                value: comprehendEntityRecognizerVpcConfigToHclTerraform(this._vpcConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComprehendEntityRecognizerVpcConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComprehendEntityRecognizer = ComprehendEntityRecognizer;
_l = JSII_RTTI_SYMBOL_1;
ComprehendEntityRecognizer[_l] = { fqn: "@cdktf/provider-aws.comprehendEntityRecognizer.ComprehendEntityRecognizer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ComprehendEntityRecognizer.tfResourceType = "aws_comprehend_entity_recognizer";
