// @cdktf/provider-aws/lib/data-aws-ecs-task-execution/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcsTaskExecution = exports.DataAwsEcsTaskExecutionPlacementStrategyList = exports.DataAwsEcsTaskExecutionPlacementStrategyOutputReference = exports.dataAwsEcsTaskExecutionPlacementStrategyToHclTerraform = exports.dataAwsEcsTaskExecutionPlacementStrategyToTerraform = exports.DataAwsEcsTaskExecutionPlacementConstraintsList = exports.DataAwsEcsTaskExecutionPlacementConstraintsOutputReference = exports.dataAwsEcsTaskExecutionPlacementConstraintsToHclTerraform = exports.dataAwsEcsTaskExecutionPlacementConstraintsToTerraform = exports.DataAwsEcsTaskExecutionOverridesOutputReference = exports.dataAwsEcsTaskExecutionOverridesToHclTerraform = exports.dataAwsEcsTaskExecutionOverridesToTerraform = exports.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList = exports.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference = exports.dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToHclTerraform = exports.dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToTerraform = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesList = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesToHclTerraform = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesToTerraform = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToHclTerraform = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToTerraform = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList = exports.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToHclTerraform = exports.dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToTerraform = exports.DataAwsEcsTaskExecutionNetworkConfigurationOutputReference = exports.dataAwsEcsTaskExecutionNetworkConfigurationToHclTerraform = exports.dataAwsEcsTaskExecutionNetworkConfigurationToTerraform = exports.DataAwsEcsTaskExecutionCapacityProviderStrategyList = exports.DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference = exports.dataAwsEcsTaskExecutionCapacityProviderStrategyToHclTerraform = exports.dataAwsEcsTaskExecutionCapacityProviderStrategyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEcsTaskExecutionCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.dataAwsEcsTaskExecutionCapacityProviderStrategyToTerraform = dataAwsEcsTaskExecutionCapacityProviderStrategyToTerraform;
function dataAwsEcsTaskExecutionCapacityProviderStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        base: {
            value: cdktf.numberToHclTerraform(struct.base),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        capacity_provider: {
            value: cdktf.stringToHclTerraform(struct.capacityProvider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weight: {
            value: cdktf.numberToHclTerraform(struct.weight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionCapacityProviderStrategyToHclTerraform = dataAwsEcsTaskExecutionCapacityProviderStrategyToHclTerraform;
class DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._base !== undefined) {
            hasAnyValues = true;
            internalValueResult.base = this._base;
        }
        if (this._capacityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProvider = this._capacityProvider;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._base = undefined;
            this._capacityProvider = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._base = value.base;
            this._capacityProvider = value.capacityProvider;
            this._weight = value.weight;
        }
    }
    get base() {
        return this.getNumberAttribute('base');
    }
    set base(value) {
        this._base = value;
    }
    resetBase() {
        this._base = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseInput() {
        return this._base;
    }
    get capacityProvider() {
        return this.getStringAttribute('capacity_provider');
    }
    set capacityProvider(value) {
        this._capacityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderInput() {
        return this._capacityProvider;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference = DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionCapacityProviderStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionCapacityProviderStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionCapacityProviderStrategyList = DataAwsEcsTaskExecutionCapacityProviderStrategyList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionCapacityProviderStrategyList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionCapacityProviderStrategyList", version: "19.15.0" };
function dataAwsEcsTaskExecutionNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.dataAwsEcsTaskExecutionNetworkConfigurationToTerraform = dataAwsEcsTaskExecutionNetworkConfigurationToTerraform;
function dataAwsEcsTaskExecutionNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        assign_public_ip: {
            value: cdktf.booleanToHclTerraform(struct.assignPublicIp),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        security_groups: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionNetworkConfigurationToHclTerraform = dataAwsEcsTaskExecutionNetworkConfigurationToHclTerraform;
class DataAwsEcsTaskExecutionNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.DataAwsEcsTaskExecutionNetworkConfigurationOutputReference = DataAwsEcsTaskExecutionNetworkConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionNetworkConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionNetworkConfigurationOutputReference", version: "19.15.0" };
function dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToTerraform;
function dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToHclTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToHclTerraform;
class DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference = DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference[_d] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList = DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList[_e] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList", version: "19.15.0" };
function dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToTerraform;
function dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToHclTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToHclTerraform;
class DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference = DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference[_f] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList = DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList[_g] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList", version: "19.15.0" };
function dataAwsEcsTaskExecutionOverridesContainerOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        command: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.command),
        cpu: cdktf.numberToTerraform(struct.cpu),
        memory: cdktf.numberToTerraform(struct.memory),
        memory_reservation: cdktf.numberToTerraform(struct.memoryReservation),
        name: cdktf.stringToTerraform(struct.name),
        environment: cdktf.listMapper(dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToTerraform, true)(struct.environment),
        resource_requirements: cdktf.listMapper(dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToTerraform, true)(struct.resourceRequirements),
    };
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesToTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesToTerraform;
function dataAwsEcsTaskExecutionOverridesContainerOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        command: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.command),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        cpu: {
            value: cdktf.numberToHclTerraform(struct.cpu),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        memory: {
            value: cdktf.numberToHclTerraform(struct.memory),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        memory_reservation: {
            value: cdktf.numberToHclTerraform(struct.memoryReservation),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        environment: {
            value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentToHclTerraform, true)(struct.environment),
            isBlock: true,
            type: "set",
            storageClassType: "DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList",
        },
        resource_requirements: {
            value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsToHclTerraform, true)(struct.resourceRequirements),
            isBlock: true,
            type: "set",
            storageClassType: "DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionOverridesContainerOverridesToHclTerraform = dataAwsEcsTaskExecutionOverridesContainerOverridesToHclTerraform;
class DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // environment - computed: false, optional: true, required: false
        this._environment = new DataAwsEcsTaskExecutionOverridesContainerOverridesEnvironmentList(this, "environment", true);
        // resource_requirements - computed: false, optional: true, required: false
        this._resourceRequirements = new DataAwsEcsTaskExecutionOverridesContainerOverridesResourceRequirementsList(this, "resource_requirements", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._command !== undefined) {
            hasAnyValues = true;
            internalValueResult.command = this._command;
        }
        if (this._cpu !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpu = this._cpu;
        }
        if (this._memory !== undefined) {
            hasAnyValues = true;
            internalValueResult.memory = this._memory;
        }
        if (this._memoryReservation !== undefined) {
            hasAnyValues = true;
            internalValueResult.memoryReservation = this._memoryReservation;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._environment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment?.internalValue;
        }
        if (this._resourceRequirements?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceRequirements = this._resourceRequirements?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._command = undefined;
            this._cpu = undefined;
            this._memory = undefined;
            this._memoryReservation = undefined;
            this._name = undefined;
            this._environment.internalValue = undefined;
            this._resourceRequirements.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._command = value.command;
            this._cpu = value.cpu;
            this._memory = value.memory;
            this._memoryReservation = value.memoryReservation;
            this._name = value.name;
            this._environment.internalValue = value.environment;
            this._resourceRequirements.internalValue = value.resourceRequirements;
        }
    }
    get command() {
        return this.getListAttribute('command');
    }
    set command(value) {
        this._command = value;
    }
    resetCommand() {
        this._command = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commandInput() {
        return this._command;
    }
    get cpu() {
        return this.getNumberAttribute('cpu');
    }
    set cpu(value) {
        this._cpu = value;
    }
    resetCpu() {
        this._cpu = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuInput() {
        return this._cpu;
    }
    get memory() {
        return this.getNumberAttribute('memory');
    }
    set memory(value) {
        this._memory = value;
    }
    resetMemory() {
        this._memory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memoryInput() {
        return this._memory;
    }
    get memoryReservation() {
        return this.getNumberAttribute('memory_reservation');
    }
    set memoryReservation(value) {
        this._memoryReservation = value;
    }
    resetMemoryReservation() {
        this._memoryReservation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memoryReservationInput() {
        return this._memoryReservation;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get environment() {
        return this._environment;
    }
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    resetEnvironment() {
        this._environment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment.internalValue;
    }
    get resourceRequirements() {
        return this._resourceRequirements;
    }
    putResourceRequirements(value) {
        this._resourceRequirements.internalValue = value;
    }
    resetResourceRequirements() {
        this._resourceRequirements.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceRequirementsInput() {
        return this._resourceRequirements.internalValue;
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference = DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference[_h] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionOverridesContainerOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionOverridesContainerOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionOverridesContainerOverridesList = DataAwsEcsTaskExecutionOverridesContainerOverridesList;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesContainerOverridesList[_j] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesContainerOverridesList", version: "19.15.0" };
function dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        device_name: cdktf.stringToTerraform(struct.deviceName),
        device_type: cdktf.stringToTerraform(struct.deviceType),
    };
}
exports.dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToTerraform = dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToTerraform;
function dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        device_name: {
            value: cdktf.stringToHclTerraform(struct.deviceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        device_type: {
            value: cdktf.stringToHclTerraform(struct.deviceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToHclTerraform = dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToHclTerraform;
class DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceName = this._deviceName;
        }
        if (this._deviceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceType = this._deviceType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deviceName = undefined;
            this._deviceType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deviceName = value.deviceName;
            this._deviceType = value.deviceType;
        }
    }
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    set deviceName(value) {
        this._deviceName = value;
    }
    resetDeviceName() {
        this._deviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceNameInput() {
        return this._deviceName;
    }
    get deviceType() {
        return this.getStringAttribute('device_type');
    }
    set deviceType(value) {
        this._deviceType = value;
    }
    resetDeviceType() {
        this._deviceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceTypeInput() {
        return this._deviceType;
    }
}
exports.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference = DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference[_k] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList = DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList;
_l = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList[_l] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList", version: "19.15.0" };
function dataAwsEcsTaskExecutionOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu: cdktf.stringToTerraform(struct.cpu),
        execution_role_arn: cdktf.stringToTerraform(struct.executionRoleArn),
        memory: cdktf.stringToTerraform(struct.memory),
        task_role_arn: cdktf.stringToTerraform(struct.taskRoleArn),
        container_overrides: cdktf.listMapper(dataAwsEcsTaskExecutionOverridesContainerOverridesToTerraform, true)(struct.containerOverrides),
        inference_accelerator_overrides: cdktf.listMapper(dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToTerraform, true)(struct.inferenceAcceleratorOverrides),
    };
}
exports.dataAwsEcsTaskExecutionOverridesToTerraform = dataAwsEcsTaskExecutionOverridesToTerraform;
function dataAwsEcsTaskExecutionOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cpu: {
            value: cdktf.stringToHclTerraform(struct.cpu),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        execution_role_arn: {
            value: cdktf.stringToHclTerraform(struct.executionRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        memory: {
            value: cdktf.stringToHclTerraform(struct.memory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        task_role_arn: {
            value: cdktf.stringToHclTerraform(struct.taskRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_overrides: {
            value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionOverridesContainerOverridesToHclTerraform, true)(struct.containerOverrides),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsEcsTaskExecutionOverridesContainerOverridesList",
        },
        inference_accelerator_overrides: {
            value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesToHclTerraform, true)(struct.inferenceAcceleratorOverrides),
            isBlock: true,
            type: "set",
            storageClassType: "DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionOverridesToHclTerraform = dataAwsEcsTaskExecutionOverridesToHclTerraform;
class DataAwsEcsTaskExecutionOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // container_overrides - computed: false, optional: true, required: false
        this._containerOverrides = new DataAwsEcsTaskExecutionOverridesContainerOverridesList(this, "container_overrides", false);
        // inference_accelerator_overrides - computed: false, optional: true, required: false
        this._inferenceAcceleratorOverrides = new DataAwsEcsTaskExecutionOverridesInferenceAcceleratorOverridesList(this, "inference_accelerator_overrides", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpu !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpu = this._cpu;
        }
        if (this._executionRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionRoleArn = this._executionRoleArn;
        }
        if (this._memory !== undefined) {
            hasAnyValues = true;
            internalValueResult.memory = this._memory;
        }
        if (this._taskRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskRoleArn = this._taskRoleArn;
        }
        if (this._containerOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerOverrides = this._containerOverrides?.internalValue;
        }
        if (this._inferenceAcceleratorOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.inferenceAcceleratorOverrides = this._inferenceAcceleratorOverrides?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpu = undefined;
            this._executionRoleArn = undefined;
            this._memory = undefined;
            this._taskRoleArn = undefined;
            this._containerOverrides.internalValue = undefined;
            this._inferenceAcceleratorOverrides.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpu = value.cpu;
            this._executionRoleArn = value.executionRoleArn;
            this._memory = value.memory;
            this._taskRoleArn = value.taskRoleArn;
            this._containerOverrides.internalValue = value.containerOverrides;
            this._inferenceAcceleratorOverrides.internalValue = value.inferenceAcceleratorOverrides;
        }
    }
    get cpu() {
        return this.getStringAttribute('cpu');
    }
    set cpu(value) {
        this._cpu = value;
    }
    resetCpu() {
        this._cpu = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuInput() {
        return this._cpu;
    }
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    resetExecutionRoleArn() {
        this._executionRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    get memory() {
        return this.getStringAttribute('memory');
    }
    set memory(value) {
        this._memory = value;
    }
    resetMemory() {
        this._memory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memoryInput() {
        return this._memory;
    }
    get taskRoleArn() {
        return this.getStringAttribute('task_role_arn');
    }
    set taskRoleArn(value) {
        this._taskRoleArn = value;
    }
    resetTaskRoleArn() {
        this._taskRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskRoleArnInput() {
        return this._taskRoleArn;
    }
    get containerOverrides() {
        return this._containerOverrides;
    }
    putContainerOverrides(value) {
        this._containerOverrides.internalValue = value;
    }
    resetContainerOverrides() {
        this._containerOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerOverridesInput() {
        return this._containerOverrides.internalValue;
    }
    get inferenceAcceleratorOverrides() {
        return this._inferenceAcceleratorOverrides;
    }
    putInferenceAcceleratorOverrides(value) {
        this._inferenceAcceleratorOverrides.internalValue = value;
    }
    resetInferenceAcceleratorOverrides() {
        this._inferenceAcceleratorOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inferenceAcceleratorOverridesInput() {
        return this._inferenceAcceleratorOverrides.internalValue;
    }
}
exports.DataAwsEcsTaskExecutionOverridesOutputReference = DataAwsEcsTaskExecutionOverridesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionOverridesOutputReference[_m] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionOverridesOutputReference", version: "19.15.0" };
function dataAwsEcsTaskExecutionPlacementConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataAwsEcsTaskExecutionPlacementConstraintsToTerraform = dataAwsEcsTaskExecutionPlacementConstraintsToTerraform;
function dataAwsEcsTaskExecutionPlacementConstraintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionPlacementConstraintsToHclTerraform = dataAwsEcsTaskExecutionPlacementConstraintsToHclTerraform;
class DataAwsEcsTaskExecutionPlacementConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expression = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expression = value.expression;
            this._type = value.type;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    resetExpression() {
        this._expression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataAwsEcsTaskExecutionPlacementConstraintsOutputReference = DataAwsEcsTaskExecutionPlacementConstraintsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionPlacementConstraintsOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionPlacementConstraintsOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionPlacementConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionPlacementConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionPlacementConstraintsList = DataAwsEcsTaskExecutionPlacementConstraintsList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionPlacementConstraintsList[_p] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionPlacementConstraintsList", version: "19.15.0" };
function dataAwsEcsTaskExecutionPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataAwsEcsTaskExecutionPlacementStrategyToTerraform = dataAwsEcsTaskExecutionPlacementStrategyToTerraform;
function dataAwsEcsTaskExecutionPlacementStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcsTaskExecutionPlacementStrategyToHclTerraform = dataAwsEcsTaskExecutionPlacementStrategyToHclTerraform;
class DataAwsEcsTaskExecutionPlacementStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._field = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._field = value.field;
            this._type = value.type;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataAwsEcsTaskExecutionPlacementStrategyOutputReference = DataAwsEcsTaskExecutionPlacementStrategyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionPlacementStrategyOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionPlacementStrategyOutputReference", version: "19.15.0" };
class DataAwsEcsTaskExecutionPlacementStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsTaskExecutionPlacementStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsTaskExecutionPlacementStrategyList = DataAwsEcsTaskExecutionPlacementStrategyList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecutionPlacementStrategyList[_r] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecutionPlacementStrategyList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecs_task_execution aws_ecs_task_execution}
*/
class DataAwsEcsTaskExecution extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEcsTaskExecution resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEcsTaskExecution to import
    * @param importFromId The id of the existing DataAwsEcsTaskExecution that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecs_task_execution#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEcsTaskExecution to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecs_task_execution", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecs_task_execution aws_ecs_task_execution} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEcsTaskExecutionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_task_execution',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // capacity_provider_strategy - computed: false, optional: true, required: false
        this._capacityProviderStrategy = new DataAwsEcsTaskExecutionCapacityProviderStrategyList(this, "capacity_provider_strategy", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new DataAwsEcsTaskExecutionNetworkConfigurationOutputReference(this, "network_configuration");
        // overrides - computed: false, optional: true, required: false
        this._overrides = new DataAwsEcsTaskExecutionOverridesOutputReference(this, "overrides");
        // placement_constraints - computed: false, optional: true, required: false
        this._placementConstraints = new DataAwsEcsTaskExecutionPlacementConstraintsList(this, "placement_constraints", true);
        // placement_strategy - computed: false, optional: true, required: false
        this._placementStrategy = new DataAwsEcsTaskExecutionPlacementStrategyList(this, "placement_strategy", false);
        this._clientToken = config.clientToken;
        this._cluster = config.cluster;
        this._desiredCount = config.desiredCount;
        this._enableEcsManagedTags = config.enableEcsManagedTags;
        this._enableExecuteCommand = config.enableExecuteCommand;
        this._group = config.group;
        this._id = config.id;
        this._launchType = config.launchType;
        this._platformVersion = config.platformVersion;
        this._propagateTags = config.propagateTags;
        this._referenceId = config.referenceId;
        this._startedBy = config.startedBy;
        this._tags = config.tags;
        this._taskDefinition = config.taskDefinition;
        this._capacityProviderStrategy.internalValue = config.capacityProviderStrategy;
        this._networkConfiguration.internalValue = config.networkConfiguration;
        this._overrides.internalValue = config.overrides;
        this._placementConstraints.internalValue = config.placementConstraints;
        this._placementStrategy.internalValue = config.placementStrategy;
    }
    get clientToken() {
        return this.getStringAttribute('client_token');
    }
    set clientToken(value) {
        this._clientToken = value;
    }
    resetClientToken() {
        this._clientToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientTokenInput() {
        return this._clientToken;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    get desiredCount() {
        return this.getNumberAttribute('desired_count');
    }
    set desiredCount(value) {
        this._desiredCount = value;
    }
    resetDesiredCount() {
        this._desiredCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredCountInput() {
        return this._desiredCount;
    }
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    get group() {
        return this.getStringAttribute('group');
    }
    set group(value) {
        this._group = value;
    }
    resetGroup() {
        this._group = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupInput() {
        return this._group;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    set launchType(value) {
        this._launchType = value;
    }
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchTypeInput() {
        return this._launchType;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    set propagateTags(value) {
        this._propagateTags = value;
    }
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propagateTagsInput() {
        return this._propagateTags;
    }
    get referenceId() {
        return this.getStringAttribute('reference_id');
    }
    set referenceId(value) {
        this._referenceId = value;
    }
    resetReferenceId() {
        this._referenceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get referenceIdInput() {
        return this._referenceId;
    }
    get startedBy() {
        return this.getStringAttribute('started_by');
    }
    set startedBy(value) {
        this._startedBy = value;
    }
    resetStartedBy() {
        this._startedBy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startedByInput() {
        return this._startedBy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // task_arns - computed: true, optional: false, required: false
    get taskArns() {
        return this.getListAttribute('task_arns');
    }
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    set taskDefinition(value) {
        this._taskDefinition = value;
    }
    // Temporarily expose input value. Use with caution.
    get taskDefinitionInput() {
        return this._taskDefinition;
    }
    get capacityProviderStrategy() {
        return this._capacityProviderStrategy;
    }
    putCapacityProviderStrategy(value) {
        this._capacityProviderStrategy.internalValue = value;
    }
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get overrides() {
        return this._overrides;
    }
    putOverrides(value) {
        this._overrides.internalValue = value;
    }
    resetOverrides() {
        this._overrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overridesInput() {
        return this._overrides.internalValue;
    }
    get placementConstraints() {
        return this._placementConstraints;
    }
    putPlacementConstraints(value) {
        this._placementConstraints.internalValue = value;
    }
    resetPlacementConstraints() {
        this._placementConstraints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementConstraintsInput() {
        return this._placementConstraints.internalValue;
    }
    get placementStrategy() {
        return this._placementStrategy;
    }
    putPlacementStrategy(value) {
        this._placementStrategy.internalValue = value;
    }
    resetPlacementStrategy() {
        this._placementStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementStrategyInput() {
        return this._placementStrategy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client_token: cdktf.stringToTerraform(this._clientToken),
            cluster: cdktf.stringToTerraform(this._cluster),
            desired_count: cdktf.numberToTerraform(this._desiredCount),
            enable_ecs_managed_tags: cdktf.booleanToTerraform(this._enableEcsManagedTags),
            enable_execute_command: cdktf.booleanToTerraform(this._enableExecuteCommand),
            group: cdktf.stringToTerraform(this._group),
            id: cdktf.stringToTerraform(this._id),
            launch_type: cdktf.stringToTerraform(this._launchType),
            platform_version: cdktf.stringToTerraform(this._platformVersion),
            propagate_tags: cdktf.stringToTerraform(this._propagateTags),
            reference_id: cdktf.stringToTerraform(this._referenceId),
            started_by: cdktf.stringToTerraform(this._startedBy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            task_definition: cdktf.stringToTerraform(this._taskDefinition),
            capacity_provider_strategy: cdktf.listMapper(dataAwsEcsTaskExecutionCapacityProviderStrategyToTerraform, true)(this._capacityProviderStrategy.internalValue),
            network_configuration: dataAwsEcsTaskExecutionNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
            overrides: dataAwsEcsTaskExecutionOverridesToTerraform(this._overrides.internalValue),
            placement_constraints: cdktf.listMapper(dataAwsEcsTaskExecutionPlacementConstraintsToTerraform, true)(this._placementConstraints.internalValue),
            placement_strategy: cdktf.listMapper(dataAwsEcsTaskExecutionPlacementStrategyToTerraform, true)(this._placementStrategy.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            client_token: {
                value: cdktf.stringToHclTerraform(this._clientToken),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cluster: {
                value: cdktf.stringToHclTerraform(this._cluster),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            desired_count: {
                value: cdktf.numberToHclTerraform(this._desiredCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            enable_ecs_managed_tags: {
                value: cdktf.booleanToHclTerraform(this._enableEcsManagedTags),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            enable_execute_command: {
                value: cdktf.booleanToHclTerraform(this._enableExecuteCommand),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            group: {
                value: cdktf.stringToHclTerraform(this._group),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            launch_type: {
                value: cdktf.stringToHclTerraform(this._launchType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            platform_version: {
                value: cdktf.stringToHclTerraform(this._platformVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            propagate_tags: {
                value: cdktf.stringToHclTerraform(this._propagateTags),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            reference_id: {
                value: cdktf.stringToHclTerraform(this._referenceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            started_by: {
                value: cdktf.stringToHclTerraform(this._startedBy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            task_definition: {
                value: cdktf.stringToHclTerraform(this._taskDefinition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            capacity_provider_strategy: {
                value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionCapacityProviderStrategyToHclTerraform, true)(this._capacityProviderStrategy.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsEcsTaskExecutionCapacityProviderStrategyList",
            },
            network_configuration: {
                value: dataAwsEcsTaskExecutionNetworkConfigurationToHclTerraform(this._networkConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsEcsTaskExecutionNetworkConfigurationList",
            },
            overrides: {
                value: dataAwsEcsTaskExecutionOverridesToHclTerraform(this._overrides.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsEcsTaskExecutionOverridesList",
            },
            placement_constraints: {
                value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionPlacementConstraintsToHclTerraform, true)(this._placementConstraints.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsEcsTaskExecutionPlacementConstraintsList",
            },
            placement_strategy: {
                value: cdktf.listMapperHcl(dataAwsEcsTaskExecutionPlacementStrategyToHclTerraform, true)(this._placementStrategy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsEcsTaskExecutionPlacementStrategyList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEcsTaskExecution = DataAwsEcsTaskExecution;
_s = JSII_RTTI_SYMBOL_1;
DataAwsEcsTaskExecution[_s] = { fqn: "@cdktf/provider-aws.dataAwsEcsTaskExecution.DataAwsEcsTaskExecution", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEcsTaskExecution.tfResourceType = "aws_ecs_task_execution";
