// @cdktf/provider-aws/lib/resourceexplorer2-view/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resourceexplorer2View = exports.Resourceexplorer2ViewIncludedPropertyList = exports.Resourceexplorer2ViewIncludedPropertyOutputReference = exports.resourceexplorer2ViewIncludedPropertyToHclTerraform = exports.resourceexplorer2ViewIncludedPropertyToTerraform = exports.Resourceexplorer2ViewFiltersList = exports.Resourceexplorer2ViewFiltersOutputReference = exports.resourceexplorer2ViewFiltersToHclTerraform = exports.resourceexplorer2ViewFiltersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function resourceexplorer2ViewFiltersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter_string: cdktf.stringToTerraform(struct.filterString),
    };
}
exports.resourceexplorer2ViewFiltersToTerraform = resourceexplorer2ViewFiltersToTerraform;
function resourceexplorer2ViewFiltersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter_string: {
            value: cdktf.stringToHclTerraform(struct.filterString),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.resourceexplorer2ViewFiltersToHclTerraform = resourceexplorer2ViewFiltersToHclTerraform;
class Resourceexplorer2ViewFiltersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filterString !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterString = this._filterString;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._filterString = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._filterString = value.filterString;
        }
    }
    get filterString() {
        return this.getStringAttribute('filter_string');
    }
    set filterString(value) {
        this._filterString = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterStringInput() {
        return this._filterString;
    }
}
exports.Resourceexplorer2ViewFiltersOutputReference = Resourceexplorer2ViewFiltersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Resourceexplorer2ViewFiltersOutputReference[_a] = { fqn: "@cdktf/provider-aws.resourceexplorer2View.Resourceexplorer2ViewFiltersOutputReference", version: "19.15.0" };
class Resourceexplorer2ViewFiltersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Resourceexplorer2ViewFiltersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Resourceexplorer2ViewFiltersList = Resourceexplorer2ViewFiltersList;
_b = JSII_RTTI_SYMBOL_1;
Resourceexplorer2ViewFiltersList[_b] = { fqn: "@cdktf/provider-aws.resourceexplorer2View.Resourceexplorer2ViewFiltersList", version: "19.15.0" };
function resourceexplorer2ViewIncludedPropertyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.resourceexplorer2ViewIncludedPropertyToTerraform = resourceexplorer2ViewIncludedPropertyToTerraform;
function resourceexplorer2ViewIncludedPropertyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.resourceexplorer2ViewIncludedPropertyToHclTerraform = resourceexplorer2ViewIncludedPropertyToHclTerraform;
class Resourceexplorer2ViewIncludedPropertyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.Resourceexplorer2ViewIncludedPropertyOutputReference = Resourceexplorer2ViewIncludedPropertyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Resourceexplorer2ViewIncludedPropertyOutputReference[_c] = { fqn: "@cdktf/provider-aws.resourceexplorer2View.Resourceexplorer2ViewIncludedPropertyOutputReference", version: "19.15.0" };
class Resourceexplorer2ViewIncludedPropertyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Resourceexplorer2ViewIncludedPropertyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Resourceexplorer2ViewIncludedPropertyList = Resourceexplorer2ViewIncludedPropertyList;
_d = JSII_RTTI_SYMBOL_1;
Resourceexplorer2ViewIncludedPropertyList[_d] = { fqn: "@cdktf/provider-aws.resourceexplorer2View.Resourceexplorer2ViewIncludedPropertyList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/resourceexplorer2_view aws_resourceexplorer2_view}
*/
class Resourceexplorer2View extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Resourceexplorer2View resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Resourceexplorer2View to import
    * @param importFromId The id of the existing Resourceexplorer2View that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/resourceexplorer2_view#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Resourceexplorer2View to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_resourceexplorer2_view", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/resourceexplorer2_view aws_resourceexplorer2_view} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Resourceexplorer2ViewConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_resourceexplorer2_view',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // filters - computed: false, optional: true, required: false
        this._filters = new Resourceexplorer2ViewFiltersList(this, "filters", false);
        // included_property - computed: false, optional: true, required: false
        this._includedProperty = new Resourceexplorer2ViewIncludedPropertyList(this, "included_property", false);
        this._defaultView = config.defaultView;
        this._name = config.name;
        this._tags = config.tags;
        this._filters.internalValue = config.filters;
        this._includedProperty.internalValue = config.includedProperty;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get defaultView() {
        return this.getBooleanAttribute('default_view');
    }
    set defaultView(value) {
        this._defaultView = value;
    }
    resetDefaultView() {
        this._defaultView = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultViewInput() {
        return this._defaultView;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get filters() {
        return this._filters;
    }
    putFilters(value) {
        this._filters.internalValue = value;
    }
    resetFilters() {
        this._filters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filtersInput() {
        return this._filters.internalValue;
    }
    get includedProperty() {
        return this._includedProperty;
    }
    putIncludedProperty(value) {
        this._includedProperty.internalValue = value;
    }
    resetIncludedProperty() {
        this._includedProperty.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includedPropertyInput() {
        return this._includedProperty.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_view: cdktf.booleanToTerraform(this._defaultView),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            filters: cdktf.listMapper(resourceexplorer2ViewFiltersToTerraform, true)(this._filters.internalValue),
            included_property: cdktf.listMapper(resourceexplorer2ViewIncludedPropertyToTerraform, true)(this._includedProperty.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            default_view: {
                value: cdktf.booleanToHclTerraform(this._defaultView),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            filters: {
                value: cdktf.listMapperHcl(resourceexplorer2ViewFiltersToHclTerraform, true)(this._filters.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Resourceexplorer2ViewFiltersList",
            },
            included_property: {
                value: cdktf.listMapperHcl(resourceexplorer2ViewIncludedPropertyToHclTerraform, true)(this._includedProperty.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Resourceexplorer2ViewIncludedPropertyList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Resourceexplorer2View = Resourceexplorer2View;
_e = JSII_RTTI_SYMBOL_1;
Resourceexplorer2View[_e] = { fqn: "@cdktf/provider-aws.resourceexplorer2View.Resourceexplorer2View", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Resourceexplorer2View.tfResourceType = "aws_resourceexplorer2_view";
