// @cdktf/provider-aws/lib/data-aws-sqs-queues/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSqsQueues = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sqs_queues aws_sqs_queues}
*/
class DataAwsSqsQueues extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSqsQueues resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSqsQueues to import
    * @param importFromId The id of the existing DataAwsSqsQueues that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sqs_queues#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSqsQueues to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sqs_queues", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sqs_queues aws_sqs_queues} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSqsQueuesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_sqs_queues',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._queueNamePrefix = config.queueNamePrefix;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get queueNamePrefix() {
        return this.getStringAttribute('queue_name_prefix');
    }
    set queueNamePrefix(value) {
        this._queueNamePrefix = value;
    }
    resetQueueNamePrefix() {
        this._queueNamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queueNamePrefixInput() {
        return this._queueNamePrefix;
    }
    // queue_urls - computed: true, optional: false, required: false
    get queueUrls() {
        return cdktf.Fn.tolist(this.getListAttribute('queue_urls'));
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            queue_name_prefix: cdktf.stringToTerraform(this._queueNamePrefix),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            queue_name_prefix: {
                value: cdktf.stringToHclTerraform(this._queueNamePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSqsQueues = DataAwsSqsQueues;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSqsQueues[_a] = { fqn: "@cdktf/provider-aws.dataAwsSqsQueues.DataAwsSqsQueues", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSqsQueues.tfResourceType = "aws_sqs_queues";
