// @cdktf/provider-aws/lib/shield-proactive-engagement/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShieldProactiveEngagement = exports.ShieldProactiveEngagementEmergencyContactList = exports.ShieldProactiveEngagementEmergencyContactOutputReference = exports.shieldProactiveEngagementEmergencyContactToHclTerraform = exports.shieldProactiveEngagementEmergencyContactToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function shieldProactiveEngagementEmergencyContactToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contact_notes: cdktf.stringToTerraform(struct.contactNotes),
        email_address: cdktf.stringToTerraform(struct.emailAddress),
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
    };
}
exports.shieldProactiveEngagementEmergencyContactToTerraform = shieldProactiveEngagementEmergencyContactToTerraform;
function shieldProactiveEngagementEmergencyContactToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contact_notes: {
            value: cdktf.stringToHclTerraform(struct.contactNotes),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        email_address: {
            value: cdktf.stringToHclTerraform(struct.emailAddress),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        phone_number: {
            value: cdktf.stringToHclTerraform(struct.phoneNumber),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.shieldProactiveEngagementEmergencyContactToHclTerraform = shieldProactiveEngagementEmergencyContactToHclTerraform;
class ShieldProactiveEngagementEmergencyContactOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contactNotes !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactNotes = this._contactNotes;
        }
        if (this._emailAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.emailAddress = this._emailAddress;
        }
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contactNotes = undefined;
            this._emailAddress = undefined;
            this._phoneNumber = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contactNotes = value.contactNotes;
            this._emailAddress = value.emailAddress;
            this._phoneNumber = value.phoneNumber;
        }
    }
    get contactNotes() {
        return this.getStringAttribute('contact_notes');
    }
    set contactNotes(value) {
        this._contactNotes = value;
    }
    resetContactNotes() {
        this._contactNotes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactNotesInput() {
        return this._contactNotes;
    }
    get emailAddress() {
        return this.getStringAttribute('email_address');
    }
    set emailAddress(value) {
        this._emailAddress = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailAddressInput() {
        return this._emailAddress;
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    resetPhoneNumber() {
        this._phoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
}
exports.ShieldProactiveEngagementEmergencyContactOutputReference = ShieldProactiveEngagementEmergencyContactOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ShieldProactiveEngagementEmergencyContactOutputReference[_a] = { fqn: "@cdktf/provider-aws.shieldProactiveEngagement.ShieldProactiveEngagementEmergencyContactOutputReference", version: "19.15.0" };
class ShieldProactiveEngagementEmergencyContactList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ShieldProactiveEngagementEmergencyContactOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ShieldProactiveEngagementEmergencyContactList = ShieldProactiveEngagementEmergencyContactList;
_b = JSII_RTTI_SYMBOL_1;
ShieldProactiveEngagementEmergencyContactList[_b] = { fqn: "@cdktf/provider-aws.shieldProactiveEngagement.ShieldProactiveEngagementEmergencyContactList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/shield_proactive_engagement aws_shield_proactive_engagement}
*/
class ShieldProactiveEngagement extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ShieldProactiveEngagement resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ShieldProactiveEngagement to import
    * @param importFromId The id of the existing ShieldProactiveEngagement that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/shield_proactive_engagement#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ShieldProactiveEngagement to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_shield_proactive_engagement", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/shield_proactive_engagement aws_shield_proactive_engagement} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ShieldProactiveEngagementConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_shield_proactive_engagement',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // emergency_contact - computed: false, optional: true, required: false
        this._emergencyContact = new ShieldProactiveEngagementEmergencyContactList(this, "emergency_contact", false);
        this._enabled = config.enabled;
        this._emergencyContact.internalValue = config.emergencyContact;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get emergencyContact() {
        return this._emergencyContact;
    }
    putEmergencyContact(value) {
        this._emergencyContact.internalValue = value;
    }
    resetEmergencyContact() {
        this._emergencyContact.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emergencyContactInput() {
        return this._emergencyContact.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enabled: cdktf.booleanToTerraform(this._enabled),
            emergency_contact: cdktf.listMapper(shieldProactiveEngagementEmergencyContactToTerraform, true)(this._emergencyContact.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            emergency_contact: {
                value: cdktf.listMapperHcl(shieldProactiveEngagementEmergencyContactToHclTerraform, true)(this._emergencyContact.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ShieldProactiveEngagementEmergencyContactList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ShieldProactiveEngagement = ShieldProactiveEngagement;
_c = JSII_RTTI_SYMBOL_1;
ShieldProactiveEngagement[_c] = { fqn: "@cdktf/provider-aws.shieldProactiveEngagement.ShieldProactiveEngagement", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ShieldProactiveEngagement.tfResourceType = "aws_shield_proactive_engagement";
