// @cdktf/provider-aws/lib/api-gateway-request-validator/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayRequestValidator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/api_gateway_request_validator aws_api_gateway_request_validator}
*/
class ApiGatewayRequestValidator extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApiGatewayRequestValidator resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApiGatewayRequestValidator to import
    * @param importFromId The id of the existing ApiGatewayRequestValidator that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/api_gateway_request_validator#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApiGatewayRequestValidator to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_api_gateway_request_validator", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/api_gateway_request_validator aws_api_gateway_request_validator} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayRequestValidatorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_request_validator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._name = config.name;
        this._restApiId = config.restApiId;
        this._validateRequestBody = config.validateRequestBody;
        this._validateRequestParameters = config.validateRequestParameters;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get restApiIdInput() {
        return this._restApiId;
    }
    get validateRequestBody() {
        return this.getBooleanAttribute('validate_request_body');
    }
    set validateRequestBody(value) {
        this._validateRequestBody = value;
    }
    resetValidateRequestBody() {
        this._validateRequestBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateRequestBodyInput() {
        return this._validateRequestBody;
    }
    get validateRequestParameters() {
        return this.getBooleanAttribute('validate_request_parameters');
    }
    set validateRequestParameters(value) {
        this._validateRequestParameters = value;
    }
    resetValidateRequestParameters() {
        this._validateRequestParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateRequestParametersInput() {
        return this._validateRequestParameters;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            validate_request_body: cdktf.booleanToTerraform(this._validateRequestBody),
            validate_request_parameters: cdktf.booleanToTerraform(this._validateRequestParameters),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rest_api_id: {
                value: cdktf.stringToHclTerraform(this._restApiId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            validate_request_body: {
                value: cdktf.booleanToHclTerraform(this._validateRequestBody),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            validate_request_parameters: {
                value: cdktf.booleanToHclTerraform(this._validateRequestParameters),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApiGatewayRequestValidator = ApiGatewayRequestValidator;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayRequestValidator[_a] = { fqn: "@cdktf/provider-aws.apiGatewayRequestValidator.ApiGatewayRequestValidator", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ApiGatewayRequestValidator.tfResourceType = "aws_api_gateway_request_validator";
