// @cdktf/provider-aws/lib/lightsail-database/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailDatabase = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_database aws_lightsail_database}
*/
class LightsailDatabase extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LightsailDatabase resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LightsailDatabase to import
    * @param importFromId The id of the existing LightsailDatabase that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_database#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LightsailDatabase to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lightsail_database", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_database aws_lightsail_database} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LightsailDatabaseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_database',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._applyImmediately = config.applyImmediately;
        this._availabilityZone = config.availabilityZone;
        this._backupRetentionEnabled = config.backupRetentionEnabled;
        this._blueprintId = config.blueprintId;
        this._bundleId = config.bundleId;
        this._finalSnapshotName = config.finalSnapshotName;
        this._id = config.id;
        this._masterDatabaseName = config.masterDatabaseName;
        this._masterPassword = config.masterPassword;
        this._masterUsername = config.masterUsername;
        this._preferredBackupWindow = config.preferredBackupWindow;
        this._preferredMaintenanceWindow = config.preferredMaintenanceWindow;
        this._publiclyAccessible = config.publiclyAccessible;
        this._relationalDatabaseName = config.relationalDatabaseName;
        this._skipFinalSnapshot = config.skipFinalSnapshot;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    get applyImmediately() {
        return this.getBooleanAttribute('apply_immediately');
    }
    set applyImmediately(value) {
        this._applyImmediately = value;
    }
    resetApplyImmediately() {
        this._applyImmediately = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applyImmediatelyInput() {
        return this._applyImmediately;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    resetAvailabilityZone() {
        this._availabilityZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    get backupRetentionEnabled() {
        return this.getBooleanAttribute('backup_retention_enabled');
    }
    set backupRetentionEnabled(value) {
        this._backupRetentionEnabled = value;
    }
    resetBackupRetentionEnabled() {
        this._backupRetentionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupRetentionEnabledInput() {
        return this._backupRetentionEnabled;
    }
    get blueprintId() {
        return this.getStringAttribute('blueprint_id');
    }
    set blueprintId(value) {
        this._blueprintId = value;
    }
    // Temporarily expose input value. Use with caution.
    get blueprintIdInput() {
        return this._blueprintId;
    }
    get bundleId() {
        return this.getStringAttribute('bundle_id');
    }
    set bundleId(value) {
        this._bundleId = value;
    }
    // Temporarily expose input value. Use with caution.
    get bundleIdInput() {
        return this._bundleId;
    }
    // ca_certificate_identifier - computed: true, optional: false, required: false
    get caCertificateIdentifier() {
        return this.getStringAttribute('ca_certificate_identifier');
    }
    // cpu_count - computed: true, optional: false, required: false
    get cpuCount() {
        return this.getNumberAttribute('cpu_count');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // engine - computed: true, optional: false, required: false
    get engine() {
        return this.getStringAttribute('engine');
    }
    // engine_version - computed: true, optional: false, required: false
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    get finalSnapshotName() {
        return this.getStringAttribute('final_snapshot_name');
    }
    set finalSnapshotName(value) {
        this._finalSnapshotName = value;
    }
    resetFinalSnapshotName() {
        this._finalSnapshotName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get finalSnapshotNameInput() {
        return this._finalSnapshotName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get masterDatabaseName() {
        return this.getStringAttribute('master_database_name');
    }
    set masterDatabaseName(value) {
        this._masterDatabaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get masterDatabaseNameInput() {
        return this._masterDatabaseName;
    }
    // master_endpoint_address - computed: true, optional: false, required: false
    get masterEndpointAddress() {
        return this.getStringAttribute('master_endpoint_address');
    }
    // master_endpoint_port - computed: true, optional: false, required: false
    get masterEndpointPort() {
        return this.getNumberAttribute('master_endpoint_port');
    }
    get masterPassword() {
        return this.getStringAttribute('master_password');
    }
    set masterPassword(value) {
        this._masterPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get masterPasswordInput() {
        return this._masterPassword;
    }
    get masterUsername() {
        return this.getStringAttribute('master_username');
    }
    set masterUsername(value) {
        this._masterUsername = value;
    }
    // Temporarily expose input value. Use with caution.
    get masterUsernameInput() {
        return this._masterUsername;
    }
    get preferredBackupWindow() {
        return this.getStringAttribute('preferred_backup_window');
    }
    set preferredBackupWindow(value) {
        this._preferredBackupWindow = value;
    }
    resetPreferredBackupWindow() {
        this._preferredBackupWindow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preferredBackupWindowInput() {
        return this._preferredBackupWindow;
    }
    get preferredMaintenanceWindow() {
        return this.getStringAttribute('preferred_maintenance_window');
    }
    set preferredMaintenanceWindow(value) {
        this._preferredMaintenanceWindow = value;
    }
    resetPreferredMaintenanceWindow() {
        this._preferredMaintenanceWindow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preferredMaintenanceWindowInput() {
        return this._preferredMaintenanceWindow;
    }
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    set publiclyAccessible(value) {
        this._publiclyAccessible = value;
    }
    resetPubliclyAccessible() {
        this._publiclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publiclyAccessibleInput() {
        return this._publiclyAccessible;
    }
    // ram_size - computed: true, optional: false, required: false
    get ramSize() {
        return this.getNumberAttribute('ram_size');
    }
    get relationalDatabaseName() {
        return this.getStringAttribute('relational_database_name');
    }
    set relationalDatabaseName(value) {
        this._relationalDatabaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get relationalDatabaseNameInput() {
        return this._relationalDatabaseName;
    }
    // secondary_availability_zone - computed: true, optional: false, required: false
    get secondaryAvailabilityZone() {
        return this.getStringAttribute('secondary_availability_zone');
    }
    get skipFinalSnapshot() {
        return this.getBooleanAttribute('skip_final_snapshot');
    }
    set skipFinalSnapshot(value) {
        this._skipFinalSnapshot = value;
    }
    resetSkipFinalSnapshot() {
        this._skipFinalSnapshot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipFinalSnapshotInput() {
        return this._skipFinalSnapshot;
    }
    // support_code - computed: true, optional: false, required: false
    get supportCode() {
        return this.getStringAttribute('support_code');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            apply_immediately: cdktf.booleanToTerraform(this._applyImmediately),
            availability_zone: cdktf.stringToTerraform(this._availabilityZone),
            backup_retention_enabled: cdktf.booleanToTerraform(this._backupRetentionEnabled),
            blueprint_id: cdktf.stringToTerraform(this._blueprintId),
            bundle_id: cdktf.stringToTerraform(this._bundleId),
            final_snapshot_name: cdktf.stringToTerraform(this._finalSnapshotName),
            id: cdktf.stringToTerraform(this._id),
            master_database_name: cdktf.stringToTerraform(this._masterDatabaseName),
            master_password: cdktf.stringToTerraform(this._masterPassword),
            master_username: cdktf.stringToTerraform(this._masterUsername),
            preferred_backup_window: cdktf.stringToTerraform(this._preferredBackupWindow),
            preferred_maintenance_window: cdktf.stringToTerraform(this._preferredMaintenanceWindow),
            publicly_accessible: cdktf.booleanToTerraform(this._publiclyAccessible),
            relational_database_name: cdktf.stringToTerraform(this._relationalDatabaseName),
            skip_final_snapshot: cdktf.booleanToTerraform(this._skipFinalSnapshot),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            apply_immediately: {
                value: cdktf.booleanToHclTerraform(this._applyImmediately),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            availability_zone: {
                value: cdktf.stringToHclTerraform(this._availabilityZone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            backup_retention_enabled: {
                value: cdktf.booleanToHclTerraform(this._backupRetentionEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            blueprint_id: {
                value: cdktf.stringToHclTerraform(this._blueprintId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bundle_id: {
                value: cdktf.stringToHclTerraform(this._bundleId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            final_snapshot_name: {
                value: cdktf.stringToHclTerraform(this._finalSnapshotName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            master_database_name: {
                value: cdktf.stringToHclTerraform(this._masterDatabaseName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            master_password: {
                value: cdktf.stringToHclTerraform(this._masterPassword),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            master_username: {
                value: cdktf.stringToHclTerraform(this._masterUsername),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preferred_backup_window: {
                value: cdktf.stringToHclTerraform(this._preferredBackupWindow),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preferred_maintenance_window: {
                value: cdktf.stringToHclTerraform(this._preferredMaintenanceWindow),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            publicly_accessible: {
                value: cdktf.booleanToHclTerraform(this._publiclyAccessible),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            relational_database_name: {
                value: cdktf.stringToHclTerraform(this._relationalDatabaseName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            skip_final_snapshot: {
                value: cdktf.booleanToHclTerraform(this._skipFinalSnapshot),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LightsailDatabase = LightsailDatabase;
_a = JSII_RTTI_SYMBOL_1;
LightsailDatabase[_a] = { fqn: "@cdktf/provider-aws.lightsailDatabase.LightsailDatabase", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LightsailDatabase.tfResourceType = "aws_lightsail_database";
