// @cdktf/provider-aws/lib/opensearchserverless-security-config/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpensearchserverlessSecurityConfig = exports.OpensearchserverlessSecurityConfigSamlOptionsOutputReference = exports.opensearchserverlessSecurityConfigSamlOptionsToHclTerraform = exports.opensearchserverlessSecurityConfigSamlOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function opensearchserverlessSecurityConfigSamlOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        group_attribute: cdktf.stringToTerraform(struct.groupAttribute),
        metadata: cdktf.stringToTerraform(struct.metadata),
        session_timeout: cdktf.numberToTerraform(struct.sessionTimeout),
        user_attribute: cdktf.stringToTerraform(struct.userAttribute),
    };
}
exports.opensearchserverlessSecurityConfigSamlOptionsToTerraform = opensearchserverlessSecurityConfigSamlOptionsToTerraform;
function opensearchserverlessSecurityConfigSamlOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        group_attribute: {
            value: cdktf.stringToHclTerraform(struct.groupAttribute),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        metadata: {
            value: cdktf.stringToHclTerraform(struct.metadata),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        session_timeout: {
            value: cdktf.numberToHclTerraform(struct.sessionTimeout),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        user_attribute: {
            value: cdktf.stringToHclTerraform(struct.userAttribute),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchserverlessSecurityConfigSamlOptionsToHclTerraform = opensearchserverlessSecurityConfigSamlOptionsToHclTerraform;
class OpensearchserverlessSecurityConfigSamlOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._groupAttribute !== undefined) {
            hasAnyValues = true;
            internalValueResult.groupAttribute = this._groupAttribute;
        }
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        if (this._sessionTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionTimeout = this._sessionTimeout;
        }
        if (this._userAttribute !== undefined) {
            hasAnyValues = true;
            internalValueResult.userAttribute = this._userAttribute;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._groupAttribute = undefined;
            this._metadata = undefined;
            this._sessionTimeout = undefined;
            this._userAttribute = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._groupAttribute = value.groupAttribute;
            this._metadata = value.metadata;
            this._sessionTimeout = value.sessionTimeout;
            this._userAttribute = value.userAttribute;
        }
    }
    get groupAttribute() {
        return this.getStringAttribute('group_attribute');
    }
    set groupAttribute(value) {
        this._groupAttribute = value;
    }
    resetGroupAttribute() {
        this._groupAttribute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupAttributeInput() {
        return this._groupAttribute;
    }
    get metadata() {
        return this.getStringAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    set sessionTimeout(value) {
        this._sessionTimeout = value;
    }
    resetSessionTimeout() {
        this._sessionTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionTimeoutInput() {
        return this._sessionTimeout;
    }
    get userAttribute() {
        return this.getStringAttribute('user_attribute');
    }
    set userAttribute(value) {
        this._userAttribute = value;
    }
    resetUserAttribute() {
        this._userAttribute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userAttributeInput() {
        return this._userAttribute;
    }
}
exports.OpensearchserverlessSecurityConfigSamlOptionsOutputReference = OpensearchserverlessSecurityConfigSamlOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
OpensearchserverlessSecurityConfigSamlOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.opensearchserverlessSecurityConfig.OpensearchserverlessSecurityConfigSamlOptionsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearchserverless_security_config aws_opensearchserverless_security_config}
*/
class OpensearchserverlessSecurityConfig extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a OpensearchserverlessSecurityConfig resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the OpensearchserverlessSecurityConfig to import
    * @param importFromId The id of the existing OpensearchserverlessSecurityConfig that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearchserverless_security_config#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the OpensearchserverlessSecurityConfig to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_opensearchserverless_security_config", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearchserverless_security_config aws_opensearchserverless_security_config} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OpensearchserverlessSecurityConfigConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opensearchserverless_security_config',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // saml_options - computed: false, optional: false, required: false
        this._samlOptions = new OpensearchserverlessSecurityConfigSamlOptionsOutputReference(this, "saml_options");
        this._description = config.description;
        this._name = config.name;
        this._type = config.type;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // config_version - computed: true, optional: false, required: false
    get configVersion() {
        return this.getStringAttribute('config_version');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get samlOptions() {
        return this._samlOptions;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            type: cdktf.stringToTerraform(this._type),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.OpensearchserverlessSecurityConfig = OpensearchserverlessSecurityConfig;
_b = JSII_RTTI_SYMBOL_1;
OpensearchserverlessSecurityConfig[_b] = { fqn: "@cdktf/provider-aws.opensearchserverlessSecurityConfig.OpensearchserverlessSecurityConfig", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
OpensearchserverlessSecurityConfig.tfResourceType = "aws_opensearchserverless_security_config";
