// @cdktf/provider-aws/lib/route53-cidr-location/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53CidrLocation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53_cidr_location aws_route53_cidr_location}
*/
class Route53CidrLocation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Route53CidrLocation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Route53CidrLocation to import
    * @param importFromId The id of the existing Route53CidrLocation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53_cidr_location#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Route53CidrLocation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_route53_cidr_location", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53_cidr_location aws_route53_cidr_location} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53CidrLocationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_cidr_location',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._cidrBlocks = config.cidrBlocks;
        this._cidrCollectionId = config.cidrCollectionId;
        this._name = config.name;
    }
    get cidrBlocks() {
        return cdktf.Fn.tolist(this.getListAttribute('cidr_blocks'));
    }
    set cidrBlocks(value) {
        this._cidrBlocks = value;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlocksInput() {
        return this._cidrBlocks;
    }
    get cidrCollectionId() {
        return this.getStringAttribute('cidr_collection_id');
    }
    set cidrCollectionId(value) {
        this._cidrCollectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get cidrCollectionIdInput() {
        return this._cidrCollectionId;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(this._cidrBlocks),
            cidr_collection_id: cdktf.stringToTerraform(this._cidrCollectionId),
            name: cdktf.stringToTerraform(this._name),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cidr_blocks: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._cidrBlocks),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            cidr_collection_id: {
                value: cdktf.stringToHclTerraform(this._cidrCollectionId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Route53CidrLocation = Route53CidrLocation;
_a = JSII_RTTI_SYMBOL_1;
Route53CidrLocation[_a] = { fqn: "@cdktf/provider-aws.route53CidrLocation.Route53CidrLocation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Route53CidrLocation.tfResourceType = "aws_route53_cidr_location";
