// @cdktf/provider-aws/lib/vpc-endpoint-connection-notification/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcEndpointConnectionNotification = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_endpoint_connection_notification aws_vpc_endpoint_connection_notification}
*/
class VpcEndpointConnectionNotification extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VpcEndpointConnectionNotification resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VpcEndpointConnectionNotification to import
    * @param importFromId The id of the existing VpcEndpointConnectionNotification that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_endpoint_connection_notification#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VpcEndpointConnectionNotification to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpc_endpoint_connection_notification", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_endpoint_connection_notification aws_vpc_endpoint_connection_notification} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcEndpointConnectionNotificationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint_connection_notification',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._connectionEvents = config.connectionEvents;
        this._connectionNotificationArn = config.connectionNotificationArn;
        this._id = config.id;
        this._vpcEndpointId = config.vpcEndpointId;
        this._vpcEndpointServiceId = config.vpcEndpointServiceId;
    }
    get connectionEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('connection_events'));
    }
    set connectionEvents(value) {
        this._connectionEvents = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionEventsInput() {
        return this._connectionEvents;
    }
    get connectionNotificationArn() {
        return this.getStringAttribute('connection_notification_arn');
    }
    set connectionNotificationArn(value) {
        this._connectionNotificationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNotificationArnInput() {
        return this._connectionNotificationArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // notification_type - computed: true, optional: false, required: false
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get vpcEndpointId() {
        return this.getStringAttribute('vpc_endpoint_id');
    }
    set vpcEndpointId(value) {
        this._vpcEndpointId = value;
    }
    resetVpcEndpointId() {
        this._vpcEndpointId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointIdInput() {
        return this._vpcEndpointId;
    }
    get vpcEndpointServiceId() {
        return this.getStringAttribute('vpc_endpoint_service_id');
    }
    set vpcEndpointServiceId(value) {
        this._vpcEndpointServiceId = value;
    }
    resetVpcEndpointServiceId() {
        this._vpcEndpointServiceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointServiceIdInput() {
        return this._vpcEndpointServiceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connection_events: cdktf.listMapper(cdktf.stringToTerraform, false)(this._connectionEvents),
            connection_notification_arn: cdktf.stringToTerraform(this._connectionNotificationArn),
            id: cdktf.stringToTerraform(this._id),
            vpc_endpoint_id: cdktf.stringToTerraform(this._vpcEndpointId),
            vpc_endpoint_service_id: cdktf.stringToTerraform(this._vpcEndpointServiceId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            connection_events: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._connectionEvents),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            connection_notification_arn: {
                value: cdktf.stringToHclTerraform(this._connectionNotificationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vpc_endpoint_id: {
                value: cdktf.stringToHclTerraform(this._vpcEndpointId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vpc_endpoint_service_id: {
                value: cdktf.stringToHclTerraform(this._vpcEndpointServiceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VpcEndpointConnectionNotification = VpcEndpointConnectionNotification;
_a = JSII_RTTI_SYMBOL_1;
VpcEndpointConnectionNotification[_a] = { fqn: "@cdktf/provider-aws.vpcEndpointConnectionNotification.VpcEndpointConnectionNotification", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VpcEndpointConnectionNotification.tfResourceType = "aws_vpc_endpoint_connection_notification";
