// @cdktf/provider-aws/lib/gamelift-game-server-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameliftGameServerGroup = exports.GameliftGameServerGroupTimeoutsOutputReference = exports.gameliftGameServerGroupTimeoutsToHclTerraform = exports.gameliftGameServerGroupTimeoutsToTerraform = exports.GameliftGameServerGroupLaunchTemplateOutputReference = exports.gameliftGameServerGroupLaunchTemplateToHclTerraform = exports.gameliftGameServerGroupLaunchTemplateToTerraform = exports.GameliftGameServerGroupInstanceDefinitionList = exports.GameliftGameServerGroupInstanceDefinitionOutputReference = exports.gameliftGameServerGroupInstanceDefinitionToHclTerraform = exports.gameliftGameServerGroupInstanceDefinitionToTerraform = exports.GameliftGameServerGroupAutoScalingPolicyOutputReference = exports.gameliftGameServerGroupAutoScalingPolicyToHclTerraform = exports.gameliftGameServerGroupAutoScalingPolicyToTerraform = exports.GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference = exports.gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToHclTerraform = exports.gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_value: cdktf.numberToTerraform(struct.targetValue),
    };
}
exports.gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToTerraform = gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToTerraform;
function gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target_value: {
            value: cdktf.numberToHclTerraform(struct.targetValue),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToHclTerraform = gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToHclTerraform;
class GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetValue = this._targetValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._targetValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._targetValue = value.targetValue;
        }
    }
    get targetValue() {
        return this.getNumberAttribute('target_value');
    }
    set targetValue(value) {
        this._targetValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetValueInput() {
        return this._targetValue;
    }
}
exports.GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference = GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference", version: "19.15.0" };
function gameliftGameServerGroupAutoScalingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        estimated_instance_warmup: cdktf.numberToTerraform(struct.estimatedInstanceWarmup),
        target_tracking_configuration: gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToTerraform(struct.targetTrackingConfiguration),
    };
}
exports.gameliftGameServerGroupAutoScalingPolicyToTerraform = gameliftGameServerGroupAutoScalingPolicyToTerraform;
function gameliftGameServerGroupAutoScalingPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        estimated_instance_warmup: {
            value: cdktf.numberToHclTerraform(struct.estimatedInstanceWarmup),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        target_tracking_configuration: {
            value: gameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationToHclTerraform(struct.targetTrackingConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameServerGroupAutoScalingPolicyToHclTerraform = gameliftGameServerGroupAutoScalingPolicyToHclTerraform;
class GameliftGameServerGroupAutoScalingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // target_tracking_configuration - computed: false, optional: false, required: true
        this._targetTrackingConfiguration = new GameliftGameServerGroupAutoScalingPolicyTargetTrackingConfigurationOutputReference(this, "target_tracking_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._estimatedInstanceWarmup !== undefined) {
            hasAnyValues = true;
            internalValueResult.estimatedInstanceWarmup = this._estimatedInstanceWarmup;
        }
        if (this._targetTrackingConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetTrackingConfiguration = this._targetTrackingConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._estimatedInstanceWarmup = undefined;
            this._targetTrackingConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._estimatedInstanceWarmup = value.estimatedInstanceWarmup;
            this._targetTrackingConfiguration.internalValue = value.targetTrackingConfiguration;
        }
    }
    get estimatedInstanceWarmup() {
        return this.getNumberAttribute('estimated_instance_warmup');
    }
    set estimatedInstanceWarmup(value) {
        this._estimatedInstanceWarmup = value;
    }
    resetEstimatedInstanceWarmup() {
        this._estimatedInstanceWarmup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get estimatedInstanceWarmupInput() {
        return this._estimatedInstanceWarmup;
    }
    get targetTrackingConfiguration() {
        return this._targetTrackingConfiguration;
    }
    putTargetTrackingConfiguration(value) {
        this._targetTrackingConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetTrackingConfigurationInput() {
        return this._targetTrackingConfiguration.internalValue;
    }
}
exports.GameliftGameServerGroupAutoScalingPolicyOutputReference = GameliftGameServerGroupAutoScalingPolicyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupAutoScalingPolicyOutputReference[_b] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupAutoScalingPolicyOutputReference", version: "19.15.0" };
function gameliftGameServerGroupInstanceDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        weighted_capacity: cdktf.stringToTerraform(struct.weightedCapacity),
    };
}
exports.gameliftGameServerGroupInstanceDefinitionToTerraform = gameliftGameServerGroupInstanceDefinitionToTerraform;
function gameliftGameServerGroupInstanceDefinitionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weighted_capacity: {
            value: cdktf.stringToHclTerraform(struct.weightedCapacity),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameServerGroupInstanceDefinitionToHclTerraform = gameliftGameServerGroupInstanceDefinitionToHclTerraform;
class GameliftGameServerGroupInstanceDefinitionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._weightedCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedCapacity = this._weightedCapacity;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._instanceType = undefined;
            this._weightedCapacity = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._instanceType = value.instanceType;
            this._weightedCapacity = value.weightedCapacity;
        }
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get weightedCapacity() {
        return this.getStringAttribute('weighted_capacity');
    }
    set weightedCapacity(value) {
        this._weightedCapacity = value;
    }
    resetWeightedCapacity() {
        this._weightedCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightedCapacityInput() {
        return this._weightedCapacity;
    }
}
exports.GameliftGameServerGroupInstanceDefinitionOutputReference = GameliftGameServerGroupInstanceDefinitionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupInstanceDefinitionOutputReference[_c] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupInstanceDefinitionOutputReference", version: "19.15.0" };
class GameliftGameServerGroupInstanceDefinitionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GameliftGameServerGroupInstanceDefinitionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GameliftGameServerGroupInstanceDefinitionList = GameliftGameServerGroupInstanceDefinitionList;
_d = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupInstanceDefinitionList[_d] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupInstanceDefinitionList", version: "19.15.0" };
function gameliftGameServerGroupLaunchTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        name: cdktf.stringToTerraform(struct.name),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.gameliftGameServerGroupLaunchTemplateToTerraform = gameliftGameServerGroupLaunchTemplateToTerraform;
function gameliftGameServerGroupLaunchTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        version: {
            value: cdktf.stringToHclTerraform(struct.version),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameServerGroupLaunchTemplateToHclTerraform = gameliftGameServerGroupLaunchTemplateToHclTerraform;
class GameliftGameServerGroupLaunchTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._id = undefined;
            this._name = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._id = value.id;
            this._name = value.name;
            this._version = value.version;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.GameliftGameServerGroupLaunchTemplateOutputReference = GameliftGameServerGroupLaunchTemplateOutputReference;
_e = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupLaunchTemplateOutputReference[_e] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupLaunchTemplateOutputReference", version: "19.15.0" };
function gameliftGameServerGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.gameliftGameServerGroupTimeoutsToTerraform = gameliftGameServerGroupTimeoutsToTerraform;
function gameliftGameServerGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameServerGroupTimeoutsToHclTerraform = gameliftGameServerGroupTimeoutsToHclTerraform;
class GameliftGameServerGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.GameliftGameServerGroupTimeoutsOutputReference = GameliftGameServerGroupTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroupTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroupTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_server_group aws_gamelift_game_server_group}
*/
class GameliftGameServerGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GameliftGameServerGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GameliftGameServerGroup to import
    * @param importFromId The id of the existing GameliftGameServerGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_server_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GameliftGameServerGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_gamelift_game_server_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_server_group aws_gamelift_game_server_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GameliftGameServerGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_gamelift_game_server_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auto_scaling_policy - computed: false, optional: true, required: false
        this._autoScalingPolicy = new GameliftGameServerGroupAutoScalingPolicyOutputReference(this, "auto_scaling_policy");
        // instance_definition - computed: false, optional: false, required: true
        this._instanceDefinition = new GameliftGameServerGroupInstanceDefinitionList(this, "instance_definition", true);
        // launch_template - computed: false, optional: false, required: true
        this._launchTemplate = new GameliftGameServerGroupLaunchTemplateOutputReference(this, "launch_template");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GameliftGameServerGroupTimeoutsOutputReference(this, "timeouts");
        this._balancingStrategy = config.balancingStrategy;
        this._gameServerGroupName = config.gameServerGroupName;
        this._gameServerProtectionPolicy = config.gameServerProtectionPolicy;
        this._id = config.id;
        this._maxSize = config.maxSize;
        this._minSize = config.minSize;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcSubnets = config.vpcSubnets;
        this._autoScalingPolicy.internalValue = config.autoScalingPolicy;
        this._instanceDefinition.internalValue = config.instanceDefinition;
        this._launchTemplate.internalValue = config.launchTemplate;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // auto_scaling_group_arn - computed: true, optional: false, required: false
    get autoScalingGroupArn() {
        return this.getStringAttribute('auto_scaling_group_arn');
    }
    get balancingStrategy() {
        return this.getStringAttribute('balancing_strategy');
    }
    set balancingStrategy(value) {
        this._balancingStrategy = value;
    }
    resetBalancingStrategy() {
        this._balancingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get balancingStrategyInput() {
        return this._balancingStrategy;
    }
    get gameServerGroupName() {
        return this.getStringAttribute('game_server_group_name');
    }
    set gameServerGroupName(value) {
        this._gameServerGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get gameServerGroupNameInput() {
        return this._gameServerGroupName;
    }
    get gameServerProtectionPolicy() {
        return this.getStringAttribute('game_server_protection_policy');
    }
    set gameServerProtectionPolicy(value) {
        this._gameServerProtectionPolicy = value;
    }
    resetGameServerProtectionPolicy() {
        this._gameServerProtectionPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gameServerProtectionPolicyInput() {
        return this._gameServerProtectionPolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    set maxSize(value) {
        this._maxSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxSizeInput() {
        return this._maxSize;
    }
    get minSize() {
        return this.getNumberAttribute('min_size');
    }
    set minSize(value) {
        this._minSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get minSizeInput() {
        return this._minSize;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get vpcSubnets() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_subnets'));
    }
    set vpcSubnets(value) {
        this._vpcSubnets = value;
    }
    resetVpcSubnets() {
        this._vpcSubnets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcSubnetsInput() {
        return this._vpcSubnets;
    }
    get autoScalingPolicy() {
        return this._autoScalingPolicy;
    }
    putAutoScalingPolicy(value) {
        this._autoScalingPolicy.internalValue = value;
    }
    resetAutoScalingPolicy() {
        this._autoScalingPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoScalingPolicyInput() {
        return this._autoScalingPolicy.internalValue;
    }
    get instanceDefinition() {
        return this._instanceDefinition;
    }
    putInstanceDefinition(value) {
        this._instanceDefinition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceDefinitionInput() {
        return this._instanceDefinition.internalValue;
    }
    get launchTemplate() {
        return this._launchTemplate;
    }
    putLaunchTemplate(value) {
        this._launchTemplate.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get launchTemplateInput() {
        return this._launchTemplate.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            balancing_strategy: cdktf.stringToTerraform(this._balancingStrategy),
            game_server_group_name: cdktf.stringToTerraform(this._gameServerGroupName),
            game_server_protection_policy: cdktf.stringToTerraform(this._gameServerProtectionPolicy),
            id: cdktf.stringToTerraform(this._id),
            max_size: cdktf.numberToTerraform(this._maxSize),
            min_size: cdktf.numberToTerraform(this._minSize),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vpc_subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(this._vpcSubnets),
            auto_scaling_policy: gameliftGameServerGroupAutoScalingPolicyToTerraform(this._autoScalingPolicy.internalValue),
            instance_definition: cdktf.listMapper(gameliftGameServerGroupInstanceDefinitionToTerraform, true)(this._instanceDefinition.internalValue),
            launch_template: gameliftGameServerGroupLaunchTemplateToTerraform(this._launchTemplate.internalValue),
            timeouts: gameliftGameServerGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            balancing_strategy: {
                value: cdktf.stringToHclTerraform(this._balancingStrategy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            game_server_group_name: {
                value: cdktf.stringToHclTerraform(this._gameServerGroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            game_server_protection_policy: {
                value: cdktf.stringToHclTerraform(this._gameServerProtectionPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            max_size: {
                value: cdktf.numberToHclTerraform(this._maxSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            min_size: {
                value: cdktf.numberToHclTerraform(this._minSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            vpc_subnets: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._vpcSubnets),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            auto_scaling_policy: {
                value: gameliftGameServerGroupAutoScalingPolicyToHclTerraform(this._autoScalingPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GameliftGameServerGroupAutoScalingPolicyList",
            },
            instance_definition: {
                value: cdktf.listMapperHcl(gameliftGameServerGroupInstanceDefinitionToHclTerraform, true)(this._instanceDefinition.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "GameliftGameServerGroupInstanceDefinitionList",
            },
            launch_template: {
                value: gameliftGameServerGroupLaunchTemplateToHclTerraform(this._launchTemplate.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GameliftGameServerGroupLaunchTemplateList",
            },
            timeouts: {
                value: gameliftGameServerGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "GameliftGameServerGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GameliftGameServerGroup = GameliftGameServerGroup;
_g = JSII_RTTI_SYMBOL_1;
GameliftGameServerGroup[_g] = { fqn: "@cdktf/provider-aws.gameliftGameServerGroup.GameliftGameServerGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GameliftGameServerGroup.tfResourceType = "aws_gamelift_game_server_group";
