// @cdktf/provider-aws/lib/data-aws-fsx-ontap-file-system/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsFsxOntapFileSystem = exports.DataAwsFsxOntapFileSystemEndpointsList = exports.DataAwsFsxOntapFileSystemEndpointsOutputReference = exports.dataAwsFsxOntapFileSystemEndpointsToHclTerraform = exports.dataAwsFsxOntapFileSystemEndpointsToTerraform = exports.DataAwsFsxOntapFileSystemEndpointsManagementList = exports.DataAwsFsxOntapFileSystemEndpointsManagementOutputReference = exports.dataAwsFsxOntapFileSystemEndpointsManagementToHclTerraform = exports.dataAwsFsxOntapFileSystemEndpointsManagementToTerraform = exports.DataAwsFsxOntapFileSystemEndpointsInterclusterList = exports.DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference = exports.dataAwsFsxOntapFileSystemEndpointsInterclusterToHclTerraform = exports.dataAwsFsxOntapFileSystemEndpointsInterclusterToTerraform = exports.DataAwsFsxOntapFileSystemDiskIopsConfigurationList = exports.DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference = exports.dataAwsFsxOntapFileSystemDiskIopsConfigurationToHclTerraform = exports.dataAwsFsxOntapFileSystemDiskIopsConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsFsxOntapFileSystemDiskIopsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapFileSystemDiskIopsConfigurationToTerraform = dataAwsFsxOntapFileSystemDiskIopsConfigurationToTerraform;
function dataAwsFsxOntapFileSystemDiskIopsConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapFileSystemDiskIopsConfigurationToHclTerraform = dataAwsFsxOntapFileSystemDiskIopsConfigurationToHclTerraform;
class DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference = DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference", version: "19.15.0" };
class DataAwsFsxOntapFileSystemDiskIopsConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapFileSystemDiskIopsConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapFileSystemDiskIopsConfigurationList = DataAwsFsxOntapFileSystemDiskIopsConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemDiskIopsConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemDiskIopsConfigurationList", version: "19.15.0" };
function dataAwsFsxOntapFileSystemEndpointsInterclusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapFileSystemEndpointsInterclusterToTerraform = dataAwsFsxOntapFileSystemEndpointsInterclusterToTerraform;
function dataAwsFsxOntapFileSystemEndpointsInterclusterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapFileSystemEndpointsInterclusterToHclTerraform = dataAwsFsxOntapFileSystemEndpointsInterclusterToHclTerraform;
class DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference = DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference", version: "19.15.0" };
class DataAwsFsxOntapFileSystemEndpointsInterclusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapFileSystemEndpointsInterclusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsInterclusterList = DataAwsFsxOntapFileSystemEndpointsInterclusterList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsInterclusterList[_d] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsInterclusterList", version: "19.15.0" };
function dataAwsFsxOntapFileSystemEndpointsManagementToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapFileSystemEndpointsManagementToTerraform = dataAwsFsxOntapFileSystemEndpointsManagementToTerraform;
function dataAwsFsxOntapFileSystemEndpointsManagementToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapFileSystemEndpointsManagementToHclTerraform = dataAwsFsxOntapFileSystemEndpointsManagementToHclTerraform;
class DataAwsFsxOntapFileSystemEndpointsManagementOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsManagementOutputReference = DataAwsFsxOntapFileSystemEndpointsManagementOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsManagementOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsManagementOutputReference", version: "19.15.0" };
class DataAwsFsxOntapFileSystemEndpointsManagementList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapFileSystemEndpointsManagementOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsManagementList = DataAwsFsxOntapFileSystemEndpointsManagementList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsManagementList[_f] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsManagementList", version: "19.15.0" };
function dataAwsFsxOntapFileSystemEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapFileSystemEndpointsToTerraform = dataAwsFsxOntapFileSystemEndpointsToTerraform;
function dataAwsFsxOntapFileSystemEndpointsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapFileSystemEndpointsToHclTerraform = dataAwsFsxOntapFileSystemEndpointsToHclTerraform;
class DataAwsFsxOntapFileSystemEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // intercluster - computed: true, optional: false, required: false
        this._intercluster = new DataAwsFsxOntapFileSystemEndpointsInterclusterList(this, "intercluster", false);
        // management - computed: true, optional: false, required: false
        this._management = new DataAwsFsxOntapFileSystemEndpointsManagementList(this, "management", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get intercluster() {
        return this._intercluster;
    }
    get management() {
        return this._management;
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsOutputReference = DataAwsFsxOntapFileSystemEndpointsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsOutputReference", version: "19.15.0" };
class DataAwsFsxOntapFileSystemEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapFileSystemEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapFileSystemEndpointsList = DataAwsFsxOntapFileSystemEndpointsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystemEndpointsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystemEndpointsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_file_system aws_fsx_ontap_file_system}
*/
class DataAwsFsxOntapFileSystem extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsFsxOntapFileSystem resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsFsxOntapFileSystem to import
    * @param importFromId The id of the existing DataAwsFsxOntapFileSystem that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_file_system#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsFsxOntapFileSystem to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_ontap_file_system", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_file_system aws_fsx_ontap_file_system} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsFsxOntapFileSystemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_ontap_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // disk_iops_configuration - computed: true, optional: false, required: false
        this._diskIopsConfiguration = new DataAwsFsxOntapFileSystemDiskIopsConfigurationList(this, "disk_iops_configuration", false);
        // endpoints - computed: true, optional: false, required: false
        this._endpoints = new DataAwsFsxOntapFileSystemEndpointsList(this, "endpoints", false);
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // automatic_backup_retention_days - computed: true, optional: false, required: false
    get automaticBackupRetentionDays() {
        return this.getNumberAttribute('automatic_backup_retention_days');
    }
    // daily_automatic_backup_start_time - computed: true, optional: false, required: false
    get dailyAutomaticBackupStartTime() {
        return this.getStringAttribute('daily_automatic_backup_start_time');
    }
    // deployment_type - computed: true, optional: false, required: false
    get deploymentType() {
        return this.getStringAttribute('deployment_type');
    }
    get diskIopsConfiguration() {
        return this._diskIopsConfiguration;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // endpoint_ip_address_range - computed: true, optional: false, required: false
    get endpointIpAddressRange() {
        return this.getStringAttribute('endpoint_ip_address_range');
    }
    get endpoints() {
        return this._endpoints;
    }
    // ha_pairs - computed: true, optional: false, required: false
    get haPairs() {
        return this.getNumberAttribute('ha_pairs');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // network_interface_ids - computed: true, optional: false, required: false
    get networkInterfaceIds() {
        return this.getListAttribute('network_interface_ids');
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // preferred_subnet_id - computed: true, optional: false, required: false
    get preferredSubnetId() {
        return this.getStringAttribute('preferred_subnet_id');
    }
    // route_table_ids - computed: true, optional: false, required: false
    get routeTableIds() {
        return cdktf.Fn.tolist(this.getListAttribute('route_table_ids'));
    }
    // storage_capacity - computed: true, optional: false, required: false
    get storageCapacity() {
        return this.getNumberAttribute('storage_capacity');
    }
    // storage_type - computed: true, optional: false, required: false
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // throughput_capacity - computed: true, optional: false, required: false
    get throughputCapacity() {
        return this.getNumberAttribute('throughput_capacity');
    }
    // throughput_capacity_per_ha_pair - computed: true, optional: false, required: false
    get throughputCapacityPerHaPair() {
        return this.getNumberAttribute('throughput_capacity_per_ha_pair');
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // weekly_maintenance_start_time - computed: true, optional: false, required: false
    get weeklyMaintenanceStartTime() {
        return this.getStringAttribute('weekly_maintenance_start_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsFsxOntapFileSystem = DataAwsFsxOntapFileSystem;
_j = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapFileSystem[_j] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapFileSystem.DataAwsFsxOntapFileSystem", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsFsxOntapFileSystem.tfResourceType = "aws_fsx_ontap_file_system";
