// @cdktf/provider-aws/lib/licensemanager-grant-accepter/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicensemanagerGrantAccepter = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_grant_accepter aws_licensemanager_grant_accepter}
*/
class LicensemanagerGrantAccepter extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LicensemanagerGrantAccepter resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LicensemanagerGrantAccepter to import
    * @param importFromId The id of the existing LicensemanagerGrantAccepter that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_grant_accepter#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LicensemanagerGrantAccepter to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_licensemanager_grant_accepter", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_grant_accepter aws_licensemanager_grant_accepter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LicensemanagerGrantAccepterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_licensemanager_grant_accepter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._grantArn = config.grantArn;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // allowed_operations - computed: true, optional: false, required: false
    get allowedOperations() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_operations'));
    }
    get grantArn() {
        return this.getStringAttribute('grant_arn');
    }
    set grantArn(value) {
        this._grantArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get grantArnInput() {
        return this._grantArn;
    }
    // home_region - computed: true, optional: false, required: false
    get homeRegion() {
        return this.getStringAttribute('home_region');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // license_arn - computed: true, optional: false, required: false
    get licenseArn() {
        return this.getStringAttribute('license_arn');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // parent_arn - computed: true, optional: false, required: false
    get parentArn() {
        return this.getStringAttribute('parent_arn');
    }
    // principal - computed: true, optional: false, required: false
    get principal() {
        return this.getStringAttribute('principal');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            grant_arn: cdktf.stringToTerraform(this._grantArn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            grant_arn: {
                value: cdktf.stringToHclTerraform(this._grantArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LicensemanagerGrantAccepter = LicensemanagerGrantAccepter;
_a = JSII_RTTI_SYMBOL_1;
LicensemanagerGrantAccepter[_a] = { fqn: "@cdktf/provider-aws.licensemanagerGrantAccepter.LicensemanagerGrantAccepter", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LicensemanagerGrantAccepter.tfResourceType = "aws_licensemanager_grant_accepter";
