// @cdktf/provider-aws/lib/codepipeline/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Codepipeline = exports.CodepipelineVariableList = exports.CodepipelineVariableOutputReference = exports.codepipelineVariableToHclTerraform = exports.codepipelineVariableToTerraform = exports.CodepipelineTriggerList = exports.CodepipelineTriggerOutputReference = exports.codepipelineTriggerToHclTerraform = exports.codepipelineTriggerToTerraform = exports.CodepipelineTriggerGitConfigurationOutputReference = exports.codepipelineTriggerGitConfigurationToHclTerraform = exports.codepipelineTriggerGitConfigurationToTerraform = exports.CodepipelineTriggerGitConfigurationPushList = exports.CodepipelineTriggerGitConfigurationPushOutputReference = exports.codepipelineTriggerGitConfigurationPushToHclTerraform = exports.codepipelineTriggerGitConfigurationPushToTerraform = exports.CodepipelineTriggerGitConfigurationPushTagsOutputReference = exports.codepipelineTriggerGitConfigurationPushTagsToHclTerraform = exports.codepipelineTriggerGitConfigurationPushTagsToTerraform = exports.CodepipelineTriggerGitConfigurationPushFilePathsOutputReference = exports.codepipelineTriggerGitConfigurationPushFilePathsToHclTerraform = exports.codepipelineTriggerGitConfigurationPushFilePathsToTerraform = exports.CodepipelineTriggerGitConfigurationPushBranchesOutputReference = exports.codepipelineTriggerGitConfigurationPushBranchesToHclTerraform = exports.codepipelineTriggerGitConfigurationPushBranchesToTerraform = exports.CodepipelineTriggerGitConfigurationPullRequestList = exports.CodepipelineTriggerGitConfigurationPullRequestOutputReference = exports.codepipelineTriggerGitConfigurationPullRequestToHclTerraform = exports.codepipelineTriggerGitConfigurationPullRequestToTerraform = exports.CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference = exports.codepipelineTriggerGitConfigurationPullRequestFilePathsToHclTerraform = exports.codepipelineTriggerGitConfigurationPullRequestFilePathsToTerraform = exports.CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference = exports.codepipelineTriggerGitConfigurationPullRequestBranchesToHclTerraform = exports.codepipelineTriggerGitConfigurationPullRequestBranchesToTerraform = exports.CodepipelineStageList = exports.CodepipelineStageOutputReference = exports.codepipelineStageToHclTerraform = exports.codepipelineStageToTerraform = exports.CodepipelineStageActionList = exports.CodepipelineStageActionOutputReference = exports.codepipelineStageActionToHclTerraform = exports.codepipelineStageActionToTerraform = exports.CodepipelineArtifactStoreList = exports.CodepipelineArtifactStoreOutputReference = exports.codepipelineArtifactStoreToHclTerraform = exports.codepipelineArtifactStoreToTerraform = exports.CodepipelineArtifactStoreEncryptionKeyOutputReference = exports.codepipelineArtifactStoreEncryptionKeyToHclTerraform = exports.codepipelineArtifactStoreEncryptionKeyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codepipelineArtifactStoreEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codepipelineArtifactStoreEncryptionKeyToTerraform = codepipelineArtifactStoreEncryptionKeyToTerraform;
function codepipelineArtifactStoreEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineArtifactStoreEncryptionKeyToHclTerraform = codepipelineArtifactStoreEncryptionKeyToHclTerraform;
class CodepipelineArtifactStoreEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._id = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._id = value.id;
            this._type = value.type;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CodepipelineArtifactStoreEncryptionKeyOutputReference = CodepipelineArtifactStoreEncryptionKeyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodepipelineArtifactStoreEncryptionKeyOutputReference[_a] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineArtifactStoreEncryptionKeyOutputReference", version: "19.15.0" };
function codepipelineArtifactStoreToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        location: cdktf.stringToTerraform(struct.location),
        region: cdktf.stringToTerraform(struct.region),
        type: cdktf.stringToTerraform(struct.type),
        encryption_key: codepipelineArtifactStoreEncryptionKeyToTerraform(struct.encryptionKey),
    };
}
exports.codepipelineArtifactStoreToTerraform = codepipelineArtifactStoreToTerraform;
function codepipelineArtifactStoreToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        encryption_key: {
            value: codepipelineArtifactStoreEncryptionKeyToHclTerraform(struct.encryptionKey),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineArtifactStoreEncryptionKeyList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineArtifactStoreToHclTerraform = codepipelineArtifactStoreToHclTerraform;
class CodepipelineArtifactStoreOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // encryption_key - computed: false, optional: true, required: false
        this._encryptionKey = new CodepipelineArtifactStoreEncryptionKeyOutputReference(this, "encryption_key");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._encryptionKey?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._location = undefined;
            this._region = undefined;
            this._type = undefined;
            this._encryptionKey.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._location = value.location;
            this._region = value.region;
            this._type = value.type;
            this._encryptionKey.internalValue = value.encryptionKey;
        }
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get encryptionKey() {
        return this._encryptionKey;
    }
    putEncryptionKey(value) {
        this._encryptionKey.internalValue = value;
    }
    resetEncryptionKey() {
        this._encryptionKey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey.internalValue;
    }
}
exports.CodepipelineArtifactStoreOutputReference = CodepipelineArtifactStoreOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CodepipelineArtifactStoreOutputReference[_b] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineArtifactStoreOutputReference", version: "19.15.0" };
class CodepipelineArtifactStoreList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineArtifactStoreOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineArtifactStoreList = CodepipelineArtifactStoreList;
_c = JSII_RTTI_SYMBOL_1;
CodepipelineArtifactStoreList[_c] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineArtifactStoreList", version: "19.15.0" };
function codepipelineStageActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        category: cdktf.stringToTerraform(struct.category),
        configuration: cdktf.hashMapper(cdktf.stringToTerraform)(struct.configuration),
        input_artifacts: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.inputArtifacts),
        name: cdktf.stringToTerraform(struct.name),
        namespace: cdktf.stringToTerraform(struct.namespace),
        output_artifacts: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.outputArtifacts),
        owner: cdktf.stringToTerraform(struct.owner),
        provider: cdktf.stringToTerraform(struct.provider),
        region: cdktf.stringToTerraform(struct.region),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        run_order: cdktf.numberToTerraform(struct.runOrder),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.codepipelineStageActionToTerraform = codepipelineStageActionToTerraform;
function codepipelineStageActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        category: {
            value: cdktf.stringToHclTerraform(struct.category),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        configuration: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.configuration),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        input_artifacts: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.inputArtifacts),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        namespace: {
            value: cdktf.stringToHclTerraform(struct.namespace),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        output_artifacts: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.outputArtifacts),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        owner: {
            value: cdktf.stringToHclTerraform(struct.owner),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        provider: {
            value: cdktf.stringToHclTerraform(struct.provider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        run_order: {
            value: cdktf.numberToHclTerraform(struct.runOrder),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        version: {
            value: cdktf.stringToHclTerraform(struct.version),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineStageActionToHclTerraform = codepipelineStageActionToHclTerraform;
class CodepipelineStageActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._category !== undefined) {
            hasAnyValues = true;
            internalValueResult.category = this._category;
        }
        if (this._configuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.configuration = this._configuration;
        }
        if (this._inputArtifacts !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputArtifacts = this._inputArtifacts;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._namespace !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespace = this._namespace;
        }
        if (this._outputArtifacts !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputArtifacts = this._outputArtifacts;
        }
        if (this._owner !== undefined) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner;
        }
        if (this._provider !== undefined) {
            hasAnyValues = true;
            internalValueResult.provider = this._provider;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._runOrder !== undefined) {
            hasAnyValues = true;
            internalValueResult.runOrder = this._runOrder;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._category = undefined;
            this._configuration = undefined;
            this._inputArtifacts = undefined;
            this._name = undefined;
            this._namespace = undefined;
            this._outputArtifacts = undefined;
            this._owner = undefined;
            this._provider = undefined;
            this._region = undefined;
            this._roleArn = undefined;
            this._runOrder = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._category = value.category;
            this._configuration = value.configuration;
            this._inputArtifacts = value.inputArtifacts;
            this._name = value.name;
            this._namespace = value.namespace;
            this._outputArtifacts = value.outputArtifacts;
            this._owner = value.owner;
            this._provider = value.provider;
            this._region = value.region;
            this._roleArn = value.roleArn;
            this._runOrder = value.runOrder;
            this._version = value.version;
        }
    }
    get category() {
        return this.getStringAttribute('category');
    }
    set category(value) {
        this._category = value;
    }
    // Temporarily expose input value. Use with caution.
    get categoryInput() {
        return this._category;
    }
    get configuration() {
        return this.getStringMapAttribute('configuration');
    }
    set configuration(value) {
        this._configuration = value;
    }
    resetConfiguration() {
        this._configuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration;
    }
    get inputArtifacts() {
        return this.getListAttribute('input_artifacts');
    }
    set inputArtifacts(value) {
        this._inputArtifacts = value;
    }
    resetInputArtifacts() {
        this._inputArtifacts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputArtifactsInput() {
        return this._inputArtifacts;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get outputArtifacts() {
        return this.getListAttribute('output_artifacts');
    }
    set outputArtifacts(value) {
        this._outputArtifacts = value;
    }
    resetOutputArtifacts() {
        this._outputArtifacts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputArtifactsInput() {
        return this._outputArtifacts;
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
    get provider() {
        return this.getStringAttribute('provider');
    }
    set provider(value) {
        this._provider = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerInput() {
        return this._provider;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get runOrder() {
        return this.getNumberAttribute('run_order');
    }
    set runOrder(value) {
        this._runOrder = value;
    }
    resetRunOrder() {
        this._runOrder = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runOrderInput() {
        return this._runOrder;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CodepipelineStageActionOutputReference = CodepipelineStageActionOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CodepipelineStageActionOutputReference[_d] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineStageActionOutputReference", version: "19.15.0" };
class CodepipelineStageActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineStageActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineStageActionList = CodepipelineStageActionList;
_e = JSII_RTTI_SYMBOL_1;
CodepipelineStageActionList[_e] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineStageActionList", version: "19.15.0" };
function codepipelineStageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        action: cdktf.listMapper(codepipelineStageActionToTerraform, true)(struct.action),
    };
}
exports.codepipelineStageToTerraform = codepipelineStageToTerraform;
function codepipelineStageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        action: {
            value: cdktf.listMapperHcl(codepipelineStageActionToHclTerraform, true)(struct.action),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineStageActionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineStageToHclTerraform = codepipelineStageToHclTerraform;
class CodepipelineStageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new CodepipelineStageActionList(this, "action", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._action.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._action.internalValue = value.action;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
}
exports.CodepipelineStageOutputReference = CodepipelineStageOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CodepipelineStageOutputReference[_f] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineStageOutputReference", version: "19.15.0" };
class CodepipelineStageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineStageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineStageList = CodepipelineStageList;
_g = JSII_RTTI_SYMBOL_1;
CodepipelineStageList[_g] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineStageList", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPullRequestBranchesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludes),
        includes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includes),
    };
}
exports.codepipelineTriggerGitConfigurationPullRequestBranchesToTerraform = codepipelineTriggerGitConfigurationPullRequestBranchesToTerraform;
function codepipelineTriggerGitConfigurationPullRequestBranchesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excludes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        includes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPullRequestBranchesToHclTerraform = codepipelineTriggerGitConfigurationPullRequestBranchesToHclTerraform;
class CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludes = this._excludes;
        }
        if (this._includes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includes = this._includes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes = undefined;
            this._includes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes = value.excludes;
            this._includes = value.includes;
        }
    }
    get excludes() {
        return this.getListAttribute('excludes');
    }
    set excludes(value) {
        this._excludes = value;
    }
    resetExcludes() {
        this._excludes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes;
    }
    get includes() {
        return this.getListAttribute('includes');
    }
    set includes(value) {
        this._includes = value;
    }
    resetIncludes() {
        this._includes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes;
    }
}
exports.CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference = CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference[_h] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPullRequestFilePathsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludes),
        includes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includes),
    };
}
exports.codepipelineTriggerGitConfigurationPullRequestFilePathsToTerraform = codepipelineTriggerGitConfigurationPullRequestFilePathsToTerraform;
function codepipelineTriggerGitConfigurationPullRequestFilePathsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excludes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        includes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPullRequestFilePathsToHclTerraform = codepipelineTriggerGitConfigurationPullRequestFilePathsToHclTerraform;
class CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludes = this._excludes;
        }
        if (this._includes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includes = this._includes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes = undefined;
            this._includes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes = value.excludes;
            this._includes = value.includes;
        }
    }
    get excludes() {
        return this.getListAttribute('excludes');
    }
    set excludes(value) {
        this._excludes = value;
    }
    resetExcludes() {
        this._excludes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes;
    }
    get includes() {
        return this.getListAttribute('includes');
    }
    set includes(value) {
        this._includes = value;
    }
    resetIncludes() {
        this._includes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes;
    }
}
exports.CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference = CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference[_j] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPullRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
        branches: codepipelineTriggerGitConfigurationPullRequestBranchesToTerraform(struct.branches),
        file_paths: codepipelineTriggerGitConfigurationPullRequestFilePathsToTerraform(struct.filePaths),
    };
}
exports.codepipelineTriggerGitConfigurationPullRequestToTerraform = codepipelineTriggerGitConfigurationPullRequestToTerraform;
function codepipelineTriggerGitConfigurationPullRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        branches: {
            value: codepipelineTriggerGitConfigurationPullRequestBranchesToHclTerraform(struct.branches),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPullRequestBranchesList",
        },
        file_paths: {
            value: codepipelineTriggerGitConfigurationPullRequestFilePathsToHclTerraform(struct.filePaths),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPullRequestFilePathsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPullRequestToHclTerraform = codepipelineTriggerGitConfigurationPullRequestToHclTerraform;
class CodepipelineTriggerGitConfigurationPullRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // branches - computed: false, optional: true, required: false
        this._branches = new CodepipelineTriggerGitConfigurationPullRequestBranchesOutputReference(this, "branches");
        // file_paths - computed: false, optional: true, required: false
        this._filePaths = new CodepipelineTriggerGitConfigurationPullRequestFilePathsOutputReference(this, "file_paths");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        if (this._branches?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.branches = this._branches?.internalValue;
        }
        if (this._filePaths?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filePaths = this._filePaths?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._events = undefined;
            this._branches.internalValue = undefined;
            this._filePaths.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._events = value.events;
            this._branches.internalValue = value.branches;
            this._filePaths.internalValue = value.filePaths;
        }
    }
    get events() {
        return this.getListAttribute('events');
    }
    set events(value) {
        this._events = value;
    }
    resetEvents() {
        this._events = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
    get branches() {
        return this._branches;
    }
    putBranches(value) {
        this._branches.internalValue = value;
    }
    resetBranches() {
        this._branches.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchesInput() {
        return this._branches.internalValue;
    }
    get filePaths() {
        return this._filePaths;
    }
    putFilePaths(value) {
        this._filePaths.internalValue = value;
    }
    resetFilePaths() {
        this._filePaths.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filePathsInput() {
        return this._filePaths.internalValue;
    }
}
exports.CodepipelineTriggerGitConfigurationPullRequestOutputReference = CodepipelineTriggerGitConfigurationPullRequestOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPullRequestOutputReference[_k] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPullRequestOutputReference", version: "19.15.0" };
class CodepipelineTriggerGitConfigurationPullRequestList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineTriggerGitConfigurationPullRequestOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineTriggerGitConfigurationPullRequestList = CodepipelineTriggerGitConfigurationPullRequestList;
_l = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPullRequestList[_l] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPullRequestList", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPushBranchesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludes),
        includes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includes),
    };
}
exports.codepipelineTriggerGitConfigurationPushBranchesToTerraform = codepipelineTriggerGitConfigurationPushBranchesToTerraform;
function codepipelineTriggerGitConfigurationPushBranchesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excludes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        includes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPushBranchesToHclTerraform = codepipelineTriggerGitConfigurationPushBranchesToHclTerraform;
class CodepipelineTriggerGitConfigurationPushBranchesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludes = this._excludes;
        }
        if (this._includes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includes = this._includes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes = undefined;
            this._includes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes = value.excludes;
            this._includes = value.includes;
        }
    }
    get excludes() {
        return this.getListAttribute('excludes');
    }
    set excludes(value) {
        this._excludes = value;
    }
    resetExcludes() {
        this._excludes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes;
    }
    get includes() {
        return this.getListAttribute('includes');
    }
    set includes(value) {
        this._includes = value;
    }
    resetIncludes() {
        this._includes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes;
    }
}
exports.CodepipelineTriggerGitConfigurationPushBranchesOutputReference = CodepipelineTriggerGitConfigurationPushBranchesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPushBranchesOutputReference[_m] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPushBranchesOutputReference", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPushFilePathsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludes),
        includes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includes),
    };
}
exports.codepipelineTriggerGitConfigurationPushFilePathsToTerraform = codepipelineTriggerGitConfigurationPushFilePathsToTerraform;
function codepipelineTriggerGitConfigurationPushFilePathsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excludes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        includes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPushFilePathsToHclTerraform = codepipelineTriggerGitConfigurationPushFilePathsToHclTerraform;
class CodepipelineTriggerGitConfigurationPushFilePathsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludes = this._excludes;
        }
        if (this._includes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includes = this._includes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes = undefined;
            this._includes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes = value.excludes;
            this._includes = value.includes;
        }
    }
    get excludes() {
        return this.getListAttribute('excludes');
    }
    set excludes(value) {
        this._excludes = value;
    }
    resetExcludes() {
        this._excludes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes;
    }
    get includes() {
        return this.getListAttribute('includes');
    }
    set includes(value) {
        this._includes = value;
    }
    resetIncludes() {
        this._includes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes;
    }
}
exports.CodepipelineTriggerGitConfigurationPushFilePathsOutputReference = CodepipelineTriggerGitConfigurationPushFilePathsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPushFilePathsOutputReference[_o] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPushFilePathsOutputReference", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPushTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludes),
        includes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includes),
    };
}
exports.codepipelineTriggerGitConfigurationPushTagsToTerraform = codepipelineTriggerGitConfigurationPushTagsToTerraform;
function codepipelineTriggerGitConfigurationPushTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excludes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        includes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPushTagsToHclTerraform = codepipelineTriggerGitConfigurationPushTagsToHclTerraform;
class CodepipelineTriggerGitConfigurationPushTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludes = this._excludes;
        }
        if (this._includes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includes = this._includes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes = undefined;
            this._includes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes = value.excludes;
            this._includes = value.includes;
        }
    }
    get excludes() {
        return this.getListAttribute('excludes');
    }
    set excludes(value) {
        this._excludes = value;
    }
    resetExcludes() {
        this._excludes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes;
    }
    get includes() {
        return this.getListAttribute('includes');
    }
    set includes(value) {
        this._includes = value;
    }
    resetIncludes() {
        this._includes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes;
    }
}
exports.CodepipelineTriggerGitConfigurationPushTagsOutputReference = CodepipelineTriggerGitConfigurationPushTagsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPushTagsOutputReference[_p] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPushTagsOutputReference", version: "19.15.0" };
function codepipelineTriggerGitConfigurationPushToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branches: codepipelineTriggerGitConfigurationPushBranchesToTerraform(struct.branches),
        file_paths: codepipelineTriggerGitConfigurationPushFilePathsToTerraform(struct.filePaths),
        tags: codepipelineTriggerGitConfigurationPushTagsToTerraform(struct.tags),
    };
}
exports.codepipelineTriggerGitConfigurationPushToTerraform = codepipelineTriggerGitConfigurationPushToTerraform;
function codepipelineTriggerGitConfigurationPushToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        branches: {
            value: codepipelineTriggerGitConfigurationPushBranchesToHclTerraform(struct.branches),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPushBranchesList",
        },
        file_paths: {
            value: codepipelineTriggerGitConfigurationPushFilePathsToHclTerraform(struct.filePaths),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPushFilePathsList",
        },
        tags: {
            value: codepipelineTriggerGitConfigurationPushTagsToHclTerraform(struct.tags),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPushTagsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationPushToHclTerraform = codepipelineTriggerGitConfigurationPushToHclTerraform;
class CodepipelineTriggerGitConfigurationPushOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // branches - computed: false, optional: true, required: false
        this._branches = new CodepipelineTriggerGitConfigurationPushBranchesOutputReference(this, "branches");
        // file_paths - computed: false, optional: true, required: false
        this._filePaths = new CodepipelineTriggerGitConfigurationPushFilePathsOutputReference(this, "file_paths");
        // tags - computed: false, optional: true, required: false
        this._tags = new CodepipelineTriggerGitConfigurationPushTagsOutputReference(this, "tags");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branches?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.branches = this._branches?.internalValue;
        }
        if (this._filePaths?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filePaths = this._filePaths?.internalValue;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._branches.internalValue = undefined;
            this._filePaths.internalValue = undefined;
            this._tags.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._branches.internalValue = value.branches;
            this._filePaths.internalValue = value.filePaths;
            this._tags.internalValue = value.tags;
        }
    }
    get branches() {
        return this._branches;
    }
    putBranches(value) {
        this._branches.internalValue = value;
    }
    resetBranches() {
        this._branches.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchesInput() {
        return this._branches.internalValue;
    }
    get filePaths() {
        return this._filePaths;
    }
    putFilePaths(value) {
        this._filePaths.internalValue = value;
    }
    resetFilePaths() {
        this._filePaths.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filePathsInput() {
        return this._filePaths.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.CodepipelineTriggerGitConfigurationPushOutputReference = CodepipelineTriggerGitConfigurationPushOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPushOutputReference[_q] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPushOutputReference", version: "19.15.0" };
class CodepipelineTriggerGitConfigurationPushList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineTriggerGitConfigurationPushOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineTriggerGitConfigurationPushList = CodepipelineTriggerGitConfigurationPushList;
_r = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationPushList[_r] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationPushList", version: "19.15.0" };
function codepipelineTriggerGitConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_action_name: cdktf.stringToTerraform(struct.sourceActionName),
        pull_request: cdktf.listMapper(codepipelineTriggerGitConfigurationPullRequestToTerraform, true)(struct.pullRequest),
        push: cdktf.listMapper(codepipelineTriggerGitConfigurationPushToTerraform, true)(struct.push),
    };
}
exports.codepipelineTriggerGitConfigurationToTerraform = codepipelineTriggerGitConfigurationToTerraform;
function codepipelineTriggerGitConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_action_name: {
            value: cdktf.stringToHclTerraform(struct.sourceActionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        pull_request: {
            value: cdktf.listMapperHcl(codepipelineTriggerGitConfigurationPullRequestToHclTerraform, true)(struct.pullRequest),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPullRequestList",
        },
        push: {
            value: cdktf.listMapperHcl(codepipelineTriggerGitConfigurationPushToHclTerraform, true)(struct.push),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationPushList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerGitConfigurationToHclTerraform = codepipelineTriggerGitConfigurationToHclTerraform;
class CodepipelineTriggerGitConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // pull_request - computed: false, optional: true, required: false
        this._pullRequest = new CodepipelineTriggerGitConfigurationPullRequestList(this, "pull_request", false);
        // push - computed: false, optional: true, required: false
        this._push = new CodepipelineTriggerGitConfigurationPushList(this, "push", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceActionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceActionName = this._sourceActionName;
        }
        if (this._pullRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.pullRequest = this._pullRequest?.internalValue;
        }
        if (this._push?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.push = this._push?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sourceActionName = undefined;
            this._pullRequest.internalValue = undefined;
            this._push.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sourceActionName = value.sourceActionName;
            this._pullRequest.internalValue = value.pullRequest;
            this._push.internalValue = value.push;
        }
    }
    get sourceActionName() {
        return this.getStringAttribute('source_action_name');
    }
    set sourceActionName(value) {
        this._sourceActionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceActionNameInput() {
        return this._sourceActionName;
    }
    get pullRequest() {
        return this._pullRequest;
    }
    putPullRequest(value) {
        this._pullRequest.internalValue = value;
    }
    resetPullRequest() {
        this._pullRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pullRequestInput() {
        return this._pullRequest.internalValue;
    }
    get push() {
        return this._push;
    }
    putPush(value) {
        this._push.internalValue = value;
    }
    resetPush() {
        this._push.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pushInput() {
        return this._push.internalValue;
    }
}
exports.CodepipelineTriggerGitConfigurationOutputReference = CodepipelineTriggerGitConfigurationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerGitConfigurationOutputReference[_s] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerGitConfigurationOutputReference", version: "19.15.0" };
function codepipelineTriggerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        provider_type: cdktf.stringToTerraform(struct.providerType),
        git_configuration: codepipelineTriggerGitConfigurationToTerraform(struct.gitConfiguration),
    };
}
exports.codepipelineTriggerToTerraform = codepipelineTriggerToTerraform;
function codepipelineTriggerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        provider_type: {
            value: cdktf.stringToHclTerraform(struct.providerType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        git_configuration: {
            value: codepipelineTriggerGitConfigurationToHclTerraform(struct.gitConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "CodepipelineTriggerGitConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineTriggerToHclTerraform = codepipelineTriggerToHclTerraform;
class CodepipelineTriggerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // git_configuration - computed: false, optional: false, required: true
        this._gitConfiguration = new CodepipelineTriggerGitConfigurationOutputReference(this, "git_configuration");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._providerType !== undefined) {
            hasAnyValues = true;
            internalValueResult.providerType = this._providerType;
        }
        if (this._gitConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gitConfiguration = this._gitConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._providerType = undefined;
            this._gitConfiguration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._providerType = value.providerType;
            this._gitConfiguration.internalValue = value.gitConfiguration;
        }
    }
    get providerType() {
        return this.getStringAttribute('provider_type');
    }
    set providerType(value) {
        this._providerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerTypeInput() {
        return this._providerType;
    }
    get gitConfiguration() {
        return this._gitConfiguration;
    }
    putGitConfiguration(value) {
        this._gitConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get gitConfigurationInput() {
        return this._gitConfiguration.internalValue;
    }
}
exports.CodepipelineTriggerOutputReference = CodepipelineTriggerOutputReference;
_t = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerOutputReference[_t] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerOutputReference", version: "19.15.0" };
class CodepipelineTriggerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineTriggerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineTriggerList = CodepipelineTriggerList;
_u = JSII_RTTI_SYMBOL_1;
CodepipelineTriggerList[_u] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineTriggerList", version: "19.15.0" };
function codepipelineVariableToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_value: cdktf.stringToTerraform(struct.defaultValue),
        description: cdktf.stringToTerraform(struct.description),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.codepipelineVariableToTerraform = codepipelineVariableToTerraform;
function codepipelineVariableToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_value: {
            value: cdktf.stringToHclTerraform(struct.defaultValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineVariableToHclTerraform = codepipelineVariableToHclTerraform;
class CodepipelineVariableOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultValue = this._defaultValue;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._defaultValue = undefined;
            this._description = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._defaultValue = value.defaultValue;
            this._description = value.description;
            this._name = value.name;
        }
    }
    get defaultValue() {
        return this.getStringAttribute('default_value');
    }
    set defaultValue(value) {
        this._defaultValue = value;
    }
    resetDefaultValue() {
        this._defaultValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultValueInput() {
        return this._defaultValue;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CodepipelineVariableOutputReference = CodepipelineVariableOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CodepipelineVariableOutputReference[_v] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineVariableOutputReference", version: "19.15.0" };
class CodepipelineVariableList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineVariableOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineVariableList = CodepipelineVariableList;
_w = JSII_RTTI_SYMBOL_1;
CodepipelineVariableList[_w] = { fqn: "@cdktf/provider-aws.codepipeline.CodepipelineVariableList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline aws_codepipeline}
*/
class Codepipeline extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Codepipeline resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Codepipeline to import
    * @param importFromId The id of the existing Codepipeline that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Codepipeline to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_codepipeline", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline aws_codepipeline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodepipelineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codepipeline',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // artifact_store - computed: false, optional: false, required: true
        this._artifactStore = new CodepipelineArtifactStoreList(this, "artifact_store", true);
        // stage - computed: false, optional: false, required: true
        this._stage = new CodepipelineStageList(this, "stage", false);
        // trigger - computed: false, optional: true, required: false
        this._trigger = new CodepipelineTriggerList(this, "trigger", false);
        // variable - computed: false, optional: true, required: false
        this._variable = new CodepipelineVariableList(this, "variable", false);
        this._executionMode = config.executionMode;
        this._id = config.id;
        this._name = config.name;
        this._pipelineType = config.pipelineType;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._artifactStore.internalValue = config.artifactStore;
        this._stage.internalValue = config.stage;
        this._trigger.internalValue = config.trigger;
        this._variable.internalValue = config.variable;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get executionMode() {
        return this.getStringAttribute('execution_mode');
    }
    set executionMode(value) {
        this._executionMode = value;
    }
    resetExecutionMode() {
        this._executionMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionModeInput() {
        return this._executionMode;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get pipelineType() {
        return this.getStringAttribute('pipeline_type');
    }
    set pipelineType(value) {
        this._pipelineType = value;
    }
    resetPipelineType() {
        this._pipelineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineTypeInput() {
        return this._pipelineType;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get artifactStore() {
        return this._artifactStore;
    }
    putArtifactStore(value) {
        this._artifactStore.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get artifactStoreInput() {
        return this._artifactStore.internalValue;
    }
    get stage() {
        return this._stage;
    }
    putStage(value) {
        this._stage.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stageInput() {
        return this._stage.internalValue;
    }
    get trigger() {
        return this._trigger;
    }
    putTrigger(value) {
        this._trigger.internalValue = value;
    }
    resetTrigger() {
        this._trigger.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggerInput() {
        return this._trigger.internalValue;
    }
    get variable() {
        return this._variable;
    }
    putVariable(value) {
        this._variable.internalValue = value;
    }
    resetVariable() {
        this._variable.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get variableInput() {
        return this._variable.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            execution_mode: cdktf.stringToTerraform(this._executionMode),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            pipeline_type: cdktf.stringToTerraform(this._pipelineType),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            artifact_store: cdktf.listMapper(codepipelineArtifactStoreToTerraform, true)(this._artifactStore.internalValue),
            stage: cdktf.listMapper(codepipelineStageToTerraform, true)(this._stage.internalValue),
            trigger: cdktf.listMapper(codepipelineTriggerToTerraform, true)(this._trigger.internalValue),
            variable: cdktf.listMapper(codepipelineVariableToTerraform, true)(this._variable.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            execution_mode: {
                value: cdktf.stringToHclTerraform(this._executionMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_type: {
                value: cdktf.stringToHclTerraform(this._pipelineType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            artifact_store: {
                value: cdktf.listMapperHcl(codepipelineArtifactStoreToHclTerraform, true)(this._artifactStore.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CodepipelineArtifactStoreList",
            },
            stage: {
                value: cdktf.listMapperHcl(codepipelineStageToHclTerraform, true)(this._stage.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineStageList",
            },
            trigger: {
                value: cdktf.listMapperHcl(codepipelineTriggerToHclTerraform, true)(this._trigger.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineTriggerList",
            },
            variable: {
                value: cdktf.listMapperHcl(codepipelineVariableToHclTerraform, true)(this._variable.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineVariableList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Codepipeline = Codepipeline;
_x = JSII_RTTI_SYMBOL_1;
Codepipeline[_x] = { fqn: "@cdktf/provider-aws.codepipeline.Codepipeline", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Codepipeline.tfResourceType = "aws_codepipeline";
