// @cdktf/provider-aws/lib/elasticsearch-domain/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticsearchDomainTimeoutsOutputReference = exports.elasticsearchDomainTimeoutsToHclTerraform = exports.elasticsearchDomainTimeoutsToTerraform = exports.ElasticsearchDomainSnapshotOptionsOutputReference = exports.elasticsearchDomainSnapshotOptionsToHclTerraform = exports.elasticsearchDomainSnapshotOptionsToTerraform = exports.ElasticsearchDomainNodeToNodeEncryptionOutputReference = exports.elasticsearchDomainNodeToNodeEncryptionToHclTerraform = exports.elasticsearchDomainNodeToNodeEncryptionToTerraform = exports.ElasticsearchDomainLogPublishingOptionsList = exports.ElasticsearchDomainLogPublishingOptionsOutputReference = exports.elasticsearchDomainLogPublishingOptionsToHclTerraform = exports.elasticsearchDomainLogPublishingOptionsToTerraform = exports.ElasticsearchDomainEncryptAtRestOutputReference = exports.elasticsearchDomainEncryptAtRestToHclTerraform = exports.elasticsearchDomainEncryptAtRestToTerraform = exports.ElasticsearchDomainEbsOptionsOutputReference = exports.elasticsearchDomainEbsOptionsToHclTerraform = exports.elasticsearchDomainEbsOptionsToTerraform = exports.ElasticsearchDomainDomainEndpointOptionsOutputReference = exports.elasticsearchDomainDomainEndpointOptionsToHclTerraform = exports.elasticsearchDomainDomainEndpointOptionsToTerraform = exports.ElasticsearchDomainCognitoOptionsOutputReference = exports.elasticsearchDomainCognitoOptionsToHclTerraform = exports.elasticsearchDomainCognitoOptionsToTerraform = exports.ElasticsearchDomainClusterConfigOutputReference = exports.elasticsearchDomainClusterConfigToHclTerraform = exports.elasticsearchDomainClusterConfigToTerraform = exports.ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference = exports.elasticsearchDomainClusterConfigZoneAwarenessConfigToHclTerraform = exports.elasticsearchDomainClusterConfigZoneAwarenessConfigToTerraform = exports.ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference = exports.elasticsearchDomainClusterConfigColdStorageOptionsToHclTerraform = exports.elasticsearchDomainClusterConfigColdStorageOptionsToTerraform = exports.ElasticsearchDomainAutoTuneOptionsOutputReference = exports.elasticsearchDomainAutoTuneOptionsToHclTerraform = exports.elasticsearchDomainAutoTuneOptionsToTerraform = exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList = exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform = exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform = exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = exports.ElasticsearchDomainAdvancedSecurityOptionsOutputReference = exports.elasticsearchDomainAdvancedSecurityOptionsToHclTerraform = exports.elasticsearchDomainAdvancedSecurityOptionsToTerraform = exports.ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference = exports.elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform = exports.elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform = void 0;
exports.ElasticsearchDomain = exports.ElasticsearchDomainVpcOptionsOutputReference = exports.elasticsearchDomainVpcOptionsToHclTerraform = exports.elasticsearchDomainVpcOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        master_user_arn: cdktf.stringToTerraform(struct.masterUserArn),
        master_user_name: cdktf.stringToTerraform(struct.masterUserName),
        master_user_password: cdktf.stringToTerraform(struct.masterUserPassword),
    };
}
exports.elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform = elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform;
function elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        master_user_arn: {
            value: cdktf.stringToHclTerraform(struct.masterUserArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        master_user_name: {
            value: cdktf.stringToHclTerraform(struct.masterUserName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        master_user_password: {
            value: cdktf.stringToHclTerraform(struct.masterUserPassword),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform = elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform;
class ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._masterUserArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserArn = this._masterUserArn;
        }
        if (this._masterUserName !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserName = this._masterUserName;
        }
        if (this._masterUserPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserPassword = this._masterUserPassword;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._masterUserArn = undefined;
            this._masterUserName = undefined;
            this._masterUserPassword = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._masterUserArn = value.masterUserArn;
            this._masterUserName = value.masterUserName;
            this._masterUserPassword = value.masterUserPassword;
        }
    }
    get masterUserArn() {
        return this.getStringAttribute('master_user_arn');
    }
    set masterUserArn(value) {
        this._masterUserArn = value;
    }
    resetMasterUserArn() {
        this._masterUserArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserArnInput() {
        return this._masterUserArn;
    }
    get masterUserName() {
        return this.getStringAttribute('master_user_name');
    }
    set masterUserName(value) {
        this._masterUserName = value;
    }
    resetMasterUserName() {
        this._masterUserName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserNameInput() {
        return this._masterUserName;
    }
    get masterUserPassword() {
        return this.getStringAttribute('master_user_password');
    }
    set masterUserPassword(value) {
        this._masterUserPassword = value;
    }
    resetMasterUserPassword() {
        this._masterUserPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserPasswordInput() {
        return this._masterUserPassword;
    }
}
exports.ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference = ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainAdvancedSecurityOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        internal_user_database_enabled: cdktf.booleanToTerraform(struct.internalUserDatabaseEnabled),
        master_user_options: elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform(struct.masterUserOptions),
    };
}
exports.elasticsearchDomainAdvancedSecurityOptionsToTerraform = elasticsearchDomainAdvancedSecurityOptionsToTerraform;
function elasticsearchDomainAdvancedSecurityOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        internal_user_database_enabled: {
            value: cdktf.booleanToHclTerraform(struct.internalUserDatabaseEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        master_user_options: {
            value: elasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform(struct.masterUserOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainAdvancedSecurityOptionsToHclTerraform = elasticsearchDomainAdvancedSecurityOptionsToHclTerraform;
class ElasticsearchDomainAdvancedSecurityOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // master_user_options - computed: false, optional: true, required: false
        this._masterUserOptions = new ElasticsearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference(this, "master_user_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._internalUserDatabaseEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.internalUserDatabaseEnabled = this._internalUserDatabaseEnabled;
        }
        if (this._masterUserOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserOptions = this._masterUserOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._internalUserDatabaseEnabled = undefined;
            this._masterUserOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._internalUserDatabaseEnabled = value.internalUserDatabaseEnabled;
            this._masterUserOptions.internalValue = value.masterUserOptions;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get internalUserDatabaseEnabled() {
        return this.getBooleanAttribute('internal_user_database_enabled');
    }
    set internalUserDatabaseEnabled(value) {
        this._internalUserDatabaseEnabled = value;
    }
    resetInternalUserDatabaseEnabled() {
        this._internalUserDatabaseEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get internalUserDatabaseEnabledInput() {
        return this._internalUserDatabaseEnabled;
    }
    get masterUserOptions() {
        return this._masterUserOptions;
    }
    putMasterUserOptions(value) {
        this._masterUserOptions.internalValue = value;
    }
    resetMasterUserOptions() {
        this._masterUserOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserOptionsInput() {
        return this._masterUserOptions.internalValue;
    }
}
exports.ElasticsearchDomainAdvancedSecurityOptionsOutputReference = ElasticsearchDomainAdvancedSecurityOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAdvancedSecurityOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAdvancedSecurityOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform;
function elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        unit: {
            value: cdktf.stringToHclTerraform(struct.unit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform = elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform;
class ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference", version: "19.15.0" };
function elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cron_expression_for_recurrence: cdktf.stringToTerraform(struct.cronExpressionForRecurrence),
        start_at: cdktf.stringToTerraform(struct.startAt),
        duration: elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform(struct.duration),
    };
}
exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform;
function elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cron_expression_for_recurrence: {
            value: cdktf.stringToHclTerraform(struct.cronExpressionForRecurrence),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_at: {
            value: cdktf.stringToHclTerraform(struct.startAt),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        duration: {
            value: elasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform(struct.duration),
            isBlock: true,
            type: "list",
            storageClassType: "ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform = elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform;
class ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // duration - computed: false, optional: false, required: true
        this._duration = new ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference(this, "duration");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cronExpressionForRecurrence !== undefined) {
            hasAnyValues = true;
            internalValueResult.cronExpressionForRecurrence = this._cronExpressionForRecurrence;
        }
        if (this._startAt !== undefined) {
            hasAnyValues = true;
            internalValueResult.startAt = this._startAt;
        }
        if (this._duration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.duration = this._duration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cronExpressionForRecurrence = undefined;
            this._startAt = undefined;
            this._duration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cronExpressionForRecurrence = value.cronExpressionForRecurrence;
            this._startAt = value.startAt;
            this._duration.internalValue = value.duration;
        }
    }
    get cronExpressionForRecurrence() {
        return this.getStringAttribute('cron_expression_for_recurrence');
    }
    set cronExpressionForRecurrence(value) {
        this._cronExpressionForRecurrence = value;
    }
    // Temporarily expose input value. Use with caution.
    get cronExpressionForRecurrenceInput() {
        return this._cronExpressionForRecurrence;
    }
    get startAt() {
        return this.getStringAttribute('start_at');
    }
    set startAt(value) {
        this._startAt = value;
    }
    // Temporarily expose input value. Use with caution.
    get startAtInput() {
        return this._startAt;
    }
    get duration() {
        return this._duration;
    }
    putDuration(value) {
        this._duration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration.internalValue;
    }
}
exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference[_d] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference", version: "19.15.0" };
class ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList = ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList;
_e = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList[_e] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList", version: "19.15.0" };
function elasticsearchDomainAutoTuneOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        desired_state: cdktf.stringToTerraform(struct.desiredState),
        rollback_on_disable: cdktf.stringToTerraform(struct.rollbackOnDisable),
        maintenance_schedule: cdktf.listMapper(elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform, true)(struct.maintenanceSchedule),
    };
}
exports.elasticsearchDomainAutoTuneOptionsToTerraform = elasticsearchDomainAutoTuneOptionsToTerraform;
function elasticsearchDomainAutoTuneOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        desired_state: {
            value: cdktf.stringToHclTerraform(struct.desiredState),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rollback_on_disable: {
            value: cdktf.stringToHclTerraform(struct.rollbackOnDisable),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        maintenance_schedule: {
            value: cdktf.listMapperHcl(elasticsearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform, true)(struct.maintenanceSchedule),
            isBlock: true,
            type: "set",
            storageClassType: "ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainAutoTuneOptionsToHclTerraform = elasticsearchDomainAutoTuneOptionsToHclTerraform;
class ElasticsearchDomainAutoTuneOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // maintenance_schedule - computed: false, optional: true, required: false
        this._maintenanceSchedule = new ElasticsearchDomainAutoTuneOptionsMaintenanceScheduleList(this, "maintenance_schedule", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._desiredState !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredState = this._desiredState;
        }
        if (this._rollbackOnDisable !== undefined) {
            hasAnyValues = true;
            internalValueResult.rollbackOnDisable = this._rollbackOnDisable;
        }
        if (this._maintenanceSchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maintenanceSchedule = this._maintenanceSchedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._desiredState = undefined;
            this._rollbackOnDisable = undefined;
            this._maintenanceSchedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._desiredState = value.desiredState;
            this._rollbackOnDisable = value.rollbackOnDisable;
            this._maintenanceSchedule.internalValue = value.maintenanceSchedule;
        }
    }
    get desiredState() {
        return this.getStringAttribute('desired_state');
    }
    set desiredState(value) {
        this._desiredState = value;
    }
    // Temporarily expose input value. Use with caution.
    get desiredStateInput() {
        return this._desiredState;
    }
    get rollbackOnDisable() {
        return this.getStringAttribute('rollback_on_disable');
    }
    set rollbackOnDisable(value) {
        this._rollbackOnDisable = value;
    }
    resetRollbackOnDisable() {
        this._rollbackOnDisable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rollbackOnDisableInput() {
        return this._rollbackOnDisable;
    }
    get maintenanceSchedule() {
        return this._maintenanceSchedule;
    }
    putMaintenanceSchedule(value) {
        this._maintenanceSchedule.internalValue = value;
    }
    resetMaintenanceSchedule() {
        this._maintenanceSchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceScheduleInput() {
        return this._maintenanceSchedule.internalValue;
    }
}
exports.ElasticsearchDomainAutoTuneOptionsOutputReference = ElasticsearchDomainAutoTuneOptionsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainAutoTuneOptionsOutputReference[_f] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainAutoTuneOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainClusterConfigColdStorageOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.elasticsearchDomainClusterConfigColdStorageOptionsToTerraform = elasticsearchDomainClusterConfigColdStorageOptionsToTerraform;
function elasticsearchDomainClusterConfigColdStorageOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainClusterConfigColdStorageOptionsToHclTerraform = elasticsearchDomainClusterConfigColdStorageOptionsToHclTerraform;
class ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference = ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference[_g] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainClusterConfigZoneAwarenessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_zone_count: cdktf.numberToTerraform(struct.availabilityZoneCount),
    };
}
exports.elasticsearchDomainClusterConfigZoneAwarenessConfigToTerraform = elasticsearchDomainClusterConfigZoneAwarenessConfigToTerraform;
function elasticsearchDomainClusterConfigZoneAwarenessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        availability_zone_count: {
            value: cdktf.numberToHclTerraform(struct.availabilityZoneCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainClusterConfigZoneAwarenessConfigToHclTerraform = elasticsearchDomainClusterConfigZoneAwarenessConfigToHclTerraform;
class ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availabilityZoneCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityZoneCount = this._availabilityZoneCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availabilityZoneCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availabilityZoneCount = value.availabilityZoneCount;
        }
    }
    get availabilityZoneCount() {
        return this.getNumberAttribute('availability_zone_count');
    }
    set availabilityZoneCount(value) {
        this._availabilityZoneCount = value;
    }
    resetAvailabilityZoneCount() {
        this._availabilityZoneCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneCountInput() {
        return this._availabilityZoneCount;
    }
}
exports.ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference = ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference", version: "19.15.0" };
function elasticsearchDomainClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dedicated_master_count: cdktf.numberToTerraform(struct.dedicatedMasterCount),
        dedicated_master_enabled: cdktf.booleanToTerraform(struct.dedicatedMasterEnabled),
        dedicated_master_type: cdktf.stringToTerraform(struct.dedicatedMasterType),
        instance_count: cdktf.numberToTerraform(struct.instanceCount),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        warm_count: cdktf.numberToTerraform(struct.warmCount),
        warm_enabled: cdktf.booleanToTerraform(struct.warmEnabled),
        warm_type: cdktf.stringToTerraform(struct.warmType),
        zone_awareness_enabled: cdktf.booleanToTerraform(struct.zoneAwarenessEnabled),
        cold_storage_options: elasticsearchDomainClusterConfigColdStorageOptionsToTerraform(struct.coldStorageOptions),
        zone_awareness_config: elasticsearchDomainClusterConfigZoneAwarenessConfigToTerraform(struct.zoneAwarenessConfig),
    };
}
exports.elasticsearchDomainClusterConfigToTerraform = elasticsearchDomainClusterConfigToTerraform;
function elasticsearchDomainClusterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dedicated_master_count: {
            value: cdktf.numberToHclTerraform(struct.dedicatedMasterCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        dedicated_master_enabled: {
            value: cdktf.booleanToHclTerraform(struct.dedicatedMasterEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        dedicated_master_type: {
            value: cdktf.stringToHclTerraform(struct.dedicatedMasterType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        instance_count: {
            value: cdktf.numberToHclTerraform(struct.instanceCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        warm_count: {
            value: cdktf.numberToHclTerraform(struct.warmCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        warm_enabled: {
            value: cdktf.booleanToHclTerraform(struct.warmEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        warm_type: {
            value: cdktf.stringToHclTerraform(struct.warmType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        zone_awareness_enabled: {
            value: cdktf.booleanToHclTerraform(struct.zoneAwarenessEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        cold_storage_options: {
            value: elasticsearchDomainClusterConfigColdStorageOptionsToHclTerraform(struct.coldStorageOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ElasticsearchDomainClusterConfigColdStorageOptionsList",
        },
        zone_awareness_config: {
            value: elasticsearchDomainClusterConfigZoneAwarenessConfigToHclTerraform(struct.zoneAwarenessConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ElasticsearchDomainClusterConfigZoneAwarenessConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainClusterConfigToHclTerraform = elasticsearchDomainClusterConfigToHclTerraform;
class ElasticsearchDomainClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cold_storage_options - computed: false, optional: true, required: false
        this._coldStorageOptions = new ElasticsearchDomainClusterConfigColdStorageOptionsOutputReference(this, "cold_storage_options");
        // zone_awareness_config - computed: false, optional: true, required: false
        this._zoneAwarenessConfig = new ElasticsearchDomainClusterConfigZoneAwarenessConfigOutputReference(this, "zone_awareness_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dedicatedMasterCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterCount = this._dedicatedMasterCount;
        }
        if (this._dedicatedMasterEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterEnabled = this._dedicatedMasterEnabled;
        }
        if (this._dedicatedMasterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterType = this._dedicatedMasterType;
        }
        if (this._instanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceCount = this._instanceCount;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._warmCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmCount = this._warmCount;
        }
        if (this._warmEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmEnabled = this._warmEnabled;
        }
        if (this._warmType !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmType = this._warmType;
        }
        if (this._zoneAwarenessEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneAwarenessEnabled = this._zoneAwarenessEnabled;
        }
        if (this._coldStorageOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.coldStorageOptions = this._coldStorageOptions?.internalValue;
        }
        if (this._zoneAwarenessConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneAwarenessConfig = this._zoneAwarenessConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dedicatedMasterCount = undefined;
            this._dedicatedMasterEnabled = undefined;
            this._dedicatedMasterType = undefined;
            this._instanceCount = undefined;
            this._instanceType = undefined;
            this._warmCount = undefined;
            this._warmEnabled = undefined;
            this._warmType = undefined;
            this._zoneAwarenessEnabled = undefined;
            this._coldStorageOptions.internalValue = undefined;
            this._zoneAwarenessConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dedicatedMasterCount = value.dedicatedMasterCount;
            this._dedicatedMasterEnabled = value.dedicatedMasterEnabled;
            this._dedicatedMasterType = value.dedicatedMasterType;
            this._instanceCount = value.instanceCount;
            this._instanceType = value.instanceType;
            this._warmCount = value.warmCount;
            this._warmEnabled = value.warmEnabled;
            this._warmType = value.warmType;
            this._zoneAwarenessEnabled = value.zoneAwarenessEnabled;
            this._coldStorageOptions.internalValue = value.coldStorageOptions;
            this._zoneAwarenessConfig.internalValue = value.zoneAwarenessConfig;
        }
    }
    get dedicatedMasterCount() {
        return this.getNumberAttribute('dedicated_master_count');
    }
    set dedicatedMasterCount(value) {
        this._dedicatedMasterCount = value;
    }
    resetDedicatedMasterCount() {
        this._dedicatedMasterCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterCountInput() {
        return this._dedicatedMasterCount;
    }
    get dedicatedMasterEnabled() {
        return this.getBooleanAttribute('dedicated_master_enabled');
    }
    set dedicatedMasterEnabled(value) {
        this._dedicatedMasterEnabled = value;
    }
    resetDedicatedMasterEnabled() {
        this._dedicatedMasterEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterEnabledInput() {
        return this._dedicatedMasterEnabled;
    }
    get dedicatedMasterType() {
        return this.getStringAttribute('dedicated_master_type');
    }
    set dedicatedMasterType(value) {
        this._dedicatedMasterType = value;
    }
    resetDedicatedMasterType() {
        this._dedicatedMasterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterTypeInput() {
        return this._dedicatedMasterType;
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get warmCount() {
        return this.getNumberAttribute('warm_count');
    }
    set warmCount(value) {
        this._warmCount = value;
    }
    resetWarmCount() {
        this._warmCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmCountInput() {
        return this._warmCount;
    }
    get warmEnabled() {
        return this.getBooleanAttribute('warm_enabled');
    }
    set warmEnabled(value) {
        this._warmEnabled = value;
    }
    resetWarmEnabled() {
        this._warmEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmEnabledInput() {
        return this._warmEnabled;
    }
    get warmType() {
        return this.getStringAttribute('warm_type');
    }
    set warmType(value) {
        this._warmType = value;
    }
    resetWarmType() {
        this._warmType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmTypeInput() {
        return this._warmType;
    }
    get zoneAwarenessEnabled() {
        return this.getBooleanAttribute('zone_awareness_enabled');
    }
    set zoneAwarenessEnabled(value) {
        this._zoneAwarenessEnabled = value;
    }
    resetZoneAwarenessEnabled() {
        this._zoneAwarenessEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneAwarenessEnabledInput() {
        return this._zoneAwarenessEnabled;
    }
    get coldStorageOptions() {
        return this._coldStorageOptions;
    }
    putColdStorageOptions(value) {
        this._coldStorageOptions.internalValue = value;
    }
    resetColdStorageOptions() {
        this._coldStorageOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coldStorageOptionsInput() {
        return this._coldStorageOptions.internalValue;
    }
    get zoneAwarenessConfig() {
        return this._zoneAwarenessConfig;
    }
    putZoneAwarenessConfig(value) {
        this._zoneAwarenessConfig.internalValue = value;
    }
    resetZoneAwarenessConfig() {
        this._zoneAwarenessConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneAwarenessConfigInput() {
        return this._zoneAwarenessConfig.internalValue;
    }
}
exports.ElasticsearchDomainClusterConfigOutputReference = ElasticsearchDomainClusterConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainClusterConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainClusterConfigOutputReference", version: "19.15.0" };
function elasticsearchDomainCognitoOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        identity_pool_id: cdktf.stringToTerraform(struct.identityPoolId),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        user_pool_id: cdktf.stringToTerraform(struct.userPoolId),
    };
}
exports.elasticsearchDomainCognitoOptionsToTerraform = elasticsearchDomainCognitoOptionsToTerraform;
function elasticsearchDomainCognitoOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        identity_pool_id: {
            value: cdktf.stringToHclTerraform(struct.identityPoolId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_pool_id: {
            value: cdktf.stringToHclTerraform(struct.userPoolId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainCognitoOptionsToHclTerraform = elasticsearchDomainCognitoOptionsToHclTerraform;
class ElasticsearchDomainCognitoOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._identityPoolId !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityPoolId = this._identityPoolId;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._userPoolId !== undefined) {
            hasAnyValues = true;
            internalValueResult.userPoolId = this._userPoolId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._identityPoolId = undefined;
            this._roleArn = undefined;
            this._userPoolId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._identityPoolId = value.identityPoolId;
            this._roleArn = value.roleArn;
            this._userPoolId = value.userPoolId;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get identityPoolId() {
        return this.getStringAttribute('identity_pool_id');
    }
    set identityPoolId(value) {
        this._identityPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityPoolIdInput() {
        return this._identityPoolId;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
    set userPoolId(value) {
        this._userPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userPoolIdInput() {
        return this._userPoolId;
    }
}
exports.ElasticsearchDomainCognitoOptionsOutputReference = ElasticsearchDomainCognitoOptionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainCognitoOptionsOutputReference[_k] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainCognitoOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainDomainEndpointOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_endpoint: cdktf.stringToTerraform(struct.customEndpoint),
        custom_endpoint_certificate_arn: cdktf.stringToTerraform(struct.customEndpointCertificateArn),
        custom_endpoint_enabled: cdktf.booleanToTerraform(struct.customEndpointEnabled),
        enforce_https: cdktf.booleanToTerraform(struct.enforceHttps),
        tls_security_policy: cdktf.stringToTerraform(struct.tlsSecurityPolicy),
    };
}
exports.elasticsearchDomainDomainEndpointOptionsToTerraform = elasticsearchDomainDomainEndpointOptionsToTerraform;
function elasticsearchDomainDomainEndpointOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        custom_endpoint: {
            value: cdktf.stringToHclTerraform(struct.customEndpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        custom_endpoint_certificate_arn: {
            value: cdktf.stringToHclTerraform(struct.customEndpointCertificateArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        custom_endpoint_enabled: {
            value: cdktf.booleanToHclTerraform(struct.customEndpointEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enforce_https: {
            value: cdktf.booleanToHclTerraform(struct.enforceHttps),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        tls_security_policy: {
            value: cdktf.stringToHclTerraform(struct.tlsSecurityPolicy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainDomainEndpointOptionsToHclTerraform = elasticsearchDomainDomainEndpointOptionsToHclTerraform;
class ElasticsearchDomainDomainEndpointOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpoint = this._customEndpoint;
        }
        if (this._customEndpointCertificateArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpointCertificateArn = this._customEndpointCertificateArn;
        }
        if (this._customEndpointEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpointEnabled = this._customEndpointEnabled;
        }
        if (this._enforceHttps !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceHttps = this._enforceHttps;
        }
        if (this._tlsSecurityPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.tlsSecurityPolicy = this._tlsSecurityPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customEndpoint = undefined;
            this._customEndpointCertificateArn = undefined;
            this._customEndpointEnabled = undefined;
            this._enforceHttps = undefined;
            this._tlsSecurityPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customEndpoint = value.customEndpoint;
            this._customEndpointCertificateArn = value.customEndpointCertificateArn;
            this._customEndpointEnabled = value.customEndpointEnabled;
            this._enforceHttps = value.enforceHttps;
            this._tlsSecurityPolicy = value.tlsSecurityPolicy;
        }
    }
    get customEndpoint() {
        return this.getStringAttribute('custom_endpoint');
    }
    set customEndpoint(value) {
        this._customEndpoint = value;
    }
    resetCustomEndpoint() {
        this._customEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointInput() {
        return this._customEndpoint;
    }
    get customEndpointCertificateArn() {
        return this.getStringAttribute('custom_endpoint_certificate_arn');
    }
    set customEndpointCertificateArn(value) {
        this._customEndpointCertificateArn = value;
    }
    resetCustomEndpointCertificateArn() {
        this._customEndpointCertificateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointCertificateArnInput() {
        return this._customEndpointCertificateArn;
    }
    get customEndpointEnabled() {
        return this.getBooleanAttribute('custom_endpoint_enabled');
    }
    set customEndpointEnabled(value) {
        this._customEndpointEnabled = value;
    }
    resetCustomEndpointEnabled() {
        this._customEndpointEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointEnabledInput() {
        return this._customEndpointEnabled;
    }
    get enforceHttps() {
        return this.getBooleanAttribute('enforce_https');
    }
    set enforceHttps(value) {
        this._enforceHttps = value;
    }
    resetEnforceHttps() {
        this._enforceHttps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceHttpsInput() {
        return this._enforceHttps;
    }
    get tlsSecurityPolicy() {
        return this.getStringAttribute('tls_security_policy');
    }
    set tlsSecurityPolicy(value) {
        this._tlsSecurityPolicy = value;
    }
    resetTlsSecurityPolicy() {
        this._tlsSecurityPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsSecurityPolicyInput() {
        return this._tlsSecurityPolicy;
    }
}
exports.ElasticsearchDomainDomainEndpointOptionsOutputReference = ElasticsearchDomainDomainEndpointOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainDomainEndpointOptionsOutputReference[_l] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainDomainEndpointOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainEbsOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ebs_enabled: cdktf.booleanToTerraform(struct.ebsEnabled),
        iops: cdktf.numberToTerraform(struct.iops),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.elasticsearchDomainEbsOptionsToTerraform = elasticsearchDomainEbsOptionsToTerraform;
function elasticsearchDomainEbsOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ebs_enabled: {
            value: cdktf.booleanToHclTerraform(struct.ebsEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        iops: {
            value: cdktf.numberToHclTerraform(struct.iops),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        throughput: {
            value: cdktf.numberToHclTerraform(struct.throughput),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        volume_size: {
            value: cdktf.numberToHclTerraform(struct.volumeSize),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        volume_type: {
            value: cdktf.stringToHclTerraform(struct.volumeType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainEbsOptionsToHclTerraform = elasticsearchDomainEbsOptionsToHclTerraform;
class ElasticsearchDomainEbsOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ebsEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsEnabled = this._ebsEnabled;
        }
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._throughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._volumeSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeSize = this._volumeSize;
        }
        if (this._volumeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeType = this._volumeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ebsEnabled = undefined;
            this._iops = undefined;
            this._throughput = undefined;
            this._volumeSize = undefined;
            this._volumeType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ebsEnabled = value.ebsEnabled;
            this._iops = value.iops;
            this._throughput = value.throughput;
            this._volumeSize = value.volumeSize;
            this._volumeType = value.volumeType;
        }
    }
    get ebsEnabled() {
        return this.getBooleanAttribute('ebs_enabled');
    }
    set ebsEnabled(value) {
        this._ebsEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get ebsEnabledInput() {
        return this._ebsEnabled;
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    set throughput(value) {
        this._throughput = value;
    }
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get throughputInput() {
        return this._throughput;
    }
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    set volumeSize(value) {
        this._volumeSize = value;
    }
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInput() {
        return this._volumeSize;
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
}
exports.ElasticsearchDomainEbsOptionsOutputReference = ElasticsearchDomainEbsOptionsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainEbsOptionsOutputReference[_m] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainEbsOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainEncryptAtRestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.elasticsearchDomainEncryptAtRestToTerraform = elasticsearchDomainEncryptAtRestToTerraform;
function elasticsearchDomainEncryptAtRestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainEncryptAtRestToHclTerraform = elasticsearchDomainEncryptAtRestToHclTerraform;
class ElasticsearchDomainEncryptAtRestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._kmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.ElasticsearchDomainEncryptAtRestOutputReference = ElasticsearchDomainEncryptAtRestOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainEncryptAtRestOutputReference[_o] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainEncryptAtRestOutputReference", version: "19.15.0" };
function elasticsearchDomainLogPublishingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_log_group_arn: cdktf.stringToTerraform(struct.cloudwatchLogGroupArn),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_type: cdktf.stringToTerraform(struct.logType),
    };
}
exports.elasticsearchDomainLogPublishingOptionsToTerraform = elasticsearchDomainLogPublishingOptionsToTerraform;
function elasticsearchDomainLogPublishingOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cloudwatch_log_group_arn: {
            value: cdktf.stringToHclTerraform(struct.cloudwatchLogGroupArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        log_type: {
            value: cdktf.stringToHclTerraform(struct.logType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainLogPublishingOptionsToHclTerraform = elasticsearchDomainLogPublishingOptionsToHclTerraform;
class ElasticsearchDomainLogPublishingOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudwatchLogGroupArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogGroupArn = this._cloudwatchLogGroupArn;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logType !== undefined) {
            hasAnyValues = true;
            internalValueResult.logType = this._logType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cloudwatchLogGroupArn = undefined;
            this._enabled = undefined;
            this._logType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cloudwatchLogGroupArn = value.cloudwatchLogGroupArn;
            this._enabled = value.enabled;
            this._logType = value.logType;
        }
    }
    get cloudwatchLogGroupArn() {
        return this.getStringAttribute('cloudwatch_log_group_arn');
    }
    set cloudwatchLogGroupArn(value) {
        this._cloudwatchLogGroupArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogGroupArnInput() {
        return this._cloudwatchLogGroupArn;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get logType() {
        return this.getStringAttribute('log_type');
    }
    set logType(value) {
        this._logType = value;
    }
    // Temporarily expose input value. Use with caution.
    get logTypeInput() {
        return this._logType;
    }
}
exports.ElasticsearchDomainLogPublishingOptionsOutputReference = ElasticsearchDomainLogPublishingOptionsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainLogPublishingOptionsOutputReference[_p] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainLogPublishingOptionsOutputReference", version: "19.15.0" };
class ElasticsearchDomainLogPublishingOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElasticsearchDomainLogPublishingOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElasticsearchDomainLogPublishingOptionsList = ElasticsearchDomainLogPublishingOptionsList;
_q = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainLogPublishingOptionsList[_q] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainLogPublishingOptionsList", version: "19.15.0" };
function elasticsearchDomainNodeToNodeEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.elasticsearchDomainNodeToNodeEncryptionToTerraform = elasticsearchDomainNodeToNodeEncryptionToTerraform;
function elasticsearchDomainNodeToNodeEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainNodeToNodeEncryptionToHclTerraform = elasticsearchDomainNodeToNodeEncryptionToHclTerraform;
class ElasticsearchDomainNodeToNodeEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.ElasticsearchDomainNodeToNodeEncryptionOutputReference = ElasticsearchDomainNodeToNodeEncryptionOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainNodeToNodeEncryptionOutputReference[_r] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainNodeToNodeEncryptionOutputReference", version: "19.15.0" };
function elasticsearchDomainSnapshotOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        automated_snapshot_start_hour: cdktf.numberToTerraform(struct.automatedSnapshotStartHour),
    };
}
exports.elasticsearchDomainSnapshotOptionsToTerraform = elasticsearchDomainSnapshotOptionsToTerraform;
function elasticsearchDomainSnapshotOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        automated_snapshot_start_hour: {
            value: cdktf.numberToHclTerraform(struct.automatedSnapshotStartHour),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainSnapshotOptionsToHclTerraform = elasticsearchDomainSnapshotOptionsToHclTerraform;
class ElasticsearchDomainSnapshotOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._automatedSnapshotStartHour !== undefined) {
            hasAnyValues = true;
            internalValueResult.automatedSnapshotStartHour = this._automatedSnapshotStartHour;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._automatedSnapshotStartHour = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._automatedSnapshotStartHour = value.automatedSnapshotStartHour;
        }
    }
    get automatedSnapshotStartHour() {
        return this.getNumberAttribute('automated_snapshot_start_hour');
    }
    set automatedSnapshotStartHour(value) {
        this._automatedSnapshotStartHour = value;
    }
    // Temporarily expose input value. Use with caution.
    get automatedSnapshotStartHourInput() {
        return this._automatedSnapshotStartHour;
    }
}
exports.ElasticsearchDomainSnapshotOptionsOutputReference = ElasticsearchDomainSnapshotOptionsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainSnapshotOptionsOutputReference[_s] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainSnapshotOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.elasticsearchDomainTimeoutsToTerraform = elasticsearchDomainTimeoutsToTerraform;
function elasticsearchDomainTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainTimeoutsToHclTerraform = elasticsearchDomainTimeoutsToHclTerraform;
class ElasticsearchDomainTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ElasticsearchDomainTimeoutsOutputReference = ElasticsearchDomainTimeoutsOutputReference;
_t = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainTimeoutsOutputReference[_t] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainTimeoutsOutputReference", version: "19.15.0" };
function elasticsearchDomainVpcOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.elasticsearchDomainVpcOptionsToTerraform = elasticsearchDomainVpcOptionsToTerraform;
function elasticsearchDomainVpcOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainVpcOptionsToHclTerraform = elasticsearchDomainVpcOptionsToHclTerraform;
class ElasticsearchDomainVpcOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    // availability_zones - computed: true, optional: false, required: false
    get availabilityZones() {
        return cdktf.Fn.tolist(this.getListAttribute('availability_zones'));
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.ElasticsearchDomainVpcOptionsOutputReference = ElasticsearchDomainVpcOptionsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainVpcOptionsOutputReference[_u] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomainVpcOptionsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain aws_elasticsearch_domain}
*/
class ElasticsearchDomain extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ElasticsearchDomain resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ElasticsearchDomain to import
    * @param importFromId The id of the existing ElasticsearchDomain that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ElasticsearchDomain to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_elasticsearch_domain", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain aws_elasticsearch_domain} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticsearchDomainConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticsearch_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // advanced_security_options - computed: false, optional: true, required: false
        this._advancedSecurityOptions = new ElasticsearchDomainAdvancedSecurityOptionsOutputReference(this, "advanced_security_options");
        // auto_tune_options - computed: false, optional: true, required: false
        this._autoTuneOptions = new ElasticsearchDomainAutoTuneOptionsOutputReference(this, "auto_tune_options");
        // cluster_config - computed: false, optional: true, required: false
        this._clusterConfig = new ElasticsearchDomainClusterConfigOutputReference(this, "cluster_config");
        // cognito_options - computed: false, optional: true, required: false
        this._cognitoOptions = new ElasticsearchDomainCognitoOptionsOutputReference(this, "cognito_options");
        // domain_endpoint_options - computed: false, optional: true, required: false
        this._domainEndpointOptions = new ElasticsearchDomainDomainEndpointOptionsOutputReference(this, "domain_endpoint_options");
        // ebs_options - computed: false, optional: true, required: false
        this._ebsOptions = new ElasticsearchDomainEbsOptionsOutputReference(this, "ebs_options");
        // encrypt_at_rest - computed: false, optional: true, required: false
        this._encryptAtRest = new ElasticsearchDomainEncryptAtRestOutputReference(this, "encrypt_at_rest");
        // log_publishing_options - computed: false, optional: true, required: false
        this._logPublishingOptions = new ElasticsearchDomainLogPublishingOptionsList(this, "log_publishing_options", true);
        // node_to_node_encryption - computed: false, optional: true, required: false
        this._nodeToNodeEncryption = new ElasticsearchDomainNodeToNodeEncryptionOutputReference(this, "node_to_node_encryption");
        // snapshot_options - computed: false, optional: true, required: false
        this._snapshotOptions = new ElasticsearchDomainSnapshotOptionsOutputReference(this, "snapshot_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ElasticsearchDomainTimeoutsOutputReference(this, "timeouts");
        // vpc_options - computed: false, optional: true, required: false
        this._vpcOptions = new ElasticsearchDomainVpcOptionsOutputReference(this, "vpc_options");
        this._accessPolicies = config.accessPolicies;
        this._advancedOptions = config.advancedOptions;
        this._domainName = config.domainName;
        this._elasticsearchVersion = config.elasticsearchVersion;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._advancedSecurityOptions.internalValue = config.advancedSecurityOptions;
        this._autoTuneOptions.internalValue = config.autoTuneOptions;
        this._clusterConfig.internalValue = config.clusterConfig;
        this._cognitoOptions.internalValue = config.cognitoOptions;
        this._domainEndpointOptions.internalValue = config.domainEndpointOptions;
        this._ebsOptions.internalValue = config.ebsOptions;
        this._encryptAtRest.internalValue = config.encryptAtRest;
        this._logPublishingOptions.internalValue = config.logPublishingOptions;
        this._nodeToNodeEncryption.internalValue = config.nodeToNodeEncryption;
        this._snapshotOptions.internalValue = config.snapshotOptions;
        this._timeouts.internalValue = config.timeouts;
        this._vpcOptions.internalValue = config.vpcOptions;
    }
    get accessPolicies() {
        return this.getStringAttribute('access_policies');
    }
    set accessPolicies(value) {
        this._accessPolicies = value;
    }
    resetAccessPolicies() {
        this._accessPolicies = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessPoliciesInput() {
        return this._accessPolicies;
    }
    get advancedOptions() {
        return this.getStringMapAttribute('advanced_options');
    }
    set advancedOptions(value) {
        this._advancedOptions = value;
    }
    resetAdvancedOptions() {
        this._advancedOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedOptionsInput() {
        return this._advancedOptions;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // domain_id - computed: true, optional: false, required: false
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get elasticsearchVersion() {
        return this.getStringAttribute('elasticsearch_version');
    }
    set elasticsearchVersion(value) {
        this._elasticsearchVersion = value;
    }
    resetElasticsearchVersion() {
        this._elasticsearchVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elasticsearchVersionInput() {
        return this._elasticsearchVersion;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kibana_endpoint - computed: true, optional: false, required: false
    get kibanaEndpoint() {
        return this.getStringAttribute('kibana_endpoint');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get advancedSecurityOptions() {
        return this._advancedSecurityOptions;
    }
    putAdvancedSecurityOptions(value) {
        this._advancedSecurityOptions.internalValue = value;
    }
    resetAdvancedSecurityOptions() {
        this._advancedSecurityOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedSecurityOptionsInput() {
        return this._advancedSecurityOptions.internalValue;
    }
    get autoTuneOptions() {
        return this._autoTuneOptions;
    }
    putAutoTuneOptions(value) {
        this._autoTuneOptions.internalValue = value;
    }
    resetAutoTuneOptions() {
        this._autoTuneOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoTuneOptionsInput() {
        return this._autoTuneOptions.internalValue;
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    putClusterConfig(value) {
        this._clusterConfig.internalValue = value;
    }
    resetClusterConfig() {
        this._clusterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterConfigInput() {
        return this._clusterConfig.internalValue;
    }
    get cognitoOptions() {
        return this._cognitoOptions;
    }
    putCognitoOptions(value) {
        this._cognitoOptions.internalValue = value;
    }
    resetCognitoOptions() {
        this._cognitoOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cognitoOptionsInput() {
        return this._cognitoOptions.internalValue;
    }
    get domainEndpointOptions() {
        return this._domainEndpointOptions;
    }
    putDomainEndpointOptions(value) {
        this._domainEndpointOptions.internalValue = value;
    }
    resetDomainEndpointOptions() {
        this._domainEndpointOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainEndpointOptionsInput() {
        return this._domainEndpointOptions.internalValue;
    }
    get ebsOptions() {
        return this._ebsOptions;
    }
    putEbsOptions(value) {
        this._ebsOptions.internalValue = value;
    }
    resetEbsOptions() {
        this._ebsOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsOptionsInput() {
        return this._ebsOptions.internalValue;
    }
    get encryptAtRest() {
        return this._encryptAtRest;
    }
    putEncryptAtRest(value) {
        this._encryptAtRest.internalValue = value;
    }
    resetEncryptAtRest() {
        this._encryptAtRest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptAtRestInput() {
        return this._encryptAtRest.internalValue;
    }
    get logPublishingOptions() {
        return this._logPublishingOptions;
    }
    putLogPublishingOptions(value) {
        this._logPublishingOptions.internalValue = value;
    }
    resetLogPublishingOptions() {
        this._logPublishingOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logPublishingOptionsInput() {
        return this._logPublishingOptions.internalValue;
    }
    get nodeToNodeEncryption() {
        return this._nodeToNodeEncryption;
    }
    putNodeToNodeEncryption(value) {
        this._nodeToNodeEncryption.internalValue = value;
    }
    resetNodeToNodeEncryption() {
        this._nodeToNodeEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeToNodeEncryptionInput() {
        return this._nodeToNodeEncryption.internalValue;
    }
    get snapshotOptions() {
        return this._snapshotOptions;
    }
    putSnapshotOptions(value) {
        this._snapshotOptions.internalValue = value;
    }
    resetSnapshotOptions() {
        this._snapshotOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotOptionsInput() {
        return this._snapshotOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcOptions() {
        return this._vpcOptions;
    }
    putVpcOptions(value) {
        this._vpcOptions.internalValue = value;
    }
    resetVpcOptions() {
        this._vpcOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcOptionsInput() {
        return this._vpcOptions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_policies: cdktf.stringToTerraform(this._accessPolicies),
            advanced_options: cdktf.hashMapper(cdktf.stringToTerraform)(this._advancedOptions),
            domain_name: cdktf.stringToTerraform(this._domainName),
            elasticsearch_version: cdktf.stringToTerraform(this._elasticsearchVersion),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            advanced_security_options: elasticsearchDomainAdvancedSecurityOptionsToTerraform(this._advancedSecurityOptions.internalValue),
            auto_tune_options: elasticsearchDomainAutoTuneOptionsToTerraform(this._autoTuneOptions.internalValue),
            cluster_config: elasticsearchDomainClusterConfigToTerraform(this._clusterConfig.internalValue),
            cognito_options: elasticsearchDomainCognitoOptionsToTerraform(this._cognitoOptions.internalValue),
            domain_endpoint_options: elasticsearchDomainDomainEndpointOptionsToTerraform(this._domainEndpointOptions.internalValue),
            ebs_options: elasticsearchDomainEbsOptionsToTerraform(this._ebsOptions.internalValue),
            encrypt_at_rest: elasticsearchDomainEncryptAtRestToTerraform(this._encryptAtRest.internalValue),
            log_publishing_options: cdktf.listMapper(elasticsearchDomainLogPublishingOptionsToTerraform, true)(this._logPublishingOptions.internalValue),
            node_to_node_encryption: elasticsearchDomainNodeToNodeEncryptionToTerraform(this._nodeToNodeEncryption.internalValue),
            snapshot_options: elasticsearchDomainSnapshotOptionsToTerraform(this._snapshotOptions.internalValue),
            timeouts: elasticsearchDomainTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_options: elasticsearchDomainVpcOptionsToTerraform(this._vpcOptions.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_policies: {
                value: cdktf.stringToHclTerraform(this._accessPolicies),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            advanced_options: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._advancedOptions),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            elasticsearch_version: {
                value: cdktf.stringToHclTerraform(this._elasticsearchVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            advanced_security_options: {
                value: elasticsearchDomainAdvancedSecurityOptionsToHclTerraform(this._advancedSecurityOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainAdvancedSecurityOptionsList",
            },
            auto_tune_options: {
                value: elasticsearchDomainAutoTuneOptionsToHclTerraform(this._autoTuneOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainAutoTuneOptionsList",
            },
            cluster_config: {
                value: elasticsearchDomainClusterConfigToHclTerraform(this._clusterConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainClusterConfigList",
            },
            cognito_options: {
                value: elasticsearchDomainCognitoOptionsToHclTerraform(this._cognitoOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainCognitoOptionsList",
            },
            domain_endpoint_options: {
                value: elasticsearchDomainDomainEndpointOptionsToHclTerraform(this._domainEndpointOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainDomainEndpointOptionsList",
            },
            ebs_options: {
                value: elasticsearchDomainEbsOptionsToHclTerraform(this._ebsOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainEbsOptionsList",
            },
            encrypt_at_rest: {
                value: elasticsearchDomainEncryptAtRestToHclTerraform(this._encryptAtRest.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainEncryptAtRestList",
            },
            log_publishing_options: {
                value: cdktf.listMapperHcl(elasticsearchDomainLogPublishingOptionsToHclTerraform, true)(this._logPublishingOptions.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ElasticsearchDomainLogPublishingOptionsList",
            },
            node_to_node_encryption: {
                value: elasticsearchDomainNodeToNodeEncryptionToHclTerraform(this._nodeToNodeEncryption.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainNodeToNodeEncryptionList",
            },
            snapshot_options: {
                value: elasticsearchDomainSnapshotOptionsToHclTerraform(this._snapshotOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainSnapshotOptionsList",
            },
            timeouts: {
                value: elasticsearchDomainTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ElasticsearchDomainTimeouts",
            },
            vpc_options: {
                value: elasticsearchDomainVpcOptionsToHclTerraform(this._vpcOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainVpcOptionsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ElasticsearchDomain = ElasticsearchDomain;
_v = JSII_RTTI_SYMBOL_1;
ElasticsearchDomain[_v] = { fqn: "@cdktf/provider-aws.elasticsearchDomain.ElasticsearchDomain", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ElasticsearchDomain.tfResourceType = "aws_elasticsearch_domain";
