// @cdktf/provider-aws/lib/data-aws-lb-listener/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLbListener = exports.DataAwsLbListenerTimeoutsOutputReference = exports.dataAwsLbListenerTimeoutsToHclTerraform = exports.dataAwsLbListenerTimeoutsToTerraform = exports.DataAwsLbListenerMutualAuthenticationList = exports.DataAwsLbListenerMutualAuthenticationOutputReference = exports.dataAwsLbListenerMutualAuthenticationToHclTerraform = exports.dataAwsLbListenerMutualAuthenticationToTerraform = exports.DataAwsLbListenerDefaultActionList = exports.DataAwsLbListenerDefaultActionOutputReference = exports.dataAwsLbListenerDefaultActionToHclTerraform = exports.dataAwsLbListenerDefaultActionToTerraform = exports.DataAwsLbListenerDefaultActionRedirectList = exports.DataAwsLbListenerDefaultActionRedirectOutputReference = exports.dataAwsLbListenerDefaultActionRedirectToHclTerraform = exports.dataAwsLbListenerDefaultActionRedirectToTerraform = exports.DataAwsLbListenerDefaultActionForwardList = exports.DataAwsLbListenerDefaultActionForwardOutputReference = exports.dataAwsLbListenerDefaultActionForwardToHclTerraform = exports.dataAwsLbListenerDefaultActionForwardToTerraform = exports.DataAwsLbListenerDefaultActionForwardTargetGroupList = exports.DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference = exports.dataAwsLbListenerDefaultActionForwardTargetGroupToHclTerraform = exports.dataAwsLbListenerDefaultActionForwardTargetGroupToTerraform = exports.DataAwsLbListenerDefaultActionForwardStickinessList = exports.DataAwsLbListenerDefaultActionForwardStickinessOutputReference = exports.dataAwsLbListenerDefaultActionForwardStickinessToHclTerraform = exports.dataAwsLbListenerDefaultActionForwardStickinessToTerraform = exports.DataAwsLbListenerDefaultActionFixedResponseList = exports.DataAwsLbListenerDefaultActionFixedResponseOutputReference = exports.dataAwsLbListenerDefaultActionFixedResponseToHclTerraform = exports.dataAwsLbListenerDefaultActionFixedResponseToTerraform = exports.DataAwsLbListenerDefaultActionAuthenticateOidcList = exports.DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference = exports.dataAwsLbListenerDefaultActionAuthenticateOidcToHclTerraform = exports.dataAwsLbListenerDefaultActionAuthenticateOidcToTerraform = exports.DataAwsLbListenerDefaultActionAuthenticateCognitoList = exports.DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference = exports.dataAwsLbListenerDefaultActionAuthenticateCognitoToHclTerraform = exports.dataAwsLbListenerDefaultActionAuthenticateCognitoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLbListenerDefaultActionAuthenticateCognitoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionAuthenticateCognitoToTerraform = dataAwsLbListenerDefaultActionAuthenticateCognitoToTerraform;
function dataAwsLbListenerDefaultActionAuthenticateCognitoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionAuthenticateCognitoToHclTerraform = dataAwsLbListenerDefaultActionAuthenticateCognitoToHclTerraform;
class DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authentication_request_extra_params - computed: true, optional: false, required: false
        this._authenticationRequestExtraParams = new cdktf.StringMap(this, "authentication_request_extra_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticationRequestExtraParams() {
        return this._authenticationRequestExtraParams;
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // user_pool_arn - computed: true, optional: false, required: false
    get userPoolArn() {
        return this.getStringAttribute('user_pool_arn');
    }
    // user_pool_client_id - computed: true, optional: false, required: false
    get userPoolClientId() {
        return this.getStringAttribute('user_pool_client_id');
    }
    // user_pool_domain - computed: true, optional: false, required: false
    get userPoolDomain() {
        return this.getStringAttribute('user_pool_domain');
    }
}
exports.DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference = DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionAuthenticateCognitoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionAuthenticateCognitoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionAuthenticateCognitoList = DataAwsLbListenerDefaultActionAuthenticateCognitoList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionAuthenticateCognitoList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionAuthenticateCognitoList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionAuthenticateOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionAuthenticateOidcToTerraform = dataAwsLbListenerDefaultActionAuthenticateOidcToTerraform;
function dataAwsLbListenerDefaultActionAuthenticateOidcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionAuthenticateOidcToHclTerraform = dataAwsLbListenerDefaultActionAuthenticateOidcToHclTerraform;
class DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authentication_request_extra_params - computed: true, optional: false, required: false
        this._authenticationRequestExtraParams = new cdktf.StringMap(this, "authentication_request_extra_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticationRequestExtraParams() {
        return this._authenticationRequestExtraParams;
    }
    // authorization_endpoint - computed: true, optional: false, required: false
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    // client_id - computed: true, optional: false, required: false
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    // client_secret - computed: true, optional: false, required: false
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // token_endpoint - computed: true, optional: false, required: false
    get tokenEndpoint() {
        return this.getStringAttribute('token_endpoint');
    }
    // user_info_endpoint - computed: true, optional: false, required: false
    get userInfoEndpoint() {
        return this.getStringAttribute('user_info_endpoint');
    }
}
exports.DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference = DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionAuthenticateOidcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionAuthenticateOidcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionAuthenticateOidcList = DataAwsLbListenerDefaultActionAuthenticateOidcList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionAuthenticateOidcList[_d] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionAuthenticateOidcList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionFixedResponseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionFixedResponseToTerraform = dataAwsLbListenerDefaultActionFixedResponseToTerraform;
function dataAwsLbListenerDefaultActionFixedResponseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionFixedResponseToHclTerraform = dataAwsLbListenerDefaultActionFixedResponseToHclTerraform;
class DataAwsLbListenerDefaultActionFixedResponseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // content_type - computed: true, optional: false, required: false
    get contentType() {
        return this.getStringAttribute('content_type');
    }
    // message_body - computed: true, optional: false, required: false
    get messageBody() {
        return this.getStringAttribute('message_body');
    }
    // status_code - computed: true, optional: false, required: false
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsLbListenerDefaultActionFixedResponseOutputReference = DataAwsLbListenerDefaultActionFixedResponseOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionFixedResponseOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionFixedResponseOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionFixedResponseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionFixedResponseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionFixedResponseList = DataAwsLbListenerDefaultActionFixedResponseList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionFixedResponseList[_f] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionFixedResponseList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionForwardStickinessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionForwardStickinessToTerraform = dataAwsLbListenerDefaultActionForwardStickinessToTerraform;
function dataAwsLbListenerDefaultActionForwardStickinessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionForwardStickinessToHclTerraform = dataAwsLbListenerDefaultActionForwardStickinessToHclTerraform;
class DataAwsLbListenerDefaultActionForwardStickinessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // duration - computed: true, optional: false, required: false
    get duration() {
        return this.getNumberAttribute('duration');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsLbListenerDefaultActionForwardStickinessOutputReference = DataAwsLbListenerDefaultActionForwardStickinessOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardStickinessOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardStickinessOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionForwardStickinessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionForwardStickinessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionForwardStickinessList = DataAwsLbListenerDefaultActionForwardStickinessList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardStickinessList[_h] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardStickinessList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionForwardTargetGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionForwardTargetGroupToTerraform = dataAwsLbListenerDefaultActionForwardTargetGroupToTerraform;
function dataAwsLbListenerDefaultActionForwardTargetGroupToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionForwardTargetGroupToHclTerraform = dataAwsLbListenerDefaultActionForwardTargetGroupToHclTerraform;
class DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // weight - computed: true, optional: false, required: false
    get weight() {
        return this.getNumberAttribute('weight');
    }
}
exports.DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference = DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionForwardTargetGroupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionForwardTargetGroupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionForwardTargetGroupList = DataAwsLbListenerDefaultActionForwardTargetGroupList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardTargetGroupList[_k] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardTargetGroupList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionForwardToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionForwardToTerraform = dataAwsLbListenerDefaultActionForwardToTerraform;
function dataAwsLbListenerDefaultActionForwardToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionForwardToHclTerraform = dataAwsLbListenerDefaultActionForwardToHclTerraform;
class DataAwsLbListenerDefaultActionForwardOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // stickiness - computed: true, optional: false, required: false
        this._stickiness = new DataAwsLbListenerDefaultActionForwardStickinessList(this, "stickiness", false);
        // target_group - computed: true, optional: false, required: false
        this._targetGroup = new DataAwsLbListenerDefaultActionForwardTargetGroupList(this, "target_group", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get stickiness() {
        return this._stickiness;
    }
    get targetGroup() {
        return this._targetGroup;
    }
}
exports.DataAwsLbListenerDefaultActionForwardOutputReference = DataAwsLbListenerDefaultActionForwardOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionForwardList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionForwardOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionForwardList = DataAwsLbListenerDefaultActionForwardList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionForwardList[_m] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionForwardList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionRedirectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionRedirectToTerraform = dataAwsLbListenerDefaultActionRedirectToTerraform;
function dataAwsLbListenerDefaultActionRedirectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionRedirectToHclTerraform = dataAwsLbListenerDefaultActionRedirectToHclTerraform;
class DataAwsLbListenerDefaultActionRedirectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getStringAttribute('port');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // query - computed: true, optional: false, required: false
    get query() {
        return this.getStringAttribute('query');
    }
    // status_code - computed: true, optional: false, required: false
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsLbListenerDefaultActionRedirectOutputReference = DataAwsLbListenerDefaultActionRedirectOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionRedirectOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionRedirectOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionRedirectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionRedirectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionRedirectList = DataAwsLbListenerDefaultActionRedirectList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionRedirectList[_p] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionRedirectList", version: "19.15.0" };
function dataAwsLbListenerDefaultActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerDefaultActionToTerraform = dataAwsLbListenerDefaultActionToTerraform;
function dataAwsLbListenerDefaultActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerDefaultActionToHclTerraform = dataAwsLbListenerDefaultActionToHclTerraform;
class DataAwsLbListenerDefaultActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authenticate_cognito - computed: true, optional: false, required: false
        this._authenticateCognito = new DataAwsLbListenerDefaultActionAuthenticateCognitoList(this, "authenticate_cognito", false);
        // authenticate_oidc - computed: true, optional: false, required: false
        this._authenticateOidc = new DataAwsLbListenerDefaultActionAuthenticateOidcList(this, "authenticate_oidc", false);
        // fixed_response - computed: true, optional: false, required: false
        this._fixedResponse = new DataAwsLbListenerDefaultActionFixedResponseList(this, "fixed_response", false);
        // forward - computed: true, optional: false, required: false
        this._forward = new DataAwsLbListenerDefaultActionForwardList(this, "forward", false);
        // redirect - computed: true, optional: false, required: false
        this._redirect = new DataAwsLbListenerDefaultActionRedirectList(this, "redirect", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticateCognito() {
        return this._authenticateCognito;
    }
    get authenticateOidc() {
        return this._authenticateOidc;
    }
    get fixedResponse() {
        return this._fixedResponse;
    }
    get forward() {
        return this._forward;
    }
    // order - computed: true, optional: false, required: false
    get order() {
        return this.getNumberAttribute('order');
    }
    get redirect() {
        return this._redirect;
    }
    // target_group_arn - computed: true, optional: false, required: false
    get targetGroupArn() {
        return this.getStringAttribute('target_group_arn');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsLbListenerDefaultActionOutputReference = DataAwsLbListenerDefaultActionOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionOutputReference", version: "19.15.0" };
class DataAwsLbListenerDefaultActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerDefaultActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerDefaultActionList = DataAwsLbListenerDefaultActionList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerDefaultActionList[_r] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerDefaultActionList", version: "19.15.0" };
function dataAwsLbListenerMutualAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLbListenerMutualAuthenticationToTerraform = dataAwsLbListenerMutualAuthenticationToTerraform;
function dataAwsLbListenerMutualAuthenticationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLbListenerMutualAuthenticationToHclTerraform = dataAwsLbListenerMutualAuthenticationToHclTerraform;
class DataAwsLbListenerMutualAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ignore_client_certificate_expiry - computed: true, optional: false, required: false
    get ignoreClientCertificateExpiry() {
        return this.getBooleanAttribute('ignore_client_certificate_expiry');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // trust_store_arn - computed: true, optional: false, required: false
    get trustStoreArn() {
        return this.getStringAttribute('trust_store_arn');
    }
}
exports.DataAwsLbListenerMutualAuthenticationOutputReference = DataAwsLbListenerMutualAuthenticationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerMutualAuthenticationOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerMutualAuthenticationOutputReference", version: "19.15.0" };
class DataAwsLbListenerMutualAuthenticationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLbListenerMutualAuthenticationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLbListenerMutualAuthenticationList = DataAwsLbListenerMutualAuthenticationList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerMutualAuthenticationList[_t] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerMutualAuthenticationList", version: "19.15.0" };
function dataAwsLbListenerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsLbListenerTimeoutsToTerraform = dataAwsLbListenerTimeoutsToTerraform;
function dataAwsLbListenerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsLbListenerTimeoutsToHclTerraform = dataAwsLbListenerTimeoutsToHclTerraform;
class DataAwsLbListenerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsLbListenerTimeoutsOutputReference = DataAwsLbListenerTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsLbListenerTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListenerTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lb_listener aws_lb_listener}
*/
class DataAwsLbListener extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsLbListener resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsLbListener to import
    * @param importFromId The id of the existing DataAwsLbListener that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lb_listener#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsLbListener to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lb_listener", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lb_listener aws_lb_listener} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLbListenerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_lb_listener',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // default_action - computed: true, optional: false, required: false
        this._defaultAction = new DataAwsLbListenerDefaultActionList(this, "default_action", false);
        // mutual_authentication - computed: true, optional: false, required: false
        this._mutualAuthentication = new DataAwsLbListenerMutualAuthenticationList(this, "mutual_authentication", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsLbListenerTimeoutsOutputReference(this, "timeouts");
        this._arn = config.arn;
        this._id = config.id;
        this._loadBalancerArn = config.loadBalancerArn;
        this._port = config.port;
        this._tags = config.tags;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // alpn_policy - computed: true, optional: false, required: false
    get alpnPolicy() {
        return this.getStringAttribute('alpn_policy');
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // certificate_arn - computed: true, optional: false, required: false
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    get defaultAction() {
        return this._defaultAction;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get loadBalancerArn() {
        return this.getStringAttribute('load_balancer_arn');
    }
    set loadBalancerArn(value) {
        this._loadBalancerArn = value;
    }
    resetLoadBalancerArn() {
        this._loadBalancerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerArnInput() {
        return this._loadBalancerArn;
    }
    get mutualAuthentication() {
        return this._mutualAuthentication;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // ssl_policy - computed: true, optional: false, required: false
    get sslPolicy() {
        return this.getStringAttribute('ssl_policy');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            load_balancer_arn: cdktf.stringToTerraform(this._loadBalancerArn),
            port: cdktf.numberToTerraform(this._port),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            timeouts: dataAwsLbListenerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancer_arn: {
                value: cdktf.stringToHclTerraform(this._loadBalancerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            port: {
                value: cdktf.numberToHclTerraform(this._port),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeouts: {
                value: dataAwsLbListenerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsLbListenerTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsLbListener = DataAwsLbListener;
_v = JSII_RTTI_SYMBOL_1;
DataAwsLbListener[_v] = { fqn: "@cdktf/provider-aws.dataAwsLbListener.DataAwsLbListener", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLbListener.tfResourceType = "aws_lb_listener";
