// @winglang/sdk/lib/ui/field.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueField = exports.Field = exports.FIELD_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const base_1 = require("./base");
const cloud_1 = require("../cloud");
const constants_1 = require("../constants");
const core_1 = require("../core");
/**
 * Global identifier for `Field`.
 */
exports.FIELD_FQN = (0, constants_1.fqnForType)("ui.Field");
/**
 * A field can be used to display a value.
 * @noinflight
 */
class Field extends base_1.VisualComponent {
    /**
     * Creates a new ui.Field instance through the app.
     * @internal
     */
    static _newField(scope, id, label, handler, props = {}) {
        return core_1.App.of(scope).newAbstract(exports.FIELD_FQN, scope, id, label, handler, props);
    }
    constructor(scope, id, label, handler, props = {}) {
        super(scope, id);
        this.label = label;
        this.refreshRate = props.refreshRate?.seconds;
        this.fn = new cloud_1.Function(this, "Handler", handler);
        this.link = props.link;
    }
    /** @internal */
    _toUIComponent() {
        return {
            kind: "field",
            label: this.label,
            handler: this.fn.node.path,
            refreshRate: this.refreshRate,
            link: this.link,
        };
    }
    /** @internal */
    _toInflight() {
        throw new Error("Method not implemented.");
    }
}
exports.Field = Field;
_a = JSII_RTTI_SYMBOL_1;
Field[_a] = { fqn: "@winglang/sdk.ui.Field", version: "0.0.0" };
/**
 * A value field can be used to display a string value.
 * @noinflight
 */
class ValueField extends Field {
    constructor(scope, id, label, value) {
        const handler = (0, core_1.lift)({ value }).inflight(async (ctx) => {
            return ctx.value;
        });
        super(scope, id, label, handler);
    }
}
exports.ValueField = ValueField;
_b = JSII_RTTI_SYMBOL_1;
ValueField[_b] = { fqn: "@winglang/sdk.ui.ValueField", version: "0.0.0" };
