// @cdktf/provider-aws/lib/iot-thing-group-membership/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThingGroupMembership = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_thing_group_membership aws_iot_thing_group_membership}
*/
class IotThingGroupMembership extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IotThingGroupMembership resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IotThingGroupMembership to import
    * @param importFromId The id of the existing IotThingGroupMembership that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_thing_group_membership#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IotThingGroupMembership to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iot_thing_group_membership", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_thing_group_membership aws_iot_thing_group_membership} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotThingGroupMembershipConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing_group_membership',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._overrideDynamicGroup = config.overrideDynamicGroup;
        this._thingGroupName = config.thingGroupName;
        this._thingName = config.thingName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get overrideDynamicGroup() {
        return this.getBooleanAttribute('override_dynamic_group');
    }
    set overrideDynamicGroup(value) {
        this._overrideDynamicGroup = value;
    }
    resetOverrideDynamicGroup() {
        this._overrideDynamicGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overrideDynamicGroupInput() {
        return this._overrideDynamicGroup;
    }
    get thingGroupName() {
        return this.getStringAttribute('thing_group_name');
    }
    set thingGroupName(value) {
        this._thingGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get thingGroupNameInput() {
        return this._thingGroupName;
    }
    get thingName() {
        return this.getStringAttribute('thing_name');
    }
    set thingName(value) {
        this._thingName = value;
    }
    // Temporarily expose input value. Use with caution.
    get thingNameInput() {
        return this._thingName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            override_dynamic_group: cdktf.booleanToTerraform(this._overrideDynamicGroup),
            thing_group_name: cdktf.stringToTerraform(this._thingGroupName),
            thing_name: cdktf.stringToTerraform(this._thingName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            override_dynamic_group: {
                value: cdktf.booleanToHclTerraform(this._overrideDynamicGroup),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            thing_group_name: {
                value: cdktf.stringToHclTerraform(this._thingGroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            thing_name: {
                value: cdktf.stringToHclTerraform(this._thingName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IotThingGroupMembership = IotThingGroupMembership;
_a = JSII_RTTI_SYMBOL_1;
IotThingGroupMembership[_a] = { fqn: "@cdktf/provider-aws.iotThingGroupMembership.IotThingGroupMembership", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IotThingGroupMembership.tfResourceType = "aws_iot_thing_group_membership";
