// @cdktf/provider-aws/lib/default-network-acl/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultNetworkAcl = exports.DefaultNetworkAclIngressList = exports.DefaultNetworkAclIngressOutputReference = exports.defaultNetworkAclIngressToHclTerraform = exports.defaultNetworkAclIngressToTerraform = exports.DefaultNetworkAclEgressList = exports.DefaultNetworkAclEgressOutputReference = exports.defaultNetworkAclEgressToHclTerraform = exports.defaultNetworkAclEgressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function defaultNetworkAclEgressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        cidr_block: cdktf.stringToTerraform(struct.cidrBlock),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        icmp_code: cdktf.numberToTerraform(struct.icmpCode),
        icmp_type: cdktf.numberToTerraform(struct.icmpType),
        ipv6_cidr_block: cdktf.stringToTerraform(struct.ipv6CidrBlock),
        protocol: cdktf.stringToTerraform(struct.protocol),
        rule_no: cdktf.numberToTerraform(struct.ruleNo),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.defaultNetworkAclEgressToTerraform = defaultNetworkAclEgressToTerraform;
function defaultNetworkAclEgressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        cidr_block: {
            value: cdktf.stringToHclTerraform(struct.cidrBlock),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        from_port: {
            value: cdktf.numberToHclTerraform(struct.fromPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        icmp_code: {
            value: cdktf.numberToHclTerraform(struct.icmpCode),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        icmp_type: {
            value: cdktf.numberToHclTerraform(struct.icmpType),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        ipv6_cidr_block: {
            value: cdktf.stringToHclTerraform(struct.ipv6CidrBlock),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rule_no: {
            value: cdktf.numberToHclTerraform(struct.ruleNo),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        to_port: {
            value: cdktf.numberToHclTerraform(struct.toPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.defaultNetworkAclEgressToHclTerraform = defaultNetworkAclEgressToHclTerraform;
class DefaultNetworkAclEgressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._cidrBlock !== undefined) {
            hasAnyValues = true;
            internalValueResult.cidrBlock = this._cidrBlock;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._icmpCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.icmpCode = this._icmpCode;
        }
        if (this._icmpType !== undefined) {
            hasAnyValues = true;
            internalValueResult.icmpType = this._icmpType;
        }
        if (this._ipv6CidrBlock !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6CidrBlock = this._ipv6CidrBlock;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._ruleNo !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleNo = this._ruleNo;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._cidrBlock = undefined;
            this._fromPort = undefined;
            this._icmpCode = undefined;
            this._icmpType = undefined;
            this._ipv6CidrBlock = undefined;
            this._protocol = undefined;
            this._ruleNo = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._cidrBlock = value.cidrBlock;
            this._fromPort = value.fromPort;
            this._icmpCode = value.icmpCode;
            this._icmpType = value.icmpType;
            this._ipv6CidrBlock = value.ipv6CidrBlock;
            this._protocol = value.protocol;
            this._ruleNo = value.ruleNo;
            this._toPort = value.toPort;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
    set cidrBlock(value) {
        this._cidrBlock = value;
    }
    resetCidrBlock() {
        this._cidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlockInput() {
        return this._cidrBlock;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get icmpCode() {
        return this.getNumberAttribute('icmp_code');
    }
    set icmpCode(value) {
        this._icmpCode = value;
    }
    resetIcmpCode() {
        this._icmpCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get icmpCodeInput() {
        return this._icmpCode;
    }
    get icmpType() {
        return this.getNumberAttribute('icmp_type');
    }
    set icmpType(value) {
        this._icmpType = value;
    }
    resetIcmpType() {
        this._icmpType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get icmpTypeInput() {
        return this._icmpType;
    }
    get ipv6CidrBlock() {
        return this.getStringAttribute('ipv6_cidr_block');
    }
    set ipv6CidrBlock(value) {
        this._ipv6CidrBlock = value;
    }
    resetIpv6CidrBlock() {
        this._ipv6CidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6CidrBlockInput() {
        return this._ipv6CidrBlock;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get ruleNo() {
        return this.getNumberAttribute('rule_no');
    }
    set ruleNo(value) {
        this._ruleNo = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleNoInput() {
        return this._ruleNo;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.DefaultNetworkAclEgressOutputReference = DefaultNetworkAclEgressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DefaultNetworkAclEgressOutputReference[_a] = { fqn: "@cdktf/provider-aws.defaultNetworkAcl.DefaultNetworkAclEgressOutputReference", version: "19.15.0" };
class DefaultNetworkAclEgressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DefaultNetworkAclEgressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DefaultNetworkAclEgressList = DefaultNetworkAclEgressList;
_b = JSII_RTTI_SYMBOL_1;
DefaultNetworkAclEgressList[_b] = { fqn: "@cdktf/provider-aws.defaultNetworkAcl.DefaultNetworkAclEgressList", version: "19.15.0" };
function defaultNetworkAclIngressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        cidr_block: cdktf.stringToTerraform(struct.cidrBlock),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        icmp_code: cdktf.numberToTerraform(struct.icmpCode),
        icmp_type: cdktf.numberToTerraform(struct.icmpType),
        ipv6_cidr_block: cdktf.stringToTerraform(struct.ipv6CidrBlock),
        protocol: cdktf.stringToTerraform(struct.protocol),
        rule_no: cdktf.numberToTerraform(struct.ruleNo),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.defaultNetworkAclIngressToTerraform = defaultNetworkAclIngressToTerraform;
function defaultNetworkAclIngressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        cidr_block: {
            value: cdktf.stringToHclTerraform(struct.cidrBlock),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        from_port: {
            value: cdktf.numberToHclTerraform(struct.fromPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        icmp_code: {
            value: cdktf.numberToHclTerraform(struct.icmpCode),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        icmp_type: {
            value: cdktf.numberToHclTerraform(struct.icmpType),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        ipv6_cidr_block: {
            value: cdktf.stringToHclTerraform(struct.ipv6CidrBlock),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rule_no: {
            value: cdktf.numberToHclTerraform(struct.ruleNo),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        to_port: {
            value: cdktf.numberToHclTerraform(struct.toPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.defaultNetworkAclIngressToHclTerraform = defaultNetworkAclIngressToHclTerraform;
class DefaultNetworkAclIngressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._cidrBlock !== undefined) {
            hasAnyValues = true;
            internalValueResult.cidrBlock = this._cidrBlock;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._icmpCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.icmpCode = this._icmpCode;
        }
        if (this._icmpType !== undefined) {
            hasAnyValues = true;
            internalValueResult.icmpType = this._icmpType;
        }
        if (this._ipv6CidrBlock !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6CidrBlock = this._ipv6CidrBlock;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._ruleNo !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleNo = this._ruleNo;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._cidrBlock = undefined;
            this._fromPort = undefined;
            this._icmpCode = undefined;
            this._icmpType = undefined;
            this._ipv6CidrBlock = undefined;
            this._protocol = undefined;
            this._ruleNo = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._cidrBlock = value.cidrBlock;
            this._fromPort = value.fromPort;
            this._icmpCode = value.icmpCode;
            this._icmpType = value.icmpType;
            this._ipv6CidrBlock = value.ipv6CidrBlock;
            this._protocol = value.protocol;
            this._ruleNo = value.ruleNo;
            this._toPort = value.toPort;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
    set cidrBlock(value) {
        this._cidrBlock = value;
    }
    resetCidrBlock() {
        this._cidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlockInput() {
        return this._cidrBlock;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get icmpCode() {
        return this.getNumberAttribute('icmp_code');
    }
    set icmpCode(value) {
        this._icmpCode = value;
    }
    resetIcmpCode() {
        this._icmpCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get icmpCodeInput() {
        return this._icmpCode;
    }
    get icmpType() {
        return this.getNumberAttribute('icmp_type');
    }
    set icmpType(value) {
        this._icmpType = value;
    }
    resetIcmpType() {
        this._icmpType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get icmpTypeInput() {
        return this._icmpType;
    }
    get ipv6CidrBlock() {
        return this.getStringAttribute('ipv6_cidr_block');
    }
    set ipv6CidrBlock(value) {
        this._ipv6CidrBlock = value;
    }
    resetIpv6CidrBlock() {
        this._ipv6CidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6CidrBlockInput() {
        return this._ipv6CidrBlock;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get ruleNo() {
        return this.getNumberAttribute('rule_no');
    }
    set ruleNo(value) {
        this._ruleNo = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleNoInput() {
        return this._ruleNo;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.DefaultNetworkAclIngressOutputReference = DefaultNetworkAclIngressOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DefaultNetworkAclIngressOutputReference[_c] = { fqn: "@cdktf/provider-aws.defaultNetworkAcl.DefaultNetworkAclIngressOutputReference", version: "19.15.0" };
class DefaultNetworkAclIngressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DefaultNetworkAclIngressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DefaultNetworkAclIngressList = DefaultNetworkAclIngressList;
_d = JSII_RTTI_SYMBOL_1;
DefaultNetworkAclIngressList[_d] = { fqn: "@cdktf/provider-aws.defaultNetworkAcl.DefaultNetworkAclIngressList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_network_acl aws_default_network_acl}
*/
class DefaultNetworkAcl extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DefaultNetworkAcl resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DefaultNetworkAcl to import
    * @param importFromId The id of the existing DefaultNetworkAcl that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_network_acl#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DefaultNetworkAcl to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_default_network_acl", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_network_acl aws_default_network_acl} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DefaultNetworkAclConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_default_network_acl',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // egress - computed: false, optional: true, required: false
        this._egress = new DefaultNetworkAclEgressList(this, "egress", true);
        // ingress - computed: false, optional: true, required: false
        this._ingress = new DefaultNetworkAclIngressList(this, "ingress", true);
        this._defaultNetworkAclId = config.defaultNetworkAclId;
        this._id = config.id;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._egress.internalValue = config.egress;
        this._ingress.internalValue = config.ingress;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get defaultNetworkAclId() {
        return this.getStringAttribute('default_network_acl_id');
    }
    set defaultNetworkAclId(value) {
        this._defaultNetworkAclId = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultNetworkAclIdInput() {
        return this._defaultNetworkAclId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    get egress() {
        return this._egress;
    }
    putEgress(value) {
        this._egress.internalValue = value;
    }
    resetEgress() {
        this._egress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress.internalValue;
    }
    get ingress() {
        return this._ingress;
    }
    putIngress(value) {
        this._ingress.internalValue = value;
    }
    resetIngress() {
        this._ingress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressInput() {
        return this._ingress.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_network_acl_id: cdktf.stringToTerraform(this._defaultNetworkAclId),
            id: cdktf.stringToTerraform(this._id),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            egress: cdktf.listMapper(defaultNetworkAclEgressToTerraform, true)(this._egress.internalValue),
            ingress: cdktf.listMapper(defaultNetworkAclIngressToTerraform, true)(this._ingress.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            default_network_acl_id: {
                value: cdktf.stringToHclTerraform(this._defaultNetworkAclId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._subnetIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            egress: {
                value: cdktf.listMapperHcl(defaultNetworkAclEgressToHclTerraform, true)(this._egress.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DefaultNetworkAclEgressList",
            },
            ingress: {
                value: cdktf.listMapperHcl(defaultNetworkAclIngressToHclTerraform, true)(this._ingress.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DefaultNetworkAclIngressList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DefaultNetworkAcl = DefaultNetworkAcl;
_e = JSII_RTTI_SYMBOL_1;
DefaultNetworkAcl[_e] = { fqn: "@cdktf/provider-aws.defaultNetworkAcl.DefaultNetworkAcl", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DefaultNetworkAcl.tfResourceType = "aws_default_network_acl";
