// @winglang/sdk/lib/shared-aws/api.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiEndpointHandler = exports.Api = exports.STAGE_NAME = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../core");
/**
 * The stage name for the API, used in its url.
 * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-custom-domains.html
 */
exports.STAGE_NAME = "prod";
/**
 * A helper class for working with AWS apis.
 */
class Api {
    /**
     * If the api is an AWS RestApi, return a helper interface for
     * working with it.
     * @param api The cloud.Api.
     */
    static from(api) {
        if (this.isAwsApi(api)) {
            return api;
        }
        return undefined;
    }
    static isAwsApi(obj) {
        return (typeof obj.restApiArn === "string" &&
            typeof obj.restApiId === "string" &&
            typeof obj.restApiName === "string" &&
            typeof obj.stageName === "string" &&
            typeof obj.invokeUrl === "string" &&
            typeof obj.deploymentId === "string");
    }
}
exports.Api = Api;
_a = JSII_RTTI_SYMBOL_1;
Api[_a] = { fqn: "@winglang/sdk.aws.Api", version: "0.0.0" };
/**
 * A helper class for working with AWS api endpoint handlers.
 */
class ApiEndpointHandler {
    /**
     * Returns a `cloud.Function` handler for handling requests from a `cloud.Api`.
     * @param handler The `onRequest` handler.
     * @param headers HTTP response headers to add to all responses (used by CORS)
     * @returns The `cloud.Function` handler.
     */
    static toFunctionHandler(handler, headers) {
        return (0, core_1.lift)({
            handler,
            headers: headers ?? {},
        }).inflight(async (ctx, request) => {
            const { apigwFunctionHandler,
            // eslint-disable-next-line @typescript-eslint/no-require-imports
             } = require("@winglang/sdk/lib/shared-aws/api-util.js");
            return apigwFunctionHandler(request, ctx.handler, ctx.headers);
        });
    }
}
exports.ApiEndpointHandler = ApiEndpointHandler;
_b = JSII_RTTI_SYMBOL_1;
ApiEndpointHandler[_b] = { fqn: "@winglang/sdk.aws.ApiEndpointHandler", version: "0.0.0" };
