// @winglang/sdk/lib/shared-aws/secret.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretClient = void 0;
const client_secrets_manager_1 = require("@aws-sdk/client-secrets-manager");
class SecretClient {
    constructor(secretArn, client = new client_secrets_manager_1.SecretsManagerClient({})) {
        this.secretArn = secretArn;
        this.client = client;
    }
    async value(options = {}) {
        if ((options.cache ?? true) && this.secretValue) {
            return this.secretValue;
        }
        const command = new client_secrets_manager_1.GetSecretValueCommand({
            SecretId: this.secretArn,
        });
        const getSecretValue = await this.client.send(command);
        if (!getSecretValue.SecretString) {
            throw new Error(`No value for secret found\n(hint: try running the 'wing secrets -t TARGET' to store secret)`);
        }
        this.secretValue = getSecretValue.SecretString;
        return this.secretValue;
    }
    async valueJson(options = {}) {
        return JSON.parse(await this.value(options));
    }
}
exports.SecretClient = SecretClient;
