// @winglang/sdk/lib/shared-aws/topic.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicClient = void 0;
const client_sns_1 = require("@aws-sdk/client-sns");
const util_1 = require("../util/util");
/**
 * Topics in AWS can receive up to 10 messages at a time
 * using the PublishBatchCommand, this constant is used
 * to generate batches respecting the limits.
 */
const CHUNK_SIZE = 10;
class TopicClient {
    constructor(topicArn, client = new client_sns_1.SNSClient({})) {
        this.topicArn = topicArn;
        this.client = client;
    }
    async publish(...messages) {
        if (messages.includes("")) {
            throw new Error("Empty messages are not allowed");
        }
        let batchMessages = [];
        for (let i = 0; i < messages.length; i += CHUNK_SIZE) {
            const chunk = messages.slice(i, i + CHUNK_SIZE);
            batchMessages.push(this.processBatchMessages(chunk, i));
        }
        for (const batch of batchMessages) {
            try {
                const command = new client_sns_1.PublishBatchCommand({
                    TopicArn: this.topicArn,
                    PublishBatchRequestEntries: batch,
                });
                await this.client.send(command);
            }
            catch (e) {
                if (e instanceof client_sns_1.InvalidBatchEntryIdException) {
                    throw new Error(`The Id of a batch entry in a batch request doesn't abide by the specification. (message=${messages}): ${e.stack})}`);
                }
                throw new Error(e.stack);
            }
        }
    }
    processBatchMessages(messages, idx) {
        let batchMessages = [];
        let index = idx;
        for (const message of messages) {
            batchMessages.push({
                Id: util_1.Util.sha256(`${message}-${++index}`),
                Message: message,
            });
        }
        return batchMessages;
    }
}
exports.TopicClient = TopicClient;
