// @cdktf/provider-aws/lib/data-aws-s3-objects/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsS3Objects = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/s3_objects aws_s3_objects}
*/
class DataAwsS3Objects extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsS3Objects resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsS3Objects to import
    * @param importFromId The id of the existing DataAwsS3Objects that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/s3_objects#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsS3Objects to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_s3_objects", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/s3_objects aws_s3_objects} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsS3ObjectsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_objects',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._bucket = config.bucket;
        this._delimiter = config.delimiter;
        this._encodingType = config.encodingType;
        this._fetchOwner = config.fetchOwner;
        this._id = config.id;
        this._maxKeys = config.maxKeys;
        this._prefix = config.prefix;
        this._requestPayer = config.requestPayer;
        this._startAfter = config.startAfter;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    // common_prefixes - computed: true, optional: false, required: false
    get commonPrefixes() {
        return this.getListAttribute('common_prefixes');
    }
    get delimiter() {
        return this.getStringAttribute('delimiter');
    }
    set delimiter(value) {
        this._delimiter = value;
    }
    resetDelimiter() {
        this._delimiter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get delimiterInput() {
        return this._delimiter;
    }
    get encodingType() {
        return this.getStringAttribute('encoding_type');
    }
    set encodingType(value) {
        this._encodingType = value;
    }
    resetEncodingType() {
        this._encodingType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encodingTypeInput() {
        return this._encodingType;
    }
    get fetchOwner() {
        return this.getBooleanAttribute('fetch_owner');
    }
    set fetchOwner(value) {
        this._fetchOwner = value;
    }
    resetFetchOwner() {
        this._fetchOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fetchOwnerInput() {
        return this._fetchOwner;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // keys - computed: true, optional: false, required: false
    get keys() {
        return this.getListAttribute('keys');
    }
    get maxKeys() {
        return this.getNumberAttribute('max_keys');
    }
    set maxKeys(value) {
        this._maxKeys = value;
    }
    resetMaxKeys() {
        this._maxKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxKeysInput() {
        return this._maxKeys;
    }
    // owners - computed: true, optional: false, required: false
    get owners() {
        return this.getListAttribute('owners');
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    // request_charged - computed: true, optional: false, required: false
    get requestCharged() {
        return this.getStringAttribute('request_charged');
    }
    get requestPayer() {
        return this.getStringAttribute('request_payer');
    }
    set requestPayer(value) {
        this._requestPayer = value;
    }
    resetRequestPayer() {
        this._requestPayer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestPayerInput() {
        return this._requestPayer;
    }
    get startAfter() {
        return this.getStringAttribute('start_after');
    }
    set startAfter(value) {
        this._startAfter = value;
    }
    resetStartAfter() {
        this._startAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startAfterInput() {
        return this._startAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            delimiter: cdktf.stringToTerraform(this._delimiter),
            encoding_type: cdktf.stringToTerraform(this._encodingType),
            fetch_owner: cdktf.booleanToTerraform(this._fetchOwner),
            id: cdktf.stringToTerraform(this._id),
            max_keys: cdktf.numberToTerraform(this._maxKeys),
            prefix: cdktf.stringToTerraform(this._prefix),
            request_payer: cdktf.stringToTerraform(this._requestPayer),
            start_after: cdktf.stringToTerraform(this._startAfter),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bucket: {
                value: cdktf.stringToHclTerraform(this._bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            delimiter: {
                value: cdktf.stringToHclTerraform(this._delimiter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            encoding_type: {
                value: cdktf.stringToHclTerraform(this._encodingType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            fetch_owner: {
                value: cdktf.booleanToHclTerraform(this._fetchOwner),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            max_keys: {
                value: cdktf.numberToHclTerraform(this._maxKeys),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            prefix: {
                value: cdktf.stringToHclTerraform(this._prefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            request_payer: {
                value: cdktf.stringToHclTerraform(this._requestPayer),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            start_after: {
                value: cdktf.stringToHclTerraform(this._startAfter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsS3Objects = DataAwsS3Objects;
_a = JSII_RTTI_SYMBOL_1;
DataAwsS3Objects[_a] = { fqn: "@cdktf/provider-aws.dataAwsS3Objects.DataAwsS3Objects", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsS3Objects.tfResourceType = "aws_s3_objects";
