// @cdktf/provider-aws/lib/data-aws-redshift-cluster-credentials/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftClusterCredentials = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_cluster_credentials aws_redshift_cluster_credentials}
*/
class DataAwsRedshiftClusterCredentials extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsRedshiftClusterCredentials resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsRedshiftClusterCredentials to import
    * @param importFromId The id of the existing DataAwsRedshiftClusterCredentials that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_cluster_credentials#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsRedshiftClusterCredentials to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshift_cluster_credentials", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_cluster_credentials aws_redshift_cluster_credentials} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRedshiftClusterCredentialsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_cluster_credentials',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._autoCreate = config.autoCreate;
        this._clusterIdentifier = config.clusterIdentifier;
        this._dbGroups = config.dbGroups;
        this._dbName = config.dbName;
        this._dbUser = config.dbUser;
        this._durationSeconds = config.durationSeconds;
        this._id = config.id;
    }
    get autoCreate() {
        return this.getBooleanAttribute('auto_create');
    }
    set autoCreate(value) {
        this._autoCreate = value;
    }
    resetAutoCreate() {
        this._autoCreate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoCreateInput() {
        return this._autoCreate;
    }
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    get dbGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('db_groups'));
    }
    set dbGroups(value) {
        this._dbGroups = value;
    }
    resetDbGroups() {
        this._dbGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbGroupsInput() {
        return this._dbGroups;
    }
    get dbName() {
        return this.getStringAttribute('db_name');
    }
    set dbName(value) {
        this._dbName = value;
    }
    resetDbName() {
        this._dbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbNameInput() {
        return this._dbName;
    }
    // db_password - computed: true, optional: false, required: false
    get dbPassword() {
        return this.getStringAttribute('db_password');
    }
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    set dbUser(value) {
        this._dbUser = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbUserInput() {
        return this._dbUser;
    }
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    // expiration - computed: true, optional: false, required: false
    get expiration() {
        return this.getStringAttribute('expiration');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_create: cdktf.booleanToTerraform(this._autoCreate),
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            db_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(this._dbGroups),
            db_name: cdktf.stringToTerraform(this._dbName),
            db_user: cdktf.stringToTerraform(this._dbUser),
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_create: {
                value: cdktf.booleanToHclTerraform(this._autoCreate),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            cluster_identifier: {
                value: cdktf.stringToHclTerraform(this._clusterIdentifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            db_groups: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._dbGroups),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            db_name: {
                value: cdktf.stringToHclTerraform(this._dbName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            db_user: {
                value: cdktf.stringToHclTerraform(this._dbUser),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            duration_seconds: {
                value: cdktf.numberToHclTerraform(this._durationSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsRedshiftClusterCredentials = DataAwsRedshiftClusterCredentials;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftClusterCredentials[_a] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftClusterCredentials.DataAwsRedshiftClusterCredentials", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRedshiftClusterCredentials.tfResourceType = "aws_redshift_cluster_credentials";
