// @cdktf/provider-aws/lib/rbin-rule/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RbinRule = exports.RbinRuleTimeoutsOutputReference = exports.rbinRuleTimeoutsToHclTerraform = exports.rbinRuleTimeoutsToTerraform = exports.RbinRuleRetentionPeriodOutputReference = exports.rbinRuleRetentionPeriodToHclTerraform = exports.rbinRuleRetentionPeriodToTerraform = exports.RbinRuleResourceTagsList = exports.RbinRuleResourceTagsOutputReference = exports.rbinRuleResourceTagsToHclTerraform = exports.rbinRuleResourceTagsToTerraform = exports.RbinRuleLockConfigurationOutputReference = exports.rbinRuleLockConfigurationToHclTerraform = exports.rbinRuleLockConfigurationToTerraform = exports.RbinRuleLockConfigurationUnlockDelayOutputReference = exports.rbinRuleLockConfigurationUnlockDelayToHclTerraform = exports.rbinRuleLockConfigurationUnlockDelayToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function rbinRuleLockConfigurationUnlockDelayToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unlock_delay_unit: cdktf.stringToTerraform(struct.unlockDelayUnit),
        unlock_delay_value: cdktf.numberToTerraform(struct.unlockDelayValue),
    };
}
exports.rbinRuleLockConfigurationUnlockDelayToTerraform = rbinRuleLockConfigurationUnlockDelayToTerraform;
function rbinRuleLockConfigurationUnlockDelayToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        unlock_delay_unit: {
            value: cdktf.stringToHclTerraform(struct.unlockDelayUnit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        unlock_delay_value: {
            value: cdktf.numberToHclTerraform(struct.unlockDelayValue),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rbinRuleLockConfigurationUnlockDelayToHclTerraform = rbinRuleLockConfigurationUnlockDelayToHclTerraform;
class RbinRuleLockConfigurationUnlockDelayOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unlockDelayUnit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unlockDelayUnit = this._unlockDelayUnit;
        }
        if (this._unlockDelayValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.unlockDelayValue = this._unlockDelayValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unlockDelayUnit = undefined;
            this._unlockDelayValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unlockDelayUnit = value.unlockDelayUnit;
            this._unlockDelayValue = value.unlockDelayValue;
        }
    }
    get unlockDelayUnit() {
        return this.getStringAttribute('unlock_delay_unit');
    }
    set unlockDelayUnit(value) {
        this._unlockDelayUnit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unlockDelayUnitInput() {
        return this._unlockDelayUnit;
    }
    get unlockDelayValue() {
        return this.getNumberAttribute('unlock_delay_value');
    }
    set unlockDelayValue(value) {
        this._unlockDelayValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get unlockDelayValueInput() {
        return this._unlockDelayValue;
    }
}
exports.RbinRuleLockConfigurationUnlockDelayOutputReference = RbinRuleLockConfigurationUnlockDelayOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RbinRuleLockConfigurationUnlockDelayOutputReference[_a] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleLockConfigurationUnlockDelayOutputReference", version: "19.15.0" };
function rbinRuleLockConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unlock_delay: rbinRuleLockConfigurationUnlockDelayToTerraform(struct.unlockDelay),
    };
}
exports.rbinRuleLockConfigurationToTerraform = rbinRuleLockConfigurationToTerraform;
function rbinRuleLockConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        unlock_delay: {
            value: rbinRuleLockConfigurationUnlockDelayToHclTerraform(struct.unlockDelay),
            isBlock: true,
            type: "list",
            storageClassType: "RbinRuleLockConfigurationUnlockDelayList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rbinRuleLockConfigurationToHclTerraform = rbinRuleLockConfigurationToHclTerraform;
class RbinRuleLockConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // unlock_delay - computed: false, optional: false, required: true
        this._unlockDelay = new RbinRuleLockConfigurationUnlockDelayOutputReference(this, "unlock_delay");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unlockDelay?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.unlockDelay = this._unlockDelay?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unlockDelay.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unlockDelay.internalValue = value.unlockDelay;
        }
    }
    get unlockDelay() {
        return this._unlockDelay;
    }
    putUnlockDelay(value) {
        this._unlockDelay.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get unlockDelayInput() {
        return this._unlockDelay.internalValue;
    }
}
exports.RbinRuleLockConfigurationOutputReference = RbinRuleLockConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
RbinRuleLockConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleLockConfigurationOutputReference", version: "19.15.0" };
function rbinRuleResourceTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_tag_key: cdktf.stringToTerraform(struct.resourceTagKey),
        resource_tag_value: cdktf.stringToTerraform(struct.resourceTagValue),
    };
}
exports.rbinRuleResourceTagsToTerraform = rbinRuleResourceTagsToTerraform;
function rbinRuleResourceTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resource_tag_key: {
            value: cdktf.stringToHclTerraform(struct.resourceTagKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resource_tag_value: {
            value: cdktf.stringToHclTerraform(struct.resourceTagValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rbinRuleResourceTagsToHclTerraform = rbinRuleResourceTagsToHclTerraform;
class RbinRuleResourceTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceTagKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceTagKey = this._resourceTagKey;
        }
        if (this._resourceTagValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceTagValue = this._resourceTagValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._resourceTagKey = undefined;
            this._resourceTagValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._resourceTagKey = value.resourceTagKey;
            this._resourceTagValue = value.resourceTagValue;
        }
    }
    get resourceTagKey() {
        return this.getStringAttribute('resource_tag_key');
    }
    set resourceTagKey(value) {
        this._resourceTagKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTagKeyInput() {
        return this._resourceTagKey;
    }
    get resourceTagValue() {
        return this.getStringAttribute('resource_tag_value');
    }
    set resourceTagValue(value) {
        this._resourceTagValue = value;
    }
    resetResourceTagValue() {
        this._resourceTagValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTagValueInput() {
        return this._resourceTagValue;
    }
}
exports.RbinRuleResourceTagsOutputReference = RbinRuleResourceTagsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RbinRuleResourceTagsOutputReference[_c] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleResourceTagsOutputReference", version: "19.15.0" };
class RbinRuleResourceTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RbinRuleResourceTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RbinRuleResourceTagsList = RbinRuleResourceTagsList;
_d = JSII_RTTI_SYMBOL_1;
RbinRuleResourceTagsList[_d] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleResourceTagsList", version: "19.15.0" };
function rbinRuleRetentionPeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        retention_period_unit: cdktf.stringToTerraform(struct.retentionPeriodUnit),
        retention_period_value: cdktf.numberToTerraform(struct.retentionPeriodValue),
    };
}
exports.rbinRuleRetentionPeriodToTerraform = rbinRuleRetentionPeriodToTerraform;
function rbinRuleRetentionPeriodToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        retention_period_unit: {
            value: cdktf.stringToHclTerraform(struct.retentionPeriodUnit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        retention_period_value: {
            value: cdktf.numberToHclTerraform(struct.retentionPeriodValue),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rbinRuleRetentionPeriodToHclTerraform = rbinRuleRetentionPeriodToHclTerraform;
class RbinRuleRetentionPeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._retentionPeriodUnit !== undefined) {
            hasAnyValues = true;
            internalValueResult.retentionPeriodUnit = this._retentionPeriodUnit;
        }
        if (this._retentionPeriodValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retentionPeriodValue = this._retentionPeriodValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._retentionPeriodUnit = undefined;
            this._retentionPeriodValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._retentionPeriodUnit = value.retentionPeriodUnit;
            this._retentionPeriodValue = value.retentionPeriodValue;
        }
    }
    get retentionPeriodUnit() {
        return this.getStringAttribute('retention_period_unit');
    }
    set retentionPeriodUnit(value) {
        this._retentionPeriodUnit = value;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodUnitInput() {
        return this._retentionPeriodUnit;
    }
    get retentionPeriodValue() {
        return this.getNumberAttribute('retention_period_value');
    }
    set retentionPeriodValue(value) {
        this._retentionPeriodValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodValueInput() {
        return this._retentionPeriodValue;
    }
}
exports.RbinRuleRetentionPeriodOutputReference = RbinRuleRetentionPeriodOutputReference;
_e = JSII_RTTI_SYMBOL_1;
RbinRuleRetentionPeriodOutputReference[_e] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleRetentionPeriodOutputReference", version: "19.15.0" };
function rbinRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.rbinRuleTimeoutsToTerraform = rbinRuleTimeoutsToTerraform;
function rbinRuleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rbinRuleTimeoutsToHclTerraform = rbinRuleTimeoutsToHclTerraform;
class RbinRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.RbinRuleTimeoutsOutputReference = RbinRuleTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
RbinRuleTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRuleTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rbin_rule aws_rbin_rule}
*/
class RbinRule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RbinRule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RbinRule to import
    * @param importFromId The id of the existing RbinRule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rbin_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RbinRule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_rbin_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rbin_rule aws_rbin_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RbinRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rbin_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // lock_configuration - computed: false, optional: true, required: false
        this._lockConfiguration = new RbinRuleLockConfigurationOutputReference(this, "lock_configuration");
        // resource_tags - computed: false, optional: true, required: false
        this._resourceTags = new RbinRuleResourceTagsList(this, "resource_tags", true);
        // retention_period - computed: false, optional: false, required: true
        this._retentionPeriod = new RbinRuleRetentionPeriodOutputReference(this, "retention_period");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RbinRuleTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._resourceType = config.resourceType;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._lockConfiguration.internalValue = config.lockConfiguration;
        this._resourceTags.internalValue = config.resourceTags;
        this._retentionPeriod.internalValue = config.retentionPeriod;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // lock_end_time - computed: true, optional: false, required: false
    get lockEndTime() {
        return this.getStringAttribute('lock_end_time');
    }
    // lock_state - computed: true, optional: false, required: false
    get lockState() {
        return this.getStringAttribute('lock_state');
    }
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    set resourceType(value) {
        this._resourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypeInput() {
        return this._resourceType;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get lockConfiguration() {
        return this._lockConfiguration;
    }
    putLockConfiguration(value) {
        this._lockConfiguration.internalValue = value;
    }
    resetLockConfiguration() {
        this._lockConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lockConfigurationInput() {
        return this._lockConfiguration.internalValue;
    }
    get resourceTags() {
        return this._resourceTags;
    }
    putResourceTags(value) {
        this._resourceTags.internalValue = value;
    }
    resetResourceTags() {
        this._resourceTags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTagsInput() {
        return this._resourceTags.internalValue;
    }
    get retentionPeriod() {
        return this._retentionPeriod;
    }
    putRetentionPeriod(value) {
        this._retentionPeriod.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            resource_type: cdktf.stringToTerraform(this._resourceType),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            lock_configuration: rbinRuleLockConfigurationToTerraform(this._lockConfiguration.internalValue),
            resource_tags: cdktf.listMapper(rbinRuleResourceTagsToTerraform, true)(this._resourceTags.internalValue),
            retention_period: rbinRuleRetentionPeriodToTerraform(this._retentionPeriod.internalValue),
            timeouts: rbinRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_type: {
                value: cdktf.stringToHclTerraform(this._resourceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            lock_configuration: {
                value: rbinRuleLockConfigurationToHclTerraform(this._lockConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RbinRuleLockConfigurationList",
            },
            resource_tags: {
                value: cdktf.listMapperHcl(rbinRuleResourceTagsToHclTerraform, true)(this._resourceTags.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "RbinRuleResourceTagsList",
            },
            retention_period: {
                value: rbinRuleRetentionPeriodToHclTerraform(this._retentionPeriod.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RbinRuleRetentionPeriodList",
            },
            timeouts: {
                value: rbinRuleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "RbinRuleTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RbinRule = RbinRule;
_g = JSII_RTTI_SYMBOL_1;
RbinRule[_g] = { fqn: "@cdktf/provider-aws.rbinRule.RbinRule", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RbinRule.tfResourceType = "aws_rbin_rule";
