// @cdktf/provider-aws/lib/medialive-multiplex/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MedialiveMultiplex = exports.MedialiveMultiplexTimeoutsOutputReference = exports.medialiveMultiplexTimeoutsToHclTerraform = exports.medialiveMultiplexTimeoutsToTerraform = exports.MedialiveMultiplexMultiplexSettingsOutputReference = exports.medialiveMultiplexMultiplexSettingsToHclTerraform = exports.medialiveMultiplexMultiplexSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function medialiveMultiplexMultiplexSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_video_buffer_delay_milliseconds: cdktf.numberToTerraform(struct.maximumVideoBufferDelayMilliseconds),
        transport_stream_bitrate: cdktf.numberToTerraform(struct.transportStreamBitrate),
        transport_stream_id: cdktf.numberToTerraform(struct.transportStreamId),
        transport_stream_reserved_bitrate: cdktf.numberToTerraform(struct.transportStreamReservedBitrate),
    };
}
exports.medialiveMultiplexMultiplexSettingsToTerraform = medialiveMultiplexMultiplexSettingsToTerraform;
function medialiveMultiplexMultiplexSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_video_buffer_delay_milliseconds: {
            value: cdktf.numberToHclTerraform(struct.maximumVideoBufferDelayMilliseconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        transport_stream_bitrate: {
            value: cdktf.numberToHclTerraform(struct.transportStreamBitrate),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        transport_stream_id: {
            value: cdktf.numberToHclTerraform(struct.transportStreamId),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        transport_stream_reserved_bitrate: {
            value: cdktf.numberToHclTerraform(struct.transportStreamReservedBitrate),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.medialiveMultiplexMultiplexSettingsToHclTerraform = medialiveMultiplexMultiplexSettingsToHclTerraform;
class MedialiveMultiplexMultiplexSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumVideoBufferDelayMilliseconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumVideoBufferDelayMilliseconds = this._maximumVideoBufferDelayMilliseconds;
        }
        if (this._transportStreamBitrate !== undefined) {
            hasAnyValues = true;
            internalValueResult.transportStreamBitrate = this._transportStreamBitrate;
        }
        if (this._transportStreamId !== undefined) {
            hasAnyValues = true;
            internalValueResult.transportStreamId = this._transportStreamId;
        }
        if (this._transportStreamReservedBitrate !== undefined) {
            hasAnyValues = true;
            internalValueResult.transportStreamReservedBitrate = this._transportStreamReservedBitrate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumVideoBufferDelayMilliseconds = undefined;
            this._transportStreamBitrate = undefined;
            this._transportStreamId = undefined;
            this._transportStreamReservedBitrate = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumVideoBufferDelayMilliseconds = value.maximumVideoBufferDelayMilliseconds;
            this._transportStreamBitrate = value.transportStreamBitrate;
            this._transportStreamId = value.transportStreamId;
            this._transportStreamReservedBitrate = value.transportStreamReservedBitrate;
        }
    }
    get maximumVideoBufferDelayMilliseconds() {
        return this.getNumberAttribute('maximum_video_buffer_delay_milliseconds');
    }
    set maximumVideoBufferDelayMilliseconds(value) {
        this._maximumVideoBufferDelayMilliseconds = value;
    }
    resetMaximumVideoBufferDelayMilliseconds() {
        this._maximumVideoBufferDelayMilliseconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumVideoBufferDelayMillisecondsInput() {
        return this._maximumVideoBufferDelayMilliseconds;
    }
    get transportStreamBitrate() {
        return this.getNumberAttribute('transport_stream_bitrate');
    }
    set transportStreamBitrate(value) {
        this._transportStreamBitrate = value;
    }
    // Temporarily expose input value. Use with caution.
    get transportStreamBitrateInput() {
        return this._transportStreamBitrate;
    }
    get transportStreamId() {
        return this.getNumberAttribute('transport_stream_id');
    }
    set transportStreamId(value) {
        this._transportStreamId = value;
    }
    // Temporarily expose input value. Use with caution.
    get transportStreamIdInput() {
        return this._transportStreamId;
    }
    get transportStreamReservedBitrate() {
        return this.getNumberAttribute('transport_stream_reserved_bitrate');
    }
    set transportStreamReservedBitrate(value) {
        this._transportStreamReservedBitrate = value;
    }
    resetTransportStreamReservedBitrate() {
        this._transportStreamReservedBitrate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transportStreamReservedBitrateInput() {
        return this._transportStreamReservedBitrate;
    }
}
exports.MedialiveMultiplexMultiplexSettingsOutputReference = MedialiveMultiplexMultiplexSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MedialiveMultiplexMultiplexSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.medialiveMultiplex.MedialiveMultiplexMultiplexSettingsOutputReference", version: "19.15.0" };
function medialiveMultiplexTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.medialiveMultiplexTimeoutsToTerraform = medialiveMultiplexTimeoutsToTerraform;
function medialiveMultiplexTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.medialiveMultiplexTimeoutsToHclTerraform = medialiveMultiplexTimeoutsToHclTerraform;
class MedialiveMultiplexTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MedialiveMultiplexTimeoutsOutputReference = MedialiveMultiplexTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MedialiveMultiplexTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.medialiveMultiplex.MedialiveMultiplexTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/medialive_multiplex aws_medialive_multiplex}
*/
class MedialiveMultiplex extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a MedialiveMultiplex resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the MedialiveMultiplex to import
    * @param importFromId The id of the existing MedialiveMultiplex that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/medialive_multiplex#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the MedialiveMultiplex to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_medialive_multiplex", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/medialive_multiplex aws_medialive_multiplex} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MedialiveMultiplexConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_medialive_multiplex',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // multiplex_settings - computed: false, optional: true, required: false
        this._multiplexSettings = new MedialiveMultiplexMultiplexSettingsOutputReference(this, "multiplex_settings");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MedialiveMultiplexTimeoutsOutputReference(this, "timeouts");
        this._availabilityZones = config.availabilityZones;
        this._id = config.id;
        this._name = config.name;
        this._startMultiplex = config.startMultiplex;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._multiplexSettings.internalValue = config.multiplexSettings;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get availabilityZones() {
        return this.getListAttribute('availability_zones');
    }
    set availabilityZones(value) {
        this._availabilityZones = value;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZonesInput() {
        return this._availabilityZones;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get startMultiplex() {
        return this.getBooleanAttribute('start_multiplex');
    }
    set startMultiplex(value) {
        this._startMultiplex = value;
    }
    resetStartMultiplex() {
        this._startMultiplex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startMultiplexInput() {
        return this._startMultiplex;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get multiplexSettings() {
        return this._multiplexSettings;
    }
    putMultiplexSettings(value) {
        this._multiplexSettings.internalValue = value;
    }
    resetMultiplexSettings() {
        this._multiplexSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multiplexSettingsInput() {
        return this._multiplexSettings.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            availability_zones: cdktf.listMapper(cdktf.stringToTerraform, false)(this._availabilityZones),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            start_multiplex: cdktf.booleanToTerraform(this._startMultiplex),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            multiplex_settings: medialiveMultiplexMultiplexSettingsToTerraform(this._multiplexSettings.internalValue),
            timeouts: medialiveMultiplexTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            availability_zones: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._availabilityZones),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            start_multiplex: {
                value: cdktf.booleanToHclTerraform(this._startMultiplex),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            multiplex_settings: {
                value: medialiveMultiplexMultiplexSettingsToHclTerraform(this._multiplexSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MedialiveMultiplexMultiplexSettingsList",
            },
            timeouts: {
                value: medialiveMultiplexTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "MedialiveMultiplexTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.MedialiveMultiplex = MedialiveMultiplex;
_c = JSII_RTTI_SYMBOL_1;
MedialiveMultiplex[_c] = { fqn: "@cdktf/provider-aws.medialiveMultiplex.MedialiveMultiplex", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
MedialiveMultiplex.tfResourceType = "aws_medialive_multiplex";
