// @winglang/sdk/lib/target-tf-gcp/schedule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedule = void 0;
const app_1 = require("./app");
const function_1 = require("./function");
const cloud_scheduler_job_1 = require("../.gen/providers/google/cloud-scheduler-job");
const service_account_1 = require("../.gen/providers/google/service-account");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
const std_1 = require("../std");
/**
 * GCP implementation of `cloud.Schedule`.
 *
 * @inflight `@winglang/sdk.cloud.IScheduleClient`
 */
class Schedule extends cloud.Schedule {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        // private readonly scheduler: CloudSchedulerJob;
        this.handlers = {};
        const { rate, cron } = props;
        /*
         * A schedule is defined using the unix-cron string format (* * * * *) which
         * is a set of five fields in a line, indicating when the job should be executed.
         * https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules
         */
        this.scheduleExpression = (rate ? `*/${rate.minutes} * * * *` : cron);
    }
    onTick(inflight, props = {}) {
        const uniqueId = this.node.addr.substring(0, 8);
        const functionHandler = core.lift({ inflight }).inflight(async (ctx) => {
            await ctx.inflight();
        });
        let cronFunction = this.handlers[inflight._id];
        if (cronFunction) {
            return cronFunction;
        }
        cronFunction = new function_1.Function(this, app_1.App.of(this).makeId(this, "OnTick"), functionHandler, props);
        this.handlers[inflight._id] = cronFunction;
        // create scheduler service account
        const schedulerServiceAccount = new service_account_1.ServiceAccount(this, "SchedulerServiceAccount", {
            accountId: `scheduler-${uniqueId}-sa`,
            displayName: `Service Account for scheduler-${uniqueId}`,
        });
        // allow scheduler service account to invoke cron function
        cronFunction._addPermissionToInvoke(schedulerServiceAccount);
        // create scheduler
        new cloud_scheduler_job_1.CloudSchedulerJob(this, "Scheduler", {
            name: `scheduler-${uniqueId}`,
            description: `Trigger ${cronFunction.functionName}`,
            schedule: this.scheduleExpression,
            timeZone: "Etc/UTC",
            attemptDeadline: "300s",
            httpTarget: {
                httpMethod: "GET",
                uri: cronFunction._getHttpsTriggerUrl(),
                oidcToken: {
                    serviceAccountEmail: schedulerServiceAccount.email,
                },
            },
        });
        std_1.Node.of(this).addConnection({
            source: this,
            sourceOp: cloud.ScheduleInflightMethods.TICK,
            target: cronFunction,
            targetOp: cloud.FunctionInflightMethods.INVOKE,
            name: "onTick()",
        });
        return cronFunction;
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-gcp", "shared-gcp"), __filename, "ScheduleClient", [`process.env["${this.envName()}"]`]);
    }
    envName() {
        return `SCHEDULE_EVENT_${this.node.addr.slice(-8)}`;
    }
}
exports.Schedule = Schedule;
