// @winglang/sdk/lib/target-tf-gcp/test-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunner = void 0;
const terraform_output_1 = require("cdktf/lib/terraform-output");
const tokens_1 = require("cdktf/lib/tokens");
const function_1 = require("./function");
const core = __importStar(require("../core"));
const std = __importStar(require("../std"));
const OUTPUT_TEST_RUNNER_FUNCTION_IDENTIFIERS = "WING_TEST_RUNNER_FUNCTION_IDENTIFIERS";
/**
 * Tf-Azure implementation of `cloud.TestRunner`.
 *
 * @inflight `@winglang/sdk.cloud.ITestRunnerClient`
 */
class TestRunner extends std.TestRunner {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        // This output is created so the CLI's `wing test` command can obtain a list
        // of all names of test functions by running `terraform output`.
        const output = new terraform_output_1.TerraformOutput(this, "TestFunctionIdentifiers", {
            value: tokens_1.Lazy.stringValue({
                produce: () => {
                    return JSON.stringify([
                        ...this.getTestFunctionIdentifiers().entries(),
                    ]);
                },
            }),
        });
        output.overrideLogicalId(OUTPUT_TEST_RUNNER_FUNCTION_IDENTIFIERS);
    }
    onLift(host, ops) {
        if (!(host instanceof function_1.Function)) {
            throw new Error("TestRunner can only be bound by tfgcp.Function for now");
        }
        const { region, projectId } = core.App.of(this);
        const testFunctions = this.getTestFunctionIdentifiers();
        host.addEnvironment(this.envTestFunctionIdentifiers(), JSON.stringify([...testFunctions.entries()]));
        host.addEnvironment(this.projectEnv(), projectId);
        host.addEnvironment(this.regionEnv(), region);
        super.onLift(host, ops);
    }
    /** @internal */
    _preSynthesize() {
        // add a dependency on each test function
        for (const test of this.findTests()) {
            if (test._fn) {
                this.node.addDependency(test._fn);
            }
        }
        super._preSynthesize();
    }
    getTestFunctionIdentifiers() {
        const arns = new Map();
        for (const test of this.findTests()) {
            if (test._fn) {
                if (!(test._fn instanceof function_1.Function)) {
                    throw new Error(`Unsupported test function type, ${test._fn.node.path} was not a tfgcp.Function`);
                }
                arns.set(test.node.path, test._fn.functionName);
            }
        }
        return arns;
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname, __filename, "TestRunnerClient", [
            `process.env["${this.envTestFunctionIdentifiers()}"]`,
            `process.env["${this.projectEnv()}"]`,
            `process.env["${this.regionEnv()}"]`,
        ]);
    }
    envTestFunctionIdentifiers() {
        return `TEST_RUNNER_FUNCTIONS_${this.node.addr.slice(-8)}`;
    }
    regionEnv() {
        return `REGION_${this.node.addr.slice(-8)}`;
    }
    projectEnv() {
        return `PROJECT_${this.node.addr.slice(-8)}`;
    }
}
exports.TestRunner = TestRunner;
