// @cdktf/provider-aws/lib/data-aws-devopsguru-notification-channel/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDevopsguruNotificationChannel = exports.DataAwsDevopsguruNotificationChannelSnsList = exports.DataAwsDevopsguruNotificationChannelSnsOutputReference = exports.dataAwsDevopsguruNotificationChannelSnsToHclTerraform = exports.dataAwsDevopsguruNotificationChannelSnsToTerraform = exports.DataAwsDevopsguruNotificationChannelFiltersList = exports.DataAwsDevopsguruNotificationChannelFiltersOutputReference = exports.dataAwsDevopsguruNotificationChannelFiltersToHclTerraform = exports.dataAwsDevopsguruNotificationChannelFiltersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsDevopsguruNotificationChannelFiltersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDevopsguruNotificationChannelFiltersToTerraform = dataAwsDevopsguruNotificationChannelFiltersToTerraform;
function dataAwsDevopsguruNotificationChannelFiltersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDevopsguruNotificationChannelFiltersToHclTerraform = dataAwsDevopsguruNotificationChannelFiltersToHclTerraform;
class DataAwsDevopsguruNotificationChannelFiltersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // message_types - computed: true, optional: false, required: false
    get messageTypes() {
        return this.getListAttribute('message_types');
    }
    // severities - computed: true, optional: false, required: false
    get severities() {
        return this.getListAttribute('severities');
    }
}
exports.DataAwsDevopsguruNotificationChannelFiltersOutputReference = DataAwsDevopsguruNotificationChannelFiltersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruNotificationChannelFiltersOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruNotificationChannel.DataAwsDevopsguruNotificationChannelFiltersOutputReference", version: "19.15.0" };
class DataAwsDevopsguruNotificationChannelFiltersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDevopsguruNotificationChannelFiltersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDevopsguruNotificationChannelFiltersList = DataAwsDevopsguruNotificationChannelFiltersList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruNotificationChannelFiltersList[_b] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruNotificationChannel.DataAwsDevopsguruNotificationChannelFiltersList", version: "19.15.0" };
function dataAwsDevopsguruNotificationChannelSnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDevopsguruNotificationChannelSnsToTerraform = dataAwsDevopsguruNotificationChannelSnsToTerraform;
function dataAwsDevopsguruNotificationChannelSnsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDevopsguruNotificationChannelSnsToHclTerraform = dataAwsDevopsguruNotificationChannelSnsToHclTerraform;
class DataAwsDevopsguruNotificationChannelSnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // topic_arn - computed: true, optional: false, required: false
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
}
exports.DataAwsDevopsguruNotificationChannelSnsOutputReference = DataAwsDevopsguruNotificationChannelSnsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruNotificationChannelSnsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruNotificationChannel.DataAwsDevopsguruNotificationChannelSnsOutputReference", version: "19.15.0" };
class DataAwsDevopsguruNotificationChannelSnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDevopsguruNotificationChannelSnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDevopsguruNotificationChannelSnsList = DataAwsDevopsguruNotificationChannelSnsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruNotificationChannelSnsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruNotificationChannel.DataAwsDevopsguruNotificationChannelSnsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_notification_channel aws_devopsguru_notification_channel}
*/
class DataAwsDevopsguruNotificationChannel extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDevopsguruNotificationChannel resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDevopsguruNotificationChannel to import
    * @param importFromId The id of the existing DataAwsDevopsguruNotificationChannel that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_notification_channel#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDevopsguruNotificationChannel to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_devopsguru_notification_channel", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_notification_channel aws_devopsguru_notification_channel} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDevopsguruNotificationChannelConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_devopsguru_notification_channel',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filters - computed: false, optional: true, required: false
        this._filters = new DataAwsDevopsguruNotificationChannelFiltersList(this, "filters", false);
        // sns - computed: false, optional: true, required: false
        this._sns = new DataAwsDevopsguruNotificationChannelSnsList(this, "sns", false);
        this._id = config.id;
        this._filters.internalValue = config.filters;
        this._sns.internalValue = config.sns;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get filters() {
        return this._filters;
    }
    putFilters(value) {
        this._filters.internalValue = value;
    }
    resetFilters() {
        this._filters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filtersInput() {
        return this._filters.internalValue;
    }
    get sns() {
        return this._sns;
    }
    putSns(value) {
        this._sns.internalValue = value;
    }
    resetSns() {
        this._sns.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snsInput() {
        return this._sns.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            filters: cdktf.listMapper(dataAwsDevopsguruNotificationChannelFiltersToTerraform, true)(this._filters.internalValue),
            sns: cdktf.listMapper(dataAwsDevopsguruNotificationChannelSnsToTerraform, true)(this._sns.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filters: {
                value: cdktf.listMapperHcl(dataAwsDevopsguruNotificationChannelFiltersToHclTerraform, true)(this._filters.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsDevopsguruNotificationChannelFiltersList",
            },
            sns: {
                value: cdktf.listMapperHcl(dataAwsDevopsguruNotificationChannelSnsToHclTerraform, true)(this._sns.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsDevopsguruNotificationChannelSnsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDevopsguruNotificationChannel = DataAwsDevopsguruNotificationChannel;
_e = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruNotificationChannel[_e] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruNotificationChannel.DataAwsDevopsguruNotificationChannel", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDevopsguruNotificationChannel.tfResourceType = "aws_devopsguru_notification_channel";
