// @cdktf/provider-aws/lib/lightsail-disk-attachment/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailDiskAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_disk_attachment aws_lightsail_disk_attachment}
*/
class LightsailDiskAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LightsailDiskAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LightsailDiskAttachment to import
    * @param importFromId The id of the existing LightsailDiskAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_disk_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LightsailDiskAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lightsail_disk_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_disk_attachment aws_lightsail_disk_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LightsailDiskAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_disk_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._diskName = config.diskName;
        this._diskPath = config.diskPath;
        this._id = config.id;
        this._instanceName = config.instanceName;
    }
    get diskName() {
        return this.getStringAttribute('disk_name');
    }
    set diskName(value) {
        this._diskName = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskNameInput() {
        return this._diskName;
    }
    get diskPath() {
        return this.getStringAttribute('disk_path');
    }
    set diskPath(value) {
        this._diskPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskPathInput() {
        return this._diskPath;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceName() {
        return this.getStringAttribute('instance_name');
    }
    set instanceName(value) {
        this._instanceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceNameInput() {
        return this._instanceName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            disk_name: cdktf.stringToTerraform(this._diskName),
            disk_path: cdktf.stringToTerraform(this._diskPath),
            id: cdktf.stringToTerraform(this._id),
            instance_name: cdktf.stringToTerraform(this._instanceName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            disk_name: {
                value: cdktf.stringToHclTerraform(this._diskName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disk_path: {
                value: cdktf.stringToHclTerraform(this._diskPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_name: {
                value: cdktf.stringToHclTerraform(this._instanceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LightsailDiskAttachment = LightsailDiskAttachment;
_a = JSII_RTTI_SYMBOL_1;
LightsailDiskAttachment[_a] = { fqn: "@cdktf/provider-aws.lightsailDiskAttachment.LightsailDiskAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LightsailDiskAttachment.tfResourceType = "aws_lightsail_disk_attachment";
