// @cdktf/provider-aws/lib/data-aws-connect-hours-of-operation/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectHoursOfOperation = exports.DataAwsConnectHoursOfOperationConfigAList = exports.DataAwsConnectHoursOfOperationConfigAOutputReference = exports.dataAwsConnectHoursOfOperationConfigAToHclTerraform = exports.dataAwsConnectHoursOfOperationConfigAToTerraform = exports.DataAwsConnectHoursOfOperationConfigStartTimeList = exports.DataAwsConnectHoursOfOperationConfigStartTimeOutputReference = exports.dataAwsConnectHoursOfOperationConfigStartTimeToHclTerraform = exports.dataAwsConnectHoursOfOperationConfigStartTimeToTerraform = exports.DataAwsConnectHoursOfOperationConfigEndTimeList = exports.DataAwsConnectHoursOfOperationConfigEndTimeOutputReference = exports.dataAwsConnectHoursOfOperationConfigEndTimeToHclTerraform = exports.dataAwsConnectHoursOfOperationConfigEndTimeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectHoursOfOperationConfigEndTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectHoursOfOperationConfigEndTimeToTerraform = dataAwsConnectHoursOfOperationConfigEndTimeToTerraform;
function dataAwsConnectHoursOfOperationConfigEndTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectHoursOfOperationConfigEndTimeToHclTerraform = dataAwsConnectHoursOfOperationConfigEndTimeToHclTerraform;
class DataAwsConnectHoursOfOperationConfigEndTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
}
exports.DataAwsConnectHoursOfOperationConfigEndTimeOutputReference = DataAwsConnectHoursOfOperationConfigEndTimeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigEndTimeOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigEndTimeOutputReference", version: "19.15.0" };
class DataAwsConnectHoursOfOperationConfigEndTimeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectHoursOfOperationConfigEndTimeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectHoursOfOperationConfigEndTimeList = DataAwsConnectHoursOfOperationConfigEndTimeList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigEndTimeList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigEndTimeList", version: "19.15.0" };
function dataAwsConnectHoursOfOperationConfigStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectHoursOfOperationConfigStartTimeToTerraform = dataAwsConnectHoursOfOperationConfigStartTimeToTerraform;
function dataAwsConnectHoursOfOperationConfigStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectHoursOfOperationConfigStartTimeToHclTerraform = dataAwsConnectHoursOfOperationConfigStartTimeToHclTerraform;
class DataAwsConnectHoursOfOperationConfigStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
}
exports.DataAwsConnectHoursOfOperationConfigStartTimeOutputReference = DataAwsConnectHoursOfOperationConfigStartTimeOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigStartTimeOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigStartTimeOutputReference", version: "19.15.0" };
class DataAwsConnectHoursOfOperationConfigStartTimeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectHoursOfOperationConfigStartTimeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectHoursOfOperationConfigStartTimeList = DataAwsConnectHoursOfOperationConfigStartTimeList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigStartTimeList[_d] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigStartTimeList", version: "19.15.0" };
function dataAwsConnectHoursOfOperationConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectHoursOfOperationConfigAToTerraform = dataAwsConnectHoursOfOperationConfigAToTerraform;
function dataAwsConnectHoursOfOperationConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectHoursOfOperationConfigAToHclTerraform = dataAwsConnectHoursOfOperationConfigAToHclTerraform;
class DataAwsConnectHoursOfOperationConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // end_time - computed: true, optional: false, required: false
        this._endTime = new DataAwsConnectHoursOfOperationConfigEndTimeList(this, "end_time", false);
        // start_time - computed: true, optional: false, required: false
        this._startTime = new DataAwsConnectHoursOfOperationConfigStartTimeList(this, "start_time", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getStringAttribute('day');
    }
    get endTime() {
        return this._endTime;
    }
    get startTime() {
        return this._startTime;
    }
}
exports.DataAwsConnectHoursOfOperationConfigAOutputReference = DataAwsConnectHoursOfOperationConfigAOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigAOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigAOutputReference", version: "19.15.0" };
class DataAwsConnectHoursOfOperationConfigAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectHoursOfOperationConfigAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectHoursOfOperationConfigAList = DataAwsConnectHoursOfOperationConfigAList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperationConfigAList[_f] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperationConfigAList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_hours_of_operation aws_connect_hours_of_operation}
*/
class DataAwsConnectHoursOfOperation extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsConnectHoursOfOperation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsConnectHoursOfOperation to import
    * @param importFromId The id of the existing DataAwsConnectHoursOfOperation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_hours_of_operation#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsConnectHoursOfOperation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_connect_hours_of_operation", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_hours_of_operation aws_connect_hours_of_operation} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectHoursOfOperationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_hours_of_operation',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: true, optional: false, required: false
        this._config = new DataAwsConnectHoursOfOperationConfigAList(this, "config", true);
        this._hoursOfOperationId = config.hoursOfOperationId;
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get config() {
        return this._config;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get hoursOfOperationId() {
        return this.getStringAttribute('hours_of_operation_id');
    }
    set hoursOfOperationId(value) {
        this._hoursOfOperationId = value;
    }
    resetHoursOfOperationId() {
        this._hoursOfOperationId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursOfOperationIdInput() {
        return this._hoursOfOperationId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // time_zone - computed: true, optional: false, required: false
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hours_of_operation_id: cdktf.stringToTerraform(this._hoursOfOperationId),
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            hours_of_operation_id: {
                value: cdktf.stringToHclTerraform(this._hoursOfOperationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsConnectHoursOfOperation = DataAwsConnectHoursOfOperation;
_g = JSII_RTTI_SYMBOL_1;
DataAwsConnectHoursOfOperation[_g] = { fqn: "@cdktf/provider-aws.dataAwsConnectHoursOfOperation.DataAwsConnectHoursOfOperation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectHoursOfOperation.tfResourceType = "aws_connect_hours_of_operation";
