// @cdktf/provider-aws/lib/data-aws-route53-resolver-firewall-rules/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRoute53ResolverFirewallRules = exports.DataAwsRoute53ResolverFirewallRulesFirewallRulesList = exports.DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference = exports.dataAwsRoute53ResolverFirewallRulesFirewallRulesToHclTerraform = exports.dataAwsRoute53ResolverFirewallRulesFirewallRulesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRoute53ResolverFirewallRulesFirewallRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsRoute53ResolverFirewallRulesFirewallRulesToTerraform = dataAwsRoute53ResolverFirewallRulesFirewallRulesToTerraform;
function dataAwsRoute53ResolverFirewallRulesFirewallRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsRoute53ResolverFirewallRulesFirewallRulesToHclTerraform = dataAwsRoute53ResolverFirewallRulesFirewallRulesToHclTerraform;
class DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // action - computed: true, optional: false, required: false
    get action() {
        return this.getStringAttribute('action');
    }
    // block_override_dns_type - computed: true, optional: false, required: false
    get blockOverrideDnsType() {
        return this.getStringAttribute('block_override_dns_type');
    }
    // block_override_domain - computed: true, optional: false, required: false
    get blockOverrideDomain() {
        return this.getStringAttribute('block_override_domain');
    }
    // block_override_ttl - computed: true, optional: false, required: false
    get blockOverrideTtl() {
        return this.getNumberAttribute('block_override_ttl');
    }
    // block_response - computed: true, optional: false, required: false
    get blockResponse() {
        return this.getStringAttribute('block_response');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // creator_request_id - computed: true, optional: false, required: false
    get creatorRequestId() {
        return this.getStringAttribute('creator_request_id');
    }
    // firewall_domain_list_id - computed: true, optional: false, required: false
    get firewallDomainListId() {
        return this.getStringAttribute('firewall_domain_list_id');
    }
    // firewall_rule_group_id - computed: true, optional: false, required: false
    get firewallRuleGroupId() {
        return this.getStringAttribute('firewall_rule_group_id');
    }
    // modification_time - computed: true, optional: false, required: false
    get modificationTime() {
        return this.getStringAttribute('modification_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
}
exports.DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference = DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsRoute53ResolverFirewallRules.DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference", version: "19.15.0" };
class DataAwsRoute53ResolverFirewallRulesFirewallRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53ResolverFirewallRulesFirewallRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53ResolverFirewallRulesFirewallRulesList = DataAwsRoute53ResolverFirewallRulesFirewallRulesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsRoute53ResolverFirewallRulesFirewallRulesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsRoute53ResolverFirewallRules.DataAwsRoute53ResolverFirewallRulesFirewallRulesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/route53_resolver_firewall_rules aws_route53_resolver_firewall_rules}
*/
class DataAwsRoute53ResolverFirewallRules extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsRoute53ResolverFirewallRules resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsRoute53ResolverFirewallRules to import
    * @param importFromId The id of the existing DataAwsRoute53ResolverFirewallRules that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/route53_resolver_firewall_rules#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsRoute53ResolverFirewallRules to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_route53_resolver_firewall_rules", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/route53_resolver_firewall_rules aws_route53_resolver_firewall_rules} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRoute53ResolverFirewallRulesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_resolver_firewall_rules',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // firewall_rules - computed: true, optional: false, required: false
        this._firewallRules = new DataAwsRoute53ResolverFirewallRulesFirewallRulesList(this, "firewall_rules", false);
        this._action = config.action;
        this._firewallRuleGroupId = config.firewallRuleGroupId;
        this._id = config.id;
        this._priority = config.priority;
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    resetAction() {
        this._action = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get firewallRuleGroupId() {
        return this.getStringAttribute('firewall_rule_group_id');
    }
    set firewallRuleGroupId(value) {
        this._firewallRuleGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get firewallRuleGroupIdInput() {
        return this._firewallRuleGroupId;
    }
    get firewallRules() {
        return this._firewallRules;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action: cdktf.stringToTerraform(this._action),
            firewall_rule_group_id: cdktf.stringToTerraform(this._firewallRuleGroupId),
            id: cdktf.stringToTerraform(this._id),
            priority: cdktf.numberToTerraform(this._priority),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action: {
                value: cdktf.stringToHclTerraform(this._action),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            firewall_rule_group_id: {
                value: cdktf.stringToHclTerraform(this._firewallRuleGroupId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            priority: {
                value: cdktf.numberToHclTerraform(this._priority),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsRoute53ResolverFirewallRules = DataAwsRoute53ResolverFirewallRules;
_c = JSII_RTTI_SYMBOL_1;
DataAwsRoute53ResolverFirewallRules[_c] = { fqn: "@cdktf/provider-aws.dataAwsRoute53ResolverFirewallRules.DataAwsRoute53ResolverFirewallRules", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRoute53ResolverFirewallRules.tfResourceType = "aws_route53_resolver_firewall_rules";
