// @cdktf/provider-aws/lib/s3control-access-grant/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3ControlAccessGrant = exports.S3ControlAccessGrantGranteeList = exports.S3ControlAccessGrantGranteeOutputReference = exports.s3ControlAccessGrantGranteeToHclTerraform = exports.s3ControlAccessGrantGranteeToTerraform = exports.S3ControlAccessGrantAccessGrantsLocationConfigurationList = exports.S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference = exports.s3ControlAccessGrantAccessGrantsLocationConfigurationToHclTerraform = exports.s3ControlAccessGrantAccessGrantsLocationConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3ControlAccessGrantAccessGrantsLocationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_sub_prefix: cdktf.stringToTerraform(struct.s3SubPrefix),
    };
}
exports.s3ControlAccessGrantAccessGrantsLocationConfigurationToTerraform = s3ControlAccessGrantAccessGrantsLocationConfigurationToTerraform;
function s3ControlAccessGrantAccessGrantsLocationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        s3_sub_prefix: {
            value: cdktf.stringToHclTerraform(struct.s3SubPrefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3ControlAccessGrantAccessGrantsLocationConfigurationToHclTerraform = s3ControlAccessGrantAccessGrantsLocationConfigurationToHclTerraform;
class S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3SubPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3SubPrefix = this._s3SubPrefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._s3SubPrefix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._s3SubPrefix = value.s3SubPrefix;
        }
    }
    get s3SubPrefix() {
        return this.getStringAttribute('s3_sub_prefix');
    }
    set s3SubPrefix(value) {
        this._s3SubPrefix = value;
    }
    resetS3SubPrefix() {
        this._s3SubPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3SubPrefixInput() {
        return this._s3SubPrefix;
    }
}
exports.S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference = S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3ControlAccessGrant.S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference", version: "19.15.0" };
class S3ControlAccessGrantAccessGrantsLocationConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3ControlAccessGrantAccessGrantsLocationConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3ControlAccessGrantAccessGrantsLocationConfigurationList = S3ControlAccessGrantAccessGrantsLocationConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
S3ControlAccessGrantAccessGrantsLocationConfigurationList[_b] = { fqn: "@cdktf/provider-aws.s3ControlAccessGrant.S3ControlAccessGrantAccessGrantsLocationConfigurationList", version: "19.15.0" };
function s3ControlAccessGrantGranteeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grantee_identifier: cdktf.stringToTerraform(struct.granteeIdentifier),
        grantee_type: cdktf.stringToTerraform(struct.granteeType),
    };
}
exports.s3ControlAccessGrantGranteeToTerraform = s3ControlAccessGrantGranteeToTerraform;
function s3ControlAccessGrantGranteeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        grantee_identifier: {
            value: cdktf.stringToHclTerraform(struct.granteeIdentifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        grantee_type: {
            value: cdktf.stringToHclTerraform(struct.granteeType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3ControlAccessGrantGranteeToHclTerraform = s3ControlAccessGrantGranteeToHclTerraform;
class S3ControlAccessGrantGranteeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._granteeIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.granteeIdentifier = this._granteeIdentifier;
        }
        if (this._granteeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.granteeType = this._granteeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._granteeIdentifier = undefined;
            this._granteeType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._granteeIdentifier = value.granteeIdentifier;
            this._granteeType = value.granteeType;
        }
    }
    get granteeIdentifier() {
        return this.getStringAttribute('grantee_identifier');
    }
    set granteeIdentifier(value) {
        this._granteeIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get granteeIdentifierInput() {
        return this._granteeIdentifier;
    }
    get granteeType() {
        return this.getStringAttribute('grantee_type');
    }
    set granteeType(value) {
        this._granteeType = value;
    }
    // Temporarily expose input value. Use with caution.
    get granteeTypeInput() {
        return this._granteeType;
    }
}
exports.S3ControlAccessGrantGranteeOutputReference = S3ControlAccessGrantGranteeOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3ControlAccessGrantGranteeOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3ControlAccessGrant.S3ControlAccessGrantGranteeOutputReference", version: "19.15.0" };
class S3ControlAccessGrantGranteeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3ControlAccessGrantGranteeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3ControlAccessGrantGranteeList = S3ControlAccessGrantGranteeList;
_d = JSII_RTTI_SYMBOL_1;
S3ControlAccessGrantGranteeList[_d] = { fqn: "@cdktf/provider-aws.s3ControlAccessGrant.S3ControlAccessGrantGranteeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3control_access_grant aws_s3control_access_grant}
*/
class S3ControlAccessGrant extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a S3ControlAccessGrant resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the S3ControlAccessGrant to import
    * @param importFromId The id of the existing S3ControlAccessGrant that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3control_access_grant#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the S3ControlAccessGrant to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_s3control_access_grant", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3control_access_grant aws_s3control_access_grant} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3ControlAccessGrantConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3control_access_grant',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // access_grants_location_configuration - computed: false, optional: true, required: false
        this._accessGrantsLocationConfiguration = new S3ControlAccessGrantAccessGrantsLocationConfigurationList(this, "access_grants_location_configuration", false);
        // grantee - computed: false, optional: true, required: false
        this._grantee = new S3ControlAccessGrantGranteeList(this, "grantee", false);
        this._accessGrantsLocationId = config.accessGrantsLocationId;
        this._accountId = config.accountId;
        this._permission = config.permission;
        this._s3PrefixType = config.s3PrefixType;
        this._tags = config.tags;
        this._accessGrantsLocationConfiguration.internalValue = config.accessGrantsLocationConfiguration;
        this._grantee.internalValue = config.grantee;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_grant_arn - computed: true, optional: false, required: false
    get accessGrantArn() {
        return this.getStringAttribute('access_grant_arn');
    }
    // access_grant_id - computed: true, optional: false, required: false
    get accessGrantId() {
        return this.getStringAttribute('access_grant_id');
    }
    get accessGrantsLocationId() {
        return this.getStringAttribute('access_grants_location_id');
    }
    set accessGrantsLocationId(value) {
        this._accessGrantsLocationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessGrantsLocationIdInput() {
        return this._accessGrantsLocationId;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    // grant_scope - computed: true, optional: false, required: false
    get grantScope() {
        return this.getStringAttribute('grant_scope');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get permission() {
        return this.getStringAttribute('permission');
    }
    set permission(value) {
        this._permission = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission;
    }
    get s3PrefixType() {
        return this.getStringAttribute('s3_prefix_type');
    }
    set s3PrefixType(value) {
        this._s3PrefixType = value;
    }
    resetS3PrefixType() {
        this._s3PrefixType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3PrefixTypeInput() {
        return this._s3PrefixType;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get accessGrantsLocationConfiguration() {
        return this._accessGrantsLocationConfiguration;
    }
    putAccessGrantsLocationConfiguration(value) {
        this._accessGrantsLocationConfiguration.internalValue = value;
    }
    resetAccessGrantsLocationConfiguration() {
        this._accessGrantsLocationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessGrantsLocationConfigurationInput() {
        return this._accessGrantsLocationConfiguration.internalValue;
    }
    get grantee() {
        return this._grantee;
    }
    putGrantee(value) {
        this._grantee.internalValue = value;
    }
    resetGrantee() {
        this._grantee.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeInput() {
        return this._grantee.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_grants_location_id: cdktf.stringToTerraform(this._accessGrantsLocationId),
            account_id: cdktf.stringToTerraform(this._accountId),
            permission: cdktf.stringToTerraform(this._permission),
            s3_prefix_type: cdktf.stringToTerraform(this._s3PrefixType),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            access_grants_location_configuration: cdktf.listMapper(s3ControlAccessGrantAccessGrantsLocationConfigurationToTerraform, true)(this._accessGrantsLocationConfiguration.internalValue),
            grantee: cdktf.listMapper(s3ControlAccessGrantGranteeToTerraform, true)(this._grantee.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_grants_location_id: {
                value: cdktf.stringToHclTerraform(this._accessGrantsLocationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            account_id: {
                value: cdktf.stringToHclTerraform(this._accountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permission: {
                value: cdktf.stringToHclTerraform(this._permission),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_prefix_type: {
                value: cdktf.stringToHclTerraform(this._s3PrefixType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            access_grants_location_configuration: {
                value: cdktf.listMapperHcl(s3ControlAccessGrantAccessGrantsLocationConfigurationToHclTerraform, true)(this._accessGrantsLocationConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3ControlAccessGrantAccessGrantsLocationConfigurationList",
            },
            grantee: {
                value: cdktf.listMapperHcl(s3ControlAccessGrantGranteeToHclTerraform, true)(this._grantee.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3ControlAccessGrantGranteeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.S3ControlAccessGrant = S3ControlAccessGrant;
_e = JSII_RTTI_SYMBOL_1;
S3ControlAccessGrant[_e] = { fqn: "@cdktf/provider-aws.s3ControlAccessGrant.S3ControlAccessGrant", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
S3ControlAccessGrant.tfResourceType = "aws_s3control_access_grant";
