// @cdktf/provider-aws/lib/verifiedpermissions-schema/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedpermissionsSchema = exports.VerifiedpermissionsSchemaDefinitionOutputReference = exports.verifiedpermissionsSchemaDefinitionToHclTerraform = exports.verifiedpermissionsSchemaDefinitionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function verifiedpermissionsSchemaDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.verifiedpermissionsSchemaDefinitionToTerraform = verifiedpermissionsSchemaDefinitionToTerraform;
function verifiedpermissionsSchemaDefinitionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedpermissionsSchemaDefinitionToHclTerraform = verifiedpermissionsSchemaDefinitionToHclTerraform;
class VerifiedpermissionsSchemaDefinitionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._value = value.value;
        }
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.VerifiedpermissionsSchemaDefinitionOutputReference = VerifiedpermissionsSchemaDefinitionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VerifiedpermissionsSchemaDefinitionOutputReference[_a] = { fqn: "@cdktf/provider-aws.verifiedpermissionsSchema.VerifiedpermissionsSchemaDefinitionOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_schema aws_verifiedpermissions_schema}
*/
class VerifiedpermissionsSchema extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedpermissionsSchema resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedpermissionsSchema to import
    * @param importFromId The id of the existing VerifiedpermissionsSchema that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_schema#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedpermissionsSchema to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedpermissions_schema", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_schema aws_verifiedpermissions_schema} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedpermissionsSchemaConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedpermissions_schema',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // definition - computed: false, optional: false, required: true
        this._definition = new VerifiedpermissionsSchemaDefinitionOutputReference(this, "definition");
        this._policyStoreId = config.policyStoreId;
        this._definition.internalValue = config.definition;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // namespaces - computed: true, optional: false, required: false
    get namespaces() {
        return cdktf.Fn.tolist(this.getListAttribute('namespaces'));
    }
    get policyStoreId() {
        return this.getStringAttribute('policy_store_id');
    }
    set policyStoreId(value) {
        this._policyStoreId = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyStoreIdInput() {
        return this._policyStoreId;
    }
    get definition() {
        return this._definition;
    }
    putDefinition(value) {
        this._definition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get definitionInput() {
        return this._definition.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            policy_store_id: cdktf.stringToTerraform(this._policyStoreId),
            definition: verifiedpermissionsSchemaDefinitionToTerraform(this._definition.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            policy_store_id: {
                value: cdktf.stringToHclTerraform(this._policyStoreId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            definition: {
                value: verifiedpermissionsSchemaDefinitionToHclTerraform(this._definition.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VerifiedpermissionsSchemaDefinition",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedpermissionsSchema = VerifiedpermissionsSchema;
_b = JSII_RTTI_SYMBOL_1;
VerifiedpermissionsSchema[_b] = { fqn: "@cdktf/provider-aws.verifiedpermissionsSchema.VerifiedpermissionsSchema", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedpermissionsSchema.tfResourceType = "aws_verifiedpermissions_schema";
