// @cdktf/provider-aws/lib/elasticsearch-domain-saml-options/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticsearchDomainSamlOptions = exports.ElasticsearchDomainSamlOptionsTimeoutsOutputReference = exports.elasticsearchDomainSamlOptionsTimeoutsToHclTerraform = exports.elasticsearchDomainSamlOptionsTimeoutsToTerraform = exports.ElasticsearchDomainSamlOptionsSamlOptionsOutputReference = exports.elasticsearchDomainSamlOptionsSamlOptionsToHclTerraform = exports.elasticsearchDomainSamlOptionsSamlOptionsToTerraform = exports.ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference = exports.elasticsearchDomainSamlOptionsSamlOptionsIdpToHclTerraform = exports.elasticsearchDomainSamlOptionsSamlOptionsIdpToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elasticsearchDomainSamlOptionsSamlOptionsIdpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        entity_id: cdktf.stringToTerraform(struct.entityId),
        metadata_content: cdktf.stringToTerraform(struct.metadataContent),
    };
}
exports.elasticsearchDomainSamlOptionsSamlOptionsIdpToTerraform = elasticsearchDomainSamlOptionsSamlOptionsIdpToTerraform;
function elasticsearchDomainSamlOptionsSamlOptionsIdpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        entity_id: {
            value: cdktf.stringToHclTerraform(struct.entityId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        metadata_content: {
            value: cdktf.stringToHclTerraform(struct.metadataContent),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainSamlOptionsSamlOptionsIdpToHclTerraform = elasticsearchDomainSamlOptionsSamlOptionsIdpToHclTerraform;
class ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._entityId !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityId = this._entityId;
        }
        if (this._metadataContent !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadataContent = this._metadataContent;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._entityId = undefined;
            this._metadataContent = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._entityId = value.entityId;
            this._metadataContent = value.metadataContent;
        }
    }
    get entityId() {
        return this.getStringAttribute('entity_id');
    }
    set entityId(value) {
        this._entityId = value;
    }
    // Temporarily expose input value. Use with caution.
    get entityIdInput() {
        return this._entityId;
    }
    get metadataContent() {
        return this.getStringAttribute('metadata_content');
    }
    set metadataContent(value) {
        this._metadataContent = value;
    }
    // Temporarily expose input value. Use with caution.
    get metadataContentInput() {
        return this._metadataContent;
    }
}
exports.ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference = ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference[_a] = { fqn: "@cdktf/provider-aws.elasticsearchDomainSamlOptions.ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference", version: "19.15.0" };
function elasticsearchDomainSamlOptionsSamlOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        master_backend_role: cdktf.stringToTerraform(struct.masterBackendRole),
        master_user_name: cdktf.stringToTerraform(struct.masterUserName),
        roles_key: cdktf.stringToTerraform(struct.rolesKey),
        session_timeout_minutes: cdktf.numberToTerraform(struct.sessionTimeoutMinutes),
        subject_key: cdktf.stringToTerraform(struct.subjectKey),
        idp: elasticsearchDomainSamlOptionsSamlOptionsIdpToTerraform(struct.idp),
    };
}
exports.elasticsearchDomainSamlOptionsSamlOptionsToTerraform = elasticsearchDomainSamlOptionsSamlOptionsToTerraform;
function elasticsearchDomainSamlOptionsSamlOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        master_backend_role: {
            value: cdktf.stringToHclTerraform(struct.masterBackendRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        master_user_name: {
            value: cdktf.stringToHclTerraform(struct.masterUserName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        roles_key: {
            value: cdktf.stringToHclTerraform(struct.rolesKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        session_timeout_minutes: {
            value: cdktf.numberToHclTerraform(struct.sessionTimeoutMinutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        subject_key: {
            value: cdktf.stringToHclTerraform(struct.subjectKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        idp: {
            value: elasticsearchDomainSamlOptionsSamlOptionsIdpToHclTerraform(struct.idp),
            isBlock: true,
            type: "list",
            storageClassType: "ElasticsearchDomainSamlOptionsSamlOptionsIdpList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainSamlOptionsSamlOptionsToHclTerraform = elasticsearchDomainSamlOptionsSamlOptionsToHclTerraform;
class ElasticsearchDomainSamlOptionsSamlOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // idp - computed: false, optional: true, required: false
        this._idp = new ElasticsearchDomainSamlOptionsSamlOptionsIdpOutputReference(this, "idp");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._masterBackendRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterBackendRole = this._masterBackendRole;
        }
        if (this._masterUserName !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserName = this._masterUserName;
        }
        if (this._rolesKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.rolesKey = this._rolesKey;
        }
        if (this._sessionTimeoutMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionTimeoutMinutes = this._sessionTimeoutMinutes;
        }
        if (this._subjectKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.subjectKey = this._subjectKey;
        }
        if (this._idp?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.idp = this._idp?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._masterBackendRole = undefined;
            this._masterUserName = undefined;
            this._rolesKey = undefined;
            this._sessionTimeoutMinutes = undefined;
            this._subjectKey = undefined;
            this._idp.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._masterBackendRole = value.masterBackendRole;
            this._masterUserName = value.masterUserName;
            this._rolesKey = value.rolesKey;
            this._sessionTimeoutMinutes = value.sessionTimeoutMinutes;
            this._subjectKey = value.subjectKey;
            this._idp.internalValue = value.idp;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get masterBackendRole() {
        return this.getStringAttribute('master_backend_role');
    }
    set masterBackendRole(value) {
        this._masterBackendRole = value;
    }
    resetMasterBackendRole() {
        this._masterBackendRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterBackendRoleInput() {
        return this._masterBackendRole;
    }
    get masterUserName() {
        return this.getStringAttribute('master_user_name');
    }
    set masterUserName(value) {
        this._masterUserName = value;
    }
    resetMasterUserName() {
        this._masterUserName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserNameInput() {
        return this._masterUserName;
    }
    get rolesKey() {
        return this.getStringAttribute('roles_key');
    }
    set rolesKey(value) {
        this._rolesKey = value;
    }
    resetRolesKey() {
        this._rolesKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rolesKeyInput() {
        return this._rolesKey;
    }
    get sessionTimeoutMinutes() {
        return this.getNumberAttribute('session_timeout_minutes');
    }
    set sessionTimeoutMinutes(value) {
        this._sessionTimeoutMinutes = value;
    }
    resetSessionTimeoutMinutes() {
        this._sessionTimeoutMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionTimeoutMinutesInput() {
        return this._sessionTimeoutMinutes;
    }
    get subjectKey() {
        return this.getStringAttribute('subject_key');
    }
    set subjectKey(value) {
        this._subjectKey = value;
    }
    resetSubjectKey() {
        this._subjectKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subjectKeyInput() {
        return this._subjectKey;
    }
    get idp() {
        return this._idp;
    }
    putIdp(value) {
        this._idp.internalValue = value;
    }
    resetIdp() {
        this._idp.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idpInput() {
        return this._idp.internalValue;
    }
}
exports.ElasticsearchDomainSamlOptionsSamlOptionsOutputReference = ElasticsearchDomainSamlOptionsSamlOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainSamlOptionsSamlOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.elasticsearchDomainSamlOptions.ElasticsearchDomainSamlOptionsSamlOptionsOutputReference", version: "19.15.0" };
function elasticsearchDomainSamlOptionsTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.elasticsearchDomainSamlOptionsTimeoutsToTerraform = elasticsearchDomainSamlOptionsTimeoutsToTerraform;
function elasticsearchDomainSamlOptionsTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticsearchDomainSamlOptionsTimeoutsToHclTerraform = elasticsearchDomainSamlOptionsTimeoutsToHclTerraform;
class ElasticsearchDomainSamlOptionsTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ElasticsearchDomainSamlOptionsTimeoutsOutputReference = ElasticsearchDomainSamlOptionsTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainSamlOptionsTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.elasticsearchDomainSamlOptions.ElasticsearchDomainSamlOptionsTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain_saml_options aws_elasticsearch_domain_saml_options}
*/
class ElasticsearchDomainSamlOptions extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ElasticsearchDomainSamlOptions resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ElasticsearchDomainSamlOptions to import
    * @param importFromId The id of the existing ElasticsearchDomainSamlOptions that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain_saml_options#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ElasticsearchDomainSamlOptions to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_elasticsearch_domain_saml_options", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elasticsearch_domain_saml_options aws_elasticsearch_domain_saml_options} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticsearchDomainSamlOptionsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticsearch_domain_saml_options',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // saml_options - computed: false, optional: true, required: false
        this._samlOptions = new ElasticsearchDomainSamlOptionsSamlOptionsOutputReference(this, "saml_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ElasticsearchDomainSamlOptionsTimeoutsOutputReference(this, "timeouts");
        this._domainName = config.domainName;
        this._id = config.id;
        this._samlOptions.internalValue = config.samlOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get samlOptions() {
        return this._samlOptions;
    }
    putSamlOptions(value) {
        this._samlOptions.internalValue = value;
    }
    resetSamlOptions() {
        this._samlOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get samlOptionsInput() {
        return this._samlOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            id: cdktf.stringToTerraform(this._id),
            saml_options: elasticsearchDomainSamlOptionsSamlOptionsToTerraform(this._samlOptions.internalValue),
            timeouts: elasticsearchDomainSamlOptionsTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            saml_options: {
                value: elasticsearchDomainSamlOptionsSamlOptionsToHclTerraform(this._samlOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticsearchDomainSamlOptionsSamlOptionsList",
            },
            timeouts: {
                value: elasticsearchDomainSamlOptionsTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ElasticsearchDomainSamlOptionsTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ElasticsearchDomainSamlOptions = ElasticsearchDomainSamlOptions;
_d = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainSamlOptions[_d] = { fqn: "@cdktf/provider-aws.elasticsearchDomainSamlOptions.ElasticsearchDomainSamlOptions", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ElasticsearchDomainSamlOptions.tfResourceType = "aws_elasticsearch_domain_saml_options";
