// @cdktf/provider-aws/lib/data-aws-connect-routing-profile/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectRoutingProfile = exports.DataAwsConnectRoutingProfileQueueConfigsList = exports.DataAwsConnectRoutingProfileQueueConfigsOutputReference = exports.dataAwsConnectRoutingProfileQueueConfigsToHclTerraform = exports.dataAwsConnectRoutingProfileQueueConfigsToTerraform = exports.DataAwsConnectRoutingProfileMediaConcurrenciesList = exports.DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference = exports.dataAwsConnectRoutingProfileMediaConcurrenciesToHclTerraform = exports.dataAwsConnectRoutingProfileMediaConcurrenciesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectRoutingProfileMediaConcurrenciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectRoutingProfileMediaConcurrenciesToTerraform = dataAwsConnectRoutingProfileMediaConcurrenciesToTerraform;
function dataAwsConnectRoutingProfileMediaConcurrenciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectRoutingProfileMediaConcurrenciesToHclTerraform = dataAwsConnectRoutingProfileMediaConcurrenciesToHclTerraform;
class DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // channel - computed: true, optional: false, required: false
    get channel() {
        return this.getStringAttribute('channel');
    }
    // concurrency - computed: true, optional: false, required: false
    get concurrency() {
        return this.getNumberAttribute('concurrency');
    }
}
exports.DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference = DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectRoutingProfile.DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference", version: "19.15.0" };
class DataAwsConnectRoutingProfileMediaConcurrenciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectRoutingProfileMediaConcurrenciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectRoutingProfileMediaConcurrenciesList = DataAwsConnectRoutingProfileMediaConcurrenciesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectRoutingProfileMediaConcurrenciesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectRoutingProfile.DataAwsConnectRoutingProfileMediaConcurrenciesList", version: "19.15.0" };
function dataAwsConnectRoutingProfileQueueConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectRoutingProfileQueueConfigsToTerraform = dataAwsConnectRoutingProfileQueueConfigsToTerraform;
function dataAwsConnectRoutingProfileQueueConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectRoutingProfileQueueConfigsToHclTerraform = dataAwsConnectRoutingProfileQueueConfigsToHclTerraform;
class DataAwsConnectRoutingProfileQueueConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // channel - computed: true, optional: false, required: false
    get channel() {
        return this.getStringAttribute('channel');
    }
    // delay - computed: true, optional: false, required: false
    get delay() {
        return this.getNumberAttribute('delay');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // queue_arn - computed: true, optional: false, required: false
    get queueArn() {
        return this.getStringAttribute('queue_arn');
    }
    // queue_id - computed: true, optional: false, required: false
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
    // queue_name - computed: true, optional: false, required: false
    get queueName() {
        return this.getStringAttribute('queue_name');
    }
}
exports.DataAwsConnectRoutingProfileQueueConfigsOutputReference = DataAwsConnectRoutingProfileQueueConfigsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectRoutingProfileQueueConfigsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectRoutingProfile.DataAwsConnectRoutingProfileQueueConfigsOutputReference", version: "19.15.0" };
class DataAwsConnectRoutingProfileQueueConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectRoutingProfileQueueConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectRoutingProfileQueueConfigsList = DataAwsConnectRoutingProfileQueueConfigsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsConnectRoutingProfileQueueConfigsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsConnectRoutingProfile.DataAwsConnectRoutingProfileQueueConfigsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_routing_profile aws_connect_routing_profile}
*/
class DataAwsConnectRoutingProfile extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsConnectRoutingProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsConnectRoutingProfile to import
    * @param importFromId The id of the existing DataAwsConnectRoutingProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_routing_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsConnectRoutingProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_connect_routing_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_routing_profile aws_connect_routing_profile} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectRoutingProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_routing_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // media_concurrencies - computed: true, optional: false, required: false
        this._mediaConcurrencies = new DataAwsConnectRoutingProfileMediaConcurrenciesList(this, "media_concurrencies", true);
        // queue_configs - computed: true, optional: false, required: false
        this._queueConfigs = new DataAwsConnectRoutingProfileQueueConfigsList(this, "queue_configs", true);
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._routingProfileId = config.routingProfileId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_outbound_queue_id - computed: true, optional: false, required: false
    get defaultOutboundQueueId() {
        return this.getStringAttribute('default_outbound_queue_id');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get mediaConcurrencies() {
        return this._mediaConcurrencies;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get queueConfigs() {
        return this._queueConfigs;
    }
    get routingProfileId() {
        return this.getStringAttribute('routing_profile_id');
    }
    set routingProfileId(value) {
        this._routingProfileId = value;
    }
    resetRoutingProfileId() {
        this._routingProfileId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routingProfileIdInput() {
        return this._routingProfileId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            routing_profile_id: cdktf.stringToTerraform(this._routingProfileId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            routing_profile_id: {
                value: cdktf.stringToHclTerraform(this._routingProfileId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsConnectRoutingProfile = DataAwsConnectRoutingProfile;
_e = JSII_RTTI_SYMBOL_1;
DataAwsConnectRoutingProfile[_e] = { fqn: "@cdktf/provider-aws.dataAwsConnectRoutingProfile.DataAwsConnectRoutingProfile", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectRoutingProfile.tfResourceType = "aws_connect_routing_profile";
