// @cdktf/provider-aws/lib/data-aws-vpc-peering-connection/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpcPeeringConnection = exports.DataAwsVpcPeeringConnectionTimeoutsOutputReference = exports.dataAwsVpcPeeringConnectionTimeoutsToHclTerraform = exports.dataAwsVpcPeeringConnectionTimeoutsToTerraform = exports.DataAwsVpcPeeringConnectionFilterList = exports.DataAwsVpcPeeringConnectionFilterOutputReference = exports.dataAwsVpcPeeringConnectionFilterToHclTerraform = exports.dataAwsVpcPeeringConnectionFilterToTerraform = exports.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList = exports.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference = exports.dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToHclTerraform = exports.dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToTerraform = exports.DataAwsVpcPeeringConnectionPeerCidrBlockSetList = exports.DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference = exports.dataAwsVpcPeeringConnectionPeerCidrBlockSetToHclTerraform = exports.dataAwsVpcPeeringConnectionPeerCidrBlockSetToTerraform = exports.DataAwsVpcPeeringConnectionIpv6CidrBlockSetList = exports.DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference = exports.dataAwsVpcPeeringConnectionIpv6CidrBlockSetToHclTerraform = exports.dataAwsVpcPeeringConnectionIpv6CidrBlockSetToTerraform = exports.DataAwsVpcPeeringConnectionCidrBlockSetList = exports.DataAwsVpcPeeringConnectionCidrBlockSetOutputReference = exports.dataAwsVpcPeeringConnectionCidrBlockSetToHclTerraform = exports.dataAwsVpcPeeringConnectionCidrBlockSetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsVpcPeeringConnectionCidrBlockSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpcPeeringConnectionCidrBlockSetToTerraform = dataAwsVpcPeeringConnectionCidrBlockSetToTerraform;
function dataAwsVpcPeeringConnectionCidrBlockSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpcPeeringConnectionCidrBlockSetToHclTerraform = dataAwsVpcPeeringConnectionCidrBlockSetToHclTerraform;
class DataAwsVpcPeeringConnectionCidrBlockSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cidr_block - computed: true, optional: false, required: false
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionCidrBlockSetOutputReference = DataAwsVpcPeeringConnectionCidrBlockSetOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionCidrBlockSetOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionCidrBlockSetOutputReference", version: "19.15.0" };
class DataAwsVpcPeeringConnectionCidrBlockSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcPeeringConnectionCidrBlockSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcPeeringConnectionCidrBlockSetList = DataAwsVpcPeeringConnectionCidrBlockSetList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionCidrBlockSetList[_b] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionCidrBlockSetList", version: "19.15.0" };
function dataAwsVpcPeeringConnectionIpv6CidrBlockSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpcPeeringConnectionIpv6CidrBlockSetToTerraform = dataAwsVpcPeeringConnectionIpv6CidrBlockSetToTerraform;
function dataAwsVpcPeeringConnectionIpv6CidrBlockSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpcPeeringConnectionIpv6CidrBlockSetToHclTerraform = dataAwsVpcPeeringConnectionIpv6CidrBlockSetToHclTerraform;
class DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ipv6_cidr_block - computed: true, optional: false, required: false
    get ipv6CidrBlock() {
        return this.getStringAttribute('ipv6_cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference = DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference", version: "19.15.0" };
class DataAwsVpcPeeringConnectionIpv6CidrBlockSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcPeeringConnectionIpv6CidrBlockSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcPeeringConnectionIpv6CidrBlockSetList = DataAwsVpcPeeringConnectionIpv6CidrBlockSetList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionIpv6CidrBlockSetList[_d] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionIpv6CidrBlockSetList", version: "19.15.0" };
function dataAwsVpcPeeringConnectionPeerCidrBlockSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpcPeeringConnectionPeerCidrBlockSetToTerraform = dataAwsVpcPeeringConnectionPeerCidrBlockSetToTerraform;
function dataAwsVpcPeeringConnectionPeerCidrBlockSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpcPeeringConnectionPeerCidrBlockSetToHclTerraform = dataAwsVpcPeeringConnectionPeerCidrBlockSetToHclTerraform;
class DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cidr_block - computed: true, optional: false, required: false
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference = DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference", version: "19.15.0" };
class DataAwsVpcPeeringConnectionPeerCidrBlockSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcPeeringConnectionPeerCidrBlockSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcPeeringConnectionPeerCidrBlockSetList = DataAwsVpcPeeringConnectionPeerCidrBlockSetList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionPeerCidrBlockSetList[_f] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionPeerCidrBlockSetList", version: "19.15.0" };
function dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToTerraform = dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToTerraform;
function dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToHclTerraform = dataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetToHclTerraform;
class DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ipv6_cidr_block - computed: true, optional: false, required: false
    get ipv6CidrBlock() {
        return this.getStringAttribute('ipv6_cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference = DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference", version: "19.15.0" };
class DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList = DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList[_h] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList", version: "19.15.0" };
function dataAwsVpcPeeringConnectionFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsVpcPeeringConnectionFilterToTerraform = dataAwsVpcPeeringConnectionFilterToTerraform;
function dataAwsVpcPeeringConnectionFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsVpcPeeringConnectionFilterToHclTerraform = dataAwsVpcPeeringConnectionFilterToHclTerraform;
class DataAwsVpcPeeringConnectionFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsVpcPeeringConnectionFilterOutputReference = DataAwsVpcPeeringConnectionFilterOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionFilterOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionFilterOutputReference", version: "19.15.0" };
class DataAwsVpcPeeringConnectionFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcPeeringConnectionFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcPeeringConnectionFilterList = DataAwsVpcPeeringConnectionFilterList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionFilterList[_k] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionFilterList", version: "19.15.0" };
function dataAwsVpcPeeringConnectionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsVpcPeeringConnectionTimeoutsToTerraform = dataAwsVpcPeeringConnectionTimeoutsToTerraform;
function dataAwsVpcPeeringConnectionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsVpcPeeringConnectionTimeoutsToHclTerraform = dataAwsVpcPeeringConnectionTimeoutsToHclTerraform;
class DataAwsVpcPeeringConnectionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsVpcPeeringConnectionTimeoutsOutputReference = DataAwsVpcPeeringConnectionTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnectionTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpc_peering_connection aws_vpc_peering_connection}
*/
class DataAwsVpcPeeringConnection extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsVpcPeeringConnection resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsVpcPeeringConnection to import
    * @param importFromId The id of the existing DataAwsVpcPeeringConnection that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpc_peering_connection#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsVpcPeeringConnection to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpc_peering_connection", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpc_peering_connection aws_vpc_peering_connection} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsVpcPeeringConnectionConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_peering_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // accepter - computed: true, optional: false, required: false
        this._accepter = new cdktf.BooleanMap(this, "accepter");
        // cidr_block_set - computed: true, optional: false, required: false
        this._cidrBlockSet = new DataAwsVpcPeeringConnectionCidrBlockSetList(this, "cidr_block_set", false);
        // ipv6_cidr_block_set - computed: true, optional: false, required: false
        this._ipv6CidrBlockSet = new DataAwsVpcPeeringConnectionIpv6CidrBlockSetList(this, "ipv6_cidr_block_set", false);
        // peer_cidr_block_set - computed: true, optional: false, required: false
        this._peerCidrBlockSet = new DataAwsVpcPeeringConnectionPeerCidrBlockSetList(this, "peer_cidr_block_set", false);
        // peer_ipv6_cidr_block_set - computed: true, optional: false, required: false
        this._peerIpv6CidrBlockSet = new DataAwsVpcPeeringConnectionPeerIpv6CidrBlockSetList(this, "peer_ipv6_cidr_block_set", false);
        // requester - computed: true, optional: false, required: false
        this._requester = new cdktf.BooleanMap(this, "requester");
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsVpcPeeringConnectionFilterList(this, "filter", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsVpcPeeringConnectionTimeoutsOutputReference(this, "timeouts");
        this._cidrBlock = config.cidrBlock;
        this._id = config.id;
        this._ownerId = config.ownerId;
        this._peerCidrBlock = config.peerCidrBlock;
        this._peerOwnerId = config.peerOwnerId;
        this._peerRegion = config.peerRegion;
        this._peerVpcId = config.peerVpcId;
        this._region = config.region;
        this._status = config.status;
        this._tags = config.tags;
        this._vpcId = config.vpcId;
        this._filter.internalValue = config.filter;
        this._timeouts.internalValue = config.timeouts;
    }
    get accepter() {
        return this._accepter;
    }
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
    set cidrBlock(value) {
        this._cidrBlock = value;
    }
    resetCidrBlock() {
        this._cidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlockInput() {
        return this._cidrBlock;
    }
    get cidrBlockSet() {
        return this._cidrBlockSet;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipv6CidrBlockSet() {
        return this._ipv6CidrBlockSet;
    }
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    set ownerId(value) {
        this._ownerId = value;
    }
    resetOwnerId() {
        this._ownerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ownerIdInput() {
        return this._ownerId;
    }
    get peerCidrBlock() {
        return this.getStringAttribute('peer_cidr_block');
    }
    set peerCidrBlock(value) {
        this._peerCidrBlock = value;
    }
    resetPeerCidrBlock() {
        this._peerCidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get peerCidrBlockInput() {
        return this._peerCidrBlock;
    }
    get peerCidrBlockSet() {
        return this._peerCidrBlockSet;
    }
    get peerIpv6CidrBlockSet() {
        return this._peerIpv6CidrBlockSet;
    }
    get peerOwnerId() {
        return this.getStringAttribute('peer_owner_id');
    }
    set peerOwnerId(value) {
        this._peerOwnerId = value;
    }
    resetPeerOwnerId() {
        this._peerOwnerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get peerOwnerIdInput() {
        return this._peerOwnerId;
    }
    get peerRegion() {
        return this.getStringAttribute('peer_region');
    }
    set peerRegion(value) {
        this._peerRegion = value;
    }
    resetPeerRegion() {
        this._peerRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get peerRegionInput() {
        return this._peerRegion;
    }
    get peerVpcId() {
        return this.getStringAttribute('peer_vpc_id');
    }
    set peerVpcId(value) {
        this._peerVpcId = value;
    }
    resetPeerVpcId() {
        this._peerVpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get peerVpcIdInput() {
        return this._peerVpcId;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get requester() {
        return this._requester;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cidr_block: cdktf.stringToTerraform(this._cidrBlock),
            id: cdktf.stringToTerraform(this._id),
            owner_id: cdktf.stringToTerraform(this._ownerId),
            peer_cidr_block: cdktf.stringToTerraform(this._peerCidrBlock),
            peer_owner_id: cdktf.stringToTerraform(this._peerOwnerId),
            peer_region: cdktf.stringToTerraform(this._peerRegion),
            peer_vpc_id: cdktf.stringToTerraform(this._peerVpcId),
            region: cdktf.stringToTerraform(this._region),
            status: cdktf.stringToTerraform(this._status),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            filter: cdktf.listMapper(dataAwsVpcPeeringConnectionFilterToTerraform, true)(this._filter.internalValue),
            timeouts: dataAwsVpcPeeringConnectionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cidr_block: {
                value: cdktf.stringToHclTerraform(this._cidrBlock),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            owner_id: {
                value: cdktf.stringToHclTerraform(this._ownerId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            peer_cidr_block: {
                value: cdktf.stringToHclTerraform(this._peerCidrBlock),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            peer_owner_id: {
                value: cdktf.stringToHclTerraform(this._peerOwnerId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            peer_region: {
                value: cdktf.stringToHclTerraform(this._peerRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            peer_vpc_id: {
                value: cdktf.stringToHclTerraform(this._peerVpcId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            vpc_id: {
                value: cdktf.stringToHclTerraform(this._vpcId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filter: {
                value: cdktf.listMapperHcl(dataAwsVpcPeeringConnectionFilterToHclTerraform, true)(this._filter.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsVpcPeeringConnectionFilterList",
            },
            timeouts: {
                value: dataAwsVpcPeeringConnectionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsVpcPeeringConnectionTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsVpcPeeringConnection = DataAwsVpcPeeringConnection;
_m = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnection[_m] = { fqn: "@cdktf/provider-aws.dataAwsVpcPeeringConnection.DataAwsVpcPeeringConnection", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsVpcPeeringConnection.tfResourceType = "aws_vpc_peering_connection";
