// @cdktf/provider-aws/lib/securityhub-organization-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubOrganizationConfiguration = exports.SecurityhubOrganizationConfigurationTimeoutsOutputReference = exports.securityhubOrganizationConfigurationTimeoutsToHclTerraform = exports.securityhubOrganizationConfigurationTimeoutsToTerraform = exports.SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference = exports.securityhubOrganizationConfigurationOrganizationConfigurationToHclTerraform = exports.securityhubOrganizationConfigurationOrganizationConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function securityhubOrganizationConfigurationOrganizationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        configuration_type: cdktf.stringToTerraform(struct.configurationType),
    };
}
exports.securityhubOrganizationConfigurationOrganizationConfigurationToTerraform = securityhubOrganizationConfigurationOrganizationConfigurationToTerraform;
function securityhubOrganizationConfigurationOrganizationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        configuration_type: {
            value: cdktf.stringToHclTerraform(struct.configurationType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securityhubOrganizationConfigurationOrganizationConfigurationToHclTerraform = securityhubOrganizationConfigurationOrganizationConfigurationToHclTerraform;
class SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._configurationType !== undefined) {
            hasAnyValues = true;
            internalValueResult.configurationType = this._configurationType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._configurationType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._configurationType = value.configurationType;
        }
    }
    get configurationType() {
        return this.getStringAttribute('configuration_type');
    }
    set configurationType(value) {
        this._configurationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get configurationTypeInput() {
        return this._configurationType;
    }
}
exports.SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference = SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.securityhubOrganizationConfiguration.SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference", version: "19.15.0" };
function securityhubOrganizationConfigurationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.securityhubOrganizationConfigurationTimeoutsToTerraform = securityhubOrganizationConfigurationTimeoutsToTerraform;
function securityhubOrganizationConfigurationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securityhubOrganizationConfigurationTimeoutsToHclTerraform = securityhubOrganizationConfigurationTimeoutsToHclTerraform;
class SecurityhubOrganizationConfigurationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.SecurityhubOrganizationConfigurationTimeoutsOutputReference = SecurityhubOrganizationConfigurationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SecurityhubOrganizationConfigurationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.securityhubOrganizationConfiguration.SecurityhubOrganizationConfigurationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_organization_configuration aws_securityhub_organization_configuration}
*/
class SecurityhubOrganizationConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecurityhubOrganizationConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecurityhubOrganizationConfiguration to import
    * @param importFromId The id of the existing SecurityhubOrganizationConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_organization_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecurityhubOrganizationConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securityhub_organization_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_organization_configuration aws_securityhub_organization_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubOrganizationConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_organization_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // organization_configuration - computed: false, optional: true, required: false
        this._organizationConfiguration = new SecurityhubOrganizationConfigurationOrganizationConfigurationOutputReference(this, "organization_configuration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new SecurityhubOrganizationConfigurationTimeoutsOutputReference(this, "timeouts");
        this._autoEnable = config.autoEnable;
        this._autoEnableStandards = config.autoEnableStandards;
        this._id = config.id;
        this._organizationConfiguration.internalValue = config.organizationConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    get autoEnable() {
        return this.getBooleanAttribute('auto_enable');
    }
    set autoEnable(value) {
        this._autoEnable = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableInput() {
        return this._autoEnable;
    }
    get autoEnableStandards() {
        return this.getStringAttribute('auto_enable_standards');
    }
    set autoEnableStandards(value) {
        this._autoEnableStandards = value;
    }
    resetAutoEnableStandards() {
        this._autoEnableStandards = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableStandardsInput() {
        return this._autoEnableStandards;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get organizationConfiguration() {
        return this._organizationConfiguration;
    }
    putOrganizationConfiguration(value) {
        this._organizationConfiguration.internalValue = value;
    }
    resetOrganizationConfiguration() {
        this._organizationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationConfigurationInput() {
        return this._organizationConfiguration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_enable: cdktf.booleanToTerraform(this._autoEnable),
            auto_enable_standards: cdktf.stringToTerraform(this._autoEnableStandards),
            id: cdktf.stringToTerraform(this._id),
            organization_configuration: securityhubOrganizationConfigurationOrganizationConfigurationToTerraform(this._organizationConfiguration.internalValue),
            timeouts: securityhubOrganizationConfigurationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_enable: {
                value: cdktf.booleanToHclTerraform(this._autoEnable),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            auto_enable_standards: {
                value: cdktf.stringToHclTerraform(this._autoEnableStandards),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organization_configuration: {
                value: securityhubOrganizationConfigurationOrganizationConfigurationToHclTerraform(this._organizationConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SecurityhubOrganizationConfigurationOrganizationConfigurationList",
            },
            timeouts: {
                value: securityhubOrganizationConfigurationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "SecurityhubOrganizationConfigurationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecurityhubOrganizationConfiguration = SecurityhubOrganizationConfiguration;
_c = JSII_RTTI_SYMBOL_1;
SecurityhubOrganizationConfiguration[_c] = { fqn: "@cdktf/provider-aws.securityhubOrganizationConfiguration.SecurityhubOrganizationConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubOrganizationConfiguration.tfResourceType = "aws_securityhub_organization_configuration";
