// @cdktf/provider-aws/lib/data-aws-globalaccelerator-custom-routing-accelerator/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsGlobalacceleratorCustomRoutingAccelerator = exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList = exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference = exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToHclTerraform = exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToTerraform = exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList = exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference = exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToHclTerraform = exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToTerraform = dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToTerraform;
function dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToHclTerraform = dataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesToHclTerraform;
class DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // flow_logs_enabled - computed: true, optional: false, required: false
    get flowLogsEnabled() {
        return this.getBooleanAttribute('flow_logs_enabled');
    }
    // flow_logs_s3_bucket - computed: true, optional: false, required: false
    get flowLogsS3Bucket() {
        return this.getStringAttribute('flow_logs_s3_bucket');
    }
    // flow_logs_s3_prefix - computed: true, optional: false, required: false
    get flowLogsS3Prefix() {
        return this.getStringAttribute('flow_logs_s3_prefix');
    }
}
exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference = DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorCustomRoutingAccelerator.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference", version: "19.15.0" };
class DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList = DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorCustomRoutingAccelerator.DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList", version: "19.15.0" };
function dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToTerraform = dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToTerraform;
function dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToHclTerraform = dataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsToHclTerraform;
class DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return this.getListAttribute('ip_addresses');
    }
    // ip_family - computed: true, optional: false, required: false
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
}
exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference = DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorCustomRoutingAccelerator.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference", version: "19.15.0" };
class DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList = DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorCustomRoutingAccelerator.DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/globalaccelerator_custom_routing_accelerator aws_globalaccelerator_custom_routing_accelerator}
*/
class DataAwsGlobalacceleratorCustomRoutingAccelerator extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsGlobalacceleratorCustomRoutingAccelerator resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsGlobalacceleratorCustomRoutingAccelerator to import
    * @param importFromId The id of the existing DataAwsGlobalacceleratorCustomRoutingAccelerator that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/globalaccelerator_custom_routing_accelerator#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsGlobalacceleratorCustomRoutingAccelerator to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_globalaccelerator_custom_routing_accelerator", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/globalaccelerator_custom_routing_accelerator aws_globalaccelerator_custom_routing_accelerator} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsGlobalacceleratorCustomRoutingAcceleratorConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_custom_routing_accelerator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attributes - computed: true, optional: false, required: false
        this._attributes = new DataAwsGlobalacceleratorCustomRoutingAcceleratorAttributesList(this, "attributes", false);
        // ip_sets - computed: true, optional: false, required: false
        this._ipSets = new DataAwsGlobalacceleratorCustomRoutingAcceleratorIpSetsList(this, "ip_sets", false);
        this._arn = config.arn;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get attributes() {
        return this._attributes;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ip_address_type - computed: true, optional: false, required: false
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    get ipSets() {
        return this._ipSets;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsGlobalacceleratorCustomRoutingAccelerator = DataAwsGlobalacceleratorCustomRoutingAccelerator;
_e = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorCustomRoutingAccelerator[_e] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorCustomRoutingAccelerator.DataAwsGlobalacceleratorCustomRoutingAccelerator", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsGlobalacceleratorCustomRoutingAccelerator.tfResourceType = "aws_globalaccelerator_custom_routing_accelerator";
