// @cdktf/provider-aws/lib/data-aws-fsx-ontap-storage-virtual-machine/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsFsxOntapStorageVirtualMachine = exports.DataAwsFsxOntapStorageVirtualMachineFilterList = exports.DataAwsFsxOntapStorageVirtualMachineFilterOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineFilterToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineFilterToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList = exports.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsList = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList = exports.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList = exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToTerraform = exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList = exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference = exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToHclTerraform = exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToTerraform = dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToTerraform;
function dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToHclTerraform = dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_ips - computed: true, optional: false, required: false
    get dnsIps() {
        return cdktf.Fn.tolist(this.getListAttribute('dns_ips'));
    }
    // domain_name - computed: true, optional: false, required: false
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // file_system_administrators_group - computed: true, optional: false, required: false
    get fileSystemAdministratorsGroup() {
        return this.getStringAttribute('file_system_administrators_group');
    }
    // organizational_unit_distinguished_name - computed: true, optional: false, required: false
    get organizationalUnitDistinguishedName() {
        return this.getStringAttribute('organizational_unit_distinguished_name');
    }
    // username - computed: true, optional: false, required: false
    get username() {
        return this.getStringAttribute('username');
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference = DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList = DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToTerraform = dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToTerraform;
function dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToHclTerraform = dataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // self_managed_active_directory_configuration - computed: true, optional: false, required: false
        this._selfManagedActiveDirectoryConfiguration = new DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationSelfManagedActiveDirectoryConfigurationList(this, "self_managed_active_directory_configuration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // netbios_name - computed: true, optional: false, required: false
    get netbiosName() {
        return this.getStringAttribute('netbios_name');
    }
    get selfManagedActiveDirectoryConfiguration() {
        return this._selfManagedActiveDirectoryConfiguration;
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference = DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList = DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList[_d] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToTerraform;
function dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToHclTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsIscsiToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference = DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList = DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList[_f] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToTerraform;
function dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToHclTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsManagementToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference = DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineEndpointsManagementOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList = DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList[_h] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToTerraform;
function dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToHclTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsNfsToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference = DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineEndpointsNfsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList = DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList[_k] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToTerraform;
function dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToHclTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsSmbToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_addresses'));
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference = DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineEndpointsSmbOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList = DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList[_m] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsToTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsToTerraform;
function dataAwsFsxOntapStorageVirtualMachineEndpointsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineEndpointsToHclTerraform = dataAwsFsxOntapStorageVirtualMachineEndpointsToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // iscsi - computed: true, optional: false, required: false
        this._iscsi = new DataAwsFsxOntapStorageVirtualMachineEndpointsIscsiList(this, "iscsi", false);
        // management - computed: true, optional: false, required: false
        this._management = new DataAwsFsxOntapStorageVirtualMachineEndpointsManagementList(this, "management", false);
        // nfs - computed: true, optional: false, required: false
        this._nfs = new DataAwsFsxOntapStorageVirtualMachineEndpointsNfsList(this, "nfs", false);
        // smb - computed: true, optional: false, required: false
        this._smb = new DataAwsFsxOntapStorageVirtualMachineEndpointsSmbList(this, "smb", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get iscsi() {
        return this._iscsi;
    }
    get management() {
        return this._management;
    }
    get nfs() {
        return this._nfs;
    }
    get smb() {
        return this._smb;
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference = DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineEndpointsList = DataAwsFsxOntapStorageVirtualMachineEndpointsList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineEndpointsList[_p] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineEndpointsList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToTerraform = dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToTerraform;
function dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToHclTerraform = dataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference = DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList = DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList[_r] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList", version: "19.15.0" };
function dataAwsFsxOntapStorageVirtualMachineFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsFsxOntapStorageVirtualMachineFilterToTerraform = dataAwsFsxOntapStorageVirtualMachineFilterToTerraform;
function dataAwsFsxOntapStorageVirtualMachineFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsFsxOntapStorageVirtualMachineFilterToHclTerraform = dataAwsFsxOntapStorageVirtualMachineFilterToHclTerraform;
class DataAwsFsxOntapStorageVirtualMachineFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineFilterOutputReference = DataAwsFsxOntapStorageVirtualMachineFilterOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineFilterOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineFilterOutputReference", version: "19.15.0" };
class DataAwsFsxOntapStorageVirtualMachineFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxOntapStorageVirtualMachineFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxOntapStorageVirtualMachineFilterList = DataAwsFsxOntapStorageVirtualMachineFilterList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachineFilterList[_t] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachineFilterList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_storage_virtual_machine aws_fsx_ontap_storage_virtual_machine}
*/
class DataAwsFsxOntapStorageVirtualMachine extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsFsxOntapStorageVirtualMachine resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsFsxOntapStorageVirtualMachine to import
    * @param importFromId The id of the existing DataAwsFsxOntapStorageVirtualMachine that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_storage_virtual_machine#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsFsxOntapStorageVirtualMachine to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_ontap_storage_virtual_machine", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_ontap_storage_virtual_machine aws_fsx_ontap_storage_virtual_machine} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsFsxOntapStorageVirtualMachineConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_ontap_storage_virtual_machine',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // active_directory_configuration - computed: true, optional: false, required: false
        this._activeDirectoryConfiguration = new DataAwsFsxOntapStorageVirtualMachineActiveDirectoryConfigurationList(this, "active_directory_configuration", false);
        // endpoints - computed: true, optional: false, required: false
        this._endpoints = new DataAwsFsxOntapStorageVirtualMachineEndpointsList(this, "endpoints", false);
        // lifecycle_transition_reason - computed: true, optional: false, required: false
        this._lifecycleTransitionReason = new DataAwsFsxOntapStorageVirtualMachineLifecycleTransitionReasonList(this, "lifecycle_transition_reason", true);
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsFsxOntapStorageVirtualMachineFilterList(this, "filter", true);
        this._id = config.id;
        this._tags = config.tags;
        this._filter.internalValue = config.filter;
    }
    get activeDirectoryConfiguration() {
        return this._activeDirectoryConfiguration;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    get endpoints() {
        return this._endpoints;
    }
    // file_system_id - computed: true, optional: false, required: false
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // lifecycle_status - computed: true, optional: false, required: false
    get lifecycleStatus() {
        return this.getStringAttribute('lifecycle_status');
    }
    get lifecycleTransitionReason() {
        return this._lifecycleTransitionReason;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // subtype - computed: true, optional: false, required: false
    get subtype() {
        return this.getStringAttribute('subtype');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // uuid - computed: true, optional: false, required: false
    get uuid() {
        return this.getStringAttribute('uuid');
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsFsxOntapStorageVirtualMachineFilterToTerraform, true)(this._filter.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            filter: {
                value: cdktf.listMapperHcl(dataAwsFsxOntapStorageVirtualMachineFilterToHclTerraform, true)(this._filter.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsFsxOntapStorageVirtualMachineFilterList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsFsxOntapStorageVirtualMachine = DataAwsFsxOntapStorageVirtualMachine;
_u = JSII_RTTI_SYMBOL_1;
DataAwsFsxOntapStorageVirtualMachine[_u] = { fqn: "@cdktf/provider-aws.dataAwsFsxOntapStorageVirtualMachine.DataAwsFsxOntapStorageVirtualMachine", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsFsxOntapStorageVirtualMachine.tfResourceType = "aws_fsx_ontap_storage_virtual_machine";
