// @cdktf/provider-aws/lib/scheduler-schedule/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchedulerSchedule = exports.SchedulerScheduleTargetOutputReference = exports.schedulerScheduleTargetToHclTerraform = exports.schedulerScheduleTargetToTerraform = exports.SchedulerScheduleTargetSqsParametersOutputReference = exports.schedulerScheduleTargetSqsParametersToHclTerraform = exports.schedulerScheduleTargetSqsParametersToTerraform = exports.SchedulerScheduleTargetSagemakerPipelineParametersOutputReference = exports.schedulerScheduleTargetSagemakerPipelineParametersToHclTerraform = exports.schedulerScheduleTargetSagemakerPipelineParametersToTerraform = exports.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList = exports.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference = exports.schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToHclTerraform = exports.schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToTerraform = exports.SchedulerScheduleTargetRetryPolicyOutputReference = exports.schedulerScheduleTargetRetryPolicyToHclTerraform = exports.schedulerScheduleTargetRetryPolicyToTerraform = exports.SchedulerScheduleTargetKinesisParametersOutputReference = exports.schedulerScheduleTargetKinesisParametersToHclTerraform = exports.schedulerScheduleTargetKinesisParametersToTerraform = exports.SchedulerScheduleTargetEventbridgeParametersOutputReference = exports.schedulerScheduleTargetEventbridgeParametersToHclTerraform = exports.schedulerScheduleTargetEventbridgeParametersToTerraform = exports.SchedulerScheduleTargetEcsParametersOutputReference = exports.schedulerScheduleTargetEcsParametersToHclTerraform = exports.schedulerScheduleTargetEcsParametersToTerraform = exports.SchedulerScheduleTargetEcsParametersPlacementStrategyList = exports.SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference = exports.schedulerScheduleTargetEcsParametersPlacementStrategyToHclTerraform = exports.schedulerScheduleTargetEcsParametersPlacementStrategyToTerraform = exports.SchedulerScheduleTargetEcsParametersPlacementConstraintsList = exports.SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference = exports.schedulerScheduleTargetEcsParametersPlacementConstraintsToHclTerraform = exports.schedulerScheduleTargetEcsParametersPlacementConstraintsToTerraform = exports.SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference = exports.schedulerScheduleTargetEcsParametersNetworkConfigurationToHclTerraform = exports.schedulerScheduleTargetEcsParametersNetworkConfigurationToTerraform = exports.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList = exports.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference = exports.schedulerScheduleTargetEcsParametersCapacityProviderStrategyToHclTerraform = exports.schedulerScheduleTargetEcsParametersCapacityProviderStrategyToTerraform = exports.SchedulerScheduleTargetDeadLetterConfigOutputReference = exports.schedulerScheduleTargetDeadLetterConfigToHclTerraform = exports.schedulerScheduleTargetDeadLetterConfigToTerraform = exports.SchedulerScheduleFlexibleTimeWindowOutputReference = exports.schedulerScheduleFlexibleTimeWindowToHclTerraform = exports.schedulerScheduleFlexibleTimeWindowToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function schedulerScheduleFlexibleTimeWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_window_in_minutes: cdktf.numberToTerraform(struct.maximumWindowInMinutes),
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.schedulerScheduleFlexibleTimeWindowToTerraform = schedulerScheduleFlexibleTimeWindowToTerraform;
function schedulerScheduleFlexibleTimeWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_window_in_minutes: {
            value: cdktf.numberToHclTerraform(struct.maximumWindowInMinutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleFlexibleTimeWindowToHclTerraform = schedulerScheduleFlexibleTimeWindowToHclTerraform;
class SchedulerScheduleFlexibleTimeWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumWindowInMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumWindowInMinutes = this._maximumWindowInMinutes;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumWindowInMinutes = undefined;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumWindowInMinutes = value.maximumWindowInMinutes;
            this._mode = value.mode;
        }
    }
    get maximumWindowInMinutes() {
        return this.getNumberAttribute('maximum_window_in_minutes');
    }
    set maximumWindowInMinutes(value) {
        this._maximumWindowInMinutes = value;
    }
    resetMaximumWindowInMinutes() {
        this._maximumWindowInMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumWindowInMinutesInput() {
        return this._maximumWindowInMinutes;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.SchedulerScheduleFlexibleTimeWindowOutputReference = SchedulerScheduleFlexibleTimeWindowOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SchedulerScheduleFlexibleTimeWindowOutputReference[_a] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleFlexibleTimeWindowOutputReference", version: "19.15.0" };
function schedulerScheduleTargetDeadLetterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
    };
}
exports.schedulerScheduleTargetDeadLetterConfigToTerraform = schedulerScheduleTargetDeadLetterConfigToTerraform;
function schedulerScheduleTargetDeadLetterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetDeadLetterConfigToHclTerraform = schedulerScheduleTargetDeadLetterConfigToHclTerraform;
class SchedulerScheduleTargetDeadLetterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
}
exports.SchedulerScheduleTargetDeadLetterConfigOutputReference = SchedulerScheduleTargetDeadLetterConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetDeadLetterConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetDeadLetterConfigOutputReference", version: "19.15.0" };
function schedulerScheduleTargetEcsParametersCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.schedulerScheduleTargetEcsParametersCapacityProviderStrategyToTerraform = schedulerScheduleTargetEcsParametersCapacityProviderStrategyToTerraform;
function schedulerScheduleTargetEcsParametersCapacityProviderStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        base: {
            value: cdktf.numberToHclTerraform(struct.base),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        capacity_provider: {
            value: cdktf.stringToHclTerraform(struct.capacityProvider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weight: {
            value: cdktf.numberToHclTerraform(struct.weight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEcsParametersCapacityProviderStrategyToHclTerraform = schedulerScheduleTargetEcsParametersCapacityProviderStrategyToHclTerraform;
class SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._base !== undefined) {
            hasAnyValues = true;
            internalValueResult.base = this._base;
        }
        if (this._capacityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProvider = this._capacityProvider;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._base = undefined;
            this._capacityProvider = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._base = value.base;
            this._capacityProvider = value.capacityProvider;
            this._weight = value.weight;
        }
    }
    get base() {
        return this.getNumberAttribute('base');
    }
    set base(value) {
        this._base = value;
    }
    resetBase() {
        this._base = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseInput() {
        return this._base;
    }
    get capacityProvider() {
        return this.getStringAttribute('capacity_provider');
    }
    set capacityProvider(value) {
        this._capacityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderInput() {
        return this._capacityProvider;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference = SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference[_c] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference", version: "19.15.0" };
class SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SchedulerScheduleTargetEcsParametersCapacityProviderStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList = SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList;
_d = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList[_d] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList", version: "19.15.0" };
function schedulerScheduleTargetEcsParametersNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.schedulerScheduleTargetEcsParametersNetworkConfigurationToTerraform = schedulerScheduleTargetEcsParametersNetworkConfigurationToTerraform;
function schedulerScheduleTargetEcsParametersNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        assign_public_ip: {
            value: cdktf.booleanToHclTerraform(struct.assignPublicIp),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        security_groups: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEcsParametersNetworkConfigurationToHclTerraform = schedulerScheduleTargetEcsParametersNetworkConfigurationToHclTerraform;
class SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference = SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference", version: "19.15.0" };
function schedulerScheduleTargetEcsParametersPlacementConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.schedulerScheduleTargetEcsParametersPlacementConstraintsToTerraform = schedulerScheduleTargetEcsParametersPlacementConstraintsToTerraform;
function schedulerScheduleTargetEcsParametersPlacementConstraintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEcsParametersPlacementConstraintsToHclTerraform = schedulerScheduleTargetEcsParametersPlacementConstraintsToHclTerraform;
class SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expression = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expression = value.expression;
            this._type = value.type;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    resetExpression() {
        this._expression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference = SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference[_f] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference", version: "19.15.0" };
class SchedulerScheduleTargetEcsParametersPlacementConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SchedulerScheduleTargetEcsParametersPlacementConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SchedulerScheduleTargetEcsParametersPlacementConstraintsList = SchedulerScheduleTargetEcsParametersPlacementConstraintsList;
_g = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersPlacementConstraintsList[_g] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersPlacementConstraintsList", version: "19.15.0" };
function schedulerScheduleTargetEcsParametersPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.schedulerScheduleTargetEcsParametersPlacementStrategyToTerraform = schedulerScheduleTargetEcsParametersPlacementStrategyToTerraform;
function schedulerScheduleTargetEcsParametersPlacementStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEcsParametersPlacementStrategyToHclTerraform = schedulerScheduleTargetEcsParametersPlacementStrategyToHclTerraform;
class SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._field = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._field = value.field;
            this._type = value.type;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference = SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference[_h] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference", version: "19.15.0" };
class SchedulerScheduleTargetEcsParametersPlacementStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SchedulerScheduleTargetEcsParametersPlacementStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SchedulerScheduleTargetEcsParametersPlacementStrategyList = SchedulerScheduleTargetEcsParametersPlacementStrategyList;
_j = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersPlacementStrategyList[_j] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersPlacementStrategyList", version: "19.15.0" };
function schedulerScheduleTargetEcsParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_ecs_managed_tags: cdktf.booleanToTerraform(struct.enableEcsManagedTags),
        enable_execute_command: cdktf.booleanToTerraform(struct.enableExecuteCommand),
        group: cdktf.stringToTerraform(struct.group),
        launch_type: cdktf.stringToTerraform(struct.launchType),
        platform_version: cdktf.stringToTerraform(struct.platformVersion),
        propagate_tags: cdktf.stringToTerraform(struct.propagateTags),
        reference_id: cdktf.stringToTerraform(struct.referenceId),
        tags: cdktf.hashMapper(cdktf.stringToTerraform)(struct.tags),
        task_count: cdktf.numberToTerraform(struct.taskCount),
        task_definition_arn: cdktf.stringToTerraform(struct.taskDefinitionArn),
        capacity_provider_strategy: cdktf.listMapper(schedulerScheduleTargetEcsParametersCapacityProviderStrategyToTerraform, true)(struct.capacityProviderStrategy),
        network_configuration: schedulerScheduleTargetEcsParametersNetworkConfigurationToTerraform(struct.networkConfiguration),
        placement_constraints: cdktf.listMapper(schedulerScheduleTargetEcsParametersPlacementConstraintsToTerraform, true)(struct.placementConstraints),
        placement_strategy: cdktf.listMapper(schedulerScheduleTargetEcsParametersPlacementStrategyToTerraform, true)(struct.placementStrategy),
    };
}
exports.schedulerScheduleTargetEcsParametersToTerraform = schedulerScheduleTargetEcsParametersToTerraform;
function schedulerScheduleTargetEcsParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_ecs_managed_tags: {
            value: cdktf.booleanToHclTerraform(struct.enableEcsManagedTags),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_execute_command: {
            value: cdktf.booleanToHclTerraform(struct.enableExecuteCommand),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        group: {
            value: cdktf.stringToHclTerraform(struct.group),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        launch_type: {
            value: cdktf.stringToHclTerraform(struct.launchType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        platform_version: {
            value: cdktf.stringToHclTerraform(struct.platformVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        propagate_tags: {
            value: cdktf.stringToHclTerraform(struct.propagateTags),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        reference_id: {
            value: cdktf.stringToHclTerraform(struct.referenceId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tags: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.tags),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        task_count: {
            value: cdktf.numberToHclTerraform(struct.taskCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        task_definition_arn: {
            value: cdktf.stringToHclTerraform(struct.taskDefinitionArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        capacity_provider_strategy: {
            value: cdktf.listMapperHcl(schedulerScheduleTargetEcsParametersCapacityProviderStrategyToHclTerraform, true)(struct.capacityProviderStrategy),
            isBlock: true,
            type: "set",
            storageClassType: "SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList",
        },
        network_configuration: {
            value: schedulerScheduleTargetEcsParametersNetworkConfigurationToHclTerraform(struct.networkConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetEcsParametersNetworkConfigurationList",
        },
        placement_constraints: {
            value: cdktf.listMapperHcl(schedulerScheduleTargetEcsParametersPlacementConstraintsToHclTerraform, true)(struct.placementConstraints),
            isBlock: true,
            type: "set",
            storageClassType: "SchedulerScheduleTargetEcsParametersPlacementConstraintsList",
        },
        placement_strategy: {
            value: cdktf.listMapperHcl(schedulerScheduleTargetEcsParametersPlacementStrategyToHclTerraform, true)(struct.placementStrategy),
            isBlock: true,
            type: "set",
            storageClassType: "SchedulerScheduleTargetEcsParametersPlacementStrategyList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEcsParametersToHclTerraform = schedulerScheduleTargetEcsParametersToHclTerraform;
class SchedulerScheduleTargetEcsParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // capacity_provider_strategy - computed: false, optional: true, required: false
        this._capacityProviderStrategy = new SchedulerScheduleTargetEcsParametersCapacityProviderStrategyList(this, "capacity_provider_strategy", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new SchedulerScheduleTargetEcsParametersNetworkConfigurationOutputReference(this, "network_configuration");
        // placement_constraints - computed: false, optional: true, required: false
        this._placementConstraints = new SchedulerScheduleTargetEcsParametersPlacementConstraintsList(this, "placement_constraints", true);
        // placement_strategy - computed: false, optional: true, required: false
        this._placementStrategy = new SchedulerScheduleTargetEcsParametersPlacementStrategyList(this, "placement_strategy", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableEcsManagedTags !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableEcsManagedTags = this._enableEcsManagedTags;
        }
        if (this._enableExecuteCommand !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableExecuteCommand = this._enableExecuteCommand;
        }
        if (this._group !== undefined) {
            hasAnyValues = true;
            internalValueResult.group = this._group;
        }
        if (this._launchType !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchType = this._launchType;
        }
        if (this._platformVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.platformVersion = this._platformVersion;
        }
        if (this._propagateTags !== undefined) {
            hasAnyValues = true;
            internalValueResult.propagateTags = this._propagateTags;
        }
        if (this._referenceId !== undefined) {
            hasAnyValues = true;
            internalValueResult.referenceId = this._referenceId;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._taskCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskCount = this._taskCount;
        }
        if (this._taskDefinitionArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskDefinitionArn = this._taskDefinitionArn;
        }
        if (this._capacityProviderStrategy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProviderStrategy = this._capacityProviderStrategy?.internalValue;
        }
        if (this._networkConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkConfiguration = this._networkConfiguration?.internalValue;
        }
        if (this._placementConstraints?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.placementConstraints = this._placementConstraints?.internalValue;
        }
        if (this._placementStrategy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.placementStrategy = this._placementStrategy?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableEcsManagedTags = undefined;
            this._enableExecuteCommand = undefined;
            this._group = undefined;
            this._launchType = undefined;
            this._platformVersion = undefined;
            this._propagateTags = undefined;
            this._referenceId = undefined;
            this._tags = undefined;
            this._taskCount = undefined;
            this._taskDefinitionArn = undefined;
            this._capacityProviderStrategy.internalValue = undefined;
            this._networkConfiguration.internalValue = undefined;
            this._placementConstraints.internalValue = undefined;
            this._placementStrategy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableEcsManagedTags = value.enableEcsManagedTags;
            this._enableExecuteCommand = value.enableExecuteCommand;
            this._group = value.group;
            this._launchType = value.launchType;
            this._platformVersion = value.platformVersion;
            this._propagateTags = value.propagateTags;
            this._referenceId = value.referenceId;
            this._tags = value.tags;
            this._taskCount = value.taskCount;
            this._taskDefinitionArn = value.taskDefinitionArn;
            this._capacityProviderStrategy.internalValue = value.capacityProviderStrategy;
            this._networkConfiguration.internalValue = value.networkConfiguration;
            this._placementConstraints.internalValue = value.placementConstraints;
            this._placementStrategy.internalValue = value.placementStrategy;
        }
    }
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    get group() {
        return this.getStringAttribute('group');
    }
    set group(value) {
        this._group = value;
    }
    resetGroup() {
        this._group = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupInput() {
        return this._group;
    }
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    set launchType(value) {
        this._launchType = value;
    }
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchTypeInput() {
        return this._launchType;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    set propagateTags(value) {
        this._propagateTags = value;
    }
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propagateTagsInput() {
        return this._propagateTags;
    }
    get referenceId() {
        return this.getStringAttribute('reference_id');
    }
    set referenceId(value) {
        this._referenceId = value;
    }
    resetReferenceId() {
        this._referenceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get referenceIdInput() {
        return this._referenceId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get taskCount() {
        return this.getNumberAttribute('task_count');
    }
    set taskCount(value) {
        this._taskCount = value;
    }
    resetTaskCount() {
        this._taskCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskCountInput() {
        return this._taskCount;
    }
    get taskDefinitionArn() {
        return this.getStringAttribute('task_definition_arn');
    }
    set taskDefinitionArn(value) {
        this._taskDefinitionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get taskDefinitionArnInput() {
        return this._taskDefinitionArn;
    }
    get capacityProviderStrategy() {
        return this._capacityProviderStrategy;
    }
    putCapacityProviderStrategy(value) {
        this._capacityProviderStrategy.internalValue = value;
    }
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get placementConstraints() {
        return this._placementConstraints;
    }
    putPlacementConstraints(value) {
        this._placementConstraints.internalValue = value;
    }
    resetPlacementConstraints() {
        this._placementConstraints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementConstraintsInput() {
        return this._placementConstraints.internalValue;
    }
    get placementStrategy() {
        return this._placementStrategy;
    }
    putPlacementStrategy(value) {
        this._placementStrategy.internalValue = value;
    }
    resetPlacementStrategy() {
        this._placementStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementStrategyInput() {
        return this._placementStrategy.internalValue;
    }
}
exports.SchedulerScheduleTargetEcsParametersOutputReference = SchedulerScheduleTargetEcsParametersOutputReference;
_k = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEcsParametersOutputReference[_k] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEcsParametersOutputReference", version: "19.15.0" };
function schedulerScheduleTargetEventbridgeParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        detail_type: cdktf.stringToTerraform(struct.detailType),
        source: cdktf.stringToTerraform(struct.source),
    };
}
exports.schedulerScheduleTargetEventbridgeParametersToTerraform = schedulerScheduleTargetEventbridgeParametersToTerraform;
function schedulerScheduleTargetEventbridgeParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        detail_type: {
            value: cdktf.stringToHclTerraform(struct.detailType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source: {
            value: cdktf.stringToHclTerraform(struct.source),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetEventbridgeParametersToHclTerraform = schedulerScheduleTargetEventbridgeParametersToHclTerraform;
class SchedulerScheduleTargetEventbridgeParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._detailType !== undefined) {
            hasAnyValues = true;
            internalValueResult.detailType = this._detailType;
        }
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._detailType = undefined;
            this._source = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._detailType = value.detailType;
            this._source = value.source;
        }
    }
    get detailType() {
        return this.getStringAttribute('detail_type');
    }
    set detailType(value) {
        this._detailType = value;
    }
    // Temporarily expose input value. Use with caution.
    get detailTypeInput() {
        return this._detailType;
    }
    get source() {
        return this.getStringAttribute('source');
    }
    set source(value) {
        this._source = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
}
exports.SchedulerScheduleTargetEventbridgeParametersOutputReference = SchedulerScheduleTargetEventbridgeParametersOutputReference;
_l = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetEventbridgeParametersOutputReference[_l] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetEventbridgeParametersOutputReference", version: "19.15.0" };
function schedulerScheduleTargetKinesisParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partition_key: cdktf.stringToTerraform(struct.partitionKey),
    };
}
exports.schedulerScheduleTargetKinesisParametersToTerraform = schedulerScheduleTargetKinesisParametersToTerraform;
function schedulerScheduleTargetKinesisParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        partition_key: {
            value: cdktf.stringToHclTerraform(struct.partitionKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetKinesisParametersToHclTerraform = schedulerScheduleTargetKinesisParametersToHclTerraform;
class SchedulerScheduleTargetKinesisParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._partitionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.partitionKey = this._partitionKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._partitionKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._partitionKey = value.partitionKey;
        }
    }
    get partitionKey() {
        return this.getStringAttribute('partition_key');
    }
    set partitionKey(value) {
        this._partitionKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get partitionKeyInput() {
        return this._partitionKey;
    }
}
exports.SchedulerScheduleTargetKinesisParametersOutputReference = SchedulerScheduleTargetKinesisParametersOutputReference;
_m = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetKinesisParametersOutputReference[_m] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetKinesisParametersOutputReference", version: "19.15.0" };
function schedulerScheduleTargetRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_event_age_in_seconds: cdktf.numberToTerraform(struct.maximumEventAgeInSeconds),
        maximum_retry_attempts: cdktf.numberToTerraform(struct.maximumRetryAttempts),
    };
}
exports.schedulerScheduleTargetRetryPolicyToTerraform = schedulerScheduleTargetRetryPolicyToTerraform;
function schedulerScheduleTargetRetryPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_event_age_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.maximumEventAgeInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        maximum_retry_attempts: {
            value: cdktf.numberToHclTerraform(struct.maximumRetryAttempts),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetRetryPolicyToHclTerraform = schedulerScheduleTargetRetryPolicyToHclTerraform;
class SchedulerScheduleTargetRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumEventAgeInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumEventAgeInSeconds = this._maximumEventAgeInSeconds;
        }
        if (this._maximumRetryAttempts !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumRetryAttempts = this._maximumRetryAttempts;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumEventAgeInSeconds = undefined;
            this._maximumRetryAttempts = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumEventAgeInSeconds = value.maximumEventAgeInSeconds;
            this._maximumRetryAttempts = value.maximumRetryAttempts;
        }
    }
    get maximumEventAgeInSeconds() {
        return this.getNumberAttribute('maximum_event_age_in_seconds');
    }
    set maximumEventAgeInSeconds(value) {
        this._maximumEventAgeInSeconds = value;
    }
    resetMaximumEventAgeInSeconds() {
        this._maximumEventAgeInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumEventAgeInSecondsInput() {
        return this._maximumEventAgeInSeconds;
    }
    get maximumRetryAttempts() {
        return this.getNumberAttribute('maximum_retry_attempts');
    }
    set maximumRetryAttempts(value) {
        this._maximumRetryAttempts = value;
    }
    resetMaximumRetryAttempts() {
        this._maximumRetryAttempts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumRetryAttemptsInput() {
        return this._maximumRetryAttempts;
    }
}
exports.SchedulerScheduleTargetRetryPolicyOutputReference = SchedulerScheduleTargetRetryPolicyOutputReference;
_o = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetRetryPolicyOutputReference[_o] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetRetryPolicyOutputReference", version: "19.15.0" };
function schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToTerraform = schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToTerraform;
function schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToHclTerraform = schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToHclTerraform;
class SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference = SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference;
_p = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference[_p] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference", version: "19.15.0" };
class SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList = SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList;
_q = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList[_q] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList", version: "19.15.0" };
function schedulerScheduleTargetSagemakerPipelineParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pipeline_parameter: cdktf.listMapper(schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToTerraform, true)(struct.pipelineParameter),
    };
}
exports.schedulerScheduleTargetSagemakerPipelineParametersToTerraform = schedulerScheduleTargetSagemakerPipelineParametersToTerraform;
function schedulerScheduleTargetSagemakerPipelineParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pipeline_parameter: {
            value: cdktf.listMapperHcl(schedulerScheduleTargetSagemakerPipelineParametersPipelineParameterToHclTerraform, true)(struct.pipelineParameter),
            isBlock: true,
            type: "set",
            storageClassType: "SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetSagemakerPipelineParametersToHclTerraform = schedulerScheduleTargetSagemakerPipelineParametersToHclTerraform;
class SchedulerScheduleTargetSagemakerPipelineParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // pipeline_parameter - computed: false, optional: true, required: false
        this._pipelineParameter = new SchedulerScheduleTargetSagemakerPipelineParametersPipelineParameterList(this, "pipeline_parameter", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pipelineParameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.pipelineParameter = this._pipelineParameter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._pipelineParameter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._pipelineParameter.internalValue = value.pipelineParameter;
        }
    }
    get pipelineParameter() {
        return this._pipelineParameter;
    }
    putPipelineParameter(value) {
        this._pipelineParameter.internalValue = value;
    }
    resetPipelineParameter() {
        this._pipelineParameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineParameterInput() {
        return this._pipelineParameter.internalValue;
    }
}
exports.SchedulerScheduleTargetSagemakerPipelineParametersOutputReference = SchedulerScheduleTargetSagemakerPipelineParametersOutputReference;
_r = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetSagemakerPipelineParametersOutputReference[_r] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetSagemakerPipelineParametersOutputReference", version: "19.15.0" };
function schedulerScheduleTargetSqsParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_group_id: cdktf.stringToTerraform(struct.messageGroupId),
    };
}
exports.schedulerScheduleTargetSqsParametersToTerraform = schedulerScheduleTargetSqsParametersToTerraform;
function schedulerScheduleTargetSqsParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        message_group_id: {
            value: cdktf.stringToHclTerraform(struct.messageGroupId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetSqsParametersToHclTerraform = schedulerScheduleTargetSqsParametersToHclTerraform;
class SchedulerScheduleTargetSqsParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.messageGroupId = this._messageGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._messageGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._messageGroupId = value.messageGroupId;
        }
    }
    get messageGroupId() {
        return this.getStringAttribute('message_group_id');
    }
    set messageGroupId(value) {
        this._messageGroupId = value;
    }
    resetMessageGroupId() {
        this._messageGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messageGroupIdInput() {
        return this._messageGroupId;
    }
}
exports.SchedulerScheduleTargetSqsParametersOutputReference = SchedulerScheduleTargetSqsParametersOutputReference;
_s = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetSqsParametersOutputReference[_s] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetSqsParametersOutputReference", version: "19.15.0" };
function schedulerScheduleTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        input: cdktf.stringToTerraform(struct.input),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        dead_letter_config: schedulerScheduleTargetDeadLetterConfigToTerraform(struct.deadLetterConfig),
        ecs_parameters: schedulerScheduleTargetEcsParametersToTerraform(struct.ecsParameters),
        eventbridge_parameters: schedulerScheduleTargetEventbridgeParametersToTerraform(struct.eventbridgeParameters),
        kinesis_parameters: schedulerScheduleTargetKinesisParametersToTerraform(struct.kinesisParameters),
        retry_policy: schedulerScheduleTargetRetryPolicyToTerraform(struct.retryPolicy),
        sagemaker_pipeline_parameters: schedulerScheduleTargetSagemakerPipelineParametersToTerraform(struct.sagemakerPipelineParameters),
        sqs_parameters: schedulerScheduleTargetSqsParametersToTerraform(struct.sqsParameters),
    };
}
exports.schedulerScheduleTargetToTerraform = schedulerScheduleTargetToTerraform;
function schedulerScheduleTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        input: {
            value: cdktf.stringToHclTerraform(struct.input),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        dead_letter_config: {
            value: schedulerScheduleTargetDeadLetterConfigToHclTerraform(struct.deadLetterConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetDeadLetterConfigList",
        },
        ecs_parameters: {
            value: schedulerScheduleTargetEcsParametersToHclTerraform(struct.ecsParameters),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetEcsParametersList",
        },
        eventbridge_parameters: {
            value: schedulerScheduleTargetEventbridgeParametersToHclTerraform(struct.eventbridgeParameters),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetEventbridgeParametersList",
        },
        kinesis_parameters: {
            value: schedulerScheduleTargetKinesisParametersToHclTerraform(struct.kinesisParameters),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetKinesisParametersList",
        },
        retry_policy: {
            value: schedulerScheduleTargetRetryPolicyToHclTerraform(struct.retryPolicy),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetRetryPolicyList",
        },
        sagemaker_pipeline_parameters: {
            value: schedulerScheduleTargetSagemakerPipelineParametersToHclTerraform(struct.sagemakerPipelineParameters),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetSagemakerPipelineParametersList",
        },
        sqs_parameters: {
            value: schedulerScheduleTargetSqsParametersToHclTerraform(struct.sqsParameters),
            isBlock: true,
            type: "list",
            storageClassType: "SchedulerScheduleTargetSqsParametersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.schedulerScheduleTargetToHclTerraform = schedulerScheduleTargetToHclTerraform;
class SchedulerScheduleTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // dead_letter_config - computed: false, optional: true, required: false
        this._deadLetterConfig = new SchedulerScheduleTargetDeadLetterConfigOutputReference(this, "dead_letter_config");
        // ecs_parameters - computed: false, optional: true, required: false
        this._ecsParameters = new SchedulerScheduleTargetEcsParametersOutputReference(this, "ecs_parameters");
        // eventbridge_parameters - computed: false, optional: true, required: false
        this._eventbridgeParameters = new SchedulerScheduleTargetEventbridgeParametersOutputReference(this, "eventbridge_parameters");
        // kinesis_parameters - computed: false, optional: true, required: false
        this._kinesisParameters = new SchedulerScheduleTargetKinesisParametersOutputReference(this, "kinesis_parameters");
        // retry_policy - computed: false, optional: true, required: false
        this._retryPolicy = new SchedulerScheduleTargetRetryPolicyOutputReference(this, "retry_policy");
        // sagemaker_pipeline_parameters - computed: false, optional: true, required: false
        this._sagemakerPipelineParameters = new SchedulerScheduleTargetSagemakerPipelineParametersOutputReference(this, "sagemaker_pipeline_parameters");
        // sqs_parameters - computed: false, optional: true, required: false
        this._sqsParameters = new SchedulerScheduleTargetSqsParametersOutputReference(this, "sqs_parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._input !== undefined) {
            hasAnyValues = true;
            internalValueResult.input = this._input;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._deadLetterConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.deadLetterConfig = this._deadLetterConfig?.internalValue;
        }
        if (this._ecsParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ecsParameters = this._ecsParameters?.internalValue;
        }
        if (this._eventbridgeParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventbridgeParameters = this._eventbridgeParameters?.internalValue;
        }
        if (this._kinesisParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisParameters = this._kinesisParameters?.internalValue;
        }
        if (this._retryPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retryPolicy = this._retryPolicy?.internalValue;
        }
        if (this._sagemakerPipelineParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sagemakerPipelineParameters = this._sagemakerPipelineParameters?.internalValue;
        }
        if (this._sqsParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqsParameters = this._sqsParameters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._input = undefined;
            this._roleArn = undefined;
            this._deadLetterConfig.internalValue = undefined;
            this._ecsParameters.internalValue = undefined;
            this._eventbridgeParameters.internalValue = undefined;
            this._kinesisParameters.internalValue = undefined;
            this._retryPolicy.internalValue = undefined;
            this._sagemakerPipelineParameters.internalValue = undefined;
            this._sqsParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._input = value.input;
            this._roleArn = value.roleArn;
            this._deadLetterConfig.internalValue = value.deadLetterConfig;
            this._ecsParameters.internalValue = value.ecsParameters;
            this._eventbridgeParameters.internalValue = value.eventbridgeParameters;
            this._kinesisParameters.internalValue = value.kinesisParameters;
            this._retryPolicy.internalValue = value.retryPolicy;
            this._sagemakerPipelineParameters.internalValue = value.sagemakerPipelineParameters;
            this._sqsParameters.internalValue = value.sqsParameters;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get input() {
        return this.getStringAttribute('input');
    }
    set input(value) {
        this._input = value;
    }
    resetInput() {
        this._input = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputInput() {
        return this._input;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get deadLetterConfig() {
        return this._deadLetterConfig;
    }
    putDeadLetterConfig(value) {
        this._deadLetterConfig.internalValue = value;
    }
    resetDeadLetterConfig() {
        this._deadLetterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deadLetterConfigInput() {
        return this._deadLetterConfig.internalValue;
    }
    get ecsParameters() {
        return this._ecsParameters;
    }
    putEcsParameters(value) {
        this._ecsParameters.internalValue = value;
    }
    resetEcsParameters() {
        this._ecsParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ecsParametersInput() {
        return this._ecsParameters.internalValue;
    }
    get eventbridgeParameters() {
        return this._eventbridgeParameters;
    }
    putEventbridgeParameters(value) {
        this._eventbridgeParameters.internalValue = value;
    }
    resetEventbridgeParameters() {
        this._eventbridgeParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventbridgeParametersInput() {
        return this._eventbridgeParameters.internalValue;
    }
    get kinesisParameters() {
        return this._kinesisParameters;
    }
    putKinesisParameters(value) {
        this._kinesisParameters.internalValue = value;
    }
    resetKinesisParameters() {
        this._kinesisParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisParametersInput() {
        return this._kinesisParameters.internalValue;
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    putRetryPolicy(value) {
        this._retryPolicy.internalValue = value;
    }
    resetRetryPolicy() {
        this._retryPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retryPolicyInput() {
        return this._retryPolicy.internalValue;
    }
    get sagemakerPipelineParameters() {
        return this._sagemakerPipelineParameters;
    }
    putSagemakerPipelineParameters(value) {
        this._sagemakerPipelineParameters.internalValue = value;
    }
    resetSagemakerPipelineParameters() {
        this._sagemakerPipelineParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sagemakerPipelineParametersInput() {
        return this._sagemakerPipelineParameters.internalValue;
    }
    get sqsParameters() {
        return this._sqsParameters;
    }
    putSqsParameters(value) {
        this._sqsParameters.internalValue = value;
    }
    resetSqsParameters() {
        this._sqsParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqsParametersInput() {
        return this._sqsParameters.internalValue;
    }
}
exports.SchedulerScheduleTargetOutputReference = SchedulerScheduleTargetOutputReference;
_t = JSII_RTTI_SYMBOL_1;
SchedulerScheduleTargetOutputReference[_t] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerScheduleTargetOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/scheduler_schedule aws_scheduler_schedule}
*/
class SchedulerSchedule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SchedulerSchedule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SchedulerSchedule to import
    * @param importFromId The id of the existing SchedulerSchedule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/scheduler_schedule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SchedulerSchedule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_scheduler_schedule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/scheduler_schedule aws_scheduler_schedule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SchedulerScheduleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_scheduler_schedule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // flexible_time_window - computed: false, optional: false, required: true
        this._flexibleTimeWindow = new SchedulerScheduleFlexibleTimeWindowOutputReference(this, "flexible_time_window");
        // target - computed: false, optional: false, required: true
        this._target = new SchedulerScheduleTargetOutputReference(this, "target");
        this._description = config.description;
        this._endDate = config.endDate;
        this._groupName = config.groupName;
        this._id = config.id;
        this._kmsKeyArn = config.kmsKeyArn;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._scheduleExpression = config.scheduleExpression;
        this._scheduleExpressionTimezone = config.scheduleExpressionTimezone;
        this._startDate = config.startDate;
        this._state = config.state;
        this._flexibleTimeWindow.internalValue = config.flexibleTimeWindow;
        this._target.internalValue = config.target;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    set endDate(value) {
        this._endDate = value;
    }
    resetEndDate() {
        this._endDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endDateInput() {
        return this._endDate;
    }
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    set groupName(value) {
        this._groupName = value;
    }
    resetGroupName() {
        this._groupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupNameInput() {
        return this._groupName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get scheduleExpression() {
        return this.getStringAttribute('schedule_expression');
    }
    set scheduleExpression(value) {
        this._scheduleExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleExpressionInput() {
        return this._scheduleExpression;
    }
    get scheduleExpressionTimezone() {
        return this.getStringAttribute('schedule_expression_timezone');
    }
    set scheduleExpressionTimezone(value) {
        this._scheduleExpressionTimezone = value;
    }
    resetScheduleExpressionTimezone() {
        this._scheduleExpressionTimezone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleExpressionTimezoneInput() {
        return this._scheduleExpressionTimezone;
    }
    get startDate() {
        return this.getStringAttribute('start_date');
    }
    set startDate(value) {
        this._startDate = value;
    }
    resetStartDate() {
        this._startDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startDateInput() {
        return this._startDate;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get flexibleTimeWindow() {
        return this._flexibleTimeWindow;
    }
    putFlexibleTimeWindow(value) {
        this._flexibleTimeWindow.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get flexibleTimeWindowInput() {
        return this._flexibleTimeWindow.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            end_date: cdktf.stringToTerraform(this._endDate),
            group_name: cdktf.stringToTerraform(this._groupName),
            id: cdktf.stringToTerraform(this._id),
            kms_key_arn: cdktf.stringToTerraform(this._kmsKeyArn),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            schedule_expression: cdktf.stringToTerraform(this._scheduleExpression),
            schedule_expression_timezone: cdktf.stringToTerraform(this._scheduleExpressionTimezone),
            start_date: cdktf.stringToTerraform(this._startDate),
            state: cdktf.stringToTerraform(this._state),
            flexible_time_window: schedulerScheduleFlexibleTimeWindowToTerraform(this._flexibleTimeWindow.internalValue),
            target: schedulerScheduleTargetToTerraform(this._target.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            end_date: {
                value: cdktf.stringToHclTerraform(this._endDate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            group_name: {
                value: cdktf.stringToHclTerraform(this._groupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_arn: {
                value: cdktf.stringToHclTerraform(this._kmsKeyArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            schedule_expression: {
                value: cdktf.stringToHclTerraform(this._scheduleExpression),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            schedule_expression_timezone: {
                value: cdktf.stringToHclTerraform(this._scheduleExpressionTimezone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            start_date: {
                value: cdktf.stringToHclTerraform(this._startDate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            state: {
                value: cdktf.stringToHclTerraform(this._state),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            flexible_time_window: {
                value: schedulerScheduleFlexibleTimeWindowToHclTerraform(this._flexibleTimeWindow.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SchedulerScheduleFlexibleTimeWindowList",
            },
            target: {
                value: schedulerScheduleTargetToHclTerraform(this._target.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SchedulerScheduleTargetList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SchedulerSchedule = SchedulerSchedule;
_u = JSII_RTTI_SYMBOL_1;
SchedulerSchedule[_u] = { fqn: "@cdktf/provider-aws.schedulerSchedule.SchedulerSchedule", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SchedulerSchedule.tfResourceType = "aws_scheduler_schedule";
