// @winglang/sdk/lib/std/set.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MutSet = exports.Set = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../core");
/**
 * Immutable Set
 *
 * @typeparam T1
 */
class Set {
    /**
     * @internal
     */
    static _toInflightType() {
        return core_1.InflightClient.forType(__filename, this.name);
    }
    constructor() { }
    /**
     * The length of the set
     * @returns the length of the set
     */
    get size() {
        throw new Error("Abstract");
    }
    /**
     * Returns a boolean indicating whether an element with the specified value exists in the set.
     * @param value The value to test for presence in the Set object.
     * @returns `true` if an element with the specified value exists in the set; otherwise `false`.
     */
    has(value) {
        value;
        throw new Error("Abstract");
    }
    /**
     * Create a mutable shallow copy of this set
     *
     * @macro new Set($self$)
     *
     * @returns a MutableSet with the same values as this set
     */
    copyMut() {
        throw new Error("Macro");
    }
    /**
     * Create an immutable array shallow copy of this set
     *
     * @macro [...($self$)]
     *
     * @returns an ImmutableArray with the same values as this set
     */
    toArray() {
        throw new Error("Macro");
    }
}
exports.Set = Set;
_a = JSII_RTTI_SYMBOL_1;
Set[_a] = { fqn: "@winglang/sdk.std.Set", version: "0.0.0" };
/**
 * Mutable Set
 *
 * @typeparam T1
 */
class MutSet {
    /**
     * @internal
     */
    static _toInflightType() {
        return core_1.InflightClient.forType(__filename, this.name);
    }
    constructor() { }
    /**
     * The length of the set
     * @returns the length of the set
     */
    get size() {
        throw new Error("Abstract");
    }
    /**
     * Add value to set
     * @param value value to add
     * @returns true if the value was added, false if it was already in the set
     */
    add(value) {
        value;
        throw new Error("Abstract");
    }
    /**
     * The clear() method removes all elements from a set.
     */
    clear() {
        throw new Error("Abstract");
    }
    /**
     * Create an immutable shallow copy of this set
     *
     * @macro new Set($self$)
     *
     * @returns an ImmutableSet with the same values as this set
     */
    copy() {
        throw new Error("Macro");
    }
    /**
     * Removes a specified value from a set, if it is in the set.
     * @param value The value to remove from the set.
     * @returns Returns `true` if `value` was already in the set; otherwise `false`.
     */
    delete(value) {
        value;
        throw new Error("Abstract");
    }
    /**
     * Returns a boolean indicating whether an element with the specified value exists in the set.
     * @param value The value to test for presence in the Set object.
     * @returns `true` if an element with the specified value exists in the set; otherwise `false`.
     */
    has(value) {
        value;
        throw new Error("Abstract");
    }
    /**
     * Create an immutable array shallow copy of this set
     *
     * @macro [...($self$)]
     *
     * @returns an ImmutableArray with the same values as this set
     */
    toArray() {
        throw new Error("Macro");
    }
}
exports.MutSet = MutSet;
_b = JSII_RTTI_SYMBOL_1;
MutSet[_b] = { fqn: "@winglang/sdk.std.MutSet", version: "0.0.0" };
