// @winglang/sdk/lib/simulator/serialization.js
"use strict";
// TODO: more robust serialization scheme
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserialize = exports.serialize = void 0;
const datetime_1 = require("../std/datetime");
function serialize(input) {
    return JSON.stringify(input, (_key, value) => {
        if (value instanceof datetime_1.Datetime) {
            return {
                $kind: "datetime",
                day: value.dayOfMonth,
                hour: value.hours,
                min: value.min,
                month: value.month,
                sec: value.sec,
                year: value.year,
                ms: value.ms,
                tz: value.timezone,
            };
        }
        return value;
    });
}
exports.serialize = serialize;
function deserialize(input) {
    return JSON.parse(input, (_key, value) => {
        // assumption: Wing APIs don't distinguish between null and undefined, so we can swap them
        if (value === null) {
            return undefined;
        }
        if (value.$kind === "datetime") {
            return datetime_1.Datetime.fromComponents({
                day: value.day,
                hour: value.hour,
                min: value.min,
                month: value.month,
                sec: value.sec,
                year: value.year,
                ms: value.ms,
                tz: value.tz,
            });
        }
        return value;
    });
}
exports.deserialize = deserialize;
