// @cdktf/provider-aws/lib/data-aws-batch-scheduling-policy/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBatchSchedulingPolicy = exports.DataAwsBatchSchedulingPolicyFairSharePolicyList = exports.DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference = exports.dataAwsBatchSchedulingPolicyFairSharePolicyToHclTerraform = exports.dataAwsBatchSchedulingPolicyFairSharePolicyToTerraform = exports.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList = exports.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference = exports.dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToHclTerraform = exports.dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToTerraform = dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToTerraform;
function dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToHclTerraform = dataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionToHclTerraform;
class DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // share_identifier - computed: true, optional: false, required: false
    get shareIdentifier() {
        return this.getStringAttribute('share_identifier');
    }
    // weight_factor - computed: true, optional: false, required: false
    get weightFactor() {
        return this.getNumberAttribute('weight_factor');
    }
}
exports.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference = DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsBatchSchedulingPolicy.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference", version: "19.15.0" };
class DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList = DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList[_b] = { fqn: "@cdktf/provider-aws.dataAwsBatchSchedulingPolicy.DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList", version: "19.15.0" };
function dataAwsBatchSchedulingPolicyFairSharePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBatchSchedulingPolicyFairSharePolicyToTerraform = dataAwsBatchSchedulingPolicyFairSharePolicyToTerraform;
function dataAwsBatchSchedulingPolicyFairSharePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBatchSchedulingPolicyFairSharePolicyToHclTerraform = dataAwsBatchSchedulingPolicyFairSharePolicyToHclTerraform;
class DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // share_distribution - computed: true, optional: false, required: false
        this._shareDistribution = new DataAwsBatchSchedulingPolicyFairSharePolicyShareDistributionList(this, "share_distribution", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // compute_reservation - computed: true, optional: false, required: false
    get computeReservation() {
        return this.getNumberAttribute('compute_reservation');
    }
    // share_decay_seconds - computed: true, optional: false, required: false
    get shareDecaySeconds() {
        return this.getNumberAttribute('share_decay_seconds');
    }
    get shareDistribution() {
        return this._shareDistribution;
    }
}
exports.DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference = DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsBatchSchedulingPolicy.DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference", version: "19.15.0" };
class DataAwsBatchSchedulingPolicyFairSharePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBatchSchedulingPolicyFairSharePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBatchSchedulingPolicyFairSharePolicyList = DataAwsBatchSchedulingPolicyFairSharePolicyList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsBatchSchedulingPolicyFairSharePolicyList[_d] = { fqn: "@cdktf/provider-aws.dataAwsBatchSchedulingPolicy.DataAwsBatchSchedulingPolicyFairSharePolicyList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/batch_scheduling_policy aws_batch_scheduling_policy}
*/
class DataAwsBatchSchedulingPolicy extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsBatchSchedulingPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsBatchSchedulingPolicy to import
    * @param importFromId The id of the existing DataAwsBatchSchedulingPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/batch_scheduling_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsBatchSchedulingPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_batch_scheduling_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/batch_scheduling_policy aws_batch_scheduling_policy} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsBatchSchedulingPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_batch_scheduling_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // fair_share_policy - computed: true, optional: false, required: false
        this._fairSharePolicy = new DataAwsBatchSchedulingPolicyFairSharePolicyList(this, "fair_share_policy", false);
        this._arn = config.arn;
        this._id = config.id;
        this._tags = config.tags;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get fairSharePolicy() {
        return this._fairSharePolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsBatchSchedulingPolicy = DataAwsBatchSchedulingPolicy;
_e = JSII_RTTI_SYMBOL_1;
DataAwsBatchSchedulingPolicy[_e] = { fqn: "@cdktf/provider-aws.dataAwsBatchSchedulingPolicy.DataAwsBatchSchedulingPolicy", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsBatchSchedulingPolicy.tfResourceType = "aws_batch_scheduling_policy";
