// @cdktf/provider-aws/lib/ecs-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcsServiceServiceConnectConfigurationServiceOutputReference = exports.ecsServiceServiceConnectConfigurationServiceToHclTerraform = exports.ecsServiceServiceConnectConfigurationServiceToTerraform = exports.EcsServiceServiceConnectConfigurationServiceTlsOutputReference = exports.ecsServiceServiceConnectConfigurationServiceTlsToHclTerraform = exports.ecsServiceServiceConnectConfigurationServiceTlsToTerraform = exports.EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference = exports.ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToHclTerraform = exports.ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToTerraform = exports.EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference = exports.ecsServiceServiceConnectConfigurationServiceTimeoutToHclTerraform = exports.ecsServiceServiceConnectConfigurationServiceTimeoutToTerraform = exports.EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference = exports.ecsServiceServiceConnectConfigurationServiceClientAliasToHclTerraform = exports.ecsServiceServiceConnectConfigurationServiceClientAliasToTerraform = exports.EcsServiceServiceConnectConfigurationLogConfigurationOutputReference = exports.ecsServiceServiceConnectConfigurationLogConfigurationToHclTerraform = exports.ecsServiceServiceConnectConfigurationLogConfigurationToTerraform = exports.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList = exports.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference = exports.ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToHclTerraform = exports.ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToTerraform = exports.EcsServicePlacementConstraintsList = exports.EcsServicePlacementConstraintsOutputReference = exports.ecsServicePlacementConstraintsToHclTerraform = exports.ecsServicePlacementConstraintsToTerraform = exports.EcsServiceOrderedPlacementStrategyList = exports.EcsServiceOrderedPlacementStrategyOutputReference = exports.ecsServiceOrderedPlacementStrategyToHclTerraform = exports.ecsServiceOrderedPlacementStrategyToTerraform = exports.EcsServiceNetworkConfigurationOutputReference = exports.ecsServiceNetworkConfigurationToHclTerraform = exports.ecsServiceNetworkConfigurationToTerraform = exports.EcsServiceLoadBalancerList = exports.EcsServiceLoadBalancerOutputReference = exports.ecsServiceLoadBalancerToHclTerraform = exports.ecsServiceLoadBalancerToTerraform = exports.EcsServiceDeploymentControllerOutputReference = exports.ecsServiceDeploymentControllerToHclTerraform = exports.ecsServiceDeploymentControllerToTerraform = exports.EcsServiceDeploymentCircuitBreakerOutputReference = exports.ecsServiceDeploymentCircuitBreakerToHclTerraform = exports.ecsServiceDeploymentCircuitBreakerToTerraform = exports.EcsServiceCapacityProviderStrategyList = exports.EcsServiceCapacityProviderStrategyOutputReference = exports.ecsServiceCapacityProviderStrategyToHclTerraform = exports.ecsServiceCapacityProviderStrategyToTerraform = exports.EcsServiceAlarmsOutputReference = exports.ecsServiceAlarmsToHclTerraform = exports.ecsServiceAlarmsToTerraform = void 0;
exports.EcsService = exports.EcsServiceTimeoutsOutputReference = exports.ecsServiceTimeoutsToHclTerraform = exports.ecsServiceTimeoutsToTerraform = exports.EcsServiceServiceRegistriesOutputReference = exports.ecsServiceServiceRegistriesToHclTerraform = exports.ecsServiceServiceRegistriesToTerraform = exports.EcsServiceServiceConnectConfigurationOutputReference = exports.ecsServiceServiceConnectConfigurationToHclTerraform = exports.ecsServiceServiceConnectConfigurationToTerraform = exports.EcsServiceServiceConnectConfigurationServiceList = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecsServiceAlarmsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        alarm_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.alarmNames),
        enable: cdktf.booleanToTerraform(struct.enable),
        rollback: cdktf.booleanToTerraform(struct.rollback),
    };
}
exports.ecsServiceAlarmsToTerraform = ecsServiceAlarmsToTerraform;
function ecsServiceAlarmsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        alarm_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.alarmNames),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        enable: {
            value: cdktf.booleanToHclTerraform(struct.enable),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        rollback: {
            value: cdktf.booleanToHclTerraform(struct.rollback),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceAlarmsToHclTerraform = ecsServiceAlarmsToHclTerraform;
class EcsServiceAlarmsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._alarmNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.alarmNames = this._alarmNames;
        }
        if (this._enable !== undefined) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._rollback !== undefined) {
            hasAnyValues = true;
            internalValueResult.rollback = this._rollback;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._alarmNames = undefined;
            this._enable = undefined;
            this._rollback = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._alarmNames = value.alarmNames;
            this._enable = value.enable;
            this._rollback = value.rollback;
        }
    }
    get alarmNames() {
        return cdktf.Fn.tolist(this.getListAttribute('alarm_names'));
    }
    set alarmNames(value) {
        this._alarmNames = value;
    }
    // Temporarily expose input value. Use with caution.
    get alarmNamesInput() {
        return this._alarmNames;
    }
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    get enableInput() {
        return this._enable;
    }
    get rollback() {
        return this.getBooleanAttribute('rollback');
    }
    set rollback(value) {
        this._rollback = value;
    }
    // Temporarily expose input value. Use with caution.
    get rollbackInput() {
        return this._rollback;
    }
}
exports.EcsServiceAlarmsOutputReference = EcsServiceAlarmsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcsServiceAlarmsOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceAlarmsOutputReference", version: "19.15.0" };
function ecsServiceCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.ecsServiceCapacityProviderStrategyToTerraform = ecsServiceCapacityProviderStrategyToTerraform;
function ecsServiceCapacityProviderStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        base: {
            value: cdktf.numberToHclTerraform(struct.base),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        capacity_provider: {
            value: cdktf.stringToHclTerraform(struct.capacityProvider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weight: {
            value: cdktf.numberToHclTerraform(struct.weight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceCapacityProviderStrategyToHclTerraform = ecsServiceCapacityProviderStrategyToHclTerraform;
class EcsServiceCapacityProviderStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._base !== undefined) {
            hasAnyValues = true;
            internalValueResult.base = this._base;
        }
        if (this._capacityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProvider = this._capacityProvider;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._base = undefined;
            this._capacityProvider = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._base = value.base;
            this._capacityProvider = value.capacityProvider;
            this._weight = value.weight;
        }
    }
    get base() {
        return this.getNumberAttribute('base');
    }
    set base(value) {
        this._base = value;
    }
    resetBase() {
        this._base = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseInput() {
        return this._base;
    }
    get capacityProvider() {
        return this.getStringAttribute('capacity_provider');
    }
    set capacityProvider(value) {
        this._capacityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderInput() {
        return this._capacityProvider;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.EcsServiceCapacityProviderStrategyOutputReference = EcsServiceCapacityProviderStrategyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EcsServiceCapacityProviderStrategyOutputReference[_b] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceCapacityProviderStrategyOutputReference", version: "19.15.0" };
class EcsServiceCapacityProviderStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceCapacityProviderStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceCapacityProviderStrategyList = EcsServiceCapacityProviderStrategyList;
_c = JSII_RTTI_SYMBOL_1;
EcsServiceCapacityProviderStrategyList[_c] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceCapacityProviderStrategyList", version: "19.15.0" };
function ecsServiceDeploymentCircuitBreakerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable: cdktf.booleanToTerraform(struct.enable),
        rollback: cdktf.booleanToTerraform(struct.rollback),
    };
}
exports.ecsServiceDeploymentCircuitBreakerToTerraform = ecsServiceDeploymentCircuitBreakerToTerraform;
function ecsServiceDeploymentCircuitBreakerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable: {
            value: cdktf.booleanToHclTerraform(struct.enable),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        rollback: {
            value: cdktf.booleanToHclTerraform(struct.rollback),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceDeploymentCircuitBreakerToHclTerraform = ecsServiceDeploymentCircuitBreakerToHclTerraform;
class EcsServiceDeploymentCircuitBreakerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enable !== undefined) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._rollback !== undefined) {
            hasAnyValues = true;
            internalValueResult.rollback = this._rollback;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enable = undefined;
            this._rollback = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enable = value.enable;
            this._rollback = value.rollback;
        }
    }
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    get enableInput() {
        return this._enable;
    }
    get rollback() {
        return this.getBooleanAttribute('rollback');
    }
    set rollback(value) {
        this._rollback = value;
    }
    // Temporarily expose input value. Use with caution.
    get rollbackInput() {
        return this._rollback;
    }
}
exports.EcsServiceDeploymentCircuitBreakerOutputReference = EcsServiceDeploymentCircuitBreakerOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentCircuitBreakerOutputReference[_d] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceDeploymentCircuitBreakerOutputReference", version: "19.15.0" };
function ecsServiceDeploymentControllerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceDeploymentControllerToTerraform = ecsServiceDeploymentControllerToTerraform;
function ecsServiceDeploymentControllerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceDeploymentControllerToHclTerraform = ecsServiceDeploymentControllerToHclTerraform;
class EcsServiceDeploymentControllerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServiceDeploymentControllerOutputReference = EcsServiceDeploymentControllerOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentControllerOutputReference[_e] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceDeploymentControllerOutputReference", version: "19.15.0" };
function ecsServiceLoadBalancerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        elb_name: cdktf.stringToTerraform(struct.elbName),
        target_group_arn: cdktf.stringToTerraform(struct.targetGroupArn),
    };
}
exports.ecsServiceLoadBalancerToTerraform = ecsServiceLoadBalancerToTerraform;
function ecsServiceLoadBalancerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_name: {
            value: cdktf.stringToHclTerraform(struct.containerName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_port: {
            value: cdktf.numberToHclTerraform(struct.containerPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        elb_name: {
            value: cdktf.stringToHclTerraform(struct.elbName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_group_arn: {
            value: cdktf.stringToHclTerraform(struct.targetGroupArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceLoadBalancerToHclTerraform = ecsServiceLoadBalancerToHclTerraform;
class EcsServiceLoadBalancerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerName = this._containerName;
        }
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._elbName !== undefined) {
            hasAnyValues = true;
            internalValueResult.elbName = this._elbName;
        }
        if (this._targetGroupArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetGroupArn = this._targetGroupArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerName = undefined;
            this._containerPort = undefined;
            this._elbName = undefined;
            this._targetGroupArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerName = value.containerName;
            this._containerPort = value.containerPort;
            this._elbName = value.elbName;
            this._targetGroupArn = value.targetGroupArn;
        }
    }
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    set containerName(value) {
        this._containerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerNameInput() {
        return this._containerName;
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get elbName() {
        return this.getStringAttribute('elb_name');
    }
    set elbName(value) {
        this._elbName = value;
    }
    resetElbName() {
        this._elbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elbNameInput() {
        return this._elbName;
    }
    get targetGroupArn() {
        return this.getStringAttribute('target_group_arn');
    }
    set targetGroupArn(value) {
        this._targetGroupArn = value;
    }
    resetTargetGroupArn() {
        this._targetGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetGroupArnInput() {
        return this._targetGroupArn;
    }
}
exports.EcsServiceLoadBalancerOutputReference = EcsServiceLoadBalancerOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EcsServiceLoadBalancerOutputReference[_f] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceLoadBalancerOutputReference", version: "19.15.0" };
class EcsServiceLoadBalancerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceLoadBalancerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceLoadBalancerList = EcsServiceLoadBalancerList;
_g = JSII_RTTI_SYMBOL_1;
EcsServiceLoadBalancerList[_g] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceLoadBalancerList", version: "19.15.0" };
function ecsServiceNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.ecsServiceNetworkConfigurationToTerraform = ecsServiceNetworkConfigurationToTerraform;
function ecsServiceNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        assign_public_ip: {
            value: cdktf.booleanToHclTerraform(struct.assignPublicIp),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        security_groups: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceNetworkConfigurationToHclTerraform = ecsServiceNetworkConfigurationToHclTerraform;
class EcsServiceNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.EcsServiceNetworkConfigurationOutputReference = EcsServiceNetworkConfigurationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EcsServiceNetworkConfigurationOutputReference[_h] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceNetworkConfigurationOutputReference", version: "19.15.0" };
function ecsServiceOrderedPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceOrderedPlacementStrategyToTerraform = ecsServiceOrderedPlacementStrategyToTerraform;
function ecsServiceOrderedPlacementStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceOrderedPlacementStrategyToHclTerraform = ecsServiceOrderedPlacementStrategyToHclTerraform;
class EcsServiceOrderedPlacementStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._field = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._field = value.field;
            this._type = value.type;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServiceOrderedPlacementStrategyOutputReference = EcsServiceOrderedPlacementStrategyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
EcsServiceOrderedPlacementStrategyOutputReference[_j] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceOrderedPlacementStrategyOutputReference", version: "19.15.0" };
class EcsServiceOrderedPlacementStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceOrderedPlacementStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceOrderedPlacementStrategyList = EcsServiceOrderedPlacementStrategyList;
_k = JSII_RTTI_SYMBOL_1;
EcsServiceOrderedPlacementStrategyList[_k] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceOrderedPlacementStrategyList", version: "19.15.0" };
function ecsServicePlacementConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServicePlacementConstraintsToTerraform = ecsServicePlacementConstraintsToTerraform;
function ecsServicePlacementConstraintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServicePlacementConstraintsToHclTerraform = ecsServicePlacementConstraintsToHclTerraform;
class EcsServicePlacementConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expression = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expression = value.expression;
            this._type = value.type;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    resetExpression() {
        this._expression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServicePlacementConstraintsOutputReference = EcsServicePlacementConstraintsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
EcsServicePlacementConstraintsOutputReference[_l] = { fqn: "@cdktf/provider-aws.ecsService.EcsServicePlacementConstraintsOutputReference", version: "19.15.0" };
class EcsServicePlacementConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServicePlacementConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServicePlacementConstraintsList = EcsServicePlacementConstraintsList;
_m = JSII_RTTI_SYMBOL_1;
EcsServicePlacementConstraintsList[_m] = { fqn: "@cdktf/provider-aws.ecsService.EcsServicePlacementConstraintsList", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value_from: cdktf.stringToTerraform(struct.valueFrom),
    };
}
exports.ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToTerraform = ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToTerraform;
function ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value_from: {
            value: cdktf.stringToHclTerraform(struct.valueFrom),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToHclTerraform = ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToHclTerraform;
class EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._valueFrom !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueFrom = this._valueFrom;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._valueFrom = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._valueFrom = value.valueFrom;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get valueFrom() {
        return this.getStringAttribute('value_from');
    }
    set valueFrom(value) {
        this._valueFrom = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueFromInput() {
        return this._valueFrom;
    }
}
exports.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference = EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference;
_o = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference[_o] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference", version: "19.15.0" };
class EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList = EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList;
_p = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList[_p] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationLogConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_driver: cdktf.stringToTerraform(struct.logDriver),
        options: cdktf.hashMapper(cdktf.stringToTerraform)(struct.options),
        secret_option: cdktf.listMapper(ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToTerraform, true)(struct.secretOption),
    };
}
exports.ecsServiceServiceConnectConfigurationLogConfigurationToTerraform = ecsServiceServiceConnectConfigurationLogConfigurationToTerraform;
function ecsServiceServiceConnectConfigurationLogConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        log_driver: {
            value: cdktf.stringToHclTerraform(struct.logDriver),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        options: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.options),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        secret_option: {
            value: cdktf.listMapperHcl(ecsServiceServiceConnectConfigurationLogConfigurationSecretOptionToHclTerraform, true)(struct.secretOption),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationLogConfigurationToHclTerraform = ecsServiceServiceConnectConfigurationLogConfigurationToHclTerraform;
class EcsServiceServiceConnectConfigurationLogConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // secret_option - computed: false, optional: true, required: false
        this._secretOption = new EcsServiceServiceConnectConfigurationLogConfigurationSecretOptionList(this, "secret_option", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logDriver !== undefined) {
            hasAnyValues = true;
            internalValueResult.logDriver = this._logDriver;
        }
        if (this._options !== undefined) {
            hasAnyValues = true;
            internalValueResult.options = this._options;
        }
        if (this._secretOption?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretOption = this._secretOption?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logDriver = undefined;
            this._options = undefined;
            this._secretOption.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logDriver = value.logDriver;
            this._options = value.options;
            this._secretOption.internalValue = value.secretOption;
        }
    }
    get logDriver() {
        return this.getStringAttribute('log_driver');
    }
    set logDriver(value) {
        this._logDriver = value;
    }
    // Temporarily expose input value. Use with caution.
    get logDriverInput() {
        return this._logDriver;
    }
    get options() {
        return this.getStringMapAttribute('options');
    }
    set options(value) {
        this._options = value;
    }
    resetOptions() {
        this._options = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options;
    }
    get secretOption() {
        return this._secretOption;
    }
    putSecretOption(value) {
        this._secretOption.internalValue = value;
    }
    resetSecretOption() {
        this._secretOption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretOptionInput() {
        return this._secretOption.internalValue;
    }
}
exports.EcsServiceServiceConnectConfigurationLogConfigurationOutputReference = EcsServiceServiceConnectConfigurationLogConfigurationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationLogConfigurationOutputReference[_q] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationLogConfigurationOutputReference", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationServiceClientAliasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dns_name: cdktf.stringToTerraform(struct.dnsName),
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.ecsServiceServiceConnectConfigurationServiceClientAliasToTerraform = ecsServiceServiceConnectConfigurationServiceClientAliasToTerraform;
function ecsServiceServiceConnectConfigurationServiceClientAliasToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dns_name: {
            value: cdktf.stringToHclTerraform(struct.dnsName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationServiceClientAliasToHclTerraform = ecsServiceServiceConnectConfigurationServiceClientAliasToHclTerraform;
class EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dnsName !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsName = this._dnsName;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dnsName = undefined;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dnsName = value.dnsName;
            this._port = value.port;
        }
    }
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    set dnsName(value) {
        this._dnsName = value;
    }
    resetDnsName() {
        this._dnsName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsNameInput() {
        return this._dnsName;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference = EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference;
_r = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference[_r] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationServiceTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle_timeout_seconds: cdktf.numberToTerraform(struct.idleTimeoutSeconds),
        per_request_timeout_seconds: cdktf.numberToTerraform(struct.perRequestTimeoutSeconds),
    };
}
exports.ecsServiceServiceConnectConfigurationServiceTimeoutToTerraform = ecsServiceServiceConnectConfigurationServiceTimeoutToTerraform;
function ecsServiceServiceConnectConfigurationServiceTimeoutToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        idle_timeout_seconds: {
            value: cdktf.numberToHclTerraform(struct.idleTimeoutSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        per_request_timeout_seconds: {
            value: cdktf.numberToHclTerraform(struct.perRequestTimeoutSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationServiceTimeoutToHclTerraform = ecsServiceServiceConnectConfigurationServiceTimeoutToHclTerraform;
class EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idleTimeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.idleTimeoutSeconds = this._idleTimeoutSeconds;
        }
        if (this._perRequestTimeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRequestTimeoutSeconds = this._perRequestTimeoutSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idleTimeoutSeconds = undefined;
            this._perRequestTimeoutSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idleTimeoutSeconds = value.idleTimeoutSeconds;
            this._perRequestTimeoutSeconds = value.perRequestTimeoutSeconds;
        }
    }
    get idleTimeoutSeconds() {
        return this.getNumberAttribute('idle_timeout_seconds');
    }
    set idleTimeoutSeconds(value) {
        this._idleTimeoutSeconds = value;
    }
    resetIdleTimeoutSeconds() {
        this._idleTimeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleTimeoutSecondsInput() {
        return this._idleTimeoutSeconds;
    }
    get perRequestTimeoutSeconds() {
        return this.getNumberAttribute('per_request_timeout_seconds');
    }
    set perRequestTimeoutSeconds(value) {
        this._perRequestTimeoutSeconds = value;
    }
    resetPerRequestTimeoutSeconds() {
        this._perRequestTimeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get perRequestTimeoutSecondsInput() {
        return this._perRequestTimeoutSeconds;
    }
}
exports.EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference = EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference;
_s = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference[_s] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        aws_pca_authority_arn: cdktf.stringToTerraform(struct.awsPcaAuthorityArn),
    };
}
exports.ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToTerraform = ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToTerraform;
function ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        aws_pca_authority_arn: {
            value: cdktf.stringToHclTerraform(struct.awsPcaAuthorityArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToHclTerraform = ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToHclTerraform;
class EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._awsPcaAuthorityArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsPcaAuthorityArn = this._awsPcaAuthorityArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._awsPcaAuthorityArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._awsPcaAuthorityArn = value.awsPcaAuthorityArn;
        }
    }
    get awsPcaAuthorityArn() {
        return this.getStringAttribute('aws_pca_authority_arn');
    }
    set awsPcaAuthorityArn(value) {
        this._awsPcaAuthorityArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get awsPcaAuthorityArnInput() {
        return this._awsPcaAuthorityArn;
    }
}
exports.EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference = EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference;
_t = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference[_t] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationServiceTlsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        issuer_cert_authority: ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToTerraform(struct.issuerCertAuthority),
    };
}
exports.ecsServiceServiceConnectConfigurationServiceTlsToTerraform = ecsServiceServiceConnectConfigurationServiceTlsToTerraform;
function ecsServiceServiceConnectConfigurationServiceTlsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        issuer_cert_authority: {
            value: ecsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityToHclTerraform(struct.issuerCertAuthority),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationServiceTlsToHclTerraform = ecsServiceServiceConnectConfigurationServiceTlsToHclTerraform;
class EcsServiceServiceConnectConfigurationServiceTlsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // issuer_cert_authority - computed: false, optional: false, required: true
        this._issuerCertAuthority = new EcsServiceServiceConnectConfigurationServiceTlsIssuerCertAuthorityOutputReference(this, "issuer_cert_authority");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._issuerCertAuthority?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuerCertAuthority = this._issuerCertAuthority?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKey = undefined;
            this._roleArn = undefined;
            this._issuerCertAuthority.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKey = value.kmsKey;
            this._roleArn = value.roleArn;
            this._issuerCertAuthority.internalValue = value.issuerCertAuthority;
        }
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get issuerCertAuthority() {
        return this._issuerCertAuthority;
    }
    putIssuerCertAuthority(value) {
        this._issuerCertAuthority.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get issuerCertAuthorityInput() {
        return this._issuerCertAuthority.internalValue;
    }
}
exports.EcsServiceServiceConnectConfigurationServiceTlsOutputReference = EcsServiceServiceConnectConfigurationServiceTlsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceTlsOutputReference[_u] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceTlsOutputReference", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationServiceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        discovery_name: cdktf.stringToTerraform(struct.discoveryName),
        ingress_port_override: cdktf.numberToTerraform(struct.ingressPortOverride),
        port_name: cdktf.stringToTerraform(struct.portName),
        client_alias: ecsServiceServiceConnectConfigurationServiceClientAliasToTerraform(struct.clientAlias),
        timeout: ecsServiceServiceConnectConfigurationServiceTimeoutToTerraform(struct.timeout),
        tls: ecsServiceServiceConnectConfigurationServiceTlsToTerraform(struct.tls),
    };
}
exports.ecsServiceServiceConnectConfigurationServiceToTerraform = ecsServiceServiceConnectConfigurationServiceToTerraform;
function ecsServiceServiceConnectConfigurationServiceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        discovery_name: {
            value: cdktf.stringToHclTerraform(struct.discoveryName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ingress_port_override: {
            value: cdktf.numberToHclTerraform(struct.ingressPortOverride),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        port_name: {
            value: cdktf.stringToHclTerraform(struct.portName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        client_alias: {
            value: ecsServiceServiceConnectConfigurationServiceClientAliasToHclTerraform(struct.clientAlias),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationServiceClientAliasList",
        },
        timeout: {
            value: ecsServiceServiceConnectConfigurationServiceTimeoutToHclTerraform(struct.timeout),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationServiceTimeoutList",
        },
        tls: {
            value: ecsServiceServiceConnectConfigurationServiceTlsToHclTerraform(struct.tls),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationServiceTlsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationServiceToHclTerraform = ecsServiceServiceConnectConfigurationServiceToHclTerraform;
class EcsServiceServiceConnectConfigurationServiceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // client_alias - computed: false, optional: true, required: false
        this._clientAlias = new EcsServiceServiceConnectConfigurationServiceClientAliasOutputReference(this, "client_alias");
        // timeout - computed: false, optional: true, required: false
        this._timeout = new EcsServiceServiceConnectConfigurationServiceTimeoutOutputReference(this, "timeout");
        // tls - computed: false, optional: true, required: false
        this._tls = new EcsServiceServiceConnectConfigurationServiceTlsOutputReference(this, "tls");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._discoveryName !== undefined) {
            hasAnyValues = true;
            internalValueResult.discoveryName = this._discoveryName;
        }
        if (this._ingressPortOverride !== undefined) {
            hasAnyValues = true;
            internalValueResult.ingressPortOverride = this._ingressPortOverride;
        }
        if (this._portName !== undefined) {
            hasAnyValues = true;
            internalValueResult.portName = this._portName;
        }
        if (this._clientAlias?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientAlias = this._clientAlias?.internalValue;
        }
        if (this._timeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout?.internalValue;
        }
        if (this._tls?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tls = this._tls?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._discoveryName = undefined;
            this._ingressPortOverride = undefined;
            this._portName = undefined;
            this._clientAlias.internalValue = undefined;
            this._timeout.internalValue = undefined;
            this._tls.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._discoveryName = value.discoveryName;
            this._ingressPortOverride = value.ingressPortOverride;
            this._portName = value.portName;
            this._clientAlias.internalValue = value.clientAlias;
            this._timeout.internalValue = value.timeout;
            this._tls.internalValue = value.tls;
        }
    }
    get discoveryName() {
        return this.getStringAttribute('discovery_name');
    }
    set discoveryName(value) {
        this._discoveryName = value;
    }
    resetDiscoveryName() {
        this._discoveryName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get discoveryNameInput() {
        return this._discoveryName;
    }
    get ingressPortOverride() {
        return this.getNumberAttribute('ingress_port_override');
    }
    set ingressPortOverride(value) {
        this._ingressPortOverride = value;
    }
    resetIngressPortOverride() {
        this._ingressPortOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressPortOverrideInput() {
        return this._ingressPortOverride;
    }
    get portName() {
        return this.getStringAttribute('port_name');
    }
    set portName(value) {
        this._portName = value;
    }
    // Temporarily expose input value. Use with caution.
    get portNameInput() {
        return this._portName;
    }
    get clientAlias() {
        return this._clientAlias;
    }
    putClientAlias(value) {
        this._clientAlias.internalValue = value;
    }
    resetClientAlias() {
        this._clientAlias.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientAliasInput() {
        return this._clientAlias.internalValue;
    }
    get timeout() {
        return this._timeout;
    }
    putTimeout(value) {
        this._timeout.internalValue = value;
    }
    resetTimeout() {
        this._timeout.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout.internalValue;
    }
    get tls() {
        return this._tls;
    }
    putTls(value) {
        this._tls.internalValue = value;
    }
    resetTls() {
        this._tls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsInput() {
        return this._tls.internalValue;
    }
}
exports.EcsServiceServiceConnectConfigurationServiceOutputReference = EcsServiceServiceConnectConfigurationServiceOutputReference;
_v = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceOutputReference[_v] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceOutputReference", version: "19.15.0" };
class EcsServiceServiceConnectConfigurationServiceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceServiceConnectConfigurationServiceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceServiceConnectConfigurationServiceList = EcsServiceServiceConnectConfigurationServiceList;
_w = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationServiceList[_w] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationServiceList", version: "19.15.0" };
function ecsServiceServiceConnectConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        namespace: cdktf.stringToTerraform(struct.namespace),
        log_configuration: ecsServiceServiceConnectConfigurationLogConfigurationToTerraform(struct.logConfiguration),
        service: cdktf.listMapper(ecsServiceServiceConnectConfigurationServiceToTerraform, true)(struct.service),
    };
}
exports.ecsServiceServiceConnectConfigurationToTerraform = ecsServiceServiceConnectConfigurationToTerraform;
function ecsServiceServiceConnectConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        namespace: {
            value: cdktf.stringToHclTerraform(struct.namespace),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        log_configuration: {
            value: ecsServiceServiceConnectConfigurationLogConfigurationToHclTerraform(struct.logConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationLogConfigurationList",
        },
        service: {
            value: cdktf.listMapperHcl(ecsServiceServiceConnectConfigurationServiceToHclTerraform, true)(struct.service),
            isBlock: true,
            type: "list",
            storageClassType: "EcsServiceServiceConnectConfigurationServiceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceConnectConfigurationToHclTerraform = ecsServiceServiceConnectConfigurationToHclTerraform;
class EcsServiceServiceConnectConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // log_configuration - computed: false, optional: true, required: false
        this._logConfiguration = new EcsServiceServiceConnectConfigurationLogConfigurationOutputReference(this, "log_configuration");
        // service - computed: false, optional: true, required: false
        this._service = new EcsServiceServiceConnectConfigurationServiceList(this, "service", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._namespace !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespace = this._namespace;
        }
        if (this._logConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.logConfiguration = this._logConfiguration?.internalValue;
        }
        if (this._service?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.service = this._service?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._namespace = undefined;
            this._logConfiguration.internalValue = undefined;
            this._service.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._namespace = value.namespace;
            this._logConfiguration.internalValue = value.logConfiguration;
            this._service.internalValue = value.service;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get logConfiguration() {
        return this._logConfiguration;
    }
    putLogConfiguration(value) {
        this._logConfiguration.internalValue = value;
    }
    resetLogConfiguration() {
        this._logConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logConfigurationInput() {
        return this._logConfiguration.internalValue;
    }
    get service() {
        return this._service;
    }
    putService(value) {
        this._service.internalValue = value;
    }
    resetService() {
        this._service.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceInput() {
        return this._service.internalValue;
    }
}
exports.EcsServiceServiceConnectConfigurationOutputReference = EcsServiceServiceConnectConfigurationOutputReference;
_x = JSII_RTTI_SYMBOL_1;
EcsServiceServiceConnectConfigurationOutputReference[_x] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceConnectConfigurationOutputReference", version: "19.15.0" };
function ecsServiceServiceRegistriesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        port: cdktf.numberToTerraform(struct.port),
        registry_arn: cdktf.stringToTerraform(struct.registryArn),
    };
}
exports.ecsServiceServiceRegistriesToTerraform = ecsServiceServiceRegistriesToTerraform;
function ecsServiceServiceRegistriesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_name: {
            value: cdktf.stringToHclTerraform(struct.containerName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_port: {
            value: cdktf.numberToHclTerraform(struct.containerPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        registry_arn: {
            value: cdktf.stringToHclTerraform(struct.registryArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceServiceRegistriesToHclTerraform = ecsServiceServiceRegistriesToHclTerraform;
class EcsServiceServiceRegistriesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerName = this._containerName;
        }
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._registryArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.registryArn = this._registryArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerName = undefined;
            this._containerPort = undefined;
            this._port = undefined;
            this._registryArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerName = value.containerName;
            this._containerPort = value.containerPort;
            this._port = value.port;
            this._registryArn = value.registryArn;
        }
    }
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    set containerName(value) {
        this._containerName = value;
    }
    resetContainerName() {
        this._containerName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerNameInput() {
        return this._containerName;
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get registryArn() {
        return this.getStringAttribute('registry_arn');
    }
    set registryArn(value) {
        this._registryArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get registryArnInput() {
        return this._registryArn;
    }
}
exports.EcsServiceServiceRegistriesOutputReference = EcsServiceServiceRegistriesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
EcsServiceServiceRegistriesOutputReference[_y] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceRegistriesOutputReference", version: "19.15.0" };
function ecsServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.ecsServiceTimeoutsToTerraform = ecsServiceTimeoutsToTerraform;
function ecsServiceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecsServiceTimeoutsToHclTerraform = ecsServiceTimeoutsToHclTerraform;
class EcsServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EcsServiceTimeoutsOutputReference = EcsServiceTimeoutsOutputReference;
_z = JSII_RTTI_SYMBOL_1;
EcsServiceTimeoutsOutputReference[_z] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecs_service aws_ecs_service}
*/
class EcsService extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EcsService resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EcsService to import
    * @param importFromId The id of the existing EcsService that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecs_service#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EcsService to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecs_service", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecs_service aws_ecs_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EcsServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // alarms - computed: false, optional: true, required: false
        this._alarms = new EcsServiceAlarmsOutputReference(this, "alarms");
        // capacity_provider_strategy - computed: false, optional: true, required: false
        this._capacityProviderStrategy = new EcsServiceCapacityProviderStrategyList(this, "capacity_provider_strategy", true);
        // deployment_circuit_breaker - computed: false, optional: true, required: false
        this._deploymentCircuitBreaker = new EcsServiceDeploymentCircuitBreakerOutputReference(this, "deployment_circuit_breaker");
        // deployment_controller - computed: false, optional: true, required: false
        this._deploymentController = new EcsServiceDeploymentControllerOutputReference(this, "deployment_controller");
        // load_balancer - computed: false, optional: true, required: false
        this._loadBalancer = new EcsServiceLoadBalancerList(this, "load_balancer", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new EcsServiceNetworkConfigurationOutputReference(this, "network_configuration");
        // ordered_placement_strategy - computed: false, optional: true, required: false
        this._orderedPlacementStrategy = new EcsServiceOrderedPlacementStrategyList(this, "ordered_placement_strategy", false);
        // placement_constraints - computed: false, optional: true, required: false
        this._placementConstraints = new EcsServicePlacementConstraintsList(this, "placement_constraints", true);
        // service_connect_configuration - computed: false, optional: true, required: false
        this._serviceConnectConfiguration = new EcsServiceServiceConnectConfigurationOutputReference(this, "service_connect_configuration");
        // service_registries - computed: false, optional: true, required: false
        this._serviceRegistries = new EcsServiceServiceRegistriesOutputReference(this, "service_registries");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EcsServiceTimeoutsOutputReference(this, "timeouts");
        this._cluster = config.cluster;
        this._deploymentMaximumPercent = config.deploymentMaximumPercent;
        this._deploymentMinimumHealthyPercent = config.deploymentMinimumHealthyPercent;
        this._desiredCount = config.desiredCount;
        this._enableEcsManagedTags = config.enableEcsManagedTags;
        this._enableExecuteCommand = config.enableExecuteCommand;
        this._forceNewDeployment = config.forceNewDeployment;
        this._healthCheckGracePeriodSeconds = config.healthCheckGracePeriodSeconds;
        this._iamRole = config.iamRole;
        this._id = config.id;
        this._launchType = config.launchType;
        this._name = config.name;
        this._platformVersion = config.platformVersion;
        this._propagateTags = config.propagateTags;
        this._schedulingStrategy = config.schedulingStrategy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._taskDefinition = config.taskDefinition;
        this._triggers = config.triggers;
        this._waitForSteadyState = config.waitForSteadyState;
        this._alarms.internalValue = config.alarms;
        this._capacityProviderStrategy.internalValue = config.capacityProviderStrategy;
        this._deploymentCircuitBreaker.internalValue = config.deploymentCircuitBreaker;
        this._deploymentController.internalValue = config.deploymentController;
        this._loadBalancer.internalValue = config.loadBalancer;
        this._networkConfiguration.internalValue = config.networkConfiguration;
        this._orderedPlacementStrategy.internalValue = config.orderedPlacementStrategy;
        this._placementConstraints.internalValue = config.placementConstraints;
        this._serviceConnectConfiguration.internalValue = config.serviceConnectConfiguration;
        this._serviceRegistries.internalValue = config.serviceRegistries;
        this._timeouts.internalValue = config.timeouts;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    resetCluster() {
        this._cluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    get deploymentMaximumPercent() {
        return this.getNumberAttribute('deployment_maximum_percent');
    }
    set deploymentMaximumPercent(value) {
        this._deploymentMaximumPercent = value;
    }
    resetDeploymentMaximumPercent() {
        this._deploymentMaximumPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentMaximumPercentInput() {
        return this._deploymentMaximumPercent;
    }
    get deploymentMinimumHealthyPercent() {
        return this.getNumberAttribute('deployment_minimum_healthy_percent');
    }
    set deploymentMinimumHealthyPercent(value) {
        this._deploymentMinimumHealthyPercent = value;
    }
    resetDeploymentMinimumHealthyPercent() {
        this._deploymentMinimumHealthyPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentMinimumHealthyPercentInput() {
        return this._deploymentMinimumHealthyPercent;
    }
    get desiredCount() {
        return this.getNumberAttribute('desired_count');
    }
    set desiredCount(value) {
        this._desiredCount = value;
    }
    resetDesiredCount() {
        this._desiredCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredCountInput() {
        return this._desiredCount;
    }
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    get forceNewDeployment() {
        return this.getBooleanAttribute('force_new_deployment');
    }
    set forceNewDeployment(value) {
        this._forceNewDeployment = value;
    }
    resetForceNewDeployment() {
        this._forceNewDeployment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceNewDeploymentInput() {
        return this._forceNewDeployment;
    }
    get healthCheckGracePeriodSeconds() {
        return this.getNumberAttribute('health_check_grace_period_seconds');
    }
    set healthCheckGracePeriodSeconds(value) {
        this._healthCheckGracePeriodSeconds = value;
    }
    resetHealthCheckGracePeriodSeconds() {
        this._healthCheckGracePeriodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckGracePeriodSecondsInput() {
        return this._healthCheckGracePeriodSeconds;
    }
    get iamRole() {
        return this.getStringAttribute('iam_role');
    }
    set iamRole(value) {
        this._iamRole = value;
    }
    resetIamRole() {
        this._iamRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleInput() {
        return this._iamRole;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    set launchType(value) {
        this._launchType = value;
    }
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchTypeInput() {
        return this._launchType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    set propagateTags(value) {
        this._propagateTags = value;
    }
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propagateTagsInput() {
        return this._propagateTags;
    }
    get schedulingStrategy() {
        return this.getStringAttribute('scheduling_strategy');
    }
    set schedulingStrategy(value) {
        this._schedulingStrategy = value;
    }
    resetSchedulingStrategy() {
        this._schedulingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schedulingStrategyInput() {
        return this._schedulingStrategy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    set taskDefinition(value) {
        this._taskDefinition = value;
    }
    resetTaskDefinition() {
        this._taskDefinition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskDefinitionInput() {
        return this._taskDefinition;
    }
    get triggers() {
        return this.getStringMapAttribute('triggers');
    }
    set triggers(value) {
        this._triggers = value;
    }
    resetTriggers() {
        this._triggers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggersInput() {
        return this._triggers;
    }
    get waitForSteadyState() {
        return this.getBooleanAttribute('wait_for_steady_state');
    }
    set waitForSteadyState(value) {
        this._waitForSteadyState = value;
    }
    resetWaitForSteadyState() {
        this._waitForSteadyState = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForSteadyStateInput() {
        return this._waitForSteadyState;
    }
    get alarms() {
        return this._alarms;
    }
    putAlarms(value) {
        this._alarms.internalValue = value;
    }
    resetAlarms() {
        this._alarms.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get alarmsInput() {
        return this._alarms.internalValue;
    }
    get capacityProviderStrategy() {
        return this._capacityProviderStrategy;
    }
    putCapacityProviderStrategy(value) {
        this._capacityProviderStrategy.internalValue = value;
    }
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy.internalValue;
    }
    get deploymentCircuitBreaker() {
        return this._deploymentCircuitBreaker;
    }
    putDeploymentCircuitBreaker(value) {
        this._deploymentCircuitBreaker.internalValue = value;
    }
    resetDeploymentCircuitBreaker() {
        this._deploymentCircuitBreaker.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentCircuitBreakerInput() {
        return this._deploymentCircuitBreaker.internalValue;
    }
    get deploymentController() {
        return this._deploymentController;
    }
    putDeploymentController(value) {
        this._deploymentController.internalValue = value;
    }
    resetDeploymentController() {
        this._deploymentController.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentControllerInput() {
        return this._deploymentController.internalValue;
    }
    get loadBalancer() {
        return this._loadBalancer;
    }
    putLoadBalancer(value) {
        this._loadBalancer.internalValue = value;
    }
    resetLoadBalancer() {
        this._loadBalancer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerInput() {
        return this._loadBalancer.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get orderedPlacementStrategy() {
        return this._orderedPlacementStrategy;
    }
    putOrderedPlacementStrategy(value) {
        this._orderedPlacementStrategy.internalValue = value;
    }
    resetOrderedPlacementStrategy() {
        this._orderedPlacementStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orderedPlacementStrategyInput() {
        return this._orderedPlacementStrategy.internalValue;
    }
    get placementConstraints() {
        return this._placementConstraints;
    }
    putPlacementConstraints(value) {
        this._placementConstraints.internalValue = value;
    }
    resetPlacementConstraints() {
        this._placementConstraints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementConstraintsInput() {
        return this._placementConstraints.internalValue;
    }
    get serviceConnectConfiguration() {
        return this._serviceConnectConfiguration;
    }
    putServiceConnectConfiguration(value) {
        this._serviceConnectConfiguration.internalValue = value;
    }
    resetServiceConnectConfiguration() {
        this._serviceConnectConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceConnectConfigurationInput() {
        return this._serviceConnectConfiguration.internalValue;
    }
    get serviceRegistries() {
        return this._serviceRegistries;
    }
    putServiceRegistries(value) {
        this._serviceRegistries.internalValue = value;
    }
    resetServiceRegistries() {
        this._serviceRegistries.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceRegistriesInput() {
        return this._serviceRegistries.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster: cdktf.stringToTerraform(this._cluster),
            deployment_maximum_percent: cdktf.numberToTerraform(this._deploymentMaximumPercent),
            deployment_minimum_healthy_percent: cdktf.numberToTerraform(this._deploymentMinimumHealthyPercent),
            desired_count: cdktf.numberToTerraform(this._desiredCount),
            enable_ecs_managed_tags: cdktf.booleanToTerraform(this._enableEcsManagedTags),
            enable_execute_command: cdktf.booleanToTerraform(this._enableExecuteCommand),
            force_new_deployment: cdktf.booleanToTerraform(this._forceNewDeployment),
            health_check_grace_period_seconds: cdktf.numberToTerraform(this._healthCheckGracePeriodSeconds),
            iam_role: cdktf.stringToTerraform(this._iamRole),
            id: cdktf.stringToTerraform(this._id),
            launch_type: cdktf.stringToTerraform(this._launchType),
            name: cdktf.stringToTerraform(this._name),
            platform_version: cdktf.stringToTerraform(this._platformVersion),
            propagate_tags: cdktf.stringToTerraform(this._propagateTags),
            scheduling_strategy: cdktf.stringToTerraform(this._schedulingStrategy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            task_definition: cdktf.stringToTerraform(this._taskDefinition),
            triggers: cdktf.hashMapper(cdktf.stringToTerraform)(this._triggers),
            wait_for_steady_state: cdktf.booleanToTerraform(this._waitForSteadyState),
            alarms: ecsServiceAlarmsToTerraform(this._alarms.internalValue),
            capacity_provider_strategy: cdktf.listMapper(ecsServiceCapacityProviderStrategyToTerraform, true)(this._capacityProviderStrategy.internalValue),
            deployment_circuit_breaker: ecsServiceDeploymentCircuitBreakerToTerraform(this._deploymentCircuitBreaker.internalValue),
            deployment_controller: ecsServiceDeploymentControllerToTerraform(this._deploymentController.internalValue),
            load_balancer: cdktf.listMapper(ecsServiceLoadBalancerToTerraform, true)(this._loadBalancer.internalValue),
            network_configuration: ecsServiceNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
            ordered_placement_strategy: cdktf.listMapper(ecsServiceOrderedPlacementStrategyToTerraform, true)(this._orderedPlacementStrategy.internalValue),
            placement_constraints: cdktf.listMapper(ecsServicePlacementConstraintsToTerraform, true)(this._placementConstraints.internalValue),
            service_connect_configuration: ecsServiceServiceConnectConfigurationToTerraform(this._serviceConnectConfiguration.internalValue),
            service_registries: ecsServiceServiceRegistriesToTerraform(this._serviceRegistries.internalValue),
            timeouts: ecsServiceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster: {
                value: cdktf.stringToHclTerraform(this._cluster),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deployment_maximum_percent: {
                value: cdktf.numberToHclTerraform(this._deploymentMaximumPercent),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            deployment_minimum_healthy_percent: {
                value: cdktf.numberToHclTerraform(this._deploymentMinimumHealthyPercent),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            desired_count: {
                value: cdktf.numberToHclTerraform(this._desiredCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            enable_ecs_managed_tags: {
                value: cdktf.booleanToHclTerraform(this._enableEcsManagedTags),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            enable_execute_command: {
                value: cdktf.booleanToHclTerraform(this._enableExecuteCommand),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            force_new_deployment: {
                value: cdktf.booleanToHclTerraform(this._forceNewDeployment),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            health_check_grace_period_seconds: {
                value: cdktf.numberToHclTerraform(this._healthCheckGracePeriodSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            iam_role: {
                value: cdktf.stringToHclTerraform(this._iamRole),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            launch_type: {
                value: cdktf.stringToHclTerraform(this._launchType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            platform_version: {
                value: cdktf.stringToHclTerraform(this._platformVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            propagate_tags: {
                value: cdktf.stringToHclTerraform(this._propagateTags),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scheduling_strategy: {
                value: cdktf.stringToHclTerraform(this._schedulingStrategy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            task_definition: {
                value: cdktf.stringToHclTerraform(this._taskDefinition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            triggers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._triggers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            wait_for_steady_state: {
                value: cdktf.booleanToHclTerraform(this._waitForSteadyState),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            alarms: {
                value: ecsServiceAlarmsToHclTerraform(this._alarms.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceAlarmsList",
            },
            capacity_provider_strategy: {
                value: cdktf.listMapperHcl(ecsServiceCapacityProviderStrategyToHclTerraform, true)(this._capacityProviderStrategy.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "EcsServiceCapacityProviderStrategyList",
            },
            deployment_circuit_breaker: {
                value: ecsServiceDeploymentCircuitBreakerToHclTerraform(this._deploymentCircuitBreaker.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceDeploymentCircuitBreakerList",
            },
            deployment_controller: {
                value: ecsServiceDeploymentControllerToHclTerraform(this._deploymentController.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceDeploymentControllerList",
            },
            load_balancer: {
                value: cdktf.listMapperHcl(ecsServiceLoadBalancerToHclTerraform, true)(this._loadBalancer.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "EcsServiceLoadBalancerList",
            },
            network_configuration: {
                value: ecsServiceNetworkConfigurationToHclTerraform(this._networkConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceNetworkConfigurationList",
            },
            ordered_placement_strategy: {
                value: cdktf.listMapperHcl(ecsServiceOrderedPlacementStrategyToHclTerraform, true)(this._orderedPlacementStrategy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceOrderedPlacementStrategyList",
            },
            placement_constraints: {
                value: cdktf.listMapperHcl(ecsServicePlacementConstraintsToHclTerraform, true)(this._placementConstraints.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "EcsServicePlacementConstraintsList",
            },
            service_connect_configuration: {
                value: ecsServiceServiceConnectConfigurationToHclTerraform(this._serviceConnectConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceServiceConnectConfigurationList",
            },
            service_registries: {
                value: ecsServiceServiceRegistriesToHclTerraform(this._serviceRegistries.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcsServiceServiceRegistriesList",
            },
            timeouts: {
                value: ecsServiceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "EcsServiceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EcsService = EcsService;
_0 = JSII_RTTI_SYMBOL_1;
EcsService[_0] = { fqn: "@cdktf/provider-aws.ecsService.EcsService", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EcsService.tfResourceType = "aws_ecs_service";
