// @cdktf/provider-aws/lib/devopsguru-notification-channel/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevopsguruNotificationChannel = exports.DevopsguruNotificationChannelSnsList = exports.DevopsguruNotificationChannelSnsOutputReference = exports.devopsguruNotificationChannelSnsToHclTerraform = exports.devopsguruNotificationChannelSnsToTerraform = exports.DevopsguruNotificationChannelFiltersList = exports.DevopsguruNotificationChannelFiltersOutputReference = exports.devopsguruNotificationChannelFiltersToHclTerraform = exports.devopsguruNotificationChannelFiltersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function devopsguruNotificationChannelFiltersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.messageTypes),
        severities: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.severities),
    };
}
exports.devopsguruNotificationChannelFiltersToTerraform = devopsguruNotificationChannelFiltersToTerraform;
function devopsguruNotificationChannelFiltersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        message_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.messageTypes),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        severities: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.severities),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.devopsguruNotificationChannelFiltersToHclTerraform = devopsguruNotificationChannelFiltersToHclTerraform;
class DevopsguruNotificationChannelFiltersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.messageTypes = this._messageTypes;
        }
        if (this._severities !== undefined) {
            hasAnyValues = true;
            internalValueResult.severities = this._severities;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._messageTypes = undefined;
            this._severities = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._messageTypes = value.messageTypes;
            this._severities = value.severities;
        }
    }
    get messageTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('message_types'));
    }
    set messageTypes(value) {
        this._messageTypes = value;
    }
    resetMessageTypes() {
        this._messageTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messageTypesInput() {
        return this._messageTypes;
    }
    get severities() {
        return cdktf.Fn.tolist(this.getListAttribute('severities'));
    }
    set severities(value) {
        this._severities = value;
    }
    resetSeverities() {
        this._severities = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get severitiesInput() {
        return this._severities;
    }
}
exports.DevopsguruNotificationChannelFiltersOutputReference = DevopsguruNotificationChannelFiltersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DevopsguruNotificationChannelFiltersOutputReference[_a] = { fqn: "@cdktf/provider-aws.devopsguruNotificationChannel.DevopsguruNotificationChannelFiltersOutputReference", version: "19.15.0" };
class DevopsguruNotificationChannelFiltersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DevopsguruNotificationChannelFiltersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DevopsguruNotificationChannelFiltersList = DevopsguruNotificationChannelFiltersList;
_b = JSII_RTTI_SYMBOL_1;
DevopsguruNotificationChannelFiltersList[_b] = { fqn: "@cdktf/provider-aws.devopsguruNotificationChannel.DevopsguruNotificationChannelFiltersList", version: "19.15.0" };
function devopsguruNotificationChannelSnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        topic_arn: cdktf.stringToTerraform(struct.topicArn),
    };
}
exports.devopsguruNotificationChannelSnsToTerraform = devopsguruNotificationChannelSnsToTerraform;
function devopsguruNotificationChannelSnsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        topic_arn: {
            value: cdktf.stringToHclTerraform(struct.topicArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.devopsguruNotificationChannelSnsToHclTerraform = devopsguruNotificationChannelSnsToHclTerraform;
class DevopsguruNotificationChannelSnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._topicArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.topicArn = this._topicArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._topicArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._topicArn = value.topicArn;
        }
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
}
exports.DevopsguruNotificationChannelSnsOutputReference = DevopsguruNotificationChannelSnsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DevopsguruNotificationChannelSnsOutputReference[_c] = { fqn: "@cdktf/provider-aws.devopsguruNotificationChannel.DevopsguruNotificationChannelSnsOutputReference", version: "19.15.0" };
class DevopsguruNotificationChannelSnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DevopsguruNotificationChannelSnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DevopsguruNotificationChannelSnsList = DevopsguruNotificationChannelSnsList;
_d = JSII_RTTI_SYMBOL_1;
DevopsguruNotificationChannelSnsList[_d] = { fqn: "@cdktf/provider-aws.devopsguruNotificationChannel.DevopsguruNotificationChannelSnsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devopsguru_notification_channel aws_devopsguru_notification_channel}
*/
class DevopsguruNotificationChannel extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DevopsguruNotificationChannel resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DevopsguruNotificationChannel to import
    * @param importFromId The id of the existing DevopsguruNotificationChannel that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devopsguru_notification_channel#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DevopsguruNotificationChannel to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_devopsguru_notification_channel", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devopsguru_notification_channel aws_devopsguru_notification_channel} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DevopsguruNotificationChannelConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_devopsguru_notification_channel',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filters - computed: false, optional: true, required: false
        this._filters = new DevopsguruNotificationChannelFiltersList(this, "filters", false);
        // sns - computed: false, optional: true, required: false
        this._sns = new DevopsguruNotificationChannelSnsList(this, "sns", false);
        this._filters.internalValue = config.filters;
        this._sns.internalValue = config.sns;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get filters() {
        return this._filters;
    }
    putFilters(value) {
        this._filters.internalValue = value;
    }
    resetFilters() {
        this._filters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filtersInput() {
        return this._filters.internalValue;
    }
    get sns() {
        return this._sns;
    }
    putSns(value) {
        this._sns.internalValue = value;
    }
    resetSns() {
        this._sns.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snsInput() {
        return this._sns.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filters: cdktf.listMapper(devopsguruNotificationChannelFiltersToTerraform, true)(this._filters.internalValue),
            sns: cdktf.listMapper(devopsguruNotificationChannelSnsToTerraform, true)(this._sns.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filters: {
                value: cdktf.listMapperHcl(devopsguruNotificationChannelFiltersToHclTerraform, true)(this._filters.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DevopsguruNotificationChannelFiltersList",
            },
            sns: {
                value: cdktf.listMapperHcl(devopsguruNotificationChannelSnsToHclTerraform, true)(this._sns.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DevopsguruNotificationChannelSnsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DevopsguruNotificationChannel = DevopsguruNotificationChannel;
_e = JSII_RTTI_SYMBOL_1;
DevopsguruNotificationChannel[_e] = { fqn: "@cdktf/provider-aws.devopsguruNotificationChannel.DevopsguruNotificationChannel", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DevopsguruNotificationChannel.tfResourceType = "aws_devopsguru_notification_channel";
