// @cdktf/provider-aws/lib/data-aws-ssmcontacts-contact-channel/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsmcontactsContactChannel = exports.DataAwsSsmcontactsContactChannelDeliveryAddressList = exports.DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference = exports.dataAwsSsmcontactsContactChannelDeliveryAddressToHclTerraform = exports.dataAwsSsmcontactsContactChannelDeliveryAddressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSsmcontactsContactChannelDeliveryAddressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsmcontactsContactChannelDeliveryAddressToTerraform = dataAwsSsmcontactsContactChannelDeliveryAddressToTerraform;
function dataAwsSsmcontactsContactChannelDeliveryAddressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsmcontactsContactChannelDeliveryAddressToHclTerraform = dataAwsSsmcontactsContactChannelDeliveryAddressToHclTerraform;
class DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // simple_address - computed: true, optional: false, required: false
    get simpleAddress() {
        return this.getStringAttribute('simple_address');
    }
}
exports.DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference = DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSsmcontactsContactChannel.DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference", version: "19.15.0" };
class DataAwsSsmcontactsContactChannelDeliveryAddressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsmcontactsContactChannelDeliveryAddressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsmcontactsContactChannelDeliveryAddressList = DataAwsSsmcontactsContactChannelDeliveryAddressList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSsmcontactsContactChannelDeliveryAddressList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSsmcontactsContactChannel.DataAwsSsmcontactsContactChannelDeliveryAddressList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssmcontacts_contact_channel aws_ssmcontacts_contact_channel}
*/
class DataAwsSsmcontactsContactChannel extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSsmcontactsContactChannel resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSsmcontactsContactChannel to import
    * @param importFromId The id of the existing DataAwsSsmcontactsContactChannel that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssmcontacts_contact_channel#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSsmcontactsContactChannel to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssmcontacts_contact_channel", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssmcontacts_contact_channel aws_ssmcontacts_contact_channel} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSsmcontactsContactChannelConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssmcontacts_contact_channel',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // delivery_address - computed: true, optional: false, required: false
        this._deliveryAddress = new DataAwsSsmcontactsContactChannelDeliveryAddressList(this, "delivery_address", false);
        this._arn = config.arn;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // activation_status - computed: true, optional: false, required: false
    get activationStatus() {
        return this.getStringAttribute('activation_status');
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // contact_id - computed: true, optional: false, required: false
    get contactId() {
        return this.getStringAttribute('contact_id');
    }
    get deliveryAddress() {
        return this._deliveryAddress;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSsmcontactsContactChannel = DataAwsSsmcontactsContactChannel;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSsmcontactsContactChannel[_c] = { fqn: "@cdktf/provider-aws.dataAwsSsmcontactsContactChannel.DataAwsSsmcontactsContactChannel", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSsmcontactsContactChannel.tfResourceType = "aws_ssmcontacts_contact_channel";
