// @cdktf/provider-aws/lib/transfer-agreement/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferAgreement = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_agreement aws_transfer_agreement}
*/
class TransferAgreement extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a TransferAgreement resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the TransferAgreement to import
    * @param importFromId The id of the existing TransferAgreement that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_agreement#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the TransferAgreement to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_transfer_agreement", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_agreement aws_transfer_agreement} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options TransferAgreementConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_transfer_agreement',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accessRole = config.accessRole;
        this._baseDirectory = config.baseDirectory;
        this._description = config.description;
        this._id = config.id;
        this._localProfileId = config.localProfileId;
        this._partnerProfileId = config.partnerProfileId;
        this._serverId = config.serverId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    get accessRole() {
        return this.getStringAttribute('access_role');
    }
    set accessRole(value) {
        this._accessRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessRoleInput() {
        return this._accessRole;
    }
    // agreement_id - computed: true, optional: false, required: false
    get agreementId() {
        return this.getStringAttribute('agreement_id');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get baseDirectory() {
        return this.getStringAttribute('base_directory');
    }
    set baseDirectory(value) {
        this._baseDirectory = value;
    }
    // Temporarily expose input value. Use with caution.
    get baseDirectoryInput() {
        return this._baseDirectory;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get localProfileId() {
        return this.getStringAttribute('local_profile_id');
    }
    set localProfileId(value) {
        this._localProfileId = value;
    }
    // Temporarily expose input value. Use with caution.
    get localProfileIdInput() {
        return this._localProfileId;
    }
    get partnerProfileId() {
        return this.getStringAttribute('partner_profile_id');
    }
    set partnerProfileId(value) {
        this._partnerProfileId = value;
    }
    // Temporarily expose input value. Use with caution.
    get partnerProfileIdInput() {
        return this._partnerProfileId;
    }
    get serverId() {
        return this.getStringAttribute('server_id');
    }
    set serverId(value) {
        this._serverId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serverIdInput() {
        return this._serverId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_role: cdktf.stringToTerraform(this._accessRole),
            base_directory: cdktf.stringToTerraform(this._baseDirectory),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            local_profile_id: cdktf.stringToTerraform(this._localProfileId),
            partner_profile_id: cdktf.stringToTerraform(this._partnerProfileId),
            server_id: cdktf.stringToTerraform(this._serverId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_role: {
                value: cdktf.stringToHclTerraform(this._accessRole),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            base_directory: {
                value: cdktf.stringToHclTerraform(this._baseDirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            local_profile_id: {
                value: cdktf.stringToHclTerraform(this._localProfileId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            partner_profile_id: {
                value: cdktf.stringToHclTerraform(this._partnerProfileId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            server_id: {
                value: cdktf.stringToHclTerraform(this._serverId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.TransferAgreement = TransferAgreement;
_a = JSII_RTTI_SYMBOL_1;
TransferAgreement[_a] = { fqn: "@cdktf/provider-aws.transferAgreement.TransferAgreement", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
TransferAgreement.tfResourceType = "aws_transfer_agreement";
