// @winglang/sdk/lib/shared-aws/api-util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apigwFunctionHandler = void 0;
const cloud_1 = require("../cloud");
async function apigwFunctionHandler(request, handlerFunction, headers) {
    const apiRequest = mapApigatewayEventToCloudApiRequest(request);
    const apiResponse = (await handlerFunction(apiRequest)) ?? {};
    return mapCloudApiResponseToApigatewayResponse(apiResponse, headers);
}
exports.apigwFunctionHandler = apigwFunctionHandler;
/**
 * Map a Cloud API response to an API Gateway response
 * @param resp Cloud API response
 * @returns API Gateway response
 */
function mapCloudApiResponseToApigatewayResponse(resp, corsHeaders) {
    return {
        statusCode: resp.status ?? cloud_1.DEFAULT_RESPONSE_STATUS,
        body: resp.body ?? "",
        headers: {
            "Content-Type": "application/json",
            ...corsHeaders,
            ...resp.headers,
        },
    };
}
/**
 * Map an API Gateway event to a Cloud API request
 * @param request API Gateway event
 * @returns Cloud API request
 */
function mapApigatewayEventToCloudApiRequest(request) {
    const query = {
        ...request.queryStringParameters,
        ...request.multiValueQueryStringParameters,
    };
    return {
        path: request.path,
        body: parseBody(request),
        headers: request.headers,
        method: (0, cloud_1.parseHttpMethod)(request.httpMethod),
        query: (0, cloud_1.sanitizeParamLikeObject)(query),
        vars: (0, cloud_1.sanitizeParamLikeObject)(request.pathParameters ?? {}),
    };
}
/**
 * Parse body to JSON or empty string
 * @param body body
 * @returns JSON body
 */
function parseBody(request) {
    if (!request.body)
        return "";
    const contentType = Object.entries(request.headers).find(([key, _]) => key.toLowerCase() === "content-type")?.[1];
    if (contentType === "application/x-www-form-urlencoded") {
        return JSON.stringify(Object.fromEntries(new URLSearchParams(request.body)));
    }
    return request.body;
}
