// @winglang/sdk/lib/shared-aws/counter.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterClient = void 0;
const client_dynamodb_1 = require("@aws-sdk/client-dynamodb");
const commons_1 = require("./commons");
const AMOUNT_TOKEN = "amount";
const INITIAL_VALUE_TOKEN = "initial";
const COUNTER_ID = "counter";
const VALUE_ATTRIBUTE = "counter_value";
const SET_VALUE = "set_value";
class CounterClient {
    constructor(tableName, initial = 0, client = new client_dynamodb_1.DynamoDBClient({})) {
        this.tableName = tableName;
        this.initial = initial;
        this.client = client;
    }
    async inc(amount = 1, key = COUNTER_ID) {
        const command = new client_dynamodb_1.UpdateItemCommand({
            TableName: this.tableName,
            Key: { [commons_1.COUNTER_HASH_KEY]: { S: key } },
            UpdateExpression: `SET ${VALUE_ATTRIBUTE} = if_not_exists(${VALUE_ATTRIBUTE}, :${INITIAL_VALUE_TOKEN}) + :${AMOUNT_TOKEN}`,
            ExpressionAttributeValues: {
                [`:${AMOUNT_TOKEN}`]: { N: `${amount}` },
                [`:${INITIAL_VALUE_TOKEN}`]: { N: `${this.initial}` },
            },
            ReturnValues: "UPDATED_NEW",
        });
        const result = await this.client.send(command);
        let newValue = result.Attributes?.[VALUE_ATTRIBUTE].N;
        if (!newValue) {
            throw new Error(`${VALUE_ATTRIBUTE} attribute not found on table.`);
        }
        // return the old value
        return parseInt(newValue) - amount;
    }
    async dec(amount = 1, key = COUNTER_ID) {
        return this.inc(-1 * amount, key);
    }
    async peek(key = COUNTER_ID) {
        const command = new client_dynamodb_1.GetItemCommand({
            TableName: this.tableName,
            Key: { [commons_1.COUNTER_HASH_KEY]: { S: key } },
        });
        const result = await this.client.send(command);
        let currentValue = result.Item?.[VALUE_ATTRIBUTE].N;
        if (currentValue === undefined) {
            return this.initial;
        }
        return parseInt(currentValue);
    }
    async set(value, key = COUNTER_ID) {
        const command = new client_dynamodb_1.UpdateItemCommand({
            TableName: this.tableName,
            Key: { [commons_1.COUNTER_HASH_KEY]: { S: key } },
            UpdateExpression: `SET ${VALUE_ATTRIBUTE} = :${SET_VALUE}`,
            ExpressionAttributeValues: {
                [`:${SET_VALUE}`]: { N: `${value}` },
            },
            ReturnValues: "UPDATED_NEW",
        });
        await this.client.send(command);
        return;
    }
}
exports.CounterClient = CounterClient;
