// @winglang/sdk/lib/shared-aws/domain.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Domain = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cloud = __importStar(require("../cloud"));
const core_1 = require("../core");
const errors_1 = require("../core/errors");
/**
 * AWS implementation of `cloud.Domain`.
 * @inflight `@winglang/sdk.cloud.IDomainClient`
 */
class Domain extends cloud.Domain {
    constructor(scope, id, props) {
        super(scope, id, props);
        const parameters = core_1.App.of(scope).parameters;
        // Domain requires parameters from the user, so we need to add the parameter schemas to the registrar
        let schema = {
            type: "object",
            oneOf: [
                {
                    required: ["iamCertificate"],
                },
                {
                    required: ["acmCertificateArn"],
                },
            ],
            required: ["hostedZoneId"],
            properties: {
                iamCertificate: {
                    type: "string",
                },
                acmCertificateArn: {
                    type: "string",
                },
                hostedZoneId: {
                    type: "string",
                },
            },
        };
        parameters.addSchemaAtPath(schema, this.node.path, true);
        const iamCertificate = parameters.value(`${this.node.path}/iamCertificate`);
        const acmCertificateArn = parameters.value(`${this.node.path}/acmCertificateArn`);
        const hostedZoneId = parameters.value(`${this.node.path}/hostedZoneId`);
        this._iamCertificate = iamCertificate;
        this._hostedZoneId = hostedZoneId;
        this._acmCertificateArn = acmCertificateArn;
    }
    /**
     * The hosted zone id value.
     */
    get hostedZoneId() {
        return this._hostedZoneId;
    }
    /**
     * The AWS Certificate Manager (ACM) certificate arn value.
     */
    get acmCertificateArn() {
        return this._acmCertificateArn;
    }
    /**
     * The IAM certificate identifier value.
     */
    get iamCertificate() {
        return this._iamCertificate;
    }
    /** @internal */
    _toInflight() {
        throw new errors_1.NotImplementedError("Domain inflight client is not implemented yet on this target.");
    }
}
exports.Domain = Domain;
_a = JSII_RTTI_SYMBOL_1;
Domain[_a] = { fqn: "@winglang/sdk.aws.Domain", version: "0.0.0" };
