// @cdktf/provider-aws/lib/data-aws-dms-replication-task/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDmsReplicationTask = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dms_replication_task aws_dms_replication_task}
*/
class DataAwsDmsReplicationTask extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDmsReplicationTask resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDmsReplicationTask to import
    * @param importFromId The id of the existing DataAwsDmsReplicationTask that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dms_replication_task#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDmsReplicationTask to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_dms_replication_task", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dms_replication_task aws_dms_replication_task} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDmsReplicationTaskConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dms_replication_task',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._replicationTaskId = config.replicationTaskId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // cdc_start_position - computed: true, optional: false, required: false
    get cdcStartPosition() {
        return this.getStringAttribute('cdc_start_position');
    }
    // cdc_start_time - computed: true, optional: false, required: false
    get cdcStartTime() {
        return this.getStringAttribute('cdc_start_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // migration_type - computed: true, optional: false, required: false
    get migrationType() {
        return this.getStringAttribute('migration_type');
    }
    // replication_instance_arn - computed: true, optional: false, required: false
    get replicationInstanceArn() {
        return this.getStringAttribute('replication_instance_arn');
    }
    // replication_task_arn - computed: true, optional: false, required: false
    get replicationTaskArn() {
        return this.getStringAttribute('replication_task_arn');
    }
    get replicationTaskId() {
        return this.getStringAttribute('replication_task_id');
    }
    set replicationTaskId(value) {
        this._replicationTaskId = value;
    }
    // Temporarily expose input value. Use with caution.
    get replicationTaskIdInput() {
        return this._replicationTaskId;
    }
    // replication_task_settings - computed: true, optional: false, required: false
    get replicationTaskSettings() {
        return this.getStringAttribute('replication_task_settings');
    }
    // source_endpoint_arn - computed: true, optional: false, required: false
    get sourceEndpointArn() {
        return this.getStringAttribute('source_endpoint_arn');
    }
    // start_replication_task - computed: true, optional: false, required: false
    get startReplicationTask() {
        return this.getBooleanAttribute('start_replication_task');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // table_mappings - computed: true, optional: false, required: false
    get tableMappings() {
        return this.getStringAttribute('table_mappings');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // target_endpoint_arn - computed: true, optional: false, required: false
    get targetEndpointArn() {
        return this.getStringAttribute('target_endpoint_arn');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            replication_task_id: cdktf.stringToTerraform(this._replicationTaskId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            replication_task_id: {
                value: cdktf.stringToHclTerraform(this._replicationTaskId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDmsReplicationTask = DataAwsDmsReplicationTask;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDmsReplicationTask[_a] = { fqn: "@cdktf/provider-aws.dataAwsDmsReplicationTask.DataAwsDmsReplicationTask", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDmsReplicationTask.tfResourceType = "aws_dms_replication_task";
