// @cdktf/provider-aws/lib/ssoadmin-instance-access-control-attributes/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminInstanceAccessControlAttributes = exports.SsoadminInstanceAccessControlAttributesAttributeList = exports.SsoadminInstanceAccessControlAttributesAttributeOutputReference = exports.ssoadminInstanceAccessControlAttributesAttributeToHclTerraform = exports.ssoadminInstanceAccessControlAttributesAttributeToTerraform = exports.SsoadminInstanceAccessControlAttributesAttributeValueList = exports.SsoadminInstanceAccessControlAttributesAttributeValueOutputReference = exports.ssoadminInstanceAccessControlAttributesAttributeValueToHclTerraform = exports.ssoadminInstanceAccessControlAttributesAttributeValueToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssoadminInstanceAccessControlAttributesAttributeValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.source),
    };
}
exports.ssoadminInstanceAccessControlAttributesAttributeValueToTerraform = ssoadminInstanceAccessControlAttributesAttributeValueToTerraform;
function ssoadminInstanceAccessControlAttributesAttributeValueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.source),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminInstanceAccessControlAttributesAttributeValueToHclTerraform = ssoadminInstanceAccessControlAttributesAttributeValueToHclTerraform;
class SsoadminInstanceAccessControlAttributesAttributeValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._source = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._source = value.source;
        }
    }
    get source() {
        return cdktf.Fn.tolist(this.getListAttribute('source'));
    }
    set source(value) {
        this._source = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
}
exports.SsoadminInstanceAccessControlAttributesAttributeValueOutputReference = SsoadminInstanceAccessControlAttributesAttributeValueOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsoadminInstanceAccessControlAttributesAttributeValueOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssoadminInstanceAccessControlAttributes.SsoadminInstanceAccessControlAttributesAttributeValueOutputReference", version: "19.15.0" };
class SsoadminInstanceAccessControlAttributesAttributeValueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminInstanceAccessControlAttributesAttributeValueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminInstanceAccessControlAttributesAttributeValueList = SsoadminInstanceAccessControlAttributesAttributeValueList;
_b = JSII_RTTI_SYMBOL_1;
SsoadminInstanceAccessControlAttributesAttributeValueList[_b] = { fqn: "@cdktf/provider-aws.ssoadminInstanceAccessControlAttributes.SsoadminInstanceAccessControlAttributesAttributeValueList", version: "19.15.0" };
function ssoadminInstanceAccessControlAttributesAttributeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.listMapper(ssoadminInstanceAccessControlAttributesAttributeValueToTerraform, true)(struct.value),
    };
}
exports.ssoadminInstanceAccessControlAttributesAttributeToTerraform = ssoadminInstanceAccessControlAttributesAttributeToTerraform;
function ssoadminInstanceAccessControlAttributesAttributeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.listMapperHcl(ssoadminInstanceAccessControlAttributesAttributeValueToHclTerraform, true)(struct.value),
            isBlock: true,
            type: "set",
            storageClassType: "SsoadminInstanceAccessControlAttributesAttributeValueList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminInstanceAccessControlAttributesAttributeToHclTerraform = ssoadminInstanceAccessControlAttributesAttributeToHclTerraform;
class SsoadminInstanceAccessControlAttributesAttributeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value - computed: false, optional: false, required: true
        this._value = new SsoadminInstanceAccessControlAttributesAttributeValueList(this, "value", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value.internalValue = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this._value;
    }
    putValue(value) {
        this._value.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value.internalValue;
    }
}
exports.SsoadminInstanceAccessControlAttributesAttributeOutputReference = SsoadminInstanceAccessControlAttributesAttributeOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SsoadminInstanceAccessControlAttributesAttributeOutputReference[_c] = { fqn: "@cdktf/provider-aws.ssoadminInstanceAccessControlAttributes.SsoadminInstanceAccessControlAttributesAttributeOutputReference", version: "19.15.0" };
class SsoadminInstanceAccessControlAttributesAttributeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminInstanceAccessControlAttributesAttributeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminInstanceAccessControlAttributesAttributeList = SsoadminInstanceAccessControlAttributesAttributeList;
_d = JSII_RTTI_SYMBOL_1;
SsoadminInstanceAccessControlAttributesAttributeList[_d] = { fqn: "@cdktf/provider-aws.ssoadminInstanceAccessControlAttributes.SsoadminInstanceAccessControlAttributesAttributeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_instance_access_control_attributes aws_ssoadmin_instance_access_control_attributes}
*/
class SsoadminInstanceAccessControlAttributes extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsoadminInstanceAccessControlAttributes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsoadminInstanceAccessControlAttributes to import
    * @param importFromId The id of the existing SsoadminInstanceAccessControlAttributes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_instance_access_control_attributes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsoadminInstanceAccessControlAttributes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_instance_access_control_attributes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_instance_access_control_attributes aws_ssoadmin_instance_access_control_attributes} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsoadminInstanceAccessControlAttributesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_instance_access_control_attributes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attribute - computed: false, optional: false, required: true
        this._attribute = new SsoadminInstanceAccessControlAttributesAttributeList(this, "attribute", true);
        this._id = config.id;
        this._instanceArn = config.instanceArn;
        this._attribute.internalValue = config.attribute;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceArnInput() {
        return this._instanceArn;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_reason - computed: true, optional: false, required: false
    get statusReason() {
        return this.getStringAttribute('status_reason');
    }
    get attribute() {
        return this._attribute;
    }
    putAttribute(value) {
        this._attribute.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get attributeInput() {
        return this._attribute.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            attribute: cdktf.listMapper(ssoadminInstanceAccessControlAttributesAttributeToTerraform, true)(this._attribute.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_arn: {
                value: cdktf.stringToHclTerraform(this._instanceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            attribute: {
                value: cdktf.listMapperHcl(ssoadminInstanceAccessControlAttributesAttributeToHclTerraform, true)(this._attribute.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "SsoadminInstanceAccessControlAttributesAttributeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsoadminInstanceAccessControlAttributes = SsoadminInstanceAccessControlAttributes;
_e = JSII_RTTI_SYMBOL_1;
SsoadminInstanceAccessControlAttributes[_e] = { fqn: "@cdktf/provider-aws.ssoadminInstanceAccessControlAttributes.SsoadminInstanceAccessControlAttributes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsoadminInstanceAccessControlAttributes.tfResourceType = "aws_ssoadmin_instance_access_control_attributes";
