// @cdktf/provider-aws/lib/vpc-dhcp-options/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcDhcpOptions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_dhcp_options aws_vpc_dhcp_options}
*/
class VpcDhcpOptions extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VpcDhcpOptions resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VpcDhcpOptions to import
    * @param importFromId The id of the existing VpcDhcpOptions that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_dhcp_options#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VpcDhcpOptions to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpc_dhcp_options", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_dhcp_options aws_vpc_dhcp_options} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcDhcpOptionsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_dhcp_options',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._domainName = config.domainName;
        this._domainNameServers = config.domainNameServers;
        this._id = config.id;
        this._ipv6AddressPreferredLeaseTime = config.ipv6AddressPreferredLeaseTime;
        this._netbiosNameServers = config.netbiosNameServers;
        this._netbiosNodeType = config.netbiosNodeType;
        this._ntpServers = config.ntpServers;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    resetDomainName() {
        this._domainName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get domainNameServers() {
        return this.getListAttribute('domain_name_servers');
    }
    set domainNameServers(value) {
        this._domainNameServers = value;
    }
    resetDomainNameServers() {
        this._domainNameServers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameServersInput() {
        return this._domainNameServers;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipv6AddressPreferredLeaseTime() {
        return this.getStringAttribute('ipv6_address_preferred_lease_time');
    }
    set ipv6AddressPreferredLeaseTime(value) {
        this._ipv6AddressPreferredLeaseTime = value;
    }
    resetIpv6AddressPreferredLeaseTime() {
        this._ipv6AddressPreferredLeaseTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressPreferredLeaseTimeInput() {
        return this._ipv6AddressPreferredLeaseTime;
    }
    get netbiosNameServers() {
        return this.getListAttribute('netbios_name_servers');
    }
    set netbiosNameServers(value) {
        this._netbiosNameServers = value;
    }
    resetNetbiosNameServers() {
        this._netbiosNameServers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get netbiosNameServersInput() {
        return this._netbiosNameServers;
    }
    get netbiosNodeType() {
        return this.getStringAttribute('netbios_node_type');
    }
    set netbiosNodeType(value) {
        this._netbiosNodeType = value;
    }
    resetNetbiosNodeType() {
        this._netbiosNodeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get netbiosNodeTypeInput() {
        return this._netbiosNodeType;
    }
    get ntpServers() {
        return this.getListAttribute('ntp_servers');
    }
    set ntpServers(value) {
        this._ntpServers = value;
    }
    resetNtpServers() {
        this._ntpServers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ntpServersInput() {
        return this._ntpServers;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            domain_name_servers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._domainNameServers),
            id: cdktf.stringToTerraform(this._id),
            ipv6_address_preferred_lease_time: cdktf.stringToTerraform(this._ipv6AddressPreferredLeaseTime),
            netbios_name_servers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._netbiosNameServers),
            netbios_node_type: cdktf.stringToTerraform(this._netbiosNodeType),
            ntp_servers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ntpServers),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            domain_name_servers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._domainNameServers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ipv6_address_preferred_lease_time: {
                value: cdktf.stringToHclTerraform(this._ipv6AddressPreferredLeaseTime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            netbios_name_servers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._netbiosNameServers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            netbios_node_type: {
                value: cdktf.stringToHclTerraform(this._netbiosNodeType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ntp_servers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._ntpServers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VpcDhcpOptions = VpcDhcpOptions;
_a = JSII_RTTI_SYMBOL_1;
VpcDhcpOptions[_a] = { fqn: "@cdktf/provider-aws.vpcDhcpOptions.VpcDhcpOptions", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VpcDhcpOptions.tfResourceType = "aws_vpc_dhcp_options";
