// @cdktf/provider-aws/lib/acmpca-certificate-authority/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcmpcaCertificateAuthority = exports.AcmpcaCertificateAuthorityTimeoutsOutputReference = exports.acmpcaCertificateAuthorityTimeoutsToHclTerraform = exports.acmpcaCertificateAuthorityTimeoutsToTerraform = exports.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference = exports.acmpcaCertificateAuthorityRevocationConfigurationToHclTerraform = exports.acmpcaCertificateAuthorityRevocationConfigurationToTerraform = exports.AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference = exports.acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToHclTerraform = exports.acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToTerraform = exports.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference = exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToHclTerraform = exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToHclTerraform = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform = exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToHclTerraform = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        common_name: cdktf.stringToTerraform(struct.commonName),
        country: cdktf.stringToTerraform(struct.country),
        distinguished_name_qualifier: cdktf.stringToTerraform(struct.distinguishedNameQualifier),
        generation_qualifier: cdktf.stringToTerraform(struct.generationQualifier),
        given_name: cdktf.stringToTerraform(struct.givenName),
        initials: cdktf.stringToTerraform(struct.initials),
        locality: cdktf.stringToTerraform(struct.locality),
        organization: cdktf.stringToTerraform(struct.organization),
        organizational_unit: cdktf.stringToTerraform(struct.organizationalUnit),
        pseudonym: cdktf.stringToTerraform(struct.pseudonym),
        state: cdktf.stringToTerraform(struct.state),
        surname: cdktf.stringToTerraform(struct.surname),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform;
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        common_name: {
            value: cdktf.stringToHclTerraform(struct.commonName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        country: {
            value: cdktf.stringToHclTerraform(struct.country),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        distinguished_name_qualifier: {
            value: cdktf.stringToHclTerraform(struct.distinguishedNameQualifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        generation_qualifier: {
            value: cdktf.stringToHclTerraform(struct.generationQualifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        given_name: {
            value: cdktf.stringToHclTerraform(struct.givenName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        initials: {
            value: cdktf.stringToHclTerraform(struct.initials),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        locality: {
            value: cdktf.stringToHclTerraform(struct.locality),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        organization: {
            value: cdktf.stringToHclTerraform(struct.organization),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        organizational_unit: {
            value: cdktf.stringToHclTerraform(struct.organizationalUnit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        pseudonym: {
            value: cdktf.stringToHclTerraform(struct.pseudonym),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        state: {
            value: cdktf.stringToHclTerraform(struct.state),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        surname: {
            value: cdktf.stringToHclTerraform(struct.surname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        title: {
            value: cdktf.stringToHclTerraform(struct.title),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToHclTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToHclTerraform;
class AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._commonName !== undefined) {
            hasAnyValues = true;
            internalValueResult.commonName = this._commonName;
        }
        if (this._country !== undefined) {
            hasAnyValues = true;
            internalValueResult.country = this._country;
        }
        if (this._distinguishedNameQualifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.distinguishedNameQualifier = this._distinguishedNameQualifier;
        }
        if (this._generationQualifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.generationQualifier = this._generationQualifier;
        }
        if (this._givenName !== undefined) {
            hasAnyValues = true;
            internalValueResult.givenName = this._givenName;
        }
        if (this._initials !== undefined) {
            hasAnyValues = true;
            internalValueResult.initials = this._initials;
        }
        if (this._locality !== undefined) {
            hasAnyValues = true;
            internalValueResult.locality = this._locality;
        }
        if (this._organization !== undefined) {
            hasAnyValues = true;
            internalValueResult.organization = this._organization;
        }
        if (this._organizationalUnit !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationalUnit = this._organizationalUnit;
        }
        if (this._pseudonym !== undefined) {
            hasAnyValues = true;
            internalValueResult.pseudonym = this._pseudonym;
        }
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        if (this._surname !== undefined) {
            hasAnyValues = true;
            internalValueResult.surname = this._surname;
        }
        if (this._title !== undefined) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._commonName = undefined;
            this._country = undefined;
            this._distinguishedNameQualifier = undefined;
            this._generationQualifier = undefined;
            this._givenName = undefined;
            this._initials = undefined;
            this._locality = undefined;
            this._organization = undefined;
            this._organizationalUnit = undefined;
            this._pseudonym = undefined;
            this._state = undefined;
            this._surname = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._commonName = value.commonName;
            this._country = value.country;
            this._distinguishedNameQualifier = value.distinguishedNameQualifier;
            this._generationQualifier = value.generationQualifier;
            this._givenName = value.givenName;
            this._initials = value.initials;
            this._locality = value.locality;
            this._organization = value.organization;
            this._organizationalUnit = value.organizationalUnit;
            this._pseudonym = value.pseudonym;
            this._state = value.state;
            this._surname = value.surname;
            this._title = value.title;
        }
    }
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    set commonName(value) {
        this._commonName = value;
    }
    resetCommonName() {
        this._commonName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commonNameInput() {
        return this._commonName;
    }
    get country() {
        return this.getStringAttribute('country');
    }
    set country(value) {
        this._country = value;
    }
    resetCountry() {
        this._country = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryInput() {
        return this._country;
    }
    get distinguishedNameQualifier() {
        return this.getStringAttribute('distinguished_name_qualifier');
    }
    set distinguishedNameQualifier(value) {
        this._distinguishedNameQualifier = value;
    }
    resetDistinguishedNameQualifier() {
        this._distinguishedNameQualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get distinguishedNameQualifierInput() {
        return this._distinguishedNameQualifier;
    }
    get generationQualifier() {
        return this.getStringAttribute('generation_qualifier');
    }
    set generationQualifier(value) {
        this._generationQualifier = value;
    }
    resetGenerationQualifier() {
        this._generationQualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get generationQualifierInput() {
        return this._generationQualifier;
    }
    get givenName() {
        return this.getStringAttribute('given_name');
    }
    set givenName(value) {
        this._givenName = value;
    }
    resetGivenName() {
        this._givenName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get givenNameInput() {
        return this._givenName;
    }
    get initials() {
        return this.getStringAttribute('initials');
    }
    set initials(value) {
        this._initials = value;
    }
    resetInitials() {
        this._initials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialsInput() {
        return this._initials;
    }
    get locality() {
        return this.getStringAttribute('locality');
    }
    set locality(value) {
        this._locality = value;
    }
    resetLocality() {
        this._locality = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localityInput() {
        return this._locality;
    }
    get organization() {
        return this.getStringAttribute('organization');
    }
    set organization(value) {
        this._organization = value;
    }
    resetOrganization() {
        this._organization = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationInput() {
        return this._organization;
    }
    get organizationalUnit() {
        return this.getStringAttribute('organizational_unit');
    }
    set organizationalUnit(value) {
        this._organizationalUnit = value;
    }
    resetOrganizationalUnit() {
        this._organizationalUnit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitInput() {
        return this._organizationalUnit;
    }
    get pseudonym() {
        return this.getStringAttribute('pseudonym');
    }
    set pseudonym(value) {
        this._pseudonym = value;
    }
    resetPseudonym() {
        this._pseudonym = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pseudonymInput() {
        return this._pseudonym;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get surname() {
        return this.getStringAttribute('surname');
    }
    set surname(value) {
        this._surname = value;
    }
    resetSurname() {
        this._surname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get surnameInput() {
        return this._surname;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
}
exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference = AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference[_a] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference", version: "19.15.0" };
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_algorithm: cdktf.stringToTerraform(struct.keyAlgorithm),
        signing_algorithm: cdktf.stringToTerraform(struct.signingAlgorithm),
        subject: acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform(struct.subject),
    };
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform;
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key_algorithm: {
            value: cdktf.stringToHclTerraform(struct.keyAlgorithm),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        signing_algorithm: {
            value: cdktf.stringToHclTerraform(struct.signingAlgorithm),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subject: {
            value: acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToHclTerraform(struct.subject),
            isBlock: true,
            type: "list",
            storageClassType: "AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToHclTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationToHclTerraform;
class AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // subject - computed: false, optional: false, required: true
        this._subject = new AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference(this, "subject");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyAlgorithm !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyAlgorithm = this._keyAlgorithm;
        }
        if (this._signingAlgorithm !== undefined) {
            hasAnyValues = true;
            internalValueResult.signingAlgorithm = this._signingAlgorithm;
        }
        if (this._subject?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.subject = this._subject?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyAlgorithm = undefined;
            this._signingAlgorithm = undefined;
            this._subject.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyAlgorithm = value.keyAlgorithm;
            this._signingAlgorithm = value.signingAlgorithm;
            this._subject.internalValue = value.subject;
        }
    }
    get keyAlgorithm() {
        return this.getStringAttribute('key_algorithm');
    }
    set keyAlgorithm(value) {
        this._keyAlgorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyAlgorithmInput() {
        return this._keyAlgorithm;
    }
    get signingAlgorithm() {
        return this.getStringAttribute('signing_algorithm');
    }
    set signingAlgorithm(value) {
        this._signingAlgorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    get signingAlgorithmInput() {
        return this._signingAlgorithm;
    }
    get subject() {
        return this._subject;
    }
    putSubject(value) {
        this._subject.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get subjectInput() {
        return this._subject.internalValue;
    }
}
exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference = AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference", version: "19.15.0" };
function acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_cname: cdktf.stringToTerraform(struct.customCname),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        expiration_in_days: cdktf.numberToTerraform(struct.expirationInDays),
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_object_acl: cdktf.stringToTerraform(struct.s3ObjectAcl),
    };
}
exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform;
function acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        custom_cname: {
            value: cdktf.stringToHclTerraform(struct.customCname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        expiration_in_days: {
            value: cdktf.numberToHclTerraform(struct.expirationInDays),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        s3_bucket_name: {
            value: cdktf.stringToHclTerraform(struct.s3BucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_object_acl: {
            value: cdktf.stringToHclTerraform(struct.s3ObjectAcl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToHclTerraform = acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToHclTerraform;
class AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customCname !== undefined) {
            hasAnyValues = true;
            internalValueResult.customCname = this._customCname;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._expirationInDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.expirationInDays = this._expirationInDays;
        }
        if (this._s3BucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3ObjectAcl !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3ObjectAcl = this._s3ObjectAcl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customCname = undefined;
            this._enabled = undefined;
            this._expirationInDays = undefined;
            this._s3BucketName = undefined;
            this._s3ObjectAcl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customCname = value.customCname;
            this._enabled = value.enabled;
            this._expirationInDays = value.expirationInDays;
            this._s3BucketName = value.s3BucketName;
            this._s3ObjectAcl = value.s3ObjectAcl;
        }
    }
    get customCname() {
        return this.getStringAttribute('custom_cname');
    }
    set customCname(value) {
        this._customCname = value;
    }
    resetCustomCname() {
        this._customCname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customCnameInput() {
        return this._customCname;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get expirationInDays() {
        return this.getNumberAttribute('expiration_in_days');
    }
    set expirationInDays(value) {
        this._expirationInDays = value;
    }
    resetExpirationInDays() {
        this._expirationInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expirationInDaysInput() {
        return this._expirationInDays;
    }
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    resetS3BucketName() {
        this._s3BucketName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    get s3ObjectAcl() {
        return this.getStringAttribute('s3_object_acl');
    }
    set s3ObjectAcl(value) {
        this._s3ObjectAcl = value;
    }
    resetS3ObjectAcl() {
        this._s3ObjectAcl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3ObjectAclInput() {
        return this._s3ObjectAcl;
    }
}
exports.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference = AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference", version: "19.15.0" };
function acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        ocsp_custom_cname: cdktf.stringToTerraform(struct.ocspCustomCname),
    };
}
exports.acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToTerraform = acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToTerraform;
function acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        ocsp_custom_cname: {
            value: cdktf.stringToHclTerraform(struct.ocspCustomCname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToHclTerraform = acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToHclTerraform;
class AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._ocspCustomCname !== undefined) {
            hasAnyValues = true;
            internalValueResult.ocspCustomCname = this._ocspCustomCname;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._ocspCustomCname = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._ocspCustomCname = value.ocspCustomCname;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get ocspCustomCname() {
        return this.getStringAttribute('ocsp_custom_cname');
    }
    set ocspCustomCname(value) {
        this._ocspCustomCname = value;
    }
    resetOcspCustomCname() {
        this._ocspCustomCname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ocspCustomCnameInput() {
        return this._ocspCustomCname;
    }
}
exports.AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference = AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference[_d] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference", version: "19.15.0" };
function acmpcaCertificateAuthorityRevocationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crl_configuration: acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform(struct.crlConfiguration),
        ocsp_configuration: acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToTerraform(struct.ocspConfiguration),
    };
}
exports.acmpcaCertificateAuthorityRevocationConfigurationToTerraform = acmpcaCertificateAuthorityRevocationConfigurationToTerraform;
function acmpcaCertificateAuthorityRevocationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        crl_configuration: {
            value: acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToHclTerraform(struct.crlConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationList",
        },
        ocsp_configuration: {
            value: acmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationToHclTerraform(struct.ocspConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityRevocationConfigurationToHclTerraform = acmpcaCertificateAuthorityRevocationConfigurationToHclTerraform;
class AcmpcaCertificateAuthorityRevocationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // crl_configuration - computed: false, optional: true, required: false
        this._crlConfiguration = new AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference(this, "crl_configuration");
        // ocsp_configuration - computed: false, optional: true, required: false
        this._ocspConfiguration = new AcmpcaCertificateAuthorityRevocationConfigurationOcspConfigurationOutputReference(this, "ocsp_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._crlConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.crlConfiguration = this._crlConfiguration?.internalValue;
        }
        if (this._ocspConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ocspConfiguration = this._ocspConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._crlConfiguration.internalValue = undefined;
            this._ocspConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._crlConfiguration.internalValue = value.crlConfiguration;
            this._ocspConfiguration.internalValue = value.ocspConfiguration;
        }
    }
    get crlConfiguration() {
        return this._crlConfiguration;
    }
    putCrlConfiguration(value) {
        this._crlConfiguration.internalValue = value;
    }
    resetCrlConfiguration() {
        this._crlConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get crlConfigurationInput() {
        return this._crlConfiguration.internalValue;
    }
    get ocspConfiguration() {
        return this._ocspConfiguration;
    }
    putOcspConfiguration(value) {
        this._ocspConfiguration.internalValue = value;
    }
    resetOcspConfiguration() {
        this._ocspConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ocspConfigurationInput() {
        return this._ocspConfiguration.internalValue;
    }
}
exports.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference = AcmpcaCertificateAuthorityRevocationConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityRevocationConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference", version: "19.15.0" };
function acmpcaCertificateAuthorityTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.acmpcaCertificateAuthorityTimeoutsToTerraform = acmpcaCertificateAuthorityTimeoutsToTerraform;
function acmpcaCertificateAuthorityTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateAuthorityTimeoutsToHclTerraform = acmpcaCertificateAuthorityTimeoutsToHclTerraform;
class AcmpcaCertificateAuthorityTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.AcmpcaCertificateAuthorityTimeoutsOutputReference = AcmpcaCertificateAuthorityTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthorityTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate_authority aws_acmpca_certificate_authority}
*/
class AcmpcaCertificateAuthority extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AcmpcaCertificateAuthority resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AcmpcaCertificateAuthority to import
    * @param importFromId The id of the existing AcmpcaCertificateAuthority that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate_authority#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AcmpcaCertificateAuthority to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_acmpca_certificate_authority", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate_authority aws_acmpca_certificate_authority} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AcmpcaCertificateAuthorityConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_acmpca_certificate_authority',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // certificate_authority_configuration - computed: false, optional: false, required: true
        this._certificateAuthorityConfiguration = new AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference(this, "certificate_authority_configuration");
        // revocation_configuration - computed: false, optional: true, required: false
        this._revocationConfiguration = new AcmpcaCertificateAuthorityRevocationConfigurationOutputReference(this, "revocation_configuration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new AcmpcaCertificateAuthorityTimeoutsOutputReference(this, "timeouts");
        this._enabled = config.enabled;
        this._id = config.id;
        this._keyStorageSecurityStandard = config.keyStorageSecurityStandard;
        this._permanentDeletionTimeInDays = config.permanentDeletionTimeInDays;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._usageMode = config.usageMode;
        this._certificateAuthorityConfiguration.internalValue = config.certificateAuthorityConfiguration;
        this._revocationConfiguration.internalValue = config.revocationConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate - computed: true, optional: false, required: false
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    // certificate_chain - computed: true, optional: false, required: false
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    // certificate_signing_request - computed: true, optional: false, required: false
    get certificateSigningRequest() {
        return this.getStringAttribute('certificate_signing_request');
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyStorageSecurityStandard() {
        return this.getStringAttribute('key_storage_security_standard');
    }
    set keyStorageSecurityStandard(value) {
        this._keyStorageSecurityStandard = value;
    }
    resetKeyStorageSecurityStandard() {
        this._keyStorageSecurityStandard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyStorageSecurityStandardInput() {
        return this._keyStorageSecurityStandard;
    }
    // not_after - computed: true, optional: false, required: false
    get notAfter() {
        return this.getStringAttribute('not_after');
    }
    // not_before - computed: true, optional: false, required: false
    get notBefore() {
        return this.getStringAttribute('not_before');
    }
    get permanentDeletionTimeInDays() {
        return this.getNumberAttribute('permanent_deletion_time_in_days');
    }
    set permanentDeletionTimeInDays(value) {
        this._permanentDeletionTimeInDays = value;
    }
    resetPermanentDeletionTimeInDays() {
        this._permanentDeletionTimeInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permanentDeletionTimeInDaysInput() {
        return this._permanentDeletionTimeInDays;
    }
    // serial - computed: true, optional: false, required: false
    get serial() {
        return this.getStringAttribute('serial');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get usageMode() {
        return this.getStringAttribute('usage_mode');
    }
    set usageMode(value) {
        this._usageMode = value;
    }
    resetUsageMode() {
        this._usageMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get usageModeInput() {
        return this._usageMode;
    }
    get certificateAuthorityConfiguration() {
        return this._certificateAuthorityConfiguration;
    }
    putCertificateAuthorityConfiguration(value) {
        this._certificateAuthorityConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get certificateAuthorityConfigurationInput() {
        return this._certificateAuthorityConfiguration.internalValue;
    }
    get revocationConfiguration() {
        return this._revocationConfiguration;
    }
    putRevocationConfiguration(value) {
        this._revocationConfiguration.internalValue = value;
    }
    resetRevocationConfiguration() {
        this._revocationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revocationConfigurationInput() {
        return this._revocationConfiguration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            key_storage_security_standard: cdktf.stringToTerraform(this._keyStorageSecurityStandard),
            permanent_deletion_time_in_days: cdktf.numberToTerraform(this._permanentDeletionTimeInDays),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            usage_mode: cdktf.stringToTerraform(this._usageMode),
            certificate_authority_configuration: acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform(this._certificateAuthorityConfiguration.internalValue),
            revocation_configuration: acmpcaCertificateAuthorityRevocationConfigurationToTerraform(this._revocationConfiguration.internalValue),
            timeouts: acmpcaCertificateAuthorityTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key_storage_security_standard: {
                value: cdktf.stringToHclTerraform(this._keyStorageSecurityStandard),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permanent_deletion_time_in_days: {
                value: cdktf.numberToHclTerraform(this._permanentDeletionTimeInDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            usage_mode: {
                value: cdktf.stringToHclTerraform(this._usageMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            certificate_authority_configuration: {
                value: acmpcaCertificateAuthorityCertificateAuthorityConfigurationToHclTerraform(this._certificateAuthorityConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AcmpcaCertificateAuthorityCertificateAuthorityConfigurationList",
            },
            revocation_configuration: {
                value: acmpcaCertificateAuthorityRevocationConfigurationToHclTerraform(this._revocationConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AcmpcaCertificateAuthorityRevocationConfigurationList",
            },
            timeouts: {
                value: acmpcaCertificateAuthorityTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "AcmpcaCertificateAuthorityTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AcmpcaCertificateAuthority = AcmpcaCertificateAuthority;
_g = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthority[_g] = { fqn: "@cdktf/provider-aws.acmpcaCertificateAuthority.AcmpcaCertificateAuthority", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AcmpcaCertificateAuthority.tfResourceType = "aws_acmpca_certificate_authority";
