// @cdktf/provider-aws/lib/lambda-function/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaFunction = exports.LambdaFunctionVpcConfigOutputReference = exports.lambdaFunctionVpcConfigToHclTerraform = exports.lambdaFunctionVpcConfigToTerraform = exports.LambdaFunctionTracingConfigOutputReference = exports.lambdaFunctionTracingConfigToHclTerraform = exports.lambdaFunctionTracingConfigToTerraform = exports.LambdaFunctionTimeoutsOutputReference = exports.lambdaFunctionTimeoutsToHclTerraform = exports.lambdaFunctionTimeoutsToTerraform = exports.LambdaFunctionSnapStartOutputReference = exports.lambdaFunctionSnapStartToHclTerraform = exports.lambdaFunctionSnapStartToTerraform = exports.LambdaFunctionLoggingConfigOutputReference = exports.lambdaFunctionLoggingConfigToHclTerraform = exports.lambdaFunctionLoggingConfigToTerraform = exports.LambdaFunctionImageConfigOutputReference = exports.lambdaFunctionImageConfigToHclTerraform = exports.lambdaFunctionImageConfigToTerraform = exports.LambdaFunctionFileSystemConfigOutputReference = exports.lambdaFunctionFileSystemConfigToHclTerraform = exports.lambdaFunctionFileSystemConfigToTerraform = exports.LambdaFunctionEphemeralStorageOutputReference = exports.lambdaFunctionEphemeralStorageToHclTerraform = exports.lambdaFunctionEphemeralStorageToTerraform = exports.LambdaFunctionEnvironmentOutputReference = exports.lambdaFunctionEnvironmentToHclTerraform = exports.lambdaFunctionEnvironmentToTerraform = exports.LambdaFunctionDeadLetterConfigOutputReference = exports.lambdaFunctionDeadLetterConfigToHclTerraform = exports.lambdaFunctionDeadLetterConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lambdaFunctionDeadLetterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_arn: cdktf.stringToTerraform(struct.targetArn),
    };
}
exports.lambdaFunctionDeadLetterConfigToTerraform = lambdaFunctionDeadLetterConfigToTerraform;
function lambdaFunctionDeadLetterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target_arn: {
            value: cdktf.stringToHclTerraform(struct.targetArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionDeadLetterConfigToHclTerraform = lambdaFunctionDeadLetterConfigToHclTerraform;
class LambdaFunctionDeadLetterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetArn = this._targetArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._targetArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._targetArn = value.targetArn;
        }
    }
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    set targetArn(value) {
        this._targetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetArnInput() {
        return this._targetArn;
    }
}
exports.LambdaFunctionDeadLetterConfigOutputReference = LambdaFunctionDeadLetterConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LambdaFunctionDeadLetterConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionDeadLetterConfigOutputReference", version: "19.15.0" };
function lambdaFunctionEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        variables: cdktf.hashMapper(cdktf.stringToTerraform)(struct.variables),
    };
}
exports.lambdaFunctionEnvironmentToTerraform = lambdaFunctionEnvironmentToTerraform;
function lambdaFunctionEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        variables: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.variables),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionEnvironmentToHclTerraform = lambdaFunctionEnvironmentToHclTerraform;
class LambdaFunctionEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._variables !== undefined) {
            hasAnyValues = true;
            internalValueResult.variables = this._variables;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._variables = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._variables = value.variables;
        }
    }
    get variables() {
        return this.getStringMapAttribute('variables');
    }
    set variables(value) {
        this._variables = value;
    }
    resetVariables() {
        this._variables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get variablesInput() {
        return this._variables;
    }
}
exports.LambdaFunctionEnvironmentOutputReference = LambdaFunctionEnvironmentOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LambdaFunctionEnvironmentOutputReference[_b] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionEnvironmentOutputReference", version: "19.15.0" };
function lambdaFunctionEphemeralStorageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        size: cdktf.numberToTerraform(struct.size),
    };
}
exports.lambdaFunctionEphemeralStorageToTerraform = lambdaFunctionEphemeralStorageToTerraform;
function lambdaFunctionEphemeralStorageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        size: {
            value: cdktf.numberToHclTerraform(struct.size),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionEphemeralStorageToHclTerraform = lambdaFunctionEphemeralStorageToHclTerraform;
class LambdaFunctionEphemeralStorageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._size = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._size = value.size;
        }
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    resetSize() {
        this._size = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
}
exports.LambdaFunctionEphemeralStorageOutputReference = LambdaFunctionEphemeralStorageOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LambdaFunctionEphemeralStorageOutputReference[_c] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionEphemeralStorageOutputReference", version: "19.15.0" };
function lambdaFunctionFileSystemConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        local_mount_path: cdktf.stringToTerraform(struct.localMountPath),
    };
}
exports.lambdaFunctionFileSystemConfigToTerraform = lambdaFunctionFileSystemConfigToTerraform;
function lambdaFunctionFileSystemConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        local_mount_path: {
            value: cdktf.stringToHclTerraform(struct.localMountPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionFileSystemConfigToHclTerraform = lambdaFunctionFileSystemConfigToHclTerraform;
class LambdaFunctionFileSystemConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._localMountPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.localMountPath = this._localMountPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._localMountPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._localMountPath = value.localMountPath;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get localMountPath() {
        return this.getStringAttribute('local_mount_path');
    }
    set localMountPath(value) {
        this._localMountPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get localMountPathInput() {
        return this._localMountPath;
    }
}
exports.LambdaFunctionFileSystemConfigOutputReference = LambdaFunctionFileSystemConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
LambdaFunctionFileSystemConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionFileSystemConfigOutputReference", version: "19.15.0" };
function lambdaFunctionImageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        command: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.command),
        entry_point: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.entryPoint),
        working_directory: cdktf.stringToTerraform(struct.workingDirectory),
    };
}
exports.lambdaFunctionImageConfigToTerraform = lambdaFunctionImageConfigToTerraform;
function lambdaFunctionImageConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        command: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.command),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        entry_point: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.entryPoint),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        working_directory: {
            value: cdktf.stringToHclTerraform(struct.workingDirectory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionImageConfigToHclTerraform = lambdaFunctionImageConfigToHclTerraform;
class LambdaFunctionImageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._command !== undefined) {
            hasAnyValues = true;
            internalValueResult.command = this._command;
        }
        if (this._entryPoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.entryPoint = this._entryPoint;
        }
        if (this._workingDirectory !== undefined) {
            hasAnyValues = true;
            internalValueResult.workingDirectory = this._workingDirectory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._command = undefined;
            this._entryPoint = undefined;
            this._workingDirectory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._command = value.command;
            this._entryPoint = value.entryPoint;
            this._workingDirectory = value.workingDirectory;
        }
    }
    get command() {
        return this.getListAttribute('command');
    }
    set command(value) {
        this._command = value;
    }
    resetCommand() {
        this._command = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commandInput() {
        return this._command;
    }
    get entryPoint() {
        return this.getListAttribute('entry_point');
    }
    set entryPoint(value) {
        this._entryPoint = value;
    }
    resetEntryPoint() {
        this._entryPoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entryPointInput() {
        return this._entryPoint;
    }
    get workingDirectory() {
        return this.getStringAttribute('working_directory');
    }
    set workingDirectory(value) {
        this._workingDirectory = value;
    }
    resetWorkingDirectory() {
        this._workingDirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workingDirectoryInput() {
        return this._workingDirectory;
    }
}
exports.LambdaFunctionImageConfigOutputReference = LambdaFunctionImageConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
LambdaFunctionImageConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionImageConfigOutputReference", version: "19.15.0" };
function lambdaFunctionLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        application_log_level: cdktf.stringToTerraform(struct.applicationLogLevel),
        log_format: cdktf.stringToTerraform(struct.logFormat),
        log_group: cdktf.stringToTerraform(struct.logGroup),
        system_log_level: cdktf.stringToTerraform(struct.systemLogLevel),
    };
}
exports.lambdaFunctionLoggingConfigToTerraform = lambdaFunctionLoggingConfigToTerraform;
function lambdaFunctionLoggingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        application_log_level: {
            value: cdktf.stringToHclTerraform(struct.applicationLogLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        log_format: {
            value: cdktf.stringToHclTerraform(struct.logFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        log_group: {
            value: cdktf.stringToHclTerraform(struct.logGroup),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        system_log_level: {
            value: cdktf.stringToHclTerraform(struct.systemLogLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionLoggingConfigToHclTerraform = lambdaFunctionLoggingConfigToHclTerraform;
class LambdaFunctionLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._applicationLogLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.applicationLogLevel = this._applicationLogLevel;
        }
        if (this._logFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.logFormat = this._logFormat;
        }
        if (this._logGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.logGroup = this._logGroup;
        }
        if (this._systemLogLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.systemLogLevel = this._systemLogLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._applicationLogLevel = undefined;
            this._logFormat = undefined;
            this._logGroup = undefined;
            this._systemLogLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._applicationLogLevel = value.applicationLogLevel;
            this._logFormat = value.logFormat;
            this._logGroup = value.logGroup;
            this._systemLogLevel = value.systemLogLevel;
        }
    }
    get applicationLogLevel() {
        return this.getStringAttribute('application_log_level');
    }
    set applicationLogLevel(value) {
        this._applicationLogLevel = value;
    }
    resetApplicationLogLevel() {
        this._applicationLogLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationLogLevelInput() {
        return this._applicationLogLevel;
    }
    get logFormat() {
        return this.getStringAttribute('log_format');
    }
    set logFormat(value) {
        this._logFormat = value;
    }
    // Temporarily expose input value. Use with caution.
    get logFormatInput() {
        return this._logFormat;
    }
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    set logGroup(value) {
        this._logGroup = value;
    }
    resetLogGroup() {
        this._logGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupInput() {
        return this._logGroup;
    }
    get systemLogLevel() {
        return this.getStringAttribute('system_log_level');
    }
    set systemLogLevel(value) {
        this._systemLogLevel = value;
    }
    resetSystemLogLevel() {
        this._systemLogLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get systemLogLevelInput() {
        return this._systemLogLevel;
    }
}
exports.LambdaFunctionLoggingConfigOutputReference = LambdaFunctionLoggingConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
LambdaFunctionLoggingConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionLoggingConfigOutputReference", version: "19.15.0" };
function lambdaFunctionSnapStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        apply_on: cdktf.stringToTerraform(struct.applyOn),
    };
}
exports.lambdaFunctionSnapStartToTerraform = lambdaFunctionSnapStartToTerraform;
function lambdaFunctionSnapStartToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        apply_on: {
            value: cdktf.stringToHclTerraform(struct.applyOn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionSnapStartToHclTerraform = lambdaFunctionSnapStartToHclTerraform;
class LambdaFunctionSnapStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._applyOn !== undefined) {
            hasAnyValues = true;
            internalValueResult.applyOn = this._applyOn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._applyOn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._applyOn = value.applyOn;
        }
    }
    get applyOn() {
        return this.getStringAttribute('apply_on');
    }
    set applyOn(value) {
        this._applyOn = value;
    }
    // Temporarily expose input value. Use with caution.
    get applyOnInput() {
        return this._applyOn;
    }
    // optimization_status - computed: true, optional: false, required: false
    get optimizationStatus() {
        return this.getStringAttribute('optimization_status');
    }
}
exports.LambdaFunctionSnapStartOutputReference = LambdaFunctionSnapStartOutputReference;
_g = JSII_RTTI_SYMBOL_1;
LambdaFunctionSnapStartOutputReference[_g] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionSnapStartOutputReference", version: "19.15.0" };
function lambdaFunctionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lambdaFunctionTimeoutsToTerraform = lambdaFunctionTimeoutsToTerraform;
function lambdaFunctionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionTimeoutsToHclTerraform = lambdaFunctionTimeoutsToHclTerraform;
class LambdaFunctionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.LambdaFunctionTimeoutsOutputReference = LambdaFunctionTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
LambdaFunctionTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionTimeoutsOutputReference", version: "19.15.0" };
function lambdaFunctionTracingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.lambdaFunctionTracingConfigToTerraform = lambdaFunctionTracingConfigToTerraform;
function lambdaFunctionTracingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionTracingConfigToHclTerraform = lambdaFunctionTracingConfigToHclTerraform;
class LambdaFunctionTracingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.LambdaFunctionTracingConfigOutputReference = LambdaFunctionTracingConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
LambdaFunctionTracingConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionTracingConfigOutputReference", version: "19.15.0" };
function lambdaFunctionVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ipv6_allowed_for_dual_stack: cdktf.booleanToTerraform(struct.ipv6AllowedForDualStack),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.lambdaFunctionVpcConfigToTerraform = lambdaFunctionVpcConfigToTerraform;
function lambdaFunctionVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ipv6_allowed_for_dual_stack: {
            value: cdktf.booleanToHclTerraform(struct.ipv6AllowedForDualStack),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaFunctionVpcConfigToHclTerraform = lambdaFunctionVpcConfigToHclTerraform;
class LambdaFunctionVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipv6AllowedForDualStack !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6AllowedForDualStack = this._ipv6AllowedForDualStack;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ipv6AllowedForDualStack = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ipv6AllowedForDualStack = value.ipv6AllowedForDualStack;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    get ipv6AllowedForDualStack() {
        return this.getBooleanAttribute('ipv6_allowed_for_dual_stack');
    }
    set ipv6AllowedForDualStack(value) {
        this._ipv6AllowedForDualStack = value;
    }
    resetIpv6AllowedForDualStack() {
        this._ipv6AllowedForDualStack = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AllowedForDualStackInput() {
        return this._ipv6AllowedForDualStack;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.LambdaFunctionVpcConfigOutputReference = LambdaFunctionVpcConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
LambdaFunctionVpcConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunctionVpcConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_function aws_lambda_function}
*/
class LambdaFunction extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LambdaFunction resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LambdaFunction to import
    * @param importFromId The id of the existing LambdaFunction that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_function#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LambdaFunction to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lambda_function", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_function aws_lambda_function} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LambdaFunctionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_function',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dead_letter_config - computed: false, optional: true, required: false
        this._deadLetterConfig = new LambdaFunctionDeadLetterConfigOutputReference(this, "dead_letter_config");
        // environment - computed: false, optional: true, required: false
        this._environment = new LambdaFunctionEnvironmentOutputReference(this, "environment");
        // ephemeral_storage - computed: false, optional: true, required: false
        this._ephemeralStorage = new LambdaFunctionEphemeralStorageOutputReference(this, "ephemeral_storage");
        // file_system_config - computed: false, optional: true, required: false
        this._fileSystemConfig = new LambdaFunctionFileSystemConfigOutputReference(this, "file_system_config");
        // image_config - computed: false, optional: true, required: false
        this._imageConfig = new LambdaFunctionImageConfigOutputReference(this, "image_config");
        // logging_config - computed: false, optional: true, required: false
        this._loggingConfig = new LambdaFunctionLoggingConfigOutputReference(this, "logging_config");
        // snap_start - computed: false, optional: true, required: false
        this._snapStart = new LambdaFunctionSnapStartOutputReference(this, "snap_start");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LambdaFunctionTimeoutsOutputReference(this, "timeouts");
        // tracing_config - computed: false, optional: true, required: false
        this._tracingConfig = new LambdaFunctionTracingConfigOutputReference(this, "tracing_config");
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new LambdaFunctionVpcConfigOutputReference(this, "vpc_config");
        this._architectures = config.architectures;
        this._codeSigningConfigArn = config.codeSigningConfigArn;
        this._description = config.description;
        this._filename = config.filename;
        this._functionName = config.functionName;
        this._handler = config.handler;
        this._id = config.id;
        this._imageUri = config.imageUri;
        this._kmsKeyArn = config.kmsKeyArn;
        this._layers = config.layers;
        this._memorySize = config.memorySize;
        this._packageType = config.packageType;
        this._publish = config.publish;
        this._replaceSecurityGroupsOnDestroy = config.replaceSecurityGroupsOnDestroy;
        this._replacementSecurityGroupIds = config.replacementSecurityGroupIds;
        this._reservedConcurrentExecutions = config.reservedConcurrentExecutions;
        this._role = config.role;
        this._runtime = config.runtime;
        this._s3Bucket = config.s3Bucket;
        this._s3Key = config.s3Key;
        this._s3ObjectVersion = config.s3ObjectVersion;
        this._skipDestroy = config.skipDestroy;
        this._sourceCodeHash = config.sourceCodeHash;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeout = config.timeout;
        this._deadLetterConfig.internalValue = config.deadLetterConfig;
        this._environment.internalValue = config.environment;
        this._ephemeralStorage.internalValue = config.ephemeralStorage;
        this._fileSystemConfig.internalValue = config.fileSystemConfig;
        this._imageConfig.internalValue = config.imageConfig;
        this._loggingConfig.internalValue = config.loggingConfig;
        this._snapStart.internalValue = config.snapStart;
        this._timeouts.internalValue = config.timeouts;
        this._tracingConfig.internalValue = config.tracingConfig;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    get architectures() {
        return this.getListAttribute('architectures');
    }
    set architectures(value) {
        this._architectures = value;
    }
    resetArchitectures() {
        this._architectures = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get architecturesInput() {
        return this._architectures;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get codeSigningConfigArn() {
        return this.getStringAttribute('code_signing_config_arn');
    }
    set codeSigningConfigArn(value) {
        this._codeSigningConfigArn = value;
    }
    resetCodeSigningConfigArn() {
        this._codeSigningConfigArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codeSigningConfigArnInput() {
        return this._codeSigningConfigArn;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get filename() {
        return this.getStringAttribute('filename');
    }
    set filename(value) {
        this._filename = value;
    }
    resetFilename() {
        this._filename = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filenameInput() {
        return this._filename;
    }
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get functionNameInput() {
        return this._functionName;
    }
    get handler() {
        return this.getStringAttribute('handler');
    }
    set handler(value) {
        this._handler = value;
    }
    resetHandler() {
        this._handler = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get handlerInput() {
        return this._handler;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get imageUri() {
        return this.getStringAttribute('image_uri');
    }
    set imageUri(value) {
        this._imageUri = value;
    }
    resetImageUri() {
        this._imageUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageUriInput() {
        return this._imageUri;
    }
    // invoke_arn - computed: true, optional: false, required: false
    get invokeArn() {
        return this.getStringAttribute('invoke_arn');
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    // last_modified - computed: true, optional: false, required: false
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    get layers() {
        return this.getListAttribute('layers');
    }
    set layers(value) {
        this._layers = value;
    }
    resetLayers() {
        this._layers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get layersInput() {
        return this._layers;
    }
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    set memorySize(value) {
        this._memorySize = value;
    }
    resetMemorySize() {
        this._memorySize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memorySizeInput() {
        return this._memorySize;
    }
    get packageType() {
        return this.getStringAttribute('package_type');
    }
    set packageType(value) {
        this._packageType = value;
    }
    resetPackageType() {
        this._packageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get packageTypeInput() {
        return this._packageType;
    }
    get publish() {
        return this.getBooleanAttribute('publish');
    }
    set publish(value) {
        this._publish = value;
    }
    resetPublish() {
        this._publish = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publishInput() {
        return this._publish;
    }
    // qualified_arn - computed: true, optional: false, required: false
    get qualifiedArn() {
        return this.getStringAttribute('qualified_arn');
    }
    // qualified_invoke_arn - computed: true, optional: false, required: false
    get qualifiedInvokeArn() {
        return this.getStringAttribute('qualified_invoke_arn');
    }
    get replaceSecurityGroupsOnDestroy() {
        return this.getBooleanAttribute('replace_security_groups_on_destroy');
    }
    set replaceSecurityGroupsOnDestroy(value) {
        this._replaceSecurityGroupsOnDestroy = value;
    }
    resetReplaceSecurityGroupsOnDestroy() {
        this._replaceSecurityGroupsOnDestroy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replaceSecurityGroupsOnDestroyInput() {
        return this._replaceSecurityGroupsOnDestroy;
    }
    get replacementSecurityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('replacement_security_group_ids'));
    }
    set replacementSecurityGroupIds(value) {
        this._replacementSecurityGroupIds = value;
    }
    resetReplacementSecurityGroupIds() {
        this._replacementSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replacementSecurityGroupIdsInput() {
        return this._replacementSecurityGroupIds;
    }
    get reservedConcurrentExecutions() {
        return this.getNumberAttribute('reserved_concurrent_executions');
    }
    set reservedConcurrentExecutions(value) {
        this._reservedConcurrentExecutions = value;
    }
    resetReservedConcurrentExecutions() {
        this._reservedConcurrentExecutions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reservedConcurrentExecutionsInput() {
        return this._reservedConcurrentExecutions;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get runtime() {
        return this.getStringAttribute('runtime');
    }
    set runtime(value) {
        this._runtime = value;
    }
    resetRuntime() {
        this._runtime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeInput() {
        return this._runtime;
    }
    get s3Bucket() {
        return this.getStringAttribute('s3_bucket');
    }
    set s3Bucket(value) {
        this._s3Bucket = value;
    }
    resetS3Bucket() {
        this._s3Bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketInput() {
        return this._s3Bucket;
    }
    get s3Key() {
        return this.getStringAttribute('s3_key');
    }
    set s3Key(value) {
        this._s3Key = value;
    }
    resetS3Key() {
        this._s3Key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3KeyInput() {
        return this._s3Key;
    }
    get s3ObjectVersion() {
        return this.getStringAttribute('s3_object_version');
    }
    set s3ObjectVersion(value) {
        this._s3ObjectVersion = value;
    }
    resetS3ObjectVersion() {
        this._s3ObjectVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3ObjectVersionInput() {
        return this._s3ObjectVersion;
    }
    // signing_job_arn - computed: true, optional: false, required: false
    get signingJobArn() {
        return this.getStringAttribute('signing_job_arn');
    }
    // signing_profile_version_arn - computed: true, optional: false, required: false
    get signingProfileVersionArn() {
        return this.getStringAttribute('signing_profile_version_arn');
    }
    get skipDestroy() {
        return this.getBooleanAttribute('skip_destroy');
    }
    set skipDestroy(value) {
        this._skipDestroy = value;
    }
    resetSkipDestroy() {
        this._skipDestroy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipDestroyInput() {
        return this._skipDestroy;
    }
    get sourceCodeHash() {
        return this.getStringAttribute('source_code_hash');
    }
    set sourceCodeHash(value) {
        this._sourceCodeHash = value;
    }
    resetSourceCodeHash() {
        this._sourceCodeHash = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceCodeHashInput() {
        return this._sourceCodeHash;
    }
    // source_code_size - computed: true, optional: false, required: false
    get sourceCodeSize() {
        return this.getNumberAttribute('source_code_size');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get deadLetterConfig() {
        return this._deadLetterConfig;
    }
    putDeadLetterConfig(value) {
        this._deadLetterConfig.internalValue = value;
    }
    resetDeadLetterConfig() {
        this._deadLetterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deadLetterConfigInput() {
        return this._deadLetterConfig.internalValue;
    }
    get environment() {
        return this._environment;
    }
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    resetEnvironment() {
        this._environment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment.internalValue;
    }
    get ephemeralStorage() {
        return this._ephemeralStorage;
    }
    putEphemeralStorage(value) {
        this._ephemeralStorage.internalValue = value;
    }
    resetEphemeralStorage() {
        this._ephemeralStorage.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ephemeralStorageInput() {
        return this._ephemeralStorage.internalValue;
    }
    get fileSystemConfig() {
        return this._fileSystemConfig;
    }
    putFileSystemConfig(value) {
        this._fileSystemConfig.internalValue = value;
    }
    resetFileSystemConfig() {
        this._fileSystemConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemConfigInput() {
        return this._fileSystemConfig.internalValue;
    }
    get imageConfig() {
        return this._imageConfig;
    }
    putImageConfig(value) {
        this._imageConfig.internalValue = value;
    }
    resetImageConfig() {
        this._imageConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigInput() {
        return this._imageConfig.internalValue;
    }
    get loggingConfig() {
        return this._loggingConfig;
    }
    putLoggingConfig(value) {
        this._loggingConfig.internalValue = value;
    }
    resetLoggingConfig() {
        this._loggingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingConfigInput() {
        return this._loggingConfig.internalValue;
    }
    get snapStart() {
        return this._snapStart;
    }
    putSnapStart(value) {
        this._snapStart.internalValue = value;
    }
    resetSnapStart() {
        this._snapStart.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapStartInput() {
        return this._snapStart.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get tracingConfig() {
        return this._tracingConfig;
    }
    putTracingConfig(value) {
        this._tracingConfig.internalValue = value;
    }
    resetTracingConfig() {
        this._tracingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tracingConfigInput() {
        return this._tracingConfig.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            architectures: cdktf.listMapper(cdktf.stringToTerraform, false)(this._architectures),
            code_signing_config_arn: cdktf.stringToTerraform(this._codeSigningConfigArn),
            description: cdktf.stringToTerraform(this._description),
            filename: cdktf.stringToTerraform(this._filename),
            function_name: cdktf.stringToTerraform(this._functionName),
            handler: cdktf.stringToTerraform(this._handler),
            id: cdktf.stringToTerraform(this._id),
            image_uri: cdktf.stringToTerraform(this._imageUri),
            kms_key_arn: cdktf.stringToTerraform(this._kmsKeyArn),
            layers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._layers),
            memory_size: cdktf.numberToTerraform(this._memorySize),
            package_type: cdktf.stringToTerraform(this._packageType),
            publish: cdktf.booleanToTerraform(this._publish),
            replace_security_groups_on_destroy: cdktf.booleanToTerraform(this._replaceSecurityGroupsOnDestroy),
            replacement_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._replacementSecurityGroupIds),
            reserved_concurrent_executions: cdktf.numberToTerraform(this._reservedConcurrentExecutions),
            role: cdktf.stringToTerraform(this._role),
            runtime: cdktf.stringToTerraform(this._runtime),
            s3_bucket: cdktf.stringToTerraform(this._s3Bucket),
            s3_key: cdktf.stringToTerraform(this._s3Key),
            s3_object_version: cdktf.stringToTerraform(this._s3ObjectVersion),
            skip_destroy: cdktf.booleanToTerraform(this._skipDestroy),
            source_code_hash: cdktf.stringToTerraform(this._sourceCodeHash),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeout: cdktf.numberToTerraform(this._timeout),
            dead_letter_config: lambdaFunctionDeadLetterConfigToTerraform(this._deadLetterConfig.internalValue),
            environment: lambdaFunctionEnvironmentToTerraform(this._environment.internalValue),
            ephemeral_storage: lambdaFunctionEphemeralStorageToTerraform(this._ephemeralStorage.internalValue),
            file_system_config: lambdaFunctionFileSystemConfigToTerraform(this._fileSystemConfig.internalValue),
            image_config: lambdaFunctionImageConfigToTerraform(this._imageConfig.internalValue),
            logging_config: lambdaFunctionLoggingConfigToTerraform(this._loggingConfig.internalValue),
            snap_start: lambdaFunctionSnapStartToTerraform(this._snapStart.internalValue),
            timeouts: lambdaFunctionTimeoutsToTerraform(this._timeouts.internalValue),
            tracing_config: lambdaFunctionTracingConfigToTerraform(this._tracingConfig.internalValue),
            vpc_config: lambdaFunctionVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            architectures: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._architectures),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            code_signing_config_arn: {
                value: cdktf.stringToHclTerraform(this._codeSigningConfigArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filename: {
                value: cdktf.stringToHclTerraform(this._filename),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            function_name: {
                value: cdktf.stringToHclTerraform(this._functionName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            handler: {
                value: cdktf.stringToHclTerraform(this._handler),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            image_uri: {
                value: cdktf.stringToHclTerraform(this._imageUri),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_arn: {
                value: cdktf.stringToHclTerraform(this._kmsKeyArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            layers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._layers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            memory_size: {
                value: cdktf.numberToHclTerraform(this._memorySize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            package_type: {
                value: cdktf.stringToHclTerraform(this._packageType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            publish: {
                value: cdktf.booleanToHclTerraform(this._publish),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            replace_security_groups_on_destroy: {
                value: cdktf.booleanToHclTerraform(this._replaceSecurityGroupsOnDestroy),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            replacement_security_group_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._replacementSecurityGroupIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            reserved_concurrent_executions: {
                value: cdktf.numberToHclTerraform(this._reservedConcurrentExecutions),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            role: {
                value: cdktf.stringToHclTerraform(this._role),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            runtime: {
                value: cdktf.stringToHclTerraform(this._runtime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_bucket: {
                value: cdktf.stringToHclTerraform(this._s3Bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_key: {
                value: cdktf.stringToHclTerraform(this._s3Key),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_object_version: {
                value: cdktf.stringToHclTerraform(this._s3ObjectVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            skip_destroy: {
                value: cdktf.booleanToHclTerraform(this._skipDestroy),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            source_code_hash: {
                value: cdktf.stringToHclTerraform(this._sourceCodeHash),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeout: {
                value: cdktf.numberToHclTerraform(this._timeout),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            dead_letter_config: {
                value: lambdaFunctionDeadLetterConfigToHclTerraform(this._deadLetterConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionDeadLetterConfigList",
            },
            environment: {
                value: lambdaFunctionEnvironmentToHclTerraform(this._environment.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionEnvironmentList",
            },
            ephemeral_storage: {
                value: lambdaFunctionEphemeralStorageToHclTerraform(this._ephemeralStorage.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionEphemeralStorageList",
            },
            file_system_config: {
                value: lambdaFunctionFileSystemConfigToHclTerraform(this._fileSystemConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionFileSystemConfigList",
            },
            image_config: {
                value: lambdaFunctionImageConfigToHclTerraform(this._imageConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionImageConfigList",
            },
            logging_config: {
                value: lambdaFunctionLoggingConfigToHclTerraform(this._loggingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionLoggingConfigList",
            },
            snap_start: {
                value: lambdaFunctionSnapStartToHclTerraform(this._snapStart.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionSnapStartList",
            },
            timeouts: {
                value: lambdaFunctionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "LambdaFunctionTimeouts",
            },
            tracing_config: {
                value: lambdaFunctionTracingConfigToHclTerraform(this._tracingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionTracingConfigList",
            },
            vpc_config: {
                value: lambdaFunctionVpcConfigToHclTerraform(this._vpcConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaFunctionVpcConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LambdaFunction = LambdaFunction;
_l = JSII_RTTI_SYMBOL_1;
LambdaFunction[_l] = { fqn: "@cdktf/provider-aws.lambdaFunction.LambdaFunction", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LambdaFunction.tfResourceType = "aws_lambda_function";
