// @cdktf/provider-aws/lib/alb-target-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlbTargetGroup = exports.AlbTargetGroupTargetHealthStateList = exports.AlbTargetGroupTargetHealthStateOutputReference = exports.albTargetGroupTargetHealthStateToHclTerraform = exports.albTargetGroupTargetHealthStateToTerraform = exports.AlbTargetGroupTargetFailoverList = exports.AlbTargetGroupTargetFailoverOutputReference = exports.albTargetGroupTargetFailoverToHclTerraform = exports.albTargetGroupTargetFailoverToTerraform = exports.AlbTargetGroupStickinessOutputReference = exports.albTargetGroupStickinessToHclTerraform = exports.albTargetGroupStickinessToTerraform = exports.AlbTargetGroupHealthCheckOutputReference = exports.albTargetGroupHealthCheckToHclTerraform = exports.albTargetGroupHealthCheckToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function albTargetGroupHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        healthy_threshold: cdktf.numberToTerraform(struct.healthyThreshold),
        interval: cdktf.numberToTerraform(struct.interval),
        matcher: cdktf.stringToTerraform(struct.matcher),
        path: cdktf.stringToTerraform(struct.path),
        port: cdktf.stringToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        timeout: cdktf.numberToTerraform(struct.timeout),
        unhealthy_threshold: cdktf.numberToTerraform(struct.unhealthyThreshold),
    };
}
exports.albTargetGroupHealthCheckToTerraform = albTargetGroupHealthCheckToTerraform;
function albTargetGroupHealthCheckToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        healthy_threshold: {
            value: cdktf.numberToHclTerraform(struct.healthyThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval: {
            value: cdktf.numberToHclTerraform(struct.interval),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        matcher: {
            value: cdktf.stringToHclTerraform(struct.matcher),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.stringToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        timeout: {
            value: cdktf.numberToHclTerraform(struct.timeout),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        unhealthy_threshold: {
            value: cdktf.numberToHclTerraform(struct.unhealthyThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.albTargetGroupHealthCheckToHclTerraform = albTargetGroupHealthCheckToHclTerraform;
class AlbTargetGroupHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._healthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthyThreshold = this._healthyThreshold;
        }
        if (this._interval !== undefined) {
            hasAnyValues = true;
            internalValueResult.interval = this._interval;
        }
        if (this._matcher !== undefined) {
            hasAnyValues = true;
            internalValueResult.matcher = this._matcher;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._unhealthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.unhealthyThreshold = this._unhealthyThreshold;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._healthyThreshold = undefined;
            this._interval = undefined;
            this._matcher = undefined;
            this._path = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._timeout = undefined;
            this._unhealthyThreshold = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._healthyThreshold = value.healthyThreshold;
            this._interval = value.interval;
            this._matcher = value.matcher;
            this._path = value.path;
            this._port = value.port;
            this._protocol = value.protocol;
            this._timeout = value.timeout;
            this._unhealthyThreshold = value.unhealthyThreshold;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    set healthyThreshold(value) {
        this._healthyThreshold = value;
    }
    resetHealthyThreshold() {
        this._healthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthyThresholdInput() {
        return this._healthyThreshold;
    }
    get interval() {
        return this.getNumberAttribute('interval');
    }
    set interval(value) {
        this._interval = value;
    }
    resetInterval() {
        this._interval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intervalInput() {
        return this._interval;
    }
    get matcher() {
        return this.getStringAttribute('matcher');
    }
    set matcher(value) {
        this._matcher = value;
    }
    resetMatcher() {
        this._matcher = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matcherInput() {
        return this._matcher;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get port() {
        return this.getStringAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
    set unhealthyThreshold(value) {
        this._unhealthyThreshold = value;
    }
    resetUnhealthyThreshold() {
        this._unhealthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unhealthyThresholdInput() {
        return this._unhealthyThreshold;
    }
}
exports.AlbTargetGroupHealthCheckOutputReference = AlbTargetGroupHealthCheckOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AlbTargetGroupHealthCheckOutputReference[_a] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupHealthCheckOutputReference", version: "19.15.0" };
function albTargetGroupStickinessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cookie_duration: cdktf.numberToTerraform(struct.cookieDuration),
        cookie_name: cdktf.stringToTerraform(struct.cookieName),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.albTargetGroupStickinessToTerraform = albTargetGroupStickinessToTerraform;
function albTargetGroupStickinessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cookie_duration: {
            value: cdktf.numberToHclTerraform(struct.cookieDuration),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        cookie_name: {
            value: cdktf.stringToHclTerraform(struct.cookieName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.albTargetGroupStickinessToHclTerraform = albTargetGroupStickinessToHclTerraform;
class AlbTargetGroupStickinessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cookieDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.cookieDuration = this._cookieDuration;
        }
        if (this._cookieName !== undefined) {
            hasAnyValues = true;
            internalValueResult.cookieName = this._cookieName;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cookieDuration = undefined;
            this._cookieName = undefined;
            this._enabled = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cookieDuration = value.cookieDuration;
            this._cookieName = value.cookieName;
            this._enabled = value.enabled;
            this._type = value.type;
        }
    }
    get cookieDuration() {
        return this.getNumberAttribute('cookie_duration');
    }
    set cookieDuration(value) {
        this._cookieDuration = value;
    }
    resetCookieDuration() {
        this._cookieDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cookieDurationInput() {
        return this._cookieDuration;
    }
    get cookieName() {
        return this.getStringAttribute('cookie_name');
    }
    set cookieName(value) {
        this._cookieName = value;
    }
    resetCookieName() {
        this._cookieName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cookieNameInput() {
        return this._cookieName;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.AlbTargetGroupStickinessOutputReference = AlbTargetGroupStickinessOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AlbTargetGroupStickinessOutputReference[_b] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupStickinessOutputReference", version: "19.15.0" };
function albTargetGroupTargetFailoverToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_deregistration: cdktf.stringToTerraform(struct.onDeregistration),
        on_unhealthy: cdktf.stringToTerraform(struct.onUnhealthy),
    };
}
exports.albTargetGroupTargetFailoverToTerraform = albTargetGroupTargetFailoverToTerraform;
function albTargetGroupTargetFailoverToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        on_deregistration: {
            value: cdktf.stringToHclTerraform(struct.onDeregistration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        on_unhealthy: {
            value: cdktf.stringToHclTerraform(struct.onUnhealthy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.albTargetGroupTargetFailoverToHclTerraform = albTargetGroupTargetFailoverToHclTerraform;
class AlbTargetGroupTargetFailoverOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onDeregistration !== undefined) {
            hasAnyValues = true;
            internalValueResult.onDeregistration = this._onDeregistration;
        }
        if (this._onUnhealthy !== undefined) {
            hasAnyValues = true;
            internalValueResult.onUnhealthy = this._onUnhealthy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._onDeregistration = undefined;
            this._onUnhealthy = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._onDeregistration = value.onDeregistration;
            this._onUnhealthy = value.onUnhealthy;
        }
    }
    get onDeregistration() {
        return this.getStringAttribute('on_deregistration');
    }
    set onDeregistration(value) {
        this._onDeregistration = value;
    }
    // Temporarily expose input value. Use with caution.
    get onDeregistrationInput() {
        return this._onDeregistration;
    }
    get onUnhealthy() {
        return this.getStringAttribute('on_unhealthy');
    }
    set onUnhealthy(value) {
        this._onUnhealthy = value;
    }
    // Temporarily expose input value. Use with caution.
    get onUnhealthyInput() {
        return this._onUnhealthy;
    }
}
exports.AlbTargetGroupTargetFailoverOutputReference = AlbTargetGroupTargetFailoverOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AlbTargetGroupTargetFailoverOutputReference[_c] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupTargetFailoverOutputReference", version: "19.15.0" };
class AlbTargetGroupTargetFailoverList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AlbTargetGroupTargetFailoverOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AlbTargetGroupTargetFailoverList = AlbTargetGroupTargetFailoverList;
_d = JSII_RTTI_SYMBOL_1;
AlbTargetGroupTargetFailoverList[_d] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupTargetFailoverList", version: "19.15.0" };
function albTargetGroupTargetHealthStateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_unhealthy_connection_termination: cdktf.booleanToTerraform(struct.enableUnhealthyConnectionTermination),
    };
}
exports.albTargetGroupTargetHealthStateToTerraform = albTargetGroupTargetHealthStateToTerraform;
function albTargetGroupTargetHealthStateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_unhealthy_connection_termination: {
            value: cdktf.booleanToHclTerraform(struct.enableUnhealthyConnectionTermination),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.albTargetGroupTargetHealthStateToHclTerraform = albTargetGroupTargetHealthStateToHclTerraform;
class AlbTargetGroupTargetHealthStateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableUnhealthyConnectionTermination !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableUnhealthyConnectionTermination = this._enableUnhealthyConnectionTermination;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._enableUnhealthyConnectionTermination = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._enableUnhealthyConnectionTermination = value.enableUnhealthyConnectionTermination;
        }
    }
    get enableUnhealthyConnectionTermination() {
        return this.getBooleanAttribute('enable_unhealthy_connection_termination');
    }
    set enableUnhealthyConnectionTermination(value) {
        this._enableUnhealthyConnectionTermination = value;
    }
    // Temporarily expose input value. Use with caution.
    get enableUnhealthyConnectionTerminationInput() {
        return this._enableUnhealthyConnectionTermination;
    }
}
exports.AlbTargetGroupTargetHealthStateOutputReference = AlbTargetGroupTargetHealthStateOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AlbTargetGroupTargetHealthStateOutputReference[_e] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupTargetHealthStateOutputReference", version: "19.15.0" };
class AlbTargetGroupTargetHealthStateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AlbTargetGroupTargetHealthStateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AlbTargetGroupTargetHealthStateList = AlbTargetGroupTargetHealthStateList;
_f = JSII_RTTI_SYMBOL_1;
AlbTargetGroupTargetHealthStateList[_f] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroupTargetHealthStateList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/alb_target_group aws_alb_target_group}
*/
class AlbTargetGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AlbTargetGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AlbTargetGroup to import
    * @param importFromId The id of the existing AlbTargetGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/alb_target_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AlbTargetGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_alb_target_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/alb_target_group aws_alb_target_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AlbTargetGroupConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_alb_target_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // health_check - computed: false, optional: true, required: false
        this._healthCheck = new AlbTargetGroupHealthCheckOutputReference(this, "health_check");
        // stickiness - computed: false, optional: true, required: false
        this._stickiness = new AlbTargetGroupStickinessOutputReference(this, "stickiness");
        // target_failover - computed: false, optional: true, required: false
        this._targetFailover = new AlbTargetGroupTargetFailoverList(this, "target_failover", false);
        // target_health_state - computed: false, optional: true, required: false
        this._targetHealthState = new AlbTargetGroupTargetHealthStateList(this, "target_health_state", false);
        this._connectionTermination = config.connectionTermination;
        this._deregistrationDelay = config.deregistrationDelay;
        this._id = config.id;
        this._ipAddressType = config.ipAddressType;
        this._lambdaMultiValueHeadersEnabled = config.lambdaMultiValueHeadersEnabled;
        this._loadBalancingAlgorithmType = config.loadBalancingAlgorithmType;
        this._loadBalancingAnomalyMitigation = config.loadBalancingAnomalyMitigation;
        this._loadBalancingCrossZoneEnabled = config.loadBalancingCrossZoneEnabled;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._port = config.port;
        this._preserveClientIp = config.preserveClientIp;
        this._protocol = config.protocol;
        this._protocolVersion = config.protocolVersion;
        this._proxyProtocolV2 = config.proxyProtocolV2;
        this._slowStart = config.slowStart;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._targetType = config.targetType;
        this._vpcId = config.vpcId;
        this._healthCheck.internalValue = config.healthCheck;
        this._stickiness.internalValue = config.stickiness;
        this._targetFailover.internalValue = config.targetFailover;
        this._targetHealthState.internalValue = config.targetHealthState;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // arn_suffix - computed: true, optional: false, required: false
    get arnSuffix() {
        return this.getStringAttribute('arn_suffix');
    }
    get connectionTermination() {
        return this.getBooleanAttribute('connection_termination');
    }
    set connectionTermination(value) {
        this._connectionTermination = value;
    }
    resetConnectionTermination() {
        this._connectionTermination = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionTerminationInput() {
        return this._connectionTermination;
    }
    get deregistrationDelay() {
        return this.getStringAttribute('deregistration_delay');
    }
    set deregistrationDelay(value) {
        this._deregistrationDelay = value;
    }
    resetDeregistrationDelay() {
        this._deregistrationDelay = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deregistrationDelayInput() {
        return this._deregistrationDelay;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    set ipAddressType(value) {
        this._ipAddressType = value;
    }
    resetIpAddressType() {
        this._ipAddressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressTypeInput() {
        return this._ipAddressType;
    }
    get lambdaMultiValueHeadersEnabled() {
        return this.getBooleanAttribute('lambda_multi_value_headers_enabled');
    }
    set lambdaMultiValueHeadersEnabled(value) {
        this._lambdaMultiValueHeadersEnabled = value;
    }
    resetLambdaMultiValueHeadersEnabled() {
        this._lambdaMultiValueHeadersEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaMultiValueHeadersEnabledInput() {
        return this._lambdaMultiValueHeadersEnabled;
    }
    // load_balancer_arns - computed: true, optional: false, required: false
    get loadBalancerArns() {
        return cdktf.Fn.tolist(this.getListAttribute('load_balancer_arns'));
    }
    get loadBalancingAlgorithmType() {
        return this.getStringAttribute('load_balancing_algorithm_type');
    }
    set loadBalancingAlgorithmType(value) {
        this._loadBalancingAlgorithmType = value;
    }
    resetLoadBalancingAlgorithmType() {
        this._loadBalancingAlgorithmType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingAlgorithmTypeInput() {
        return this._loadBalancingAlgorithmType;
    }
    get loadBalancingAnomalyMitigation() {
        return this.getStringAttribute('load_balancing_anomaly_mitigation');
    }
    set loadBalancingAnomalyMitigation(value) {
        this._loadBalancingAnomalyMitigation = value;
    }
    resetLoadBalancingAnomalyMitigation() {
        this._loadBalancingAnomalyMitigation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingAnomalyMitigationInput() {
        return this._loadBalancingAnomalyMitigation;
    }
    get loadBalancingCrossZoneEnabled() {
        return this.getStringAttribute('load_balancing_cross_zone_enabled');
    }
    set loadBalancingCrossZoneEnabled(value) {
        this._loadBalancingCrossZoneEnabled = value;
    }
    resetLoadBalancingCrossZoneEnabled() {
        this._loadBalancingCrossZoneEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingCrossZoneEnabledInput() {
        return this._loadBalancingCrossZoneEnabled;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get preserveClientIp() {
        return this.getStringAttribute('preserve_client_ip');
    }
    set preserveClientIp(value) {
        this._preserveClientIp = value;
    }
    resetPreserveClientIp() {
        this._preserveClientIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preserveClientIpInput() {
        return this._preserveClientIp;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get protocolVersion() {
        return this.getStringAttribute('protocol_version');
    }
    set protocolVersion(value) {
        this._protocolVersion = value;
    }
    resetProtocolVersion() {
        this._protocolVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolVersionInput() {
        return this._protocolVersion;
    }
    get proxyProtocolV2() {
        return this.getBooleanAttribute('proxy_protocol_v2');
    }
    set proxyProtocolV2(value) {
        this._proxyProtocolV2 = value;
    }
    resetProxyProtocolV2() {
        this._proxyProtocolV2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get proxyProtocolV2Input() {
        return this._proxyProtocolV2;
    }
    get slowStart() {
        return this.getNumberAttribute('slow_start');
    }
    set slowStart(value) {
        this._slowStart = value;
    }
    resetSlowStart() {
        this._slowStart = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get slowStartInput() {
        return this._slowStart;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get targetType() {
        return this.getStringAttribute('target_type');
    }
    set targetType(value) {
        this._targetType = value;
    }
    resetTargetType() {
        this._targetType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetTypeInput() {
        return this._targetType;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get healthCheck() {
        return this._healthCheck;
    }
    putHealthCheck(value) {
        this._healthCheck.internalValue = value;
    }
    resetHealthCheck() {
        this._healthCheck.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck.internalValue;
    }
    get stickiness() {
        return this._stickiness;
    }
    putStickiness(value) {
        this._stickiness.internalValue = value;
    }
    resetStickiness() {
        this._stickiness.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stickinessInput() {
        return this._stickiness.internalValue;
    }
    get targetFailover() {
        return this._targetFailover;
    }
    putTargetFailover(value) {
        this._targetFailover.internalValue = value;
    }
    resetTargetFailover() {
        this._targetFailover.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetFailoverInput() {
        return this._targetFailover.internalValue;
    }
    get targetHealthState() {
        return this._targetHealthState;
    }
    putTargetHealthState(value) {
        this._targetHealthState.internalValue = value;
    }
    resetTargetHealthState() {
        this._targetHealthState.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetHealthStateInput() {
        return this._targetHealthState.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connection_termination: cdktf.booleanToTerraform(this._connectionTermination),
            deregistration_delay: cdktf.stringToTerraform(this._deregistrationDelay),
            id: cdktf.stringToTerraform(this._id),
            ip_address_type: cdktf.stringToTerraform(this._ipAddressType),
            lambda_multi_value_headers_enabled: cdktf.booleanToTerraform(this._lambdaMultiValueHeadersEnabled),
            load_balancing_algorithm_type: cdktf.stringToTerraform(this._loadBalancingAlgorithmType),
            load_balancing_anomaly_mitigation: cdktf.stringToTerraform(this._loadBalancingAnomalyMitigation),
            load_balancing_cross_zone_enabled: cdktf.stringToTerraform(this._loadBalancingCrossZoneEnabled),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            port: cdktf.numberToTerraform(this._port),
            preserve_client_ip: cdktf.stringToTerraform(this._preserveClientIp),
            protocol: cdktf.stringToTerraform(this._protocol),
            protocol_version: cdktf.stringToTerraform(this._protocolVersion),
            proxy_protocol_v2: cdktf.booleanToTerraform(this._proxyProtocolV2),
            slow_start: cdktf.numberToTerraform(this._slowStart),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            target_type: cdktf.stringToTerraform(this._targetType),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            health_check: albTargetGroupHealthCheckToTerraform(this._healthCheck.internalValue),
            stickiness: albTargetGroupStickinessToTerraform(this._stickiness.internalValue),
            target_failover: cdktf.listMapper(albTargetGroupTargetFailoverToTerraform, true)(this._targetFailover.internalValue),
            target_health_state: cdktf.listMapper(albTargetGroupTargetHealthStateToTerraform, true)(this._targetHealthState.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            connection_termination: {
                value: cdktf.booleanToHclTerraform(this._connectionTermination),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            deregistration_delay: {
                value: cdktf.stringToHclTerraform(this._deregistrationDelay),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ip_address_type: {
                value: cdktf.stringToHclTerraform(this._ipAddressType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            lambda_multi_value_headers_enabled: {
                value: cdktf.booleanToHclTerraform(this._lambdaMultiValueHeadersEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            load_balancing_algorithm_type: {
                value: cdktf.stringToHclTerraform(this._loadBalancingAlgorithmType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancing_anomaly_mitigation: {
                value: cdktf.stringToHclTerraform(this._loadBalancingAnomalyMitigation),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancing_cross_zone_enabled: {
                value: cdktf.stringToHclTerraform(this._loadBalancingCrossZoneEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            port: {
                value: cdktf.numberToHclTerraform(this._port),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            preserve_client_ip: {
                value: cdktf.stringToHclTerraform(this._preserveClientIp),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            protocol: {
                value: cdktf.stringToHclTerraform(this._protocol),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            protocol_version: {
                value: cdktf.stringToHclTerraform(this._protocolVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            proxy_protocol_v2: {
                value: cdktf.booleanToHclTerraform(this._proxyProtocolV2),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            slow_start: {
                value: cdktf.numberToHclTerraform(this._slowStart),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            target_type: {
                value: cdktf.stringToHclTerraform(this._targetType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vpc_id: {
                value: cdktf.stringToHclTerraform(this._vpcId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            health_check: {
                value: albTargetGroupHealthCheckToHclTerraform(this._healthCheck.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AlbTargetGroupHealthCheckList",
            },
            stickiness: {
                value: albTargetGroupStickinessToHclTerraform(this._stickiness.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AlbTargetGroupStickinessList",
            },
            target_failover: {
                value: cdktf.listMapperHcl(albTargetGroupTargetFailoverToHclTerraform, true)(this._targetFailover.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AlbTargetGroupTargetFailoverList",
            },
            target_health_state: {
                value: cdktf.listMapperHcl(albTargetGroupTargetHealthStateToHclTerraform, true)(this._targetHealthState.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AlbTargetGroupTargetHealthStateList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AlbTargetGroup = AlbTargetGroup;
_g = JSII_RTTI_SYMBOL_1;
AlbTargetGroup[_g] = { fqn: "@cdktf/provider-aws.albTargetGroup.AlbTargetGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AlbTargetGroup.tfResourceType = "aws_alb_target_group";
