// @cdktf/provider-aws/lib/redshift-data-share-consumer-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftDataShareConsumerAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_data_share_consumer_association aws_redshift_data_share_consumer_association}
*/
class RedshiftDataShareConsumerAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftDataShareConsumerAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftDataShareConsumerAssociation to import
    * @param importFromId The id of the existing RedshiftDataShareConsumerAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_data_share_consumer_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftDataShareConsumerAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshift_data_share_consumer_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_data_share_consumer_association aws_redshift_data_share_consumer_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftDataShareConsumerAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_data_share_consumer_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._allowWrites = config.allowWrites;
        this._associateEntireAccount = config.associateEntireAccount;
        this._consumerArn = config.consumerArn;
        this._consumerRegion = config.consumerRegion;
        this._dataShareArn = config.dataShareArn;
    }
    get allowWrites() {
        return this.getBooleanAttribute('allow_writes');
    }
    set allowWrites(value) {
        this._allowWrites = value;
    }
    resetAllowWrites() {
        this._allowWrites = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowWritesInput() {
        return this._allowWrites;
    }
    get associateEntireAccount() {
        return this.getBooleanAttribute('associate_entire_account');
    }
    set associateEntireAccount(value) {
        this._associateEntireAccount = value;
    }
    resetAssociateEntireAccount() {
        this._associateEntireAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get associateEntireAccountInput() {
        return this._associateEntireAccount;
    }
    get consumerArn() {
        return this.getStringAttribute('consumer_arn');
    }
    set consumerArn(value) {
        this._consumerArn = value;
    }
    resetConsumerArn() {
        this._consumerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get consumerArnInput() {
        return this._consumerArn;
    }
    get consumerRegion() {
        return this.getStringAttribute('consumer_region');
    }
    set consumerRegion(value) {
        this._consumerRegion = value;
    }
    resetConsumerRegion() {
        this._consumerRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get consumerRegionInput() {
        return this._consumerRegion;
    }
    get dataShareArn() {
        return this.getStringAttribute('data_share_arn');
    }
    set dataShareArn(value) {
        this._dataShareArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataShareArnInput() {
        return this._dataShareArn;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // managed_by - computed: true, optional: false, required: false
    get managedBy() {
        return this.getStringAttribute('managed_by');
    }
    // producer_arn - computed: true, optional: false, required: false
    get producerArn() {
        return this.getStringAttribute('producer_arn');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allow_writes: cdktf.booleanToTerraform(this._allowWrites),
            associate_entire_account: cdktf.booleanToTerraform(this._associateEntireAccount),
            consumer_arn: cdktf.stringToTerraform(this._consumerArn),
            consumer_region: cdktf.stringToTerraform(this._consumerRegion),
            data_share_arn: cdktf.stringToTerraform(this._dataShareArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            allow_writes: {
                value: cdktf.booleanToHclTerraform(this._allowWrites),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            associate_entire_account: {
                value: cdktf.booleanToHclTerraform(this._associateEntireAccount),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            consumer_arn: {
                value: cdktf.stringToHclTerraform(this._consumerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            consumer_region: {
                value: cdktf.stringToHclTerraform(this._consumerRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_share_arn: {
                value: cdktf.stringToHclTerraform(this._dataShareArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftDataShareConsumerAssociation = RedshiftDataShareConsumerAssociation;
_a = JSII_RTTI_SYMBOL_1;
RedshiftDataShareConsumerAssociation[_a] = { fqn: "@cdktf/provider-aws.redshiftDataShareConsumerAssociation.RedshiftDataShareConsumerAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftDataShareConsumerAssociation.tfResourceType = "aws_redshift_data_share_consumer_association";
