// @cdktf/provider-aws/lib/fsx-openzfs-file-system/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxOpenzfsFileSystem = exports.FsxOpenzfsFileSystemTimeoutsOutputReference = exports.fsxOpenzfsFileSystemTimeoutsToHclTerraform = exports.fsxOpenzfsFileSystemTimeoutsToTerraform = exports.FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference = exports.fsxOpenzfsFileSystemRootVolumeConfigurationToHclTerraform = exports.fsxOpenzfsFileSystemRootVolumeConfigurationToTerraform = exports.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList = exports.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference = exports.fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToHclTerraform = exports.fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToTerraform = exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference = exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToHclTerraform = exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToTerraform = exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList = exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference = exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToHclTerraform = exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToTerraform = exports.FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference = exports.fsxOpenzfsFileSystemDiskIopsConfigurationToHclTerraform = exports.fsxOpenzfsFileSystemDiskIopsConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxOpenzfsFileSystemDiskIopsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.fsxOpenzfsFileSystemDiskIopsConfigurationToTerraform = fsxOpenzfsFileSystemDiskIopsConfigurationToTerraform;
function fsxOpenzfsFileSystemDiskIopsConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iops: {
            value: cdktf.numberToHclTerraform(struct.iops),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemDiskIopsConfigurationToHclTerraform = fsxOpenzfsFileSystemDiskIopsConfigurationToHclTerraform;
class FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iops = undefined;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iops = value.iops;
            this._mode = value.mode;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference = FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference", version: "19.15.0" };
function fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        clients: cdktf.stringToTerraform(struct.clients),
        options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.options),
    };
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToTerraform;
function fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        clients: {
            value: cdktf.stringToHclTerraform(struct.clients),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.options),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToHclTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToHclTerraform;
class FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clients !== undefined) {
            hasAnyValues = true;
            internalValueResult.clients = this._clients;
        }
        if (this._options !== undefined) {
            hasAnyValues = true;
            internalValueResult.options = this._options;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._clients = undefined;
            this._options = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._clients = value.clients;
            this._options = value.options;
        }
    }
    get clients() {
        return this.getStringAttribute('clients');
    }
    set clients(value) {
        this._clients = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientsInput() {
        return this._clients;
    }
    get options() {
        return this.getListAttribute('options');
    }
    set options(value) {
        this._options = value;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options;
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference = FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference[_b] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference", version: "19.15.0" };
class FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList = FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList;
_c = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList[_c] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList", version: "19.15.0" };
function fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_configurations: cdktf.listMapper(fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToTerraform, true)(struct.clientConfigurations),
    };
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToTerraform;
function fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        client_configurations: {
            value: cdktf.listMapperHcl(fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsToHclTerraform, true)(struct.clientConfigurations),
            isBlock: true,
            type: "set",
            storageClassType: "FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToHclTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToHclTerraform;
class FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // client_configurations - computed: false, optional: false, required: true
        this._clientConfigurations = new FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsClientConfigurationsList(this, "client_configurations", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientConfigurations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientConfigurations = this._clientConfigurations?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientConfigurations.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientConfigurations.internalValue = value.clientConfigurations;
        }
    }
    get clientConfigurations() {
        return this._clientConfigurations;
    }
    putClientConfigurations(value) {
        this._clientConfigurations.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientConfigurationsInput() {
        return this._clientConfigurations.internalValue;
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference = FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference[_d] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference", version: "19.15.0" };
function fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.numberToTerraform(struct.id),
        storage_capacity_quota_gib: cdktf.numberToTerraform(struct.storageCapacityQuotaGib),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToTerraform;
function fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.numberToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        storage_capacity_quota_gib: {
            value: cdktf.numberToHclTerraform(struct.storageCapacityQuotaGib),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToHclTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToHclTerraform;
class FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._storageCapacityQuotaGib !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageCapacityQuotaGib = this._storageCapacityQuotaGib;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._storageCapacityQuotaGib = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._storageCapacityQuotaGib = value.storageCapacityQuotaGib;
            this._type = value.type;
        }
    }
    get id() {
        return this.getNumberAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get storageCapacityQuotaGib() {
        return this.getNumberAttribute('storage_capacity_quota_gib');
    }
    set storageCapacityQuotaGib(value) {
        this._storageCapacityQuotaGib = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageCapacityQuotaGibInput() {
        return this._storageCapacityQuotaGib;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference = FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference;
_e = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference[_e] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference", version: "19.15.0" };
class FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList = FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList;
_f = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList[_f] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList", version: "19.15.0" };
function fsxOpenzfsFileSystemRootVolumeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        copy_tags_to_snapshots: cdktf.booleanToTerraform(struct.copyTagsToSnapshots),
        data_compression_type: cdktf.stringToTerraform(struct.dataCompressionType),
        read_only: cdktf.booleanToTerraform(struct.readOnly),
        record_size_kib: cdktf.numberToTerraform(struct.recordSizeKib),
        nfs_exports: fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToTerraform(struct.nfsExports),
        user_and_group_quotas: cdktf.listMapper(fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToTerraform, true)(struct.userAndGroupQuotas),
    };
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationToTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationToTerraform;
function fsxOpenzfsFileSystemRootVolumeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        copy_tags_to_snapshots: {
            value: cdktf.booleanToHclTerraform(struct.copyTagsToSnapshots),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        data_compression_type: {
            value: cdktf.stringToHclTerraform(struct.dataCompressionType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        read_only: {
            value: cdktf.booleanToHclTerraform(struct.readOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        record_size_kib: {
            value: cdktf.numberToHclTerraform(struct.recordSizeKib),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        nfs_exports: {
            value: fsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsToHclTerraform(struct.nfsExports),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsList",
        },
        user_and_group_quotas: {
            value: cdktf.listMapperHcl(fsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasToHclTerraform, true)(struct.userAndGroupQuotas),
            isBlock: true,
            type: "set",
            storageClassType: "FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemRootVolumeConfigurationToHclTerraform = fsxOpenzfsFileSystemRootVolumeConfigurationToHclTerraform;
class FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // nfs_exports - computed: false, optional: true, required: false
        this._nfsExports = new FsxOpenzfsFileSystemRootVolumeConfigurationNfsExportsOutputReference(this, "nfs_exports");
        // user_and_group_quotas - computed: false, optional: true, required: false
        this._userAndGroupQuotas = new FsxOpenzfsFileSystemRootVolumeConfigurationUserAndGroupQuotasList(this, "user_and_group_quotas", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._copyTagsToSnapshots !== undefined) {
            hasAnyValues = true;
            internalValueResult.copyTagsToSnapshots = this._copyTagsToSnapshots;
        }
        if (this._dataCompressionType !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataCompressionType = this._dataCompressionType;
        }
        if (this._readOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.readOnly = this._readOnly;
        }
        if (this._recordSizeKib !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordSizeKib = this._recordSizeKib;
        }
        if (this._nfsExports?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.nfsExports = this._nfsExports?.internalValue;
        }
        if (this._userAndGroupQuotas?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.userAndGroupQuotas = this._userAndGroupQuotas?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._copyTagsToSnapshots = undefined;
            this._dataCompressionType = undefined;
            this._readOnly = undefined;
            this._recordSizeKib = undefined;
            this._nfsExports.internalValue = undefined;
            this._userAndGroupQuotas.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._copyTagsToSnapshots = value.copyTagsToSnapshots;
            this._dataCompressionType = value.dataCompressionType;
            this._readOnly = value.readOnly;
            this._recordSizeKib = value.recordSizeKib;
            this._nfsExports.internalValue = value.nfsExports;
            this._userAndGroupQuotas.internalValue = value.userAndGroupQuotas;
        }
    }
    get copyTagsToSnapshots() {
        return this.getBooleanAttribute('copy_tags_to_snapshots');
    }
    set copyTagsToSnapshots(value) {
        this._copyTagsToSnapshots = value;
    }
    resetCopyTagsToSnapshots() {
        this._copyTagsToSnapshots = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyTagsToSnapshotsInput() {
        return this._copyTagsToSnapshots;
    }
    get dataCompressionType() {
        return this.getStringAttribute('data_compression_type');
    }
    set dataCompressionType(value) {
        this._dataCompressionType = value;
    }
    resetDataCompressionType() {
        this._dataCompressionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataCompressionTypeInput() {
        return this._dataCompressionType;
    }
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    set readOnly(value) {
        this._readOnly = value;
    }
    resetReadOnly() {
        this._readOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readOnlyInput() {
        return this._readOnly;
    }
    get recordSizeKib() {
        return this.getNumberAttribute('record_size_kib');
    }
    set recordSizeKib(value) {
        this._recordSizeKib = value;
    }
    resetRecordSizeKib() {
        this._recordSizeKib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordSizeKibInput() {
        return this._recordSizeKib;
    }
    get nfsExports() {
        return this._nfsExports;
    }
    putNfsExports(value) {
        this._nfsExports.internalValue = value;
    }
    resetNfsExports() {
        this._nfsExports.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nfsExportsInput() {
        return this._nfsExports.internalValue;
    }
    get userAndGroupQuotas() {
        return this._userAndGroupQuotas;
    }
    putUserAndGroupQuotas(value) {
        this._userAndGroupQuotas.internalValue = value;
    }
    resetUserAndGroupQuotas() {
        this._userAndGroupQuotas.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userAndGroupQuotasInput() {
        return this._userAndGroupQuotas.internalValue;
    }
}
exports.FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference = FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference", version: "19.15.0" };
function fsxOpenzfsFileSystemTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxOpenzfsFileSystemTimeoutsToTerraform = fsxOpenzfsFileSystemTimeoutsToTerraform;
function fsxOpenzfsFileSystemTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOpenzfsFileSystemTimeoutsToHclTerraform = fsxOpenzfsFileSystemTimeoutsToHclTerraform;
class FsxOpenzfsFileSystemTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FsxOpenzfsFileSystemTimeoutsOutputReference = FsxOpenzfsFileSystemTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystemTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystemTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_openzfs_file_system aws_fsx_openzfs_file_system}
*/
class FsxOpenzfsFileSystem extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FsxOpenzfsFileSystem resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FsxOpenzfsFileSystem to import
    * @param importFromId The id of the existing FsxOpenzfsFileSystem that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_openzfs_file_system#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FsxOpenzfsFileSystem to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_openzfs_file_system", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_openzfs_file_system aws_fsx_openzfs_file_system} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FsxOpenzfsFileSystemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_openzfs_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // disk_iops_configuration - computed: false, optional: true, required: false
        this._diskIopsConfiguration = new FsxOpenzfsFileSystemDiskIopsConfigurationOutputReference(this, "disk_iops_configuration");
        // root_volume_configuration - computed: false, optional: true, required: false
        this._rootVolumeConfiguration = new FsxOpenzfsFileSystemRootVolumeConfigurationOutputReference(this, "root_volume_configuration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxOpenzfsFileSystemTimeoutsOutputReference(this, "timeouts");
        this._automaticBackupRetentionDays = config.automaticBackupRetentionDays;
        this._backupId = config.backupId;
        this._copyTagsToBackups = config.copyTagsToBackups;
        this._copyTagsToVolumes = config.copyTagsToVolumes;
        this._dailyAutomaticBackupStartTime = config.dailyAutomaticBackupStartTime;
        this._deploymentType = config.deploymentType;
        this._endpointIpAddressRange = config.endpointIpAddressRange;
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._preferredSubnetId = config.preferredSubnetId;
        this._routeTableIds = config.routeTableIds;
        this._securityGroupIds = config.securityGroupIds;
        this._skipFinalBackup = config.skipFinalBackup;
        this._storageCapacity = config.storageCapacity;
        this._storageType = config.storageType;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._throughputCapacity = config.throughputCapacity;
        this._weeklyMaintenanceStartTime = config.weeklyMaintenanceStartTime;
        this._diskIopsConfiguration.internalValue = config.diskIopsConfiguration;
        this._rootVolumeConfiguration.internalValue = config.rootVolumeConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get automaticBackupRetentionDays() {
        return this.getNumberAttribute('automatic_backup_retention_days');
    }
    set automaticBackupRetentionDays(value) {
        this._automaticBackupRetentionDays = value;
    }
    resetAutomaticBackupRetentionDays() {
        this._automaticBackupRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get automaticBackupRetentionDaysInput() {
        return this._automaticBackupRetentionDays;
    }
    get backupId() {
        return this.getStringAttribute('backup_id');
    }
    set backupId(value) {
        this._backupId = value;
    }
    resetBackupId() {
        this._backupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupIdInput() {
        return this._backupId;
    }
    get copyTagsToBackups() {
        return this.getBooleanAttribute('copy_tags_to_backups');
    }
    set copyTagsToBackups(value) {
        this._copyTagsToBackups = value;
    }
    resetCopyTagsToBackups() {
        this._copyTagsToBackups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyTagsToBackupsInput() {
        return this._copyTagsToBackups;
    }
    get copyTagsToVolumes() {
        return this.getBooleanAttribute('copy_tags_to_volumes');
    }
    set copyTagsToVolumes(value) {
        this._copyTagsToVolumes = value;
    }
    resetCopyTagsToVolumes() {
        this._copyTagsToVolumes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyTagsToVolumesInput() {
        return this._copyTagsToVolumes;
    }
    get dailyAutomaticBackupStartTime() {
        return this.getStringAttribute('daily_automatic_backup_start_time');
    }
    set dailyAutomaticBackupStartTime(value) {
        this._dailyAutomaticBackupStartTime = value;
    }
    resetDailyAutomaticBackupStartTime() {
        this._dailyAutomaticBackupStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dailyAutomaticBackupStartTimeInput() {
        return this._dailyAutomaticBackupStartTime;
    }
    get deploymentType() {
        return this.getStringAttribute('deployment_type');
    }
    set deploymentType(value) {
        this._deploymentType = value;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentTypeInput() {
        return this._deploymentType;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // endpoint_ip_address - computed: true, optional: false, required: false
    get endpointIpAddress() {
        return this.getStringAttribute('endpoint_ip_address');
    }
    get endpointIpAddressRange() {
        return this.getStringAttribute('endpoint_ip_address_range');
    }
    set endpointIpAddressRange(value) {
        this._endpointIpAddressRange = value;
    }
    resetEndpointIpAddressRange() {
        this._endpointIpAddressRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointIpAddressRangeInput() {
        return this._endpointIpAddressRange;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    // network_interface_ids - computed: true, optional: false, required: false
    get networkInterfaceIds() {
        return this.getListAttribute('network_interface_ids');
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get preferredSubnetId() {
        return this.getStringAttribute('preferred_subnet_id');
    }
    set preferredSubnetId(value) {
        this._preferredSubnetId = value;
    }
    resetPreferredSubnetId() {
        this._preferredSubnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preferredSubnetIdInput() {
        return this._preferredSubnetId;
    }
    // root_volume_id - computed: true, optional: false, required: false
    get rootVolumeId() {
        return this.getStringAttribute('root_volume_id');
    }
    get routeTableIds() {
        return cdktf.Fn.tolist(this.getListAttribute('route_table_ids'));
    }
    set routeTableIds(value) {
        this._routeTableIds = value;
    }
    resetRouteTableIds() {
        this._routeTableIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routeTableIdsInput() {
        return this._routeTableIds;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get skipFinalBackup() {
        return this.getBooleanAttribute('skip_final_backup');
    }
    set skipFinalBackup(value) {
        this._skipFinalBackup = value;
    }
    resetSkipFinalBackup() {
        this._skipFinalBackup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipFinalBackupInput() {
        return this._skipFinalBackup;
    }
    get storageCapacity() {
        return this.getNumberAttribute('storage_capacity');
    }
    set storageCapacity(value) {
        this._storageCapacity = value;
    }
    resetStorageCapacity() {
        this._storageCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageCapacityInput() {
        return this._storageCapacity;
    }
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    set storageType(value) {
        this._storageType = value;
    }
    resetStorageType() {
        this._storageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageTypeInput() {
        return this._storageType;
    }
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get throughputCapacity() {
        return this.getNumberAttribute('throughput_capacity');
    }
    set throughputCapacity(value) {
        this._throughputCapacity = value;
    }
    // Temporarily expose input value. Use with caution.
    get throughputCapacityInput() {
        return this._throughputCapacity;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    get weeklyMaintenanceStartTime() {
        return this.getStringAttribute('weekly_maintenance_start_time');
    }
    set weeklyMaintenanceStartTime(value) {
        this._weeklyMaintenanceStartTime = value;
    }
    resetWeeklyMaintenanceStartTime() {
        this._weeklyMaintenanceStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weeklyMaintenanceStartTimeInput() {
        return this._weeklyMaintenanceStartTime;
    }
    get diskIopsConfiguration() {
        return this._diskIopsConfiguration;
    }
    putDiskIopsConfiguration(value) {
        this._diskIopsConfiguration.internalValue = value;
    }
    resetDiskIopsConfiguration() {
        this._diskIopsConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskIopsConfigurationInput() {
        return this._diskIopsConfiguration.internalValue;
    }
    get rootVolumeConfiguration() {
        return this._rootVolumeConfiguration;
    }
    putRootVolumeConfiguration(value) {
        this._rootVolumeConfiguration.internalValue = value;
    }
    resetRootVolumeConfiguration() {
        this._rootVolumeConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rootVolumeConfigurationInput() {
        return this._rootVolumeConfiguration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            automatic_backup_retention_days: cdktf.numberToTerraform(this._automaticBackupRetentionDays),
            backup_id: cdktf.stringToTerraform(this._backupId),
            copy_tags_to_backups: cdktf.booleanToTerraform(this._copyTagsToBackups),
            copy_tags_to_volumes: cdktf.booleanToTerraform(this._copyTagsToVolumes),
            daily_automatic_backup_start_time: cdktf.stringToTerraform(this._dailyAutomaticBackupStartTime),
            deployment_type: cdktf.stringToTerraform(this._deploymentType),
            endpoint_ip_address_range: cdktf.stringToTerraform(this._endpointIpAddressRange),
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            preferred_subnet_id: cdktf.stringToTerraform(this._preferredSubnetId),
            route_table_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._routeTableIds),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            skip_final_backup: cdktf.booleanToTerraform(this._skipFinalBackup),
            storage_capacity: cdktf.numberToTerraform(this._storageCapacity),
            storage_type: cdktf.stringToTerraform(this._storageType),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            throughput_capacity: cdktf.numberToTerraform(this._throughputCapacity),
            weekly_maintenance_start_time: cdktf.stringToTerraform(this._weeklyMaintenanceStartTime),
            disk_iops_configuration: fsxOpenzfsFileSystemDiskIopsConfigurationToTerraform(this._diskIopsConfiguration.internalValue),
            root_volume_configuration: fsxOpenzfsFileSystemRootVolumeConfigurationToTerraform(this._rootVolumeConfiguration.internalValue),
            timeouts: fsxOpenzfsFileSystemTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            automatic_backup_retention_days: {
                value: cdktf.numberToHclTerraform(this._automaticBackupRetentionDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            backup_id: {
                value: cdktf.stringToHclTerraform(this._backupId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            copy_tags_to_backups: {
                value: cdktf.booleanToHclTerraform(this._copyTagsToBackups),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            copy_tags_to_volumes: {
                value: cdktf.booleanToHclTerraform(this._copyTagsToVolumes),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            daily_automatic_backup_start_time: {
                value: cdktf.stringToHclTerraform(this._dailyAutomaticBackupStartTime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deployment_type: {
                value: cdktf.stringToHclTerraform(this._deploymentType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_ip_address_range: {
                value: cdktf.stringToHclTerraform(this._endpointIpAddressRange),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_id: {
                value: cdktf.stringToHclTerraform(this._kmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preferred_subnet_id: {
                value: cdktf.stringToHclTerraform(this._preferredSubnetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            route_table_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._routeTableIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            security_group_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._securityGroupIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            skip_final_backup: {
                value: cdktf.booleanToHclTerraform(this._skipFinalBackup),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            storage_capacity: {
                value: cdktf.numberToHclTerraform(this._storageCapacity),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            storage_type: {
                value: cdktf.stringToHclTerraform(this._storageType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._subnetIds),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            throughput_capacity: {
                value: cdktf.numberToHclTerraform(this._throughputCapacity),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            weekly_maintenance_start_time: {
                value: cdktf.stringToHclTerraform(this._weeklyMaintenanceStartTime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disk_iops_configuration: {
                value: fsxOpenzfsFileSystemDiskIopsConfigurationToHclTerraform(this._diskIopsConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FsxOpenzfsFileSystemDiskIopsConfigurationList",
            },
            root_volume_configuration: {
                value: fsxOpenzfsFileSystemRootVolumeConfigurationToHclTerraform(this._rootVolumeConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FsxOpenzfsFileSystemRootVolumeConfigurationList",
            },
            timeouts: {
                value: fsxOpenzfsFileSystemTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FsxOpenzfsFileSystemTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FsxOpenzfsFileSystem = FsxOpenzfsFileSystem;
_j = JSII_RTTI_SYMBOL_1;
FsxOpenzfsFileSystem[_j] = { fqn: "@cdktf/provider-aws.fsxOpenzfsFileSystem.FsxOpenzfsFileSystem", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FsxOpenzfsFileSystem.tfResourceType = "aws_fsx_openzfs_file_system";
