// @cdktf/provider-aws/lib/ssm-default-patch-baseline/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmDefaultPatchBaseline = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_default_patch_baseline aws_ssm_default_patch_baseline}
*/
class SsmDefaultPatchBaseline extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsmDefaultPatchBaseline resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsmDefaultPatchBaseline to import
    * @param importFromId The id of the existing SsmDefaultPatchBaseline that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_default_patch_baseline#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsmDefaultPatchBaseline to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssm_default_patch_baseline", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_default_patch_baseline aws_ssm_default_patch_baseline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsmDefaultPatchBaselineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_default_patch_baseline',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._baselineId = config.baselineId;
        this._id = config.id;
        this._operatingSystem = config.operatingSystem;
    }
    get baselineId() {
        return this.getStringAttribute('baseline_id');
    }
    set baselineId(value) {
        this._baselineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get baselineIdInput() {
        return this._baselineId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get operatingSystem() {
        return this.getStringAttribute('operating_system');
    }
    set operatingSystem(value) {
        this._operatingSystem = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatingSystemInput() {
        return this._operatingSystem;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            baseline_id: cdktf.stringToTerraform(this._baselineId),
            id: cdktf.stringToTerraform(this._id),
            operating_system: cdktf.stringToTerraform(this._operatingSystem),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            baseline_id: {
                value: cdktf.stringToHclTerraform(this._baselineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            operating_system: {
                value: cdktf.stringToHclTerraform(this._operatingSystem),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsmDefaultPatchBaseline = SsmDefaultPatchBaseline;
_a = JSII_RTTI_SYMBOL_1;
SsmDefaultPatchBaseline[_a] = { fqn: "@cdktf/provider-aws.ssmDefaultPatchBaseline.SsmDefaultPatchBaseline", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsmDefaultPatchBaseline.tfResourceType = "aws_ssm_default_patch_baseline";
