// @cdktf/provider-aws/lib/vpc-ipam-organization-admin-account/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcIpamOrganizationAdminAccount = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_organization_admin_account aws_vpc_ipam_organization_admin_account}
*/
class VpcIpamOrganizationAdminAccount extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VpcIpamOrganizationAdminAccount resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VpcIpamOrganizationAdminAccount to import
    * @param importFromId The id of the existing VpcIpamOrganizationAdminAccount that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_organization_admin_account#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VpcIpamOrganizationAdminAccount to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpc_ipam_organization_admin_account", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_organization_admin_account aws_vpc_ipam_organization_admin_account} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcIpamOrganizationAdminAccountConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam_organization_admin_account',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._delegatedAdminAccountId = config.delegatedAdminAccountId;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get delegatedAdminAccountId() {
        return this.getStringAttribute('delegated_admin_account_id');
    }
    set delegatedAdminAccountId(value) {
        this._delegatedAdminAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get delegatedAdminAccountIdInput() {
        return this._delegatedAdminAccountId;
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // service_principal - computed: true, optional: false, required: false
    get servicePrincipal() {
        return this.getStringAttribute('service_principal');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            delegated_admin_account_id: cdktf.stringToTerraform(this._delegatedAdminAccountId),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            delegated_admin_account_id: {
                value: cdktf.stringToHclTerraform(this._delegatedAdminAccountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VpcIpamOrganizationAdminAccount = VpcIpamOrganizationAdminAccount;
_a = JSII_RTTI_SYMBOL_1;
VpcIpamOrganizationAdminAccount[_a] = { fqn: "@cdktf/provider-aws.vpcIpamOrganizationAdminAccount.VpcIpamOrganizationAdminAccount", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VpcIpamOrganizationAdminAccount.tfResourceType = "aws_vpc_ipam_organization_admin_account";
