// @cdktf/provider-aws/lib/securityhub-account/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubAccount = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_account aws_securityhub_account}
*/
class SecurityhubAccount extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecurityhubAccount resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecurityhubAccount to import
    * @param importFromId The id of the existing SecurityhubAccount that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_account#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecurityhubAccount to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securityhub_account", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_account aws_securityhub_account} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubAccountConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_account',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._autoEnableControls = config.autoEnableControls;
        this._controlFindingGenerator = config.controlFindingGenerator;
        this._enableDefaultStandards = config.enableDefaultStandards;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoEnableControls() {
        return this.getBooleanAttribute('auto_enable_controls');
    }
    set autoEnableControls(value) {
        this._autoEnableControls = value;
    }
    resetAutoEnableControls() {
        this._autoEnableControls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableControlsInput() {
        return this._autoEnableControls;
    }
    get controlFindingGenerator() {
        return this.getStringAttribute('control_finding_generator');
    }
    set controlFindingGenerator(value) {
        this._controlFindingGenerator = value;
    }
    resetControlFindingGenerator() {
        this._controlFindingGenerator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlFindingGeneratorInput() {
        return this._controlFindingGenerator;
    }
    get enableDefaultStandards() {
        return this.getBooleanAttribute('enable_default_standards');
    }
    set enableDefaultStandards(value) {
        this._enableDefaultStandards = value;
    }
    resetEnableDefaultStandards() {
        this._enableDefaultStandards = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableDefaultStandardsInput() {
        return this._enableDefaultStandards;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_enable_controls: cdktf.booleanToTerraform(this._autoEnableControls),
            control_finding_generator: cdktf.stringToTerraform(this._controlFindingGenerator),
            enable_default_standards: cdktf.booleanToTerraform(this._enableDefaultStandards),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_enable_controls: {
                value: cdktf.booleanToHclTerraform(this._autoEnableControls),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            control_finding_generator: {
                value: cdktf.stringToHclTerraform(this._controlFindingGenerator),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enable_default_standards: {
                value: cdktf.booleanToHclTerraform(this._enableDefaultStandards),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecurityhubAccount = SecurityhubAccount;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubAccount[_a] = { fqn: "@cdktf/provider-aws.securityhubAccount.SecurityhubAccount", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubAccount.tfResourceType = "aws_securityhub_account";
