// @winglang/sdk/lib/target-tf-azure/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
const bucket_1 = require("./bucket");
const counter_1 = require("./counter");
const function_1 = require("./function");
const internal_1 = require("./internal");
const test_runner_1 = require("./test-runner");
const application_insights_1 = require("../.gen/providers/azurerm/application-insights");
const log_analytics_workspace_1 = require("../.gen/providers/azurerm/log-analytics-workspace");
const provider_1 = require("../.gen/providers/azurerm/provider");
const resource_group_1 = require("../.gen/providers/azurerm/resource-group");
const service_plan_1 = require("../.gen/providers/azurerm/service-plan");
const storage_account_1 = require("../.gen/providers/azurerm/storage-account");
const cloud_1 = require("../cloud");
const resource_names_1 = require("../shared/resource-names");
const app_1 = require("../shared-tf/app");
const std_1 = require("../std");
/**
 * ResourceGroup names are limited to 90 characters.
 * You can use alphanumeric characters, hyphens, and underscores,
 * parentheses and periods.
 */
const RESOURCEGROUP_NAME_OPTS = {
    maxLen: 90,
    disallowedRegex: /([^a-zA-Z0-9\-\_\(\)\.]+)/g,
};
/**
 * Configuration options for generating a name for Azure Log Analytics Workspace.
 *
 * - The workspace name must be between 4 and 63 characters.
 * - The workspace name can contain only letters, numbers, and hyphens (`"-"`).
 * - The hyphen (`"-"`) should not be the first or the last character in the name.
 */
const LOG_ANALYTICS_WORKSPACE_NAME_OPTS = {
    maxLen: 63,
    disallowedRegex: /([^a-zA-Z0-9\-]+)/g,
};
/**
 * StorageAccount names are limited to 24 characters.
 * You can only use alphanumeric characters.
 */
const STORAGEACCOUNT_NAME_OPTS = {
    maxLen: 24,
    case: resource_names_1.CaseConventions.LOWERCASE,
    disallowedRegex: /([^a-z0-9]+)/g,
    sep: "",
};
const SERVICEPLAN_NAME_OPTS = {
    maxLen: 50,
    disallowedRegex: /([^a-zA-Z0-9\-]+)/g,
};
/**
 * An app that knows how to synthesize constructs into a Terraform configuration
 * for Azure resources.
 */
class App extends app_1.CdktfApp {
    constructor(props) {
        super(props);
        this._target = "tf-azure";
        this.location = props.location ?? process.env.AZURE_LOCATION;
        test_runner_1.TestRunner._createTree(this, props.rootConstruct);
        // Using env variable for location is work around until we are
        // able to implement https://github.com/winglang/wing/issues/493 (policy as infrastructure)
        if (this.location === undefined) {
            throw new Error("Location must be specified in the AZURE_LOCATION environment variable");
        }
        new provider_1.AzurermProvider(this, "azure", {
            features: {
                // To be able to run terraform destroy during tests, and in a reasonable time
                resourceGroup: { preventDeletionIfContainsResources: false },
            },
        });
        Object.defineProperty(this, internal_1.APP_AZURE_TF_SYMBOL, {
            value: this,
            enumerable: false,
            writable: false,
        });
    }
    get logAnalyticsWorkspace() {
        if (!this._logAnalyticsWorkspace) {
            this._logAnalyticsWorkspace = new log_analytics_workspace_1.LogAnalyticsWorkspace(this, "LogAnalyticsWorkspace", {
                location: this.location,
                resourceGroupName: this.resourceGroup.name,
                name: resource_names_1.ResourceNames.generateName(this, LOG_ANALYTICS_WORKSPACE_NAME_OPTS),
            });
        }
        return this._logAnalyticsWorkspace;
    }
    get applicationInsights() {
        if (!this._applicationInsights) {
            this._applicationInsights = new application_insights_1.ApplicationInsights(this, `ApplicationInsights`, {
                name: `application-insights`,
                resourceGroupName: this.resourceGroup.name,
                location: this.resourceGroup.location,
                applicationType: "web",
                workspaceId: this.logAnalyticsWorkspace.id,
            });
        }
        return this._applicationInsights;
    }
    /**
     * Get resource group using lazy initialization
     */
    get resourceGroup() {
        if (!this._resourceGroup) {
            this._resourceGroup = new resource_group_1.ResourceGroup(this, "ResourceGroup", {
                location: this.location,
                name: resource_names_1.ResourceNames.generateName(this, RESOURCEGROUP_NAME_OPTS),
            });
        }
        return this._resourceGroup;
    }
    /**
     * Get storage account using lazy initialization
     */
    get storageAccount() {
        if (!this._storageAccount) {
            this._storageAccount = new storage_account_1.StorageAccount(this, "StorageAccount", {
                name: resource_names_1.ResourceNames.generateName(this, STORAGEACCOUNT_NAME_OPTS),
                resourceGroupName: this.resourceGroup.name,
                location: this.resourceGroup.location,
                accountTier: "Standard",
                accountReplicationType: "LRS",
            });
        }
        return this._storageAccount;
    }
    /**
     * Get service plan using lazy initialization
     */
    get servicePlan() {
        if (!this._servicePlan) {
            this._servicePlan = new service_plan_1.ServicePlan(this, "ServicePlan", {
                name: resource_names_1.ResourceNames.generateName(this, SERVICEPLAN_NAME_OPTS),
                resourceGroupName: this.resourceGroup.name,
                location: this.resourceGroup.location,
                osType: "Linux",
                // Dynamic Stock Keeping Unit (SKU)
                // https://learn.microsoft.com/en-us/partner-center/developer/product-resources#sku
                skuName: "Y1",
            });
        }
        return this._servicePlan;
    }
    typeForFqn(fqn) {
        switch (fqn) {
            case std_1.TEST_RUNNER_FQN:
                return test_runner_1.TestRunner;
            case cloud_1.FUNCTION_FQN:
                return function_1.Function;
            case cloud_1.BUCKET_FQN:
                return bucket_1.Bucket;
            case cloud_1.COUNTER_FQN:
                return counter_1.Counter;
        }
        return undefined;
    }
}
exports.App = App;
