// @winglang/sdk/lib/target-tf-azure/function.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Function = void 0;
const fs = __importStar(require("fs"));
const cdktf_1 = require("cdktf");
const app_1 = require("./app");
const bucket_1 = require("./bucket");
const __1 = require("..");
const linux_function_app_1 = require("../.gen/providers/azurerm/linux-function-app");
const role_assignment_1 = require("../.gen/providers/azurerm/role-assignment");
const storage_blob_1 = require("../.gen/providers/azurerm/storage-blob");
const cloud = __importStar(require("../cloud"));
const errors_1 = require("../core/errors");
const bundling_1 = require("../shared/bundling");
const resource_names_1 = require("../shared/resource-names");
/**
 * Function names are limited to 32 characters.
 * You can use alphanumeric characters.
 * https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftweb
 */
const FUNCTION_NAME_OPTS = {
    maxLen: 32,
    disallowedRegex: /[^a-z0-9]+/g,
    case: resource_names_1.CaseConventions.LOWERCASE,
};
/**
 * Azure implementation of `cloud.Function`.
 *
 * @inflight `@winglang/sdk.cloud.IFunctionClient`
 */
class Function extends cloud.Function {
    constructor(scope, id, inflight, props = {}) {
        super(scope, id, inflight, props);
        this.permissions = new Map();
        const app = app_1.App.of(this);
        this.storageAccount = app.storageAccount;
        this.resourceGroup = app.resourceGroup;
        this.servicePlan = app.servicePlan;
        this.applicationInsights = app.applicationInsights;
        this.functionName = resource_names_1.ResourceNames.generateName(this, FUNCTION_NAME_OPTS);
        const functionIdentityType = "SystemAssigned";
        const functionRuntime = "node";
        const functionNodeVersion = "20"; // support fetch
        // Create Bucket to store function code
        const functionCodeBucket = new bucket_1.Bucket(this, "FunctionBucket");
        if (props.concurrency != null) {
            throw new errors_1.NotImplementedError("Function concurrency isn't implemented yet on the current target.");
        }
        // throw an error if props.memory is defined for an Azure function
        if (props.memory) {
            throw new errors_1.NotImplementedError("memory is an invalid parameter on Azure", {
                resource: this.constructor.name,
                operation: "memory",
            });
        }
        // TODO: will be uncommented when fixing https://github.com/winglang/wing/issues/4494
        // const timeout = props.timeout ?? Duration.fromMinutes(1);
        if (props.timeout) {
            throw new errors_1.NotImplementedError("Function.timeout is not implemented yet on tf-azure target.", {
                issue: "https://github.com/winglang/wing/issues/4494",
                resource: this.constructor.name,
                operation: "timeout",
            });
        }
        // Upload zip asset to storage account
        const functionCodeBlob = new storage_blob_1.StorageBlob(this, "CodeBlob", {
            name: `${this.functionName}.zip`,
            storageAccountName: this.storageAccount.name,
            storageContainerName: functionCodeBucket.storageContainer.name,
            type: "Block",
            source: cdktf_1.Lazy.stringValue({
                produce: () => {
                    if (!this.assetPath) {
                        throw new Error("assetPath was not set");
                    }
                    return this.assetPath;
                },
            }),
        });
        // Create the function
        this.function = new linux_function_app_1.LinuxFunctionApp(this, "Function", {
            name: this.functionName,
            resourceGroupName: this.resourceGroup.name,
            location: this.resourceGroup.location,
            servicePlanId: this.servicePlan.id,
            storageAccountName: this.storageAccount.name,
            identity: {
                type: functionIdentityType,
            },
            storageAccountAccessKey: this.storageAccount.primaryAccessKey,
            siteConfig: {
                applicationStack: {
                    nodeVersion: functionNodeVersion,
                },
                applicationInsightsConnectionString: this.applicationInsights.connectionString,
                applicationInsightsKey: this.applicationInsights.instrumentationKey,
            },
            httpsOnly: true,
            appSettings: cdktf_1.Lazy.anyValue({
                produce: () => ({
                    ...this.env,
                    WEBSITE_RUN_FROM_PACKAGE: `https://${this.storageAccount.name}.blob.core.windows.net/${functionCodeBucket.storageContainer.name}/${functionCodeBlob.name}`,
                    FUNCTIONS_WORKER_RUNTIME: functionRuntime,
                }),
            }),
        });
        // Apply permissions from bound resources
        for (const key of this.permissions.keys() || []) {
            const scopedRoleAssignment = this.permissions?.get(key);
            new role_assignment_1.RoleAssignment(this, `RoleAssignment${key}`, {
                scope: scopedRoleAssignment.scope,
                roleDefinitionName: scopedRoleAssignment.roleDefinitionName,
                principalId: this.function.identity.principalId,
            });
        }
        const roleAssignment = {
            principalId: this.function.identity.principalId,
            roleDefinitionName: bucket_1.StorageAccountPermissions.READ,
            scope: this.storageAccount.id,
        };
        // Add permissions to read function code
        new role_assignment_1.RoleAssignment(this, `ReadLambdaCodeAssignment`, roleAssignment);
        this.permissions.set(`${this.storageAccount.node.addr.substring(-8)}_${bucket_1.StorageAccountPermissions.READ}`, roleAssignment);
    }
    /** @internal */
    _preSynthesize() {
        super._preSynthesize();
        const bundle = (0, bundling_1.createBundle)(this.entrypoint);
        const codeDir = bundle.directory;
        // Package up code in azure expected format
        const outDir = `${codeDir}/${this.functionName}`;
        // Move index.js to function name directory. Every Azure function in a function app
        // must be in its own folder containing an index.js and function.json files
        fs.mkdirSync(`${codeDir}/${this.functionName}`);
        fs.renameSync(bundle.outfilePath, `${outDir}/index.js`);
        // As per documentation "a function must have exactly one trigger" so for now
        // by default a function will support http get requests
        // when we lift other resources like queues or topics this function.json will need to
        // be overwritten with the correct trigger
        // https://learn.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings?tabs=csharp
        fs.writeFileSync(`${outDir}/function.json`, JSON.stringify({
            bindings: [
                {
                    authLevel: "anonymous", // TODO: this auth level will be changed with https://github.com/winglang/wing/issues/4497
                    type: "httpTrigger",
                    direction: "in",
                    name: "req",
                    methods: ["get", "post"],
                },
                {
                    type: "http",
                    direction: "out",
                    name: "res",
                },
            ],
        }));
        // Write host.json file to set function timeout (must be set in root of function app)
        // https://learn.microsoft.com/en-us/azure/azure-functions/functions-host-json
        // this means that timeout is set for all functions in the function app
        fs.writeFileSync(`${codeDir}/host.json`, JSON.stringify({
            version: "2.0",
            //TODO: need to read the documentation and parse the number in the right way,
            // while not surpassing the limits, since it will be resulted in a hard to detect error
            functionTimeout: `00:01:00`,
        }));
        // Create zip asset from function code
        const asset = new cdktf_1.TerraformAsset(this, "Asset", {
            path: `${codeDir}`,
            type: cdktf_1.AssetType.ARCHIVE,
        });
        this.assetPath = asset.path;
    }
    /**
     * Function name, used for invocation
     */
    get name() {
        return this.function.name;
    }
    /**
     *  Adds role to function for given azure scope
     *
     * @param scopedResource - The resource to which the role assignment will be scoped.
     * @param scopedRoleAssignment - The mapping of azure scope to role definition name.
     */
    addPermission(scopedResource, scopedRoleAssignment) {
        const uniqueId = scopedResource.node.addr.substring(-8);
        const permissionsKey = `${uniqueId}_${scopedRoleAssignment.roleDefinitionName}`;
        // If the function has already been initialized attach the role assignment directly
        if (this.function) {
            if (this.permissions.has(permissionsKey)) {
                return; // already exists
            }
            new role_assignment_1.RoleAssignment(this, `RoleAssignment${permissionsKey}`, {
                scope: scopedRoleAssignment.scope,
                roleDefinitionName: scopedRoleAssignment.roleDefinitionName,
                principalId: this.function.identity.principalId,
            });
        }
        this.permissions.set(permissionsKey, scopedRoleAssignment);
    }
    _getCodeLines(handler) {
        const inflightClient = handler._toInflight();
        const lines = new Array();
        lines.push('"use strict";');
        lines.push("module.exports = async function(context, req) {");
        lines.push(`  const body = await (${inflightClient}).handle(context.req.body ?? "");`);
        lines.push(`  context.res = { body };`);
        lines.push(`};`);
        return lines;
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.FunctionInflightMethods.INVOKE]: [[this.handler, ["handle"]]],
        };
    }
    onLift(host, ops) {
        //TODO: add permissions here when changing auth level: https://github.com/winglang/wing/issues/4497
        host.addEnvironment(this.envName(), this.function.name);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return __1.core.InflightClient.for(__dirname.replace("target-tf-azure", "shared-azure"), __filename, "FunctionClient", [`process.env["${this.envName()}"]`]);
    }
    envName() {
        return `FUNCTION_NAME_${this.node.addr.slice(-8)}`;
    }
}
exports.Function = Function;
