// @cdktf/provider-aws/lib/redshift-authentication-profile/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftAuthenticationProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_authentication_profile aws_redshift_authentication_profile}
*/
class RedshiftAuthenticationProfile extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftAuthenticationProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftAuthenticationProfile to import
    * @param importFromId The id of the existing RedshiftAuthenticationProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_authentication_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftAuthenticationProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshift_authentication_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_authentication_profile aws_redshift_authentication_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftAuthenticationProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_authentication_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._authenticationProfileContent = config.authenticationProfileContent;
        this._authenticationProfileName = config.authenticationProfileName;
        this._id = config.id;
    }
    get authenticationProfileContent() {
        return this.getStringAttribute('authentication_profile_content');
    }
    set authenticationProfileContent(value) {
        this._authenticationProfileContent = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationProfileContentInput() {
        return this._authenticationProfileContent;
    }
    get authenticationProfileName() {
        return this.getStringAttribute('authentication_profile_name');
    }
    set authenticationProfileName(value) {
        this._authenticationProfileName = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationProfileNameInput() {
        return this._authenticationProfileName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authentication_profile_content: cdktf.stringToTerraform(this._authenticationProfileContent),
            authentication_profile_name: cdktf.stringToTerraform(this._authenticationProfileName),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            authentication_profile_content: {
                value: cdktf.stringToHclTerraform(this._authenticationProfileContent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authentication_profile_name: {
                value: cdktf.stringToHclTerraform(this._authenticationProfileName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftAuthenticationProfile = RedshiftAuthenticationProfile;
_a = JSII_RTTI_SYMBOL_1;
RedshiftAuthenticationProfile[_a] = { fqn: "@cdktf/provider-aws.redshiftAuthenticationProfile.RedshiftAuthenticationProfile", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftAuthenticationProfile.tfResourceType = "aws_redshift_authentication_profile";
