// @winglang/sdk/lib/target-tf-gcp/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
const bucket_1 = require("./bucket");
const counter_1 = require("./counter");
const function_1 = require("./function");
const schedule_1 = require("./schedule");
const table_1 = require("./table");
const test_runner_1 = require("./test-runner");
const provider_1 = require("../.gen/providers/google/provider");
const provider_2 = require("../.gen/providers/random/provider");
const cloud_1 = require("../cloud");
const ex_1 = require("../ex");
const app_1 = require("../shared-tf/app");
const std_1 = require("../std");
/**
 * An app that knows how to synthesize constructs into a Terraform configuration
 * for GCP resources.
 */
class App extends app_1.CdktfApp {
    constructor(props) {
        super(props);
        this._target = "tf-gcp";
        this.projectId = props.projectId ?? process.env.GOOGLE_PROJECT_ID;
        if (this.projectId === undefined) {
            throw new Error("A Google Cloud project ID must be specified through the GOOGLE_PROJECT_ID environment variable.");
        }
        this.region = props.region ?? process.env.GOOGLE_REGION;
        if (this.region === undefined) {
            throw new Error("A Google Cloud region must be specified through the GOOGLE_REGION environment variable.");
        }
        this.zone = props.zone ?? `${this.region}-a`;
        new provider_1.GoogleProvider(this, "google", {
            project: this.projectId,
            region: this.region,
        });
        new provider_2.RandomProvider(this, "random");
        test_runner_1.TestRunner._createTree(this, props.rootConstruct);
    }
    typeForFqn(fqn) {
        switch (fqn) {
            case std_1.TEST_RUNNER_FQN:
                return test_runner_1.TestRunner;
            case cloud_1.BUCKET_FQN:
                return bucket_1.Bucket;
            case cloud_1.FUNCTION_FQN:
                return function_1.Function;
            case ex_1.TABLE_FQN:
                return table_1.Table;
            case cloud_1.COUNTER_FQN:
                return counter_1.Counter;
            case cloud_1.SCHEDULE_FQN:
                return schedule_1.Schedule;
        }
        return undefined;
    }
}
exports.App = App;
