// @winglang/sdk/lib/target-tf-gcp/counter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Counter = void 0;
const app_1 = require("./app");
const function_1 = require("./function");
const permissions_1 = require("./permissions");
const firestore_database_1 = require("../.gen/providers/google/firestore-database");
const project_service_1 = require("../.gen/providers/google/project-service");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
const resource_names_1 = require("../shared/resource-names");
/**
 * Valid database IDs include `(default)` and IDs that conform to the following:
 * - Includes only letters, numbers, and hyphen (-) characters.
 * - The first character must be a letter.
 * - The last character must be a letter or number.
 * - Minimum of 4 characters.
 * - Maximum of 63 characters.
 */
const NAME_OPTS = {
    maxLen: 63,
    case: resource_names_1.CaseConventions.LOWERCASE,
    disallowedRegex: /[^a-zA-Z0-9\-]+/g,
    prefix: "wing-counter-",
};
/**
 * GCP implementation of `cloud.Counter`.
 *
 * @inflight `@winglang/sdk.cloud.ICounterClient`
 */
class Counter extends cloud.Counter {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        const cloudFirestoreApi = new project_service_1.ProjectService(this, "CloudFirestoreAPI", {
            service: "firestore.googleapis.com",
            disableOnDestroy: false,
        });
        this.database = new firestore_database_1.FirestoreDatabase(this, "Default", {
            name: resource_names_1.ResourceNames.generateName(this, NAME_OPTS),
            locationId: app_1.App.of(this).region,
            type: "DATASTORE_MODE",
            concurrencyMode: "OPTIMISTIC",
            appEngineIntegrationMode: "DISABLED",
            pointInTimeRecoveryEnablement: "POINT_IN_TIME_RECOVERY_DISABLED",
            deleteProtectionState: "DELETE_PROTECTION_DISABLED",
            deletionPolicy: "DELETE",
            dependsOn: [cloudFirestoreApi],
        });
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.CounterInflightMethods.PEEK]: [],
            [cloud.CounterInflightMethods.INC]: [],
            [cloud.CounterInflightMethods.DEC]: [],
            [cloud.CounterInflightMethods.SET]: [],
        };
    }
    onLift(host, ops) {
        if (!(host instanceof function_1.Function)) {
            throw new Error("counters can only be bound by tfgcp.Function for now");
        }
        const permissions = (0, permissions_1.calculateCounterPermissions)(ops);
        host.addPermissions(permissions);
        host.addEnvironment(this.envName(), this.database.name);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-gcp", "shared-gcp"), __filename, "CounterClient", [`process.env["${this.envName()}"]`, `${this.initial}`]);
    }
    envName() {
        return `FIRESTORE_DATABASE_NAME_${this.node.addr.slice(-8)}`;
    }
}
exports.Counter = Counter;
