// @winglang/sdk/lib/target-tf-gcp/table.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = void 0;
const app_1 = require("./app");
const bigtable_instance_1 = require("../.gen/providers/google/bigtable-instance");
const bigtable_table_1 = require("../.gen/providers/google/bigtable-table");
const errors_1 = require("../core/errors");
const ex = __importStar(require("../ex"));
const resource_names_1 = require("../shared/resource-names");
const TABLE_NAME_OPTS = {
    maxLen: 22,
    disallowedRegex: /[a-z0-9\-\.\_]+/g,
    sep: "a",
};
const INSTANCE_NAME_OPTS = {
    maxLen: 22,
    disallowedRegex: /[a-z0-9\-\.\_]+/g,
    sep: "a",
    case: resource_names_1.CaseConventions.LOWERCASE,
};
/**
 * GCP implementation of `ex.Table`.
 *
 * @inflight `@winglang/sdk.ex.ITableClient`
 */
class Table extends ex.Table {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        if (props.initialRows) {
            throw new errors_1.NotImplementedError(`Property initialRows is not supported for the GCP target`, { resource: this.constructor.name, operation: "initialRows" });
        }
        const app = app_1.App.of(this);
        const tableName = resource_names_1.ResourceNames.generateName(this, TABLE_NAME_OPTS);
        const instanceName = resource_names_1.ResourceNames.generateName(this, INSTANCE_NAME_OPTS);
        const columnsFamily = [];
        for (let key in this.columns) {
            columnsFamily.push({ family: key });
        }
        const autoscalingConfig = {
            minNodes: 1,
            maxNodes: 3,
            cpuTarget: 35,
        };
        const instanceCluster = {
            clusterId: "default",
            storageType: "SSD",
            zone: app.zone,
            autoscalingConfig: autoscalingConfig,
        };
        const instanceConfig = {
            name: instanceName,
            cluster: [instanceCluster],
        };
        let instance = new bigtable_instance_1.BigtableInstance(this, "Instance", instanceConfig);
        const tableConfig = {
            name: tableName,
            instanceName: instance.name,
            columnFamily: columnsFamily,
            project: app.projectId,
        };
        new bigtable_table_1.BigtableTable(this, "Default", tableConfig);
    }
    addRow(_key, _row) {
        throw new errors_1.NotImplementedError("Method is not supported as a preflight for the GCP target.", { resource: this.constructor.name, operation: "addRow" });
    }
    onLift(_host, _ops) {
        throw new errors_1.NotImplementedError("Method not implemented.");
    }
    _toInflight() {
        throw new errors_1.NotImplementedError("Table's Inflight client is not implemented yet on tf-GCP target");
    }
}
exports.Table = Table;
