// @winglang/sdk/lib/target-sim/counter.inflight.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterBackend = void 0;
const fs = __importStar(require("fs"));
const path_1 = require("path");
const util_1 = require("./util");
const VALUES_FILENAME = "values.json";
class CounterBackend {
    constructor(ctx, props) {
        this.ctx = ctx;
        this.initial = props.initial ?? 0;
        this.values = new Map().set("default", this.initial);
    }
    async onStart() {
        // Load the values from disk
        const valuesFile = (0, path_1.join)(await this.ctx.statedir(), VALUES_FILENAME);
        const valueFilesExists = await (0, util_1.exists)(valuesFile);
        if (valueFilesExists) {
            const valuesContents = await fs.promises.readFile(valuesFile, "utf-8");
            const values = JSON.parse(valuesContents);
            this.values = new Map(values);
        }
    }
    async onStop() {
        // Save the values to disk
        fs.writeFileSync((0, path_1.join)(await this.ctx.statedir(), VALUES_FILENAME), JSON.stringify(Array.from(this.values.entries())));
    }
    async inc(amount = 1, key = "default") {
        const prev = this.values.get(key) ?? this.initial;
        this.values.set(key, prev + amount);
        return prev;
    }
    async dec(amount = 1, key = "default") {
        const prev = this.values.get(key) ?? this.initial;
        this.values.set(key, prev - amount);
        return prev;
    }
    async peek(key = "default") {
        return this.values.get(key) ?? this.initial;
    }
    async set(value, key = "default") {
        this.values.set(key, value);
    }
}
exports.CounterBackend = CounterBackend;
