// @winglang/sdk/lib/target-sim/endpoint.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = void 0;
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
/**
 * Simulator implementation of `cloud.Endpoint`
 */
class Endpoint extends cloud.Endpoint {
    constructor(scope, id, url, props = {}) {
        super(scope, id, url, props);
        this._inputUrl = url;
    }
    /**
     * @internal
     */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
    toSimulator() {
        const props = {
            inputUrl: this._inputUrl,
            url: this.url,
            label: this.label,
            browserSupport: this.browserSupport,
        };
        return {
            type: cloud.ENDPOINT_FQN,
            props,
        };
    }
}
exports.Endpoint = Endpoint;
