// @winglang/sdk/lib/target-sim/event-mapping.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventMapping = void 0;
const simulator_1 = require("../simulator/simulator");
class EventMapping {
    constructor(props) {
        this.publisher = props.publisher;
        this.subscriber = props.subscriber;
        this.eventSubscription = props.subscriptionProps;
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        const client = this.context.getClient(this.publisher, true);
        await client.addEventSubscription(this.subscriber, this.eventSubscription);
        return {};
    }
    async cleanup() {
        const client = this.context.getClient(this.publisher, true);
        await client.removeEventSubscription(this.subscriber);
    }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
}
exports.EventMapping = EventMapping;
