// @winglang/sdk/lib/target-sim/policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Policy = exports.POLICY_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const tokens_1 = require("./tokens");
const constants_1 = require("../constants");
const std_1 = require("../std");
exports.POLICY_FQN = (0, constants_1.fqnForType)("sim.Policy");
/**
 * Implementation of `sim.Policy`.
 *
 * @inflight `@winglang/sdk.sim.IPolicyClient`
 */
class Policy extends std_1.Resource {
    constructor(scope, id, props) {
        super(scope, id);
        this.statements = new Map();
        this.principal = props.principal;
        std_1.Node.of(this).hidden = true;
        std_1.Node.of(this).title = "Policy";
        std_1.Node.of(this).description = "A simulated resource policy";
    }
    /** @internal */
    get _liftMap() {
        return {};
    }
    /**
     * Adds a statement to the policy.
     */
    addStatement(resource, op) {
        if (!this.statements.has(resource)) {
            this.statements.set(resource, new Set());
        }
        this.statements.get(resource).add(op);
    }
    toSimulator() {
        const statements = [];
        for (const [resource, ops] of this.statements.entries()) {
            for (const op of ops) {
                statements.push({
                    resourceHandle: (0, tokens_1.simulatorHandleToken)(resource),
                    operation: op,
                });
            }
        }
        const props = {
            principal: (0, tokens_1.simulatorHandleToken)(this.principal),
            statements,
        };
        return {
            type: exports.POLICY_FQN,
            props,
        };
    }
}
exports.Policy = Policy;
_a = JSII_RTTI_SYMBOL_1;
Policy[_a] = { fqn: "@winglang/sdk.sim.Policy", version: "0.0.0" };
