// @winglang/sdk/lib/target-sim/secret.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secret = void 0;
const cloud_1 = require("../cloud");
const simulator_1 = require("../simulator/simulator");
const std_1 = require("../std");
class Secret {
    constructor(props) {
        this.name = props.name;
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        return {};
    }
    async cleanup() { }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async value() {
        this.context.addTrace({
            data: {
                message: "Get value",
            },
            level: std_1.LogLevel.VERBOSE,
            sourcePath: this.context.resourcePath,
            sourceType: cloud_1.SECRET_FQN,
            type: std_1.TraceType.RESOURCE,
            timestamp: new Date().toISOString(),
        });
        const secretValue = process.env[this.name];
        if (!secretValue) {
            throw new Error(`No value for secret ${this.name}\n(hint: try running the "wing secrets -t TARGET" to store secret)`);
        }
        return secretValue;
    }
    async valueJson() {
        return JSON.parse(await this.value());
    }
}
exports.Secret = Secret;
