// @winglang/sdk/lib/target-sim/state.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateInflightMethods = exports.State = exports.STATE_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const constants_1 = require("../constants");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `State`.
 */
exports.STATE_FQN = (0, constants_1.fqnForType)("sim.State");
/**
 * Key/value in-memory state for the simulator.
 *
 * Use the preflight method `token(key)` to obtain a token that can be used to reference the value
 * of the state at runtime.
 *
 * During deployment (i.e. `cloud.OnDeploy` or `cloud.Service` startup), you must call the inflight
 * method `set(key, value)` to set the runtime value. The value will be available at runtime through
 * the inflight method `get(key)` (or resolved as a token).
 *
 * See tests for examples.
 *
 * @inflight `@winglang/sdk.sim.IStateClient`
 */
class State extends std_1.Resource {
    /**
     * Returns a token that can be used to retrieve the value of the state after the simulation has
     * run.
     * @param key The object key retrieved through the inflight `state.get()`.
     */
    token(key) {
        return (0, tokens_1.simulatorAttrToken)(this, key);
    }
    /** @internal */
    get _liftMap() {
        return {
            [StateInflightMethods.GET]: [],
            [StateInflightMethods.SET]: [],
            [StateInflightMethods.TRY_GET]: [],
        };
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    toSimulator() {
        const props = {};
        return {
            type: exports.STATE_FQN,
            props,
        };
    }
}
exports.State = State;
_a = JSII_RTTI_SYMBOL_1;
State[_a] = { fqn: "@winglang/sdk.sim.State", version: "0.0.0" };
/**
 * List of inflight operations available for `State`.
 * @internal
 */
var StateInflightMethods;
(function (StateInflightMethods) {
    /** `State.set` */
    StateInflightMethods["SET"] = "set";
    /** `State.get` */
    StateInflightMethods["GET"] = "get";
    /**`State.tryGet` */
    StateInflightMethods["TRY_GET"] = "tryGet";
})(StateInflightMethods || (exports.StateInflightMethods = StateInflightMethods = {}));
