// @winglang/sdk/lib/target-sim/table.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = void 0;
const util_1 = require("./util");
const ex = __importStar(require("../ex"));
/**
 * Simulator implementation of `ex.Table`.
 *
 * @inflight `@winglang/sdk.ex.ITableClient`
 */
class Table extends ex.Table {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        this.initialRows = {};
    }
    addRow(key, row) {
        this.initialRows[key] = { ...row, [this.primaryKey]: key };
    }
    toSimulator() {
        const props = {
            name: this.name,
            columns: this.columns,
            primaryKey: this.primaryKey,
            initialRows: this.initialRows,
        };
        return {
            type: ex.TABLE_FQN,
            props,
        };
    }
    /** @internal */
    get _liftMap() {
        return {
            [ex.TableInflightMethods.INSERT]: [],
            [ex.TableInflightMethods.UPSERT]: [],
            [ex.TableInflightMethods.UPDATE]: [],
            [ex.TableInflightMethods.DELETE]: [],
            [ex.TableInflightMethods.GET]: [],
            [ex.TableInflightMethods.TRYGET]: [],
            [ex.TableInflightMethods.LIST]: [],
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Table = Table;
