// @winglang/sdk/lib/target-sim/test-runner.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunner = void 0;
const simulator_1 = require("../simulator");
const std_1 = require("../std");
class TestRunner {
    constructor(props) {
        this.tests = new Map(Object.entries(props.tests));
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        return {};
    }
    async cleanup() {
        return;
    }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async listTests() {
        return Array.from(this.tests.keys());
    }
    async runTest(path) {
        let functionHandle = this.tests.get(path);
        if (!functionHandle) {
            throw new Error(`No test found at path "${path}"`);
        }
        const fnClient = this.context.getClient(functionHandle, true);
        let pass = false;
        let error;
        const previousTraces = this.context.listTraces().length;
        try {
            await fnClient.invoke();
            pass = true;
        }
        catch (e) {
            error = e.stack;
        }
        // only return traces that were added after the test was run
        const newTraces = this.context.listTraces().slice(previousTraces);
        // as well as any log trace prior to that- https://github.com/winglang/wing/issues/4995
        const logTraces = this.context
            .listTraces()
            .slice(0, previousTraces)
            .filter((trace) => trace.type === std_1.TraceType.LOG);
        return {
            path,
            pass,
            error,
            traces: [...logTraces, ...newTraces],
        };
    }
}
exports.TestRunner = TestRunner;
