// @winglang/sdk/lib/target-sim/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimTokens = exports.isSimulatorToken = exports.SIMULATOR_TOKEN_REGEX_FULL = exports.SIMULATOR_TOKEN_REGEX = exports.simulatorAttrToken = exports.simulatorHandleToken = void 0;
const tokens_1 = require("../core/tokens");
/**
 * Produce a token that will be replaced with the handle of a resource
 * when the simulator is started. This can be inserted to an environment variable
 * so that the real value can be used by an inflight function.
 */
function simulatorHandleToken(resource) {
    return simulatorAttrToken(resource, "handle");
}
exports.simulatorHandleToken = simulatorHandleToken;
/**
 * Produce a token that will be replaced with a deploy-time resource attribute
 * when the simulator is started.
 */
function simulatorAttrToken(resource, attrName) {
    return `\${wsim#${resource.node.path}#attrs.${attrName}}`;
}
exports.simulatorAttrToken = simulatorAttrToken;
/**
 * Regular expression that matches a simulator token. This is a sequence of
 * characters that can appear in the middle of any string in the following format:
 *
 * ${wsim#path/to/resource#property.path}
 */
exports.SIMULATOR_TOKEN_REGEX = /\$\{wsim#[^#\{\}]+#[a-zA-Z0-9_\-\/\.]+\}/;
/**
 * The same as SIMULATOR_TOKEN_REGEX, but it must match the entire string.
 */
exports.SIMULATOR_TOKEN_REGEX_FULL = new RegExp(`^${exports.SIMULATOR_TOKEN_REGEX.source}$`);
/**
 * Returns true is the given value is a Simulator token.
 */
function isSimulatorToken(value) {
    return exports.SIMULATOR_TOKEN_REGEX.test(value);
}
exports.isSimulatorToken = isSimulatorToken;
/**
 * Represents values that can only be resolved after the app is synthesized.
 * Tokens values are captured as environment variable, and resolved through the compilation target token mechanism.
 */
class SimTokens {
    /**
     * Returns true is the given value is a Simulator token.
     */
    isToken(value) {
        if (typeof value === "string") {
            return isSimulatorToken(value);
        }
        return false;
    }
    /**
     * Lifts a value into an inflight context.
     */
    lift(value) {
        switch (typeof value) {
            case "string":
                return `process.env[${JSON.stringify((0, tokens_1.tokenEnvName)(value))}]`;
            default:
                throw new Error(`Unsupported token type`);
        }
    }
    /**
     * Lifts the given token to the host.
     */
    onLiftValue(host, value) {
        switch (typeof value) {
            case "string":
                const envName = (0, tokens_1.tokenEnvName)(value);
                host.addEnvironment(envName, value);
                break;
            default:
                throw new Error(`Unable to lift token ${value}`);
        }
    }
}
exports.SimTokens = SimTokens;
