// @winglang/sdk/lib/target-sim/topic.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Topic = void 0;
const cloud_1 = require("../cloud");
const simulator_1 = require("../simulator/simulator");
const std_1 = require("../std");
class Topic {
    constructor(_props) {
        this.subscribers = new Array();
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        return {};
    }
    async cleanup() { }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async publishMessage(message) {
        for (const subscriber of this.subscribers) {
            const fnClient = this.context.getClient(subscriber.functionHandle);
            this.context.addTrace({
                type: std_1.TraceType.RESOURCE,
                level: std_1.LogLevel.VERBOSE,
                data: {
                    message: `Sending message (message=${message}, subscriber=${subscriber.functionHandle}).`,
                },
                sourcePath: this.context.resourcePath,
                sourceType: cloud_1.TOPIC_FQN,
                timestamp: new Date().toISOString(),
            });
            await fnClient.invokeAsync(message);
        }
    }
    async addEventSubscription(subscriber, subscriptionProps) {
        let s = {
            functionHandle: subscriber,
            ...subscriptionProps,
        };
        this.subscribers.push(s);
    }
    async removeEventSubscription(subscriber) {
        const index = this.subscribers.findIndex((s) => s.functionHandle === subscriber);
        if (index >= 0) {
            this.subscribers.splice(index, 1);
        }
    }
    publish(...messages) {
        return this.context.withTrace({
            message: `Publish (messages=${messages}).`,
            activity: async () => {
                if (messages.includes("")) {
                    throw new Error("Empty messages are not allowed");
                }
                let publishAll = [];
                for (const message of messages) {
                    publishAll.push(this.publishMessage(message));
                }
                return Promise.all(publishAll);
            },
        });
    }
}
exports.Topic = Topic;
