// @winglang/sdk/lib/target-sim/website.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Website = void 0;
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
const std_1 = require("../std");
/**
 * A static website.
 */
class Website extends cloud.Website {
    constructor(scope, id, props) {
        super(scope, id, props);
        this.fileRoutes = {};
        this.endpoint = new cloud.Endpoint(this, "Endpoint", (0, tokens_1.simulatorAttrToken)(this, "url"), { label: `Website ${this.node.path}`, browserSupport: true });
        std_1.Node.of(this.endpoint).hidden = true;
        this.errorDocument = props.errorDocument;
    }
    get _endpoint() {
        return this.endpoint;
    }
    /** Adds a file to the website during deployment */
    addFile(path, data, options) {
        this.fileRoutes[path] = {
            data,
            contentType: options?.contentType ?? "text/plain",
        };
        return `${this.url}/${path}`;
    }
    /** Returns sim schema */
    toSimulator() {
        const props = {
            staticFilesPath: this.path,
            fileRoutes: this.fileRoutes,
            errorDocument: this.errorDocument,
        };
        return {
            type: cloud.WEBSITE_FQN,
            props,
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Website = Website;
