// @cdktf/provider-aws/lib/datasync-location-hdfs/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncLocationHdfs = exports.DatasyncLocationHdfsQopConfigurationOutputReference = exports.datasyncLocationHdfsQopConfigurationToHclTerraform = exports.datasyncLocationHdfsQopConfigurationToTerraform = exports.DatasyncLocationHdfsNameNodeList = exports.DatasyncLocationHdfsNameNodeOutputReference = exports.datasyncLocationHdfsNameNodeToHclTerraform = exports.datasyncLocationHdfsNameNodeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncLocationHdfsNameNodeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hostname: cdktf.stringToTerraform(struct.hostname),
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.datasyncLocationHdfsNameNodeToTerraform = datasyncLocationHdfsNameNodeToTerraform;
function datasyncLocationHdfsNameNodeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hostname: {
            value: cdktf.stringToHclTerraform(struct.hostname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncLocationHdfsNameNodeToHclTerraform = datasyncLocationHdfsNameNodeToHclTerraform;
class DatasyncLocationHdfsNameNodeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostname = this._hostname;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._hostname = undefined;
            this._port = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._hostname = value.hostname;
            this._port = value.port;
        }
    }
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    set hostname(value) {
        this._hostname = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.DatasyncLocationHdfsNameNodeOutputReference = DatasyncLocationHdfsNameNodeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncLocationHdfsNameNodeOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasyncLocationHdfs.DatasyncLocationHdfsNameNodeOutputReference", version: "19.15.0" };
class DatasyncLocationHdfsNameNodeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatasyncLocationHdfsNameNodeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatasyncLocationHdfsNameNodeList = DatasyncLocationHdfsNameNodeList;
_b = JSII_RTTI_SYMBOL_1;
DatasyncLocationHdfsNameNodeList[_b] = { fqn: "@cdktf/provider-aws.datasyncLocationHdfs.DatasyncLocationHdfsNameNodeList", version: "19.15.0" };
function datasyncLocationHdfsQopConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_transfer_protection: cdktf.stringToTerraform(struct.dataTransferProtection),
        rpc_protection: cdktf.stringToTerraform(struct.rpcProtection),
    };
}
exports.datasyncLocationHdfsQopConfigurationToTerraform = datasyncLocationHdfsQopConfigurationToTerraform;
function datasyncLocationHdfsQopConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_transfer_protection: {
            value: cdktf.stringToHclTerraform(struct.dataTransferProtection),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rpc_protection: {
            value: cdktf.stringToHclTerraform(struct.rpcProtection),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncLocationHdfsQopConfigurationToHclTerraform = datasyncLocationHdfsQopConfigurationToHclTerraform;
class DatasyncLocationHdfsQopConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataTransferProtection !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataTransferProtection = this._dataTransferProtection;
        }
        if (this._rpcProtection !== undefined) {
            hasAnyValues = true;
            internalValueResult.rpcProtection = this._rpcProtection;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataTransferProtection = undefined;
            this._rpcProtection = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataTransferProtection = value.dataTransferProtection;
            this._rpcProtection = value.rpcProtection;
        }
    }
    get dataTransferProtection() {
        return this.getStringAttribute('data_transfer_protection');
    }
    set dataTransferProtection(value) {
        this._dataTransferProtection = value;
    }
    resetDataTransferProtection() {
        this._dataTransferProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataTransferProtectionInput() {
        return this._dataTransferProtection;
    }
    get rpcProtection() {
        return this.getStringAttribute('rpc_protection');
    }
    set rpcProtection(value) {
        this._rpcProtection = value;
    }
    resetRpcProtection() {
        this._rpcProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rpcProtectionInput() {
        return this._rpcProtection;
    }
}
exports.DatasyncLocationHdfsQopConfigurationOutputReference = DatasyncLocationHdfsQopConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DatasyncLocationHdfsQopConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.datasyncLocationHdfs.DatasyncLocationHdfsQopConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_hdfs aws_datasync_location_hdfs}
*/
class DatasyncLocationHdfs extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DatasyncLocationHdfs resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DatasyncLocationHdfs to import
    * @param importFromId The id of the existing DatasyncLocationHdfs that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_hdfs#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DatasyncLocationHdfs to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_datasync_location_hdfs", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_hdfs aws_datasync_location_hdfs} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DatasyncLocationHdfsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_location_hdfs',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // name_node - computed: false, optional: false, required: true
        this._nameNode = new DatasyncLocationHdfsNameNodeList(this, "name_node", true);
        // qop_configuration - computed: false, optional: true, required: false
        this._qopConfiguration = new DatasyncLocationHdfsQopConfigurationOutputReference(this, "qop_configuration");
        this._agentArns = config.agentArns;
        this._authenticationType = config.authenticationType;
        this._blockSize = config.blockSize;
        this._id = config.id;
        this._kerberosKeytab = config.kerberosKeytab;
        this._kerberosKeytabBase64 = config.kerberosKeytabBase64;
        this._kerberosKrb5Conf = config.kerberosKrb5Conf;
        this._kerberosKrb5ConfBase64 = config.kerberosKrb5ConfBase64;
        this._kerberosPrincipal = config.kerberosPrincipal;
        this._kmsKeyProviderUri = config.kmsKeyProviderUri;
        this._replicationFactor = config.replicationFactor;
        this._simpleUser = config.simpleUser;
        this._subdirectory = config.subdirectory;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._nameNode.internalValue = config.nameNode;
        this._qopConfiguration.internalValue = config.qopConfiguration;
    }
    get agentArns() {
        return cdktf.Fn.tolist(this.getListAttribute('agent_arns'));
    }
    set agentArns(value) {
        this._agentArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get agentArnsInput() {
        return this._agentArns;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authenticationType() {
        return this.getStringAttribute('authentication_type');
    }
    set authenticationType(value) {
        this._authenticationType = value;
    }
    resetAuthenticationType() {
        this._authenticationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationTypeInput() {
        return this._authenticationType;
    }
    get blockSize() {
        return this.getNumberAttribute('block_size');
    }
    set blockSize(value) {
        this._blockSize = value;
    }
    resetBlockSize() {
        this._blockSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blockSizeInput() {
        return this._blockSize;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kerberosKeytab() {
        return this.getStringAttribute('kerberos_keytab');
    }
    set kerberosKeytab(value) {
        this._kerberosKeytab = value;
    }
    resetKerberosKeytab() {
        this._kerberosKeytab = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosKeytabInput() {
        return this._kerberosKeytab;
    }
    get kerberosKeytabBase64() {
        return this.getStringAttribute('kerberos_keytab_base64');
    }
    set kerberosKeytabBase64(value) {
        this._kerberosKeytabBase64 = value;
    }
    resetKerberosKeytabBase64() {
        this._kerberosKeytabBase64 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosKeytabBase64Input() {
        return this._kerberosKeytabBase64;
    }
    get kerberosKrb5Conf() {
        return this.getStringAttribute('kerberos_krb5_conf');
    }
    set kerberosKrb5Conf(value) {
        this._kerberosKrb5Conf = value;
    }
    resetKerberosKrb5Conf() {
        this._kerberosKrb5Conf = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosKrb5ConfInput() {
        return this._kerberosKrb5Conf;
    }
    get kerberosKrb5ConfBase64() {
        return this.getStringAttribute('kerberos_krb5_conf_base64');
    }
    set kerberosKrb5ConfBase64(value) {
        this._kerberosKrb5ConfBase64 = value;
    }
    resetKerberosKrb5ConfBase64() {
        this._kerberosKrb5ConfBase64 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosKrb5ConfBase64Input() {
        return this._kerberosKrb5ConfBase64;
    }
    get kerberosPrincipal() {
        return this.getStringAttribute('kerberos_principal');
    }
    set kerberosPrincipal(value) {
        this._kerberosPrincipal = value;
    }
    resetKerberosPrincipal() {
        this._kerberosPrincipal = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosPrincipalInput() {
        return this._kerberosPrincipal;
    }
    get kmsKeyProviderUri() {
        return this.getStringAttribute('kms_key_provider_uri');
    }
    set kmsKeyProviderUri(value) {
        this._kmsKeyProviderUri = value;
    }
    resetKmsKeyProviderUri() {
        this._kmsKeyProviderUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyProviderUriInput() {
        return this._kmsKeyProviderUri;
    }
    get replicationFactor() {
        return this.getNumberAttribute('replication_factor');
    }
    set replicationFactor(value) {
        this._replicationFactor = value;
    }
    resetReplicationFactor() {
        this._replicationFactor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationFactorInput() {
        return this._replicationFactor;
    }
    get simpleUser() {
        return this.getStringAttribute('simple_user');
    }
    set simpleUser(value) {
        this._simpleUser = value;
    }
    resetSimpleUser() {
        this._simpleUser = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get simpleUserInput() {
        return this._simpleUser;
    }
    get subdirectory() {
        return this.getStringAttribute('subdirectory');
    }
    set subdirectory(value) {
        this._subdirectory = value;
    }
    resetSubdirectory() {
        this._subdirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subdirectoryInput() {
        return this._subdirectory;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
    get nameNode() {
        return this._nameNode;
    }
    putNameNode(value) {
        this._nameNode.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameNodeInput() {
        return this._nameNode.internalValue;
    }
    get qopConfiguration() {
        return this._qopConfiguration;
    }
    putQopConfiguration(value) {
        this._qopConfiguration.internalValue = value;
    }
    resetQopConfiguration() {
        this._qopConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get qopConfigurationInput() {
        return this._qopConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            agent_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._agentArns),
            authentication_type: cdktf.stringToTerraform(this._authenticationType),
            block_size: cdktf.numberToTerraform(this._blockSize),
            id: cdktf.stringToTerraform(this._id),
            kerberos_keytab: cdktf.stringToTerraform(this._kerberosKeytab),
            kerberos_keytab_base64: cdktf.stringToTerraform(this._kerberosKeytabBase64),
            kerberos_krb5_conf: cdktf.stringToTerraform(this._kerberosKrb5Conf),
            kerberos_krb5_conf_base64: cdktf.stringToTerraform(this._kerberosKrb5ConfBase64),
            kerberos_principal: cdktf.stringToTerraform(this._kerberosPrincipal),
            kms_key_provider_uri: cdktf.stringToTerraform(this._kmsKeyProviderUri),
            replication_factor: cdktf.numberToTerraform(this._replicationFactor),
            simple_user: cdktf.stringToTerraform(this._simpleUser),
            subdirectory: cdktf.stringToTerraform(this._subdirectory),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            name_node: cdktf.listMapper(datasyncLocationHdfsNameNodeToTerraform, true)(this._nameNode.internalValue),
            qop_configuration: datasyncLocationHdfsQopConfigurationToTerraform(this._qopConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            agent_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._agentArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            authentication_type: {
                value: cdktf.stringToHclTerraform(this._authenticationType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            block_size: {
                value: cdktf.numberToHclTerraform(this._blockSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_keytab: {
                value: cdktf.stringToHclTerraform(this._kerberosKeytab),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_keytab_base64: {
                value: cdktf.stringToHclTerraform(this._kerberosKeytabBase64),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_krb5_conf: {
                value: cdktf.stringToHclTerraform(this._kerberosKrb5Conf),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_krb5_conf_base64: {
                value: cdktf.stringToHclTerraform(this._kerberosKrb5ConfBase64),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_principal: {
                value: cdktf.stringToHclTerraform(this._kerberosPrincipal),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_provider_uri: {
                value: cdktf.stringToHclTerraform(this._kmsKeyProviderUri),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            replication_factor: {
                value: cdktf.numberToHclTerraform(this._replicationFactor),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            simple_user: {
                value: cdktf.stringToHclTerraform(this._simpleUser),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subdirectory: {
                value: cdktf.stringToHclTerraform(this._subdirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            name_node: {
                value: cdktf.listMapperHcl(datasyncLocationHdfsNameNodeToHclTerraform, true)(this._nameNode.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DatasyncLocationHdfsNameNodeList",
            },
            qop_configuration: {
                value: datasyncLocationHdfsQopConfigurationToHclTerraform(this._qopConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncLocationHdfsQopConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DatasyncLocationHdfs = DatasyncLocationHdfs;
_d = JSII_RTTI_SYMBOL_1;
DatasyncLocationHdfs[_d] = { fqn: "@cdktf/provider-aws.datasyncLocationHdfs.DatasyncLocationHdfs", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DatasyncLocationHdfs.tfResourceType = "aws_datasync_location_hdfs";
