// @cdktf/provider-aws/lib/s3-bucket-logging/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketLoggingA = exports.S3BucketLoggingTargetObjectKeyFormatOutputReference = exports.s3BucketLoggingTargetObjectKeyFormatToHclTerraform = exports.s3BucketLoggingTargetObjectKeyFormatToTerraform = exports.S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference = exports.s3BucketLoggingTargetObjectKeyFormatSimplePrefixToHclTerraform = exports.s3BucketLoggingTargetObjectKeyFormatSimplePrefixToTerraform = exports.S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference = exports.s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToHclTerraform = exports.s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToTerraform = exports.S3BucketLoggingTargetGrantList = exports.S3BucketLoggingTargetGrantOutputReference = exports.s3BucketLoggingTargetGrantToHclTerraform = exports.s3BucketLoggingTargetGrantToTerraform = exports.S3BucketLoggingTargetGrantGranteeOutputReference = exports.s3BucketLoggingTargetGrantGranteeToHclTerraform = exports.s3BucketLoggingTargetGrantGranteeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketLoggingTargetGrantGranteeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        email_address: cdktf.stringToTerraform(struct.emailAddress),
        id: cdktf.stringToTerraform(struct.id),
        type: cdktf.stringToTerraform(struct.type),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.s3BucketLoggingTargetGrantGranteeToTerraform = s3BucketLoggingTargetGrantGranteeToTerraform;
function s3BucketLoggingTargetGrantGranteeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        email_address: {
            value: cdktf.stringToHclTerraform(struct.emailAddress),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketLoggingTargetGrantGranteeToHclTerraform = s3BucketLoggingTargetGrantGranteeToHclTerraform;
class S3BucketLoggingTargetGrantGranteeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._emailAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.emailAddress = this._emailAddress;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._emailAddress = undefined;
            this._id = undefined;
            this._type = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._emailAddress = value.emailAddress;
            this._id = value.id;
            this._type = value.type;
            this._uri = value.uri;
        }
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get emailAddress() {
        return this.getStringAttribute('email_address');
    }
    set emailAddress(value) {
        this._emailAddress = value;
    }
    resetEmailAddress() {
        this._emailAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailAddressInput() {
        return this._emailAddress;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.S3BucketLoggingTargetGrantGranteeOutputReference = S3BucketLoggingTargetGrantGranteeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetGrantGranteeOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetGrantGranteeOutputReference", version: "19.15.0" };
function s3BucketLoggingTargetGrantToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        permission: cdktf.stringToTerraform(struct.permission),
        grantee: s3BucketLoggingTargetGrantGranteeToTerraform(struct.grantee),
    };
}
exports.s3BucketLoggingTargetGrantToTerraform = s3BucketLoggingTargetGrantToTerraform;
function s3BucketLoggingTargetGrantToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        permission: {
            value: cdktf.stringToHclTerraform(struct.permission),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        grantee: {
            value: s3BucketLoggingTargetGrantGranteeToHclTerraform(struct.grantee),
            isBlock: true,
            type: "list",
            storageClassType: "S3BucketLoggingTargetGrantGranteeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketLoggingTargetGrantToHclTerraform = s3BucketLoggingTargetGrantToHclTerraform;
class S3BucketLoggingTargetGrantOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // grantee - computed: false, optional: false, required: true
        this._grantee = new S3BucketLoggingTargetGrantGranteeOutputReference(this, "grantee");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._permission !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission;
        }
        if (this._grantee?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grantee = this._grantee?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._permission = undefined;
            this._grantee.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._permission = value.permission;
            this._grantee.internalValue = value.grantee;
        }
    }
    get permission() {
        return this.getStringAttribute('permission');
    }
    set permission(value) {
        this._permission = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission;
    }
    get grantee() {
        return this._grantee;
    }
    putGrantee(value) {
        this._grantee.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get granteeInput() {
        return this._grantee.internalValue;
    }
}
exports.S3BucketLoggingTargetGrantOutputReference = S3BucketLoggingTargetGrantOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetGrantOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetGrantOutputReference", version: "19.15.0" };
class S3BucketLoggingTargetGrantList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketLoggingTargetGrantOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketLoggingTargetGrantList = S3BucketLoggingTargetGrantList;
_c = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetGrantList[_c] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetGrantList", version: "19.15.0" };
function s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partition_date_source: cdktf.stringToTerraform(struct.partitionDateSource),
    };
}
exports.s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToTerraform = s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToTerraform;
function s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        partition_date_source: {
            value: cdktf.stringToHclTerraform(struct.partitionDateSource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToHclTerraform = s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToHclTerraform;
class S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._partitionDateSource !== undefined) {
            hasAnyValues = true;
            internalValueResult.partitionDateSource = this._partitionDateSource;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._partitionDateSource = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._partitionDateSource = value.partitionDateSource;
        }
    }
    get partitionDateSource() {
        return this.getStringAttribute('partition_date_source');
    }
    set partitionDateSource(value) {
        this._partitionDateSource = value;
    }
    // Temporarily expose input value. Use with caution.
    get partitionDateSourceInput() {
        return this._partitionDateSource;
    }
}
exports.S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference = S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference", version: "19.15.0" };
function s3BucketLoggingTargetObjectKeyFormatSimplePrefixToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.s3BucketLoggingTargetObjectKeyFormatSimplePrefixToTerraform = s3BucketLoggingTargetObjectKeyFormatSimplePrefixToTerraform;
function s3BucketLoggingTargetObjectKeyFormatSimplePrefixToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.s3BucketLoggingTargetObjectKeyFormatSimplePrefixToHclTerraform = s3BucketLoggingTargetObjectKeyFormatSimplePrefixToHclTerraform;
class S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference = S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference", version: "19.15.0" };
function s3BucketLoggingTargetObjectKeyFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partitioned_prefix: s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToTerraform(struct.partitionedPrefix),
        simple_prefix: s3BucketLoggingTargetObjectKeyFormatSimplePrefixToTerraform(struct.simplePrefix),
    };
}
exports.s3BucketLoggingTargetObjectKeyFormatToTerraform = s3BucketLoggingTargetObjectKeyFormatToTerraform;
function s3BucketLoggingTargetObjectKeyFormatToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        partitioned_prefix: {
            value: s3BucketLoggingTargetObjectKeyFormatPartitionedPrefixToHclTerraform(struct.partitionedPrefix),
            isBlock: true,
            type: "list",
            storageClassType: "S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixList",
        },
        simple_prefix: {
            value: s3BucketLoggingTargetObjectKeyFormatSimplePrefixToHclTerraform(struct.simplePrefix),
            isBlock: true,
            type: "list",
            storageClassType: "S3BucketLoggingTargetObjectKeyFormatSimplePrefixList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketLoggingTargetObjectKeyFormatToHclTerraform = s3BucketLoggingTargetObjectKeyFormatToHclTerraform;
class S3BucketLoggingTargetObjectKeyFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // partitioned_prefix - computed: false, optional: true, required: false
        this._partitionedPrefix = new S3BucketLoggingTargetObjectKeyFormatPartitionedPrefixOutputReference(this, "partitioned_prefix");
        // simple_prefix - computed: false, optional: true, required: false
        this._simplePrefix = new S3BucketLoggingTargetObjectKeyFormatSimplePrefixOutputReference(this, "simple_prefix");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._partitionedPrefix?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.partitionedPrefix = this._partitionedPrefix?.internalValue;
        }
        if (this._simplePrefix?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.simplePrefix = this._simplePrefix?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._partitionedPrefix.internalValue = undefined;
            this._simplePrefix.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._partitionedPrefix.internalValue = value.partitionedPrefix;
            this._simplePrefix.internalValue = value.simplePrefix;
        }
    }
    get partitionedPrefix() {
        return this._partitionedPrefix;
    }
    putPartitionedPrefix(value) {
        this._partitionedPrefix.internalValue = value;
    }
    resetPartitionedPrefix() {
        this._partitionedPrefix.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get partitionedPrefixInput() {
        return this._partitionedPrefix.internalValue;
    }
    get simplePrefix() {
        return this._simplePrefix;
    }
    putSimplePrefix(value) {
        this._simplePrefix.internalValue = value;
    }
    resetSimplePrefix() {
        this._simplePrefix.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get simplePrefixInput() {
        return this._simplePrefix.internalValue;
    }
}
exports.S3BucketLoggingTargetObjectKeyFormatOutputReference = S3BucketLoggingTargetObjectKeyFormatOutputReference;
_f = JSII_RTTI_SYMBOL_1;
S3BucketLoggingTargetObjectKeyFormatOutputReference[_f] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingTargetObjectKeyFormatOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_logging aws_s3_bucket_logging}
*/
class S3BucketLoggingA extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a S3BucketLoggingA resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the S3BucketLoggingA to import
    * @param importFromId The id of the existing S3BucketLoggingA that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_logging#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the S3BucketLoggingA to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_s3_bucket_logging", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_logging aws_s3_bucket_logging} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketLoggingAConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_logging',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // target_grant - computed: false, optional: true, required: false
        this._targetGrant = new S3BucketLoggingTargetGrantList(this, "target_grant", true);
        // target_object_key_format - computed: false, optional: true, required: false
        this._targetObjectKeyFormat = new S3BucketLoggingTargetObjectKeyFormatOutputReference(this, "target_object_key_format");
        this._bucket = config.bucket;
        this._expectedBucketOwner = config.expectedBucketOwner;
        this._id = config.id;
        this._targetBucket = config.targetBucket;
        this._targetPrefix = config.targetPrefix;
        this._targetGrant.internalValue = config.targetGrant;
        this._targetObjectKeyFormat.internalValue = config.targetObjectKeyFormat;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get expectedBucketOwner() {
        return this.getStringAttribute('expected_bucket_owner');
    }
    set expectedBucketOwner(value) {
        this._expectedBucketOwner = value;
    }
    resetExpectedBucketOwner() {
        this._expectedBucketOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expectedBucketOwnerInput() {
        return this._expectedBucketOwner;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get targetBucket() {
        return this.getStringAttribute('target_bucket');
    }
    set targetBucket(value) {
        this._targetBucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetBucketInput() {
        return this._targetBucket;
    }
    get targetPrefix() {
        return this.getStringAttribute('target_prefix');
    }
    set targetPrefix(value) {
        this._targetPrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetPrefixInput() {
        return this._targetPrefix;
    }
    get targetGrant() {
        return this._targetGrant;
    }
    putTargetGrant(value) {
        this._targetGrant.internalValue = value;
    }
    resetTargetGrant() {
        this._targetGrant.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetGrantInput() {
        return this._targetGrant.internalValue;
    }
    get targetObjectKeyFormat() {
        return this._targetObjectKeyFormat;
    }
    putTargetObjectKeyFormat(value) {
        this._targetObjectKeyFormat.internalValue = value;
    }
    resetTargetObjectKeyFormat() {
        this._targetObjectKeyFormat.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetObjectKeyFormatInput() {
        return this._targetObjectKeyFormat.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            expected_bucket_owner: cdktf.stringToTerraform(this._expectedBucketOwner),
            id: cdktf.stringToTerraform(this._id),
            target_bucket: cdktf.stringToTerraform(this._targetBucket),
            target_prefix: cdktf.stringToTerraform(this._targetPrefix),
            target_grant: cdktf.listMapper(s3BucketLoggingTargetGrantToTerraform, true)(this._targetGrant.internalValue),
            target_object_key_format: s3BucketLoggingTargetObjectKeyFormatToTerraform(this._targetObjectKeyFormat.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bucket: {
                value: cdktf.stringToHclTerraform(this._bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            expected_bucket_owner: {
                value: cdktf.stringToHclTerraform(this._expectedBucketOwner),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_bucket: {
                value: cdktf.stringToHclTerraform(this._targetBucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_prefix: {
                value: cdktf.stringToHclTerraform(this._targetPrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_grant: {
                value: cdktf.listMapperHcl(s3BucketLoggingTargetGrantToHclTerraform, true)(this._targetGrant.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "S3BucketLoggingTargetGrantList",
            },
            target_object_key_format: {
                value: s3BucketLoggingTargetObjectKeyFormatToHclTerraform(this._targetObjectKeyFormat.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3BucketLoggingTargetObjectKeyFormatList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.S3BucketLoggingA = S3BucketLoggingA;
_g = JSII_RTTI_SYMBOL_1;
S3BucketLoggingA[_g] = { fqn: "@cdktf/provider-aws.s3BucketLogging.S3BucketLoggingA", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
S3BucketLoggingA.tfResourceType = "aws_s3_bucket_logging";
