// @cdktf/provider-aws/lib/data-aws-networkmanager-core-network-policy-document/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocument = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToTerraform = exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToHclTerraform = exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        association_method: cdktf.stringToTerraform(struct.associationMethod),
        require_acceptance: cdktf.booleanToTerraform(struct.requireAcceptance),
        segment: cdktf.stringToTerraform(struct.segment),
        tag_value_of_key: cdktf.stringToTerraform(struct.tagValueOfKey),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        association_method: {
            value: cdktf.stringToHclTerraform(struct.associationMethod),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        require_acceptance: {
            value: cdktf.booleanToHclTerraform(struct.requireAcceptance),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        segment: {
            value: cdktf.stringToHclTerraform(struct.segment),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tag_value_of_key: {
            value: cdktf.stringToHclTerraform(struct.tagValueOfKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._associationMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.associationMethod = this._associationMethod;
        }
        if (this._requireAcceptance !== undefined) {
            hasAnyValues = true;
            internalValueResult.requireAcceptance = this._requireAcceptance;
        }
        if (this._segment !== undefined) {
            hasAnyValues = true;
            internalValueResult.segment = this._segment;
        }
        if (this._tagValueOfKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagValueOfKey = this._tagValueOfKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._associationMethod = undefined;
            this._requireAcceptance = undefined;
            this._segment = undefined;
            this._tagValueOfKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._associationMethod = value.associationMethod;
            this._requireAcceptance = value.requireAcceptance;
            this._segment = value.segment;
            this._tagValueOfKey = value.tagValueOfKey;
        }
    }
    get associationMethod() {
        return this.getStringAttribute('association_method');
    }
    set associationMethod(value) {
        this._associationMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    get associationMethodInput() {
        return this._associationMethod;
    }
    get requireAcceptance() {
        return this.getBooleanAttribute('require_acceptance');
    }
    set requireAcceptance(value) {
        this._requireAcceptance = value;
    }
    resetRequireAcceptance() {
        this._requireAcceptance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requireAcceptanceInput() {
        return this._requireAcceptance;
    }
    get segment() {
        return this.getStringAttribute('segment');
    }
    set segment(value) {
        this._segment = value;
    }
    resetSegment() {
        this._segment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentInput() {
        return this._segment;
    }
    get tagValueOfKey() {
        return this.getStringAttribute('tag_value_of_key');
    }
    set tagValueOfKey(value) {
        this._tagValueOfKey = value;
    }
    resetTagValueOfKey() {
        this._tagValueOfKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagValueOfKeyInput() {
        return this._tagValueOfKey;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        operator: cdktf.stringToTerraform(struct.operator),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._operator = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._operator = value.operator;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    resetOperator() {
        this._operator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList;
_c = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList[_c] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        condition_logic: cdktf.stringToTerraform(struct.conditionLogic),
        description: cdktf.stringToTerraform(struct.description),
        rule_number: cdktf.numberToTerraform(struct.ruleNumber),
        action: dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToTerraform(struct.action),
        conditions: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToTerraform, true)(struct.conditions),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        condition_logic: {
            value: cdktf.stringToHclTerraform(struct.conditionLogic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rule_number: {
            value: cdktf.numberToHclTerraform(struct.ruleNumber),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        action: {
            value: dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionToHclTerraform(struct.action),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionList",
        },
        conditions: {
            value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsToHclTerraform, true)(struct.conditions),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesActionOutputReference(this, "action");
        // conditions - computed: false, optional: false, required: true
        this._conditions = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesConditionsList(this, "conditions", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._conditionLogic !== undefined) {
            hasAnyValues = true;
            internalValueResult.conditionLogic = this._conditionLogic;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._ruleNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleNumber = this._ruleNumber;
        }
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._conditions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.conditions = this._conditions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._conditionLogic = undefined;
            this._description = undefined;
            this._ruleNumber = undefined;
            this._action.internalValue = undefined;
            this._conditions.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._conditionLogic = value.conditionLogic;
            this._description = value.description;
            this._ruleNumber = value.ruleNumber;
            this._action.internalValue = value.action;
            this._conditions.internalValue = value.conditions;
        }
    }
    get conditionLogic() {
        return this.getStringAttribute('condition_logic');
    }
    set conditionLogic(value) {
        this._conditionLogic = value;
    }
    resetConditionLogic() {
        this._conditionLogic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conditionLogicInput() {
        return this._conditionLogic;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get ruleNumber() {
        return this.getNumberAttribute('rule_number');
    }
    set ruleNumber(value) {
        this._ruleNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleNumberInput() {
        return this._ruleNumber;
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get conditions() {
        return this._conditions;
    }
    putConditions(value) {
        this._conditions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get conditionsInput() {
        return this._conditions.internalValue;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference[_d] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList;
_e = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList[_e] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        asn: cdktf.stringToTerraform(struct.asn),
        inside_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.insideCidrBlocks),
        location: cdktf.stringToTerraform(struct.location),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        asn: {
            value: cdktf.stringToHclTerraform(struct.asn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        inside_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.insideCidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._asn !== undefined) {
            hasAnyValues = true;
            internalValueResult.asn = this._asn;
        }
        if (this._insideCidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.insideCidrBlocks = this._insideCidrBlocks;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._asn = undefined;
            this._insideCidrBlocks = undefined;
            this._location = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._asn = value.asn;
            this._insideCidrBlocks = value.insideCidrBlocks;
            this._location = value.location;
        }
    }
    get asn() {
        return this.getStringAttribute('asn');
    }
    set asn(value) {
        this._asn = value;
    }
    resetAsn() {
        this._asn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get asnInput() {
        return this._asn;
    }
    get insideCidrBlocks() {
        return this.getListAttribute('inside_cidr_blocks');
    }
    set insideCidrBlocks(value) {
        this._insideCidrBlocks = value;
    }
    resetInsideCidrBlocks() {
        this._insideCidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get insideCidrBlocksInput() {
        return this._insideCidrBlocks;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference[_f] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList;
_g = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList[_g] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        asn_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.asnRanges),
        inside_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.insideCidrBlocks),
        vpn_ecmp_support: cdktf.booleanToTerraform(struct.vpnEcmpSupport),
        edge_locations: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToTerraform, true)(struct.edgeLocations),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        asn_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.asnRanges),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        inside_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.insideCidrBlocks),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        vpn_ecmp_support: {
            value: cdktf.booleanToHclTerraform(struct.vpnEcmpSupport),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        edge_locations: {
            value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsToHclTerraform, true)(struct.edgeLocations),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // edge_locations - computed: false, optional: false, required: true
        this._edgeLocations = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationEdgeLocationsList(this, "edge_locations", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._asnRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.asnRanges = this._asnRanges;
        }
        if (this._insideCidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.insideCidrBlocks = this._insideCidrBlocks;
        }
        if (this._vpnEcmpSupport !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpnEcmpSupport = this._vpnEcmpSupport;
        }
        if (this._edgeLocations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.edgeLocations = this._edgeLocations?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._asnRanges = undefined;
            this._insideCidrBlocks = undefined;
            this._vpnEcmpSupport = undefined;
            this._edgeLocations.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._asnRanges = value.asnRanges;
            this._insideCidrBlocks = value.insideCidrBlocks;
            this._vpnEcmpSupport = value.vpnEcmpSupport;
            this._edgeLocations.internalValue = value.edgeLocations;
        }
    }
    get asnRanges() {
        return cdktf.Fn.tolist(this.getListAttribute('asn_ranges'));
    }
    set asnRanges(value) {
        this._asnRanges = value;
    }
    // Temporarily expose input value. Use with caution.
    get asnRangesInput() {
        return this._asnRanges;
    }
    get insideCidrBlocks() {
        return cdktf.Fn.tolist(this.getListAttribute('inside_cidr_blocks'));
    }
    set insideCidrBlocks(value) {
        this._insideCidrBlocks = value;
    }
    resetInsideCidrBlocks() {
        this._insideCidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get insideCidrBlocksInput() {
        return this._insideCidrBlocks;
    }
    get vpnEcmpSupport() {
        return this.getBooleanAttribute('vpn_ecmp_support');
    }
    set vpnEcmpSupport(value) {
        this._vpnEcmpSupport = value;
    }
    resetVpnEcmpSupport() {
        this._vpnEcmpSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpnEcmpSupportInput() {
        return this._vpnEcmpSupport;
    }
    get edgeLocations() {
        return this._edgeLocations;
    }
    putEdgeLocations(value) {
        this._edgeLocations.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get edgeLocationsInput() {
        return this._edgeLocations.internalValue;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference[_h] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList;
_j = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList[_j] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        description: cdktf.stringToTerraform(struct.description),
        destination_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.destinationCidrBlocks),
        destinations: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.destinations),
        mode: cdktf.stringToTerraform(struct.mode),
        segment: cdktf.stringToTerraform(struct.segment),
        share_with: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.shareWith),
        share_with_except: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.shareWithExcept),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        destination_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.destinationCidrBlocks),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        destinations: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.destinations),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        segment: {
            value: cdktf.stringToHclTerraform(struct.segment),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        share_with: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.shareWith),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        share_with_except: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.shareWithExcept),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._destinationCidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationCidrBlocks = this._destinationCidrBlocks;
        }
        if (this._destinations !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinations = this._destinations;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._segment !== undefined) {
            hasAnyValues = true;
            internalValueResult.segment = this._segment;
        }
        if (this._shareWith !== undefined) {
            hasAnyValues = true;
            internalValueResult.shareWith = this._shareWith;
        }
        if (this._shareWithExcept !== undefined) {
            hasAnyValues = true;
            internalValueResult.shareWithExcept = this._shareWithExcept;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._description = undefined;
            this._destinationCidrBlocks = undefined;
            this._destinations = undefined;
            this._mode = undefined;
            this._segment = undefined;
            this._shareWith = undefined;
            this._shareWithExcept = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._description = value.description;
            this._destinationCidrBlocks = value.destinationCidrBlocks;
            this._destinations = value.destinations;
            this._mode = value.mode;
            this._segment = value.segment;
            this._shareWith = value.shareWith;
            this._shareWithExcept = value.shareWithExcept;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get destinationCidrBlocks() {
        return cdktf.Fn.tolist(this.getListAttribute('destination_cidr_blocks'));
    }
    set destinationCidrBlocks(value) {
        this._destinationCidrBlocks = value;
    }
    resetDestinationCidrBlocks() {
        this._destinationCidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationCidrBlocksInput() {
        return this._destinationCidrBlocks;
    }
    get destinations() {
        return cdktf.Fn.tolist(this.getListAttribute('destinations'));
    }
    set destinations(value) {
        this._destinations = value;
    }
    resetDestinations() {
        this._destinations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationsInput() {
        return this._destinations;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get segment() {
        return this.getStringAttribute('segment');
    }
    set segment(value) {
        this._segment = value;
    }
    // Temporarily expose input value. Use with caution.
    get segmentInput() {
        return this._segment;
    }
    get shareWith() {
        return cdktf.Fn.tolist(this.getListAttribute('share_with'));
    }
    set shareWith(value) {
        this._shareWith = value;
    }
    resetShareWith() {
        this._shareWith = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shareWithInput() {
        return this._shareWith;
    }
    get shareWithExcept() {
        return cdktf.Fn.tolist(this.getListAttribute('share_with_except'));
    }
    set shareWithExcept(value) {
        this._shareWithExcept = value;
    }
    resetShareWithExcept() {
        this._shareWithExcept = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shareWithExceptInput() {
        return this._shareWithExcept;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference[_k] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList;
_l = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList[_l] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList", version: "19.15.0" };
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allow_filter: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.allowFilter),
        deny_filter: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.denyFilter),
        description: cdktf.stringToTerraform(struct.description),
        edge_locations: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.edgeLocations),
        isolate_attachments: cdktf.booleanToTerraform(struct.isolateAttachments),
        name: cdktf.stringToTerraform(struct.name),
        require_attachment_acceptance: cdktf.booleanToTerraform(struct.requireAttachmentAcceptance),
    };
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToTerraform;
function dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        allow_filter: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.allowFilter),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        deny_filter: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.denyFilter),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        edge_locations: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.edgeLocations),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        isolate_attachments: {
            value: cdktf.booleanToHclTerraform(struct.isolateAttachments),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        require_attachment_acceptance: {
            value: cdktf.booleanToHclTerraform(struct.requireAttachmentAcceptance),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToHclTerraform = dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToHclTerraform;
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowFilter !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowFilter = this._allowFilter;
        }
        if (this._denyFilter !== undefined) {
            hasAnyValues = true;
            internalValueResult.denyFilter = this._denyFilter;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._edgeLocations !== undefined) {
            hasAnyValues = true;
            internalValueResult.edgeLocations = this._edgeLocations;
        }
        if (this._isolateAttachments !== undefined) {
            hasAnyValues = true;
            internalValueResult.isolateAttachments = this._isolateAttachments;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._requireAttachmentAcceptance !== undefined) {
            hasAnyValues = true;
            internalValueResult.requireAttachmentAcceptance = this._requireAttachmentAcceptance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._allowFilter = undefined;
            this._denyFilter = undefined;
            this._description = undefined;
            this._edgeLocations = undefined;
            this._isolateAttachments = undefined;
            this._name = undefined;
            this._requireAttachmentAcceptance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._allowFilter = value.allowFilter;
            this._denyFilter = value.denyFilter;
            this._description = value.description;
            this._edgeLocations = value.edgeLocations;
            this._isolateAttachments = value.isolateAttachments;
            this._name = value.name;
            this._requireAttachmentAcceptance = value.requireAttachmentAcceptance;
        }
    }
    get allowFilter() {
        return cdktf.Fn.tolist(this.getListAttribute('allow_filter'));
    }
    set allowFilter(value) {
        this._allowFilter = value;
    }
    resetAllowFilter() {
        this._allowFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowFilterInput() {
        return this._allowFilter;
    }
    get denyFilter() {
        return cdktf.Fn.tolist(this.getListAttribute('deny_filter'));
    }
    set denyFilter(value) {
        this._denyFilter = value;
    }
    resetDenyFilter() {
        this._denyFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get denyFilterInput() {
        return this._denyFilter;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get edgeLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('edge_locations'));
    }
    set edgeLocations(value) {
        this._edgeLocations = value;
    }
    resetEdgeLocations() {
        this._edgeLocations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get edgeLocationsInput() {
        return this._edgeLocations;
    }
    get isolateAttachments() {
        return this.getBooleanAttribute('isolate_attachments');
    }
    set isolateAttachments(value) {
        this._isolateAttachments = value;
    }
    resetIsolateAttachments() {
        this._isolateAttachments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isolateAttachmentsInput() {
        return this._isolateAttachments;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get requireAttachmentAcceptance() {
        return this.getBooleanAttribute('require_attachment_acceptance');
    }
    set requireAttachmentAcceptance(value) {
        this._requireAttachmentAcceptance = value;
    }
    resetRequireAttachmentAcceptance() {
        this._requireAttachmentAcceptance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requireAttachmentAcceptanceInput() {
        return this._requireAttachmentAcceptance;
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference = DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference[_m] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference", version: "19.15.0" };
class DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList = DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList;
_o = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList[_o] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/networkmanager_core_network_policy_document aws_networkmanager_core_network_policy_document}
*/
class DataAwsNetworkmanagerCoreNetworkPolicyDocument extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsNetworkmanagerCoreNetworkPolicyDocument resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsNetworkmanagerCoreNetworkPolicyDocument to import
    * @param importFromId The id of the existing DataAwsNetworkmanagerCoreNetworkPolicyDocument that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/networkmanager_core_network_policy_document#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsNetworkmanagerCoreNetworkPolicyDocument to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_networkmanager_core_network_policy_document", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/networkmanager_core_network_policy_document aws_networkmanager_core_network_policy_document} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsNetworkmanagerCoreNetworkPolicyDocumentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkmanager_core_network_policy_document',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attachment_policies - computed: false, optional: true, required: false
        this._attachmentPolicies = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList(this, "attachment_policies", false);
        // core_network_configuration - computed: false, optional: false, required: true
        this._coreNetworkConfiguration = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList(this, "core_network_configuration", false);
        // segment_actions - computed: false, optional: true, required: false
        this._segmentActions = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList(this, "segment_actions", false);
        // segments - computed: false, optional: false, required: true
        this._segments = new DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList(this, "segments", false);
        this._id = config.id;
        this._version = config.version;
        this._attachmentPolicies.internalValue = config.attachmentPolicies;
        this._coreNetworkConfiguration.internalValue = config.coreNetworkConfiguration;
        this._segmentActions.internalValue = config.segmentActions;
        this._segments.internalValue = config.segments;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // json - computed: true, optional: false, required: false
    get json() {
        return this.getStringAttribute('json');
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get attachmentPolicies() {
        return this._attachmentPolicies;
    }
    putAttachmentPolicies(value) {
        this._attachmentPolicies.internalValue = value;
    }
    resetAttachmentPolicies() {
        this._attachmentPolicies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attachmentPoliciesInput() {
        return this._attachmentPolicies.internalValue;
    }
    get coreNetworkConfiguration() {
        return this._coreNetworkConfiguration;
    }
    putCoreNetworkConfiguration(value) {
        this._coreNetworkConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get coreNetworkConfigurationInput() {
        return this._coreNetworkConfiguration.internalValue;
    }
    get segmentActions() {
        return this._segmentActions;
    }
    putSegmentActions(value) {
        this._segmentActions.internalValue = value;
    }
    resetSegmentActions() {
        this._segmentActions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentActionsInput() {
        return this._segmentActions.internalValue;
    }
    get segments() {
        return this._segments;
    }
    putSegments(value) {
        this._segments.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get segmentsInput() {
        return this._segments.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            version: cdktf.stringToTerraform(this._version),
            attachment_policies: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToTerraform, true)(this._attachmentPolicies.internalValue),
            core_network_configuration: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToTerraform, true)(this._coreNetworkConfiguration.internalValue),
            segment_actions: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToTerraform, true)(this._segmentActions.internalValue),
            segments: cdktf.listMapper(dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToTerraform, true)(this._segments.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version: {
                value: cdktf.stringToHclTerraform(this._version),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            attachment_policies: {
                value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesToHclTerraform, true)(this._attachmentPolicies.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentAttachmentPoliciesList",
            },
            core_network_configuration: {
                value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationToHclTerraform, true)(this._coreNetworkConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentCoreNetworkConfigurationList",
            },
            segment_actions: {
                value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsToHclTerraform, true)(this._segmentActions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentActionsList",
            },
            segments: {
                value: cdktf.listMapperHcl(dataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsToHclTerraform, true)(this._segments.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsNetworkmanagerCoreNetworkPolicyDocumentSegmentsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsNetworkmanagerCoreNetworkPolicyDocument = DataAwsNetworkmanagerCoreNetworkPolicyDocument;
_p = JSII_RTTI_SYMBOL_1;
DataAwsNetworkmanagerCoreNetworkPolicyDocument[_p] = { fqn: "@cdktf/provider-aws.dataAwsNetworkmanagerCoreNetworkPolicyDocument.DataAwsNetworkmanagerCoreNetworkPolicyDocument", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsNetworkmanagerCoreNetworkPolicyDocument.tfResourceType = "aws_networkmanager_core_network_policy_document";
