// @cdktf/provider-aws/lib/apigatewayv2-route/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Route = exports.Apigatewayv2RouteRequestParameterList = exports.Apigatewayv2RouteRequestParameterOutputReference = exports.apigatewayv2RouteRequestParameterToHclTerraform = exports.apigatewayv2RouteRequestParameterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2RouteRequestParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        request_parameter_key: cdktf.stringToTerraform(struct.requestParameterKey),
        required: cdktf.booleanToTerraform(struct.required),
    };
}
exports.apigatewayv2RouteRequestParameterToTerraform = apigatewayv2RouteRequestParameterToTerraform;
function apigatewayv2RouteRequestParameterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        request_parameter_key: {
            value: cdktf.stringToHclTerraform(struct.requestParameterKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        required: {
            value: cdktf.booleanToHclTerraform(struct.required),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apigatewayv2RouteRequestParameterToHclTerraform = apigatewayv2RouteRequestParameterToHclTerraform;
class Apigatewayv2RouteRequestParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._requestParameterKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestParameterKey = this._requestParameterKey;
        }
        if (this._required !== undefined) {
            hasAnyValues = true;
            internalValueResult.required = this._required;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._requestParameterKey = undefined;
            this._required = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._requestParameterKey = value.requestParameterKey;
            this._required = value.required;
        }
    }
    get requestParameterKey() {
        return this.getStringAttribute('request_parameter_key');
    }
    set requestParameterKey(value) {
        this._requestParameterKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get requestParameterKeyInput() {
        return this._requestParameterKey;
    }
    get required() {
        return this.getBooleanAttribute('required');
    }
    set required(value) {
        this._required = value;
    }
    // Temporarily expose input value. Use with caution.
    get requiredInput() {
        return this._required;
    }
}
exports.Apigatewayv2RouteRequestParameterOutputReference = Apigatewayv2RouteRequestParameterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2RouteRequestParameterOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2Route.Apigatewayv2RouteRequestParameterOutputReference", version: "19.15.0" };
class Apigatewayv2RouteRequestParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Apigatewayv2RouteRequestParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Apigatewayv2RouteRequestParameterList = Apigatewayv2RouteRequestParameterList;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2RouteRequestParameterList[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2Route.Apigatewayv2RouteRequestParameterList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apigatewayv2_route aws_apigatewayv2_route}
*/
class Apigatewayv2Route extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Apigatewayv2Route resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Apigatewayv2Route to import
    * @param importFromId The id of the existing Apigatewayv2Route that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apigatewayv2_route#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Apigatewayv2Route to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_apigatewayv2_route", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apigatewayv2_route aws_apigatewayv2_route} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Apigatewayv2RouteConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_route',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // request_parameter - computed: false, optional: true, required: false
        this._requestParameter = new Apigatewayv2RouteRequestParameterList(this, "request_parameter", true);
        this._apiId = config.apiId;
        this._apiKeyRequired = config.apiKeyRequired;
        this._authorizationScopes = config.authorizationScopes;
        this._authorizationType = config.authorizationType;
        this._authorizerId = config.authorizerId;
        this._id = config.id;
        this._modelSelectionExpression = config.modelSelectionExpression;
        this._operationName = config.operationName;
        this._requestModels = config.requestModels;
        this._routeKey = config.routeKey;
        this._routeResponseSelectionExpression = config.routeResponseSelectionExpression;
        this._target = config.target;
        this._requestParameter.internalValue = config.requestParameter;
    }
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get apiIdInput() {
        return this._apiId;
    }
    get apiKeyRequired() {
        return this.getBooleanAttribute('api_key_required');
    }
    set apiKeyRequired(value) {
        this._apiKeyRequired = value;
    }
    resetApiKeyRequired() {
        this._apiKeyRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiKeyRequiredInput() {
        return this._apiKeyRequired;
    }
    get authorizationScopes() {
        return cdktf.Fn.tolist(this.getListAttribute('authorization_scopes'));
    }
    set authorizationScopes(value) {
        this._authorizationScopes = value;
    }
    resetAuthorizationScopes() {
        this._authorizationScopes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationScopesInput() {
        return this._authorizationScopes;
    }
    get authorizationType() {
        return this.getStringAttribute('authorization_type');
    }
    set authorizationType(value) {
        this._authorizationType = value;
    }
    resetAuthorizationType() {
        this._authorizationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationTypeInput() {
        return this._authorizationType;
    }
    get authorizerId() {
        return this.getStringAttribute('authorizer_id');
    }
    set authorizerId(value) {
        this._authorizerId = value;
    }
    resetAuthorizerId() {
        this._authorizerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerIdInput() {
        return this._authorizerId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get modelSelectionExpression() {
        return this.getStringAttribute('model_selection_expression');
    }
    set modelSelectionExpression(value) {
        this._modelSelectionExpression = value;
    }
    resetModelSelectionExpression() {
        this._modelSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelSelectionExpressionInput() {
        return this._modelSelectionExpression;
    }
    get operationName() {
        return this.getStringAttribute('operation_name');
    }
    set operationName(value) {
        this._operationName = value;
    }
    resetOperationName() {
        this._operationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operationNameInput() {
        return this._operationName;
    }
    get requestModels() {
        return this.getStringMapAttribute('request_models');
    }
    set requestModels(value) {
        this._requestModels = value;
    }
    resetRequestModels() {
        this._requestModels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestModelsInput() {
        return this._requestModels;
    }
    get routeKey() {
        return this.getStringAttribute('route_key');
    }
    set routeKey(value) {
        this._routeKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get routeKeyInput() {
        return this._routeKey;
    }
    get routeResponseSelectionExpression() {
        return this.getStringAttribute('route_response_selection_expression');
    }
    set routeResponseSelectionExpression(value) {
        this._routeResponseSelectionExpression = value;
    }
    resetRouteResponseSelectionExpression() {
        this._routeResponseSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routeResponseSelectionExpressionInput() {
        return this._routeResponseSelectionExpression;
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get requestParameter() {
        return this._requestParameter;
    }
    putRequestParameter(value) {
        this._requestParameter.internalValue = value;
    }
    resetRequestParameter() {
        this._requestParameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestParameterInput() {
        return this._requestParameter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            api_key_required: cdktf.booleanToTerraform(this._apiKeyRequired),
            authorization_scopes: cdktf.listMapper(cdktf.stringToTerraform, false)(this._authorizationScopes),
            authorization_type: cdktf.stringToTerraform(this._authorizationType),
            authorizer_id: cdktf.stringToTerraform(this._authorizerId),
            id: cdktf.stringToTerraform(this._id),
            model_selection_expression: cdktf.stringToTerraform(this._modelSelectionExpression),
            operation_name: cdktf.stringToTerraform(this._operationName),
            request_models: cdktf.hashMapper(cdktf.stringToTerraform)(this._requestModels),
            route_key: cdktf.stringToTerraform(this._routeKey),
            route_response_selection_expression: cdktf.stringToTerraform(this._routeResponseSelectionExpression),
            target: cdktf.stringToTerraform(this._target),
            request_parameter: cdktf.listMapper(apigatewayv2RouteRequestParameterToTerraform, true)(this._requestParameter.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            api_id: {
                value: cdktf.stringToHclTerraform(this._apiId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            api_key_required: {
                value: cdktf.booleanToHclTerraform(this._apiKeyRequired),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            authorization_scopes: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._authorizationScopes),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            authorization_type: {
                value: cdktf.stringToHclTerraform(this._authorizationType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authorizer_id: {
                value: cdktf.stringToHclTerraform(this._authorizerId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            model_selection_expression: {
                value: cdktf.stringToHclTerraform(this._modelSelectionExpression),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            operation_name: {
                value: cdktf.stringToHclTerraform(this._operationName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            request_models: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._requestModels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            route_key: {
                value: cdktf.stringToHclTerraform(this._routeKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            route_response_selection_expression: {
                value: cdktf.stringToHclTerraform(this._routeResponseSelectionExpression),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target: {
                value: cdktf.stringToHclTerraform(this._target),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            request_parameter: {
                value: cdktf.listMapperHcl(apigatewayv2RouteRequestParameterToHclTerraform, true)(this._requestParameter.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "Apigatewayv2RouteRequestParameterList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Apigatewayv2Route = Apigatewayv2Route;
_c = JSII_RTTI_SYMBOL_1;
Apigatewayv2Route[_c] = { fqn: "@cdktf/provider-aws.apigatewayv2Route.Apigatewayv2Route", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Apigatewayv2Route.tfResourceType = "aws_apigatewayv2_route";
