// @cdktf/provider-aws/lib/ssm-patch-group/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmPatchGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_patch_group aws_ssm_patch_group}
*/
class SsmPatchGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsmPatchGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsmPatchGroup to import
    * @param importFromId The id of the existing SsmPatchGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_patch_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsmPatchGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssm_patch_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_patch_group aws_ssm_patch_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsmPatchGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_patch_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._baselineId = config.baselineId;
        this._id = config.id;
        this._patchGroup = config.patchGroup;
    }
    get baselineId() {
        return this.getStringAttribute('baseline_id');
    }
    set baselineId(value) {
        this._baselineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get baselineIdInput() {
        return this._baselineId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get patchGroup() {
        return this.getStringAttribute('patch_group');
    }
    set patchGroup(value) {
        this._patchGroup = value;
    }
    // Temporarily expose input value. Use with caution.
    get patchGroupInput() {
        return this._patchGroup;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            baseline_id: cdktf.stringToTerraform(this._baselineId),
            id: cdktf.stringToTerraform(this._id),
            patch_group: cdktf.stringToTerraform(this._patchGroup),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            baseline_id: {
                value: cdktf.stringToHclTerraform(this._baselineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            patch_group: {
                value: cdktf.stringToHclTerraform(this._patchGroup),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsmPatchGroup = SsmPatchGroup;
_a = JSII_RTTI_SYMBOL_1;
SsmPatchGroup[_a] = { fqn: "@cdktf/provider-aws.ssmPatchGroup.SsmPatchGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsmPatchGroup.tfResourceType = "aws_ssm_patch_group";
