// @cdktf/provider-aws/lib/data-aws-serverlessapplicationrepository-application/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServerlessapplicationrepositoryApplication = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/serverlessapplicationrepository_application aws_serverlessapplicationrepository_application}
*/
class DataAwsServerlessapplicationrepositoryApplication extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsServerlessapplicationrepositoryApplication resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsServerlessapplicationrepositoryApplication to import
    * @param importFromId The id of the existing DataAwsServerlessapplicationrepositoryApplication that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/serverlessapplicationrepository_application#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsServerlessapplicationrepositoryApplication to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_serverlessapplicationrepository_application", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/serverlessapplicationrepository_application aws_serverlessapplicationrepository_application} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsServerlessapplicationrepositoryApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_serverlessapplicationrepository_application',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._applicationId = config.applicationId;
        this._id = config.id;
        this._semanticVersion = config.semanticVersion;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // required_capabilities - computed: true, optional: false, required: false
    get requiredCapabilities() {
        return cdktf.Fn.tolist(this.getListAttribute('required_capabilities'));
    }
    get semanticVersion() {
        return this.getStringAttribute('semantic_version');
    }
    set semanticVersion(value) {
        this._semanticVersion = value;
    }
    resetSemanticVersion() {
        this._semanticVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get semanticVersionInput() {
        return this._semanticVersion;
    }
    // source_code_url - computed: true, optional: false, required: false
    get sourceCodeUrl() {
        return this.getStringAttribute('source_code_url');
    }
    // template_url - computed: true, optional: false, required: false
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            id: cdktf.stringToTerraform(this._id),
            semantic_version: cdktf.stringToTerraform(this._semanticVersion),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_id: {
                value: cdktf.stringToHclTerraform(this._applicationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            semantic_version: {
                value: cdktf.stringToHclTerraform(this._semanticVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsServerlessapplicationrepositoryApplication = DataAwsServerlessapplicationrepositoryApplication;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServerlessapplicationrepositoryApplication[_a] = { fqn: "@cdktf/provider-aws.dataAwsServerlessapplicationrepositoryApplication.DataAwsServerlessapplicationrepositoryApplication", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsServerlessapplicationrepositoryApplication.tfResourceType = "aws_serverlessapplicationrepository_application";
