// @cdktf/provider-aws/lib/data-aws-servicequotas-service-quota/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServicequotasServiceQuota = exports.DataAwsServicequotasServiceQuotaUsageMetricList = exports.DataAwsServicequotasServiceQuotaUsageMetricOutputReference = exports.dataAwsServicequotasServiceQuotaUsageMetricToHclTerraform = exports.dataAwsServicequotasServiceQuotaUsageMetricToTerraform = exports.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList = exports.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference = exports.dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform = exports.dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform = dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform;
function dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform = dataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform;
class DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // class - computed: true, optional: false, required: false
    get class() {
        return this.getStringAttribute('class');
    }
    // resource - computed: true, optional: false, required: false
    get resource() {
        return this.getStringAttribute('resource');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference = DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsServicequotasServiceQuota.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference", version: "19.15.0" };
class DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList = DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsServicequotasServiceQuota.DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList", version: "19.15.0" };
function dataAwsServicequotasServiceQuotaUsageMetricToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServicequotasServiceQuotaUsageMetricToTerraform = dataAwsServicequotasServiceQuotaUsageMetricToTerraform;
function dataAwsServicequotasServiceQuotaUsageMetricToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsServicequotasServiceQuotaUsageMetricToHclTerraform = dataAwsServicequotasServiceQuotaUsageMetricToHclTerraform;
class DataAwsServicequotasServiceQuotaUsageMetricOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metric_dimensions - computed: true, optional: false, required: false
        this._metricDimensions = new DataAwsServicequotasServiceQuotaUsageMetricMetricDimensionsList(this, "metric_dimensions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get metricDimensions() {
        return this._metricDimensions;
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // metric_namespace - computed: true, optional: false, required: false
    get metricNamespace() {
        return this.getStringAttribute('metric_namespace');
    }
    // metric_statistic_recommendation - computed: true, optional: false, required: false
    get metricStatisticRecommendation() {
        return this.getStringAttribute('metric_statistic_recommendation');
    }
}
exports.DataAwsServicequotasServiceQuotaUsageMetricOutputReference = DataAwsServicequotasServiceQuotaUsageMetricOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsServicequotasServiceQuotaUsageMetricOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsServicequotasServiceQuota.DataAwsServicequotasServiceQuotaUsageMetricOutputReference", version: "19.15.0" };
class DataAwsServicequotasServiceQuotaUsageMetricList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServicequotasServiceQuotaUsageMetricOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServicequotasServiceQuotaUsageMetricList = DataAwsServicequotasServiceQuotaUsageMetricList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsServicequotasServiceQuotaUsageMetricList[_d] = { fqn: "@cdktf/provider-aws.dataAwsServicequotasServiceQuota.DataAwsServicequotasServiceQuotaUsageMetricList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/servicequotas_service_quota aws_servicequotas_service_quota}
*/
class DataAwsServicequotasServiceQuota extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsServicequotasServiceQuota resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsServicequotasServiceQuota to import
    * @param importFromId The id of the existing DataAwsServicequotasServiceQuota that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/servicequotas_service_quota#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsServicequotasServiceQuota to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_servicequotas_service_quota", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/servicequotas_service_quota aws_servicequotas_service_quota} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsServicequotasServiceQuotaConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicequotas_service_quota',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // usage_metric - computed: true, optional: false, required: false
        this._usageMetric = new DataAwsServicequotasServiceQuotaUsageMetricList(this, "usage_metric", false);
        this._id = config.id;
        this._quotaCode = config.quotaCode;
        this._quotaName = config.quotaName;
        this._serviceCode = config.serviceCode;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // adjustable - computed: true, optional: false, required: false
    get adjustable() {
        return this.getBooleanAttribute('adjustable');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_value - computed: true, optional: false, required: false
    get defaultValue() {
        return this.getNumberAttribute('default_value');
    }
    // global_quota - computed: true, optional: false, required: false
    get globalQuota() {
        return this.getBooleanAttribute('global_quota');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get quotaCode() {
        return this.getStringAttribute('quota_code');
    }
    set quotaCode(value) {
        this._quotaCode = value;
    }
    resetQuotaCode() {
        this._quotaCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get quotaCodeInput() {
        return this._quotaCode;
    }
    get quotaName() {
        return this.getStringAttribute('quota_name');
    }
    set quotaName(value) {
        this._quotaName = value;
    }
    resetQuotaName() {
        this._quotaName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get quotaNameInput() {
        return this._quotaName;
    }
    get serviceCode() {
        return this.getStringAttribute('service_code');
    }
    set serviceCode(value) {
        this._serviceCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceCodeInput() {
        return this._serviceCode;
    }
    // service_name - computed: true, optional: false, required: false
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    get usageMetric() {
        return this._usageMetric;
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getNumberAttribute('value');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            quota_code: cdktf.stringToTerraform(this._quotaCode),
            quota_name: cdktf.stringToTerraform(this._quotaName),
            service_code: cdktf.stringToTerraform(this._serviceCode),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_code: {
                value: cdktf.stringToHclTerraform(this._quotaCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_name: {
                value: cdktf.stringToHclTerraform(this._quotaName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_code: {
                value: cdktf.stringToHclTerraform(this._serviceCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsServicequotasServiceQuota = DataAwsServicequotasServiceQuota;
_e = JSII_RTTI_SYMBOL_1;
DataAwsServicequotasServiceQuota[_e] = { fqn: "@cdktf/provider-aws.dataAwsServicequotasServiceQuota.DataAwsServicequotasServiceQuota", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsServicequotasServiceQuota.tfResourceType = "aws_servicequotas_service_quota";
