// @cdktf/provider-aws/lib/autoscaling-notification/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoscalingNotification = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/autoscaling_notification aws_autoscaling_notification}
*/
class AutoscalingNotification extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AutoscalingNotification resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AutoscalingNotification to import
    * @param importFromId The id of the existing AutoscalingNotification that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/autoscaling_notification#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AutoscalingNotification to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_autoscaling_notification", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/autoscaling_notification aws_autoscaling_notification} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AutoscalingNotificationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_autoscaling_notification',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._groupNames = config.groupNames;
        this._id = config.id;
        this._notifications = config.notifications;
        this._topicArn = config.topicArn;
    }
    get groupNames() {
        return cdktf.Fn.tolist(this.getListAttribute('group_names'));
    }
    set groupNames(value) {
        this._groupNames = value;
    }
    // Temporarily expose input value. Use with caution.
    get groupNamesInput() {
        return this._groupNames;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get notifications() {
        return cdktf.Fn.tolist(this.getListAttribute('notifications'));
    }
    set notifications(value) {
        this._notifications = value;
    }
    // Temporarily expose input value. Use with caution.
    get notificationsInput() {
        return this._notifications;
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            group_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._groupNames),
            id: cdktf.stringToTerraform(this._id),
            notifications: cdktf.listMapper(cdktf.stringToTerraform, false)(this._notifications),
            topic_arn: cdktf.stringToTerraform(this._topicArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            group_names: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._groupNames),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notifications: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._notifications),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            topic_arn: {
                value: cdktf.stringToHclTerraform(this._topicArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AutoscalingNotification = AutoscalingNotification;
_a = JSII_RTTI_SYMBOL_1;
AutoscalingNotification[_a] = { fqn: "@cdktf/provider-aws.autoscalingNotification.AutoscalingNotification", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AutoscalingNotification.tfResourceType = "aws_autoscaling_notification";
