// @cdktf/provider-aws/lib/ecr-pull-through-cache-rule/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcrPullThroughCacheRule = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_pull_through_cache_rule aws_ecr_pull_through_cache_rule}
*/
class EcrPullThroughCacheRule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EcrPullThroughCacheRule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EcrPullThroughCacheRule to import
    * @param importFromId The id of the existing EcrPullThroughCacheRule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_pull_through_cache_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EcrPullThroughCacheRule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecr_pull_through_cache_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_pull_through_cache_rule aws_ecr_pull_through_cache_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EcrPullThroughCacheRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_pull_through_cache_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._credentialArn = config.credentialArn;
        this._ecrRepositoryPrefix = config.ecrRepositoryPrefix;
        this._id = config.id;
        this._upstreamRegistryUrl = config.upstreamRegistryUrl;
    }
    get credentialArn() {
        return this.getStringAttribute('credential_arn');
    }
    set credentialArn(value) {
        this._credentialArn = value;
    }
    resetCredentialArn() {
        this._credentialArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get credentialArnInput() {
        return this._credentialArn;
    }
    get ecrRepositoryPrefix() {
        return this.getStringAttribute('ecr_repository_prefix');
    }
    set ecrRepositoryPrefix(value) {
        this._ecrRepositoryPrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get ecrRepositoryPrefixInput() {
        return this._ecrRepositoryPrefix;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // registry_id - computed: true, optional: false, required: false
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    get upstreamRegistryUrl() {
        return this.getStringAttribute('upstream_registry_url');
    }
    set upstreamRegistryUrl(value) {
        this._upstreamRegistryUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get upstreamRegistryUrlInput() {
        return this._upstreamRegistryUrl;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            credential_arn: cdktf.stringToTerraform(this._credentialArn),
            ecr_repository_prefix: cdktf.stringToTerraform(this._ecrRepositoryPrefix),
            id: cdktf.stringToTerraform(this._id),
            upstream_registry_url: cdktf.stringToTerraform(this._upstreamRegistryUrl),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            credential_arn: {
                value: cdktf.stringToHclTerraform(this._credentialArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ecr_repository_prefix: {
                value: cdktf.stringToHclTerraform(this._ecrRepositoryPrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            upstream_registry_url: {
                value: cdktf.stringToHclTerraform(this._upstreamRegistryUrl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EcrPullThroughCacheRule = EcrPullThroughCacheRule;
_a = JSII_RTTI_SYMBOL_1;
EcrPullThroughCacheRule[_a] = { fqn: "@cdktf/provider-aws.ecrPullThroughCacheRule.EcrPullThroughCacheRule", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EcrPullThroughCacheRule.tfResourceType = "aws_ecr_pull_through_cache_rule";
