// @cdktf/provider-aws/lib/cognito-identity-pool-roles-attachment/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CognitoIdentityPoolRolesAttachment = exports.CognitoIdentityPoolRolesAttachmentRoleMappingList = exports.CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference = exports.cognitoIdentityPoolRolesAttachmentRoleMappingToHclTerraform = exports.cognitoIdentityPoolRolesAttachmentRoleMappingToTerraform = exports.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList = exports.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference = exports.cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToHclTerraform = exports.cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        claim: cdktf.stringToTerraform(struct.claim),
        match_type: cdktf.stringToTerraform(struct.matchType),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToTerraform = cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToTerraform;
function cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        claim: {
            value: cdktf.stringToHclTerraform(struct.claim),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_type: {
            value: cdktf.stringToHclTerraform(struct.matchType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToHclTerraform = cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToHclTerraform;
class CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._claim !== undefined) {
            hasAnyValues = true;
            internalValueResult.claim = this._claim;
        }
        if (this._matchType !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchType = this._matchType;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._claim = undefined;
            this._matchType = undefined;
            this._roleArn = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._claim = value.claim;
            this._matchType = value.matchType;
            this._roleArn = value.roleArn;
            this._value = value.value;
        }
    }
    get claim() {
        return this.getStringAttribute('claim');
    }
    set claim(value) {
        this._claim = value;
    }
    // Temporarily expose input value. Use with caution.
    get claimInput() {
        return this._claim;
    }
    get matchType() {
        return this.getStringAttribute('match_type');
    }
    set matchType(value) {
        this._matchType = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchTypeInput() {
        return this._matchType;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference = CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference[_a] = { fqn: "@cdktf/provider-aws.cognitoIdentityPoolRolesAttachment.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference", version: "19.15.0" };
class CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList = CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList;
_b = JSII_RTTI_SYMBOL_1;
CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList[_b] = { fqn: "@cdktf/provider-aws.cognitoIdentityPoolRolesAttachment.CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList", version: "19.15.0" };
function cognitoIdentityPoolRolesAttachmentRoleMappingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ambiguous_role_resolution: cdktf.stringToTerraform(struct.ambiguousRoleResolution),
        identity_provider: cdktf.stringToTerraform(struct.identityProvider),
        type: cdktf.stringToTerraform(struct.type),
        mapping_rule: cdktf.listMapper(cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToTerraform, true)(struct.mappingRule),
    };
}
exports.cognitoIdentityPoolRolesAttachmentRoleMappingToTerraform = cognitoIdentityPoolRolesAttachmentRoleMappingToTerraform;
function cognitoIdentityPoolRolesAttachmentRoleMappingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ambiguous_role_resolution: {
            value: cdktf.stringToHclTerraform(struct.ambiguousRoleResolution),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        identity_provider: {
            value: cdktf.stringToHclTerraform(struct.identityProvider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mapping_rule: {
            value: cdktf.listMapperHcl(cognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleToHclTerraform, true)(struct.mappingRule),
            isBlock: true,
            type: "list",
            storageClassType: "CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cognitoIdentityPoolRolesAttachmentRoleMappingToHclTerraform = cognitoIdentityPoolRolesAttachmentRoleMappingToHclTerraform;
class CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // mapping_rule - computed: false, optional: true, required: false
        this._mappingRule = new CognitoIdentityPoolRolesAttachmentRoleMappingMappingRuleList(this, "mapping_rule", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ambiguousRoleResolution !== undefined) {
            hasAnyValues = true;
            internalValueResult.ambiguousRoleResolution = this._ambiguousRoleResolution;
        }
        if (this._identityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityProvider = this._identityProvider;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._mappingRule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mappingRule = this._mappingRule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._ambiguousRoleResolution = undefined;
            this._identityProvider = undefined;
            this._type = undefined;
            this._mappingRule.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._ambiguousRoleResolution = value.ambiguousRoleResolution;
            this._identityProvider = value.identityProvider;
            this._type = value.type;
            this._mappingRule.internalValue = value.mappingRule;
        }
    }
    get ambiguousRoleResolution() {
        return this.getStringAttribute('ambiguous_role_resolution');
    }
    set ambiguousRoleResolution(value) {
        this._ambiguousRoleResolution = value;
    }
    resetAmbiguousRoleResolution() {
        this._ambiguousRoleResolution = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ambiguousRoleResolutionInput() {
        return this._ambiguousRoleResolution;
    }
    get identityProvider() {
        return this.getStringAttribute('identity_provider');
    }
    set identityProvider(value) {
        this._identityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityProviderInput() {
        return this._identityProvider;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get mappingRule() {
        return this._mappingRule;
    }
    putMappingRule(value) {
        this._mappingRule.internalValue = value;
    }
    resetMappingRule() {
        this._mappingRule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mappingRuleInput() {
        return this._mappingRule.internalValue;
    }
}
exports.CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference = CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference[_c] = { fqn: "@cdktf/provider-aws.cognitoIdentityPoolRolesAttachment.CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference", version: "19.15.0" };
class CognitoIdentityPoolRolesAttachmentRoleMappingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CognitoIdentityPoolRolesAttachmentRoleMappingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CognitoIdentityPoolRolesAttachmentRoleMappingList = CognitoIdentityPoolRolesAttachmentRoleMappingList;
_d = JSII_RTTI_SYMBOL_1;
CognitoIdentityPoolRolesAttachmentRoleMappingList[_d] = { fqn: "@cdktf/provider-aws.cognitoIdentityPoolRolesAttachment.CognitoIdentityPoolRolesAttachmentRoleMappingList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_pool_roles_attachment aws_cognito_identity_pool_roles_attachment}
*/
class CognitoIdentityPoolRolesAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CognitoIdentityPoolRolesAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CognitoIdentityPoolRolesAttachment to import
    * @param importFromId The id of the existing CognitoIdentityPoolRolesAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_pool_roles_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CognitoIdentityPoolRolesAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cognito_identity_pool_roles_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_pool_roles_attachment aws_cognito_identity_pool_roles_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CognitoIdentityPoolRolesAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cognito_identity_pool_roles_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // role_mapping - computed: false, optional: true, required: false
        this._roleMapping = new CognitoIdentityPoolRolesAttachmentRoleMappingList(this, "role_mapping", true);
        this._id = config.id;
        this._identityPoolId = config.identityPoolId;
        this._roles = config.roles;
        this._roleMapping.internalValue = config.roleMapping;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityPoolId() {
        return this.getStringAttribute('identity_pool_id');
    }
    set identityPoolId(value) {
        this._identityPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityPoolIdInput() {
        return this._identityPoolId;
    }
    get roles() {
        return this.getStringMapAttribute('roles');
    }
    set roles(value) {
        this._roles = value;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles;
    }
    get roleMapping() {
        return this._roleMapping;
    }
    putRoleMapping(value) {
        this._roleMapping.internalValue = value;
    }
    resetRoleMapping() {
        this._roleMapping.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleMappingInput() {
        return this._roleMapping.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            identity_pool_id: cdktf.stringToTerraform(this._identityPoolId),
            roles: cdktf.hashMapper(cdktf.stringToTerraform)(this._roles),
            role_mapping: cdktf.listMapper(cognitoIdentityPoolRolesAttachmentRoleMappingToTerraform, true)(this._roleMapping.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identity_pool_id: {
                value: cdktf.stringToHclTerraform(this._identityPoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            roles: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._roles),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            role_mapping: {
                value: cdktf.listMapperHcl(cognitoIdentityPoolRolesAttachmentRoleMappingToHclTerraform, true)(this._roleMapping.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CognitoIdentityPoolRolesAttachmentRoleMappingList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CognitoIdentityPoolRolesAttachment = CognitoIdentityPoolRolesAttachment;
_e = JSII_RTTI_SYMBOL_1;
CognitoIdentityPoolRolesAttachment[_e] = { fqn: "@cdktf/provider-aws.cognitoIdentityPoolRolesAttachment.CognitoIdentityPoolRolesAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CognitoIdentityPoolRolesAttachment.tfResourceType = "aws_cognito_identity_pool_roles_attachment";
