// @winglang/sdk/lib/shared-gcp/counter.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterClient = void 0;
const datastore_1 = require("@google-cloud/datastore");
const DEFAULT_COUNTER_KEY = "counter";
const COUNTER_ENTITY_KIND = "Counter";
class CounterClient {
    constructor(databaseName, initial = 0) {
        this.databaseName = databaseName;
        this.initial = initial;
        this.client = new datastore_1.Datastore({ databaseId: this.databaseName });
    }
    async inc(amount = 1, key = DEFAULT_COUNTER_KEY) {
        const currentValue = await this._getCounterValue(key);
        const newValue = currentValue + amount;
        await this._updateCounter(key, newValue);
        // Return the previous value before the increment
        return currentValue;
    }
    async dec(amount = 1, key = DEFAULT_COUNTER_KEY) {
        const currentValue = await this._getCounterValue(key);
        const newValue = currentValue - amount;
        await this._updateCounter(key, newValue);
        // Return the previous value before the decrement
        return currentValue;
    }
    async set(value, key = DEFAULT_COUNTER_KEY) {
        await this._updateCounter(key, value);
    }
    async peek(key = DEFAULT_COUNTER_KEY) {
        return this._getCounterValue(key);
    }
    async _getCounterValue(key) {
        const counterKey = this.client.key([COUNTER_ENTITY_KIND, key]);
        // Fetch the counter from the datastore
        const [existingCounter] = await this.client.get(counterKey);
        // If the counter exists, return its current count
        // Else, initialize it with the `initial` value
        if (existingCounter) {
            return existingCounter.count;
        }
        else {
            await this._initCounter(key);
            return this.initial;
        }
    }
    async _initCounter(key) {
        const counterEntity = {
            key: this.client.key([COUNTER_ENTITY_KIND, key]),
            data: { count: this.initial },
        };
        await this.client.insert(counterEntity);
    }
    async _updateCounter(key, newValue) {
        const counterEntity = {
            key: this.client.key([COUNTER_ENTITY_KIND, key]),
            data: { count: newValue },
        };
        await this.client.save(counterEntity);
    }
}
exports.CounterClient = CounterClient;
