// @winglang/sdk/lib/shared-gcp/function.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionClient = void 0;
const google_auth_library_1 = require("google-auth-library");
const http_1 = require("../http");
class FunctionClient {
    constructor(functionName, projectId, region) {
        this.functionName = functionName;
        this.projectId = projectId;
        this.region = region;
    }
    get functionURL() {
        return `https://${this.region}-${this.projectId}.cloudfunctions.net/${this.functionName}`;
    }
    async _invokeLocally(payload, token) {
        try {
            const res = await http_1.Util.post(this.functionURL, {
                headers: {
                    body: payload,
                    "Content-Type": "text/plain",
                    authorization: `bearer ${token}`,
                },
                redirect: http_1.RequestRedirect.FOLLOW,
            });
            if (!res.ok) {
                throw new Error(res.body);
            }
            return res.body ?? "";
        }
        catch (error) {
            throw new Error(`Error while invoking the function ${this.functionName}:\n${error.message}`);
        }
    }
    /**
     * Invoke the function, passing the given payload as an argument.
     *  @returns the function returned payload only
     */
    async invoke(payload) {
        try {
            const auth = new google_auth_library_1.GoogleAuth();
            const client = await auth.getIdTokenClient(this.functionURL);
            const res = await client.request({
                url: this.functionURL,
                method: "POST",
                body: payload,
                headers: {
                    "Content-Type": "text/plain",
                },
                retryConfig: {
                    retry: 8,
                    retryDelay: 15 * 1000,
                    shouldRetry: ({ code }) => {
                        // unauthorized - means that the permissions for the invocation are not yet established
                        return code === "403";
                    },
                },
            });
            return res.data;
        }
        catch (error) {
            throw new Error(`Error while invoking the function ${this.functionName}:\n${error.message}`);
        }
    }
    async invokeAsync(_payload) {
        throw new Error("invokeAsync not implemented");
    }
    /**
     * Invoke the function, passing the given payload as an argument.
     *
     * @returns the function returned payload and logs
     */
    async invokeWithLogs(payload, token) {
        console.error("Test invocation on tf-gcp doesn't include logs yet");
        //TODO: add traces to tf-gcp tests- https://github.com/winglang/wing/issues/4904
        return [await this._invokeLocally(payload, token), []];
    }
}
exports.FunctionClient = FunctionClient;
