// @cdktf/provider-aws/lib/data-aws-redshift-producer-data-shares/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftProducerDataShares = exports.DataAwsRedshiftProducerDataSharesDataSharesList = exports.DataAwsRedshiftProducerDataSharesDataSharesOutputReference = exports.dataAwsRedshiftProducerDataSharesDataSharesToHclTerraform = exports.dataAwsRedshiftProducerDataSharesDataSharesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRedshiftProducerDataSharesDataSharesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsRedshiftProducerDataSharesDataSharesToTerraform = dataAwsRedshiftProducerDataSharesDataSharesToTerraform;
function dataAwsRedshiftProducerDataSharesDataSharesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsRedshiftProducerDataSharesDataSharesToHclTerraform = dataAwsRedshiftProducerDataSharesDataSharesToHclTerraform;
class DataAwsRedshiftProducerDataSharesDataSharesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // data_share_arn - computed: true, optional: false, required: false
    get dataShareArn() {
        return this.getStringAttribute('data_share_arn');
    }
    // managed_by - computed: true, optional: false, required: false
    get managedBy() {
        return this.getStringAttribute('managed_by');
    }
    // producer_arn - computed: true, optional: false, required: false
    get producerArn() {
        return this.getStringAttribute('producer_arn');
    }
}
exports.DataAwsRedshiftProducerDataSharesDataSharesOutputReference = DataAwsRedshiftProducerDataSharesDataSharesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftProducerDataSharesDataSharesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftProducerDataShares.DataAwsRedshiftProducerDataSharesDataSharesOutputReference", version: "19.15.0" };
class DataAwsRedshiftProducerDataSharesDataSharesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRedshiftProducerDataSharesDataSharesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRedshiftProducerDataSharesDataSharesList = DataAwsRedshiftProducerDataSharesDataSharesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftProducerDataSharesDataSharesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftProducerDataShares.DataAwsRedshiftProducerDataSharesDataSharesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_producer_data_shares aws_redshift_producer_data_shares}
*/
class DataAwsRedshiftProducerDataShares extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsRedshiftProducerDataShares resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsRedshiftProducerDataShares to import
    * @param importFromId The id of the existing DataAwsRedshiftProducerDataShares that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_producer_data_shares#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsRedshiftProducerDataShares to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshift_producer_data_shares", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshift_producer_data_shares aws_redshift_producer_data_shares} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRedshiftProducerDataSharesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_producer_data_shares',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // data_shares - computed: false, optional: true, required: false
        this._dataShares = new DataAwsRedshiftProducerDataSharesDataSharesList(this, "data_shares", false);
        this._producerArn = config.producerArn;
        this._status = config.status;
        this._dataShares.internalValue = config.dataShares;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get producerArn() {
        return this.getStringAttribute('producer_arn');
    }
    set producerArn(value) {
        this._producerArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get producerArnInput() {
        return this._producerArn;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get dataShares() {
        return this._dataShares;
    }
    putDataShares(value) {
        this._dataShares.internalValue = value;
    }
    resetDataShares() {
        this._dataShares.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataSharesInput() {
        return this._dataShares.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            producer_arn: cdktf.stringToTerraform(this._producerArn),
            status: cdktf.stringToTerraform(this._status),
            data_shares: cdktf.listMapper(dataAwsRedshiftProducerDataSharesDataSharesToTerraform, true)(this._dataShares.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            producer_arn: {
                value: cdktf.stringToHclTerraform(this._producerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_shares: {
                value: cdktf.listMapperHcl(dataAwsRedshiftProducerDataSharesDataSharesToHclTerraform, true)(this._dataShares.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsRedshiftProducerDataSharesDataSharesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsRedshiftProducerDataShares = DataAwsRedshiftProducerDataShares;
_c = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftProducerDataShares[_c] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftProducerDataShares.DataAwsRedshiftProducerDataShares", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRedshiftProducerDataShares.tfResourceType = "aws_redshift_producer_data_shares";
