// @cdktf/provider-aws/lib/data-aws-medialive-input/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsMedialiveInput = exports.DataAwsMedialiveInputSourcesList = exports.DataAwsMedialiveInputSourcesOutputReference = exports.dataAwsMedialiveInputSourcesToHclTerraform = exports.dataAwsMedialiveInputSourcesToTerraform = exports.DataAwsMedialiveInputMediaConnectFlowsList = exports.DataAwsMedialiveInputMediaConnectFlowsOutputReference = exports.dataAwsMedialiveInputMediaConnectFlowsToHclTerraform = exports.dataAwsMedialiveInputMediaConnectFlowsToTerraform = exports.DataAwsMedialiveInputInputDevicesList = exports.DataAwsMedialiveInputInputDevicesOutputReference = exports.dataAwsMedialiveInputInputDevicesToHclTerraform = exports.dataAwsMedialiveInputInputDevicesToTerraform = exports.DataAwsMedialiveInputDestinationsList = exports.DataAwsMedialiveInputDestinationsOutputReference = exports.dataAwsMedialiveInputDestinationsToHclTerraform = exports.dataAwsMedialiveInputDestinationsToTerraform = exports.DataAwsMedialiveInputDestinationsVpcList = exports.DataAwsMedialiveInputDestinationsVpcOutputReference = exports.dataAwsMedialiveInputDestinationsVpcToHclTerraform = exports.dataAwsMedialiveInputDestinationsVpcToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsMedialiveInputDestinationsVpcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMedialiveInputDestinationsVpcToTerraform = dataAwsMedialiveInputDestinationsVpcToTerraform;
function dataAwsMedialiveInputDestinationsVpcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMedialiveInputDestinationsVpcToHclTerraform = dataAwsMedialiveInputDestinationsVpcToHclTerraform;
class DataAwsMedialiveInputDestinationsVpcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // network_interface_id - computed: true, optional: false, required: false
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
}
exports.DataAwsMedialiveInputDestinationsVpcOutputReference = DataAwsMedialiveInputDestinationsVpcOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputDestinationsVpcOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputDestinationsVpcOutputReference", version: "19.15.0" };
class DataAwsMedialiveInputDestinationsVpcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMedialiveInputDestinationsVpcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMedialiveInputDestinationsVpcList = DataAwsMedialiveInputDestinationsVpcList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputDestinationsVpcList[_b] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputDestinationsVpcList", version: "19.15.0" };
function dataAwsMedialiveInputDestinationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMedialiveInputDestinationsToTerraform = dataAwsMedialiveInputDestinationsToTerraform;
function dataAwsMedialiveInputDestinationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMedialiveInputDestinationsToHclTerraform = dataAwsMedialiveInputDestinationsToHclTerraform;
class DataAwsMedialiveInputDestinationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // vpc - computed: true, optional: false, required: false
        this._vpc = new DataAwsMedialiveInputDestinationsVpcList(this, "vpc", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip - computed: true, optional: false, required: false
    get ip() {
        return this.getStringAttribute('ip');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getStringAttribute('port');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get vpc() {
        return this._vpc;
    }
}
exports.DataAwsMedialiveInputDestinationsOutputReference = DataAwsMedialiveInputDestinationsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputDestinationsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputDestinationsOutputReference", version: "19.15.0" };
class DataAwsMedialiveInputDestinationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMedialiveInputDestinationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMedialiveInputDestinationsList = DataAwsMedialiveInputDestinationsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputDestinationsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputDestinationsList", version: "19.15.0" };
function dataAwsMedialiveInputInputDevicesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMedialiveInputInputDevicesToTerraform = dataAwsMedialiveInputInputDevicesToTerraform;
function dataAwsMedialiveInputInputDevicesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMedialiveInputInputDevicesToHclTerraform = dataAwsMedialiveInputInputDevicesToHclTerraform;
class DataAwsMedialiveInputInputDevicesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
}
exports.DataAwsMedialiveInputInputDevicesOutputReference = DataAwsMedialiveInputInputDevicesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputInputDevicesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputInputDevicesOutputReference", version: "19.15.0" };
class DataAwsMedialiveInputInputDevicesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMedialiveInputInputDevicesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMedialiveInputInputDevicesList = DataAwsMedialiveInputInputDevicesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputInputDevicesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputInputDevicesList", version: "19.15.0" };
function dataAwsMedialiveInputMediaConnectFlowsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMedialiveInputMediaConnectFlowsToTerraform = dataAwsMedialiveInputMediaConnectFlowsToTerraform;
function dataAwsMedialiveInputMediaConnectFlowsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMedialiveInputMediaConnectFlowsToHclTerraform = dataAwsMedialiveInputMediaConnectFlowsToHclTerraform;
class DataAwsMedialiveInputMediaConnectFlowsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // flow_arn - computed: true, optional: false, required: false
    get flowArn() {
        return this.getStringAttribute('flow_arn');
    }
}
exports.DataAwsMedialiveInputMediaConnectFlowsOutputReference = DataAwsMedialiveInputMediaConnectFlowsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputMediaConnectFlowsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputMediaConnectFlowsOutputReference", version: "19.15.0" };
class DataAwsMedialiveInputMediaConnectFlowsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMedialiveInputMediaConnectFlowsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMedialiveInputMediaConnectFlowsList = DataAwsMedialiveInputMediaConnectFlowsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputMediaConnectFlowsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputMediaConnectFlowsList", version: "19.15.0" };
function dataAwsMedialiveInputSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMedialiveInputSourcesToTerraform = dataAwsMedialiveInputSourcesToTerraform;
function dataAwsMedialiveInputSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMedialiveInputSourcesToHclTerraform = dataAwsMedialiveInputSourcesToHclTerraform;
class DataAwsMedialiveInputSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // password_param - computed: true, optional: false, required: false
    get passwordParam() {
        return this.getStringAttribute('password_param');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    // username - computed: true, optional: false, required: false
    get username() {
        return this.getStringAttribute('username');
    }
}
exports.DataAwsMedialiveInputSourcesOutputReference = DataAwsMedialiveInputSourcesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputSourcesOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputSourcesOutputReference", version: "19.15.0" };
class DataAwsMedialiveInputSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMedialiveInputSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMedialiveInputSourcesList = DataAwsMedialiveInputSourcesList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInputSourcesList[_k] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInputSourcesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/medialive_input aws_medialive_input}
*/
class DataAwsMedialiveInput extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsMedialiveInput resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsMedialiveInput to import
    * @param importFromId The id of the existing DataAwsMedialiveInput that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/medialive_input#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsMedialiveInput to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_medialive_input", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/medialive_input aws_medialive_input} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsMedialiveInputConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_medialive_input',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // destinations - computed: true, optional: false, required: false
        this._destinations = new DataAwsMedialiveInputDestinationsList(this, "destinations", false);
        // input_devices - computed: true, optional: false, required: false
        this._inputDevices = new DataAwsMedialiveInputInputDevicesList(this, "input_devices", false);
        // media_connect_flows - computed: true, optional: false, required: false
        this._mediaConnectFlows = new DataAwsMedialiveInputMediaConnectFlowsList(this, "media_connect_flows", false);
        // sources - computed: true, optional: false, required: false
        this._sources = new DataAwsMedialiveInputSourcesList(this, "sources", false);
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attached_channels - computed: true, optional: false, required: false
    get attachedChannels() {
        return this.getListAttribute('attached_channels');
    }
    get destinations() {
        return this._destinations;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // input_class - computed: true, optional: false, required: false
    get inputClass() {
        return this.getStringAttribute('input_class');
    }
    get inputDevices() {
        return this._inputDevices;
    }
    // input_partner_ids - computed: true, optional: false, required: false
    get inputPartnerIds() {
        return this.getListAttribute('input_partner_ids');
    }
    // input_source_type - computed: true, optional: false, required: false
    get inputSourceType() {
        return this.getStringAttribute('input_source_type');
    }
    get mediaConnectFlows() {
        return this._mediaConnectFlows;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // security_groups - computed: true, optional: false, required: false
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    get sources() {
        return this._sources;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this._tags;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsMedialiveInput = DataAwsMedialiveInput;
_l = JSII_RTTI_SYMBOL_1;
DataAwsMedialiveInput[_l] = { fqn: "@cdktf/provider-aws.dataAwsMedialiveInput.DataAwsMedialiveInput", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsMedialiveInput.tfResourceType = "aws_medialive_input";
