// @cdktf/provider-aws/lib/iam-user-login-profile/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamUserLoginProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_user_login_profile aws_iam_user_login_profile}
*/
class IamUserLoginProfile extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamUserLoginProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamUserLoginProfile to import
    * @param importFromId The id of the existing IamUserLoginProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_user_login_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamUserLoginProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iam_user_login_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_user_login_profile aws_iam_user_login_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamUserLoginProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_user_login_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._passwordLength = config.passwordLength;
        this._passwordResetRequired = config.passwordResetRequired;
        this._pgpKey = config.pgpKey;
        this._user = config.user;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // encrypted_password - computed: true, optional: false, required: false
    get encryptedPassword() {
        return this.getStringAttribute('encrypted_password');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // key_fingerprint - computed: true, optional: false, required: false
    get keyFingerprint() {
        return this.getStringAttribute('key_fingerprint');
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    get passwordLength() {
        return this.getNumberAttribute('password_length');
    }
    set passwordLength(value) {
        this._passwordLength = value;
    }
    resetPasswordLength() {
        this._passwordLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordLengthInput() {
        return this._passwordLength;
    }
    get passwordResetRequired() {
        return this.getBooleanAttribute('password_reset_required');
    }
    set passwordResetRequired(value) {
        this._passwordResetRequired = value;
    }
    resetPasswordResetRequired() {
        this._passwordResetRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordResetRequiredInput() {
        return this._passwordResetRequired;
    }
    get pgpKey() {
        return this.getStringAttribute('pgp_key');
    }
    set pgpKey(value) {
        this._pgpKey = value;
    }
    resetPgpKey() {
        this._pgpKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pgpKeyInput() {
        return this._pgpKey;
    }
    get user() {
        return this.getStringAttribute('user');
    }
    set user(value) {
        this._user = value;
    }
    // Temporarily expose input value. Use with caution.
    get userInput() {
        return this._user;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            password_length: cdktf.numberToTerraform(this._passwordLength),
            password_reset_required: cdktf.booleanToTerraform(this._passwordResetRequired),
            pgp_key: cdktf.stringToTerraform(this._pgpKey),
            user: cdktf.stringToTerraform(this._user),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            password_length: {
                value: cdktf.numberToHclTerraform(this._passwordLength),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            password_reset_required: {
                value: cdktf.booleanToHclTerraform(this._passwordResetRequired),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            pgp_key: {
                value: cdktf.stringToHclTerraform(this._pgpKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user: {
                value: cdktf.stringToHclTerraform(this._user),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamUserLoginProfile = IamUserLoginProfile;
_a = JSII_RTTI_SYMBOL_1;
IamUserLoginProfile[_a] = { fqn: "@cdktf/provider-aws.iamUserLoginProfile.IamUserLoginProfile", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IamUserLoginProfile.tfResourceType = "aws_iam_user_login_profile";
