// @cdktf/provider-aws/lib/data-aws-guardduty-detector/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsGuarddutyDetector = exports.DataAwsGuarddutyDetectorFeaturesList = exports.DataAwsGuarddutyDetectorFeaturesOutputReference = exports.dataAwsGuarddutyDetectorFeaturesToHclTerraform = exports.dataAwsGuarddutyDetectorFeaturesToTerraform = exports.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList = exports.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference = exports.dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToHclTerraform = exports.dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToTerraform = dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToTerraform;
function dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToHclTerraform = dataAwsGuarddutyDetectorFeaturesAdditionalConfigurationToHclTerraform;
class DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference = DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsGuarddutyDetector.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference", version: "19.15.0" };
class DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList = DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsGuarddutyDetector.DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList", version: "19.15.0" };
function dataAwsGuarddutyDetectorFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGuarddutyDetectorFeaturesToTerraform = dataAwsGuarddutyDetectorFeaturesToTerraform;
function dataAwsGuarddutyDetectorFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGuarddutyDetectorFeaturesToHclTerraform = dataAwsGuarddutyDetectorFeaturesToHclTerraform;
class DataAwsGuarddutyDetectorFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // additional_configuration - computed: true, optional: false, required: false
        this._additionalConfiguration = new DataAwsGuarddutyDetectorFeaturesAdditionalConfigurationList(this, "additional_configuration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get additionalConfiguration() {
        return this._additionalConfiguration;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.DataAwsGuarddutyDetectorFeaturesOutputReference = DataAwsGuarddutyDetectorFeaturesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsGuarddutyDetectorFeaturesOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsGuarddutyDetector.DataAwsGuarddutyDetectorFeaturesOutputReference", version: "19.15.0" };
class DataAwsGuarddutyDetectorFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGuarddutyDetectorFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGuarddutyDetectorFeaturesList = DataAwsGuarddutyDetectorFeaturesList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsGuarddutyDetectorFeaturesList[_d] = { fqn: "@cdktf/provider-aws.dataAwsGuarddutyDetector.DataAwsGuarddutyDetectorFeaturesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/guardduty_detector aws_guardduty_detector}
*/
class DataAwsGuarddutyDetector extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsGuarddutyDetector resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsGuarddutyDetector to import
    * @param importFromId The id of the existing DataAwsGuarddutyDetector that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/guardduty_detector#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsGuarddutyDetector to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_guardduty_detector", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/guardduty_detector aws_guardduty_detector} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsGuarddutyDetectorConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_detector',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // features - computed: true, optional: false, required: false
        this._features = new DataAwsGuarddutyDetectorFeaturesList(this, "features", false);
        this._id = config.id;
    }
    get features() {
        return this._features;
    }
    // finding_publishing_frequency - computed: true, optional: false, required: false
    get findingPublishingFrequency() {
        return this.getStringAttribute('finding_publishing_frequency');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // service_role_arn - computed: true, optional: false, required: false
    get serviceRoleArn() {
        return this.getStringAttribute('service_role_arn');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsGuarddutyDetector = DataAwsGuarddutyDetector;
_e = JSII_RTTI_SYMBOL_1;
DataAwsGuarddutyDetector[_e] = { fqn: "@cdktf/provider-aws.dataAwsGuarddutyDetector.DataAwsGuarddutyDetector", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsGuarddutyDetector.tfResourceType = "aws_guardduty_detector";
