// @cdktf/provider-aws/lib/ssm-maintenance-window/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmMaintenanceWindow = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_maintenance_window aws_ssm_maintenance_window}
*/
class SsmMaintenanceWindow extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsmMaintenanceWindow resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsmMaintenanceWindow to import
    * @param importFromId The id of the existing SsmMaintenanceWindow that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_maintenance_window#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsmMaintenanceWindow to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssm_maintenance_window", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssm_maintenance_window aws_ssm_maintenance_window} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsmMaintenanceWindowConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_maintenance_window',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._allowUnassociatedTargets = config.allowUnassociatedTargets;
        this._cutoff = config.cutoff;
        this._description = config.description;
        this._duration = config.duration;
        this._enabled = config.enabled;
        this._endDate = config.endDate;
        this._id = config.id;
        this._name = config.name;
        this._schedule = config.schedule;
        this._scheduleOffset = config.scheduleOffset;
        this._scheduleTimezone = config.scheduleTimezone;
        this._startDate = config.startDate;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    get allowUnassociatedTargets() {
        return this.getBooleanAttribute('allow_unassociated_targets');
    }
    set allowUnassociatedTargets(value) {
        this._allowUnassociatedTargets = value;
    }
    resetAllowUnassociatedTargets() {
        this._allowUnassociatedTargets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowUnassociatedTargetsInput() {
        return this._allowUnassociatedTargets;
    }
    get cutoff() {
        return this.getNumberAttribute('cutoff');
    }
    set cutoff(value) {
        this._cutoff = value;
    }
    // Temporarily expose input value. Use with caution.
    get cutoffInput() {
        return this._cutoff;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get duration() {
        return this.getNumberAttribute('duration');
    }
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    set endDate(value) {
        this._endDate = value;
    }
    resetEndDate() {
        this._endDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endDateInput() {
        return this._endDate;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
    get scheduleOffset() {
        return this.getNumberAttribute('schedule_offset');
    }
    set scheduleOffset(value) {
        this._scheduleOffset = value;
    }
    resetScheduleOffset() {
        this._scheduleOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleOffsetInput() {
        return this._scheduleOffset;
    }
    get scheduleTimezone() {
        return this.getStringAttribute('schedule_timezone');
    }
    set scheduleTimezone(value) {
        this._scheduleTimezone = value;
    }
    resetScheduleTimezone() {
        this._scheduleTimezone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleTimezoneInput() {
        return this._scheduleTimezone;
    }
    get startDate() {
        return this.getStringAttribute('start_date');
    }
    set startDate(value) {
        this._startDate = value;
    }
    resetStartDate() {
        this._startDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startDateInput() {
        return this._startDate;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allow_unassociated_targets: cdktf.booleanToTerraform(this._allowUnassociatedTargets),
            cutoff: cdktf.numberToTerraform(this._cutoff),
            description: cdktf.stringToTerraform(this._description),
            duration: cdktf.numberToTerraform(this._duration),
            enabled: cdktf.booleanToTerraform(this._enabled),
            end_date: cdktf.stringToTerraform(this._endDate),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            schedule: cdktf.stringToTerraform(this._schedule),
            schedule_offset: cdktf.numberToTerraform(this._scheduleOffset),
            schedule_timezone: cdktf.stringToTerraform(this._scheduleTimezone),
            start_date: cdktf.stringToTerraform(this._startDate),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            allow_unassociated_targets: {
                value: cdktf.booleanToHclTerraform(this._allowUnassociatedTargets),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            cutoff: {
                value: cdktf.numberToHclTerraform(this._cutoff),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            duration: {
                value: cdktf.numberToHclTerraform(this._duration),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            end_date: {
                value: cdktf.stringToHclTerraform(this._endDate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            schedule: {
                value: cdktf.stringToHclTerraform(this._schedule),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            schedule_offset: {
                value: cdktf.numberToHclTerraform(this._scheduleOffset),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            schedule_timezone: {
                value: cdktf.stringToHclTerraform(this._scheduleTimezone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            start_date: {
                value: cdktf.stringToHclTerraform(this._startDate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsmMaintenanceWindow = SsmMaintenanceWindow;
_a = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindow[_a] = { fqn: "@cdktf/provider-aws.ssmMaintenanceWindow.SsmMaintenanceWindow", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsmMaintenanceWindow.tfResourceType = "aws_ssm_maintenance_window";
