// @cdktf/provider-aws/lib/glue-user-defined-function/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueUserDefinedFunction = exports.GlueUserDefinedFunctionResourceUrisList = exports.GlueUserDefinedFunctionResourceUrisOutputReference = exports.glueUserDefinedFunctionResourceUrisToHclTerraform = exports.glueUserDefinedFunctionResourceUrisToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueUserDefinedFunctionResourceUrisToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_type: cdktf.stringToTerraform(struct.resourceType),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.glueUserDefinedFunctionResourceUrisToTerraform = glueUserDefinedFunctionResourceUrisToTerraform;
function glueUserDefinedFunctionResourceUrisToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resource_type: {
            value: cdktf.stringToHclTerraform(struct.resourceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.glueUserDefinedFunctionResourceUrisToHclTerraform = glueUserDefinedFunctionResourceUrisToHclTerraform;
class GlueUserDefinedFunctionResourceUrisOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceType = this._resourceType;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._resourceType = undefined;
            this._uri = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._resourceType = value.resourceType;
            this._uri = value.uri;
        }
    }
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    set resourceType(value) {
        this._resourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypeInput() {
        return this._resourceType;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.GlueUserDefinedFunctionResourceUrisOutputReference = GlueUserDefinedFunctionResourceUrisOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueUserDefinedFunctionResourceUrisOutputReference[_a] = { fqn: "@cdktf/provider-aws.glueUserDefinedFunction.GlueUserDefinedFunctionResourceUrisOutputReference", version: "19.15.0" };
class GlueUserDefinedFunctionResourceUrisList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueUserDefinedFunctionResourceUrisOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueUserDefinedFunctionResourceUrisList = GlueUserDefinedFunctionResourceUrisList;
_b = JSII_RTTI_SYMBOL_1;
GlueUserDefinedFunctionResourceUrisList[_b] = { fqn: "@cdktf/provider-aws.glueUserDefinedFunction.GlueUserDefinedFunctionResourceUrisList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_user_defined_function aws_glue_user_defined_function}
*/
class GlueUserDefinedFunction extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GlueUserDefinedFunction resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GlueUserDefinedFunction to import
    * @param importFromId The id of the existing GlueUserDefinedFunction that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_user_defined_function#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GlueUserDefinedFunction to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_glue_user_defined_function", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_user_defined_function aws_glue_user_defined_function} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueUserDefinedFunctionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_user_defined_function',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // resource_uris - computed: false, optional: true, required: false
        this._resourceUris = new GlueUserDefinedFunctionResourceUrisList(this, "resource_uris", true);
        this._catalogId = config.catalogId;
        this._className = config.className;
        this._databaseName = config.databaseName;
        this._id = config.id;
        this._name = config.name;
        this._ownerName = config.ownerName;
        this._ownerType = config.ownerType;
        this._resourceUris.internalValue = config.resourceUris;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get className() {
        return this.getStringAttribute('class_name');
    }
    set className(value) {
        this._className = value;
    }
    // Temporarily expose input value. Use with caution.
    get classNameInput() {
        return this._className;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ownerName() {
        return this.getStringAttribute('owner_name');
    }
    set ownerName(value) {
        this._ownerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerNameInput() {
        return this._ownerName;
    }
    get ownerType() {
        return this.getStringAttribute('owner_type');
    }
    set ownerType(value) {
        this._ownerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerTypeInput() {
        return this._ownerType;
    }
    get resourceUris() {
        return this._resourceUris;
    }
    putResourceUris(value) {
        this._resourceUris.internalValue = value;
    }
    resetResourceUris() {
        this._resourceUris.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceUrisInput() {
        return this._resourceUris.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            class_name: cdktf.stringToTerraform(this._className),
            database_name: cdktf.stringToTerraform(this._databaseName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            owner_name: cdktf.stringToTerraform(this._ownerName),
            owner_type: cdktf.stringToTerraform(this._ownerType),
            resource_uris: cdktf.listMapper(glueUserDefinedFunctionResourceUrisToTerraform, true)(this._resourceUris.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            catalog_id: {
                value: cdktf.stringToHclTerraform(this._catalogId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            class_name: {
                value: cdktf.stringToHclTerraform(this._className),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            database_name: {
                value: cdktf.stringToHclTerraform(this._databaseName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            owner_name: {
                value: cdktf.stringToHclTerraform(this._ownerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            owner_type: {
                value: cdktf.stringToHclTerraform(this._ownerType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_uris: {
                value: cdktf.listMapperHcl(glueUserDefinedFunctionResourceUrisToHclTerraform, true)(this._resourceUris.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "GlueUserDefinedFunctionResourceUrisList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GlueUserDefinedFunction = GlueUserDefinedFunction;
_c = JSII_RTTI_SYMBOL_1;
GlueUserDefinedFunction[_c] = { fqn: "@cdktf/provider-aws.glueUserDefinedFunction.GlueUserDefinedFunction", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GlueUserDefinedFunction.tfResourceType = "aws_glue_user_defined_function";
