// @cdktf/provider-aws/lib/chime-voice-connector-termination/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChimeVoiceConnectorTermination = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chime_voice_connector_termination aws_chime_voice_connector_termination}
*/
class ChimeVoiceConnectorTermination extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ChimeVoiceConnectorTermination resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ChimeVoiceConnectorTermination to import
    * @param importFromId The id of the existing ChimeVoiceConnectorTermination that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chime_voice_connector_termination#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ChimeVoiceConnectorTermination to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_chime_voice_connector_termination", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chime_voice_connector_termination aws_chime_voice_connector_termination} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ChimeVoiceConnectorTerminationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_chime_voice_connector_termination',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._callingRegions = config.callingRegions;
        this._cidrAllowList = config.cidrAllowList;
        this._cpsLimit = config.cpsLimit;
        this._defaultPhoneNumber = config.defaultPhoneNumber;
        this._disabled = config.disabled;
        this._id = config.id;
        this._voiceConnectorId = config.voiceConnectorId;
    }
    get callingRegions() {
        return cdktf.Fn.tolist(this.getListAttribute('calling_regions'));
    }
    set callingRegions(value) {
        this._callingRegions = value;
    }
    // Temporarily expose input value. Use with caution.
    get callingRegionsInput() {
        return this._callingRegions;
    }
    get cidrAllowList() {
        return cdktf.Fn.tolist(this.getListAttribute('cidr_allow_list'));
    }
    set cidrAllowList(value) {
        this._cidrAllowList = value;
    }
    // Temporarily expose input value. Use with caution.
    get cidrAllowListInput() {
        return this._cidrAllowList;
    }
    get cpsLimit() {
        return this.getNumberAttribute('cps_limit');
    }
    set cpsLimit(value) {
        this._cpsLimit = value;
    }
    resetCpsLimit() {
        this._cpsLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpsLimitInput() {
        return this._cpsLimit;
    }
    get defaultPhoneNumber() {
        return this.getStringAttribute('default_phone_number');
    }
    set defaultPhoneNumber(value) {
        this._defaultPhoneNumber = value;
    }
    resetDefaultPhoneNumber() {
        this._defaultPhoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultPhoneNumberInput() {
        return this._defaultPhoneNumber;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get voiceConnectorId() {
        return this.getStringAttribute('voice_connector_id');
    }
    set voiceConnectorId(value) {
        this._voiceConnectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    get voiceConnectorIdInput() {
        return this._voiceConnectorId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            calling_regions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._callingRegions),
            cidr_allow_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._cidrAllowList),
            cps_limit: cdktf.numberToTerraform(this._cpsLimit),
            default_phone_number: cdktf.stringToTerraform(this._defaultPhoneNumber),
            disabled: cdktf.booleanToTerraform(this._disabled),
            id: cdktf.stringToTerraform(this._id),
            voice_connector_id: cdktf.stringToTerraform(this._voiceConnectorId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            calling_regions: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._callingRegions),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            cidr_allow_list: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._cidrAllowList),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            cps_limit: {
                value: cdktf.numberToHclTerraform(this._cpsLimit),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            default_phone_number: {
                value: cdktf.stringToHclTerraform(this._defaultPhoneNumber),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disabled: {
                value: cdktf.booleanToHclTerraform(this._disabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            voice_connector_id: {
                value: cdktf.stringToHclTerraform(this._voiceConnectorId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ChimeVoiceConnectorTermination = ChimeVoiceConnectorTermination;
_a = JSII_RTTI_SYMBOL_1;
ChimeVoiceConnectorTermination[_a] = { fqn: "@cdktf/provider-aws.chimeVoiceConnectorTermination.ChimeVoiceConnectorTermination", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ChimeVoiceConnectorTermination.tfResourceType = "aws_chime_voice_connector_termination";
