// @cdktf/provider-aws/lib/data-aws-datapipeline-pipeline-definition/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDatapipelinePipelineDefinition = exports.DataAwsDatapipelinePipelineDefinitionParameterValueList = exports.DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference = exports.dataAwsDatapipelinePipelineDefinitionParameterValueToHclTerraform = exports.dataAwsDatapipelinePipelineDefinitionParameterValueToTerraform = exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectList = exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference = exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectToHclTerraform = exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectToTerraform = exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList = exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference = exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToHclTerraform = exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToTerraform = exports.DataAwsDatapipelinePipelineDefinitionParameterObjectList = exports.DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference = exports.dataAwsDatapipelinePipelineDefinitionParameterObjectToHclTerraform = exports.dataAwsDatapipelinePipelineDefinitionParameterObjectToTerraform = exports.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList = exports.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference = exports.dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToHclTerraform = exports.dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToTerraform = dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToTerraform;
function dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToHclTerraform = dataAwsDatapipelinePipelineDefinitionParameterObjectAttributeToHclTerraform;
class DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // string_value - computed: true, optional: false, required: false
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference = DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference", version: "19.15.0" };
class DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList = DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList[_b] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList", version: "19.15.0" };
function dataAwsDatapipelinePipelineDefinitionParameterObjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDatapipelinePipelineDefinitionParameterObjectToTerraform = dataAwsDatapipelinePipelineDefinitionParameterObjectToTerraform;
function dataAwsDatapipelinePipelineDefinitionParameterObjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDatapipelinePipelineDefinitionParameterObjectToHclTerraform = dataAwsDatapipelinePipelineDefinitionParameterObjectToHclTerraform;
class DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // attribute - computed: true, optional: false, required: false
        this._attribute = new DataAwsDatapipelinePipelineDefinitionParameterObjectAttributeList(this, "attribute", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get attribute() {
        return this._attribute;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference = DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference", version: "19.15.0" };
class DataAwsDatapipelinePipelineDefinitionParameterObjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDatapipelinePipelineDefinitionParameterObjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterObjectList = DataAwsDatapipelinePipelineDefinitionParameterObjectList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterObjectList[_d] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterObjectList", version: "19.15.0" };
function dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToTerraform = dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToTerraform;
function dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToHclTerraform = dataAwsDatapipelinePipelineDefinitionPipelineObjectFieldToHclTerraform;
class DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // ref_value - computed: true, optional: false, required: false
    get refValue() {
        return this.getStringAttribute('ref_value');
    }
    // string_value - computed: true, optional: false, required: false
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
}
exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference = DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference", version: "19.15.0" };
class DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList = DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList[_f] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList", version: "19.15.0" };
function dataAwsDatapipelinePipelineDefinitionPipelineObjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectToTerraform = dataAwsDatapipelinePipelineDefinitionPipelineObjectToTerraform;
function dataAwsDatapipelinePipelineDefinitionPipelineObjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDatapipelinePipelineDefinitionPipelineObjectToHclTerraform = dataAwsDatapipelinePipelineDefinitionPipelineObjectToHclTerraform;
class DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // field - computed: true, optional: false, required: false
        this._field = new DataAwsDatapipelinePipelineDefinitionPipelineObjectFieldList(this, "field", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get field() {
        return this._field;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference = DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference", version: "19.15.0" };
class DataAwsDatapipelinePipelineDefinitionPipelineObjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDatapipelinePipelineDefinitionPipelineObjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDatapipelinePipelineDefinitionPipelineObjectList = DataAwsDatapipelinePipelineDefinitionPipelineObjectList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionPipelineObjectList[_h] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionPipelineObjectList", version: "19.15.0" };
function dataAwsDatapipelinePipelineDefinitionParameterValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDatapipelinePipelineDefinitionParameterValueToTerraform = dataAwsDatapipelinePipelineDefinitionParameterValueToTerraform;
function dataAwsDatapipelinePipelineDefinitionParameterValueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDatapipelinePipelineDefinitionParameterValueToHclTerraform = dataAwsDatapipelinePipelineDefinitionParameterValueToHclTerraform;
class DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // string_value - computed: true, optional: false, required: false
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference = DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference", version: "19.15.0" };
class DataAwsDatapipelinePipelineDefinitionParameterValueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDatapipelinePipelineDefinitionParameterValueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDatapipelinePipelineDefinitionParameterValueList = DataAwsDatapipelinePipelineDefinitionParameterValueList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinitionParameterValueList[_k] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinitionParameterValueList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/datapipeline_pipeline_definition aws_datapipeline_pipeline_definition}
*/
class DataAwsDatapipelinePipelineDefinition extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDatapipelinePipelineDefinition resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDatapipelinePipelineDefinition to import
    * @param importFromId The id of the existing DataAwsDatapipelinePipelineDefinition that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/datapipeline_pipeline_definition#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDatapipelinePipelineDefinition to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_datapipeline_pipeline_definition", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/datapipeline_pipeline_definition aws_datapipeline_pipeline_definition} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDatapipelinePipelineDefinitionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datapipeline_pipeline_definition',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // parameter_object - computed: true, optional: false, required: false
        this._parameterObject = new DataAwsDatapipelinePipelineDefinitionParameterObjectList(this, "parameter_object", true);
        // pipeline_object - computed: true, optional: false, required: false
        this._pipelineObject = new DataAwsDatapipelinePipelineDefinitionPipelineObjectList(this, "pipeline_object", true);
        // parameter_value - computed: false, optional: true, required: false
        this._parameterValue = new DataAwsDatapipelinePipelineDefinitionParameterValueList(this, "parameter_value", true);
        this._id = config.id;
        this._pipelineId = config.pipelineId;
        this._parameterValue.internalValue = config.parameterValue;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get parameterObject() {
        return this._parameterObject;
    }
    get pipelineId() {
        return this.getStringAttribute('pipeline_id');
    }
    set pipelineId(value) {
        this._pipelineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineIdInput() {
        return this._pipelineId;
    }
    get pipelineObject() {
        return this._pipelineObject;
    }
    get parameterValue() {
        return this._parameterValue;
    }
    putParameterValue(value) {
        this._parameterValue.internalValue = value;
    }
    resetParameterValue() {
        this._parameterValue.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterValueInput() {
        return this._parameterValue.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            pipeline_id: cdktf.stringToTerraform(this._pipelineId),
            parameter_value: cdktf.listMapper(dataAwsDatapipelinePipelineDefinitionParameterValueToTerraform, true)(this._parameterValue.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_id: {
                value: cdktf.stringToHclTerraform(this._pipelineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parameter_value: {
                value: cdktf.listMapperHcl(dataAwsDatapipelinePipelineDefinitionParameterValueToHclTerraform, true)(this._parameterValue.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsDatapipelinePipelineDefinitionParameterValueList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDatapipelinePipelineDefinition = DataAwsDatapipelinePipelineDefinition;
_l = JSII_RTTI_SYMBOL_1;
DataAwsDatapipelinePipelineDefinition[_l] = { fqn: "@cdktf/provider-aws.dataAwsDatapipelinePipelineDefinition.DataAwsDatapipelinePipelineDefinition", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDatapipelinePipelineDefinition.tfResourceType = "aws_datapipeline_pipeline_definition";
