// @cdktf/provider-aws/lib/data-aws-quicksight-theme/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsQuicksightTheme = exports.DataAwsQuicksightThemePermissionsList = exports.DataAwsQuicksightThemePermissionsOutputReference = exports.dataAwsQuicksightThemePermissionsToHclTerraform = exports.dataAwsQuicksightThemePermissionsToTerraform = exports.DataAwsQuicksightThemeConfigurationList = exports.DataAwsQuicksightThemeConfigurationOutputReference = exports.dataAwsQuicksightThemeConfigurationToHclTerraform = exports.dataAwsQuicksightThemeConfigurationToTerraform = exports.DataAwsQuicksightThemeConfigurationUiColorPaletteList = exports.DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference = exports.dataAwsQuicksightThemeConfigurationUiColorPaletteToHclTerraform = exports.dataAwsQuicksightThemeConfigurationUiColorPaletteToTerraform = exports.DataAwsQuicksightThemeConfigurationTypographyList = exports.DataAwsQuicksightThemeConfigurationTypographyOutputReference = exports.dataAwsQuicksightThemeConfigurationTypographyToHclTerraform = exports.dataAwsQuicksightThemeConfigurationTypographyToTerraform = exports.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList = exports.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference = exports.dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToHclTerraform = exports.dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetList = exports.DataAwsQuicksightThemeConfigurationSheetOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutList = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList = exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetTileList = exports.DataAwsQuicksightThemeConfigurationSheetTileOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetTileToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetTileToTerraform = exports.DataAwsQuicksightThemeConfigurationSheetTileBorderList = exports.DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference = exports.dataAwsQuicksightThemeConfigurationSheetTileBorderToHclTerraform = exports.dataAwsQuicksightThemeConfigurationSheetTileBorderToTerraform = exports.DataAwsQuicksightThemeConfigurationDataColorPaletteList = exports.DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference = exports.dataAwsQuicksightThemeConfigurationDataColorPaletteToHclTerraform = exports.dataAwsQuicksightThemeConfigurationDataColorPaletteToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsQuicksightThemeConfigurationDataColorPaletteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationDataColorPaletteToTerraform = dataAwsQuicksightThemeConfigurationDataColorPaletteToTerraform;
function dataAwsQuicksightThemeConfigurationDataColorPaletteToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationDataColorPaletteToHclTerraform = dataAwsQuicksightThemeConfigurationDataColorPaletteToHclTerraform;
class DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // colors - computed: true, optional: false, required: false
    get colors() {
        return this.getListAttribute('colors');
    }
    // empty_fill_color - computed: true, optional: false, required: false
    get emptyFillColor() {
        return this.getStringAttribute('empty_fill_color');
    }
    // min_max_gradient - computed: true, optional: false, required: false
    get minMaxGradient() {
        return this.getListAttribute('min_max_gradient');
    }
}
exports.DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference = DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationDataColorPaletteList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationDataColorPaletteOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationDataColorPaletteList = DataAwsQuicksightThemeConfigurationDataColorPaletteList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationDataColorPaletteList[_b] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationDataColorPaletteList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetTileBorderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetTileBorderToTerraform = dataAwsQuicksightThemeConfigurationSheetTileBorderToTerraform;
function dataAwsQuicksightThemeConfigurationSheetTileBorderToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetTileBorderToHclTerraform = dataAwsQuicksightThemeConfigurationSheetTileBorderToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // show - computed: true, optional: false, required: false
    get show() {
        return this.getBooleanAttribute('show');
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference = DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetTileBorderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetTileBorderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileBorderList = DataAwsQuicksightThemeConfigurationSheetTileBorderList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileBorderList[_d] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileBorderList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetTileToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetTileToTerraform = dataAwsQuicksightThemeConfigurationSheetTileToTerraform;
function dataAwsQuicksightThemeConfigurationSheetTileToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetTileToHclTerraform = dataAwsQuicksightThemeConfigurationSheetTileToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetTileOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // border - computed: true, optional: false, required: false
        this._border = new DataAwsQuicksightThemeConfigurationSheetTileBorderList(this, "border", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get border() {
        return this._border;
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileOutputReference = DataAwsQuicksightThemeConfigurationSheetTileOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetTileList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetTileOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileList = DataAwsQuicksightThemeConfigurationSheetTileList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileList[_f] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToTerraform;
function dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToHclTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutGutterToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // show - computed: true, optional: false, required: false
    get show() {
        return this.getBooleanAttribute('show');
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference = DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList = DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList[_h] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToTerraform;
function dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToHclTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutMarginToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // show - computed: true, optional: false, required: false
    get show() {
        return this.getBooleanAttribute('show');
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference = DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList = DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList[_k] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetTileLayoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutToTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutToTerraform;
function dataAwsQuicksightThemeConfigurationSheetTileLayoutToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetTileLayoutToHclTerraform = dataAwsQuicksightThemeConfigurationSheetTileLayoutToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // gutter - computed: true, optional: false, required: false
        this._gutter = new DataAwsQuicksightThemeConfigurationSheetTileLayoutGutterList(this, "gutter", false);
        // margin - computed: true, optional: false, required: false
        this._margin = new DataAwsQuicksightThemeConfigurationSheetTileLayoutMarginList(this, "margin", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get gutter() {
        return this._gutter;
    }
    get margin() {
        return this._margin;
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference = DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetTileLayoutList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetTileLayoutOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetTileLayoutList = DataAwsQuicksightThemeConfigurationSheetTileLayoutList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetTileLayoutList[_m] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetTileLayoutList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationSheetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationSheetToTerraform = dataAwsQuicksightThemeConfigurationSheetToTerraform;
function dataAwsQuicksightThemeConfigurationSheetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationSheetToHclTerraform = dataAwsQuicksightThemeConfigurationSheetToHclTerraform;
class DataAwsQuicksightThemeConfigurationSheetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // tile - computed: true, optional: false, required: false
        this._tile = new DataAwsQuicksightThemeConfigurationSheetTileList(this, "tile", false);
        // tile_layout - computed: true, optional: false, required: false
        this._tileLayout = new DataAwsQuicksightThemeConfigurationSheetTileLayoutList(this, "tile_layout", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get tile() {
        return this._tile;
    }
    get tileLayout() {
        return this._tileLayout;
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetOutputReference = DataAwsQuicksightThemeConfigurationSheetOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationSheetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationSheetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationSheetList = DataAwsQuicksightThemeConfigurationSheetList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationSheetList[_p] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationSheetList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToTerraform = dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToTerraform;
function dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToHclTerraform = dataAwsQuicksightThemeConfigurationTypographyFontFamiliesToHclTerraform;
class DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // font_family - computed: true, optional: false, required: false
    get fontFamily() {
        return this.getStringAttribute('font_family');
    }
}
exports.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference = DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationTypographyFontFamiliesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList = DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList[_r] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationTypographyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationTypographyToTerraform = dataAwsQuicksightThemeConfigurationTypographyToTerraform;
function dataAwsQuicksightThemeConfigurationTypographyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationTypographyToHclTerraform = dataAwsQuicksightThemeConfigurationTypographyToHclTerraform;
class DataAwsQuicksightThemeConfigurationTypographyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // font_families - computed: true, optional: false, required: false
        this._fontFamilies = new DataAwsQuicksightThemeConfigurationTypographyFontFamiliesList(this, "font_families", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get fontFamilies() {
        return this._fontFamilies;
    }
}
exports.DataAwsQuicksightThemeConfigurationTypographyOutputReference = DataAwsQuicksightThemeConfigurationTypographyOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationTypographyOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationTypographyOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationTypographyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationTypographyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationTypographyList = DataAwsQuicksightThemeConfigurationTypographyList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationTypographyList[_t] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationTypographyList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationUiColorPaletteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationUiColorPaletteToTerraform = dataAwsQuicksightThemeConfigurationUiColorPaletteToTerraform;
function dataAwsQuicksightThemeConfigurationUiColorPaletteToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationUiColorPaletteToHclTerraform = dataAwsQuicksightThemeConfigurationUiColorPaletteToHclTerraform;
class DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // accent - computed: true, optional: false, required: false
    get accent() {
        return this.getStringAttribute('accent');
    }
    // accent_foreground - computed: true, optional: false, required: false
    get accentForeground() {
        return this.getStringAttribute('accent_foreground');
    }
    // danger - computed: true, optional: false, required: false
    get danger() {
        return this.getStringAttribute('danger');
    }
    // danger_foreground - computed: true, optional: false, required: false
    get dangerForeground() {
        return this.getStringAttribute('danger_foreground');
    }
    // dimension - computed: true, optional: false, required: false
    get dimension() {
        return this.getStringAttribute('dimension');
    }
    // dimension_foreground - computed: true, optional: false, required: false
    get dimensionForeground() {
        return this.getStringAttribute('dimension_foreground');
    }
    // measure - computed: true, optional: false, required: false
    get measure() {
        return this.getStringAttribute('measure');
    }
    // measure_foreground - computed: true, optional: false, required: false
    get measureForeground() {
        return this.getStringAttribute('measure_foreground');
    }
    // primary_background - computed: true, optional: false, required: false
    get primaryBackground() {
        return this.getStringAttribute('primary_background');
    }
    // primary_foreground - computed: true, optional: false, required: false
    get primaryForeground() {
        return this.getStringAttribute('primary_foreground');
    }
    // secondary_background - computed: true, optional: false, required: false
    get secondaryBackground() {
        return this.getStringAttribute('secondary_background');
    }
    // secondary_foreground - computed: true, optional: false, required: false
    get secondaryForeground() {
        return this.getStringAttribute('secondary_foreground');
    }
    // success - computed: true, optional: false, required: false
    get success() {
        return this.getStringAttribute('success');
    }
    // success_foreground - computed: true, optional: false, required: false
    get successForeground() {
        return this.getStringAttribute('success_foreground');
    }
    // warning - computed: true, optional: false, required: false
    get warning() {
        return this.getStringAttribute('warning');
    }
    // warning_foreground - computed: true, optional: false, required: false
    get warningForeground() {
        return this.getStringAttribute('warning_foreground');
    }
}
exports.DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference = DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationUiColorPaletteList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationUiColorPaletteOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationUiColorPaletteList = DataAwsQuicksightThemeConfigurationUiColorPaletteList;
_v = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationUiColorPaletteList[_v] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationUiColorPaletteList", version: "19.15.0" };
function dataAwsQuicksightThemeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemeConfigurationToTerraform = dataAwsQuicksightThemeConfigurationToTerraform;
function dataAwsQuicksightThemeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemeConfigurationToHclTerraform = dataAwsQuicksightThemeConfigurationToHclTerraform;
class DataAwsQuicksightThemeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // data_color_palette - computed: true, optional: false, required: false
        this._dataColorPalette = new DataAwsQuicksightThemeConfigurationDataColorPaletteList(this, "data_color_palette", false);
        // sheet - computed: true, optional: false, required: false
        this._sheet = new DataAwsQuicksightThemeConfigurationSheetList(this, "sheet", false);
        // typography - computed: true, optional: false, required: false
        this._typography = new DataAwsQuicksightThemeConfigurationTypographyList(this, "typography", false);
        // ui_color_palette - computed: true, optional: false, required: false
        this._uiColorPalette = new DataAwsQuicksightThemeConfigurationUiColorPaletteList(this, "ui_color_palette", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dataColorPalette() {
        return this._dataColorPalette;
    }
    get sheet() {
        return this._sheet;
    }
    get typography() {
        return this._typography;
    }
    get uiColorPalette() {
        return this._uiColorPalette;
    }
}
exports.DataAwsQuicksightThemeConfigurationOutputReference = DataAwsQuicksightThemeConfigurationOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationOutputReference[_w] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemeConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemeConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemeConfigurationList = DataAwsQuicksightThemeConfigurationList;
_x = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemeConfigurationList[_x] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemeConfigurationList", version: "19.15.0" };
function dataAwsQuicksightThemePermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsQuicksightThemePermissionsToTerraform = dataAwsQuicksightThemePermissionsToTerraform;
function dataAwsQuicksightThemePermissionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsQuicksightThemePermissionsToHclTerraform = dataAwsQuicksightThemePermissionsToHclTerraform;
class DataAwsQuicksightThemePermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // actions - computed: true, optional: false, required: false
    get actions() {
        return cdktf.Fn.tolist(this.getListAttribute('actions'));
    }
    // principal - computed: true, optional: false, required: false
    get principal() {
        return this.getStringAttribute('principal');
    }
}
exports.DataAwsQuicksightThemePermissionsOutputReference = DataAwsQuicksightThemePermissionsOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemePermissionsOutputReference[_y] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemePermissionsOutputReference", version: "19.15.0" };
class DataAwsQuicksightThemePermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsQuicksightThemePermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsQuicksightThemePermissionsList = DataAwsQuicksightThemePermissionsList;
_z = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightThemePermissionsList[_z] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightThemePermissionsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/quicksight_theme aws_quicksight_theme}
*/
class DataAwsQuicksightTheme extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsQuicksightTheme resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsQuicksightTheme to import
    * @param importFromId The id of the existing DataAwsQuicksightTheme that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/quicksight_theme#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsQuicksightTheme to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_quicksight_theme", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/quicksight_theme aws_quicksight_theme} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsQuicksightThemeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_quicksight_theme',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // configuration - computed: true, optional: false, required: false
        this._configuration = new DataAwsQuicksightThemeConfigurationList(this, "configuration", false);
        // permissions - computed: true, optional: false, required: false
        this._permissions = new DataAwsQuicksightThemePermissionsList(this, "permissions", false);
        this._awsAccountId = config.awsAccountId;
        this._id = config.id;
        this._tags = config.tags;
        this._themeId = config.themeId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    set awsAccountId(value) {
        this._awsAccountId = value;
    }
    resetAwsAccountId() {
        this._awsAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccountIdInput() {
        return this._awsAccountId;
    }
    // base_theme_id - computed: true, optional: false, required: false
    get baseThemeId() {
        return this.getStringAttribute('base_theme_id');
    }
    get configuration() {
        return this._configuration;
    }
    // created_time - computed: true, optional: false, required: false
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_time - computed: true, optional: false, required: false
    get lastUpdatedTime() {
        return this.getStringAttribute('last_updated_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get permissions() {
        return this._permissions;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get themeId() {
        return this.getStringAttribute('theme_id');
    }
    set themeId(value) {
        this._themeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get themeIdInput() {
        return this._themeId;
    }
    // version_description - computed: true, optional: false, required: false
    get versionDescription() {
        return this.getStringAttribute('version_description');
    }
    // version_number - computed: true, optional: false, required: false
    get versionNumber() {
        return this.getNumberAttribute('version_number');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            aws_account_id: cdktf.stringToTerraform(this._awsAccountId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            theme_id: cdktf.stringToTerraform(this._themeId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            aws_account_id: {
                value: cdktf.stringToHclTerraform(this._awsAccountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            theme_id: {
                value: cdktf.stringToHclTerraform(this._themeId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsQuicksightTheme = DataAwsQuicksightTheme;
_0 = JSII_RTTI_SYMBOL_1;
DataAwsQuicksightTheme[_0] = { fqn: "@cdktf/provider-aws.dataAwsQuicksightTheme.DataAwsQuicksightTheme", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsQuicksightTheme.tfResourceType = "aws_quicksight_theme";
