// @cdktf/provider-aws/lib/redshiftserverless-snapshot/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessSnapshot = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_snapshot aws_redshiftserverless_snapshot}
*/
class RedshiftserverlessSnapshot extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftserverlessSnapshot resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftserverlessSnapshot to import
    * @param importFromId The id of the existing RedshiftserverlessSnapshot that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_snapshot#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftserverlessSnapshot to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshiftserverless_snapshot", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_snapshot aws_redshiftserverless_snapshot} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessSnapshotConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_snapshot',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._namespaceName = config.namespaceName;
        this._retentionPeriod = config.retentionPeriod;
        this._snapshotName = config.snapshotName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // accounts_with_provisioned_restore_access - computed: true, optional: false, required: false
    get accountsWithProvisionedRestoreAccess() {
        return cdktf.Fn.tolist(this.getListAttribute('accounts_with_provisioned_restore_access'));
    }
    // accounts_with_restore_access - computed: true, optional: false, required: false
    get accountsWithRestoreAccess() {
        return cdktf.Fn.tolist(this.getListAttribute('accounts_with_restore_access'));
    }
    // admin_username - computed: true, optional: false, required: false
    get adminUsername() {
        return this.getStringAttribute('admin_username');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // namespace_arn - computed: true, optional: false, required: false
    get namespaceArn() {
        return this.getStringAttribute('namespace_arn');
    }
    get namespaceName() {
        return this.getStringAttribute('namespace_name');
    }
    set namespaceName(value) {
        this._namespaceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceNameInput() {
        return this._namespaceName;
    }
    // owner_account - computed: true, optional: false, required: false
    get ownerAccount() {
        return this.getStringAttribute('owner_account');
    }
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    set retentionPeriod(value) {
        this._retentionPeriod = value;
    }
    resetRetentionPeriod() {
        this._retentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod;
    }
    get snapshotName() {
        return this.getStringAttribute('snapshot_name');
    }
    set snapshotName(value) {
        this._snapshotName = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotNameInput() {
        return this._snapshotName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            namespace_name: cdktf.stringToTerraform(this._namespaceName),
            retention_period: cdktf.numberToTerraform(this._retentionPeriod),
            snapshot_name: cdktf.stringToTerraform(this._snapshotName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            namespace_name: {
                value: cdktf.stringToHclTerraform(this._namespaceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            retention_period: {
                value: cdktf.numberToHclTerraform(this._retentionPeriod),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            snapshot_name: {
                value: cdktf.stringToHclTerraform(this._snapshotName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftserverlessSnapshot = RedshiftserverlessSnapshot;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessSnapshot[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessSnapshot.RedshiftserverlessSnapshot", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessSnapshot.tfResourceType = "aws_redshiftserverless_snapshot";
