// @winglang/sdk/lib/target-tf-aws/counter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Counter = void 0;
const dynamodb_table_1 = require("../.gen/providers/aws/dynamodb-table");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
const resource_names_1 = require("../shared/resource-names");
const shared_aws_1 = require("../shared-aws");
const commons_1 = require("../shared-aws/commons");
const permissions_1 = require("../shared-aws/permissions");
/**
 * Counter (Table) names must be between 3 and 255 characters.
 * You can use alphanumeric characters, dot (.), dash (-), and underscores (_).
 */
const NAME_OPTS = {
    maxLen: 255,
    disallowedRegex: /[^a-zA-Z0-9\_\.\-]+/g,
    prefix: "wing-counter-",
};
/**
 * AWS implementation of `cloud.Counter`.
 *
 * @inflight `@winglang/sdk.cloud.ICounterClient`
 */
class Counter extends cloud.Counter {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        this.table = new dynamodb_table_1.DynamodbTable(this, "Default", {
            name: resource_names_1.ResourceNames.generateName(this, NAME_OPTS),
            attribute: [{ name: commons_1.COUNTER_HASH_KEY, type: "S" }],
            hashKey: commons_1.COUNTER_HASH_KEY,
            billingMode: "PAY_PER_REQUEST",
        });
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.CounterInflightMethods.INC]: [],
            [cloud.CounterInflightMethods.DEC]: [],
            [cloud.CounterInflightMethods.PEEK]: [],
            [cloud.CounterInflightMethods.SET]: [],
        };
    }
    onLift(host, ops) {
        if (!shared_aws_1.AwsInflightHost.isAwsInflightHost(host)) {
            throw new Error("Host is expected to implement `IAwsInfightHost`");
        }
        host.addPolicyStatements(...(0, permissions_1.calculateCounterPermissions)(this.table.arn, ops));
        host.addEnvironment(this.envName(), this.table.name);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-aws", "shared-aws"), __filename, "CounterClient", [`process.env["${this.envName()}"]`, `${this.initial}`]);
    }
    envName() {
        return `DYNAMODB_TABLE_NAME_${this.node.addr.slice(-8)}`;
    }
    get dynamoTableArn() {
        return this.table.arn;
    }
    get dynamoTableName() {
        return this.table.name;
    }
}
exports.Counter = Counter;
