// @winglang/sdk/lib/target-tf-aws/on-deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnDeploy = void 0;
const data_aws_lambda_invocation_1 = require("../.gen/providers/aws/data-aws-lambda-invocation");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
/**
 * AWS implementation of `cloud.OnDeploy`.
 *
 * @inflight `@winglang/sdk.cloud.IOnDeployClient`
 */
class OnDeploy extends cloud.OnDeploy {
    constructor(scope, id, handler, props = {}) {
        super(scope, id, handler, props);
        let fn = new cloud.Function(this, "Function", handler, props);
        const awsFn = fn;
        // add all of the children of the construct to the dependencies
        const dependsOn = [];
        for (const c of props.executeAfter ?? []) {
            for (const child of c.node.findAll()) {
                if (isTerraformDependable(child)) {
                    dependsOn.push(child);
                }
            }
            this.node.addDependency(c);
        }
        // currently using the aws_lambda_invocation *data source* since it runs on every terraform apply.
        // if we want OnDeploy to only run code conditionally,
        // we can use the aws_lambda_invocation *resource* instead.
        const lambdaInvocation = new data_aws_lambda_invocation_1.DataAwsLambdaInvocation(this, "Invocation", {
            functionName: awsFn.functionName,
            input: JSON.stringify({}), // call the function with an empty object
            dependsOn,
        });
        for (const c of props.executeBefore ?? []) {
            // add the invocation as a dependency on all of the children of the construct
            for (const child of c.node.findAll()) {
                if (isTerraformResource(child)) {
                    if (child.dependsOn === undefined) {
                        child.dependsOn = [];
                    }
                    child.dependsOn.push(lambdaInvocation.fqn);
                }
            }
            c.node.addDependency(this);
        }
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-aws", "shared-aws"), __filename, "OnDeployClient", []);
    }
}
exports.OnDeploy = OnDeploy;
function isTerraformDependable(x) {
    return "fqn" in x;
}
function isTerraformResource(x) {
    return "terraformResourceType" in x;
}
