// @winglang/sdk/lib/target-tf-aws/platform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Platform = void 0;
const app_1 = require("./app");
/**
 * AWS Terraform Platform
 */
class Platform {
    constructor() {
        /** Platform model */
        this.target = "tf-aws";
        this.parameters = {
            $id: "tf-aws-parameters",
            type: "object",
            properties: {
                "tf-aws": {
                    type: "object",
                    properties: {
                        vpc: {
                            type: "string",
                            description: "Determine whether to create a new VPC or use an existing one",
                            enum: ["new", "existing"],
                        },
                        vpc_id: {
                            description: "If using an existing VPC, provide the VPC ID",
                            type: "string",
                        },
                        private_subnet_ids: {
                            description: "If using an existing VPC, provide the private subnet ID",
                            type: "array",
                            items: {
                                type: "string",
                            },
                        },
                        public_subnet_ids: {
                            description: "If using an existing VPC, provide the public subnet ID",
                            type: "array",
                            items: {
                                type: "string",
                            },
                        },
                        vpc_api_gateway: {
                            description: "Whether Api gateways should be deployed in a VPC",
                            type: "boolean",
                        },
                        vpc_lambda: {
                            description: "Whether Lambda functions should be deployed in a VPC",
                            type: "boolean",
                        },
                    },
                    allOf: [
                        {
                            $comment: "if vpc is existing, then we need to require the vpc_id, private_subnet_ids",
                            if: { properties: { vpc: { const: "existing" } } },
                            then: {
                                required: ["vpc_id", "private_subnet_ids"],
                            },
                        },
                    ],
                },
            },
        };
    }
    newApp(appProps) {
        return new app_1.App(appProps);
    }
    async storeSecrets(secrets) {
        const { SecretsManagerClient, GetSecretValueCommand, CreateSecretCommand, UpdateSecretCommand, } = await Promise.resolve().then(() => __importStar(require("@aws-sdk/client-secrets-manager")));
        console.log("Storing secrets in AWS Secrets Manager");
        const client = new SecretsManagerClient({});
        for (const [name, value] of Object.entries(secrets)) {
            try {
                // Attempt to retrieve the secret to check if it exists
                await client.send(new GetSecretValueCommand({ SecretId: name }));
                console.log(`Secret ${name} exists, updating it.`);
                await client.send(new UpdateSecretCommand({
                    SecretId: name,
                    SecretString: value,
                }));
            }
            catch (error) {
                if (error.name === "ResourceNotFoundException") {
                    // If the secret does not exist, create it
                    console.log(`Secret ${name} does not exist, creating it.`);
                    await client.send(new CreateSecretCommand({
                        Name: name,
                        SecretString: value,
                    }));
                }
                else {
                    console.error(`Failed to store secret ${name}:`, error);
                    throw error;
                }
            }
        }
        console.log(`${Object.keys(secrets).length} secret(s) stored AWS Secrets Manager`);
    }
}
exports.Platform = Platform;
