// @winglang/sdk/lib/target-tf-aws/schedule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedule = void 0;
const app_1 = require("./app");
const function_1 = require("./function");
const cloudwatch_event_rule_1 = require("../.gen/providers/aws/cloudwatch-event-rule");
const cloudwatch_event_target_1 = require("../.gen/providers/aws/cloudwatch-event-target");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
const schedule_1 = require("../shared-aws/schedule");
const std_1 = require("../std");
/**
 * AWS implementation of `cloud.Schedule`.
 *
 * @inflight `@winglang/sdk.cloud.IScheduleClient`
 */
class Schedule extends cloud.Schedule {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        this.handlers = {};
        const { rate, cron } = props;
        this.scheduleExpression = rate
            ? rate.minutes === 1
                ? `rate(${rate.minutes} minute)`
                : `rate(${rate.minutes} minutes)`
            : `cron(${(0, schedule_1.convertUnixCronToAWSCron)(cron)})`;
        this.rule = new cloudwatch_event_rule_1.CloudwatchEventRule(this, "Schedule", {
            scheduleExpression: this.scheduleExpression,
        });
    }
    onTick(inflight, props = {}) {
        const functionHandler = schedule_1.ScheduleOnTickHandler.toFunctionHandler(inflight);
        let fn = this.handlers[inflight._id];
        if (fn) {
            return fn;
        }
        fn = new function_1.Function(this, app_1.App.of(this).makeId(this, "OnTick"), functionHandler, props);
        this.handlers[inflight._id] = fn;
        // TODO: remove this constraint by adding generic permission APIs to cloud.Function
        if (!(fn instanceof function_1.Function)) {
            throw new Error("Schedule only supports creating tfaws.Function right now");
        }
        fn.addPermissionToInvoke(this, "events.amazonaws.com", this.rule.arn);
        new cloudwatch_event_target_1.CloudwatchEventTarget(this, app_1.App.of(this).makeId(this, "ScheduleTarget"), {
            arn: fn.qualifiedArn,
            rule: this.rule.name,
        });
        std_1.Node.of(this).addConnection({
            source: this,
            sourceOp: cloud.ScheduleInflightMethods.TICK,
            target: fn,
            targetOp: cloud.FunctionInflightMethods.INVOKE,
            name: "onTick()",
        });
        return fn;
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-aws", "shared-aws"), __filename, "ScheduleClient", [`process.env["${this.envName()}"]`]);
    }
    envName() {
        return `SCHEDULE_EVENT_${this.node.addr.slice(-8)}`;
    }
}
exports.Schedule = Schedule;
