// @cdktf/provider-aws/lib/load-balancer-backend-server-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancerBackendServerPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_backend_server_policy aws_load_balancer_backend_server_policy}
*/
class LoadBalancerBackendServerPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LoadBalancerBackendServerPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LoadBalancerBackendServerPolicy to import
    * @param importFromId The id of the existing LoadBalancerBackendServerPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_backend_server_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LoadBalancerBackendServerPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_load_balancer_backend_server_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_backend_server_policy aws_load_balancer_backend_server_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LoadBalancerBackendServerPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_load_balancer_backend_server_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._instancePort = config.instancePort;
        this._loadBalancerName = config.loadBalancerName;
        this._policyNames = config.policyNames;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instancePort() {
        return this.getNumberAttribute('instance_port');
    }
    set instancePort(value) {
        this._instancePort = value;
    }
    // Temporarily expose input value. Use with caution.
    get instancePortInput() {
        return this._instancePort;
    }
    get loadBalancerName() {
        return this.getStringAttribute('load_balancer_name');
    }
    set loadBalancerName(value) {
        this._loadBalancerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerNameInput() {
        return this._loadBalancerName;
    }
    get policyNames() {
        return cdktf.Fn.tolist(this.getListAttribute('policy_names'));
    }
    set policyNames(value) {
        this._policyNames = value;
    }
    resetPolicyNames() {
        this._policyNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyNamesInput() {
        return this._policyNames;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_port: cdktf.numberToTerraform(this._instancePort),
            load_balancer_name: cdktf.stringToTerraform(this._loadBalancerName),
            policy_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._policyNames),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_port: {
                value: cdktf.numberToHclTerraform(this._instancePort),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            load_balancer_name: {
                value: cdktf.stringToHclTerraform(this._loadBalancerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy_names: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._policyNames),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LoadBalancerBackendServerPolicy = LoadBalancerBackendServerPolicy;
_a = JSII_RTTI_SYMBOL_1;
LoadBalancerBackendServerPolicy[_a] = { fqn: "@cdktf/provider-aws.loadBalancerBackendServerPolicy.LoadBalancerBackendServerPolicy", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LoadBalancerBackendServerPolicy.tfResourceType = "aws_load_balancer_backend_server_policy";
