// @cdktf/provider-aws/lib/appconfig-extension-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppconfigExtensionAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_extension_association aws_appconfig_extension_association}
*/
class AppconfigExtensionAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AppconfigExtensionAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AppconfigExtensionAssociation to import
    * @param importFromId The id of the existing AppconfigExtensionAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_extension_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AppconfigExtensionAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_appconfig_extension_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_extension_association aws_appconfig_extension_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppconfigExtensionAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appconfig_extension_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._extensionArn = config.extensionArn;
        this._id = config.id;
        this._parameters = config.parameters;
        this._resourceArn = config.resourceArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get extensionArn() {
        return this.getStringAttribute('extension_arn');
    }
    set extensionArn(value) {
        this._extensionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get extensionArnInput() {
        return this._extensionArn;
    }
    // extension_version - computed: true, optional: false, required: false
    get extensionVersion() {
        return this.getNumberAttribute('extension_version');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            extension_arn: cdktf.stringToTerraform(this._extensionArn),
            id: cdktf.stringToTerraform(this._id),
            parameters: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameters),
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            extension_arn: {
                value: cdktf.stringToHclTerraform(this._extensionArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parameters: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._parameters),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            resource_arn: {
                value: cdktf.stringToHclTerraform(this._resourceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AppconfigExtensionAssociation = AppconfigExtensionAssociation;
_a = JSII_RTTI_SYMBOL_1;
AppconfigExtensionAssociation[_a] = { fqn: "@cdktf/provider-aws.appconfigExtensionAssociation.AppconfigExtensionAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AppconfigExtensionAssociation.tfResourceType = "aws_appconfig_extension_association";
