// @cdktf/provider-aws/lib/gamelift-game-session-queue/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameliftGameSessionQueue = exports.GameliftGameSessionQueuePlayerLatencyPolicyList = exports.GameliftGameSessionQueuePlayerLatencyPolicyOutputReference = exports.gameliftGameSessionQueuePlayerLatencyPolicyToHclTerraform = exports.gameliftGameSessionQueuePlayerLatencyPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gameliftGameSessionQueuePlayerLatencyPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_individual_player_latency_milliseconds: cdktf.numberToTerraform(struct.maximumIndividualPlayerLatencyMilliseconds),
        policy_duration_seconds: cdktf.numberToTerraform(struct.policyDurationSeconds),
    };
}
exports.gameliftGameSessionQueuePlayerLatencyPolicyToTerraform = gameliftGameSessionQueuePlayerLatencyPolicyToTerraform;
function gameliftGameSessionQueuePlayerLatencyPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_individual_player_latency_milliseconds: {
            value: cdktf.numberToHclTerraform(struct.maximumIndividualPlayerLatencyMilliseconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        policy_duration_seconds: {
            value: cdktf.numberToHclTerraform(struct.policyDurationSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gameliftGameSessionQueuePlayerLatencyPolicyToHclTerraform = gameliftGameSessionQueuePlayerLatencyPolicyToHclTerraform;
class GameliftGameSessionQueuePlayerLatencyPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumIndividualPlayerLatencyMilliseconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumIndividualPlayerLatencyMilliseconds = this._maximumIndividualPlayerLatencyMilliseconds;
        }
        if (this._policyDurationSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.policyDurationSeconds = this._policyDurationSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._maximumIndividualPlayerLatencyMilliseconds = undefined;
            this._policyDurationSeconds = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._maximumIndividualPlayerLatencyMilliseconds = value.maximumIndividualPlayerLatencyMilliseconds;
            this._policyDurationSeconds = value.policyDurationSeconds;
        }
    }
    get maximumIndividualPlayerLatencyMilliseconds() {
        return this.getNumberAttribute('maximum_individual_player_latency_milliseconds');
    }
    set maximumIndividualPlayerLatencyMilliseconds(value) {
        this._maximumIndividualPlayerLatencyMilliseconds = value;
    }
    // Temporarily expose input value. Use with caution.
    get maximumIndividualPlayerLatencyMillisecondsInput() {
        return this._maximumIndividualPlayerLatencyMilliseconds;
    }
    get policyDurationSeconds() {
        return this.getNumberAttribute('policy_duration_seconds');
    }
    set policyDurationSeconds(value) {
        this._policyDurationSeconds = value;
    }
    resetPolicyDurationSeconds() {
        this._policyDurationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyDurationSecondsInput() {
        return this._policyDurationSeconds;
    }
}
exports.GameliftGameSessionQueuePlayerLatencyPolicyOutputReference = GameliftGameSessionQueuePlayerLatencyPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GameliftGameSessionQueuePlayerLatencyPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.gameliftGameSessionQueue.GameliftGameSessionQueuePlayerLatencyPolicyOutputReference", version: "19.15.0" };
class GameliftGameSessionQueuePlayerLatencyPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GameliftGameSessionQueuePlayerLatencyPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GameliftGameSessionQueuePlayerLatencyPolicyList = GameliftGameSessionQueuePlayerLatencyPolicyList;
_b = JSII_RTTI_SYMBOL_1;
GameliftGameSessionQueuePlayerLatencyPolicyList[_b] = { fqn: "@cdktf/provider-aws.gameliftGameSessionQueue.GameliftGameSessionQueuePlayerLatencyPolicyList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_session_queue aws_gamelift_game_session_queue}
*/
class GameliftGameSessionQueue extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GameliftGameSessionQueue resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GameliftGameSessionQueue to import
    * @param importFromId The id of the existing GameliftGameSessionQueue that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_session_queue#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GameliftGameSessionQueue to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_gamelift_game_session_queue", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/gamelift_game_session_queue aws_gamelift_game_session_queue} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GameliftGameSessionQueueConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_gamelift_game_session_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // player_latency_policy - computed: false, optional: true, required: false
        this._playerLatencyPolicy = new GameliftGameSessionQueuePlayerLatencyPolicyList(this, "player_latency_policy", false);
        this._customEventData = config.customEventData;
        this._destinations = config.destinations;
        this._id = config.id;
        this._name = config.name;
        this._notificationTarget = config.notificationTarget;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeoutInSeconds = config.timeoutInSeconds;
        this._playerLatencyPolicy.internalValue = config.playerLatencyPolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get customEventData() {
        return this.getStringAttribute('custom_event_data');
    }
    set customEventData(value) {
        this._customEventData = value;
    }
    resetCustomEventData() {
        this._customEventData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEventDataInput() {
        return this._customEventData;
    }
    get destinations() {
        return this.getListAttribute('destinations');
    }
    set destinations(value) {
        this._destinations = value;
    }
    resetDestinations() {
        this._destinations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationsInput() {
        return this._destinations;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get notificationTarget() {
        return this.getStringAttribute('notification_target');
    }
    set notificationTarget(value) {
        this._notificationTarget = value;
    }
    resetNotificationTarget() {
        this._notificationTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationTargetInput() {
        return this._notificationTarget;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeoutInSeconds() {
        return this.getNumberAttribute('timeout_in_seconds');
    }
    set timeoutInSeconds(value) {
        this._timeoutInSeconds = value;
    }
    resetTimeoutInSeconds() {
        this._timeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInSecondsInput() {
        return this._timeoutInSeconds;
    }
    get playerLatencyPolicy() {
        return this._playerLatencyPolicy;
    }
    putPlayerLatencyPolicy(value) {
        this._playerLatencyPolicy.internalValue = value;
    }
    resetPlayerLatencyPolicy() {
        this._playerLatencyPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get playerLatencyPolicyInput() {
        return this._playerLatencyPolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            custom_event_data: cdktf.stringToTerraform(this._customEventData),
            destinations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._destinations),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            notification_target: cdktf.stringToTerraform(this._notificationTarget),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeout_in_seconds: cdktf.numberToTerraform(this._timeoutInSeconds),
            player_latency_policy: cdktf.listMapper(gameliftGameSessionQueuePlayerLatencyPolicyToTerraform, true)(this._playerLatencyPolicy.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            custom_event_data: {
                value: cdktf.stringToHclTerraform(this._customEventData),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destinations: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._destinations),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notification_target: {
                value: cdktf.stringToHclTerraform(this._notificationTarget),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeout_in_seconds: {
                value: cdktf.numberToHclTerraform(this._timeoutInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            player_latency_policy: {
                value: cdktf.listMapperHcl(gameliftGameSessionQueuePlayerLatencyPolicyToHclTerraform, true)(this._playerLatencyPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GameliftGameSessionQueuePlayerLatencyPolicyList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GameliftGameSessionQueue = GameliftGameSessionQueue;
_c = JSII_RTTI_SYMBOL_1;
GameliftGameSessionQueue[_c] = { fqn: "@cdktf/provider-aws.gameliftGameSessionQueue.GameliftGameSessionQueue", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GameliftGameSessionQueue.tfResourceType = "aws_gamelift_game_session_queue";
