// @cdktf/provider-aws/lib/inspector-assessment-template/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorAssessmentTemplate = exports.InspectorAssessmentTemplateEventSubscriptionList = exports.InspectorAssessmentTemplateEventSubscriptionOutputReference = exports.inspectorAssessmentTemplateEventSubscriptionToHclTerraform = exports.inspectorAssessmentTemplateEventSubscriptionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function inspectorAssessmentTemplateEventSubscriptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event: cdktf.stringToTerraform(struct.event),
        topic_arn: cdktf.stringToTerraform(struct.topicArn),
    };
}
exports.inspectorAssessmentTemplateEventSubscriptionToTerraform = inspectorAssessmentTemplateEventSubscriptionToTerraform;
function inspectorAssessmentTemplateEventSubscriptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        event: {
            value: cdktf.stringToHclTerraform(struct.event),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        topic_arn: {
            value: cdktf.stringToHclTerraform(struct.topicArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.inspectorAssessmentTemplateEventSubscriptionToHclTerraform = inspectorAssessmentTemplateEventSubscriptionToHclTerraform;
class InspectorAssessmentTemplateEventSubscriptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._event !== undefined) {
            hasAnyValues = true;
            internalValueResult.event = this._event;
        }
        if (this._topicArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.topicArn = this._topicArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._event = undefined;
            this._topicArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._event = value.event;
            this._topicArn = value.topicArn;
        }
    }
    get event() {
        return this.getStringAttribute('event');
    }
    set event(value) {
        this._event = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventInput() {
        return this._event;
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
}
exports.InspectorAssessmentTemplateEventSubscriptionOutputReference = InspectorAssessmentTemplateEventSubscriptionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
InspectorAssessmentTemplateEventSubscriptionOutputReference[_a] = { fqn: "@cdktf/provider-aws.inspectorAssessmentTemplate.InspectorAssessmentTemplateEventSubscriptionOutputReference", version: "19.15.0" };
class InspectorAssessmentTemplateEventSubscriptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new InspectorAssessmentTemplateEventSubscriptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.InspectorAssessmentTemplateEventSubscriptionList = InspectorAssessmentTemplateEventSubscriptionList;
_b = JSII_RTTI_SYMBOL_1;
InspectorAssessmentTemplateEventSubscriptionList[_b] = { fqn: "@cdktf/provider-aws.inspectorAssessmentTemplate.InspectorAssessmentTemplateEventSubscriptionList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector_assessment_template aws_inspector_assessment_template}
*/
class InspectorAssessmentTemplate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a InspectorAssessmentTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the InspectorAssessmentTemplate to import
    * @param importFromId The id of the existing InspectorAssessmentTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector_assessment_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the InspectorAssessmentTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_inspector_assessment_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector_assessment_template aws_inspector_assessment_template} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options InspectorAssessmentTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_inspector_assessment_template',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // event_subscription - computed: false, optional: true, required: false
        this._eventSubscription = new InspectorAssessmentTemplateEventSubscriptionList(this, "event_subscription", true);
        this._duration = config.duration;
        this._id = config.id;
        this._name = config.name;
        this._rulesPackageArns = config.rulesPackageArns;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._targetArn = config.targetArn;
        this._eventSubscription.internalValue = config.eventSubscription;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get duration() {
        return this.getNumberAttribute('duration');
    }
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get rulesPackageArns() {
        return cdktf.Fn.tolist(this.getListAttribute('rules_package_arns'));
    }
    set rulesPackageArns(value) {
        this._rulesPackageArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get rulesPackageArnsInput() {
        return this._rulesPackageArns;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    set targetArn(value) {
        this._targetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetArnInput() {
        return this._targetArn;
    }
    get eventSubscription() {
        return this._eventSubscription;
    }
    putEventSubscription(value) {
        this._eventSubscription.internalValue = value;
    }
    resetEventSubscription() {
        this._eventSubscription.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventSubscriptionInput() {
        return this._eventSubscription.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            duration: cdktf.numberToTerraform(this._duration),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            rules_package_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._rulesPackageArns),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            target_arn: cdktf.stringToTerraform(this._targetArn),
            event_subscription: cdktf.listMapper(inspectorAssessmentTemplateEventSubscriptionToTerraform, true)(this._eventSubscription.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            duration: {
                value: cdktf.numberToHclTerraform(this._duration),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rules_package_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._rulesPackageArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            target_arn: {
                value: cdktf.stringToHclTerraform(this._targetArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            event_subscription: {
                value: cdktf.listMapperHcl(inspectorAssessmentTemplateEventSubscriptionToHclTerraform, true)(this._eventSubscription.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "InspectorAssessmentTemplateEventSubscriptionList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.InspectorAssessmentTemplate = InspectorAssessmentTemplate;
_c = JSII_RTTI_SYMBOL_1;
InspectorAssessmentTemplate[_c] = { fqn: "@cdktf/provider-aws.inspectorAssessmentTemplate.InspectorAssessmentTemplate", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
InspectorAssessmentTemplate.tfResourceType = "aws_inspector_assessment_template";
