// @cdktf/provider-aws/lib/pinpoint-app/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinpointApp = exports.PinpointAppQuietTimeOutputReference = exports.pinpointAppQuietTimeToHclTerraform = exports.pinpointAppQuietTimeToTerraform = exports.PinpointAppLimitsOutputReference = exports.pinpointAppLimitsToHclTerraform = exports.pinpointAppLimitsToTerraform = exports.PinpointAppCampaignHookOutputReference = exports.pinpointAppCampaignHookToHclTerraform = exports.pinpointAppCampaignHookToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function pinpointAppCampaignHookToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lambda_function_name: cdktf.stringToTerraform(struct.lambdaFunctionName),
        mode: cdktf.stringToTerraform(struct.mode),
        web_url: cdktf.stringToTerraform(struct.webUrl),
    };
}
exports.pinpointAppCampaignHookToTerraform = pinpointAppCampaignHookToTerraform;
function pinpointAppCampaignHookToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        lambda_function_name: {
            value: cdktf.stringToHclTerraform(struct.lambdaFunctionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        web_url: {
            value: cdktf.stringToHclTerraform(struct.webUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.pinpointAppCampaignHookToHclTerraform = pinpointAppCampaignHookToHclTerraform;
class PinpointAppCampaignHookOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lambdaFunctionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambdaFunctionName = this._lambdaFunctionName;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._webUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.webUrl = this._webUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lambdaFunctionName = undefined;
            this._mode = undefined;
            this._webUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lambdaFunctionName = value.lambdaFunctionName;
            this._mode = value.mode;
            this._webUrl = value.webUrl;
        }
    }
    get lambdaFunctionName() {
        return this.getStringAttribute('lambda_function_name');
    }
    set lambdaFunctionName(value) {
        this._lambdaFunctionName = value;
    }
    resetLambdaFunctionName() {
        this._lambdaFunctionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaFunctionNameInput() {
        return this._lambdaFunctionName;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get webUrl() {
        return this.getStringAttribute('web_url');
    }
    set webUrl(value) {
        this._webUrl = value;
    }
    resetWebUrl() {
        this._webUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get webUrlInput() {
        return this._webUrl;
    }
}
exports.PinpointAppCampaignHookOutputReference = PinpointAppCampaignHookOutputReference;
_a = JSII_RTTI_SYMBOL_1;
PinpointAppCampaignHookOutputReference[_a] = { fqn: "@cdktf/provider-aws.pinpointApp.PinpointAppCampaignHookOutputReference", version: "19.15.0" };
function pinpointAppLimitsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        daily: cdktf.numberToTerraform(struct.daily),
        maximum_duration: cdktf.numberToTerraform(struct.maximumDuration),
        messages_per_second: cdktf.numberToTerraform(struct.messagesPerSecond),
        total: cdktf.numberToTerraform(struct.total),
    };
}
exports.pinpointAppLimitsToTerraform = pinpointAppLimitsToTerraform;
function pinpointAppLimitsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        daily: {
            value: cdktf.numberToHclTerraform(struct.daily),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        maximum_duration: {
            value: cdktf.numberToHclTerraform(struct.maximumDuration),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        messages_per_second: {
            value: cdktf.numberToHclTerraform(struct.messagesPerSecond),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        total: {
            value: cdktf.numberToHclTerraform(struct.total),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.pinpointAppLimitsToHclTerraform = pinpointAppLimitsToHclTerraform;
class PinpointAppLimitsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daily !== undefined) {
            hasAnyValues = true;
            internalValueResult.daily = this._daily;
        }
        if (this._maximumDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumDuration = this._maximumDuration;
        }
        if (this._messagesPerSecond !== undefined) {
            hasAnyValues = true;
            internalValueResult.messagesPerSecond = this._messagesPerSecond;
        }
        if (this._total !== undefined) {
            hasAnyValues = true;
            internalValueResult.total = this._total;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._daily = undefined;
            this._maximumDuration = undefined;
            this._messagesPerSecond = undefined;
            this._total = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._daily = value.daily;
            this._maximumDuration = value.maximumDuration;
            this._messagesPerSecond = value.messagesPerSecond;
            this._total = value.total;
        }
    }
    get daily() {
        return this.getNumberAttribute('daily');
    }
    set daily(value) {
        this._daily = value;
    }
    resetDaily() {
        this._daily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dailyInput() {
        return this._daily;
    }
    get maximumDuration() {
        return this.getNumberAttribute('maximum_duration');
    }
    set maximumDuration(value) {
        this._maximumDuration = value;
    }
    resetMaximumDuration() {
        this._maximumDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumDurationInput() {
        return this._maximumDuration;
    }
    get messagesPerSecond() {
        return this.getNumberAttribute('messages_per_second');
    }
    set messagesPerSecond(value) {
        this._messagesPerSecond = value;
    }
    resetMessagesPerSecond() {
        this._messagesPerSecond = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messagesPerSecondInput() {
        return this._messagesPerSecond;
    }
    get total() {
        return this.getNumberAttribute('total');
    }
    set total(value) {
        this._total = value;
    }
    resetTotal() {
        this._total = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get totalInput() {
        return this._total;
    }
}
exports.PinpointAppLimitsOutputReference = PinpointAppLimitsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
PinpointAppLimitsOutputReference[_b] = { fqn: "@cdktf/provider-aws.pinpointApp.PinpointAppLimitsOutputReference", version: "19.15.0" };
function pinpointAppQuietTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end: cdktf.stringToTerraform(struct.end),
        start: cdktf.stringToTerraform(struct.start),
    };
}
exports.pinpointAppQuietTimeToTerraform = pinpointAppQuietTimeToTerraform;
function pinpointAppQuietTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        end: {
            value: cdktf.stringToHclTerraform(struct.end),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start: {
            value: cdktf.stringToHclTerraform(struct.start),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.pinpointAppQuietTimeToHclTerraform = pinpointAppQuietTimeToHclTerraform;
class PinpointAppQuietTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._end !== undefined) {
            hasAnyValues = true;
            internalValueResult.end = this._end;
        }
        if (this._start !== undefined) {
            hasAnyValues = true;
            internalValueResult.start = this._start;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._end = undefined;
            this._start = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._end = value.end;
            this._start = value.start;
        }
    }
    get end() {
        return this.getStringAttribute('end');
    }
    set end(value) {
        this._end = value;
    }
    resetEnd() {
        this._end = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endInput() {
        return this._end;
    }
    get start() {
        return this.getStringAttribute('start');
    }
    set start(value) {
        this._start = value;
    }
    resetStart() {
        this._start = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startInput() {
        return this._start;
    }
}
exports.PinpointAppQuietTimeOutputReference = PinpointAppQuietTimeOutputReference;
_c = JSII_RTTI_SYMBOL_1;
PinpointAppQuietTimeOutputReference[_c] = { fqn: "@cdktf/provider-aws.pinpointApp.PinpointAppQuietTimeOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/pinpoint_app aws_pinpoint_app}
*/
class PinpointApp extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a PinpointApp resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the PinpointApp to import
    * @param importFromId The id of the existing PinpointApp that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/pinpoint_app#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the PinpointApp to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_pinpoint_app", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/pinpoint_app aws_pinpoint_app} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options PinpointAppConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_pinpoint_app',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // campaign_hook - computed: false, optional: true, required: false
        this._campaignHook = new PinpointAppCampaignHookOutputReference(this, "campaign_hook");
        // limits - computed: false, optional: true, required: false
        this._limits = new PinpointAppLimitsOutputReference(this, "limits");
        // quiet_time - computed: false, optional: true, required: false
        this._quietTime = new PinpointAppQuietTimeOutputReference(this, "quiet_time");
        this._id = config.id;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._campaignHook.internalValue = config.campaignHook;
        this._limits.internalValue = config.limits;
        this._quietTime.internalValue = config.quietTime;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // application_id - computed: true, optional: false, required: false
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get campaignHook() {
        return this._campaignHook;
    }
    putCampaignHook(value) {
        this._campaignHook.internalValue = value;
    }
    resetCampaignHook() {
        this._campaignHook.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get campaignHookInput() {
        return this._campaignHook.internalValue;
    }
    get limits() {
        return this._limits;
    }
    putLimits(value) {
        this._limits.internalValue = value;
    }
    resetLimits() {
        this._limits.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get limitsInput() {
        return this._limits.internalValue;
    }
    get quietTime() {
        return this._quietTime;
    }
    putQuietTime(value) {
        this._quietTime.internalValue = value;
    }
    resetQuietTime() {
        this._quietTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get quietTimeInput() {
        return this._quietTime.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            campaign_hook: pinpointAppCampaignHookToTerraform(this._campaignHook.internalValue),
            limits: pinpointAppLimitsToTerraform(this._limits.internalValue),
            quiet_time: pinpointAppQuietTimeToTerraform(this._quietTime.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            campaign_hook: {
                value: pinpointAppCampaignHookToHclTerraform(this._campaignHook.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "PinpointAppCampaignHookList",
            },
            limits: {
                value: pinpointAppLimitsToHclTerraform(this._limits.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "PinpointAppLimitsList",
            },
            quiet_time: {
                value: pinpointAppQuietTimeToHclTerraform(this._quietTime.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "PinpointAppQuietTimeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.PinpointApp = PinpointApp;
_d = JSII_RTTI_SYMBOL_1;
PinpointApp[_d] = { fqn: "@cdktf/provider-aws.pinpointApp.PinpointApp", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
PinpointApp.tfResourceType = "aws_pinpoint_app";
