// @winglang/sdk/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.util = exports.ui = exports.sim = exports.std = exports.simulator = exports.aws = exports.platform = exports.math = exports.http = exports.helpers = exports.fs = exports.expect = exports.ex = exports.core = exports.cloud = void 0;
// this file must only export the standard library types and not target-specific types.
exports.cloud = __importStar(require("./cloud"));
exports.core = __importStar(require("./core"));
exports.ex = __importStar(require("./ex"));
exports.expect = __importStar(require("./expect"));
exports.fs = __importStar(require("./fs"));
exports.helpers = __importStar(require("./helpers"));
exports.http = __importStar(require("./http"));
exports.math = __importStar(require("./math"));
exports.platform = __importStar(require("./platform"));
exports.aws = __importStar(require("./shared-aws"));
exports.simulator = __importStar(require("./simulator"));
exports.std = __importStar(require("./std"));
exports.sim = __importStar(require("./target-sim"));
exports.ui = __importStar(require("./ui"));
exports.util = __importStar(require("./util"));
