// @cdktf/provider-aws/lib/fis-experiment-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FisExperimentTemplate = exports.FisExperimentTemplateTimeoutsOutputReference = exports.fisExperimentTemplateTimeoutsToHclTerraform = exports.fisExperimentTemplateTimeoutsToTerraform = exports.FisExperimentTemplateTargetList = exports.FisExperimentTemplateTargetOutputReference = exports.fisExperimentTemplateTargetToHclTerraform = exports.fisExperimentTemplateTargetToTerraform = exports.FisExperimentTemplateTargetResourceTagList = exports.FisExperimentTemplateTargetResourceTagOutputReference = exports.fisExperimentTemplateTargetResourceTagToHclTerraform = exports.fisExperimentTemplateTargetResourceTagToTerraform = exports.FisExperimentTemplateTargetFilterList = exports.FisExperimentTemplateTargetFilterOutputReference = exports.fisExperimentTemplateTargetFilterToHclTerraform = exports.fisExperimentTemplateTargetFilterToTerraform = exports.FisExperimentTemplateStopConditionList = exports.FisExperimentTemplateStopConditionOutputReference = exports.fisExperimentTemplateStopConditionToHclTerraform = exports.fisExperimentTemplateStopConditionToTerraform = exports.FisExperimentTemplateLogConfigurationOutputReference = exports.fisExperimentTemplateLogConfigurationToHclTerraform = exports.fisExperimentTemplateLogConfigurationToTerraform = exports.FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference = exports.fisExperimentTemplateLogConfigurationS3ConfigurationToHclTerraform = exports.fisExperimentTemplateLogConfigurationS3ConfigurationToTerraform = exports.FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference = exports.fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToHclTerraform = exports.fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToTerraform = exports.FisExperimentTemplateActionList = exports.FisExperimentTemplateActionOutputReference = exports.fisExperimentTemplateActionToHclTerraform = exports.fisExperimentTemplateActionToTerraform = exports.FisExperimentTemplateActionTargetOutputReference = exports.fisExperimentTemplateActionTargetToHclTerraform = exports.fisExperimentTemplateActionTargetToTerraform = exports.FisExperimentTemplateActionParameterList = exports.FisExperimentTemplateActionParameterOutputReference = exports.fisExperimentTemplateActionParameterToHclTerraform = exports.fisExperimentTemplateActionParameterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fisExperimentTemplateActionParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.fisExperimentTemplateActionParameterToTerraform = fisExperimentTemplateActionParameterToTerraform;
function fisExperimentTemplateActionParameterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateActionParameterToHclTerraform = fisExperimentTemplateActionParameterToHclTerraform;
class FisExperimentTemplateActionParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FisExperimentTemplateActionParameterOutputReference = FisExperimentTemplateActionParameterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateActionParameterOutputReference[_a] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateActionParameterOutputReference", version: "19.15.0" };
class FisExperimentTemplateActionParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateActionParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateActionParameterList = FisExperimentTemplateActionParameterList;
_b = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateActionParameterList[_b] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateActionParameterList", version: "19.15.0" };
function fisExperimentTemplateActionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.fisExperimentTemplateActionTargetToTerraform = fisExperimentTemplateActionTargetToTerraform;
function fisExperimentTemplateActionTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateActionTargetToHclTerraform = fisExperimentTemplateActionTargetToHclTerraform;
class FisExperimentTemplateActionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FisExperimentTemplateActionTargetOutputReference = FisExperimentTemplateActionTargetOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateActionTargetOutputReference[_c] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateActionTargetOutputReference", version: "19.15.0" };
function fisExperimentTemplateActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_id: cdktf.stringToTerraform(struct.actionId),
        description: cdktf.stringToTerraform(struct.description),
        name: cdktf.stringToTerraform(struct.name),
        start_after: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.startAfter),
        parameter: cdktf.listMapper(fisExperimentTemplateActionParameterToTerraform, true)(struct.parameter),
        target: fisExperimentTemplateActionTargetToTerraform(struct.target),
    };
}
exports.fisExperimentTemplateActionToTerraform = fisExperimentTemplateActionToTerraform;
function fisExperimentTemplateActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action_id: {
            value: cdktf.stringToHclTerraform(struct.actionId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_after: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.startAfter),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        parameter: {
            value: cdktf.listMapperHcl(fisExperimentTemplateActionParameterToHclTerraform, true)(struct.parameter),
            isBlock: true,
            type: "set",
            storageClassType: "FisExperimentTemplateActionParameterList",
        },
        target: {
            value: fisExperimentTemplateActionTargetToHclTerraform(struct.target),
            isBlock: true,
            type: "list",
            storageClassType: "FisExperimentTemplateActionTargetList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateActionToHclTerraform = fisExperimentTemplateActionToHclTerraform;
class FisExperimentTemplateActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameter - computed: false, optional: true, required: false
        this._parameter = new FisExperimentTemplateActionParameterList(this, "parameter", true);
        // target - computed: false, optional: true, required: false
        this._target = new FisExperimentTemplateActionTargetOutputReference(this, "target");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionId !== undefined) {
            hasAnyValues = true;
            internalValueResult.actionId = this._actionId;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._startAfter !== undefined) {
            hasAnyValues = true;
            internalValueResult.startAfter = this._startAfter;
        }
        if (this._parameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter?.internalValue;
        }
        if (this._target?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._actionId = undefined;
            this._description = undefined;
            this._name = undefined;
            this._startAfter = undefined;
            this._parameter.internalValue = undefined;
            this._target.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._actionId = value.actionId;
            this._description = value.description;
            this._name = value.name;
            this._startAfter = value.startAfter;
            this._parameter.internalValue = value.parameter;
            this._target.internalValue = value.target;
        }
    }
    get actionId() {
        return this.getStringAttribute('action_id');
    }
    set actionId(value) {
        this._actionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionIdInput() {
        return this._actionId;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get startAfter() {
        return cdktf.Fn.tolist(this.getListAttribute('start_after'));
    }
    set startAfter(value) {
        this._startAfter = value;
    }
    resetStartAfter() {
        this._startAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startAfterInput() {
        return this._startAfter;
    }
    get parameter() {
        return this._parameter;
    }
    putParameter(value) {
        this._parameter.internalValue = value;
    }
    resetParameter() {
        this._parameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterInput() {
        return this._parameter.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    resetTarget() {
        this._target.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
}
exports.FisExperimentTemplateActionOutputReference = FisExperimentTemplateActionOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateActionOutputReference[_d] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateActionOutputReference", version: "19.15.0" };
class FisExperimentTemplateActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateActionList = FisExperimentTemplateActionList;
_e = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateActionList[_e] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateActionList", version: "19.15.0" };
function fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_group_arn: cdktf.stringToTerraform(struct.logGroupArn),
    };
}
exports.fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToTerraform = fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToTerraform;
function fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        log_group_arn: {
            value: cdktf.stringToHclTerraform(struct.logGroupArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToHclTerraform = fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToHclTerraform;
class FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logGroupArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.logGroupArn = this._logGroupArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logGroupArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logGroupArn = value.logGroupArn;
        }
    }
    get logGroupArn() {
        return this.getStringAttribute('log_group_arn');
    }
    set logGroupArn(value) {
        this._logGroupArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupArnInput() {
        return this._logGroupArn;
    }
}
exports.FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference = FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference[_f] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference", version: "19.15.0" };
function fisExperimentTemplateLogConfigurationS3ConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.fisExperimentTemplateLogConfigurationS3ConfigurationToTerraform = fisExperimentTemplateLogConfigurationS3ConfigurationToTerraform;
function fisExperimentTemplateLogConfigurationS3ConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_name: {
            value: cdktf.stringToHclTerraform(struct.bucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateLogConfigurationS3ConfigurationToHclTerraform = fisExperimentTemplateLogConfigurationS3ConfigurationToHclTerraform;
class FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._prefix = value.prefix;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
}
exports.FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference = FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference", version: "19.15.0" };
function fisExperimentTemplateLogConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_schema_version: cdktf.numberToTerraform(struct.logSchemaVersion),
        cloudwatch_logs_configuration: fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToTerraform(struct.cloudwatchLogsConfiguration),
        s3_configuration: fisExperimentTemplateLogConfigurationS3ConfigurationToTerraform(struct.s3Configuration),
    };
}
exports.fisExperimentTemplateLogConfigurationToTerraform = fisExperimentTemplateLogConfigurationToTerraform;
function fisExperimentTemplateLogConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        log_schema_version: {
            value: cdktf.numberToHclTerraform(struct.logSchemaVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        cloudwatch_logs_configuration: {
            value: fisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationToHclTerraform(struct.cloudwatchLogsConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationList",
        },
        s3_configuration: {
            value: fisExperimentTemplateLogConfigurationS3ConfigurationToHclTerraform(struct.s3Configuration),
            isBlock: true,
            type: "list",
            storageClassType: "FisExperimentTemplateLogConfigurationS3ConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateLogConfigurationToHclTerraform = fisExperimentTemplateLogConfigurationToHclTerraform;
class FisExperimentTemplateLogConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cloudwatch_logs_configuration - computed: false, optional: true, required: false
        this._cloudwatchLogsConfiguration = new FisExperimentTemplateLogConfigurationCloudwatchLogsConfigurationOutputReference(this, "cloudwatch_logs_configuration");
        // s3_configuration - computed: false, optional: true, required: false
        this._s3Configuration = new FisExperimentTemplateLogConfigurationS3ConfigurationOutputReference(this, "s3_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logSchemaVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.logSchemaVersion = this._logSchemaVersion;
        }
        if (this._cloudwatchLogsConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogsConfiguration = this._cloudwatchLogsConfiguration?.internalValue;
        }
        if (this._s3Configuration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Configuration = this._s3Configuration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logSchemaVersion = undefined;
            this._cloudwatchLogsConfiguration.internalValue = undefined;
            this._s3Configuration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logSchemaVersion = value.logSchemaVersion;
            this._cloudwatchLogsConfiguration.internalValue = value.cloudwatchLogsConfiguration;
            this._s3Configuration.internalValue = value.s3Configuration;
        }
    }
    get logSchemaVersion() {
        return this.getNumberAttribute('log_schema_version');
    }
    set logSchemaVersion(value) {
        this._logSchemaVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get logSchemaVersionInput() {
        return this._logSchemaVersion;
    }
    get cloudwatchLogsConfiguration() {
        return this._cloudwatchLogsConfiguration;
    }
    putCloudwatchLogsConfiguration(value) {
        this._cloudwatchLogsConfiguration.internalValue = value;
    }
    resetCloudwatchLogsConfiguration() {
        this._cloudwatchLogsConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogsConfigurationInput() {
        return this._cloudwatchLogsConfiguration.internalValue;
    }
    get s3Configuration() {
        return this._s3Configuration;
    }
    putS3Configuration(value) {
        this._s3Configuration.internalValue = value;
    }
    resetS3Configuration() {
        this._s3Configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3ConfigurationInput() {
        return this._s3Configuration.internalValue;
    }
}
exports.FisExperimentTemplateLogConfigurationOutputReference = FisExperimentTemplateLogConfigurationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateLogConfigurationOutputReference[_h] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateLogConfigurationOutputReference", version: "19.15.0" };
function fisExperimentTemplateStopConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source: cdktf.stringToTerraform(struct.source),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.fisExperimentTemplateStopConditionToTerraform = fisExperimentTemplateStopConditionToTerraform;
function fisExperimentTemplateStopConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source: {
            value: cdktf.stringToHclTerraform(struct.source),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateStopConditionToHclTerraform = fisExperimentTemplateStopConditionToHclTerraform;
class FisExperimentTemplateStopConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._source = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._source = value.source;
            this._value = value.value;
        }
    }
    get source() {
        return this.getStringAttribute('source');
    }
    set source(value) {
        this._source = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FisExperimentTemplateStopConditionOutputReference = FisExperimentTemplateStopConditionOutputReference;
_j = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateStopConditionOutputReference[_j] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateStopConditionOutputReference", version: "19.15.0" };
class FisExperimentTemplateStopConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateStopConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateStopConditionList = FisExperimentTemplateStopConditionList;
_k = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateStopConditionList[_k] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateStopConditionList", version: "19.15.0" };
function fisExperimentTemplateTargetFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.fisExperimentTemplateTargetFilterToTerraform = fisExperimentTemplateTargetFilterToTerraform;
function fisExperimentTemplateTargetFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateTargetFilterToHclTerraform = fisExperimentTemplateTargetFilterToHclTerraform;
class FisExperimentTemplateTargetFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._path = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._path = value.path;
            this._values = value.values;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.FisExperimentTemplateTargetFilterOutputReference = FisExperimentTemplateTargetFilterOutputReference;
_l = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetFilterOutputReference[_l] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetFilterOutputReference", version: "19.15.0" };
class FisExperimentTemplateTargetFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateTargetFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateTargetFilterList = FisExperimentTemplateTargetFilterList;
_m = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetFilterList[_m] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetFilterList", version: "19.15.0" };
function fisExperimentTemplateTargetResourceTagToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.fisExperimentTemplateTargetResourceTagToTerraform = fisExperimentTemplateTargetResourceTagToTerraform;
function fisExperimentTemplateTargetResourceTagToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateTargetResourceTagToHclTerraform = fisExperimentTemplateTargetResourceTagToHclTerraform;
class FisExperimentTemplateTargetResourceTagOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FisExperimentTemplateTargetResourceTagOutputReference = FisExperimentTemplateTargetResourceTagOutputReference;
_o = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetResourceTagOutputReference[_o] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetResourceTagOutputReference", version: "19.15.0" };
class FisExperimentTemplateTargetResourceTagList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateTargetResourceTagOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateTargetResourceTagList = FisExperimentTemplateTargetResourceTagList;
_p = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetResourceTagList[_p] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetResourceTagList", version: "19.15.0" };
function fisExperimentTemplateTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.parameters),
        resource_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resourceArns),
        resource_type: cdktf.stringToTerraform(struct.resourceType),
        selection_mode: cdktf.stringToTerraform(struct.selectionMode),
        filter: cdktf.listMapper(fisExperimentTemplateTargetFilterToTerraform, true)(struct.filter),
        resource_tag: cdktf.listMapper(fisExperimentTemplateTargetResourceTagToTerraform, true)(struct.resourceTag),
    };
}
exports.fisExperimentTemplateTargetToTerraform = fisExperimentTemplateTargetToTerraform;
function fisExperimentTemplateTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.parameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        resource_arns: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resourceArns),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        resource_type: {
            value: cdktf.stringToHclTerraform(struct.resourceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        selection_mode: {
            value: cdktf.stringToHclTerraform(struct.selectionMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        filter: {
            value: cdktf.listMapperHcl(fisExperimentTemplateTargetFilterToHclTerraform, true)(struct.filter),
            isBlock: true,
            type: "list",
            storageClassType: "FisExperimentTemplateTargetFilterList",
        },
        resource_tag: {
            value: cdktf.listMapperHcl(fisExperimentTemplateTargetResourceTagToHclTerraform, true)(struct.resourceTag),
            isBlock: true,
            type: "set",
            storageClassType: "FisExperimentTemplateTargetResourceTagList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateTargetToHclTerraform = fisExperimentTemplateTargetToHclTerraform;
class FisExperimentTemplateTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // filter - computed: false, optional: true, required: false
        this._filter = new FisExperimentTemplateTargetFilterList(this, "filter", false);
        // resource_tag - computed: false, optional: true, required: false
        this._resourceTag = new FisExperimentTemplateTargetResourceTagList(this, "resource_tag", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._parameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._resourceArns !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArns = this._resourceArns;
        }
        if (this._resourceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceType = this._resourceType;
        }
        if (this._selectionMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.selectionMode = this._selectionMode;
        }
        if (this._filter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter?.internalValue;
        }
        if (this._resourceTag?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceTag = this._resourceTag?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._parameters = undefined;
            this._resourceArns = undefined;
            this._resourceType = undefined;
            this._selectionMode = undefined;
            this._filter.internalValue = undefined;
            this._resourceTag.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._parameters = value.parameters;
            this._resourceArns = value.resourceArns;
            this._resourceType = value.resourceType;
            this._selectionMode = value.selectionMode;
            this._filter.internalValue = value.filter;
            this._resourceTag.internalValue = value.resourceTag;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get resourceArns() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_arns'));
    }
    set resourceArns(value) {
        this._resourceArns = value;
    }
    resetResourceArns() {
        this._resourceArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnsInput() {
        return this._resourceArns;
    }
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    set resourceType(value) {
        this._resourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypeInput() {
        return this._resourceType;
    }
    get selectionMode() {
        return this.getStringAttribute('selection_mode');
    }
    set selectionMode(value) {
        this._selectionMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get selectionModeInput() {
        return this._selectionMode;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get resourceTag() {
        return this._resourceTag;
    }
    putResourceTag(value) {
        this._resourceTag.internalValue = value;
    }
    resetResourceTag() {
        this._resourceTag.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTagInput() {
        return this._resourceTag.internalValue;
    }
}
exports.FisExperimentTemplateTargetOutputReference = FisExperimentTemplateTargetOutputReference;
_q = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetOutputReference[_q] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetOutputReference", version: "19.15.0" };
class FisExperimentTemplateTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FisExperimentTemplateTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FisExperimentTemplateTargetList = FisExperimentTemplateTargetList;
_r = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTargetList[_r] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTargetList", version: "19.15.0" };
function fisExperimentTemplateTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fisExperimentTemplateTimeoutsToTerraform = fisExperimentTemplateTimeoutsToTerraform;
function fisExperimentTemplateTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fisExperimentTemplateTimeoutsToHclTerraform = fisExperimentTemplateTimeoutsToHclTerraform;
class FisExperimentTemplateTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FisExperimentTemplateTimeoutsOutputReference = FisExperimentTemplateTimeoutsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
FisExperimentTemplateTimeoutsOutputReference[_s] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplateTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fis_experiment_template aws_fis_experiment_template}
*/
class FisExperimentTemplate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FisExperimentTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FisExperimentTemplate to import
    * @param importFromId The id of the existing FisExperimentTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fis_experiment_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FisExperimentTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fis_experiment_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fis_experiment_template aws_fis_experiment_template} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FisExperimentTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fis_experiment_template',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // action - computed: false, optional: false, required: true
        this._action = new FisExperimentTemplateActionList(this, "action", true);
        // log_configuration - computed: false, optional: true, required: false
        this._logConfiguration = new FisExperimentTemplateLogConfigurationOutputReference(this, "log_configuration");
        // stop_condition - computed: false, optional: false, required: true
        this._stopCondition = new FisExperimentTemplateStopConditionList(this, "stop_condition", true);
        // target - computed: false, optional: true, required: false
        this._target = new FisExperimentTemplateTargetList(this, "target", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FisExperimentTemplateTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._action.internalValue = config.action;
        this._logConfiguration.internalValue = config.logConfiguration;
        this._stopCondition.internalValue = config.stopCondition;
        this._target.internalValue = config.target;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get logConfiguration() {
        return this._logConfiguration;
    }
    putLogConfiguration(value) {
        this._logConfiguration.internalValue = value;
    }
    resetLogConfiguration() {
        this._logConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logConfigurationInput() {
        return this._logConfiguration.internalValue;
    }
    get stopCondition() {
        return this._stopCondition;
    }
    putStopCondition(value) {
        this._stopCondition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stopConditionInput() {
        return this._stopCondition.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    resetTarget() {
        this._target.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            action: cdktf.listMapper(fisExperimentTemplateActionToTerraform, true)(this._action.internalValue),
            log_configuration: fisExperimentTemplateLogConfigurationToTerraform(this._logConfiguration.internalValue),
            stop_condition: cdktf.listMapper(fisExperimentTemplateStopConditionToTerraform, true)(this._stopCondition.internalValue),
            target: cdktf.listMapper(fisExperimentTemplateTargetToTerraform, true)(this._target.internalValue),
            timeouts: fisExperimentTemplateTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            action: {
                value: cdktf.listMapperHcl(fisExperimentTemplateActionToHclTerraform, true)(this._action.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "FisExperimentTemplateActionList",
            },
            log_configuration: {
                value: fisExperimentTemplateLogConfigurationToHclTerraform(this._logConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FisExperimentTemplateLogConfigurationList",
            },
            stop_condition: {
                value: cdktf.listMapperHcl(fisExperimentTemplateStopConditionToHclTerraform, true)(this._stopCondition.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "FisExperimentTemplateStopConditionList",
            },
            target: {
                value: cdktf.listMapperHcl(fisExperimentTemplateTargetToHclTerraform, true)(this._target.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "FisExperimentTemplateTargetList",
            },
            timeouts: {
                value: fisExperimentTemplateTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FisExperimentTemplateTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FisExperimentTemplate = FisExperimentTemplate;
_t = JSII_RTTI_SYMBOL_1;
FisExperimentTemplate[_t] = { fqn: "@cdktf/provider-aws.fisExperimentTemplate.FisExperimentTemplate", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FisExperimentTemplate.tfResourceType = "aws_fis_experiment_template";
