// @cdktf/provider-aws/lib/lakeformation-data-cells-filter/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LakeformationDataCellsFilter = exports.LakeformationDataCellsFilterTimeoutsOutputReference = exports.lakeformationDataCellsFilterTimeoutsToHclTerraform = exports.lakeformationDataCellsFilterTimeoutsToTerraform = exports.LakeformationDataCellsFilterTableDataList = exports.LakeformationDataCellsFilterTableDataOutputReference = exports.lakeformationDataCellsFilterTableDataToHclTerraform = exports.lakeformationDataCellsFilterTableDataToTerraform = exports.LakeformationDataCellsFilterTableDataRowFilterList = exports.LakeformationDataCellsFilterTableDataRowFilterOutputReference = exports.lakeformationDataCellsFilterTableDataRowFilterToHclTerraform = exports.lakeformationDataCellsFilterTableDataRowFilterToTerraform = exports.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList = exports.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference = exports.lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToHclTerraform = exports.lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToTerraform = exports.LakeformationDataCellsFilterTableDataColumnWildcardList = exports.LakeformationDataCellsFilterTableDataColumnWildcardOutputReference = exports.lakeformationDataCellsFilterTableDataColumnWildcardToHclTerraform = exports.lakeformationDataCellsFilterTableDataColumnWildcardToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lakeformationDataCellsFilterTableDataColumnWildcardToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excluded_column_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludedColumnNames),
    };
}
exports.lakeformationDataCellsFilterTableDataColumnWildcardToTerraform = lakeformationDataCellsFilterTableDataColumnWildcardToTerraform;
function lakeformationDataCellsFilterTableDataColumnWildcardToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        excluded_column_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludedColumnNames),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lakeformationDataCellsFilterTableDataColumnWildcardToHclTerraform = lakeformationDataCellsFilterTableDataColumnWildcardToHclTerraform;
class LakeformationDataCellsFilterTableDataColumnWildcardOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludedColumnNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludedColumnNames = this._excludedColumnNames;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._excludedColumnNames = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._excludedColumnNames = value.excludedColumnNames;
        }
    }
    get excludedColumnNames() {
        return this.getListAttribute('excluded_column_names');
    }
    set excludedColumnNames(value) {
        this._excludedColumnNames = value;
    }
    resetExcludedColumnNames() {
        this._excludedColumnNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludedColumnNamesInput() {
        return this._excludedColumnNames;
    }
}
exports.LakeformationDataCellsFilterTableDataColumnWildcardOutputReference = LakeformationDataCellsFilterTableDataColumnWildcardOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataColumnWildcardOutputReference[_a] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataColumnWildcardOutputReference", version: "19.15.0" };
class LakeformationDataCellsFilterTableDataColumnWildcardList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LakeformationDataCellsFilterTableDataColumnWildcardOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LakeformationDataCellsFilterTableDataColumnWildcardList = LakeformationDataCellsFilterTableDataColumnWildcardList;
_b = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataColumnWildcardList[_b] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataColumnWildcardList", version: "19.15.0" };
function lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToTerraform = lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToTerraform;
function lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToHclTerraform = lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToHclTerraform;
class LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
}
exports.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference = LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference[_c] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference", version: "19.15.0" };
class LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList = LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList;
_d = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList[_d] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList", version: "19.15.0" };
function lakeformationDataCellsFilterTableDataRowFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter_expression: cdktf.stringToTerraform(struct.filterExpression),
        all_rows_wildcard: cdktf.listMapper(lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToTerraform, true)(struct.allRowsWildcard),
    };
}
exports.lakeformationDataCellsFilterTableDataRowFilterToTerraform = lakeformationDataCellsFilterTableDataRowFilterToTerraform;
function lakeformationDataCellsFilterTableDataRowFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter_expression: {
            value: cdktf.stringToHclTerraform(struct.filterExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        all_rows_wildcard: {
            value: cdktf.listMapperHcl(lakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardToHclTerraform, true)(struct.allRowsWildcard),
            isBlock: true,
            type: "list",
            storageClassType: "LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lakeformationDataCellsFilterTableDataRowFilterToHclTerraform = lakeformationDataCellsFilterTableDataRowFilterToHclTerraform;
class LakeformationDataCellsFilterTableDataRowFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // all_rows_wildcard - computed: false, optional: true, required: false
        this._allRowsWildcard = new LakeformationDataCellsFilterTableDataRowFilterAllRowsWildcardList(this, "all_rows_wildcard", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filterExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterExpression = this._filterExpression;
        }
        if (this._allRowsWildcard?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.allRowsWildcard = this._allRowsWildcard?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._filterExpression = undefined;
            this._allRowsWildcard.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._filterExpression = value.filterExpression;
            this._allRowsWildcard.internalValue = value.allRowsWildcard;
        }
    }
    get filterExpression() {
        return this.getStringAttribute('filter_expression');
    }
    set filterExpression(value) {
        this._filterExpression = value;
    }
    resetFilterExpression() {
        this._filterExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterExpressionInput() {
        return this._filterExpression;
    }
    get allRowsWildcard() {
        return this._allRowsWildcard;
    }
    putAllRowsWildcard(value) {
        this._allRowsWildcard.internalValue = value;
    }
    resetAllRowsWildcard() {
        this._allRowsWildcard.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allRowsWildcardInput() {
        return this._allRowsWildcard.internalValue;
    }
}
exports.LakeformationDataCellsFilterTableDataRowFilterOutputReference = LakeformationDataCellsFilterTableDataRowFilterOutputReference;
_e = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataRowFilterOutputReference[_e] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataRowFilterOutputReference", version: "19.15.0" };
class LakeformationDataCellsFilterTableDataRowFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LakeformationDataCellsFilterTableDataRowFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LakeformationDataCellsFilterTableDataRowFilterList = LakeformationDataCellsFilterTableDataRowFilterList;
_f = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataRowFilterList[_f] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataRowFilterList", version: "19.15.0" };
function lakeformationDataCellsFilterTableDataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        column_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.columnNames),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        name: cdktf.stringToTerraform(struct.name),
        table_catalog_id: cdktf.stringToTerraform(struct.tableCatalogId),
        table_name: cdktf.stringToTerraform(struct.tableName),
        version_id: cdktf.stringToTerraform(struct.versionId),
        column_wildcard: cdktf.listMapper(lakeformationDataCellsFilterTableDataColumnWildcardToTerraform, true)(struct.columnWildcard),
        row_filter: cdktf.listMapper(lakeformationDataCellsFilterTableDataRowFilterToTerraform, true)(struct.rowFilter),
    };
}
exports.lakeformationDataCellsFilterTableDataToTerraform = lakeformationDataCellsFilterTableDataToTerraform;
function lakeformationDataCellsFilterTableDataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        column_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.columnNames),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        database_name: {
            value: cdktf.stringToHclTerraform(struct.databaseName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        table_catalog_id: {
            value: cdktf.stringToHclTerraform(struct.tableCatalogId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        table_name: {
            value: cdktf.stringToHclTerraform(struct.tableName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        version_id: {
            value: cdktf.stringToHclTerraform(struct.versionId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        column_wildcard: {
            value: cdktf.listMapperHcl(lakeformationDataCellsFilterTableDataColumnWildcardToHclTerraform, true)(struct.columnWildcard),
            isBlock: true,
            type: "list",
            storageClassType: "LakeformationDataCellsFilterTableDataColumnWildcardList",
        },
        row_filter: {
            value: cdktf.listMapperHcl(lakeformationDataCellsFilterTableDataRowFilterToHclTerraform, true)(struct.rowFilter),
            isBlock: true,
            type: "list",
            storageClassType: "LakeformationDataCellsFilterTableDataRowFilterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lakeformationDataCellsFilterTableDataToHclTerraform = lakeformationDataCellsFilterTableDataToHclTerraform;
class LakeformationDataCellsFilterTableDataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // column_wildcard - computed: false, optional: true, required: false
        this._columnWildcard = new LakeformationDataCellsFilterTableDataColumnWildcardList(this, "column_wildcard", false);
        // row_filter - computed: false, optional: true, required: false
        this._rowFilter = new LakeformationDataCellsFilterTableDataRowFilterList(this, "row_filter", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._columnNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.columnNames = this._columnNames;
        }
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._tableCatalogId !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableCatalogId = this._tableCatalogId;
        }
        if (this._tableName !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        if (this._versionId !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionId = this._versionId;
        }
        if (this._columnWildcard?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.columnWildcard = this._columnWildcard?.internalValue;
        }
        if (this._rowFilter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rowFilter = this._rowFilter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._columnNames = undefined;
            this._databaseName = undefined;
            this._name = undefined;
            this._tableCatalogId = undefined;
            this._tableName = undefined;
            this._versionId = undefined;
            this._columnWildcard.internalValue = undefined;
            this._rowFilter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._columnNames = value.columnNames;
            this._databaseName = value.databaseName;
            this._name = value.name;
            this._tableCatalogId = value.tableCatalogId;
            this._tableName = value.tableName;
            this._versionId = value.versionId;
            this._columnWildcard.internalValue = value.columnWildcard;
            this._rowFilter.internalValue = value.rowFilter;
        }
    }
    get columnNames() {
        return cdktf.Fn.tolist(this.getListAttribute('column_names'));
    }
    set columnNames(value) {
        this._columnNames = value;
    }
    resetColumnNames() {
        this._columnNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get columnNamesInput() {
        return this._columnNames;
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tableCatalogId() {
        return this.getStringAttribute('table_catalog_id');
    }
    set tableCatalogId(value) {
        this._tableCatalogId = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableCatalogIdInput() {
        return this._tableCatalogId;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    set versionId(value) {
        this._versionId = value;
    }
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionIdInput() {
        return this._versionId;
    }
    get columnWildcard() {
        return this._columnWildcard;
    }
    putColumnWildcard(value) {
        this._columnWildcard.internalValue = value;
    }
    resetColumnWildcard() {
        this._columnWildcard.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get columnWildcardInput() {
        return this._columnWildcard.internalValue;
    }
    get rowFilter() {
        return this._rowFilter;
    }
    putRowFilter(value) {
        this._rowFilter.internalValue = value;
    }
    resetRowFilter() {
        this._rowFilter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rowFilterInput() {
        return this._rowFilter.internalValue;
    }
}
exports.LakeformationDataCellsFilterTableDataOutputReference = LakeformationDataCellsFilterTableDataOutputReference;
_g = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataOutputReference[_g] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataOutputReference", version: "19.15.0" };
class LakeformationDataCellsFilterTableDataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LakeformationDataCellsFilterTableDataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LakeformationDataCellsFilterTableDataList = LakeformationDataCellsFilterTableDataList;
_h = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTableDataList[_h] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTableDataList", version: "19.15.0" };
function lakeformationDataCellsFilterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.lakeformationDataCellsFilterTimeoutsToTerraform = lakeformationDataCellsFilterTimeoutsToTerraform;
function lakeformationDataCellsFilterTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lakeformationDataCellsFilterTimeoutsToHclTerraform = lakeformationDataCellsFilterTimeoutsToHclTerraform;
class LakeformationDataCellsFilterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.LakeformationDataCellsFilterTimeoutsOutputReference = LakeformationDataCellsFilterTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilterTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilterTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_data_cells_filter aws_lakeformation_data_cells_filter}
*/
class LakeformationDataCellsFilter extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LakeformationDataCellsFilter resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LakeformationDataCellsFilter to import
    * @param importFromId The id of the existing LakeformationDataCellsFilter that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_data_cells_filter#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LakeformationDataCellsFilter to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lakeformation_data_cells_filter", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_data_cells_filter aws_lakeformation_data_cells_filter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LakeformationDataCellsFilterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_lakeformation_data_cells_filter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // table_data - computed: false, optional: true, required: false
        this._tableData = new LakeformationDataCellsFilterTableDataList(this, "table_data", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LakeformationDataCellsFilterTimeoutsOutputReference(this, "timeouts");
        this._tableData.internalValue = config.tableData;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get tableData() {
        return this._tableData;
    }
    putTableData(value) {
        this._tableData.internalValue = value;
    }
    resetTableData() {
        this._tableData.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tableDataInput() {
        return this._tableData.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            table_data: cdktf.listMapper(lakeformationDataCellsFilterTableDataToTerraform, true)(this._tableData.internalValue),
            timeouts: lakeformationDataCellsFilterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            table_data: {
                value: cdktf.listMapperHcl(lakeformationDataCellsFilterTableDataToHclTerraform, true)(this._tableData.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LakeformationDataCellsFilterTableDataList",
            },
            timeouts: {
                value: lakeformationDataCellsFilterTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "LakeformationDataCellsFilterTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LakeformationDataCellsFilter = LakeformationDataCellsFilter;
_k = JSII_RTTI_SYMBOL_1;
LakeformationDataCellsFilter[_k] = { fqn: "@cdktf/provider-aws.lakeformationDataCellsFilter.LakeformationDataCellsFilter", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LakeformationDataCellsFilter.tfResourceType = "aws_lakeformation_data_cells_filter";
