// @cdktf/provider-aws/lib/storagegateway-smb-file-share/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoragegatewaySmbFileShare = exports.StoragegatewaySmbFileShareTimeoutsOutputReference = exports.storagegatewaySmbFileShareTimeoutsToHclTerraform = exports.storagegatewaySmbFileShareTimeoutsToTerraform = exports.StoragegatewaySmbFileShareCacheAttributesOutputReference = exports.storagegatewaySmbFileShareCacheAttributesToHclTerraform = exports.storagegatewaySmbFileShareCacheAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function storagegatewaySmbFileShareCacheAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cache_stale_timeout_in_seconds: cdktf.numberToTerraform(struct.cacheStaleTimeoutInSeconds),
    };
}
exports.storagegatewaySmbFileShareCacheAttributesToTerraform = storagegatewaySmbFileShareCacheAttributesToTerraform;
function storagegatewaySmbFileShareCacheAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cache_stale_timeout_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.cacheStaleTimeoutInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storagegatewaySmbFileShareCacheAttributesToHclTerraform = storagegatewaySmbFileShareCacheAttributesToHclTerraform;
class StoragegatewaySmbFileShareCacheAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cacheStaleTimeoutInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.cacheStaleTimeoutInSeconds = this._cacheStaleTimeoutInSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cacheStaleTimeoutInSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cacheStaleTimeoutInSeconds = value.cacheStaleTimeoutInSeconds;
        }
    }
    get cacheStaleTimeoutInSeconds() {
        return this.getNumberAttribute('cache_stale_timeout_in_seconds');
    }
    set cacheStaleTimeoutInSeconds(value) {
        this._cacheStaleTimeoutInSeconds = value;
    }
    resetCacheStaleTimeoutInSeconds() {
        this._cacheStaleTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cacheStaleTimeoutInSecondsInput() {
        return this._cacheStaleTimeoutInSeconds;
    }
}
exports.StoragegatewaySmbFileShareCacheAttributesOutputReference = StoragegatewaySmbFileShareCacheAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
StoragegatewaySmbFileShareCacheAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.storagegatewaySmbFileShare.StoragegatewaySmbFileShareCacheAttributesOutputReference", version: "19.15.0" };
function storagegatewaySmbFileShareTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.storagegatewaySmbFileShareTimeoutsToTerraform = storagegatewaySmbFileShareTimeoutsToTerraform;
function storagegatewaySmbFileShareTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storagegatewaySmbFileShareTimeoutsToHclTerraform = storagegatewaySmbFileShareTimeoutsToHclTerraform;
class StoragegatewaySmbFileShareTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.StoragegatewaySmbFileShareTimeoutsOutputReference = StoragegatewaySmbFileShareTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
StoragegatewaySmbFileShareTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.storagegatewaySmbFileShare.StoragegatewaySmbFileShareTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/storagegateway_smb_file_share aws_storagegateway_smb_file_share}
*/
class StoragegatewaySmbFileShare extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a StoragegatewaySmbFileShare resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the StoragegatewaySmbFileShare to import
    * @param importFromId The id of the existing StoragegatewaySmbFileShare that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/storagegateway_smb_file_share#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the StoragegatewaySmbFileShare to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_storagegateway_smb_file_share", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/storagegateway_smb_file_share aws_storagegateway_smb_file_share} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options StoragegatewaySmbFileShareConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_storagegateway_smb_file_share',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cache_attributes - computed: false, optional: true, required: false
        this._cacheAttributes = new StoragegatewaySmbFileShareCacheAttributesOutputReference(this, "cache_attributes");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new StoragegatewaySmbFileShareTimeoutsOutputReference(this, "timeouts");
        this._accessBasedEnumeration = config.accessBasedEnumeration;
        this._adminUserList = config.adminUserList;
        this._auditDestinationArn = config.auditDestinationArn;
        this._authentication = config.authentication;
        this._bucketRegion = config.bucketRegion;
        this._caseSensitivity = config.caseSensitivity;
        this._defaultStorageClass = config.defaultStorageClass;
        this._fileShareName = config.fileShareName;
        this._gatewayArn = config.gatewayArn;
        this._guessMimeTypeEnabled = config.guessMimeTypeEnabled;
        this._id = config.id;
        this._invalidUserList = config.invalidUserList;
        this._kmsEncrypted = config.kmsEncrypted;
        this._kmsKeyArn = config.kmsKeyArn;
        this._locationArn = config.locationArn;
        this._notificationPolicy = config.notificationPolicy;
        this._objectAcl = config.objectAcl;
        this._oplocksEnabled = config.oplocksEnabled;
        this._readOnly = config.readOnly;
        this._requesterPays = config.requesterPays;
        this._roleArn = config.roleArn;
        this._smbAclEnabled = config.smbAclEnabled;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._validUserList = config.validUserList;
        this._vpcEndpointDnsName = config.vpcEndpointDnsName;
        this._cacheAttributes.internalValue = config.cacheAttributes;
        this._timeouts.internalValue = config.timeouts;
    }
    get accessBasedEnumeration() {
        return this.getBooleanAttribute('access_based_enumeration');
    }
    set accessBasedEnumeration(value) {
        this._accessBasedEnumeration = value;
    }
    resetAccessBasedEnumeration() {
        this._accessBasedEnumeration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessBasedEnumerationInput() {
        return this._accessBasedEnumeration;
    }
    get adminUserList() {
        return cdktf.Fn.tolist(this.getListAttribute('admin_user_list'));
    }
    set adminUserList(value) {
        this._adminUserList = value;
    }
    resetAdminUserList() {
        this._adminUserList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUserListInput() {
        return this._adminUserList;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get auditDestinationArn() {
        return this.getStringAttribute('audit_destination_arn');
    }
    set auditDestinationArn(value) {
        this._auditDestinationArn = value;
    }
    resetAuditDestinationArn() {
        this._auditDestinationArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditDestinationArnInput() {
        return this._auditDestinationArn;
    }
    get authentication() {
        return this.getStringAttribute('authentication');
    }
    set authentication(value) {
        this._authentication = value;
    }
    resetAuthentication() {
        this._authentication = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationInput() {
        return this._authentication;
    }
    get bucketRegion() {
        return this.getStringAttribute('bucket_region');
    }
    set bucketRegion(value) {
        this._bucketRegion = value;
    }
    resetBucketRegion() {
        this._bucketRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketRegionInput() {
        return this._bucketRegion;
    }
    get caseSensitivity() {
        return this.getStringAttribute('case_sensitivity');
    }
    set caseSensitivity(value) {
        this._caseSensitivity = value;
    }
    resetCaseSensitivity() {
        this._caseSensitivity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get caseSensitivityInput() {
        return this._caseSensitivity;
    }
    get defaultStorageClass() {
        return this.getStringAttribute('default_storage_class');
    }
    set defaultStorageClass(value) {
        this._defaultStorageClass = value;
    }
    resetDefaultStorageClass() {
        this._defaultStorageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultStorageClassInput() {
        return this._defaultStorageClass;
    }
    get fileShareName() {
        return this.getStringAttribute('file_share_name');
    }
    set fileShareName(value) {
        this._fileShareName = value;
    }
    resetFileShareName() {
        this._fileShareName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileShareNameInput() {
        return this._fileShareName;
    }
    // fileshare_id - computed: true, optional: false, required: false
    get fileshareId() {
        return this.getStringAttribute('fileshare_id');
    }
    get gatewayArn() {
        return this.getStringAttribute('gateway_arn');
    }
    set gatewayArn(value) {
        this._gatewayArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get gatewayArnInput() {
        return this._gatewayArn;
    }
    get guessMimeTypeEnabled() {
        return this.getBooleanAttribute('guess_mime_type_enabled');
    }
    set guessMimeTypeEnabled(value) {
        this._guessMimeTypeEnabled = value;
    }
    resetGuessMimeTypeEnabled() {
        this._guessMimeTypeEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get guessMimeTypeEnabledInput() {
        return this._guessMimeTypeEnabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get invalidUserList() {
        return cdktf.Fn.tolist(this.getListAttribute('invalid_user_list'));
    }
    set invalidUserList(value) {
        this._invalidUserList = value;
    }
    resetInvalidUserList() {
        this._invalidUserList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invalidUserListInput() {
        return this._invalidUserList;
    }
    get kmsEncrypted() {
        return this.getBooleanAttribute('kms_encrypted');
    }
    set kmsEncrypted(value) {
        this._kmsEncrypted = value;
    }
    resetKmsEncrypted() {
        this._kmsEncrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsEncryptedInput() {
        return this._kmsEncrypted;
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    get locationArn() {
        return this.getStringAttribute('location_arn');
    }
    set locationArn(value) {
        this._locationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationArnInput() {
        return this._locationArn;
    }
    get notificationPolicy() {
        return this.getStringAttribute('notification_policy');
    }
    set notificationPolicy(value) {
        this._notificationPolicy = value;
    }
    resetNotificationPolicy() {
        this._notificationPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationPolicyInput() {
        return this._notificationPolicy;
    }
    get objectAcl() {
        return this.getStringAttribute('object_acl');
    }
    set objectAcl(value) {
        this._objectAcl = value;
    }
    resetObjectAcl() {
        this._objectAcl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get objectAclInput() {
        return this._objectAcl;
    }
    get oplocksEnabled() {
        return this.getBooleanAttribute('oplocks_enabled');
    }
    set oplocksEnabled(value) {
        this._oplocksEnabled = value;
    }
    resetOplocksEnabled() {
        this._oplocksEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oplocksEnabledInput() {
        return this._oplocksEnabled;
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    set readOnly(value) {
        this._readOnly = value;
    }
    resetReadOnly() {
        this._readOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readOnlyInput() {
        return this._readOnly;
    }
    get requesterPays() {
        return this.getBooleanAttribute('requester_pays');
    }
    set requesterPays(value) {
        this._requesterPays = value;
    }
    resetRequesterPays() {
        this._requesterPays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requesterPaysInput() {
        return this._requesterPays;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get smbAclEnabled() {
        return this.getBooleanAttribute('smb_acl_enabled');
    }
    set smbAclEnabled(value) {
        this._smbAclEnabled = value;
    }
    resetSmbAclEnabled() {
        this._smbAclEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get smbAclEnabledInput() {
        return this._smbAclEnabled;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get validUserList() {
        return cdktf.Fn.tolist(this.getListAttribute('valid_user_list'));
    }
    set validUserList(value) {
        this._validUserList = value;
    }
    resetValidUserList() {
        this._validUserList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validUserListInput() {
        return this._validUserList;
    }
    get vpcEndpointDnsName() {
        return this.getStringAttribute('vpc_endpoint_dns_name');
    }
    set vpcEndpointDnsName(value) {
        this._vpcEndpointDnsName = value;
    }
    resetVpcEndpointDnsName() {
        this._vpcEndpointDnsName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointDnsNameInput() {
        return this._vpcEndpointDnsName;
    }
    get cacheAttributes() {
        return this._cacheAttributes;
    }
    putCacheAttributes(value) {
        this._cacheAttributes.internalValue = value;
    }
    resetCacheAttributes() {
        this._cacheAttributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cacheAttributesInput() {
        return this._cacheAttributes.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_based_enumeration: cdktf.booleanToTerraform(this._accessBasedEnumeration),
            admin_user_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._adminUserList),
            audit_destination_arn: cdktf.stringToTerraform(this._auditDestinationArn),
            authentication: cdktf.stringToTerraform(this._authentication),
            bucket_region: cdktf.stringToTerraform(this._bucketRegion),
            case_sensitivity: cdktf.stringToTerraform(this._caseSensitivity),
            default_storage_class: cdktf.stringToTerraform(this._defaultStorageClass),
            file_share_name: cdktf.stringToTerraform(this._fileShareName),
            gateway_arn: cdktf.stringToTerraform(this._gatewayArn),
            guess_mime_type_enabled: cdktf.booleanToTerraform(this._guessMimeTypeEnabled),
            id: cdktf.stringToTerraform(this._id),
            invalid_user_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._invalidUserList),
            kms_encrypted: cdktf.booleanToTerraform(this._kmsEncrypted),
            kms_key_arn: cdktf.stringToTerraform(this._kmsKeyArn),
            location_arn: cdktf.stringToTerraform(this._locationArn),
            notification_policy: cdktf.stringToTerraform(this._notificationPolicy),
            object_acl: cdktf.stringToTerraform(this._objectAcl),
            oplocks_enabled: cdktf.booleanToTerraform(this._oplocksEnabled),
            read_only: cdktf.booleanToTerraform(this._readOnly),
            requester_pays: cdktf.booleanToTerraform(this._requesterPays),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            smb_acl_enabled: cdktf.booleanToTerraform(this._smbAclEnabled),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            valid_user_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._validUserList),
            vpc_endpoint_dns_name: cdktf.stringToTerraform(this._vpcEndpointDnsName),
            cache_attributes: storagegatewaySmbFileShareCacheAttributesToTerraform(this._cacheAttributes.internalValue),
            timeouts: storagegatewaySmbFileShareTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_based_enumeration: {
                value: cdktf.booleanToHclTerraform(this._accessBasedEnumeration),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            admin_user_list: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._adminUserList),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            audit_destination_arn: {
                value: cdktf.stringToHclTerraform(this._auditDestinationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authentication: {
                value: cdktf.stringToHclTerraform(this._authentication),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bucket_region: {
                value: cdktf.stringToHclTerraform(this._bucketRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            case_sensitivity: {
                value: cdktf.stringToHclTerraform(this._caseSensitivity),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            default_storage_class: {
                value: cdktf.stringToHclTerraform(this._defaultStorageClass),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            file_share_name: {
                value: cdktf.stringToHclTerraform(this._fileShareName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gateway_arn: {
                value: cdktf.stringToHclTerraform(this._gatewayArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            guess_mime_type_enabled: {
                value: cdktf.booleanToHclTerraform(this._guessMimeTypeEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            invalid_user_list: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._invalidUserList),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            kms_encrypted: {
                value: cdktf.booleanToHclTerraform(this._kmsEncrypted),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            kms_key_arn: {
                value: cdktf.stringToHclTerraform(this._kmsKeyArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location_arn: {
                value: cdktf.stringToHclTerraform(this._locationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notification_policy: {
                value: cdktf.stringToHclTerraform(this._notificationPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            object_acl: {
                value: cdktf.stringToHclTerraform(this._objectAcl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            oplocks_enabled: {
                value: cdktf.booleanToHclTerraform(this._oplocksEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            read_only: {
                value: cdktf.booleanToHclTerraform(this._readOnly),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            requester_pays: {
                value: cdktf.booleanToHclTerraform(this._requesterPays),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            smb_acl_enabled: {
                value: cdktf.booleanToHclTerraform(this._smbAclEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            valid_user_list: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._validUserList),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            vpc_endpoint_dns_name: {
                value: cdktf.stringToHclTerraform(this._vpcEndpointDnsName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cache_attributes: {
                value: storagegatewaySmbFileShareCacheAttributesToHclTerraform(this._cacheAttributes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "StoragegatewaySmbFileShareCacheAttributesList",
            },
            timeouts: {
                value: storagegatewaySmbFileShareTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "StoragegatewaySmbFileShareTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.StoragegatewaySmbFileShare = StoragegatewaySmbFileShare;
_c = JSII_RTTI_SYMBOL_1;
StoragegatewaySmbFileShare[_c] = { fqn: "@cdktf/provider-aws.storagegatewaySmbFileShare.StoragegatewaySmbFileShare", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
StoragegatewaySmbFileShare.tfResourceType = "aws_storagegateway_smb_file_share";
