// @winglang/sdk/lib/shared/resource-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceNames = exports.CaseConventions = void 0;
var CaseConventions;
(function (CaseConventions) {
    CaseConventions["LOWERCASE"] = "lowercase";
    CaseConventions["UPPERCASE"] = "uppercase";
})(CaseConventions || (exports.CaseConventions = CaseConventions = {}));
class ResourceNames {
    static generateName(resource, props) {
        const sep = props.sep ?? "-";
        const maxLen = props.maxLen;
        if (maxLen && maxLen < 8) {
            throw new Error("maxLen must be at least 8");
        }
        let name = resource.node.id;
        name = applyCaseConversion(name, props.case);
        if (props.prefix) {
            name = `${props.prefix}${name}`;
        }
        let includeHash = props.includeHash ?? true;
        let hash = includeHash ? sep + resource.node.addr.substring(0, 8) : "";
        let suffix = props.suffix ?? "";
        name = name.replace(props.disallowedRegex, sep);
        // TODO: allow customizing where we "trim" the name, e.g.
        // 1. trim from the end
        // 2. trim from the beginning
        // 3. trim from the middle
        if (maxLen) {
            name = name.substring(0, maxLen - hash.length - suffix.length);
        }
        name = `${name}${hash}${suffix}`;
        // apply case conversion again in case the prefix, suffix, or hash is not
        // case-conformant
        name = applyCaseConversion(name, props.case);
        return name;
    }
}
exports.ResourceNames = ResourceNames;
function applyCaseConversion(name, caseConventions) {
    if (caseConventions === CaseConventions.LOWERCASE) {
        return name.toLocaleLowerCase();
    }
    if (caseConventions === CaseConventions.UPPERCASE) {
        return name.toLocaleUpperCase();
    }
    return name;
}
