// @cdktf/provider-aws/lib/rolesanywhere-trust-anchor/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesanywhereTrustAnchor = exports.RolesanywhereTrustAnchorSourceOutputReference = exports.rolesanywhereTrustAnchorSourceToHclTerraform = exports.rolesanywhereTrustAnchorSourceToTerraform = exports.RolesanywhereTrustAnchorSourceSourceDataOutputReference = exports.rolesanywhereTrustAnchorSourceSourceDataToHclTerraform = exports.rolesanywhereTrustAnchorSourceSourceDataToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function rolesanywhereTrustAnchorSourceSourceDataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        acm_pca_arn: cdktf.stringToTerraform(struct.acmPcaArn),
        x509_certificate_data: cdktf.stringToTerraform(struct.x509CertificateData),
    };
}
exports.rolesanywhereTrustAnchorSourceSourceDataToTerraform = rolesanywhereTrustAnchorSourceSourceDataToTerraform;
function rolesanywhereTrustAnchorSourceSourceDataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        acm_pca_arn: {
            value: cdktf.stringToHclTerraform(struct.acmPcaArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        x509_certificate_data: {
            value: cdktf.stringToHclTerraform(struct.x509CertificateData),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rolesanywhereTrustAnchorSourceSourceDataToHclTerraform = rolesanywhereTrustAnchorSourceSourceDataToHclTerraform;
class RolesanywhereTrustAnchorSourceSourceDataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._acmPcaArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.acmPcaArn = this._acmPcaArn;
        }
        if (this._x509CertificateData !== undefined) {
            hasAnyValues = true;
            internalValueResult.x509CertificateData = this._x509CertificateData;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._acmPcaArn = undefined;
            this._x509CertificateData = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._acmPcaArn = value.acmPcaArn;
            this._x509CertificateData = value.x509CertificateData;
        }
    }
    get acmPcaArn() {
        return this.getStringAttribute('acm_pca_arn');
    }
    set acmPcaArn(value) {
        this._acmPcaArn = value;
    }
    resetAcmPcaArn() {
        this._acmPcaArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acmPcaArnInput() {
        return this._acmPcaArn;
    }
    get x509CertificateData() {
        return this.getStringAttribute('x509_certificate_data');
    }
    set x509CertificateData(value) {
        this._x509CertificateData = value;
    }
    resetX509CertificateData() {
        this._x509CertificateData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get x509CertificateDataInput() {
        return this._x509CertificateData;
    }
}
exports.RolesanywhereTrustAnchorSourceSourceDataOutputReference = RolesanywhereTrustAnchorSourceSourceDataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RolesanywhereTrustAnchorSourceSourceDataOutputReference[_a] = { fqn: "@cdktf/provider-aws.rolesanywhereTrustAnchor.RolesanywhereTrustAnchorSourceSourceDataOutputReference", version: "19.15.0" };
function rolesanywhereTrustAnchorSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_type: cdktf.stringToTerraform(struct.sourceType),
        source_data: rolesanywhereTrustAnchorSourceSourceDataToTerraform(struct.sourceData),
    };
}
exports.rolesanywhereTrustAnchorSourceToTerraform = rolesanywhereTrustAnchorSourceToTerraform;
function rolesanywhereTrustAnchorSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_type: {
            value: cdktf.stringToHclTerraform(struct.sourceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_data: {
            value: rolesanywhereTrustAnchorSourceSourceDataToHclTerraform(struct.sourceData),
            isBlock: true,
            type: "list",
            storageClassType: "RolesanywhereTrustAnchorSourceSourceDataList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.rolesanywhereTrustAnchorSourceToHclTerraform = rolesanywhereTrustAnchorSourceToHclTerraform;
class RolesanywhereTrustAnchorSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // source_data - computed: false, optional: false, required: true
        this._sourceData = new RolesanywhereTrustAnchorSourceSourceDataOutputReference(this, "source_data");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceType = this._sourceType;
        }
        if (this._sourceData?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceData = this._sourceData?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sourceType = undefined;
            this._sourceData.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sourceType = value.sourceType;
            this._sourceData.internalValue = value.sourceData;
        }
    }
    get sourceType() {
        return this.getStringAttribute('source_type');
    }
    set sourceType(value) {
        this._sourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceTypeInput() {
        return this._sourceType;
    }
    get sourceData() {
        return this._sourceData;
    }
    putSourceData(value) {
        this._sourceData.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDataInput() {
        return this._sourceData.internalValue;
    }
}
exports.RolesanywhereTrustAnchorSourceOutputReference = RolesanywhereTrustAnchorSourceOutputReference;
_b = JSII_RTTI_SYMBOL_1;
RolesanywhereTrustAnchorSourceOutputReference[_b] = { fqn: "@cdktf/provider-aws.rolesanywhereTrustAnchor.RolesanywhereTrustAnchorSourceOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_trust_anchor aws_rolesanywhere_trust_anchor}
*/
class RolesanywhereTrustAnchor extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RolesanywhereTrustAnchor resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RolesanywhereTrustAnchor to import
    * @param importFromId The id of the existing RolesanywhereTrustAnchor that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_trust_anchor#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RolesanywhereTrustAnchor to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_rolesanywhere_trust_anchor", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_trust_anchor aws_rolesanywhere_trust_anchor} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RolesanywhereTrustAnchorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rolesanywhere_trust_anchor',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // source - computed: false, optional: false, required: true
        this._source = new RolesanywhereTrustAnchorSourceOutputReference(this, "source");
        this._enabled = config.enabled;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._source.internalValue = config.source;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get source() {
        return this._source;
    }
    putSource(value) {
        this._source.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            source: rolesanywhereTrustAnchorSourceToTerraform(this._source.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            source: {
                value: rolesanywhereTrustAnchorSourceToHclTerraform(this._source.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RolesanywhereTrustAnchorSourceList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RolesanywhereTrustAnchor = RolesanywhereTrustAnchor;
_c = JSII_RTTI_SYMBOL_1;
RolesanywhereTrustAnchor[_c] = { fqn: "@cdktf/provider-aws.rolesanywhereTrustAnchor.RolesanywhereTrustAnchor", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RolesanywhereTrustAnchor.tfResourceType = "aws_rolesanywhere_trust_anchor";
