// @cdktf/provider-aws/lib/data-aws-glue-catalog-table/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsGlueCatalogTable = exports.DataAwsGlueCatalogTableTargetTableList = exports.DataAwsGlueCatalogTableTargetTableOutputReference = exports.dataAwsGlueCatalogTableTargetTableToHclTerraform = exports.dataAwsGlueCatalogTableTargetTableToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorList = exports.DataAwsGlueCatalogTableStorageDescriptorOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorSortColumnsList = exports.DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorSortColumnsToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorSortColumnsToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList = exports.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList = exports.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList = exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList = exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToTerraform = exports.DataAwsGlueCatalogTableStorageDescriptorColumnsList = exports.DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference = exports.dataAwsGlueCatalogTableStorageDescriptorColumnsToHclTerraform = exports.dataAwsGlueCatalogTableStorageDescriptorColumnsToTerraform = exports.DataAwsGlueCatalogTablePartitionKeysList = exports.DataAwsGlueCatalogTablePartitionKeysOutputReference = exports.dataAwsGlueCatalogTablePartitionKeysToHclTerraform = exports.dataAwsGlueCatalogTablePartitionKeysToTerraform = exports.DataAwsGlueCatalogTablePartitionIndexList = exports.DataAwsGlueCatalogTablePartitionIndexOutputReference = exports.dataAwsGlueCatalogTablePartitionIndexToHclTerraform = exports.dataAwsGlueCatalogTablePartitionIndexToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsGlueCatalogTablePartitionIndexToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTablePartitionIndexToTerraform = dataAwsGlueCatalogTablePartitionIndexToTerraform;
function dataAwsGlueCatalogTablePartitionIndexToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTablePartitionIndexToHclTerraform = dataAwsGlueCatalogTablePartitionIndexToHclTerraform;
class DataAwsGlueCatalogTablePartitionIndexOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // index_name - computed: true, optional: false, required: false
    get indexName() {
        return this.getStringAttribute('index_name');
    }
    // index_status - computed: true, optional: false, required: false
    get indexStatus() {
        return this.getStringAttribute('index_status');
    }
    // keys - computed: true, optional: false, required: false
    get keys() {
        return this.getListAttribute('keys');
    }
}
exports.DataAwsGlueCatalogTablePartitionIndexOutputReference = DataAwsGlueCatalogTablePartitionIndexOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTablePartitionIndexOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTablePartitionIndexOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTablePartitionIndexList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTablePartitionIndexOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTablePartitionIndexList = DataAwsGlueCatalogTablePartitionIndexList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTablePartitionIndexList[_b] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTablePartitionIndexList", version: "19.15.0" };
function dataAwsGlueCatalogTablePartitionKeysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTablePartitionKeysToTerraform = dataAwsGlueCatalogTablePartitionKeysToTerraform;
function dataAwsGlueCatalogTablePartitionKeysToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTablePartitionKeysToHclTerraform = dataAwsGlueCatalogTablePartitionKeysToHclTerraform;
class DataAwsGlueCatalogTablePartitionKeysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // comment - computed: true, optional: false, required: false
    get comment() {
        return this.getStringAttribute('comment');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsGlueCatalogTablePartitionKeysOutputReference = DataAwsGlueCatalogTablePartitionKeysOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTablePartitionKeysOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTablePartitionKeysOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTablePartitionKeysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTablePartitionKeysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTablePartitionKeysList = DataAwsGlueCatalogTablePartitionKeysList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTablePartitionKeysList[_d] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTablePartitionKeysList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorColumnsToTerraform = dataAwsGlueCatalogTableStorageDescriptorColumnsToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorColumnsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorColumnsToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorColumnsToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameters - computed: true, optional: false, required: false
        this._parameters = new cdktf.StringMap(this, "parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // comment - computed: true, optional: false, required: false
    get comment() {
        return this.getStringAttribute('comment');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parameters() {
        return this._parameters;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference = DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorColumnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorColumnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorColumnsList = DataAwsGlueCatalogTableStorageDescriptorColumnsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorColumnsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorColumnsList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToTerraform = dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // registry_name - computed: true, optional: false, required: false
    get registryName() {
        return this.getStringAttribute('registry_name');
    }
    // schema_arn - computed: true, optional: false, required: false
    get schemaArn() {
        return this.getStringAttribute('schema_arn');
    }
    // schema_name - computed: true, optional: false, required: false
    get schemaName() {
        return this.getStringAttribute('schema_name');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference = DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList = DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList[_h] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToTerraform = dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorSchemaReferenceToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // schema_id - computed: true, optional: false, required: false
        this._schemaId = new DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceSchemaIdList(this, "schema_id", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get schemaId() {
        return this._schemaId;
    }
    // schema_version_id - computed: true, optional: false, required: false
    get schemaVersionId() {
        return this.getStringAttribute('schema_version_id');
    }
    // schema_version_number - computed: true, optional: false, required: false
    get schemaVersionNumber() {
        return this.getNumberAttribute('schema_version_number');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference = DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList = DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList[_k] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToTerraform = dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorSerDeInfoToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameters - computed: true, optional: false, required: false
        this._parameters = new cdktf.StringMap(this, "parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parameters() {
        return this._parameters;
    }
    // serialization_library - computed: true, optional: false, required: false
    get serializationLibrary() {
        return this.getStringAttribute('serialization_library');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference = DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorSerDeInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList = DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList[_m] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToTerraform = dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorSkewedInfoToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // skewed_column_value_location_maps - computed: true, optional: false, required: false
        this._skewedColumnValueLocationMaps = new cdktf.StringMap(this, "skewed_column_value_location_maps");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // skewed_column_names - computed: true, optional: false, required: false
    get skewedColumnNames() {
        return this.getListAttribute('skewed_column_names');
    }
    get skewedColumnValueLocationMaps() {
        return this._skewedColumnValueLocationMaps;
    }
    // skewed_column_values - computed: true, optional: false, required: false
    get skewedColumnValues() {
        return this.getListAttribute('skewed_column_values');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference = DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorSkewedInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList = DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList[_p] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorSortColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorSortColumnsToTerraform = dataAwsGlueCatalogTableStorageDescriptorSortColumnsToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorSortColumnsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorSortColumnsToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorSortColumnsToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // column - computed: true, optional: false, required: false
    get column() {
        return this.getStringAttribute('column');
    }
    // sort_order - computed: true, optional: false, required: false
    get sortOrder() {
        return this.getNumberAttribute('sort_order');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference = DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorSortColumnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorSortColumnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorSortColumnsList = DataAwsGlueCatalogTableStorageDescriptorSortColumnsList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorSortColumnsList[_r] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorSortColumnsList", version: "19.15.0" };
function dataAwsGlueCatalogTableStorageDescriptorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableStorageDescriptorToTerraform = dataAwsGlueCatalogTableStorageDescriptorToTerraform;
function dataAwsGlueCatalogTableStorageDescriptorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableStorageDescriptorToHclTerraform = dataAwsGlueCatalogTableStorageDescriptorToHclTerraform;
class DataAwsGlueCatalogTableStorageDescriptorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // columns - computed: true, optional: false, required: false
        this._columns = new DataAwsGlueCatalogTableStorageDescriptorColumnsList(this, "columns", false);
        // parameters - computed: true, optional: false, required: false
        this._parameters = new cdktf.StringMap(this, "parameters");
        // schema_reference - computed: true, optional: false, required: false
        this._schemaReference = new DataAwsGlueCatalogTableStorageDescriptorSchemaReferenceList(this, "schema_reference", false);
        // ser_de_info - computed: true, optional: false, required: false
        this._serDeInfo = new DataAwsGlueCatalogTableStorageDescriptorSerDeInfoList(this, "ser_de_info", false);
        // skewed_info - computed: true, optional: false, required: false
        this._skewedInfo = new DataAwsGlueCatalogTableStorageDescriptorSkewedInfoList(this, "skewed_info", false);
        // sort_columns - computed: true, optional: false, required: false
        this._sortColumns = new DataAwsGlueCatalogTableStorageDescriptorSortColumnsList(this, "sort_columns", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket_columns - computed: true, optional: false, required: false
    get bucketColumns() {
        return this.getListAttribute('bucket_columns');
    }
    get columns() {
        return this._columns;
    }
    // compressed - computed: true, optional: false, required: false
    get compressed() {
        return this.getBooleanAttribute('compressed');
    }
    // input_format - computed: true, optional: false, required: false
    get inputFormat() {
        return this.getStringAttribute('input_format');
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // number_of_buckets - computed: true, optional: false, required: false
    get numberOfBuckets() {
        return this.getNumberAttribute('number_of_buckets');
    }
    // output_format - computed: true, optional: false, required: false
    get outputFormat() {
        return this.getStringAttribute('output_format');
    }
    get parameters() {
        return this._parameters;
    }
    get schemaReference() {
        return this._schemaReference;
    }
    get serDeInfo() {
        return this._serDeInfo;
    }
    get skewedInfo() {
        return this._skewedInfo;
    }
    get sortColumns() {
        return this._sortColumns;
    }
    // stored_as_sub_directories - computed: true, optional: false, required: false
    get storedAsSubDirectories() {
        return this.getBooleanAttribute('stored_as_sub_directories');
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorOutputReference = DataAwsGlueCatalogTableStorageDescriptorOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableStorageDescriptorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableStorageDescriptorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableStorageDescriptorList = DataAwsGlueCatalogTableStorageDescriptorList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableStorageDescriptorList[_t] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableStorageDescriptorList", version: "19.15.0" };
function dataAwsGlueCatalogTableTargetTableToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlueCatalogTableTargetTableToTerraform = dataAwsGlueCatalogTableTargetTableToTerraform;
function dataAwsGlueCatalogTableTargetTableToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsGlueCatalogTableTargetTableToHclTerraform = dataAwsGlueCatalogTableTargetTableToHclTerraform;
class DataAwsGlueCatalogTableTargetTableOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // catalog_id - computed: true, optional: false, required: false
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsGlueCatalogTableTargetTableOutputReference = DataAwsGlueCatalogTableTargetTableOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableTargetTableOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableTargetTableOutputReference", version: "19.15.0" };
class DataAwsGlueCatalogTableTargetTableList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlueCatalogTableTargetTableOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlueCatalogTableTargetTableList = DataAwsGlueCatalogTableTargetTableList;
_v = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTableTargetTableList[_v] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTableTargetTableList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/glue_catalog_table aws_glue_catalog_table}
*/
class DataAwsGlueCatalogTable extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsGlueCatalogTable resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsGlueCatalogTable to import
    * @param importFromId The id of the existing DataAwsGlueCatalogTable that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/glue_catalog_table#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsGlueCatalogTable to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_glue_catalog_table", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/glue_catalog_table aws_glue_catalog_table} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsGlueCatalogTableConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_catalog_table',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // parameters - computed: true, optional: false, required: false
        this._parameters = new cdktf.StringMap(this, "parameters");
        // partition_index - computed: true, optional: false, required: false
        this._partitionIndex = new DataAwsGlueCatalogTablePartitionIndexList(this, "partition_index", false);
        // partition_keys - computed: true, optional: false, required: false
        this._partitionKeys = new DataAwsGlueCatalogTablePartitionKeysList(this, "partition_keys", false);
        // storage_descriptor - computed: true, optional: false, required: false
        this._storageDescriptor = new DataAwsGlueCatalogTableStorageDescriptorList(this, "storage_descriptor", false);
        // target_table - computed: true, optional: false, required: false
        this._targetTable = new DataAwsGlueCatalogTableTargetTableList(this, "target_table", false);
        this._catalogId = config.catalogId;
        this._databaseName = config.databaseName;
        this._id = config.id;
        this._name = config.name;
        this._queryAsOfTime = config.queryAsOfTime;
        this._transactionId = config.transactionId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // owner - computed: true, optional: false, required: false
    get owner() {
        return this.getStringAttribute('owner');
    }
    get parameters() {
        return this._parameters;
    }
    get partitionIndex() {
        return this._partitionIndex;
    }
    get partitionKeys() {
        return this._partitionKeys;
    }
    get queryAsOfTime() {
        return this.getStringAttribute('query_as_of_time');
    }
    set queryAsOfTime(value) {
        this._queryAsOfTime = value;
    }
    resetQueryAsOfTime() {
        this._queryAsOfTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryAsOfTimeInput() {
        return this._queryAsOfTime;
    }
    // retention - computed: true, optional: false, required: false
    get retention() {
        return this.getNumberAttribute('retention');
    }
    get storageDescriptor() {
        return this._storageDescriptor;
    }
    // table_type - computed: true, optional: false, required: false
    get tableType() {
        return this.getStringAttribute('table_type');
    }
    get targetTable() {
        return this._targetTable;
    }
    get transactionId() {
        return this.getNumberAttribute('transaction_id');
    }
    set transactionId(value) {
        this._transactionId = value;
    }
    resetTransactionId() {
        this._transactionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transactionIdInput() {
        return this._transactionId;
    }
    // view_expanded_text - computed: true, optional: false, required: false
    get viewExpandedText() {
        return this.getStringAttribute('view_expanded_text');
    }
    // view_original_text - computed: true, optional: false, required: false
    get viewOriginalText() {
        return this.getStringAttribute('view_original_text');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            database_name: cdktf.stringToTerraform(this._databaseName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            query_as_of_time: cdktf.stringToTerraform(this._queryAsOfTime),
            transaction_id: cdktf.numberToTerraform(this._transactionId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            catalog_id: {
                value: cdktf.stringToHclTerraform(this._catalogId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            database_name: {
                value: cdktf.stringToHclTerraform(this._databaseName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            query_as_of_time: {
                value: cdktf.stringToHclTerraform(this._queryAsOfTime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            transaction_id: {
                value: cdktf.numberToHclTerraform(this._transactionId),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsGlueCatalogTable = DataAwsGlueCatalogTable;
_w = JSII_RTTI_SYMBOL_1;
DataAwsGlueCatalogTable[_w] = { fqn: "@cdktf/provider-aws.dataAwsGlueCatalogTable.DataAwsGlueCatalogTable", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsGlueCatalogTable.tfResourceType = "aws_glue_catalog_table";
