// @cdktf/provider-aws/lib/amplify-domain-association/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyDomainAssociation = exports.AmplifyDomainAssociationSubDomainList = exports.AmplifyDomainAssociationSubDomainOutputReference = exports.amplifyDomainAssociationSubDomainToHclTerraform = exports.amplifyDomainAssociationSubDomainToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function amplifyDomainAssociationSubDomainToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch_name: cdktf.stringToTerraform(struct.branchName),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.amplifyDomainAssociationSubDomainToTerraform = amplifyDomainAssociationSubDomainToTerraform;
function amplifyDomainAssociationSubDomainToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        branch_name: {
            value: cdktf.stringToHclTerraform(struct.branchName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.amplifyDomainAssociationSubDomainToHclTerraform = amplifyDomainAssociationSubDomainToHclTerraform;
class AmplifyDomainAssociationSubDomainOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branchName !== undefined) {
            hasAnyValues = true;
            internalValueResult.branchName = this._branchName;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._branchName = undefined;
            this._prefix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._branchName = value.branchName;
            this._prefix = value.prefix;
        }
    }
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    set branchName(value) {
        this._branchName = value;
    }
    // Temporarily expose input value. Use with caution.
    get branchNameInput() {
        return this._branchName;
    }
    // dns_record - computed: true, optional: false, required: false
    get dnsRecord() {
        return this.getStringAttribute('dns_record');
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    // verified - computed: true, optional: false, required: false
    get verified() {
        return this.getBooleanAttribute('verified');
    }
}
exports.AmplifyDomainAssociationSubDomainOutputReference = AmplifyDomainAssociationSubDomainOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AmplifyDomainAssociationSubDomainOutputReference[_a] = { fqn: "@cdktf/provider-aws.amplifyDomainAssociation.AmplifyDomainAssociationSubDomainOutputReference", version: "19.15.0" };
class AmplifyDomainAssociationSubDomainList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AmplifyDomainAssociationSubDomainOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AmplifyDomainAssociationSubDomainList = AmplifyDomainAssociationSubDomainList;
_b = JSII_RTTI_SYMBOL_1;
AmplifyDomainAssociationSubDomainList[_b] = { fqn: "@cdktf/provider-aws.amplifyDomainAssociation.AmplifyDomainAssociationSubDomainList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/amplify_domain_association aws_amplify_domain_association}
*/
class AmplifyDomainAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AmplifyDomainAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AmplifyDomainAssociation to import
    * @param importFromId The id of the existing AmplifyDomainAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/amplify_domain_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AmplifyDomainAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_amplify_domain_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/amplify_domain_association aws_amplify_domain_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AmplifyDomainAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_amplify_domain_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // sub_domain - computed: false, optional: false, required: true
        this._subDomain = new AmplifyDomainAssociationSubDomainList(this, "sub_domain", true);
        this._appId = config.appId;
        this._domainName = config.domainName;
        this._enableAutoSubDomain = config.enableAutoSubDomain;
        this._id = config.id;
        this._waitForVerification = config.waitForVerification;
        this._subDomain.internalValue = config.subDomain;
    }
    get appId() {
        return this.getStringAttribute('app_id');
    }
    set appId(value) {
        this._appId = value;
    }
    // Temporarily expose input value. Use with caution.
    get appIdInput() {
        return this._appId;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_verification_dns_record - computed: true, optional: false, required: false
    get certificateVerificationDnsRecord() {
        return this.getStringAttribute('certificate_verification_dns_record');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get enableAutoSubDomain() {
        return this.getBooleanAttribute('enable_auto_sub_domain');
    }
    set enableAutoSubDomain(value) {
        this._enableAutoSubDomain = value;
    }
    resetEnableAutoSubDomain() {
        this._enableAutoSubDomain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableAutoSubDomainInput() {
        return this._enableAutoSubDomain;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get waitForVerification() {
        return this.getBooleanAttribute('wait_for_verification');
    }
    set waitForVerification(value) {
        this._waitForVerification = value;
    }
    resetWaitForVerification() {
        this._waitForVerification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForVerificationInput() {
        return this._waitForVerification;
    }
    get subDomain() {
        return this._subDomain;
    }
    putSubDomain(value) {
        this._subDomain.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get subDomainInput() {
        return this._subDomain.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            app_id: cdktf.stringToTerraform(this._appId),
            domain_name: cdktf.stringToTerraform(this._domainName),
            enable_auto_sub_domain: cdktf.booleanToTerraform(this._enableAutoSubDomain),
            id: cdktf.stringToTerraform(this._id),
            wait_for_verification: cdktf.booleanToTerraform(this._waitForVerification),
            sub_domain: cdktf.listMapper(amplifyDomainAssociationSubDomainToTerraform, true)(this._subDomain.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            app_id: {
                value: cdktf.stringToHclTerraform(this._appId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enable_auto_sub_domain: {
                value: cdktf.booleanToHclTerraform(this._enableAutoSubDomain),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            wait_for_verification: {
                value: cdktf.booleanToHclTerraform(this._waitForVerification),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            sub_domain: {
                value: cdktf.listMapperHcl(amplifyDomainAssociationSubDomainToHclTerraform, true)(this._subDomain.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "AmplifyDomainAssociationSubDomainList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AmplifyDomainAssociation = AmplifyDomainAssociation;
_c = JSII_RTTI_SYMBOL_1;
AmplifyDomainAssociation[_c] = { fqn: "@cdktf/provider-aws.amplifyDomainAssociation.AmplifyDomainAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AmplifyDomainAssociation.tfResourceType = "aws_amplify_domain_association";
