// @cdktf/provider-aws/lib/ssoadmin-permissions-boundary-attachment/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminPermissionsBoundaryAttachment = exports.SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference = exports.ssoadminPermissionsBoundaryAttachmentTimeoutsToHclTerraform = exports.ssoadminPermissionsBoundaryAttachmentTimeoutsToTerraform = exports.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference = exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToHclTerraform = exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToTerraform = exports.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference = exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToHclTerraform = exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToTerraform = ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToTerraform;
function ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToHclTerraform = ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToHclTerraform;
class SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._path = value.path;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference = SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssoadminPermissionsBoundaryAttachment.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference", version: "19.15.0" };
function ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        managed_policy_arn: cdktf.stringToTerraform(struct.managedPolicyArn),
        customer_managed_policy_reference: ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToTerraform(struct.customerManagedPolicyReference),
    };
}
exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToTerraform = ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToTerraform;
function ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        managed_policy_arn: {
            value: cdktf.stringToHclTerraform(struct.managedPolicyArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        customer_managed_policy_reference: {
            value: ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceToHclTerraform(struct.customerManagedPolicyReference),
            isBlock: true,
            type: "list",
            storageClassType: "SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToHclTerraform = ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToHclTerraform;
class SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // customer_managed_policy_reference - computed: false, optional: true, required: false
        this._customerManagedPolicyReference = new SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryCustomerManagedPolicyReferenceOutputReference(this, "customer_managed_policy_reference");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._managedPolicyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.managedPolicyArn = this._managedPolicyArn;
        }
        if (this._customerManagedPolicyReference?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.customerManagedPolicyReference = this._customerManagedPolicyReference?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._managedPolicyArn = undefined;
            this._customerManagedPolicyReference.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._managedPolicyArn = value.managedPolicyArn;
            this._customerManagedPolicyReference.internalValue = value.customerManagedPolicyReference;
        }
    }
    get managedPolicyArn() {
        return this.getStringAttribute('managed_policy_arn');
    }
    set managedPolicyArn(value) {
        this._managedPolicyArn = value;
    }
    resetManagedPolicyArn() {
        this._managedPolicyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedPolicyArnInput() {
        return this._managedPolicyArn;
    }
    get customerManagedPolicyReference() {
        return this._customerManagedPolicyReference;
    }
    putCustomerManagedPolicyReference(value) {
        this._customerManagedPolicyReference.internalValue = value;
    }
    resetCustomerManagedPolicyReference() {
        this._customerManagedPolicyReference.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customerManagedPolicyReferenceInput() {
        return this._customerManagedPolicyReference.internalValue;
    }
}
exports.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference = SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference[_b] = { fqn: "@cdktf/provider-aws.ssoadminPermissionsBoundaryAttachment.SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference", version: "19.15.0" };
function ssoadminPermissionsBoundaryAttachmentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.ssoadminPermissionsBoundaryAttachmentTimeoutsToTerraform = ssoadminPermissionsBoundaryAttachmentTimeoutsToTerraform;
function ssoadminPermissionsBoundaryAttachmentTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminPermissionsBoundaryAttachmentTimeoutsToHclTerraform = ssoadminPermissionsBoundaryAttachmentTimeoutsToHclTerraform;
class SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference = SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.ssoadminPermissionsBoundaryAttachment.SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_permissions_boundary_attachment aws_ssoadmin_permissions_boundary_attachment}
*/
class SsoadminPermissionsBoundaryAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsoadminPermissionsBoundaryAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsoadminPermissionsBoundaryAttachment to import
    * @param importFromId The id of the existing SsoadminPermissionsBoundaryAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_permissions_boundary_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsoadminPermissionsBoundaryAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_permissions_boundary_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_permissions_boundary_attachment aws_ssoadmin_permissions_boundary_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsoadminPermissionsBoundaryAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_permissions_boundary_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // permissions_boundary - computed: false, optional: false, required: true
        this._permissionsBoundary = new SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryOutputReference(this, "permissions_boundary");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new SsoadminPermissionsBoundaryAttachmentTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._instanceArn = config.instanceArn;
        this._permissionSetArn = config.permissionSetArn;
        this._permissionsBoundary.internalValue = config.permissionsBoundary;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceArnInput() {
        return this._instanceArn;
    }
    get permissionSetArn() {
        return this.getStringAttribute('permission_set_arn');
    }
    set permissionSetArn(value) {
        this._permissionSetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionSetArnInput() {
        return this._permissionSetArn;
    }
    get permissionsBoundary() {
        return this._permissionsBoundary;
    }
    putPermissionsBoundary(value) {
        this._permissionsBoundary.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionsBoundaryInput() {
        return this._permissionsBoundary.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            permission_set_arn: cdktf.stringToTerraform(this._permissionSetArn),
            permissions_boundary: ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToTerraform(this._permissionsBoundary.internalValue),
            timeouts: ssoadminPermissionsBoundaryAttachmentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_arn: {
                value: cdktf.stringToHclTerraform(this._instanceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permission_set_arn: {
                value: cdktf.stringToHclTerraform(this._permissionSetArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permissions_boundary: {
                value: ssoadminPermissionsBoundaryAttachmentPermissionsBoundaryToHclTerraform(this._permissionsBoundary.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SsoadminPermissionsBoundaryAttachmentPermissionsBoundaryList",
            },
            timeouts: {
                value: ssoadminPermissionsBoundaryAttachmentTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "SsoadminPermissionsBoundaryAttachmentTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsoadminPermissionsBoundaryAttachment = SsoadminPermissionsBoundaryAttachment;
_d = JSII_RTTI_SYMBOL_1;
SsoadminPermissionsBoundaryAttachment[_d] = { fqn: "@cdktf/provider-aws.ssoadminPermissionsBoundaryAttachment.SsoadminPermissionsBoundaryAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsoadminPermissionsBoundaryAttachment.tfResourceType = "aws_ssoadmin_permissions_boundary_attachment";
