// @cdktf/provider-aws/lib/elastic-beanstalk-application/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticBeanstalkApplication = exports.ElasticBeanstalkApplicationAppversionLifecycleOutputReference = exports.elasticBeanstalkApplicationAppversionLifecycleToHclTerraform = exports.elasticBeanstalkApplicationAppversionLifecycleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elasticBeanstalkApplicationAppversionLifecycleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_source_from_s3: cdktf.booleanToTerraform(struct.deleteSourceFromS3),
        max_age_in_days: cdktf.numberToTerraform(struct.maxAgeInDays),
        max_count: cdktf.numberToTerraform(struct.maxCount),
        service_role: cdktf.stringToTerraform(struct.serviceRole),
    };
}
exports.elasticBeanstalkApplicationAppversionLifecycleToTerraform = elasticBeanstalkApplicationAppversionLifecycleToTerraform;
function elasticBeanstalkApplicationAppversionLifecycleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete_source_from_s3: {
            value: cdktf.booleanToHclTerraform(struct.deleteSourceFromS3),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        max_age_in_days: {
            value: cdktf.numberToHclTerraform(struct.maxAgeInDays),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_count: {
            value: cdktf.numberToHclTerraform(struct.maxCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        service_role: {
            value: cdktf.stringToHclTerraform(struct.serviceRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticBeanstalkApplicationAppversionLifecycleToHclTerraform = elasticBeanstalkApplicationAppversionLifecycleToHclTerraform;
class ElasticBeanstalkApplicationAppversionLifecycleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteSourceFromS3 !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteSourceFromS3 = this._deleteSourceFromS3;
        }
        if (this._maxAgeInDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxAgeInDays = this._maxAgeInDays;
        }
        if (this._maxCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxCount = this._maxCount;
        }
        if (this._serviceRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceRole = this._serviceRole;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteSourceFromS3 = undefined;
            this._maxAgeInDays = undefined;
            this._maxCount = undefined;
            this._serviceRole = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteSourceFromS3 = value.deleteSourceFromS3;
            this._maxAgeInDays = value.maxAgeInDays;
            this._maxCount = value.maxCount;
            this._serviceRole = value.serviceRole;
        }
    }
    get deleteSourceFromS3() {
        return this.getBooleanAttribute('delete_source_from_s3');
    }
    set deleteSourceFromS3(value) {
        this._deleteSourceFromS3 = value;
    }
    resetDeleteSourceFromS3() {
        this._deleteSourceFromS3 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteSourceFromS3Input() {
        return this._deleteSourceFromS3;
    }
    get maxAgeInDays() {
        return this.getNumberAttribute('max_age_in_days');
    }
    set maxAgeInDays(value) {
        this._maxAgeInDays = value;
    }
    resetMaxAgeInDays() {
        this._maxAgeInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxAgeInDaysInput() {
        return this._maxAgeInDays;
    }
    get maxCount() {
        return this.getNumberAttribute('max_count');
    }
    set maxCount(value) {
        this._maxCount = value;
    }
    resetMaxCount() {
        this._maxCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxCountInput() {
        return this._maxCount;
    }
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    set serviceRole(value) {
        this._serviceRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceRoleInput() {
        return this._serviceRole;
    }
}
exports.ElasticBeanstalkApplicationAppversionLifecycleOutputReference = ElasticBeanstalkApplicationAppversionLifecycleOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkApplicationAppversionLifecycleOutputReference[_a] = { fqn: "@cdktf/provider-aws.elasticBeanstalkApplication.ElasticBeanstalkApplicationAppversionLifecycleOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_application aws_elastic_beanstalk_application}
*/
class ElasticBeanstalkApplication extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ElasticBeanstalkApplication resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ElasticBeanstalkApplication to import
    * @param importFromId The id of the existing ElasticBeanstalkApplication that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_application#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ElasticBeanstalkApplication to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_elastic_beanstalk_application", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_application aws_elastic_beanstalk_application} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticBeanstalkApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastic_beanstalk_application',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // appversion_lifecycle - computed: false, optional: true, required: false
        this._appversionLifecycle = new ElasticBeanstalkApplicationAppversionLifecycleOutputReference(this, "appversion_lifecycle");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._appversionLifecycle.internalValue = config.appversionLifecycle;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get appversionLifecycle() {
        return this._appversionLifecycle;
    }
    putAppversionLifecycle(value) {
        this._appversionLifecycle.internalValue = value;
    }
    resetAppversionLifecycle() {
        this._appversionLifecycle.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get appversionLifecycleInput() {
        return this._appversionLifecycle.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            appversion_lifecycle: elasticBeanstalkApplicationAppversionLifecycleToTerraform(this._appversionLifecycle.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            appversion_lifecycle: {
                value: elasticBeanstalkApplicationAppversionLifecycleToHclTerraform(this._appversionLifecycle.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ElasticBeanstalkApplicationAppversionLifecycleList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ElasticBeanstalkApplication = ElasticBeanstalkApplication;
_b = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkApplication[_b] = { fqn: "@cdktf/provider-aws.elasticBeanstalkApplication.ElasticBeanstalkApplication", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ElasticBeanstalkApplication.tfResourceType = "aws_elastic_beanstalk_application";
