// @winglang/sdk/lib/shared-azure/counter.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterClient = void 0;
const data_tables_1 = require("@azure/data-tables");
const COUNTER_ID = "counter";
const PARTITION_KEY = "partitionKey";
class CounterClient {
    constructor(storageAccountName, storageTableName, accountKeyVariable, initial = 0) {
        this.storageAccountName = storageAccountName;
        this.storageTableName = storageTableName;
        this.accountKeyVariable = accountKeyVariable;
        this.initial = initial;
        if (!process.env[this.accountKeyVariable]) {
            throw new Error("missing storage account key");
        }
        const credentials = new data_tables_1.AzureNamedKeyCredential(this.storageAccountName, process.env[this.accountKeyVariable]);
        this.client = new data_tables_1.TableClient(`https://${this.storageAccountName}.table.core.windows.net`, this.storageTableName, credentials);
    }
    async inc(amount = 1, key = COUNTER_ID) {
        const entity = await this._getEntity(key);
        const currentValue = entity?.counterValue ?? this.initial;
        const newValue = currentValue + amount;
        await this._upsertEntity(key, newValue);
        return currentValue;
    }
    async dec(amount = 1, key = COUNTER_ID) {
        return this.inc(-1 * amount, key);
    }
    async peek(key = COUNTER_ID) {
        const entity = await this._getEntity(key);
        return entity?.counterValue ?? this.initial;
    }
    async set(value, key = COUNTER_ID) {
        await this._upsertEntity(key, value);
    }
    async _getEntity(key) {
        try {
            return await this.client.getEntity(PARTITION_KEY, key);
        }
        catch (error) {
            return undefined;
        }
    }
    async _upsertEntity(key, value) {
        const entity = {
            partitionKey: PARTITION_KEY,
            rowKey: key,
            counterValue: value,
        };
        await this.client.upsertEntity(entity);
    }
}
exports.CounterClient = CounterClient;
