// @winglang/sdk/lib/shared-azure/function.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionClient = void 0;
const http_1 = require("../http");
class FunctionClient {
    constructor(functionName) {
        this.functionName = functionName;
    }
    /**
     * Invoke the function, passing the given payload as an argument.
     *  @returns the function returned payload only
     */
    async invoke(payload) {
        try {
            const res = await http_1.Util.post(`https://${this.functionName}.azurewebsites.net/api/${this.functionName}`, { body: payload });
            if (!res.ok) {
                throw new Error(res.body || "Invocation Error");
            }
            return res.body || undefined;
        }
        catch (error) {
            throw new Error(`Error while invoking the function ${this.functionName}:\n${error.message}`);
        }
    }
    /**
     * Invokes the function asynchronously, passing the given payload as an argument.
     * @returns immediately once the event has been handed off to AWS Lambda.
     */
    async invokeAsync(payload) {
        payload;
        throw new Error("invokeAsync is not implemented");
    }
    /**
     * Invoke the function, passing the given payload as an argument.
     *
     * @returns the function returned payload and logs
     */
    async invokeWithLogs(payload) {
        console.error("Test invocation on tf-azure doesn't include logs yet");
        //TODO: add traces to tf-azure tests- https://github.com/winglang/wing/issues/4574
        return [await this.invoke(payload), []];
    }
}
exports.FunctionClient = FunctionClient;
