// @cdktf/provider-aws/lib/appstream-fleet/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamFleet = exports.AppstreamFleetVpcConfigOutputReference = exports.appstreamFleetVpcConfigToHclTerraform = exports.appstreamFleetVpcConfigToTerraform = exports.AppstreamFleetDomainJoinInfoOutputReference = exports.appstreamFleetDomainJoinInfoToHclTerraform = exports.appstreamFleetDomainJoinInfoToTerraform = exports.AppstreamFleetComputeCapacityOutputReference = exports.appstreamFleetComputeCapacityToHclTerraform = exports.appstreamFleetComputeCapacityToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appstreamFleetComputeCapacityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        desired_instances: cdktf.numberToTerraform(struct.desiredInstances),
        desired_sessions: cdktf.numberToTerraform(struct.desiredSessions),
    };
}
exports.appstreamFleetComputeCapacityToTerraform = appstreamFleetComputeCapacityToTerraform;
function appstreamFleetComputeCapacityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        desired_instances: {
            value: cdktf.numberToHclTerraform(struct.desiredInstances),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        desired_sessions: {
            value: cdktf.numberToHclTerraform(struct.desiredSessions),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.appstreamFleetComputeCapacityToHclTerraform = appstreamFleetComputeCapacityToHclTerraform;
class AppstreamFleetComputeCapacityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._desiredInstances !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredInstances = this._desiredInstances;
        }
        if (this._desiredSessions !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredSessions = this._desiredSessions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._desiredInstances = undefined;
            this._desiredSessions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._desiredInstances = value.desiredInstances;
            this._desiredSessions = value.desiredSessions;
        }
    }
    // available - computed: true, optional: false, required: false
    get available() {
        return this.getNumberAttribute('available');
    }
    get desiredInstances() {
        return this.getNumberAttribute('desired_instances');
    }
    set desiredInstances(value) {
        this._desiredInstances = value;
    }
    resetDesiredInstances() {
        this._desiredInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredInstancesInput() {
        return this._desiredInstances;
    }
    get desiredSessions() {
        return this.getNumberAttribute('desired_sessions');
    }
    set desiredSessions(value) {
        this._desiredSessions = value;
    }
    resetDesiredSessions() {
        this._desiredSessions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredSessionsInput() {
        return this._desiredSessions;
    }
    // in_use - computed: true, optional: false, required: false
    get inUse() {
        return this.getNumberAttribute('in_use');
    }
    // running - computed: true, optional: false, required: false
    get running() {
        return this.getNumberAttribute('running');
    }
}
exports.AppstreamFleetComputeCapacityOutputReference = AppstreamFleetComputeCapacityOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppstreamFleetComputeCapacityOutputReference[_a] = { fqn: "@cdktf/provider-aws.appstreamFleet.AppstreamFleetComputeCapacityOutputReference", version: "19.15.0" };
function appstreamFleetDomainJoinInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        directory_name: cdktf.stringToTerraform(struct.directoryName),
        organizational_unit_distinguished_name: cdktf.stringToTerraform(struct.organizationalUnitDistinguishedName),
    };
}
exports.appstreamFleetDomainJoinInfoToTerraform = appstreamFleetDomainJoinInfoToTerraform;
function appstreamFleetDomainJoinInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        directory_name: {
            value: cdktf.stringToHclTerraform(struct.directoryName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        organizational_unit_distinguished_name: {
            value: cdktf.stringToHclTerraform(struct.organizationalUnitDistinguishedName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.appstreamFleetDomainJoinInfoToHclTerraform = appstreamFleetDomainJoinInfoToHclTerraform;
class AppstreamFleetDomainJoinInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._directoryName !== undefined) {
            hasAnyValues = true;
            internalValueResult.directoryName = this._directoryName;
        }
        if (this._organizationalUnitDistinguishedName !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationalUnitDistinguishedName = this._organizationalUnitDistinguishedName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._directoryName = undefined;
            this._organizationalUnitDistinguishedName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._directoryName = value.directoryName;
            this._organizationalUnitDistinguishedName = value.organizationalUnitDistinguishedName;
        }
    }
    get directoryName() {
        return this.getStringAttribute('directory_name');
    }
    set directoryName(value) {
        this._directoryName = value;
    }
    resetDirectoryName() {
        this._directoryName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directoryNameInput() {
        return this._directoryName;
    }
    get organizationalUnitDistinguishedName() {
        return this.getStringAttribute('organizational_unit_distinguished_name');
    }
    set organizationalUnitDistinguishedName(value) {
        this._organizationalUnitDistinguishedName = value;
    }
    resetOrganizationalUnitDistinguishedName() {
        this._organizationalUnitDistinguishedName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitDistinguishedNameInput() {
        return this._organizationalUnitDistinguishedName;
    }
}
exports.AppstreamFleetDomainJoinInfoOutputReference = AppstreamFleetDomainJoinInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AppstreamFleetDomainJoinInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.appstreamFleet.AppstreamFleetDomainJoinInfoOutputReference", version: "19.15.0" };
function appstreamFleetVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.appstreamFleetVpcConfigToTerraform = appstreamFleetVpcConfigToTerraform;
function appstreamFleetVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.appstreamFleetVpcConfigToHclTerraform = appstreamFleetVpcConfigToHclTerraform;
class AppstreamFleetVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.AppstreamFleetVpcConfigOutputReference = AppstreamFleetVpcConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AppstreamFleetVpcConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.appstreamFleet.AppstreamFleetVpcConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appstream_fleet aws_appstream_fleet}
*/
class AppstreamFleet extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AppstreamFleet resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AppstreamFleet to import
    * @param importFromId The id of the existing AppstreamFleet that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appstream_fleet#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AppstreamFleet to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_appstream_fleet", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appstream_fleet aws_appstream_fleet} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppstreamFleetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_fleet',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // compute_capacity - computed: false, optional: false, required: true
        this._computeCapacity = new AppstreamFleetComputeCapacityOutputReference(this, "compute_capacity");
        // domain_join_info - computed: false, optional: true, required: false
        this._domainJoinInfo = new AppstreamFleetDomainJoinInfoOutputReference(this, "domain_join_info");
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new AppstreamFleetVpcConfigOutputReference(this, "vpc_config");
        this._description = config.description;
        this._disconnectTimeoutInSeconds = config.disconnectTimeoutInSeconds;
        this._displayName = config.displayName;
        this._enableDefaultInternetAccess = config.enableDefaultInternetAccess;
        this._fleetType = config.fleetType;
        this._iamRoleArn = config.iamRoleArn;
        this._id = config.id;
        this._idleDisconnectTimeoutInSeconds = config.idleDisconnectTimeoutInSeconds;
        this._imageArn = config.imageArn;
        this._imageName = config.imageName;
        this._instanceType = config.instanceType;
        this._maxSessionsPerInstance = config.maxSessionsPerInstance;
        this._maxUserDurationInSeconds = config.maxUserDurationInSeconds;
        this._name = config.name;
        this._streamView = config.streamView;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._computeCapacity.internalValue = config.computeCapacity;
        this._domainJoinInfo.internalValue = config.domainJoinInfo;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_time - computed: true, optional: false, required: false
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disconnectTimeoutInSeconds() {
        return this.getNumberAttribute('disconnect_timeout_in_seconds');
    }
    set disconnectTimeoutInSeconds(value) {
        this._disconnectTimeoutInSeconds = value;
    }
    resetDisconnectTimeoutInSeconds() {
        this._disconnectTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disconnectTimeoutInSecondsInput() {
        return this._disconnectTimeoutInSeconds;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get enableDefaultInternetAccess() {
        return this.getBooleanAttribute('enable_default_internet_access');
    }
    set enableDefaultInternetAccess(value) {
        this._enableDefaultInternetAccess = value;
    }
    resetEnableDefaultInternetAccess() {
        this._enableDefaultInternetAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableDefaultInternetAccessInput() {
        return this._enableDefaultInternetAccess;
    }
    get fleetType() {
        return this.getStringAttribute('fleet_type');
    }
    set fleetType(value) {
        this._fleetType = value;
    }
    resetFleetType() {
        this._fleetType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fleetTypeInput() {
        return this._fleetType;
    }
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    set iamRoleArn(value) {
        this._iamRoleArn = value;
    }
    resetIamRoleArn() {
        this._iamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleArnInput() {
        return this._iamRoleArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get idleDisconnectTimeoutInSeconds() {
        return this.getNumberAttribute('idle_disconnect_timeout_in_seconds');
    }
    set idleDisconnectTimeoutInSeconds(value) {
        this._idleDisconnectTimeoutInSeconds = value;
    }
    resetIdleDisconnectTimeoutInSeconds() {
        this._idleDisconnectTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleDisconnectTimeoutInSecondsInput() {
        return this._idleDisconnectTimeoutInSeconds;
    }
    get imageArn() {
        return this.getStringAttribute('image_arn');
    }
    set imageArn(value) {
        this._imageArn = value;
    }
    resetImageArn() {
        this._imageArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageArnInput() {
        return this._imageArn;
    }
    get imageName() {
        return this.getStringAttribute('image_name');
    }
    set imageName(value) {
        this._imageName = value;
    }
    resetImageName() {
        this._imageName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageNameInput() {
        return this._imageName;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get maxSessionsPerInstance() {
        return this.getNumberAttribute('max_sessions_per_instance');
    }
    set maxSessionsPerInstance(value) {
        this._maxSessionsPerInstance = value;
    }
    resetMaxSessionsPerInstance() {
        this._maxSessionsPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSessionsPerInstanceInput() {
        return this._maxSessionsPerInstance;
    }
    get maxUserDurationInSeconds() {
        return this.getNumberAttribute('max_user_duration_in_seconds');
    }
    set maxUserDurationInSeconds(value) {
        this._maxUserDurationInSeconds = value;
    }
    resetMaxUserDurationInSeconds() {
        this._maxUserDurationInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUserDurationInSecondsInput() {
        return this._maxUserDurationInSeconds;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get streamView() {
        return this.getStringAttribute('stream_view');
    }
    set streamView(value) {
        this._streamView = value;
    }
    resetStreamView() {
        this._streamView = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get streamViewInput() {
        return this._streamView;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get computeCapacity() {
        return this._computeCapacity;
    }
    putComputeCapacity(value) {
        this._computeCapacity.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get computeCapacityInput() {
        return this._computeCapacity.internalValue;
    }
    get domainJoinInfo() {
        return this._domainJoinInfo;
    }
    putDomainJoinInfo(value) {
        this._domainJoinInfo.internalValue = value;
    }
    resetDomainJoinInfo() {
        this._domainJoinInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainJoinInfoInput() {
        return this._domainJoinInfo.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            disconnect_timeout_in_seconds: cdktf.numberToTerraform(this._disconnectTimeoutInSeconds),
            display_name: cdktf.stringToTerraform(this._displayName),
            enable_default_internet_access: cdktf.booleanToTerraform(this._enableDefaultInternetAccess),
            fleet_type: cdktf.stringToTerraform(this._fleetType),
            iam_role_arn: cdktf.stringToTerraform(this._iamRoleArn),
            id: cdktf.stringToTerraform(this._id),
            idle_disconnect_timeout_in_seconds: cdktf.numberToTerraform(this._idleDisconnectTimeoutInSeconds),
            image_arn: cdktf.stringToTerraform(this._imageArn),
            image_name: cdktf.stringToTerraform(this._imageName),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            max_sessions_per_instance: cdktf.numberToTerraform(this._maxSessionsPerInstance),
            max_user_duration_in_seconds: cdktf.numberToTerraform(this._maxUserDurationInSeconds),
            name: cdktf.stringToTerraform(this._name),
            stream_view: cdktf.stringToTerraform(this._streamView),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            compute_capacity: appstreamFleetComputeCapacityToTerraform(this._computeCapacity.internalValue),
            domain_join_info: appstreamFleetDomainJoinInfoToTerraform(this._domainJoinInfo.internalValue),
            vpc_config: appstreamFleetVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disconnect_timeout_in_seconds: {
                value: cdktf.numberToHclTerraform(this._disconnectTimeoutInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enable_default_internet_access: {
                value: cdktf.booleanToHclTerraform(this._enableDefaultInternetAccess),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            fleet_type: {
                value: cdktf.stringToHclTerraform(this._fleetType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_role_arn: {
                value: cdktf.stringToHclTerraform(this._iamRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            idle_disconnect_timeout_in_seconds: {
                value: cdktf.numberToHclTerraform(this._idleDisconnectTimeoutInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            image_arn: {
                value: cdktf.stringToHclTerraform(this._imageArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            image_name: {
                value: cdktf.stringToHclTerraform(this._imageName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_type: {
                value: cdktf.stringToHclTerraform(this._instanceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            max_sessions_per_instance: {
                value: cdktf.numberToHclTerraform(this._maxSessionsPerInstance),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            max_user_duration_in_seconds: {
                value: cdktf.numberToHclTerraform(this._maxUserDurationInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            stream_view: {
                value: cdktf.stringToHclTerraform(this._streamView),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            compute_capacity: {
                value: appstreamFleetComputeCapacityToHclTerraform(this._computeCapacity.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AppstreamFleetComputeCapacityList",
            },
            domain_join_info: {
                value: appstreamFleetDomainJoinInfoToHclTerraform(this._domainJoinInfo.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AppstreamFleetDomainJoinInfoList",
            },
            vpc_config: {
                value: appstreamFleetVpcConfigToHclTerraform(this._vpcConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AppstreamFleetVpcConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AppstreamFleet = AppstreamFleet;
_d = JSII_RTTI_SYMBOL_1;
AppstreamFleet[_d] = { fqn: "@cdktf/provider-aws.appstreamFleet.AppstreamFleet", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AppstreamFleet.tfResourceType = "aws_appstream_fleet";
