// @winglang/sdk/lib/target-sim/event-mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventMapping = exports.EVENT_MAPPING_FQN = void 0;
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const constants_1 = require("../constants");
const std_1 = require("../std");
exports.EVENT_MAPPING_FQN = (0, constants_1.fqnForType)("sim.EventMapping");
/**
 * Implementation of `sim.EventMapping`.
 *
 * @inflight `@winglang/sdk.sim.EventMapping`
 */
class EventMapping extends std_1.Resource {
    constructor(scope, id, props) {
        super(scope, id);
        this._eventProps = props;
        std_1.Node.of(this).hidden = true;
        // Add dependencies to the publisher and subscriber
        this.node.addDependency(props.subscriber);
        this.node.addDependency(props.publisher);
    }
    /** @internal */
    get _liftMap() {
        return {};
    }
    get eventProps() {
        return this._eventProps;
    }
    toSimulator() {
        const props = {
            subscriber: (0, tokens_1.simulatorHandleToken)(this.eventProps.subscriber),
            publisher: (0, tokens_1.simulatorHandleToken)(this.eventProps.publisher),
            subscriptionProps: this.eventProps.subscriptionProps,
        };
        return {
            type: exports.EVENT_MAPPING_FQN,
            props,
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.EventMapping = EventMapping;
