// @winglang/sdk/lib/target-sim/function.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionInflightMethods = exports.Function = exports.ENV_WING_SIM_INFLIGHT_RESOURCE_TYPE = exports.ENV_WING_SIM_INFLIGHT_RESOURCE_PATH = void 0;
const path_1 = require("path");
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
const core_1 = require("../core");
const duration_1 = require("../std/duration");
exports.ENV_WING_SIM_INFLIGHT_RESOURCE_PATH = "WING_SIM_INFLIGHT_RESOURCE_PATH";
exports.ENV_WING_SIM_INFLIGHT_RESOURCE_TYPE = "WING_SIM_INFLIGHT_RESOURCE_TYPE";
/**
 * Simulator implementation of `cloud.Function`.
 *
 * @inflight `@winglang/sdk.cloud.IFunctionClient`
 */
class Function extends cloud.Function {
    constructor(scope, id, inflight, props = {}) {
        super(scope, id, inflight, props);
        this.permissions = [];
        // props.memory is unused since we are not simulating it
        this.timeout = props.timeout ?? duration_1.Duration.fromMinutes(1);
        this.concurrency = props.concurrency ?? 100;
    }
    addPermission(resource, op) {
        this.permissions.push([resource, op]);
    }
    toSimulator() {
        const outdir = core_1.App.of(this).outdir;
        const policy = [];
        for (const [resource, operation] of this.permissions) {
            policy.push({
                operation,
                resourceHandle: (0, tokens_1.simulatorHandleToken)(resource),
            });
        }
        const props = {
            sourceCodeFile: (0, path_1.relative)(outdir, this.entrypoint),
            sourceCodeLanguage: "javascript",
            environmentVariables: this.env,
            timeout: this.timeout.seconds * 1000,
            concurrency: this.concurrency,
        };
        return {
            type: cloud.FUNCTION_FQN,
            props,
            policy,
        };
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.FunctionInflightMethods.INVOKE]: [[this.handler, ["handle"]]],
            [cloud.FunctionInflightMethods.INVOKE_ASYNC]: [
                [this.handler, ["handle"]],
            ],
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Function = Function;
/**
 * Simulator-specific inflight methods for `cloud.Function`.
 */
var FunctionInflightMethods;
(function (FunctionInflightMethods) {
    FunctionInflightMethods["HAS_AVAILABLE_WORKERS"] = "hasAvailableWorkers";
})(FunctionInflightMethods || (exports.FunctionInflightMethods = FunctionInflightMethods = {}));
