// @winglang/sdk/lib/target-sim/on-deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnDeploy = void 0;
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
const std_1 = require("../std");
class OnDeploy extends cloud.OnDeploy {
    constructor(scope, id, handler, props = {}) {
        super(scope, id, handler, props);
        this.fn = new cloud.Function(this, "Function", handler, props);
        std_1.Node.of(this.fn).sourceModule = std_1.SDK_SOURCE_MODULE;
        this.node.addDependency(this.fn);
        for (const c of props.executeBefore ?? []) {
            c.node.addDependency(this);
        }
        for (const c of props.executeAfter ?? []) {
            this.node.addDependency(c);
        }
    }
    toSimulator() {
        const props = {
            functionHandle: (0, tokens_1.simulatorHandleToken)(this.fn),
        };
        return {
            type: cloud.ON_DEPLOY_FQN,
            props,
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.OnDeploy = OnDeploy;
