// @winglang/sdk/lib/target-sim/queue.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueSetConsumerHandler = exports.Queue = void 0;
const app_1 = require("./app");
const event_mapping_1 = require("./event-mapping");
const function_1 = require("./function");
const policy_1 = require("./policy");
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
const core_1 = require("../core");
const errors_1 = require("../core/errors");
const std_1 = require("../std");
/**
 * Simulator implementation of `cloud.Queue`.
 *
 * @inflight `@winglang/sdk.cloud.IQueueClient`
 */
class Queue extends cloud.Queue {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        if (props.timeout) {
            throw new errors_1.NotImplementedError("Queue.timeout is not supported on the sim platform yet.", {
                issue: "https://github.com/winglang/wing/issues/1980",
                resource: this.constructor.name,
                operation: "timeout",
            });
        }
        this.timeout = props.timeout ?? std_1.Duration.fromSeconds(30);
        this.retentionPeriod = props.retentionPeriod ?? std_1.Duration.fromHours(1);
        if (this.retentionPeriod.seconds < this.timeout.seconds) {
            throw new Error("Retention period must be greater than or equal to timeout");
        }
        this.policy = new policy_1.Policy(this, "Policy", { principal: this });
        if (props.dlq && props.dlq.queue) {
            this.dlq = props.dlq;
            this.policy.addStatement(this.dlq.queue, cloud.QueueInflightMethods.PUSH);
            std_1.Node.of(this).addConnection({
                source: this,
                sourceOp: cloud.QueueInflightMethods.POP,
                target: this.dlq.queue,
                targetOp: cloud.QueueInflightMethods.PUSH,
                name: "dead-letter queue",
            });
        }
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.QueueInflightMethods.PUSH]: [],
            [cloud.QueueInflightMethods.PURGE]: [],
            [cloud.QueueInflightMethods.APPROX_SIZE]: [],
            [cloud.QueueInflightMethods.POP]: [],
        };
    }
    setConsumer(inflight, props = {}) {
        const functionHandler = QueueSetConsumerHandler.toFunctionHandler(inflight);
        const fn = new cloud.Function(this, app_1.App.of(this).makeId(this, "Consumer"), functionHandler, props);
        const fnNode = std_1.Node.of(fn);
        fnNode.sourceModule = std_1.SDK_SOURCE_MODULE;
        fnNode.title = "setConsumer()";
        const mapping = new event_mapping_1.EventMapping(this, app_1.App.of(this).makeId(this, "QueueEventMapping"), {
            subscriber: fn,
            publisher: this,
            subscriptionProps: {
                batchSize: props.batchSize ?? 1,
            },
        });
        this.policy.addStatement(fn, cloud.FunctionInflightMethods.INVOKE);
        this.policy.addStatement(fn, function_1.FunctionInflightMethods.HAS_AVAILABLE_WORKERS);
        mapping.node.addDependency(this.policy);
        std_1.Node.of(this).addConnection({
            source: this,
            sourceOp: cloud.QueueInflightMethods.PUSH,
            target: fn,
            targetOp: cloud.FunctionInflightMethods.INVOKE,
            name: "setConsumer()",
        });
        return fn;
    }
    toSimulator() {
        const props = {
            timeout: this.timeout.seconds,
            retentionPeriod: this.retentionPeriod.seconds,
            dlq: this.dlq
                ? {
                    dlqHandler: (0, tokens_1.simulatorHandleToken)(this.dlq.queue),
                    maxDeliveryAttempts: this.dlq.maxDeliveryAttempts ?? cloud.DEFAULT_DELIVERY_ATTEMPTS,
                }
                : undefined,
        };
        return {
            type: cloud.QUEUE_FQN,
            props,
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Queue = Queue;
/**
 * Utility class to work with queue set consumer handlers.
 */
class QueueSetConsumerHandler {
    /**
     * Converts from a `cloud.IQueueSetConsumerHandler` to a `cloud.IFunctionHandler`.
     * @param handler The handler to convert.
     * @returns The function handler.
     */
    static toFunctionHandler(handler) {
        return (0, core_1.lift)({ handler }).inflight(async (ctx, event) => {
            const batchItemFailures = [];
            let parsed = JSON.parse(event ?? "{}");
            if (!parsed.messages)
                throw new Error('No "messages" field in event.');
            for (const $message of parsed.messages) {
                try {
                    await ctx.handler($message.payload);
                }
                catch (error) {
                    // TODO: an error from user code is getting dropped - bad! https://github.com/winglang/wing/issues/6445
                    batchItemFailures.push($message);
                }
            }
            return batchItemFailures.length > 0
                ? JSON.stringify(batchItemFailures)
                : undefined;
        });
    }
}
exports.QueueSetConsumerHandler = QueueSetConsumerHandler;
