// @winglang/sdk/lib/target-sim/redis.inflight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Redis = void 0;
const ioredis_1 = __importDefault(require("ioredis"));
const ex_1 = require("../ex");
const simulator_1 = require("../simulator/simulator");
class Redis extends ex_1.RedisClientBase {
    constructor(props) {
        super();
        this.props = props;
        this.connectionUrl = undefined;
        this.isCleanedUp = false;
    }
    async init(_context) {
        try {
            if (this.isCleanedUp) {
                return {};
            }
            // redis url based on host port
            this.connectionUrl = `redis://0.0.0.0:${this.props.port}`;
            return {};
        }
        catch (e) {
            throw Error(`Error setting up Redis resource simulation (${e})
      - Make sure you have docker installed and running`);
        }
    }
    async cleanup() {
        this.isCleanedUp = true;
        // disconnect from the redis server
        await this.connection?.quit();
        this.connection?.disconnect();
    }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async rawClient() {
        if (this.connection) {
            return this.connection;
        }
        if (this.connectionUrl) {
            this.connection = new ioredis_1.default(this.connectionUrl);
            return this.connection;
        }
        throw new Error("Redis server not initialized");
    }
    async url() {
        if (this.connectionUrl != undefined) {
            return this.connectionUrl;
        }
        else {
            throw new Error("Redis server not initialized");
        }
    }
}
exports.Redis = Redis;
