// @winglang/sdk/lib/target-sim/redis.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Redis = void 0;
const container_1 = require("./container");
const util_1 = require("./util");
const ex = __importStar(require("../ex"));
/**
 * Simulator implementation of `redis.Redis`.
 *
 * @inflight `@winglang/sdk.redis.IRedisClient`
 */
class Redis extends ex.Redis {
    constructor(scope, id) {
        super(scope, id);
        this.WING_REDIS_IMAGE = process.env.WING_REDIS_IMAGE ??
            // Redis version 7.0.9
            "redis@sha256:e50c7e23f79ae81351beacb20e004720d4bed657415e68c2b1a2b5557c075ce0";
        const c = new container_1.Container(this, "Container", {
            name: "redis",
            image: this.WING_REDIS_IMAGE,
            containerPort: 6379,
        });
        if (!c.hostPort) {
            throw new Error("Failed to get host port for the redis container");
        }
        this.hostPort = c.hostPort;
    }
    toSimulator() {
        const props = {
            port: this.hostPort,
        };
        return {
            type: ex.REDIS_FQN,
            props,
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    /** @internal */
    get _liftMap() {
        return {
            [ex.RedisInflightMethods.URL]: [],
            [ex.RedisInflightMethods.SET]: [],
            [ex.RedisInflightMethods.GET]: [],
            [ex.RedisInflightMethods.HSET]: [],
            [ex.RedisInflightMethods.HGET]: [],
            [ex.RedisInflightMethods.SADD]: [],
            [ex.RedisInflightMethods.SMEMBERS]: [],
            [ex.RedisInflightMethods.DEL]: [],
        };
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Redis = Redis;
