// @winglang/sdk/lib/target-sim/schedule.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedule = void 0;
const cron_parser_1 = require("cron-parser");
const cloud_1 = require("../cloud");
const simulator_1 = require("../simulator");
const std_1 = require("../std");
class Schedule {
    constructor(props) {
        this.tasks = new Array();
        this.interval = (0, cron_parser_1.parseExpression)(props.cronExpression, { utc: true });
        this.scheduleFunction();
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    // Calculate the delay for the next execution
    nextDelay(interval) {
        return interval.next().toDate().getTime() - Date.now();
    }
    // Recursively schedule the function to be executed
    scheduleFunction() {
        this.intervalTimeout = setTimeout(() => {
            this.runTasks();
            this.scheduleFunction();
        }, this.nextDelay(this.interval));
    }
    async init(context) {
        this._context = context;
        return {};
    }
    async cleanup() {
        clearTimeout(this.intervalTimeout);
    }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async addEventSubscription(subscriber, subscriptionProps) {
        const task = {
            functionHandle: subscriber,
            ...subscriptionProps,
        };
        this.tasks.push(task);
    }
    async removeEventSubscription(subscriber) {
        const index = this.tasks.findIndex((s) => s.functionHandle === subscriber);
        if (index >= 0) {
            this.tasks.splice(index, 1);
        }
    }
    runTasks() {
        for (const task of this.tasks) {
            const fnClient = this.context.getClient(task.functionHandle);
            if (!fnClient) {
                throw new Error("No function client found for task.");
            }
            this.context.addTrace({
                type: std_1.TraceType.RESOURCE,
                level: std_1.LogLevel.VERBOSE,
                data: {
                    message: `Running task with function handle: ${task.functionHandle}.`,
                },
                sourcePath: this.context.resourcePath,
                sourceType: cloud_1.SCHEDULE_FQN,
                timestamp: new Date().toISOString(),
            });
            void fnClient.invoke().catch((err) => {
                this.context.addTrace({
                    data: {
                        message: `Schedule error: ${err}`,
                    },
                    sourcePath: this.context.resourcePath,
                    sourceType: cloud_1.SCHEDULE_FQN,
                    timestamp: new Date().toISOString(),
                    type: std_1.TraceType.RESOURCE,
                    level: std_1.LogLevel.ERROR,
                });
            });
        }
    }
}
exports.Schedule = Schedule;
