// @winglang/sdk/lib/target-sim/schedule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScheduleOnTickHandler = exports.Schedule = void 0;
const app_1 = require("./app");
const event_mapping_1 = require("./event-mapping");
const policy_1 = require("./policy");
const util_1 = require("./util");
const cloud_1 = require("../cloud");
const cloud = __importStar(require("../cloud"));
const core_1 = require("../core");
const std_1 = require("../std");
/**
 * Simulator implementation of `cloud.Schedule`.
 *
 * @inflight `@winglang/sdk.cloud.IScheduleClient`
 */
class Schedule extends cloud.Schedule {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        const { rate, cron } = props;
        this.cronExpression = cron ?? (0, util_1.convertDurationToCronExpression)(rate);
        this.policy = new policy_1.Policy(this, "Policy", { principal: this });
    }
    onTick(inflight, props = {}) {
        const functionHandler = ScheduleOnTickHandler.toFunctionHandler(inflight);
        const fn = new cloud_1.Function(this, app_1.App.of(this).makeId(this, "OnTick"), functionHandler, props);
        std_1.Node.of(fn).sourceModule = std_1.SDK_SOURCE_MODULE;
        std_1.Node.of(fn).title = "onTick()";
        new event_mapping_1.EventMapping(this, app_1.App.of(this).makeId(this, "OnTickMapping"), {
            subscriber: fn,
            publisher: this,
            subscriptionProps: {},
        });
        std_1.Node.of(this).addConnection({
            source: this,
            sourceOp: cloud.ScheduleInflightMethods.TICK,
            target: fn,
            targetOp: cloud.FunctionInflightMethods.INVOKE,
            name: "onTick()",
        });
        this.policy.addStatement(fn, cloud.FunctionInflightMethods.INVOKE);
        return fn;
    }
    toSimulator() {
        const props = {
            cronExpression: this.cronExpression,
        };
        return {
            type: cloud.SCHEDULE_FQN,
            props,
        };
    }
    /** @internal */
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
}
exports.Schedule = Schedule;
/**
 * Utility class to work with schedule on tick handlers.
 */
class ScheduleOnTickHandler {
    /**
     * Converts a `cloud.IScheduleOnTickHandler` to a `cloud.IFunctionHandler`.
     * @param handler the handler to convert
     * @returns the function handler
     */
    static toFunctionHandler(handler) {
        return (0, core_1.lift)({ handler }).inflight(async (ctx) => {
            await ctx.handler();
            return undefined;
        });
    }
}
exports.ScheduleOnTickHandler = ScheduleOnTickHandler;
