// @winglang/sdk/lib/target-sim/service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Service = void 0;
const path_1 = require("path");
const tokens_1 = require("./tokens");
const util_1 = require("./util");
const cloud = __importStar(require("../cloud"));
const core_1 = require("../core");
class Service extends cloud.Service {
    constructor(scope, id, handler, props = {}) {
        super(scope, id, handler, props);
        this.permissions = [];
        this.autoStart = props.autoStart ?? true;
    }
    addPermission(resource, op) {
        this.permissions.push([resource, op]);
    }
    toSimulator() {
        const policy = [];
        for (const [resource, operation] of this.permissions) {
            policy.push({
                operation,
                resourceHandle: (0, tokens_1.simulatorHandleToken)(resource),
            });
        }
        const props = {
            environmentVariables: this.env,
            sourceCodeFile: (0, path_1.relative)(core_1.App.of(this).outdir, this.entrypoint),
            autoStart: this.autoStart,
        };
        return {
            type: cloud.SERVICE_FQN,
            props,
            policy,
        };
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.ServiceInflightMethods.START]: [],
            [cloud.ServiceInflightMethods.STOP]: [],
            [cloud.ServiceInflightMethods.STARTED]: [],
        };
    }
    onLift(host, ops) {
        (0, util_1.bindSimulatorResource)(__filename, this, host, ops);
        super.onLift(host, ops);
    }
    _toInflight() {
        return (0, util_1.makeSimulatorJsClient)(__filename, this);
    }
}
exports.Service = Service;
