// @winglang/sdk/lib/target-sim/state.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.State = void 0;
const simulator_1 = require("../simulator");
class State {
    constructor(_props) { }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        return {};
    }
    async cleanup() { }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    async set(key, value) {
        this.context.setResourceAttributes(this.context.resourcePath, {
            [key]: value,
        });
    }
    async get(key) {
        const value = await this.tryGet(key);
        if (value === undefined) {
            throw new Error(`Key "${key}" not found`);
        }
        return value;
    }
    async tryGet(key) {
        return this.context.resourceAttributes(this.context.resourcePath)[key];
    }
}
exports.State = State;
