// @winglang/sdk/lib/target-sim/website.inflight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Website = void 0;
const path_1 = require("path");
const express_1 = __importDefault(require("express"));
const cloud_1 = require("../cloud");
const simulator_1 = require("../simulator/simulator");
const std_1 = require("../std");
const LOCALHOST_ADDRESS = "127.0.0.1";
class Website {
    constructor(props) {
        // Set up an express server that handles the routes.
        this.app = (0, express_1.default)();
        // Use static directory
        this.app.use(express_1.default.static(props.staticFilesPath));
        this.initiateFileRoutes(props.fileRoutes);
        if (props.errorDocument) {
            let errorDocument = props.errorDocument;
            this.app.get("*", function (_, res) {
                return res.sendFile((0, path_1.join)(props.staticFilesPath, errorDocument));
            });
        }
    }
    initiateFileRoutes(routes) {
        for (const path in routes) {
            this.app.get(`/${path}`, (_, res) => {
                res.setHeader("Content-Type", routes[path].contentType);
                res.send(routes[path].data);
            });
        }
    }
    get context() {
        if (!this._context) {
            throw new Error("Cannot access context during class construction");
        }
        return this._context;
    }
    async init(context) {
        this._context = context;
        // `server.address()` returns `null` until the server is listening
        // on a port. We use a promise to wait for the server to start
        // listening before returning the URL.
        const addrInfo = await new Promise((resolve, reject) => {
            this.server = this.app.listen(0, LOCALHOST_ADDRESS, () => {
                const addr = this.server?.address();
                if (addr && typeof addr === "object" && addr.port) {
                    resolve(addr);
                }
                else {
                    reject(new Error("No address found"));
                }
            });
        });
        this.url = `http://${addrInfo.address}:${addrInfo.port}`;
        this.addTrace(`Website server listening on ${this.url}`, std_1.LogLevel.INFO);
        return {
            url: this.url,
        };
    }
    async cleanup() {
        this.addTrace(`Closing server on ${this.url}`, std_1.LogLevel.VERBOSE);
        this.server?.close();
        this.server?.closeAllConnections();
    }
    async save() { }
    async plan() {
        return simulator_1.UpdatePlan.AUTO;
    }
    addTrace(message, level) {
        this.context.addTrace({
            type: std_1.TraceType.RESOURCE,
            data: {
                message,
            },
            sourcePath: this.context.resourcePath,
            sourceType: cloud_1.WEBSITE_FQN,
            timestamp: new Date().toISOString(),
            level,
        });
    }
}
exports.Website = Website;
