// @cdktf/provider-aws/lib/apprunner-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApprunnerService = exports.ApprunnerServiceSourceConfigurationOutputReference = exports.apprunnerServiceSourceConfigurationToHclTerraform = exports.apprunnerServiceSourceConfigurationToTerraform = exports.ApprunnerServiceSourceConfigurationImageRepositoryOutputReference = exports.apprunnerServiceSourceConfigurationImageRepositoryToHclTerraform = exports.apprunnerServiceSourceConfigurationImageRepositoryToTerraform = exports.ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference = exports.apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToHclTerraform = exports.apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToTerraform = exports.ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference = exports.apprunnerServiceSourceConfigurationCodeRepositoryToHclTerraform = exports.apprunnerServiceSourceConfigurationCodeRepositoryToTerraform = exports.ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference = exports.apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToHclTerraform = exports.apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToTerraform = exports.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference = exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToHclTerraform = exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToTerraform = exports.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference = exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToHclTerraform = exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToTerraform = exports.ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference = exports.apprunnerServiceSourceConfigurationAuthenticationConfigurationToHclTerraform = exports.apprunnerServiceSourceConfigurationAuthenticationConfigurationToTerraform = exports.ApprunnerServiceObservabilityConfigurationOutputReference = exports.apprunnerServiceObservabilityConfigurationToHclTerraform = exports.apprunnerServiceObservabilityConfigurationToTerraform = exports.ApprunnerServiceNetworkConfigurationOutputReference = exports.apprunnerServiceNetworkConfigurationToHclTerraform = exports.apprunnerServiceNetworkConfigurationToTerraform = exports.ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference = exports.apprunnerServiceNetworkConfigurationIngressConfigurationToHclTerraform = exports.apprunnerServiceNetworkConfigurationIngressConfigurationToTerraform = exports.ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference = exports.apprunnerServiceNetworkConfigurationEgressConfigurationToHclTerraform = exports.apprunnerServiceNetworkConfigurationEgressConfigurationToTerraform = exports.ApprunnerServiceInstanceConfigurationOutputReference = exports.apprunnerServiceInstanceConfigurationToHclTerraform = exports.apprunnerServiceInstanceConfigurationToTerraform = exports.ApprunnerServiceHealthCheckConfigurationOutputReference = exports.apprunnerServiceHealthCheckConfigurationToHclTerraform = exports.apprunnerServiceHealthCheckConfigurationToTerraform = exports.ApprunnerServiceEncryptionConfigurationOutputReference = exports.apprunnerServiceEncryptionConfigurationToHclTerraform = exports.apprunnerServiceEncryptionConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apprunnerServiceEncryptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
    };
}
exports.apprunnerServiceEncryptionConfigurationToTerraform = apprunnerServiceEncryptionConfigurationToTerraform;
function apprunnerServiceEncryptionConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceEncryptionConfigurationToHclTerraform = apprunnerServiceEncryptionConfigurationToHclTerraform;
class ApprunnerServiceEncryptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKey = value.kmsKey;
        }
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
}
exports.ApprunnerServiceEncryptionConfigurationOutputReference = ApprunnerServiceEncryptionConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApprunnerServiceEncryptionConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceEncryptionConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceHealthCheckConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        healthy_threshold: cdktf.numberToTerraform(struct.healthyThreshold),
        interval: cdktf.numberToTerraform(struct.interval),
        path: cdktf.stringToTerraform(struct.path),
        protocol: cdktf.stringToTerraform(struct.protocol),
        timeout: cdktf.numberToTerraform(struct.timeout),
        unhealthy_threshold: cdktf.numberToTerraform(struct.unhealthyThreshold),
    };
}
exports.apprunnerServiceHealthCheckConfigurationToTerraform = apprunnerServiceHealthCheckConfigurationToTerraform;
function apprunnerServiceHealthCheckConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        healthy_threshold: {
            value: cdktf.numberToHclTerraform(struct.healthyThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval: {
            value: cdktf.numberToHclTerraform(struct.interval),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        timeout: {
            value: cdktf.numberToHclTerraform(struct.timeout),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        unhealthy_threshold: {
            value: cdktf.numberToHclTerraform(struct.unhealthyThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceHealthCheckConfigurationToHclTerraform = apprunnerServiceHealthCheckConfigurationToHclTerraform;
class ApprunnerServiceHealthCheckConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthyThreshold = this._healthyThreshold;
        }
        if (this._interval !== undefined) {
            hasAnyValues = true;
            internalValueResult.interval = this._interval;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._unhealthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.unhealthyThreshold = this._unhealthyThreshold;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._healthyThreshold = undefined;
            this._interval = undefined;
            this._path = undefined;
            this._protocol = undefined;
            this._timeout = undefined;
            this._unhealthyThreshold = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._healthyThreshold = value.healthyThreshold;
            this._interval = value.interval;
            this._path = value.path;
            this._protocol = value.protocol;
            this._timeout = value.timeout;
            this._unhealthyThreshold = value.unhealthyThreshold;
        }
    }
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    set healthyThreshold(value) {
        this._healthyThreshold = value;
    }
    resetHealthyThreshold() {
        this._healthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthyThresholdInput() {
        return this._healthyThreshold;
    }
    get interval() {
        return this.getNumberAttribute('interval');
    }
    set interval(value) {
        this._interval = value;
    }
    resetInterval() {
        this._interval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intervalInput() {
        return this._interval;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
    set unhealthyThreshold(value) {
        this._unhealthyThreshold = value;
    }
    resetUnhealthyThreshold() {
        this._unhealthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unhealthyThresholdInput() {
        return this._unhealthyThreshold;
    }
}
exports.ApprunnerServiceHealthCheckConfigurationOutputReference = ApprunnerServiceHealthCheckConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ApprunnerServiceHealthCheckConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceHealthCheckConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceInstanceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu: cdktf.stringToTerraform(struct.cpu),
        instance_role_arn: cdktf.stringToTerraform(struct.instanceRoleArn),
        memory: cdktf.stringToTerraform(struct.memory),
    };
}
exports.apprunnerServiceInstanceConfigurationToTerraform = apprunnerServiceInstanceConfigurationToTerraform;
function apprunnerServiceInstanceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cpu: {
            value: cdktf.stringToHclTerraform(struct.cpu),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        instance_role_arn: {
            value: cdktf.stringToHclTerraform(struct.instanceRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        memory: {
            value: cdktf.stringToHclTerraform(struct.memory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceInstanceConfigurationToHclTerraform = apprunnerServiceInstanceConfigurationToHclTerraform;
class ApprunnerServiceInstanceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpu !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpu = this._cpu;
        }
        if (this._instanceRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceRoleArn = this._instanceRoleArn;
        }
        if (this._memory !== undefined) {
            hasAnyValues = true;
            internalValueResult.memory = this._memory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpu = undefined;
            this._instanceRoleArn = undefined;
            this._memory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpu = value.cpu;
            this._instanceRoleArn = value.instanceRoleArn;
            this._memory = value.memory;
        }
    }
    get cpu() {
        return this.getStringAttribute('cpu');
    }
    set cpu(value) {
        this._cpu = value;
    }
    resetCpu() {
        this._cpu = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuInput() {
        return this._cpu;
    }
    get instanceRoleArn() {
        return this.getStringAttribute('instance_role_arn');
    }
    set instanceRoleArn(value) {
        this._instanceRoleArn = value;
    }
    resetInstanceRoleArn() {
        this._instanceRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceRoleArnInput() {
        return this._instanceRoleArn;
    }
    get memory() {
        return this.getStringAttribute('memory');
    }
    set memory(value) {
        this._memory = value;
    }
    resetMemory() {
        this._memory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memoryInput() {
        return this._memory;
    }
}
exports.ApprunnerServiceInstanceConfigurationOutputReference = ApprunnerServiceInstanceConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ApprunnerServiceInstanceConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceInstanceConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceNetworkConfigurationEgressConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        egress_type: cdktf.stringToTerraform(struct.egressType),
        vpc_connector_arn: cdktf.stringToTerraform(struct.vpcConnectorArn),
    };
}
exports.apprunnerServiceNetworkConfigurationEgressConfigurationToTerraform = apprunnerServiceNetworkConfigurationEgressConfigurationToTerraform;
function apprunnerServiceNetworkConfigurationEgressConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        egress_type: {
            value: cdktf.stringToHclTerraform(struct.egressType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vpc_connector_arn: {
            value: cdktf.stringToHclTerraform(struct.vpcConnectorArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceNetworkConfigurationEgressConfigurationToHclTerraform = apprunnerServiceNetworkConfigurationEgressConfigurationToHclTerraform;
class ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._egressType !== undefined) {
            hasAnyValues = true;
            internalValueResult.egressType = this._egressType;
        }
        if (this._vpcConnectorArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcConnectorArn = this._vpcConnectorArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._egressType = undefined;
            this._vpcConnectorArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._egressType = value.egressType;
            this._vpcConnectorArn = value.vpcConnectorArn;
        }
    }
    get egressType() {
        return this.getStringAttribute('egress_type');
    }
    set egressType(value) {
        this._egressType = value;
    }
    resetEgressType() {
        this._egressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressTypeInput() {
        return this._egressType;
    }
    get vpcConnectorArn() {
        return this.getStringAttribute('vpc_connector_arn');
    }
    set vpcConnectorArn(value) {
        this._vpcConnectorArn = value;
    }
    resetVpcConnectorArn() {
        this._vpcConnectorArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConnectorArnInput() {
        return this._vpcConnectorArn;
    }
}
exports.ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference = ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference[_d] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceNetworkConfigurationIngressConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_publicly_accessible: cdktf.booleanToTerraform(struct.isPubliclyAccessible),
    };
}
exports.apprunnerServiceNetworkConfigurationIngressConfigurationToTerraform = apprunnerServiceNetworkConfigurationIngressConfigurationToTerraform;
function apprunnerServiceNetworkConfigurationIngressConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        is_publicly_accessible: {
            value: cdktf.booleanToHclTerraform(struct.isPubliclyAccessible),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceNetworkConfigurationIngressConfigurationToHclTerraform = apprunnerServiceNetworkConfigurationIngressConfigurationToHclTerraform;
class ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isPubliclyAccessible !== undefined) {
            hasAnyValues = true;
            internalValueResult.isPubliclyAccessible = this._isPubliclyAccessible;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._isPubliclyAccessible = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._isPubliclyAccessible = value.isPubliclyAccessible;
        }
    }
    get isPubliclyAccessible() {
        return this.getBooleanAttribute('is_publicly_accessible');
    }
    set isPubliclyAccessible(value) {
        this._isPubliclyAccessible = value;
    }
    resetIsPubliclyAccessible() {
        this._isPubliclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isPubliclyAccessibleInput() {
        return this._isPubliclyAccessible;
    }
}
exports.ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference = ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ip_address_type: cdktf.stringToTerraform(struct.ipAddressType),
        egress_configuration: apprunnerServiceNetworkConfigurationEgressConfigurationToTerraform(struct.egressConfiguration),
        ingress_configuration: apprunnerServiceNetworkConfigurationIngressConfigurationToTerraform(struct.ingressConfiguration),
    };
}
exports.apprunnerServiceNetworkConfigurationToTerraform = apprunnerServiceNetworkConfigurationToTerraform;
function apprunnerServiceNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ip_address_type: {
            value: cdktf.stringToHclTerraform(struct.ipAddressType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        egress_configuration: {
            value: apprunnerServiceNetworkConfigurationEgressConfigurationToHclTerraform(struct.egressConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceNetworkConfigurationEgressConfigurationList",
        },
        ingress_configuration: {
            value: apprunnerServiceNetworkConfigurationIngressConfigurationToHclTerraform(struct.ingressConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceNetworkConfigurationIngressConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceNetworkConfigurationToHclTerraform = apprunnerServiceNetworkConfigurationToHclTerraform;
class ApprunnerServiceNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // egress_configuration - computed: false, optional: true, required: false
        this._egressConfiguration = new ApprunnerServiceNetworkConfigurationEgressConfigurationOutputReference(this, "egress_configuration");
        // ingress_configuration - computed: false, optional: true, required: false
        this._ingressConfiguration = new ApprunnerServiceNetworkConfigurationIngressConfigurationOutputReference(this, "ingress_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipAddressType !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipAddressType = this._ipAddressType;
        }
        if (this._egressConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.egressConfiguration = this._egressConfiguration?.internalValue;
        }
        if (this._ingressConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ingressConfiguration = this._ingressConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ipAddressType = undefined;
            this._egressConfiguration.internalValue = undefined;
            this._ingressConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ipAddressType = value.ipAddressType;
            this._egressConfiguration.internalValue = value.egressConfiguration;
            this._ingressConfiguration.internalValue = value.ingressConfiguration;
        }
    }
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    set ipAddressType(value) {
        this._ipAddressType = value;
    }
    resetIpAddressType() {
        this._ipAddressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressTypeInput() {
        return this._ipAddressType;
    }
    get egressConfiguration() {
        return this._egressConfiguration;
    }
    putEgressConfiguration(value) {
        this._egressConfiguration.internalValue = value;
    }
    resetEgressConfiguration() {
        this._egressConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressConfigurationInput() {
        return this._egressConfiguration.internalValue;
    }
    get ingressConfiguration() {
        return this._ingressConfiguration;
    }
    putIngressConfiguration(value) {
        this._ingressConfiguration.internalValue = value;
    }
    resetIngressConfiguration() {
        this._ingressConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressConfigurationInput() {
        return this._ingressConfiguration.internalValue;
    }
}
exports.ApprunnerServiceNetworkConfigurationOutputReference = ApprunnerServiceNetworkConfigurationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ApprunnerServiceNetworkConfigurationOutputReference[_f] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceNetworkConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceObservabilityConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        observability_configuration_arn: cdktf.stringToTerraform(struct.observabilityConfigurationArn),
        observability_enabled: cdktf.booleanToTerraform(struct.observabilityEnabled),
    };
}
exports.apprunnerServiceObservabilityConfigurationToTerraform = apprunnerServiceObservabilityConfigurationToTerraform;
function apprunnerServiceObservabilityConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        observability_configuration_arn: {
            value: cdktf.stringToHclTerraform(struct.observabilityConfigurationArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        observability_enabled: {
            value: cdktf.booleanToHclTerraform(struct.observabilityEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceObservabilityConfigurationToHclTerraform = apprunnerServiceObservabilityConfigurationToHclTerraform;
class ApprunnerServiceObservabilityConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._observabilityConfigurationArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.observabilityConfigurationArn = this._observabilityConfigurationArn;
        }
        if (this._observabilityEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.observabilityEnabled = this._observabilityEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._observabilityConfigurationArn = undefined;
            this._observabilityEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._observabilityConfigurationArn = value.observabilityConfigurationArn;
            this._observabilityEnabled = value.observabilityEnabled;
        }
    }
    get observabilityConfigurationArn() {
        return this.getStringAttribute('observability_configuration_arn');
    }
    set observabilityConfigurationArn(value) {
        this._observabilityConfigurationArn = value;
    }
    resetObservabilityConfigurationArn() {
        this._observabilityConfigurationArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get observabilityConfigurationArnInput() {
        return this._observabilityConfigurationArn;
    }
    get observabilityEnabled() {
        return this.getBooleanAttribute('observability_enabled');
    }
    set observabilityEnabled(value) {
        this._observabilityEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get observabilityEnabledInput() {
        return this._observabilityEnabled;
    }
}
exports.ApprunnerServiceObservabilityConfigurationOutputReference = ApprunnerServiceObservabilityConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ApprunnerServiceObservabilityConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceObservabilityConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationAuthenticationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_role_arn: cdktf.stringToTerraform(struct.accessRoleArn),
        connection_arn: cdktf.stringToTerraform(struct.connectionArn),
    };
}
exports.apprunnerServiceSourceConfigurationAuthenticationConfigurationToTerraform = apprunnerServiceSourceConfigurationAuthenticationConfigurationToTerraform;
function apprunnerServiceSourceConfigurationAuthenticationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_role_arn: {
            value: cdktf.stringToHclTerraform(struct.accessRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        connection_arn: {
            value: cdktf.stringToHclTerraform(struct.connectionArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationAuthenticationConfigurationToHclTerraform = apprunnerServiceSourceConfigurationAuthenticationConfigurationToHclTerraform;
class ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessRoleArn = this._accessRoleArn;
        }
        if (this._connectionArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionArn = this._connectionArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accessRoleArn = undefined;
            this._connectionArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accessRoleArn = value.accessRoleArn;
            this._connectionArn = value.connectionArn;
        }
    }
    get accessRoleArn() {
        return this.getStringAttribute('access_role_arn');
    }
    set accessRoleArn(value) {
        this._accessRoleArn = value;
    }
    resetAccessRoleArn() {
        this._accessRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessRoleArnInput() {
        return this._accessRoleArn;
    }
    get connectionArn() {
        return this.getStringAttribute('connection_arn');
    }
    set connectionArn(value) {
        this._connectionArn = value;
    }
    resetConnectionArn() {
        this._connectionArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionArnInput() {
        return this._connectionArn;
    }
}
exports.ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference = ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference[_h] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        build_command: cdktf.stringToTerraform(struct.buildCommand),
        port: cdktf.stringToTerraform(struct.port),
        runtime: cdktf.stringToTerraform(struct.runtime),
        runtime_environment_secrets: cdktf.hashMapper(cdktf.stringToTerraform)(struct.runtimeEnvironmentSecrets),
        runtime_environment_variables: cdktf.hashMapper(cdktf.stringToTerraform)(struct.runtimeEnvironmentVariables),
        start_command: cdktf.stringToTerraform(struct.startCommand),
    };
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToTerraform = apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToTerraform;
function apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        build_command: {
            value: cdktf.stringToHclTerraform(struct.buildCommand),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.stringToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        runtime: {
            value: cdktf.stringToHclTerraform(struct.runtime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        runtime_environment_secrets: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.runtimeEnvironmentSecrets),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        runtime_environment_variables: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.runtimeEnvironmentVariables),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        start_command: {
            value: cdktf.stringToHclTerraform(struct.startCommand),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToHclTerraform = apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToHclTerraform;
class ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._buildCommand !== undefined) {
            hasAnyValues = true;
            internalValueResult.buildCommand = this._buildCommand;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._runtime !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtime = this._runtime;
        }
        if (this._runtimeEnvironmentSecrets !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeEnvironmentSecrets = this._runtimeEnvironmentSecrets;
        }
        if (this._runtimeEnvironmentVariables !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeEnvironmentVariables = this._runtimeEnvironmentVariables;
        }
        if (this._startCommand !== undefined) {
            hasAnyValues = true;
            internalValueResult.startCommand = this._startCommand;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._buildCommand = undefined;
            this._port = undefined;
            this._runtime = undefined;
            this._runtimeEnvironmentSecrets = undefined;
            this._runtimeEnvironmentVariables = undefined;
            this._startCommand = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._buildCommand = value.buildCommand;
            this._port = value.port;
            this._runtime = value.runtime;
            this._runtimeEnvironmentSecrets = value.runtimeEnvironmentSecrets;
            this._runtimeEnvironmentVariables = value.runtimeEnvironmentVariables;
            this._startCommand = value.startCommand;
        }
    }
    get buildCommand() {
        return this.getStringAttribute('build_command');
    }
    set buildCommand(value) {
        this._buildCommand = value;
    }
    resetBuildCommand() {
        this._buildCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get buildCommandInput() {
        return this._buildCommand;
    }
    get port() {
        return this.getStringAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get runtime() {
        return this.getStringAttribute('runtime');
    }
    set runtime(value) {
        this._runtime = value;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeInput() {
        return this._runtime;
    }
    get runtimeEnvironmentSecrets() {
        return this.getStringMapAttribute('runtime_environment_secrets');
    }
    set runtimeEnvironmentSecrets(value) {
        this._runtimeEnvironmentSecrets = value;
    }
    resetRuntimeEnvironmentSecrets() {
        this._runtimeEnvironmentSecrets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeEnvironmentSecretsInput() {
        return this._runtimeEnvironmentSecrets;
    }
    get runtimeEnvironmentVariables() {
        return this.getStringMapAttribute('runtime_environment_variables');
    }
    set runtimeEnvironmentVariables(value) {
        this._runtimeEnvironmentVariables = value;
    }
    resetRuntimeEnvironmentVariables() {
        this._runtimeEnvironmentVariables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeEnvironmentVariablesInput() {
        return this._runtimeEnvironmentVariables;
    }
    get startCommand() {
        return this.getStringAttribute('start_command');
    }
    set startCommand(value) {
        this._startCommand = value;
    }
    resetStartCommand() {
        this._startCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startCommandInput() {
        return this._startCommand;
    }
}
exports.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference = ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference[_j] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        configuration_source: cdktf.stringToTerraform(struct.configurationSource),
        code_configuration_values: apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToTerraform(struct.codeConfigurationValues),
    };
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToTerraform = apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToTerraform;
function apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        configuration_source: {
            value: cdktf.stringToHclTerraform(struct.configurationSource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        code_configuration_values: {
            value: apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesToHclTerraform(struct.codeConfigurationValues),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToHclTerraform = apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToHclTerraform;
class ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // code_configuration_values - computed: false, optional: true, required: false
        this._codeConfigurationValues = new ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationCodeConfigurationValuesOutputReference(this, "code_configuration_values");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._configurationSource !== undefined) {
            hasAnyValues = true;
            internalValueResult.configurationSource = this._configurationSource;
        }
        if (this._codeConfigurationValues?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.codeConfigurationValues = this._codeConfigurationValues?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._configurationSource = undefined;
            this._codeConfigurationValues.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._configurationSource = value.configurationSource;
            this._codeConfigurationValues.internalValue = value.codeConfigurationValues;
        }
    }
    get configurationSource() {
        return this.getStringAttribute('configuration_source');
    }
    set configurationSource(value) {
        this._configurationSource = value;
    }
    // Temporarily expose input value. Use with caution.
    get configurationSourceInput() {
        return this._configurationSource;
    }
    get codeConfigurationValues() {
        return this._codeConfigurationValues;
    }
    putCodeConfigurationValues(value) {
        this._codeConfigurationValues.internalValue = value;
    }
    resetCodeConfigurationValues() {
        this._codeConfigurationValues.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codeConfigurationValuesInput() {
        return this._codeConfigurationValues.internalValue;
    }
}
exports.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference = ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference[_k] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToTerraform = apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToTerraform;
function apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToHclTerraform = apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToHclTerraform;
class ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference = ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference[_l] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationCodeRepositoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_url: cdktf.stringToTerraform(struct.repositoryUrl),
        source_directory: cdktf.stringToTerraform(struct.sourceDirectory),
        code_configuration: apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToTerraform(struct.codeConfiguration),
        source_code_version: apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToTerraform(struct.sourceCodeVersion),
    };
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryToTerraform = apprunnerServiceSourceConfigurationCodeRepositoryToTerraform;
function apprunnerServiceSourceConfigurationCodeRepositoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository_url: {
            value: cdktf.stringToHclTerraform(struct.repositoryUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_directory: {
            value: cdktf.stringToHclTerraform(struct.sourceDirectory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        code_configuration: {
            value: apprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationToHclTerraform(struct.codeConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationList",
        },
        source_code_version: {
            value: apprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionToHclTerraform(struct.sourceCodeVersion),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationCodeRepositoryToHclTerraform = apprunnerServiceSourceConfigurationCodeRepositoryToHclTerraform;
class ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // code_configuration - computed: false, optional: true, required: false
        this._codeConfiguration = new ApprunnerServiceSourceConfigurationCodeRepositoryCodeConfigurationOutputReference(this, "code_configuration");
        // source_code_version - computed: false, optional: false, required: true
        this._sourceCodeVersion = new ApprunnerServiceSourceConfigurationCodeRepositorySourceCodeVersionOutputReference(this, "source_code_version");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryUrl = this._repositoryUrl;
        }
        if (this._sourceDirectory !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceDirectory = this._sourceDirectory;
        }
        if (this._codeConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.codeConfiguration = this._codeConfiguration?.internalValue;
        }
        if (this._sourceCodeVersion?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceCodeVersion = this._sourceCodeVersion?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryUrl = undefined;
            this._sourceDirectory = undefined;
            this._codeConfiguration.internalValue = undefined;
            this._sourceCodeVersion.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryUrl = value.repositoryUrl;
            this._sourceDirectory = value.sourceDirectory;
            this._codeConfiguration.internalValue = value.codeConfiguration;
            this._sourceCodeVersion.internalValue = value.sourceCodeVersion;
        }
    }
    get repositoryUrl() {
        return this.getStringAttribute('repository_url');
    }
    set repositoryUrl(value) {
        this._repositoryUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryUrlInput() {
        return this._repositoryUrl;
    }
    get sourceDirectory() {
        return this.getStringAttribute('source_directory');
    }
    set sourceDirectory(value) {
        this._sourceDirectory = value;
    }
    resetSourceDirectory() {
        this._sourceDirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDirectoryInput() {
        return this._sourceDirectory;
    }
    get codeConfiguration() {
        return this._codeConfiguration;
    }
    putCodeConfiguration(value) {
        this._codeConfiguration.internalValue = value;
    }
    resetCodeConfiguration() {
        this._codeConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codeConfigurationInput() {
        return this._codeConfiguration.internalValue;
    }
    get sourceCodeVersion() {
        return this._sourceCodeVersion;
    }
    putSourceCodeVersion(value) {
        this._sourceCodeVersion.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceCodeVersionInput() {
        return this._sourceCodeVersion.internalValue;
    }
}
exports.ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference = ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference[_m] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.stringToTerraform(struct.port),
        runtime_environment_secrets: cdktf.hashMapper(cdktf.stringToTerraform)(struct.runtimeEnvironmentSecrets),
        runtime_environment_variables: cdktf.hashMapper(cdktf.stringToTerraform)(struct.runtimeEnvironmentVariables),
        start_command: cdktf.stringToTerraform(struct.startCommand),
    };
}
exports.apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToTerraform = apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToTerraform;
function apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        port: {
            value: cdktf.stringToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        runtime_environment_secrets: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.runtimeEnvironmentSecrets),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        runtime_environment_variables: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.runtimeEnvironmentVariables),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        start_command: {
            value: cdktf.stringToHclTerraform(struct.startCommand),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToHclTerraform = apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToHclTerraform;
class ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._runtimeEnvironmentSecrets !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeEnvironmentSecrets = this._runtimeEnvironmentSecrets;
        }
        if (this._runtimeEnvironmentVariables !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeEnvironmentVariables = this._runtimeEnvironmentVariables;
        }
        if (this._startCommand !== undefined) {
            hasAnyValues = true;
            internalValueResult.startCommand = this._startCommand;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
            this._runtimeEnvironmentSecrets = undefined;
            this._runtimeEnvironmentVariables = undefined;
            this._startCommand = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
            this._runtimeEnvironmentSecrets = value.runtimeEnvironmentSecrets;
            this._runtimeEnvironmentVariables = value.runtimeEnvironmentVariables;
            this._startCommand = value.startCommand;
        }
    }
    get port() {
        return this.getStringAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get runtimeEnvironmentSecrets() {
        return this.getStringMapAttribute('runtime_environment_secrets');
    }
    set runtimeEnvironmentSecrets(value) {
        this._runtimeEnvironmentSecrets = value;
    }
    resetRuntimeEnvironmentSecrets() {
        this._runtimeEnvironmentSecrets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeEnvironmentSecretsInput() {
        return this._runtimeEnvironmentSecrets;
    }
    get runtimeEnvironmentVariables() {
        return this.getStringMapAttribute('runtime_environment_variables');
    }
    set runtimeEnvironmentVariables(value) {
        this._runtimeEnvironmentVariables = value;
    }
    resetRuntimeEnvironmentVariables() {
        this._runtimeEnvironmentVariables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeEnvironmentVariablesInput() {
        return this._runtimeEnvironmentVariables;
    }
    get startCommand() {
        return this.getStringAttribute('start_command');
    }
    set startCommand(value) {
        this._startCommand = value;
    }
    resetStartCommand() {
        this._startCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startCommandInput() {
        return this._startCommand;
    }
}
exports.ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference = ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference[_o] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationImageRepositoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        image_identifier: cdktf.stringToTerraform(struct.imageIdentifier),
        image_repository_type: cdktf.stringToTerraform(struct.imageRepositoryType),
        image_configuration: apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToTerraform(struct.imageConfiguration),
    };
}
exports.apprunnerServiceSourceConfigurationImageRepositoryToTerraform = apprunnerServiceSourceConfigurationImageRepositoryToTerraform;
function apprunnerServiceSourceConfigurationImageRepositoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        image_identifier: {
            value: cdktf.stringToHclTerraform(struct.imageIdentifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        image_repository_type: {
            value: cdktf.stringToHclTerraform(struct.imageRepositoryType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        image_configuration: {
            value: apprunnerServiceSourceConfigurationImageRepositoryImageConfigurationToHclTerraform(struct.imageConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationImageRepositoryToHclTerraform = apprunnerServiceSourceConfigurationImageRepositoryToHclTerraform;
class ApprunnerServiceSourceConfigurationImageRepositoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // image_configuration - computed: false, optional: true, required: false
        this._imageConfiguration = new ApprunnerServiceSourceConfigurationImageRepositoryImageConfigurationOutputReference(this, "image_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._imageIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageIdentifier = this._imageIdentifier;
        }
        if (this._imageRepositoryType !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageRepositoryType = this._imageRepositoryType;
        }
        if (this._imageConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageConfiguration = this._imageConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._imageIdentifier = undefined;
            this._imageRepositoryType = undefined;
            this._imageConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._imageIdentifier = value.imageIdentifier;
            this._imageRepositoryType = value.imageRepositoryType;
            this._imageConfiguration.internalValue = value.imageConfiguration;
        }
    }
    get imageIdentifier() {
        return this.getStringAttribute('image_identifier');
    }
    set imageIdentifier(value) {
        this._imageIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageIdentifierInput() {
        return this._imageIdentifier;
    }
    get imageRepositoryType() {
        return this.getStringAttribute('image_repository_type');
    }
    set imageRepositoryType(value) {
        this._imageRepositoryType = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageRepositoryTypeInput() {
        return this._imageRepositoryType;
    }
    get imageConfiguration() {
        return this._imageConfiguration;
    }
    putImageConfiguration(value) {
        this._imageConfiguration.internalValue = value;
    }
    resetImageConfiguration() {
        this._imageConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigurationInput() {
        return this._imageConfiguration.internalValue;
    }
}
exports.ApprunnerServiceSourceConfigurationImageRepositoryOutputReference = ApprunnerServiceSourceConfigurationImageRepositoryOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationImageRepositoryOutputReference[_p] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationImageRepositoryOutputReference", version: "19.15.0" };
function apprunnerServiceSourceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_deployments_enabled: cdktf.booleanToTerraform(struct.autoDeploymentsEnabled),
        authentication_configuration: apprunnerServiceSourceConfigurationAuthenticationConfigurationToTerraform(struct.authenticationConfiguration),
        code_repository: apprunnerServiceSourceConfigurationCodeRepositoryToTerraform(struct.codeRepository),
        image_repository: apprunnerServiceSourceConfigurationImageRepositoryToTerraform(struct.imageRepository),
    };
}
exports.apprunnerServiceSourceConfigurationToTerraform = apprunnerServiceSourceConfigurationToTerraform;
function apprunnerServiceSourceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_deployments_enabled: {
            value: cdktf.booleanToHclTerraform(struct.autoDeploymentsEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        authentication_configuration: {
            value: apprunnerServiceSourceConfigurationAuthenticationConfigurationToHclTerraform(struct.authenticationConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationAuthenticationConfigurationList",
        },
        code_repository: {
            value: apprunnerServiceSourceConfigurationCodeRepositoryToHclTerraform(struct.codeRepository),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationCodeRepositoryList",
        },
        image_repository: {
            value: apprunnerServiceSourceConfigurationImageRepositoryToHclTerraform(struct.imageRepository),
            isBlock: true,
            type: "list",
            storageClassType: "ApprunnerServiceSourceConfigurationImageRepositoryList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerServiceSourceConfigurationToHclTerraform = apprunnerServiceSourceConfigurationToHclTerraform;
class ApprunnerServiceSourceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // authentication_configuration - computed: false, optional: true, required: false
        this._authenticationConfiguration = new ApprunnerServiceSourceConfigurationAuthenticationConfigurationOutputReference(this, "authentication_configuration");
        // code_repository - computed: false, optional: true, required: false
        this._codeRepository = new ApprunnerServiceSourceConfigurationCodeRepositoryOutputReference(this, "code_repository");
        // image_repository - computed: false, optional: true, required: false
        this._imageRepository = new ApprunnerServiceSourceConfigurationImageRepositoryOutputReference(this, "image_repository");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoDeploymentsEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDeploymentsEnabled = this._autoDeploymentsEnabled;
        }
        if (this._authenticationConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.authenticationConfiguration = this._authenticationConfiguration?.internalValue;
        }
        if (this._codeRepository?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.codeRepository = this._codeRepository?.internalValue;
        }
        if (this._imageRepository?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageRepository = this._imageRepository?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoDeploymentsEnabled = undefined;
            this._authenticationConfiguration.internalValue = undefined;
            this._codeRepository.internalValue = undefined;
            this._imageRepository.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoDeploymentsEnabled = value.autoDeploymentsEnabled;
            this._authenticationConfiguration.internalValue = value.authenticationConfiguration;
            this._codeRepository.internalValue = value.codeRepository;
            this._imageRepository.internalValue = value.imageRepository;
        }
    }
    get autoDeploymentsEnabled() {
        return this.getBooleanAttribute('auto_deployments_enabled');
    }
    set autoDeploymentsEnabled(value) {
        this._autoDeploymentsEnabled = value;
    }
    resetAutoDeploymentsEnabled() {
        this._autoDeploymentsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeploymentsEnabledInput() {
        return this._autoDeploymentsEnabled;
    }
    get authenticationConfiguration() {
        return this._authenticationConfiguration;
    }
    putAuthenticationConfiguration(value) {
        this._authenticationConfiguration.internalValue = value;
    }
    resetAuthenticationConfiguration() {
        this._authenticationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationConfigurationInput() {
        return this._authenticationConfiguration.internalValue;
    }
    get codeRepository() {
        return this._codeRepository;
    }
    putCodeRepository(value) {
        this._codeRepository.internalValue = value;
    }
    resetCodeRepository() {
        this._codeRepository.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codeRepositoryInput() {
        return this._codeRepository.internalValue;
    }
    get imageRepository() {
        return this._imageRepository;
    }
    putImageRepository(value) {
        this._imageRepository.internalValue = value;
    }
    resetImageRepository() {
        this._imageRepository.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageRepositoryInput() {
        return this._imageRepository.internalValue;
    }
}
exports.ApprunnerServiceSourceConfigurationOutputReference = ApprunnerServiceSourceConfigurationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ApprunnerServiceSourceConfigurationOutputReference[_q] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerServiceSourceConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_service aws_apprunner_service}
*/
class ApprunnerService extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApprunnerService resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApprunnerService to import
    * @param importFromId The id of the existing ApprunnerService that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_service#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApprunnerService to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_apprunner_service", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_service aws_apprunner_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApprunnerServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apprunner_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // encryption_configuration - computed: false, optional: true, required: false
        this._encryptionConfiguration = new ApprunnerServiceEncryptionConfigurationOutputReference(this, "encryption_configuration");
        // health_check_configuration - computed: false, optional: true, required: false
        this._healthCheckConfiguration = new ApprunnerServiceHealthCheckConfigurationOutputReference(this, "health_check_configuration");
        // instance_configuration - computed: false, optional: true, required: false
        this._instanceConfiguration = new ApprunnerServiceInstanceConfigurationOutputReference(this, "instance_configuration");
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new ApprunnerServiceNetworkConfigurationOutputReference(this, "network_configuration");
        // observability_configuration - computed: false, optional: true, required: false
        this._observabilityConfiguration = new ApprunnerServiceObservabilityConfigurationOutputReference(this, "observability_configuration");
        // source_configuration - computed: false, optional: false, required: true
        this._sourceConfiguration = new ApprunnerServiceSourceConfigurationOutputReference(this, "source_configuration");
        this._autoScalingConfigurationArn = config.autoScalingConfigurationArn;
        this._id = config.id;
        this._serviceName = config.serviceName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._encryptionConfiguration.internalValue = config.encryptionConfiguration;
        this._healthCheckConfiguration.internalValue = config.healthCheckConfiguration;
        this._instanceConfiguration.internalValue = config.instanceConfiguration;
        this._networkConfiguration.internalValue = config.networkConfiguration;
        this._observabilityConfiguration.internalValue = config.observabilityConfiguration;
        this._sourceConfiguration.internalValue = config.sourceConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoScalingConfigurationArn() {
        return this.getStringAttribute('auto_scaling_configuration_arn');
    }
    set autoScalingConfigurationArn(value) {
        this._autoScalingConfigurationArn = value;
    }
    resetAutoScalingConfigurationArn() {
        this._autoScalingConfigurationArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoScalingConfigurationArnInput() {
        return this._autoScalingConfigurationArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // service_id - computed: true, optional: false, required: false
    get serviceId() {
        return this.getStringAttribute('service_id');
    }
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    set serviceName(value) {
        this._serviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNameInput() {
        return this._serviceName;
    }
    // service_url - computed: true, optional: false, required: false
    get serviceUrl() {
        return this.getStringAttribute('service_url');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get encryptionConfiguration() {
        return this._encryptionConfiguration;
    }
    putEncryptionConfiguration(value) {
        this._encryptionConfiguration.internalValue = value;
    }
    resetEncryptionConfiguration() {
        this._encryptionConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionConfigurationInput() {
        return this._encryptionConfiguration.internalValue;
    }
    get healthCheckConfiguration() {
        return this._healthCheckConfiguration;
    }
    putHealthCheckConfiguration(value) {
        this._healthCheckConfiguration.internalValue = value;
    }
    resetHealthCheckConfiguration() {
        this._healthCheckConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckConfigurationInput() {
        return this._healthCheckConfiguration.internalValue;
    }
    get instanceConfiguration() {
        return this._instanceConfiguration;
    }
    putInstanceConfiguration(value) {
        this._instanceConfiguration.internalValue = value;
    }
    resetInstanceConfiguration() {
        this._instanceConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceConfigurationInput() {
        return this._instanceConfiguration.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get observabilityConfiguration() {
        return this._observabilityConfiguration;
    }
    putObservabilityConfiguration(value) {
        this._observabilityConfiguration.internalValue = value;
    }
    resetObservabilityConfiguration() {
        this._observabilityConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get observabilityConfigurationInput() {
        return this._observabilityConfiguration.internalValue;
    }
    get sourceConfiguration() {
        return this._sourceConfiguration;
    }
    putSourceConfiguration(value) {
        this._sourceConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceConfigurationInput() {
        return this._sourceConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_scaling_configuration_arn: cdktf.stringToTerraform(this._autoScalingConfigurationArn),
            id: cdktf.stringToTerraform(this._id),
            service_name: cdktf.stringToTerraform(this._serviceName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            encryption_configuration: apprunnerServiceEncryptionConfigurationToTerraform(this._encryptionConfiguration.internalValue),
            health_check_configuration: apprunnerServiceHealthCheckConfigurationToTerraform(this._healthCheckConfiguration.internalValue),
            instance_configuration: apprunnerServiceInstanceConfigurationToTerraform(this._instanceConfiguration.internalValue),
            network_configuration: apprunnerServiceNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
            observability_configuration: apprunnerServiceObservabilityConfigurationToTerraform(this._observabilityConfiguration.internalValue),
            source_configuration: apprunnerServiceSourceConfigurationToTerraform(this._sourceConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_scaling_configuration_arn: {
                value: cdktf.stringToHclTerraform(this._autoScalingConfigurationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_name: {
                value: cdktf.stringToHclTerraform(this._serviceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            encryption_configuration: {
                value: apprunnerServiceEncryptionConfigurationToHclTerraform(this._encryptionConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceEncryptionConfigurationList",
            },
            health_check_configuration: {
                value: apprunnerServiceHealthCheckConfigurationToHclTerraform(this._healthCheckConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceHealthCheckConfigurationList",
            },
            instance_configuration: {
                value: apprunnerServiceInstanceConfigurationToHclTerraform(this._instanceConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceInstanceConfigurationList",
            },
            network_configuration: {
                value: apprunnerServiceNetworkConfigurationToHclTerraform(this._networkConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceNetworkConfigurationList",
            },
            observability_configuration: {
                value: apprunnerServiceObservabilityConfigurationToHclTerraform(this._observabilityConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceObservabilityConfigurationList",
            },
            source_configuration: {
                value: apprunnerServiceSourceConfigurationToHclTerraform(this._sourceConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerServiceSourceConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApprunnerService = ApprunnerService;
_r = JSII_RTTI_SYMBOL_1;
ApprunnerService[_r] = { fqn: "@cdktf/provider-aws.apprunnerService.ApprunnerService", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ApprunnerService.tfResourceType = "aws_apprunner_service";
