// @cdktf/provider-aws/lib/iot-authorizer/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotAuthorizer = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_authorizer aws_iot_authorizer}
*/
class IotAuthorizer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IotAuthorizer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IotAuthorizer to import
    * @param importFromId The id of the existing IotAuthorizer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_authorizer#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IotAuthorizer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iot_authorizer", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_authorizer aws_iot_authorizer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotAuthorizerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_authorizer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._authorizerFunctionArn = config.authorizerFunctionArn;
        this._enableCachingForHttp = config.enableCachingForHttp;
        this._id = config.id;
        this._name = config.name;
        this._signingDisabled = config.signingDisabled;
        this._status = config.status;
        this._tokenKeyName = config.tokenKeyName;
        this._tokenSigningPublicKeys = config.tokenSigningPublicKeys;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authorizerFunctionArn() {
        return this.getStringAttribute('authorizer_function_arn');
    }
    set authorizerFunctionArn(value) {
        this._authorizerFunctionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerFunctionArnInput() {
        return this._authorizerFunctionArn;
    }
    get enableCachingForHttp() {
        return this.getBooleanAttribute('enable_caching_for_http');
    }
    set enableCachingForHttp(value) {
        this._enableCachingForHttp = value;
    }
    resetEnableCachingForHttp() {
        this._enableCachingForHttp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableCachingForHttpInput() {
        return this._enableCachingForHttp;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get signingDisabled() {
        return this.getBooleanAttribute('signing_disabled');
    }
    set signingDisabled(value) {
        this._signingDisabled = value;
    }
    resetSigningDisabled() {
        this._signingDisabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get signingDisabledInput() {
        return this._signingDisabled;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get tokenKeyName() {
        return this.getStringAttribute('token_key_name');
    }
    set tokenKeyName(value) {
        this._tokenKeyName = value;
    }
    resetTokenKeyName() {
        this._tokenKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tokenKeyNameInput() {
        return this._tokenKeyName;
    }
    get tokenSigningPublicKeys() {
        return this.getStringMapAttribute('token_signing_public_keys');
    }
    set tokenSigningPublicKeys(value) {
        this._tokenSigningPublicKeys = value;
    }
    resetTokenSigningPublicKeys() {
        this._tokenSigningPublicKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tokenSigningPublicKeysInput() {
        return this._tokenSigningPublicKeys;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authorizer_function_arn: cdktf.stringToTerraform(this._authorizerFunctionArn),
            enable_caching_for_http: cdktf.booleanToTerraform(this._enableCachingForHttp),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            signing_disabled: cdktf.booleanToTerraform(this._signingDisabled),
            status: cdktf.stringToTerraform(this._status),
            token_key_name: cdktf.stringToTerraform(this._tokenKeyName),
            token_signing_public_keys: cdktf.hashMapper(cdktf.stringToTerraform)(this._tokenSigningPublicKeys),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            authorizer_function_arn: {
                value: cdktf.stringToHclTerraform(this._authorizerFunctionArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enable_caching_for_http: {
                value: cdktf.booleanToHclTerraform(this._enableCachingForHttp),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            signing_disabled: {
                value: cdktf.booleanToHclTerraform(this._signingDisabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            token_key_name: {
                value: cdktf.stringToHclTerraform(this._tokenKeyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            token_signing_public_keys: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tokenSigningPublicKeys),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IotAuthorizer = IotAuthorizer;
_a = JSII_RTTI_SYMBOL_1;
IotAuthorizer[_a] = { fqn: "@cdktf/provider-aws.iotAuthorizer.IotAuthorizer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IotAuthorizer.tfResourceType = "aws_iot_authorizer";
