// @cdktf/provider-aws/lib/elastic-beanstalk-environment/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticBeanstalkEnvironment = exports.ElasticBeanstalkEnvironmentSettingList = exports.ElasticBeanstalkEnvironmentSettingOutputReference = exports.elasticBeanstalkEnvironmentSettingToHclTerraform = exports.elasticBeanstalkEnvironmentSettingToTerraform = exports.ElasticBeanstalkEnvironmentAllSettingsList = exports.ElasticBeanstalkEnvironmentAllSettingsOutputReference = exports.elasticBeanstalkEnvironmentAllSettingsToHclTerraform = exports.elasticBeanstalkEnvironmentAllSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elasticBeanstalkEnvironmentAllSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.elasticBeanstalkEnvironmentAllSettingsToTerraform = elasticBeanstalkEnvironmentAllSettingsToTerraform;
function elasticBeanstalkEnvironmentAllSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.elasticBeanstalkEnvironmentAllSettingsToHclTerraform = elasticBeanstalkEnvironmentAllSettingsToHclTerraform;
class ElasticBeanstalkEnvironmentAllSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // resource - computed: true, optional: false, required: false
    get resource() {
        return this.getStringAttribute('resource');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.ElasticBeanstalkEnvironmentAllSettingsOutputReference = ElasticBeanstalkEnvironmentAllSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkEnvironmentAllSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.elasticBeanstalkEnvironment.ElasticBeanstalkEnvironmentAllSettingsOutputReference", version: "19.15.0" };
class ElasticBeanstalkEnvironmentAllSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElasticBeanstalkEnvironmentAllSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElasticBeanstalkEnvironmentAllSettingsList = ElasticBeanstalkEnvironmentAllSettingsList;
_b = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkEnvironmentAllSettingsList[_b] = { fqn: "@cdktf/provider-aws.elasticBeanstalkEnvironment.ElasticBeanstalkEnvironmentAllSettingsList", version: "19.15.0" };
function elasticBeanstalkEnvironmentSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        namespace: cdktf.stringToTerraform(struct.namespace),
        resource: cdktf.stringToTerraform(struct.resource),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.elasticBeanstalkEnvironmentSettingToTerraform = elasticBeanstalkEnvironmentSettingToTerraform;
function elasticBeanstalkEnvironmentSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        namespace: {
            value: cdktf.stringToHclTerraform(struct.namespace),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resource: {
            value: cdktf.stringToHclTerraform(struct.resource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.elasticBeanstalkEnvironmentSettingToHclTerraform = elasticBeanstalkEnvironmentSettingToHclTerraform;
class ElasticBeanstalkEnvironmentSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._namespace !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespace = this._namespace;
        }
        if (this._resource !== undefined) {
            hasAnyValues = true;
            internalValueResult.resource = this._resource;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._namespace = undefined;
            this._resource = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._namespace = value.namespace;
            this._resource = value.resource;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get resource() {
        return this.getStringAttribute('resource');
    }
    set resource(value) {
        this._resource = value;
    }
    resetResource() {
        this._resource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceInput() {
        return this._resource;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ElasticBeanstalkEnvironmentSettingOutputReference = ElasticBeanstalkEnvironmentSettingOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkEnvironmentSettingOutputReference[_c] = { fqn: "@cdktf/provider-aws.elasticBeanstalkEnvironment.ElasticBeanstalkEnvironmentSettingOutputReference", version: "19.15.0" };
class ElasticBeanstalkEnvironmentSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElasticBeanstalkEnvironmentSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElasticBeanstalkEnvironmentSettingList = ElasticBeanstalkEnvironmentSettingList;
_d = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkEnvironmentSettingList[_d] = { fqn: "@cdktf/provider-aws.elasticBeanstalkEnvironment.ElasticBeanstalkEnvironmentSettingList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_environment aws_elastic_beanstalk_environment}
*/
class ElasticBeanstalkEnvironment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ElasticBeanstalkEnvironment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ElasticBeanstalkEnvironment to import
    * @param importFromId The id of the existing ElasticBeanstalkEnvironment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_environment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ElasticBeanstalkEnvironment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_elastic_beanstalk_environment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/elastic_beanstalk_environment aws_elastic_beanstalk_environment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticBeanstalkEnvironmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastic_beanstalk_environment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // all_settings - computed: true, optional: false, required: false
        this._allSettings = new ElasticBeanstalkEnvironmentAllSettingsList(this, "all_settings", true);
        // setting - computed: false, optional: true, required: false
        this._setting = new ElasticBeanstalkEnvironmentSettingList(this, "setting", true);
        this._application = config.application;
        this._cnamePrefix = config.cnamePrefix;
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._platformArn = config.platformArn;
        this._pollInterval = config.pollInterval;
        this._solutionStackName = config.solutionStackName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._templateName = config.templateName;
        this._tier = config.tier;
        this._versionLabel = config.versionLabel;
        this._waitForReadyTimeout = config.waitForReadyTimeout;
        this._setting.internalValue = config.setting;
    }
    get allSettings() {
        return this._allSettings;
    }
    get application() {
        return this.getStringAttribute('application');
    }
    set application(value) {
        this._application = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationInput() {
        return this._application;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // autoscaling_groups - computed: true, optional: false, required: false
    get autoscalingGroups() {
        return this.getListAttribute('autoscaling_groups');
    }
    // cname - computed: true, optional: false, required: false
    get cname() {
        return this.getStringAttribute('cname');
    }
    get cnamePrefix() {
        return this.getStringAttribute('cname_prefix');
    }
    set cnamePrefix(value) {
        this._cnamePrefix = value;
    }
    resetCnamePrefix() {
        this._cnamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cnamePrefixInput() {
        return this._cnamePrefix;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // endpoint_url - computed: true, optional: false, required: false
    get endpointUrl() {
        return this.getStringAttribute('endpoint_url');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instances - computed: true, optional: false, required: false
    get instances() {
        return this.getListAttribute('instances');
    }
    // launch_configurations - computed: true, optional: false, required: false
    get launchConfigurations() {
        return this.getListAttribute('launch_configurations');
    }
    // load_balancers - computed: true, optional: false, required: false
    get loadBalancers() {
        return this.getListAttribute('load_balancers');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get platformArn() {
        return this.getStringAttribute('platform_arn');
    }
    set platformArn(value) {
        this._platformArn = value;
    }
    resetPlatformArn() {
        this._platformArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformArnInput() {
        return this._platformArn;
    }
    get pollInterval() {
        return this.getStringAttribute('poll_interval');
    }
    set pollInterval(value) {
        this._pollInterval = value;
    }
    resetPollInterval() {
        this._pollInterval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pollIntervalInput() {
        return this._pollInterval;
    }
    // queues - computed: true, optional: false, required: false
    get queues() {
        return this.getListAttribute('queues');
    }
    get solutionStackName() {
        return this.getStringAttribute('solution_stack_name');
    }
    set solutionStackName(value) {
        this._solutionStackName = value;
    }
    resetSolutionStackName() {
        this._solutionStackName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get solutionStackNameInput() {
        return this._solutionStackName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get templateName() {
        return this.getStringAttribute('template_name');
    }
    set templateName(value) {
        this._templateName = value;
    }
    resetTemplateName() {
        this._templateName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateNameInput() {
        return this._templateName;
    }
    get tier() {
        return this.getStringAttribute('tier');
    }
    set tier(value) {
        this._tier = value;
    }
    resetTier() {
        this._tier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tierInput() {
        return this._tier;
    }
    // triggers - computed: true, optional: false, required: false
    get triggers() {
        return this.getListAttribute('triggers');
    }
    get versionLabel() {
        return this.getStringAttribute('version_label');
    }
    set versionLabel(value) {
        this._versionLabel = value;
    }
    resetVersionLabel() {
        this._versionLabel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionLabelInput() {
        return this._versionLabel;
    }
    get waitForReadyTimeout() {
        return this.getStringAttribute('wait_for_ready_timeout');
    }
    set waitForReadyTimeout(value) {
        this._waitForReadyTimeout = value;
    }
    resetWaitForReadyTimeout() {
        this._waitForReadyTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForReadyTimeoutInput() {
        return this._waitForReadyTimeout;
    }
    get setting() {
        return this._setting;
    }
    putSetting(value) {
        this._setting.internalValue = value;
    }
    resetSetting() {
        this._setting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get settingInput() {
        return this._setting.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application: cdktf.stringToTerraform(this._application),
            cname_prefix: cdktf.stringToTerraform(this._cnamePrefix),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            platform_arn: cdktf.stringToTerraform(this._platformArn),
            poll_interval: cdktf.stringToTerraform(this._pollInterval),
            solution_stack_name: cdktf.stringToTerraform(this._solutionStackName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            template_name: cdktf.stringToTerraform(this._templateName),
            tier: cdktf.stringToTerraform(this._tier),
            version_label: cdktf.stringToTerraform(this._versionLabel),
            wait_for_ready_timeout: cdktf.stringToTerraform(this._waitForReadyTimeout),
            setting: cdktf.listMapper(elasticBeanstalkEnvironmentSettingToTerraform, true)(this._setting.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application: {
                value: cdktf.stringToHclTerraform(this._application),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cname_prefix: {
                value: cdktf.stringToHclTerraform(this._cnamePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            platform_arn: {
                value: cdktf.stringToHclTerraform(this._platformArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            poll_interval: {
                value: cdktf.stringToHclTerraform(this._pollInterval),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            solution_stack_name: {
                value: cdktf.stringToHclTerraform(this._solutionStackName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            template_name: {
                value: cdktf.stringToHclTerraform(this._templateName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tier: {
                value: cdktf.stringToHclTerraform(this._tier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version_label: {
                value: cdktf.stringToHclTerraform(this._versionLabel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            wait_for_ready_timeout: {
                value: cdktf.stringToHclTerraform(this._waitForReadyTimeout),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            setting: {
                value: cdktf.listMapperHcl(elasticBeanstalkEnvironmentSettingToHclTerraform, true)(this._setting.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ElasticBeanstalkEnvironmentSettingList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ElasticBeanstalkEnvironment = ElasticBeanstalkEnvironment;
_e = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkEnvironment[_e] = { fqn: "@cdktf/provider-aws.elasticBeanstalkEnvironment.ElasticBeanstalkEnvironment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ElasticBeanstalkEnvironment.tfResourceType = "aws_elastic_beanstalk_environment";
