// @cdktf/provider-aws/lib/emr-block-public-access-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmrBlockPublicAccessConfiguration = exports.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList = exports.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference = exports.emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToHclTerraform = exports.emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_range: cdktf.numberToTerraform(struct.maxRange),
        min_range: cdktf.numberToTerraform(struct.minRange),
    };
}
exports.emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToTerraform = emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToTerraform;
function emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_range: {
            value: cdktf.numberToHclTerraform(struct.maxRange),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        min_range: {
            value: cdktf.numberToHclTerraform(struct.minRange),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToHclTerraform = emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToHclTerraform;
class EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRange = this._maxRange;
        }
        if (this._minRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.minRange = this._minRange;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._maxRange = undefined;
            this._minRange = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._maxRange = value.maxRange;
            this._minRange = value.minRange;
        }
    }
    get maxRange() {
        return this.getNumberAttribute('max_range');
    }
    set maxRange(value) {
        this._maxRange = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxRangeInput() {
        return this._maxRange;
    }
    get minRange() {
        return this.getNumberAttribute('min_range');
    }
    set minRange(value) {
        this._minRange = value;
    }
    // Temporarily expose input value. Use with caution.
    get minRangeInput() {
        return this._minRange;
    }
}
exports.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference = EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference[_a] = { fqn: "@cdktf/provider-aws.emrBlockPublicAccessConfiguration.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference", version: "19.15.0" };
class EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList = EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList;
_b = JSII_RTTI_SYMBOL_1;
EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList[_b] = { fqn: "@cdktf/provider-aws.emrBlockPublicAccessConfiguration.EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_block_public_access_configuration aws_emr_block_public_access_configuration}
*/
class EmrBlockPublicAccessConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EmrBlockPublicAccessConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EmrBlockPublicAccessConfiguration to import
    * @param importFromId The id of the existing EmrBlockPublicAccessConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_block_public_access_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EmrBlockPublicAccessConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_emr_block_public_access_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_block_public_access_configuration aws_emr_block_public_access_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EmrBlockPublicAccessConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emr_block_public_access_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // permitted_public_security_group_rule_range - computed: false, optional: true, required: false
        this._permittedPublicSecurityGroupRuleRange = new EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList(this, "permitted_public_security_group_rule_range", false);
        this._blockPublicSecurityGroupRules = config.blockPublicSecurityGroupRules;
        this._id = config.id;
        this._permittedPublicSecurityGroupRuleRange.internalValue = config.permittedPublicSecurityGroupRuleRange;
    }
    get blockPublicSecurityGroupRules() {
        return this.getBooleanAttribute('block_public_security_group_rules');
    }
    set blockPublicSecurityGroupRules(value) {
        this._blockPublicSecurityGroupRules = value;
    }
    // Temporarily expose input value. Use with caution.
    get blockPublicSecurityGroupRulesInput() {
        return this._blockPublicSecurityGroupRules;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get permittedPublicSecurityGroupRuleRange() {
        return this._permittedPublicSecurityGroupRuleRange;
    }
    putPermittedPublicSecurityGroupRuleRange(value) {
        this._permittedPublicSecurityGroupRuleRange.internalValue = value;
    }
    resetPermittedPublicSecurityGroupRuleRange() {
        this._permittedPublicSecurityGroupRuleRange.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permittedPublicSecurityGroupRuleRangeInput() {
        return this._permittedPublicSecurityGroupRuleRange.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            block_public_security_group_rules: cdktf.booleanToTerraform(this._blockPublicSecurityGroupRules),
            id: cdktf.stringToTerraform(this._id),
            permitted_public_security_group_rule_range: cdktf.listMapper(emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToTerraform, true)(this._permittedPublicSecurityGroupRuleRange.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            block_public_security_group_rules: {
                value: cdktf.booleanToHclTerraform(this._blockPublicSecurityGroupRules),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permitted_public_security_group_rule_range: {
                value: cdktf.listMapperHcl(emrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeToHclTerraform, true)(this._permittedPublicSecurityGroupRuleRange.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EmrBlockPublicAccessConfigurationPermittedPublicSecurityGroupRuleRangeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EmrBlockPublicAccessConfiguration = EmrBlockPublicAccessConfiguration;
_c = JSII_RTTI_SYMBOL_1;
EmrBlockPublicAccessConfiguration[_c] = { fqn: "@cdktf/provider-aws.emrBlockPublicAccessConfiguration.EmrBlockPublicAccessConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EmrBlockPublicAccessConfiguration.tfResourceType = "aws_emr_block_public_access_configuration";
