// @cdktf/provider-aws/lib/rds-cluster-activity-stream/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RdsClusterActivityStream = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rds_cluster_activity_stream aws_rds_cluster_activity_stream}
*/
class RdsClusterActivityStream extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RdsClusterActivityStream resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RdsClusterActivityStream to import
    * @param importFromId The id of the existing RdsClusterActivityStream that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rds_cluster_activity_stream#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RdsClusterActivityStream to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_rds_cluster_activity_stream", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rds_cluster_activity_stream aws_rds_cluster_activity_stream} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RdsClusterActivityStreamConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_cluster_activity_stream',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._engineNativeAuditFieldsIncluded = config.engineNativeAuditFieldsIncluded;
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._mode = config.mode;
        this._resourceArn = config.resourceArn;
    }
    get engineNativeAuditFieldsIncluded() {
        return this.getBooleanAttribute('engine_native_audit_fields_included');
    }
    set engineNativeAuditFieldsIncluded(value) {
        this._engineNativeAuditFieldsIncluded = value;
    }
    resetEngineNativeAuditFieldsIncluded() {
        this._engineNativeAuditFieldsIncluded = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engineNativeAuditFieldsIncludedInput() {
        return this._engineNativeAuditFieldsIncluded;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kinesis_stream_name - computed: true, optional: false, required: false
    get kinesisStreamName() {
        return this.getStringAttribute('kinesis_stream_name');
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            engine_native_audit_fields_included: cdktf.booleanToTerraform(this._engineNativeAuditFieldsIncluded),
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            mode: cdktf.stringToTerraform(this._mode),
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            engine_native_audit_fields_included: {
                value: cdktf.booleanToHclTerraform(this._engineNativeAuditFieldsIncluded),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_id: {
                value: cdktf.stringToHclTerraform(this._kmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            mode: {
                value: cdktf.stringToHclTerraform(this._mode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_arn: {
                value: cdktf.stringToHclTerraform(this._resourceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RdsClusterActivityStream = RdsClusterActivityStream;
_a = JSII_RTTI_SYMBOL_1;
RdsClusterActivityStream[_a] = { fqn: "@cdktf/provider-aws.rdsClusterActivityStream.RdsClusterActivityStream", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RdsClusterActivityStream.tfResourceType = "aws_rds_cluster_activity_stream";
