// @cdktf/provider-aws/lib/data-aws-connect-quick-connect/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectQuickConnect = exports.DataAwsConnectQuickConnectQuickConnectConfigList = exports.DataAwsConnectQuickConnectQuickConnectConfigOutputReference = exports.dataAwsConnectQuickConnectQuickConnectConfigToHclTerraform = exports.dataAwsConnectQuickConnectQuickConnectConfigToTerraform = exports.DataAwsConnectQuickConnectQuickConnectConfigUserConfigList = exports.DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference = exports.dataAwsConnectQuickConnectQuickConnectConfigUserConfigToHclTerraform = exports.dataAwsConnectQuickConnectQuickConnectConfigUserConfigToTerraform = exports.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList = exports.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference = exports.dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToHclTerraform = exports.dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToTerraform = exports.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList = exports.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference = exports.dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToHclTerraform = exports.dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToTerraform = dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToTerraform;
function dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToHclTerraform = dataAwsConnectQuickConnectQuickConnectConfigPhoneConfigToHclTerraform;
class DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // phone_number - computed: true, optional: false, required: false
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference = DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference", version: "19.15.0" };
class DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList = DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList", version: "19.15.0" };
function dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToTerraform = dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToTerraform;
function dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToHclTerraform = dataAwsConnectQuickConnectQuickConnectConfigQueueConfigToHclTerraform;
class DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // contact_flow_id - computed: true, optional: false, required: false
    get contactFlowId() {
        return this.getStringAttribute('contact_flow_id');
    }
    // queue_id - computed: true, optional: false, required: false
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference = DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference", version: "19.15.0" };
class DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectQuickConnectQuickConnectConfigQueueConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList = DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList[_d] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList", version: "19.15.0" };
function dataAwsConnectQuickConnectQuickConnectConfigUserConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectQuickConnectQuickConnectConfigUserConfigToTerraform = dataAwsConnectQuickConnectQuickConnectConfigUserConfigToTerraform;
function dataAwsConnectQuickConnectQuickConnectConfigUserConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectQuickConnectQuickConnectConfigUserConfigToHclTerraform = dataAwsConnectQuickConnectQuickConnectConfigUserConfigToHclTerraform;
class DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // contact_flow_id - computed: true, optional: false, required: false
    get contactFlowId() {
        return this.getStringAttribute('contact_flow_id');
    }
    // user_id - computed: true, optional: false, required: false
    get userId() {
        return this.getStringAttribute('user_id');
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference = DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference", version: "19.15.0" };
class DataAwsConnectQuickConnectQuickConnectConfigUserConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectQuickConnectQuickConnectConfigUserConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigUserConfigList = DataAwsConnectQuickConnectQuickConnectConfigUserConfigList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigUserConfigList[_f] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigUserConfigList", version: "19.15.0" };
function dataAwsConnectQuickConnectQuickConnectConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectQuickConnectQuickConnectConfigToTerraform = dataAwsConnectQuickConnectQuickConnectConfigToTerraform;
function dataAwsConnectQuickConnectQuickConnectConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsConnectQuickConnectQuickConnectConfigToHclTerraform = dataAwsConnectQuickConnectQuickConnectConfigToHclTerraform;
class DataAwsConnectQuickConnectQuickConnectConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // phone_config - computed: true, optional: false, required: false
        this._phoneConfig = new DataAwsConnectQuickConnectQuickConnectConfigPhoneConfigList(this, "phone_config", false);
        // queue_config - computed: true, optional: false, required: false
        this._queueConfig = new DataAwsConnectQuickConnectQuickConnectConfigQueueConfigList(this, "queue_config", false);
        // user_config - computed: true, optional: false, required: false
        this._userConfig = new DataAwsConnectQuickConnectQuickConnectConfigUserConfigList(this, "user_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get phoneConfig() {
        return this._phoneConfig;
    }
    get queueConfig() {
        return this._queueConfig;
    }
    // quick_connect_type - computed: true, optional: false, required: false
    get quickConnectType() {
        return this.getStringAttribute('quick_connect_type');
    }
    get userConfig() {
        return this._userConfig;
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigOutputReference = DataAwsConnectQuickConnectQuickConnectConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigOutputReference", version: "19.15.0" };
class DataAwsConnectQuickConnectQuickConnectConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectQuickConnectQuickConnectConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectQuickConnectQuickConnectConfigList = DataAwsConnectQuickConnectQuickConnectConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnectQuickConnectConfigList[_h] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnectQuickConnectConfigList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_quick_connect aws_connect_quick_connect}
*/
class DataAwsConnectQuickConnect extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsConnectQuickConnect resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsConnectQuickConnect to import
    * @param importFromId The id of the existing DataAwsConnectQuickConnect that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_quick_connect#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsConnectQuickConnect to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_connect_quick_connect", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_quick_connect aws_connect_quick_connect} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectQuickConnectConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_quick_connect',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // quick_connect_config - computed: true, optional: false, required: false
        this._quickConnectConfig = new DataAwsConnectQuickConnectQuickConnectConfigList(this, "quick_connect_config", false);
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._quickConnectId = config.quickConnectId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get quickConnectConfig() {
        return this._quickConnectConfig;
    }
    get quickConnectId() {
        return this.getStringAttribute('quick_connect_id');
    }
    set quickConnectId(value) {
        this._quickConnectId = value;
    }
    resetQuickConnectId() {
        this._quickConnectId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get quickConnectIdInput() {
        return this._quickConnectId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            quick_connect_id: cdktf.stringToTerraform(this._quickConnectId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quick_connect_id: {
                value: cdktf.stringToHclTerraform(this._quickConnectId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsConnectQuickConnect = DataAwsConnectQuickConnect;
_j = JSII_RTTI_SYMBOL_1;
DataAwsConnectQuickConnect[_j] = { fqn: "@cdktf/provider-aws.dataAwsConnectQuickConnect.DataAwsConnectQuickConnect", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectQuickConnect.tfResourceType = "aws_connect_quick_connect";
