// @cdktf/provider-aws/lib/servicequotas-service-quota/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicequotasServiceQuota = exports.ServicequotasServiceQuotaUsageMetricList = exports.ServicequotasServiceQuotaUsageMetricOutputReference = exports.servicequotasServiceQuotaUsageMetricToHclTerraform = exports.servicequotasServiceQuotaUsageMetricToTerraform = exports.ServicequotasServiceQuotaUsageMetricMetricDimensionsList = exports.ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference = exports.servicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform = exports.servicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function servicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.servicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform = servicequotasServiceQuotaUsageMetricMetricDimensionsToTerraform;
function servicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.servicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform = servicequotasServiceQuotaUsageMetricMetricDimensionsToHclTerraform;
class ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // class - computed: true, optional: false, required: false
    get class() {
        return this.getStringAttribute('class');
    }
    // resource - computed: true, optional: false, required: false
    get resource() {
        return this.getStringAttribute('resource');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference = ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.servicequotasServiceQuota.ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference", version: "19.15.0" };
class ServicequotasServiceQuotaUsageMetricMetricDimensionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ServicequotasServiceQuotaUsageMetricMetricDimensionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ServicequotasServiceQuotaUsageMetricMetricDimensionsList = ServicequotasServiceQuotaUsageMetricMetricDimensionsList;
_b = JSII_RTTI_SYMBOL_1;
ServicequotasServiceQuotaUsageMetricMetricDimensionsList[_b] = { fqn: "@cdktf/provider-aws.servicequotasServiceQuota.ServicequotasServiceQuotaUsageMetricMetricDimensionsList", version: "19.15.0" };
function servicequotasServiceQuotaUsageMetricToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.servicequotasServiceQuotaUsageMetricToTerraform = servicequotasServiceQuotaUsageMetricToTerraform;
function servicequotasServiceQuotaUsageMetricToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.servicequotasServiceQuotaUsageMetricToHclTerraform = servicequotasServiceQuotaUsageMetricToHclTerraform;
class ServicequotasServiceQuotaUsageMetricOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metric_dimensions - computed: true, optional: false, required: false
        this._metricDimensions = new ServicequotasServiceQuotaUsageMetricMetricDimensionsList(this, "metric_dimensions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get metricDimensions() {
        return this._metricDimensions;
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // metric_namespace - computed: true, optional: false, required: false
    get metricNamespace() {
        return this.getStringAttribute('metric_namespace');
    }
    // metric_statistic_recommendation - computed: true, optional: false, required: false
    get metricStatisticRecommendation() {
        return this.getStringAttribute('metric_statistic_recommendation');
    }
}
exports.ServicequotasServiceQuotaUsageMetricOutputReference = ServicequotasServiceQuotaUsageMetricOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ServicequotasServiceQuotaUsageMetricOutputReference[_c] = { fqn: "@cdktf/provider-aws.servicequotasServiceQuota.ServicequotasServiceQuotaUsageMetricOutputReference", version: "19.15.0" };
class ServicequotasServiceQuotaUsageMetricList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ServicequotasServiceQuotaUsageMetricOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ServicequotasServiceQuotaUsageMetricList = ServicequotasServiceQuotaUsageMetricList;
_d = JSII_RTTI_SYMBOL_1;
ServicequotasServiceQuotaUsageMetricList[_d] = { fqn: "@cdktf/provider-aws.servicequotasServiceQuota.ServicequotasServiceQuotaUsageMetricList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/servicequotas_service_quota aws_servicequotas_service_quota}
*/
class ServicequotasServiceQuota extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ServicequotasServiceQuota resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ServicequotasServiceQuota to import
    * @param importFromId The id of the existing ServicequotasServiceQuota that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/servicequotas_service_quota#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ServicequotasServiceQuota to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_servicequotas_service_quota", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/servicequotas_service_quota aws_servicequotas_service_quota} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ServicequotasServiceQuotaConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicequotas_service_quota',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // usage_metric - computed: true, optional: false, required: false
        this._usageMetric = new ServicequotasServiceQuotaUsageMetricList(this, "usage_metric", false);
        this._id = config.id;
        this._quotaCode = config.quotaCode;
        this._serviceCode = config.serviceCode;
        this._value = config.value;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // adjustable - computed: true, optional: false, required: false
    get adjustable() {
        return this.getBooleanAttribute('adjustable');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_value - computed: true, optional: false, required: false
    get defaultValue() {
        return this.getNumberAttribute('default_value');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get quotaCode() {
        return this.getStringAttribute('quota_code');
    }
    set quotaCode(value) {
        this._quotaCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get quotaCodeInput() {
        return this._quotaCode;
    }
    // quota_name - computed: true, optional: false, required: false
    get quotaName() {
        return this.getStringAttribute('quota_name');
    }
    // request_id - computed: true, optional: false, required: false
    get requestId() {
        return this.getStringAttribute('request_id');
    }
    // request_status - computed: true, optional: false, required: false
    get requestStatus() {
        return this.getStringAttribute('request_status');
    }
    get serviceCode() {
        return this.getStringAttribute('service_code');
    }
    set serviceCode(value) {
        this._serviceCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceCodeInput() {
        return this._serviceCode;
    }
    // service_name - computed: true, optional: false, required: false
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    get usageMetric() {
        return this._usageMetric;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            quota_code: cdktf.stringToTerraform(this._quotaCode),
            service_code: cdktf.stringToTerraform(this._serviceCode),
            value: cdktf.numberToTerraform(this._value),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_code: {
                value: cdktf.stringToHclTerraform(this._quotaCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_code: {
                value: cdktf.stringToHclTerraform(this._serviceCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            value: {
                value: cdktf.numberToHclTerraform(this._value),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ServicequotasServiceQuota = ServicequotasServiceQuota;
_e = JSII_RTTI_SYMBOL_1;
ServicequotasServiceQuota[_e] = { fqn: "@cdktf/provider-aws.servicequotasServiceQuota.ServicequotasServiceQuota", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ServicequotasServiceQuota.tfResourceType = "aws_servicequotas_service_quota";
