// @cdktf/provider-aws/lib/load-balancer-listener-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancerListenerPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_listener_policy aws_load_balancer_listener_policy}
*/
class LoadBalancerListenerPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LoadBalancerListenerPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LoadBalancerListenerPolicy to import
    * @param importFromId The id of the existing LoadBalancerListenerPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_listener_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LoadBalancerListenerPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_load_balancer_listener_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/load_balancer_listener_policy aws_load_balancer_listener_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LoadBalancerListenerPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_load_balancer_listener_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._loadBalancerName = config.loadBalancerName;
        this._loadBalancerPort = config.loadBalancerPort;
        this._policyNames = config.policyNames;
        this._triggers = config.triggers;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get loadBalancerName() {
        return this.getStringAttribute('load_balancer_name');
    }
    set loadBalancerName(value) {
        this._loadBalancerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerNameInput() {
        return this._loadBalancerName;
    }
    get loadBalancerPort() {
        return this.getNumberAttribute('load_balancer_port');
    }
    set loadBalancerPort(value) {
        this._loadBalancerPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerPortInput() {
        return this._loadBalancerPort;
    }
    get policyNames() {
        return cdktf.Fn.tolist(this.getListAttribute('policy_names'));
    }
    set policyNames(value) {
        this._policyNames = value;
    }
    resetPolicyNames() {
        this._policyNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyNamesInput() {
        return this._policyNames;
    }
    get triggers() {
        return this.getStringMapAttribute('triggers');
    }
    set triggers(value) {
        this._triggers = value;
    }
    resetTriggers() {
        this._triggers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggersInput() {
        return this._triggers;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            load_balancer_name: cdktf.stringToTerraform(this._loadBalancerName),
            load_balancer_port: cdktf.numberToTerraform(this._loadBalancerPort),
            policy_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._policyNames),
            triggers: cdktf.hashMapper(cdktf.stringToTerraform)(this._triggers),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancer_name: {
                value: cdktf.stringToHclTerraform(this._loadBalancerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancer_port: {
                value: cdktf.numberToHclTerraform(this._loadBalancerPort),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            policy_names: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._policyNames),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            triggers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._triggers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LoadBalancerListenerPolicy = LoadBalancerListenerPolicy;
_a = JSII_RTTI_SYMBOL_1;
LoadBalancerListenerPolicy[_a] = { fqn: "@cdktf/provider-aws.loadBalancerListenerPolicy.LoadBalancerListenerPolicy", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LoadBalancerListenerPolicy.tfResourceType = "aws_load_balancer_listener_policy";
