// @cdktf/provider-aws/lib/wafregional-rate-based-rule/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafregionalRateBasedRule = exports.WafregionalRateBasedRulePredicateList = exports.WafregionalRateBasedRulePredicateOutputReference = exports.wafregionalRateBasedRulePredicateToHclTerraform = exports.wafregionalRateBasedRulePredicateToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafregionalRateBasedRulePredicateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_id: cdktf.stringToTerraform(struct.dataId),
        negated: cdktf.booleanToTerraform(struct.negated),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafregionalRateBasedRulePredicateToTerraform = wafregionalRateBasedRulePredicateToTerraform;
function wafregionalRateBasedRulePredicateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_id: {
            value: cdktf.stringToHclTerraform(struct.dataId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        negated: {
            value: cdktf.booleanToHclTerraform(struct.negated),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.wafregionalRateBasedRulePredicateToHclTerraform = wafregionalRateBasedRulePredicateToHclTerraform;
class WafregionalRateBasedRulePredicateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataId !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataId = this._dataId;
        }
        if (this._negated !== undefined) {
            hasAnyValues = true;
            internalValueResult.negated = this._negated;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dataId = undefined;
            this._negated = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dataId = value.dataId;
            this._negated = value.negated;
            this._type = value.type;
        }
    }
    get dataId() {
        return this.getStringAttribute('data_id');
    }
    set dataId(value) {
        this._dataId = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataIdInput() {
        return this._dataId;
    }
    get negated() {
        return this.getBooleanAttribute('negated');
    }
    set negated(value) {
        this._negated = value;
    }
    // Temporarily expose input value. Use with caution.
    get negatedInput() {
        return this._negated;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.WafregionalRateBasedRulePredicateOutputReference = WafregionalRateBasedRulePredicateOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafregionalRateBasedRulePredicateOutputReference[_a] = { fqn: "@cdktf/provider-aws.wafregionalRateBasedRule.WafregionalRateBasedRulePredicateOutputReference", version: "19.15.0" };
class WafregionalRateBasedRulePredicateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WafregionalRateBasedRulePredicateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WafregionalRateBasedRulePredicateList = WafregionalRateBasedRulePredicateList;
_b = JSII_RTTI_SYMBOL_1;
WafregionalRateBasedRulePredicateList[_b] = { fqn: "@cdktf/provider-aws.wafregionalRateBasedRule.WafregionalRateBasedRulePredicateList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/wafregional_rate_based_rule aws_wafregional_rate_based_rule}
*/
class WafregionalRateBasedRule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a WafregionalRateBasedRule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the WafregionalRateBasedRule to import
    * @param importFromId The id of the existing WafregionalRateBasedRule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/wafregional_rate_based_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the WafregionalRateBasedRule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_wafregional_rate_based_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/wafregional_rate_based_rule aws_wafregional_rate_based_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options WafregionalRateBasedRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafregional_rate_based_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // predicate - computed: false, optional: true, required: false
        this._predicate = new WafregionalRateBasedRulePredicateList(this, "predicate", true);
        this._id = config.id;
        this._metricName = config.metricName;
        this._name = config.name;
        this._rateKey = config.rateKey;
        this._rateLimit = config.rateLimit;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._predicate.internalValue = config.predicate;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    set metricName(value) {
        this._metricName = value;
    }
    // Temporarily expose input value. Use with caution.
    get metricNameInput() {
        return this._metricName;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get rateKey() {
        return this.getStringAttribute('rate_key');
    }
    set rateKey(value) {
        this._rateKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get rateKeyInput() {
        return this._rateKey;
    }
    get rateLimit() {
        return this.getNumberAttribute('rate_limit');
    }
    set rateLimit(value) {
        this._rateLimit = value;
    }
    // Temporarily expose input value. Use with caution.
    get rateLimitInput() {
        return this._rateLimit;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get predicate() {
        return this._predicate;
    }
    putPredicate(value) {
        this._predicate.internalValue = value;
    }
    resetPredicate() {
        this._predicate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predicateInput() {
        return this._predicate.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            metric_name: cdktf.stringToTerraform(this._metricName),
            name: cdktf.stringToTerraform(this._name),
            rate_key: cdktf.stringToTerraform(this._rateKey),
            rate_limit: cdktf.numberToTerraform(this._rateLimit),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            predicate: cdktf.listMapper(wafregionalRateBasedRulePredicateToTerraform, true)(this._predicate.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            metric_name: {
                value: cdktf.stringToHclTerraform(this._metricName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rate_key: {
                value: cdktf.stringToHclTerraform(this._rateKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rate_limit: {
                value: cdktf.numberToHclTerraform(this._rateLimit),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            predicate: {
                value: cdktf.listMapperHcl(wafregionalRateBasedRulePredicateToHclTerraform, true)(this._predicate.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "WafregionalRateBasedRulePredicateList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.WafregionalRateBasedRule = WafregionalRateBasedRule;
_c = JSII_RTTI_SYMBOL_1;
WafregionalRateBasedRule[_c] = { fqn: "@cdktf/provider-aws.wafregionalRateBasedRule.WafregionalRateBasedRule", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
WafregionalRateBasedRule.tfResourceType = "aws_wafregional_rate_based_rule";
