// @cdktf/provider-aws/lib/location-place-index/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocationPlaceIndex = exports.LocationPlaceIndexDataSourceConfigurationOutputReference = exports.locationPlaceIndexDataSourceConfigurationToHclTerraform = exports.locationPlaceIndexDataSourceConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function locationPlaceIndexDataSourceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        intended_use: cdktf.stringToTerraform(struct.intendedUse),
    };
}
exports.locationPlaceIndexDataSourceConfigurationToTerraform = locationPlaceIndexDataSourceConfigurationToTerraform;
function locationPlaceIndexDataSourceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        intended_use: {
            value: cdktf.stringToHclTerraform(struct.intendedUse),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.locationPlaceIndexDataSourceConfigurationToHclTerraform = locationPlaceIndexDataSourceConfigurationToHclTerraform;
class LocationPlaceIndexDataSourceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._intendedUse !== undefined) {
            hasAnyValues = true;
            internalValueResult.intendedUse = this._intendedUse;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._intendedUse = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._intendedUse = value.intendedUse;
        }
    }
    get intendedUse() {
        return this.getStringAttribute('intended_use');
    }
    set intendedUse(value) {
        this._intendedUse = value;
    }
    resetIntendedUse() {
        this._intendedUse = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intendedUseInput() {
        return this._intendedUse;
    }
}
exports.LocationPlaceIndexDataSourceConfigurationOutputReference = LocationPlaceIndexDataSourceConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LocationPlaceIndexDataSourceConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.locationPlaceIndex.LocationPlaceIndexDataSourceConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/location_place_index aws_location_place_index}
*/
class LocationPlaceIndex extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LocationPlaceIndex resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LocationPlaceIndex to import
    * @param importFromId The id of the existing LocationPlaceIndex that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/location_place_index#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LocationPlaceIndex to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_location_place_index", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/location_place_index aws_location_place_index} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LocationPlaceIndexConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_location_place_index',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // data_source_configuration - computed: false, optional: true, required: false
        this._dataSourceConfiguration = new LocationPlaceIndexDataSourceConfigurationOutputReference(this, "data_source_configuration");
        this._dataSource = config.dataSource;
        this._description = config.description;
        this._id = config.id;
        this._indexName = config.indexName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._dataSourceConfiguration.internalValue = config.dataSourceConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dataSource() {
        return this.getStringAttribute('data_source');
    }
    set dataSource(value) {
        this._dataSource = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataSourceInput() {
        return this._dataSource;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // index_arn - computed: true, optional: false, required: false
    get indexArn() {
        return this.getStringAttribute('index_arn');
    }
    get indexName() {
        return this.getStringAttribute('index_name');
    }
    set indexName(value) {
        this._indexName = value;
    }
    // Temporarily expose input value. Use with caution.
    get indexNameInput() {
        return this._indexName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get dataSourceConfiguration() {
        return this._dataSourceConfiguration;
    }
    putDataSourceConfiguration(value) {
        this._dataSourceConfiguration.internalValue = value;
    }
    resetDataSourceConfiguration() {
        this._dataSourceConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataSourceConfigurationInput() {
        return this._dataSourceConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            data_source: cdktf.stringToTerraform(this._dataSource),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            index_name: cdktf.stringToTerraform(this._indexName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            data_source_configuration: locationPlaceIndexDataSourceConfigurationToTerraform(this._dataSourceConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            data_source: {
                value: cdktf.stringToHclTerraform(this._dataSource),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            index_name: {
                value: cdktf.stringToHclTerraform(this._indexName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            data_source_configuration: {
                value: locationPlaceIndexDataSourceConfigurationToHclTerraform(this._dataSourceConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LocationPlaceIndexDataSourceConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LocationPlaceIndex = LocationPlaceIndex;
_b = JSII_RTTI_SYMBOL_1;
LocationPlaceIndex[_b] = { fqn: "@cdktf/provider-aws.locationPlaceIndex.LocationPlaceIndex", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LocationPlaceIndex.tfResourceType = "aws_location_place_index";
