// @cdktf/provider-aws/lib/lb-trust-store-revocation/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LbTrustStoreRevocation = exports.LbTrustStoreRevocationTimeoutsOutputReference = exports.lbTrustStoreRevocationTimeoutsToHclTerraform = exports.lbTrustStoreRevocationTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lbTrustStoreRevocationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.lbTrustStoreRevocationTimeoutsToTerraform = lbTrustStoreRevocationTimeoutsToTerraform;
function lbTrustStoreRevocationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lbTrustStoreRevocationTimeoutsToHclTerraform = lbTrustStoreRevocationTimeoutsToHclTerraform;
class LbTrustStoreRevocationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.LbTrustStoreRevocationTimeoutsOutputReference = LbTrustStoreRevocationTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LbTrustStoreRevocationTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.lbTrustStoreRevocation.LbTrustStoreRevocationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lb_trust_store_revocation aws_lb_trust_store_revocation}
*/
class LbTrustStoreRevocation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LbTrustStoreRevocation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LbTrustStoreRevocation to import
    * @param importFromId The id of the existing LbTrustStoreRevocation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lb_trust_store_revocation#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LbTrustStoreRevocation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lb_trust_store_revocation", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lb_trust_store_revocation aws_lb_trust_store_revocation} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LbTrustStoreRevocationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lb_trust_store_revocation',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LbTrustStoreRevocationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._revocationsS3Bucket = config.revocationsS3Bucket;
        this._revocationsS3Key = config.revocationsS3Key;
        this._revocationsS3ObjectVersion = config.revocationsS3ObjectVersion;
        this._trustStoreArn = config.trustStoreArn;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // revocation_id - computed: true, optional: false, required: false
    get revocationId() {
        return this.getNumberAttribute('revocation_id');
    }
    get revocationsS3Bucket() {
        return this.getStringAttribute('revocations_s3_bucket');
    }
    set revocationsS3Bucket(value) {
        this._revocationsS3Bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get revocationsS3BucketInput() {
        return this._revocationsS3Bucket;
    }
    get revocationsS3Key() {
        return this.getStringAttribute('revocations_s3_key');
    }
    set revocationsS3Key(value) {
        this._revocationsS3Key = value;
    }
    // Temporarily expose input value. Use with caution.
    get revocationsS3KeyInput() {
        return this._revocationsS3Key;
    }
    get revocationsS3ObjectVersion() {
        return this.getStringAttribute('revocations_s3_object_version');
    }
    set revocationsS3ObjectVersion(value) {
        this._revocationsS3ObjectVersion = value;
    }
    resetRevocationsS3ObjectVersion() {
        this._revocationsS3ObjectVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revocationsS3ObjectVersionInput() {
        return this._revocationsS3ObjectVersion;
    }
    get trustStoreArn() {
        return this.getStringAttribute('trust_store_arn');
    }
    set trustStoreArn(value) {
        this._trustStoreArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustStoreArnInput() {
        return this._trustStoreArn;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            revocations_s3_bucket: cdktf.stringToTerraform(this._revocationsS3Bucket),
            revocations_s3_key: cdktf.stringToTerraform(this._revocationsS3Key),
            revocations_s3_object_version: cdktf.stringToTerraform(this._revocationsS3ObjectVersion),
            trust_store_arn: cdktf.stringToTerraform(this._trustStoreArn),
            timeouts: lbTrustStoreRevocationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            revocations_s3_bucket: {
                value: cdktf.stringToHclTerraform(this._revocationsS3Bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            revocations_s3_key: {
                value: cdktf.stringToHclTerraform(this._revocationsS3Key),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            revocations_s3_object_version: {
                value: cdktf.stringToHclTerraform(this._revocationsS3ObjectVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_store_arn: {
                value: cdktf.stringToHclTerraform(this._trustStoreArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: lbTrustStoreRevocationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "LbTrustStoreRevocationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LbTrustStoreRevocation = LbTrustStoreRevocation;
_b = JSII_RTTI_SYMBOL_1;
LbTrustStoreRevocation[_b] = { fqn: "@cdktf/provider-aws.lbTrustStoreRevocation.LbTrustStoreRevocation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LbTrustStoreRevocation.tfResourceType = "aws_lb_trust_store_revocation";
