// @cdktf/provider-aws/lib/data-aws-iam-principal-policy-simulation/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIamPrincipalPolicySimulation = exports.DataAwsIamPrincipalPolicySimulationContextList = exports.DataAwsIamPrincipalPolicySimulationContextOutputReference = exports.dataAwsIamPrincipalPolicySimulationContextToHclTerraform = exports.dataAwsIamPrincipalPolicySimulationContextToTerraform = exports.DataAwsIamPrincipalPolicySimulationResultsList = exports.DataAwsIamPrincipalPolicySimulationResultsOutputReference = exports.dataAwsIamPrincipalPolicySimulationResultsToHclTerraform = exports.dataAwsIamPrincipalPolicySimulationResultsToTerraform = exports.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList = exports.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference = exports.dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToHclTerraform = exports.dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToTerraform = dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToTerraform;
function dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToHclTerraform = dataAwsIamPrincipalPolicySimulationResultsMatchedStatementsToHclTerraform;
class DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // source_policy_id - computed: true, optional: false, required: false
    get sourcePolicyId() {
        return this.getStringAttribute('source_policy_id');
    }
    // source_policy_type - computed: true, optional: false, required: false
    get sourcePolicyType() {
        return this.getStringAttribute('source_policy_type');
    }
}
exports.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference = DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference", version: "19.15.0" };
class DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList = DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList", version: "19.15.0" };
function dataAwsIamPrincipalPolicySimulationResultsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsIamPrincipalPolicySimulationResultsToTerraform = dataAwsIamPrincipalPolicySimulationResultsToTerraform;
function dataAwsIamPrincipalPolicySimulationResultsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsIamPrincipalPolicySimulationResultsToHclTerraform = dataAwsIamPrincipalPolicySimulationResultsToHclTerraform;
class DataAwsIamPrincipalPolicySimulationResultsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // decision_details - computed: true, optional: false, required: false
        this._decisionDetails = new cdktf.StringMap(this, "decision_details");
        // matched_statements - computed: true, optional: false, required: false
        this._matchedStatements = new DataAwsIamPrincipalPolicySimulationResultsMatchedStatementsList(this, "matched_statements", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // action_name - computed: true, optional: false, required: false
    get actionName() {
        return this.getStringAttribute('action_name');
    }
    // allowed - computed: true, optional: false, required: false
    get allowed() {
        return this.getBooleanAttribute('allowed');
    }
    // decision - computed: true, optional: false, required: false
    get decision() {
        return this.getStringAttribute('decision');
    }
    get decisionDetails() {
        return this._decisionDetails;
    }
    get matchedStatements() {
        return this._matchedStatements;
    }
    // missing_context_keys - computed: true, optional: false, required: false
    get missingContextKeys() {
        return cdktf.Fn.tolist(this.getListAttribute('missing_context_keys'));
    }
    // resource_arn - computed: true, optional: false, required: false
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
}
exports.DataAwsIamPrincipalPolicySimulationResultsOutputReference = DataAwsIamPrincipalPolicySimulationResultsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationResultsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationResultsOutputReference", version: "19.15.0" };
class DataAwsIamPrincipalPolicySimulationResultsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsIamPrincipalPolicySimulationResultsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsIamPrincipalPolicySimulationResultsList = DataAwsIamPrincipalPolicySimulationResultsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationResultsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationResultsList", version: "19.15.0" };
function dataAwsIamPrincipalPolicySimulationContextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        type: cdktf.stringToTerraform(struct.type),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsIamPrincipalPolicySimulationContextToTerraform = dataAwsIamPrincipalPolicySimulationContextToTerraform;
function dataAwsIamPrincipalPolicySimulationContextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsIamPrincipalPolicySimulationContextToHclTerraform = dataAwsIamPrincipalPolicySimulationContextToHclTerraform;
class DataAwsIamPrincipalPolicySimulationContextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._type = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._type = value.type;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsIamPrincipalPolicySimulationContextOutputReference = DataAwsIamPrincipalPolicySimulationContextOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationContextOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationContextOutputReference", version: "19.15.0" };
class DataAwsIamPrincipalPolicySimulationContextList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsIamPrincipalPolicySimulationContextOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsIamPrincipalPolicySimulationContextList = DataAwsIamPrincipalPolicySimulationContextList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulationContextList[_f] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulationContextList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/iam_principal_policy_simulation aws_iam_principal_policy_simulation}
*/
class DataAwsIamPrincipalPolicySimulation extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsIamPrincipalPolicySimulation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsIamPrincipalPolicySimulation to import
    * @param importFromId The id of the existing DataAwsIamPrincipalPolicySimulation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/iam_principal_policy_simulation#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsIamPrincipalPolicySimulation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iam_principal_policy_simulation", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/iam_principal_policy_simulation aws_iam_principal_policy_simulation} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsIamPrincipalPolicySimulationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_principal_policy_simulation',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // results - computed: true, optional: false, required: false
        this._results = new DataAwsIamPrincipalPolicySimulationResultsList(this, "results", true);
        // context - computed: false, optional: true, required: false
        this._context = new DataAwsIamPrincipalPolicySimulationContextList(this, "context", true);
        this._actionNames = config.actionNames;
        this._additionalPoliciesJson = config.additionalPoliciesJson;
        this._callerArn = config.callerArn;
        this._permissionsBoundaryPoliciesJson = config.permissionsBoundaryPoliciesJson;
        this._policySourceArn = config.policySourceArn;
        this._resourceArns = config.resourceArns;
        this._resourceHandlingOption = config.resourceHandlingOption;
        this._resourceOwnerAccountId = config.resourceOwnerAccountId;
        this._resourcePolicyJson = config.resourcePolicyJson;
        this._context.internalValue = config.context;
    }
    get actionNames() {
        return cdktf.Fn.tolist(this.getListAttribute('action_names'));
    }
    set actionNames(value) {
        this._actionNames = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionNamesInput() {
        return this._actionNames;
    }
    get additionalPoliciesJson() {
        return cdktf.Fn.tolist(this.getListAttribute('additional_policies_json'));
    }
    set additionalPoliciesJson(value) {
        this._additionalPoliciesJson = value;
    }
    resetAdditionalPoliciesJson() {
        this._additionalPoliciesJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalPoliciesJsonInput() {
        return this._additionalPoliciesJson;
    }
    // all_allowed - computed: true, optional: false, required: false
    get allAllowed() {
        return this.getBooleanAttribute('all_allowed');
    }
    get callerArn() {
        return this.getStringAttribute('caller_arn');
    }
    set callerArn(value) {
        this._callerArn = value;
    }
    resetCallerArn() {
        this._callerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get callerArnInput() {
        return this._callerArn;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get permissionsBoundaryPoliciesJson() {
        return cdktf.Fn.tolist(this.getListAttribute('permissions_boundary_policies_json'));
    }
    set permissionsBoundaryPoliciesJson(value) {
        this._permissionsBoundaryPoliciesJson = value;
    }
    resetPermissionsBoundaryPoliciesJson() {
        this._permissionsBoundaryPoliciesJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionsBoundaryPoliciesJsonInput() {
        return this._permissionsBoundaryPoliciesJson;
    }
    get policySourceArn() {
        return this.getStringAttribute('policy_source_arn');
    }
    set policySourceArn(value) {
        this._policySourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get policySourceArnInput() {
        return this._policySourceArn;
    }
    get resourceArns() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_arns'));
    }
    set resourceArns(value) {
        this._resourceArns = value;
    }
    resetResourceArns() {
        this._resourceArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnsInput() {
        return this._resourceArns;
    }
    get resourceHandlingOption() {
        return this.getStringAttribute('resource_handling_option');
    }
    set resourceHandlingOption(value) {
        this._resourceHandlingOption = value;
    }
    resetResourceHandlingOption() {
        this._resourceHandlingOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceHandlingOptionInput() {
        return this._resourceHandlingOption;
    }
    get resourceOwnerAccountId() {
        return this.getStringAttribute('resource_owner_account_id');
    }
    set resourceOwnerAccountId(value) {
        this._resourceOwnerAccountId = value;
    }
    resetResourceOwnerAccountId() {
        this._resourceOwnerAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceOwnerAccountIdInput() {
        return this._resourceOwnerAccountId;
    }
    get resourcePolicyJson() {
        return this.getStringAttribute('resource_policy_json');
    }
    set resourcePolicyJson(value) {
        this._resourcePolicyJson = value;
    }
    resetResourcePolicyJson() {
        this._resourcePolicyJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePolicyJsonInput() {
        return this._resourcePolicyJson;
    }
    get results() {
        return this._results;
    }
    get context() {
        return this._context;
    }
    putContext(value) {
        this._context.internalValue = value;
    }
    resetContext() {
        this._context.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contextInput() {
        return this._context.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._actionNames),
            additional_policies_json: cdktf.listMapper(cdktf.stringToTerraform, false)(this._additionalPoliciesJson),
            caller_arn: cdktf.stringToTerraform(this._callerArn),
            permissions_boundary_policies_json: cdktf.listMapper(cdktf.stringToTerraform, false)(this._permissionsBoundaryPoliciesJson),
            policy_source_arn: cdktf.stringToTerraform(this._policySourceArn),
            resource_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._resourceArns),
            resource_handling_option: cdktf.stringToTerraform(this._resourceHandlingOption),
            resource_owner_account_id: cdktf.stringToTerraform(this._resourceOwnerAccountId),
            resource_policy_json: cdktf.stringToTerraform(this._resourcePolicyJson),
            context: cdktf.listMapper(dataAwsIamPrincipalPolicySimulationContextToTerraform, true)(this._context.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action_names: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._actionNames),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            additional_policies_json: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._additionalPoliciesJson),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            caller_arn: {
                value: cdktf.stringToHclTerraform(this._callerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            permissions_boundary_policies_json: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._permissionsBoundaryPoliciesJson),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            policy_source_arn: {
                value: cdktf.stringToHclTerraform(this._policySourceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._resourceArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            resource_handling_option: {
                value: cdktf.stringToHclTerraform(this._resourceHandlingOption),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_owner_account_id: {
                value: cdktf.stringToHclTerraform(this._resourceOwnerAccountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_policy_json: {
                value: cdktf.stringToHclTerraform(this._resourcePolicyJson),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            context: {
                value: cdktf.listMapperHcl(dataAwsIamPrincipalPolicySimulationContextToHclTerraform, true)(this._context.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsIamPrincipalPolicySimulationContextList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsIamPrincipalPolicySimulation = DataAwsIamPrincipalPolicySimulation;
_g = JSII_RTTI_SYMBOL_1;
DataAwsIamPrincipalPolicySimulation[_g] = { fqn: "@cdktf/provider-aws.dataAwsIamPrincipalPolicySimulation.DataAwsIamPrincipalPolicySimulation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsIamPrincipalPolicySimulation.tfResourceType = "aws_iam_principal_policy_simulation";
