// @cdktf/provider-aws/lib/flow-log/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowLog = exports.FlowLogDestinationOptionsOutputReference = exports.flowLogDestinationOptionsToHclTerraform = exports.flowLogDestinationOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function flowLogDestinationOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file_format: cdktf.stringToTerraform(struct.fileFormat),
        hive_compatible_partitions: cdktf.booleanToTerraform(struct.hiveCompatiblePartitions),
        per_hour_partition: cdktf.booleanToTerraform(struct.perHourPartition),
    };
}
exports.flowLogDestinationOptionsToTerraform = flowLogDestinationOptionsToTerraform;
function flowLogDestinationOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        file_format: {
            value: cdktf.stringToHclTerraform(struct.fileFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        hive_compatible_partitions: {
            value: cdktf.booleanToHclTerraform(struct.hiveCompatiblePartitions),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        per_hour_partition: {
            value: cdktf.booleanToHclTerraform(struct.perHourPartition),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.flowLogDestinationOptionsToHclTerraform = flowLogDestinationOptionsToHclTerraform;
class FlowLogDestinationOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fileFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileFormat = this._fileFormat;
        }
        if (this._hiveCompatiblePartitions !== undefined) {
            hasAnyValues = true;
            internalValueResult.hiveCompatiblePartitions = this._hiveCompatiblePartitions;
        }
        if (this._perHourPartition !== undefined) {
            hasAnyValues = true;
            internalValueResult.perHourPartition = this._perHourPartition;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fileFormat = undefined;
            this._hiveCompatiblePartitions = undefined;
            this._perHourPartition = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fileFormat = value.fileFormat;
            this._hiveCompatiblePartitions = value.hiveCompatiblePartitions;
            this._perHourPartition = value.perHourPartition;
        }
    }
    get fileFormat() {
        return this.getStringAttribute('file_format');
    }
    set fileFormat(value) {
        this._fileFormat = value;
    }
    resetFileFormat() {
        this._fileFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileFormatInput() {
        return this._fileFormat;
    }
    get hiveCompatiblePartitions() {
        return this.getBooleanAttribute('hive_compatible_partitions');
    }
    set hiveCompatiblePartitions(value) {
        this._hiveCompatiblePartitions = value;
    }
    resetHiveCompatiblePartitions() {
        this._hiveCompatiblePartitions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hiveCompatiblePartitionsInput() {
        return this._hiveCompatiblePartitions;
    }
    get perHourPartition() {
        return this.getBooleanAttribute('per_hour_partition');
    }
    set perHourPartition(value) {
        this._perHourPartition = value;
    }
    resetPerHourPartition() {
        this._perHourPartition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get perHourPartitionInput() {
        return this._perHourPartition;
    }
}
exports.FlowLogDestinationOptionsOutputReference = FlowLogDestinationOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FlowLogDestinationOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.flowLog.FlowLogDestinationOptionsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/flow_log aws_flow_log}
*/
class FlowLog extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FlowLog resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FlowLog to import
    * @param importFromId The id of the existing FlowLog that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/flow_log#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FlowLog to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_flow_log", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/flow_log aws_flow_log} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FlowLogConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_flow_log',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // destination_options - computed: false, optional: true, required: false
        this._destinationOptions = new FlowLogDestinationOptionsOutputReference(this, "destination_options");
        this._deliverCrossAccountRole = config.deliverCrossAccountRole;
        this._eniId = config.eniId;
        this._iamRoleArn = config.iamRoleArn;
        this._id = config.id;
        this._logDestination = config.logDestination;
        this._logDestinationType = config.logDestinationType;
        this._logFormat = config.logFormat;
        this._logGroupName = config.logGroupName;
        this._maxAggregationInterval = config.maxAggregationInterval;
        this._subnetId = config.subnetId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._trafficType = config.trafficType;
        this._transitGatewayAttachmentId = config.transitGatewayAttachmentId;
        this._transitGatewayId = config.transitGatewayId;
        this._vpcId = config.vpcId;
        this._destinationOptions.internalValue = config.destinationOptions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get deliverCrossAccountRole() {
        return this.getStringAttribute('deliver_cross_account_role');
    }
    set deliverCrossAccountRole(value) {
        this._deliverCrossAccountRole = value;
    }
    resetDeliverCrossAccountRole() {
        this._deliverCrossAccountRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deliverCrossAccountRoleInput() {
        return this._deliverCrossAccountRole;
    }
    get eniId() {
        return this.getStringAttribute('eni_id');
    }
    set eniId(value) {
        this._eniId = value;
    }
    resetEniId() {
        this._eniId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eniIdInput() {
        return this._eniId;
    }
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    set iamRoleArn(value) {
        this._iamRoleArn = value;
    }
    resetIamRoleArn() {
        this._iamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleArnInput() {
        return this._iamRoleArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get logDestination() {
        return this.getStringAttribute('log_destination');
    }
    set logDestination(value) {
        this._logDestination = value;
    }
    resetLogDestination() {
        this._logDestination = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logDestinationInput() {
        return this._logDestination;
    }
    get logDestinationType() {
        return this.getStringAttribute('log_destination_type');
    }
    set logDestinationType(value) {
        this._logDestinationType = value;
    }
    resetLogDestinationType() {
        this._logDestinationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logDestinationTypeInput() {
        return this._logDestinationType;
    }
    get logFormat() {
        return this.getStringAttribute('log_format');
    }
    set logFormat(value) {
        this._logFormat = value;
    }
    resetLogFormat() {
        this._logFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logFormatInput() {
        return this._logFormat;
    }
    get logGroupName() {
        return this.getStringAttribute('log_group_name');
    }
    set logGroupName(value) {
        this._logGroupName = value;
    }
    resetLogGroupName() {
        this._logGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupNameInput() {
        return this._logGroupName;
    }
    get maxAggregationInterval() {
        return this.getNumberAttribute('max_aggregation_interval');
    }
    set maxAggregationInterval(value) {
        this._maxAggregationInterval = value;
    }
    resetMaxAggregationInterval() {
        this._maxAggregationInterval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxAggregationIntervalInput() {
        return this._maxAggregationInterval;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get trafficType() {
        return this.getStringAttribute('traffic_type');
    }
    set trafficType(value) {
        this._trafficType = value;
    }
    resetTrafficType() {
        this._trafficType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trafficTypeInput() {
        return this._trafficType;
    }
    get transitGatewayAttachmentId() {
        return this.getStringAttribute('transit_gateway_attachment_id');
    }
    set transitGatewayAttachmentId(value) {
        this._transitGatewayAttachmentId = value;
    }
    resetTransitGatewayAttachmentId() {
        this._transitGatewayAttachmentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitGatewayAttachmentIdInput() {
        return this._transitGatewayAttachmentId;
    }
    get transitGatewayId() {
        return this.getStringAttribute('transit_gateway_id');
    }
    set transitGatewayId(value) {
        this._transitGatewayId = value;
    }
    resetTransitGatewayId() {
        this._transitGatewayId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitGatewayIdInput() {
        return this._transitGatewayId;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get destinationOptions() {
        return this._destinationOptions;
    }
    putDestinationOptions(value) {
        this._destinationOptions.internalValue = value;
    }
    resetDestinationOptions() {
        this._destinationOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationOptionsInput() {
        return this._destinationOptions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            deliver_cross_account_role: cdktf.stringToTerraform(this._deliverCrossAccountRole),
            eni_id: cdktf.stringToTerraform(this._eniId),
            iam_role_arn: cdktf.stringToTerraform(this._iamRoleArn),
            id: cdktf.stringToTerraform(this._id),
            log_destination: cdktf.stringToTerraform(this._logDestination),
            log_destination_type: cdktf.stringToTerraform(this._logDestinationType),
            log_format: cdktf.stringToTerraform(this._logFormat),
            log_group_name: cdktf.stringToTerraform(this._logGroupName),
            max_aggregation_interval: cdktf.numberToTerraform(this._maxAggregationInterval),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            traffic_type: cdktf.stringToTerraform(this._trafficType),
            transit_gateway_attachment_id: cdktf.stringToTerraform(this._transitGatewayAttachmentId),
            transit_gateway_id: cdktf.stringToTerraform(this._transitGatewayId),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            destination_options: flowLogDestinationOptionsToTerraform(this._destinationOptions.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            deliver_cross_account_role: {
                value: cdktf.stringToHclTerraform(this._deliverCrossAccountRole),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            eni_id: {
                value: cdktf.stringToHclTerraform(this._eniId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_role_arn: {
                value: cdktf.stringToHclTerraform(this._iamRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            log_destination: {
                value: cdktf.stringToHclTerraform(this._logDestination),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            log_destination_type: {
                value: cdktf.stringToHclTerraform(this._logDestinationType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            log_format: {
                value: cdktf.stringToHclTerraform(this._logFormat),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            log_group_name: {
                value: cdktf.stringToHclTerraform(this._logGroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            max_aggregation_interval: {
                value: cdktf.numberToHclTerraform(this._maxAggregationInterval),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            subnet_id: {
                value: cdktf.stringToHclTerraform(this._subnetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            traffic_type: {
                value: cdktf.stringToHclTerraform(this._trafficType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            transit_gateway_attachment_id: {
                value: cdktf.stringToHclTerraform(this._transitGatewayAttachmentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            transit_gateway_id: {
                value: cdktf.stringToHclTerraform(this._transitGatewayId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vpc_id: {
                value: cdktf.stringToHclTerraform(this._vpcId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destination_options: {
                value: flowLogDestinationOptionsToHclTerraform(this._destinationOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FlowLogDestinationOptionsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FlowLog = FlowLog;
_b = JSII_RTTI_SYMBOL_1;
FlowLog[_b] = { fqn: "@cdktf/provider-aws.flowLog.FlowLog", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FlowLog.tfResourceType = "aws_flow_log";
