// @cdktf/provider-aws/lib/ses-event-destination/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesEventDestination = exports.SesEventDestinationSnsDestinationOutputReference = exports.sesEventDestinationSnsDestinationToHclTerraform = exports.sesEventDestinationSnsDestinationToTerraform = exports.SesEventDestinationKinesisDestinationOutputReference = exports.sesEventDestinationKinesisDestinationToHclTerraform = exports.sesEventDestinationKinesisDestinationToTerraform = exports.SesEventDestinationCloudwatchDestinationList = exports.SesEventDestinationCloudwatchDestinationOutputReference = exports.sesEventDestinationCloudwatchDestinationToHclTerraform = exports.sesEventDestinationCloudwatchDestinationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sesEventDestinationCloudwatchDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_value: cdktf.stringToTerraform(struct.defaultValue),
        dimension_name: cdktf.stringToTerraform(struct.dimensionName),
        value_source: cdktf.stringToTerraform(struct.valueSource),
    };
}
exports.sesEventDestinationCloudwatchDestinationToTerraform = sesEventDestinationCloudwatchDestinationToTerraform;
function sesEventDestinationCloudwatchDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_value: {
            value: cdktf.stringToHclTerraform(struct.defaultValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        dimension_name: {
            value: cdktf.stringToHclTerraform(struct.dimensionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value_source: {
            value: cdktf.stringToHclTerraform(struct.valueSource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sesEventDestinationCloudwatchDestinationToHclTerraform = sesEventDestinationCloudwatchDestinationToHclTerraform;
class SesEventDestinationCloudwatchDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultValue = this._defaultValue;
        }
        if (this._dimensionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimensionName = this._dimensionName;
        }
        if (this._valueSource !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueSource = this._valueSource;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._defaultValue = undefined;
            this._dimensionName = undefined;
            this._valueSource = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._defaultValue = value.defaultValue;
            this._dimensionName = value.dimensionName;
            this._valueSource = value.valueSource;
        }
    }
    get defaultValue() {
        return this.getStringAttribute('default_value');
    }
    set defaultValue(value) {
        this._defaultValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultValueInput() {
        return this._defaultValue;
    }
    get dimensionName() {
        return this.getStringAttribute('dimension_name');
    }
    set dimensionName(value) {
        this._dimensionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionNameInput() {
        return this._dimensionName;
    }
    get valueSource() {
        return this.getStringAttribute('value_source');
    }
    set valueSource(value) {
        this._valueSource = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueSourceInput() {
        return this._valueSource;
    }
}
exports.SesEventDestinationCloudwatchDestinationOutputReference = SesEventDestinationCloudwatchDestinationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SesEventDestinationCloudwatchDestinationOutputReference[_a] = { fqn: "@cdktf/provider-aws.sesEventDestination.SesEventDestinationCloudwatchDestinationOutputReference", version: "19.15.0" };
class SesEventDestinationCloudwatchDestinationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SesEventDestinationCloudwatchDestinationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SesEventDestinationCloudwatchDestinationList = SesEventDestinationCloudwatchDestinationList;
_b = JSII_RTTI_SYMBOL_1;
SesEventDestinationCloudwatchDestinationList[_b] = { fqn: "@cdktf/provider-aws.sesEventDestination.SesEventDestinationCloudwatchDestinationList", version: "19.15.0" };
function sesEventDestinationKinesisDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        stream_arn: cdktf.stringToTerraform(struct.streamArn),
    };
}
exports.sesEventDestinationKinesisDestinationToTerraform = sesEventDestinationKinesisDestinationToTerraform;
function sesEventDestinationKinesisDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        stream_arn: {
            value: cdktf.stringToHclTerraform(struct.streamArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sesEventDestinationKinesisDestinationToHclTerraform = sesEventDestinationKinesisDestinationToHclTerraform;
class SesEventDestinationKinesisDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._streamArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.streamArn = this._streamArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._roleArn = undefined;
            this._streamArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._roleArn = value.roleArn;
            this._streamArn = value.streamArn;
        }
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    set streamArn(value) {
        this._streamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamArnInput() {
        return this._streamArn;
    }
}
exports.SesEventDestinationKinesisDestinationOutputReference = SesEventDestinationKinesisDestinationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SesEventDestinationKinesisDestinationOutputReference[_c] = { fqn: "@cdktf/provider-aws.sesEventDestination.SesEventDestinationKinesisDestinationOutputReference", version: "19.15.0" };
function sesEventDestinationSnsDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        topic_arn: cdktf.stringToTerraform(struct.topicArn),
    };
}
exports.sesEventDestinationSnsDestinationToTerraform = sesEventDestinationSnsDestinationToTerraform;
function sesEventDestinationSnsDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        topic_arn: {
            value: cdktf.stringToHclTerraform(struct.topicArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sesEventDestinationSnsDestinationToHclTerraform = sesEventDestinationSnsDestinationToHclTerraform;
class SesEventDestinationSnsDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._topicArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.topicArn = this._topicArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._topicArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._topicArn = value.topicArn;
        }
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
}
exports.SesEventDestinationSnsDestinationOutputReference = SesEventDestinationSnsDestinationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SesEventDestinationSnsDestinationOutputReference[_d] = { fqn: "@cdktf/provider-aws.sesEventDestination.SesEventDestinationSnsDestinationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ses_event_destination aws_ses_event_destination}
*/
class SesEventDestination extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SesEventDestination resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SesEventDestination to import
    * @param importFromId The id of the existing SesEventDestination that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ses_event_destination#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SesEventDestination to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ses_event_destination", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ses_event_destination aws_ses_event_destination} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SesEventDestinationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_event_destination',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cloudwatch_destination - computed: false, optional: true, required: false
        this._cloudwatchDestination = new SesEventDestinationCloudwatchDestinationList(this, "cloudwatch_destination", true);
        // kinesis_destination - computed: false, optional: true, required: false
        this._kinesisDestination = new SesEventDestinationKinesisDestinationOutputReference(this, "kinesis_destination");
        // sns_destination - computed: false, optional: true, required: false
        this._snsDestination = new SesEventDestinationSnsDestinationOutputReference(this, "sns_destination");
        this._configurationSetName = config.configurationSetName;
        this._enabled = config.enabled;
        this._id = config.id;
        this._matchingTypes = config.matchingTypes;
        this._name = config.name;
        this._cloudwatchDestination.internalValue = config.cloudwatchDestination;
        this._kinesisDestination.internalValue = config.kinesisDestination;
        this._snsDestination.internalValue = config.snsDestination;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get configurationSetName() {
        return this.getStringAttribute('configuration_set_name');
    }
    set configurationSetName(value) {
        this._configurationSetName = value;
    }
    // Temporarily expose input value. Use with caution.
    get configurationSetNameInput() {
        return this._configurationSetName;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get matchingTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('matching_types'));
    }
    set matchingTypes(value) {
        this._matchingTypes = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchingTypesInput() {
        return this._matchingTypes;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get cloudwatchDestination() {
        return this._cloudwatchDestination;
    }
    putCloudwatchDestination(value) {
        this._cloudwatchDestination.internalValue = value;
    }
    resetCloudwatchDestination() {
        this._cloudwatchDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchDestinationInput() {
        return this._cloudwatchDestination.internalValue;
    }
    get kinesisDestination() {
        return this._kinesisDestination;
    }
    putKinesisDestination(value) {
        this._kinesisDestination.internalValue = value;
    }
    resetKinesisDestination() {
        this._kinesisDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisDestinationInput() {
        return this._kinesisDestination.internalValue;
    }
    get snsDestination() {
        return this._snsDestination;
    }
    putSnsDestination(value) {
        this._snsDestination.internalValue = value;
    }
    resetSnsDestination() {
        this._snsDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snsDestinationInput() {
        return this._snsDestination.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            configuration_set_name: cdktf.stringToTerraform(this._configurationSetName),
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            matching_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._matchingTypes),
            name: cdktf.stringToTerraform(this._name),
            cloudwatch_destination: cdktf.listMapper(sesEventDestinationCloudwatchDestinationToTerraform, true)(this._cloudwatchDestination.internalValue),
            kinesis_destination: sesEventDestinationKinesisDestinationToTerraform(this._kinesisDestination.internalValue),
            sns_destination: sesEventDestinationSnsDestinationToTerraform(this._snsDestination.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            configuration_set_name: {
                value: cdktf.stringToHclTerraform(this._configurationSetName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            matching_types: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._matchingTypes),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloudwatch_destination: {
                value: cdktf.listMapperHcl(sesEventDestinationCloudwatchDestinationToHclTerraform, true)(this._cloudwatchDestination.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "SesEventDestinationCloudwatchDestinationList",
            },
            kinesis_destination: {
                value: sesEventDestinationKinesisDestinationToHclTerraform(this._kinesisDestination.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SesEventDestinationKinesisDestinationList",
            },
            sns_destination: {
                value: sesEventDestinationSnsDestinationToHclTerraform(this._snsDestination.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SesEventDestinationSnsDestinationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SesEventDestination = SesEventDestination;
_e = JSII_RTTI_SYMBOL_1;
SesEventDestination[_e] = { fqn: "@cdktf/provider-aws.sesEventDestination.SesEventDestination", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SesEventDestination.tfResourceType = "aws_ses_event_destination";
