// @cdktf/provider-aws/lib/ce-cost-category/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CeCostCategoryRuleRuleTagsOutputReference = exports.ceCostCategoryRuleRuleTagsToHclTerraform = exports.ceCostCategoryRuleRuleTagsToTerraform = exports.CeCostCategoryRuleRuleOrList = exports.CeCostCategoryRuleRuleOrOutputReference = exports.ceCostCategoryRuleRuleOrToHclTerraform = exports.ceCostCategoryRuleRuleOrToTerraform = exports.CeCostCategoryRuleRuleOrTagsOutputReference = exports.ceCostCategoryRuleRuleOrTagsToHclTerraform = exports.ceCostCategoryRuleRuleOrTagsToTerraform = exports.CeCostCategoryRuleRuleOrDimensionOutputReference = exports.ceCostCategoryRuleRuleOrDimensionToHclTerraform = exports.ceCostCategoryRuleRuleOrDimensionToTerraform = exports.CeCostCategoryRuleRuleOrCostCategoryOutputReference = exports.ceCostCategoryRuleRuleOrCostCategoryToHclTerraform = exports.ceCostCategoryRuleRuleOrCostCategoryToTerraform = exports.CeCostCategoryRuleRuleNotOutputReference = exports.ceCostCategoryRuleRuleNotToHclTerraform = exports.ceCostCategoryRuleRuleNotToTerraform = exports.CeCostCategoryRuleRuleNotTagsOutputReference = exports.ceCostCategoryRuleRuleNotTagsToHclTerraform = exports.ceCostCategoryRuleRuleNotTagsToTerraform = exports.CeCostCategoryRuleRuleNotDimensionOutputReference = exports.ceCostCategoryRuleRuleNotDimensionToHclTerraform = exports.ceCostCategoryRuleRuleNotDimensionToTerraform = exports.CeCostCategoryRuleRuleNotCostCategoryOutputReference = exports.ceCostCategoryRuleRuleNotCostCategoryToHclTerraform = exports.ceCostCategoryRuleRuleNotCostCategoryToTerraform = exports.CeCostCategoryRuleRuleDimensionOutputReference = exports.ceCostCategoryRuleRuleDimensionToHclTerraform = exports.ceCostCategoryRuleRuleDimensionToTerraform = exports.CeCostCategoryRuleRuleCostCategoryOutputReference = exports.ceCostCategoryRuleRuleCostCategoryToHclTerraform = exports.ceCostCategoryRuleRuleCostCategoryToTerraform = exports.CeCostCategoryRuleRuleAndList = exports.CeCostCategoryRuleRuleAndOutputReference = exports.ceCostCategoryRuleRuleAndToHclTerraform = exports.ceCostCategoryRuleRuleAndToTerraform = exports.CeCostCategoryRuleRuleAndTagsOutputReference = exports.ceCostCategoryRuleRuleAndTagsToHclTerraform = exports.ceCostCategoryRuleRuleAndTagsToTerraform = exports.CeCostCategoryRuleRuleAndDimensionOutputReference = exports.ceCostCategoryRuleRuleAndDimensionToHclTerraform = exports.ceCostCategoryRuleRuleAndDimensionToTerraform = exports.CeCostCategoryRuleRuleAndCostCategoryOutputReference = exports.ceCostCategoryRuleRuleAndCostCategoryToHclTerraform = exports.ceCostCategoryRuleRuleAndCostCategoryToTerraform = exports.CeCostCategoryRuleInheritedValueOutputReference = exports.ceCostCategoryRuleInheritedValueToHclTerraform = exports.ceCostCategoryRuleInheritedValueToTerraform = void 0;
exports.CeCostCategory = exports.CeCostCategorySplitChargeRuleList = exports.CeCostCategorySplitChargeRuleOutputReference = exports.ceCostCategorySplitChargeRuleToHclTerraform = exports.ceCostCategorySplitChargeRuleToTerraform = exports.CeCostCategorySplitChargeRuleParameterList = exports.CeCostCategorySplitChargeRuleParameterOutputReference = exports.ceCostCategorySplitChargeRuleParameterToHclTerraform = exports.ceCostCategorySplitChargeRuleParameterToTerraform = exports.CeCostCategoryRuleList = exports.CeCostCategoryRuleOutputReference = exports.ceCostCategoryRuleToHclTerraform = exports.ceCostCategoryRuleToTerraform = exports.CeCostCategoryRuleRuleOutputReference = exports.ceCostCategoryRuleRuleToHclTerraform = exports.ceCostCategoryRuleRuleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ceCostCategoryRuleInheritedValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dimension_key: cdktf.stringToTerraform(struct.dimensionKey),
        dimension_name: cdktf.stringToTerraform(struct.dimensionName),
    };
}
exports.ceCostCategoryRuleInheritedValueToTerraform = ceCostCategoryRuleInheritedValueToTerraform;
function ceCostCategoryRuleInheritedValueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dimension_key: {
            value: cdktf.stringToHclTerraform(struct.dimensionKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        dimension_name: {
            value: cdktf.stringToHclTerraform(struct.dimensionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleInheritedValueToHclTerraform = ceCostCategoryRuleInheritedValueToHclTerraform;
class CeCostCategoryRuleInheritedValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dimensionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimensionKey = this._dimensionKey;
        }
        if (this._dimensionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimensionName = this._dimensionName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dimensionKey = undefined;
            this._dimensionName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dimensionKey = value.dimensionKey;
            this._dimensionName = value.dimensionName;
        }
    }
    get dimensionKey() {
        return this.getStringAttribute('dimension_key');
    }
    set dimensionKey(value) {
        this._dimensionKey = value;
    }
    resetDimensionKey() {
        this._dimensionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionKeyInput() {
        return this._dimensionKey;
    }
    get dimensionName() {
        return this.getStringAttribute('dimension_name');
    }
    set dimensionName(value) {
        this._dimensionName = value;
    }
    resetDimensionName() {
        this._dimensionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionNameInput() {
        return this._dimensionName;
    }
}
exports.CeCostCategoryRuleInheritedValueOutputReference = CeCostCategoryRuleInheritedValueOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleInheritedValueOutputReference[_a] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleInheritedValueOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleAndCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleAndCostCategoryToTerraform = ceCostCategoryRuleRuleAndCostCategoryToTerraform;
function ceCostCategoryRuleRuleAndCostCategoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleAndCostCategoryToHclTerraform = ceCostCategoryRuleRuleAndCostCategoryToHclTerraform;
class CeCostCategoryRuleRuleAndCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleAndCostCategoryOutputReference = CeCostCategoryRuleRuleAndCostCategoryOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleAndCostCategoryOutputReference[_b] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleAndCostCategoryOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleAndDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleAndDimensionToTerraform = ceCostCategoryRuleRuleAndDimensionToTerraform;
function ceCostCategoryRuleRuleAndDimensionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleAndDimensionToHclTerraform = ceCostCategoryRuleRuleAndDimensionToHclTerraform;
class CeCostCategoryRuleRuleAndDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleAndDimensionOutputReference = CeCostCategoryRuleRuleAndDimensionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleAndDimensionOutputReference[_c] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleAndDimensionOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleAndTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleAndTagsToTerraform = ceCostCategoryRuleRuleAndTagsToTerraform;
function ceCostCategoryRuleRuleAndTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleAndTagsToHclTerraform = ceCostCategoryRuleRuleAndTagsToHclTerraform;
class CeCostCategoryRuleRuleAndTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleAndTagsOutputReference = CeCostCategoryRuleRuleAndTagsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleAndTagsOutputReference[_d] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleAndTagsOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleAndToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cost_category: ceCostCategoryRuleRuleAndCostCategoryToTerraform(struct.costCategory),
        dimension: ceCostCategoryRuleRuleAndDimensionToTerraform(struct.dimension),
        tags: ceCostCategoryRuleRuleAndTagsToTerraform(struct.tags),
    };
}
exports.ceCostCategoryRuleRuleAndToTerraform = ceCostCategoryRuleRuleAndToTerraform;
function ceCostCategoryRuleRuleAndToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cost_category: {
            value: ceCostCategoryRuleRuleAndCostCategoryToHclTerraform(struct.costCategory),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleAndCostCategoryList",
        },
        dimension: {
            value: ceCostCategoryRuleRuleAndDimensionToHclTerraform(struct.dimension),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleAndDimensionList",
        },
        tags: {
            value: ceCostCategoryRuleRuleAndTagsToHclTerraform(struct.tags),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleAndTagsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleAndToHclTerraform = ceCostCategoryRuleRuleAndToHclTerraform;
class CeCostCategoryRuleRuleAndOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cost_category - computed: false, optional: true, required: false
        this._costCategory = new CeCostCategoryRuleRuleAndCostCategoryOutputReference(this, "cost_category");
        // dimension - computed: false, optional: true, required: false
        this._dimension = new CeCostCategoryRuleRuleAndDimensionOutputReference(this, "dimension");
        // tags - computed: false, optional: true, required: false
        this._tags = new CeCostCategoryRuleRuleAndTagsOutputReference(this, "tags");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._costCategory?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.costCategory = this._costCategory?.internalValue;
        }
        if (this._dimension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimension = this._dimension?.internalValue;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._costCategory.internalValue = undefined;
            this._dimension.internalValue = undefined;
            this._tags.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._costCategory.internalValue = value.costCategory;
            this._dimension.internalValue = value.dimension;
            this._tags.internalValue = value.tags;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    putCostCategory(value) {
        this._costCategory.internalValue = value;
    }
    resetCostCategory() {
        this._costCategory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get costCategoryInput() {
        return this._costCategory.internalValue;
    }
    get dimension() {
        return this._dimension;
    }
    putDimension(value) {
        this._dimension.internalValue = value;
    }
    resetDimension() {
        this._dimension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionInput() {
        return this._dimension.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.CeCostCategoryRuleRuleAndOutputReference = CeCostCategoryRuleRuleAndOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleAndOutputReference[_e] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleAndOutputReference", version: "19.15.0" };
class CeCostCategoryRuleRuleAndList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CeCostCategoryRuleRuleAndOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CeCostCategoryRuleRuleAndList = CeCostCategoryRuleRuleAndList;
_f = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleAndList[_f] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleAndList", version: "19.15.0" };
function ceCostCategoryRuleRuleCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleCostCategoryToTerraform = ceCostCategoryRuleRuleCostCategoryToTerraform;
function ceCostCategoryRuleRuleCostCategoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleCostCategoryToHclTerraform = ceCostCategoryRuleRuleCostCategoryToHclTerraform;
class CeCostCategoryRuleRuleCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleCostCategoryOutputReference = CeCostCategoryRuleRuleCostCategoryOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleCostCategoryOutputReference[_g] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleCostCategoryOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleDimensionToTerraform = ceCostCategoryRuleRuleDimensionToTerraform;
function ceCostCategoryRuleRuleDimensionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleDimensionToHclTerraform = ceCostCategoryRuleRuleDimensionToHclTerraform;
class CeCostCategoryRuleRuleDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleDimensionOutputReference = CeCostCategoryRuleRuleDimensionOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleDimensionOutputReference[_h] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleDimensionOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleNotCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleNotCostCategoryToTerraform = ceCostCategoryRuleRuleNotCostCategoryToTerraform;
function ceCostCategoryRuleRuleNotCostCategoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleNotCostCategoryToHclTerraform = ceCostCategoryRuleRuleNotCostCategoryToHclTerraform;
class CeCostCategoryRuleRuleNotCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleNotCostCategoryOutputReference = CeCostCategoryRuleRuleNotCostCategoryOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleNotCostCategoryOutputReference[_j] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleNotCostCategoryOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleNotDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleNotDimensionToTerraform = ceCostCategoryRuleRuleNotDimensionToTerraform;
function ceCostCategoryRuleRuleNotDimensionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleNotDimensionToHclTerraform = ceCostCategoryRuleRuleNotDimensionToHclTerraform;
class CeCostCategoryRuleRuleNotDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleNotDimensionOutputReference = CeCostCategoryRuleRuleNotDimensionOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleNotDimensionOutputReference[_k] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleNotDimensionOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleNotTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleNotTagsToTerraform = ceCostCategoryRuleRuleNotTagsToTerraform;
function ceCostCategoryRuleRuleNotTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleNotTagsToHclTerraform = ceCostCategoryRuleRuleNotTagsToHclTerraform;
class CeCostCategoryRuleRuleNotTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleNotTagsOutputReference = CeCostCategoryRuleRuleNotTagsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleNotTagsOutputReference[_l] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleNotTagsOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleNotToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cost_category: ceCostCategoryRuleRuleNotCostCategoryToTerraform(struct.costCategory),
        dimension: ceCostCategoryRuleRuleNotDimensionToTerraform(struct.dimension),
        tags: ceCostCategoryRuleRuleNotTagsToTerraform(struct.tags),
    };
}
exports.ceCostCategoryRuleRuleNotToTerraform = ceCostCategoryRuleRuleNotToTerraform;
function ceCostCategoryRuleRuleNotToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cost_category: {
            value: ceCostCategoryRuleRuleNotCostCategoryToHclTerraform(struct.costCategory),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleNotCostCategoryList",
        },
        dimension: {
            value: ceCostCategoryRuleRuleNotDimensionToHclTerraform(struct.dimension),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleNotDimensionList",
        },
        tags: {
            value: ceCostCategoryRuleRuleNotTagsToHclTerraform(struct.tags),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleNotTagsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleNotToHclTerraform = ceCostCategoryRuleRuleNotToHclTerraform;
class CeCostCategoryRuleRuleNotOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cost_category - computed: false, optional: true, required: false
        this._costCategory = new CeCostCategoryRuleRuleNotCostCategoryOutputReference(this, "cost_category");
        // dimension - computed: false, optional: true, required: false
        this._dimension = new CeCostCategoryRuleRuleNotDimensionOutputReference(this, "dimension");
        // tags - computed: false, optional: true, required: false
        this._tags = new CeCostCategoryRuleRuleNotTagsOutputReference(this, "tags");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._costCategory?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.costCategory = this._costCategory?.internalValue;
        }
        if (this._dimension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimension = this._dimension?.internalValue;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._costCategory.internalValue = undefined;
            this._dimension.internalValue = undefined;
            this._tags.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._costCategory.internalValue = value.costCategory;
            this._dimension.internalValue = value.dimension;
            this._tags.internalValue = value.tags;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    putCostCategory(value) {
        this._costCategory.internalValue = value;
    }
    resetCostCategory() {
        this._costCategory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get costCategoryInput() {
        return this._costCategory.internalValue;
    }
    get dimension() {
        return this._dimension;
    }
    putDimension(value) {
        this._dimension.internalValue = value;
    }
    resetDimension() {
        this._dimension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionInput() {
        return this._dimension.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.CeCostCategoryRuleRuleNotOutputReference = CeCostCategoryRuleRuleNotOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleNotOutputReference[_m] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleNotOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleOrCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleOrCostCategoryToTerraform = ceCostCategoryRuleRuleOrCostCategoryToTerraform;
function ceCostCategoryRuleRuleOrCostCategoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleOrCostCategoryToHclTerraform = ceCostCategoryRuleRuleOrCostCategoryToHclTerraform;
class CeCostCategoryRuleRuleOrCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleOrCostCategoryOutputReference = CeCostCategoryRuleRuleOrCostCategoryOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOrCostCategoryOutputReference[_o] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOrCostCategoryOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleOrDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleOrDimensionToTerraform = ceCostCategoryRuleRuleOrDimensionToTerraform;
function ceCostCategoryRuleRuleOrDimensionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleOrDimensionToHclTerraform = ceCostCategoryRuleRuleOrDimensionToHclTerraform;
class CeCostCategoryRuleRuleOrDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleOrDimensionOutputReference = CeCostCategoryRuleRuleOrDimensionOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOrDimensionOutputReference[_p] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOrDimensionOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleOrTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleOrTagsToTerraform = ceCostCategoryRuleRuleOrTagsToTerraform;
function ceCostCategoryRuleRuleOrTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleOrTagsToHclTerraform = ceCostCategoryRuleRuleOrTagsToHclTerraform;
class CeCostCategoryRuleRuleOrTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleOrTagsOutputReference = CeCostCategoryRuleRuleOrTagsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOrTagsOutputReference[_q] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOrTagsOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleOrToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cost_category: ceCostCategoryRuleRuleOrCostCategoryToTerraform(struct.costCategory),
        dimension: ceCostCategoryRuleRuleOrDimensionToTerraform(struct.dimension),
        tags: ceCostCategoryRuleRuleOrTagsToTerraform(struct.tags),
    };
}
exports.ceCostCategoryRuleRuleOrToTerraform = ceCostCategoryRuleRuleOrToTerraform;
function ceCostCategoryRuleRuleOrToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cost_category: {
            value: ceCostCategoryRuleRuleOrCostCategoryToHclTerraform(struct.costCategory),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleOrCostCategoryList",
        },
        dimension: {
            value: ceCostCategoryRuleRuleOrDimensionToHclTerraform(struct.dimension),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleOrDimensionList",
        },
        tags: {
            value: ceCostCategoryRuleRuleOrTagsToHclTerraform(struct.tags),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleOrTagsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleOrToHclTerraform = ceCostCategoryRuleRuleOrToHclTerraform;
class CeCostCategoryRuleRuleOrOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cost_category - computed: false, optional: true, required: false
        this._costCategory = new CeCostCategoryRuleRuleOrCostCategoryOutputReference(this, "cost_category");
        // dimension - computed: false, optional: true, required: false
        this._dimension = new CeCostCategoryRuleRuleOrDimensionOutputReference(this, "dimension");
        // tags - computed: false, optional: true, required: false
        this._tags = new CeCostCategoryRuleRuleOrTagsOutputReference(this, "tags");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._costCategory?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.costCategory = this._costCategory?.internalValue;
        }
        if (this._dimension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimension = this._dimension?.internalValue;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._costCategory.internalValue = undefined;
            this._dimension.internalValue = undefined;
            this._tags.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._costCategory.internalValue = value.costCategory;
            this._dimension.internalValue = value.dimension;
            this._tags.internalValue = value.tags;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    putCostCategory(value) {
        this._costCategory.internalValue = value;
    }
    resetCostCategory() {
        this._costCategory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get costCategoryInput() {
        return this._costCategory.internalValue;
    }
    get dimension() {
        return this._dimension;
    }
    putDimension(value) {
        this._dimension.internalValue = value;
    }
    resetDimension() {
        this._dimension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionInput() {
        return this._dimension.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.CeCostCategoryRuleRuleOrOutputReference = CeCostCategoryRuleRuleOrOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOrOutputReference[_r] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOrOutputReference", version: "19.15.0" };
class CeCostCategoryRuleRuleOrList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CeCostCategoryRuleRuleOrOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CeCostCategoryRuleRuleOrList = CeCostCategoryRuleRuleOrList;
_s = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOrList[_s] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOrList", version: "19.15.0" };
function ceCostCategoryRuleRuleTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        match_options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.matchOptions),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategoryRuleRuleTagsToTerraform = ceCostCategoryRuleRuleTagsToTerraform;
function ceCostCategoryRuleRuleTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        match_options: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.matchOptions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleTagsToHclTerraform = ceCostCategoryRuleRuleTagsToHclTerraform;
class CeCostCategoryRuleRuleTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._matchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchOptions = this._matchOptions;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._matchOptions = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._matchOptions = value.matchOptions;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    set matchOptions(value) {
        this._matchOptions = value;
    }
    resetMatchOptions() {
        this._matchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchOptionsInput() {
        return this._matchOptions;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategoryRuleRuleTagsOutputReference = CeCostCategoryRuleRuleTagsOutputReference;
_t = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleTagsOutputReference[_t] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleTagsOutputReference", version: "19.15.0" };
function ceCostCategoryRuleRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        and: cdktf.listMapper(ceCostCategoryRuleRuleAndToTerraform, true)(struct.and),
        cost_category: ceCostCategoryRuleRuleCostCategoryToTerraform(struct.costCategory),
        dimension: ceCostCategoryRuleRuleDimensionToTerraform(struct.dimension),
        not: ceCostCategoryRuleRuleNotToTerraform(struct.not),
        or: cdktf.listMapper(ceCostCategoryRuleRuleOrToTerraform, true)(struct.or),
        tags: ceCostCategoryRuleRuleTagsToTerraform(struct.tags),
    };
}
exports.ceCostCategoryRuleRuleToTerraform = ceCostCategoryRuleRuleToTerraform;
function ceCostCategoryRuleRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        and: {
            value: cdktf.listMapperHcl(ceCostCategoryRuleRuleAndToHclTerraform, true)(struct.and),
            isBlock: true,
            type: "set",
            storageClassType: "CeCostCategoryRuleRuleAndList",
        },
        cost_category: {
            value: ceCostCategoryRuleRuleCostCategoryToHclTerraform(struct.costCategory),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleCostCategoryList",
        },
        dimension: {
            value: ceCostCategoryRuleRuleDimensionToHclTerraform(struct.dimension),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleDimensionList",
        },
        not: {
            value: ceCostCategoryRuleRuleNotToHclTerraform(struct.not),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleNotList",
        },
        or: {
            value: cdktf.listMapperHcl(ceCostCategoryRuleRuleOrToHclTerraform, true)(struct.or),
            isBlock: true,
            type: "set",
            storageClassType: "CeCostCategoryRuleRuleOrList",
        },
        tags: {
            value: ceCostCategoryRuleRuleTagsToHclTerraform(struct.tags),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleTagsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleRuleToHclTerraform = ceCostCategoryRuleRuleToHclTerraform;
class CeCostCategoryRuleRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // and - computed: false, optional: true, required: false
        this._and = new CeCostCategoryRuleRuleAndList(this, "and", true);
        // cost_category - computed: false, optional: true, required: false
        this._costCategory = new CeCostCategoryRuleRuleCostCategoryOutputReference(this, "cost_category");
        // dimension - computed: false, optional: true, required: false
        this._dimension = new CeCostCategoryRuleRuleDimensionOutputReference(this, "dimension");
        // not - computed: false, optional: true, required: false
        this._not = new CeCostCategoryRuleRuleNotOutputReference(this, "not");
        // or - computed: false, optional: true, required: false
        this._or = new CeCostCategoryRuleRuleOrList(this, "or", true);
        // tags - computed: false, optional: true, required: false
        this._tags = new CeCostCategoryRuleRuleTagsOutputReference(this, "tags");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._and?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.and = this._and?.internalValue;
        }
        if (this._costCategory?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.costCategory = this._costCategory?.internalValue;
        }
        if (this._dimension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimension = this._dimension?.internalValue;
        }
        if (this._not?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.not = this._not?.internalValue;
        }
        if (this._or?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.or = this._or?.internalValue;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._and.internalValue = undefined;
            this._costCategory.internalValue = undefined;
            this._dimension.internalValue = undefined;
            this._not.internalValue = undefined;
            this._or.internalValue = undefined;
            this._tags.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._and.internalValue = value.and;
            this._costCategory.internalValue = value.costCategory;
            this._dimension.internalValue = value.dimension;
            this._not.internalValue = value.not;
            this._or.internalValue = value.or;
            this._tags.internalValue = value.tags;
        }
    }
    get and() {
        return this._and;
    }
    putAnd(value) {
        this._and.internalValue = value;
    }
    resetAnd() {
        this._and.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get andInput() {
        return this._and.internalValue;
    }
    get costCategory() {
        return this._costCategory;
    }
    putCostCategory(value) {
        this._costCategory.internalValue = value;
    }
    resetCostCategory() {
        this._costCategory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get costCategoryInput() {
        return this._costCategory.internalValue;
    }
    get dimension() {
        return this._dimension;
    }
    putDimension(value) {
        this._dimension.internalValue = value;
    }
    resetDimension() {
        this._dimension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionInput() {
        return this._dimension.internalValue;
    }
    get not() {
        return this._not;
    }
    putNot(value) {
        this._not.internalValue = value;
    }
    resetNot() {
        this._not.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notInput() {
        return this._not.internalValue;
    }
    get or() {
        return this._or;
    }
    putOr(value) {
        this._or.internalValue = value;
    }
    resetOr() {
        this._or.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orInput() {
        return this._or.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.CeCostCategoryRuleRuleOutputReference = CeCostCategoryRuleRuleOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleRuleOutputReference[_u] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleRuleOutputReference", version: "19.15.0" };
function ceCostCategoryRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
        inherited_value: ceCostCategoryRuleInheritedValueToTerraform(struct.inheritedValue),
        rule: ceCostCategoryRuleRuleToTerraform(struct.rule),
    };
}
exports.ceCostCategoryRuleToTerraform = ceCostCategoryRuleToTerraform;
function ceCostCategoryRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        inherited_value: {
            value: ceCostCategoryRuleInheritedValueToHclTerraform(struct.inheritedValue),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleInheritedValueList",
        },
        rule: {
            value: ceCostCategoryRuleRuleToHclTerraform(struct.rule),
            isBlock: true,
            type: "list",
            storageClassType: "CeCostCategoryRuleRuleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategoryRuleToHclTerraform = ceCostCategoryRuleToHclTerraform;
class CeCostCategoryRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // inherited_value - computed: false, optional: true, required: false
        this._inheritedValue = new CeCostCategoryRuleInheritedValueOutputReference(this, "inherited_value");
        // rule - computed: false, optional: true, required: false
        this._rule = new CeCostCategoryRuleRuleOutputReference(this, "rule");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        if (this._inheritedValue?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.inheritedValue = this._inheritedValue?.internalValue;
        }
        if (this._rule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rule = this._rule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._value = undefined;
            this._inheritedValue.internalValue = undefined;
            this._rule.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._value = value.value;
            this._inheritedValue.internalValue = value.inheritedValue;
            this._rule.internalValue = value.rule;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    get inheritedValue() {
        return this._inheritedValue;
    }
    putInheritedValue(value) {
        this._inheritedValue.internalValue = value;
    }
    resetInheritedValue() {
        this._inheritedValue.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inheritedValueInput() {
        return this._inheritedValue.internalValue;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
}
exports.CeCostCategoryRuleOutputReference = CeCostCategoryRuleOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleOutputReference[_v] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleOutputReference", version: "19.15.0" };
class CeCostCategoryRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CeCostCategoryRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CeCostCategoryRuleList = CeCostCategoryRuleList;
_w = JSII_RTTI_SYMBOL_1;
CeCostCategoryRuleList[_w] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategoryRuleList", version: "19.15.0" };
function ceCostCategorySplitChargeRuleParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.ceCostCategorySplitChargeRuleParameterToTerraform = ceCostCategorySplitChargeRuleParameterToTerraform;
function ceCostCategorySplitChargeRuleParameterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategorySplitChargeRuleParameterToHclTerraform = ceCostCategorySplitChargeRuleParameterToHclTerraform;
class CeCostCategorySplitChargeRuleParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._values = value.values;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CeCostCategorySplitChargeRuleParameterOutputReference = CeCostCategorySplitChargeRuleParameterOutputReference;
_x = JSII_RTTI_SYMBOL_1;
CeCostCategorySplitChargeRuleParameterOutputReference[_x] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategorySplitChargeRuleParameterOutputReference", version: "19.15.0" };
class CeCostCategorySplitChargeRuleParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CeCostCategorySplitChargeRuleParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CeCostCategorySplitChargeRuleParameterList = CeCostCategorySplitChargeRuleParameterList;
_y = JSII_RTTI_SYMBOL_1;
CeCostCategorySplitChargeRuleParameterList[_y] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategorySplitChargeRuleParameterList", version: "19.15.0" };
function ceCostCategorySplitChargeRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        method: cdktf.stringToTerraform(struct.method),
        source: cdktf.stringToTerraform(struct.source),
        targets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.targets),
        parameter: cdktf.listMapper(ceCostCategorySplitChargeRuleParameterToTerraform, true)(struct.parameter),
    };
}
exports.ceCostCategorySplitChargeRuleToTerraform = ceCostCategorySplitChargeRuleToTerraform;
function ceCostCategorySplitChargeRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        method: {
            value: cdktf.stringToHclTerraform(struct.method),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source: {
            value: cdktf.stringToHclTerraform(struct.source),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        targets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.targets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        parameter: {
            value: cdktf.listMapperHcl(ceCostCategorySplitChargeRuleParameterToHclTerraform, true)(struct.parameter),
            isBlock: true,
            type: "set",
            storageClassType: "CeCostCategorySplitChargeRuleParameterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ceCostCategorySplitChargeRuleToHclTerraform = ceCostCategorySplitChargeRuleToHclTerraform;
class CeCostCategorySplitChargeRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameter - computed: false, optional: true, required: false
        this._parameter = new CeCostCategorySplitChargeRuleParameterList(this, "parameter", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._method !== undefined) {
            hasAnyValues = true;
            internalValueResult.method = this._method;
        }
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        if (this._targets !== undefined) {
            hasAnyValues = true;
            internalValueResult.targets = this._targets;
        }
        if (this._parameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._method = undefined;
            this._source = undefined;
            this._targets = undefined;
            this._parameter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._method = value.method;
            this._source = value.source;
            this._targets = value.targets;
            this._parameter.internalValue = value.parameter;
        }
    }
    get method() {
        return this.getStringAttribute('method');
    }
    set method(value) {
        this._method = value;
    }
    // Temporarily expose input value. Use with caution.
    get methodInput() {
        return this._method;
    }
    get source() {
        return this.getStringAttribute('source');
    }
    set source(value) {
        this._source = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
    get targets() {
        return cdktf.Fn.tolist(this.getListAttribute('targets'));
    }
    set targets(value) {
        this._targets = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetsInput() {
        return this._targets;
    }
    get parameter() {
        return this._parameter;
    }
    putParameter(value) {
        this._parameter.internalValue = value;
    }
    resetParameter() {
        this._parameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterInput() {
        return this._parameter.internalValue;
    }
}
exports.CeCostCategorySplitChargeRuleOutputReference = CeCostCategorySplitChargeRuleOutputReference;
_z = JSII_RTTI_SYMBOL_1;
CeCostCategorySplitChargeRuleOutputReference[_z] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategorySplitChargeRuleOutputReference", version: "19.15.0" };
class CeCostCategorySplitChargeRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CeCostCategorySplitChargeRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CeCostCategorySplitChargeRuleList = CeCostCategorySplitChargeRuleList;
_0 = JSII_RTTI_SYMBOL_1;
CeCostCategorySplitChargeRuleList[_0] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategorySplitChargeRuleList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_cost_category aws_ce_cost_category}
*/
class CeCostCategory extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CeCostCategory resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CeCostCategory to import
    * @param importFromId The id of the existing CeCostCategory that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_cost_category#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CeCostCategory to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ce_cost_category", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_cost_category aws_ce_cost_category} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CeCostCategoryConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ce_cost_category',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: false, optional: false, required: true
        this._rule = new CeCostCategoryRuleList(this, "rule", true);
        // split_charge_rule - computed: false, optional: true, required: false
        this._splitChargeRule = new CeCostCategorySplitChargeRuleList(this, "split_charge_rule", true);
        this._defaultValue = config.defaultValue;
        this._effectiveStart = config.effectiveStart;
        this._id = config.id;
        this._name = config.name;
        this._ruleVersion = config.ruleVersion;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._rule.internalValue = config.rule;
        this._splitChargeRule.internalValue = config.splitChargeRule;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get defaultValue() {
        return this.getStringAttribute('default_value');
    }
    set defaultValue(value) {
        this._defaultValue = value;
    }
    resetDefaultValue() {
        this._defaultValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultValueInput() {
        return this._defaultValue;
    }
    // effective_end - computed: true, optional: false, required: false
    get effectiveEnd() {
        return this.getStringAttribute('effective_end');
    }
    get effectiveStart() {
        return this.getStringAttribute('effective_start');
    }
    set effectiveStart(value) {
        this._effectiveStart = value;
    }
    resetEffectiveStart() {
        this._effectiveStart = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get effectiveStartInput() {
        return this._effectiveStart;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ruleVersion() {
        return this.getStringAttribute('rule_version');
    }
    set ruleVersion(value) {
        this._ruleVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleVersionInput() {
        return this._ruleVersion;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    get splitChargeRule() {
        return this._splitChargeRule;
    }
    putSplitChargeRule(value) {
        this._splitChargeRule.internalValue = value;
    }
    resetSplitChargeRule() {
        this._splitChargeRule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get splitChargeRuleInput() {
        return this._splitChargeRule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_value: cdktf.stringToTerraform(this._defaultValue),
            effective_start: cdktf.stringToTerraform(this._effectiveStart),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            rule_version: cdktf.stringToTerraform(this._ruleVersion),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            rule: cdktf.listMapper(ceCostCategoryRuleToTerraform, true)(this._rule.internalValue),
            split_charge_rule: cdktf.listMapper(ceCostCategorySplitChargeRuleToTerraform, true)(this._splitChargeRule.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            default_value: {
                value: cdktf.stringToHclTerraform(this._defaultValue),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            effective_start: {
                value: cdktf.stringToHclTerraform(this._effectiveStart),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rule_version: {
                value: cdktf.stringToHclTerraform(this._ruleVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            rule: {
                value: cdktf.listMapperHcl(ceCostCategoryRuleToHclTerraform, true)(this._rule.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CeCostCategoryRuleList",
            },
            split_charge_rule: {
                value: cdktf.listMapperHcl(ceCostCategorySplitChargeRuleToHclTerraform, true)(this._splitChargeRule.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CeCostCategorySplitChargeRuleList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CeCostCategory = CeCostCategory;
_1 = JSII_RTTI_SYMBOL_1;
CeCostCategory[_1] = { fqn: "@cdktf/provider-aws.ceCostCategory.CeCostCategory", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CeCostCategory.tfResourceType = "aws_ce_cost_category";
