// @cdktf/provider-aws/lib/glue-catalog-database/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueCatalogDatabase = exports.GlueCatalogDatabaseTargetDatabaseOutputReference = exports.glueCatalogDatabaseTargetDatabaseToHclTerraform = exports.glueCatalogDatabaseTargetDatabaseToTerraform = exports.GlueCatalogDatabaseFederatedDatabaseOutputReference = exports.glueCatalogDatabaseFederatedDatabaseToHclTerraform = exports.glueCatalogDatabaseFederatedDatabaseToTerraform = exports.GlueCatalogDatabaseCreateTableDefaultPermissionList = exports.GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference = exports.glueCatalogDatabaseCreateTableDefaultPermissionToHclTerraform = exports.glueCatalogDatabaseCreateTableDefaultPermissionToTerraform = exports.GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference = exports.glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToHclTerraform = exports.glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_lake_principal_identifier: cdktf.stringToTerraform(struct.dataLakePrincipalIdentifier),
    };
}
exports.glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToTerraform = glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToTerraform;
function glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_lake_principal_identifier: {
            value: cdktf.stringToHclTerraform(struct.dataLakePrincipalIdentifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToHclTerraform = glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToHclTerraform;
class GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataLakePrincipalIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataLakePrincipalIdentifier = this._dataLakePrincipalIdentifier;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataLakePrincipalIdentifier = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataLakePrincipalIdentifier = value.dataLakePrincipalIdentifier;
        }
    }
    get dataLakePrincipalIdentifier() {
        return this.getStringAttribute('data_lake_principal_identifier');
    }
    set dataLakePrincipalIdentifier(value) {
        this._dataLakePrincipalIdentifier = value;
    }
    resetDataLakePrincipalIdentifier() {
        this._dataLakePrincipalIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataLakePrincipalIdentifierInput() {
        return this._dataLakePrincipalIdentifier;
    }
}
exports.GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference = GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference[_a] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference", version: "19.15.0" };
function glueCatalogDatabaseCreateTableDefaultPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        permissions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.permissions),
        principal: glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToTerraform(struct.principal),
    };
}
exports.glueCatalogDatabaseCreateTableDefaultPermissionToTerraform = glueCatalogDatabaseCreateTableDefaultPermissionToTerraform;
function glueCatalogDatabaseCreateTableDefaultPermissionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        permissions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.permissions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        principal: {
            value: glueCatalogDatabaseCreateTableDefaultPermissionPrincipalToHclTerraform(struct.principal),
            isBlock: true,
            type: "list",
            storageClassType: "GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.glueCatalogDatabaseCreateTableDefaultPermissionToHclTerraform = glueCatalogDatabaseCreateTableDefaultPermissionToHclTerraform;
class GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // principal - computed: false, optional: true, required: false
        this._principal = new GlueCatalogDatabaseCreateTableDefaultPermissionPrincipalOutputReference(this, "principal");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._permissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.permissions = this._permissions;
        }
        if (this._principal?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.principal = this._principal?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._permissions = undefined;
            this._principal.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._permissions = value.permissions;
            this._principal.internalValue = value.principal;
        }
    }
    get permissions() {
        return cdktf.Fn.tolist(this.getListAttribute('permissions'));
    }
    set permissions(value) {
        this._permissions = value;
    }
    resetPermissions() {
        this._permissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionsInput() {
        return this._permissions;
    }
    get principal() {
        return this._principal;
    }
    putPrincipal(value) {
        this._principal.internalValue = value;
    }
    resetPrincipal() {
        this._principal.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get principalInput() {
        return this._principal.internalValue;
    }
}
exports.GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference = GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference[_b] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference", version: "19.15.0" };
class GlueCatalogDatabaseCreateTableDefaultPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCatalogDatabaseCreateTableDefaultPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCatalogDatabaseCreateTableDefaultPermissionList = GlueCatalogDatabaseCreateTableDefaultPermissionList;
_c = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabaseCreateTableDefaultPermissionList[_c] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabaseCreateTableDefaultPermissionList", version: "19.15.0" };
function glueCatalogDatabaseFederatedDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        identifier: cdktf.stringToTerraform(struct.identifier),
    };
}
exports.glueCatalogDatabaseFederatedDatabaseToTerraform = glueCatalogDatabaseFederatedDatabaseToTerraform;
function glueCatalogDatabaseFederatedDatabaseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        connection_name: {
            value: cdktf.stringToHclTerraform(struct.connectionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        identifier: {
            value: cdktf.stringToHclTerraform(struct.identifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.glueCatalogDatabaseFederatedDatabaseToHclTerraform = glueCatalogDatabaseFederatedDatabaseToHclTerraform;
class GlueCatalogDatabaseFederatedDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionName = this._connectionName;
        }
        if (this._identifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.identifier = this._identifier;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connectionName = undefined;
            this._identifier = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connectionName = value.connectionName;
            this._identifier = value.identifier;
        }
    }
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    set connectionName(value) {
        this._connectionName = value;
    }
    resetConnectionName() {
        this._connectionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNameInput() {
        return this._connectionName;
    }
    get identifier() {
        return this.getStringAttribute('identifier');
    }
    set identifier(value) {
        this._identifier = value;
    }
    resetIdentifier() {
        this._identifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identifierInput() {
        return this._identifier;
    }
}
exports.GlueCatalogDatabaseFederatedDatabaseOutputReference = GlueCatalogDatabaseFederatedDatabaseOutputReference;
_d = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabaseFederatedDatabaseOutputReference[_d] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabaseFederatedDatabaseOutputReference", version: "19.15.0" };
function glueCatalogDatabaseTargetDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        region: cdktf.stringToTerraform(struct.region),
    };
}
exports.glueCatalogDatabaseTargetDatabaseToTerraform = glueCatalogDatabaseTargetDatabaseToTerraform;
function glueCatalogDatabaseTargetDatabaseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        catalog_id: {
            value: cdktf.stringToHclTerraform(struct.catalogId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        database_name: {
            value: cdktf.stringToHclTerraform(struct.databaseName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.glueCatalogDatabaseTargetDatabaseToHclTerraform = glueCatalogDatabaseTargetDatabaseToHclTerraform;
class GlueCatalogDatabaseTargetDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._catalogId !== undefined) {
            hasAnyValues = true;
            internalValueResult.catalogId = this._catalogId;
        }
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._catalogId = undefined;
            this._databaseName = undefined;
            this._region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._catalogId = value.catalogId;
            this._databaseName = value.databaseName;
            this._region = value.region;
        }
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
}
exports.GlueCatalogDatabaseTargetDatabaseOutputReference = GlueCatalogDatabaseTargetDatabaseOutputReference;
_e = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabaseTargetDatabaseOutputReference[_e] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabaseTargetDatabaseOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_catalog_database aws_glue_catalog_database}
*/
class GlueCatalogDatabase extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GlueCatalogDatabase resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GlueCatalogDatabase to import
    * @param importFromId The id of the existing GlueCatalogDatabase that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_catalog_database#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GlueCatalogDatabase to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_glue_catalog_database", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_catalog_database aws_glue_catalog_database} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueCatalogDatabaseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_catalog_database',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // create_table_default_permission - computed: false, optional: true, required: false
        this._createTableDefaultPermission = new GlueCatalogDatabaseCreateTableDefaultPermissionList(this, "create_table_default_permission", false);
        // federated_database - computed: false, optional: true, required: false
        this._federatedDatabase = new GlueCatalogDatabaseFederatedDatabaseOutputReference(this, "federated_database");
        // target_database - computed: false, optional: true, required: false
        this._targetDatabase = new GlueCatalogDatabaseTargetDatabaseOutputReference(this, "target_database");
        this._catalogId = config.catalogId;
        this._description = config.description;
        this._id = config.id;
        this._locationUri = config.locationUri;
        this._name = config.name;
        this._parameters = config.parameters;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._createTableDefaultPermission.internalValue = config.createTableDefaultPermission;
        this._federatedDatabase.internalValue = config.federatedDatabase;
        this._targetDatabase.internalValue = config.targetDatabase;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get locationUri() {
        return this.getStringAttribute('location_uri');
    }
    set locationUri(value) {
        this._locationUri = value;
    }
    resetLocationUri() {
        this._locationUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationUriInput() {
        return this._locationUri;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get createTableDefaultPermission() {
        return this._createTableDefaultPermission;
    }
    putCreateTableDefaultPermission(value) {
        this._createTableDefaultPermission.internalValue = value;
    }
    resetCreateTableDefaultPermission() {
        this._createTableDefaultPermission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createTableDefaultPermissionInput() {
        return this._createTableDefaultPermission.internalValue;
    }
    get federatedDatabase() {
        return this._federatedDatabase;
    }
    putFederatedDatabase(value) {
        this._federatedDatabase.internalValue = value;
    }
    resetFederatedDatabase() {
        this._federatedDatabase.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get federatedDatabaseInput() {
        return this._federatedDatabase.internalValue;
    }
    get targetDatabase() {
        return this._targetDatabase;
    }
    putTargetDatabase(value) {
        this._targetDatabase.internalValue = value;
    }
    resetTargetDatabase() {
        this._targetDatabase.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetDatabaseInput() {
        return this._targetDatabase.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            location_uri: cdktf.stringToTerraform(this._locationUri),
            name: cdktf.stringToTerraform(this._name),
            parameters: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameters),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            create_table_default_permission: cdktf.listMapper(glueCatalogDatabaseCreateTableDefaultPermissionToTerraform, true)(this._createTableDefaultPermission.internalValue),
            federated_database: glueCatalogDatabaseFederatedDatabaseToTerraform(this._federatedDatabase.internalValue),
            target_database: glueCatalogDatabaseTargetDatabaseToTerraform(this._targetDatabase.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            catalog_id: {
                value: cdktf.stringToHclTerraform(this._catalogId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location_uri: {
                value: cdktf.stringToHclTerraform(this._locationUri),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parameters: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._parameters),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            create_table_default_permission: {
                value: cdktf.listMapperHcl(glueCatalogDatabaseCreateTableDefaultPermissionToHclTerraform, true)(this._createTableDefaultPermission.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GlueCatalogDatabaseCreateTableDefaultPermissionList",
            },
            federated_database: {
                value: glueCatalogDatabaseFederatedDatabaseToHclTerraform(this._federatedDatabase.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GlueCatalogDatabaseFederatedDatabaseList",
            },
            target_database: {
                value: glueCatalogDatabaseTargetDatabaseToHclTerraform(this._targetDatabase.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GlueCatalogDatabaseTargetDatabaseList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GlueCatalogDatabase = GlueCatalogDatabase;
_f = JSII_RTTI_SYMBOL_1;
GlueCatalogDatabase[_f] = { fqn: "@cdktf/provider-aws.glueCatalogDatabase.GlueCatalogDatabase", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GlueCatalogDatabase.tfResourceType = "aws_glue_catalog_database";
