// @cdktf/provider-aws/lib/cognito-resource-server/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CognitoResourceServer = exports.CognitoResourceServerScopeList = exports.CognitoResourceServerScopeOutputReference = exports.cognitoResourceServerScopeToHclTerraform = exports.cognitoResourceServerScopeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cognitoResourceServerScopeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        scope_description: cdktf.stringToTerraform(struct.scopeDescription),
        scope_name: cdktf.stringToTerraform(struct.scopeName),
    };
}
exports.cognitoResourceServerScopeToTerraform = cognitoResourceServerScopeToTerraform;
function cognitoResourceServerScopeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        scope_description: {
            value: cdktf.stringToHclTerraform(struct.scopeDescription),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        scope_name: {
            value: cdktf.stringToHclTerraform(struct.scopeName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cognitoResourceServerScopeToHclTerraform = cognitoResourceServerScopeToHclTerraform;
class CognitoResourceServerScopeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._scopeDescription !== undefined) {
            hasAnyValues = true;
            internalValueResult.scopeDescription = this._scopeDescription;
        }
        if (this._scopeName !== undefined) {
            hasAnyValues = true;
            internalValueResult.scopeName = this._scopeName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._scopeDescription = undefined;
            this._scopeName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._scopeDescription = value.scopeDescription;
            this._scopeName = value.scopeName;
        }
    }
    get scopeDescription() {
        return this.getStringAttribute('scope_description');
    }
    set scopeDescription(value) {
        this._scopeDescription = value;
    }
    // Temporarily expose input value. Use with caution.
    get scopeDescriptionInput() {
        return this._scopeDescription;
    }
    get scopeName() {
        return this.getStringAttribute('scope_name');
    }
    set scopeName(value) {
        this._scopeName = value;
    }
    // Temporarily expose input value. Use with caution.
    get scopeNameInput() {
        return this._scopeName;
    }
}
exports.CognitoResourceServerScopeOutputReference = CognitoResourceServerScopeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CognitoResourceServerScopeOutputReference[_a] = { fqn: "@cdktf/provider-aws.cognitoResourceServer.CognitoResourceServerScopeOutputReference", version: "19.15.0" };
class CognitoResourceServerScopeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CognitoResourceServerScopeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CognitoResourceServerScopeList = CognitoResourceServerScopeList;
_b = JSII_RTTI_SYMBOL_1;
CognitoResourceServerScopeList[_b] = { fqn: "@cdktf/provider-aws.cognitoResourceServer.CognitoResourceServerScopeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_resource_server aws_cognito_resource_server}
*/
class CognitoResourceServer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CognitoResourceServer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CognitoResourceServer to import
    * @param importFromId The id of the existing CognitoResourceServer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_resource_server#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CognitoResourceServer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cognito_resource_server", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_resource_server aws_cognito_resource_server} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CognitoResourceServerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cognito_resource_server',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // scope - computed: false, optional: true, required: false
        this._scope = new CognitoResourceServerScopeList(this, "scope", true);
        this._id = config.id;
        this._identifier = config.identifier;
        this._name = config.name;
        this._userPoolId = config.userPoolId;
        this._scope.internalValue = config.scope;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identifier() {
        return this.getStringAttribute('identifier');
    }
    set identifier(value) {
        this._identifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get identifierInput() {
        return this._identifier;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // scope_identifiers - computed: true, optional: false, required: false
    get scopeIdentifiers() {
        return this.getListAttribute('scope_identifiers');
    }
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
    set userPoolId(value) {
        this._userPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userPoolIdInput() {
        return this._userPoolId;
    }
    get scope() {
        return this._scope;
    }
    putScope(value) {
        this._scope.internalValue = value;
    }
    resetScope() {
        this._scope.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            identifier: cdktf.stringToTerraform(this._identifier),
            name: cdktf.stringToTerraform(this._name),
            user_pool_id: cdktf.stringToTerraform(this._userPoolId),
            scope: cdktf.listMapper(cognitoResourceServerScopeToTerraform, true)(this._scope.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identifier: {
                value: cdktf.stringToHclTerraform(this._identifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_pool_id: {
                value: cdktf.stringToHclTerraform(this._userPoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scope: {
                value: cdktf.listMapperHcl(cognitoResourceServerScopeToHclTerraform, true)(this._scope.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CognitoResourceServerScopeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CognitoResourceServer = CognitoResourceServer;
_c = JSII_RTTI_SYMBOL_1;
CognitoResourceServer[_c] = { fqn: "@cdktf/provider-aws.cognitoResourceServer.CognitoResourceServer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CognitoResourceServer.tfResourceType = "aws_cognito_resource_server";
