// @cdktf/provider-aws/lib/securitylake-subscriber-notification/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecuritylakeSubscriberNotification = exports.SecuritylakeSubscriberNotificationConfigurationList = exports.SecuritylakeSubscriberNotificationConfigurationOutputReference = exports.securitylakeSubscriberNotificationConfigurationToHclTerraform = exports.securitylakeSubscriberNotificationConfigurationToTerraform = exports.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList = exports.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference = exports.securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToHclTerraform = exports.securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToTerraform = exports.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList = exports.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference = exports.securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToHclTerraform = exports.securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authorization_api_key_name: cdktf.stringToTerraform(struct.authorizationApiKeyName),
        authorization_api_key_value: cdktf.stringToTerraform(struct.authorizationApiKeyValue),
        endpoint: cdktf.stringToTerraform(struct.endpoint),
        http_method: cdktf.stringToTerraform(struct.httpMethod),
        target_role_arn: cdktf.stringToTerraform(struct.targetRoleArn),
    };
}
exports.securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToTerraform = securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToTerraform;
function securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        authorization_api_key_name: {
            value: cdktf.stringToHclTerraform(struct.authorizationApiKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        authorization_api_key_value: {
            value: cdktf.stringToHclTerraform(struct.authorizationApiKeyValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        endpoint: {
            value: cdktf.stringToHclTerraform(struct.endpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        http_method: {
            value: cdktf.stringToHclTerraform(struct.httpMethod),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_role_arn: {
            value: cdktf.stringToHclTerraform(struct.targetRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToHclTerraform = securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToHclTerraform;
class SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authorizationApiKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.authorizationApiKeyName = this._authorizationApiKeyName;
        }
        if (this._authorizationApiKeyValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.authorizationApiKeyValue = this._authorizationApiKeyValue;
        }
        if (this._endpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpoint = this._endpoint;
        }
        if (this._httpMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpMethod = this._httpMethod;
        }
        if (this._targetRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRoleArn = this._targetRoleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._authorizationApiKeyName = undefined;
            this._authorizationApiKeyValue = undefined;
            this._endpoint = undefined;
            this._httpMethod = undefined;
            this._targetRoleArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._authorizationApiKeyName = value.authorizationApiKeyName;
            this._authorizationApiKeyValue = value.authorizationApiKeyValue;
            this._endpoint = value.endpoint;
            this._httpMethod = value.httpMethod;
            this._targetRoleArn = value.targetRoleArn;
        }
    }
    get authorizationApiKeyName() {
        return this.getStringAttribute('authorization_api_key_name');
    }
    set authorizationApiKeyName(value) {
        this._authorizationApiKeyName = value;
    }
    resetAuthorizationApiKeyName() {
        this._authorizationApiKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationApiKeyNameInput() {
        return this._authorizationApiKeyName;
    }
    get authorizationApiKeyValue() {
        return this.getStringAttribute('authorization_api_key_value');
    }
    set authorizationApiKeyValue(value) {
        this._authorizationApiKeyValue = value;
    }
    resetAuthorizationApiKeyValue() {
        this._authorizationApiKeyValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationApiKeyValueInput() {
        return this._authorizationApiKeyValue;
    }
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    set endpoint(value) {
        this._endpoint = value;
    }
    resetEndpoint() {
        this._endpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointInput() {
        return this._endpoint;
    }
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    set httpMethod(value) {
        this._httpMethod = value;
    }
    resetHttpMethod() {
        this._httpMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpMethodInput() {
        return this._httpMethod;
    }
    get targetRoleArn() {
        return this.getStringAttribute('target_role_arn');
    }
    set targetRoleArn(value) {
        this._targetRoleArn = value;
    }
    resetTargetRoleArn() {
        this._targetRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetRoleArnInput() {
        return this._targetRoleArn;
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference = SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList = SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList[_b] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList", version: "19.15.0" };
function securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToTerraform = securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToTerraform;
function securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToHclTerraform = securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToHclTerraform;
class SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference = SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList = SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList[_d] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList", version: "19.15.0" };
function securitylakeSubscriberNotificationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        https_notification_configuration: cdktf.listMapper(securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToTerraform, true)(struct.httpsNotificationConfiguration),
        sqs_notification_configuration: cdktf.listMapper(securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToTerraform, true)(struct.sqsNotificationConfiguration),
    };
}
exports.securitylakeSubscriberNotificationConfigurationToTerraform = securitylakeSubscriberNotificationConfigurationToTerraform;
function securitylakeSubscriberNotificationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        https_notification_configuration: {
            value: cdktf.listMapperHcl(securitylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationToHclTerraform, true)(struct.httpsNotificationConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList",
        },
        sqs_notification_configuration: {
            value: cdktf.listMapperHcl(securitylakeSubscriberNotificationConfigurationSqsNotificationConfigurationToHclTerraform, true)(struct.sqsNotificationConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeSubscriberNotificationConfigurationToHclTerraform = securitylakeSubscriberNotificationConfigurationToHclTerraform;
class SecuritylakeSubscriberNotificationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // https_notification_configuration - computed: false, optional: true, required: false
        this._httpsNotificationConfiguration = new SecuritylakeSubscriberNotificationConfigurationHttpsNotificationConfigurationList(this, "https_notification_configuration", false);
        // sqs_notification_configuration - computed: false, optional: true, required: false
        this._sqsNotificationConfiguration = new SecuritylakeSubscriberNotificationConfigurationSqsNotificationConfigurationList(this, "sqs_notification_configuration", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpsNotificationConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpsNotificationConfiguration = this._httpsNotificationConfiguration?.internalValue;
        }
        if (this._sqsNotificationConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqsNotificationConfiguration = this._sqsNotificationConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._httpsNotificationConfiguration.internalValue = undefined;
            this._sqsNotificationConfiguration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._httpsNotificationConfiguration.internalValue = value.httpsNotificationConfiguration;
            this._sqsNotificationConfiguration.internalValue = value.sqsNotificationConfiguration;
        }
    }
    get httpsNotificationConfiguration() {
        return this._httpsNotificationConfiguration;
    }
    putHttpsNotificationConfiguration(value) {
        this._httpsNotificationConfiguration.internalValue = value;
    }
    resetHttpsNotificationConfiguration() {
        this._httpsNotificationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpsNotificationConfigurationInput() {
        return this._httpsNotificationConfiguration.internalValue;
    }
    get sqsNotificationConfiguration() {
        return this._sqsNotificationConfiguration;
    }
    putSqsNotificationConfiguration(value) {
        this._sqsNotificationConfiguration.internalValue = value;
    }
    resetSqsNotificationConfiguration() {
        this._sqsNotificationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqsNotificationConfigurationInput() {
        return this._sqsNotificationConfiguration.internalValue;
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationOutputReference = SecuritylakeSubscriberNotificationConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeSubscriberNotificationConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeSubscriberNotificationConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeSubscriberNotificationConfigurationList = SecuritylakeSubscriberNotificationConfigurationList;
_f = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotificationConfigurationList[_f] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotificationConfigurationList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_subscriber_notification aws_securitylake_subscriber_notification}
*/
class SecuritylakeSubscriberNotification extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecuritylakeSubscriberNotification resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecuritylakeSubscriberNotification to import
    * @param importFromId The id of the existing SecuritylakeSubscriberNotification that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_subscriber_notification#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecuritylakeSubscriberNotification to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securitylake_subscriber_notification", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_subscriber_notification aws_securitylake_subscriber_notification} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecuritylakeSubscriberNotificationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securitylake_subscriber_notification',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // configuration - computed: false, optional: true, required: false
        this._configuration = new SecuritylakeSubscriberNotificationConfigurationList(this, "configuration", false);
        this._subscriberId = config.subscriberId;
        this._configuration.internalValue = config.configuration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // endpoint_id - computed: true, optional: false, required: false
    get endpointId() {
        return this.getStringAttribute('endpoint_id');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get subscriberId() {
        return this.getStringAttribute('subscriber_id');
    }
    set subscriberId(value) {
        this._subscriberId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subscriberIdInput() {
        return this._subscriberId;
    }
    get configuration() {
        return this._configuration;
    }
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            subscriber_id: cdktf.stringToTerraform(this._subscriberId),
            configuration: cdktf.listMapper(securitylakeSubscriberNotificationConfigurationToTerraform, true)(this._configuration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            subscriber_id: {
                value: cdktf.stringToHclTerraform(this._subscriberId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            configuration: {
                value: cdktf.listMapperHcl(securitylakeSubscriberNotificationConfigurationToHclTerraform, true)(this._configuration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SecuritylakeSubscriberNotificationConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecuritylakeSubscriberNotification = SecuritylakeSubscriberNotification;
_g = JSII_RTTI_SYMBOL_1;
SecuritylakeSubscriberNotification[_g] = { fqn: "@cdktf/provider-aws.securitylakeSubscriberNotification.SecuritylakeSubscriberNotification", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecuritylakeSubscriberNotification.tfResourceType = "aws_securitylake_subscriber_notification";
