// @cdktf/provider-aws/lib/fsx-ontap-volume/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxOntapVolume = exports.FsxOntapVolumeTimeoutsOutputReference = exports.fsxOntapVolumeTimeoutsToHclTerraform = exports.fsxOntapVolumeTimeoutsToTerraform = exports.FsxOntapVolumeTieringPolicyOutputReference = exports.fsxOntapVolumeTieringPolicyToHclTerraform = exports.fsxOntapVolumeTieringPolicyToTerraform = exports.FsxOntapVolumeSnaplockConfigurationOutputReference = exports.fsxOntapVolumeSnaplockConfigurationToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationToTerraform = exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodToTerraform = exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToTerraform = exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToTerraform = exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToTerraform = exports.FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference = exports.fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToHclTerraform = exports.fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToTerraform = fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToTerraform;
function fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToHclTerraform = fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference = FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference", version: "19.15.0" };
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToTerraform;
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToHclTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference = FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference[_b] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference", version: "19.15.0" };
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToTerraform;
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToHclTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference = FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference[_c] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference", version: "19.15.0" };
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToTerraform;
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToHclTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference = FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference[_d] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference", version: "19.15.0" };
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_retention: fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToTerraform(struct.defaultRetention),
        maximum_retention: fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToTerraform(struct.maximumRetention),
        minimum_retention: fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToTerraform(struct.minimumRetention),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodToTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodToTerraform;
function fsxOntapVolumeSnaplockConfigurationRetentionPeriodToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_retention: {
            value: fsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionToHclTerraform(struct.defaultRetention),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionList",
        },
        maximum_retention: {
            value: fsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionToHclTerraform(struct.maximumRetention),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionList",
        },
        minimum_retention: {
            value: fsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionToHclTerraform(struct.minimumRetention),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationRetentionPeriodToHclTerraform = fsxOntapVolumeSnaplockConfigurationRetentionPeriodToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // default_retention - computed: false, optional: true, required: false
        this._defaultRetention = new FsxOntapVolumeSnaplockConfigurationRetentionPeriodDefaultRetentionOutputReference(this, "default_retention");
        // maximum_retention - computed: false, optional: true, required: false
        this._maximumRetention = new FsxOntapVolumeSnaplockConfigurationRetentionPeriodMaximumRetentionOutputReference(this, "maximum_retention");
        // minimum_retention - computed: false, optional: true, required: false
        this._minimumRetention = new FsxOntapVolumeSnaplockConfigurationRetentionPeriodMinimumRetentionOutputReference(this, "minimum_retention");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultRetention?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultRetention = this._defaultRetention?.internalValue;
        }
        if (this._maximumRetention?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumRetention = this._maximumRetention?.internalValue;
        }
        if (this._minimumRetention?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimumRetention = this._minimumRetention?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultRetention.internalValue = undefined;
            this._maximumRetention.internalValue = undefined;
            this._minimumRetention.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultRetention.internalValue = value.defaultRetention;
            this._maximumRetention.internalValue = value.maximumRetention;
            this._minimumRetention.internalValue = value.minimumRetention;
        }
    }
    get defaultRetention() {
        return this._defaultRetention;
    }
    putDefaultRetention(value) {
        this._defaultRetention.internalValue = value;
    }
    resetDefaultRetention() {
        this._defaultRetention.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultRetentionInput() {
        return this._defaultRetention.internalValue;
    }
    get maximumRetention() {
        return this._maximumRetention;
    }
    putMaximumRetention(value) {
        this._maximumRetention.internalValue = value;
    }
    resetMaximumRetention() {
        this._maximumRetention.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumRetentionInput() {
        return this._maximumRetention.internalValue;
    }
    get minimumRetention() {
        return this._minimumRetention;
    }
    putMinimumRetention(value) {
        this._minimumRetention.internalValue = value;
    }
    resetMinimumRetention() {
        this._minimumRetention.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minimumRetentionInput() {
        return this._minimumRetention.internalValue;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference = FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference;
_e = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference[_e] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference", version: "19.15.0" };
function fsxOntapVolumeSnaplockConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit_log_volume: cdktf.booleanToTerraform(struct.auditLogVolume),
        privileged_delete: cdktf.stringToTerraform(struct.privilegedDelete),
        snaplock_type: cdktf.stringToTerraform(struct.snaplockType),
        volume_append_mode_enabled: cdktf.booleanToTerraform(struct.volumeAppendModeEnabled),
        autocommit_period: fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToTerraform(struct.autocommitPeriod),
        retention_period: fsxOntapVolumeSnaplockConfigurationRetentionPeriodToTerraform(struct.retentionPeriod),
    };
}
exports.fsxOntapVolumeSnaplockConfigurationToTerraform = fsxOntapVolumeSnaplockConfigurationToTerraform;
function fsxOntapVolumeSnaplockConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audit_log_volume: {
            value: cdktf.booleanToHclTerraform(struct.auditLogVolume),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        privileged_delete: {
            value: cdktf.stringToHclTerraform(struct.privilegedDelete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        snaplock_type: {
            value: cdktf.stringToHclTerraform(struct.snaplockType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        volume_append_mode_enabled: {
            value: cdktf.booleanToHclTerraform(struct.volumeAppendModeEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        autocommit_period: {
            value: fsxOntapVolumeSnaplockConfigurationAutocommitPeriodToHclTerraform(struct.autocommitPeriod),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOntapVolumeSnaplockConfigurationAutocommitPeriodList",
        },
        retention_period: {
            value: fsxOntapVolumeSnaplockConfigurationRetentionPeriodToHclTerraform(struct.retentionPeriod),
            isBlock: true,
            type: "list",
            storageClassType: "FsxOntapVolumeSnaplockConfigurationRetentionPeriodList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeSnaplockConfigurationToHclTerraform = fsxOntapVolumeSnaplockConfigurationToHclTerraform;
class FsxOntapVolumeSnaplockConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // autocommit_period - computed: false, optional: true, required: false
        this._autocommitPeriod = new FsxOntapVolumeSnaplockConfigurationAutocommitPeriodOutputReference(this, "autocommit_period");
        // retention_period - computed: false, optional: true, required: false
        this._retentionPeriod = new FsxOntapVolumeSnaplockConfigurationRetentionPeriodOutputReference(this, "retention_period");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._auditLogVolume !== undefined) {
            hasAnyValues = true;
            internalValueResult.auditLogVolume = this._auditLogVolume;
        }
        if (this._privilegedDelete !== undefined) {
            hasAnyValues = true;
            internalValueResult.privilegedDelete = this._privilegedDelete;
        }
        if (this._snaplockType !== undefined) {
            hasAnyValues = true;
            internalValueResult.snaplockType = this._snaplockType;
        }
        if (this._volumeAppendModeEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeAppendModeEnabled = this._volumeAppendModeEnabled;
        }
        if (this._autocommitPeriod?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autocommitPeriod = this._autocommitPeriod?.internalValue;
        }
        if (this._retentionPeriod?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retentionPeriod = this._retentionPeriod?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._auditLogVolume = undefined;
            this._privilegedDelete = undefined;
            this._snaplockType = undefined;
            this._volumeAppendModeEnabled = undefined;
            this._autocommitPeriod.internalValue = undefined;
            this._retentionPeriod.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._auditLogVolume = value.auditLogVolume;
            this._privilegedDelete = value.privilegedDelete;
            this._snaplockType = value.snaplockType;
            this._volumeAppendModeEnabled = value.volumeAppendModeEnabled;
            this._autocommitPeriod.internalValue = value.autocommitPeriod;
            this._retentionPeriod.internalValue = value.retentionPeriod;
        }
    }
    get auditLogVolume() {
        return this.getBooleanAttribute('audit_log_volume');
    }
    set auditLogVolume(value) {
        this._auditLogVolume = value;
    }
    resetAuditLogVolume() {
        this._auditLogVolume = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditLogVolumeInput() {
        return this._auditLogVolume;
    }
    get privilegedDelete() {
        return this.getStringAttribute('privileged_delete');
    }
    set privilegedDelete(value) {
        this._privilegedDelete = value;
    }
    resetPrivilegedDelete() {
        this._privilegedDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privilegedDeleteInput() {
        return this._privilegedDelete;
    }
    get snaplockType() {
        return this.getStringAttribute('snaplock_type');
    }
    set snaplockType(value) {
        this._snaplockType = value;
    }
    // Temporarily expose input value. Use with caution.
    get snaplockTypeInput() {
        return this._snaplockType;
    }
    get volumeAppendModeEnabled() {
        return this.getBooleanAttribute('volume_append_mode_enabled');
    }
    set volumeAppendModeEnabled(value) {
        this._volumeAppendModeEnabled = value;
    }
    resetVolumeAppendModeEnabled() {
        this._volumeAppendModeEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeAppendModeEnabledInput() {
        return this._volumeAppendModeEnabled;
    }
    get autocommitPeriod() {
        return this._autocommitPeriod;
    }
    putAutocommitPeriod(value) {
        this._autocommitPeriod.internalValue = value;
    }
    resetAutocommitPeriod() {
        this._autocommitPeriod.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autocommitPeriodInput() {
        return this._autocommitPeriod.internalValue;
    }
    get retentionPeriod() {
        return this._retentionPeriod;
    }
    putRetentionPeriod(value) {
        this._retentionPeriod.internalValue = value;
    }
    resetRetentionPeriod() {
        this._retentionPeriod.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod.internalValue;
    }
}
exports.FsxOntapVolumeSnaplockConfigurationOutputReference = FsxOntapVolumeSnaplockConfigurationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeSnaplockConfigurationOutputReference[_f] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeSnaplockConfigurationOutputReference", version: "19.15.0" };
function fsxOntapVolumeTieringPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cooling_period: cdktf.numberToTerraform(struct.coolingPeriod),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.fsxOntapVolumeTieringPolicyToTerraform = fsxOntapVolumeTieringPolicyToTerraform;
function fsxOntapVolumeTieringPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cooling_period: {
            value: cdktf.numberToHclTerraform(struct.coolingPeriod),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeTieringPolicyToHclTerraform = fsxOntapVolumeTieringPolicyToHclTerraform;
class FsxOntapVolumeTieringPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coolingPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.coolingPeriod = this._coolingPeriod;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coolingPeriod = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coolingPeriod = value.coolingPeriod;
            this._name = value.name;
        }
    }
    get coolingPeriod() {
        return this.getNumberAttribute('cooling_period');
    }
    set coolingPeriod(value) {
        this._coolingPeriod = value;
    }
    resetCoolingPeriod() {
        this._coolingPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coolingPeriodInput() {
        return this._coolingPeriod;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.FsxOntapVolumeTieringPolicyOutputReference = FsxOntapVolumeTieringPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeTieringPolicyOutputReference[_g] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeTieringPolicyOutputReference", version: "19.15.0" };
function fsxOntapVolumeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxOntapVolumeTimeoutsToTerraform = fsxOntapVolumeTimeoutsToTerraform;
function fsxOntapVolumeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxOntapVolumeTimeoutsToHclTerraform = fsxOntapVolumeTimeoutsToHclTerraform;
class FsxOntapVolumeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FsxOntapVolumeTimeoutsOutputReference = FsxOntapVolumeTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_ontap_volume aws_fsx_ontap_volume}
*/
class FsxOntapVolume extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FsxOntapVolume resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FsxOntapVolume to import
    * @param importFromId The id of the existing FsxOntapVolume that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_ontap_volume#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FsxOntapVolume to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_ontap_volume", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_ontap_volume aws_fsx_ontap_volume} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FsxOntapVolumeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_ontap_volume',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // snaplock_configuration - computed: false, optional: true, required: false
        this._snaplockConfiguration = new FsxOntapVolumeSnaplockConfigurationOutputReference(this, "snaplock_configuration");
        // tiering_policy - computed: false, optional: true, required: false
        this._tieringPolicy = new FsxOntapVolumeTieringPolicyOutputReference(this, "tiering_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxOntapVolumeTimeoutsOutputReference(this, "timeouts");
        this._bypassSnaplockEnterpriseRetention = config.bypassSnaplockEnterpriseRetention;
        this._copyTagsToBackups = config.copyTagsToBackups;
        this._id = config.id;
        this._junctionPath = config.junctionPath;
        this._name = config.name;
        this._ontapVolumeType = config.ontapVolumeType;
        this._securityStyle = config.securityStyle;
        this._sizeInMegabytes = config.sizeInMegabytes;
        this._skipFinalBackup = config.skipFinalBackup;
        this._snapshotPolicy = config.snapshotPolicy;
        this._storageEfficiencyEnabled = config.storageEfficiencyEnabled;
        this._storageVirtualMachineId = config.storageVirtualMachineId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._volumeType = config.volumeType;
        this._snaplockConfiguration.internalValue = config.snaplockConfiguration;
        this._tieringPolicy.internalValue = config.tieringPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get bypassSnaplockEnterpriseRetention() {
        return this.getBooleanAttribute('bypass_snaplock_enterprise_retention');
    }
    set bypassSnaplockEnterpriseRetention(value) {
        this._bypassSnaplockEnterpriseRetention = value;
    }
    resetBypassSnaplockEnterpriseRetention() {
        this._bypassSnaplockEnterpriseRetention = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bypassSnaplockEnterpriseRetentionInput() {
        return this._bypassSnaplockEnterpriseRetention;
    }
    get copyTagsToBackups() {
        return this.getBooleanAttribute('copy_tags_to_backups');
    }
    set copyTagsToBackups(value) {
        this._copyTagsToBackups = value;
    }
    resetCopyTagsToBackups() {
        this._copyTagsToBackups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyTagsToBackupsInput() {
        return this._copyTagsToBackups;
    }
    // file_system_id - computed: true, optional: false, required: false
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    // flexcache_endpoint_type - computed: true, optional: false, required: false
    get flexcacheEndpointType() {
        return this.getStringAttribute('flexcache_endpoint_type');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get junctionPath() {
        return this.getStringAttribute('junction_path');
    }
    set junctionPath(value) {
        this._junctionPath = value;
    }
    resetJunctionPath() {
        this._junctionPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get junctionPathInput() {
        return this._junctionPath;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ontapVolumeType() {
        return this.getStringAttribute('ontap_volume_type');
    }
    set ontapVolumeType(value) {
        this._ontapVolumeType = value;
    }
    resetOntapVolumeType() {
        this._ontapVolumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ontapVolumeTypeInput() {
        return this._ontapVolumeType;
    }
    get securityStyle() {
        return this.getStringAttribute('security_style');
    }
    set securityStyle(value) {
        this._securityStyle = value;
    }
    resetSecurityStyle() {
        this._securityStyle = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityStyleInput() {
        return this._securityStyle;
    }
    get sizeInMegabytes() {
        return this.getNumberAttribute('size_in_megabytes');
    }
    set sizeInMegabytes(value) {
        this._sizeInMegabytes = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInMegabytesInput() {
        return this._sizeInMegabytes;
    }
    get skipFinalBackup() {
        return this.getBooleanAttribute('skip_final_backup');
    }
    set skipFinalBackup(value) {
        this._skipFinalBackup = value;
    }
    resetSkipFinalBackup() {
        this._skipFinalBackup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipFinalBackupInput() {
        return this._skipFinalBackup;
    }
    get snapshotPolicy() {
        return this.getStringAttribute('snapshot_policy');
    }
    set snapshotPolicy(value) {
        this._snapshotPolicy = value;
    }
    resetSnapshotPolicy() {
        this._snapshotPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotPolicyInput() {
        return this._snapshotPolicy;
    }
    get storageEfficiencyEnabled() {
        return this.getBooleanAttribute('storage_efficiency_enabled');
    }
    set storageEfficiencyEnabled(value) {
        this._storageEfficiencyEnabled = value;
    }
    resetStorageEfficiencyEnabled() {
        this._storageEfficiencyEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageEfficiencyEnabledInput() {
        return this._storageEfficiencyEnabled;
    }
    get storageVirtualMachineId() {
        return this.getStringAttribute('storage_virtual_machine_id');
    }
    set storageVirtualMachineId(value) {
        this._storageVirtualMachineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageVirtualMachineIdInput() {
        return this._storageVirtualMachineId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uuid - computed: true, optional: false, required: false
    get uuid() {
        return this.getStringAttribute('uuid');
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
    get snaplockConfiguration() {
        return this._snaplockConfiguration;
    }
    putSnaplockConfiguration(value) {
        this._snaplockConfiguration.internalValue = value;
    }
    resetSnaplockConfiguration() {
        this._snaplockConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snaplockConfigurationInput() {
        return this._snaplockConfiguration.internalValue;
    }
    get tieringPolicy() {
        return this._tieringPolicy;
    }
    putTieringPolicy(value) {
        this._tieringPolicy.internalValue = value;
    }
    resetTieringPolicy() {
        this._tieringPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tieringPolicyInput() {
        return this._tieringPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bypass_snaplock_enterprise_retention: cdktf.booleanToTerraform(this._bypassSnaplockEnterpriseRetention),
            copy_tags_to_backups: cdktf.booleanToTerraform(this._copyTagsToBackups),
            id: cdktf.stringToTerraform(this._id),
            junction_path: cdktf.stringToTerraform(this._junctionPath),
            name: cdktf.stringToTerraform(this._name),
            ontap_volume_type: cdktf.stringToTerraform(this._ontapVolumeType),
            security_style: cdktf.stringToTerraform(this._securityStyle),
            size_in_megabytes: cdktf.numberToTerraform(this._sizeInMegabytes),
            skip_final_backup: cdktf.booleanToTerraform(this._skipFinalBackup),
            snapshot_policy: cdktf.stringToTerraform(this._snapshotPolicy),
            storage_efficiency_enabled: cdktf.booleanToTerraform(this._storageEfficiencyEnabled),
            storage_virtual_machine_id: cdktf.stringToTerraform(this._storageVirtualMachineId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            volume_type: cdktf.stringToTerraform(this._volumeType),
            snaplock_configuration: fsxOntapVolumeSnaplockConfigurationToTerraform(this._snaplockConfiguration.internalValue),
            tiering_policy: fsxOntapVolumeTieringPolicyToTerraform(this._tieringPolicy.internalValue),
            timeouts: fsxOntapVolumeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bypass_snaplock_enterprise_retention: {
                value: cdktf.booleanToHclTerraform(this._bypassSnaplockEnterpriseRetention),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            copy_tags_to_backups: {
                value: cdktf.booleanToHclTerraform(this._copyTagsToBackups),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            junction_path: {
                value: cdktf.stringToHclTerraform(this._junctionPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ontap_volume_type: {
                value: cdktf.stringToHclTerraform(this._ontapVolumeType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            security_style: {
                value: cdktf.stringToHclTerraform(this._securityStyle),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            size_in_megabytes: {
                value: cdktf.numberToHclTerraform(this._sizeInMegabytes),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            skip_final_backup: {
                value: cdktf.booleanToHclTerraform(this._skipFinalBackup),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            snapshot_policy: {
                value: cdktf.stringToHclTerraform(this._snapshotPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            storage_efficiency_enabled: {
                value: cdktf.booleanToHclTerraform(this._storageEfficiencyEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            storage_virtual_machine_id: {
                value: cdktf.stringToHclTerraform(this._storageVirtualMachineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            volume_type: {
                value: cdktf.stringToHclTerraform(this._volumeType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            snaplock_configuration: {
                value: fsxOntapVolumeSnaplockConfigurationToHclTerraform(this._snaplockConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FsxOntapVolumeSnaplockConfigurationList",
            },
            tiering_policy: {
                value: fsxOntapVolumeTieringPolicyToHclTerraform(this._tieringPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FsxOntapVolumeTieringPolicyList",
            },
            timeouts: {
                value: fsxOntapVolumeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FsxOntapVolumeTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FsxOntapVolume = FsxOntapVolume;
_j = JSII_RTTI_SYMBOL_1;
FsxOntapVolume[_j] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolume", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FsxOntapVolume.tfResourceType = "aws_fsx_ontap_volume";
