// @cdktf/provider-aws/lib/cloudformation-stack-set/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudformationStackSet = exports.CloudformationStackSetTimeoutsOutputReference = exports.cloudformationStackSetTimeoutsToHclTerraform = exports.cloudformationStackSetTimeoutsToTerraform = exports.CloudformationStackSetOperationPreferencesOutputReference = exports.cloudformationStackSetOperationPreferencesToHclTerraform = exports.cloudformationStackSetOperationPreferencesToTerraform = exports.CloudformationStackSetManagedExecutionOutputReference = exports.cloudformationStackSetManagedExecutionToHclTerraform = exports.cloudformationStackSetManagedExecutionToTerraform = exports.CloudformationStackSetAutoDeploymentOutputReference = exports.cloudformationStackSetAutoDeploymentToHclTerraform = exports.cloudformationStackSetAutoDeploymentToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudformationStackSetAutoDeploymentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        retain_stacks_on_account_removal: cdktf.booleanToTerraform(struct.retainStacksOnAccountRemoval),
    };
}
exports.cloudformationStackSetAutoDeploymentToTerraform = cloudformationStackSetAutoDeploymentToTerraform;
function cloudformationStackSetAutoDeploymentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        retain_stacks_on_account_removal: {
            value: cdktf.booleanToHclTerraform(struct.retainStacksOnAccountRemoval),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudformationStackSetAutoDeploymentToHclTerraform = cloudformationStackSetAutoDeploymentToHclTerraform;
class CloudformationStackSetAutoDeploymentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._retainStacksOnAccountRemoval !== undefined) {
            hasAnyValues = true;
            internalValueResult.retainStacksOnAccountRemoval = this._retainStacksOnAccountRemoval;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._retainStacksOnAccountRemoval = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._retainStacksOnAccountRemoval = value.retainStacksOnAccountRemoval;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get retainStacksOnAccountRemoval() {
        return this.getBooleanAttribute('retain_stacks_on_account_removal');
    }
    set retainStacksOnAccountRemoval(value) {
        this._retainStacksOnAccountRemoval = value;
    }
    resetRetainStacksOnAccountRemoval() {
        this._retainStacksOnAccountRemoval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retainStacksOnAccountRemovalInput() {
        return this._retainStacksOnAccountRemoval;
    }
}
exports.CloudformationStackSetAutoDeploymentOutputReference = CloudformationStackSetAutoDeploymentOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudformationStackSetAutoDeploymentOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudformationStackSet.CloudformationStackSetAutoDeploymentOutputReference", version: "19.15.0" };
function cloudformationStackSetManagedExecutionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        active: cdktf.booleanToTerraform(struct.active),
    };
}
exports.cloudformationStackSetManagedExecutionToTerraform = cloudformationStackSetManagedExecutionToTerraform;
function cloudformationStackSetManagedExecutionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        active: {
            value: cdktf.booleanToHclTerraform(struct.active),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudformationStackSetManagedExecutionToHclTerraform = cloudformationStackSetManagedExecutionToHclTerraform;
class CloudformationStackSetManagedExecutionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._active !== undefined) {
            hasAnyValues = true;
            internalValueResult.active = this._active;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._active = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._active = value.active;
        }
    }
    get active() {
        return this.getBooleanAttribute('active');
    }
    set active(value) {
        this._active = value;
    }
    resetActive() {
        this._active = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get activeInput() {
        return this._active;
    }
}
exports.CloudformationStackSetManagedExecutionOutputReference = CloudformationStackSetManagedExecutionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudformationStackSetManagedExecutionOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudformationStackSet.CloudformationStackSetManagedExecutionOutputReference", version: "19.15.0" };
function cloudformationStackSetOperationPreferencesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_tolerance_count: cdktf.numberToTerraform(struct.failureToleranceCount),
        failure_tolerance_percentage: cdktf.numberToTerraform(struct.failureTolerancePercentage),
        max_concurrent_count: cdktf.numberToTerraform(struct.maxConcurrentCount),
        max_concurrent_percentage: cdktf.numberToTerraform(struct.maxConcurrentPercentage),
        region_concurrency_type: cdktf.stringToTerraform(struct.regionConcurrencyType),
        region_order: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.regionOrder),
    };
}
exports.cloudformationStackSetOperationPreferencesToTerraform = cloudformationStackSetOperationPreferencesToTerraform;
function cloudformationStackSetOperationPreferencesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        failure_tolerance_count: {
            value: cdktf.numberToHclTerraform(struct.failureToleranceCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        failure_tolerance_percentage: {
            value: cdktf.numberToHclTerraform(struct.failureTolerancePercentage),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_concurrent_count: {
            value: cdktf.numberToHclTerraform(struct.maxConcurrentCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_concurrent_percentage: {
            value: cdktf.numberToHclTerraform(struct.maxConcurrentPercentage),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        region_concurrency_type: {
            value: cdktf.stringToHclTerraform(struct.regionConcurrencyType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        region_order: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.regionOrder),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudformationStackSetOperationPreferencesToHclTerraform = cloudformationStackSetOperationPreferencesToHclTerraform;
class CloudformationStackSetOperationPreferencesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureToleranceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureToleranceCount = this._failureToleranceCount;
        }
        if (this._failureTolerancePercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureTolerancePercentage = this._failureTolerancePercentage;
        }
        if (this._maxConcurrentCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentCount = this._maxConcurrentCount;
        }
        if (this._maxConcurrentPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentPercentage = this._maxConcurrentPercentage;
        }
        if (this._regionConcurrencyType !== undefined) {
            hasAnyValues = true;
            internalValueResult.regionConcurrencyType = this._regionConcurrencyType;
        }
        if (this._regionOrder !== undefined) {
            hasAnyValues = true;
            internalValueResult.regionOrder = this._regionOrder;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureToleranceCount = undefined;
            this._failureTolerancePercentage = undefined;
            this._maxConcurrentCount = undefined;
            this._maxConcurrentPercentage = undefined;
            this._regionConcurrencyType = undefined;
            this._regionOrder = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureToleranceCount = value.failureToleranceCount;
            this._failureTolerancePercentage = value.failureTolerancePercentage;
            this._maxConcurrentCount = value.maxConcurrentCount;
            this._maxConcurrentPercentage = value.maxConcurrentPercentage;
            this._regionConcurrencyType = value.regionConcurrencyType;
            this._regionOrder = value.regionOrder;
        }
    }
    get failureToleranceCount() {
        return this.getNumberAttribute('failure_tolerance_count');
    }
    set failureToleranceCount(value) {
        this._failureToleranceCount = value;
    }
    resetFailureToleranceCount() {
        this._failureToleranceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureToleranceCountInput() {
        return this._failureToleranceCount;
    }
    get failureTolerancePercentage() {
        return this.getNumberAttribute('failure_tolerance_percentage');
    }
    set failureTolerancePercentage(value) {
        this._failureTolerancePercentage = value;
    }
    resetFailureTolerancePercentage() {
        this._failureTolerancePercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureTolerancePercentageInput() {
        return this._failureTolerancePercentage;
    }
    get maxConcurrentCount() {
        return this.getNumberAttribute('max_concurrent_count');
    }
    set maxConcurrentCount(value) {
        this._maxConcurrentCount = value;
    }
    resetMaxConcurrentCount() {
        this._maxConcurrentCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentCountInput() {
        return this._maxConcurrentCount;
    }
    get maxConcurrentPercentage() {
        return this.getNumberAttribute('max_concurrent_percentage');
    }
    set maxConcurrentPercentage(value) {
        this._maxConcurrentPercentage = value;
    }
    resetMaxConcurrentPercentage() {
        this._maxConcurrentPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentPercentageInput() {
        return this._maxConcurrentPercentage;
    }
    get regionConcurrencyType() {
        return this.getStringAttribute('region_concurrency_type');
    }
    set regionConcurrencyType(value) {
        this._regionConcurrencyType = value;
    }
    resetRegionConcurrencyType() {
        this._regionConcurrencyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionConcurrencyTypeInput() {
        return this._regionConcurrencyType;
    }
    get regionOrder() {
        return this.getListAttribute('region_order');
    }
    set regionOrder(value) {
        this._regionOrder = value;
    }
    resetRegionOrder() {
        this._regionOrder = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionOrderInput() {
        return this._regionOrder;
    }
}
exports.CloudformationStackSetOperationPreferencesOutputReference = CloudformationStackSetOperationPreferencesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudformationStackSetOperationPreferencesOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudformationStackSet.CloudformationStackSetOperationPreferencesOutputReference", version: "19.15.0" };
function cloudformationStackSetTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudformationStackSetTimeoutsToTerraform = cloudformationStackSetTimeoutsToTerraform;
function cloudformationStackSetTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudformationStackSetTimeoutsToHclTerraform = cloudformationStackSetTimeoutsToHclTerraform;
class CloudformationStackSetTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._update = value.update;
        }
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudformationStackSetTimeoutsOutputReference = CloudformationStackSetTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudformationStackSetTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.cloudformationStackSet.CloudformationStackSetTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudformation_stack_set aws_cloudformation_stack_set}
*/
class CloudformationStackSet extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudformationStackSet resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudformationStackSet to import
    * @param importFromId The id of the existing CloudformationStackSet that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudformation_stack_set#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudformationStackSet to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudformation_stack_set", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudformation_stack_set aws_cloudformation_stack_set} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudformationStackSetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudformation_stack_set',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auto_deployment - computed: false, optional: true, required: false
        this._autoDeployment = new CloudformationStackSetAutoDeploymentOutputReference(this, "auto_deployment");
        // managed_execution - computed: false, optional: true, required: false
        this._managedExecution = new CloudformationStackSetManagedExecutionOutputReference(this, "managed_execution");
        // operation_preferences - computed: false, optional: true, required: false
        this._operationPreferences = new CloudformationStackSetOperationPreferencesOutputReference(this, "operation_preferences");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudformationStackSetTimeoutsOutputReference(this, "timeouts");
        this._administrationRoleArn = config.administrationRoleArn;
        this._callAs = config.callAs;
        this._capabilities = config.capabilities;
        this._description = config.description;
        this._executionRoleName = config.executionRoleName;
        this._id = config.id;
        this._name = config.name;
        this._parameters = config.parameters;
        this._permissionModel = config.permissionModel;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._templateBody = config.templateBody;
        this._templateUrl = config.templateUrl;
        this._autoDeployment.internalValue = config.autoDeployment;
        this._managedExecution.internalValue = config.managedExecution;
        this._operationPreferences.internalValue = config.operationPreferences;
        this._timeouts.internalValue = config.timeouts;
    }
    get administrationRoleArn() {
        return this.getStringAttribute('administration_role_arn');
    }
    set administrationRoleArn(value) {
        this._administrationRoleArn = value;
    }
    resetAdministrationRoleArn() {
        this._administrationRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get administrationRoleArnInput() {
        return this._administrationRoleArn;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get callAs() {
        return this.getStringAttribute('call_as');
    }
    set callAs(value) {
        this._callAs = value;
    }
    resetCallAs() {
        this._callAs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get callAsInput() {
        return this._callAs;
    }
    get capabilities() {
        return cdktf.Fn.tolist(this.getListAttribute('capabilities'));
    }
    set capabilities(value) {
        this._capabilities = value;
    }
    resetCapabilities() {
        this._capabilities = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capabilitiesInput() {
        return this._capabilities;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get executionRoleName() {
        return this.getStringAttribute('execution_role_name');
    }
    set executionRoleName(value) {
        this._executionRoleName = value;
    }
    resetExecutionRoleName() {
        this._executionRoleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleNameInput() {
        return this._executionRoleName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get permissionModel() {
        return this.getStringAttribute('permission_model');
    }
    set permissionModel(value) {
        this._permissionModel = value;
    }
    resetPermissionModel() {
        this._permissionModel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionModelInput() {
        return this._permissionModel;
    }
    // stack_set_id - computed: true, optional: false, required: false
    get stackSetId() {
        return this.getStringAttribute('stack_set_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get templateBody() {
        return this.getStringAttribute('template_body');
    }
    set templateBody(value) {
        this._templateBody = value;
    }
    resetTemplateBody() {
        this._templateBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateBodyInput() {
        return this._templateBody;
    }
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    set templateUrl(value) {
        this._templateUrl = value;
    }
    resetTemplateUrl() {
        this._templateUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateUrlInput() {
        return this._templateUrl;
    }
    get autoDeployment() {
        return this._autoDeployment;
    }
    putAutoDeployment(value) {
        this._autoDeployment.internalValue = value;
    }
    resetAutoDeployment() {
        this._autoDeployment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeploymentInput() {
        return this._autoDeployment.internalValue;
    }
    get managedExecution() {
        return this._managedExecution;
    }
    putManagedExecution(value) {
        this._managedExecution.internalValue = value;
    }
    resetManagedExecution() {
        this._managedExecution.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedExecutionInput() {
        return this._managedExecution.internalValue;
    }
    get operationPreferences() {
        return this._operationPreferences;
    }
    putOperationPreferences(value) {
        this._operationPreferences.internalValue = value;
    }
    resetOperationPreferences() {
        this._operationPreferences.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operationPreferencesInput() {
        return this._operationPreferences.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            administration_role_arn: cdktf.stringToTerraform(this._administrationRoleArn),
            call_as: cdktf.stringToTerraform(this._callAs),
            capabilities: cdktf.listMapper(cdktf.stringToTerraform, false)(this._capabilities),
            description: cdktf.stringToTerraform(this._description),
            execution_role_name: cdktf.stringToTerraform(this._executionRoleName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parameters: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameters),
            permission_model: cdktf.stringToTerraform(this._permissionModel),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            template_body: cdktf.stringToTerraform(this._templateBody),
            template_url: cdktf.stringToTerraform(this._templateUrl),
            auto_deployment: cloudformationStackSetAutoDeploymentToTerraform(this._autoDeployment.internalValue),
            managed_execution: cloudformationStackSetManagedExecutionToTerraform(this._managedExecution.internalValue),
            operation_preferences: cloudformationStackSetOperationPreferencesToTerraform(this._operationPreferences.internalValue),
            timeouts: cloudformationStackSetTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            administration_role_arn: {
                value: cdktf.stringToHclTerraform(this._administrationRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            call_as: {
                value: cdktf.stringToHclTerraform(this._callAs),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            capabilities: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._capabilities),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            execution_role_name: {
                value: cdktf.stringToHclTerraform(this._executionRoleName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parameters: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._parameters),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            permission_model: {
                value: cdktf.stringToHclTerraform(this._permissionModel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            template_body: {
                value: cdktf.stringToHclTerraform(this._templateBody),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            template_url: {
                value: cdktf.stringToHclTerraform(this._templateUrl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            auto_deployment: {
                value: cloudformationStackSetAutoDeploymentToHclTerraform(this._autoDeployment.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudformationStackSetAutoDeploymentList",
            },
            managed_execution: {
                value: cloudformationStackSetManagedExecutionToHclTerraform(this._managedExecution.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudformationStackSetManagedExecutionList",
            },
            operation_preferences: {
                value: cloudformationStackSetOperationPreferencesToHclTerraform(this._operationPreferences.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudformationStackSetOperationPreferencesList",
            },
            timeouts: {
                value: cloudformationStackSetTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CloudformationStackSetTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudformationStackSet = CloudformationStackSet;
_e = JSII_RTTI_SYMBOL_1;
CloudformationStackSet[_e] = { fqn: "@cdktf/provider-aws.cloudformationStackSet.CloudformationStackSet", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudformationStackSet.tfResourceType = "aws_cloudformation_stack_set";
