// @winglang/sdk/lib/target-tf-aws/redis.inflight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisClient = void 0;
const client_elasticache_1 = require("@aws-sdk/client-elasticache");
const ex_1 = require("../ex");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const IoRedis = require("ioredis");
class RedisClient extends ex_1.RedisClientBase {
    constructor(clusterId, connection, elasticacheClient = new client_elasticache_1.ElastiCacheClient({})) {
        super();
        this.clusterId = clusterId;
        this.connection = connection;
        this.elasticacheClient = elasticacheClient;
    }
    /**
     * The Redis cluster endpoint is not available to inject from the cdktf resource,
     * therefore we need to query the AWS API to get it. This is a helper function that will
     * use the cluster id to get the endpoint.
     *
     * @returns The Redis cluster endpoint
     */
    async getEndpoint() {
        const command = new client_elasticache_1.DescribeCacheClustersCommand({
            CacheClusterId: this.clusterId,
            ShowCacheNodeInfo: true,
        });
        const resp = await this.elasticacheClient.send(command);
        if (!resp.CacheClusters) {
            throw new Error("No Redis cache clusters found");
        }
        const cluster = resp.CacheClusters[0];
        if (!cluster.CacheNodes) {
            throw new Error("No Redis cache nodes found");
        }
        const cacheNode = cluster.CacheNodes[0];
        if (!cacheNode.Endpoint) {
            throw new Error("No Redis cache node endpoint found");
        }
        return cacheNode.Endpoint.Address;
    }
    async rawClient() {
        if (!this.clusterId) {
            throw new Error("No cluster id provided");
        }
        if (this.connection) {
            return this.connection;
        }
        if (!this.connectionUrl) {
            this.connectionUrl = await this.getEndpoint();
        }
        this.connection = new IoRedis(this.connectionUrl);
        return this.connection;
    }
    async url() {
        if (this.connectionUrl == undefined) {
            this.connectionUrl = await this.getEndpoint();
        }
        return this.connectionUrl;
    }
}
exports.RedisClient = RedisClient;
