// @winglang/sdk/lib/target-tf-aws/secret.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secret = void 0;
const cdktf_1 = require("cdktf");
const data_aws_secretsmanager_secret_1 = require("../.gen/providers/aws/data-aws-secretsmanager-secret");
const secretsmanager_secret_1 = require("../.gen/providers/aws/secretsmanager-secret");
const cloud = __importStar(require("../cloud"));
const core = __importStar(require("../core"));
const resource_names_1 = require("../shared/resource-names");
const shared_aws_1 = require("../shared-aws");
const permissions_1 = require("../shared-aws/permissions");
/**
 * The secret name can contain ASCII letters, numbers, and the following characters: /_+=.@-
 */
const NAME_OPTS = {
    disallowedRegex: /[^\w/+=.@-]+/g,
};
/**
 * AWS implementation of `cloud.Secret`
 *
 * @inflight `@winglang/sdk.cloud.ISecretClient`
 */
class Secret extends cloud.Secret {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        if (props.name) {
            this.secret = new data_aws_secretsmanager_secret_1.DataAwsSecretsmanagerSecret(this, "Default", {
                name: props.name,
            });
        }
        else {
            (this._name = resource_names_1.ResourceNames.generateName(this, NAME_OPTS)),
                (this.secret = new secretsmanager_secret_1.SecretsmanagerSecret(this, "Default", {
                    name: this._name,
                }));
            new cdktf_1.TerraformOutput(this, "SecretArn", {
                value: this.secret.arn,
            });
        }
    }
    /** @internal */
    get _liftMap() {
        return {
            [cloud.SecretInflightMethods.VALUE]: [],
            [cloud.SecretInflightMethods.VALUE_JSON]: [],
        };
    }
    onLift(host, ops) {
        if (!shared_aws_1.AwsInflightHost.isAwsInflightHost(host)) {
            throw new Error("Host is expected to implement `IAwsInfightHost`");
        }
        host.addPolicyStatements(...(0, permissions_1.calculateSecretPermissions)(this.secret.arn, ops));
        host.addEnvironment(this.envName(), this.secret.arn);
        super.onLift(host, ops);
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-aws", "shared-aws"), __filename, "SecretClient", [`process.env["${this.envName()}"]`]);
    }
    envName() {
        return `SECRET_ARN_${this.node.addr.slice(-8)}`;
    }
}
exports.Secret = Secret;
