// @winglang/sdk/lib/target-tf-aws/test-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunner = void 0;
const terraform_output_1 = require("cdktf/lib/terraform-output");
const tokens_1 = require("cdktf/lib/tokens");
const function_1 = require("./function");
const core = __importStar(require("../core"));
const std = __importStar(require("../std"));
const OUTPUT_TEST_RUNNER_FUNCTION_IDENTIFIERS = "WING_TEST_RUNNER_FUNCTION_IDENTIFIERS";
/**
 * AWS implementation of `cloud.TestRunner`.
 *
 * @inflight `@winglang/sdk.cloud.ITestRunnerClient`
 */
class TestRunner extends std.TestRunner {
    constructor(scope, id, props = {}) {
        super(scope, id, props);
        // This output is created so the CLI's `wing test` command can obtain a list
        // of all ARNs of test functions by running `terraform output`.
        const output = new terraform_output_1.TerraformOutput(this, "TestFunctionArns", {
            value: tokens_1.Lazy.stringValue({
                produce: () => {
                    return JSON.stringify([...this.getTestFunctionArns().entries()]);
                },
            }),
        });
        output.overrideLogicalId(OUTPUT_TEST_RUNNER_FUNCTION_IDENTIFIERS);
    }
    onLift(host, ops) {
        // Collect all of the test functions and their ARNs, and pass them to the
        // test engine so they can be invoked inflight.
        // TODO: are we going to run into AWS's 4KB environment variable limit here?
        // some solutions:
        // - base64 encode the string value
        // - move the logic for picking one test from each isolated environment to
        //   here so that if there are N tests in the original app and N
        //   environments, we only need to output N test function ARNs instead of
        //   N * N
        const testFunctions = this.getTestFunctionArns();
        host.addEnvironment(this.envTestFunctionArns(), JSON.stringify([...testFunctions.entries()]));
        super.onLift(host, ops);
    }
    /** @internal */
    _preSynthesize() {
        // add a dependency on each test function
        for (const test of this.findTests()) {
            if (test._fn) {
                this.node.addDependency(test._fn);
            }
        }
        super._preSynthesize();
    }
    getTestFunctionArns() {
        const arns = new Map();
        for (const test of this.findTests()) {
            if (test._fn) {
                if (!(test._fn instanceof function_1.Function)) {
                    throw new Error(`Unsupported test function type, ${test._fn.node.path} was not a tfaws.Function`);
                }
                arns.set(test.node.path, test._fn.functionArn);
            }
        }
        return arns;
    }
    /** @internal */
    _toInflight() {
        return core.InflightClient.for(__dirname.replace("target-tf-aws", "shared-aws"), __filename, "TestRunnerClient", [`process.env["${this.envTestFunctionArns()}"]`]);
    }
    envTestFunctionArns() {
        return `TEST_RUNNER_FUNCTIONS_${this.node.addr.slice(-8)}`;
    }
}
exports.TestRunner = TestRunner;
