// @cdktf/provider-aws/lib/sagemaker-model/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerModel = exports.SagemakerModelVpcConfigOutputReference = exports.sagemakerModelVpcConfigToHclTerraform = exports.sagemakerModelVpcConfigToTerraform = exports.SagemakerModelPrimaryContainerOutputReference = exports.sagemakerModelPrimaryContainerToHclTerraform = exports.sagemakerModelPrimaryContainerToTerraform = exports.SagemakerModelPrimaryContainerModelDataSourceOutputReference = exports.sagemakerModelPrimaryContainerModelDataSourceToHclTerraform = exports.sagemakerModelPrimaryContainerModelDataSourceToTerraform = exports.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList = exports.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference = exports.sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToHclTerraform = exports.sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToTerraform = exports.SagemakerModelPrimaryContainerImageConfigOutputReference = exports.sagemakerModelPrimaryContainerImageConfigToHclTerraform = exports.sagemakerModelPrimaryContainerImageConfigToTerraform = exports.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference = exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToHclTerraform = exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform = exports.SagemakerModelInferenceExecutionConfigOutputReference = exports.sagemakerModelInferenceExecutionConfigToHclTerraform = exports.sagemakerModelInferenceExecutionConfigToTerraform = exports.SagemakerModelContainerList = exports.SagemakerModelContainerOutputReference = exports.sagemakerModelContainerToHclTerraform = exports.sagemakerModelContainerToTerraform = exports.SagemakerModelContainerModelDataSourceOutputReference = exports.sagemakerModelContainerModelDataSourceToHclTerraform = exports.sagemakerModelContainerModelDataSourceToTerraform = exports.SagemakerModelContainerModelDataSourceS3DataSourceList = exports.SagemakerModelContainerModelDataSourceS3DataSourceOutputReference = exports.sagemakerModelContainerModelDataSourceS3DataSourceToHclTerraform = exports.sagemakerModelContainerModelDataSourceS3DataSourceToTerraform = exports.SagemakerModelContainerImageConfigOutputReference = exports.sagemakerModelContainerImageConfigToHclTerraform = exports.sagemakerModelContainerImageConfigToTerraform = exports.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference = exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToHclTerraform = exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_credentials_provider_arn: cdktf.stringToTerraform(struct.repositoryCredentialsProviderArn),
    };
}
exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform = sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform;
function sagemakerModelContainerImageConfigRepositoryAuthConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository_credentials_provider_arn: {
            value: cdktf.stringToHclTerraform(struct.repositoryCredentialsProviderArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToHclTerraform = sagemakerModelContainerImageConfigRepositoryAuthConfigToHclTerraform;
class SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryCredentialsProviderArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryCredentialsProviderArn = this._repositoryCredentialsProviderArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryCredentialsProviderArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryCredentialsProviderArn = value.repositoryCredentialsProviderArn;
        }
    }
    get repositoryCredentialsProviderArn() {
        return this.getStringAttribute('repository_credentials_provider_arn');
    }
    set repositoryCredentialsProviderArn(value) {
        this._repositoryCredentialsProviderArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCredentialsProviderArnInput() {
        return this._repositoryCredentialsProviderArn;
    }
}
exports.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference = SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference", version: "19.15.0" };
function sagemakerModelContainerImageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_access_mode: cdktf.stringToTerraform(struct.repositoryAccessMode),
        repository_auth_config: sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform(struct.repositoryAuthConfig),
    };
}
exports.sagemakerModelContainerImageConfigToTerraform = sagemakerModelContainerImageConfigToTerraform;
function sagemakerModelContainerImageConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository_access_mode: {
            value: cdktf.stringToHclTerraform(struct.repositoryAccessMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        repository_auth_config: {
            value: sagemakerModelContainerImageConfigRepositoryAuthConfigToHclTerraform(struct.repositoryAuthConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelContainerImageConfigRepositoryAuthConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelContainerImageConfigToHclTerraform = sagemakerModelContainerImageConfigToHclTerraform;
class SagemakerModelContainerImageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // repository_auth_config - computed: false, optional: true, required: false
        this._repositoryAuthConfig = new SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference(this, "repository_auth_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryAccessMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAccessMode = this._repositoryAccessMode;
        }
        if (this._repositoryAuthConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAuthConfig = this._repositoryAuthConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryAccessMode = undefined;
            this._repositoryAuthConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryAccessMode = value.repositoryAccessMode;
            this._repositoryAuthConfig.internalValue = value.repositoryAuthConfig;
        }
    }
    get repositoryAccessMode() {
        return this.getStringAttribute('repository_access_mode');
    }
    set repositoryAccessMode(value) {
        this._repositoryAccessMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAccessModeInput() {
        return this._repositoryAccessMode;
    }
    get repositoryAuthConfig() {
        return this._repositoryAuthConfig;
    }
    putRepositoryAuthConfig(value) {
        this._repositoryAuthConfig.internalValue = value;
    }
    resetRepositoryAuthConfig() {
        this._repositoryAuthConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAuthConfigInput() {
        return this._repositoryAuthConfig.internalValue;
    }
}
exports.SagemakerModelContainerImageConfigOutputReference = SagemakerModelContainerImageConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerImageConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerImageConfigOutputReference", version: "19.15.0" };
function sagemakerModelContainerModelDataSourceS3DataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        compression_type: cdktf.stringToTerraform(struct.compressionType),
        s3_data_type: cdktf.stringToTerraform(struct.s3DataType),
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
    };
}
exports.sagemakerModelContainerModelDataSourceS3DataSourceToTerraform = sagemakerModelContainerModelDataSourceS3DataSourceToTerraform;
function sagemakerModelContainerModelDataSourceS3DataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        compression_type: {
            value: cdktf.stringToHclTerraform(struct.compressionType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_data_type: {
            value: cdktf.stringToHclTerraform(struct.s3DataType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelContainerModelDataSourceS3DataSourceToHclTerraform = sagemakerModelContainerModelDataSourceS3DataSourceToHclTerraform;
class SagemakerModelContainerModelDataSourceS3DataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._compressionType !== undefined) {
            hasAnyValues = true;
            internalValueResult.compressionType = this._compressionType;
        }
        if (this._s3DataType !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3DataType = this._s3DataType;
        }
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._compressionType = undefined;
            this._s3DataType = undefined;
            this._s3Uri = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._compressionType = value.compressionType;
            this._s3DataType = value.s3DataType;
            this._s3Uri = value.s3Uri;
        }
    }
    get compressionType() {
        return this.getStringAttribute('compression_type');
    }
    set compressionType(value) {
        this._compressionType = value;
    }
    // Temporarily expose input value. Use with caution.
    get compressionTypeInput() {
        return this._compressionType;
    }
    get s3DataType() {
        return this.getStringAttribute('s3_data_type');
    }
    set s3DataType(value) {
        this._s3DataType = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3DataTypeInput() {
        return this._s3DataType;
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
}
exports.SagemakerModelContainerModelDataSourceS3DataSourceOutputReference = SagemakerModelContainerModelDataSourceS3DataSourceOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerModelDataSourceS3DataSourceOutputReference[_c] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerModelDataSourceS3DataSourceOutputReference", version: "19.15.0" };
class SagemakerModelContainerModelDataSourceS3DataSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerModelContainerModelDataSourceS3DataSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerModelContainerModelDataSourceS3DataSourceList = SagemakerModelContainerModelDataSourceS3DataSourceList;
_d = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerModelDataSourceS3DataSourceList[_d] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerModelDataSourceS3DataSourceList", version: "19.15.0" };
function sagemakerModelContainerModelDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_data_source: cdktf.listMapper(sagemakerModelContainerModelDataSourceS3DataSourceToTerraform, true)(struct.s3DataSource),
    };
}
exports.sagemakerModelContainerModelDataSourceToTerraform = sagemakerModelContainerModelDataSourceToTerraform;
function sagemakerModelContainerModelDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        s3_data_source: {
            value: cdktf.listMapperHcl(sagemakerModelContainerModelDataSourceS3DataSourceToHclTerraform, true)(struct.s3DataSource),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelContainerModelDataSourceS3DataSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelContainerModelDataSourceToHclTerraform = sagemakerModelContainerModelDataSourceToHclTerraform;
class SagemakerModelContainerModelDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // s3_data_source - computed: false, optional: false, required: true
        this._s3DataSource = new SagemakerModelContainerModelDataSourceS3DataSourceList(this, "s3_data_source", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3DataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3DataSource = this._s3DataSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3DataSource.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3DataSource.internalValue = value.s3DataSource;
        }
    }
    get s3DataSource() {
        return this._s3DataSource;
    }
    putS3DataSource(value) {
        this._s3DataSource.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3DataSourceInput() {
        return this._s3DataSource.internalValue;
    }
}
exports.SagemakerModelContainerModelDataSourceOutputReference = SagemakerModelContainerModelDataSourceOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerModelDataSourceOutputReference[_e] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerModelDataSourceOutputReference", version: "19.15.0" };
function sagemakerModelContainerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_hostname: cdktf.stringToTerraform(struct.containerHostname),
        environment: cdktf.hashMapper(cdktf.stringToTerraform)(struct.environment),
        image: cdktf.stringToTerraform(struct.image),
        mode: cdktf.stringToTerraform(struct.mode),
        model_data_url: cdktf.stringToTerraform(struct.modelDataUrl),
        model_package_name: cdktf.stringToTerraform(struct.modelPackageName),
        image_config: sagemakerModelContainerImageConfigToTerraform(struct.imageConfig),
        model_data_source: sagemakerModelContainerModelDataSourceToTerraform(struct.modelDataSource),
    };
}
exports.sagemakerModelContainerToTerraform = sagemakerModelContainerToTerraform;
function sagemakerModelContainerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_hostname: {
            value: cdktf.stringToHclTerraform(struct.containerHostname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        environment: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.environment),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        image: {
            value: cdktf.stringToHclTerraform(struct.image),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_data_url: {
            value: cdktf.stringToHclTerraform(struct.modelDataUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_package_name: {
            value: cdktf.stringToHclTerraform(struct.modelPackageName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        image_config: {
            value: sagemakerModelContainerImageConfigToHclTerraform(struct.imageConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelContainerImageConfigList",
        },
        model_data_source: {
            value: sagemakerModelContainerModelDataSourceToHclTerraform(struct.modelDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelContainerModelDataSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelContainerToHclTerraform = sagemakerModelContainerToHclTerraform;
class SagemakerModelContainerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // image_config - computed: false, optional: true, required: false
        this._imageConfig = new SagemakerModelContainerImageConfigOutputReference(this, "image_config");
        // model_data_source - computed: false, optional: true, required: false
        this._modelDataSource = new SagemakerModelContainerModelDataSourceOutputReference(this, "model_data_source");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerHostname = this._containerHostname;
        }
        if (this._environment !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._modelDataUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataUrl = this._modelDataUrl;
        }
        if (this._modelPackageName !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelPackageName = this._modelPackageName;
        }
        if (this._imageConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageConfig = this._imageConfig?.internalValue;
        }
        if (this._modelDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataSource = this._modelDataSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerHostname = undefined;
            this._environment = undefined;
            this._image = undefined;
            this._mode = undefined;
            this._modelDataUrl = undefined;
            this._modelPackageName = undefined;
            this._imageConfig.internalValue = undefined;
            this._modelDataSource.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerHostname = value.containerHostname;
            this._environment = value.environment;
            this._image = value.image;
            this._mode = value.mode;
            this._modelDataUrl = value.modelDataUrl;
            this._modelPackageName = value.modelPackageName;
            this._imageConfig.internalValue = value.imageConfig;
            this._modelDataSource.internalValue = value.modelDataSource;
        }
    }
    get containerHostname() {
        return this.getStringAttribute('container_hostname');
    }
    set containerHostname(value) {
        this._containerHostname = value;
    }
    resetContainerHostname() {
        this._containerHostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerHostnameInput() {
        return this._containerHostname;
    }
    get environment() {
        return this.getStringMapAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    resetImage() {
        this._image = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get modelDataUrl() {
        return this.getStringAttribute('model_data_url');
    }
    set modelDataUrl(value) {
        this._modelDataUrl = value;
    }
    resetModelDataUrl() {
        this._modelDataUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataUrlInput() {
        return this._modelDataUrl;
    }
    get modelPackageName() {
        return this.getStringAttribute('model_package_name');
    }
    set modelPackageName(value) {
        this._modelPackageName = value;
    }
    resetModelPackageName() {
        this._modelPackageName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelPackageNameInput() {
        return this._modelPackageName;
    }
    get imageConfig() {
        return this._imageConfig;
    }
    putImageConfig(value) {
        this._imageConfig.internalValue = value;
    }
    resetImageConfig() {
        this._imageConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigInput() {
        return this._imageConfig.internalValue;
    }
    get modelDataSource() {
        return this._modelDataSource;
    }
    putModelDataSource(value) {
        this._modelDataSource.internalValue = value;
    }
    resetModelDataSource() {
        this._modelDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataSourceInput() {
        return this._modelDataSource.internalValue;
    }
}
exports.SagemakerModelContainerOutputReference = SagemakerModelContainerOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerOutputReference[_f] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerOutputReference", version: "19.15.0" };
class SagemakerModelContainerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerModelContainerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerModelContainerList = SagemakerModelContainerList;
_g = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerList[_g] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerList", version: "19.15.0" };
function sagemakerModelInferenceExecutionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.sagemakerModelInferenceExecutionConfigToTerraform = sagemakerModelInferenceExecutionConfigToTerraform;
function sagemakerModelInferenceExecutionConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelInferenceExecutionConfigToHclTerraform = sagemakerModelInferenceExecutionConfigToHclTerraform;
class SagemakerModelInferenceExecutionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.SagemakerModelInferenceExecutionConfigOutputReference = SagemakerModelInferenceExecutionConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SagemakerModelInferenceExecutionConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelInferenceExecutionConfigOutputReference", version: "19.15.0" };
function sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_credentials_provider_arn: cdktf.stringToTerraform(struct.repositoryCredentialsProviderArn),
    };
}
exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform = sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform;
function sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository_credentials_provider_arn: {
            value: cdktf.stringToHclTerraform(struct.repositoryCredentialsProviderArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToHclTerraform = sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToHclTerraform;
class SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryCredentialsProviderArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryCredentialsProviderArn = this._repositoryCredentialsProviderArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryCredentialsProviderArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryCredentialsProviderArn = value.repositoryCredentialsProviderArn;
        }
    }
    get repositoryCredentialsProviderArn() {
        return this.getStringAttribute('repository_credentials_provider_arn');
    }
    set repositoryCredentialsProviderArn(value) {
        this._repositoryCredentialsProviderArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCredentialsProviderArnInput() {
        return this._repositoryCredentialsProviderArn;
    }
}
exports.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference = SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference", version: "19.15.0" };
function sagemakerModelPrimaryContainerImageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_access_mode: cdktf.stringToTerraform(struct.repositoryAccessMode),
        repository_auth_config: sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform(struct.repositoryAuthConfig),
    };
}
exports.sagemakerModelPrimaryContainerImageConfigToTerraform = sagemakerModelPrimaryContainerImageConfigToTerraform;
function sagemakerModelPrimaryContainerImageConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository_access_mode: {
            value: cdktf.stringToHclTerraform(struct.repositoryAccessMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        repository_auth_config: {
            value: sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToHclTerraform(struct.repositoryAuthConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelPrimaryContainerImageConfigToHclTerraform = sagemakerModelPrimaryContainerImageConfigToHclTerraform;
class SagemakerModelPrimaryContainerImageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // repository_auth_config - computed: false, optional: true, required: false
        this._repositoryAuthConfig = new SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference(this, "repository_auth_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryAccessMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAccessMode = this._repositoryAccessMode;
        }
        if (this._repositoryAuthConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAuthConfig = this._repositoryAuthConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryAccessMode = undefined;
            this._repositoryAuthConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryAccessMode = value.repositoryAccessMode;
            this._repositoryAuthConfig.internalValue = value.repositoryAuthConfig;
        }
    }
    get repositoryAccessMode() {
        return this.getStringAttribute('repository_access_mode');
    }
    set repositoryAccessMode(value) {
        this._repositoryAccessMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAccessModeInput() {
        return this._repositoryAccessMode;
    }
    get repositoryAuthConfig() {
        return this._repositoryAuthConfig;
    }
    putRepositoryAuthConfig(value) {
        this._repositoryAuthConfig.internalValue = value;
    }
    resetRepositoryAuthConfig() {
        this._repositoryAuthConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAuthConfigInput() {
        return this._repositoryAuthConfig.internalValue;
    }
}
exports.SagemakerModelPrimaryContainerImageConfigOutputReference = SagemakerModelPrimaryContainerImageConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerImageConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerImageConfigOutputReference", version: "19.15.0" };
function sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        compression_type: cdktf.stringToTerraform(struct.compressionType),
        s3_data_type: cdktf.stringToTerraform(struct.s3DataType),
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
    };
}
exports.sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToTerraform = sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToTerraform;
function sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        compression_type: {
            value: cdktf.stringToHclTerraform(struct.compressionType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_data_type: {
            value: cdktf.stringToHclTerraform(struct.s3DataType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_uri: {
            value: cdktf.stringToHclTerraform(struct.s3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToHclTerraform = sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToHclTerraform;
class SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._compressionType !== undefined) {
            hasAnyValues = true;
            internalValueResult.compressionType = this._compressionType;
        }
        if (this._s3DataType !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3DataType = this._s3DataType;
        }
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._compressionType = undefined;
            this._s3DataType = undefined;
            this._s3Uri = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._compressionType = value.compressionType;
            this._s3DataType = value.s3DataType;
            this._s3Uri = value.s3Uri;
        }
    }
    get compressionType() {
        return this.getStringAttribute('compression_type');
    }
    set compressionType(value) {
        this._compressionType = value;
    }
    // Temporarily expose input value. Use with caution.
    get compressionTypeInput() {
        return this._compressionType;
    }
    get s3DataType() {
        return this.getStringAttribute('s3_data_type');
    }
    set s3DataType(value) {
        this._s3DataType = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3DataTypeInput() {
        return this._s3DataType;
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
}
exports.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference = SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference;
_l = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference[_l] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference", version: "19.15.0" };
class SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerModelPrimaryContainerModelDataSourceS3DataSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList = SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList;
_m = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList[_m] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList", version: "19.15.0" };
function sagemakerModelPrimaryContainerModelDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_data_source: cdktf.listMapper(sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToTerraform, true)(struct.s3DataSource),
    };
}
exports.sagemakerModelPrimaryContainerModelDataSourceToTerraform = sagemakerModelPrimaryContainerModelDataSourceToTerraform;
function sagemakerModelPrimaryContainerModelDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        s3_data_source: {
            value: cdktf.listMapperHcl(sagemakerModelPrimaryContainerModelDataSourceS3DataSourceToHclTerraform, true)(struct.s3DataSource),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelPrimaryContainerModelDataSourceToHclTerraform = sagemakerModelPrimaryContainerModelDataSourceToHclTerraform;
class SagemakerModelPrimaryContainerModelDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // s3_data_source - computed: false, optional: false, required: true
        this._s3DataSource = new SagemakerModelPrimaryContainerModelDataSourceS3DataSourceList(this, "s3_data_source", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3DataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3DataSource = this._s3DataSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3DataSource.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3DataSource.internalValue = value.s3DataSource;
        }
    }
    get s3DataSource() {
        return this._s3DataSource;
    }
    putS3DataSource(value) {
        this._s3DataSource.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3DataSourceInput() {
        return this._s3DataSource.internalValue;
    }
}
exports.SagemakerModelPrimaryContainerModelDataSourceOutputReference = SagemakerModelPrimaryContainerModelDataSourceOutputReference;
_o = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerModelDataSourceOutputReference[_o] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerModelDataSourceOutputReference", version: "19.15.0" };
function sagemakerModelPrimaryContainerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_hostname: cdktf.stringToTerraform(struct.containerHostname),
        environment: cdktf.hashMapper(cdktf.stringToTerraform)(struct.environment),
        image: cdktf.stringToTerraform(struct.image),
        mode: cdktf.stringToTerraform(struct.mode),
        model_data_url: cdktf.stringToTerraform(struct.modelDataUrl),
        model_package_name: cdktf.stringToTerraform(struct.modelPackageName),
        image_config: sagemakerModelPrimaryContainerImageConfigToTerraform(struct.imageConfig),
        model_data_source: sagemakerModelPrimaryContainerModelDataSourceToTerraform(struct.modelDataSource),
    };
}
exports.sagemakerModelPrimaryContainerToTerraform = sagemakerModelPrimaryContainerToTerraform;
function sagemakerModelPrimaryContainerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_hostname: {
            value: cdktf.stringToHclTerraform(struct.containerHostname),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        environment: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.environment),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        image: {
            value: cdktf.stringToHclTerraform(struct.image),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_data_url: {
            value: cdktf.stringToHclTerraform(struct.modelDataUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_package_name: {
            value: cdktf.stringToHclTerraform(struct.modelPackageName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        image_config: {
            value: sagemakerModelPrimaryContainerImageConfigToHclTerraform(struct.imageConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelPrimaryContainerImageConfigList",
        },
        model_data_source: {
            value: sagemakerModelPrimaryContainerModelDataSourceToHclTerraform(struct.modelDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerModelPrimaryContainerModelDataSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelPrimaryContainerToHclTerraform = sagemakerModelPrimaryContainerToHclTerraform;
class SagemakerModelPrimaryContainerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // image_config - computed: false, optional: true, required: false
        this._imageConfig = new SagemakerModelPrimaryContainerImageConfigOutputReference(this, "image_config");
        // model_data_source - computed: false, optional: true, required: false
        this._modelDataSource = new SagemakerModelPrimaryContainerModelDataSourceOutputReference(this, "model_data_source");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerHostname = this._containerHostname;
        }
        if (this._environment !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._modelDataUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataUrl = this._modelDataUrl;
        }
        if (this._modelPackageName !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelPackageName = this._modelPackageName;
        }
        if (this._imageConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageConfig = this._imageConfig?.internalValue;
        }
        if (this._modelDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataSource = this._modelDataSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerHostname = undefined;
            this._environment = undefined;
            this._image = undefined;
            this._mode = undefined;
            this._modelDataUrl = undefined;
            this._modelPackageName = undefined;
            this._imageConfig.internalValue = undefined;
            this._modelDataSource.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerHostname = value.containerHostname;
            this._environment = value.environment;
            this._image = value.image;
            this._mode = value.mode;
            this._modelDataUrl = value.modelDataUrl;
            this._modelPackageName = value.modelPackageName;
            this._imageConfig.internalValue = value.imageConfig;
            this._modelDataSource.internalValue = value.modelDataSource;
        }
    }
    get containerHostname() {
        return this.getStringAttribute('container_hostname');
    }
    set containerHostname(value) {
        this._containerHostname = value;
    }
    resetContainerHostname() {
        this._containerHostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerHostnameInput() {
        return this._containerHostname;
    }
    get environment() {
        return this.getStringMapAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    resetImage() {
        this._image = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get modelDataUrl() {
        return this.getStringAttribute('model_data_url');
    }
    set modelDataUrl(value) {
        this._modelDataUrl = value;
    }
    resetModelDataUrl() {
        this._modelDataUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataUrlInput() {
        return this._modelDataUrl;
    }
    get modelPackageName() {
        return this.getStringAttribute('model_package_name');
    }
    set modelPackageName(value) {
        this._modelPackageName = value;
    }
    resetModelPackageName() {
        this._modelPackageName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelPackageNameInput() {
        return this._modelPackageName;
    }
    get imageConfig() {
        return this._imageConfig;
    }
    putImageConfig(value) {
        this._imageConfig.internalValue = value;
    }
    resetImageConfig() {
        this._imageConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigInput() {
        return this._imageConfig.internalValue;
    }
    get modelDataSource() {
        return this._modelDataSource;
    }
    putModelDataSource(value) {
        this._modelDataSource.internalValue = value;
    }
    resetModelDataSource() {
        this._modelDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataSourceInput() {
        return this._modelDataSource.internalValue;
    }
}
exports.SagemakerModelPrimaryContainerOutputReference = SagemakerModelPrimaryContainerOutputReference;
_p = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerOutputReference[_p] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerOutputReference", version: "19.15.0" };
function sagemakerModelVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.sagemakerModelVpcConfigToTerraform = sagemakerModelVpcConfigToTerraform;
function sagemakerModelVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerModelVpcConfigToHclTerraform = sagemakerModelVpcConfigToHclTerraform;
class SagemakerModelVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnets = value.subnets;
        }
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.SagemakerModelVpcConfigOutputReference = SagemakerModelVpcConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
SagemakerModelVpcConfigOutputReference[_q] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelVpcConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model aws_sagemaker_model}
*/
class SagemakerModel extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SagemakerModel resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SagemakerModel to import
    * @param importFromId The id of the existing SagemakerModel that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SagemakerModel to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sagemaker_model", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model aws_sagemaker_model} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerModelConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_model',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // container - computed: false, optional: true, required: false
        this._container = new SagemakerModelContainerList(this, "container", false);
        // inference_execution_config - computed: false, optional: true, required: false
        this._inferenceExecutionConfig = new SagemakerModelInferenceExecutionConfigOutputReference(this, "inference_execution_config");
        // primary_container - computed: false, optional: true, required: false
        this._primaryContainer = new SagemakerModelPrimaryContainerOutputReference(this, "primary_container");
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new SagemakerModelVpcConfigOutputReference(this, "vpc_config");
        this._enableNetworkIsolation = config.enableNetworkIsolation;
        this._executionRoleArn = config.executionRoleArn;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._container.internalValue = config.container;
        this._inferenceExecutionConfig.internalValue = config.inferenceExecutionConfig;
        this._primaryContainer.internalValue = config.primaryContainer;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get enableNetworkIsolation() {
        return this.getBooleanAttribute('enable_network_isolation');
    }
    set enableNetworkIsolation(value) {
        this._enableNetworkIsolation = value;
    }
    resetEnableNetworkIsolation() {
        this._enableNetworkIsolation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableNetworkIsolationInput() {
        return this._enableNetworkIsolation;
    }
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get container() {
        return this._container;
    }
    putContainer(value) {
        this._container.internalValue = value;
    }
    resetContainer() {
        this._container.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerInput() {
        return this._container.internalValue;
    }
    get inferenceExecutionConfig() {
        return this._inferenceExecutionConfig;
    }
    putInferenceExecutionConfig(value) {
        this._inferenceExecutionConfig.internalValue = value;
    }
    resetInferenceExecutionConfig() {
        this._inferenceExecutionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inferenceExecutionConfigInput() {
        return this._inferenceExecutionConfig.internalValue;
    }
    get primaryContainer() {
        return this._primaryContainer;
    }
    putPrimaryContainer(value) {
        this._primaryContainer.internalValue = value;
    }
    resetPrimaryContainer() {
        this._primaryContainer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryContainerInput() {
        return this._primaryContainer.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enable_network_isolation: cdktf.booleanToTerraform(this._enableNetworkIsolation),
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            container: cdktf.listMapper(sagemakerModelContainerToTerraform, true)(this._container.internalValue),
            inference_execution_config: sagemakerModelInferenceExecutionConfigToTerraform(this._inferenceExecutionConfig.internalValue),
            primary_container: sagemakerModelPrimaryContainerToTerraform(this._primaryContainer.internalValue),
            vpc_config: sagemakerModelVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            enable_network_isolation: {
                value: cdktf.booleanToHclTerraform(this._enableNetworkIsolation),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            execution_role_arn: {
                value: cdktf.stringToHclTerraform(this._executionRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            container: {
                value: cdktf.listMapperHcl(sagemakerModelContainerToHclTerraform, true)(this._container.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerModelContainerList",
            },
            inference_execution_config: {
                value: sagemakerModelInferenceExecutionConfigToHclTerraform(this._inferenceExecutionConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerModelInferenceExecutionConfigList",
            },
            primary_container: {
                value: sagemakerModelPrimaryContainerToHclTerraform(this._primaryContainer.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerModelPrimaryContainerList",
            },
            vpc_config: {
                value: sagemakerModelVpcConfigToHclTerraform(this._vpcConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerModelVpcConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SagemakerModel = SagemakerModel;
_r = JSII_RTTI_SYMBOL_1;
SagemakerModel[_r] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModel", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SagemakerModel.tfResourceType = "aws_sagemaker_model";
