// @winglang/sdk/lib/std/struct.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Struct = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const json_schema_1 = require("./json_schema");
const core_1 = require("../core");
/**
 * Shared behavior for all structs
 *
 * @typeparam T1
 */
class Struct {
    /**
     * @internal
     */
    static _toInflightType() {
        return core_1.InflightClient.forType(__filename, this.name);
    }
    /**
     * Converts a Json to a Struct
     *
     * @macro $self$._fromJson($args$)
     */
    static fromJson(json, options) {
        json;
        options;
        throw new Error("Macro");
    }
    /**
     * Converts a Json to a Struct, returning nil if the Json is not valid
     *
     * @macro $self$._tryFromJson($args$)
     */
    static tryFromJson(json) {
        json;
        throw new Error("Macro");
    }
    /**
     * Parse a Json string into a Struct
     *
     * @macro $self$._fromJson(JSON.parse($args$))
     */
    static parseJson(json) {
        json;
        throw new Error("Macro");
    }
    /**
     * Parse a Json string into a Struct, returning nil if the Json is not valid
     *
     * @macro $self$._tryParseJson($args$)
     */
    static tryParseJson(json) {
        json;
        throw new Error("Macro");
    }
    /**
     * Retrieve the schema for this struct
     * @macro $self$
     */
    static schema() {
        throw new Error("Macro");
    }
    /**
     * Create an instance of a StructSchema from a JsonSchema
     *
     * @internal
     */
    static _createJsonSchema(schema) {
        return new json_schema_1.JsonSchema(schema);
    }
    constructor() { }
}
exports.Struct = Struct;
_a = JSII_RTTI_SYMBOL_1;
Struct[_a] = { fqn: "@winglang/sdk.std.Struct", version: "0.0.0" };
