// @cdktf/provider-aws/lib/data-aws-devopsguru-resource-collection/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDevopsguruResourceCollection = exports.DataAwsDevopsguruResourceCollectionTagsList = exports.DataAwsDevopsguruResourceCollectionTagsOutputReference = exports.dataAwsDevopsguruResourceCollectionTagsToHclTerraform = exports.dataAwsDevopsguruResourceCollectionTagsToTerraform = exports.DataAwsDevopsguruResourceCollectionCloudformationList = exports.DataAwsDevopsguruResourceCollectionCloudformationOutputReference = exports.dataAwsDevopsguruResourceCollectionCloudformationToHclTerraform = exports.dataAwsDevopsguruResourceCollectionCloudformationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsDevopsguruResourceCollectionCloudformationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDevopsguruResourceCollectionCloudformationToTerraform = dataAwsDevopsguruResourceCollectionCloudformationToTerraform;
function dataAwsDevopsguruResourceCollectionCloudformationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDevopsguruResourceCollectionCloudformationToHclTerraform = dataAwsDevopsguruResourceCollectionCloudformationToHclTerraform;
class DataAwsDevopsguruResourceCollectionCloudformationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // stack_names - computed: true, optional: false, required: false
    get stackNames() {
        return this.getListAttribute('stack_names');
    }
}
exports.DataAwsDevopsguruResourceCollectionCloudformationOutputReference = DataAwsDevopsguruResourceCollectionCloudformationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruResourceCollectionCloudformationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruResourceCollection.DataAwsDevopsguruResourceCollectionCloudformationOutputReference", version: "19.15.0" };
class DataAwsDevopsguruResourceCollectionCloudformationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDevopsguruResourceCollectionCloudformationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDevopsguruResourceCollectionCloudformationList = DataAwsDevopsguruResourceCollectionCloudformationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruResourceCollectionCloudformationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruResourceCollection.DataAwsDevopsguruResourceCollectionCloudformationList", version: "19.15.0" };
function dataAwsDevopsguruResourceCollectionTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDevopsguruResourceCollectionTagsToTerraform = dataAwsDevopsguruResourceCollectionTagsToTerraform;
function dataAwsDevopsguruResourceCollectionTagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDevopsguruResourceCollectionTagsToHclTerraform = dataAwsDevopsguruResourceCollectionTagsToHclTerraform;
class DataAwsDevopsguruResourceCollectionTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // app_boundary_key - computed: true, optional: false, required: false
    get appBoundaryKey() {
        return this.getStringAttribute('app_boundary_key');
    }
    // tag_values - computed: true, optional: false, required: false
    get tagValues() {
        return this.getListAttribute('tag_values');
    }
}
exports.DataAwsDevopsguruResourceCollectionTagsOutputReference = DataAwsDevopsguruResourceCollectionTagsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruResourceCollectionTagsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruResourceCollection.DataAwsDevopsguruResourceCollectionTagsOutputReference", version: "19.15.0" };
class DataAwsDevopsguruResourceCollectionTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDevopsguruResourceCollectionTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDevopsguruResourceCollectionTagsList = DataAwsDevopsguruResourceCollectionTagsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruResourceCollectionTagsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruResourceCollection.DataAwsDevopsguruResourceCollectionTagsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_resource_collection aws_devopsguru_resource_collection}
*/
class DataAwsDevopsguruResourceCollection extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDevopsguruResourceCollection resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDevopsguruResourceCollection to import
    * @param importFromId The id of the existing DataAwsDevopsguruResourceCollection that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_resource_collection#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDevopsguruResourceCollection to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_devopsguru_resource_collection", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/devopsguru_resource_collection aws_devopsguru_resource_collection} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDevopsguruResourceCollectionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_devopsguru_resource_collection',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cloudformation - computed: false, optional: true, required: false
        this._cloudformation = new DataAwsDevopsguruResourceCollectionCloudformationList(this, "cloudformation", false);
        // tags - computed: false, optional: true, required: false
        this._tags = new DataAwsDevopsguruResourceCollectionTagsList(this, "tags", false);
        this._type = config.type;
        this._cloudformation.internalValue = config.cloudformation;
        this._tags.internalValue = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get cloudformation() {
        return this._cloudformation;
    }
    putCloudformation(value) {
        this._cloudformation.internalValue = value;
    }
    resetCloudformation() {
        this._cloudformation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudformationInput() {
        return this._cloudformation.internalValue;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            type: cdktf.stringToTerraform(this._type),
            cloudformation: cdktf.listMapper(dataAwsDevopsguruResourceCollectionCloudformationToTerraform, true)(this._cloudformation.internalValue),
            tags: cdktf.listMapper(dataAwsDevopsguruResourceCollectionTagsToTerraform, true)(this._tags.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloudformation: {
                value: cdktf.listMapperHcl(dataAwsDevopsguruResourceCollectionCloudformationToHclTerraform, true)(this._cloudformation.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsDevopsguruResourceCollectionCloudformationList",
            },
            tags: {
                value: cdktf.listMapperHcl(dataAwsDevopsguruResourceCollectionTagsToHclTerraform, true)(this._tags.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsDevopsguruResourceCollectionTagsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDevopsguruResourceCollection = DataAwsDevopsguruResourceCollection;
_e = JSII_RTTI_SYMBOL_1;
DataAwsDevopsguruResourceCollection[_e] = { fqn: "@cdktf/provider-aws.dataAwsDevopsguruResourceCollection.DataAwsDevopsguruResourceCollection", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDevopsguruResourceCollection.tfResourceType = "aws_devopsguru_resource_collection";
