// @cdktf/provider-aws/lib/msk-scram-secret-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskScramSecretAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/msk_scram_secret_association aws_msk_scram_secret_association}
*/
class MskScramSecretAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a MskScramSecretAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the MskScramSecretAssociation to import
    * @param importFromId The id of the existing MskScramSecretAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/msk_scram_secret_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the MskScramSecretAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_msk_scram_secret_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/msk_scram_secret_association aws_msk_scram_secret_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskScramSecretAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_scram_secret_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._clusterArn = config.clusterArn;
        this._id = config.id;
        this._secretArnList = config.secretArnList;
    }
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterArnInput() {
        return this._clusterArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get secretArnList() {
        return cdktf.Fn.tolist(this.getListAttribute('secret_arn_list'));
    }
    set secretArnList(value) {
        this._secretArnList = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretArnListInput() {
        return this._secretArnList;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
            id: cdktf.stringToTerraform(this._id),
            secret_arn_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._secretArnList),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_arn: {
                value: cdktf.stringToHclTerraform(this._clusterArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secret_arn_list: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._secretArnList),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.MskScramSecretAssociation = MskScramSecretAssociation;
_a = JSII_RTTI_SYMBOL_1;
MskScramSecretAssociation[_a] = { fqn: "@cdktf/provider-aws.mskScramSecretAssociation.MskScramSecretAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
MskScramSecretAssociation.tfResourceType = "aws_msk_scram_secret_association";
