// @cdktf/provider-aws/lib/iam-security-token-service-preferences/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamSecurityTokenServicePreferences = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_security_token_service_preferences aws_iam_security_token_service_preferences}
*/
class IamSecurityTokenServicePreferences extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamSecurityTokenServicePreferences resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamSecurityTokenServicePreferences to import
    * @param importFromId The id of the existing IamSecurityTokenServicePreferences that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_security_token_service_preferences#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamSecurityTokenServicePreferences to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iam_security_token_service_preferences", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iam_security_token_service_preferences aws_iam_security_token_service_preferences} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamSecurityTokenServicePreferencesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_security_token_service_preferences',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._globalEndpointTokenVersion = config.globalEndpointTokenVersion;
        this._id = config.id;
    }
    get globalEndpointTokenVersion() {
        return this.getStringAttribute('global_endpoint_token_version');
    }
    set globalEndpointTokenVersion(value) {
        this._globalEndpointTokenVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get globalEndpointTokenVersionInput() {
        return this._globalEndpointTokenVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            global_endpoint_token_version: cdktf.stringToTerraform(this._globalEndpointTokenVersion),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            global_endpoint_token_version: {
                value: cdktf.stringToHclTerraform(this._globalEndpointTokenVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamSecurityTokenServicePreferences = IamSecurityTokenServicePreferences;
_a = JSII_RTTI_SYMBOL_1;
IamSecurityTokenServicePreferences[_a] = { fqn: "@cdktf/provider-aws.iamSecurityTokenServicePreferences.IamSecurityTokenServicePreferences", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IamSecurityTokenServicePreferences.tfResourceType = "aws_iam_security_token_service_preferences";
