// cdktf/lib/backends/consul-backend.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTerraformRemoteStateConsul = exports.ConsulBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_backend_1 = require("../terraform-backend");
const util_1 = require("../util");
const terraform_remote_state_1 = require("../terraform-remote-state");
// eslint-disable-next-line jsdoc/require-jsdoc
class ConsulBackend extends terraform_backend_1.TerraformBackend {
    constructor(scope, props) {
        super(scope, "backend", "consul");
        this.props = props;
    }
    synthesizeHclAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    synthesizeAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    getRemoteStateDataSource(scope, name, _fromStack) {
        return new DataTerraformRemoteStateConsul(scope, name, {
            ...this.props,
            workspace: "${terraform.workspace}",
        });
    }
}
exports.ConsulBackend = ConsulBackend;
_a = JSII_RTTI_SYMBOL_1;
ConsulBackend[_a] = { fqn: "cdktf.ConsulBackend", version: "0.20.7" };
// eslint-disable-next-line jsdoc/require-jsdoc
class DataTerraformRemoteStateConsul extends terraform_remote_state_1.TerraformRemoteState {
    constructor(scope, id, config) {
        super(scope, id, "consul", config);
    }
}
exports.DataTerraformRemoteStateConsul = DataTerraformRemoteStateConsul;
_b = JSII_RTTI_SYMBOL_1;
DataTerraformRemoteStateConsul[_b] = { fqn: "cdktf.DataTerraformRemoteStateConsul", version: "0.20.7" };
