// @cdktf/provider-aws/lib/data-aws-availability-zones/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAvailabilityZones = exports.DataAwsAvailabilityZonesTimeoutsOutputReference = exports.dataAwsAvailabilityZonesTimeoutsToHclTerraform = exports.dataAwsAvailabilityZonesTimeoutsToTerraform = exports.DataAwsAvailabilityZonesFilterList = exports.DataAwsAvailabilityZonesFilterOutputReference = exports.dataAwsAvailabilityZonesFilterToHclTerraform = exports.dataAwsAvailabilityZonesFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAvailabilityZonesFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsAvailabilityZonesFilterToTerraform = dataAwsAvailabilityZonesFilterToTerraform;
function dataAwsAvailabilityZonesFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsAvailabilityZonesFilterToHclTerraform = dataAwsAvailabilityZonesFilterToHclTerraform;
class DataAwsAvailabilityZonesFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsAvailabilityZonesFilterOutputReference = DataAwsAvailabilityZonesFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZonesFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsAvailabilityZones.DataAwsAvailabilityZonesFilterOutputReference", version: "19.15.0" };
class DataAwsAvailabilityZonesFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAvailabilityZonesFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAvailabilityZonesFilterList = DataAwsAvailabilityZonesFilterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZonesFilterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsAvailabilityZones.DataAwsAvailabilityZonesFilterList", version: "19.15.0" };
function dataAwsAvailabilityZonesTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsAvailabilityZonesTimeoutsToTerraform = dataAwsAvailabilityZonesTimeoutsToTerraform;
function dataAwsAvailabilityZonesTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsAvailabilityZonesTimeoutsToHclTerraform = dataAwsAvailabilityZonesTimeoutsToHclTerraform;
class DataAwsAvailabilityZonesTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsAvailabilityZonesTimeoutsOutputReference = DataAwsAvailabilityZonesTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZonesTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsAvailabilityZones.DataAwsAvailabilityZonesTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/availability_zones aws_availability_zones}
*/
class DataAwsAvailabilityZones extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsAvailabilityZones resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsAvailabilityZones to import
    * @param importFromId The id of the existing DataAwsAvailabilityZones that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/availability_zones#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsAvailabilityZones to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_availability_zones", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/availability_zones aws_availability_zones} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAvailabilityZonesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_availability_zones',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsAvailabilityZonesFilterList(this, "filter", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsAvailabilityZonesTimeoutsOutputReference(this, "timeouts");
        this._allAvailabilityZones = config.allAvailabilityZones;
        this._excludeNames = config.excludeNames;
        this._excludeZoneIds = config.excludeZoneIds;
        this._id = config.id;
        this._state = config.state;
        this._filter.internalValue = config.filter;
        this._timeouts.internalValue = config.timeouts;
    }
    get allAvailabilityZones() {
        return this.getBooleanAttribute('all_availability_zones');
    }
    set allAvailabilityZones(value) {
        this._allAvailabilityZones = value;
    }
    resetAllAvailabilityZones() {
        this._allAvailabilityZones = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allAvailabilityZonesInput() {
        return this._allAvailabilityZones;
    }
    get excludeNames() {
        return cdktf.Fn.tolist(this.getListAttribute('exclude_names'));
    }
    set excludeNames(value) {
        this._excludeNames = value;
    }
    resetExcludeNames() {
        this._excludeNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeNamesInput() {
        return this._excludeNames;
    }
    get excludeZoneIds() {
        return cdktf.Fn.tolist(this.getListAttribute('exclude_zone_ids'));
    }
    set excludeZoneIds(value) {
        this._excludeZoneIds = value;
    }
    resetExcludeZoneIds() {
        this._excludeZoneIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeZoneIdsInput() {
        return this._excludeZoneIds;
    }
    // group_names - computed: true, optional: false, required: false
    get groupNames() {
        return cdktf.Fn.tolist(this.getListAttribute('group_names'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // names - computed: true, optional: false, required: false
    get names() {
        return this.getListAttribute('names');
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    // zone_ids - computed: true, optional: false, required: false
    get zoneIds() {
        return this.getListAttribute('zone_ids');
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            all_availability_zones: cdktf.booleanToTerraform(this._allAvailabilityZones),
            exclude_names: cdktf.listMapper(cdktf.stringToTerraform, false)(this._excludeNames),
            exclude_zone_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._excludeZoneIds),
            id: cdktf.stringToTerraform(this._id),
            state: cdktf.stringToTerraform(this._state),
            filter: cdktf.listMapper(dataAwsAvailabilityZonesFilterToTerraform, true)(this._filter.internalValue),
            timeouts: dataAwsAvailabilityZonesTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            all_availability_zones: {
                value: cdktf.booleanToHclTerraform(this._allAvailabilityZones),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            exclude_names: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._excludeNames),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            exclude_zone_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._excludeZoneIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            state: {
                value: cdktf.stringToHclTerraform(this._state),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filter: {
                value: cdktf.listMapperHcl(dataAwsAvailabilityZonesFilterToHclTerraform, true)(this._filter.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsAvailabilityZonesFilterList",
            },
            timeouts: {
                value: dataAwsAvailabilityZonesTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsAvailabilityZonesTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsAvailabilityZones = DataAwsAvailabilityZones;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZones[_d] = { fqn: "@cdktf/provider-aws.dataAwsAvailabilityZones.DataAwsAvailabilityZones", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAvailabilityZones.tfResourceType = "aws_availability_zones";
