// @cdktf/provider-aws/lib/eks-access-policy-association/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksAccessPolicyAssociation = exports.EksAccessPolicyAssociationTimeoutsOutputReference = exports.eksAccessPolicyAssociationTimeoutsToHclTerraform = exports.eksAccessPolicyAssociationTimeoutsToTerraform = exports.EksAccessPolicyAssociationAccessScopeOutputReference = exports.eksAccessPolicyAssociationAccessScopeToHclTerraform = exports.eksAccessPolicyAssociationAccessScopeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function eksAccessPolicyAssociationAccessScopeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        namespaces: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.namespaces),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.eksAccessPolicyAssociationAccessScopeToTerraform = eksAccessPolicyAssociationAccessScopeToTerraform;
function eksAccessPolicyAssociationAccessScopeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        namespaces: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.namespaces),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.eksAccessPolicyAssociationAccessScopeToHclTerraform = eksAccessPolicyAssociationAccessScopeToHclTerraform;
class EksAccessPolicyAssociationAccessScopeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namespaces !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespaces = this._namespaces;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namespaces = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namespaces = value.namespaces;
            this._type = value.type;
        }
    }
    get namespaces() {
        return cdktf.Fn.tolist(this.getListAttribute('namespaces'));
    }
    set namespaces(value) {
        this._namespaces = value;
    }
    resetNamespaces() {
        this._namespaces = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespacesInput() {
        return this._namespaces;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EksAccessPolicyAssociationAccessScopeOutputReference = EksAccessPolicyAssociationAccessScopeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EksAccessPolicyAssociationAccessScopeOutputReference[_a] = { fqn: "@cdktf/provider-aws.eksAccessPolicyAssociation.EksAccessPolicyAssociationAccessScopeOutputReference", version: "19.15.0" };
function eksAccessPolicyAssociationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.eksAccessPolicyAssociationTimeoutsToTerraform = eksAccessPolicyAssociationTimeoutsToTerraform;
function eksAccessPolicyAssociationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.eksAccessPolicyAssociationTimeoutsToHclTerraform = eksAccessPolicyAssociationTimeoutsToHclTerraform;
class EksAccessPolicyAssociationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.EksAccessPolicyAssociationTimeoutsOutputReference = EksAccessPolicyAssociationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EksAccessPolicyAssociationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.eksAccessPolicyAssociation.EksAccessPolicyAssociationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_access_policy_association aws_eks_access_policy_association}
*/
class EksAccessPolicyAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EksAccessPolicyAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EksAccessPolicyAssociation to import
    * @param importFromId The id of the existing EksAccessPolicyAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_access_policy_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EksAccessPolicyAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_eks_access_policy_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_access_policy_association aws_eks_access_policy_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EksAccessPolicyAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_access_policy_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_scope - computed: false, optional: false, required: true
        this._accessScope = new EksAccessPolicyAssociationAccessScopeOutputReference(this, "access_scope");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksAccessPolicyAssociationTimeoutsOutputReference(this, "timeouts");
        this._clusterName = config.clusterName;
        this._id = config.id;
        this._policyArn = config.policyArn;
        this._principalArn = config.principalArn;
        this._accessScope.internalValue = config.accessScope;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // associated_at - computed: true, optional: false, required: false
    get associatedAt() {
        return this.getStringAttribute('associated_at');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // modified_at - computed: true, optional: false, required: false
    get modifiedAt() {
        return this.getStringAttribute('modified_at');
    }
    get policyArn() {
        return this.getStringAttribute('policy_arn');
    }
    set policyArn(value) {
        this._policyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyArnInput() {
        return this._policyArn;
    }
    get principalArn() {
        return this.getStringAttribute('principal_arn');
    }
    set principalArn(value) {
        this._principalArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get principalArnInput() {
        return this._principalArn;
    }
    get accessScope() {
        return this._accessScope;
    }
    putAccessScope(value) {
        this._accessScope.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessScopeInput() {
        return this._accessScope.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            id: cdktf.stringToTerraform(this._id),
            policy_arn: cdktf.stringToTerraform(this._policyArn),
            principal_arn: cdktf.stringToTerraform(this._principalArn),
            access_scope: eksAccessPolicyAssociationAccessScopeToTerraform(this._accessScope.internalValue),
            timeouts: eksAccessPolicyAssociationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_name: {
                value: cdktf.stringToHclTerraform(this._clusterName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy_arn: {
                value: cdktf.stringToHclTerraform(this._policyArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            principal_arn: {
                value: cdktf.stringToHclTerraform(this._principalArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_scope: {
                value: eksAccessPolicyAssociationAccessScopeToHclTerraform(this._accessScope.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EksAccessPolicyAssociationAccessScopeList",
            },
            timeouts: {
                value: eksAccessPolicyAssociationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "EksAccessPolicyAssociationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EksAccessPolicyAssociation = EksAccessPolicyAssociation;
_c = JSII_RTTI_SYMBOL_1;
EksAccessPolicyAssociation[_c] = { fqn: "@cdktf/provider-aws.eksAccessPolicyAssociation.EksAccessPolicyAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EksAccessPolicyAssociation.tfResourceType = "aws_eks_access_policy_association";
