// @winglang/sdk/lib/.gen/providers/google/provider/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/google/5.10.0/docs
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleProvider = exports.googleProviderBatchingToHclTerraform = exports.googleProviderBatchingToTerraform = void 0;
const cdktf = __importStar(require("cdktf"));
function googleProviderBatchingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_batching: cdktf.booleanToTerraform(struct.enableBatching),
        send_after: cdktf.stringToTerraform(struct.sendAfter),
    };
}
exports.googleProviderBatchingToTerraform = googleProviderBatchingToTerraform;
function googleProviderBatchingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_batching: {
            value: cdktf.booleanToHclTerraform(struct.enableBatching),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        send_after: {
            value: cdktf.stringToHclTerraform(struct.sendAfter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.googleProviderBatchingToHclTerraform = googleProviderBatchingToHclTerraform;
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/5.10.0/docs google}
*/
class GoogleProvider extends cdktf.TerraformProvider {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GoogleProvider resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GoogleProvider to import
    * @param importFromId The id of the existing GoogleProvider that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/5.10.0/docs#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GoogleProvider to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/5.10.0/docs google} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GoogleProviderConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '5.10.0',
                providerVersionConstraint: '~>5.10.0'
            },
            terraformProviderSource: 'google'
        });
        this._accessApprovalCustomEndpoint = config.accessApprovalCustomEndpoint;
        this._accessContextManagerCustomEndpoint = config.accessContextManagerCustomEndpoint;
        this._accessToken = config.accessToken;
        this._activeDirectoryCustomEndpoint = config.activeDirectoryCustomEndpoint;
        this._alloydbCustomEndpoint = config.alloydbCustomEndpoint;
        this._apigeeCustomEndpoint = config.apigeeCustomEndpoint;
        this._apikeysCustomEndpoint = config.apikeysCustomEndpoint;
        this._appEngineCustomEndpoint = config.appEngineCustomEndpoint;
        this._artifactRegistryCustomEndpoint = config.artifactRegistryCustomEndpoint;
        this._assuredWorkloadsCustomEndpoint = config.assuredWorkloadsCustomEndpoint;
        this._beyondcorpCustomEndpoint = config.beyondcorpCustomEndpoint;
        this._bigQueryCustomEndpoint = config.bigQueryCustomEndpoint;
        this._biglakeCustomEndpoint = config.biglakeCustomEndpoint;
        this._bigqueryAnalyticsHubCustomEndpoint = config.bigqueryAnalyticsHubCustomEndpoint;
        this._bigqueryConnectionCustomEndpoint = config.bigqueryConnectionCustomEndpoint;
        this._bigqueryDataTransferCustomEndpoint = config.bigqueryDataTransferCustomEndpoint;
        this._bigqueryDatapolicyCustomEndpoint = config.bigqueryDatapolicyCustomEndpoint;
        this._bigqueryReservationCustomEndpoint = config.bigqueryReservationCustomEndpoint;
        this._bigtableCustomEndpoint = config.bigtableCustomEndpoint;
        this._billingCustomEndpoint = config.billingCustomEndpoint;
        this._billingProject = config.billingProject;
        this._binaryAuthorizationCustomEndpoint = config.binaryAuthorizationCustomEndpoint;
        this._certificateManagerCustomEndpoint = config.certificateManagerCustomEndpoint;
        this._cloudAssetCustomEndpoint = config.cloudAssetCustomEndpoint;
        this._cloudBillingCustomEndpoint = config.cloudBillingCustomEndpoint;
        this._cloudBuildCustomEndpoint = config.cloudBuildCustomEndpoint;
        this._cloudBuildWorkerPoolCustomEndpoint = config.cloudBuildWorkerPoolCustomEndpoint;
        this._cloudFunctionsCustomEndpoint = config.cloudFunctionsCustomEndpoint;
        this._cloudIdentityCustomEndpoint = config.cloudIdentityCustomEndpoint;
        this._cloudIdsCustomEndpoint = config.cloudIdsCustomEndpoint;
        this._cloudResourceManagerCustomEndpoint = config.cloudResourceManagerCustomEndpoint;
        this._cloudRunCustomEndpoint = config.cloudRunCustomEndpoint;
        this._cloudRunV2CustomEndpoint = config.cloudRunV2CustomEndpoint;
        this._cloudSchedulerCustomEndpoint = config.cloudSchedulerCustomEndpoint;
        this._cloudTasksCustomEndpoint = config.cloudTasksCustomEndpoint;
        this._cloudbuildv2CustomEndpoint = config.cloudbuildv2CustomEndpoint;
        this._clouddeployCustomEndpoint = config.clouddeployCustomEndpoint;
        this._cloudfunctions2CustomEndpoint = config.cloudfunctions2CustomEndpoint;
        this._composerCustomEndpoint = config.composerCustomEndpoint;
        this._computeCustomEndpoint = config.computeCustomEndpoint;
        this._containerAnalysisCustomEndpoint = config.containerAnalysisCustomEndpoint;
        this._containerAttachedCustomEndpoint = config.containerAttachedCustomEndpoint;
        this._containerAwsCustomEndpoint = config.containerAwsCustomEndpoint;
        this._containerAzureCustomEndpoint = config.containerAzureCustomEndpoint;
        this._containerCustomEndpoint = config.containerCustomEndpoint;
        this._coreBillingCustomEndpoint = config.coreBillingCustomEndpoint;
        this._credentials = config.credentials;
        this._dataCatalogCustomEndpoint = config.dataCatalogCustomEndpoint;
        this._dataFusionCustomEndpoint = config.dataFusionCustomEndpoint;
        this._dataLossPreventionCustomEndpoint = config.dataLossPreventionCustomEndpoint;
        this._dataPipelineCustomEndpoint = config.dataPipelineCustomEndpoint;
        this._databaseMigrationServiceCustomEndpoint = config.databaseMigrationServiceCustomEndpoint;
        this._dataflowCustomEndpoint = config.dataflowCustomEndpoint;
        this._dataplexCustomEndpoint = config.dataplexCustomEndpoint;
        this._dataprocCustomEndpoint = config.dataprocCustomEndpoint;
        this._dataprocMetastoreCustomEndpoint = config.dataprocMetastoreCustomEndpoint;
        this._datastoreCustomEndpoint = config.datastoreCustomEndpoint;
        this._datastreamCustomEndpoint = config.datastreamCustomEndpoint;
        this._defaultLabels = config.defaultLabels;
        this._deploymentManagerCustomEndpoint = config.deploymentManagerCustomEndpoint;
        this._dialogflowCustomEndpoint = config.dialogflowCustomEndpoint;
        this._dialogflowCxCustomEndpoint = config.dialogflowCxCustomEndpoint;
        this._dnsCustomEndpoint = config.dnsCustomEndpoint;
        this._documentAiCustomEndpoint = config.documentAiCustomEndpoint;
        this._documentAiWarehouseCustomEndpoint = config.documentAiWarehouseCustomEndpoint;
        this._edgecontainerCustomEndpoint = config.edgecontainerCustomEndpoint;
        this._edgenetworkCustomEndpoint = config.edgenetworkCustomEndpoint;
        this._essentialContactsCustomEndpoint = config.essentialContactsCustomEndpoint;
        this._eventarcCustomEndpoint = config.eventarcCustomEndpoint;
        this._filestoreCustomEndpoint = config.filestoreCustomEndpoint;
        this._firebaserulesCustomEndpoint = config.firebaserulesCustomEndpoint;
        this._firestoreCustomEndpoint = config.firestoreCustomEndpoint;
        this._gkeBackupCustomEndpoint = config.gkeBackupCustomEndpoint;
        this._gkeHub2CustomEndpoint = config.gkeHub2CustomEndpoint;
        this._gkeHubCustomEndpoint = config.gkeHubCustomEndpoint;
        this._gkehubFeatureCustomEndpoint = config.gkehubFeatureCustomEndpoint;
        this._gkeonpremCustomEndpoint = config.gkeonpremCustomEndpoint;
        this._healthcareCustomEndpoint = config.healthcareCustomEndpoint;
        this._iam2CustomEndpoint = config.iam2CustomEndpoint;
        this._iamBetaCustomEndpoint = config.iamBetaCustomEndpoint;
        this._iamCredentialsCustomEndpoint = config.iamCredentialsCustomEndpoint;
        this._iamCustomEndpoint = config.iamCustomEndpoint;
        this._iamWorkforcePoolCustomEndpoint = config.iamWorkforcePoolCustomEndpoint;
        this._iapCustomEndpoint = config.iapCustomEndpoint;
        this._identityPlatformCustomEndpoint = config.identityPlatformCustomEndpoint;
        this._impersonateServiceAccount = config.impersonateServiceAccount;
        this._impersonateServiceAccountDelegates = config.impersonateServiceAccountDelegates;
        this._integrationConnectorsCustomEndpoint = config.integrationConnectorsCustomEndpoint;
        this._kmsCustomEndpoint = config.kmsCustomEndpoint;
        this._loggingCustomEndpoint = config.loggingCustomEndpoint;
        this._lookerCustomEndpoint = config.lookerCustomEndpoint;
        this._memcacheCustomEndpoint = config.memcacheCustomEndpoint;
        this._migrationCenterCustomEndpoint = config.migrationCenterCustomEndpoint;
        this._mlEngineCustomEndpoint = config.mlEngineCustomEndpoint;
        this._monitoringCustomEndpoint = config.monitoringCustomEndpoint;
        this._netappCustomEndpoint = config.netappCustomEndpoint;
        this._networkConnectivityCustomEndpoint = config.networkConnectivityCustomEndpoint;
        this._networkManagementCustomEndpoint = config.networkManagementCustomEndpoint;
        this._networkSecurityCustomEndpoint = config.networkSecurityCustomEndpoint;
        this._networkServicesCustomEndpoint = config.networkServicesCustomEndpoint;
        this._notebooksCustomEndpoint = config.notebooksCustomEndpoint;
        this._orgPolicyCustomEndpoint = config.orgPolicyCustomEndpoint;
        this._osConfigCustomEndpoint = config.osConfigCustomEndpoint;
        this._osLoginCustomEndpoint = config.osLoginCustomEndpoint;
        this._privatecaCustomEndpoint = config.privatecaCustomEndpoint;
        this._project = config.project;
        this._publicCaCustomEndpoint = config.publicCaCustomEndpoint;
        this._pubsubCustomEndpoint = config.pubsubCustomEndpoint;
        this._pubsubLiteCustomEndpoint = config.pubsubLiteCustomEndpoint;
        this._recaptchaEnterpriseCustomEndpoint = config.recaptchaEnterpriseCustomEndpoint;
        this._redisCustomEndpoint = config.redisCustomEndpoint;
        this._region = config.region;
        this._requestReason = config.requestReason;
        this._requestTimeout = config.requestTimeout;
        this._resourceManagerCustomEndpoint = config.resourceManagerCustomEndpoint;
        this._resourceManagerV3CustomEndpoint = config.resourceManagerV3CustomEndpoint;
        this._scopes = config.scopes;
        this._secretManagerCustomEndpoint = config.secretManagerCustomEndpoint;
        this._secureSourceManagerCustomEndpoint = config.secureSourceManagerCustomEndpoint;
        this._securityCenterCustomEndpoint = config.securityCenterCustomEndpoint;
        this._serviceManagementCustomEndpoint = config.serviceManagementCustomEndpoint;
        this._serviceNetworkingCustomEndpoint = config.serviceNetworkingCustomEndpoint;
        this._serviceUsageCustomEndpoint = config.serviceUsageCustomEndpoint;
        this._sourceRepoCustomEndpoint = config.sourceRepoCustomEndpoint;
        this._spannerCustomEndpoint = config.spannerCustomEndpoint;
        this._sqlCustomEndpoint = config.sqlCustomEndpoint;
        this._storageCustomEndpoint = config.storageCustomEndpoint;
        this._storageInsightsCustomEndpoint = config.storageInsightsCustomEndpoint;
        this._storageTransferCustomEndpoint = config.storageTransferCustomEndpoint;
        this._tagsCustomEndpoint = config.tagsCustomEndpoint;
        this._tagsLocationCustomEndpoint = config.tagsLocationCustomEndpoint;
        this._tpuCustomEndpoint = config.tpuCustomEndpoint;
        this._universeDomain = config.universeDomain;
        this._userProjectOverride = config.userProjectOverride;
        this._vertexAiCustomEndpoint = config.vertexAiCustomEndpoint;
        this._vmwareengineCustomEndpoint = config.vmwareengineCustomEndpoint;
        this._vpcAccessCustomEndpoint = config.vpcAccessCustomEndpoint;
        this._workbenchCustomEndpoint = config.workbenchCustomEndpoint;
        this._workflowsCustomEndpoint = config.workflowsCustomEndpoint;
        this._zone = config.zone;
        this._alias = config.alias;
        this._batching = config.batching;
    }
    get accessApprovalCustomEndpoint() {
        return this._accessApprovalCustomEndpoint;
    }
    set accessApprovalCustomEndpoint(value) {
        this._accessApprovalCustomEndpoint = value;
    }
    resetAccessApprovalCustomEndpoint() {
        this._accessApprovalCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessApprovalCustomEndpointInput() {
        return this._accessApprovalCustomEndpoint;
    }
    get accessContextManagerCustomEndpoint() {
        return this._accessContextManagerCustomEndpoint;
    }
    set accessContextManagerCustomEndpoint(value) {
        this._accessContextManagerCustomEndpoint = value;
    }
    resetAccessContextManagerCustomEndpoint() {
        this._accessContextManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessContextManagerCustomEndpointInput() {
        return this._accessContextManagerCustomEndpoint;
    }
    get accessToken() {
        return this._accessToken;
    }
    set accessToken(value) {
        this._accessToken = value;
    }
    resetAccessToken() {
        this._accessToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessTokenInput() {
        return this._accessToken;
    }
    get activeDirectoryCustomEndpoint() {
        return this._activeDirectoryCustomEndpoint;
    }
    set activeDirectoryCustomEndpoint(value) {
        this._activeDirectoryCustomEndpoint = value;
    }
    resetActiveDirectoryCustomEndpoint() {
        this._activeDirectoryCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get activeDirectoryCustomEndpointInput() {
        return this._activeDirectoryCustomEndpoint;
    }
    get alloydbCustomEndpoint() {
        return this._alloydbCustomEndpoint;
    }
    set alloydbCustomEndpoint(value) {
        this._alloydbCustomEndpoint = value;
    }
    resetAlloydbCustomEndpoint() {
        this._alloydbCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get alloydbCustomEndpointInput() {
        return this._alloydbCustomEndpoint;
    }
    get apigeeCustomEndpoint() {
        return this._apigeeCustomEndpoint;
    }
    set apigeeCustomEndpoint(value) {
        this._apigeeCustomEndpoint = value;
    }
    resetApigeeCustomEndpoint() {
        this._apigeeCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apigeeCustomEndpointInput() {
        return this._apigeeCustomEndpoint;
    }
    get apikeysCustomEndpoint() {
        return this._apikeysCustomEndpoint;
    }
    set apikeysCustomEndpoint(value) {
        this._apikeysCustomEndpoint = value;
    }
    resetApikeysCustomEndpoint() {
        this._apikeysCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apikeysCustomEndpointInput() {
        return this._apikeysCustomEndpoint;
    }
    get appEngineCustomEndpoint() {
        return this._appEngineCustomEndpoint;
    }
    set appEngineCustomEndpoint(value) {
        this._appEngineCustomEndpoint = value;
    }
    resetAppEngineCustomEndpoint() {
        this._appEngineCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get appEngineCustomEndpointInput() {
        return this._appEngineCustomEndpoint;
    }
    get artifactRegistryCustomEndpoint() {
        return this._artifactRegistryCustomEndpoint;
    }
    set artifactRegistryCustomEndpoint(value) {
        this._artifactRegistryCustomEndpoint = value;
    }
    resetArtifactRegistryCustomEndpoint() {
        this._artifactRegistryCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get artifactRegistryCustomEndpointInput() {
        return this._artifactRegistryCustomEndpoint;
    }
    get assuredWorkloadsCustomEndpoint() {
        return this._assuredWorkloadsCustomEndpoint;
    }
    set assuredWorkloadsCustomEndpoint(value) {
        this._assuredWorkloadsCustomEndpoint = value;
    }
    resetAssuredWorkloadsCustomEndpoint() {
        this._assuredWorkloadsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assuredWorkloadsCustomEndpointInput() {
        return this._assuredWorkloadsCustomEndpoint;
    }
    get beyondcorpCustomEndpoint() {
        return this._beyondcorpCustomEndpoint;
    }
    set beyondcorpCustomEndpoint(value) {
        this._beyondcorpCustomEndpoint = value;
    }
    resetBeyondcorpCustomEndpoint() {
        this._beyondcorpCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get beyondcorpCustomEndpointInput() {
        return this._beyondcorpCustomEndpoint;
    }
    get bigQueryCustomEndpoint() {
        return this._bigQueryCustomEndpoint;
    }
    set bigQueryCustomEndpoint(value) {
        this._bigQueryCustomEndpoint = value;
    }
    resetBigQueryCustomEndpoint() {
        this._bigQueryCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigQueryCustomEndpointInput() {
        return this._bigQueryCustomEndpoint;
    }
    get biglakeCustomEndpoint() {
        return this._biglakeCustomEndpoint;
    }
    set biglakeCustomEndpoint(value) {
        this._biglakeCustomEndpoint = value;
    }
    resetBiglakeCustomEndpoint() {
        this._biglakeCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get biglakeCustomEndpointInput() {
        return this._biglakeCustomEndpoint;
    }
    get bigqueryAnalyticsHubCustomEndpoint() {
        return this._bigqueryAnalyticsHubCustomEndpoint;
    }
    set bigqueryAnalyticsHubCustomEndpoint(value) {
        this._bigqueryAnalyticsHubCustomEndpoint = value;
    }
    resetBigqueryAnalyticsHubCustomEndpoint() {
        this._bigqueryAnalyticsHubCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryAnalyticsHubCustomEndpointInput() {
        return this._bigqueryAnalyticsHubCustomEndpoint;
    }
    get bigqueryConnectionCustomEndpoint() {
        return this._bigqueryConnectionCustomEndpoint;
    }
    set bigqueryConnectionCustomEndpoint(value) {
        this._bigqueryConnectionCustomEndpoint = value;
    }
    resetBigqueryConnectionCustomEndpoint() {
        this._bigqueryConnectionCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryConnectionCustomEndpointInput() {
        return this._bigqueryConnectionCustomEndpoint;
    }
    get bigqueryDataTransferCustomEndpoint() {
        return this._bigqueryDataTransferCustomEndpoint;
    }
    set bigqueryDataTransferCustomEndpoint(value) {
        this._bigqueryDataTransferCustomEndpoint = value;
    }
    resetBigqueryDataTransferCustomEndpoint() {
        this._bigqueryDataTransferCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryDataTransferCustomEndpointInput() {
        return this._bigqueryDataTransferCustomEndpoint;
    }
    get bigqueryDatapolicyCustomEndpoint() {
        return this._bigqueryDatapolicyCustomEndpoint;
    }
    set bigqueryDatapolicyCustomEndpoint(value) {
        this._bigqueryDatapolicyCustomEndpoint = value;
    }
    resetBigqueryDatapolicyCustomEndpoint() {
        this._bigqueryDatapolicyCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryDatapolicyCustomEndpointInput() {
        return this._bigqueryDatapolicyCustomEndpoint;
    }
    get bigqueryReservationCustomEndpoint() {
        return this._bigqueryReservationCustomEndpoint;
    }
    set bigqueryReservationCustomEndpoint(value) {
        this._bigqueryReservationCustomEndpoint = value;
    }
    resetBigqueryReservationCustomEndpoint() {
        this._bigqueryReservationCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryReservationCustomEndpointInput() {
        return this._bigqueryReservationCustomEndpoint;
    }
    get bigtableCustomEndpoint() {
        return this._bigtableCustomEndpoint;
    }
    set bigtableCustomEndpoint(value) {
        this._bigtableCustomEndpoint = value;
    }
    resetBigtableCustomEndpoint() {
        this._bigtableCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigtableCustomEndpointInput() {
        return this._bigtableCustomEndpoint;
    }
    get billingCustomEndpoint() {
        return this._billingCustomEndpoint;
    }
    set billingCustomEndpoint(value) {
        this._billingCustomEndpoint = value;
    }
    resetBillingCustomEndpoint() {
        this._billingCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get billingCustomEndpointInput() {
        return this._billingCustomEndpoint;
    }
    get billingProject() {
        return this._billingProject;
    }
    set billingProject(value) {
        this._billingProject = value;
    }
    resetBillingProject() {
        this._billingProject = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get billingProjectInput() {
        return this._billingProject;
    }
    get binaryAuthorizationCustomEndpoint() {
        return this._binaryAuthorizationCustomEndpoint;
    }
    set binaryAuthorizationCustomEndpoint(value) {
        this._binaryAuthorizationCustomEndpoint = value;
    }
    resetBinaryAuthorizationCustomEndpoint() {
        this._binaryAuthorizationCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get binaryAuthorizationCustomEndpointInput() {
        return this._binaryAuthorizationCustomEndpoint;
    }
    get certificateManagerCustomEndpoint() {
        return this._certificateManagerCustomEndpoint;
    }
    set certificateManagerCustomEndpoint(value) {
        this._certificateManagerCustomEndpoint = value;
    }
    resetCertificateManagerCustomEndpoint() {
        this._certificateManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificateManagerCustomEndpointInput() {
        return this._certificateManagerCustomEndpoint;
    }
    get cloudAssetCustomEndpoint() {
        return this._cloudAssetCustomEndpoint;
    }
    set cloudAssetCustomEndpoint(value) {
        this._cloudAssetCustomEndpoint = value;
    }
    resetCloudAssetCustomEndpoint() {
        this._cloudAssetCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudAssetCustomEndpointInput() {
        return this._cloudAssetCustomEndpoint;
    }
    get cloudBillingCustomEndpoint() {
        return this._cloudBillingCustomEndpoint;
    }
    set cloudBillingCustomEndpoint(value) {
        this._cloudBillingCustomEndpoint = value;
    }
    resetCloudBillingCustomEndpoint() {
        this._cloudBillingCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudBillingCustomEndpointInput() {
        return this._cloudBillingCustomEndpoint;
    }
    get cloudBuildCustomEndpoint() {
        return this._cloudBuildCustomEndpoint;
    }
    set cloudBuildCustomEndpoint(value) {
        this._cloudBuildCustomEndpoint = value;
    }
    resetCloudBuildCustomEndpoint() {
        this._cloudBuildCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudBuildCustomEndpointInput() {
        return this._cloudBuildCustomEndpoint;
    }
    get cloudBuildWorkerPoolCustomEndpoint() {
        return this._cloudBuildWorkerPoolCustomEndpoint;
    }
    set cloudBuildWorkerPoolCustomEndpoint(value) {
        this._cloudBuildWorkerPoolCustomEndpoint = value;
    }
    resetCloudBuildWorkerPoolCustomEndpoint() {
        this._cloudBuildWorkerPoolCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudBuildWorkerPoolCustomEndpointInput() {
        return this._cloudBuildWorkerPoolCustomEndpoint;
    }
    get cloudFunctionsCustomEndpoint() {
        return this._cloudFunctionsCustomEndpoint;
    }
    set cloudFunctionsCustomEndpoint(value) {
        this._cloudFunctionsCustomEndpoint = value;
    }
    resetCloudFunctionsCustomEndpoint() {
        this._cloudFunctionsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudFunctionsCustomEndpointInput() {
        return this._cloudFunctionsCustomEndpoint;
    }
    get cloudIdentityCustomEndpoint() {
        return this._cloudIdentityCustomEndpoint;
    }
    set cloudIdentityCustomEndpoint(value) {
        this._cloudIdentityCustomEndpoint = value;
    }
    resetCloudIdentityCustomEndpoint() {
        this._cloudIdentityCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudIdentityCustomEndpointInput() {
        return this._cloudIdentityCustomEndpoint;
    }
    get cloudIdsCustomEndpoint() {
        return this._cloudIdsCustomEndpoint;
    }
    set cloudIdsCustomEndpoint(value) {
        this._cloudIdsCustomEndpoint = value;
    }
    resetCloudIdsCustomEndpoint() {
        this._cloudIdsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudIdsCustomEndpointInput() {
        return this._cloudIdsCustomEndpoint;
    }
    get cloudResourceManagerCustomEndpoint() {
        return this._cloudResourceManagerCustomEndpoint;
    }
    set cloudResourceManagerCustomEndpoint(value) {
        this._cloudResourceManagerCustomEndpoint = value;
    }
    resetCloudResourceManagerCustomEndpoint() {
        this._cloudResourceManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudResourceManagerCustomEndpointInput() {
        return this._cloudResourceManagerCustomEndpoint;
    }
    get cloudRunCustomEndpoint() {
        return this._cloudRunCustomEndpoint;
    }
    set cloudRunCustomEndpoint(value) {
        this._cloudRunCustomEndpoint = value;
    }
    resetCloudRunCustomEndpoint() {
        this._cloudRunCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudRunCustomEndpointInput() {
        return this._cloudRunCustomEndpoint;
    }
    get cloudRunV2CustomEndpoint() {
        return this._cloudRunV2CustomEndpoint;
    }
    set cloudRunV2CustomEndpoint(value) {
        this._cloudRunV2CustomEndpoint = value;
    }
    resetCloudRunV2CustomEndpoint() {
        this._cloudRunV2CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudRunV2CustomEndpointInput() {
        return this._cloudRunV2CustomEndpoint;
    }
    get cloudSchedulerCustomEndpoint() {
        return this._cloudSchedulerCustomEndpoint;
    }
    set cloudSchedulerCustomEndpoint(value) {
        this._cloudSchedulerCustomEndpoint = value;
    }
    resetCloudSchedulerCustomEndpoint() {
        this._cloudSchedulerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSchedulerCustomEndpointInput() {
        return this._cloudSchedulerCustomEndpoint;
    }
    get cloudTasksCustomEndpoint() {
        return this._cloudTasksCustomEndpoint;
    }
    set cloudTasksCustomEndpoint(value) {
        this._cloudTasksCustomEndpoint = value;
    }
    resetCloudTasksCustomEndpoint() {
        this._cloudTasksCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudTasksCustomEndpointInput() {
        return this._cloudTasksCustomEndpoint;
    }
    get cloudbuildv2CustomEndpoint() {
        return this._cloudbuildv2CustomEndpoint;
    }
    set cloudbuildv2CustomEndpoint(value) {
        this._cloudbuildv2CustomEndpoint = value;
    }
    resetCloudbuildv2CustomEndpoint() {
        this._cloudbuildv2CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudbuildv2CustomEndpointInput() {
        return this._cloudbuildv2CustomEndpoint;
    }
    get clouddeployCustomEndpoint() {
        return this._clouddeployCustomEndpoint;
    }
    set clouddeployCustomEndpoint(value) {
        this._clouddeployCustomEndpoint = value;
    }
    resetClouddeployCustomEndpoint() {
        this._clouddeployCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clouddeployCustomEndpointInput() {
        return this._clouddeployCustomEndpoint;
    }
    get cloudfunctions2CustomEndpoint() {
        return this._cloudfunctions2CustomEndpoint;
    }
    set cloudfunctions2CustomEndpoint(value) {
        this._cloudfunctions2CustomEndpoint = value;
    }
    resetCloudfunctions2CustomEndpoint() {
        this._cloudfunctions2CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudfunctions2CustomEndpointInput() {
        return this._cloudfunctions2CustomEndpoint;
    }
    get composerCustomEndpoint() {
        return this._composerCustomEndpoint;
    }
    set composerCustomEndpoint(value) {
        this._composerCustomEndpoint = value;
    }
    resetComposerCustomEndpoint() {
        this._composerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get composerCustomEndpointInput() {
        return this._composerCustomEndpoint;
    }
    get computeCustomEndpoint() {
        return this._computeCustomEndpoint;
    }
    set computeCustomEndpoint(value) {
        this._computeCustomEndpoint = value;
    }
    resetComputeCustomEndpoint() {
        this._computeCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get computeCustomEndpointInput() {
        return this._computeCustomEndpoint;
    }
    get containerAnalysisCustomEndpoint() {
        return this._containerAnalysisCustomEndpoint;
    }
    set containerAnalysisCustomEndpoint(value) {
        this._containerAnalysisCustomEndpoint = value;
    }
    resetContainerAnalysisCustomEndpoint() {
        this._containerAnalysisCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerAnalysisCustomEndpointInput() {
        return this._containerAnalysisCustomEndpoint;
    }
    get containerAttachedCustomEndpoint() {
        return this._containerAttachedCustomEndpoint;
    }
    set containerAttachedCustomEndpoint(value) {
        this._containerAttachedCustomEndpoint = value;
    }
    resetContainerAttachedCustomEndpoint() {
        this._containerAttachedCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerAttachedCustomEndpointInput() {
        return this._containerAttachedCustomEndpoint;
    }
    get containerAwsCustomEndpoint() {
        return this._containerAwsCustomEndpoint;
    }
    set containerAwsCustomEndpoint(value) {
        this._containerAwsCustomEndpoint = value;
    }
    resetContainerAwsCustomEndpoint() {
        this._containerAwsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerAwsCustomEndpointInput() {
        return this._containerAwsCustomEndpoint;
    }
    get containerAzureCustomEndpoint() {
        return this._containerAzureCustomEndpoint;
    }
    set containerAzureCustomEndpoint(value) {
        this._containerAzureCustomEndpoint = value;
    }
    resetContainerAzureCustomEndpoint() {
        this._containerAzureCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerAzureCustomEndpointInput() {
        return this._containerAzureCustomEndpoint;
    }
    get containerCustomEndpoint() {
        return this._containerCustomEndpoint;
    }
    set containerCustomEndpoint(value) {
        this._containerCustomEndpoint = value;
    }
    resetContainerCustomEndpoint() {
        this._containerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerCustomEndpointInput() {
        return this._containerCustomEndpoint;
    }
    get coreBillingCustomEndpoint() {
        return this._coreBillingCustomEndpoint;
    }
    set coreBillingCustomEndpoint(value) {
        this._coreBillingCustomEndpoint = value;
    }
    resetCoreBillingCustomEndpoint() {
        this._coreBillingCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreBillingCustomEndpointInput() {
        return this._coreBillingCustomEndpoint;
    }
    get credentials() {
        return this._credentials;
    }
    set credentials(value) {
        this._credentials = value;
    }
    resetCredentials() {
        this._credentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get credentialsInput() {
        return this._credentials;
    }
    get dataCatalogCustomEndpoint() {
        return this._dataCatalogCustomEndpoint;
    }
    set dataCatalogCustomEndpoint(value) {
        this._dataCatalogCustomEndpoint = value;
    }
    resetDataCatalogCustomEndpoint() {
        this._dataCatalogCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataCatalogCustomEndpointInput() {
        return this._dataCatalogCustomEndpoint;
    }
    get dataFusionCustomEndpoint() {
        return this._dataFusionCustomEndpoint;
    }
    set dataFusionCustomEndpoint(value) {
        this._dataFusionCustomEndpoint = value;
    }
    resetDataFusionCustomEndpoint() {
        this._dataFusionCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataFusionCustomEndpointInput() {
        return this._dataFusionCustomEndpoint;
    }
    get dataLossPreventionCustomEndpoint() {
        return this._dataLossPreventionCustomEndpoint;
    }
    set dataLossPreventionCustomEndpoint(value) {
        this._dataLossPreventionCustomEndpoint = value;
    }
    resetDataLossPreventionCustomEndpoint() {
        this._dataLossPreventionCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataLossPreventionCustomEndpointInput() {
        return this._dataLossPreventionCustomEndpoint;
    }
    get dataPipelineCustomEndpoint() {
        return this._dataPipelineCustomEndpoint;
    }
    set dataPipelineCustomEndpoint(value) {
        this._dataPipelineCustomEndpoint = value;
    }
    resetDataPipelineCustomEndpoint() {
        this._dataPipelineCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataPipelineCustomEndpointInput() {
        return this._dataPipelineCustomEndpoint;
    }
    get databaseMigrationServiceCustomEndpoint() {
        return this._databaseMigrationServiceCustomEndpoint;
    }
    set databaseMigrationServiceCustomEndpoint(value) {
        this._databaseMigrationServiceCustomEndpoint = value;
    }
    resetDatabaseMigrationServiceCustomEndpoint() {
        this._databaseMigrationServiceCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseMigrationServiceCustomEndpointInput() {
        return this._databaseMigrationServiceCustomEndpoint;
    }
    get dataflowCustomEndpoint() {
        return this._dataflowCustomEndpoint;
    }
    set dataflowCustomEndpoint(value) {
        this._dataflowCustomEndpoint = value;
    }
    resetDataflowCustomEndpoint() {
        this._dataflowCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataflowCustomEndpointInput() {
        return this._dataflowCustomEndpoint;
    }
    get dataplexCustomEndpoint() {
        return this._dataplexCustomEndpoint;
    }
    set dataplexCustomEndpoint(value) {
        this._dataplexCustomEndpoint = value;
    }
    resetDataplexCustomEndpoint() {
        this._dataplexCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataplexCustomEndpointInput() {
        return this._dataplexCustomEndpoint;
    }
    get dataprocCustomEndpoint() {
        return this._dataprocCustomEndpoint;
    }
    set dataprocCustomEndpoint(value) {
        this._dataprocCustomEndpoint = value;
    }
    resetDataprocCustomEndpoint() {
        this._dataprocCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataprocCustomEndpointInput() {
        return this._dataprocCustomEndpoint;
    }
    get dataprocMetastoreCustomEndpoint() {
        return this._dataprocMetastoreCustomEndpoint;
    }
    set dataprocMetastoreCustomEndpoint(value) {
        this._dataprocMetastoreCustomEndpoint = value;
    }
    resetDataprocMetastoreCustomEndpoint() {
        this._dataprocMetastoreCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataprocMetastoreCustomEndpointInput() {
        return this._dataprocMetastoreCustomEndpoint;
    }
    get datastoreCustomEndpoint() {
        return this._datastoreCustomEndpoint;
    }
    set datastoreCustomEndpoint(value) {
        this._datastoreCustomEndpoint = value;
    }
    resetDatastoreCustomEndpoint() {
        this._datastoreCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get datastoreCustomEndpointInput() {
        return this._datastoreCustomEndpoint;
    }
    get datastreamCustomEndpoint() {
        return this._datastreamCustomEndpoint;
    }
    set datastreamCustomEndpoint(value) {
        this._datastreamCustomEndpoint = value;
    }
    resetDatastreamCustomEndpoint() {
        this._datastreamCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get datastreamCustomEndpointInput() {
        return this._datastreamCustomEndpoint;
    }
    get defaultLabels() {
        return this._defaultLabels;
    }
    set defaultLabels(value) {
        this._defaultLabels = value;
    }
    resetDefaultLabels() {
        this._defaultLabels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultLabelsInput() {
        return this._defaultLabels;
    }
    get deploymentManagerCustomEndpoint() {
        return this._deploymentManagerCustomEndpoint;
    }
    set deploymentManagerCustomEndpoint(value) {
        this._deploymentManagerCustomEndpoint = value;
    }
    resetDeploymentManagerCustomEndpoint() {
        this._deploymentManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentManagerCustomEndpointInput() {
        return this._deploymentManagerCustomEndpoint;
    }
    get dialogflowCustomEndpoint() {
        return this._dialogflowCustomEndpoint;
    }
    set dialogflowCustomEndpoint(value) {
        this._dialogflowCustomEndpoint = value;
    }
    resetDialogflowCustomEndpoint() {
        this._dialogflowCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dialogflowCustomEndpointInput() {
        return this._dialogflowCustomEndpoint;
    }
    get dialogflowCxCustomEndpoint() {
        return this._dialogflowCxCustomEndpoint;
    }
    set dialogflowCxCustomEndpoint(value) {
        this._dialogflowCxCustomEndpoint = value;
    }
    resetDialogflowCxCustomEndpoint() {
        this._dialogflowCxCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dialogflowCxCustomEndpointInput() {
        return this._dialogflowCxCustomEndpoint;
    }
    get dnsCustomEndpoint() {
        return this._dnsCustomEndpoint;
    }
    set dnsCustomEndpoint(value) {
        this._dnsCustomEndpoint = value;
    }
    resetDnsCustomEndpoint() {
        this._dnsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsCustomEndpointInput() {
        return this._dnsCustomEndpoint;
    }
    get documentAiCustomEndpoint() {
        return this._documentAiCustomEndpoint;
    }
    set documentAiCustomEndpoint(value) {
        this._documentAiCustomEndpoint = value;
    }
    resetDocumentAiCustomEndpoint() {
        this._documentAiCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentAiCustomEndpointInput() {
        return this._documentAiCustomEndpoint;
    }
    get documentAiWarehouseCustomEndpoint() {
        return this._documentAiWarehouseCustomEndpoint;
    }
    set documentAiWarehouseCustomEndpoint(value) {
        this._documentAiWarehouseCustomEndpoint = value;
    }
    resetDocumentAiWarehouseCustomEndpoint() {
        this._documentAiWarehouseCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentAiWarehouseCustomEndpointInput() {
        return this._documentAiWarehouseCustomEndpoint;
    }
    get edgecontainerCustomEndpoint() {
        return this._edgecontainerCustomEndpoint;
    }
    set edgecontainerCustomEndpoint(value) {
        this._edgecontainerCustomEndpoint = value;
    }
    resetEdgecontainerCustomEndpoint() {
        this._edgecontainerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get edgecontainerCustomEndpointInput() {
        return this._edgecontainerCustomEndpoint;
    }
    get edgenetworkCustomEndpoint() {
        return this._edgenetworkCustomEndpoint;
    }
    set edgenetworkCustomEndpoint(value) {
        this._edgenetworkCustomEndpoint = value;
    }
    resetEdgenetworkCustomEndpoint() {
        this._edgenetworkCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get edgenetworkCustomEndpointInput() {
        return this._edgenetworkCustomEndpoint;
    }
    get essentialContactsCustomEndpoint() {
        return this._essentialContactsCustomEndpoint;
    }
    set essentialContactsCustomEndpoint(value) {
        this._essentialContactsCustomEndpoint = value;
    }
    resetEssentialContactsCustomEndpoint() {
        this._essentialContactsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get essentialContactsCustomEndpointInput() {
        return this._essentialContactsCustomEndpoint;
    }
    get eventarcCustomEndpoint() {
        return this._eventarcCustomEndpoint;
    }
    set eventarcCustomEndpoint(value) {
        this._eventarcCustomEndpoint = value;
    }
    resetEventarcCustomEndpoint() {
        this._eventarcCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventarcCustomEndpointInput() {
        return this._eventarcCustomEndpoint;
    }
    get filestoreCustomEndpoint() {
        return this._filestoreCustomEndpoint;
    }
    set filestoreCustomEndpoint(value) {
        this._filestoreCustomEndpoint = value;
    }
    resetFilestoreCustomEndpoint() {
        this._filestoreCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filestoreCustomEndpointInput() {
        return this._filestoreCustomEndpoint;
    }
    get firebaserulesCustomEndpoint() {
        return this._firebaserulesCustomEndpoint;
    }
    set firebaserulesCustomEndpoint(value) {
        this._firebaserulesCustomEndpoint = value;
    }
    resetFirebaserulesCustomEndpoint() {
        this._firebaserulesCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firebaserulesCustomEndpointInput() {
        return this._firebaserulesCustomEndpoint;
    }
    get firestoreCustomEndpoint() {
        return this._firestoreCustomEndpoint;
    }
    set firestoreCustomEndpoint(value) {
        this._firestoreCustomEndpoint = value;
    }
    resetFirestoreCustomEndpoint() {
        this._firestoreCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firestoreCustomEndpointInput() {
        return this._firestoreCustomEndpoint;
    }
    get gkeBackupCustomEndpoint() {
        return this._gkeBackupCustomEndpoint;
    }
    set gkeBackupCustomEndpoint(value) {
        this._gkeBackupCustomEndpoint = value;
    }
    resetGkeBackupCustomEndpoint() {
        this._gkeBackupCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gkeBackupCustomEndpointInput() {
        return this._gkeBackupCustomEndpoint;
    }
    get gkeHub2CustomEndpoint() {
        return this._gkeHub2CustomEndpoint;
    }
    set gkeHub2CustomEndpoint(value) {
        this._gkeHub2CustomEndpoint = value;
    }
    resetGkeHub2CustomEndpoint() {
        this._gkeHub2CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gkeHub2CustomEndpointInput() {
        return this._gkeHub2CustomEndpoint;
    }
    get gkeHubCustomEndpoint() {
        return this._gkeHubCustomEndpoint;
    }
    set gkeHubCustomEndpoint(value) {
        this._gkeHubCustomEndpoint = value;
    }
    resetGkeHubCustomEndpoint() {
        this._gkeHubCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gkeHubCustomEndpointInput() {
        return this._gkeHubCustomEndpoint;
    }
    get gkehubFeatureCustomEndpoint() {
        return this._gkehubFeatureCustomEndpoint;
    }
    set gkehubFeatureCustomEndpoint(value) {
        this._gkehubFeatureCustomEndpoint = value;
    }
    resetGkehubFeatureCustomEndpoint() {
        this._gkehubFeatureCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gkehubFeatureCustomEndpointInput() {
        return this._gkehubFeatureCustomEndpoint;
    }
    get gkeonpremCustomEndpoint() {
        return this._gkeonpremCustomEndpoint;
    }
    set gkeonpremCustomEndpoint(value) {
        this._gkeonpremCustomEndpoint = value;
    }
    resetGkeonpremCustomEndpoint() {
        this._gkeonpremCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gkeonpremCustomEndpointInput() {
        return this._gkeonpremCustomEndpoint;
    }
    get healthcareCustomEndpoint() {
        return this._healthcareCustomEndpoint;
    }
    set healthcareCustomEndpoint(value) {
        this._healthcareCustomEndpoint = value;
    }
    resetHealthcareCustomEndpoint() {
        this._healthcareCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthcareCustomEndpointInput() {
        return this._healthcareCustomEndpoint;
    }
    get iam2CustomEndpoint() {
        return this._iam2CustomEndpoint;
    }
    set iam2CustomEndpoint(value) {
        this._iam2CustomEndpoint = value;
    }
    resetIam2CustomEndpoint() {
        this._iam2CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iam2CustomEndpointInput() {
        return this._iam2CustomEndpoint;
    }
    get iamBetaCustomEndpoint() {
        return this._iamBetaCustomEndpoint;
    }
    set iamBetaCustomEndpoint(value) {
        this._iamBetaCustomEndpoint = value;
    }
    resetIamBetaCustomEndpoint() {
        this._iamBetaCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamBetaCustomEndpointInput() {
        return this._iamBetaCustomEndpoint;
    }
    get iamCredentialsCustomEndpoint() {
        return this._iamCredentialsCustomEndpoint;
    }
    set iamCredentialsCustomEndpoint(value) {
        this._iamCredentialsCustomEndpoint = value;
    }
    resetIamCredentialsCustomEndpoint() {
        this._iamCredentialsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamCredentialsCustomEndpointInput() {
        return this._iamCredentialsCustomEndpoint;
    }
    get iamCustomEndpoint() {
        return this._iamCustomEndpoint;
    }
    set iamCustomEndpoint(value) {
        this._iamCustomEndpoint = value;
    }
    resetIamCustomEndpoint() {
        this._iamCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamCustomEndpointInput() {
        return this._iamCustomEndpoint;
    }
    get iamWorkforcePoolCustomEndpoint() {
        return this._iamWorkforcePoolCustomEndpoint;
    }
    set iamWorkforcePoolCustomEndpoint(value) {
        this._iamWorkforcePoolCustomEndpoint = value;
    }
    resetIamWorkforcePoolCustomEndpoint() {
        this._iamWorkforcePoolCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamWorkforcePoolCustomEndpointInput() {
        return this._iamWorkforcePoolCustomEndpoint;
    }
    get iapCustomEndpoint() {
        return this._iapCustomEndpoint;
    }
    set iapCustomEndpoint(value) {
        this._iapCustomEndpoint = value;
    }
    resetIapCustomEndpoint() {
        this._iapCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iapCustomEndpointInput() {
        return this._iapCustomEndpoint;
    }
    get identityPlatformCustomEndpoint() {
        return this._identityPlatformCustomEndpoint;
    }
    set identityPlatformCustomEndpoint(value) {
        this._identityPlatformCustomEndpoint = value;
    }
    resetIdentityPlatformCustomEndpoint() {
        this._identityPlatformCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identityPlatformCustomEndpointInput() {
        return this._identityPlatformCustomEndpoint;
    }
    get impersonateServiceAccount() {
        return this._impersonateServiceAccount;
    }
    set impersonateServiceAccount(value) {
        this._impersonateServiceAccount = value;
    }
    resetImpersonateServiceAccount() {
        this._impersonateServiceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get impersonateServiceAccountInput() {
        return this._impersonateServiceAccount;
    }
    get impersonateServiceAccountDelegates() {
        return this._impersonateServiceAccountDelegates;
    }
    set impersonateServiceAccountDelegates(value) {
        this._impersonateServiceAccountDelegates = value;
    }
    resetImpersonateServiceAccountDelegates() {
        this._impersonateServiceAccountDelegates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get impersonateServiceAccountDelegatesInput() {
        return this._impersonateServiceAccountDelegates;
    }
    get integrationConnectorsCustomEndpoint() {
        return this._integrationConnectorsCustomEndpoint;
    }
    set integrationConnectorsCustomEndpoint(value) {
        this._integrationConnectorsCustomEndpoint = value;
    }
    resetIntegrationConnectorsCustomEndpoint() {
        this._integrationConnectorsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get integrationConnectorsCustomEndpointInput() {
        return this._integrationConnectorsCustomEndpoint;
    }
    get kmsCustomEndpoint() {
        return this._kmsCustomEndpoint;
    }
    set kmsCustomEndpoint(value) {
        this._kmsCustomEndpoint = value;
    }
    resetKmsCustomEndpoint() {
        this._kmsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsCustomEndpointInput() {
        return this._kmsCustomEndpoint;
    }
    get loggingCustomEndpoint() {
        return this._loggingCustomEndpoint;
    }
    set loggingCustomEndpoint(value) {
        this._loggingCustomEndpoint = value;
    }
    resetLoggingCustomEndpoint() {
        this._loggingCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingCustomEndpointInput() {
        return this._loggingCustomEndpoint;
    }
    get lookerCustomEndpoint() {
        return this._lookerCustomEndpoint;
    }
    set lookerCustomEndpoint(value) {
        this._lookerCustomEndpoint = value;
    }
    resetLookerCustomEndpoint() {
        this._lookerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lookerCustomEndpointInput() {
        return this._lookerCustomEndpoint;
    }
    get memcacheCustomEndpoint() {
        return this._memcacheCustomEndpoint;
    }
    set memcacheCustomEndpoint(value) {
        this._memcacheCustomEndpoint = value;
    }
    resetMemcacheCustomEndpoint() {
        this._memcacheCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memcacheCustomEndpointInput() {
        return this._memcacheCustomEndpoint;
    }
    get migrationCenterCustomEndpoint() {
        return this._migrationCenterCustomEndpoint;
    }
    set migrationCenterCustomEndpoint(value) {
        this._migrationCenterCustomEndpoint = value;
    }
    resetMigrationCenterCustomEndpoint() {
        this._migrationCenterCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get migrationCenterCustomEndpointInput() {
        return this._migrationCenterCustomEndpoint;
    }
    get mlEngineCustomEndpoint() {
        return this._mlEngineCustomEndpoint;
    }
    set mlEngineCustomEndpoint(value) {
        this._mlEngineCustomEndpoint = value;
    }
    resetMlEngineCustomEndpoint() {
        this._mlEngineCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mlEngineCustomEndpointInput() {
        return this._mlEngineCustomEndpoint;
    }
    get monitoringCustomEndpoint() {
        return this._monitoringCustomEndpoint;
    }
    set monitoringCustomEndpoint(value) {
        this._monitoringCustomEndpoint = value;
    }
    resetMonitoringCustomEndpoint() {
        this._monitoringCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monitoringCustomEndpointInput() {
        return this._monitoringCustomEndpoint;
    }
    get netappCustomEndpoint() {
        return this._netappCustomEndpoint;
    }
    set netappCustomEndpoint(value) {
        this._netappCustomEndpoint = value;
    }
    resetNetappCustomEndpoint() {
        this._netappCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get netappCustomEndpointInput() {
        return this._netappCustomEndpoint;
    }
    get networkConnectivityCustomEndpoint() {
        return this._networkConnectivityCustomEndpoint;
    }
    set networkConnectivityCustomEndpoint(value) {
        this._networkConnectivityCustomEndpoint = value;
    }
    resetNetworkConnectivityCustomEndpoint() {
        this._networkConnectivityCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConnectivityCustomEndpointInput() {
        return this._networkConnectivityCustomEndpoint;
    }
    get networkManagementCustomEndpoint() {
        return this._networkManagementCustomEndpoint;
    }
    set networkManagementCustomEndpoint(value) {
        this._networkManagementCustomEndpoint = value;
    }
    resetNetworkManagementCustomEndpoint() {
        this._networkManagementCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkManagementCustomEndpointInput() {
        return this._networkManagementCustomEndpoint;
    }
    get networkSecurityCustomEndpoint() {
        return this._networkSecurityCustomEndpoint;
    }
    set networkSecurityCustomEndpoint(value) {
        this._networkSecurityCustomEndpoint = value;
    }
    resetNetworkSecurityCustomEndpoint() {
        this._networkSecurityCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkSecurityCustomEndpointInput() {
        return this._networkSecurityCustomEndpoint;
    }
    get networkServicesCustomEndpoint() {
        return this._networkServicesCustomEndpoint;
    }
    set networkServicesCustomEndpoint(value) {
        this._networkServicesCustomEndpoint = value;
    }
    resetNetworkServicesCustomEndpoint() {
        this._networkServicesCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkServicesCustomEndpointInput() {
        return this._networkServicesCustomEndpoint;
    }
    get notebooksCustomEndpoint() {
        return this._notebooksCustomEndpoint;
    }
    set notebooksCustomEndpoint(value) {
        this._notebooksCustomEndpoint = value;
    }
    resetNotebooksCustomEndpoint() {
        this._notebooksCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notebooksCustomEndpointInput() {
        return this._notebooksCustomEndpoint;
    }
    get orgPolicyCustomEndpoint() {
        return this._orgPolicyCustomEndpoint;
    }
    set orgPolicyCustomEndpoint(value) {
        this._orgPolicyCustomEndpoint = value;
    }
    resetOrgPolicyCustomEndpoint() {
        this._orgPolicyCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orgPolicyCustomEndpointInput() {
        return this._orgPolicyCustomEndpoint;
    }
    get osConfigCustomEndpoint() {
        return this._osConfigCustomEndpoint;
    }
    set osConfigCustomEndpoint(value) {
        this._osConfigCustomEndpoint = value;
    }
    resetOsConfigCustomEndpoint() {
        this._osConfigCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get osConfigCustomEndpointInput() {
        return this._osConfigCustomEndpoint;
    }
    get osLoginCustomEndpoint() {
        return this._osLoginCustomEndpoint;
    }
    set osLoginCustomEndpoint(value) {
        this._osLoginCustomEndpoint = value;
    }
    resetOsLoginCustomEndpoint() {
        this._osLoginCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get osLoginCustomEndpointInput() {
        return this._osLoginCustomEndpoint;
    }
    get privatecaCustomEndpoint() {
        return this._privatecaCustomEndpoint;
    }
    set privatecaCustomEndpoint(value) {
        this._privatecaCustomEndpoint = value;
    }
    resetPrivatecaCustomEndpoint() {
        this._privatecaCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privatecaCustomEndpointInput() {
        return this._privatecaCustomEndpoint;
    }
    get project() {
        return this._project;
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get publicCaCustomEndpoint() {
        return this._publicCaCustomEndpoint;
    }
    set publicCaCustomEndpoint(value) {
        this._publicCaCustomEndpoint = value;
    }
    resetPublicCaCustomEndpoint() {
        this._publicCaCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicCaCustomEndpointInput() {
        return this._publicCaCustomEndpoint;
    }
    get pubsubCustomEndpoint() {
        return this._pubsubCustomEndpoint;
    }
    set pubsubCustomEndpoint(value) {
        this._pubsubCustomEndpoint = value;
    }
    resetPubsubCustomEndpoint() {
        this._pubsubCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pubsubCustomEndpointInput() {
        return this._pubsubCustomEndpoint;
    }
    get pubsubLiteCustomEndpoint() {
        return this._pubsubLiteCustomEndpoint;
    }
    set pubsubLiteCustomEndpoint(value) {
        this._pubsubLiteCustomEndpoint = value;
    }
    resetPubsubLiteCustomEndpoint() {
        this._pubsubLiteCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pubsubLiteCustomEndpointInput() {
        return this._pubsubLiteCustomEndpoint;
    }
    get recaptchaEnterpriseCustomEndpoint() {
        return this._recaptchaEnterpriseCustomEndpoint;
    }
    set recaptchaEnterpriseCustomEndpoint(value) {
        this._recaptchaEnterpriseCustomEndpoint = value;
    }
    resetRecaptchaEnterpriseCustomEndpoint() {
        this._recaptchaEnterpriseCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recaptchaEnterpriseCustomEndpointInput() {
        return this._recaptchaEnterpriseCustomEndpoint;
    }
    get redisCustomEndpoint() {
        return this._redisCustomEndpoint;
    }
    set redisCustomEndpoint(value) {
        this._redisCustomEndpoint = value;
    }
    resetRedisCustomEndpoint() {
        this._redisCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redisCustomEndpointInput() {
        return this._redisCustomEndpoint;
    }
    get region() {
        return this._region;
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get requestReason() {
        return this._requestReason;
    }
    set requestReason(value) {
        this._requestReason = value;
    }
    resetRequestReason() {
        this._requestReason = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestReasonInput() {
        return this._requestReason;
    }
    get requestTimeout() {
        return this._requestTimeout;
    }
    set requestTimeout(value) {
        this._requestTimeout = value;
    }
    resetRequestTimeout() {
        this._requestTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestTimeoutInput() {
        return this._requestTimeout;
    }
    get resourceManagerCustomEndpoint() {
        return this._resourceManagerCustomEndpoint;
    }
    set resourceManagerCustomEndpoint(value) {
        this._resourceManagerCustomEndpoint = value;
    }
    resetResourceManagerCustomEndpoint() {
        this._resourceManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceManagerCustomEndpointInput() {
        return this._resourceManagerCustomEndpoint;
    }
    get resourceManagerV3CustomEndpoint() {
        return this._resourceManagerV3CustomEndpoint;
    }
    set resourceManagerV3CustomEndpoint(value) {
        this._resourceManagerV3CustomEndpoint = value;
    }
    resetResourceManagerV3CustomEndpoint() {
        this._resourceManagerV3CustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceManagerV3CustomEndpointInput() {
        return this._resourceManagerV3CustomEndpoint;
    }
    get scopes() {
        return this._scopes;
    }
    set scopes(value) {
        this._scopes = value;
    }
    resetScopes() {
        this._scopes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scopesInput() {
        return this._scopes;
    }
    get secretManagerCustomEndpoint() {
        return this._secretManagerCustomEndpoint;
    }
    set secretManagerCustomEndpoint(value) {
        this._secretManagerCustomEndpoint = value;
    }
    resetSecretManagerCustomEndpoint() {
        this._secretManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretManagerCustomEndpointInput() {
        return this._secretManagerCustomEndpoint;
    }
    get secureSourceManagerCustomEndpoint() {
        return this._secureSourceManagerCustomEndpoint;
    }
    set secureSourceManagerCustomEndpoint(value) {
        this._secureSourceManagerCustomEndpoint = value;
    }
    resetSecureSourceManagerCustomEndpoint() {
        this._secureSourceManagerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secureSourceManagerCustomEndpointInput() {
        return this._secureSourceManagerCustomEndpoint;
    }
    get securityCenterCustomEndpoint() {
        return this._securityCenterCustomEndpoint;
    }
    set securityCenterCustomEndpoint(value) {
        this._securityCenterCustomEndpoint = value;
    }
    resetSecurityCenterCustomEndpoint() {
        this._securityCenterCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityCenterCustomEndpointInput() {
        return this._securityCenterCustomEndpoint;
    }
    get serviceManagementCustomEndpoint() {
        return this._serviceManagementCustomEndpoint;
    }
    set serviceManagementCustomEndpoint(value) {
        this._serviceManagementCustomEndpoint = value;
    }
    resetServiceManagementCustomEndpoint() {
        this._serviceManagementCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceManagementCustomEndpointInput() {
        return this._serviceManagementCustomEndpoint;
    }
    get serviceNetworkingCustomEndpoint() {
        return this._serviceNetworkingCustomEndpoint;
    }
    set serviceNetworkingCustomEndpoint(value) {
        this._serviceNetworkingCustomEndpoint = value;
    }
    resetServiceNetworkingCustomEndpoint() {
        this._serviceNetworkingCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNetworkingCustomEndpointInput() {
        return this._serviceNetworkingCustomEndpoint;
    }
    get serviceUsageCustomEndpoint() {
        return this._serviceUsageCustomEndpoint;
    }
    set serviceUsageCustomEndpoint(value) {
        this._serviceUsageCustomEndpoint = value;
    }
    resetServiceUsageCustomEndpoint() {
        this._serviceUsageCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceUsageCustomEndpointInput() {
        return this._serviceUsageCustomEndpoint;
    }
    get sourceRepoCustomEndpoint() {
        return this._sourceRepoCustomEndpoint;
    }
    set sourceRepoCustomEndpoint(value) {
        this._sourceRepoCustomEndpoint = value;
    }
    resetSourceRepoCustomEndpoint() {
        this._sourceRepoCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceRepoCustomEndpointInput() {
        return this._sourceRepoCustomEndpoint;
    }
    get spannerCustomEndpoint() {
        return this._spannerCustomEndpoint;
    }
    set spannerCustomEndpoint(value) {
        this._spannerCustomEndpoint = value;
    }
    resetSpannerCustomEndpoint() {
        this._spannerCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get spannerCustomEndpointInput() {
        return this._spannerCustomEndpoint;
    }
    get sqlCustomEndpoint() {
        return this._sqlCustomEndpoint;
    }
    set sqlCustomEndpoint(value) {
        this._sqlCustomEndpoint = value;
    }
    resetSqlCustomEndpoint() {
        this._sqlCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlCustomEndpointInput() {
        return this._sqlCustomEndpoint;
    }
    get storageCustomEndpoint() {
        return this._storageCustomEndpoint;
    }
    set storageCustomEndpoint(value) {
        this._storageCustomEndpoint = value;
    }
    resetStorageCustomEndpoint() {
        this._storageCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageCustomEndpointInput() {
        return this._storageCustomEndpoint;
    }
    get storageInsightsCustomEndpoint() {
        return this._storageInsightsCustomEndpoint;
    }
    set storageInsightsCustomEndpoint(value) {
        this._storageInsightsCustomEndpoint = value;
    }
    resetStorageInsightsCustomEndpoint() {
        this._storageInsightsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageInsightsCustomEndpointInput() {
        return this._storageInsightsCustomEndpoint;
    }
    get storageTransferCustomEndpoint() {
        return this._storageTransferCustomEndpoint;
    }
    set storageTransferCustomEndpoint(value) {
        this._storageTransferCustomEndpoint = value;
    }
    resetStorageTransferCustomEndpoint() {
        this._storageTransferCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageTransferCustomEndpointInput() {
        return this._storageTransferCustomEndpoint;
    }
    get tagsCustomEndpoint() {
        return this._tagsCustomEndpoint;
    }
    set tagsCustomEndpoint(value) {
        this._tagsCustomEndpoint = value;
    }
    resetTagsCustomEndpoint() {
        this._tagsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsCustomEndpointInput() {
        return this._tagsCustomEndpoint;
    }
    get tagsLocationCustomEndpoint() {
        return this._tagsLocationCustomEndpoint;
    }
    set tagsLocationCustomEndpoint(value) {
        this._tagsLocationCustomEndpoint = value;
    }
    resetTagsLocationCustomEndpoint() {
        this._tagsLocationCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsLocationCustomEndpointInput() {
        return this._tagsLocationCustomEndpoint;
    }
    get tpuCustomEndpoint() {
        return this._tpuCustomEndpoint;
    }
    set tpuCustomEndpoint(value) {
        this._tpuCustomEndpoint = value;
    }
    resetTpuCustomEndpoint() {
        this._tpuCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tpuCustomEndpointInput() {
        return this._tpuCustomEndpoint;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    set universeDomain(value) {
        this._universeDomain = value;
    }
    resetUniverseDomain() {
        this._universeDomain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get universeDomainInput() {
        return this._universeDomain;
    }
    get userProjectOverride() {
        return this._userProjectOverride;
    }
    set userProjectOverride(value) {
        this._userProjectOverride = value;
    }
    resetUserProjectOverride() {
        this._userProjectOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userProjectOverrideInput() {
        return this._userProjectOverride;
    }
    get vertexAiCustomEndpoint() {
        return this._vertexAiCustomEndpoint;
    }
    set vertexAiCustomEndpoint(value) {
        this._vertexAiCustomEndpoint = value;
    }
    resetVertexAiCustomEndpoint() {
        this._vertexAiCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vertexAiCustomEndpointInput() {
        return this._vertexAiCustomEndpoint;
    }
    get vmwareengineCustomEndpoint() {
        return this._vmwareengineCustomEndpoint;
    }
    set vmwareengineCustomEndpoint(value) {
        this._vmwareengineCustomEndpoint = value;
    }
    resetVmwareengineCustomEndpoint() {
        this._vmwareengineCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmwareengineCustomEndpointInput() {
        return this._vmwareengineCustomEndpoint;
    }
    get vpcAccessCustomEndpoint() {
        return this._vpcAccessCustomEndpoint;
    }
    set vpcAccessCustomEndpoint(value) {
        this._vpcAccessCustomEndpoint = value;
    }
    resetVpcAccessCustomEndpoint() {
        this._vpcAccessCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcAccessCustomEndpointInput() {
        return this._vpcAccessCustomEndpoint;
    }
    get workbenchCustomEndpoint() {
        return this._workbenchCustomEndpoint;
    }
    set workbenchCustomEndpoint(value) {
        this._workbenchCustomEndpoint = value;
    }
    resetWorkbenchCustomEndpoint() {
        this._workbenchCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workbenchCustomEndpointInput() {
        return this._workbenchCustomEndpoint;
    }
    get workflowsCustomEndpoint() {
        return this._workflowsCustomEndpoint;
    }
    set workflowsCustomEndpoint(value) {
        this._workflowsCustomEndpoint = value;
    }
    resetWorkflowsCustomEndpoint() {
        this._workflowsCustomEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workflowsCustomEndpointInput() {
        return this._workflowsCustomEndpoint;
    }
    get zone() {
        return this._zone;
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get alias() {
        return this._alias;
    }
    set alias(value) {
        this._alias = value;
    }
    resetAlias() {
        this._alias = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    get batching() {
        return this._batching;
    }
    set batching(value) {
        this._batching = value;
    }
    resetBatching() {
        this._batching = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchingInput() {
        return this._batching;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_approval_custom_endpoint: cdktf.stringToTerraform(this._accessApprovalCustomEndpoint),
            access_context_manager_custom_endpoint: cdktf.stringToTerraform(this._accessContextManagerCustomEndpoint),
            access_token: cdktf.stringToTerraform(this._accessToken),
            active_directory_custom_endpoint: cdktf.stringToTerraform(this._activeDirectoryCustomEndpoint),
            alloydb_custom_endpoint: cdktf.stringToTerraform(this._alloydbCustomEndpoint),
            apigee_custom_endpoint: cdktf.stringToTerraform(this._apigeeCustomEndpoint),
            apikeys_custom_endpoint: cdktf.stringToTerraform(this._apikeysCustomEndpoint),
            app_engine_custom_endpoint: cdktf.stringToTerraform(this._appEngineCustomEndpoint),
            artifact_registry_custom_endpoint: cdktf.stringToTerraform(this._artifactRegistryCustomEndpoint),
            assured_workloads_custom_endpoint: cdktf.stringToTerraform(this._assuredWorkloadsCustomEndpoint),
            beyondcorp_custom_endpoint: cdktf.stringToTerraform(this._beyondcorpCustomEndpoint),
            big_query_custom_endpoint: cdktf.stringToTerraform(this._bigQueryCustomEndpoint),
            biglake_custom_endpoint: cdktf.stringToTerraform(this._biglakeCustomEndpoint),
            bigquery_analytics_hub_custom_endpoint: cdktf.stringToTerraform(this._bigqueryAnalyticsHubCustomEndpoint),
            bigquery_connection_custom_endpoint: cdktf.stringToTerraform(this._bigqueryConnectionCustomEndpoint),
            bigquery_data_transfer_custom_endpoint: cdktf.stringToTerraform(this._bigqueryDataTransferCustomEndpoint),
            bigquery_datapolicy_custom_endpoint: cdktf.stringToTerraform(this._bigqueryDatapolicyCustomEndpoint),
            bigquery_reservation_custom_endpoint: cdktf.stringToTerraform(this._bigqueryReservationCustomEndpoint),
            bigtable_custom_endpoint: cdktf.stringToTerraform(this._bigtableCustomEndpoint),
            billing_custom_endpoint: cdktf.stringToTerraform(this._billingCustomEndpoint),
            billing_project: cdktf.stringToTerraform(this._billingProject),
            binary_authorization_custom_endpoint: cdktf.stringToTerraform(this._binaryAuthorizationCustomEndpoint),
            certificate_manager_custom_endpoint: cdktf.stringToTerraform(this._certificateManagerCustomEndpoint),
            cloud_asset_custom_endpoint: cdktf.stringToTerraform(this._cloudAssetCustomEndpoint),
            cloud_billing_custom_endpoint: cdktf.stringToTerraform(this._cloudBillingCustomEndpoint),
            cloud_build_custom_endpoint: cdktf.stringToTerraform(this._cloudBuildCustomEndpoint),
            cloud_build_worker_pool_custom_endpoint: cdktf.stringToTerraform(this._cloudBuildWorkerPoolCustomEndpoint),
            cloud_functions_custom_endpoint: cdktf.stringToTerraform(this._cloudFunctionsCustomEndpoint),
            cloud_identity_custom_endpoint: cdktf.stringToTerraform(this._cloudIdentityCustomEndpoint),
            cloud_ids_custom_endpoint: cdktf.stringToTerraform(this._cloudIdsCustomEndpoint),
            cloud_resource_manager_custom_endpoint: cdktf.stringToTerraform(this._cloudResourceManagerCustomEndpoint),
            cloud_run_custom_endpoint: cdktf.stringToTerraform(this._cloudRunCustomEndpoint),
            cloud_run_v2_custom_endpoint: cdktf.stringToTerraform(this._cloudRunV2CustomEndpoint),
            cloud_scheduler_custom_endpoint: cdktf.stringToTerraform(this._cloudSchedulerCustomEndpoint),
            cloud_tasks_custom_endpoint: cdktf.stringToTerraform(this._cloudTasksCustomEndpoint),
            cloudbuildv2_custom_endpoint: cdktf.stringToTerraform(this._cloudbuildv2CustomEndpoint),
            clouddeploy_custom_endpoint: cdktf.stringToTerraform(this._clouddeployCustomEndpoint),
            cloudfunctions2_custom_endpoint: cdktf.stringToTerraform(this._cloudfunctions2CustomEndpoint),
            composer_custom_endpoint: cdktf.stringToTerraform(this._composerCustomEndpoint),
            compute_custom_endpoint: cdktf.stringToTerraform(this._computeCustomEndpoint),
            container_analysis_custom_endpoint: cdktf.stringToTerraform(this._containerAnalysisCustomEndpoint),
            container_attached_custom_endpoint: cdktf.stringToTerraform(this._containerAttachedCustomEndpoint),
            container_aws_custom_endpoint: cdktf.stringToTerraform(this._containerAwsCustomEndpoint),
            container_azure_custom_endpoint: cdktf.stringToTerraform(this._containerAzureCustomEndpoint),
            container_custom_endpoint: cdktf.stringToTerraform(this._containerCustomEndpoint),
            core_billing_custom_endpoint: cdktf.stringToTerraform(this._coreBillingCustomEndpoint),
            credentials: cdktf.stringToTerraform(this._credentials),
            data_catalog_custom_endpoint: cdktf.stringToTerraform(this._dataCatalogCustomEndpoint),
            data_fusion_custom_endpoint: cdktf.stringToTerraform(this._dataFusionCustomEndpoint),
            data_loss_prevention_custom_endpoint: cdktf.stringToTerraform(this._dataLossPreventionCustomEndpoint),
            data_pipeline_custom_endpoint: cdktf.stringToTerraform(this._dataPipelineCustomEndpoint),
            database_migration_service_custom_endpoint: cdktf.stringToTerraform(this._databaseMigrationServiceCustomEndpoint),
            dataflow_custom_endpoint: cdktf.stringToTerraform(this._dataflowCustomEndpoint),
            dataplex_custom_endpoint: cdktf.stringToTerraform(this._dataplexCustomEndpoint),
            dataproc_custom_endpoint: cdktf.stringToTerraform(this._dataprocCustomEndpoint),
            dataproc_metastore_custom_endpoint: cdktf.stringToTerraform(this._dataprocMetastoreCustomEndpoint),
            datastore_custom_endpoint: cdktf.stringToTerraform(this._datastoreCustomEndpoint),
            datastream_custom_endpoint: cdktf.stringToTerraform(this._datastreamCustomEndpoint),
            default_labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._defaultLabels),
            deployment_manager_custom_endpoint: cdktf.stringToTerraform(this._deploymentManagerCustomEndpoint),
            dialogflow_custom_endpoint: cdktf.stringToTerraform(this._dialogflowCustomEndpoint),
            dialogflow_cx_custom_endpoint: cdktf.stringToTerraform(this._dialogflowCxCustomEndpoint),
            dns_custom_endpoint: cdktf.stringToTerraform(this._dnsCustomEndpoint),
            document_ai_custom_endpoint: cdktf.stringToTerraform(this._documentAiCustomEndpoint),
            document_ai_warehouse_custom_endpoint: cdktf.stringToTerraform(this._documentAiWarehouseCustomEndpoint),
            edgecontainer_custom_endpoint: cdktf.stringToTerraform(this._edgecontainerCustomEndpoint),
            edgenetwork_custom_endpoint: cdktf.stringToTerraform(this._edgenetworkCustomEndpoint),
            essential_contacts_custom_endpoint: cdktf.stringToTerraform(this._essentialContactsCustomEndpoint),
            eventarc_custom_endpoint: cdktf.stringToTerraform(this._eventarcCustomEndpoint),
            filestore_custom_endpoint: cdktf.stringToTerraform(this._filestoreCustomEndpoint),
            firebaserules_custom_endpoint: cdktf.stringToTerraform(this._firebaserulesCustomEndpoint),
            firestore_custom_endpoint: cdktf.stringToTerraform(this._firestoreCustomEndpoint),
            gke_backup_custom_endpoint: cdktf.stringToTerraform(this._gkeBackupCustomEndpoint),
            gke_hub2_custom_endpoint: cdktf.stringToTerraform(this._gkeHub2CustomEndpoint),
            gke_hub_custom_endpoint: cdktf.stringToTerraform(this._gkeHubCustomEndpoint),
            gkehub_feature_custom_endpoint: cdktf.stringToTerraform(this._gkehubFeatureCustomEndpoint),
            gkeonprem_custom_endpoint: cdktf.stringToTerraform(this._gkeonpremCustomEndpoint),
            healthcare_custom_endpoint: cdktf.stringToTerraform(this._healthcareCustomEndpoint),
            iam2_custom_endpoint: cdktf.stringToTerraform(this._iam2CustomEndpoint),
            iam_beta_custom_endpoint: cdktf.stringToTerraform(this._iamBetaCustomEndpoint),
            iam_credentials_custom_endpoint: cdktf.stringToTerraform(this._iamCredentialsCustomEndpoint),
            iam_custom_endpoint: cdktf.stringToTerraform(this._iamCustomEndpoint),
            iam_workforce_pool_custom_endpoint: cdktf.stringToTerraform(this._iamWorkforcePoolCustomEndpoint),
            iap_custom_endpoint: cdktf.stringToTerraform(this._iapCustomEndpoint),
            identity_platform_custom_endpoint: cdktf.stringToTerraform(this._identityPlatformCustomEndpoint),
            impersonate_service_account: cdktf.stringToTerraform(this._impersonateServiceAccount),
            impersonate_service_account_delegates: cdktf.listMapper(cdktf.stringToTerraform, false)(this._impersonateServiceAccountDelegates),
            integration_connectors_custom_endpoint: cdktf.stringToTerraform(this._integrationConnectorsCustomEndpoint),
            kms_custom_endpoint: cdktf.stringToTerraform(this._kmsCustomEndpoint),
            logging_custom_endpoint: cdktf.stringToTerraform(this._loggingCustomEndpoint),
            looker_custom_endpoint: cdktf.stringToTerraform(this._lookerCustomEndpoint),
            memcache_custom_endpoint: cdktf.stringToTerraform(this._memcacheCustomEndpoint),
            migration_center_custom_endpoint: cdktf.stringToTerraform(this._migrationCenterCustomEndpoint),
            ml_engine_custom_endpoint: cdktf.stringToTerraform(this._mlEngineCustomEndpoint),
            monitoring_custom_endpoint: cdktf.stringToTerraform(this._monitoringCustomEndpoint),
            netapp_custom_endpoint: cdktf.stringToTerraform(this._netappCustomEndpoint),
            network_connectivity_custom_endpoint: cdktf.stringToTerraform(this._networkConnectivityCustomEndpoint),
            network_management_custom_endpoint: cdktf.stringToTerraform(this._networkManagementCustomEndpoint),
            network_security_custom_endpoint: cdktf.stringToTerraform(this._networkSecurityCustomEndpoint),
            network_services_custom_endpoint: cdktf.stringToTerraform(this._networkServicesCustomEndpoint),
            notebooks_custom_endpoint: cdktf.stringToTerraform(this._notebooksCustomEndpoint),
            org_policy_custom_endpoint: cdktf.stringToTerraform(this._orgPolicyCustomEndpoint),
            os_config_custom_endpoint: cdktf.stringToTerraform(this._osConfigCustomEndpoint),
            os_login_custom_endpoint: cdktf.stringToTerraform(this._osLoginCustomEndpoint),
            privateca_custom_endpoint: cdktf.stringToTerraform(this._privatecaCustomEndpoint),
            project: cdktf.stringToTerraform(this._project),
            public_ca_custom_endpoint: cdktf.stringToTerraform(this._publicCaCustomEndpoint),
            pubsub_custom_endpoint: cdktf.stringToTerraform(this._pubsubCustomEndpoint),
            pubsub_lite_custom_endpoint: cdktf.stringToTerraform(this._pubsubLiteCustomEndpoint),
            recaptcha_enterprise_custom_endpoint: cdktf.stringToTerraform(this._recaptchaEnterpriseCustomEndpoint),
            redis_custom_endpoint: cdktf.stringToTerraform(this._redisCustomEndpoint),
            region: cdktf.stringToTerraform(this._region),
            request_reason: cdktf.stringToTerraform(this._requestReason),
            request_timeout: cdktf.stringToTerraform(this._requestTimeout),
            resource_manager_custom_endpoint: cdktf.stringToTerraform(this._resourceManagerCustomEndpoint),
            resource_manager_v3_custom_endpoint: cdktf.stringToTerraform(this._resourceManagerV3CustomEndpoint),
            scopes: cdktf.listMapper(cdktf.stringToTerraform, false)(this._scopes),
            secret_manager_custom_endpoint: cdktf.stringToTerraform(this._secretManagerCustomEndpoint),
            secure_source_manager_custom_endpoint: cdktf.stringToTerraform(this._secureSourceManagerCustomEndpoint),
            security_center_custom_endpoint: cdktf.stringToTerraform(this._securityCenterCustomEndpoint),
            service_management_custom_endpoint: cdktf.stringToTerraform(this._serviceManagementCustomEndpoint),
            service_networking_custom_endpoint: cdktf.stringToTerraform(this._serviceNetworkingCustomEndpoint),
            service_usage_custom_endpoint: cdktf.stringToTerraform(this._serviceUsageCustomEndpoint),
            source_repo_custom_endpoint: cdktf.stringToTerraform(this._sourceRepoCustomEndpoint),
            spanner_custom_endpoint: cdktf.stringToTerraform(this._spannerCustomEndpoint),
            sql_custom_endpoint: cdktf.stringToTerraform(this._sqlCustomEndpoint),
            storage_custom_endpoint: cdktf.stringToTerraform(this._storageCustomEndpoint),
            storage_insights_custom_endpoint: cdktf.stringToTerraform(this._storageInsightsCustomEndpoint),
            storage_transfer_custom_endpoint: cdktf.stringToTerraform(this._storageTransferCustomEndpoint),
            tags_custom_endpoint: cdktf.stringToTerraform(this._tagsCustomEndpoint),
            tags_location_custom_endpoint: cdktf.stringToTerraform(this._tagsLocationCustomEndpoint),
            tpu_custom_endpoint: cdktf.stringToTerraform(this._tpuCustomEndpoint),
            universe_domain: cdktf.stringToTerraform(this._universeDomain),
            user_project_override: cdktf.booleanToTerraform(this._userProjectOverride),
            vertex_ai_custom_endpoint: cdktf.stringToTerraform(this._vertexAiCustomEndpoint),
            vmwareengine_custom_endpoint: cdktf.stringToTerraform(this._vmwareengineCustomEndpoint),
            vpc_access_custom_endpoint: cdktf.stringToTerraform(this._vpcAccessCustomEndpoint),
            workbench_custom_endpoint: cdktf.stringToTerraform(this._workbenchCustomEndpoint),
            workflows_custom_endpoint: cdktf.stringToTerraform(this._workflowsCustomEndpoint),
            zone: cdktf.stringToTerraform(this._zone),
            alias: cdktf.stringToTerraform(this._alias),
            batching: googleProviderBatchingToTerraform(this._batching),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_approval_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._accessApprovalCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_context_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._accessContextManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_token: {
                value: cdktf.stringToHclTerraform(this._accessToken),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            active_directory_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._activeDirectoryCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            alloydb_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._alloydbCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            apigee_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._apigeeCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            apikeys_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._apikeysCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            app_engine_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._appEngineCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            artifact_registry_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._artifactRegistryCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            assured_workloads_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._assuredWorkloadsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            beyondcorp_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._beyondcorpCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            big_query_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigQueryCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            biglake_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._biglakeCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigquery_analytics_hub_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigqueryAnalyticsHubCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigquery_connection_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigqueryConnectionCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigquery_data_transfer_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigqueryDataTransferCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigquery_datapolicy_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigqueryDatapolicyCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigquery_reservation_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigqueryReservationCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bigtable_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._bigtableCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            billing_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._billingCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            billing_project: {
                value: cdktf.stringToHclTerraform(this._billingProject),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            binary_authorization_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._binaryAuthorizationCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            certificate_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._certificateManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_asset_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudAssetCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_billing_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudBillingCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_build_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudBuildCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_build_worker_pool_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudBuildWorkerPoolCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_functions_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudFunctionsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_identity_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudIdentityCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_ids_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudIdsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_resource_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudResourceManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_run_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudRunCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_run_v2_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudRunV2CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_scheduler_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudSchedulerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloud_tasks_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudTasksCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloudbuildv2_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudbuildv2CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            clouddeploy_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._clouddeployCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cloudfunctions2_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._cloudfunctions2CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            composer_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._composerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            compute_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._computeCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_analysis_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._containerAnalysisCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_attached_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._containerAttachedCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_aws_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._containerAwsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_azure_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._containerAzureCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._containerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            core_billing_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._coreBillingCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            credentials: {
                value: cdktf.stringToHclTerraform(this._credentials),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_catalog_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataCatalogCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_fusion_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataFusionCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_loss_prevention_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataLossPreventionCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_pipeline_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataPipelineCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            database_migration_service_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._databaseMigrationServiceCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dataflow_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataflowCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dataplex_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataplexCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dataproc_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataprocCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dataproc_metastore_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dataprocMetastoreCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            datastore_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._datastoreCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            datastream_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._datastreamCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            default_labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._defaultLabels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            deployment_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._deploymentManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dialogflow_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dialogflowCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dialogflow_cx_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dialogflowCxCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dns_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._dnsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            document_ai_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._documentAiCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            document_ai_warehouse_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._documentAiWarehouseCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            edgecontainer_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._edgecontainerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            edgenetwork_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._edgenetworkCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            essential_contacts_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._essentialContactsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            eventarc_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._eventarcCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filestore_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._filestoreCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            firebaserules_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._firebaserulesCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            firestore_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._firestoreCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gke_backup_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._gkeBackupCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gke_hub2_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._gkeHub2CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gke_hub_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._gkeHubCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gkehub_feature_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._gkehubFeatureCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gkeonprem_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._gkeonpremCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            healthcare_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._healthcareCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam2_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iam2CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_beta_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iamBetaCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_credentials_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iamCredentialsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iamCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_workforce_pool_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iamWorkforcePoolCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iap_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._iapCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identity_platform_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._identityPlatformCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            impersonate_service_account: {
                value: cdktf.stringToHclTerraform(this._impersonateServiceAccount),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            impersonate_service_account_delegates: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._impersonateServiceAccountDelegates),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            integration_connectors_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._integrationConnectorsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._kmsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            logging_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._loggingCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            looker_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._lookerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            memcache_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._memcacheCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            migration_center_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._migrationCenterCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ml_engine_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._mlEngineCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            monitoring_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._monitoringCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            netapp_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._netappCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network_connectivity_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._networkConnectivityCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network_management_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._networkManagementCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network_security_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._networkSecurityCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network_services_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._networkServicesCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notebooks_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._notebooksCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            org_policy_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._orgPolicyCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            os_config_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._osConfigCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            os_login_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._osLoginCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            privateca_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._privatecaCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            public_ca_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._publicCaCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pubsub_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._pubsubCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pubsub_lite_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._pubsubLiteCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            recaptcha_enterprise_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._recaptchaEnterpriseCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            redis_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._redisCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            request_reason: {
                value: cdktf.stringToHclTerraform(this._requestReason),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            request_timeout: {
                value: cdktf.stringToHclTerraform(this._requestTimeout),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._resourceManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_manager_v3_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._resourceManagerV3CustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scopes: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._scopes),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            secret_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._secretManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secure_source_manager_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._secureSourceManagerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            security_center_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._securityCenterCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_management_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._serviceManagementCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_networking_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._serviceNetworkingCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_usage_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._serviceUsageCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            source_repo_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._sourceRepoCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            spanner_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._spannerCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            sql_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._sqlCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            storage_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._storageCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            storage_insights_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._storageInsightsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            storage_transfer_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._storageTransferCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._tagsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags_location_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._tagsLocationCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tpu_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._tpuCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            universe_domain: {
                value: cdktf.stringToHclTerraform(this._universeDomain),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_project_override: {
                value: cdktf.booleanToHclTerraform(this._userProjectOverride),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            vertex_ai_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._vertexAiCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vmwareengine_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._vmwareengineCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vpc_access_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._vpcAccessCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workbench_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._workbenchCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workflows_custom_endpoint: {
                value: cdktf.stringToHclTerraform(this._workflowsCustomEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            alias: {
                value: cdktf.stringToHclTerraform(this._alias),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            batching: {
                value: googleProviderBatchingToHclTerraform(this._batching),
                isBlock: true,
                type: "list",
                storageClassType: "GoogleProviderBatchingList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GoogleProvider = GoogleProvider;
// =================
// STATIC PROPERTIES
// =================
GoogleProvider.tfResourceType = "google";
