// @cdktf/provider-aws/lib/data-aws-kendra-query-suggestions-block-list/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKendraQuerySuggestionsBlockList = exports.DataAwsKendraQuerySuggestionsBlockListSourceS3PathList = exports.DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference = exports.dataAwsKendraQuerySuggestionsBlockListSourceS3PathToHclTerraform = exports.dataAwsKendraQuerySuggestionsBlockListSourceS3PathToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsKendraQuerySuggestionsBlockListSourceS3PathToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraQuerySuggestionsBlockListSourceS3PathToTerraform = dataAwsKendraQuerySuggestionsBlockListSourceS3PathToTerraform;
function dataAwsKendraQuerySuggestionsBlockListSourceS3PathToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraQuerySuggestionsBlockListSourceS3PathToHclTerraform = dataAwsKendraQuerySuggestionsBlockListSourceS3PathToHclTerraform;
class DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
}
exports.DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference = DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsKendraQuerySuggestionsBlockList.DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference", version: "19.15.0" };
class DataAwsKendraQuerySuggestionsBlockListSourceS3PathList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraQuerySuggestionsBlockListSourceS3PathOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraQuerySuggestionsBlockListSourceS3PathList = DataAwsKendraQuerySuggestionsBlockListSourceS3PathList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKendraQuerySuggestionsBlockListSourceS3PathList[_b] = { fqn: "@cdktf/provider-aws.dataAwsKendraQuerySuggestionsBlockList.DataAwsKendraQuerySuggestionsBlockListSourceS3PathList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_query_suggestions_block_list aws_kendra_query_suggestions_block_list}
*/
class DataAwsKendraQuerySuggestionsBlockList extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsKendraQuerySuggestionsBlockList resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsKendraQuerySuggestionsBlockList to import
    * @param importFromId The id of the existing DataAwsKendraQuerySuggestionsBlockList that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_query_suggestions_block_list#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsKendraQuerySuggestionsBlockList to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_kendra_query_suggestions_block_list", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_query_suggestions_block_list aws_kendra_query_suggestions_block_list} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsKendraQuerySuggestionsBlockListConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kendra_query_suggestions_block_list',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // source_s3_path - computed: true, optional: false, required: false
        this._sourceS3Path = new DataAwsKendraQuerySuggestionsBlockListSourceS3PathList(this, "source_s3_path", false);
        this._id = config.id;
        this._indexId = config.indexId;
        this._querySuggestionsBlockListId = config.querySuggestionsBlockListId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // error_message - computed: true, optional: false, required: false
    get errorMessage() {
        return this.getStringAttribute('error_message');
    }
    // file_size_bytes - computed: true, optional: false, required: false
    get fileSizeBytes() {
        return this.getNumberAttribute('file_size_bytes');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get indexId() {
        return this.getStringAttribute('index_id');
    }
    set indexId(value) {
        this._indexId = value;
    }
    // Temporarily expose input value. Use with caution.
    get indexIdInput() {
        return this._indexId;
    }
    // item_count - computed: true, optional: false, required: false
    get itemCount() {
        return this.getNumberAttribute('item_count');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get querySuggestionsBlockListId() {
        return this.getStringAttribute('query_suggestions_block_list_id');
    }
    set querySuggestionsBlockListId(value) {
        this._querySuggestionsBlockListId = value;
    }
    // Temporarily expose input value. Use with caution.
    get querySuggestionsBlockListIdInput() {
        return this._querySuggestionsBlockListId;
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    get sourceS3Path() {
        return this._sourceS3Path;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // updated_at - computed: true, optional: false, required: false
    get updatedAt() {
        return this.getStringAttribute('updated_at');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            index_id: cdktf.stringToTerraform(this._indexId),
            query_suggestions_block_list_id: cdktf.stringToTerraform(this._querySuggestionsBlockListId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            index_id: {
                value: cdktf.stringToHclTerraform(this._indexId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            query_suggestions_block_list_id: {
                value: cdktf.stringToHclTerraform(this._querySuggestionsBlockListId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsKendraQuerySuggestionsBlockList = DataAwsKendraQuerySuggestionsBlockList;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKendraQuerySuggestionsBlockList[_c] = { fqn: "@cdktf/provider-aws.dataAwsKendraQuerySuggestionsBlockList.DataAwsKendraQuerySuggestionsBlockList", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsKendraQuerySuggestionsBlockList.tfResourceType = "aws_kendra_query_suggestions_block_list";
