// @cdktf/provider-aws/lib/db-option-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbOptionGroup = exports.DbOptionGroupTimeoutsOutputReference = exports.dbOptionGroupTimeoutsToHclTerraform = exports.dbOptionGroupTimeoutsToTerraform = exports.DbOptionGroupOptionList = exports.DbOptionGroupOptionOutputReference = exports.dbOptionGroupOptionToHclTerraform = exports.dbOptionGroupOptionToTerraform = exports.DbOptionGroupOptionOptionSettingsList = exports.DbOptionGroupOptionOptionSettingsOutputReference = exports.dbOptionGroupOptionOptionSettingsToHclTerraform = exports.dbOptionGroupOptionOptionSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbOptionGroupOptionOptionSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dbOptionGroupOptionOptionSettingsToTerraform = dbOptionGroupOptionOptionSettingsToTerraform;
function dbOptionGroupOptionOptionSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dbOptionGroupOptionOptionSettingsToHclTerraform = dbOptionGroupOptionOptionSettingsToHclTerraform;
class DbOptionGroupOptionOptionSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DbOptionGroupOptionOptionSettingsOutputReference = DbOptionGroupOptionOptionSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbOptionGroupOptionOptionSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroupOptionOptionSettingsOutputReference", version: "19.15.0" };
class DbOptionGroupOptionOptionSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DbOptionGroupOptionOptionSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DbOptionGroupOptionOptionSettingsList = DbOptionGroupOptionOptionSettingsList;
_b = JSII_RTTI_SYMBOL_1;
DbOptionGroupOptionOptionSettingsList[_b] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroupOptionOptionSettingsList", version: "19.15.0" };
function dbOptionGroupOptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        db_security_group_memberships: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.dbSecurityGroupMemberships),
        option_name: cdktf.stringToTerraform(struct.optionName),
        port: cdktf.numberToTerraform(struct.port),
        version: cdktf.stringToTerraform(struct.version),
        vpc_security_group_memberships: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.vpcSecurityGroupMemberships),
        option_settings: cdktf.listMapper(dbOptionGroupOptionOptionSettingsToTerraform, true)(struct.optionSettings),
    };
}
exports.dbOptionGroupOptionToTerraform = dbOptionGroupOptionToTerraform;
function dbOptionGroupOptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        db_security_group_memberships: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.dbSecurityGroupMemberships),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        option_name: {
            value: cdktf.stringToHclTerraform(struct.optionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        version: {
            value: cdktf.stringToHclTerraform(struct.version),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vpc_security_group_memberships: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.vpcSecurityGroupMemberships),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        option_settings: {
            value: cdktf.listMapperHcl(dbOptionGroupOptionOptionSettingsToHclTerraform, true)(struct.optionSettings),
            isBlock: true,
            type: "set",
            storageClassType: "DbOptionGroupOptionOptionSettingsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dbOptionGroupOptionToHclTerraform = dbOptionGroupOptionToHclTerraform;
class DbOptionGroupOptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // option_settings - computed: false, optional: true, required: false
        this._optionSettings = new DbOptionGroupOptionOptionSettingsList(this, "option_settings", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dbSecurityGroupMemberships !== undefined) {
            hasAnyValues = true;
            internalValueResult.dbSecurityGroupMemberships = this._dbSecurityGroupMemberships;
        }
        if (this._optionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.optionName = this._optionName;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._vpcSecurityGroupMemberships !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcSecurityGroupMemberships = this._vpcSecurityGroupMemberships;
        }
        if (this._optionSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.optionSettings = this._optionSettings?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dbSecurityGroupMemberships = undefined;
            this._optionName = undefined;
            this._port = undefined;
            this._version = undefined;
            this._vpcSecurityGroupMemberships = undefined;
            this._optionSettings.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dbSecurityGroupMemberships = value.dbSecurityGroupMemberships;
            this._optionName = value.optionName;
            this._port = value.port;
            this._version = value.version;
            this._vpcSecurityGroupMemberships = value.vpcSecurityGroupMemberships;
            this._optionSettings.internalValue = value.optionSettings;
        }
    }
    get dbSecurityGroupMemberships() {
        return cdktf.Fn.tolist(this.getListAttribute('db_security_group_memberships'));
    }
    set dbSecurityGroupMemberships(value) {
        this._dbSecurityGroupMemberships = value;
    }
    resetDbSecurityGroupMemberships() {
        this._dbSecurityGroupMemberships = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbSecurityGroupMembershipsInput() {
        return this._dbSecurityGroupMemberships;
    }
    get optionName() {
        return this.getStringAttribute('option_name');
    }
    set optionName(value) {
        this._optionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get optionNameInput() {
        return this._optionName;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get vpcSecurityGroupMemberships() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_security_group_memberships'));
    }
    set vpcSecurityGroupMemberships(value) {
        this._vpcSecurityGroupMemberships = value;
    }
    resetVpcSecurityGroupMemberships() {
        this._vpcSecurityGroupMemberships = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcSecurityGroupMembershipsInput() {
        return this._vpcSecurityGroupMemberships;
    }
    get optionSettings() {
        return this._optionSettings;
    }
    putOptionSettings(value) {
        this._optionSettings.internalValue = value;
    }
    resetOptionSettings() {
        this._optionSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionSettingsInput() {
        return this._optionSettings.internalValue;
    }
}
exports.DbOptionGroupOptionOutputReference = DbOptionGroupOptionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DbOptionGroupOptionOutputReference[_c] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroupOptionOutputReference", version: "19.15.0" };
class DbOptionGroupOptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DbOptionGroupOptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DbOptionGroupOptionList = DbOptionGroupOptionList;
_d = JSII_RTTI_SYMBOL_1;
DbOptionGroupOptionList[_d] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroupOptionList", version: "19.15.0" };
function dbOptionGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.dbOptionGroupTimeoutsToTerraform = dbOptionGroupTimeoutsToTerraform;
function dbOptionGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dbOptionGroupTimeoutsToHclTerraform = dbOptionGroupTimeoutsToHclTerraform;
class DbOptionGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._delete = value.delete;
        }
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.DbOptionGroupTimeoutsOutputReference = DbOptionGroupTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DbOptionGroupTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroupTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_option_group aws_db_option_group}
*/
class DbOptionGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DbOptionGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DbOptionGroup to import
    * @param importFromId The id of the existing DbOptionGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_option_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DbOptionGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_db_option_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_option_group aws_db_option_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DbOptionGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_option_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // option - computed: false, optional: true, required: false
        this._option = new DbOptionGroupOptionList(this, "option", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbOptionGroupTimeoutsOutputReference(this, "timeouts");
        this._engineName = config.engineName;
        this._id = config.id;
        this._majorEngineVersion = config.majorEngineVersion;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._optionGroupDescription = config.optionGroupDescription;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._option.internalValue = config.option;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get engineName() {
        return this.getStringAttribute('engine_name');
    }
    set engineName(value) {
        this._engineName = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineNameInput() {
        return this._engineName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get majorEngineVersion() {
        return this.getStringAttribute('major_engine_version');
    }
    set majorEngineVersion(value) {
        this._majorEngineVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get majorEngineVersionInput() {
        return this._majorEngineVersion;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get optionGroupDescription() {
        return this.getStringAttribute('option_group_description');
    }
    set optionGroupDescription(value) {
        this._optionGroupDescription = value;
    }
    resetOptionGroupDescription() {
        this._optionGroupDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionGroupDescriptionInput() {
        return this._optionGroupDescription;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get option() {
        return this._option;
    }
    putOption(value) {
        this._option.internalValue = value;
    }
    resetOption() {
        this._option.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionInput() {
        return this._option.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            engine_name: cdktf.stringToTerraform(this._engineName),
            id: cdktf.stringToTerraform(this._id),
            major_engine_version: cdktf.stringToTerraform(this._majorEngineVersion),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            option_group_description: cdktf.stringToTerraform(this._optionGroupDescription),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            option: cdktf.listMapper(dbOptionGroupOptionToTerraform, true)(this._option.internalValue),
            timeouts: dbOptionGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            engine_name: {
                value: cdktf.stringToHclTerraform(this._engineName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            major_engine_version: {
                value: cdktf.stringToHclTerraform(this._majorEngineVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            option_group_description: {
                value: cdktf.stringToHclTerraform(this._optionGroupDescription),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            option: {
                value: cdktf.listMapperHcl(dbOptionGroupOptionToHclTerraform, true)(this._option.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DbOptionGroupOptionList",
            },
            timeouts: {
                value: dbOptionGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DbOptionGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DbOptionGroup = DbOptionGroup;
_f = JSII_RTTI_SYMBOL_1;
DbOptionGroup[_f] = { fqn: "@cdktf/provider-aws.dbOptionGroup.DbOptionGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DbOptionGroup.tfResourceType = "aws_db_option_group";
