// @cdktf/provider-aws/lib/securitylake-data-lake/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecuritylakeDataLake = exports.SecuritylakeDataLakeTimeoutsOutputReference = exports.securitylakeDataLakeTimeoutsToHclTerraform = exports.securitylakeDataLakeTimeoutsToTerraform = exports.SecuritylakeDataLakeConfigurationList = exports.SecuritylakeDataLakeConfigurationOutputReference = exports.securitylakeDataLakeConfigurationToHclTerraform = exports.securitylakeDataLakeConfigurationToTerraform = exports.SecuritylakeDataLakeConfigurationReplicationConfigurationList = exports.SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference = exports.securitylakeDataLakeConfigurationReplicationConfigurationToHclTerraform = exports.securitylakeDataLakeConfigurationReplicationConfigurationToTerraform = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationList = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference = exports.securitylakeDataLakeConfigurationLifecycleConfigurationToHclTerraform = exports.securitylakeDataLakeConfigurationLifecycleConfigurationToTerraform = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference = exports.securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToHclTerraform = exports.securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToTerraform = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList = exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference = exports.securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToHclTerraform = exports.securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToTerraform = exports.SecuritylakeDataLakeConfigurationEncryptionConfigurationList = exports.SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference = exports.securitylakeDataLakeConfigurationEncryptionConfigurationToHclTerraform = exports.securitylakeDataLakeConfigurationEncryptionConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function securitylakeDataLakeConfigurationEncryptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.securitylakeDataLakeConfigurationEncryptionConfigurationToTerraform = securitylakeDataLakeConfigurationEncryptionConfigurationToTerraform;
function securitylakeDataLakeConfigurationEncryptionConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationEncryptionConfigurationToHclTerraform = securitylakeDataLakeConfigurationEncryptionConfigurationToHclTerraform;
class SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._kmsKeyId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference = SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationEncryptionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationEncryptionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationEncryptionConfigurationList = SecuritylakeDataLakeConfigurationEncryptionConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationEncryptionConfigurationList[_b] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationEncryptionConfigurationList", version: "19.15.0" };
function securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        days: cdktf.numberToTerraform(struct.days),
    };
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToTerraform;
function securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        days: {
            value: cdktf.numberToHclTerraform(struct.days),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToHclTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToHclTerraform;
class SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._days !== undefined) {
            hasAnyValues = true;
            internalValueResult.days = this._days;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._days = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._days = value.days;
        }
    }
    get days() {
        return this.getNumberAttribute('days');
    }
    set days(value) {
        this._days = value;
    }
    resetDays() {
        this._days = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get daysInput() {
        return this._days;
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference = SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference[_c] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList = SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList;
_d = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList[_d] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList", version: "19.15.0" };
function securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        days: cdktf.numberToTerraform(struct.days),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
    };
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToTerraform;
function securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        days: {
            value: cdktf.numberToHclTerraform(struct.days),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        storage_class: {
            value: cdktf.stringToHclTerraform(struct.storageClass),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToHclTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToHclTerraform;
class SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._days !== undefined) {
            hasAnyValues = true;
            internalValueResult.days = this._days;
        }
        if (this._storageClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._days = undefined;
            this._storageClass = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._days = value.days;
            this._storageClass = value.storageClass;
        }
    }
    get days() {
        return this.getNumberAttribute('days');
    }
    set days(value) {
        this._days = value;
    }
    resetDays() {
        this._days = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get daysInput() {
        return this._days;
    }
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    set storageClass(value) {
        this._storageClass = value;
    }
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageClassInput() {
        return this._storageClass;
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference = SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference[_e] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList = SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList;
_f = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList[_f] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList", version: "19.15.0" };
function securitylakeDataLakeConfigurationLifecycleConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expiration: cdktf.listMapper(securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToTerraform, true)(struct.expiration),
        transition: cdktf.listMapper(securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToTerraform, true)(struct.transition),
    };
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationToTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationToTerraform;
function securitylakeDataLakeConfigurationLifecycleConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expiration: {
            value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationLifecycleConfigurationExpirationToHclTerraform, true)(struct.expiration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList",
        },
        transition: {
            value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationLifecycleConfigurationTransitionToHclTerraform, true)(struct.transition),
            isBlock: true,
            type: "set",
            storageClassType: "SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationLifecycleConfigurationToHclTerraform = securitylakeDataLakeConfigurationLifecycleConfigurationToHclTerraform;
class SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // expiration - computed: false, optional: true, required: false
        this._expiration = new SecuritylakeDataLakeConfigurationLifecycleConfigurationExpirationList(this, "expiration", false);
        // transition - computed: false, optional: true, required: false
        this._transition = new SecuritylakeDataLakeConfigurationLifecycleConfigurationTransitionList(this, "transition", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expiration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.expiration = this._expiration?.internalValue;
        }
        if (this._transition?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.transition = this._transition?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expiration.internalValue = undefined;
            this._transition.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expiration.internalValue = value.expiration;
            this._transition.internalValue = value.transition;
        }
    }
    get expiration() {
        return this._expiration;
    }
    putExpiration(value) {
        this._expiration.internalValue = value;
    }
    resetExpiration() {
        this._expiration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expirationInput() {
        return this._expiration.internalValue;
    }
    get transition() {
        return this._transition;
    }
    putTransition(value) {
        this._transition.internalValue = value;
    }
    resetTransition() {
        this._transition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitionInput() {
        return this._transition.internalValue;
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference = SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationLifecycleConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationLifecycleConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationLifecycleConfigurationList = SecuritylakeDataLakeConfigurationLifecycleConfigurationList;
_h = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationLifecycleConfigurationList[_h] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationLifecycleConfigurationList", version: "19.15.0" };
function securitylakeDataLakeConfigurationReplicationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        regions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.regions),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.securitylakeDataLakeConfigurationReplicationConfigurationToTerraform = securitylakeDataLakeConfigurationReplicationConfigurationToTerraform;
function securitylakeDataLakeConfigurationReplicationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        regions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.regions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationReplicationConfigurationToHclTerraform = securitylakeDataLakeConfigurationReplicationConfigurationToHclTerraform;
class SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._regions !== undefined) {
            hasAnyValues = true;
            internalValueResult.regions = this._regions;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._regions = undefined;
            this._roleArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._regions = value.regions;
            this._roleArn = value.roleArn;
        }
    }
    get regions() {
        return cdktf.Fn.tolist(this.getListAttribute('regions'));
    }
    set regions(value) {
        this._regions = value;
    }
    resetRegions() {
        this._regions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionsInput() {
        return this._regions;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference = SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference[_j] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationReplicationConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationReplicationConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationReplicationConfigurationList = SecuritylakeDataLakeConfigurationReplicationConfigurationList;
_k = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationReplicationConfigurationList[_k] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationReplicationConfigurationList", version: "19.15.0" };
function securitylakeDataLakeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_configuration: cdktf.listMapper(securitylakeDataLakeConfigurationEncryptionConfigurationToTerraform, false)(struct.encryptionConfiguration),
        region: cdktf.stringToTerraform(struct.region),
        lifecycle_configuration: cdktf.listMapper(securitylakeDataLakeConfigurationLifecycleConfigurationToTerraform, true)(struct.lifecycleConfiguration),
        replication_configuration: cdktf.listMapper(securitylakeDataLakeConfigurationReplicationConfigurationToTerraform, true)(struct.replicationConfiguration),
    };
}
exports.securitylakeDataLakeConfigurationToTerraform = securitylakeDataLakeConfigurationToTerraform;
function securitylakeDataLakeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        encryption_configuration: {
            value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationEncryptionConfigurationToHclTerraform, false)(struct.encryptionConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeDataLakeConfigurationEncryptionConfigurationList",
        },
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        lifecycle_configuration: {
            value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationLifecycleConfigurationToHclTerraform, true)(struct.lifecycleConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeDataLakeConfigurationLifecycleConfigurationList",
        },
        replication_configuration: {
            value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationReplicationConfigurationToHclTerraform, true)(struct.replicationConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SecuritylakeDataLakeConfigurationReplicationConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeConfigurationToHclTerraform = securitylakeDataLakeConfigurationToHclTerraform;
class SecuritylakeDataLakeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // encryption_configuration - computed: true, optional: true, required: false
        this._encryptionConfiguration = new SecuritylakeDataLakeConfigurationEncryptionConfigurationList(this, "encryption_configuration", false);
        // lifecycle_configuration - computed: false, optional: true, required: false
        this._lifecycleConfiguration = new SecuritylakeDataLakeConfigurationLifecycleConfigurationList(this, "lifecycle_configuration", false);
        // replication_configuration - computed: false, optional: true, required: false
        this._replicationConfiguration = new SecuritylakeDataLakeConfigurationReplicationConfigurationList(this, "replication_configuration", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionConfiguration = this._encryptionConfiguration?.internalValue;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._lifecycleConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.lifecycleConfiguration = this._lifecycleConfiguration?.internalValue;
        }
        if (this._replicationConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicationConfiguration = this._replicationConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._encryptionConfiguration.internalValue = undefined;
            this._region = undefined;
            this._lifecycleConfiguration.internalValue = undefined;
            this._replicationConfiguration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._encryptionConfiguration.internalValue = value.encryptionConfiguration;
            this._region = value.region;
            this._lifecycleConfiguration.internalValue = value.lifecycleConfiguration;
            this._replicationConfiguration.internalValue = value.replicationConfiguration;
        }
    }
    get encryptionConfiguration() {
        return this._encryptionConfiguration;
    }
    putEncryptionConfiguration(value) {
        this._encryptionConfiguration.internalValue = value;
    }
    resetEncryptionConfiguration() {
        this._encryptionConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionConfigurationInput() {
        return this._encryptionConfiguration.internalValue;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get lifecycleConfiguration() {
        return this._lifecycleConfiguration;
    }
    putLifecycleConfiguration(value) {
        this._lifecycleConfiguration.internalValue = value;
    }
    resetLifecycleConfiguration() {
        this._lifecycleConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lifecycleConfigurationInput() {
        return this._lifecycleConfiguration.internalValue;
    }
    get replicationConfiguration() {
        return this._replicationConfiguration;
    }
    putReplicationConfiguration(value) {
        this._replicationConfiguration.internalValue = value;
    }
    resetReplicationConfiguration() {
        this._replicationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationConfigurationInput() {
        return this._replicationConfiguration.internalValue;
    }
}
exports.SecuritylakeDataLakeConfigurationOutputReference = SecuritylakeDataLakeConfigurationOutputReference;
_l = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationOutputReference[_l] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationOutputReference", version: "19.15.0" };
class SecuritylakeDataLakeConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeDataLakeConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeDataLakeConfigurationList = SecuritylakeDataLakeConfigurationList;
_m = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeConfigurationList[_m] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeConfigurationList", version: "19.15.0" };
function securitylakeDataLakeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.securitylakeDataLakeTimeoutsToTerraform = securitylakeDataLakeTimeoutsToTerraform;
function securitylakeDataLakeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeDataLakeTimeoutsToHclTerraform = securitylakeDataLakeTimeoutsToHclTerraform;
class SecuritylakeDataLakeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.SecuritylakeDataLakeTimeoutsOutputReference = SecuritylakeDataLakeTimeoutsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLakeTimeoutsOutputReference[_o] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLakeTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_data_lake aws_securitylake_data_lake}
*/
class SecuritylakeDataLake extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecuritylakeDataLake resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecuritylakeDataLake to import
    * @param importFromId The id of the existing SecuritylakeDataLake that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_data_lake#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecuritylakeDataLake to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securitylake_data_lake", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_data_lake aws_securitylake_data_lake} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecuritylakeDataLakeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securitylake_data_lake',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // configuration - computed: false, optional: true, required: false
        this._configuration = new SecuritylakeDataLakeConfigurationList(this, "configuration", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new SecuritylakeDataLakeTimeoutsOutputReference(this, "timeouts");
        this._metaStoreManagerRoleArn = config.metaStoreManagerRoleArn;
        this._tags = config.tags;
        this._configuration.internalValue = config.configuration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get metaStoreManagerRoleArn() {
        return this.getStringAttribute('meta_store_manager_role_arn');
    }
    set metaStoreManagerRoleArn(value) {
        this._metaStoreManagerRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get metaStoreManagerRoleArnInput() {
        return this._metaStoreManagerRoleArn;
    }
    // s3_bucket_arn - computed: true, optional: false, required: false
    get s3BucketArn() {
        return this.getStringAttribute('s3_bucket_arn');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get configuration() {
        return this._configuration;
    }
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            meta_store_manager_role_arn: cdktf.stringToTerraform(this._metaStoreManagerRoleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            configuration: cdktf.listMapper(securitylakeDataLakeConfigurationToTerraform, true)(this._configuration.internalValue),
            timeouts: securitylakeDataLakeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            meta_store_manager_role_arn: {
                value: cdktf.stringToHclTerraform(this._metaStoreManagerRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            configuration: {
                value: cdktf.listMapperHcl(securitylakeDataLakeConfigurationToHclTerraform, true)(this._configuration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SecuritylakeDataLakeConfigurationList",
            },
            timeouts: {
                value: securitylakeDataLakeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "SecuritylakeDataLakeTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecuritylakeDataLake = SecuritylakeDataLake;
_p = JSII_RTTI_SYMBOL_1;
SecuritylakeDataLake[_p] = { fqn: "@cdktf/provider-aws.securitylakeDataLake.SecuritylakeDataLake", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecuritylakeDataLake.tfResourceType = "aws_securitylake_data_lake";
