// @cdktf/provider-aws/lib/ivschat-room/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IvschatRoom = exports.IvschatRoomTimeoutsOutputReference = exports.ivschatRoomTimeoutsToHclTerraform = exports.ivschatRoomTimeoutsToTerraform = exports.IvschatRoomMessageReviewHandlerOutputReference = exports.ivschatRoomMessageReviewHandlerToHclTerraform = exports.ivschatRoomMessageReviewHandlerToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ivschatRoomMessageReviewHandlerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fallback_result: cdktf.stringToTerraform(struct.fallbackResult),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.ivschatRoomMessageReviewHandlerToTerraform = ivschatRoomMessageReviewHandlerToTerraform;
function ivschatRoomMessageReviewHandlerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        fallback_result: {
            value: cdktf.stringToHclTerraform(struct.fallbackResult),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ivschatRoomMessageReviewHandlerToHclTerraform = ivschatRoomMessageReviewHandlerToHclTerraform;
class IvschatRoomMessageReviewHandlerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fallbackResult !== undefined) {
            hasAnyValues = true;
            internalValueResult.fallbackResult = this._fallbackResult;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fallbackResult = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fallbackResult = value.fallbackResult;
            this._uri = value.uri;
        }
    }
    get fallbackResult() {
        return this.getStringAttribute('fallback_result');
    }
    set fallbackResult(value) {
        this._fallbackResult = value;
    }
    resetFallbackResult() {
        this._fallbackResult = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fallbackResultInput() {
        return this._fallbackResult;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.IvschatRoomMessageReviewHandlerOutputReference = IvschatRoomMessageReviewHandlerOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IvschatRoomMessageReviewHandlerOutputReference[_a] = { fqn: "@cdktf/provider-aws.ivschatRoom.IvschatRoomMessageReviewHandlerOutputReference", version: "19.15.0" };
function ivschatRoomTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.ivschatRoomTimeoutsToTerraform = ivschatRoomTimeoutsToTerraform;
function ivschatRoomTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ivschatRoomTimeoutsToHclTerraform = ivschatRoomTimeoutsToHclTerraform;
class IvschatRoomTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IvschatRoomTimeoutsOutputReference = IvschatRoomTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IvschatRoomTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.ivschatRoom.IvschatRoomTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ivschat_room aws_ivschat_room}
*/
class IvschatRoom extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IvschatRoom resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IvschatRoom to import
    * @param importFromId The id of the existing IvschatRoom that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ivschat_room#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IvschatRoom to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ivschat_room", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ivschat_room aws_ivschat_room} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IvschatRoomConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ivschat_room',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // message_review_handler - computed: false, optional: true, required: false
        this._messageReviewHandler = new IvschatRoomMessageReviewHandlerOutputReference(this, "message_review_handler");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IvschatRoomTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._loggingConfigurationIdentifiers = config.loggingConfigurationIdentifiers;
        this._maximumMessageLength = config.maximumMessageLength;
        this._maximumMessageRatePerSecond = config.maximumMessageRatePerSecond;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._messageReviewHandler.internalValue = config.messageReviewHandler;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get loggingConfigurationIdentifiers() {
        return this.getListAttribute('logging_configuration_identifiers');
    }
    set loggingConfigurationIdentifiers(value) {
        this._loggingConfigurationIdentifiers = value;
    }
    resetLoggingConfigurationIdentifiers() {
        this._loggingConfigurationIdentifiers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingConfigurationIdentifiersInput() {
        return this._loggingConfigurationIdentifiers;
    }
    get maximumMessageLength() {
        return this.getNumberAttribute('maximum_message_length');
    }
    set maximumMessageLength(value) {
        this._maximumMessageLength = value;
    }
    resetMaximumMessageLength() {
        this._maximumMessageLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumMessageLengthInput() {
        return this._maximumMessageLength;
    }
    get maximumMessageRatePerSecond() {
        return this.getNumberAttribute('maximum_message_rate_per_second');
    }
    set maximumMessageRatePerSecond(value) {
        this._maximumMessageRatePerSecond = value;
    }
    resetMaximumMessageRatePerSecond() {
        this._maximumMessageRatePerSecond = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumMessageRatePerSecondInput() {
        return this._maximumMessageRatePerSecond;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get messageReviewHandler() {
        return this._messageReviewHandler;
    }
    putMessageReviewHandler(value) {
        this._messageReviewHandler.internalValue = value;
    }
    resetMessageReviewHandler() {
        this._messageReviewHandler.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messageReviewHandlerInput() {
        return this._messageReviewHandler.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            logging_configuration_identifiers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._loggingConfigurationIdentifiers),
            maximum_message_length: cdktf.numberToTerraform(this._maximumMessageLength),
            maximum_message_rate_per_second: cdktf.numberToTerraform(this._maximumMessageRatePerSecond),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            message_review_handler: ivschatRoomMessageReviewHandlerToTerraform(this._messageReviewHandler.internalValue),
            timeouts: ivschatRoomTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            logging_configuration_identifiers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._loggingConfigurationIdentifiers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            maximum_message_length: {
                value: cdktf.numberToHclTerraform(this._maximumMessageLength),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            maximum_message_rate_per_second: {
                value: cdktf.numberToHclTerraform(this._maximumMessageRatePerSecond),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            message_review_handler: {
                value: ivschatRoomMessageReviewHandlerToHclTerraform(this._messageReviewHandler.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IvschatRoomMessageReviewHandlerList",
            },
            timeouts: {
                value: ivschatRoomTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IvschatRoomTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IvschatRoom = IvschatRoom;
_c = JSII_RTTI_SYMBOL_1;
IvschatRoom[_c] = { fqn: "@cdktf/provider-aws.ivschatRoom.IvschatRoom", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IvschatRoom.tfResourceType = "aws_ivschat_room";
