// @cdktf/provider-aws/lib/kendra-experience/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KendraExperience = exports.KendraExperienceTimeoutsOutputReference = exports.kendraExperienceTimeoutsToHclTerraform = exports.kendraExperienceTimeoutsToTerraform = exports.KendraExperienceConfigurationOutputReference = exports.kendraExperienceConfigurationToHclTerraform = exports.kendraExperienceConfigurationToTerraform = exports.KendraExperienceConfigurationUserIdentityConfigurationOutputReference = exports.kendraExperienceConfigurationUserIdentityConfigurationToHclTerraform = exports.kendraExperienceConfigurationUserIdentityConfigurationToTerraform = exports.KendraExperienceConfigurationContentSourceConfigurationOutputReference = exports.kendraExperienceConfigurationContentSourceConfigurationToHclTerraform = exports.kendraExperienceConfigurationContentSourceConfigurationToTerraform = exports.KendraExperienceEndpointsList = exports.KendraExperienceEndpointsOutputReference = exports.kendraExperienceEndpointsToHclTerraform = exports.kendraExperienceEndpointsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kendraExperienceEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kendraExperienceEndpointsToTerraform = kendraExperienceEndpointsToTerraform;
function kendraExperienceEndpointsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.kendraExperienceEndpointsToHclTerraform = kendraExperienceEndpointsToHclTerraform;
class KendraExperienceEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // endpoint_type - computed: true, optional: false, required: false
    get endpointType() {
        return this.getStringAttribute('endpoint_type');
    }
}
exports.KendraExperienceEndpointsOutputReference = KendraExperienceEndpointsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KendraExperienceEndpointsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceEndpointsOutputReference", version: "19.15.0" };
class KendraExperienceEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KendraExperienceEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KendraExperienceEndpointsList = KendraExperienceEndpointsList;
_b = JSII_RTTI_SYMBOL_1;
KendraExperienceEndpointsList[_b] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceEndpointsList", version: "19.15.0" };
function kendraExperienceConfigurationContentSourceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_source_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.dataSourceIds),
        direct_put_content: cdktf.booleanToTerraform(struct.directPutContent),
        faq_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.faqIds),
    };
}
exports.kendraExperienceConfigurationContentSourceConfigurationToTerraform = kendraExperienceConfigurationContentSourceConfigurationToTerraform;
function kendraExperienceConfigurationContentSourceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_source_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.dataSourceIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        direct_put_content: {
            value: cdktf.booleanToHclTerraform(struct.directPutContent),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        faq_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.faqIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.kendraExperienceConfigurationContentSourceConfigurationToHclTerraform = kendraExperienceConfigurationContentSourceConfigurationToHclTerraform;
class KendraExperienceConfigurationContentSourceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataSourceIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataSourceIds = this._dataSourceIds;
        }
        if (this._directPutContent !== undefined) {
            hasAnyValues = true;
            internalValueResult.directPutContent = this._directPutContent;
        }
        if (this._faqIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.faqIds = this._faqIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataSourceIds = undefined;
            this._directPutContent = undefined;
            this._faqIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataSourceIds = value.dataSourceIds;
            this._directPutContent = value.directPutContent;
            this._faqIds = value.faqIds;
        }
    }
    get dataSourceIds() {
        return cdktf.Fn.tolist(this.getListAttribute('data_source_ids'));
    }
    set dataSourceIds(value) {
        this._dataSourceIds = value;
    }
    resetDataSourceIds() {
        this._dataSourceIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataSourceIdsInput() {
        return this._dataSourceIds;
    }
    get directPutContent() {
        return this.getBooleanAttribute('direct_put_content');
    }
    set directPutContent(value) {
        this._directPutContent = value;
    }
    resetDirectPutContent() {
        this._directPutContent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directPutContentInput() {
        return this._directPutContent;
    }
    get faqIds() {
        return cdktf.Fn.tolist(this.getListAttribute('faq_ids'));
    }
    set faqIds(value) {
        this._faqIds = value;
    }
    resetFaqIds() {
        this._faqIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get faqIdsInput() {
        return this._faqIds;
    }
}
exports.KendraExperienceConfigurationContentSourceConfigurationOutputReference = KendraExperienceConfigurationContentSourceConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
KendraExperienceConfigurationContentSourceConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceConfigurationContentSourceConfigurationOutputReference", version: "19.15.0" };
function kendraExperienceConfigurationUserIdentityConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        identity_attribute_name: cdktf.stringToTerraform(struct.identityAttributeName),
    };
}
exports.kendraExperienceConfigurationUserIdentityConfigurationToTerraform = kendraExperienceConfigurationUserIdentityConfigurationToTerraform;
function kendraExperienceConfigurationUserIdentityConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        identity_attribute_name: {
            value: cdktf.stringToHclTerraform(struct.identityAttributeName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.kendraExperienceConfigurationUserIdentityConfigurationToHclTerraform = kendraExperienceConfigurationUserIdentityConfigurationToHclTerraform;
class KendraExperienceConfigurationUserIdentityConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._identityAttributeName !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityAttributeName = this._identityAttributeName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._identityAttributeName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._identityAttributeName = value.identityAttributeName;
        }
    }
    get identityAttributeName() {
        return this.getStringAttribute('identity_attribute_name');
    }
    set identityAttributeName(value) {
        this._identityAttributeName = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityAttributeNameInput() {
        return this._identityAttributeName;
    }
}
exports.KendraExperienceConfigurationUserIdentityConfigurationOutputReference = KendraExperienceConfigurationUserIdentityConfigurationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
KendraExperienceConfigurationUserIdentityConfigurationOutputReference[_d] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceConfigurationUserIdentityConfigurationOutputReference", version: "19.15.0" };
function kendraExperienceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content_source_configuration: kendraExperienceConfigurationContentSourceConfigurationToTerraform(struct.contentSourceConfiguration),
        user_identity_configuration: kendraExperienceConfigurationUserIdentityConfigurationToTerraform(struct.userIdentityConfiguration),
    };
}
exports.kendraExperienceConfigurationToTerraform = kendraExperienceConfigurationToTerraform;
function kendraExperienceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        content_source_configuration: {
            value: kendraExperienceConfigurationContentSourceConfigurationToHclTerraform(struct.contentSourceConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "KendraExperienceConfigurationContentSourceConfigurationList",
        },
        user_identity_configuration: {
            value: kendraExperienceConfigurationUserIdentityConfigurationToHclTerraform(struct.userIdentityConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "KendraExperienceConfigurationUserIdentityConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.kendraExperienceConfigurationToHclTerraform = kendraExperienceConfigurationToHclTerraform;
class KendraExperienceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // content_source_configuration - computed: false, optional: true, required: false
        this._contentSourceConfiguration = new KendraExperienceConfigurationContentSourceConfigurationOutputReference(this, "content_source_configuration");
        // user_identity_configuration - computed: false, optional: true, required: false
        this._userIdentityConfiguration = new KendraExperienceConfigurationUserIdentityConfigurationOutputReference(this, "user_identity_configuration");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contentSourceConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.contentSourceConfiguration = this._contentSourceConfiguration?.internalValue;
        }
        if (this._userIdentityConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.userIdentityConfiguration = this._userIdentityConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._contentSourceConfiguration.internalValue = undefined;
            this._userIdentityConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._contentSourceConfiguration.internalValue = value.contentSourceConfiguration;
            this._userIdentityConfiguration.internalValue = value.userIdentityConfiguration;
        }
    }
    get contentSourceConfiguration() {
        return this._contentSourceConfiguration;
    }
    putContentSourceConfiguration(value) {
        this._contentSourceConfiguration.internalValue = value;
    }
    resetContentSourceConfiguration() {
        this._contentSourceConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contentSourceConfigurationInput() {
        return this._contentSourceConfiguration.internalValue;
    }
    get userIdentityConfiguration() {
        return this._userIdentityConfiguration;
    }
    putUserIdentityConfiguration(value) {
        this._userIdentityConfiguration.internalValue = value;
    }
    resetUserIdentityConfiguration() {
        this._userIdentityConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userIdentityConfigurationInput() {
        return this._userIdentityConfiguration.internalValue;
    }
}
exports.KendraExperienceConfigurationOutputReference = KendraExperienceConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
KendraExperienceConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceConfigurationOutputReference", version: "19.15.0" };
function kendraExperienceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.kendraExperienceTimeoutsToTerraform = kendraExperienceTimeoutsToTerraform;
function kendraExperienceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.kendraExperienceTimeoutsToHclTerraform = kendraExperienceTimeoutsToHclTerraform;
class KendraExperienceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.KendraExperienceTimeoutsOutputReference = KendraExperienceTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
KendraExperienceTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperienceTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kendra_experience aws_kendra_experience}
*/
class KendraExperience extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a KendraExperience resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the KendraExperience to import
    * @param importFromId The id of the existing KendraExperience that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kendra_experience#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the KendraExperience to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_kendra_experience", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kendra_experience aws_kendra_experience} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KendraExperienceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kendra_experience',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoints - computed: true, optional: false, required: false
        this._endpoints = new KendraExperienceEndpointsList(this, "endpoints", true);
        // configuration - computed: false, optional: true, required: false
        this._configuration = new KendraExperienceConfigurationOutputReference(this, "configuration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new KendraExperienceTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._indexId = config.indexId;
        this._name = config.name;
        this._roleArn = config.roleArn;
        this._configuration.internalValue = config.configuration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get endpoints() {
        return this._endpoints;
    }
    // experience_id - computed: true, optional: false, required: false
    get experienceId() {
        return this.getStringAttribute('experience_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get indexId() {
        return this.getStringAttribute('index_id');
    }
    set indexId(value) {
        this._indexId = value;
    }
    // Temporarily expose input value. Use with caution.
    get indexIdInput() {
        return this._indexId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get configuration() {
        return this._configuration;
    }
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            index_id: cdktf.stringToTerraform(this._indexId),
            name: cdktf.stringToTerraform(this._name),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            configuration: kendraExperienceConfigurationToTerraform(this._configuration.internalValue),
            timeouts: kendraExperienceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            index_id: {
                value: cdktf.stringToHclTerraform(this._indexId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            configuration: {
                value: kendraExperienceConfigurationToHclTerraform(this._configuration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "KendraExperienceConfigurationList",
            },
            timeouts: {
                value: kendraExperienceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "KendraExperienceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.KendraExperience = KendraExperience;
_g = JSII_RTTI_SYMBOL_1;
KendraExperience[_g] = { fqn: "@cdktf/provider-aws.kendraExperience.KendraExperience", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
KendraExperience.tfResourceType = "aws_kendra_experience";
