// @cdktf/provider-aws/lib/ami-launch-permission/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmiLaunchPermission = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ami_launch_permission aws_ami_launch_permission}
*/
class AmiLaunchPermission extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AmiLaunchPermission resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AmiLaunchPermission to import
    * @param importFromId The id of the existing AmiLaunchPermission that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ami_launch_permission#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AmiLaunchPermission to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ami_launch_permission", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ami_launch_permission aws_ami_launch_permission} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AmiLaunchPermissionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ami_launch_permission',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accountId = config.accountId;
        this._group = config.group;
        this._id = config.id;
        this._imageId = config.imageId;
        this._organizationArn = config.organizationArn;
        this._organizationalUnitArn = config.organizationalUnitArn;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get group() {
        return this.getStringAttribute('group');
    }
    set group(value) {
        this._group = value;
    }
    resetGroup() {
        this._group = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupInput() {
        return this._group;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    set imageId(value) {
        this._imageId = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageIdInput() {
        return this._imageId;
    }
    get organizationArn() {
        return this.getStringAttribute('organization_arn');
    }
    set organizationArn(value) {
        this._organizationArn = value;
    }
    resetOrganizationArn() {
        this._organizationArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationArnInput() {
        return this._organizationArn;
    }
    get organizationalUnitArn() {
        return this.getStringAttribute('organizational_unit_arn');
    }
    set organizationalUnitArn(value) {
        this._organizationalUnitArn = value;
    }
    resetOrganizationalUnitArn() {
        this._organizationalUnitArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitArnInput() {
        return this._organizationalUnitArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            group: cdktf.stringToTerraform(this._group),
            id: cdktf.stringToTerraform(this._id),
            image_id: cdktf.stringToTerraform(this._imageId),
            organization_arn: cdktf.stringToTerraform(this._organizationArn),
            organizational_unit_arn: cdktf.stringToTerraform(this._organizationalUnitArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            account_id: {
                value: cdktf.stringToHclTerraform(this._accountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            group: {
                value: cdktf.stringToHclTerraform(this._group),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            image_id: {
                value: cdktf.stringToHclTerraform(this._imageId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organization_arn: {
                value: cdktf.stringToHclTerraform(this._organizationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organizational_unit_arn: {
                value: cdktf.stringToHclTerraform(this._organizationalUnitArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AmiLaunchPermission = AmiLaunchPermission;
_a = JSII_RTTI_SYMBOL_1;
AmiLaunchPermission[_a] = { fqn: "@cdktf/provider-aws.amiLaunchPermission.AmiLaunchPermission", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AmiLaunchPermission.tfResourceType = "aws_ami_launch_permission";
