// @cdktf/provider-aws/lib/data-aws-alb-listener/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAlbListener = exports.DataAwsAlbListenerTimeoutsOutputReference = exports.dataAwsAlbListenerTimeoutsToHclTerraform = exports.dataAwsAlbListenerTimeoutsToTerraform = exports.DataAwsAlbListenerMutualAuthenticationList = exports.DataAwsAlbListenerMutualAuthenticationOutputReference = exports.dataAwsAlbListenerMutualAuthenticationToHclTerraform = exports.dataAwsAlbListenerMutualAuthenticationToTerraform = exports.DataAwsAlbListenerDefaultActionList = exports.DataAwsAlbListenerDefaultActionOutputReference = exports.dataAwsAlbListenerDefaultActionToHclTerraform = exports.dataAwsAlbListenerDefaultActionToTerraform = exports.DataAwsAlbListenerDefaultActionRedirectList = exports.DataAwsAlbListenerDefaultActionRedirectOutputReference = exports.dataAwsAlbListenerDefaultActionRedirectToHclTerraform = exports.dataAwsAlbListenerDefaultActionRedirectToTerraform = exports.DataAwsAlbListenerDefaultActionForwardList = exports.DataAwsAlbListenerDefaultActionForwardOutputReference = exports.dataAwsAlbListenerDefaultActionForwardToHclTerraform = exports.dataAwsAlbListenerDefaultActionForwardToTerraform = exports.DataAwsAlbListenerDefaultActionForwardTargetGroupList = exports.DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference = exports.dataAwsAlbListenerDefaultActionForwardTargetGroupToHclTerraform = exports.dataAwsAlbListenerDefaultActionForwardTargetGroupToTerraform = exports.DataAwsAlbListenerDefaultActionForwardStickinessList = exports.DataAwsAlbListenerDefaultActionForwardStickinessOutputReference = exports.dataAwsAlbListenerDefaultActionForwardStickinessToHclTerraform = exports.dataAwsAlbListenerDefaultActionForwardStickinessToTerraform = exports.DataAwsAlbListenerDefaultActionFixedResponseList = exports.DataAwsAlbListenerDefaultActionFixedResponseOutputReference = exports.dataAwsAlbListenerDefaultActionFixedResponseToHclTerraform = exports.dataAwsAlbListenerDefaultActionFixedResponseToTerraform = exports.DataAwsAlbListenerDefaultActionAuthenticateOidcList = exports.DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference = exports.dataAwsAlbListenerDefaultActionAuthenticateOidcToHclTerraform = exports.dataAwsAlbListenerDefaultActionAuthenticateOidcToTerraform = exports.DataAwsAlbListenerDefaultActionAuthenticateCognitoList = exports.DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference = exports.dataAwsAlbListenerDefaultActionAuthenticateCognitoToHclTerraform = exports.dataAwsAlbListenerDefaultActionAuthenticateCognitoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAlbListenerDefaultActionAuthenticateCognitoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionAuthenticateCognitoToTerraform = dataAwsAlbListenerDefaultActionAuthenticateCognitoToTerraform;
function dataAwsAlbListenerDefaultActionAuthenticateCognitoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionAuthenticateCognitoToHclTerraform = dataAwsAlbListenerDefaultActionAuthenticateCognitoToHclTerraform;
class DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authentication_request_extra_params - computed: true, optional: false, required: false
        this._authenticationRequestExtraParams = new cdktf.StringMap(this, "authentication_request_extra_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticationRequestExtraParams() {
        return this._authenticationRequestExtraParams;
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // user_pool_arn - computed: true, optional: false, required: false
    get userPoolArn() {
        return this.getStringAttribute('user_pool_arn');
    }
    // user_pool_client_id - computed: true, optional: false, required: false
    get userPoolClientId() {
        return this.getStringAttribute('user_pool_client_id');
    }
    // user_pool_domain - computed: true, optional: false, required: false
    get userPoolDomain() {
        return this.getStringAttribute('user_pool_domain');
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference = DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionAuthenticateCognitoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionAuthenticateCognitoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateCognitoList = DataAwsAlbListenerDefaultActionAuthenticateCognitoList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateCognitoList[_b] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionAuthenticateCognitoList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionAuthenticateOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionAuthenticateOidcToTerraform = dataAwsAlbListenerDefaultActionAuthenticateOidcToTerraform;
function dataAwsAlbListenerDefaultActionAuthenticateOidcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionAuthenticateOidcToHclTerraform = dataAwsAlbListenerDefaultActionAuthenticateOidcToHclTerraform;
class DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authentication_request_extra_params - computed: true, optional: false, required: false
        this._authenticationRequestExtraParams = new cdktf.StringMap(this, "authentication_request_extra_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticationRequestExtraParams() {
        return this._authenticationRequestExtraParams;
    }
    // authorization_endpoint - computed: true, optional: false, required: false
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    // client_id - computed: true, optional: false, required: false
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    // client_secret - computed: true, optional: false, required: false
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // token_endpoint - computed: true, optional: false, required: false
    get tokenEndpoint() {
        return this.getStringAttribute('token_endpoint');
    }
    // user_info_endpoint - computed: true, optional: false, required: false
    get userInfoEndpoint() {
        return this.getStringAttribute('user_info_endpoint');
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference = DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionAuthenticateOidcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionAuthenticateOidcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateOidcList = DataAwsAlbListenerDefaultActionAuthenticateOidcList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateOidcList[_d] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionAuthenticateOidcList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionFixedResponseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionFixedResponseToTerraform = dataAwsAlbListenerDefaultActionFixedResponseToTerraform;
function dataAwsAlbListenerDefaultActionFixedResponseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionFixedResponseToHclTerraform = dataAwsAlbListenerDefaultActionFixedResponseToHclTerraform;
class DataAwsAlbListenerDefaultActionFixedResponseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // content_type - computed: true, optional: false, required: false
    get contentType() {
        return this.getStringAttribute('content_type');
    }
    // message_body - computed: true, optional: false, required: false
    get messageBody() {
        return this.getStringAttribute('message_body');
    }
    // status_code - computed: true, optional: false, required: false
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsAlbListenerDefaultActionFixedResponseOutputReference = DataAwsAlbListenerDefaultActionFixedResponseOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionFixedResponseOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionFixedResponseOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionFixedResponseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionFixedResponseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionFixedResponseList = DataAwsAlbListenerDefaultActionFixedResponseList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionFixedResponseList[_f] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionFixedResponseList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionForwardStickinessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionForwardStickinessToTerraform = dataAwsAlbListenerDefaultActionForwardStickinessToTerraform;
function dataAwsAlbListenerDefaultActionForwardStickinessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionForwardStickinessToHclTerraform = dataAwsAlbListenerDefaultActionForwardStickinessToHclTerraform;
class DataAwsAlbListenerDefaultActionForwardStickinessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // duration - computed: true, optional: false, required: false
    get duration() {
        return this.getNumberAttribute('duration');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsAlbListenerDefaultActionForwardStickinessOutputReference = DataAwsAlbListenerDefaultActionForwardStickinessOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardStickinessOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardStickinessOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionForwardStickinessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionForwardStickinessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionForwardStickinessList = DataAwsAlbListenerDefaultActionForwardStickinessList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardStickinessList[_h] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardStickinessList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionForwardTargetGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionForwardTargetGroupToTerraform = dataAwsAlbListenerDefaultActionForwardTargetGroupToTerraform;
function dataAwsAlbListenerDefaultActionForwardTargetGroupToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionForwardTargetGroupToHclTerraform = dataAwsAlbListenerDefaultActionForwardTargetGroupToHclTerraform;
class DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // weight - computed: true, optional: false, required: false
    get weight() {
        return this.getNumberAttribute('weight');
    }
}
exports.DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference = DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionForwardTargetGroupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionForwardTargetGroupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionForwardTargetGroupList = DataAwsAlbListenerDefaultActionForwardTargetGroupList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardTargetGroupList[_k] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardTargetGroupList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionForwardToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionForwardToTerraform = dataAwsAlbListenerDefaultActionForwardToTerraform;
function dataAwsAlbListenerDefaultActionForwardToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionForwardToHclTerraform = dataAwsAlbListenerDefaultActionForwardToHclTerraform;
class DataAwsAlbListenerDefaultActionForwardOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // stickiness - computed: true, optional: false, required: false
        this._stickiness = new DataAwsAlbListenerDefaultActionForwardStickinessList(this, "stickiness", false);
        // target_group - computed: true, optional: false, required: false
        this._targetGroup = new DataAwsAlbListenerDefaultActionForwardTargetGroupList(this, "target_group", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get stickiness() {
        return this._stickiness;
    }
    get targetGroup() {
        return this._targetGroup;
    }
}
exports.DataAwsAlbListenerDefaultActionForwardOutputReference = DataAwsAlbListenerDefaultActionForwardOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionForwardList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionForwardOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionForwardList = DataAwsAlbListenerDefaultActionForwardList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardList[_m] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionForwardList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionRedirectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionRedirectToTerraform = dataAwsAlbListenerDefaultActionRedirectToTerraform;
function dataAwsAlbListenerDefaultActionRedirectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionRedirectToHclTerraform = dataAwsAlbListenerDefaultActionRedirectToHclTerraform;
class DataAwsAlbListenerDefaultActionRedirectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getStringAttribute('port');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // query - computed: true, optional: false, required: false
    get query() {
        return this.getStringAttribute('query');
    }
    // status_code - computed: true, optional: false, required: false
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsAlbListenerDefaultActionRedirectOutputReference = DataAwsAlbListenerDefaultActionRedirectOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionRedirectOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionRedirectOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionRedirectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionRedirectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionRedirectList = DataAwsAlbListenerDefaultActionRedirectList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionRedirectList[_p] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionRedirectList", version: "19.15.0" };
function dataAwsAlbListenerDefaultActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerDefaultActionToTerraform = dataAwsAlbListenerDefaultActionToTerraform;
function dataAwsAlbListenerDefaultActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerDefaultActionToHclTerraform = dataAwsAlbListenerDefaultActionToHclTerraform;
class DataAwsAlbListenerDefaultActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authenticate_cognito - computed: true, optional: false, required: false
        this._authenticateCognito = new DataAwsAlbListenerDefaultActionAuthenticateCognitoList(this, "authenticate_cognito", false);
        // authenticate_oidc - computed: true, optional: false, required: false
        this._authenticateOidc = new DataAwsAlbListenerDefaultActionAuthenticateOidcList(this, "authenticate_oidc", false);
        // fixed_response - computed: true, optional: false, required: false
        this._fixedResponse = new DataAwsAlbListenerDefaultActionFixedResponseList(this, "fixed_response", false);
        // forward - computed: true, optional: false, required: false
        this._forward = new DataAwsAlbListenerDefaultActionForwardList(this, "forward", false);
        // redirect - computed: true, optional: false, required: false
        this._redirect = new DataAwsAlbListenerDefaultActionRedirectList(this, "redirect", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authenticateCognito() {
        return this._authenticateCognito;
    }
    get authenticateOidc() {
        return this._authenticateOidc;
    }
    get fixedResponse() {
        return this._fixedResponse;
    }
    get forward() {
        return this._forward;
    }
    // order - computed: true, optional: false, required: false
    get order() {
        return this.getNumberAttribute('order');
    }
    get redirect() {
        return this._redirect;
    }
    // target_group_arn - computed: true, optional: false, required: false
    get targetGroupArn() {
        return this.getStringAttribute('target_group_arn');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsAlbListenerDefaultActionOutputReference = DataAwsAlbListenerDefaultActionOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionOutputReference", version: "19.15.0" };
class DataAwsAlbListenerDefaultActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerDefaultActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerDefaultActionList = DataAwsAlbListenerDefaultActionList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionList[_r] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerDefaultActionList", version: "19.15.0" };
function dataAwsAlbListenerMutualAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAlbListenerMutualAuthenticationToTerraform = dataAwsAlbListenerMutualAuthenticationToTerraform;
function dataAwsAlbListenerMutualAuthenticationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAlbListenerMutualAuthenticationToHclTerraform = dataAwsAlbListenerMutualAuthenticationToHclTerraform;
class DataAwsAlbListenerMutualAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ignore_client_certificate_expiry - computed: true, optional: false, required: false
    get ignoreClientCertificateExpiry() {
        return this.getBooleanAttribute('ignore_client_certificate_expiry');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // trust_store_arn - computed: true, optional: false, required: false
    get trustStoreArn() {
        return this.getStringAttribute('trust_store_arn');
    }
}
exports.DataAwsAlbListenerMutualAuthenticationOutputReference = DataAwsAlbListenerMutualAuthenticationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerMutualAuthenticationOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerMutualAuthenticationOutputReference", version: "19.15.0" };
class DataAwsAlbListenerMutualAuthenticationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAlbListenerMutualAuthenticationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAlbListenerMutualAuthenticationList = DataAwsAlbListenerMutualAuthenticationList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerMutualAuthenticationList[_t] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerMutualAuthenticationList", version: "19.15.0" };
function dataAwsAlbListenerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsAlbListenerTimeoutsToTerraform = dataAwsAlbListenerTimeoutsToTerraform;
function dataAwsAlbListenerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsAlbListenerTimeoutsToHclTerraform = dataAwsAlbListenerTimeoutsToHclTerraform;
class DataAwsAlbListenerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsAlbListenerTimeoutsOutputReference = DataAwsAlbListenerTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListenerTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/alb_listener aws_alb_listener}
*/
class DataAwsAlbListener extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsAlbListener resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsAlbListener to import
    * @param importFromId The id of the existing DataAwsAlbListener that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/alb_listener#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsAlbListener to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_alb_listener", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/alb_listener aws_alb_listener} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAlbListenerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_alb_listener',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // default_action - computed: true, optional: false, required: false
        this._defaultAction = new DataAwsAlbListenerDefaultActionList(this, "default_action", false);
        // mutual_authentication - computed: true, optional: false, required: false
        this._mutualAuthentication = new DataAwsAlbListenerMutualAuthenticationList(this, "mutual_authentication", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsAlbListenerTimeoutsOutputReference(this, "timeouts");
        this._arn = config.arn;
        this._id = config.id;
        this._loadBalancerArn = config.loadBalancerArn;
        this._port = config.port;
        this._tags = config.tags;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // alpn_policy - computed: true, optional: false, required: false
    get alpnPolicy() {
        return this.getStringAttribute('alpn_policy');
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // certificate_arn - computed: true, optional: false, required: false
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    get defaultAction() {
        return this._defaultAction;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get loadBalancerArn() {
        return this.getStringAttribute('load_balancer_arn');
    }
    set loadBalancerArn(value) {
        this._loadBalancerArn = value;
    }
    resetLoadBalancerArn() {
        this._loadBalancerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerArnInput() {
        return this._loadBalancerArn;
    }
    get mutualAuthentication() {
        return this._mutualAuthentication;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // ssl_policy - computed: true, optional: false, required: false
    get sslPolicy() {
        return this.getStringAttribute('ssl_policy');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            load_balancer_arn: cdktf.stringToTerraform(this._loadBalancerArn),
            port: cdktf.numberToTerraform(this._port),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            timeouts: dataAwsAlbListenerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancer_arn: {
                value: cdktf.stringToHclTerraform(this._loadBalancerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            port: {
                value: cdktf.numberToHclTerraform(this._port),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeouts: {
                value: dataAwsAlbListenerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsAlbListenerTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsAlbListener = DataAwsAlbListener;
_v = JSII_RTTI_SYMBOL_1;
DataAwsAlbListener[_v] = { fqn: "@cdktf/provider-aws.dataAwsAlbListener.DataAwsAlbListener", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAlbListener.tfResourceType = "aws_alb_listener";
