// @cdktf/provider-aws/lib/redshift-cluster-iam-roles/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftClusterIamRoles = exports.RedshiftClusterIamRolesTimeoutsOutputReference = exports.redshiftClusterIamRolesTimeoutsToHclTerraform = exports.redshiftClusterIamRolesTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function redshiftClusterIamRolesTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.redshiftClusterIamRolesTimeoutsToTerraform = redshiftClusterIamRolesTimeoutsToTerraform;
function redshiftClusterIamRolesTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redshiftClusterIamRolesTimeoutsToHclTerraform = redshiftClusterIamRolesTimeoutsToHclTerraform;
class RedshiftClusterIamRolesTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.RedshiftClusterIamRolesTimeoutsOutputReference = RedshiftClusterIamRolesTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RedshiftClusterIamRolesTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.redshiftClusterIamRoles.RedshiftClusterIamRolesTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_cluster_iam_roles aws_redshift_cluster_iam_roles}
*/
class RedshiftClusterIamRoles extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftClusterIamRoles resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftClusterIamRoles to import
    * @param importFromId The id of the existing RedshiftClusterIamRoles that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_cluster_iam_roles#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftClusterIamRoles to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshift_cluster_iam_roles", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshift_cluster_iam_roles aws_redshift_cluster_iam_roles} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftClusterIamRolesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_cluster_iam_roles',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RedshiftClusterIamRolesTimeoutsOutputReference(this, "timeouts");
        this._clusterIdentifier = config.clusterIdentifier;
        this._defaultIamRoleArn = config.defaultIamRoleArn;
        this._iamRoleArns = config.iamRoleArns;
        this._id = config.id;
        this._timeouts.internalValue = config.timeouts;
    }
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    get defaultIamRoleArn() {
        return this.getStringAttribute('default_iam_role_arn');
    }
    set defaultIamRoleArn(value) {
        this._defaultIamRoleArn = value;
    }
    resetDefaultIamRoleArn() {
        this._defaultIamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultIamRoleArnInput() {
        return this._defaultIamRoleArn;
    }
    get iamRoleArns() {
        return cdktf.Fn.tolist(this.getListAttribute('iam_role_arns'));
    }
    set iamRoleArns(value) {
        this._iamRoleArns = value;
    }
    resetIamRoleArns() {
        this._iamRoleArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleArnsInput() {
        return this._iamRoleArns;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            default_iam_role_arn: cdktf.stringToTerraform(this._defaultIamRoleArn),
            iam_role_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._iamRoleArns),
            id: cdktf.stringToTerraform(this._id),
            timeouts: redshiftClusterIamRolesTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_identifier: {
                value: cdktf.stringToHclTerraform(this._clusterIdentifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            default_iam_role_arn: {
                value: cdktf.stringToHclTerraform(this._defaultIamRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_role_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._iamRoleArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: redshiftClusterIamRolesTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "RedshiftClusterIamRolesTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftClusterIamRoles = RedshiftClusterIamRoles;
_b = JSII_RTTI_SYMBOL_1;
RedshiftClusterIamRoles[_b] = { fqn: "@cdktf/provider-aws.redshiftClusterIamRoles.RedshiftClusterIamRoles", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftClusterIamRoles.tfResourceType = "aws_redshift_cluster_iam_roles";
