// @cdktf/provider-aws/lib/glacier-vault-lock/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlacierVaultLock = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glacier_vault_lock aws_glacier_vault_lock}
*/
class GlacierVaultLock extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GlacierVaultLock resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GlacierVaultLock to import
    * @param importFromId The id of the existing GlacierVaultLock that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glacier_vault_lock#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GlacierVaultLock to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_glacier_vault_lock", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glacier_vault_lock aws_glacier_vault_lock} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlacierVaultLockConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glacier_vault_lock',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._completeLock = config.completeLock;
        this._id = config.id;
        this._ignoreDeletionError = config.ignoreDeletionError;
        this._policy = config.policy;
        this._vaultName = config.vaultName;
    }
    get completeLock() {
        return this.getBooleanAttribute('complete_lock');
    }
    set completeLock(value) {
        this._completeLock = value;
    }
    // Temporarily expose input value. Use with caution.
    get completeLockInput() {
        return this._completeLock;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoreDeletionError() {
        return this.getBooleanAttribute('ignore_deletion_error');
    }
    set ignoreDeletionError(value) {
        this._ignoreDeletionError = value;
    }
    resetIgnoreDeletionError() {
        this._ignoreDeletionError = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreDeletionErrorInput() {
        return this._ignoreDeletionError;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    get vaultName() {
        return this.getStringAttribute('vault_name');
    }
    set vaultName(value) {
        this._vaultName = value;
    }
    // Temporarily expose input value. Use with caution.
    get vaultNameInput() {
        return this._vaultName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            complete_lock: cdktf.booleanToTerraform(this._completeLock),
            id: cdktf.stringToTerraform(this._id),
            ignore_deletion_error: cdktf.booleanToTerraform(this._ignoreDeletionError),
            policy: cdktf.stringToTerraform(this._policy),
            vault_name: cdktf.stringToTerraform(this._vaultName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            complete_lock: {
                value: cdktf.booleanToHclTerraform(this._completeLock),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ignore_deletion_error: {
                value: cdktf.booleanToHclTerraform(this._ignoreDeletionError),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            policy: {
                value: cdktf.stringToHclTerraform(this._policy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vault_name: {
                value: cdktf.stringToHclTerraform(this._vaultName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GlacierVaultLock = GlacierVaultLock;
_a = JSII_RTTI_SYMBOL_1;
GlacierVaultLock[_a] = { fqn: "@cdktf/provider-aws.glacierVaultLock.GlacierVaultLock", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GlacierVaultLock.tfResourceType = "aws_glacier_vault_lock";
