// @cdktf/provider-aws/lib/codepipeline-custom-action-type/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodepipelineCustomActionType = exports.CodepipelineCustomActionTypeSettingsOutputReference = exports.codepipelineCustomActionTypeSettingsToHclTerraform = exports.codepipelineCustomActionTypeSettingsToTerraform = exports.CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference = exports.codepipelineCustomActionTypeOutputArtifactDetailsToHclTerraform = exports.codepipelineCustomActionTypeOutputArtifactDetailsToTerraform = exports.CodepipelineCustomActionTypeInputArtifactDetailsOutputReference = exports.codepipelineCustomActionTypeInputArtifactDetailsToHclTerraform = exports.codepipelineCustomActionTypeInputArtifactDetailsToTerraform = exports.CodepipelineCustomActionTypeConfigurationPropertyList = exports.CodepipelineCustomActionTypeConfigurationPropertyOutputReference = exports.codepipelineCustomActionTypeConfigurationPropertyToHclTerraform = exports.codepipelineCustomActionTypeConfigurationPropertyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codepipelineCustomActionTypeConfigurationPropertyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        key: cdktf.booleanToTerraform(struct.key),
        name: cdktf.stringToTerraform(struct.name),
        queryable: cdktf.booleanToTerraform(struct.queryable),
        required: cdktf.booleanToTerraform(struct.required),
        secret: cdktf.booleanToTerraform(struct.secret),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codepipelineCustomActionTypeConfigurationPropertyToTerraform = codepipelineCustomActionTypeConfigurationPropertyToTerraform;
function codepipelineCustomActionTypeConfigurationPropertyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        key: {
            value: cdktf.booleanToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        queryable: {
            value: cdktf.booleanToHclTerraform(struct.queryable),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        required: {
            value: cdktf.booleanToHclTerraform(struct.required),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        secret: {
            value: cdktf.booleanToHclTerraform(struct.secret),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineCustomActionTypeConfigurationPropertyToHclTerraform = codepipelineCustomActionTypeConfigurationPropertyToHclTerraform;
class CodepipelineCustomActionTypeConfigurationPropertyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._queryable !== undefined) {
            hasAnyValues = true;
            internalValueResult.queryable = this._queryable;
        }
        if (this._required !== undefined) {
            hasAnyValues = true;
            internalValueResult.required = this._required;
        }
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._key = undefined;
            this._name = undefined;
            this._queryable = undefined;
            this._required = undefined;
            this._secret = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._key = value.key;
            this._name = value.name;
            this._queryable = value.queryable;
            this._required = value.required;
            this._secret = value.secret;
            this._type = value.type;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get key() {
        return this.getBooleanAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get queryable() {
        return this.getBooleanAttribute('queryable');
    }
    set queryable(value) {
        this._queryable = value;
    }
    resetQueryable() {
        this._queryable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryableInput() {
        return this._queryable;
    }
    get required() {
        return this.getBooleanAttribute('required');
    }
    set required(value) {
        this._required = value;
    }
    // Temporarily expose input value. Use with caution.
    get requiredInput() {
        return this._required;
    }
    get secret() {
        return this.getBooleanAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CodepipelineCustomActionTypeConfigurationPropertyOutputReference = CodepipelineCustomActionTypeConfigurationPropertyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionTypeConfigurationPropertyOutputReference[_a] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionTypeConfigurationPropertyOutputReference", version: "19.15.0" };
class CodepipelineCustomActionTypeConfigurationPropertyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineCustomActionTypeConfigurationPropertyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineCustomActionTypeConfigurationPropertyList = CodepipelineCustomActionTypeConfigurationPropertyList;
_b = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionTypeConfigurationPropertyList[_b] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionTypeConfigurationPropertyList", version: "19.15.0" };
function codepipelineCustomActionTypeInputArtifactDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_count: cdktf.numberToTerraform(struct.maximumCount),
        minimum_count: cdktf.numberToTerraform(struct.minimumCount),
    };
}
exports.codepipelineCustomActionTypeInputArtifactDetailsToTerraform = codepipelineCustomActionTypeInputArtifactDetailsToTerraform;
function codepipelineCustomActionTypeInputArtifactDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_count: {
            value: cdktf.numberToHclTerraform(struct.maximumCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minimum_count: {
            value: cdktf.numberToHclTerraform(struct.minimumCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineCustomActionTypeInputArtifactDetailsToHclTerraform = codepipelineCustomActionTypeInputArtifactDetailsToHclTerraform;
class CodepipelineCustomActionTypeInputArtifactDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumCount = this._maximumCount;
        }
        if (this._minimumCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimumCount = this._minimumCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumCount = undefined;
            this._minimumCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumCount = value.maximumCount;
            this._minimumCount = value.minimumCount;
        }
    }
    get maximumCount() {
        return this.getNumberAttribute('maximum_count');
    }
    set maximumCount(value) {
        this._maximumCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get maximumCountInput() {
        return this._maximumCount;
    }
    get minimumCount() {
        return this.getNumberAttribute('minimum_count');
    }
    set minimumCount(value) {
        this._minimumCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get minimumCountInput() {
        return this._minimumCount;
    }
}
exports.CodepipelineCustomActionTypeInputArtifactDetailsOutputReference = CodepipelineCustomActionTypeInputArtifactDetailsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionTypeInputArtifactDetailsOutputReference[_c] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionTypeInputArtifactDetailsOutputReference", version: "19.15.0" };
function codepipelineCustomActionTypeOutputArtifactDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_count: cdktf.numberToTerraform(struct.maximumCount),
        minimum_count: cdktf.numberToTerraform(struct.minimumCount),
    };
}
exports.codepipelineCustomActionTypeOutputArtifactDetailsToTerraform = codepipelineCustomActionTypeOutputArtifactDetailsToTerraform;
function codepipelineCustomActionTypeOutputArtifactDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_count: {
            value: cdktf.numberToHclTerraform(struct.maximumCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minimum_count: {
            value: cdktf.numberToHclTerraform(struct.minimumCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineCustomActionTypeOutputArtifactDetailsToHclTerraform = codepipelineCustomActionTypeOutputArtifactDetailsToHclTerraform;
class CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumCount = this._maximumCount;
        }
        if (this._minimumCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimumCount = this._minimumCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumCount = undefined;
            this._minimumCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumCount = value.maximumCount;
            this._minimumCount = value.minimumCount;
        }
    }
    get maximumCount() {
        return this.getNumberAttribute('maximum_count');
    }
    set maximumCount(value) {
        this._maximumCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get maximumCountInput() {
        return this._maximumCount;
    }
    get minimumCount() {
        return this.getNumberAttribute('minimum_count');
    }
    set minimumCount(value) {
        this._minimumCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get minimumCountInput() {
        return this._minimumCount;
    }
}
exports.CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference = CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference[_d] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference", version: "19.15.0" };
function codepipelineCustomActionTypeSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        entity_url_template: cdktf.stringToTerraform(struct.entityUrlTemplate),
        execution_url_template: cdktf.stringToTerraform(struct.executionUrlTemplate),
        revision_url_template: cdktf.stringToTerraform(struct.revisionUrlTemplate),
        third_party_configuration_url: cdktf.stringToTerraform(struct.thirdPartyConfigurationUrl),
    };
}
exports.codepipelineCustomActionTypeSettingsToTerraform = codepipelineCustomActionTypeSettingsToTerraform;
function codepipelineCustomActionTypeSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        entity_url_template: {
            value: cdktf.stringToHclTerraform(struct.entityUrlTemplate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        execution_url_template: {
            value: cdktf.stringToHclTerraform(struct.executionUrlTemplate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        revision_url_template: {
            value: cdktf.stringToHclTerraform(struct.revisionUrlTemplate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        third_party_configuration_url: {
            value: cdktf.stringToHclTerraform(struct.thirdPartyConfigurationUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.codepipelineCustomActionTypeSettingsToHclTerraform = codepipelineCustomActionTypeSettingsToHclTerraform;
class CodepipelineCustomActionTypeSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._entityUrlTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityUrlTemplate = this._entityUrlTemplate;
        }
        if (this._executionUrlTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionUrlTemplate = this._executionUrlTemplate;
        }
        if (this._revisionUrlTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.revisionUrlTemplate = this._revisionUrlTemplate;
        }
        if (this._thirdPartyConfigurationUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.thirdPartyConfigurationUrl = this._thirdPartyConfigurationUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._entityUrlTemplate = undefined;
            this._executionUrlTemplate = undefined;
            this._revisionUrlTemplate = undefined;
            this._thirdPartyConfigurationUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._entityUrlTemplate = value.entityUrlTemplate;
            this._executionUrlTemplate = value.executionUrlTemplate;
            this._revisionUrlTemplate = value.revisionUrlTemplate;
            this._thirdPartyConfigurationUrl = value.thirdPartyConfigurationUrl;
        }
    }
    get entityUrlTemplate() {
        return this.getStringAttribute('entity_url_template');
    }
    set entityUrlTemplate(value) {
        this._entityUrlTemplate = value;
    }
    resetEntityUrlTemplate() {
        this._entityUrlTemplate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entityUrlTemplateInput() {
        return this._entityUrlTemplate;
    }
    get executionUrlTemplate() {
        return this.getStringAttribute('execution_url_template');
    }
    set executionUrlTemplate(value) {
        this._executionUrlTemplate = value;
    }
    resetExecutionUrlTemplate() {
        this._executionUrlTemplate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionUrlTemplateInput() {
        return this._executionUrlTemplate;
    }
    get revisionUrlTemplate() {
        return this.getStringAttribute('revision_url_template');
    }
    set revisionUrlTemplate(value) {
        this._revisionUrlTemplate = value;
    }
    resetRevisionUrlTemplate() {
        this._revisionUrlTemplate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revisionUrlTemplateInput() {
        return this._revisionUrlTemplate;
    }
    get thirdPartyConfigurationUrl() {
        return this.getStringAttribute('third_party_configuration_url');
    }
    set thirdPartyConfigurationUrl(value) {
        this._thirdPartyConfigurationUrl = value;
    }
    resetThirdPartyConfigurationUrl() {
        this._thirdPartyConfigurationUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thirdPartyConfigurationUrlInput() {
        return this._thirdPartyConfigurationUrl;
    }
}
exports.CodepipelineCustomActionTypeSettingsOutputReference = CodepipelineCustomActionTypeSettingsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionTypeSettingsOutputReference[_e] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionTypeSettingsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline_custom_action_type aws_codepipeline_custom_action_type}
*/
class CodepipelineCustomActionType extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CodepipelineCustomActionType resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CodepipelineCustomActionType to import
    * @param importFromId The id of the existing CodepipelineCustomActionType that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline_custom_action_type#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CodepipelineCustomActionType to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_codepipeline_custom_action_type", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codepipeline_custom_action_type aws_codepipeline_custom_action_type} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodepipelineCustomActionTypeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codepipeline_custom_action_type',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // configuration_property - computed: false, optional: true, required: false
        this._configurationProperty = new CodepipelineCustomActionTypeConfigurationPropertyList(this, "configuration_property", false);
        // input_artifact_details - computed: false, optional: false, required: true
        this._inputArtifactDetails = new CodepipelineCustomActionTypeInputArtifactDetailsOutputReference(this, "input_artifact_details");
        // output_artifact_details - computed: false, optional: false, required: true
        this._outputArtifactDetails = new CodepipelineCustomActionTypeOutputArtifactDetailsOutputReference(this, "output_artifact_details");
        // settings - computed: false, optional: true, required: false
        this._settings = new CodepipelineCustomActionTypeSettingsOutputReference(this, "settings");
        this._category = config.category;
        this._id = config.id;
        this._providerName = config.providerName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._version = config.version;
        this._configurationProperty.internalValue = config.configurationProperty;
        this._inputArtifactDetails.internalValue = config.inputArtifactDetails;
        this._outputArtifactDetails.internalValue = config.outputArtifactDetails;
        this._settings.internalValue = config.settings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get category() {
        return this.getStringAttribute('category');
    }
    set category(value) {
        this._category = value;
    }
    // Temporarily expose input value. Use with caution.
    get categoryInput() {
        return this._category;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner - computed: true, optional: false, required: false
    get owner() {
        return this.getStringAttribute('owner');
    }
    get providerName() {
        return this.getStringAttribute('provider_name');
    }
    set providerName(value) {
        this._providerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerNameInput() {
        return this._providerName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get configurationProperty() {
        return this._configurationProperty;
    }
    putConfigurationProperty(value) {
        this._configurationProperty.internalValue = value;
    }
    resetConfigurationProperty() {
        this._configurationProperty.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationPropertyInput() {
        return this._configurationProperty.internalValue;
    }
    get inputArtifactDetails() {
        return this._inputArtifactDetails;
    }
    putInputArtifactDetails(value) {
        this._inputArtifactDetails.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputArtifactDetailsInput() {
        return this._inputArtifactDetails.internalValue;
    }
    get outputArtifactDetails() {
        return this._outputArtifactDetails;
    }
    putOutputArtifactDetails(value) {
        this._outputArtifactDetails.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get outputArtifactDetailsInput() {
        return this._outputArtifactDetails.internalValue;
    }
    get settings() {
        return this._settings;
    }
    putSettings(value) {
        this._settings.internalValue = value;
    }
    resetSettings() {
        this._settings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get settingsInput() {
        return this._settings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            category: cdktf.stringToTerraform(this._category),
            id: cdktf.stringToTerraform(this._id),
            provider_name: cdktf.stringToTerraform(this._providerName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            version: cdktf.stringToTerraform(this._version),
            configuration_property: cdktf.listMapper(codepipelineCustomActionTypeConfigurationPropertyToTerraform, true)(this._configurationProperty.internalValue),
            input_artifact_details: codepipelineCustomActionTypeInputArtifactDetailsToTerraform(this._inputArtifactDetails.internalValue),
            output_artifact_details: codepipelineCustomActionTypeOutputArtifactDetailsToTerraform(this._outputArtifactDetails.internalValue),
            settings: codepipelineCustomActionTypeSettingsToTerraform(this._settings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            category: {
                value: cdktf.stringToHclTerraform(this._category),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            provider_name: {
                value: cdktf.stringToHclTerraform(this._providerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            version: {
                value: cdktf.stringToHclTerraform(this._version),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            configuration_property: {
                value: cdktf.listMapperHcl(codepipelineCustomActionTypeConfigurationPropertyToHclTerraform, true)(this._configurationProperty.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineCustomActionTypeConfigurationPropertyList",
            },
            input_artifact_details: {
                value: codepipelineCustomActionTypeInputArtifactDetailsToHclTerraform(this._inputArtifactDetails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineCustomActionTypeInputArtifactDetailsList",
            },
            output_artifact_details: {
                value: codepipelineCustomActionTypeOutputArtifactDetailsToHclTerraform(this._outputArtifactDetails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineCustomActionTypeOutputArtifactDetailsList",
            },
            settings: {
                value: codepipelineCustomActionTypeSettingsToHclTerraform(this._settings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CodepipelineCustomActionTypeSettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CodepipelineCustomActionType = CodepipelineCustomActionType;
_f = JSII_RTTI_SYMBOL_1;
CodepipelineCustomActionType[_f] = { fqn: "@cdktf/provider-aws.codepipelineCustomActionType.CodepipelineCustomActionType", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CodepipelineCustomActionType.tfResourceType = "aws_codepipeline_custom_action_type";
