// @cdktf/provider-aws/lib/ssoadmin-trusted-token-issuer/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminTrustedTokenIssuer = exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList = exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference = exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToHclTerraform = exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToTerraform = exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList = exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference = exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToHclTerraform = exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        claim_attribute_path: cdktf.stringToTerraform(struct.claimAttributePath),
        identity_store_attribute_path: cdktf.stringToTerraform(struct.identityStoreAttributePath),
        issuer_url: cdktf.stringToTerraform(struct.issuerUrl),
        jwks_retrieval_option: cdktf.stringToTerraform(struct.jwksRetrievalOption),
    };
}
exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToTerraform = ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToTerraform;
function ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        claim_attribute_path: {
            value: cdktf.stringToHclTerraform(struct.claimAttributePath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        identity_store_attribute_path: {
            value: cdktf.stringToHclTerraform(struct.identityStoreAttributePath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        issuer_url: {
            value: cdktf.stringToHclTerraform(struct.issuerUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        jwks_retrieval_option: {
            value: cdktf.stringToHclTerraform(struct.jwksRetrievalOption),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToHclTerraform = ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToHclTerraform;
class SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._claimAttributePath !== undefined) {
            hasAnyValues = true;
            internalValueResult.claimAttributePath = this._claimAttributePath;
        }
        if (this._identityStoreAttributePath !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityStoreAttributePath = this._identityStoreAttributePath;
        }
        if (this._issuerUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuerUrl = this._issuerUrl;
        }
        if (this._jwksRetrievalOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.jwksRetrievalOption = this._jwksRetrievalOption;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._claimAttributePath = undefined;
            this._identityStoreAttributePath = undefined;
            this._issuerUrl = undefined;
            this._jwksRetrievalOption = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._claimAttributePath = value.claimAttributePath;
            this._identityStoreAttributePath = value.identityStoreAttributePath;
            this._issuerUrl = value.issuerUrl;
            this._jwksRetrievalOption = value.jwksRetrievalOption;
        }
    }
    get claimAttributePath() {
        return this.getStringAttribute('claim_attribute_path');
    }
    set claimAttributePath(value) {
        this._claimAttributePath = value;
    }
    // Temporarily expose input value. Use with caution.
    get claimAttributePathInput() {
        return this._claimAttributePath;
    }
    get identityStoreAttributePath() {
        return this.getStringAttribute('identity_store_attribute_path');
    }
    set identityStoreAttributePath(value) {
        this._identityStoreAttributePath = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityStoreAttributePathInput() {
        return this._identityStoreAttributePath;
    }
    get issuerUrl() {
        return this.getStringAttribute('issuer_url');
    }
    set issuerUrl(value) {
        this._issuerUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get issuerUrlInput() {
        return this._issuerUrl;
    }
    get jwksRetrievalOption() {
        return this.getStringAttribute('jwks_retrieval_option');
    }
    set jwksRetrievalOption(value) {
        this._jwksRetrievalOption = value;
    }
    // Temporarily expose input value. Use with caution.
    get jwksRetrievalOptionInput() {
        return this._jwksRetrievalOption;
    }
}
exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference = SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssoadminTrustedTokenIssuer.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference", version: "19.15.0" };
class SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList = SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList[_b] = { fqn: "@cdktf/provider-aws.ssoadminTrustedTokenIssuer.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList", version: "19.15.0" };
function ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        oidc_jwt_configuration: cdktf.listMapper(ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToTerraform, true)(struct.oidcJwtConfiguration),
    };
}
exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToTerraform = ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToTerraform;
function ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        oidc_jwt_configuration: {
            value: cdktf.listMapperHcl(ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationToHclTerraform, true)(struct.oidcJwtConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToHclTerraform = ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToHclTerraform;
class SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // oidc_jwt_configuration - computed: false, optional: true, required: false
        this._oidcJwtConfiguration = new SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOidcJwtConfigurationList(this, "oidc_jwt_configuration", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._oidcJwtConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.oidcJwtConfiguration = this._oidcJwtConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._oidcJwtConfiguration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._oidcJwtConfiguration.internalValue = value.oidcJwtConfiguration;
        }
    }
    get oidcJwtConfiguration() {
        return this._oidcJwtConfiguration;
    }
    putOidcJwtConfiguration(value) {
        this._oidcJwtConfiguration.internalValue = value;
    }
    resetOidcJwtConfiguration() {
        this._oidcJwtConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oidcJwtConfigurationInput() {
        return this._oidcJwtConfiguration.internalValue;
    }
}
exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference = SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.ssoadminTrustedTokenIssuer.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference", version: "19.15.0" };
class SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList = SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList[_d] = { fqn: "@cdktf/provider-aws.ssoadminTrustedTokenIssuer.SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_trusted_token_issuer aws_ssoadmin_trusted_token_issuer}
*/
class SsoadminTrustedTokenIssuer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsoadminTrustedTokenIssuer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsoadminTrustedTokenIssuer to import
    * @param importFromId The id of the existing SsoadminTrustedTokenIssuer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_trusted_token_issuer#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsoadminTrustedTokenIssuer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_trusted_token_issuer", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_trusted_token_issuer aws_ssoadmin_trusted_token_issuer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsoadminTrustedTokenIssuerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_trusted_token_issuer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // trusted_token_issuer_configuration - computed: false, optional: true, required: false
        this._trustedTokenIssuerConfiguration = new SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList(this, "trusted_token_issuer_configuration", false);
        this._clientToken = config.clientToken;
        this._instanceArn = config.instanceArn;
        this._name = config.name;
        this._tags = config.tags;
        this._trustedTokenIssuerType = config.trustedTokenIssuerType;
        this._trustedTokenIssuerConfiguration.internalValue = config.trustedTokenIssuerConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get clientToken() {
        return this.getStringAttribute('client_token');
    }
    set clientToken(value) {
        this._clientToken = value;
    }
    resetClientToken() {
        this._clientToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientTokenInput() {
        return this._clientToken;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceArnInput() {
        return this._instanceArn;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get trustedTokenIssuerType() {
        return this.getStringAttribute('trusted_token_issuer_type');
    }
    set trustedTokenIssuerType(value) {
        this._trustedTokenIssuerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustedTokenIssuerTypeInput() {
        return this._trustedTokenIssuerType;
    }
    get trustedTokenIssuerConfiguration() {
        return this._trustedTokenIssuerConfiguration;
    }
    putTrustedTokenIssuerConfiguration(value) {
        this._trustedTokenIssuerConfiguration.internalValue = value;
    }
    resetTrustedTokenIssuerConfiguration() {
        this._trustedTokenIssuerConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trustedTokenIssuerConfigurationInput() {
        return this._trustedTokenIssuerConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client_token: cdktf.stringToTerraform(this._clientToken),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            trusted_token_issuer_type: cdktf.stringToTerraform(this._trustedTokenIssuerType),
            trusted_token_issuer_configuration: cdktf.listMapper(ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToTerraform, true)(this._trustedTokenIssuerConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            client_token: {
                value: cdktf.stringToHclTerraform(this._clientToken),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_arn: {
                value: cdktf.stringToHclTerraform(this._instanceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            trusted_token_issuer_type: {
                value: cdktf.stringToHclTerraform(this._trustedTokenIssuerType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trusted_token_issuer_configuration: {
                value: cdktf.listMapperHcl(ssoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationToHclTerraform, true)(this._trustedTokenIssuerConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SsoadminTrustedTokenIssuerTrustedTokenIssuerConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsoadminTrustedTokenIssuer = SsoadminTrustedTokenIssuer;
_e = JSII_RTTI_SYMBOL_1;
SsoadminTrustedTokenIssuer[_e] = { fqn: "@cdktf/provider-aws.ssoadminTrustedTokenIssuer.SsoadminTrustedTokenIssuer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsoadminTrustedTokenIssuer.tfResourceType = "aws_ssoadmin_trusted_token_issuer";
