// @cdktf/provider-aws/lib/media-convert-queue/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaConvertQueue = exports.MediaConvertQueueReservationPlanSettingsOutputReference = exports.mediaConvertQueueReservationPlanSettingsToHclTerraform = exports.mediaConvertQueueReservationPlanSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mediaConvertQueueReservationPlanSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        commitment: cdktf.stringToTerraform(struct.commitment),
        renewal_type: cdktf.stringToTerraform(struct.renewalType),
        reserved_slots: cdktf.numberToTerraform(struct.reservedSlots),
    };
}
exports.mediaConvertQueueReservationPlanSettingsToTerraform = mediaConvertQueueReservationPlanSettingsToTerraform;
function mediaConvertQueueReservationPlanSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        commitment: {
            value: cdktf.stringToHclTerraform(struct.commitment),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        renewal_type: {
            value: cdktf.stringToHclTerraform(struct.renewalType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        reserved_slots: {
            value: cdktf.numberToHclTerraform(struct.reservedSlots),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mediaConvertQueueReservationPlanSettingsToHclTerraform = mediaConvertQueueReservationPlanSettingsToHclTerraform;
class MediaConvertQueueReservationPlanSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._commitment !== undefined) {
            hasAnyValues = true;
            internalValueResult.commitment = this._commitment;
        }
        if (this._renewalType !== undefined) {
            hasAnyValues = true;
            internalValueResult.renewalType = this._renewalType;
        }
        if (this._reservedSlots !== undefined) {
            hasAnyValues = true;
            internalValueResult.reservedSlots = this._reservedSlots;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._commitment = undefined;
            this._renewalType = undefined;
            this._reservedSlots = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._commitment = value.commitment;
            this._renewalType = value.renewalType;
            this._reservedSlots = value.reservedSlots;
        }
    }
    get commitment() {
        return this.getStringAttribute('commitment');
    }
    set commitment(value) {
        this._commitment = value;
    }
    // Temporarily expose input value. Use with caution.
    get commitmentInput() {
        return this._commitment;
    }
    get renewalType() {
        return this.getStringAttribute('renewal_type');
    }
    set renewalType(value) {
        this._renewalType = value;
    }
    // Temporarily expose input value. Use with caution.
    get renewalTypeInput() {
        return this._renewalType;
    }
    get reservedSlots() {
        return this.getNumberAttribute('reserved_slots');
    }
    set reservedSlots(value) {
        this._reservedSlots = value;
    }
    // Temporarily expose input value. Use with caution.
    get reservedSlotsInput() {
        return this._reservedSlots;
    }
}
exports.MediaConvertQueueReservationPlanSettingsOutputReference = MediaConvertQueueReservationPlanSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MediaConvertQueueReservationPlanSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.mediaConvertQueue.MediaConvertQueueReservationPlanSettingsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/media_convert_queue aws_media_convert_queue}
*/
class MediaConvertQueue extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a MediaConvertQueue resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the MediaConvertQueue to import
    * @param importFromId The id of the existing MediaConvertQueue that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/media_convert_queue#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the MediaConvertQueue to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_media_convert_queue", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/media_convert_queue aws_media_convert_queue} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MediaConvertQueueConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_media_convert_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // reservation_plan_settings - computed: false, optional: true, required: false
        this._reservationPlanSettings = new MediaConvertQueueReservationPlanSettingsOutputReference(this, "reservation_plan_settings");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._pricingPlan = config.pricingPlan;
        this._status = config.status;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._reservationPlanSettings.internalValue = config.reservationPlanSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get pricingPlan() {
        return this.getStringAttribute('pricing_plan');
    }
    set pricingPlan(value) {
        this._pricingPlan = value;
    }
    resetPricingPlan() {
        this._pricingPlan = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pricingPlanInput() {
        return this._pricingPlan;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get reservationPlanSettings() {
        return this._reservationPlanSettings;
    }
    putReservationPlanSettings(value) {
        this._reservationPlanSettings.internalValue = value;
    }
    resetReservationPlanSettings() {
        this._reservationPlanSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reservationPlanSettingsInput() {
        return this._reservationPlanSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            pricing_plan: cdktf.stringToTerraform(this._pricingPlan),
            status: cdktf.stringToTerraform(this._status),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            reservation_plan_settings: mediaConvertQueueReservationPlanSettingsToTerraform(this._reservationPlanSettings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pricing_plan: {
                value: cdktf.stringToHclTerraform(this._pricingPlan),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            reservation_plan_settings: {
                value: mediaConvertQueueReservationPlanSettingsToHclTerraform(this._reservationPlanSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MediaConvertQueueReservationPlanSettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.MediaConvertQueue = MediaConvertQueue;
_b = JSII_RTTI_SYMBOL_1;
MediaConvertQueue[_b] = { fqn: "@cdktf/provider-aws.mediaConvertQueue.MediaConvertQueue", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
MediaConvertQueue.tfResourceType = "aws_media_convert_queue";
