// @cdktf/provider-aws/lib/directory-service-trust/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectoryServiceTrust = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/directory_service_trust aws_directory_service_trust}
*/
class DirectoryServiceTrust extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DirectoryServiceTrust resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DirectoryServiceTrust to import
    * @param importFromId The id of the existing DirectoryServiceTrust that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/directory_service_trust#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DirectoryServiceTrust to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_directory_service_trust", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/directory_service_trust aws_directory_service_trust} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DirectoryServiceTrustConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_directory_service_trust',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._conditionalForwarderIpAddrs = config.conditionalForwarderIpAddrs;
        this._deleteAssociatedConditionalForwarder = config.deleteAssociatedConditionalForwarder;
        this._directoryId = config.directoryId;
        this._remoteDomainName = config.remoteDomainName;
        this._selectiveAuth = config.selectiveAuth;
        this._trustDirection = config.trustDirection;
        this._trustPassword = config.trustPassword;
        this._trustType = config.trustType;
    }
    get conditionalForwarderIpAddrs() {
        return cdktf.Fn.tolist(this.getListAttribute('conditional_forwarder_ip_addrs'));
    }
    set conditionalForwarderIpAddrs(value) {
        this._conditionalForwarderIpAddrs = value;
    }
    resetConditionalForwarderIpAddrs() {
        this._conditionalForwarderIpAddrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conditionalForwarderIpAddrsInput() {
        return this._conditionalForwarderIpAddrs;
    }
    // created_date_time - computed: true, optional: false, required: false
    get createdDateTime() {
        return this.getStringAttribute('created_date_time');
    }
    get deleteAssociatedConditionalForwarder() {
        return this.getBooleanAttribute('delete_associated_conditional_forwarder');
    }
    set deleteAssociatedConditionalForwarder(value) {
        this._deleteAssociatedConditionalForwarder = value;
    }
    resetDeleteAssociatedConditionalForwarder() {
        this._deleteAssociatedConditionalForwarder = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteAssociatedConditionalForwarderInput() {
        return this._deleteAssociatedConditionalForwarder;
    }
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    set directoryId(value) {
        this._directoryId = value;
    }
    // Temporarily expose input value. Use with caution.
    get directoryIdInput() {
        return this._directoryId;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_date_time - computed: true, optional: false, required: false
    get lastUpdatedDateTime() {
        return this.getStringAttribute('last_updated_date_time');
    }
    get remoteDomainName() {
        return this.getStringAttribute('remote_domain_name');
    }
    set remoteDomainName(value) {
        this._remoteDomainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get remoteDomainNameInput() {
        return this._remoteDomainName;
    }
    get selectiveAuth() {
        return this.getStringAttribute('selective_auth');
    }
    set selectiveAuth(value) {
        this._selectiveAuth = value;
    }
    resetSelectiveAuth() {
        this._selectiveAuth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selectiveAuthInput() {
        return this._selectiveAuth;
    }
    // state_last_updated_date_time - computed: true, optional: false, required: false
    get stateLastUpdatedDateTime() {
        return this.getStringAttribute('state_last_updated_date_time');
    }
    get trustDirection() {
        return this.getStringAttribute('trust_direction');
    }
    set trustDirection(value) {
        this._trustDirection = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustDirectionInput() {
        return this._trustDirection;
    }
    get trustPassword() {
        return this.getStringAttribute('trust_password');
    }
    set trustPassword(value) {
        this._trustPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustPasswordInput() {
        return this._trustPassword;
    }
    // trust_state - computed: true, optional: false, required: false
    get trustState() {
        return this.getStringAttribute('trust_state');
    }
    // trust_state_reason - computed: true, optional: false, required: false
    get trustStateReason() {
        return this.getStringAttribute('trust_state_reason');
    }
    get trustType() {
        return this.getStringAttribute('trust_type');
    }
    set trustType(value) {
        this._trustType = value;
    }
    resetTrustType() {
        this._trustType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trustTypeInput() {
        return this._trustType;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            conditional_forwarder_ip_addrs: cdktf.listMapper(cdktf.stringToTerraform, false)(this._conditionalForwarderIpAddrs),
            delete_associated_conditional_forwarder: cdktf.booleanToTerraform(this._deleteAssociatedConditionalForwarder),
            directory_id: cdktf.stringToTerraform(this._directoryId),
            remote_domain_name: cdktf.stringToTerraform(this._remoteDomainName),
            selective_auth: cdktf.stringToTerraform(this._selectiveAuth),
            trust_direction: cdktf.stringToTerraform(this._trustDirection),
            trust_password: cdktf.stringToTerraform(this._trustPassword),
            trust_type: cdktf.stringToTerraform(this._trustType),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            conditional_forwarder_ip_addrs: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._conditionalForwarderIpAddrs),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            delete_associated_conditional_forwarder: {
                value: cdktf.booleanToHclTerraform(this._deleteAssociatedConditionalForwarder),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            directory_id: {
                value: cdktf.stringToHclTerraform(this._directoryId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            remote_domain_name: {
                value: cdktf.stringToHclTerraform(this._remoteDomainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            selective_auth: {
                value: cdktf.stringToHclTerraform(this._selectiveAuth),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_direction: {
                value: cdktf.stringToHclTerraform(this._trustDirection),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_password: {
                value: cdktf.stringToHclTerraform(this._trustPassword),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_type: {
                value: cdktf.stringToHclTerraform(this._trustType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DirectoryServiceTrust = DirectoryServiceTrust;
_a = JSII_RTTI_SYMBOL_1;
DirectoryServiceTrust[_a] = { fqn: "@cdktf/provider-aws.directoryServiceTrust.DirectoryServiceTrust", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DirectoryServiceTrust.tfResourceType = "aws_directory_service_trust";
