// @cdktf/provider-aws/lib/data-aws-auditmanager-control/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAuditmanagerControl = exports.DataAwsAuditmanagerControlControlMappingSourcesList = exports.DataAwsAuditmanagerControlControlMappingSourcesOutputReference = exports.dataAwsAuditmanagerControlControlMappingSourcesToHclTerraform = exports.dataAwsAuditmanagerControlControlMappingSourcesToTerraform = exports.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList = exports.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference = exports.dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform = exports.dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToTerraform = dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToTerraform;
function dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform = dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform;
class DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // keyword_input_type - computed: true, optional: false, required: false
    get keywordInputType() {
        return this.getStringAttribute('keyword_input_type');
    }
    // keyword_value - computed: true, optional: false, required: false
    get keywordValue() {
        return this.getStringAttribute('keyword_value');
    }
}
exports.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference = DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsAuditmanagerControl.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference", version: "19.15.0" };
class DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList = DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList[_b] = { fqn: "@cdktf/provider-aws.dataAwsAuditmanagerControl.DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList", version: "19.15.0" };
function dataAwsAuditmanagerControlControlMappingSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_keyword: cdktf.listMapper(dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToTerraform, true)(struct.sourceKeyword),
    };
}
exports.dataAwsAuditmanagerControlControlMappingSourcesToTerraform = dataAwsAuditmanagerControlControlMappingSourcesToTerraform;
function dataAwsAuditmanagerControlControlMappingSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_keyword: {
            value: cdktf.listMapperHcl(dataAwsAuditmanagerControlControlMappingSourcesSourceKeywordToHclTerraform, true)(struct.sourceKeyword),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsAuditmanagerControlControlMappingSourcesToHclTerraform = dataAwsAuditmanagerControlControlMappingSourcesToHclTerraform;
class DataAwsAuditmanagerControlControlMappingSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // source_keyword - computed: false, optional: true, required: false
        this._sourceKeyword = new DataAwsAuditmanagerControlControlMappingSourcesSourceKeywordList(this, "source_keyword", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceKeyword?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceKeyword = this._sourceKeyword?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._sourceKeyword.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._sourceKeyword.internalValue = value.sourceKeyword;
        }
    }
    // source_description - computed: true, optional: false, required: false
    get sourceDescription() {
        return this.getStringAttribute('source_description');
    }
    // source_frequency - computed: true, optional: false, required: false
    get sourceFrequency() {
        return this.getStringAttribute('source_frequency');
    }
    // source_id - computed: true, optional: false, required: false
    get sourceId() {
        return this.getStringAttribute('source_id');
    }
    // source_name - computed: true, optional: false, required: false
    get sourceName() {
        return this.getStringAttribute('source_name');
    }
    // source_set_up_option - computed: true, optional: false, required: false
    get sourceSetUpOption() {
        return this.getStringAttribute('source_set_up_option');
    }
    // source_type - computed: true, optional: false, required: false
    get sourceType() {
        return this.getStringAttribute('source_type');
    }
    // troubleshooting_text - computed: true, optional: false, required: false
    get troubleshootingText() {
        return this.getStringAttribute('troubleshooting_text');
    }
    get sourceKeyword() {
        return this._sourceKeyword;
    }
    putSourceKeyword(value) {
        this._sourceKeyword.internalValue = value;
    }
    resetSourceKeyword() {
        this._sourceKeyword.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceKeywordInput() {
        return this._sourceKeyword.internalValue;
    }
}
exports.DataAwsAuditmanagerControlControlMappingSourcesOutputReference = DataAwsAuditmanagerControlControlMappingSourcesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAuditmanagerControlControlMappingSourcesOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsAuditmanagerControl.DataAwsAuditmanagerControlControlMappingSourcesOutputReference", version: "19.15.0" };
class DataAwsAuditmanagerControlControlMappingSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAuditmanagerControlControlMappingSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAuditmanagerControlControlMappingSourcesList = DataAwsAuditmanagerControlControlMappingSourcesList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAuditmanagerControlControlMappingSourcesList[_d] = { fqn: "@cdktf/provider-aws.dataAwsAuditmanagerControl.DataAwsAuditmanagerControlControlMappingSourcesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/auditmanager_control aws_auditmanager_control}
*/
class DataAwsAuditmanagerControl extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsAuditmanagerControl resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsAuditmanagerControl to import
    * @param importFromId The id of the existing DataAwsAuditmanagerControl that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/auditmanager_control#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsAuditmanagerControl to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_auditmanager_control", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/auditmanager_control aws_auditmanager_control} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAuditmanagerControlConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_auditmanager_control',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
        // control_mapping_sources - computed: false, optional: true, required: false
        this._controlMappingSources = new DataAwsAuditmanagerControlControlMappingSourcesList(this, "control_mapping_sources", true);
        this._name = config.name;
        this._type = config.type;
        this._controlMappingSources.internalValue = config.controlMappingSources;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // action_plan_instructions - computed: true, optional: false, required: false
    get actionPlanInstructions() {
        return this.getStringAttribute('action_plan_instructions');
    }
    // action_plan_title - computed: true, optional: false, required: false
    get actionPlanTitle() {
        return this.getStringAttribute('action_plan_title');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this._tags;
    }
    // testing_information - computed: true, optional: false, required: false
    get testingInformation() {
        return this.getStringAttribute('testing_information');
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get controlMappingSources() {
        return this._controlMappingSources;
    }
    putControlMappingSources(value) {
        this._controlMappingSources.internalValue = value;
    }
    resetControlMappingSources() {
        this._controlMappingSources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlMappingSourcesInput() {
        return this._controlMappingSources.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            type: cdktf.stringToTerraform(this._type),
            control_mapping_sources: cdktf.listMapper(dataAwsAuditmanagerControlControlMappingSourcesToTerraform, true)(this._controlMappingSources.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            control_mapping_sources: {
                value: cdktf.listMapperHcl(dataAwsAuditmanagerControlControlMappingSourcesToHclTerraform, true)(this._controlMappingSources.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsAuditmanagerControlControlMappingSourcesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsAuditmanagerControl = DataAwsAuditmanagerControl;
_e = JSII_RTTI_SYMBOL_1;
DataAwsAuditmanagerControl[_e] = { fqn: "@cdktf/provider-aws.dataAwsAuditmanagerControl.DataAwsAuditmanagerControl", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAuditmanagerControl.tfResourceType = "aws_auditmanager_control";
