// @cdktf/provider-aws/lib/devicefarm-instance-profile/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevicefarmInstanceProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devicefarm_instance_profile aws_devicefarm_instance_profile}
*/
class DevicefarmInstanceProfile extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DevicefarmInstanceProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DevicefarmInstanceProfile to import
    * @param importFromId The id of the existing DevicefarmInstanceProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devicefarm_instance_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DevicefarmInstanceProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_devicefarm_instance_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/devicefarm_instance_profile aws_devicefarm_instance_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DevicefarmInstanceProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_devicefarm_instance_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._excludeAppPackagesFromCleanup = config.excludeAppPackagesFromCleanup;
        this._id = config.id;
        this._name = config.name;
        this._packageCleanup = config.packageCleanup;
        this._rebootAfterUse = config.rebootAfterUse;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get excludeAppPackagesFromCleanup() {
        return cdktf.Fn.tolist(this.getListAttribute('exclude_app_packages_from_cleanup'));
    }
    set excludeAppPackagesFromCleanup(value) {
        this._excludeAppPackagesFromCleanup = value;
    }
    resetExcludeAppPackagesFromCleanup() {
        this._excludeAppPackagesFromCleanup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeAppPackagesFromCleanupInput() {
        return this._excludeAppPackagesFromCleanup;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get packageCleanup() {
        return this.getBooleanAttribute('package_cleanup');
    }
    set packageCleanup(value) {
        this._packageCleanup = value;
    }
    resetPackageCleanup() {
        this._packageCleanup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get packageCleanupInput() {
        return this._packageCleanup;
    }
    get rebootAfterUse() {
        return this.getBooleanAttribute('reboot_after_use');
    }
    set rebootAfterUse(value) {
        this._rebootAfterUse = value;
    }
    resetRebootAfterUse() {
        this._rebootAfterUse = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rebootAfterUseInput() {
        return this._rebootAfterUse;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            exclude_app_packages_from_cleanup: cdktf.listMapper(cdktf.stringToTerraform, false)(this._excludeAppPackagesFromCleanup),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            package_cleanup: cdktf.booleanToTerraform(this._packageCleanup),
            reboot_after_use: cdktf.booleanToTerraform(this._rebootAfterUse),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            exclude_app_packages_from_cleanup: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._excludeAppPackagesFromCleanup),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            package_cleanup: {
                value: cdktf.booleanToHclTerraform(this._packageCleanup),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            reboot_after_use: {
                value: cdktf.booleanToHclTerraform(this._rebootAfterUse),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DevicefarmInstanceProfile = DevicefarmInstanceProfile;
_a = JSII_RTTI_SYMBOL_1;
DevicefarmInstanceProfile[_a] = { fqn: "@cdktf/provider-aws.devicefarmInstanceProfile.DevicefarmInstanceProfile", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DevicefarmInstanceProfile.tfResourceType = "aws_devicefarm_instance_profile";
