// @winglang/sdk/lib/cloud/counter.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterInflightMethods = exports.Counter = exports.COUNTER_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constants_1 = require("../constants");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `Counter`.
 */
exports.COUNTER_FQN = (0, constants_1.fqnForType)("cloud.Counter");
/**
 * A distributed atomic counter.
 * @inflight `@winglang/sdk.cloud.ICounterClient`
 * @abstract
 */
class Counter extends std_1.Resource {
    constructor(scope, id, props = {}) {
        if (new.target === Counter) {
            return std_1.Resource._newFromFactory(exports.COUNTER_FQN, scope, id, props);
        }
        super(scope, id);
        std_1.Node.of(this).title = "Counter";
        std_1.Node.of(this).description = "A distributed atomic counter";
        this.initial = props.initial ?? 0;
    }
}
exports.Counter = Counter;
_a = JSII_RTTI_SYMBOL_1;
Counter[_a] = { fqn: "@winglang/sdk.cloud.Counter", version: "0.0.0" };
/**
 * List of inflight operations available for `Counter`.
 * @internal
 */
var CounterInflightMethods;
(function (CounterInflightMethods) {
    /** `Counter.inc` */
    CounterInflightMethods["INC"] = "inc";
    /** `Counter.dec` */
    CounterInflightMethods["DEC"] = "dec";
    /** `Counter.peek` */
    CounterInflightMethods["PEEK"] = "peek";
    /** `Counter.set` */
    CounterInflightMethods["SET"] = "set";
})(CounterInflightMethods || (exports.CounterInflightMethods = CounterInflightMethods = {}));
