// @winglang/sdk/lib/cloud/schedule.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScheduleInflightMethods = exports.Schedule = exports.SCHEDULE_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cron_validator_1 = require("cron-validator");
const constants_1 = require("../constants");
const errors_1 = require("../core/errors");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `Schedule`.
 */
exports.SCHEDULE_FQN = (0, constants_1.fqnForType)("cloud.Schedule");
/**
 * A schedule.
 *
 * @inflight `@winglang/sdk.cloud.IScheduleClient`
 * @abstract
 */
class Schedule extends std_1.Resource {
    constructor(scope, id, props = {}) {
        if (new.target === Schedule) {
            return std_1.Resource._newFromFactory(exports.SCHEDULE_FQN, scope, id, props);
        }
        super(scope, id);
        std_1.Node.of(this).title = "Schedule";
        std_1.Node.of(this).description =
            "A cloud schedule to trigger events at regular intervals";
        const { cron, rate } = props;
        if (rate && cron) {
            throw new Error("rate and cron cannot be configured simultaneously.");
        }
        if (!rate && !cron) {
            throw new Error("rate or cron need to be filled.");
        }
        if (rate && rate.seconds < 60) {
            throw new Error("rate can not be set to less than 1 minute.");
        }
        // Check for valid UNIX cron format
        // https://www.ibm.com/docs/en/db2/11.5?topic=task-unix-cron-format
        if (cron &&
            !(0, cron_validator_1.isValidCron)(cron, {
                alias: true,
                allowSevenAsSunday: true,
                allowBlankDay: false,
                seconds: false,
            })) {
            throw new Error("cron string must be in UNIX cron format");
        }
    }
    /**
     * Create a function that runs when receiving the scheduled event.
     * @abstract
     */
    onTick(inflight, props) {
        inflight;
        props;
        throw new errors_1.AbstractMemberError();
    }
}
exports.Schedule = Schedule;
_a = JSII_RTTI_SYMBOL_1;
Schedule[_a] = { fqn: "@winglang/sdk.cloud.Schedule", version: "0.0.0" };
/**
 * List of inflight operations available for `Schedule`.
 * @internal
 */
var ScheduleInflightMethods;
(function (ScheduleInflightMethods) {
    /** When the schedule runs its scheduled actions. */
    ScheduleInflightMethods["TICK"] = "tick";
})(ScheduleInflightMethods || (exports.ScheduleInflightMethods = ScheduleInflightMethods = {}));
