// @winglang/sdk/lib/cloud/secret.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretInflightMethods = exports.Secret = exports.SECRET_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constants_1 = require("../constants");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `Secret`.
 */
exports.SECRET_FQN = (0, constants_1.fqnForType)("cloud.Secret");
/**
 * A cloud secret.
 *
 * @inflight `@winglang/sdk.cloud.ISecretClient`
 * @abstract
 */
class Secret extends std_1.Resource {
    constructor(scope, id, props = {}) {
        if (new.target === Secret) {
            return std_1.Resource._newFromFactory(exports.SECRET_FQN, scope, id, props);
        }
        super(scope, id);
        /** @internal */
        this[_b] = true;
        std_1.Node.of(this).title = "Secret";
        std_1.Node.of(this).description = "A cloud secret";
        this._name = props.name;
    }
    /** Get secret name */
    get name() {
        return this._name;
    }
}
exports.Secret = Secret;
_a = JSII_RTTI_SYMBOL_1, _b = types_1.SECRET_SYMBOL;
Secret[_a] = { fqn: "@winglang/sdk.cloud.Secret", version: "0.0.0" };
/**
 * List of inflight operations available for `Secret`.
 * @internal
 */
var SecretInflightMethods;
(function (SecretInflightMethods) {
    /** `Secret.value` */
    SecretInflightMethods["VALUE"] = "value";
    /** `Secret.valueJson` */
    SecretInflightMethods["VALUE_JSON"] = "valueJson";
})(SecretInflightMethods || (exports.SecretInflightMethods = SecretInflightMethods = {}));
