// @winglang/sdk/lib/cloud/topic.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicInflightMethods = exports.Topic = exports.TOPIC_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constants_1 = require("../constants");
const errors_1 = require("../core/errors");
const types_1 = require("../core/types");
const std_1 = require("../std");
exports.TOPIC_FQN = (0, constants_1.fqnForType)("cloud.Topic");
/**
 * A topic.
 *
 * @inflight `@winglang/sdk.cloud.ITopicClient`
 * @abstract
 */
class Topic extends std_1.Resource {
    constructor(scope, id, props = {}) {
        if (new.target === Topic) {
            return std_1.Resource._newFromFactory(exports.TOPIC_FQN, scope, id, props);
        }
        super(scope, id);
        std_1.Node.of(this).title = "Topic";
        std_1.Node.of(this).description = "A pub/sub notification topic";
        props;
    }
    /**
     * Run an inflight whenever an message is published to the topic.
     * @abstract
     */
    onMessage(inflight, props) {
        inflight;
        props;
        throw new errors_1.AbstractMemberError();
    }
    /**
     * Subscribing queue to the topic
     * @abstract
     */
    subscribeQueue(queue, props) {
        queue;
        props;
        throw new errors_1.AbstractMemberError();
    }
}
exports.Topic = Topic;
_a = JSII_RTTI_SYMBOL_1;
Topic[_a] = { fqn: "@winglang/sdk.cloud.Topic", version: "0.0.0" };
/**
 * List of inflight operations available for `Topic`.
 * @internal
 */
var TopicInflightMethods;
(function (TopicInflightMethods) {
    /** `Topic.publish` */
    TopicInflightMethods["PUBLISH"] = "publish";
})(TopicInflightMethods || (exports.TopicInflightMethods = TopicInflightMethods = {}));
