// @winglang/sdk/lib/.gen/providers/aws/api-gateway-rest-api/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_rest_api
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayRestApi = exports.ApiGatewayRestApiEndpointConfigurationOutputReference = exports.apiGatewayRestApiEndpointConfigurationToHclTerraform = exports.apiGatewayRestApiEndpointConfigurationToTerraform = void 0;
const cdktf = __importStar(require("cdktf"));
function apiGatewayRestApiEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.types),
        vpc_endpoint_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.vpcEndpointIds),
    };
}
exports.apiGatewayRestApiEndpointConfigurationToTerraform = apiGatewayRestApiEndpointConfigurationToTerraform;
function apiGatewayRestApiEndpointConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.types),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        vpc_endpoint_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.vpcEndpointIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apiGatewayRestApiEndpointConfigurationToHclTerraform = apiGatewayRestApiEndpointConfigurationToHclTerraform;
class ApiGatewayRestApiEndpointConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._types !== undefined) {
            hasAnyValues = true;
            internalValueResult.types = this._types;
        }
        if (this._vpcEndpointIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcEndpointIds = this._vpcEndpointIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._types = undefined;
            this._vpcEndpointIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._types = value.types;
            this._vpcEndpointIds = value.vpcEndpointIds;
        }
    }
    get types() {
        return this.getListAttribute('types');
    }
    set types(value) {
        this._types = value;
    }
    // Temporarily expose input value. Use with caution.
    get typesInput() {
        return this._types;
    }
    get vpcEndpointIds() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_endpoint_ids'));
    }
    set vpcEndpointIds(value) {
        this._vpcEndpointIds = value;
    }
    resetVpcEndpointIds() {
        this._vpcEndpointIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointIdsInput() {
        return this._vpcEndpointIds;
    }
}
exports.ApiGatewayRestApiEndpointConfigurationOutputReference = ApiGatewayRestApiEndpointConfigurationOutputReference;
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_rest_api aws_api_gateway_rest_api}
*/
class ApiGatewayRestApi extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApiGatewayRestApi resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApiGatewayRestApi to import
    * @param importFromId The id of the existing ApiGatewayRestApi that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_rest_api#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApiGatewayRestApi to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_api_gateway_rest_api", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_rest_api aws_api_gateway_rest_api} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayRestApiConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_rest_api',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.31.0',
                providerVersionConstraint: '~>5.31.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint_configuration - computed: false, optional: true, required: false
        this._endpointConfiguration = new ApiGatewayRestApiEndpointConfigurationOutputReference(this, "endpoint_configuration");
        this._apiKeySource = config.apiKeySource;
        this._binaryMediaTypes = config.binaryMediaTypes;
        this._body = config.body;
        this._description = config.description;
        this._disableExecuteApiEndpoint = config.disableExecuteApiEndpoint;
        this._failOnWarnings = config.failOnWarnings;
        this._id = config.id;
        this._minimumCompressionSize = config.minimumCompressionSize;
        this._name = config.name;
        this._parameters = config.parameters;
        this._policy = config.policy;
        this._putRestApiMode = config.putRestApiMode;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._endpointConfiguration.internalValue = config.endpointConfiguration;
    }
    get apiKeySource() {
        return this.getStringAttribute('api_key_source');
    }
    set apiKeySource(value) {
        this._apiKeySource = value;
    }
    resetApiKeySource() {
        this._apiKeySource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiKeySourceInput() {
        return this._apiKeySource;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get binaryMediaTypes() {
        return this.getListAttribute('binary_media_types');
    }
    set binaryMediaTypes(value) {
        this._binaryMediaTypes = value;
    }
    resetBinaryMediaTypes() {
        this._binaryMediaTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get binaryMediaTypesInput() {
        return this._binaryMediaTypes;
    }
    get body() {
        return this.getStringAttribute('body');
    }
    set body(value) {
        this._body = value;
    }
    resetBody() {
        this._body = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bodyInput() {
        return this._body;
    }
    // created_date - computed: true, optional: false, required: false
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disableExecuteApiEndpoint() {
        return this.getBooleanAttribute('disable_execute_api_endpoint');
    }
    set disableExecuteApiEndpoint(value) {
        this._disableExecuteApiEndpoint = value;
    }
    resetDisableExecuteApiEndpoint() {
        this._disableExecuteApiEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableExecuteApiEndpointInput() {
        return this._disableExecuteApiEndpoint;
    }
    // execution_arn - computed: true, optional: false, required: false
    get executionArn() {
        return this.getStringAttribute('execution_arn');
    }
    get failOnWarnings() {
        return this.getBooleanAttribute('fail_on_warnings');
    }
    set failOnWarnings(value) {
        this._failOnWarnings = value;
    }
    resetFailOnWarnings() {
        this._failOnWarnings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failOnWarningsInput() {
        return this._failOnWarnings;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get minimumCompressionSize() {
        return this.getStringAttribute('minimum_compression_size');
    }
    set minimumCompressionSize(value) {
        this._minimumCompressionSize = value;
    }
    resetMinimumCompressionSize() {
        this._minimumCompressionSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minimumCompressionSizeInput() {
        return this._minimumCompressionSize;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    resetPolicy() {
        this._policy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    get putRestApiMode() {
        return this.getStringAttribute('put_rest_api_mode');
    }
    set putRestApiMode(value) {
        this._putRestApiMode = value;
    }
    resetPutRestApiMode() {
        this._putRestApiMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get putRestApiModeInput() {
        return this._putRestApiMode;
    }
    // root_resource_id - computed: true, optional: false, required: false
    get rootResourceId() {
        return this.getStringAttribute('root_resource_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get endpointConfiguration() {
        return this._endpointConfiguration;
    }
    putEndpointConfiguration(value) {
        this._endpointConfiguration.internalValue = value;
    }
    resetEndpointConfiguration() {
        this._endpointConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointConfigurationInput() {
        return this._endpointConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_key_source: cdktf.stringToTerraform(this._apiKeySource),
            binary_media_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._binaryMediaTypes),
            body: cdktf.stringToTerraform(this._body),
            description: cdktf.stringToTerraform(this._description),
            disable_execute_api_endpoint: cdktf.booleanToTerraform(this._disableExecuteApiEndpoint),
            fail_on_warnings: cdktf.booleanToTerraform(this._failOnWarnings),
            id: cdktf.stringToTerraform(this._id),
            minimum_compression_size: cdktf.stringToTerraform(this._minimumCompressionSize),
            name: cdktf.stringToTerraform(this._name),
            parameters: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameters),
            policy: cdktf.stringToTerraform(this._policy),
            put_rest_api_mode: cdktf.stringToTerraform(this._putRestApiMode),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            endpoint_configuration: apiGatewayRestApiEndpointConfigurationToTerraform(this._endpointConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            api_key_source: {
                value: cdktf.stringToHclTerraform(this._apiKeySource),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            binary_media_types: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._binaryMediaTypes),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            body: {
                value: cdktf.stringToHclTerraform(this._body),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disable_execute_api_endpoint: {
                value: cdktf.booleanToHclTerraform(this._disableExecuteApiEndpoint),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            fail_on_warnings: {
                value: cdktf.booleanToHclTerraform(this._failOnWarnings),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            minimum_compression_size: {
                value: cdktf.stringToHclTerraform(this._minimumCompressionSize),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parameters: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._parameters),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            policy: {
                value: cdktf.stringToHclTerraform(this._policy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            put_rest_api_mode: {
                value: cdktf.stringToHclTerraform(this._putRestApiMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            endpoint_configuration: {
                value: apiGatewayRestApiEndpointConfigurationToHclTerraform(this._endpointConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApiGatewayRestApiEndpointConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApiGatewayRestApi = ApiGatewayRestApi;
// =================
// STATIC PROPERTIES
// =================
ApiGatewayRestApi.tfResourceType = "aws_api_gateway_rest_api";
