// @cdktf/provider-aws/lib/cloudtrail-event-data-store/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudtrailEventDataStore = exports.CloudtrailEventDataStoreTimeoutsOutputReference = exports.cloudtrailEventDataStoreTimeoutsToHclTerraform = exports.cloudtrailEventDataStoreTimeoutsToTerraform = exports.CloudtrailEventDataStoreAdvancedEventSelectorList = exports.CloudtrailEventDataStoreAdvancedEventSelectorOutputReference = exports.cloudtrailEventDataStoreAdvancedEventSelectorToHclTerraform = exports.cloudtrailEventDataStoreAdvancedEventSelectorToTerraform = exports.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList = exports.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference = exports.cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToHclTerraform = exports.cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ends_with: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.endsWith),
        equals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.equalTo),
        field: cdktf.stringToTerraform(struct.field),
        not_ends_with: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.notEndsWith),
        not_equals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.notEquals),
        not_starts_with: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.notStartsWith),
        starts_with: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.startsWith),
    };
}
exports.cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToTerraform = cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToTerraform;
function cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ends_with: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.endsWith),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        equals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.equalTo),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        not_ends_with: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.notEndsWith),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        not_equals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.notEquals),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        not_starts_with: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.notStartsWith),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        starts_with: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.startsWith),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToHclTerraform = cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToHclTerraform;
class CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endsWith !== undefined) {
            hasAnyValues = true;
            internalValueResult.endsWith = this._endsWith;
        }
        if (this._equals !== undefined) {
            hasAnyValues = true;
            internalValueResult.equalTo = this._equals;
        }
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._notEndsWith !== undefined) {
            hasAnyValues = true;
            internalValueResult.notEndsWith = this._notEndsWith;
        }
        if (this._notEquals !== undefined) {
            hasAnyValues = true;
            internalValueResult.notEquals = this._notEquals;
        }
        if (this._notStartsWith !== undefined) {
            hasAnyValues = true;
            internalValueResult.notStartsWith = this._notStartsWith;
        }
        if (this._startsWith !== undefined) {
            hasAnyValues = true;
            internalValueResult.startsWith = this._startsWith;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._endsWith = undefined;
            this._equals = undefined;
            this._field = undefined;
            this._notEndsWith = undefined;
            this._notEquals = undefined;
            this._notStartsWith = undefined;
            this._startsWith = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._endsWith = value.endsWith;
            this._equals = value.equalTo;
            this._field = value.field;
            this._notEndsWith = value.notEndsWith;
            this._notEquals = value.notEquals;
            this._notStartsWith = value.notStartsWith;
            this._startsWith = value.startsWith;
        }
    }
    get endsWith() {
        return this.getListAttribute('ends_with');
    }
    set endsWith(value) {
        this._endsWith = value;
    }
    resetEndsWith() {
        this._endsWith = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endsWithInput() {
        return this._endsWith;
    }
    get equalTo() {
        return this.getListAttribute('equals');
    }
    set equalTo(value) {
        this._equals = value;
    }
    resetEqualTo() {
        this._equals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get equalToInput() {
        return this._equals;
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get notEndsWith() {
        return this.getListAttribute('not_ends_with');
    }
    set notEndsWith(value) {
        this._notEndsWith = value;
    }
    resetNotEndsWith() {
        this._notEndsWith = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notEndsWithInput() {
        return this._notEndsWith;
    }
    get notEquals() {
        return this.getListAttribute('not_equals');
    }
    set notEquals(value) {
        this._notEquals = value;
    }
    resetNotEquals() {
        this._notEquals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notEqualsInput() {
        return this._notEquals;
    }
    get notStartsWith() {
        return this.getListAttribute('not_starts_with');
    }
    set notStartsWith(value) {
        this._notStartsWith = value;
    }
    resetNotStartsWith() {
        this._notStartsWith = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notStartsWithInput() {
        return this._notStartsWith;
    }
    get startsWith() {
        return this.getListAttribute('starts_with');
    }
    set startsWith(value) {
        this._startsWith = value;
    }
    resetStartsWith() {
        this._startsWith = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startsWithInput() {
        return this._startsWith;
    }
}
exports.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference = CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference", version: "19.15.0" };
class CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList = CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList;
_b = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList[_b] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList", version: "19.15.0" };
function cloudtrailEventDataStoreAdvancedEventSelectorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        field_selector: cdktf.listMapper(cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToTerraform, true)(struct.fieldSelector),
    };
}
exports.cloudtrailEventDataStoreAdvancedEventSelectorToTerraform = cloudtrailEventDataStoreAdvancedEventSelectorToTerraform;
function cloudtrailEventDataStoreAdvancedEventSelectorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        field_selector: {
            value: cdktf.listMapperHcl(cloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorToHclTerraform, true)(struct.fieldSelector),
            isBlock: true,
            type: "set",
            storageClassType: "CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudtrailEventDataStoreAdvancedEventSelectorToHclTerraform = cloudtrailEventDataStoreAdvancedEventSelectorToHclTerraform;
class CloudtrailEventDataStoreAdvancedEventSelectorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // field_selector - computed: false, optional: true, required: false
        this._fieldSelector = new CloudtrailEventDataStoreAdvancedEventSelectorFieldSelectorList(this, "field_selector", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._fieldSelector?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.fieldSelector = this._fieldSelector?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._fieldSelector.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._fieldSelector.internalValue = value.fieldSelector;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get fieldSelector() {
        return this._fieldSelector;
    }
    putFieldSelector(value) {
        this._fieldSelector.internalValue = value;
    }
    resetFieldSelector() {
        this._fieldSelector.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldSelectorInput() {
        return this._fieldSelector.internalValue;
    }
}
exports.CloudtrailEventDataStoreAdvancedEventSelectorOutputReference = CloudtrailEventDataStoreAdvancedEventSelectorOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStoreAdvancedEventSelectorOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStoreAdvancedEventSelectorOutputReference", version: "19.15.0" };
class CloudtrailEventDataStoreAdvancedEventSelectorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudtrailEventDataStoreAdvancedEventSelectorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudtrailEventDataStoreAdvancedEventSelectorList = CloudtrailEventDataStoreAdvancedEventSelectorList;
_d = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStoreAdvancedEventSelectorList[_d] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStoreAdvancedEventSelectorList", version: "19.15.0" };
function cloudtrailEventDataStoreTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudtrailEventDataStoreTimeoutsToTerraform = cloudtrailEventDataStoreTimeoutsToTerraform;
function cloudtrailEventDataStoreTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudtrailEventDataStoreTimeoutsToHclTerraform = cloudtrailEventDataStoreTimeoutsToHclTerraform;
class CloudtrailEventDataStoreTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudtrailEventDataStoreTimeoutsOutputReference = CloudtrailEventDataStoreTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStoreTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStoreTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudtrail_event_data_store aws_cloudtrail_event_data_store}
*/
class CloudtrailEventDataStore extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudtrailEventDataStore resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudtrailEventDataStore to import
    * @param importFromId The id of the existing CloudtrailEventDataStore that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudtrail_event_data_store#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudtrailEventDataStore to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudtrail_event_data_store", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudtrail_event_data_store aws_cloudtrail_event_data_store} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudtrailEventDataStoreConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudtrail_event_data_store',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // advanced_event_selector - computed: false, optional: true, required: false
        this._advancedEventSelector = new CloudtrailEventDataStoreAdvancedEventSelectorList(this, "advanced_event_selector", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudtrailEventDataStoreTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._multiRegionEnabled = config.multiRegionEnabled;
        this._name = config.name;
        this._organizationEnabled = config.organizationEnabled;
        this._retentionPeriod = config.retentionPeriod;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._terminationProtectionEnabled = config.terminationProtectionEnabled;
        this._advancedEventSelector.internalValue = config.advancedEventSelector;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get multiRegionEnabled() {
        return this.getBooleanAttribute('multi_region_enabled');
    }
    set multiRegionEnabled(value) {
        this._multiRegionEnabled = value;
    }
    resetMultiRegionEnabled() {
        this._multiRegionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multiRegionEnabledInput() {
        return this._multiRegionEnabled;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get organizationEnabled() {
        return this.getBooleanAttribute('organization_enabled');
    }
    set organizationEnabled(value) {
        this._organizationEnabled = value;
    }
    resetOrganizationEnabled() {
        this._organizationEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationEnabledInput() {
        return this._organizationEnabled;
    }
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    set retentionPeriod(value) {
        this._retentionPeriod = value;
    }
    resetRetentionPeriod() {
        this._retentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get terminationProtectionEnabled() {
        return this.getBooleanAttribute('termination_protection_enabled');
    }
    set terminationProtectionEnabled(value) {
        this._terminationProtectionEnabled = value;
    }
    resetTerminationProtectionEnabled() {
        this._terminationProtectionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get terminationProtectionEnabledInput() {
        return this._terminationProtectionEnabled;
    }
    get advancedEventSelector() {
        return this._advancedEventSelector;
    }
    putAdvancedEventSelector(value) {
        this._advancedEventSelector.internalValue = value;
    }
    resetAdvancedEventSelector() {
        this._advancedEventSelector.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedEventSelectorInput() {
        return this._advancedEventSelector.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            multi_region_enabled: cdktf.booleanToTerraform(this._multiRegionEnabled),
            name: cdktf.stringToTerraform(this._name),
            organization_enabled: cdktf.booleanToTerraform(this._organizationEnabled),
            retention_period: cdktf.numberToTerraform(this._retentionPeriod),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            termination_protection_enabled: cdktf.booleanToTerraform(this._terminationProtectionEnabled),
            advanced_event_selector: cdktf.listMapper(cloudtrailEventDataStoreAdvancedEventSelectorToTerraform, true)(this._advancedEventSelector.internalValue),
            timeouts: cloudtrailEventDataStoreTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_id: {
                value: cdktf.stringToHclTerraform(this._kmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            multi_region_enabled: {
                value: cdktf.booleanToHclTerraform(this._multiRegionEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organization_enabled: {
                value: cdktf.booleanToHclTerraform(this._organizationEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            retention_period: {
                value: cdktf.numberToHclTerraform(this._retentionPeriod),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            termination_protection_enabled: {
                value: cdktf.booleanToHclTerraform(this._terminationProtectionEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            advanced_event_selector: {
                value: cdktf.listMapperHcl(cloudtrailEventDataStoreAdvancedEventSelectorToHclTerraform, true)(this._advancedEventSelector.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudtrailEventDataStoreAdvancedEventSelectorList",
            },
            timeouts: {
                value: cloudtrailEventDataStoreTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CloudtrailEventDataStoreTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudtrailEventDataStore = CloudtrailEventDataStore;
_f = JSII_RTTI_SYMBOL_1;
CloudtrailEventDataStore[_f] = { fqn: "@cdktf/provider-aws.cloudtrailEventDataStore.CloudtrailEventDataStore", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudtrailEventDataStore.tfResourceType = "aws_cloudtrail_event_data_store";
