// @cdktf/provider-aws/lib/data-aws-ssm-patch-baseline/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsmPatchBaseline = exports.DataAwsSsmPatchBaselineSourceList = exports.DataAwsSsmPatchBaselineSourceOutputReference = exports.dataAwsSsmPatchBaselineSourceToHclTerraform = exports.dataAwsSsmPatchBaselineSourceToTerraform = exports.DataAwsSsmPatchBaselineGlobalFilterList = exports.DataAwsSsmPatchBaselineGlobalFilterOutputReference = exports.dataAwsSsmPatchBaselineGlobalFilterToHclTerraform = exports.dataAwsSsmPatchBaselineGlobalFilterToTerraform = exports.DataAwsSsmPatchBaselineApprovalRuleList = exports.DataAwsSsmPatchBaselineApprovalRuleOutputReference = exports.dataAwsSsmPatchBaselineApprovalRuleToHclTerraform = exports.dataAwsSsmPatchBaselineApprovalRuleToTerraform = exports.DataAwsSsmPatchBaselineApprovalRulePatchFilterList = exports.DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference = exports.dataAwsSsmPatchBaselineApprovalRulePatchFilterToHclTerraform = exports.dataAwsSsmPatchBaselineApprovalRulePatchFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSsmPatchBaselineApprovalRulePatchFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsmPatchBaselineApprovalRulePatchFilterToTerraform = dataAwsSsmPatchBaselineApprovalRulePatchFilterToTerraform;
function dataAwsSsmPatchBaselineApprovalRulePatchFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsmPatchBaselineApprovalRulePatchFilterToHclTerraform = dataAwsSsmPatchBaselineApprovalRulePatchFilterToHclTerraform;
class DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference = DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference", version: "19.15.0" };
class DataAwsSsmPatchBaselineApprovalRulePatchFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsmPatchBaselineApprovalRulePatchFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsmPatchBaselineApprovalRulePatchFilterList = DataAwsSsmPatchBaselineApprovalRulePatchFilterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineApprovalRulePatchFilterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineApprovalRulePatchFilterList", version: "19.15.0" };
function dataAwsSsmPatchBaselineApprovalRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsmPatchBaselineApprovalRuleToTerraform = dataAwsSsmPatchBaselineApprovalRuleToTerraform;
function dataAwsSsmPatchBaselineApprovalRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsmPatchBaselineApprovalRuleToHclTerraform = dataAwsSsmPatchBaselineApprovalRuleToHclTerraform;
class DataAwsSsmPatchBaselineApprovalRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // patch_filter - computed: true, optional: false, required: false
        this._patchFilter = new DataAwsSsmPatchBaselineApprovalRulePatchFilterList(this, "patch_filter", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // approve_after_days - computed: true, optional: false, required: false
    get approveAfterDays() {
        return this.getNumberAttribute('approve_after_days');
    }
    // approve_until_date - computed: true, optional: false, required: false
    get approveUntilDate() {
        return this.getStringAttribute('approve_until_date');
    }
    // compliance_level - computed: true, optional: false, required: false
    get complianceLevel() {
        return this.getStringAttribute('compliance_level');
    }
    // enable_non_security - computed: true, optional: false, required: false
    get enableNonSecurity() {
        return this.getBooleanAttribute('enable_non_security');
    }
    get patchFilter() {
        return this._patchFilter;
    }
}
exports.DataAwsSsmPatchBaselineApprovalRuleOutputReference = DataAwsSsmPatchBaselineApprovalRuleOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineApprovalRuleOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineApprovalRuleOutputReference", version: "19.15.0" };
class DataAwsSsmPatchBaselineApprovalRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsmPatchBaselineApprovalRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsmPatchBaselineApprovalRuleList = DataAwsSsmPatchBaselineApprovalRuleList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineApprovalRuleList[_d] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineApprovalRuleList", version: "19.15.0" };
function dataAwsSsmPatchBaselineGlobalFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsmPatchBaselineGlobalFilterToTerraform = dataAwsSsmPatchBaselineGlobalFilterToTerraform;
function dataAwsSsmPatchBaselineGlobalFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsmPatchBaselineGlobalFilterToHclTerraform = dataAwsSsmPatchBaselineGlobalFilterToHclTerraform;
class DataAwsSsmPatchBaselineGlobalFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataAwsSsmPatchBaselineGlobalFilterOutputReference = DataAwsSsmPatchBaselineGlobalFilterOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineGlobalFilterOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineGlobalFilterOutputReference", version: "19.15.0" };
class DataAwsSsmPatchBaselineGlobalFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsmPatchBaselineGlobalFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsmPatchBaselineGlobalFilterList = DataAwsSsmPatchBaselineGlobalFilterList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineGlobalFilterList[_f] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineGlobalFilterList", version: "19.15.0" };
function dataAwsSsmPatchBaselineSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsmPatchBaselineSourceToTerraform = dataAwsSsmPatchBaselineSourceToTerraform;
function dataAwsSsmPatchBaselineSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsmPatchBaselineSourceToHclTerraform = dataAwsSsmPatchBaselineSourceToHclTerraform;
class DataAwsSsmPatchBaselineSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // configuration - computed: true, optional: false, required: false
    get configuration() {
        return this.getStringAttribute('configuration');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // products - computed: true, optional: false, required: false
    get products() {
        return this.getListAttribute('products');
    }
}
exports.DataAwsSsmPatchBaselineSourceOutputReference = DataAwsSsmPatchBaselineSourceOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineSourceOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineSourceOutputReference", version: "19.15.0" };
class DataAwsSsmPatchBaselineSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsmPatchBaselineSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsmPatchBaselineSourceList = DataAwsSsmPatchBaselineSourceList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaselineSourceList[_h] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaselineSourceList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssm_patch_baseline aws_ssm_patch_baseline}
*/
class DataAwsSsmPatchBaseline extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSsmPatchBaseline resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSsmPatchBaseline to import
    * @param importFromId The id of the existing DataAwsSsmPatchBaseline that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssm_patch_baseline#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSsmPatchBaseline to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssm_patch_baseline", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssm_patch_baseline aws_ssm_patch_baseline} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSsmPatchBaselineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_patch_baseline',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // approval_rule - computed: true, optional: false, required: false
        this._approvalRule = new DataAwsSsmPatchBaselineApprovalRuleList(this, "approval_rule", false);
        // global_filter - computed: true, optional: false, required: false
        this._globalFilter = new DataAwsSsmPatchBaselineGlobalFilterList(this, "global_filter", false);
        // source - computed: true, optional: false, required: false
        this._source = new DataAwsSsmPatchBaselineSourceList(this, "source", false);
        this._defaultBaseline = config.defaultBaseline;
        this._id = config.id;
        this._namePrefix = config.namePrefix;
        this._operatingSystem = config.operatingSystem;
        this._owner = config.owner;
    }
    get approvalRule() {
        return this._approvalRule;
    }
    // approved_patches - computed: true, optional: false, required: false
    get approvedPatches() {
        return this.getListAttribute('approved_patches');
    }
    // approved_patches_compliance_level - computed: true, optional: false, required: false
    get approvedPatchesComplianceLevel() {
        return this.getStringAttribute('approved_patches_compliance_level');
    }
    // approved_patches_enable_non_security - computed: true, optional: false, required: false
    get approvedPatchesEnableNonSecurity() {
        return this.getBooleanAttribute('approved_patches_enable_non_security');
    }
    get defaultBaseline() {
        return this.getBooleanAttribute('default_baseline');
    }
    set defaultBaseline(value) {
        this._defaultBaseline = value;
    }
    resetDefaultBaseline() {
        this._defaultBaseline = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultBaselineInput() {
        return this._defaultBaseline;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get globalFilter() {
        return this._globalFilter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // json - computed: true, optional: false, required: false
    get json() {
        return this.getStringAttribute('json');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get operatingSystem() {
        return this.getStringAttribute('operating_system');
    }
    set operatingSystem(value) {
        this._operatingSystem = value;
    }
    resetOperatingSystem() {
        this._operatingSystem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operatingSystemInput() {
        return this._operatingSystem;
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
    // rejected_patches - computed: true, optional: false, required: false
    get rejectedPatches() {
        return this.getListAttribute('rejected_patches');
    }
    // rejected_patches_action - computed: true, optional: false, required: false
    get rejectedPatchesAction() {
        return this.getStringAttribute('rejected_patches_action');
    }
    get source() {
        return this._source;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_baseline: cdktf.booleanToTerraform(this._defaultBaseline),
            id: cdktf.stringToTerraform(this._id),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            operating_system: cdktf.stringToTerraform(this._operatingSystem),
            owner: cdktf.stringToTerraform(this._owner),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            default_baseline: {
                value: cdktf.booleanToHclTerraform(this._defaultBaseline),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            operating_system: {
                value: cdktf.stringToHclTerraform(this._operatingSystem),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            owner: {
                value: cdktf.stringToHclTerraform(this._owner),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSsmPatchBaseline = DataAwsSsmPatchBaseline;
_j = JSII_RTTI_SYMBOL_1;
DataAwsSsmPatchBaseline[_j] = { fqn: "@cdktf/provider-aws.dataAwsSsmPatchBaseline.DataAwsSsmPatchBaseline", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSsmPatchBaseline.tfResourceType = "aws_ssm_patch_baseline";
