// @cdktf/provider-aws/lib/securitylake-aws-log-source/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecuritylakeAwsLogSource = exports.SecuritylakeAwsLogSourceSourceList = exports.SecuritylakeAwsLogSourceSourceOutputReference = exports.securitylakeAwsLogSourceSourceToHclTerraform = exports.securitylakeAwsLogSourceSourceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function securitylakeAwsLogSourceSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accounts: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.accounts),
        regions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.regions),
        source_name: cdktf.stringToTerraform(struct.sourceName),
        source_version: cdktf.stringToTerraform(struct.sourceVersion),
    };
}
exports.securitylakeAwsLogSourceSourceToTerraform = securitylakeAwsLogSourceSourceToTerraform;
function securitylakeAwsLogSourceSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        accounts: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.accounts),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        regions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.regions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        source_name: {
            value: cdktf.stringToHclTerraform(struct.sourceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_version: {
            value: cdktf.stringToHclTerraform(struct.sourceVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.securitylakeAwsLogSourceSourceToHclTerraform = securitylakeAwsLogSourceSourceToHclTerraform;
class SecuritylakeAwsLogSourceSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accounts !== undefined) {
            hasAnyValues = true;
            internalValueResult.accounts = this._accounts;
        }
        if (this._regions !== undefined) {
            hasAnyValues = true;
            internalValueResult.regions = this._regions;
        }
        if (this._sourceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceName = this._sourceName;
        }
        if (this._sourceVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceVersion = this._sourceVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._accounts = undefined;
            this._regions = undefined;
            this._sourceName = undefined;
            this._sourceVersion = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._accounts = value.accounts;
            this._regions = value.regions;
            this._sourceName = value.sourceName;
            this._sourceVersion = value.sourceVersion;
        }
    }
    get accounts() {
        return cdktf.Fn.tolist(this.getListAttribute('accounts'));
    }
    set accounts(value) {
        this._accounts = value;
    }
    resetAccounts() {
        this._accounts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountsInput() {
        return this._accounts;
    }
    get regions() {
        return cdktf.Fn.tolist(this.getListAttribute('regions'));
    }
    set regions(value) {
        this._regions = value;
    }
    // Temporarily expose input value. Use with caution.
    get regionsInput() {
        return this._regions;
    }
    get sourceName() {
        return this.getStringAttribute('source_name');
    }
    set sourceName(value) {
        this._sourceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceNameInput() {
        return this._sourceName;
    }
    get sourceVersion() {
        return this.getStringAttribute('source_version');
    }
    set sourceVersion(value) {
        this._sourceVersion = value;
    }
    resetSourceVersion() {
        this._sourceVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceVersionInput() {
        return this._sourceVersion;
    }
}
exports.SecuritylakeAwsLogSourceSourceOutputReference = SecuritylakeAwsLogSourceSourceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SecuritylakeAwsLogSourceSourceOutputReference[_a] = { fqn: "@cdktf/provider-aws.securitylakeAwsLogSource.SecuritylakeAwsLogSourceSourceOutputReference", version: "19.15.0" };
class SecuritylakeAwsLogSourceSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SecuritylakeAwsLogSourceSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SecuritylakeAwsLogSourceSourceList = SecuritylakeAwsLogSourceSourceList;
_b = JSII_RTTI_SYMBOL_1;
SecuritylakeAwsLogSourceSourceList[_b] = { fqn: "@cdktf/provider-aws.securitylakeAwsLogSource.SecuritylakeAwsLogSourceSourceList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_aws_log_source aws_securitylake_aws_log_source}
*/
class SecuritylakeAwsLogSource extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecuritylakeAwsLogSource resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecuritylakeAwsLogSource to import
    * @param importFromId The id of the existing SecuritylakeAwsLogSource that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_aws_log_source#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecuritylakeAwsLogSource to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securitylake_aws_log_source", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securitylake_aws_log_source aws_securitylake_aws_log_source} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecuritylakeAwsLogSourceConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_securitylake_aws_log_source',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // source - computed: false, optional: true, required: false
        this._source = new SecuritylakeAwsLogSourceSourceList(this, "source", false);
        this._source.internalValue = config.source;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get source() {
        return this._source;
    }
    putSource(value) {
        this._source.internalValue = value;
    }
    resetSource() {
        this._source.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            source: cdktf.listMapper(securitylakeAwsLogSourceSourceToTerraform, true)(this._source.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            source: {
                value: cdktf.listMapperHcl(securitylakeAwsLogSourceSourceToHclTerraform, true)(this._source.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SecuritylakeAwsLogSourceSourceList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecuritylakeAwsLogSource = SecuritylakeAwsLogSource;
_c = JSII_RTTI_SYMBOL_1;
SecuritylakeAwsLogSource[_c] = { fqn: "@cdktf/provider-aws.securitylakeAwsLogSource.SecuritylakeAwsLogSource", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecuritylakeAwsLogSource.tfResourceType = "aws_securitylake_aws_log_source";
