// @cdktf/provider-aws/lib/data-aws-emr-supported-instance-types/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEmrSupportedInstanceTypes = exports.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList = exports.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference = exports.dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToHclTerraform = exports.dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToTerraform = dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToTerraform;
function dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToHclTerraform = dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToHclTerraform;
class DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // architecture - computed: true, optional: false, required: false
    get architecture() {
        return this.getStringAttribute('architecture');
    }
    // ebs_optimized_available - computed: true, optional: false, required: false
    get ebsOptimizedAvailable() {
        return this.getBooleanAttribute('ebs_optimized_available');
    }
    // ebs_optimized_by_default - computed: true, optional: false, required: false
    get ebsOptimizedByDefault() {
        return this.getBooleanAttribute('ebs_optimized_by_default');
    }
    // ebs_storage_only - computed: true, optional: false, required: false
    get ebsStorageOnly() {
        return this.getBooleanAttribute('ebs_storage_only');
    }
    // instance_family_id - computed: true, optional: false, required: false
    get instanceFamilyId() {
        return this.getStringAttribute('instance_family_id');
    }
    // is_64_bits_only - computed: true, optional: false, required: false
    get is64BitsOnly() {
        return this.getBooleanAttribute('is_64_bits_only');
    }
    // memory_gb - computed: true, optional: false, required: false
    get memoryGb() {
        return this.getNumberAttribute('memory_gb');
    }
    // number_of_disks - computed: true, optional: false, required: false
    get numberOfDisks() {
        return this.getNumberAttribute('number_of_disks');
    }
    // storage_gb - computed: true, optional: false, required: false
    get storageGb() {
        return this.getNumberAttribute('storage_gb');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // vcpu - computed: true, optional: false, required: false
    get vcpu() {
        return this.getNumberAttribute('vcpu');
    }
}
exports.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference = DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEmrSupportedInstanceTypes.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference", version: "19.15.0" };
class DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList = DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEmrSupportedInstanceTypes.DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emr_supported_instance_types aws_emr_supported_instance_types}
*/
class DataAwsEmrSupportedInstanceTypes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEmrSupportedInstanceTypes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEmrSupportedInstanceTypes to import
    * @param importFromId The id of the existing DataAwsEmrSupportedInstanceTypes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emr_supported_instance_types#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEmrSupportedInstanceTypes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_emr_supported_instance_types", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emr_supported_instance_types aws_emr_supported_instance_types} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEmrSupportedInstanceTypesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emr_supported_instance_types',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // supported_instance_types - computed: false, optional: true, required: false
        this._supportedInstanceTypes = new DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList(this, "supported_instance_types", false);
        this._releaseLabel = config.releaseLabel;
        this._supportedInstanceTypes.internalValue = config.supportedInstanceTypes;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get releaseLabel() {
        return this.getStringAttribute('release_label');
    }
    set releaseLabel(value) {
        this._releaseLabel = value;
    }
    // Temporarily expose input value. Use with caution.
    get releaseLabelInput() {
        return this._releaseLabel;
    }
    get supportedInstanceTypes() {
        return this._supportedInstanceTypes;
    }
    putSupportedInstanceTypes(value) {
        this._supportedInstanceTypes.internalValue = value;
    }
    resetSupportedInstanceTypes() {
        this._supportedInstanceTypes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get supportedInstanceTypesInput() {
        return this._supportedInstanceTypes.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            release_label: cdktf.stringToTerraform(this._releaseLabel),
            supported_instance_types: cdktf.listMapper(dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToTerraform, true)(this._supportedInstanceTypes.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            release_label: {
                value: cdktf.stringToHclTerraform(this._releaseLabel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            supported_instance_types: {
                value: cdktf.listMapperHcl(dataAwsEmrSupportedInstanceTypesSupportedInstanceTypesToHclTerraform, true)(this._supportedInstanceTypes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsEmrSupportedInstanceTypesSupportedInstanceTypesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEmrSupportedInstanceTypes = DataAwsEmrSupportedInstanceTypes;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEmrSupportedInstanceTypes[_c] = { fqn: "@cdktf/provider-aws.dataAwsEmrSupportedInstanceTypes.DataAwsEmrSupportedInstanceTypes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEmrSupportedInstanceTypes.tfResourceType = "aws_emr_supported_instance_types";
