// @winglang/sdk/lib/cloud/queue.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueInflightMethods = exports.Queue = exports.DEFAULT_DELIVERY_ATTEMPTS = exports.QUEUE_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constants_1 = require("../constants");
const errors_1 = require("../core/errors");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `Queue`.
 */
exports.QUEUE_FQN = (0, constants_1.fqnForType)("cloud.Queue");
/**
 * Dead-letter queue default retries
 */
exports.DEFAULT_DELIVERY_ATTEMPTS = 1;
/**
 * A queue.
 *
 * @inflight `@winglang/sdk.cloud.IQueueClient`
 * @abstract
 */
class Queue extends std_1.Resource {
    constructor(scope, id, props = {}) {
        if (new.target === Queue) {
            return std_1.Resource._newFromFactory(exports.QUEUE_FQN, scope, id, props);
        }
        super(scope, id);
        std_1.Node.of(this).title = "Queue";
        std_1.Node.of(this).description = "A distributed message queue";
        props;
    }
    /**
     * Create a function to consume messages from this queue.
     * @abstract
     */
    setConsumer(handler, props) {
        handler;
        props;
        throw new errors_1.AbstractMemberError();
    }
}
exports.Queue = Queue;
_a = JSII_RTTI_SYMBOL_1;
Queue[_a] = { fqn: "@winglang/sdk.cloud.Queue", version: "0.0.0" };
/**
 * List of inflight operations available for `Queue`.
 * @internal
 */
var QueueInflightMethods;
(function (QueueInflightMethods) {
    /** `Queue.push` */
    QueueInflightMethods["PUSH"] = "push";
    /** `Queue.purge` */
    QueueInflightMethods["PURGE"] = "purge";
    /** `Queue.approxSize` */
    QueueInflightMethods["APPROX_SIZE"] = "approxSize";
    /** `Queue.pop` */
    QueueInflightMethods["POP"] = "pop";
})(QueueInflightMethods || (exports.QueueInflightMethods = QueueInflightMethods = {}));
