// @winglang/sdk/lib/cloud/website.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Website = exports.WEBSITE_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const path_1 = require("path");
const constants_1 = require("../constants");
const core_1 = require("../core");
const errors_1 = require("../core/errors");
const types_1 = require("../core/types");
const std_1 = require("../std");
/**
 * Global identifier for `Website`.
 */
exports.WEBSITE_FQN = (0, constants_1.fqnForType)("cloud.Website");
/**
 * A cloud static website.
 *
 * @inflight `@winglang/sdk.cloud.IWebsiteClient`
 * @abstract
 */
class Website extends std_1.Resource {
    constructor(scope, id, props) {
        if (new.target === Website) {
            return std_1.Resource._newFromFactory(exports.WEBSITE_FQN, scope, id, props);
        }
        super(scope, id);
        std_1.Node.of(this).title = "Website";
        std_1.Node.of(this).description = "A static website";
        this._path = (0, path_1.isAbsolute)(props.path)
            ? props.path
            : (0, path_1.resolve)(core_1.App.of(scope).entrypointDir, props.path);
        this._domain = props.domain;
    }
    /**
     * Absolute local path to the website's static files.
     */
    get path() {
        return this._path;
    }
    /**
     * The website's url.
     */
    get url() {
        return this._endpoint.url;
    }
    /**
     * The Endpoint of the Website.
     * @abstract
     * @internal
     */
    get _endpoint() {
        throw new errors_1.AbstractMemberError();
    }
    /**
     * Add a JSON file with custom values during the website's deployment.
     * If the path conflicts with file path from the website's static assets, an error will be thrown.
     * @param path the file path it will be uploaded as
     * @param data the data to write to the file
     */
    addJson(path, data) {
        if (!path.endsWith(".json")) {
            throw new Error(`key must have a .json suffix. (current: "${path.split(".").pop()}")`);
        }
        return this.addFile(path, JSON.stringify(data), {
            contentType: "application/json",
        });
    }
    /**
     * Add a file to the website during deployment.
     * If the path conflicts with file path from the website's static assets, an error will be thrown.
     * @param path the file path it will be uploaded as
     * @param data the data to write to the file
     * @param options configure the file's options
     * @abstract
     */
    addFile(path, data, options) {
        path;
        data;
        options;
        throw new errors_1.AbstractMemberError();
    }
}
exports.Website = Website;
_a = JSII_RTTI_SYMBOL_1;
Website[_a] = { fqn: "@winglang/sdk.cloud.Website", version: "0.0.0" };
