// @cdktf/provider-aws/lib/data-aws-sesv2-email-identity-mail-from-attributes/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSesv2EmailIdentityMailFromAttributes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity_mail_from_attributes aws_sesv2_email_identity_mail_from_attributes}
*/
class DataAwsSesv2EmailIdentityMailFromAttributes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSesv2EmailIdentityMailFromAttributes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSesv2EmailIdentityMailFromAttributes to import
    * @param importFromId The id of the existing DataAwsSesv2EmailIdentityMailFromAttributes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity_mail_from_attributes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSesv2EmailIdentityMailFromAttributes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sesv2_email_identity_mail_from_attributes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity_mail_from_attributes aws_sesv2_email_identity_mail_from_attributes} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSesv2EmailIdentityMailFromAttributesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sesv2_email_identity_mail_from_attributes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._emailIdentity = config.emailIdentity;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // behavior_on_mx_failure - computed: true, optional: false, required: false
    get behaviorOnMxFailure() {
        return this.getStringAttribute('behavior_on_mx_failure');
    }
    get emailIdentity() {
        return this.getStringAttribute('email_identity');
    }
    set emailIdentity(value) {
        this._emailIdentity = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailIdentityInput() {
        return this._emailIdentity;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // mail_from_domain - computed: true, optional: false, required: false
    get mailFromDomain() {
        return this.getStringAttribute('mail_from_domain');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            email_identity: cdktf.stringToTerraform(this._emailIdentity),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            email_identity: {
                value: cdktf.stringToHclTerraform(this._emailIdentity),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSesv2EmailIdentityMailFromAttributes = DataAwsSesv2EmailIdentityMailFromAttributes;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSesv2EmailIdentityMailFromAttributes[_a] = { fqn: "@cdktf/provider-aws.dataAwsSesv2EmailIdentityMailFromAttributes.DataAwsSesv2EmailIdentityMailFromAttributes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSesv2EmailIdentityMailFromAttributes.tfResourceType = "aws_sesv2_email_identity_mail_from_attributes";
