// @cdktf/provider-aws/lib/fsx-data-repository-association/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxDataRepositoryAssociation = exports.FsxDataRepositoryAssociationTimeoutsOutputReference = exports.fsxDataRepositoryAssociationTimeoutsToHclTerraform = exports.fsxDataRepositoryAssociationTimeoutsToTerraform = exports.FsxDataRepositoryAssociationS3OutputReference = exports.fsxDataRepositoryAssociationS3ToHclTerraform = exports.fsxDataRepositoryAssociationS3ToTerraform = exports.FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference = exports.fsxDataRepositoryAssociationS3AutoImportPolicyToHclTerraform = exports.fsxDataRepositoryAssociationS3AutoImportPolicyToTerraform = exports.FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference = exports.fsxDataRepositoryAssociationS3AutoExportPolicyToHclTerraform = exports.fsxDataRepositoryAssociationS3AutoExportPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxDataRepositoryAssociationS3AutoExportPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
    };
}
exports.fsxDataRepositoryAssociationS3AutoExportPolicyToTerraform = fsxDataRepositoryAssociationS3AutoExportPolicyToTerraform;
function fsxDataRepositoryAssociationS3AutoExportPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxDataRepositoryAssociationS3AutoExportPolicyToHclTerraform = fsxDataRepositoryAssociationS3AutoExportPolicyToHclTerraform;
class FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._events = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._events = value.events;
        }
    }
    get events() {
        return this.getListAttribute('events');
    }
    set events(value) {
        this._events = value;
    }
    resetEvents() {
        this._events = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
}
exports.FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference = FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsxDataRepositoryAssociation.FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference", version: "19.15.0" };
function fsxDataRepositoryAssociationS3AutoImportPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
    };
}
exports.fsxDataRepositoryAssociationS3AutoImportPolicyToTerraform = fsxDataRepositoryAssociationS3AutoImportPolicyToTerraform;
function fsxDataRepositoryAssociationS3AutoImportPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxDataRepositoryAssociationS3AutoImportPolicyToHclTerraform = fsxDataRepositoryAssociationS3AutoImportPolicyToHclTerraform;
class FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._events = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._events = value.events;
        }
    }
    get events() {
        return this.getListAttribute('events');
    }
    set events(value) {
        this._events = value;
    }
    resetEvents() {
        this._events = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
}
exports.FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference = FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference[_b] = { fqn: "@cdktf/provider-aws.fsxDataRepositoryAssociation.FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference", version: "19.15.0" };
function fsxDataRepositoryAssociationS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_export_policy: fsxDataRepositoryAssociationS3AutoExportPolicyToTerraform(struct.autoExportPolicy),
        auto_import_policy: fsxDataRepositoryAssociationS3AutoImportPolicyToTerraform(struct.autoImportPolicy),
    };
}
exports.fsxDataRepositoryAssociationS3ToTerraform = fsxDataRepositoryAssociationS3ToTerraform;
function fsxDataRepositoryAssociationS3ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_export_policy: {
            value: fsxDataRepositoryAssociationS3AutoExportPolicyToHclTerraform(struct.autoExportPolicy),
            isBlock: true,
            type: "list",
            storageClassType: "FsxDataRepositoryAssociationS3AutoExportPolicyList",
        },
        auto_import_policy: {
            value: fsxDataRepositoryAssociationS3AutoImportPolicyToHclTerraform(struct.autoImportPolicy),
            isBlock: true,
            type: "list",
            storageClassType: "FsxDataRepositoryAssociationS3AutoImportPolicyList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxDataRepositoryAssociationS3ToHclTerraform = fsxDataRepositoryAssociationS3ToHclTerraform;
class FsxDataRepositoryAssociationS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // auto_export_policy - computed: false, optional: true, required: false
        this._autoExportPolicy = new FsxDataRepositoryAssociationS3AutoExportPolicyOutputReference(this, "auto_export_policy");
        // auto_import_policy - computed: false, optional: true, required: false
        this._autoImportPolicy = new FsxDataRepositoryAssociationS3AutoImportPolicyOutputReference(this, "auto_import_policy");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoExportPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoExportPolicy = this._autoExportPolicy?.internalValue;
        }
        if (this._autoImportPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoImportPolicy = this._autoImportPolicy?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoExportPolicy.internalValue = undefined;
            this._autoImportPolicy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoExportPolicy.internalValue = value.autoExportPolicy;
            this._autoImportPolicy.internalValue = value.autoImportPolicy;
        }
    }
    get autoExportPolicy() {
        return this._autoExportPolicy;
    }
    putAutoExportPolicy(value) {
        this._autoExportPolicy.internalValue = value;
    }
    resetAutoExportPolicy() {
        this._autoExportPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoExportPolicyInput() {
        return this._autoExportPolicy.internalValue;
    }
    get autoImportPolicy() {
        return this._autoImportPolicy;
    }
    putAutoImportPolicy(value) {
        this._autoImportPolicy.internalValue = value;
    }
    resetAutoImportPolicy() {
        this._autoImportPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoImportPolicyInput() {
        return this._autoImportPolicy.internalValue;
    }
}
exports.FsxDataRepositoryAssociationS3OutputReference = FsxDataRepositoryAssociationS3OutputReference;
_c = JSII_RTTI_SYMBOL_1;
FsxDataRepositoryAssociationS3OutputReference[_c] = { fqn: "@cdktf/provider-aws.fsxDataRepositoryAssociation.FsxDataRepositoryAssociationS3OutputReference", version: "19.15.0" };
function fsxDataRepositoryAssociationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxDataRepositoryAssociationTimeoutsToTerraform = fsxDataRepositoryAssociationTimeoutsToTerraform;
function fsxDataRepositoryAssociationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.fsxDataRepositoryAssociationTimeoutsToHclTerraform = fsxDataRepositoryAssociationTimeoutsToHclTerraform;
class FsxDataRepositoryAssociationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FsxDataRepositoryAssociationTimeoutsOutputReference = FsxDataRepositoryAssociationTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FsxDataRepositoryAssociationTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.fsxDataRepositoryAssociation.FsxDataRepositoryAssociationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_data_repository_association aws_fsx_data_repository_association}
*/
class FsxDataRepositoryAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FsxDataRepositoryAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FsxDataRepositoryAssociation to import
    * @param importFromId The id of the existing FsxDataRepositoryAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_data_repository_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FsxDataRepositoryAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_data_repository_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/fsx_data_repository_association aws_fsx_data_repository_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FsxDataRepositoryAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_data_repository_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // s3 - computed: false, optional: true, required: false
        this._s3 = new FsxDataRepositoryAssociationS3OutputReference(this, "s3");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxDataRepositoryAssociationTimeoutsOutputReference(this, "timeouts");
        this._batchImportMetaDataOnCreate = config.batchImportMetaDataOnCreate;
        this._dataRepositoryPath = config.dataRepositoryPath;
        this._deleteDataInFilesystem = config.deleteDataInFilesystem;
        this._fileSystemId = config.fileSystemId;
        this._fileSystemPath = config.fileSystemPath;
        this._id = config.id;
        this._importedFileChunkSize = config.importedFileChunkSize;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._s3.internalValue = config.s3;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // association_id - computed: true, optional: false, required: false
    get associationId() {
        return this.getStringAttribute('association_id');
    }
    get batchImportMetaDataOnCreate() {
        return this.getBooleanAttribute('batch_import_meta_data_on_create');
    }
    set batchImportMetaDataOnCreate(value) {
        this._batchImportMetaDataOnCreate = value;
    }
    resetBatchImportMetaDataOnCreate() {
        this._batchImportMetaDataOnCreate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchImportMetaDataOnCreateInput() {
        return this._batchImportMetaDataOnCreate;
    }
    get dataRepositoryPath() {
        return this.getStringAttribute('data_repository_path');
    }
    set dataRepositoryPath(value) {
        this._dataRepositoryPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataRepositoryPathInput() {
        return this._dataRepositoryPath;
    }
    get deleteDataInFilesystem() {
        return this.getBooleanAttribute('delete_data_in_filesystem');
    }
    set deleteDataInFilesystem(value) {
        this._deleteDataInFilesystem = value;
    }
    resetDeleteDataInFilesystem() {
        this._deleteDataInFilesystem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteDataInFilesystemInput() {
        return this._deleteDataInFilesystem;
    }
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    get fileSystemPath() {
        return this.getStringAttribute('file_system_path');
    }
    set fileSystemPath(value) {
        this._fileSystemPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemPathInput() {
        return this._fileSystemPath;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get importedFileChunkSize() {
        return this.getNumberAttribute('imported_file_chunk_size');
    }
    set importedFileChunkSize(value) {
        this._importedFileChunkSize = value;
    }
    resetImportedFileChunkSize() {
        this._importedFileChunkSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get importedFileChunkSizeInput() {
        return this._importedFileChunkSize;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            batch_import_meta_data_on_create: cdktf.booleanToTerraform(this._batchImportMetaDataOnCreate),
            data_repository_path: cdktf.stringToTerraform(this._dataRepositoryPath),
            delete_data_in_filesystem: cdktf.booleanToTerraform(this._deleteDataInFilesystem),
            file_system_id: cdktf.stringToTerraform(this._fileSystemId),
            file_system_path: cdktf.stringToTerraform(this._fileSystemPath),
            id: cdktf.stringToTerraform(this._id),
            imported_file_chunk_size: cdktf.numberToTerraform(this._importedFileChunkSize),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            s3: fsxDataRepositoryAssociationS3ToTerraform(this._s3.internalValue),
            timeouts: fsxDataRepositoryAssociationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            batch_import_meta_data_on_create: {
                value: cdktf.booleanToHclTerraform(this._batchImportMetaDataOnCreate),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            data_repository_path: {
                value: cdktf.stringToHclTerraform(this._dataRepositoryPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            delete_data_in_filesystem: {
                value: cdktf.booleanToHclTerraform(this._deleteDataInFilesystem),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            file_system_id: {
                value: cdktf.stringToHclTerraform(this._fileSystemId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            file_system_path: {
                value: cdktf.stringToHclTerraform(this._fileSystemPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            imported_file_chunk_size: {
                value: cdktf.numberToHclTerraform(this._importedFileChunkSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            s3: {
                value: fsxDataRepositoryAssociationS3ToHclTerraform(this._s3.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FsxDataRepositoryAssociationS3List",
            },
            timeouts: {
                value: fsxDataRepositoryAssociationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FsxDataRepositoryAssociationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FsxDataRepositoryAssociation = FsxDataRepositoryAssociation;
_e = JSII_RTTI_SYMBOL_1;
FsxDataRepositoryAssociation[_e] = { fqn: "@cdktf/provider-aws.fsxDataRepositoryAssociation.FsxDataRepositoryAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FsxDataRepositoryAssociation.tfResourceType = "aws_fsx_data_repository_association";
