// @cdktf/provider-aws/lib/data-aws-eks-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEksCluster = exports.DataAwsEksClusterVpcConfigList = exports.DataAwsEksClusterVpcConfigOutputReference = exports.dataAwsEksClusterVpcConfigToHclTerraform = exports.dataAwsEksClusterVpcConfigToTerraform = exports.DataAwsEksClusterOutpostConfigList = exports.DataAwsEksClusterOutpostConfigOutputReference = exports.dataAwsEksClusterOutpostConfigToHclTerraform = exports.dataAwsEksClusterOutpostConfigToTerraform = exports.DataAwsEksClusterOutpostConfigControlPlanePlacementList = exports.DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference = exports.dataAwsEksClusterOutpostConfigControlPlanePlacementToHclTerraform = exports.dataAwsEksClusterOutpostConfigControlPlanePlacementToTerraform = exports.DataAwsEksClusterKubernetesNetworkConfigList = exports.DataAwsEksClusterKubernetesNetworkConfigOutputReference = exports.dataAwsEksClusterKubernetesNetworkConfigToHclTerraform = exports.dataAwsEksClusterKubernetesNetworkConfigToTerraform = exports.DataAwsEksClusterIdentityList = exports.DataAwsEksClusterIdentityOutputReference = exports.dataAwsEksClusterIdentityToHclTerraform = exports.dataAwsEksClusterIdentityToTerraform = exports.DataAwsEksClusterIdentityOidcList = exports.DataAwsEksClusterIdentityOidcOutputReference = exports.dataAwsEksClusterIdentityOidcToHclTerraform = exports.dataAwsEksClusterIdentityOidcToTerraform = exports.DataAwsEksClusterCertificateAuthorityList = exports.DataAwsEksClusterCertificateAuthorityOutputReference = exports.dataAwsEksClusterCertificateAuthorityToHclTerraform = exports.dataAwsEksClusterCertificateAuthorityToTerraform = exports.DataAwsEksClusterAccessConfigList = exports.DataAwsEksClusterAccessConfigOutputReference = exports.dataAwsEksClusterAccessConfigToHclTerraform = exports.dataAwsEksClusterAccessConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEksClusterAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterAccessConfigToTerraform = dataAwsEksClusterAccessConfigToTerraform;
function dataAwsEksClusterAccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterAccessConfigToHclTerraform = dataAwsEksClusterAccessConfigToHclTerraform;
class DataAwsEksClusterAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // authentication_mode - computed: true, optional: false, required: false
    get authenticationMode() {
        return this.getStringAttribute('authentication_mode');
    }
}
exports.DataAwsEksClusterAccessConfigOutputReference = DataAwsEksClusterAccessConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterAccessConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterAccessConfigOutputReference", version: "19.15.0" };
class DataAwsEksClusterAccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterAccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterAccessConfigList = DataAwsEksClusterAccessConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterAccessConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterAccessConfigList", version: "19.15.0" };
function dataAwsEksClusterCertificateAuthorityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterCertificateAuthorityToTerraform = dataAwsEksClusterCertificateAuthorityToTerraform;
function dataAwsEksClusterCertificateAuthorityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterCertificateAuthorityToHclTerraform = dataAwsEksClusterCertificateAuthorityToHclTerraform;
class DataAwsEksClusterCertificateAuthorityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data - computed: true, optional: false, required: false
    get data() {
        return this.getStringAttribute('data');
    }
}
exports.DataAwsEksClusterCertificateAuthorityOutputReference = DataAwsEksClusterCertificateAuthorityOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterCertificateAuthorityOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterCertificateAuthorityOutputReference", version: "19.15.0" };
class DataAwsEksClusterCertificateAuthorityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterCertificateAuthorityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterCertificateAuthorityList = DataAwsEksClusterCertificateAuthorityList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterCertificateAuthorityList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterCertificateAuthorityList", version: "19.15.0" };
function dataAwsEksClusterIdentityOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterIdentityOidcToTerraform = dataAwsEksClusterIdentityOidcToTerraform;
function dataAwsEksClusterIdentityOidcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterIdentityOidcToHclTerraform = dataAwsEksClusterIdentityOidcToHclTerraform;
class DataAwsEksClusterIdentityOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
}
exports.DataAwsEksClusterIdentityOidcOutputReference = DataAwsEksClusterIdentityOidcOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOidcOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOidcOutputReference", version: "19.15.0" };
class DataAwsEksClusterIdentityOidcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterIdentityOidcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterIdentityOidcList = DataAwsEksClusterIdentityOidcList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOidcList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOidcList", version: "19.15.0" };
function dataAwsEksClusterIdentityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterIdentityToTerraform = dataAwsEksClusterIdentityToTerraform;
function dataAwsEksClusterIdentityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterIdentityToHclTerraform = dataAwsEksClusterIdentityToHclTerraform;
class DataAwsEksClusterIdentityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // oidc - computed: true, optional: false, required: false
        this._oidc = new DataAwsEksClusterIdentityOidcList(this, "oidc", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get oidc() {
        return this._oidc;
    }
}
exports.DataAwsEksClusterIdentityOutputReference = DataAwsEksClusterIdentityOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOutputReference", version: "19.15.0" };
class DataAwsEksClusterIdentityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterIdentityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterIdentityList = DataAwsEksClusterIdentityList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityList[_h] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityList", version: "19.15.0" };
function dataAwsEksClusterKubernetesNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterKubernetesNetworkConfigToTerraform = dataAwsEksClusterKubernetesNetworkConfigToTerraform;
function dataAwsEksClusterKubernetesNetworkConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterKubernetesNetworkConfigToHclTerraform = dataAwsEksClusterKubernetesNetworkConfigToHclTerraform;
class DataAwsEksClusterKubernetesNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_family - computed: true, optional: false, required: false
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
    // service_ipv4_cidr - computed: true, optional: false, required: false
    get serviceIpv4Cidr() {
        return this.getStringAttribute('service_ipv4_cidr');
    }
    // service_ipv6_cidr - computed: true, optional: false, required: false
    get serviceIpv6Cidr() {
        return this.getStringAttribute('service_ipv6_cidr');
    }
}
exports.DataAwsEksClusterKubernetesNetworkConfigOutputReference = DataAwsEksClusterKubernetesNetworkConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterKubernetesNetworkConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterKubernetesNetworkConfigOutputReference", version: "19.15.0" };
class DataAwsEksClusterKubernetesNetworkConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterKubernetesNetworkConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterKubernetesNetworkConfigList = DataAwsEksClusterKubernetesNetworkConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterKubernetesNetworkConfigList[_k] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterKubernetesNetworkConfigList", version: "19.15.0" };
function dataAwsEksClusterOutpostConfigControlPlanePlacementToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterOutpostConfigControlPlanePlacementToTerraform = dataAwsEksClusterOutpostConfigControlPlanePlacementToTerraform;
function dataAwsEksClusterOutpostConfigControlPlanePlacementToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterOutpostConfigControlPlanePlacementToHclTerraform = dataAwsEksClusterOutpostConfigControlPlanePlacementToHclTerraform;
class DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // group_name - computed: true, optional: false, required: false
    get groupName() {
        return this.getStringAttribute('group_name');
    }
}
exports.DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference = DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference", version: "19.15.0" };
class DataAwsEksClusterOutpostConfigControlPlanePlacementList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterOutpostConfigControlPlanePlacementOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterOutpostConfigControlPlanePlacementList = DataAwsEksClusterOutpostConfigControlPlanePlacementList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigControlPlanePlacementList[_m] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigControlPlanePlacementList", version: "19.15.0" };
function dataAwsEksClusterOutpostConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterOutpostConfigToTerraform = dataAwsEksClusterOutpostConfigToTerraform;
function dataAwsEksClusterOutpostConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterOutpostConfigToHclTerraform = dataAwsEksClusterOutpostConfigToHclTerraform;
class DataAwsEksClusterOutpostConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // control_plane_placement - computed: true, optional: false, required: false
        this._controlPlanePlacement = new DataAwsEksClusterOutpostConfigControlPlanePlacementList(this, "control_plane_placement", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // control_plane_instance_type - computed: true, optional: false, required: false
    get controlPlaneInstanceType() {
        return this.getStringAttribute('control_plane_instance_type');
    }
    get controlPlanePlacement() {
        return this._controlPlanePlacement;
    }
    // outpost_arns - computed: true, optional: false, required: false
    get outpostArns() {
        return cdktf.Fn.tolist(this.getListAttribute('outpost_arns'));
    }
}
exports.DataAwsEksClusterOutpostConfigOutputReference = DataAwsEksClusterOutpostConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigOutputReference", version: "19.15.0" };
class DataAwsEksClusterOutpostConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterOutpostConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterOutpostConfigList = DataAwsEksClusterOutpostConfigList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigList[_p] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigList", version: "19.15.0" };
function dataAwsEksClusterVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterVpcConfigToTerraform = dataAwsEksClusterVpcConfigToTerraform;
function dataAwsEksClusterVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksClusterVpcConfigToHclTerraform = dataAwsEksClusterVpcConfigToHclTerraform;
class DataAwsEksClusterVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cluster_security_group_id - computed: true, optional: false, required: false
    get clusterSecurityGroupId() {
        return this.getStringAttribute('cluster_security_group_id');
    }
    // endpoint_private_access - computed: true, optional: false, required: false
    get endpointPrivateAccess() {
        return this.getBooleanAttribute('endpoint_private_access');
    }
    // endpoint_public_access - computed: true, optional: false, required: false
    get endpointPublicAccess() {
        return this.getBooleanAttribute('endpoint_public_access');
    }
    // public_access_cidrs - computed: true, optional: false, required: false
    get publicAccessCidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('public_access_cidrs'));
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsEksClusterVpcConfigOutputReference = DataAwsEksClusterVpcConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterVpcConfigOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterVpcConfigOutputReference", version: "19.15.0" };
class DataAwsEksClusterVpcConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterVpcConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterVpcConfigList = DataAwsEksClusterVpcConfigList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterVpcConfigList[_r] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterVpcConfigList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_cluster aws_eks_cluster}
*/
class DataAwsEksCluster extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEksCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEksCluster to import
    * @param importFromId The id of the existing DataAwsEksCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEksCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_eks_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_cluster aws_eks_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEksClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // access_config - computed: true, optional: false, required: false
        this._accessConfig = new DataAwsEksClusterAccessConfigList(this, "access_config", false);
        // certificate_authority - computed: true, optional: false, required: false
        this._certificateAuthority = new DataAwsEksClusterCertificateAuthorityList(this, "certificate_authority", false);
        // identity - computed: true, optional: false, required: false
        this._identity = new DataAwsEksClusterIdentityList(this, "identity", false);
        // kubernetes_network_config - computed: true, optional: false, required: false
        this._kubernetesNetworkConfig = new DataAwsEksClusterKubernetesNetworkConfigList(this, "kubernetes_network_config", false);
        // outpost_config - computed: true, optional: false, required: false
        this._outpostConfig = new DataAwsEksClusterOutpostConfigList(this, "outpost_config", false);
        // vpc_config - computed: true, optional: false, required: false
        this._vpcConfig = new DataAwsEksClusterVpcConfigList(this, "vpc_config", false);
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    get accessConfig() {
        return this._accessConfig;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get certificateAuthority() {
        return this._certificateAuthority;
    }
    // cluster_id - computed: true, optional: false, required: false
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // enabled_cluster_log_types - computed: true, optional: false, required: false
    get enabledClusterLogTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_cluster_log_types'));
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identity() {
        return this._identity;
    }
    get kubernetesNetworkConfig() {
        return this._kubernetesNetworkConfig;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get outpostConfig() {
        return this._outpostConfig;
    }
    // platform_version - computed: true, optional: false, required: false
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEksCluster = DataAwsEksCluster;
_s = JSII_RTTI_SYMBOL_1;
DataAwsEksCluster[_s] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksCluster", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEksCluster.tfResourceType = "aws_eks_cluster";
