// @cdktf/provider-aws/lib/xray-group/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrayGroup = exports.XrayGroupInsightsConfigurationOutputReference = exports.xrayGroupInsightsConfigurationToHclTerraform = exports.xrayGroupInsightsConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function xrayGroupInsightsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        insights_enabled: cdktf.booleanToTerraform(struct.insightsEnabled),
        notifications_enabled: cdktf.booleanToTerraform(struct.notificationsEnabled),
    };
}
exports.xrayGroupInsightsConfigurationToTerraform = xrayGroupInsightsConfigurationToTerraform;
function xrayGroupInsightsConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        insights_enabled: {
            value: cdktf.booleanToHclTerraform(struct.insightsEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        notifications_enabled: {
            value: cdktf.booleanToHclTerraform(struct.notificationsEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.xrayGroupInsightsConfigurationToHclTerraform = xrayGroupInsightsConfigurationToHclTerraform;
class XrayGroupInsightsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._insightsEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.insightsEnabled = this._insightsEnabled;
        }
        if (this._notificationsEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.notificationsEnabled = this._notificationsEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._insightsEnabled = undefined;
            this._notificationsEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._insightsEnabled = value.insightsEnabled;
            this._notificationsEnabled = value.notificationsEnabled;
        }
    }
    get insightsEnabled() {
        return this.getBooleanAttribute('insights_enabled');
    }
    set insightsEnabled(value) {
        this._insightsEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get insightsEnabledInput() {
        return this._insightsEnabled;
    }
    get notificationsEnabled() {
        return this.getBooleanAttribute('notifications_enabled');
    }
    set notificationsEnabled(value) {
        this._notificationsEnabled = value;
    }
    resetNotificationsEnabled() {
        this._notificationsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationsEnabledInput() {
        return this._notificationsEnabled;
    }
}
exports.XrayGroupInsightsConfigurationOutputReference = XrayGroupInsightsConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
XrayGroupInsightsConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.xrayGroup.XrayGroupInsightsConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/xray_group aws_xray_group}
*/
class XrayGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a XrayGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the XrayGroup to import
    * @param importFromId The id of the existing XrayGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/xray_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the XrayGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_xray_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/xray_group aws_xray_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options XrayGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_xray_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // insights_configuration - computed: false, optional: true, required: false
        this._insightsConfiguration = new XrayGroupInsightsConfigurationOutputReference(this, "insights_configuration");
        this._filterExpression = config.filterExpression;
        this._groupName = config.groupName;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._insightsConfiguration.internalValue = config.insightsConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get filterExpression() {
        return this.getStringAttribute('filter_expression');
    }
    set filterExpression(value) {
        this._filterExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterExpressionInput() {
        return this._filterExpression;
    }
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    set groupName(value) {
        this._groupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get groupNameInput() {
        return this._groupName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get insightsConfiguration() {
        return this._insightsConfiguration;
    }
    putInsightsConfiguration(value) {
        this._insightsConfiguration.internalValue = value;
    }
    resetInsightsConfiguration() {
        this._insightsConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get insightsConfigurationInput() {
        return this._insightsConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter_expression: cdktf.stringToTerraform(this._filterExpression),
            group_name: cdktf.stringToTerraform(this._groupName),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            insights_configuration: xrayGroupInsightsConfigurationToTerraform(this._insightsConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter_expression: {
                value: cdktf.stringToHclTerraform(this._filterExpression),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            group_name: {
                value: cdktf.stringToHclTerraform(this._groupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            insights_configuration: {
                value: xrayGroupInsightsConfigurationToHclTerraform(this._insightsConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "XrayGroupInsightsConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.XrayGroup = XrayGroup;
_b = JSII_RTTI_SYMBOL_1;
XrayGroup[_b] = { fqn: "@cdktf/provider-aws.xrayGroup.XrayGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
XrayGroup.tfResourceType = "aws_xray_group";
