// @cdktf/provider-aws/lib/lexv2models-bot-version/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lexv2ModelsBotVersion = exports.Lexv2ModelsBotVersionTimeoutsOutputReference = exports.lexv2ModelsBotVersionTimeoutsToHclTerraform = exports.lexv2ModelsBotVersionTimeoutsToTerraform = exports.Lexv2ModelsBotVersionLocaleSpecificationMap = exports.Lexv2ModelsBotVersionLocaleSpecificationOutputReference = exports.lexv2ModelsBotVersionLocaleSpecificationToHclTerraform = exports.lexv2ModelsBotVersionLocaleSpecificationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lexv2ModelsBotVersionLocaleSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_bot_version: cdktf.stringToTerraform(struct.sourceBotVersion),
    };
}
exports.lexv2ModelsBotVersionLocaleSpecificationToTerraform = lexv2ModelsBotVersionLocaleSpecificationToTerraform;
function lexv2ModelsBotVersionLocaleSpecificationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_bot_version: {
            value: cdktf.stringToHclTerraform(struct.sourceBotVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsBotVersionLocaleSpecificationToHclTerraform = lexv2ModelsBotVersionLocaleSpecificationToHclTerraform;
class Lexv2ModelsBotVersionLocaleSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectKey the key of this item in the map
    */
    constructor(terraformResource, terraformAttribute, complexObjectKey) {
        super(terraformResource, terraformAttribute, false, complexObjectKey);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceBotVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceBotVersion = this._sourceBotVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._sourceBotVersion = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._sourceBotVersion = value.sourceBotVersion;
        }
    }
    get sourceBotVersion() {
        return this.getStringAttribute('source_bot_version');
    }
    set sourceBotVersion(value) {
        this._sourceBotVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceBotVersionInput() {
        return this._sourceBotVersion;
    }
}
exports.Lexv2ModelsBotVersionLocaleSpecificationOutputReference = Lexv2ModelsBotVersionLocaleSpecificationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotVersionLocaleSpecificationOutputReference[_a] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotVersion.Lexv2ModelsBotVersionLocaleSpecificationOutputReference", version: "19.15.0" };
class Lexv2ModelsBotVersionLocaleSpecificationMap extends cdktf.ComplexMap {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
    }
    /**
    * @param key the key of the item to return
    */
    get(key) {
        return new Lexv2ModelsBotVersionLocaleSpecificationOutputReference(this.terraformResource, this.terraformAttribute, key);
    }
}
exports.Lexv2ModelsBotVersionLocaleSpecificationMap = Lexv2ModelsBotVersionLocaleSpecificationMap;
_b = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotVersionLocaleSpecificationMap[_b] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotVersion.Lexv2ModelsBotVersionLocaleSpecificationMap", version: "19.15.0" };
function lexv2ModelsBotVersionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.lexv2ModelsBotVersionTimeoutsToTerraform = lexv2ModelsBotVersionTimeoutsToTerraform;
function lexv2ModelsBotVersionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsBotVersionTimeoutsToHclTerraform = lexv2ModelsBotVersionTimeoutsToHclTerraform;
class Lexv2ModelsBotVersionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.Lexv2ModelsBotVersionTimeoutsOutputReference = Lexv2ModelsBotVersionTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotVersionTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotVersion.Lexv2ModelsBotVersionTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_version aws_lexv2models_bot_version}
*/
class Lexv2ModelsBotVersion extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Lexv2ModelsBotVersion resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Lexv2ModelsBotVersion to import
    * @param importFromId The id of the existing Lexv2ModelsBotVersion that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_version#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Lexv2ModelsBotVersion to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lexv2models_bot_version", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_version aws_lexv2models_bot_version} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Lexv2ModelsBotVersionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lexv2models_bot_version',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // locale_specification - computed: false, optional: false, required: true
        this._localeSpecification = new Lexv2ModelsBotVersionLocaleSpecificationMap(this, "locale_specification");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Lexv2ModelsBotVersionTimeoutsOutputReference(this, "timeouts");
        this._botId = config.botId;
        this._botVersion = config.botVersion;
        this._description = config.description;
        this._localeSpecification.internalValue = config.localeSpecification;
        this._timeouts.internalValue = config.timeouts;
    }
    get botId() {
        return this.getStringAttribute('bot_id');
    }
    set botId(value) {
        this._botId = value;
    }
    // Temporarily expose input value. Use with caution.
    get botIdInput() {
        return this._botId;
    }
    get botVersion() {
        return this.getStringAttribute('bot_version');
    }
    set botVersion(value) {
        this._botVersion = value;
    }
    resetBotVersion() {
        this._botVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get botVersionInput() {
        return this._botVersion;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get localeSpecification() {
        return this._localeSpecification;
    }
    putLocaleSpecification(value) {
        this._localeSpecification.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get localeSpecificationInput() {
        return this._localeSpecification.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bot_id: cdktf.stringToTerraform(this._botId),
            bot_version: cdktf.stringToTerraform(this._botVersion),
            description: cdktf.stringToTerraform(this._description),
            locale_specification: cdktf.hashMapper(lexv2ModelsBotVersionLocaleSpecificationToTerraform)(this._localeSpecification.internalValue),
            timeouts: lexv2ModelsBotVersionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bot_id: {
                value: cdktf.stringToHclTerraform(this._botId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bot_version: {
                value: cdktf.stringToHclTerraform(this._botVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            locale_specification: {
                value: cdktf.hashMapperHcl(lexv2ModelsBotVersionLocaleSpecificationToHclTerraform)(this._localeSpecification.internalValue),
                isBlock: true,
                type: "map",
                storageClassType: "Lexv2ModelsBotVersionLocaleSpecificationMap",
            },
            timeouts: {
                value: lexv2ModelsBotVersionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "Lexv2ModelsBotVersionTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Lexv2ModelsBotVersion = Lexv2ModelsBotVersion;
_d = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotVersion[_d] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotVersion.Lexv2ModelsBotVersion", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Lexv2ModelsBotVersion.tfResourceType = "aws_lexv2models_bot_version";
