// @cdktf/provider-aws/lib/data-aws-eks-node-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEksNodeGroup = exports.DataAwsEksNodeGroupTaintsList = exports.DataAwsEksNodeGroupTaintsOutputReference = exports.dataAwsEksNodeGroupTaintsToHclTerraform = exports.dataAwsEksNodeGroupTaintsToTerraform = exports.DataAwsEksNodeGroupScalingConfigList = exports.DataAwsEksNodeGroupScalingConfigOutputReference = exports.dataAwsEksNodeGroupScalingConfigToHclTerraform = exports.dataAwsEksNodeGroupScalingConfigToTerraform = exports.DataAwsEksNodeGroupResourcesList = exports.DataAwsEksNodeGroupResourcesOutputReference = exports.dataAwsEksNodeGroupResourcesToHclTerraform = exports.dataAwsEksNodeGroupResourcesToTerraform = exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsList = exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference = exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToHclTerraform = exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform = exports.DataAwsEksNodeGroupRemoteAccessList = exports.DataAwsEksNodeGroupRemoteAccessOutputReference = exports.dataAwsEksNodeGroupRemoteAccessToHclTerraform = exports.dataAwsEksNodeGroupRemoteAccessToTerraform = exports.DataAwsEksNodeGroupLaunchTemplateList = exports.DataAwsEksNodeGroupLaunchTemplateOutputReference = exports.dataAwsEksNodeGroupLaunchTemplateToHclTerraform = exports.dataAwsEksNodeGroupLaunchTemplateToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEksNodeGroupLaunchTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupLaunchTemplateToTerraform = dataAwsEksNodeGroupLaunchTemplateToTerraform;
function dataAwsEksNodeGroupLaunchTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupLaunchTemplateToHclTerraform = dataAwsEksNodeGroupLaunchTemplateToHclTerraform;
class DataAwsEksNodeGroupLaunchTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataAwsEksNodeGroupLaunchTemplateOutputReference = DataAwsEksNodeGroupLaunchTemplateOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupLaunchTemplateOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupLaunchTemplateOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupLaunchTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupLaunchTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupLaunchTemplateList = DataAwsEksNodeGroupLaunchTemplateList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupLaunchTemplateList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupLaunchTemplateList", version: "19.15.0" };
function dataAwsEksNodeGroupRemoteAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupRemoteAccessToTerraform = dataAwsEksNodeGroupRemoteAccessToTerraform;
function dataAwsEksNodeGroupRemoteAccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupRemoteAccessToHclTerraform = dataAwsEksNodeGroupRemoteAccessToHclTerraform;
class DataAwsEksNodeGroupRemoteAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ec2_ssh_key - computed: true, optional: false, required: false
    get ec2SshKey() {
        return this.getStringAttribute('ec2_ssh_key');
    }
    // source_security_group_ids - computed: true, optional: false, required: false
    get sourceSecurityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('source_security_group_ids'));
    }
}
exports.DataAwsEksNodeGroupRemoteAccessOutputReference = DataAwsEksNodeGroupRemoteAccessOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupRemoteAccessOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupRemoteAccessOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupRemoteAccessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupRemoteAccessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupRemoteAccessList = DataAwsEksNodeGroupRemoteAccessList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupRemoteAccessList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupRemoteAccessList", version: "19.15.0" };
function dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform = dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform;
function dataAwsEksNodeGroupResourcesAutoscalingGroupsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToHclTerraform = dataAwsEksNodeGroupResourcesAutoscalingGroupsToHclTerraform;
class DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference = DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupResourcesAutoscalingGroupsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsList = DataAwsEksNodeGroupResourcesAutoscalingGroupsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesAutoscalingGroupsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesAutoscalingGroupsList", version: "19.15.0" };
function dataAwsEksNodeGroupResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupResourcesToTerraform = dataAwsEksNodeGroupResourcesToTerraform;
function dataAwsEksNodeGroupResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupResourcesToHclTerraform = dataAwsEksNodeGroupResourcesToHclTerraform;
class DataAwsEksNodeGroupResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // autoscaling_groups - computed: true, optional: false, required: false
        this._autoscalingGroups = new DataAwsEksNodeGroupResourcesAutoscalingGroupsList(this, "autoscaling_groups", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get autoscalingGroups() {
        return this._autoscalingGroups;
    }
    // remote_access_security_group_id - computed: true, optional: false, required: false
    get remoteAccessSecurityGroupId() {
        return this.getStringAttribute('remote_access_security_group_id');
    }
}
exports.DataAwsEksNodeGroupResourcesOutputReference = DataAwsEksNodeGroupResourcesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupResourcesList = DataAwsEksNodeGroupResourcesList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesList[_h] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesList", version: "19.15.0" };
function dataAwsEksNodeGroupScalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupScalingConfigToTerraform = dataAwsEksNodeGroupScalingConfigToTerraform;
function dataAwsEksNodeGroupScalingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupScalingConfigToHclTerraform = dataAwsEksNodeGroupScalingConfigToHclTerraform;
class DataAwsEksNodeGroupScalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // desired_size - computed: true, optional: false, required: false
    get desiredSize() {
        return this.getNumberAttribute('desired_size');
    }
    // max_size - computed: true, optional: false, required: false
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    // min_size - computed: true, optional: false, required: false
    get minSize() {
        return this.getNumberAttribute('min_size');
    }
}
exports.DataAwsEksNodeGroupScalingConfigOutputReference = DataAwsEksNodeGroupScalingConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupScalingConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupScalingConfigOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupScalingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupScalingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupScalingConfigList = DataAwsEksNodeGroupScalingConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupScalingConfigList[_k] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupScalingConfigList", version: "19.15.0" };
function dataAwsEksNodeGroupTaintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupTaintsToTerraform = dataAwsEksNodeGroupTaintsToTerraform;
function dataAwsEksNodeGroupTaintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEksNodeGroupTaintsToHclTerraform = dataAwsEksNodeGroupTaintsToHclTerraform;
class DataAwsEksNodeGroupTaintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // effect - computed: true, optional: false, required: false
    get effect() {
        return this.getStringAttribute('effect');
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsEksNodeGroupTaintsOutputReference = DataAwsEksNodeGroupTaintsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupTaintsOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupTaintsOutputReference", version: "19.15.0" };
class DataAwsEksNodeGroupTaintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupTaintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupTaintsList = DataAwsEksNodeGroupTaintsList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupTaintsList[_m] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupTaintsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_node_group aws_eks_node_group}
*/
class DataAwsEksNodeGroup extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEksNodeGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEksNodeGroup to import
    * @param importFromId The id of the existing DataAwsEksNodeGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_node_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEksNodeGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_eks_node_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/eks_node_group aws_eks_node_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEksNodeGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_node_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // launch_template - computed: true, optional: false, required: false
        this._launchTemplate = new DataAwsEksNodeGroupLaunchTemplateList(this, "launch_template", false);
        // remote_access - computed: true, optional: false, required: false
        this._remoteAccess = new DataAwsEksNodeGroupRemoteAccessList(this, "remote_access", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataAwsEksNodeGroupResourcesList(this, "resources", false);
        // scaling_config - computed: true, optional: false, required: false
        this._scalingConfig = new DataAwsEksNodeGroupScalingConfigList(this, "scaling_config", false);
        // taints - computed: true, optional: false, required: false
        this._taints = new DataAwsEksNodeGroupTaintsList(this, "taints", false);
        this._clusterName = config.clusterName;
        this._id = config.id;
        this._nodeGroupName = config.nodeGroupName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ami_type - computed: true, optional: false, required: false
    get amiType() {
        return this.getStringAttribute('ami_type');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // capacity_type - computed: true, optional: false, required: false
    get capacityType() {
        return this.getStringAttribute('capacity_type');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_types - computed: true, optional: false, required: false
    get instanceTypes() {
        return this.getListAttribute('instance_types');
    }
    get labels() {
        return this._labels;
    }
    get launchTemplate() {
        return this._launchTemplate;
    }
    get nodeGroupName() {
        return this.getStringAttribute('node_group_name');
    }
    set nodeGroupName(value) {
        this._nodeGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get nodeGroupNameInput() {
        return this._nodeGroupName;
    }
    // node_role_arn - computed: true, optional: false, required: false
    get nodeRoleArn() {
        return this.getStringAttribute('node_role_arn');
    }
    // release_version - computed: true, optional: false, required: false
    get releaseVersion() {
        return this.getStringAttribute('release_version');
    }
    get remoteAccess() {
        return this._remoteAccess;
    }
    get resources() {
        return this._resources;
    }
    get scalingConfig() {
        return this._scalingConfig;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get taints() {
        return this._taints;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            id: cdktf.stringToTerraform(this._id),
            node_group_name: cdktf.stringToTerraform(this._nodeGroupName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_name: {
                value: cdktf.stringToHclTerraform(this._clusterName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            node_group_name: {
                value: cdktf.stringToHclTerraform(this._nodeGroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEksNodeGroup = DataAwsEksNodeGroup;
_o = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroup[_o] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEksNodeGroup.tfResourceType = "aws_eks_node_group";
