// @winglang/sdk/lib/.gen/providers/aws/api-gateway-stage/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_stage
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayStage = exports.ApiGatewayStageCanarySettingsOutputReference = exports.apiGatewayStageCanarySettingsToHclTerraform = exports.apiGatewayStageCanarySettingsToTerraform = exports.ApiGatewayStageAccessLogSettingsOutputReference = exports.apiGatewayStageAccessLogSettingsToHclTerraform = exports.apiGatewayStageAccessLogSettingsToTerraform = void 0;
const cdktf = __importStar(require("cdktf"));
function apiGatewayStageAccessLogSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_arn: cdktf.stringToTerraform(struct.destinationArn),
        format: cdktf.stringToTerraform(struct.format),
    };
}
exports.apiGatewayStageAccessLogSettingsToTerraform = apiGatewayStageAccessLogSettingsToTerraform;
function apiGatewayStageAccessLogSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination_arn: {
            value: cdktf.stringToHclTerraform(struct.destinationArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        format: {
            value: cdktf.stringToHclTerraform(struct.format),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apiGatewayStageAccessLogSettingsToHclTerraform = apiGatewayStageAccessLogSettingsToHclTerraform;
class ApiGatewayStageAccessLogSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationArn = this._destinationArn;
        }
        if (this._format !== undefined) {
            hasAnyValues = true;
            internalValueResult.format = this._format;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationArn = undefined;
            this._format = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationArn = value.destinationArn;
            this._format = value.format;
        }
    }
    get destinationArn() {
        return this.getStringAttribute('destination_arn');
    }
    set destinationArn(value) {
        this._destinationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationArnInput() {
        return this._destinationArn;
    }
    get format() {
        return this.getStringAttribute('format');
    }
    set format(value) {
        this._format = value;
    }
    // Temporarily expose input value. Use with caution.
    get formatInput() {
        return this._format;
    }
}
exports.ApiGatewayStageAccessLogSettingsOutputReference = ApiGatewayStageAccessLogSettingsOutputReference;
function apiGatewayStageCanarySettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        percent_traffic: cdktf.numberToTerraform(struct.percentTraffic),
        stage_variable_overrides: cdktf.hashMapper(cdktf.stringToTerraform)(struct.stageVariableOverrides),
        use_stage_cache: cdktf.booleanToTerraform(struct.useStageCache),
    };
}
exports.apiGatewayStageCanarySettingsToTerraform = apiGatewayStageCanarySettingsToTerraform;
function apiGatewayStageCanarySettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        percent_traffic: {
            value: cdktf.numberToHclTerraform(struct.percentTraffic),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        stage_variable_overrides: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.stageVariableOverrides),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        use_stage_cache: {
            value: cdktf.booleanToHclTerraform(struct.useStageCache),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apiGatewayStageCanarySettingsToHclTerraform = apiGatewayStageCanarySettingsToHclTerraform;
class ApiGatewayStageCanarySettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._percentTraffic !== undefined) {
            hasAnyValues = true;
            internalValueResult.percentTraffic = this._percentTraffic;
        }
        if (this._stageVariableOverrides !== undefined) {
            hasAnyValues = true;
            internalValueResult.stageVariableOverrides = this._stageVariableOverrides;
        }
        if (this._useStageCache !== undefined) {
            hasAnyValues = true;
            internalValueResult.useStageCache = this._useStageCache;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._percentTraffic = undefined;
            this._stageVariableOverrides = undefined;
            this._useStageCache = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._percentTraffic = value.percentTraffic;
            this._stageVariableOverrides = value.stageVariableOverrides;
            this._useStageCache = value.useStageCache;
        }
    }
    get percentTraffic() {
        return this.getNumberAttribute('percent_traffic');
    }
    set percentTraffic(value) {
        this._percentTraffic = value;
    }
    resetPercentTraffic() {
        this._percentTraffic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get percentTrafficInput() {
        return this._percentTraffic;
    }
    get stageVariableOverrides() {
        return this.getStringMapAttribute('stage_variable_overrides');
    }
    set stageVariableOverrides(value) {
        this._stageVariableOverrides = value;
    }
    resetStageVariableOverrides() {
        this._stageVariableOverrides = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stageVariableOverridesInput() {
        return this._stageVariableOverrides;
    }
    get useStageCache() {
        return this.getBooleanAttribute('use_stage_cache');
    }
    set useStageCache(value) {
        this._useStageCache = value;
    }
    resetUseStageCache() {
        this._useStageCache = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useStageCacheInput() {
        return this._useStageCache;
    }
}
exports.ApiGatewayStageCanarySettingsOutputReference = ApiGatewayStageCanarySettingsOutputReference;
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_stage aws_api_gateway_stage}
*/
class ApiGatewayStage extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApiGatewayStage resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApiGatewayStage to import
    * @param importFromId The id of the existing ApiGatewayStage that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_stage#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApiGatewayStage to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_api_gateway_stage", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/api_gateway_stage aws_api_gateway_stage} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayStageConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_stage',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.31.0',
                providerVersionConstraint: '~>5.31.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_log_settings - computed: false, optional: true, required: false
        this._accessLogSettings = new ApiGatewayStageAccessLogSettingsOutputReference(this, "access_log_settings");
        // canary_settings - computed: false, optional: true, required: false
        this._canarySettings = new ApiGatewayStageCanarySettingsOutputReference(this, "canary_settings");
        this._cacheClusterEnabled = config.cacheClusterEnabled;
        this._cacheClusterSize = config.cacheClusterSize;
        this._clientCertificateId = config.clientCertificateId;
        this._deploymentId = config.deploymentId;
        this._description = config.description;
        this._documentationVersion = config.documentationVersion;
        this._id = config.id;
        this._restApiId = config.restApiId;
        this._stageName = config.stageName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._variables = config.variables;
        this._xrayTracingEnabled = config.xrayTracingEnabled;
        this._accessLogSettings.internalValue = config.accessLogSettings;
        this._canarySettings.internalValue = config.canarySettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get cacheClusterEnabled() {
        return this.getBooleanAttribute('cache_cluster_enabled');
    }
    set cacheClusterEnabled(value) {
        this._cacheClusterEnabled = value;
    }
    resetCacheClusterEnabled() {
        this._cacheClusterEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cacheClusterEnabledInput() {
        return this._cacheClusterEnabled;
    }
    get cacheClusterSize() {
        return this.getStringAttribute('cache_cluster_size');
    }
    set cacheClusterSize(value) {
        this._cacheClusterSize = value;
    }
    resetCacheClusterSize() {
        this._cacheClusterSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cacheClusterSizeInput() {
        return this._cacheClusterSize;
    }
    get clientCertificateId() {
        return this.getStringAttribute('client_certificate_id');
    }
    set clientCertificateId(value) {
        this._clientCertificateId = value;
    }
    resetClientCertificateId() {
        this._clientCertificateId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientCertificateIdInput() {
        return this._clientCertificateId;
    }
    get deploymentId() {
        return this.getStringAttribute('deployment_id');
    }
    set deploymentId(value) {
        this._deploymentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentIdInput() {
        return this._deploymentId;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get documentationVersion() {
        return this.getStringAttribute('documentation_version');
    }
    set documentationVersion(value) {
        this._documentationVersion = value;
    }
    resetDocumentationVersion() {
        this._documentationVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentationVersionInput() {
        return this._documentationVersion;
    }
    // execution_arn - computed: true, optional: false, required: false
    get executionArn() {
        return this.getStringAttribute('execution_arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // invoke_url - computed: true, optional: false, required: false
    get invokeUrl() {
        return this.getStringAttribute('invoke_url');
    }
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get restApiIdInput() {
        return this._restApiId;
    }
    get stageName() {
        return this.getStringAttribute('stage_name');
    }
    set stageName(value) {
        this._stageName = value;
    }
    // Temporarily expose input value. Use with caution.
    get stageNameInput() {
        return this._stageName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get variables() {
        return this.getStringMapAttribute('variables');
    }
    set variables(value) {
        this._variables = value;
    }
    resetVariables() {
        this._variables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get variablesInput() {
        return this._variables;
    }
    // web_acl_arn - computed: true, optional: false, required: false
    get webAclArn() {
        return this.getStringAttribute('web_acl_arn');
    }
    get xrayTracingEnabled() {
        return this.getBooleanAttribute('xray_tracing_enabled');
    }
    set xrayTracingEnabled(value) {
        this._xrayTracingEnabled = value;
    }
    resetXrayTracingEnabled() {
        this._xrayTracingEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get xrayTracingEnabledInput() {
        return this._xrayTracingEnabled;
    }
    get accessLogSettings() {
        return this._accessLogSettings;
    }
    putAccessLogSettings(value) {
        this._accessLogSettings.internalValue = value;
    }
    resetAccessLogSettings() {
        this._accessLogSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessLogSettingsInput() {
        return this._accessLogSettings.internalValue;
    }
    get canarySettings() {
        return this._canarySettings;
    }
    putCanarySettings(value) {
        this._canarySettings.internalValue = value;
    }
    resetCanarySettings() {
        this._canarySettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get canarySettingsInput() {
        return this._canarySettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cache_cluster_enabled: cdktf.booleanToTerraform(this._cacheClusterEnabled),
            cache_cluster_size: cdktf.stringToTerraform(this._cacheClusterSize),
            client_certificate_id: cdktf.stringToTerraform(this._clientCertificateId),
            deployment_id: cdktf.stringToTerraform(this._deploymentId),
            description: cdktf.stringToTerraform(this._description),
            documentation_version: cdktf.stringToTerraform(this._documentationVersion),
            id: cdktf.stringToTerraform(this._id),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            stage_name: cdktf.stringToTerraform(this._stageName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            variables: cdktf.hashMapper(cdktf.stringToTerraform)(this._variables),
            xray_tracing_enabled: cdktf.booleanToTerraform(this._xrayTracingEnabled),
            access_log_settings: apiGatewayStageAccessLogSettingsToTerraform(this._accessLogSettings.internalValue),
            canary_settings: apiGatewayStageCanarySettingsToTerraform(this._canarySettings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cache_cluster_enabled: {
                value: cdktf.booleanToHclTerraform(this._cacheClusterEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            cache_cluster_size: {
                value: cdktf.stringToHclTerraform(this._cacheClusterSize),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            client_certificate_id: {
                value: cdktf.stringToHclTerraform(this._clientCertificateId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deployment_id: {
                value: cdktf.stringToHclTerraform(this._deploymentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            documentation_version: {
                value: cdktf.stringToHclTerraform(this._documentationVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rest_api_id: {
                value: cdktf.stringToHclTerraform(this._restApiId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            stage_name: {
                value: cdktf.stringToHclTerraform(this._stageName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            variables: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._variables),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            xray_tracing_enabled: {
                value: cdktf.booleanToHclTerraform(this._xrayTracingEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            access_log_settings: {
                value: apiGatewayStageAccessLogSettingsToHclTerraform(this._accessLogSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApiGatewayStageAccessLogSettingsList",
            },
            canary_settings: {
                value: apiGatewayStageCanarySettingsToHclTerraform(this._canarySettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApiGatewayStageCanarySettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApiGatewayStage = ApiGatewayStage;
// =================
// STATIC PROPERTIES
// =================
ApiGatewayStage.tfResourceType = "aws_api_gateway_stage";
