// @cdktf/provider-aws/lib/data-aws-ecr-image/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcrImage = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_image aws_ecr_image}
*/
class DataAwsEcrImage extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEcrImage resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEcrImage to import
    * @param importFromId The id of the existing DataAwsEcrImage that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_image#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEcrImage to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecr_image", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_image aws_ecr_image} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEcrImageConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_image',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._imageDigest = config.imageDigest;
        this._imageTag = config.imageTag;
        this._mostRecent = config.mostRecent;
        this._registryId = config.registryId;
        this._repositoryName = config.repositoryName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get imageDigest() {
        return this.getStringAttribute('image_digest');
    }
    set imageDigest(value) {
        this._imageDigest = value;
    }
    resetImageDigest() {
        this._imageDigest = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageDigestInput() {
        return this._imageDigest;
    }
    // image_pushed_at - computed: true, optional: false, required: false
    get imagePushedAt() {
        return this.getNumberAttribute('image_pushed_at');
    }
    // image_size_in_bytes - computed: true, optional: false, required: false
    get imageSizeInBytes() {
        return this.getNumberAttribute('image_size_in_bytes');
    }
    get imageTag() {
        return this.getStringAttribute('image_tag');
    }
    set imageTag(value) {
        this._imageTag = value;
    }
    resetImageTag() {
        this._imageTag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageTagInput() {
        return this._imageTag;
    }
    // image_tags - computed: true, optional: false, required: false
    get imageTags() {
        return this.getListAttribute('image_tags');
    }
    // image_uri - computed: true, optional: false, required: false
    get imageUri() {
        return this.getStringAttribute('image_uri');
    }
    get mostRecent() {
        return this.getBooleanAttribute('most_recent');
    }
    set mostRecent(value) {
        this._mostRecent = value;
    }
    resetMostRecent() {
        this._mostRecent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostRecentInput() {
        return this._mostRecent;
    }
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    set registryId(value) {
        this._registryId = value;
    }
    resetRegistryId() {
        this._registryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get registryIdInput() {
        return this._registryId;
    }
    get repositoryName() {
        return this.getStringAttribute('repository_name');
    }
    set repositoryName(value) {
        this._repositoryName = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryNameInput() {
        return this._repositoryName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            image_digest: cdktf.stringToTerraform(this._imageDigest),
            image_tag: cdktf.stringToTerraform(this._imageTag),
            most_recent: cdktf.booleanToTerraform(this._mostRecent),
            registry_id: cdktf.stringToTerraform(this._registryId),
            repository_name: cdktf.stringToTerraform(this._repositoryName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            image_digest: {
                value: cdktf.stringToHclTerraform(this._imageDigest),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            image_tag: {
                value: cdktf.stringToHclTerraform(this._imageTag),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            most_recent: {
                value: cdktf.booleanToHclTerraform(this._mostRecent),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            registry_id: {
                value: cdktf.stringToHclTerraform(this._registryId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            repository_name: {
                value: cdktf.stringToHclTerraform(this._repositoryName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEcrImage = DataAwsEcrImage;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcrImage[_a] = { fqn: "@cdktf/provider-aws.dataAwsEcrImage.DataAwsEcrImage", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEcrImage.tfResourceType = "aws_ecr_image";
