// @cdktf/provider-aws/lib/ssoadmin-application/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminApplication = exports.SsoadminApplicationPortalOptionsList = exports.SsoadminApplicationPortalOptionsOutputReference = exports.ssoadminApplicationPortalOptionsToHclTerraform = exports.ssoadminApplicationPortalOptionsToTerraform = exports.SsoadminApplicationPortalOptionsSignInOptionsList = exports.SsoadminApplicationPortalOptionsSignInOptionsOutputReference = exports.ssoadminApplicationPortalOptionsSignInOptionsToHclTerraform = exports.ssoadminApplicationPortalOptionsSignInOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssoadminApplicationPortalOptionsSignInOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        application_url: cdktf.stringToTerraform(struct.applicationUrl),
        origin: cdktf.stringToTerraform(struct.origin),
    };
}
exports.ssoadminApplicationPortalOptionsSignInOptionsToTerraform = ssoadminApplicationPortalOptionsSignInOptionsToTerraform;
function ssoadminApplicationPortalOptionsSignInOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        application_url: {
            value: cdktf.stringToHclTerraform(struct.applicationUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        origin: {
            value: cdktf.stringToHclTerraform(struct.origin),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminApplicationPortalOptionsSignInOptionsToHclTerraform = ssoadminApplicationPortalOptionsSignInOptionsToHclTerraform;
class SsoadminApplicationPortalOptionsSignInOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._applicationUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.applicationUrl = this._applicationUrl;
        }
        if (this._origin !== undefined) {
            hasAnyValues = true;
            internalValueResult.origin = this._origin;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._applicationUrl = undefined;
            this._origin = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._applicationUrl = value.applicationUrl;
            this._origin = value.origin;
        }
    }
    get applicationUrl() {
        return this.getStringAttribute('application_url');
    }
    set applicationUrl(value) {
        this._applicationUrl = value;
    }
    resetApplicationUrl() {
        this._applicationUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationUrlInput() {
        return this._applicationUrl;
    }
    get origin() {
        return this.getStringAttribute('origin');
    }
    set origin(value) {
        this._origin = value;
    }
    // Temporarily expose input value. Use with caution.
    get originInput() {
        return this._origin;
    }
}
exports.SsoadminApplicationPortalOptionsSignInOptionsOutputReference = SsoadminApplicationPortalOptionsSignInOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsoadminApplicationPortalOptionsSignInOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssoadminApplication.SsoadminApplicationPortalOptionsSignInOptionsOutputReference", version: "19.15.0" };
class SsoadminApplicationPortalOptionsSignInOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminApplicationPortalOptionsSignInOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminApplicationPortalOptionsSignInOptionsList = SsoadminApplicationPortalOptionsSignInOptionsList;
_b = JSII_RTTI_SYMBOL_1;
SsoadminApplicationPortalOptionsSignInOptionsList[_b] = { fqn: "@cdktf/provider-aws.ssoadminApplication.SsoadminApplicationPortalOptionsSignInOptionsList", version: "19.15.0" };
function ssoadminApplicationPortalOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        visibility: cdktf.stringToTerraform(struct.visibility),
        sign_in_options: cdktf.listMapper(ssoadminApplicationPortalOptionsSignInOptionsToTerraform, true)(struct.signInOptions),
    };
}
exports.ssoadminApplicationPortalOptionsToTerraform = ssoadminApplicationPortalOptionsToTerraform;
function ssoadminApplicationPortalOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        visibility: {
            value: cdktf.stringToHclTerraform(struct.visibility),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sign_in_options: {
            value: cdktf.listMapperHcl(ssoadminApplicationPortalOptionsSignInOptionsToHclTerraform, true)(struct.signInOptions),
            isBlock: true,
            type: "list",
            storageClassType: "SsoadminApplicationPortalOptionsSignInOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ssoadminApplicationPortalOptionsToHclTerraform = ssoadminApplicationPortalOptionsToHclTerraform;
class SsoadminApplicationPortalOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // sign_in_options - computed: false, optional: true, required: false
        this._signInOptions = new SsoadminApplicationPortalOptionsSignInOptionsList(this, "sign_in_options", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._visibility !== undefined) {
            hasAnyValues = true;
            internalValueResult.visibility = this._visibility;
        }
        if (this._signInOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.signInOptions = this._signInOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._visibility = undefined;
            this._signInOptions.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._visibility = value.visibility;
            this._signInOptions.internalValue = value.signInOptions;
        }
    }
    get visibility() {
        return this.getStringAttribute('visibility');
    }
    set visibility(value) {
        this._visibility = value;
    }
    resetVisibility() {
        this._visibility = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get visibilityInput() {
        return this._visibility;
    }
    get signInOptions() {
        return this._signInOptions;
    }
    putSignInOptions(value) {
        this._signInOptions.internalValue = value;
    }
    resetSignInOptions() {
        this._signInOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get signInOptionsInput() {
        return this._signInOptions.internalValue;
    }
}
exports.SsoadminApplicationPortalOptionsOutputReference = SsoadminApplicationPortalOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SsoadminApplicationPortalOptionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.ssoadminApplication.SsoadminApplicationPortalOptionsOutputReference", version: "19.15.0" };
class SsoadminApplicationPortalOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SsoadminApplicationPortalOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SsoadminApplicationPortalOptionsList = SsoadminApplicationPortalOptionsList;
_d = JSII_RTTI_SYMBOL_1;
SsoadminApplicationPortalOptionsList[_d] = { fqn: "@cdktf/provider-aws.ssoadminApplication.SsoadminApplicationPortalOptionsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_application aws_ssoadmin_application}
*/
class SsoadminApplication extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SsoadminApplication resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SsoadminApplication to import
    * @param importFromId The id of the existing SsoadminApplication that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_application#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SsoadminApplication to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_application", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ssoadmin_application aws_ssoadmin_application} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsoadminApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_application',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // portal_options - computed: false, optional: true, required: false
        this._portalOptions = new SsoadminApplicationPortalOptionsList(this, "portal_options", false);
        this._applicationProviderArn = config.applicationProviderArn;
        this._clientToken = config.clientToken;
        this._description = config.description;
        this._instanceArn = config.instanceArn;
        this._name = config.name;
        this._status = config.status;
        this._tags = config.tags;
        this._portalOptions.internalValue = config.portalOptions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // application_account - computed: true, optional: false, required: false
    get applicationAccount() {
        return this.getStringAttribute('application_account');
    }
    // application_arn - computed: true, optional: false, required: false
    get applicationArn() {
        return this.getStringAttribute('application_arn');
    }
    get applicationProviderArn() {
        return this.getStringAttribute('application_provider_arn');
    }
    set applicationProviderArn(value) {
        this._applicationProviderArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationProviderArnInput() {
        return this._applicationProviderArn;
    }
    get clientToken() {
        return this.getStringAttribute('client_token');
    }
    set clientToken(value) {
        this._clientToken = value;
    }
    resetClientToken() {
        this._clientToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientTokenInput() {
        return this._clientToken;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceArnInput() {
        return this._instanceArn;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get portalOptions() {
        return this._portalOptions;
    }
    putPortalOptions(value) {
        this._portalOptions.internalValue = value;
    }
    resetPortalOptions() {
        this._portalOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portalOptionsInput() {
        return this._portalOptions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_provider_arn: cdktf.stringToTerraform(this._applicationProviderArn),
            client_token: cdktf.stringToTerraform(this._clientToken),
            description: cdktf.stringToTerraform(this._description),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            name: cdktf.stringToTerraform(this._name),
            status: cdktf.stringToTerraform(this._status),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            portal_options: cdktf.listMapper(ssoadminApplicationPortalOptionsToTerraform, true)(this._portalOptions.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_provider_arn: {
                value: cdktf.stringToHclTerraform(this._applicationProviderArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            client_token: {
                value: cdktf.stringToHclTerraform(this._clientToken),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_arn: {
                value: cdktf.stringToHclTerraform(this._instanceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            portal_options: {
                value: cdktf.listMapperHcl(ssoadminApplicationPortalOptionsToHclTerraform, true)(this._portalOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SsoadminApplicationPortalOptionsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SsoadminApplication = SsoadminApplication;
_e = JSII_RTTI_SYMBOL_1;
SsoadminApplication[_e] = { fqn: "@cdktf/provider-aws.ssoadminApplication.SsoadminApplication", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SsoadminApplication.tfResourceType = "aws_ssoadmin_application";
