// @cdktf/provider-aws/lib/networkmanager-connect-attachment/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkmanagerConnectAttachment = exports.NetworkmanagerConnectAttachmentTimeoutsOutputReference = exports.networkmanagerConnectAttachmentTimeoutsToHclTerraform = exports.networkmanagerConnectAttachmentTimeoutsToTerraform = exports.NetworkmanagerConnectAttachmentOptionsOutputReference = exports.networkmanagerConnectAttachmentOptionsToHclTerraform = exports.networkmanagerConnectAttachmentOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkmanagerConnectAttachmentOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        protocol: cdktf.stringToTerraform(struct.protocol),
    };
}
exports.networkmanagerConnectAttachmentOptionsToTerraform = networkmanagerConnectAttachmentOptionsToTerraform;
function networkmanagerConnectAttachmentOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerConnectAttachmentOptionsToHclTerraform = networkmanagerConnectAttachmentOptionsToHclTerraform;
class NetworkmanagerConnectAttachmentOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._protocol = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._protocol = value.protocol;
        }
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
}
exports.NetworkmanagerConnectAttachmentOptionsOutputReference = NetworkmanagerConnectAttachmentOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectAttachmentOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkmanagerConnectAttachment.NetworkmanagerConnectAttachmentOptionsOutputReference", version: "19.15.0" };
function networkmanagerConnectAttachmentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.networkmanagerConnectAttachmentTimeoutsToTerraform = networkmanagerConnectAttachmentTimeoutsToTerraform;
function networkmanagerConnectAttachmentTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerConnectAttachmentTimeoutsToHclTerraform = networkmanagerConnectAttachmentTimeoutsToHclTerraform;
class NetworkmanagerConnectAttachmentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.NetworkmanagerConnectAttachmentTimeoutsOutputReference = NetworkmanagerConnectAttachmentTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectAttachmentTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.networkmanagerConnectAttachment.NetworkmanagerConnectAttachmentTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_attachment aws_networkmanager_connect_attachment}
*/
class NetworkmanagerConnectAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkmanagerConnectAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkmanagerConnectAttachment to import
    * @param importFromId The id of the existing NetworkmanagerConnectAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkmanagerConnectAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_networkmanager_connect_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_attachment aws_networkmanager_connect_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkmanagerConnectAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkmanager_connect_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // options - computed: false, optional: false, required: true
        this._options = new NetworkmanagerConnectAttachmentOptionsOutputReference(this, "options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkmanagerConnectAttachmentTimeoutsOutputReference(this, "timeouts");
        this._coreNetworkId = config.coreNetworkId;
        this._edgeLocation = config.edgeLocation;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._transportAttachmentId = config.transportAttachmentId;
        this._options.internalValue = config.options;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attachment_id - computed: true, optional: false, required: false
    get attachmentId() {
        return this.getStringAttribute('attachment_id');
    }
    // attachment_policy_rule_number - computed: true, optional: false, required: false
    get attachmentPolicyRuleNumber() {
        return this.getNumberAttribute('attachment_policy_rule_number');
    }
    // attachment_type - computed: true, optional: false, required: false
    get attachmentType() {
        return this.getStringAttribute('attachment_type');
    }
    // core_network_arn - computed: true, optional: false, required: false
    get coreNetworkArn() {
        return this.getStringAttribute('core_network_arn');
    }
    get coreNetworkId() {
        return this.getStringAttribute('core_network_id');
    }
    set coreNetworkId(value) {
        this._coreNetworkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get coreNetworkIdInput() {
        return this._coreNetworkId;
    }
    get edgeLocation() {
        return this.getStringAttribute('edge_location');
    }
    set edgeLocation(value) {
        this._edgeLocation = value;
    }
    // Temporarily expose input value. Use with caution.
    get edgeLocationInput() {
        return this._edgeLocation;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_account_id - computed: true, optional: false, required: false
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    // resource_arn - computed: true, optional: false, required: false
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    // segment_name - computed: true, optional: false, required: false
    get segmentName() {
        return this.getStringAttribute('segment_name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get transportAttachmentId() {
        return this.getStringAttribute('transport_attachment_id');
    }
    set transportAttachmentId(value) {
        this._transportAttachmentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get transportAttachmentIdInput() {
        return this._transportAttachmentId;
    }
    get options() {
        return this._options;
    }
    putOptions(value) {
        this._options.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            core_network_id: cdktf.stringToTerraform(this._coreNetworkId),
            edge_location: cdktf.stringToTerraform(this._edgeLocation),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            transport_attachment_id: cdktf.stringToTerraform(this._transportAttachmentId),
            options: networkmanagerConnectAttachmentOptionsToTerraform(this._options.internalValue),
            timeouts: networkmanagerConnectAttachmentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            core_network_id: {
                value: cdktf.stringToHclTerraform(this._coreNetworkId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            edge_location: {
                value: cdktf.stringToHclTerraform(this._edgeLocation),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            transport_attachment_id: {
                value: cdktf.stringToHclTerraform(this._transportAttachmentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            options: {
                value: networkmanagerConnectAttachmentOptionsToHclTerraform(this._options.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkmanagerConnectAttachmentOptionsList",
            },
            timeouts: {
                value: networkmanagerConnectAttachmentTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkmanagerConnectAttachmentTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkmanagerConnectAttachment = NetworkmanagerConnectAttachment;
_c = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectAttachment[_c] = { fqn: "@cdktf/provider-aws.networkmanagerConnectAttachment.NetworkmanagerConnectAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
NetworkmanagerConnectAttachment.tfResourceType = "aws_networkmanager_connect_attachment";
