// @cdktf/provider-aws/lib/data-aws-elasticache-user/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElasticacheUser = exports.DataAwsElasticacheUserAuthenticationModeList = exports.DataAwsElasticacheUserAuthenticationModeOutputReference = exports.dataAwsElasticacheUserAuthenticationModeToHclTerraform = exports.dataAwsElasticacheUserAuthenticationModeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsElasticacheUserAuthenticationModeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        password_count: cdktf.numberToTerraform(struct.passwordCount),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataAwsElasticacheUserAuthenticationModeToTerraform = dataAwsElasticacheUserAuthenticationModeToTerraform;
function dataAwsElasticacheUserAuthenticationModeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        password_count: {
            value: cdktf.numberToHclTerraform(struct.passwordCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsElasticacheUserAuthenticationModeToHclTerraform = dataAwsElasticacheUserAuthenticationModeToHclTerraform;
class DataAwsElasticacheUserAuthenticationModeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._passwordCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.passwordCount = this._passwordCount;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._passwordCount = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._passwordCount = value.passwordCount;
            this._type = value.type;
        }
    }
    get passwordCount() {
        return this.getNumberAttribute('password_count');
    }
    set passwordCount(value) {
        this._passwordCount = value;
    }
    resetPasswordCount() {
        this._passwordCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordCountInput() {
        return this._passwordCount;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataAwsElasticacheUserAuthenticationModeOutputReference = DataAwsElasticacheUserAuthenticationModeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheUserAuthenticationModeOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsElasticacheUser.DataAwsElasticacheUserAuthenticationModeOutputReference", version: "19.15.0" };
class DataAwsElasticacheUserAuthenticationModeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsElasticacheUserAuthenticationModeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsElasticacheUserAuthenticationModeList = DataAwsElasticacheUserAuthenticationModeList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheUserAuthenticationModeList[_b] = { fqn: "@cdktf/provider-aws.dataAwsElasticacheUser.DataAwsElasticacheUserAuthenticationModeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/elasticache_user aws_elasticache_user}
*/
class DataAwsElasticacheUser extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsElasticacheUser resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsElasticacheUser to import
    * @param importFromId The id of the existing DataAwsElasticacheUser that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/elasticache_user#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsElasticacheUser to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_elasticache_user", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/elasticache_user aws_elasticache_user} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsElasticacheUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_user',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // authentication_mode - computed: false, optional: true, required: false
        this._authenticationMode = new DataAwsElasticacheUserAuthenticationModeList(this, "authentication_mode", false);
        this._accessString = config.accessString;
        this._engine = config.engine;
        this._id = config.id;
        this._noPasswordRequired = config.noPasswordRequired;
        this._passwords = config.passwords;
        this._userId = config.userId;
        this._userName = config.userName;
        this._authenticationMode.internalValue = config.authenticationMode;
    }
    get accessString() {
        return this.getStringAttribute('access_string');
    }
    set accessString(value) {
        this._accessString = value;
    }
    resetAccessString() {
        this._accessString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessStringInput() {
        return this._accessString;
    }
    get engine() {
        return this.getStringAttribute('engine');
    }
    set engine(value) {
        this._engine = value;
    }
    resetEngine() {
        this._engine = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engineInput() {
        return this._engine;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get noPasswordRequired() {
        return this.getBooleanAttribute('no_password_required');
    }
    set noPasswordRequired(value) {
        this._noPasswordRequired = value;
    }
    resetNoPasswordRequired() {
        this._noPasswordRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get noPasswordRequiredInput() {
        return this._noPasswordRequired;
    }
    get passwords() {
        return cdktf.Fn.tolist(this.getListAttribute('passwords'));
    }
    set passwords(value) {
        this._passwords = value;
    }
    resetPasswords() {
        this._passwords = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordsInput() {
        return this._passwords;
    }
    get userId() {
        return this.getStringAttribute('user_id');
    }
    set userId(value) {
        this._userId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userIdInput() {
        return this._userId;
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    resetUserName() {
        this._userName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    get authenticationMode() {
        return this._authenticationMode;
    }
    putAuthenticationMode(value) {
        this._authenticationMode.internalValue = value;
    }
    resetAuthenticationMode() {
        this._authenticationMode.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationModeInput() {
        return this._authenticationMode.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_string: cdktf.stringToTerraform(this._accessString),
            engine: cdktf.stringToTerraform(this._engine),
            id: cdktf.stringToTerraform(this._id),
            no_password_required: cdktf.booleanToTerraform(this._noPasswordRequired),
            passwords: cdktf.listMapper(cdktf.stringToTerraform, false)(this._passwords),
            user_id: cdktf.stringToTerraform(this._userId),
            user_name: cdktf.stringToTerraform(this._userName),
            authentication_mode: cdktf.listMapper(dataAwsElasticacheUserAuthenticationModeToTerraform, true)(this._authenticationMode.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_string: {
                value: cdktf.stringToHclTerraform(this._accessString),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            engine: {
                value: cdktf.stringToHclTerraform(this._engine),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            no_password_required: {
                value: cdktf.booleanToHclTerraform(this._noPasswordRequired),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            passwords: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._passwords),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            user_id: {
                value: cdktf.stringToHclTerraform(this._userId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_name: {
                value: cdktf.stringToHclTerraform(this._userName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authentication_mode: {
                value: cdktf.listMapperHcl(dataAwsElasticacheUserAuthenticationModeToHclTerraform, true)(this._authenticationMode.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsElasticacheUserAuthenticationModeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsElasticacheUser = DataAwsElasticacheUser;
_c = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheUser[_c] = { fqn: "@cdktf/provider-aws.dataAwsElasticacheUser.DataAwsElasticacheUser", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsElasticacheUser.tfResourceType = "aws_elasticache_user";
