// @cdktf/provider-aws/lib/bedrock-provisioned-model-throughput/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BedrockProvisionedModelThroughput = exports.BedrockProvisionedModelThroughputTimeoutsOutputReference = exports.bedrockProvisionedModelThroughputTimeoutsToHclTerraform = exports.bedrockProvisionedModelThroughputTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function bedrockProvisionedModelThroughputTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.bedrockProvisionedModelThroughputTimeoutsToTerraform = bedrockProvisionedModelThroughputTimeoutsToTerraform;
function bedrockProvisionedModelThroughputTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bedrockProvisionedModelThroughputTimeoutsToHclTerraform = bedrockProvisionedModelThroughputTimeoutsToHclTerraform;
class BedrockProvisionedModelThroughputTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.BedrockProvisionedModelThroughputTimeoutsOutputReference = BedrockProvisionedModelThroughputTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BedrockProvisionedModelThroughputTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.bedrockProvisionedModelThroughput.BedrockProvisionedModelThroughputTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/bedrock_provisioned_model_throughput aws_bedrock_provisioned_model_throughput}
*/
class BedrockProvisionedModelThroughput extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BedrockProvisionedModelThroughput resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BedrockProvisionedModelThroughput to import
    * @param importFromId The id of the existing BedrockProvisionedModelThroughput that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/bedrock_provisioned_model_throughput#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BedrockProvisionedModelThroughput to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_bedrock_provisioned_model_throughput", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/bedrock_provisioned_model_throughput aws_bedrock_provisioned_model_throughput} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BedrockProvisionedModelThroughputConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_bedrock_provisioned_model_throughput',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BedrockProvisionedModelThroughputTimeoutsOutputReference(this, "timeouts");
        this._commitmentDuration = config.commitmentDuration;
        this._modelArn = config.modelArn;
        this._modelUnits = config.modelUnits;
        this._provisionedModelName = config.provisionedModelName;
        this._tags = config.tags;
        this._timeouts.internalValue = config.timeouts;
    }
    get commitmentDuration() {
        return this.getStringAttribute('commitment_duration');
    }
    set commitmentDuration(value) {
        this._commitmentDuration = value;
    }
    resetCommitmentDuration() {
        this._commitmentDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commitmentDurationInput() {
        return this._commitmentDuration;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get modelArn() {
        return this.getStringAttribute('model_arn');
    }
    set modelArn(value) {
        this._modelArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get modelArnInput() {
        return this._modelArn;
    }
    get modelUnits() {
        return this.getNumberAttribute('model_units');
    }
    set modelUnits(value) {
        this._modelUnits = value;
    }
    // Temporarily expose input value. Use with caution.
    get modelUnitsInput() {
        return this._modelUnits;
    }
    // provisioned_model_arn - computed: true, optional: false, required: false
    get provisionedModelArn() {
        return this.getStringAttribute('provisioned_model_arn');
    }
    get provisionedModelName() {
        return this.getStringAttribute('provisioned_model_name');
    }
    set provisionedModelName(value) {
        this._provisionedModelName = value;
    }
    // Temporarily expose input value. Use with caution.
    get provisionedModelNameInput() {
        return this._provisionedModelName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            commitment_duration: cdktf.stringToTerraform(this._commitmentDuration),
            model_arn: cdktf.stringToTerraform(this._modelArn),
            model_units: cdktf.numberToTerraform(this._modelUnits),
            provisioned_model_name: cdktf.stringToTerraform(this._provisionedModelName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            timeouts: bedrockProvisionedModelThroughputTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            commitment_duration: {
                value: cdktf.stringToHclTerraform(this._commitmentDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            model_arn: {
                value: cdktf.stringToHclTerraform(this._modelArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            model_units: {
                value: cdktf.numberToHclTerraform(this._modelUnits),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            provisioned_model_name: {
                value: cdktf.stringToHclTerraform(this._provisionedModelName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeouts: {
                value: bedrockProvisionedModelThroughputTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BedrockProvisionedModelThroughputTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BedrockProvisionedModelThroughput = BedrockProvisionedModelThroughput;
_b = JSII_RTTI_SYMBOL_1;
BedrockProvisionedModelThroughput[_b] = { fqn: "@cdktf/provider-aws.bedrockProvisionedModelThroughput.BedrockProvisionedModelThroughput", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
BedrockProvisionedModelThroughput.tfResourceType = "aws_bedrock_provisioned_model_throughput";
