// @cdktf/provider-aws/lib/data-aws-sesv2-email-identity/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSesv2EmailIdentity = exports.DataAwsSesv2EmailIdentityDkimSigningAttributesList = exports.DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference = exports.dataAwsSesv2EmailIdentityDkimSigningAttributesToHclTerraform = exports.dataAwsSesv2EmailIdentityDkimSigningAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSesv2EmailIdentityDkimSigningAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSesv2EmailIdentityDkimSigningAttributesToTerraform = dataAwsSesv2EmailIdentityDkimSigningAttributesToTerraform;
function dataAwsSesv2EmailIdentityDkimSigningAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSesv2EmailIdentityDkimSigningAttributesToHclTerraform = dataAwsSesv2EmailIdentityDkimSigningAttributesToHclTerraform;
class DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // current_signing_key_length - computed: true, optional: false, required: false
    get currentSigningKeyLength() {
        return this.getStringAttribute('current_signing_key_length');
    }
    // domain_signing_private_key - computed: true, optional: false, required: false
    get domainSigningPrivateKey() {
        return this.getStringAttribute('domain_signing_private_key');
    }
    // domain_signing_selector - computed: true, optional: false, required: false
    get domainSigningSelector() {
        return this.getStringAttribute('domain_signing_selector');
    }
    // last_key_generation_timestamp - computed: true, optional: false, required: false
    get lastKeyGenerationTimestamp() {
        return this.getStringAttribute('last_key_generation_timestamp');
    }
    // next_signing_key_length - computed: true, optional: false, required: false
    get nextSigningKeyLength() {
        return this.getStringAttribute('next_signing_key_length');
    }
    // signing_attributes_origin - computed: true, optional: false, required: false
    get signingAttributesOrigin() {
        return this.getStringAttribute('signing_attributes_origin');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // tokens - computed: true, optional: false, required: false
    get tokens() {
        return this.getListAttribute('tokens');
    }
}
exports.DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference = DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSesv2EmailIdentity.DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference", version: "19.15.0" };
class DataAwsSesv2EmailIdentityDkimSigningAttributesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSesv2EmailIdentityDkimSigningAttributesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSesv2EmailIdentityDkimSigningAttributesList = DataAwsSesv2EmailIdentityDkimSigningAttributesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSesv2EmailIdentityDkimSigningAttributesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSesv2EmailIdentity.DataAwsSesv2EmailIdentityDkimSigningAttributesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity aws_sesv2_email_identity}
*/
class DataAwsSesv2EmailIdentity extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSesv2EmailIdentity resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSesv2EmailIdentity to import
    * @param importFromId The id of the existing DataAwsSesv2EmailIdentity that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSesv2EmailIdentity to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sesv2_email_identity", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/sesv2_email_identity aws_sesv2_email_identity} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSesv2EmailIdentityConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sesv2_email_identity',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dkim_signing_attributes - computed: true, optional: false, required: false
        this._dkimSigningAttributes = new DataAwsSesv2EmailIdentityDkimSigningAttributesList(this, "dkim_signing_attributes", false);
        this._emailIdentity = config.emailIdentity;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // configuration_set_name - computed: true, optional: false, required: false
    get configurationSetName() {
        return this.getStringAttribute('configuration_set_name');
    }
    get dkimSigningAttributes() {
        return this._dkimSigningAttributes;
    }
    get emailIdentity() {
        return this.getStringAttribute('email_identity');
    }
    set emailIdentity(value) {
        this._emailIdentity = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailIdentityInput() {
        return this._emailIdentity;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // identity_type - computed: true, optional: false, required: false
    get identityType() {
        return this.getStringAttribute('identity_type');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // verified_for_sending_status - computed: true, optional: false, required: false
    get verifiedForSendingStatus() {
        return this.getBooleanAttribute('verified_for_sending_status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            email_identity: cdktf.stringToTerraform(this._emailIdentity),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            email_identity: {
                value: cdktf.stringToHclTerraform(this._emailIdentity),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSesv2EmailIdentity = DataAwsSesv2EmailIdentity;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSesv2EmailIdentity[_c] = { fqn: "@cdktf/provider-aws.dataAwsSesv2EmailIdentity.DataAwsSesv2EmailIdentity", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSesv2EmailIdentity.tfResourceType = "aws_sesv2_email_identity";
