// @cdktf/provider-aws/lib/data-aws-emrcontainers-virtual-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEmrcontainersVirtualCluster = exports.DataAwsEmrcontainersVirtualClusterContainerProviderList = exports.DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference = exports.dataAwsEmrcontainersVirtualClusterContainerProviderToHclTerraform = exports.dataAwsEmrcontainersVirtualClusterContainerProviderToTerraform = exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoList = exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference = exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoToHclTerraform = exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoToTerraform = exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList = exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference = exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToHclTerraform = exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform;
function dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToHclTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoToHclTerraform;
class DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference = DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference", version: "19.15.0" };
class DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList = DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList", version: "19.15.0" };
function dataAwsEmrcontainersVirtualClusterContainerProviderInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoToTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderInfoToTerraform;
function dataAwsEmrcontainersVirtualClusterContainerProviderInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderInfoToHclTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderInfoToHclTerraform;
class DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // eks_info - computed: true, optional: false, required: false
        this._eksInfo = new DataAwsEmrcontainersVirtualClusterContainerProviderInfoEksInfoList(this, "eks_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get eksInfo() {
        return this._eksInfo;
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference = DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference", version: "19.15.0" };
class DataAwsEmrcontainersVirtualClusterContainerProviderInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEmrcontainersVirtualClusterContainerProviderInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderInfoList = DataAwsEmrcontainersVirtualClusterContainerProviderInfoList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderInfoList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderInfoList", version: "19.15.0" };
function dataAwsEmrcontainersVirtualClusterContainerProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderToTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderToTerraform;
function dataAwsEmrcontainersVirtualClusterContainerProviderToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEmrcontainersVirtualClusterContainerProviderToHclTerraform = dataAwsEmrcontainersVirtualClusterContainerProviderToHclTerraform;
class DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // info - computed: true, optional: false, required: false
        this._info = new DataAwsEmrcontainersVirtualClusterContainerProviderInfoList(this, "info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get info() {
        return this._info;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference = DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference", version: "19.15.0" };
class DataAwsEmrcontainersVirtualClusterContainerProviderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEmrcontainersVirtualClusterContainerProviderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEmrcontainersVirtualClusterContainerProviderList = DataAwsEmrcontainersVirtualClusterContainerProviderList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualClusterContainerProviderList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualClusterContainerProviderList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emrcontainers_virtual_cluster aws_emrcontainers_virtual_cluster}
*/
class DataAwsEmrcontainersVirtualCluster extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEmrcontainersVirtualCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEmrcontainersVirtualCluster to import
    * @param importFromId The id of the existing DataAwsEmrcontainersVirtualCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emrcontainers_virtual_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEmrcontainersVirtualCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_emrcontainers_virtual_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/emrcontainers_virtual_cluster aws_emrcontainers_virtual_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEmrcontainersVirtualClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emrcontainers_virtual_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // container_provider - computed: true, optional: false, required: false
        this._containerProvider = new DataAwsEmrcontainersVirtualClusterContainerProviderList(this, "container_provider", false);
        this._id = config.id;
        this._tags = config.tags;
        this._virtualClusterId = config.virtualClusterId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get containerProvider() {
        return this._containerProvider;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get virtualClusterId() {
        return this.getStringAttribute('virtual_cluster_id');
    }
    set virtualClusterId(value) {
        this._virtualClusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualClusterIdInput() {
        return this._virtualClusterId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            virtual_cluster_id: cdktf.stringToTerraform(this._virtualClusterId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            virtual_cluster_id: {
                value: cdktf.stringToHclTerraform(this._virtualClusterId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEmrcontainersVirtualCluster = DataAwsEmrcontainersVirtualCluster;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEmrcontainersVirtualCluster[_g] = { fqn: "@cdktf/provider-aws.dataAwsEmrcontainersVirtualCluster.DataAwsEmrcontainersVirtualCluster", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEmrcontainersVirtualCluster.tfResourceType = "aws_emrcontainers_virtual_cluster";
