// cdktf/lib/backends/pg-backend.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTerraformRemoteStatePg = exports.PgBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_backend_1 = require("../terraform-backend");
const util_1 = require("../util");
const terraform_remote_state_1 = require("../terraform-remote-state");
// eslint-disable-next-line jsdoc/require-jsdoc
class PgBackend extends terraform_backend_1.TerraformBackend {
    constructor(scope, props) {
        super(scope, "backend", "pg");
        this.props = props;
    }
    synthesizeAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    synthesizeHclAttributes() {
        return (0, util_1.keysToSnakeCase)({ ...this.props });
    }
    getRemoteStateDataSource(scope, name, _fromStack) {
        return new DataTerraformRemoteStatePg(scope, name, {
            ...this.props,
            workspace: "${terraform.workspace}",
        });
    }
}
exports.PgBackend = PgBackend;
_a = JSII_RTTI_SYMBOL_1;
PgBackend[_a] = { fqn: "cdktf.PgBackend", version: "0.20.3" };
// eslint-disable-next-line jsdoc/require-jsdoc
class DataTerraformRemoteStatePg extends terraform_remote_state_1.TerraformRemoteState {
    constructor(scope, id, config) {
        super(scope, id, "pg", config);
    }
}
exports.DataTerraformRemoteStatePg = DataTerraformRemoteStatePg;
_b = JSII_RTTI_SYMBOL_1;
DataTerraformRemoteStatePg[_b] = { fqn: "cdktf.DataTerraformRemoteStatePg", version: "0.20.3" };
