// @cdktf/provider-aws/lib/iot-ca-certificate/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotCaCertificate = exports.IotCaCertificateRegistrationConfigOutputReference = exports.iotCaCertificateRegistrationConfigToHclTerraform = exports.iotCaCertificateRegistrationConfigToTerraform = exports.IotCaCertificateValidityList = exports.IotCaCertificateValidityOutputReference = exports.iotCaCertificateValidityToHclTerraform = exports.iotCaCertificateValidityToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iotCaCertificateValidityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.iotCaCertificateValidityToTerraform = iotCaCertificateValidityToTerraform;
function iotCaCertificateValidityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.iotCaCertificateValidityToHclTerraform = iotCaCertificateValidityToHclTerraform;
class IotCaCertificateValidityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // not_after - computed: true, optional: false, required: false
    get notAfter() {
        return this.getStringAttribute('not_after');
    }
    // not_before - computed: true, optional: false, required: false
    get notBefore() {
        return this.getStringAttribute('not_before');
    }
}
exports.IotCaCertificateValidityOutputReference = IotCaCertificateValidityOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IotCaCertificateValidityOutputReference[_a] = { fqn: "@cdktf/provider-aws.iotCaCertificate.IotCaCertificateValidityOutputReference", version: "19.15.0" };
class IotCaCertificateValidityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IotCaCertificateValidityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IotCaCertificateValidityList = IotCaCertificateValidityList;
_b = JSII_RTTI_SYMBOL_1;
IotCaCertificateValidityList[_b] = { fqn: "@cdktf/provider-aws.iotCaCertificate.IotCaCertificateValidityList", version: "19.15.0" };
function iotCaCertificateRegistrationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        template_body: cdktf.stringToTerraform(struct.templateBody),
        template_name: cdktf.stringToTerraform(struct.templateName),
    };
}
exports.iotCaCertificateRegistrationConfigToTerraform = iotCaCertificateRegistrationConfigToTerraform;
function iotCaCertificateRegistrationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        template_body: {
            value: cdktf.stringToHclTerraform(struct.templateBody),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        template_name: {
            value: cdktf.stringToHclTerraform(struct.templateName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iotCaCertificateRegistrationConfigToHclTerraform = iotCaCertificateRegistrationConfigToHclTerraform;
class IotCaCertificateRegistrationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._templateBody !== undefined) {
            hasAnyValues = true;
            internalValueResult.templateBody = this._templateBody;
        }
        if (this._templateName !== undefined) {
            hasAnyValues = true;
            internalValueResult.templateName = this._templateName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._roleArn = undefined;
            this._templateBody = undefined;
            this._templateName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._roleArn = value.roleArn;
            this._templateBody = value.templateBody;
            this._templateName = value.templateName;
        }
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get templateBody() {
        return this.getStringAttribute('template_body');
    }
    set templateBody(value) {
        this._templateBody = value;
    }
    resetTemplateBody() {
        this._templateBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateBodyInput() {
        return this._templateBody;
    }
    get templateName() {
        return this.getStringAttribute('template_name');
    }
    set templateName(value) {
        this._templateName = value;
    }
    resetTemplateName() {
        this._templateName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateNameInput() {
        return this._templateName;
    }
}
exports.IotCaCertificateRegistrationConfigOutputReference = IotCaCertificateRegistrationConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IotCaCertificateRegistrationConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.iotCaCertificate.IotCaCertificateRegistrationConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_ca_certificate aws_iot_ca_certificate}
*/
class IotCaCertificate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IotCaCertificate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IotCaCertificate to import
    * @param importFromId The id of the existing IotCaCertificate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_ca_certificate#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IotCaCertificate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_iot_ca_certificate", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/iot_ca_certificate aws_iot_ca_certificate} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotCaCertificateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_ca_certificate',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // validity - computed: true, optional: false, required: false
        this._validity = new IotCaCertificateValidityList(this, "validity", false);
        // registration_config - computed: false, optional: true, required: false
        this._registrationConfig = new IotCaCertificateRegistrationConfigOutputReference(this, "registration_config");
        this._active = config.active;
        this._allowAutoRegistration = config.allowAutoRegistration;
        this._caCertificatePem = config.caCertificatePem;
        this._certificateMode = config.certificateMode;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._verificationCertificatePem = config.verificationCertificatePem;
        this._registrationConfig.internalValue = config.registrationConfig;
    }
    get active() {
        return this.getBooleanAttribute('active');
    }
    set active(value) {
        this._active = value;
    }
    // Temporarily expose input value. Use with caution.
    get activeInput() {
        return this._active;
    }
    get allowAutoRegistration() {
        return this.getBooleanAttribute('allow_auto_registration');
    }
    set allowAutoRegistration(value) {
        this._allowAutoRegistration = value;
    }
    // Temporarily expose input value. Use with caution.
    get allowAutoRegistrationInput() {
        return this._allowAutoRegistration;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get caCertificatePem() {
        return this.getStringAttribute('ca_certificate_pem');
    }
    set caCertificatePem(value) {
        this._caCertificatePem = value;
    }
    // Temporarily expose input value. Use with caution.
    get caCertificatePemInput() {
        return this._caCertificatePem;
    }
    get certificateMode() {
        return this.getStringAttribute('certificate_mode');
    }
    set certificateMode(value) {
        this._certificateMode = value;
    }
    resetCertificateMode() {
        this._certificateMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificateModeInput() {
        return this._certificateMode;
    }
    // customer_version - computed: true, optional: false, required: false
    get customerVersion() {
        return this.getNumberAttribute('customer_version');
    }
    // generation_id - computed: true, optional: false, required: false
    get generationId() {
        return this.getStringAttribute('generation_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get validity() {
        return this._validity;
    }
    get verificationCertificatePem() {
        return this.getStringAttribute('verification_certificate_pem');
    }
    set verificationCertificatePem(value) {
        this._verificationCertificatePem = value;
    }
    resetVerificationCertificatePem() {
        this._verificationCertificatePem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get verificationCertificatePemInput() {
        return this._verificationCertificatePem;
    }
    get registrationConfig() {
        return this._registrationConfig;
    }
    putRegistrationConfig(value) {
        this._registrationConfig.internalValue = value;
    }
    resetRegistrationConfig() {
        this._registrationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get registrationConfigInput() {
        return this._registrationConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            active: cdktf.booleanToTerraform(this._active),
            allow_auto_registration: cdktf.booleanToTerraform(this._allowAutoRegistration),
            ca_certificate_pem: cdktf.stringToTerraform(this._caCertificatePem),
            certificate_mode: cdktf.stringToTerraform(this._certificateMode),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            verification_certificate_pem: cdktf.stringToTerraform(this._verificationCertificatePem),
            registration_config: iotCaCertificateRegistrationConfigToTerraform(this._registrationConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            active: {
                value: cdktf.booleanToHclTerraform(this._active),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            allow_auto_registration: {
                value: cdktf.booleanToHclTerraform(this._allowAutoRegistration),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            ca_certificate_pem: {
                value: cdktf.stringToHclTerraform(this._caCertificatePem),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            certificate_mode: {
                value: cdktf.stringToHclTerraform(this._certificateMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            verification_certificate_pem: {
                value: cdktf.stringToHclTerraform(this._verificationCertificatePem),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            registration_config: {
                value: iotCaCertificateRegistrationConfigToHclTerraform(this._registrationConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IotCaCertificateRegistrationConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IotCaCertificate = IotCaCertificate;
_d = JSII_RTTI_SYMBOL_1;
IotCaCertificate[_d] = { fqn: "@cdktf/provider-aws.iotCaCertificate.IotCaCertificate", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IotCaCertificate.tfResourceType = "aws_iot_ca_certificate";
