// @cdktf/provider-aws/lib/lightsail-instance/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailInstance = exports.LightsailInstanceAddOnOutputReference = exports.lightsailInstanceAddOnToHclTerraform = exports.lightsailInstanceAddOnToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lightsailInstanceAddOnToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        snapshot_time: cdktf.stringToTerraform(struct.snapshotTime),
        status: cdktf.stringToTerraform(struct.status),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.lightsailInstanceAddOnToTerraform = lightsailInstanceAddOnToTerraform;
function lightsailInstanceAddOnToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        snapshot_time: {
            value: cdktf.stringToHclTerraform(struct.snapshotTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        status: {
            value: cdktf.stringToHclTerraform(struct.status),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lightsailInstanceAddOnToHclTerraform = lightsailInstanceAddOnToHclTerraform;
class LightsailInstanceAddOnOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._snapshotTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotTime = this._snapshotTime;
        }
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._snapshotTime = undefined;
            this._status = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._snapshotTime = value.snapshotTime;
            this._status = value.status;
            this._type = value.type;
        }
    }
    get snapshotTime() {
        return this.getStringAttribute('snapshot_time');
    }
    set snapshotTime(value) {
        this._snapshotTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotTimeInput() {
        return this._snapshotTime;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.LightsailInstanceAddOnOutputReference = LightsailInstanceAddOnOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LightsailInstanceAddOnOutputReference[_a] = { fqn: "@cdktf/provider-aws.lightsailInstance.LightsailInstanceAddOnOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_instance aws_lightsail_instance}
*/
class LightsailInstance extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LightsailInstance resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LightsailInstance to import
    * @param importFromId The id of the existing LightsailInstance that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_instance#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LightsailInstance to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lightsail_instance", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lightsail_instance aws_lightsail_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LightsailInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // add_on - computed: false, optional: true, required: false
        this._addOn = new LightsailInstanceAddOnOutputReference(this, "add_on");
        this._availabilityZone = config.availabilityZone;
        this._blueprintId = config.blueprintId;
        this._bundleId = config.bundleId;
        this._id = config.id;
        this._ipAddressType = config.ipAddressType;
        this._keyPairName = config.keyPairName;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._userData = config.userData;
        this._addOn.internalValue = config.addOn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    get blueprintId() {
        return this.getStringAttribute('blueprint_id');
    }
    set blueprintId(value) {
        this._blueprintId = value;
    }
    // Temporarily expose input value. Use with caution.
    get blueprintIdInput() {
        return this._blueprintId;
    }
    get bundleId() {
        return this.getStringAttribute('bundle_id');
    }
    set bundleId(value) {
        this._bundleId = value;
    }
    // Temporarily expose input value. Use with caution.
    get bundleIdInput() {
        return this._bundleId;
    }
    // cpu_count - computed: true, optional: false, required: false
    get cpuCount() {
        return this.getNumberAttribute('cpu_count');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    set ipAddressType(value) {
        this._ipAddressType = value;
    }
    resetIpAddressType() {
        this._ipAddressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressTypeInput() {
        return this._ipAddressType;
    }
    // ipv6_addresses - computed: true, optional: false, required: false
    get ipv6Addresses() {
        return this.getListAttribute('ipv6_addresses');
    }
    // is_static_ip - computed: true, optional: false, required: false
    get isStaticIp() {
        return this.getBooleanAttribute('is_static_ip');
    }
    get keyPairName() {
        return this.getStringAttribute('key_pair_name');
    }
    set keyPairName(value) {
        this._keyPairName = value;
    }
    resetKeyPairName() {
        this._keyPairName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyPairNameInput() {
        return this._keyPairName;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // private_ip_address - computed: true, optional: false, required: false
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // public_ip_address - computed: true, optional: false, required: false
    get publicIpAddress() {
        return this.getStringAttribute('public_ip_address');
    }
    // ram_size - computed: true, optional: false, required: false
    get ramSize() {
        return this.getNumberAttribute('ram_size');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get userData() {
        return this.getStringAttribute('user_data');
    }
    set userData(value) {
        this._userData = value;
    }
    resetUserData() {
        this._userData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userDataInput() {
        return this._userData;
    }
    // username - computed: true, optional: false, required: false
    get username() {
        return this.getStringAttribute('username');
    }
    get addOn() {
        return this._addOn;
    }
    putAddOn(value) {
        this._addOn.internalValue = value;
    }
    resetAddOn() {
        this._addOn.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addOnInput() {
        return this._addOn.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            availability_zone: cdktf.stringToTerraform(this._availabilityZone),
            blueprint_id: cdktf.stringToTerraform(this._blueprintId),
            bundle_id: cdktf.stringToTerraform(this._bundleId),
            id: cdktf.stringToTerraform(this._id),
            ip_address_type: cdktf.stringToTerraform(this._ipAddressType),
            key_pair_name: cdktf.stringToTerraform(this._keyPairName),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            user_data: cdktf.stringToTerraform(this._userData),
            add_on: lightsailInstanceAddOnToTerraform(this._addOn.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            availability_zone: {
                value: cdktf.stringToHclTerraform(this._availabilityZone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            blueprint_id: {
                value: cdktf.stringToHclTerraform(this._blueprintId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bundle_id: {
                value: cdktf.stringToHclTerraform(this._bundleId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ip_address_type: {
                value: cdktf.stringToHclTerraform(this._ipAddressType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key_pair_name: {
                value: cdktf.stringToHclTerraform(this._keyPairName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            user_data: {
                value: cdktf.stringToHclTerraform(this._userData),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            add_on: {
                value: lightsailInstanceAddOnToHclTerraform(this._addOn.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LightsailInstanceAddOnList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LightsailInstance = LightsailInstance;
_b = JSII_RTTI_SYMBOL_1;
LightsailInstance[_b] = { fqn: "@cdktf/provider-aws.lightsailInstance.LightsailInstance", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LightsailInstance.tfResourceType = "aws_lightsail_instance";
