// @cdktf/provider-aws/lib/identitystore-user/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentitystoreUser = exports.IdentitystoreUserPhoneNumbersOutputReference = exports.identitystoreUserPhoneNumbersToHclTerraform = exports.identitystoreUserPhoneNumbersToTerraform = exports.IdentitystoreUserNameOutputReference = exports.identitystoreUserNameToHclTerraform = exports.identitystoreUserNameToTerraform = exports.IdentitystoreUserEmailsOutputReference = exports.identitystoreUserEmailsToHclTerraform = exports.identitystoreUserEmailsToTerraform = exports.IdentitystoreUserAddressesOutputReference = exports.identitystoreUserAddressesToHclTerraform = exports.identitystoreUserAddressesToTerraform = exports.IdentitystoreUserExternalIdsList = exports.IdentitystoreUserExternalIdsOutputReference = exports.identitystoreUserExternalIdsToHclTerraform = exports.identitystoreUserExternalIdsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function identitystoreUserExternalIdsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.identitystoreUserExternalIdsToTerraform = identitystoreUserExternalIdsToTerraform;
function identitystoreUserExternalIdsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.identitystoreUserExternalIdsToHclTerraform = identitystoreUserExternalIdsToHclTerraform;
class IdentitystoreUserExternalIdsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
}
exports.IdentitystoreUserExternalIdsOutputReference = IdentitystoreUserExternalIdsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IdentitystoreUserExternalIdsOutputReference[_a] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserExternalIdsOutputReference", version: "19.15.0" };
class IdentitystoreUserExternalIdsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IdentitystoreUserExternalIdsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IdentitystoreUserExternalIdsList = IdentitystoreUserExternalIdsList;
_b = JSII_RTTI_SYMBOL_1;
IdentitystoreUserExternalIdsList[_b] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserExternalIdsList", version: "19.15.0" };
function identitystoreUserAddressesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        country: cdktf.stringToTerraform(struct.country),
        formatted: cdktf.stringToTerraform(struct.formatted),
        locality: cdktf.stringToTerraform(struct.locality),
        postal_code: cdktf.stringToTerraform(struct.postalCode),
        primary: cdktf.booleanToTerraform(struct.primary),
        region: cdktf.stringToTerraform(struct.region),
        street_address: cdktf.stringToTerraform(struct.streetAddress),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.identitystoreUserAddressesToTerraform = identitystoreUserAddressesToTerraform;
function identitystoreUserAddressesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        country: {
            value: cdktf.stringToHclTerraform(struct.country),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        formatted: {
            value: cdktf.stringToHclTerraform(struct.formatted),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        locality: {
            value: cdktf.stringToHclTerraform(struct.locality),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        postal_code: {
            value: cdktf.stringToHclTerraform(struct.postalCode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        primary: {
            value: cdktf.booleanToHclTerraform(struct.primary),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        street_address: {
            value: cdktf.stringToHclTerraform(struct.streetAddress),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.identitystoreUserAddressesToHclTerraform = identitystoreUserAddressesToHclTerraform;
class IdentitystoreUserAddressesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._country !== undefined) {
            hasAnyValues = true;
            internalValueResult.country = this._country;
        }
        if (this._formatted !== undefined) {
            hasAnyValues = true;
            internalValueResult.formatted = this._formatted;
        }
        if (this._locality !== undefined) {
            hasAnyValues = true;
            internalValueResult.locality = this._locality;
        }
        if (this._postalCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.postalCode = this._postalCode;
        }
        if (this._primary !== undefined) {
            hasAnyValues = true;
            internalValueResult.primary = this._primary;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._streetAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.streetAddress = this._streetAddress;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._country = undefined;
            this._formatted = undefined;
            this._locality = undefined;
            this._postalCode = undefined;
            this._primary = undefined;
            this._region = undefined;
            this._streetAddress = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._country = value.country;
            this._formatted = value.formatted;
            this._locality = value.locality;
            this._postalCode = value.postalCode;
            this._primary = value.primary;
            this._region = value.region;
            this._streetAddress = value.streetAddress;
            this._type = value.type;
        }
    }
    get country() {
        return this.getStringAttribute('country');
    }
    set country(value) {
        this._country = value;
    }
    resetCountry() {
        this._country = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryInput() {
        return this._country;
    }
    get formatted() {
        return this.getStringAttribute('formatted');
    }
    set formatted(value) {
        this._formatted = value;
    }
    resetFormatted() {
        this._formatted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get formattedInput() {
        return this._formatted;
    }
    get locality() {
        return this.getStringAttribute('locality');
    }
    set locality(value) {
        this._locality = value;
    }
    resetLocality() {
        this._locality = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localityInput() {
        return this._locality;
    }
    get postalCode() {
        return this.getStringAttribute('postal_code');
    }
    set postalCode(value) {
        this._postalCode = value;
    }
    resetPostalCode() {
        this._postalCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postalCodeInput() {
        return this._postalCode;
    }
    get primary() {
        return this.getBooleanAttribute('primary');
    }
    set primary(value) {
        this._primary = value;
    }
    resetPrimary() {
        this._primary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryInput() {
        return this._primary;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get streetAddress() {
        return this.getStringAttribute('street_address');
    }
    set streetAddress(value) {
        this._streetAddress = value;
    }
    resetStreetAddress() {
        this._streetAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get streetAddressInput() {
        return this._streetAddress;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.IdentitystoreUserAddressesOutputReference = IdentitystoreUserAddressesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IdentitystoreUserAddressesOutputReference[_c] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserAddressesOutputReference", version: "19.15.0" };
function identitystoreUserEmailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        primary: cdktf.booleanToTerraform(struct.primary),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.identitystoreUserEmailsToTerraform = identitystoreUserEmailsToTerraform;
function identitystoreUserEmailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        primary: {
            value: cdktf.booleanToHclTerraform(struct.primary),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.identitystoreUserEmailsToHclTerraform = identitystoreUserEmailsToHclTerraform;
class IdentitystoreUserEmailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._primary !== undefined) {
            hasAnyValues = true;
            internalValueResult.primary = this._primary;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._primary = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._primary = value.primary;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get primary() {
        return this.getBooleanAttribute('primary');
    }
    set primary(value) {
        this._primary = value;
    }
    resetPrimary() {
        this._primary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryInput() {
        return this._primary;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.IdentitystoreUserEmailsOutputReference = IdentitystoreUserEmailsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IdentitystoreUserEmailsOutputReference[_d] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserEmailsOutputReference", version: "19.15.0" };
function identitystoreUserNameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        family_name: cdktf.stringToTerraform(struct.familyName),
        formatted: cdktf.stringToTerraform(struct.formatted),
        given_name: cdktf.stringToTerraform(struct.givenName),
        honorific_prefix: cdktf.stringToTerraform(struct.honorificPrefix),
        honorific_suffix: cdktf.stringToTerraform(struct.honorificSuffix),
        middle_name: cdktf.stringToTerraform(struct.middleName),
    };
}
exports.identitystoreUserNameToTerraform = identitystoreUserNameToTerraform;
function identitystoreUserNameToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        family_name: {
            value: cdktf.stringToHclTerraform(struct.familyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        formatted: {
            value: cdktf.stringToHclTerraform(struct.formatted),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        given_name: {
            value: cdktf.stringToHclTerraform(struct.givenName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        honorific_prefix: {
            value: cdktf.stringToHclTerraform(struct.honorificPrefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        honorific_suffix: {
            value: cdktf.stringToHclTerraform(struct.honorificSuffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        middle_name: {
            value: cdktf.stringToHclTerraform(struct.middleName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.identitystoreUserNameToHclTerraform = identitystoreUserNameToHclTerraform;
class IdentitystoreUserNameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._familyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.familyName = this._familyName;
        }
        if (this._formatted !== undefined) {
            hasAnyValues = true;
            internalValueResult.formatted = this._formatted;
        }
        if (this._givenName !== undefined) {
            hasAnyValues = true;
            internalValueResult.givenName = this._givenName;
        }
        if (this._honorificPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.honorificPrefix = this._honorificPrefix;
        }
        if (this._honorificSuffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.honorificSuffix = this._honorificSuffix;
        }
        if (this._middleName !== undefined) {
            hasAnyValues = true;
            internalValueResult.middleName = this._middleName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._familyName = undefined;
            this._formatted = undefined;
            this._givenName = undefined;
            this._honorificPrefix = undefined;
            this._honorificSuffix = undefined;
            this._middleName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._familyName = value.familyName;
            this._formatted = value.formatted;
            this._givenName = value.givenName;
            this._honorificPrefix = value.honorificPrefix;
            this._honorificSuffix = value.honorificSuffix;
            this._middleName = value.middleName;
        }
    }
    get familyName() {
        return this.getStringAttribute('family_name');
    }
    set familyName(value) {
        this._familyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get familyNameInput() {
        return this._familyName;
    }
    get formatted() {
        return this.getStringAttribute('formatted');
    }
    set formatted(value) {
        this._formatted = value;
    }
    resetFormatted() {
        this._formatted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get formattedInput() {
        return this._formatted;
    }
    get givenName() {
        return this.getStringAttribute('given_name');
    }
    set givenName(value) {
        this._givenName = value;
    }
    // Temporarily expose input value. Use with caution.
    get givenNameInput() {
        return this._givenName;
    }
    get honorificPrefix() {
        return this.getStringAttribute('honorific_prefix');
    }
    set honorificPrefix(value) {
        this._honorificPrefix = value;
    }
    resetHonorificPrefix() {
        this._honorificPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get honorificPrefixInput() {
        return this._honorificPrefix;
    }
    get honorificSuffix() {
        return this.getStringAttribute('honorific_suffix');
    }
    set honorificSuffix(value) {
        this._honorificSuffix = value;
    }
    resetHonorificSuffix() {
        this._honorificSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get honorificSuffixInput() {
        return this._honorificSuffix;
    }
    get middleName() {
        return this.getStringAttribute('middle_name');
    }
    set middleName(value) {
        this._middleName = value;
    }
    resetMiddleName() {
        this._middleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get middleNameInput() {
        return this._middleName;
    }
}
exports.IdentitystoreUserNameOutputReference = IdentitystoreUserNameOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IdentitystoreUserNameOutputReference[_e] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserNameOutputReference", version: "19.15.0" };
function identitystoreUserPhoneNumbersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        primary: cdktf.booleanToTerraform(struct.primary),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.identitystoreUserPhoneNumbersToTerraform = identitystoreUserPhoneNumbersToTerraform;
function identitystoreUserPhoneNumbersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        primary: {
            value: cdktf.booleanToHclTerraform(struct.primary),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.identitystoreUserPhoneNumbersToHclTerraform = identitystoreUserPhoneNumbersToHclTerraform;
class IdentitystoreUserPhoneNumbersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._primary !== undefined) {
            hasAnyValues = true;
            internalValueResult.primary = this._primary;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._primary = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._primary = value.primary;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get primary() {
        return this.getBooleanAttribute('primary');
    }
    set primary(value) {
        this._primary = value;
    }
    resetPrimary() {
        this._primary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryInput() {
        return this._primary;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.IdentitystoreUserPhoneNumbersOutputReference = IdentitystoreUserPhoneNumbersOutputReference;
_f = JSII_RTTI_SYMBOL_1;
IdentitystoreUserPhoneNumbersOutputReference[_f] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUserPhoneNumbersOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/identitystore_user aws_identitystore_user}
*/
class IdentitystoreUser extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IdentitystoreUser resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IdentitystoreUser to import
    * @param importFromId The id of the existing IdentitystoreUser that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/identitystore_user#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IdentitystoreUser to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_identitystore_user", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/identitystore_user aws_identitystore_user} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IdentitystoreUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_identitystore_user',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // external_ids - computed: true, optional: false, required: false
        this._externalIds = new IdentitystoreUserExternalIdsList(this, "external_ids", false);
        // addresses - computed: false, optional: true, required: false
        this._addresses = new IdentitystoreUserAddressesOutputReference(this, "addresses");
        // emails - computed: false, optional: true, required: false
        this._emails = new IdentitystoreUserEmailsOutputReference(this, "emails");
        // name - computed: false, optional: false, required: true
        this._name = new IdentitystoreUserNameOutputReference(this, "name");
        // phone_numbers - computed: false, optional: true, required: false
        this._phoneNumbers = new IdentitystoreUserPhoneNumbersOutputReference(this, "phone_numbers");
        this._displayName = config.displayName;
        this._id = config.id;
        this._identityStoreId = config.identityStoreId;
        this._locale = config.locale;
        this._nickname = config.nickname;
        this._preferredLanguage = config.preferredLanguage;
        this._profileUrl = config.profileUrl;
        this._timezone = config.timezone;
        this._title = config.title;
        this._userName = config.userName;
        this._userType = config.userType;
        this._addresses.internalValue = config.addresses;
        this._emails.internalValue = config.emails;
        this._name.internalValue = config.name;
        this._phoneNumbers.internalValue = config.phoneNumbers;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get externalIds() {
        return this._externalIds;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityStoreId() {
        return this.getStringAttribute('identity_store_id');
    }
    set identityStoreId(value) {
        this._identityStoreId = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityStoreIdInput() {
        return this._identityStoreId;
    }
    get locale() {
        return this.getStringAttribute('locale');
    }
    set locale(value) {
        this._locale = value;
    }
    resetLocale() {
        this._locale = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localeInput() {
        return this._locale;
    }
    get nickname() {
        return this.getStringAttribute('nickname');
    }
    set nickname(value) {
        this._nickname = value;
    }
    resetNickname() {
        this._nickname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nicknameInput() {
        return this._nickname;
    }
    get preferredLanguage() {
        return this.getStringAttribute('preferred_language');
    }
    set preferredLanguage(value) {
        this._preferredLanguage = value;
    }
    resetPreferredLanguage() {
        this._preferredLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preferredLanguageInput() {
        return this._preferredLanguage;
    }
    get profileUrl() {
        return this.getStringAttribute('profile_url');
    }
    set profileUrl(value) {
        this._profileUrl = value;
    }
    resetProfileUrl() {
        this._profileUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get profileUrlInput() {
        return this._profileUrl;
    }
    get timezone() {
        return this.getStringAttribute('timezone');
    }
    set timezone(value) {
        this._timezone = value;
    }
    resetTimezone() {
        this._timezone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timezoneInput() {
        return this._timezone;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
    // user_id - computed: true, optional: false, required: false
    get userId() {
        return this.getStringAttribute('user_id');
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    get userType() {
        return this.getStringAttribute('user_type');
    }
    set userType(value) {
        this._userType = value;
    }
    resetUserType() {
        this._userType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userTypeInput() {
        return this._userType;
    }
    get addresses() {
        return this._addresses;
    }
    putAddresses(value) {
        this._addresses.internalValue = value;
    }
    resetAddresses() {
        this._addresses.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressesInput() {
        return this._addresses.internalValue;
    }
    get emails() {
        return this._emails;
    }
    putEmails(value) {
        this._emails.internalValue = value;
    }
    resetEmails() {
        this._emails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailsInput() {
        return this._emails.internalValue;
    }
    get name() {
        return this._name;
    }
    putName(value) {
        this._name.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name.internalValue;
    }
    get phoneNumbers() {
        return this._phoneNumbers;
    }
    putPhoneNumbers(value) {
        this._phoneNumbers.internalValue = value;
    }
    resetPhoneNumbers() {
        this._phoneNumbers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumbersInput() {
        return this._phoneNumbers.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            identity_store_id: cdktf.stringToTerraform(this._identityStoreId),
            locale: cdktf.stringToTerraform(this._locale),
            nickname: cdktf.stringToTerraform(this._nickname),
            preferred_language: cdktf.stringToTerraform(this._preferredLanguage),
            profile_url: cdktf.stringToTerraform(this._profileUrl),
            timezone: cdktf.stringToTerraform(this._timezone),
            title: cdktf.stringToTerraform(this._title),
            user_name: cdktf.stringToTerraform(this._userName),
            user_type: cdktf.stringToTerraform(this._userType),
            addresses: identitystoreUserAddressesToTerraform(this._addresses.internalValue),
            emails: identitystoreUserEmailsToTerraform(this._emails.internalValue),
            name: identitystoreUserNameToTerraform(this._name.internalValue),
            phone_numbers: identitystoreUserPhoneNumbersToTerraform(this._phoneNumbers.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identity_store_id: {
                value: cdktf.stringToHclTerraform(this._identityStoreId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            locale: {
                value: cdktf.stringToHclTerraform(this._locale),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            nickname: {
                value: cdktf.stringToHclTerraform(this._nickname),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preferred_language: {
                value: cdktf.stringToHclTerraform(this._preferredLanguage),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            profile_url: {
                value: cdktf.stringToHclTerraform(this._profileUrl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timezone: {
                value: cdktf.stringToHclTerraform(this._timezone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            title: {
                value: cdktf.stringToHclTerraform(this._title),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_name: {
                value: cdktf.stringToHclTerraform(this._userName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_type: {
                value: cdktf.stringToHclTerraform(this._userType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            addresses: {
                value: identitystoreUserAddressesToHclTerraform(this._addresses.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IdentitystoreUserAddressesList",
            },
            emails: {
                value: identitystoreUserEmailsToHclTerraform(this._emails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IdentitystoreUserEmailsList",
            },
            name: {
                value: identitystoreUserNameToHclTerraform(this._name.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IdentitystoreUserNameList",
            },
            phone_numbers: {
                value: identitystoreUserPhoneNumbersToHclTerraform(this._phoneNumbers.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IdentitystoreUserPhoneNumbersList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IdentitystoreUser = IdentitystoreUser;
_g = JSII_RTTI_SYMBOL_1;
IdentitystoreUser[_g] = { fqn: "@cdktf/provider-aws.identitystoreUser.IdentitystoreUser", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
IdentitystoreUser.tfResourceType = "aws_identitystore_user";
