// @cdktf/provider-aws/lib/data-aws-fsx-windows-file-system/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsFsxWindowsFileSystem = exports.DataAwsFsxWindowsFileSystemDiskIopsConfigurationList = exports.DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference = exports.dataAwsFsxWindowsFileSystemDiskIopsConfigurationToHclTerraform = exports.dataAwsFsxWindowsFileSystemDiskIopsConfigurationToTerraform = exports.DataAwsFsxWindowsFileSystemAuditLogConfigurationList = exports.DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference = exports.dataAwsFsxWindowsFileSystemAuditLogConfigurationToHclTerraform = exports.dataAwsFsxWindowsFileSystemAuditLogConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsFsxWindowsFileSystemAuditLogConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxWindowsFileSystemAuditLogConfigurationToTerraform = dataAwsFsxWindowsFileSystemAuditLogConfigurationToTerraform;
function dataAwsFsxWindowsFileSystemAuditLogConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxWindowsFileSystemAuditLogConfigurationToHclTerraform = dataAwsFsxWindowsFileSystemAuditLogConfigurationToHclTerraform;
class DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // audit_log_destination - computed: true, optional: false, required: false
    get auditLogDestination() {
        return this.getStringAttribute('audit_log_destination');
    }
    // file_access_audit_log_level - computed: true, optional: false, required: false
    get fileAccessAuditLogLevel() {
        return this.getStringAttribute('file_access_audit_log_level');
    }
    // file_share_access_audit_log_level - computed: true, optional: false, required: false
    get fileShareAccessAuditLogLevel() {
        return this.getStringAttribute('file_share_access_audit_log_level');
    }
}
exports.DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference = DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsFsxWindowsFileSystem.DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference", version: "19.15.0" };
class DataAwsFsxWindowsFileSystemAuditLogConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxWindowsFileSystemAuditLogConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxWindowsFileSystemAuditLogConfigurationList = DataAwsFsxWindowsFileSystemAuditLogConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsFsxWindowsFileSystemAuditLogConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsFsxWindowsFileSystem.DataAwsFsxWindowsFileSystemAuditLogConfigurationList", version: "19.15.0" };
function dataAwsFsxWindowsFileSystemDiskIopsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsFsxWindowsFileSystemDiskIopsConfigurationToTerraform = dataAwsFsxWindowsFileSystemDiskIopsConfigurationToTerraform;
function dataAwsFsxWindowsFileSystemDiskIopsConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsFsxWindowsFileSystemDiskIopsConfigurationToHclTerraform = dataAwsFsxWindowsFileSystemDiskIopsConfigurationToHclTerraform;
class DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference = DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsFsxWindowsFileSystem.DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference", version: "19.15.0" };
class DataAwsFsxWindowsFileSystemDiskIopsConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsFsxWindowsFileSystemDiskIopsConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsFsxWindowsFileSystemDiskIopsConfigurationList = DataAwsFsxWindowsFileSystemDiskIopsConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsFsxWindowsFileSystemDiskIopsConfigurationList[_d] = { fqn: "@cdktf/provider-aws.dataAwsFsxWindowsFileSystem.DataAwsFsxWindowsFileSystemDiskIopsConfigurationList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_windows_file_system aws_fsx_windows_file_system}
*/
class DataAwsFsxWindowsFileSystem extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsFsxWindowsFileSystem resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsFsxWindowsFileSystem to import
    * @param importFromId The id of the existing DataAwsFsxWindowsFileSystem that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_windows_file_system#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsFsxWindowsFileSystem to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_fsx_windows_file_system", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/fsx_windows_file_system aws_fsx_windows_file_system} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsFsxWindowsFileSystemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_windows_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // audit_log_configuration - computed: true, optional: false, required: false
        this._auditLogConfiguration = new DataAwsFsxWindowsFileSystemAuditLogConfigurationList(this, "audit_log_configuration", false);
        // disk_iops_configuration - computed: true, optional: false, required: false
        this._diskIopsConfiguration = new DataAwsFsxWindowsFileSystemDiskIopsConfigurationList(this, "disk_iops_configuration", false);
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // active_directory_id - computed: true, optional: false, required: false
    get activeDirectoryId() {
        return this.getStringAttribute('active_directory_id');
    }
    // aliases - computed: true, optional: false, required: false
    get aliases() {
        return cdktf.Fn.tolist(this.getListAttribute('aliases'));
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get auditLogConfiguration() {
        return this._auditLogConfiguration;
    }
    // automatic_backup_retention_days - computed: true, optional: false, required: false
    get automaticBackupRetentionDays() {
        return this.getNumberAttribute('automatic_backup_retention_days');
    }
    // backup_id - computed: true, optional: false, required: false
    get backupId() {
        return this.getStringAttribute('backup_id');
    }
    // copy_tags_to_backups - computed: true, optional: false, required: false
    get copyTagsToBackups() {
        return this.getBooleanAttribute('copy_tags_to_backups');
    }
    // daily_automatic_backup_start_time - computed: true, optional: false, required: false
    get dailyAutomaticBackupStartTime() {
        return this.getStringAttribute('daily_automatic_backup_start_time');
    }
    // deployment_type - computed: true, optional: false, required: false
    get deploymentType() {
        return this.getStringAttribute('deployment_type');
    }
    get diskIopsConfiguration() {
        return this._diskIopsConfiguration;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // network_interface_ids - computed: true, optional: false, required: false
    get networkInterfaceIds() {
        return cdktf.Fn.tolist(this.getListAttribute('network_interface_ids'));
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // preferred_file_server_ip - computed: true, optional: false, required: false
    get preferredFileServerIp() {
        return this.getStringAttribute('preferred_file_server_ip');
    }
    // preferred_subnet_id - computed: true, optional: false, required: false
    get preferredSubnetId() {
        return this.getStringAttribute('preferred_subnet_id');
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // skip_final_backup - computed: true, optional: false, required: false
    get skipFinalBackup() {
        return this.getBooleanAttribute('skip_final_backup');
    }
    // storage_capacity - computed: true, optional: false, required: false
    get storageCapacity() {
        return this.getNumberAttribute('storage_capacity');
    }
    // storage_type - computed: true, optional: false, required: false
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // throughput_capacity - computed: true, optional: false, required: false
    get throughputCapacity() {
        return this.getNumberAttribute('throughput_capacity');
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // weekly_maintenance_start_time - computed: true, optional: false, required: false
    get weeklyMaintenanceStartTime() {
        return this.getStringAttribute('weekly_maintenance_start_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsFsxWindowsFileSystem = DataAwsFsxWindowsFileSystem;
_e = JSII_RTTI_SYMBOL_1;
DataAwsFsxWindowsFileSystem[_e] = { fqn: "@cdktf/provider-aws.dataAwsFsxWindowsFileSystem.DataAwsFsxWindowsFileSystem", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsFsxWindowsFileSystem.tfResourceType = "aws_fsx_windows_file_system";
