// @cdktf/provider-aws/lib/chimesdkvoice-sip-rule/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChimesdkvoiceSipRule = exports.ChimesdkvoiceSipRuleTargetApplicationsList = exports.ChimesdkvoiceSipRuleTargetApplicationsOutputReference = exports.chimesdkvoiceSipRuleTargetApplicationsToHclTerraform = exports.chimesdkvoiceSipRuleTargetApplicationsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function chimesdkvoiceSipRuleTargetApplicationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        aws_region: cdktf.stringToTerraform(struct.awsRegion),
        priority: cdktf.numberToTerraform(struct.priority),
        sip_media_application_id: cdktf.stringToTerraform(struct.sipMediaApplicationId),
    };
}
exports.chimesdkvoiceSipRuleTargetApplicationsToTerraform = chimesdkvoiceSipRuleTargetApplicationsToTerraform;
function chimesdkvoiceSipRuleTargetApplicationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        aws_region: {
            value: cdktf.stringToHclTerraform(struct.awsRegion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        priority: {
            value: cdktf.numberToHclTerraform(struct.priority),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        sip_media_application_id: {
            value: cdktf.stringToHclTerraform(struct.sipMediaApplicationId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.chimesdkvoiceSipRuleTargetApplicationsToHclTerraform = chimesdkvoiceSipRuleTargetApplicationsToHclTerraform;
class ChimesdkvoiceSipRuleTargetApplicationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._awsRegion !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsRegion = this._awsRegion;
        }
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._sipMediaApplicationId !== undefined) {
            hasAnyValues = true;
            internalValueResult.sipMediaApplicationId = this._sipMediaApplicationId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._awsRegion = undefined;
            this._priority = undefined;
            this._sipMediaApplicationId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._awsRegion = value.awsRegion;
            this._priority = value.priority;
            this._sipMediaApplicationId = value.sipMediaApplicationId;
        }
    }
    get awsRegion() {
        return this.getStringAttribute('aws_region');
    }
    set awsRegion(value) {
        this._awsRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    get awsRegionInput() {
        return this._awsRegion;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get sipMediaApplicationId() {
        return this.getStringAttribute('sip_media_application_id');
    }
    set sipMediaApplicationId(value) {
        this._sipMediaApplicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get sipMediaApplicationIdInput() {
        return this._sipMediaApplicationId;
    }
}
exports.ChimesdkvoiceSipRuleTargetApplicationsOutputReference = ChimesdkvoiceSipRuleTargetApplicationsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ChimesdkvoiceSipRuleTargetApplicationsOutputReference[_a] = { fqn: "@cdktf/provider-aws.chimesdkvoiceSipRule.ChimesdkvoiceSipRuleTargetApplicationsOutputReference", version: "19.15.0" };
class ChimesdkvoiceSipRuleTargetApplicationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ChimesdkvoiceSipRuleTargetApplicationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ChimesdkvoiceSipRuleTargetApplicationsList = ChimesdkvoiceSipRuleTargetApplicationsList;
_b = JSII_RTTI_SYMBOL_1;
ChimesdkvoiceSipRuleTargetApplicationsList[_b] = { fqn: "@cdktf/provider-aws.chimesdkvoiceSipRule.ChimesdkvoiceSipRuleTargetApplicationsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chimesdkvoice_sip_rule aws_chimesdkvoice_sip_rule}
*/
class ChimesdkvoiceSipRule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ChimesdkvoiceSipRule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ChimesdkvoiceSipRule to import
    * @param importFromId The id of the existing ChimesdkvoiceSipRule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chimesdkvoice_sip_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ChimesdkvoiceSipRule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_chimesdkvoice_sip_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/chimesdkvoice_sip_rule aws_chimesdkvoice_sip_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ChimesdkvoiceSipRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_chimesdkvoice_sip_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // target_applications - computed: false, optional: false, required: true
        this._targetApplications = new ChimesdkvoiceSipRuleTargetApplicationsList(this, "target_applications", true);
        this._disabled = config.disabled;
        this._id = config.id;
        this._name = config.name;
        this._triggerType = config.triggerType;
        this._triggerValue = config.triggerValue;
        this._targetApplications.internalValue = config.targetApplications;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get triggerType() {
        return this.getStringAttribute('trigger_type');
    }
    set triggerType(value) {
        this._triggerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get triggerTypeInput() {
        return this._triggerType;
    }
    get triggerValue() {
        return this.getStringAttribute('trigger_value');
    }
    set triggerValue(value) {
        this._triggerValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get triggerValueInput() {
        return this._triggerValue;
    }
    get targetApplications() {
        return this._targetApplications;
    }
    putTargetApplications(value) {
        this._targetApplications.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetApplicationsInput() {
        return this._targetApplications.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            disabled: cdktf.booleanToTerraform(this._disabled),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            trigger_type: cdktf.stringToTerraform(this._triggerType),
            trigger_value: cdktf.stringToTerraform(this._triggerValue),
            target_applications: cdktf.listMapper(chimesdkvoiceSipRuleTargetApplicationsToTerraform, true)(this._targetApplications.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            disabled: {
                value: cdktf.booleanToHclTerraform(this._disabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trigger_type: {
                value: cdktf.stringToHclTerraform(this._triggerType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trigger_value: {
                value: cdktf.stringToHclTerraform(this._triggerValue),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_applications: {
                value: cdktf.listMapperHcl(chimesdkvoiceSipRuleTargetApplicationsToHclTerraform, true)(this._targetApplications.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ChimesdkvoiceSipRuleTargetApplicationsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ChimesdkvoiceSipRule = ChimesdkvoiceSipRule;
_c = JSII_RTTI_SYMBOL_1;
ChimesdkvoiceSipRule[_c] = { fqn: "@cdktf/provider-aws.chimesdkvoiceSipRule.ChimesdkvoiceSipRule", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ChimesdkvoiceSipRule.tfResourceType = "aws_chimesdkvoice_sip_rule";
