// @cdktf/provider-aws/lib/guardduty-filter/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuarddutyFilter = exports.GuarddutyFilterFindingCriteriaOutputReference = exports.guarddutyFilterFindingCriteriaToHclTerraform = exports.guarddutyFilterFindingCriteriaToTerraform = exports.GuarddutyFilterFindingCriteriaCriterionList = exports.GuarddutyFilterFindingCriteriaCriterionOutputReference = exports.guarddutyFilterFindingCriteriaCriterionToHclTerraform = exports.guarddutyFilterFindingCriteriaCriterionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function guarddutyFilterFindingCriteriaCriterionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        equals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.equalTo),
        field: cdktf.stringToTerraform(struct.field),
        greater_than: cdktf.stringToTerraform(struct.greaterThan),
        greater_than_or_equal: cdktf.stringToTerraform(struct.greaterThanOrEqual),
        less_than: cdktf.stringToTerraform(struct.lessThan),
        less_than_or_equal: cdktf.stringToTerraform(struct.lessThanOrEqual),
        not_equals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.notEquals),
    };
}
exports.guarddutyFilterFindingCriteriaCriterionToTerraform = guarddutyFilterFindingCriteriaCriterionToTerraform;
function guarddutyFilterFindingCriteriaCriterionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        equals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.equalTo),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        greater_than: {
            value: cdktf.stringToHclTerraform(struct.greaterThan),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        greater_than_or_equal: {
            value: cdktf.stringToHclTerraform(struct.greaterThanOrEqual),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        less_than: {
            value: cdktf.stringToHclTerraform(struct.lessThan),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        less_than_or_equal: {
            value: cdktf.stringToHclTerraform(struct.lessThanOrEqual),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        not_equals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.notEquals),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.guarddutyFilterFindingCriteriaCriterionToHclTerraform = guarddutyFilterFindingCriteriaCriterionToHclTerraform;
class GuarddutyFilterFindingCriteriaCriterionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._equals !== undefined) {
            hasAnyValues = true;
            internalValueResult.equalTo = this._equals;
        }
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._greaterThan !== undefined) {
            hasAnyValues = true;
            internalValueResult.greaterThan = this._greaterThan;
        }
        if (this._greaterThanOrEqual !== undefined) {
            hasAnyValues = true;
            internalValueResult.greaterThanOrEqual = this._greaterThanOrEqual;
        }
        if (this._lessThan !== undefined) {
            hasAnyValues = true;
            internalValueResult.lessThan = this._lessThan;
        }
        if (this._lessThanOrEqual !== undefined) {
            hasAnyValues = true;
            internalValueResult.lessThanOrEqual = this._lessThanOrEqual;
        }
        if (this._notEquals !== undefined) {
            hasAnyValues = true;
            internalValueResult.notEquals = this._notEquals;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._equals = undefined;
            this._field = undefined;
            this._greaterThan = undefined;
            this._greaterThanOrEqual = undefined;
            this._lessThan = undefined;
            this._lessThanOrEqual = undefined;
            this._notEquals = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._equals = value.equalTo;
            this._field = value.field;
            this._greaterThan = value.greaterThan;
            this._greaterThanOrEqual = value.greaterThanOrEqual;
            this._lessThan = value.lessThan;
            this._lessThanOrEqual = value.lessThanOrEqual;
            this._notEquals = value.notEquals;
        }
    }
    get equalTo() {
        return this.getListAttribute('equals');
    }
    set equalTo(value) {
        this._equals = value;
    }
    resetEqualTo() {
        this._equals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get equalToInput() {
        return this._equals;
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get greaterThan() {
        return this.getStringAttribute('greater_than');
    }
    set greaterThan(value) {
        this._greaterThan = value;
    }
    resetGreaterThan() {
        this._greaterThan = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get greaterThanInput() {
        return this._greaterThan;
    }
    get greaterThanOrEqual() {
        return this.getStringAttribute('greater_than_or_equal');
    }
    set greaterThanOrEqual(value) {
        this._greaterThanOrEqual = value;
    }
    resetGreaterThanOrEqual() {
        this._greaterThanOrEqual = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get greaterThanOrEqualInput() {
        return this._greaterThanOrEqual;
    }
    get lessThan() {
        return this.getStringAttribute('less_than');
    }
    set lessThan(value) {
        this._lessThan = value;
    }
    resetLessThan() {
        this._lessThan = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lessThanInput() {
        return this._lessThan;
    }
    get lessThanOrEqual() {
        return this.getStringAttribute('less_than_or_equal');
    }
    set lessThanOrEqual(value) {
        this._lessThanOrEqual = value;
    }
    resetLessThanOrEqual() {
        this._lessThanOrEqual = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lessThanOrEqualInput() {
        return this._lessThanOrEqual;
    }
    get notEquals() {
        return this.getListAttribute('not_equals');
    }
    set notEquals(value) {
        this._notEquals = value;
    }
    resetNotEquals() {
        this._notEquals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notEqualsInput() {
        return this._notEquals;
    }
}
exports.GuarddutyFilterFindingCriteriaCriterionOutputReference = GuarddutyFilterFindingCriteriaCriterionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GuarddutyFilterFindingCriteriaCriterionOutputReference[_a] = { fqn: "@cdktf/provider-aws.guarddutyFilter.GuarddutyFilterFindingCriteriaCriterionOutputReference", version: "19.15.0" };
class GuarddutyFilterFindingCriteriaCriterionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GuarddutyFilterFindingCriteriaCriterionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GuarddutyFilterFindingCriteriaCriterionList = GuarddutyFilterFindingCriteriaCriterionList;
_b = JSII_RTTI_SYMBOL_1;
GuarddutyFilterFindingCriteriaCriterionList[_b] = { fqn: "@cdktf/provider-aws.guarddutyFilter.GuarddutyFilterFindingCriteriaCriterionList", version: "19.15.0" };
function guarddutyFilterFindingCriteriaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        criterion: cdktf.listMapper(guarddutyFilterFindingCriteriaCriterionToTerraform, true)(struct.criterion),
    };
}
exports.guarddutyFilterFindingCriteriaToTerraform = guarddutyFilterFindingCriteriaToTerraform;
function guarddutyFilterFindingCriteriaToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        criterion: {
            value: cdktf.listMapperHcl(guarddutyFilterFindingCriteriaCriterionToHclTerraform, true)(struct.criterion),
            isBlock: true,
            type: "set",
            storageClassType: "GuarddutyFilterFindingCriteriaCriterionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.guarddutyFilterFindingCriteriaToHclTerraform = guarddutyFilterFindingCriteriaToHclTerraform;
class GuarddutyFilterFindingCriteriaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // criterion - computed: false, optional: false, required: true
        this._criterion = new GuarddutyFilterFindingCriteriaCriterionList(this, "criterion", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._criterion?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.criterion = this._criterion?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._criterion.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._criterion.internalValue = value.criterion;
        }
    }
    get criterion() {
        return this._criterion;
    }
    putCriterion(value) {
        this._criterion.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get criterionInput() {
        return this._criterion.internalValue;
    }
}
exports.GuarddutyFilterFindingCriteriaOutputReference = GuarddutyFilterFindingCriteriaOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GuarddutyFilterFindingCriteriaOutputReference[_c] = { fqn: "@cdktf/provider-aws.guarddutyFilter.GuarddutyFilterFindingCriteriaOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_filter aws_guardduty_filter}
*/
class GuarddutyFilter extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GuarddutyFilter resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GuarddutyFilter to import
    * @param importFromId The id of the existing GuarddutyFilter that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_filter#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GuarddutyFilter to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_guardduty_filter", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_filter aws_guardduty_filter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GuarddutyFilterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_filter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // finding_criteria - computed: false, optional: false, required: true
        this._findingCriteria = new GuarddutyFilterFindingCriteriaOutputReference(this, "finding_criteria");
        this._action = config.action;
        this._description = config.description;
        this._detectorId = config.detectorId;
        this._id = config.id;
        this._name = config.name;
        this._rank = config.rank;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._findingCriteria.internalValue = config.findingCriteria;
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get detectorId() {
        return this.getStringAttribute('detector_id');
    }
    set detectorId(value) {
        this._detectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    get detectorIdInput() {
        return this._detectorId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get rank() {
        return this.getNumberAttribute('rank');
    }
    set rank(value) {
        this._rank = value;
    }
    // Temporarily expose input value. Use with caution.
    get rankInput() {
        return this._rank;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get findingCriteria() {
        return this._findingCriteria;
    }
    putFindingCriteria(value) {
        this._findingCriteria.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get findingCriteriaInput() {
        return this._findingCriteria.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action: cdktf.stringToTerraform(this._action),
            description: cdktf.stringToTerraform(this._description),
            detector_id: cdktf.stringToTerraform(this._detectorId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            rank: cdktf.numberToTerraform(this._rank),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            finding_criteria: guarddutyFilterFindingCriteriaToTerraform(this._findingCriteria.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action: {
                value: cdktf.stringToHclTerraform(this._action),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            detector_id: {
                value: cdktf.stringToHclTerraform(this._detectorId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rank: {
                value: cdktf.numberToHclTerraform(this._rank),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            finding_criteria: {
                value: guarddutyFilterFindingCriteriaToHclTerraform(this._findingCriteria.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GuarddutyFilterFindingCriteriaList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GuarddutyFilter = GuarddutyFilter;
_d = JSII_RTTI_SYMBOL_1;
GuarddutyFilter[_d] = { fqn: "@cdktf/provider-aws.guarddutyFilter.GuarddutyFilter", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GuarddutyFilter.tfResourceType = "aws_guardduty_filter";
