// @cdktf/provider-aws/lib/verifiedpermissions-policy-store/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedpermissionsPolicyStore = exports.VerifiedpermissionsPolicyStoreValidationSettingsList = exports.VerifiedpermissionsPolicyStoreValidationSettingsOutputReference = exports.verifiedpermissionsPolicyStoreValidationSettingsToHclTerraform = exports.verifiedpermissionsPolicyStoreValidationSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function verifiedpermissionsPolicyStoreValidationSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.verifiedpermissionsPolicyStoreValidationSettingsToTerraform = verifiedpermissionsPolicyStoreValidationSettingsToTerraform;
function verifiedpermissionsPolicyStoreValidationSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedpermissionsPolicyStoreValidationSettingsToHclTerraform = verifiedpermissionsPolicyStoreValidationSettingsToHclTerraform;
class VerifiedpermissionsPolicyStoreValidationSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mode = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.VerifiedpermissionsPolicyStoreValidationSettingsOutputReference = VerifiedpermissionsPolicyStoreValidationSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VerifiedpermissionsPolicyStoreValidationSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.verifiedpermissionsPolicyStore.VerifiedpermissionsPolicyStoreValidationSettingsOutputReference", version: "19.15.0" };
class VerifiedpermissionsPolicyStoreValidationSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VerifiedpermissionsPolicyStoreValidationSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VerifiedpermissionsPolicyStoreValidationSettingsList = VerifiedpermissionsPolicyStoreValidationSettingsList;
_b = JSII_RTTI_SYMBOL_1;
VerifiedpermissionsPolicyStoreValidationSettingsList[_b] = { fqn: "@cdktf/provider-aws.verifiedpermissionsPolicyStore.VerifiedpermissionsPolicyStoreValidationSettingsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_policy_store aws_verifiedpermissions_policy_store}
*/
class VerifiedpermissionsPolicyStore extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedpermissionsPolicyStore resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedpermissionsPolicyStore to import
    * @param importFromId The id of the existing VerifiedpermissionsPolicyStore that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_policy_store#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedpermissionsPolicyStore to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedpermissions_policy_store", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedpermissions_policy_store aws_verifiedpermissions_policy_store} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedpermissionsPolicyStoreConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedpermissions_policy_store',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // validation_settings - computed: false, optional: true, required: false
        this._validationSettings = new VerifiedpermissionsPolicyStoreValidationSettingsList(this, "validation_settings", false);
        this._description = config.description;
        this._validationSettings.internalValue = config.validationSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // policy_store_id - computed: true, optional: false, required: false
    get policyStoreId() {
        return this.getStringAttribute('policy_store_id');
    }
    get validationSettings() {
        return this._validationSettings;
    }
    putValidationSettings(value) {
        this._validationSettings.internalValue = value;
    }
    resetValidationSettings() {
        this._validationSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validationSettingsInput() {
        return this._validationSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            validation_settings: cdktf.listMapper(verifiedpermissionsPolicyStoreValidationSettingsToTerraform, true)(this._validationSettings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            validation_settings: {
                value: cdktf.listMapperHcl(verifiedpermissionsPolicyStoreValidationSettingsToHclTerraform, true)(this._validationSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedpermissionsPolicyStoreValidationSettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedpermissionsPolicyStore = VerifiedpermissionsPolicyStore;
_c = JSII_RTTI_SYMBOL_1;
VerifiedpermissionsPolicyStore[_c] = { fqn: "@cdktf/provider-aws.verifiedpermissionsPolicyStore.VerifiedpermissionsPolicyStore", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedpermissionsPolicyStore.tfResourceType = "aws_verifiedpermissions_policy_store";
