// @cdktf/provider-aws/lib/data-aws-vpclattice-listener/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpclatticeListener = exports.DataAwsVpclatticeListenerDefaultActionList = exports.DataAwsVpclatticeListenerDefaultActionOutputReference = exports.dataAwsVpclatticeListenerDefaultActionToHclTerraform = exports.dataAwsVpclatticeListenerDefaultActionToTerraform = exports.DataAwsVpclatticeListenerDefaultActionForwardList = exports.DataAwsVpclatticeListenerDefaultActionForwardOutputReference = exports.dataAwsVpclatticeListenerDefaultActionForwardToHclTerraform = exports.dataAwsVpclatticeListenerDefaultActionForwardToTerraform = exports.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList = exports.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference = exports.dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToHclTerraform = exports.dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToTerraform = exports.DataAwsVpclatticeListenerDefaultActionFixedResponseList = exports.DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference = exports.dataAwsVpclatticeListenerDefaultActionFixedResponseToHclTerraform = exports.dataAwsVpclatticeListenerDefaultActionFixedResponseToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsVpclatticeListenerDefaultActionFixedResponseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpclatticeListenerDefaultActionFixedResponseToTerraform = dataAwsVpclatticeListenerDefaultActionFixedResponseToTerraform;
function dataAwsVpclatticeListenerDefaultActionFixedResponseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpclatticeListenerDefaultActionFixedResponseToHclTerraform = dataAwsVpclatticeListenerDefaultActionFixedResponseToHclTerraform;
class DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status_code - computed: true, optional: false, required: false
    get statusCode() {
        return this.getNumberAttribute('status_code');
    }
}
exports.DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference = DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference", version: "19.15.0" };
class DataAwsVpclatticeListenerDefaultActionFixedResponseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpclatticeListenerDefaultActionFixedResponseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpclatticeListenerDefaultActionFixedResponseList = DataAwsVpclatticeListenerDefaultActionFixedResponseList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionFixedResponseList[_b] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionFixedResponseList", version: "19.15.0" };
function dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToTerraform = dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToTerraform;
function dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToHclTerraform = dataAwsVpclatticeListenerDefaultActionForwardTargetGroupsToHclTerraform;
class DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // target_group_identifier - computed: true, optional: false, required: false
    get targetGroupIdentifier() {
        return this.getStringAttribute('target_group_identifier');
    }
    // weight - computed: true, optional: false, required: false
    get weight() {
        return this.getNumberAttribute('weight');
    }
}
exports.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference = DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference", version: "19.15.0" };
class DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList = DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList", version: "19.15.0" };
function dataAwsVpclatticeListenerDefaultActionForwardToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpclatticeListenerDefaultActionForwardToTerraform = dataAwsVpclatticeListenerDefaultActionForwardToTerraform;
function dataAwsVpclatticeListenerDefaultActionForwardToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpclatticeListenerDefaultActionForwardToHclTerraform = dataAwsVpclatticeListenerDefaultActionForwardToHclTerraform;
class DataAwsVpclatticeListenerDefaultActionForwardOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_groups - computed: true, optional: false, required: false
        this._targetGroups = new DataAwsVpclatticeListenerDefaultActionForwardTargetGroupsList(this, "target_groups", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get targetGroups() {
        return this._targetGroups;
    }
}
exports.DataAwsVpclatticeListenerDefaultActionForwardOutputReference = DataAwsVpclatticeListenerDefaultActionForwardOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionForwardOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionForwardOutputReference", version: "19.15.0" };
class DataAwsVpclatticeListenerDefaultActionForwardList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpclatticeListenerDefaultActionForwardOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpclatticeListenerDefaultActionForwardList = DataAwsVpclatticeListenerDefaultActionForwardList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionForwardList[_f] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionForwardList", version: "19.15.0" };
function dataAwsVpclatticeListenerDefaultActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpclatticeListenerDefaultActionToTerraform = dataAwsVpclatticeListenerDefaultActionToTerraform;
function dataAwsVpclatticeListenerDefaultActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsVpclatticeListenerDefaultActionToHclTerraform = dataAwsVpclatticeListenerDefaultActionToHclTerraform;
class DataAwsVpclatticeListenerDefaultActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // fixed_response - computed: true, optional: false, required: false
        this._fixedResponse = new DataAwsVpclatticeListenerDefaultActionFixedResponseList(this, "fixed_response", false);
        // forward - computed: true, optional: false, required: false
        this._forward = new DataAwsVpclatticeListenerDefaultActionForwardList(this, "forward", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get fixedResponse() {
        return this._fixedResponse;
    }
    get forward() {
        return this._forward;
    }
}
exports.DataAwsVpclatticeListenerDefaultActionOutputReference = DataAwsVpclatticeListenerDefaultActionOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionOutputReference", version: "19.15.0" };
class DataAwsVpclatticeListenerDefaultActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpclatticeListenerDefaultActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpclatticeListenerDefaultActionList = DataAwsVpclatticeListenerDefaultActionList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListenerDefaultActionList[_h] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListenerDefaultActionList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpclattice_listener aws_vpclattice_listener}
*/
class DataAwsVpclatticeListener extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsVpclatticeListener resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsVpclatticeListener to import
    * @param importFromId The id of the existing DataAwsVpclatticeListener that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpclattice_listener#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsVpclatticeListener to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpclattice_listener", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/vpclattice_listener aws_vpclattice_listener} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsVpclatticeListenerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpclattice_listener',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // default_action - computed: true, optional: false, required: false
        this._defaultAction = new DataAwsVpclatticeListenerDefaultActionList(this, "default_action", false);
        this._id = config.id;
        this._listenerIdentifier = config.listenerIdentifier;
        this._serviceIdentifier = config.serviceIdentifier;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get defaultAction() {
        return this._defaultAction;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_at - computed: true, optional: false, required: false
    get lastUpdatedAt() {
        return this.getStringAttribute('last_updated_at');
    }
    // listener_id - computed: true, optional: false, required: false
    get listenerId() {
        return this.getStringAttribute('listener_id');
    }
    get listenerIdentifier() {
        return this.getStringAttribute('listener_identifier');
    }
    set listenerIdentifier(value) {
        this._listenerIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get listenerIdentifierInput() {
        return this._listenerIdentifier;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // service_arn - computed: true, optional: false, required: false
    get serviceArn() {
        return this.getStringAttribute('service_arn');
    }
    // service_id - computed: true, optional: false, required: false
    get serviceId() {
        return this.getStringAttribute('service_id');
    }
    get serviceIdentifier() {
        return this.getStringAttribute('service_identifier');
    }
    set serviceIdentifier(value) {
        this._serviceIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIdentifierInput() {
        return this._serviceIdentifier;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            listener_identifier: cdktf.stringToTerraform(this._listenerIdentifier),
            service_identifier: cdktf.stringToTerraform(this._serviceIdentifier),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            listener_identifier: {
                value: cdktf.stringToHclTerraform(this._listenerIdentifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_identifier: {
                value: cdktf.stringToHclTerraform(this._serviceIdentifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsVpclatticeListener = DataAwsVpclatticeListener;
_j = JSII_RTTI_SYMBOL_1;
DataAwsVpclatticeListener[_j] = { fqn: "@cdktf/provider-aws.dataAwsVpclatticeListener.DataAwsVpclatticeListener", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsVpclatticeListener.tfResourceType = "aws_vpclattice_listener";
