// @cdktf/provider-aws/lib/evidently-launch/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvidentlyLaunch = exports.EvidentlyLaunchTimeoutsOutputReference = exports.evidentlyLaunchTimeoutsToHclTerraform = exports.evidentlyLaunchTimeoutsToTerraform = exports.EvidentlyLaunchScheduledSplitsConfigOutputReference = exports.evidentlyLaunchScheduledSplitsConfigToHclTerraform = exports.evidentlyLaunchScheduledSplitsConfigToTerraform = exports.EvidentlyLaunchScheduledSplitsConfigStepsList = exports.EvidentlyLaunchScheduledSplitsConfigStepsOutputReference = exports.evidentlyLaunchScheduledSplitsConfigStepsToHclTerraform = exports.evidentlyLaunchScheduledSplitsConfigStepsToTerraform = exports.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList = exports.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference = exports.evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToHclTerraform = exports.evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToTerraform = exports.EvidentlyLaunchMetricMonitorsList = exports.EvidentlyLaunchMetricMonitorsOutputReference = exports.evidentlyLaunchMetricMonitorsToHclTerraform = exports.evidentlyLaunchMetricMonitorsToTerraform = exports.EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference = exports.evidentlyLaunchMetricMonitorsMetricDefinitionToHclTerraform = exports.evidentlyLaunchMetricMonitorsMetricDefinitionToTerraform = exports.EvidentlyLaunchGroupsList = exports.EvidentlyLaunchGroupsOutputReference = exports.evidentlyLaunchGroupsToHclTerraform = exports.evidentlyLaunchGroupsToTerraform = exports.EvidentlyLaunchExecutionList = exports.EvidentlyLaunchExecutionOutputReference = exports.evidentlyLaunchExecutionToHclTerraform = exports.evidentlyLaunchExecutionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function evidentlyLaunchExecutionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.evidentlyLaunchExecutionToTerraform = evidentlyLaunchExecutionToTerraform;
function evidentlyLaunchExecutionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.evidentlyLaunchExecutionToHclTerraform = evidentlyLaunchExecutionToHclTerraform;
class EvidentlyLaunchExecutionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ended_time - computed: true, optional: false, required: false
    get endedTime() {
        return this.getStringAttribute('ended_time');
    }
    // started_time - computed: true, optional: false, required: false
    get startedTime() {
        return this.getStringAttribute('started_time');
    }
}
exports.EvidentlyLaunchExecutionOutputReference = EvidentlyLaunchExecutionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchExecutionOutputReference[_a] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchExecutionOutputReference", version: "19.15.0" };
class EvidentlyLaunchExecutionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EvidentlyLaunchExecutionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EvidentlyLaunchExecutionList = EvidentlyLaunchExecutionList;
_b = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchExecutionList[_b] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchExecutionList", version: "19.15.0" };
function evidentlyLaunchGroupsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        feature: cdktf.stringToTerraform(struct.feature),
        name: cdktf.stringToTerraform(struct.name),
        variation: cdktf.stringToTerraform(struct.variation),
    };
}
exports.evidentlyLaunchGroupsToTerraform = evidentlyLaunchGroupsToTerraform;
function evidentlyLaunchGroupsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        feature: {
            value: cdktf.stringToHclTerraform(struct.feature),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        variation: {
            value: cdktf.stringToHclTerraform(struct.variation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchGroupsToHclTerraform = evidentlyLaunchGroupsToHclTerraform;
class EvidentlyLaunchGroupsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._feature !== undefined) {
            hasAnyValues = true;
            internalValueResult.feature = this._feature;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._variation !== undefined) {
            hasAnyValues = true;
            internalValueResult.variation = this._variation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._feature = undefined;
            this._name = undefined;
            this._variation = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._feature = value.feature;
            this._name = value.name;
            this._variation = value.variation;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get feature() {
        return this.getStringAttribute('feature');
    }
    set feature(value) {
        this._feature = value;
    }
    // Temporarily expose input value. Use with caution.
    get featureInput() {
        return this._feature;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get variation() {
        return this.getStringAttribute('variation');
    }
    set variation(value) {
        this._variation = value;
    }
    // Temporarily expose input value. Use with caution.
    get variationInput() {
        return this._variation;
    }
}
exports.EvidentlyLaunchGroupsOutputReference = EvidentlyLaunchGroupsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchGroupsOutputReference[_c] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchGroupsOutputReference", version: "19.15.0" };
class EvidentlyLaunchGroupsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EvidentlyLaunchGroupsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EvidentlyLaunchGroupsList = EvidentlyLaunchGroupsList;
_d = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchGroupsList[_d] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchGroupsList", version: "19.15.0" };
function evidentlyLaunchMetricMonitorsMetricDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        entity_id_key: cdktf.stringToTerraform(struct.entityIdKey),
        event_pattern: cdktf.stringToTerraform(struct.eventPattern),
        name: cdktf.stringToTerraform(struct.name),
        unit_label: cdktf.stringToTerraform(struct.unitLabel),
        value_key: cdktf.stringToTerraform(struct.valueKey),
    };
}
exports.evidentlyLaunchMetricMonitorsMetricDefinitionToTerraform = evidentlyLaunchMetricMonitorsMetricDefinitionToTerraform;
function evidentlyLaunchMetricMonitorsMetricDefinitionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        entity_id_key: {
            value: cdktf.stringToHclTerraform(struct.entityIdKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        event_pattern: {
            value: cdktf.stringToHclTerraform(struct.eventPattern),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        unit_label: {
            value: cdktf.stringToHclTerraform(struct.unitLabel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value_key: {
            value: cdktf.stringToHclTerraform(struct.valueKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchMetricMonitorsMetricDefinitionToHclTerraform = evidentlyLaunchMetricMonitorsMetricDefinitionToHclTerraform;
class EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._entityIdKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityIdKey = this._entityIdKey;
        }
        if (this._eventPattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventPattern = this._eventPattern;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._unitLabel !== undefined) {
            hasAnyValues = true;
            internalValueResult.unitLabel = this._unitLabel;
        }
        if (this._valueKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueKey = this._valueKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._entityIdKey = undefined;
            this._eventPattern = undefined;
            this._name = undefined;
            this._unitLabel = undefined;
            this._valueKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._entityIdKey = value.entityIdKey;
            this._eventPattern = value.eventPattern;
            this._name = value.name;
            this._unitLabel = value.unitLabel;
            this._valueKey = value.valueKey;
        }
    }
    get entityIdKey() {
        return this.getStringAttribute('entity_id_key');
    }
    set entityIdKey(value) {
        this._entityIdKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get entityIdKeyInput() {
        return this._entityIdKey;
    }
    get eventPattern() {
        return this.getStringAttribute('event_pattern');
    }
    set eventPattern(value) {
        this._eventPattern = value;
    }
    resetEventPattern() {
        this._eventPattern = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventPatternInput() {
        return this._eventPattern;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get unitLabel() {
        return this.getStringAttribute('unit_label');
    }
    set unitLabel(value) {
        this._unitLabel = value;
    }
    resetUnitLabel() {
        this._unitLabel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unitLabelInput() {
        return this._unitLabel;
    }
    get valueKey() {
        return this.getStringAttribute('value_key');
    }
    set valueKey(value) {
        this._valueKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueKeyInput() {
        return this._valueKey;
    }
}
exports.EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference = EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference[_e] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference", version: "19.15.0" };
function evidentlyLaunchMetricMonitorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metric_definition: evidentlyLaunchMetricMonitorsMetricDefinitionToTerraform(struct.metricDefinition),
    };
}
exports.evidentlyLaunchMetricMonitorsToTerraform = evidentlyLaunchMetricMonitorsToTerraform;
function evidentlyLaunchMetricMonitorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metric_definition: {
            value: evidentlyLaunchMetricMonitorsMetricDefinitionToHclTerraform(struct.metricDefinition),
            isBlock: true,
            type: "list",
            storageClassType: "EvidentlyLaunchMetricMonitorsMetricDefinitionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchMetricMonitorsToHclTerraform = evidentlyLaunchMetricMonitorsToHclTerraform;
class EvidentlyLaunchMetricMonitorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metric_definition - computed: false, optional: false, required: true
        this._metricDefinition = new EvidentlyLaunchMetricMonitorsMetricDefinitionOutputReference(this, "metric_definition");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metricDefinition?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.metricDefinition = this._metricDefinition?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._metricDefinition.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._metricDefinition.internalValue = value.metricDefinition;
        }
    }
    get metricDefinition() {
        return this._metricDefinition;
    }
    putMetricDefinition(value) {
        this._metricDefinition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get metricDefinitionInput() {
        return this._metricDefinition.internalValue;
    }
}
exports.EvidentlyLaunchMetricMonitorsOutputReference = EvidentlyLaunchMetricMonitorsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchMetricMonitorsOutputReference[_f] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchMetricMonitorsOutputReference", version: "19.15.0" };
class EvidentlyLaunchMetricMonitorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EvidentlyLaunchMetricMonitorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EvidentlyLaunchMetricMonitorsList = EvidentlyLaunchMetricMonitorsList;
_g = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchMetricMonitorsList[_g] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchMetricMonitorsList", version: "19.15.0" };
function evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        evaluation_order: cdktf.numberToTerraform(struct.evaluationOrder),
        segment: cdktf.stringToTerraform(struct.segment),
        weights: cdktf.hashMapper(cdktf.numberToTerraform)(struct.weights),
    };
}
exports.evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToTerraform = evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToTerraform;
function evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        evaluation_order: {
            value: cdktf.numberToHclTerraform(struct.evaluationOrder),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        segment: {
            value: cdktf.stringToHclTerraform(struct.segment),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weights: {
            value: cdktf.hashMapperHcl(cdktf.numberToHclTerraform)(struct.weights),
            isBlock: false,
            type: "map",
            storageClassType: "numberMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToHclTerraform = evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToHclTerraform;
class EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._evaluationOrder !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluationOrder = this._evaluationOrder;
        }
        if (this._segment !== undefined) {
            hasAnyValues = true;
            internalValueResult.segment = this._segment;
        }
        if (this._weights !== undefined) {
            hasAnyValues = true;
            internalValueResult.weights = this._weights;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._evaluationOrder = undefined;
            this._segment = undefined;
            this._weights = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._evaluationOrder = value.evaluationOrder;
            this._segment = value.segment;
            this._weights = value.weights;
        }
    }
    get evaluationOrder() {
        return this.getNumberAttribute('evaluation_order');
    }
    set evaluationOrder(value) {
        this._evaluationOrder = value;
    }
    // Temporarily expose input value. Use with caution.
    get evaluationOrderInput() {
        return this._evaluationOrder;
    }
    get segment() {
        return this.getStringAttribute('segment');
    }
    set segment(value) {
        this._segment = value;
    }
    // Temporarily expose input value. Use with caution.
    get segmentInput() {
        return this._segment;
    }
    get weights() {
        return this.getNumberMapAttribute('weights');
    }
    set weights(value) {
        this._weights = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightsInput() {
        return this._weights;
    }
}
exports.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference = EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference[_h] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference", version: "19.15.0" };
class EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList = EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList;
_j = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList[_j] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList", version: "19.15.0" };
function evidentlyLaunchScheduledSplitsConfigStepsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        group_weights: cdktf.hashMapper(cdktf.numberToTerraform)(struct.groupWeights),
        start_time: cdktf.stringToTerraform(struct.startTime),
        segment_overrides: cdktf.listMapper(evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToTerraform, true)(struct.segmentOverrides),
    };
}
exports.evidentlyLaunchScheduledSplitsConfigStepsToTerraform = evidentlyLaunchScheduledSplitsConfigStepsToTerraform;
function evidentlyLaunchScheduledSplitsConfigStepsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        group_weights: {
            value: cdktf.hashMapperHcl(cdktf.numberToHclTerraform)(struct.groupWeights),
            isBlock: false,
            type: "map",
            storageClassType: "numberMap",
        },
        start_time: {
            value: cdktf.stringToHclTerraform(struct.startTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        segment_overrides: {
            value: cdktf.listMapperHcl(evidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesToHclTerraform, true)(struct.segmentOverrides),
            isBlock: true,
            type: "list",
            storageClassType: "EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchScheduledSplitsConfigStepsToHclTerraform = evidentlyLaunchScheduledSplitsConfigStepsToHclTerraform;
class EvidentlyLaunchScheduledSplitsConfigStepsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // segment_overrides - computed: false, optional: true, required: false
        this._segmentOverrides = new EvidentlyLaunchScheduledSplitsConfigStepsSegmentOverridesList(this, "segment_overrides", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._groupWeights !== undefined) {
            hasAnyValues = true;
            internalValueResult.groupWeights = this._groupWeights;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        if (this._segmentOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.segmentOverrides = this._segmentOverrides?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._groupWeights = undefined;
            this._startTime = undefined;
            this._segmentOverrides.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._groupWeights = value.groupWeights;
            this._startTime = value.startTime;
            this._segmentOverrides.internalValue = value.segmentOverrides;
        }
    }
    get groupWeights() {
        return this.getNumberMapAttribute('group_weights');
    }
    set groupWeights(value) {
        this._groupWeights = value;
    }
    // Temporarily expose input value. Use with caution.
    get groupWeightsInput() {
        return this._groupWeights;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
    get segmentOverrides() {
        return this._segmentOverrides;
    }
    putSegmentOverrides(value) {
        this._segmentOverrides.internalValue = value;
    }
    resetSegmentOverrides() {
        this._segmentOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentOverridesInput() {
        return this._segmentOverrides.internalValue;
    }
}
exports.EvidentlyLaunchScheduledSplitsConfigStepsOutputReference = EvidentlyLaunchScheduledSplitsConfigStepsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchScheduledSplitsConfigStepsOutputReference[_k] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchScheduledSplitsConfigStepsOutputReference", version: "19.15.0" };
class EvidentlyLaunchScheduledSplitsConfigStepsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EvidentlyLaunchScheduledSplitsConfigStepsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EvidentlyLaunchScheduledSplitsConfigStepsList = EvidentlyLaunchScheduledSplitsConfigStepsList;
_l = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchScheduledSplitsConfigStepsList[_l] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchScheduledSplitsConfigStepsList", version: "19.15.0" };
function evidentlyLaunchScheduledSplitsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        steps: cdktf.listMapper(evidentlyLaunchScheduledSplitsConfigStepsToTerraform, true)(struct.steps),
    };
}
exports.evidentlyLaunchScheduledSplitsConfigToTerraform = evidentlyLaunchScheduledSplitsConfigToTerraform;
function evidentlyLaunchScheduledSplitsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        steps: {
            value: cdktf.listMapperHcl(evidentlyLaunchScheduledSplitsConfigStepsToHclTerraform, true)(struct.steps),
            isBlock: true,
            type: "list",
            storageClassType: "EvidentlyLaunchScheduledSplitsConfigStepsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchScheduledSplitsConfigToHclTerraform = evidentlyLaunchScheduledSplitsConfigToHclTerraform;
class EvidentlyLaunchScheduledSplitsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // steps - computed: false, optional: false, required: true
        this._steps = new EvidentlyLaunchScheduledSplitsConfigStepsList(this, "steps", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._steps?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.steps = this._steps?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._steps.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._steps.internalValue = value.steps;
        }
    }
    get steps() {
        return this._steps;
    }
    putSteps(value) {
        this._steps.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stepsInput() {
        return this._steps.internalValue;
    }
}
exports.EvidentlyLaunchScheduledSplitsConfigOutputReference = EvidentlyLaunchScheduledSplitsConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchScheduledSplitsConfigOutputReference[_m] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchScheduledSplitsConfigOutputReference", version: "19.15.0" };
function evidentlyLaunchTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.evidentlyLaunchTimeoutsToTerraform = evidentlyLaunchTimeoutsToTerraform;
function evidentlyLaunchTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.evidentlyLaunchTimeoutsToHclTerraform = evidentlyLaunchTimeoutsToHclTerraform;
class EvidentlyLaunchTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EvidentlyLaunchTimeoutsOutputReference = EvidentlyLaunchTimeoutsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
EvidentlyLaunchTimeoutsOutputReference[_o] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunchTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/evidently_launch aws_evidently_launch}
*/
class EvidentlyLaunch extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EvidentlyLaunch resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EvidentlyLaunch to import
    * @param importFromId The id of the existing EvidentlyLaunch that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/evidently_launch#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EvidentlyLaunch to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_evidently_launch", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/evidently_launch aws_evidently_launch} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EvidentlyLaunchConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_evidently_launch',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // execution - computed: true, optional: false, required: false
        this._execution = new EvidentlyLaunchExecutionList(this, "execution", false);
        // groups - computed: false, optional: false, required: true
        this._groups = new EvidentlyLaunchGroupsList(this, "groups", false);
        // metric_monitors - computed: false, optional: true, required: false
        this._metricMonitors = new EvidentlyLaunchMetricMonitorsList(this, "metric_monitors", false);
        // scheduled_splits_config - computed: false, optional: true, required: false
        this._scheduledSplitsConfig = new EvidentlyLaunchScheduledSplitsConfigOutputReference(this, "scheduled_splits_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EvidentlyLaunchTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._randomizationSalt = config.randomizationSalt;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._groups.internalValue = config.groups;
        this._metricMonitors.internalValue = config.metricMonitors;
        this._scheduledSplitsConfig.internalValue = config.scheduledSplitsConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_time - computed: true, optional: false, required: false
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get execution() {
        return this._execution;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_time - computed: true, optional: false, required: false
    get lastUpdatedTime() {
        return this.getStringAttribute('last_updated_time');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get randomizationSalt() {
        return this.getStringAttribute('randomization_salt');
    }
    set randomizationSalt(value) {
        this._randomizationSalt = value;
    }
    resetRandomizationSalt() {
        this._randomizationSalt = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get randomizationSaltInput() {
        return this._randomizationSalt;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_reason - computed: true, optional: false, required: false
    get statusReason() {
        return this.getStringAttribute('status_reason');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    get groups() {
        return this._groups;
    }
    putGroups(value) {
        this._groups.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get groupsInput() {
        return this._groups.internalValue;
    }
    get metricMonitors() {
        return this._metricMonitors;
    }
    putMetricMonitors(value) {
        this._metricMonitors.internalValue = value;
    }
    resetMetricMonitors() {
        this._metricMonitors.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metricMonitorsInput() {
        return this._metricMonitors.internalValue;
    }
    get scheduledSplitsConfig() {
        return this._scheduledSplitsConfig;
    }
    putScheduledSplitsConfig(value) {
        this._scheduledSplitsConfig.internalValue = value;
    }
    resetScheduledSplitsConfig() {
        this._scheduledSplitsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduledSplitsConfigInput() {
        return this._scheduledSplitsConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            randomization_salt: cdktf.stringToTerraform(this._randomizationSalt),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            groups: cdktf.listMapper(evidentlyLaunchGroupsToTerraform, true)(this._groups.internalValue),
            metric_monitors: cdktf.listMapper(evidentlyLaunchMetricMonitorsToTerraform, true)(this._metricMonitors.internalValue),
            scheduled_splits_config: evidentlyLaunchScheduledSplitsConfigToTerraform(this._scheduledSplitsConfig.internalValue),
            timeouts: evidentlyLaunchTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            randomization_salt: {
                value: cdktf.stringToHclTerraform(this._randomizationSalt),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            groups: {
                value: cdktf.listMapperHcl(evidentlyLaunchGroupsToHclTerraform, true)(this._groups.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EvidentlyLaunchGroupsList",
            },
            metric_monitors: {
                value: cdktf.listMapperHcl(evidentlyLaunchMetricMonitorsToHclTerraform, true)(this._metricMonitors.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EvidentlyLaunchMetricMonitorsList",
            },
            scheduled_splits_config: {
                value: evidentlyLaunchScheduledSplitsConfigToHclTerraform(this._scheduledSplitsConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EvidentlyLaunchScheduledSplitsConfigList",
            },
            timeouts: {
                value: evidentlyLaunchTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "EvidentlyLaunchTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EvidentlyLaunch = EvidentlyLaunch;
_p = JSII_RTTI_SYMBOL_1;
EvidentlyLaunch[_p] = { fqn: "@cdktf/provider-aws.evidentlyLaunch.EvidentlyLaunch", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EvidentlyLaunch.tfResourceType = "aws_evidently_launch";
