// @cdktf/provider-aws/lib/data-aws-prometheus-workspaces/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsPrometheusWorkspaces = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/prometheus_workspaces aws_prometheus_workspaces}
*/
class DataAwsPrometheusWorkspaces extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsPrometheusWorkspaces resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsPrometheusWorkspaces to import
    * @param importFromId The id of the existing DataAwsPrometheusWorkspaces that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/prometheus_workspaces#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsPrometheusWorkspaces to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_prometheus_workspaces", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/prometheus_workspaces aws_prometheus_workspaces} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsPrometheusWorkspacesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_prometheus_workspaces',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._aliasPrefix = config.aliasPrefix;
        this._id = config.id;
    }
    get aliasPrefix() {
        return this.getStringAttribute('alias_prefix');
    }
    set aliasPrefix(value) {
        this._aliasPrefix = value;
    }
    resetAliasPrefix() {
        this._aliasPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aliasPrefixInput() {
        return this._aliasPrefix;
    }
    // aliases - computed: true, optional: false, required: false
    get aliases() {
        return this.getListAttribute('aliases');
    }
    // arns - computed: true, optional: false, required: false
    get arns() {
        return this.getListAttribute('arns');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // workspace_ids - computed: true, optional: false, required: false
    get workspaceIds() {
        return this.getListAttribute('workspace_ids');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias_prefix: cdktf.stringToTerraform(this._aliasPrefix),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            alias_prefix: {
                value: cdktf.stringToHclTerraform(this._aliasPrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsPrometheusWorkspaces = DataAwsPrometheusWorkspaces;
_a = JSII_RTTI_SYMBOL_1;
DataAwsPrometheusWorkspaces[_a] = { fqn: "@cdktf/provider-aws.dataAwsPrometheusWorkspaces.DataAwsPrometheusWorkspaces", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsPrometheusWorkspaces.tfResourceType = "aws_prometheus_workspaces";
