// @cdktf/provider-aws/lib/quicksight-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuicksightTemplate = exports.QuicksightTemplateTimeoutsOutputReference = exports.quicksightTemplateTimeoutsToHclTerraform = exports.quicksightTemplateTimeoutsToTerraform = exports.QuicksightTemplateSourceEntityOutputReference = exports.quicksightTemplateSourceEntityToHclTerraform = exports.quicksightTemplateSourceEntityToTerraform = exports.QuicksightTemplateSourceEntitySourceTemplateOutputReference = exports.quicksightTemplateSourceEntitySourceTemplateToHclTerraform = exports.quicksightTemplateSourceEntitySourceTemplateToTerraform = exports.QuicksightTemplateSourceEntitySourceAnalysisOutputReference = exports.quicksightTemplateSourceEntitySourceAnalysisToHclTerraform = exports.quicksightTemplateSourceEntitySourceAnalysisToTerraform = exports.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList = exports.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference = exports.quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToHclTerraform = exports.quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToTerraform = exports.QuicksightTemplatePermissionsList = exports.QuicksightTemplatePermissionsOutputReference = exports.quicksightTemplatePermissionsToHclTerraform = exports.quicksightTemplatePermissionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function quicksightTemplatePermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        actions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.actions),
        principal: cdktf.stringToTerraform(struct.principal),
    };
}
exports.quicksightTemplatePermissionsToTerraform = quicksightTemplatePermissionsToTerraform;
function quicksightTemplatePermissionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        actions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.actions),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        principal: {
            value: cdktf.stringToHclTerraform(struct.principal),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplatePermissionsToHclTerraform = quicksightTemplatePermissionsToHclTerraform;
class QuicksightTemplatePermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actions !== undefined) {
            hasAnyValues = true;
            internalValueResult.actions = this._actions;
        }
        if (this._principal !== undefined) {
            hasAnyValues = true;
            internalValueResult.principal = this._principal;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._actions = undefined;
            this._principal = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._actions = value.actions;
            this._principal = value.principal;
        }
    }
    get actions() {
        return cdktf.Fn.tolist(this.getListAttribute('actions'));
    }
    set actions(value) {
        this._actions = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionsInput() {
        return this._actions;
    }
    get principal() {
        return this.getStringAttribute('principal');
    }
    set principal(value) {
        this._principal = value;
    }
    // Temporarily expose input value. Use with caution.
    get principalInput() {
        return this._principal;
    }
}
exports.QuicksightTemplatePermissionsOutputReference = QuicksightTemplatePermissionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
QuicksightTemplatePermissionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplatePermissionsOutputReference", version: "19.15.0" };
class QuicksightTemplatePermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new QuicksightTemplatePermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.QuicksightTemplatePermissionsList = QuicksightTemplatePermissionsList;
_b = JSII_RTTI_SYMBOL_1;
QuicksightTemplatePermissionsList[_b] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplatePermissionsList", version: "19.15.0" };
function quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_set_arn: cdktf.stringToTerraform(struct.dataSetArn),
        data_set_placeholder: cdktf.stringToTerraform(struct.dataSetPlaceholder),
    };
}
exports.quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToTerraform = quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToTerraform;
function quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data_set_arn: {
            value: cdktf.stringToHclTerraform(struct.dataSetArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        data_set_placeholder: {
            value: cdktf.stringToHclTerraform(struct.dataSetPlaceholder),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToHclTerraform = quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToHclTerraform;
class QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataSetArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataSetArn = this._dataSetArn;
        }
        if (this._dataSetPlaceholder !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataSetPlaceholder = this._dataSetPlaceholder;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dataSetArn = undefined;
            this._dataSetPlaceholder = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dataSetArn = value.dataSetArn;
            this._dataSetPlaceholder = value.dataSetPlaceholder;
        }
    }
    get dataSetArn() {
        return this.getStringAttribute('data_set_arn');
    }
    set dataSetArn(value) {
        this._dataSetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataSetArnInput() {
        return this._dataSetArn;
    }
    get dataSetPlaceholder() {
        return this.getStringAttribute('data_set_placeholder');
    }
    set dataSetPlaceholder(value) {
        this._dataSetPlaceholder = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataSetPlaceholderInput() {
        return this._dataSetPlaceholder;
    }
}
exports.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference = QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference[_c] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference", version: "19.15.0" };
class QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList = QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList;
_d = JSII_RTTI_SYMBOL_1;
QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList[_d] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList", version: "19.15.0" };
function quicksightTemplateSourceEntitySourceAnalysisToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        data_set_references: cdktf.listMapper(quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToTerraform, true)(struct.dataSetReferences),
    };
}
exports.quicksightTemplateSourceEntitySourceAnalysisToTerraform = quicksightTemplateSourceEntitySourceAnalysisToTerraform;
function quicksightTemplateSourceEntitySourceAnalysisToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        data_set_references: {
            value: cdktf.listMapperHcl(quicksightTemplateSourceEntitySourceAnalysisDataSetReferencesToHclTerraform, true)(struct.dataSetReferences),
            isBlock: true,
            type: "list",
            storageClassType: "QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplateSourceEntitySourceAnalysisToHclTerraform = quicksightTemplateSourceEntitySourceAnalysisToHclTerraform;
class QuicksightTemplateSourceEntitySourceAnalysisOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // data_set_references - computed: false, optional: false, required: true
        this._dataSetReferences = new QuicksightTemplateSourceEntitySourceAnalysisDataSetReferencesList(this, "data_set_references", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._dataSetReferences?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataSetReferences = this._dataSetReferences?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._dataSetReferences.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._dataSetReferences.internalValue = value.dataSetReferences;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get dataSetReferences() {
        return this._dataSetReferences;
    }
    putDataSetReferences(value) {
        this._dataSetReferences.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataSetReferencesInput() {
        return this._dataSetReferences.internalValue;
    }
}
exports.QuicksightTemplateSourceEntitySourceAnalysisOutputReference = QuicksightTemplateSourceEntitySourceAnalysisOutputReference;
_e = JSII_RTTI_SYMBOL_1;
QuicksightTemplateSourceEntitySourceAnalysisOutputReference[_e] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateSourceEntitySourceAnalysisOutputReference", version: "19.15.0" };
function quicksightTemplateSourceEntitySourceTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
    };
}
exports.quicksightTemplateSourceEntitySourceTemplateToTerraform = quicksightTemplateSourceEntitySourceTemplateToTerraform;
function quicksightTemplateSourceEntitySourceTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplateSourceEntitySourceTemplateToHclTerraform = quicksightTemplateSourceEntitySourceTemplateToHclTerraform;
class QuicksightTemplateSourceEntitySourceTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
}
exports.QuicksightTemplateSourceEntitySourceTemplateOutputReference = QuicksightTemplateSourceEntitySourceTemplateOutputReference;
_f = JSII_RTTI_SYMBOL_1;
QuicksightTemplateSourceEntitySourceTemplateOutputReference[_f] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateSourceEntitySourceTemplateOutputReference", version: "19.15.0" };
function quicksightTemplateSourceEntityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_analysis: quicksightTemplateSourceEntitySourceAnalysisToTerraform(struct.sourceAnalysis),
        source_template: quicksightTemplateSourceEntitySourceTemplateToTerraform(struct.sourceTemplate),
    };
}
exports.quicksightTemplateSourceEntityToTerraform = quicksightTemplateSourceEntityToTerraform;
function quicksightTemplateSourceEntityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_analysis: {
            value: quicksightTemplateSourceEntitySourceAnalysisToHclTerraform(struct.sourceAnalysis),
            isBlock: true,
            type: "list",
            storageClassType: "QuicksightTemplateSourceEntitySourceAnalysisList",
        },
        source_template: {
            value: quicksightTemplateSourceEntitySourceTemplateToHclTerraform(struct.sourceTemplate),
            isBlock: true,
            type: "list",
            storageClassType: "QuicksightTemplateSourceEntitySourceTemplateList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplateSourceEntityToHclTerraform = quicksightTemplateSourceEntityToHclTerraform;
class QuicksightTemplateSourceEntityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // source_analysis - computed: false, optional: true, required: false
        this._sourceAnalysis = new QuicksightTemplateSourceEntitySourceAnalysisOutputReference(this, "source_analysis");
        // source_template - computed: false, optional: true, required: false
        this._sourceTemplate = new QuicksightTemplateSourceEntitySourceTemplateOutputReference(this, "source_template");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceAnalysis?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceAnalysis = this._sourceAnalysis?.internalValue;
        }
        if (this._sourceTemplate?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceTemplate = this._sourceTemplate?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sourceAnalysis.internalValue = undefined;
            this._sourceTemplate.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sourceAnalysis.internalValue = value.sourceAnalysis;
            this._sourceTemplate.internalValue = value.sourceTemplate;
        }
    }
    get sourceAnalysis() {
        return this._sourceAnalysis;
    }
    putSourceAnalysis(value) {
        this._sourceAnalysis.internalValue = value;
    }
    resetSourceAnalysis() {
        this._sourceAnalysis.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceAnalysisInput() {
        return this._sourceAnalysis.internalValue;
    }
    get sourceTemplate() {
        return this._sourceTemplate;
    }
    putSourceTemplate(value) {
        this._sourceTemplate.internalValue = value;
    }
    resetSourceTemplate() {
        this._sourceTemplate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceTemplateInput() {
        return this._sourceTemplate.internalValue;
    }
}
exports.QuicksightTemplateSourceEntityOutputReference = QuicksightTemplateSourceEntityOutputReference;
_g = JSII_RTTI_SYMBOL_1;
QuicksightTemplateSourceEntityOutputReference[_g] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateSourceEntityOutputReference", version: "19.15.0" };
function quicksightTemplateTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.quicksightTemplateTimeoutsToTerraform = quicksightTemplateTimeoutsToTerraform;
function quicksightTemplateTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightTemplateTimeoutsToHclTerraform = quicksightTemplateTimeoutsToHclTerraform;
class QuicksightTemplateTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.QuicksightTemplateTimeoutsOutputReference = QuicksightTemplateTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
QuicksightTemplateTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplateTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_template aws_quicksight_template}
*/
class QuicksightTemplate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a QuicksightTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the QuicksightTemplate to import
    * @param importFromId The id of the existing QuicksightTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the QuicksightTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_quicksight_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_template aws_quicksight_template} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options QuicksightTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_quicksight_template',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // permissions - computed: false, optional: true, required: false
        this._permissions = new QuicksightTemplatePermissionsList(this, "permissions", true);
        // source_entity - computed: false, optional: true, required: false
        this._sourceEntity = new QuicksightTemplateSourceEntityOutputReference(this, "source_entity");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new QuicksightTemplateTimeoutsOutputReference(this, "timeouts");
        this._awsAccountId = config.awsAccountId;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._templateId = config.templateId;
        this._versionDescription = config.versionDescription;
        this._definition = config.definition;
        this._permissions.internalValue = config.permissions;
        this._sourceEntity.internalValue = config.sourceEntity;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    set awsAccountId(value) {
        this._awsAccountId = value;
    }
    resetAwsAccountId() {
        this._awsAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccountIdInput() {
        return this._awsAccountId;
    }
    // created_time - computed: true, optional: false, required: false
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_time - computed: true, optional: false, required: false
    get lastUpdatedTime() {
        return this.getStringAttribute('last_updated_time');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // source_entity_arn - computed: true, optional: false, required: false
    get sourceEntityArn() {
        return this.getStringAttribute('source_entity_arn');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get templateId() {
        return this.getStringAttribute('template_id');
    }
    set templateId(value) {
        this._templateId = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateIdInput() {
        return this._templateId;
    }
    get versionDescription() {
        return this.getStringAttribute('version_description');
    }
    set versionDescription(value) {
        this._versionDescription = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionDescriptionInput() {
        return this._versionDescription;
    }
    // version_number - computed: true, optional: false, required: false
    get versionNumber() {
        return this.getNumberAttribute('version_number');
    }
    get definition() {
        return this.interpolationForAttribute('definition');
    }
    set definition(value) {
        this._definition = value;
    }
    resetDefinition() {
        this._definition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get definitionInput() {
        return this._definition;
    }
    get permissions() {
        return this._permissions;
    }
    putPermissions(value) {
        this._permissions.internalValue = value;
    }
    resetPermissions() {
        this._permissions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionsInput() {
        return this._permissions.internalValue;
    }
    get sourceEntity() {
        return this._sourceEntity;
    }
    putSourceEntity(value) {
        this._sourceEntity.internalValue = value;
    }
    resetSourceEntity() {
        this._sourceEntity.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceEntityInput() {
        return this._sourceEntity.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            aws_account_id: cdktf.stringToTerraform(this._awsAccountId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            template_id: cdktf.stringToTerraform(this._templateId),
            version_description: cdktf.stringToTerraform(this._versionDescription),
            definition: cdktf.anyToTerraform(this._definition),
            permissions: cdktf.listMapper(quicksightTemplatePermissionsToTerraform, true)(this._permissions.internalValue),
            source_entity: quicksightTemplateSourceEntityToTerraform(this._sourceEntity.internalValue),
            timeouts: quicksightTemplateTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            aws_account_id: {
                value: cdktf.stringToHclTerraform(this._awsAccountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            template_id: {
                value: cdktf.stringToHclTerraform(this._templateId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version_description: {
                value: cdktf.stringToHclTerraform(this._versionDescription),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            definition: {
                value: cdktf.anyToHclTerraform(this._definition),
                isBlock: false,
                type: "simple",
                storageClassType: "any",
            },
            permissions: {
                value: cdktf.listMapperHcl(quicksightTemplatePermissionsToHclTerraform, true)(this._permissions.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "QuicksightTemplatePermissionsList",
            },
            source_entity: {
                value: quicksightTemplateSourceEntityToHclTerraform(this._sourceEntity.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "QuicksightTemplateSourceEntityList",
            },
            timeouts: {
                value: quicksightTemplateTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "QuicksightTemplateTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.QuicksightTemplate = QuicksightTemplate;
_j = JSII_RTTI_SYMBOL_1;
QuicksightTemplate[_j] = { fqn: "@cdktf/provider-aws.quicksightTemplate.QuicksightTemplate", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
QuicksightTemplate.tfResourceType = "aws_quicksight_template";
