// @cdktf/provider-aws/lib/sesv2-account-vdm-attributes/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sesv2AccountVdmAttributes = exports.Sesv2AccountVdmAttributesGuardianAttributesOutputReference = exports.sesv2AccountVdmAttributesGuardianAttributesToHclTerraform = exports.sesv2AccountVdmAttributesGuardianAttributesToTerraform = exports.Sesv2AccountVdmAttributesDashboardAttributesOutputReference = exports.sesv2AccountVdmAttributesDashboardAttributesToHclTerraform = exports.sesv2AccountVdmAttributesDashboardAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sesv2AccountVdmAttributesDashboardAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        engagement_metrics: cdktf.stringToTerraform(struct.engagementMetrics),
    };
}
exports.sesv2AccountVdmAttributesDashboardAttributesToTerraform = sesv2AccountVdmAttributesDashboardAttributesToTerraform;
function sesv2AccountVdmAttributesDashboardAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        engagement_metrics: {
            value: cdktf.stringToHclTerraform(struct.engagementMetrics),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sesv2AccountVdmAttributesDashboardAttributesToHclTerraform = sesv2AccountVdmAttributesDashboardAttributesToHclTerraform;
class Sesv2AccountVdmAttributesDashboardAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._engagementMetrics !== undefined) {
            hasAnyValues = true;
            internalValueResult.engagementMetrics = this._engagementMetrics;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._engagementMetrics = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._engagementMetrics = value.engagementMetrics;
        }
    }
    get engagementMetrics() {
        return this.getStringAttribute('engagement_metrics');
    }
    set engagementMetrics(value) {
        this._engagementMetrics = value;
    }
    resetEngagementMetrics() {
        this._engagementMetrics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engagementMetricsInput() {
        return this._engagementMetrics;
    }
}
exports.Sesv2AccountVdmAttributesDashboardAttributesOutputReference = Sesv2AccountVdmAttributesDashboardAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Sesv2AccountVdmAttributesDashboardAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.sesv2AccountVdmAttributes.Sesv2AccountVdmAttributesDashboardAttributesOutputReference", version: "19.15.0" };
function sesv2AccountVdmAttributesGuardianAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        optimized_shared_delivery: cdktf.stringToTerraform(struct.optimizedSharedDelivery),
    };
}
exports.sesv2AccountVdmAttributesGuardianAttributesToTerraform = sesv2AccountVdmAttributesGuardianAttributesToTerraform;
function sesv2AccountVdmAttributesGuardianAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        optimized_shared_delivery: {
            value: cdktf.stringToHclTerraform(struct.optimizedSharedDelivery),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sesv2AccountVdmAttributesGuardianAttributesToHclTerraform = sesv2AccountVdmAttributesGuardianAttributesToHclTerraform;
class Sesv2AccountVdmAttributesGuardianAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._optimizedSharedDelivery !== undefined) {
            hasAnyValues = true;
            internalValueResult.optimizedSharedDelivery = this._optimizedSharedDelivery;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._optimizedSharedDelivery = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._optimizedSharedDelivery = value.optimizedSharedDelivery;
        }
    }
    get optimizedSharedDelivery() {
        return this.getStringAttribute('optimized_shared_delivery');
    }
    set optimizedSharedDelivery(value) {
        this._optimizedSharedDelivery = value;
    }
    resetOptimizedSharedDelivery() {
        this._optimizedSharedDelivery = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optimizedSharedDeliveryInput() {
        return this._optimizedSharedDelivery;
    }
}
exports.Sesv2AccountVdmAttributesGuardianAttributesOutputReference = Sesv2AccountVdmAttributesGuardianAttributesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Sesv2AccountVdmAttributesGuardianAttributesOutputReference[_b] = { fqn: "@cdktf/provider-aws.sesv2AccountVdmAttributes.Sesv2AccountVdmAttributesGuardianAttributesOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_account_vdm_attributes aws_sesv2_account_vdm_attributes}
*/
class Sesv2AccountVdmAttributes extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Sesv2AccountVdmAttributes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Sesv2AccountVdmAttributes to import
    * @param importFromId The id of the existing Sesv2AccountVdmAttributes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_account_vdm_attributes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Sesv2AccountVdmAttributes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sesv2_account_vdm_attributes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_account_vdm_attributes aws_sesv2_account_vdm_attributes} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Sesv2AccountVdmAttributesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sesv2_account_vdm_attributes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dashboard_attributes - computed: false, optional: true, required: false
        this._dashboardAttributes = new Sesv2AccountVdmAttributesDashboardAttributesOutputReference(this, "dashboard_attributes");
        // guardian_attributes - computed: false, optional: true, required: false
        this._guardianAttributes = new Sesv2AccountVdmAttributesGuardianAttributesOutputReference(this, "guardian_attributes");
        this._id = config.id;
        this._vdmEnabled = config.vdmEnabled;
        this._dashboardAttributes.internalValue = config.dashboardAttributes;
        this._guardianAttributes.internalValue = config.guardianAttributes;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get vdmEnabled() {
        return this.getStringAttribute('vdm_enabled');
    }
    set vdmEnabled(value) {
        this._vdmEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get vdmEnabledInput() {
        return this._vdmEnabled;
    }
    get dashboardAttributes() {
        return this._dashboardAttributes;
    }
    putDashboardAttributes(value) {
        this._dashboardAttributes.internalValue = value;
    }
    resetDashboardAttributes() {
        this._dashboardAttributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dashboardAttributesInput() {
        return this._dashboardAttributes.internalValue;
    }
    get guardianAttributes() {
        return this._guardianAttributes;
    }
    putGuardianAttributes(value) {
        this._guardianAttributes.internalValue = value;
    }
    resetGuardianAttributes() {
        this._guardianAttributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get guardianAttributesInput() {
        return this._guardianAttributes.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            vdm_enabled: cdktf.stringToTerraform(this._vdmEnabled),
            dashboard_attributes: sesv2AccountVdmAttributesDashboardAttributesToTerraform(this._dashboardAttributes.internalValue),
            guardian_attributes: sesv2AccountVdmAttributesGuardianAttributesToTerraform(this._guardianAttributes.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            vdm_enabled: {
                value: cdktf.stringToHclTerraform(this._vdmEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dashboard_attributes: {
                value: sesv2AccountVdmAttributesDashboardAttributesToHclTerraform(this._dashboardAttributes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Sesv2AccountVdmAttributesDashboardAttributesList",
            },
            guardian_attributes: {
                value: sesv2AccountVdmAttributesGuardianAttributesToHclTerraform(this._guardianAttributes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Sesv2AccountVdmAttributesGuardianAttributesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Sesv2AccountVdmAttributes = Sesv2AccountVdmAttributes;
_c = JSII_RTTI_SYMBOL_1;
Sesv2AccountVdmAttributes[_c] = { fqn: "@cdktf/provider-aws.sesv2AccountVdmAttributes.Sesv2AccountVdmAttributes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Sesv2AccountVdmAttributes.tfResourceType = "aws_sesv2_account_vdm_attributes";
