// @cdktf/provider-aws/lib/data-aws-ec2-instance-type/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEc2InstanceType = exports.DataAwsEc2InstanceTypeTimeoutsOutputReference = exports.dataAwsEc2InstanceTypeTimeoutsToHclTerraform = exports.dataAwsEc2InstanceTypeTimeoutsToTerraform = exports.DataAwsEc2InstanceTypeInstanceDisksList = exports.DataAwsEc2InstanceTypeInstanceDisksOutputReference = exports.dataAwsEc2InstanceTypeInstanceDisksToHclTerraform = exports.dataAwsEc2InstanceTypeInstanceDisksToTerraform = exports.DataAwsEc2InstanceTypeInferenceAcceleratorsList = exports.DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference = exports.dataAwsEc2InstanceTypeInferenceAcceleratorsToHclTerraform = exports.dataAwsEc2InstanceTypeInferenceAcceleratorsToTerraform = exports.DataAwsEc2InstanceTypeGpusList = exports.DataAwsEc2InstanceTypeGpusOutputReference = exports.dataAwsEc2InstanceTypeGpusToHclTerraform = exports.dataAwsEc2InstanceTypeGpusToTerraform = exports.DataAwsEc2InstanceTypeFpgasList = exports.DataAwsEc2InstanceTypeFpgasOutputReference = exports.dataAwsEc2InstanceTypeFpgasToHclTerraform = exports.dataAwsEc2InstanceTypeFpgasToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEc2InstanceTypeFpgasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEc2InstanceTypeFpgasToTerraform = dataAwsEc2InstanceTypeFpgasToTerraform;
function dataAwsEc2InstanceTypeFpgasToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEc2InstanceTypeFpgasToHclTerraform = dataAwsEc2InstanceTypeFpgasToHclTerraform;
class DataAwsEc2InstanceTypeFpgasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // manufacturer - computed: true, optional: false, required: false
    get manufacturer() {
        return this.getStringAttribute('manufacturer');
    }
    // memory_size - computed: true, optional: false, required: false
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsEc2InstanceTypeFpgasOutputReference = DataAwsEc2InstanceTypeFpgasOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeFpgasOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeFpgasOutputReference", version: "19.15.0" };
class DataAwsEc2InstanceTypeFpgasList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEc2InstanceTypeFpgasOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEc2InstanceTypeFpgasList = DataAwsEc2InstanceTypeFpgasList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeFpgasList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeFpgasList", version: "19.15.0" };
function dataAwsEc2InstanceTypeGpusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEc2InstanceTypeGpusToTerraform = dataAwsEc2InstanceTypeGpusToTerraform;
function dataAwsEc2InstanceTypeGpusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEc2InstanceTypeGpusToHclTerraform = dataAwsEc2InstanceTypeGpusToHclTerraform;
class DataAwsEc2InstanceTypeGpusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // manufacturer - computed: true, optional: false, required: false
    get manufacturer() {
        return this.getStringAttribute('manufacturer');
    }
    // memory_size - computed: true, optional: false, required: false
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsEc2InstanceTypeGpusOutputReference = DataAwsEc2InstanceTypeGpusOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeGpusOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeGpusOutputReference", version: "19.15.0" };
class DataAwsEc2InstanceTypeGpusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEc2InstanceTypeGpusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEc2InstanceTypeGpusList = DataAwsEc2InstanceTypeGpusList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeGpusList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeGpusList", version: "19.15.0" };
function dataAwsEc2InstanceTypeInferenceAcceleratorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEc2InstanceTypeInferenceAcceleratorsToTerraform = dataAwsEc2InstanceTypeInferenceAcceleratorsToTerraform;
function dataAwsEc2InstanceTypeInferenceAcceleratorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEc2InstanceTypeInferenceAcceleratorsToHclTerraform = dataAwsEc2InstanceTypeInferenceAcceleratorsToHclTerraform;
class DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // manufacturer - computed: true, optional: false, required: false
    get manufacturer() {
        return this.getStringAttribute('manufacturer');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference = DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference", version: "19.15.0" };
class DataAwsEc2InstanceTypeInferenceAcceleratorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEc2InstanceTypeInferenceAcceleratorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEc2InstanceTypeInferenceAcceleratorsList = DataAwsEc2InstanceTypeInferenceAcceleratorsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeInferenceAcceleratorsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeInferenceAcceleratorsList", version: "19.15.0" };
function dataAwsEc2InstanceTypeInstanceDisksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEc2InstanceTypeInstanceDisksToTerraform = dataAwsEc2InstanceTypeInstanceDisksToTerraform;
function dataAwsEc2InstanceTypeInstanceDisksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsEc2InstanceTypeInstanceDisksToHclTerraform = dataAwsEc2InstanceTypeInstanceDisksToHclTerraform;
class DataAwsEc2InstanceTypeInstanceDisksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsEc2InstanceTypeInstanceDisksOutputReference = DataAwsEc2InstanceTypeInstanceDisksOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeInstanceDisksOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeInstanceDisksOutputReference", version: "19.15.0" };
class DataAwsEc2InstanceTypeInstanceDisksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEc2InstanceTypeInstanceDisksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEc2InstanceTypeInstanceDisksList = DataAwsEc2InstanceTypeInstanceDisksList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeInstanceDisksList[_h] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeInstanceDisksList", version: "19.15.0" };
function dataAwsEc2InstanceTypeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsEc2InstanceTypeTimeoutsToTerraform = dataAwsEc2InstanceTypeTimeoutsToTerraform;
function dataAwsEc2InstanceTypeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEc2InstanceTypeTimeoutsToHclTerraform = dataAwsEc2InstanceTypeTimeoutsToHclTerraform;
class DataAwsEc2InstanceTypeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsEc2InstanceTypeTimeoutsOutputReference = DataAwsEc2InstanceTypeTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceTypeTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceTypeTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ec2_instance_type aws_ec2_instance_type}
*/
class DataAwsEc2InstanceType extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEc2InstanceType resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEc2InstanceType to import
    * @param importFromId The id of the existing DataAwsEc2InstanceType that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ec2_instance_type#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEc2InstanceType to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ec2_instance_type", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ec2_instance_type aws_ec2_instance_type} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEc2InstanceTypeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_instance_type',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // fpgas - computed: true, optional: false, required: false
        this._fpgas = new DataAwsEc2InstanceTypeFpgasList(this, "fpgas", true);
        // gpus - computed: true, optional: false, required: false
        this._gpus = new DataAwsEc2InstanceTypeGpusList(this, "gpus", true);
        // inference_accelerators - computed: true, optional: false, required: false
        this._inferenceAccelerators = new DataAwsEc2InstanceTypeInferenceAcceleratorsList(this, "inference_accelerators", true);
        // instance_disks - computed: true, optional: false, required: false
        this._instanceDisks = new DataAwsEc2InstanceTypeInstanceDisksList(this, "instance_disks", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsEc2InstanceTypeTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._instanceType = config.instanceType;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // auto_recovery_supported - computed: true, optional: false, required: false
    get autoRecoverySupported() {
        return this.getBooleanAttribute('auto_recovery_supported');
    }
    // bare_metal - computed: true, optional: false, required: false
    get bareMetal() {
        return this.getBooleanAttribute('bare_metal');
    }
    // burstable_performance_supported - computed: true, optional: false, required: false
    get burstablePerformanceSupported() {
        return this.getBooleanAttribute('burstable_performance_supported');
    }
    // current_generation - computed: true, optional: false, required: false
    get currentGeneration() {
        return this.getBooleanAttribute('current_generation');
    }
    // dedicated_hosts_supported - computed: true, optional: false, required: false
    get dedicatedHostsSupported() {
        return this.getBooleanAttribute('dedicated_hosts_supported');
    }
    // default_cores - computed: true, optional: false, required: false
    get defaultCores() {
        return this.getNumberAttribute('default_cores');
    }
    // default_threads_per_core - computed: true, optional: false, required: false
    get defaultThreadsPerCore() {
        return this.getNumberAttribute('default_threads_per_core');
    }
    // default_vcpus - computed: true, optional: false, required: false
    get defaultVcpus() {
        return this.getNumberAttribute('default_vcpus');
    }
    // ebs_encryption_support - computed: true, optional: false, required: false
    get ebsEncryptionSupport() {
        return this.getStringAttribute('ebs_encryption_support');
    }
    // ebs_nvme_support - computed: true, optional: false, required: false
    get ebsNvmeSupport() {
        return this.getStringAttribute('ebs_nvme_support');
    }
    // ebs_optimized_support - computed: true, optional: false, required: false
    get ebsOptimizedSupport() {
        return this.getStringAttribute('ebs_optimized_support');
    }
    // ebs_performance_baseline_bandwidth - computed: true, optional: false, required: false
    get ebsPerformanceBaselineBandwidth() {
        return this.getNumberAttribute('ebs_performance_baseline_bandwidth');
    }
    // ebs_performance_baseline_iops - computed: true, optional: false, required: false
    get ebsPerformanceBaselineIops() {
        return this.getNumberAttribute('ebs_performance_baseline_iops');
    }
    // ebs_performance_baseline_throughput - computed: true, optional: false, required: false
    get ebsPerformanceBaselineThroughput() {
        return this.getNumberAttribute('ebs_performance_baseline_throughput');
    }
    // ebs_performance_maximum_bandwidth - computed: true, optional: false, required: false
    get ebsPerformanceMaximumBandwidth() {
        return this.getNumberAttribute('ebs_performance_maximum_bandwidth');
    }
    // ebs_performance_maximum_iops - computed: true, optional: false, required: false
    get ebsPerformanceMaximumIops() {
        return this.getNumberAttribute('ebs_performance_maximum_iops');
    }
    // ebs_performance_maximum_throughput - computed: true, optional: false, required: false
    get ebsPerformanceMaximumThroughput() {
        return this.getNumberAttribute('ebs_performance_maximum_throughput');
    }
    // efa_supported - computed: true, optional: false, required: false
    get efaSupported() {
        return this.getBooleanAttribute('efa_supported');
    }
    // ena_support - computed: true, optional: false, required: false
    get enaSupport() {
        return this.getStringAttribute('ena_support');
    }
    // encryption_in_transit_supported - computed: true, optional: false, required: false
    get encryptionInTransitSupported() {
        return this.getBooleanAttribute('encryption_in_transit_supported');
    }
    get fpgas() {
        return this._fpgas;
    }
    // free_tier_eligible - computed: true, optional: false, required: false
    get freeTierEligible() {
        return this.getBooleanAttribute('free_tier_eligible');
    }
    get gpus() {
        return this._gpus;
    }
    // hibernation_supported - computed: true, optional: false, required: false
    get hibernationSupported() {
        return this.getBooleanAttribute('hibernation_supported');
    }
    // hypervisor - computed: true, optional: false, required: false
    get hypervisor() {
        return this.getStringAttribute('hypervisor');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get inferenceAccelerators() {
        return this._inferenceAccelerators;
    }
    get instanceDisks() {
        return this._instanceDisks;
    }
    // instance_storage_supported - computed: true, optional: false, required: false
    get instanceStorageSupported() {
        return this.getBooleanAttribute('instance_storage_supported');
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    // ipv6_supported - computed: true, optional: false, required: false
    get ipv6Supported() {
        return this.getBooleanAttribute('ipv6_supported');
    }
    // maximum_ipv4_addresses_per_interface - computed: true, optional: false, required: false
    get maximumIpv4AddressesPerInterface() {
        return this.getNumberAttribute('maximum_ipv4_addresses_per_interface');
    }
    // maximum_ipv6_addresses_per_interface - computed: true, optional: false, required: false
    get maximumIpv6AddressesPerInterface() {
        return this.getNumberAttribute('maximum_ipv6_addresses_per_interface');
    }
    // maximum_network_cards - computed: true, optional: false, required: false
    get maximumNetworkCards() {
        return this.getNumberAttribute('maximum_network_cards');
    }
    // maximum_network_interfaces - computed: true, optional: false, required: false
    get maximumNetworkInterfaces() {
        return this.getNumberAttribute('maximum_network_interfaces');
    }
    // memory_size - computed: true, optional: false, required: false
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    // network_performance - computed: true, optional: false, required: false
    get networkPerformance() {
        return this.getStringAttribute('network_performance');
    }
    // supported_architectures - computed: true, optional: false, required: false
    get supportedArchitectures() {
        return this.getListAttribute('supported_architectures');
    }
    // supported_placement_strategies - computed: true, optional: false, required: false
    get supportedPlacementStrategies() {
        return this.getListAttribute('supported_placement_strategies');
    }
    // supported_root_device_types - computed: true, optional: false, required: false
    get supportedRootDeviceTypes() {
        return this.getListAttribute('supported_root_device_types');
    }
    // supported_usages_classes - computed: true, optional: false, required: false
    get supportedUsagesClasses() {
        return this.getListAttribute('supported_usages_classes');
    }
    // supported_virtualization_types - computed: true, optional: false, required: false
    get supportedVirtualizationTypes() {
        return this.getListAttribute('supported_virtualization_types');
    }
    // sustained_clock_speed - computed: true, optional: false, required: false
    get sustainedClockSpeed() {
        return this.getNumberAttribute('sustained_clock_speed');
    }
    // total_fpga_memory - computed: true, optional: false, required: false
    get totalFpgaMemory() {
        return this.getNumberAttribute('total_fpga_memory');
    }
    // total_gpu_memory - computed: true, optional: false, required: false
    get totalGpuMemory() {
        return this.getNumberAttribute('total_gpu_memory');
    }
    // total_instance_storage - computed: true, optional: false, required: false
    get totalInstanceStorage() {
        return this.getNumberAttribute('total_instance_storage');
    }
    // valid_cores - computed: true, optional: false, required: false
    get validCores() {
        return this.getNumberListAttribute('valid_cores');
    }
    // valid_threads_per_core - computed: true, optional: false, required: false
    get validThreadsPerCore() {
        return this.getNumberListAttribute('valid_threads_per_core');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            timeouts: dataAwsEc2InstanceTypeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_type: {
                value: cdktf.stringToHclTerraform(this._instanceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: dataAwsEc2InstanceTypeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsEc2InstanceTypeTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEc2InstanceType = DataAwsEc2InstanceType;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEc2InstanceType[_k] = { fqn: "@cdktf/provider-aws.dataAwsEc2InstanceType.DataAwsEc2InstanceType", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEc2InstanceType.tfResourceType = "aws_ec2_instance_type";
