// @cdktf/provider-aws/lib/guardduty-organization-configuration-feature/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuarddutyOrganizationConfigurationFeature = exports.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList = exports.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference = exports.guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToHclTerraform = exports.guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_enable: cdktf.stringToTerraform(struct.autoEnable),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToTerraform = guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToTerraform;
function guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_enable: {
            value: cdktf.stringToHclTerraform(struct.autoEnable),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToHclTerraform = guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToHclTerraform;
class GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoEnable !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoEnable = this._autoEnable;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._autoEnable = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._autoEnable = value.autoEnable;
            this._name = value.name;
        }
    }
    get autoEnable() {
        return this.getStringAttribute('auto_enable');
    }
    set autoEnable(value) {
        this._autoEnable = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableInput() {
        return this._autoEnable;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference = GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.guarddutyOrganizationConfigurationFeature.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference", version: "19.15.0" };
class GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList = GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList[_b] = { fqn: "@cdktf/provider-aws.guarddutyOrganizationConfigurationFeature.GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_organization_configuration_feature aws_guardduty_organization_configuration_feature}
*/
class GuarddutyOrganizationConfigurationFeature extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GuarddutyOrganizationConfigurationFeature resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GuarddutyOrganizationConfigurationFeature to import
    * @param importFromId The id of the existing GuarddutyOrganizationConfigurationFeature that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_organization_configuration_feature#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GuarddutyOrganizationConfigurationFeature to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_guardduty_organization_configuration_feature", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/guardduty_organization_configuration_feature aws_guardduty_organization_configuration_feature} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GuarddutyOrganizationConfigurationFeatureConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_organization_configuration_feature',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // additional_configuration - computed: false, optional: true, required: false
        this._additionalConfiguration = new GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList(this, "additional_configuration", false);
        this._autoEnable = config.autoEnable;
        this._detectorId = config.detectorId;
        this._id = config.id;
        this._name = config.name;
        this._additionalConfiguration.internalValue = config.additionalConfiguration;
    }
    get autoEnable() {
        return this.getStringAttribute('auto_enable');
    }
    set autoEnable(value) {
        this._autoEnable = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableInput() {
        return this._autoEnable;
    }
    get detectorId() {
        return this.getStringAttribute('detector_id');
    }
    set detectorId(value) {
        this._detectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    get detectorIdInput() {
        return this._detectorId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get additionalConfiguration() {
        return this._additionalConfiguration;
    }
    putAdditionalConfiguration(value) {
        this._additionalConfiguration.internalValue = value;
    }
    resetAdditionalConfiguration() {
        this._additionalConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalConfigurationInput() {
        return this._additionalConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_enable: cdktf.stringToTerraform(this._autoEnable),
            detector_id: cdktf.stringToTerraform(this._detectorId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            additional_configuration: cdktf.listMapper(guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToTerraform, true)(this._additionalConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_enable: {
                value: cdktf.stringToHclTerraform(this._autoEnable),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            detector_id: {
                value: cdktf.stringToHclTerraform(this._detectorId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            additional_configuration: {
                value: cdktf.listMapperHcl(guarddutyOrganizationConfigurationFeatureAdditionalConfigurationToHclTerraform, true)(this._additionalConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GuarddutyOrganizationConfigurationFeatureAdditionalConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GuarddutyOrganizationConfigurationFeature = GuarddutyOrganizationConfigurationFeature;
_c = JSII_RTTI_SYMBOL_1;
GuarddutyOrganizationConfigurationFeature[_c] = { fqn: "@cdktf/provider-aws.guarddutyOrganizationConfigurationFeature.GuarddutyOrganizationConfigurationFeature", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GuarddutyOrganizationConfigurationFeature.tfResourceType = "aws_guardduty_organization_configuration_feature";
