// @cdktf/provider-aws/lib/data-aws-dynamodb-table-item/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDynamodbTableItem = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dynamodb_table_item aws_dynamodb_table_item}
*/
class DataAwsDynamodbTableItem extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDynamodbTableItem resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDynamodbTableItem to import
    * @param importFromId The id of the existing DataAwsDynamodbTableItem that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dynamodb_table_item#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDynamodbTableItem to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_dynamodb_table_item", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dynamodb_table_item aws_dynamodb_table_item} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDynamodbTableItemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table_item',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._expressionAttributeNames = config.expressionAttributeNames;
        this._id = config.id;
        this._key = config.key;
        this._projectionExpression = config.projectionExpression;
        this._tableName = config.tableName;
    }
    get expressionAttributeNames() {
        return this.getStringMapAttribute('expression_attribute_names');
    }
    set expressionAttributeNames(value) {
        this._expressionAttributeNames = value;
    }
    resetExpressionAttributeNames() {
        this._expressionAttributeNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionAttributeNamesInput() {
        return this._expressionAttributeNames;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // item - computed: true, optional: false, required: false
    get item() {
        return this.getStringAttribute('item');
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get projectionExpression() {
        return this.getStringAttribute('projection_expression');
    }
    set projectionExpression(value) {
        this._projectionExpression = value;
    }
    resetProjectionExpression() {
        this._projectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectionExpressionInput() {
        return this._projectionExpression;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            expression_attribute_names: cdktf.hashMapper(cdktf.stringToTerraform)(this._expressionAttributeNames),
            id: cdktf.stringToTerraform(this._id),
            key: cdktf.stringToTerraform(this._key),
            projection_expression: cdktf.stringToTerraform(this._projectionExpression),
            table_name: cdktf.stringToTerraform(this._tableName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            expression_attribute_names: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._expressionAttributeNames),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key: {
                value: cdktf.stringToHclTerraform(this._key),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            projection_expression: {
                value: cdktf.stringToHclTerraform(this._projectionExpression),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            table_name: {
                value: cdktf.stringToHclTerraform(this._tableName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDynamodbTableItem = DataAwsDynamodbTableItem;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableItem[_a] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTableItem.DataAwsDynamodbTableItem", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDynamodbTableItem.tfResourceType = "aws_dynamodb_table_item";
