// @cdktf/provider-aws/lib/verifiedaccess-endpoint/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedaccessEndpoint = exports.VerifiedaccessEndpointTimeoutsOutputReference = exports.verifiedaccessEndpointTimeoutsToHclTerraform = exports.verifiedaccessEndpointTimeoutsToTerraform = exports.VerifiedaccessEndpointSseSpecificationOutputReference = exports.verifiedaccessEndpointSseSpecificationToHclTerraform = exports.verifiedaccessEndpointSseSpecificationToTerraform = exports.VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference = exports.verifiedaccessEndpointNetworkInterfaceOptionsToHclTerraform = exports.verifiedaccessEndpointNetworkInterfaceOptionsToTerraform = exports.VerifiedaccessEndpointLoadBalancerOptionsOutputReference = exports.verifiedaccessEndpointLoadBalancerOptionsToHclTerraform = exports.verifiedaccessEndpointLoadBalancerOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function verifiedaccessEndpointLoadBalancerOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        load_balancer_arn: cdktf.stringToTerraform(struct.loadBalancerArn),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.verifiedaccessEndpointLoadBalancerOptionsToTerraform = verifiedaccessEndpointLoadBalancerOptionsToTerraform;
function verifiedaccessEndpointLoadBalancerOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        load_balancer_arn: {
            value: cdktf.stringToHclTerraform(struct.loadBalancerArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessEndpointLoadBalancerOptionsToHclTerraform = verifiedaccessEndpointLoadBalancerOptionsToHclTerraform;
class VerifiedaccessEndpointLoadBalancerOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._loadBalancerArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.loadBalancerArn = this._loadBalancerArn;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._loadBalancerArn = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._loadBalancerArn = value.loadBalancerArn;
            this._port = value.port;
            this._protocol = value.protocol;
            this._subnetIds = value.subnetIds;
        }
    }
    get loadBalancerArn() {
        return this.getStringAttribute('load_balancer_arn');
    }
    set loadBalancerArn(value) {
        this._loadBalancerArn = value;
    }
    resetLoadBalancerArn() {
        this._loadBalancerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerArnInput() {
        return this._loadBalancerArn;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.VerifiedaccessEndpointLoadBalancerOptionsOutputReference = VerifiedaccessEndpointLoadBalancerOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VerifiedaccessEndpointLoadBalancerOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.verifiedaccessEndpoint.VerifiedaccessEndpointLoadBalancerOptionsOutputReference", version: "19.15.0" };
function verifiedaccessEndpointNetworkInterfaceOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        network_interface_id: cdktf.stringToTerraform(struct.networkInterfaceId),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
    };
}
exports.verifiedaccessEndpointNetworkInterfaceOptionsToTerraform = verifiedaccessEndpointNetworkInterfaceOptionsToTerraform;
function verifiedaccessEndpointNetworkInterfaceOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        network_interface_id: {
            value: cdktf.stringToHclTerraform(struct.networkInterfaceId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessEndpointNetworkInterfaceOptionsToHclTerraform = verifiedaccessEndpointNetworkInterfaceOptionsToHclTerraform;
class VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._networkInterfaceId !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkInterfaceId = this._networkInterfaceId;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._networkInterfaceId = undefined;
            this._port = undefined;
            this._protocol = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._networkInterfaceId = value.networkInterfaceId;
            this._port = value.port;
            this._protocol = value.protocol;
        }
    }
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    set networkInterfaceId(value) {
        this._networkInterfaceId = value;
    }
    resetNetworkInterfaceId() {
        this._networkInterfaceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfaceIdInput() {
        return this._networkInterfaceId;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
}
exports.VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference = VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.verifiedaccessEndpoint.VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference", version: "19.15.0" };
function verifiedaccessEndpointSseSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        customer_managed_key_enabled: cdktf.booleanToTerraform(struct.customerManagedKeyEnabled),
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
    };
}
exports.verifiedaccessEndpointSseSpecificationToTerraform = verifiedaccessEndpointSseSpecificationToTerraform;
function verifiedaccessEndpointSseSpecificationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        customer_managed_key_enabled: {
            value: cdktf.booleanToHclTerraform(struct.customerManagedKeyEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kms_key_arn: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessEndpointSseSpecificationToHclTerraform = verifiedaccessEndpointSseSpecificationToHclTerraform;
class VerifiedaccessEndpointSseSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customerManagedKeyEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.customerManagedKeyEnabled = this._customerManagedKeyEnabled;
        }
        if (this._kmsKeyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyArn = this._kmsKeyArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customerManagedKeyEnabled = undefined;
            this._kmsKeyArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customerManagedKeyEnabled = value.customerManagedKeyEnabled;
            this._kmsKeyArn = value.kmsKeyArn;
        }
    }
    get customerManagedKeyEnabled() {
        return this.getBooleanAttribute('customer_managed_key_enabled');
    }
    set customerManagedKeyEnabled(value) {
        this._customerManagedKeyEnabled = value;
    }
    resetCustomerManagedKeyEnabled() {
        this._customerManagedKeyEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customerManagedKeyEnabledInput() {
        return this._customerManagedKeyEnabled;
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
}
exports.VerifiedaccessEndpointSseSpecificationOutputReference = VerifiedaccessEndpointSseSpecificationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VerifiedaccessEndpointSseSpecificationOutputReference[_c] = { fqn: "@cdktf/provider-aws.verifiedaccessEndpoint.VerifiedaccessEndpointSseSpecificationOutputReference", version: "19.15.0" };
function verifiedaccessEndpointTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.verifiedaccessEndpointTimeoutsToTerraform = verifiedaccessEndpointTimeoutsToTerraform;
function verifiedaccessEndpointTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessEndpointTimeoutsToHclTerraform = verifiedaccessEndpointTimeoutsToHclTerraform;
class VerifiedaccessEndpointTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VerifiedaccessEndpointTimeoutsOutputReference = VerifiedaccessEndpointTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
VerifiedaccessEndpointTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.verifiedaccessEndpoint.VerifiedaccessEndpointTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_endpoint aws_verifiedaccess_endpoint}
*/
class VerifiedaccessEndpoint extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedaccessEndpoint resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedaccessEndpoint to import
    * @param importFromId The id of the existing VerifiedaccessEndpoint that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_endpoint#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedaccessEndpoint to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedaccess_endpoint", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_endpoint aws_verifiedaccess_endpoint} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedaccessEndpointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedaccess_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // load_balancer_options - computed: false, optional: true, required: false
        this._loadBalancerOptions = new VerifiedaccessEndpointLoadBalancerOptionsOutputReference(this, "load_balancer_options");
        // network_interface_options - computed: false, optional: true, required: false
        this._networkInterfaceOptions = new VerifiedaccessEndpointNetworkInterfaceOptionsOutputReference(this, "network_interface_options");
        // sse_specification - computed: false, optional: true, required: false
        this._sseSpecification = new VerifiedaccessEndpointSseSpecificationOutputReference(this, "sse_specification");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VerifiedaccessEndpointTimeoutsOutputReference(this, "timeouts");
        this._applicationDomain = config.applicationDomain;
        this._attachmentType = config.attachmentType;
        this._description = config.description;
        this._domainCertificateArn = config.domainCertificateArn;
        this._endpointDomainPrefix = config.endpointDomainPrefix;
        this._endpointType = config.endpointType;
        this._id = config.id;
        this._policyDocument = config.policyDocument;
        this._securityGroupIds = config.securityGroupIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._verifiedAccessGroupId = config.verifiedAccessGroupId;
        this._loadBalancerOptions.internalValue = config.loadBalancerOptions;
        this._networkInterfaceOptions.internalValue = config.networkInterfaceOptions;
        this._sseSpecification.internalValue = config.sseSpecification;
        this._timeouts.internalValue = config.timeouts;
    }
    get applicationDomain() {
        return this.getStringAttribute('application_domain');
    }
    set applicationDomain(value) {
        this._applicationDomain = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationDomainInput() {
        return this._applicationDomain;
    }
    get attachmentType() {
        return this.getStringAttribute('attachment_type');
    }
    set attachmentType(value) {
        this._attachmentType = value;
    }
    // Temporarily expose input value. Use with caution.
    get attachmentTypeInput() {
        return this._attachmentType;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // device_validation_domain - computed: true, optional: false, required: false
    get deviceValidationDomain() {
        return this.getStringAttribute('device_validation_domain');
    }
    get domainCertificateArn() {
        return this.getStringAttribute('domain_certificate_arn');
    }
    set domainCertificateArn(value) {
        this._domainCertificateArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainCertificateArnInput() {
        return this._domainCertificateArn;
    }
    // endpoint_domain - computed: true, optional: false, required: false
    get endpointDomain() {
        return this.getStringAttribute('endpoint_domain');
    }
    get endpointDomainPrefix() {
        return this.getStringAttribute('endpoint_domain_prefix');
    }
    set endpointDomainPrefix(value) {
        this._endpointDomainPrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointDomainPrefixInput() {
        return this._endpointDomainPrefix;
    }
    get endpointType() {
        return this.getStringAttribute('endpoint_type');
    }
    set endpointType(value) {
        this._endpointType = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointTypeInput() {
        return this._endpointType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get policyDocument() {
        return this.getStringAttribute('policy_document');
    }
    set policyDocument(value) {
        this._policyDocument = value;
    }
    resetPolicyDocument() {
        this._policyDocument = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyDocumentInput() {
        return this._policyDocument;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get verifiedAccessGroupId() {
        return this.getStringAttribute('verified_access_group_id');
    }
    set verifiedAccessGroupId(value) {
        this._verifiedAccessGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get verifiedAccessGroupIdInput() {
        return this._verifiedAccessGroupId;
    }
    // verified_access_instance_id - computed: true, optional: false, required: false
    get verifiedAccessInstanceId() {
        return this.getStringAttribute('verified_access_instance_id');
    }
    get loadBalancerOptions() {
        return this._loadBalancerOptions;
    }
    putLoadBalancerOptions(value) {
        this._loadBalancerOptions.internalValue = value;
    }
    resetLoadBalancerOptions() {
        this._loadBalancerOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerOptionsInput() {
        return this._loadBalancerOptions.internalValue;
    }
    get networkInterfaceOptions() {
        return this._networkInterfaceOptions;
    }
    putNetworkInterfaceOptions(value) {
        this._networkInterfaceOptions.internalValue = value;
    }
    resetNetworkInterfaceOptions() {
        this._networkInterfaceOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfaceOptionsInput() {
        return this._networkInterfaceOptions.internalValue;
    }
    get sseSpecification() {
        return this._sseSpecification;
    }
    putSseSpecification(value) {
        this._sseSpecification.internalValue = value;
    }
    resetSseSpecification() {
        this._sseSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sseSpecificationInput() {
        return this._sseSpecification.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_domain: cdktf.stringToTerraform(this._applicationDomain),
            attachment_type: cdktf.stringToTerraform(this._attachmentType),
            description: cdktf.stringToTerraform(this._description),
            domain_certificate_arn: cdktf.stringToTerraform(this._domainCertificateArn),
            endpoint_domain_prefix: cdktf.stringToTerraform(this._endpointDomainPrefix),
            endpoint_type: cdktf.stringToTerraform(this._endpointType),
            id: cdktf.stringToTerraform(this._id),
            policy_document: cdktf.stringToTerraform(this._policyDocument),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            verified_access_group_id: cdktf.stringToTerraform(this._verifiedAccessGroupId),
            load_balancer_options: verifiedaccessEndpointLoadBalancerOptionsToTerraform(this._loadBalancerOptions.internalValue),
            network_interface_options: verifiedaccessEndpointNetworkInterfaceOptionsToTerraform(this._networkInterfaceOptions.internalValue),
            sse_specification: verifiedaccessEndpointSseSpecificationToTerraform(this._sseSpecification.internalValue),
            timeouts: verifiedaccessEndpointTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_domain: {
                value: cdktf.stringToHclTerraform(this._applicationDomain),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            attachment_type: {
                value: cdktf.stringToHclTerraform(this._attachmentType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            domain_certificate_arn: {
                value: cdktf.stringToHclTerraform(this._domainCertificateArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_domain_prefix: {
                value: cdktf.stringToHclTerraform(this._endpointDomainPrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_type: {
                value: cdktf.stringToHclTerraform(this._endpointType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy_document: {
                value: cdktf.stringToHclTerraform(this._policyDocument),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            security_group_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._securityGroupIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            verified_access_group_id: {
                value: cdktf.stringToHclTerraform(this._verifiedAccessGroupId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            load_balancer_options: {
                value: verifiedaccessEndpointLoadBalancerOptionsToHclTerraform(this._loadBalancerOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessEndpointLoadBalancerOptionsList",
            },
            network_interface_options: {
                value: verifiedaccessEndpointNetworkInterfaceOptionsToHclTerraform(this._networkInterfaceOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessEndpointNetworkInterfaceOptionsList",
            },
            sse_specification: {
                value: verifiedaccessEndpointSseSpecificationToHclTerraform(this._sseSpecification.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessEndpointSseSpecificationList",
            },
            timeouts: {
                value: verifiedaccessEndpointTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VerifiedaccessEndpointTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedaccessEndpoint = VerifiedaccessEndpoint;
_e = JSII_RTTI_SYMBOL_1;
VerifiedaccessEndpoint[_e] = { fqn: "@cdktf/provider-aws.verifiedaccessEndpoint.VerifiedaccessEndpoint", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedaccessEndpoint.tfResourceType = "aws_verifiedaccess_endpoint";
