// @cdktf/provider-aws/lib/auditmanager-framework/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditmanagerFramework = exports.AuditmanagerFrameworkControlSetsList = exports.AuditmanagerFrameworkControlSetsOutputReference = exports.auditmanagerFrameworkControlSetsToHclTerraform = exports.auditmanagerFrameworkControlSetsToTerraform = exports.AuditmanagerFrameworkControlSetsControlsList = exports.AuditmanagerFrameworkControlSetsControlsOutputReference = exports.auditmanagerFrameworkControlSetsControlsToHclTerraform = exports.auditmanagerFrameworkControlSetsControlsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function auditmanagerFrameworkControlSetsControlsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
    };
}
exports.auditmanagerFrameworkControlSetsControlsToTerraform = auditmanagerFrameworkControlSetsControlsToTerraform;
function auditmanagerFrameworkControlSetsControlsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerFrameworkControlSetsControlsToHclTerraform = auditmanagerFrameworkControlSetsControlsToHclTerraform;
class AuditmanagerFrameworkControlSetsControlsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
}
exports.AuditmanagerFrameworkControlSetsControlsOutputReference = AuditmanagerFrameworkControlSetsControlsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AuditmanagerFrameworkControlSetsControlsOutputReference[_a] = { fqn: "@cdktf/provider-aws.auditmanagerFramework.AuditmanagerFrameworkControlSetsControlsOutputReference", version: "19.15.0" };
class AuditmanagerFrameworkControlSetsControlsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerFrameworkControlSetsControlsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerFrameworkControlSetsControlsList = AuditmanagerFrameworkControlSetsControlsList;
_b = JSII_RTTI_SYMBOL_1;
AuditmanagerFrameworkControlSetsControlsList[_b] = { fqn: "@cdktf/provider-aws.auditmanagerFramework.AuditmanagerFrameworkControlSetsControlsList", version: "19.15.0" };
function auditmanagerFrameworkControlSetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        controls: cdktf.listMapper(auditmanagerFrameworkControlSetsControlsToTerraform, true)(struct.controls),
    };
}
exports.auditmanagerFrameworkControlSetsToTerraform = auditmanagerFrameworkControlSetsToTerraform;
function auditmanagerFrameworkControlSetsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        controls: {
            value: cdktf.listMapperHcl(auditmanagerFrameworkControlSetsControlsToHclTerraform, true)(struct.controls),
            isBlock: true,
            type: "set",
            storageClassType: "AuditmanagerFrameworkControlSetsControlsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerFrameworkControlSetsToHclTerraform = auditmanagerFrameworkControlSetsToHclTerraform;
class AuditmanagerFrameworkControlSetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // controls - computed: false, optional: true, required: false
        this._controls = new AuditmanagerFrameworkControlSetsControlsList(this, "controls", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._controls?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.controls = this._controls?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._controls.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._controls.internalValue = value.controls;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get controls() {
        return this._controls;
    }
    putControls(value) {
        this._controls.internalValue = value;
    }
    resetControls() {
        this._controls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlsInput() {
        return this._controls.internalValue;
    }
}
exports.AuditmanagerFrameworkControlSetsOutputReference = AuditmanagerFrameworkControlSetsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AuditmanagerFrameworkControlSetsOutputReference[_c] = { fqn: "@cdktf/provider-aws.auditmanagerFramework.AuditmanagerFrameworkControlSetsOutputReference", version: "19.15.0" };
class AuditmanagerFrameworkControlSetsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerFrameworkControlSetsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerFrameworkControlSetsList = AuditmanagerFrameworkControlSetsList;
_d = JSII_RTTI_SYMBOL_1;
AuditmanagerFrameworkControlSetsList[_d] = { fqn: "@cdktf/provider-aws.auditmanagerFramework.AuditmanagerFrameworkControlSetsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework aws_auditmanager_framework}
*/
class AuditmanagerFramework extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AuditmanagerFramework resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AuditmanagerFramework to import
    * @param importFromId The id of the existing AuditmanagerFramework that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AuditmanagerFramework to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_auditmanager_framework", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework aws_auditmanager_framework} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AuditmanagerFrameworkConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_auditmanager_framework',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // control_sets - computed: false, optional: true, required: false
        this._controlSets = new AuditmanagerFrameworkControlSetsList(this, "control_sets", true);
        this._complianceType = config.complianceType;
        this._description = config.description;
        this._name = config.name;
        this._tags = config.tags;
        this._controlSets.internalValue = config.controlSets;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get complianceType() {
        return this.getStringAttribute('compliance_type');
    }
    set complianceType(value) {
        this._complianceType = value;
    }
    resetComplianceType() {
        this._complianceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get complianceTypeInput() {
        return this._complianceType;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // framework_type - computed: true, optional: false, required: false
    get frameworkType() {
        return this.getStringAttribute('framework_type');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get controlSets() {
        return this._controlSets;
    }
    putControlSets(value) {
        this._controlSets.internalValue = value;
    }
    resetControlSets() {
        this._controlSets.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlSetsInput() {
        return this._controlSets.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            compliance_type: cdktf.stringToTerraform(this._complianceType),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            control_sets: cdktf.listMapper(auditmanagerFrameworkControlSetsToTerraform, true)(this._controlSets.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            compliance_type: {
                value: cdktf.stringToHclTerraform(this._complianceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            control_sets: {
                value: cdktf.listMapperHcl(auditmanagerFrameworkControlSetsToHclTerraform, true)(this._controlSets.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "AuditmanagerFrameworkControlSetsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AuditmanagerFramework = AuditmanagerFramework;
_e = JSII_RTTI_SYMBOL_1;
AuditmanagerFramework[_e] = { fqn: "@cdktf/provider-aws.auditmanagerFramework.AuditmanagerFramework", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AuditmanagerFramework.tfResourceType = "aws_auditmanager_framework";
