// @cdktf/provider-aws/lib/datasync-location-azure-blob/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncLocationAzureBlob = exports.DatasyncLocationAzureBlobSasConfigurationOutputReference = exports.datasyncLocationAzureBlobSasConfigurationToHclTerraform = exports.datasyncLocationAzureBlobSasConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncLocationAzureBlobSasConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        token: cdktf.stringToTerraform(struct.token),
    };
}
exports.datasyncLocationAzureBlobSasConfigurationToTerraform = datasyncLocationAzureBlobSasConfigurationToTerraform;
function datasyncLocationAzureBlobSasConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        token: {
            value: cdktf.stringToHclTerraform(struct.token),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncLocationAzureBlobSasConfigurationToHclTerraform = datasyncLocationAzureBlobSasConfigurationToHclTerraform;
class DatasyncLocationAzureBlobSasConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._token !== undefined) {
            hasAnyValues = true;
            internalValueResult.token = this._token;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._token = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._token = value.token;
        }
    }
    get token() {
        return this.getStringAttribute('token');
    }
    set token(value) {
        this._token = value;
    }
    // Temporarily expose input value. Use with caution.
    get tokenInput() {
        return this._token;
    }
}
exports.DatasyncLocationAzureBlobSasConfigurationOutputReference = DatasyncLocationAzureBlobSasConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncLocationAzureBlobSasConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasyncLocationAzureBlob.DatasyncLocationAzureBlobSasConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_azure_blob aws_datasync_location_azure_blob}
*/
class DatasyncLocationAzureBlob extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DatasyncLocationAzureBlob resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DatasyncLocationAzureBlob to import
    * @param importFromId The id of the existing DatasyncLocationAzureBlob that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_azure_blob#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DatasyncLocationAzureBlob to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_datasync_location_azure_blob", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_azure_blob aws_datasync_location_azure_blob} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DatasyncLocationAzureBlobConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_location_azure_blob',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // sas_configuration - computed: false, optional: true, required: false
        this._sasConfiguration = new DatasyncLocationAzureBlobSasConfigurationOutputReference(this, "sas_configuration");
        this._accessTier = config.accessTier;
        this._agentArns = config.agentArns;
        this._authenticationType = config.authenticationType;
        this._blobType = config.blobType;
        this._containerUrl = config.containerUrl;
        this._id = config.id;
        this._subdirectory = config.subdirectory;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._sasConfiguration.internalValue = config.sasConfiguration;
    }
    get accessTier() {
        return this.getStringAttribute('access_tier');
    }
    set accessTier(value) {
        this._accessTier = value;
    }
    resetAccessTier() {
        this._accessTier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessTierInput() {
        return this._accessTier;
    }
    get agentArns() {
        return cdktf.Fn.tolist(this.getListAttribute('agent_arns'));
    }
    set agentArns(value) {
        this._agentArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get agentArnsInput() {
        return this._agentArns;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authenticationType() {
        return this.getStringAttribute('authentication_type');
    }
    set authenticationType(value) {
        this._authenticationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationTypeInput() {
        return this._authenticationType;
    }
    get blobType() {
        return this.getStringAttribute('blob_type');
    }
    set blobType(value) {
        this._blobType = value;
    }
    resetBlobType() {
        this._blobType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blobTypeInput() {
        return this._blobType;
    }
    get containerUrl() {
        return this.getStringAttribute('container_url');
    }
    set containerUrl(value) {
        this._containerUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerUrlInput() {
        return this._containerUrl;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get subdirectory() {
        return this.getStringAttribute('subdirectory');
    }
    set subdirectory(value) {
        this._subdirectory = value;
    }
    resetSubdirectory() {
        this._subdirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subdirectoryInput() {
        return this._subdirectory;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
    get sasConfiguration() {
        return this._sasConfiguration;
    }
    putSasConfiguration(value) {
        this._sasConfiguration.internalValue = value;
    }
    resetSasConfiguration() {
        this._sasConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sasConfigurationInput() {
        return this._sasConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_tier: cdktf.stringToTerraform(this._accessTier),
            agent_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._agentArns),
            authentication_type: cdktf.stringToTerraform(this._authenticationType),
            blob_type: cdktf.stringToTerraform(this._blobType),
            container_url: cdktf.stringToTerraform(this._containerUrl),
            id: cdktf.stringToTerraform(this._id),
            subdirectory: cdktf.stringToTerraform(this._subdirectory),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            sas_configuration: datasyncLocationAzureBlobSasConfigurationToTerraform(this._sasConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_tier: {
                value: cdktf.stringToHclTerraform(this._accessTier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            agent_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._agentArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            authentication_type: {
                value: cdktf.stringToHclTerraform(this._authenticationType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            blob_type: {
                value: cdktf.stringToHclTerraform(this._blobType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            container_url: {
                value: cdktf.stringToHclTerraform(this._containerUrl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subdirectory: {
                value: cdktf.stringToHclTerraform(this._subdirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            sas_configuration: {
                value: datasyncLocationAzureBlobSasConfigurationToHclTerraform(this._sasConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncLocationAzureBlobSasConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DatasyncLocationAzureBlob = DatasyncLocationAzureBlob;
_b = JSII_RTTI_SYMBOL_1;
DatasyncLocationAzureBlob[_b] = { fqn: "@cdktf/provider-aws.datasyncLocationAzureBlob.DatasyncLocationAzureBlob", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DatasyncLocationAzureBlob.tfResourceType = "aws_datasync_location_azure_blob";
