// @cdktf/provider-aws/lib/s3-bucket-notification/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketNotification = exports.S3BucketNotificationTopicList = exports.S3BucketNotificationTopicOutputReference = exports.s3BucketNotificationTopicToHclTerraform = exports.s3BucketNotificationTopicToTerraform = exports.S3BucketNotificationQueueList = exports.S3BucketNotificationQueueOutputReference = exports.s3BucketNotificationQueueToHclTerraform = exports.s3BucketNotificationQueueToTerraform = exports.S3BucketNotificationLambdaFunctionList = exports.S3BucketNotificationLambdaFunctionOutputReference = exports.s3BucketNotificationLambdaFunctionToHclTerraform = exports.s3BucketNotificationLambdaFunctionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketNotificationLambdaFunctionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
        filter_prefix: cdktf.stringToTerraform(struct.filterPrefix),
        filter_suffix: cdktf.stringToTerraform(struct.filterSuffix),
        id: cdktf.stringToTerraform(struct.id),
        lambda_function_arn: cdktf.stringToTerraform(struct.lambdaFunctionArn),
    };
}
exports.s3BucketNotificationLambdaFunctionToTerraform = s3BucketNotificationLambdaFunctionToTerraform;
function s3BucketNotificationLambdaFunctionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        filter_prefix: {
            value: cdktf.stringToHclTerraform(struct.filterPrefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        filter_suffix: {
            value: cdktf.stringToHclTerraform(struct.filterSuffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        lambda_function_arn: {
            value: cdktf.stringToHclTerraform(struct.lambdaFunctionArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketNotificationLambdaFunctionToHclTerraform = s3BucketNotificationLambdaFunctionToHclTerraform;
class S3BucketNotificationLambdaFunctionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        if (this._filterPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterPrefix = this._filterPrefix;
        }
        if (this._filterSuffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterSuffix = this._filterSuffix;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._lambdaFunctionArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambdaFunctionArn = this._lambdaFunctionArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._events = undefined;
            this._filterPrefix = undefined;
            this._filterSuffix = undefined;
            this._id = undefined;
            this._lambdaFunctionArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._events = value.events;
            this._filterPrefix = value.filterPrefix;
            this._filterSuffix = value.filterSuffix;
            this._id = value.id;
            this._lambdaFunctionArn = value.lambdaFunctionArn;
        }
    }
    get events() {
        return cdktf.Fn.tolist(this.getListAttribute('events'));
    }
    set events(value) {
        this._events = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
    get filterPrefix() {
        return this.getStringAttribute('filter_prefix');
    }
    set filterPrefix(value) {
        this._filterPrefix = value;
    }
    resetFilterPrefix() {
        this._filterPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterPrefixInput() {
        return this._filterPrefix;
    }
    get filterSuffix() {
        return this.getStringAttribute('filter_suffix');
    }
    set filterSuffix(value) {
        this._filterSuffix = value;
    }
    resetFilterSuffix() {
        this._filterSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterSuffixInput() {
        return this._filterSuffix;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lambdaFunctionArn() {
        return this.getStringAttribute('lambda_function_arn');
    }
    set lambdaFunctionArn(value) {
        this._lambdaFunctionArn = value;
    }
    resetLambdaFunctionArn() {
        this._lambdaFunctionArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaFunctionArnInput() {
        return this._lambdaFunctionArn;
    }
}
exports.S3BucketNotificationLambdaFunctionOutputReference = S3BucketNotificationLambdaFunctionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketNotificationLambdaFunctionOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationLambdaFunctionOutputReference", version: "19.15.0" };
class S3BucketNotificationLambdaFunctionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketNotificationLambdaFunctionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketNotificationLambdaFunctionList = S3BucketNotificationLambdaFunctionList;
_b = JSII_RTTI_SYMBOL_1;
S3BucketNotificationLambdaFunctionList[_b] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationLambdaFunctionList", version: "19.15.0" };
function s3BucketNotificationQueueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
        filter_prefix: cdktf.stringToTerraform(struct.filterPrefix),
        filter_suffix: cdktf.stringToTerraform(struct.filterSuffix),
        id: cdktf.stringToTerraform(struct.id),
        queue_arn: cdktf.stringToTerraform(struct.queueArn),
    };
}
exports.s3BucketNotificationQueueToTerraform = s3BucketNotificationQueueToTerraform;
function s3BucketNotificationQueueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        filter_prefix: {
            value: cdktf.stringToHclTerraform(struct.filterPrefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        filter_suffix: {
            value: cdktf.stringToHclTerraform(struct.filterSuffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        queue_arn: {
            value: cdktf.stringToHclTerraform(struct.queueArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketNotificationQueueToHclTerraform = s3BucketNotificationQueueToHclTerraform;
class S3BucketNotificationQueueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        if (this._filterPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterPrefix = this._filterPrefix;
        }
        if (this._filterSuffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterSuffix = this._filterSuffix;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._queueArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.queueArn = this._queueArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._events = undefined;
            this._filterPrefix = undefined;
            this._filterSuffix = undefined;
            this._id = undefined;
            this._queueArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._events = value.events;
            this._filterPrefix = value.filterPrefix;
            this._filterSuffix = value.filterSuffix;
            this._id = value.id;
            this._queueArn = value.queueArn;
        }
    }
    get events() {
        return cdktf.Fn.tolist(this.getListAttribute('events'));
    }
    set events(value) {
        this._events = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
    get filterPrefix() {
        return this.getStringAttribute('filter_prefix');
    }
    set filterPrefix(value) {
        this._filterPrefix = value;
    }
    resetFilterPrefix() {
        this._filterPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterPrefixInput() {
        return this._filterPrefix;
    }
    get filterSuffix() {
        return this.getStringAttribute('filter_suffix');
    }
    set filterSuffix(value) {
        this._filterSuffix = value;
    }
    resetFilterSuffix() {
        this._filterSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterSuffixInput() {
        return this._filterSuffix;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get queueArn() {
        return this.getStringAttribute('queue_arn');
    }
    set queueArn(value) {
        this._queueArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get queueArnInput() {
        return this._queueArn;
    }
}
exports.S3BucketNotificationQueueOutputReference = S3BucketNotificationQueueOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3BucketNotificationQueueOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationQueueOutputReference", version: "19.15.0" };
class S3BucketNotificationQueueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketNotificationQueueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketNotificationQueueList = S3BucketNotificationQueueList;
_d = JSII_RTTI_SYMBOL_1;
S3BucketNotificationQueueList[_d] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationQueueList", version: "19.15.0" };
function s3BucketNotificationTopicToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.events),
        filter_prefix: cdktf.stringToTerraform(struct.filterPrefix),
        filter_suffix: cdktf.stringToTerraform(struct.filterSuffix),
        id: cdktf.stringToTerraform(struct.id),
        topic_arn: cdktf.stringToTerraform(struct.topicArn),
    };
}
exports.s3BucketNotificationTopicToTerraform = s3BucketNotificationTopicToTerraform;
function s3BucketNotificationTopicToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.events),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        filter_prefix: {
            value: cdktf.stringToHclTerraform(struct.filterPrefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        filter_suffix: {
            value: cdktf.stringToHclTerraform(struct.filterSuffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        topic_arn: {
            value: cdktf.stringToHclTerraform(struct.topicArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketNotificationTopicToHclTerraform = s3BucketNotificationTopicToHclTerraform;
class S3BucketNotificationTopicOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._events !== undefined) {
            hasAnyValues = true;
            internalValueResult.events = this._events;
        }
        if (this._filterPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterPrefix = this._filterPrefix;
        }
        if (this._filterSuffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterSuffix = this._filterSuffix;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._topicArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.topicArn = this._topicArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._events = undefined;
            this._filterPrefix = undefined;
            this._filterSuffix = undefined;
            this._id = undefined;
            this._topicArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._events = value.events;
            this._filterPrefix = value.filterPrefix;
            this._filterSuffix = value.filterSuffix;
            this._id = value.id;
            this._topicArn = value.topicArn;
        }
    }
    get events() {
        return cdktf.Fn.tolist(this.getListAttribute('events'));
    }
    set events(value) {
        this._events = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventsInput() {
        return this._events;
    }
    get filterPrefix() {
        return this.getStringAttribute('filter_prefix');
    }
    set filterPrefix(value) {
        this._filterPrefix = value;
    }
    resetFilterPrefix() {
        this._filterPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterPrefixInput() {
        return this._filterPrefix;
    }
    get filterSuffix() {
        return this.getStringAttribute('filter_suffix');
    }
    set filterSuffix(value) {
        this._filterSuffix = value;
    }
    resetFilterSuffix() {
        this._filterSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterSuffixInput() {
        return this._filterSuffix;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
}
exports.S3BucketNotificationTopicOutputReference = S3BucketNotificationTopicOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketNotificationTopicOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationTopicOutputReference", version: "19.15.0" };
class S3BucketNotificationTopicList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketNotificationTopicOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketNotificationTopicList = S3BucketNotificationTopicList;
_f = JSII_RTTI_SYMBOL_1;
S3BucketNotificationTopicList[_f] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotificationTopicList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_notification aws_s3_bucket_notification}
*/
class S3BucketNotification extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a S3BucketNotification resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the S3BucketNotification to import
    * @param importFromId The id of the existing S3BucketNotification that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_notification#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the S3BucketNotification to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_s3_bucket_notification", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_notification aws_s3_bucket_notification} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketNotificationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_notification',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // lambda_function - computed: false, optional: true, required: false
        this._lambdaFunction = new S3BucketNotificationLambdaFunctionList(this, "lambda_function", false);
        // queue - computed: false, optional: true, required: false
        this._queue = new S3BucketNotificationQueueList(this, "queue", false);
        // topic - computed: false, optional: true, required: false
        this._topic = new S3BucketNotificationTopicList(this, "topic", false);
        this._bucket = config.bucket;
        this._eventbridge = config.eventbridge;
        this._id = config.id;
        this._lambdaFunction.internalValue = config.lambdaFunction;
        this._queue.internalValue = config.queue;
        this._topic.internalValue = config.topic;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get eventbridge() {
        return this.getBooleanAttribute('eventbridge');
    }
    set eventbridge(value) {
        this._eventbridge = value;
    }
    resetEventbridge() {
        this._eventbridge = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventbridgeInput() {
        return this._eventbridge;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lambdaFunction() {
        return this._lambdaFunction;
    }
    putLambdaFunction(value) {
        this._lambdaFunction.internalValue = value;
    }
    resetLambdaFunction() {
        this._lambdaFunction.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaFunctionInput() {
        return this._lambdaFunction.internalValue;
    }
    get queue() {
        return this._queue;
    }
    putQueue(value) {
        this._queue.internalValue = value;
    }
    resetQueue() {
        this._queue.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queueInput() {
        return this._queue.internalValue;
    }
    get topic() {
        return this._topic;
    }
    putTopic(value) {
        this._topic.internalValue = value;
    }
    resetTopic() {
        this._topic.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get topicInput() {
        return this._topic.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            eventbridge: cdktf.booleanToTerraform(this._eventbridge),
            id: cdktf.stringToTerraform(this._id),
            lambda_function: cdktf.listMapper(s3BucketNotificationLambdaFunctionToTerraform, true)(this._lambdaFunction.internalValue),
            queue: cdktf.listMapper(s3BucketNotificationQueueToTerraform, true)(this._queue.internalValue),
            topic: cdktf.listMapper(s3BucketNotificationTopicToTerraform, true)(this._topic.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bucket: {
                value: cdktf.stringToHclTerraform(this._bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            eventbridge: {
                value: cdktf.booleanToHclTerraform(this._eventbridge),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            lambda_function: {
                value: cdktf.listMapperHcl(s3BucketNotificationLambdaFunctionToHclTerraform, true)(this._lambdaFunction.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3BucketNotificationLambdaFunctionList",
            },
            queue: {
                value: cdktf.listMapperHcl(s3BucketNotificationQueueToHclTerraform, true)(this._queue.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3BucketNotificationQueueList",
            },
            topic: {
                value: cdktf.listMapperHcl(s3BucketNotificationTopicToHclTerraform, true)(this._topic.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3BucketNotificationTopicList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.S3BucketNotification = S3BucketNotification;
_g = JSII_RTTI_SYMBOL_1;
S3BucketNotification[_g] = { fqn: "@cdktf/provider-aws.s3BucketNotification.S3BucketNotification", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
S3BucketNotification.tfResourceType = "aws_s3_bucket_notification";
