// @cdktf/provider-aws/lib/grafana-workspace/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrafanaWorkspace = exports.GrafanaWorkspaceVpcConfigurationOutputReference = exports.grafanaWorkspaceVpcConfigurationToHclTerraform = exports.grafanaWorkspaceVpcConfigurationToTerraform = exports.GrafanaWorkspaceTimeoutsOutputReference = exports.grafanaWorkspaceTimeoutsToHclTerraform = exports.grafanaWorkspaceTimeoutsToTerraform = exports.GrafanaWorkspaceNetworkAccessControlOutputReference = exports.grafanaWorkspaceNetworkAccessControlToHclTerraform = exports.grafanaWorkspaceNetworkAccessControlToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function grafanaWorkspaceNetworkAccessControlToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix_list_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.prefixListIds),
        vpce_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.vpceIds),
    };
}
exports.grafanaWorkspaceNetworkAccessControlToTerraform = grafanaWorkspaceNetworkAccessControlToTerraform;
function grafanaWorkspaceNetworkAccessControlToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        prefix_list_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.prefixListIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        vpce_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.vpceIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.grafanaWorkspaceNetworkAccessControlToHclTerraform = grafanaWorkspaceNetworkAccessControlToHclTerraform;
class GrafanaWorkspaceNetworkAccessControlOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefixListIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefixListIds = this._prefixListIds;
        }
        if (this._vpceIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpceIds = this._vpceIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefixListIds = undefined;
            this._vpceIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefixListIds = value.prefixListIds;
            this._vpceIds = value.vpceIds;
        }
    }
    get prefixListIds() {
        return cdktf.Fn.tolist(this.getListAttribute('prefix_list_ids'));
    }
    set prefixListIds(value) {
        this._prefixListIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get prefixListIdsInput() {
        return this._prefixListIds;
    }
    get vpceIds() {
        return cdktf.Fn.tolist(this.getListAttribute('vpce_ids'));
    }
    set vpceIds(value) {
        this._vpceIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpceIdsInput() {
        return this._vpceIds;
    }
}
exports.GrafanaWorkspaceNetworkAccessControlOutputReference = GrafanaWorkspaceNetworkAccessControlOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GrafanaWorkspaceNetworkAccessControlOutputReference[_a] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspaceNetworkAccessControlOutputReference", version: "19.15.0" };
function grafanaWorkspaceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.grafanaWorkspaceTimeoutsToTerraform = grafanaWorkspaceTimeoutsToTerraform;
function grafanaWorkspaceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.grafanaWorkspaceTimeoutsToHclTerraform = grafanaWorkspaceTimeoutsToHclTerraform;
class GrafanaWorkspaceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GrafanaWorkspaceTimeoutsOutputReference = GrafanaWorkspaceTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GrafanaWorkspaceTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspaceTimeoutsOutputReference", version: "19.15.0" };
function grafanaWorkspaceVpcConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.grafanaWorkspaceVpcConfigurationToTerraform = grafanaWorkspaceVpcConfigurationToTerraform;
function grafanaWorkspaceVpcConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.grafanaWorkspaceVpcConfigurationToHclTerraform = grafanaWorkspaceVpcConfigurationToHclTerraform;
class GrafanaWorkspaceVpcConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.GrafanaWorkspaceVpcConfigurationOutputReference = GrafanaWorkspaceVpcConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GrafanaWorkspaceVpcConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspaceVpcConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/grafana_workspace aws_grafana_workspace}
*/
class GrafanaWorkspace extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GrafanaWorkspace resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GrafanaWorkspace to import
    * @param importFromId The id of the existing GrafanaWorkspace that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/grafana_workspace#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GrafanaWorkspace to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_grafana_workspace", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/grafana_workspace aws_grafana_workspace} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GrafanaWorkspaceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_grafana_workspace',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // network_access_control - computed: false, optional: true, required: false
        this._networkAccessControl = new GrafanaWorkspaceNetworkAccessControlOutputReference(this, "network_access_control");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GrafanaWorkspaceTimeoutsOutputReference(this, "timeouts");
        // vpc_configuration - computed: false, optional: true, required: false
        this._vpcConfiguration = new GrafanaWorkspaceVpcConfigurationOutputReference(this, "vpc_configuration");
        this._accountAccessType = config.accountAccessType;
        this._authenticationProviders = config.authenticationProviders;
        this._configuration = config.configuration;
        this._dataSources = config.dataSources;
        this._description = config.description;
        this._grafanaVersion = config.grafanaVersion;
        this._id = config.id;
        this._name = config.name;
        this._notificationDestinations = config.notificationDestinations;
        this._organizationRoleName = config.organizationRoleName;
        this._organizationalUnits = config.organizationalUnits;
        this._permissionType = config.permissionType;
        this._roleArn = config.roleArn;
        this._stackSetName = config.stackSetName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._networkAccessControl.internalValue = config.networkAccessControl;
        this._timeouts.internalValue = config.timeouts;
        this._vpcConfiguration.internalValue = config.vpcConfiguration;
    }
    get accountAccessType() {
        return this.getStringAttribute('account_access_type');
    }
    set accountAccessType(value) {
        this._accountAccessType = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountAccessTypeInput() {
        return this._accountAccessType;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authenticationProviders() {
        return this.getListAttribute('authentication_providers');
    }
    set authenticationProviders(value) {
        this._authenticationProviders = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationProvidersInput() {
        return this._authenticationProviders;
    }
    get configuration() {
        return this.getStringAttribute('configuration');
    }
    set configuration(value) {
        this._configuration = value;
    }
    resetConfiguration() {
        this._configuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration;
    }
    get dataSources() {
        return this.getListAttribute('data_sources');
    }
    set dataSources(value) {
        this._dataSources = value;
    }
    resetDataSources() {
        this._dataSources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataSourcesInput() {
        return this._dataSources;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get grafanaVersion() {
        return this.getStringAttribute('grafana_version');
    }
    set grafanaVersion(value) {
        this._grafanaVersion = value;
    }
    resetGrafanaVersion() {
        this._grafanaVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grafanaVersionInput() {
        return this._grafanaVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get notificationDestinations() {
        return this.getListAttribute('notification_destinations');
    }
    set notificationDestinations(value) {
        this._notificationDestinations = value;
    }
    resetNotificationDestinations() {
        this._notificationDestinations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationDestinationsInput() {
        return this._notificationDestinations;
    }
    get organizationRoleName() {
        return this.getStringAttribute('organization_role_name');
    }
    set organizationRoleName(value) {
        this._organizationRoleName = value;
    }
    resetOrganizationRoleName() {
        this._organizationRoleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationRoleNameInput() {
        return this._organizationRoleName;
    }
    get organizationalUnits() {
        return this.getListAttribute('organizational_units');
    }
    set organizationalUnits(value) {
        this._organizationalUnits = value;
    }
    resetOrganizationalUnits() {
        this._organizationalUnits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitsInput() {
        return this._organizationalUnits;
    }
    get permissionType() {
        return this.getStringAttribute('permission_type');
    }
    set permissionType(value) {
        this._permissionType = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionTypeInput() {
        return this._permissionType;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    // saml_configuration_status - computed: true, optional: false, required: false
    get samlConfigurationStatus() {
        return this.getStringAttribute('saml_configuration_status');
    }
    get stackSetName() {
        return this.getStringAttribute('stack_set_name');
    }
    set stackSetName(value) {
        this._stackSetName = value;
    }
    resetStackSetName() {
        this._stackSetName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stackSetNameInput() {
        return this._stackSetName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get networkAccessControl() {
        return this._networkAccessControl;
    }
    putNetworkAccessControl(value) {
        this._networkAccessControl.internalValue = value;
    }
    resetNetworkAccessControl() {
        this._networkAccessControl.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkAccessControlInput() {
        return this._networkAccessControl.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcConfiguration() {
        return this._vpcConfiguration;
    }
    putVpcConfiguration(value) {
        this._vpcConfiguration.internalValue = value;
    }
    resetVpcConfiguration() {
        this._vpcConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigurationInput() {
        return this._vpcConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_access_type: cdktf.stringToTerraform(this._accountAccessType),
            authentication_providers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._authenticationProviders),
            configuration: cdktf.stringToTerraform(this._configuration),
            data_sources: cdktf.listMapper(cdktf.stringToTerraform, false)(this._dataSources),
            description: cdktf.stringToTerraform(this._description),
            grafana_version: cdktf.stringToTerraform(this._grafanaVersion),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            notification_destinations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._notificationDestinations),
            organization_role_name: cdktf.stringToTerraform(this._organizationRoleName),
            organizational_units: cdktf.listMapper(cdktf.stringToTerraform, false)(this._organizationalUnits),
            permission_type: cdktf.stringToTerraform(this._permissionType),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            stack_set_name: cdktf.stringToTerraform(this._stackSetName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            network_access_control: grafanaWorkspaceNetworkAccessControlToTerraform(this._networkAccessControl.internalValue),
            timeouts: grafanaWorkspaceTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_configuration: grafanaWorkspaceVpcConfigurationToTerraform(this._vpcConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            account_access_type: {
                value: cdktf.stringToHclTerraform(this._accountAccessType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authentication_providers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._authenticationProviders),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            configuration: {
                value: cdktf.stringToHclTerraform(this._configuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_sources: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._dataSources),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            grafana_version: {
                value: cdktf.stringToHclTerraform(this._grafanaVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notification_destinations: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._notificationDestinations),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            organization_role_name: {
                value: cdktf.stringToHclTerraform(this._organizationRoleName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organizational_units: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._organizationalUnits),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            permission_type: {
                value: cdktf.stringToHclTerraform(this._permissionType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            stack_set_name: {
                value: cdktf.stringToHclTerraform(this._stackSetName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            network_access_control: {
                value: grafanaWorkspaceNetworkAccessControlToHclTerraform(this._networkAccessControl.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GrafanaWorkspaceNetworkAccessControlList",
            },
            timeouts: {
                value: grafanaWorkspaceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "GrafanaWorkspaceTimeouts",
            },
            vpc_configuration: {
                value: grafanaWorkspaceVpcConfigurationToHclTerraform(this._vpcConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GrafanaWorkspaceVpcConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GrafanaWorkspace = GrafanaWorkspace;
_d = JSII_RTTI_SYMBOL_1;
GrafanaWorkspace[_d] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspace", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GrafanaWorkspace.tfResourceType = "aws_grafana_workspace";
