// @cdktf/provider-aws/lib/eks-addon/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksAddon = exports.EksAddonTimeoutsOutputReference = exports.eksAddonTimeoutsToHclTerraform = exports.eksAddonTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function eksAddonTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.eksAddonTimeoutsToTerraform = eksAddonTimeoutsToTerraform;
function eksAddonTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.eksAddonTimeoutsToHclTerraform = eksAddonTimeoutsToHclTerraform;
class EksAddonTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EksAddonTimeoutsOutputReference = EksAddonTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EksAddonTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.eksAddon.EksAddonTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_addon aws_eks_addon}
*/
class EksAddon extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EksAddon resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EksAddon to import
    * @param importFromId The id of the existing EksAddon that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_addon#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EksAddon to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_eks_addon", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/eks_addon aws_eks_addon} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EksAddonConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_addon',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksAddonTimeoutsOutputReference(this, "timeouts");
        this._addonName = config.addonName;
        this._addonVersion = config.addonVersion;
        this._clusterName = config.clusterName;
        this._configurationValues = config.configurationValues;
        this._id = config.id;
        this._preserve = config.preserve;
        this._resolveConflicts = config.resolveConflicts;
        this._resolveConflictsOnCreate = config.resolveConflictsOnCreate;
        this._resolveConflictsOnUpdate = config.resolveConflictsOnUpdate;
        this._serviceAccountRoleArn = config.serviceAccountRoleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeouts.internalValue = config.timeouts;
    }
    get addonName() {
        return this.getStringAttribute('addon_name');
    }
    set addonName(value) {
        this._addonName = value;
    }
    // Temporarily expose input value. Use with caution.
    get addonNameInput() {
        return this._addonName;
    }
    get addonVersion() {
        return this.getStringAttribute('addon_version');
    }
    set addonVersion(value) {
        this._addonVersion = value;
    }
    resetAddonVersion() {
        this._addonVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addonVersionInput() {
        return this._addonVersion;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    get configurationValues() {
        return this.getStringAttribute('configuration_values');
    }
    set configurationValues(value) {
        this._configurationValues = value;
    }
    resetConfigurationValues() {
        this._configurationValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationValuesInput() {
        return this._configurationValues;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // modified_at - computed: true, optional: false, required: false
    get modifiedAt() {
        return this.getStringAttribute('modified_at');
    }
    get preserve() {
        return this.getBooleanAttribute('preserve');
    }
    set preserve(value) {
        this._preserve = value;
    }
    resetPreserve() {
        this._preserve = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preserveInput() {
        return this._preserve;
    }
    get resolveConflicts() {
        return this.getStringAttribute('resolve_conflicts');
    }
    set resolveConflicts(value) {
        this._resolveConflicts = value;
    }
    resetResolveConflicts() {
        this._resolveConflicts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resolveConflictsInput() {
        return this._resolveConflicts;
    }
    get resolveConflictsOnCreate() {
        return this.getStringAttribute('resolve_conflicts_on_create');
    }
    set resolveConflictsOnCreate(value) {
        this._resolveConflictsOnCreate = value;
    }
    resetResolveConflictsOnCreate() {
        this._resolveConflictsOnCreate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resolveConflictsOnCreateInput() {
        return this._resolveConflictsOnCreate;
    }
    get resolveConflictsOnUpdate() {
        return this.getStringAttribute('resolve_conflicts_on_update');
    }
    set resolveConflictsOnUpdate(value) {
        this._resolveConflictsOnUpdate = value;
    }
    resetResolveConflictsOnUpdate() {
        this._resolveConflictsOnUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resolveConflictsOnUpdateInput() {
        return this._resolveConflictsOnUpdate;
    }
    get serviceAccountRoleArn() {
        return this.getStringAttribute('service_account_role_arn');
    }
    set serviceAccountRoleArn(value) {
        this._serviceAccountRoleArn = value;
    }
    resetServiceAccountRoleArn() {
        this._serviceAccountRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountRoleArnInput() {
        return this._serviceAccountRoleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            addon_name: cdktf.stringToTerraform(this._addonName),
            addon_version: cdktf.stringToTerraform(this._addonVersion),
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            configuration_values: cdktf.stringToTerraform(this._configurationValues),
            id: cdktf.stringToTerraform(this._id),
            preserve: cdktf.booleanToTerraform(this._preserve),
            resolve_conflicts: cdktf.stringToTerraform(this._resolveConflicts),
            resolve_conflicts_on_create: cdktf.stringToTerraform(this._resolveConflictsOnCreate),
            resolve_conflicts_on_update: cdktf.stringToTerraform(this._resolveConflictsOnUpdate),
            service_account_role_arn: cdktf.stringToTerraform(this._serviceAccountRoleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeouts: eksAddonTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            addon_name: {
                value: cdktf.stringToHclTerraform(this._addonName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            addon_version: {
                value: cdktf.stringToHclTerraform(this._addonVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cluster_name: {
                value: cdktf.stringToHclTerraform(this._clusterName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            configuration_values: {
                value: cdktf.stringToHclTerraform(this._configurationValues),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preserve: {
                value: cdktf.booleanToHclTerraform(this._preserve),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            resolve_conflicts: {
                value: cdktf.stringToHclTerraform(this._resolveConflicts),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resolve_conflicts_on_create: {
                value: cdktf.stringToHclTerraform(this._resolveConflictsOnCreate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resolve_conflicts_on_update: {
                value: cdktf.stringToHclTerraform(this._resolveConflictsOnUpdate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_account_role_arn: {
                value: cdktf.stringToHclTerraform(this._serviceAccountRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            timeouts: {
                value: eksAddonTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "EksAddonTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EksAddon = EksAddon;
_b = JSII_RTTI_SYMBOL_1;
EksAddon[_b] = { fqn: "@cdktf/provider-aws.eksAddon.EksAddon", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EksAddon.tfResourceType = "aws_eks_addon";
