// @cdktf/provider-aws/lib/cognito-identity-provider/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CognitoIdentityProvider = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_provider aws_cognito_identity_provider}
*/
class CognitoIdentityProvider extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CognitoIdentityProvider resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CognitoIdentityProvider to import
    * @param importFromId The id of the existing CognitoIdentityProvider that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_provider#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CognitoIdentityProvider to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cognito_identity_provider", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cognito_identity_provider aws_cognito_identity_provider} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CognitoIdentityProviderConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cognito_identity_provider',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._attributeMapping = config.attributeMapping;
        this._id = config.id;
        this._idpIdentifiers = config.idpIdentifiers;
        this._providerDetails = config.providerDetails;
        this._providerName = config.providerName;
        this._providerType = config.providerType;
        this._userPoolId = config.userPoolId;
    }
    get attributeMapping() {
        return this.getStringMapAttribute('attribute_mapping');
    }
    set attributeMapping(value) {
        this._attributeMapping = value;
    }
    resetAttributeMapping() {
        this._attributeMapping = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributeMappingInput() {
        return this._attributeMapping;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get idpIdentifiers() {
        return this.getListAttribute('idp_identifiers');
    }
    set idpIdentifiers(value) {
        this._idpIdentifiers = value;
    }
    resetIdpIdentifiers() {
        this._idpIdentifiers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idpIdentifiersInput() {
        return this._idpIdentifiers;
    }
    get providerDetails() {
        return this.getStringMapAttribute('provider_details');
    }
    set providerDetails(value) {
        this._providerDetails = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerDetailsInput() {
        return this._providerDetails;
    }
    get providerName() {
        return this.getStringAttribute('provider_name');
    }
    set providerName(value) {
        this._providerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerNameInput() {
        return this._providerName;
    }
    get providerType() {
        return this.getStringAttribute('provider_type');
    }
    set providerType(value) {
        this._providerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerTypeInput() {
        return this._providerType;
    }
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
    set userPoolId(value) {
        this._userPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userPoolIdInput() {
        return this._userPoolId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            attribute_mapping: cdktf.hashMapper(cdktf.stringToTerraform)(this._attributeMapping),
            id: cdktf.stringToTerraform(this._id),
            idp_identifiers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._idpIdentifiers),
            provider_details: cdktf.hashMapper(cdktf.stringToTerraform)(this._providerDetails),
            provider_name: cdktf.stringToTerraform(this._providerName),
            provider_type: cdktf.stringToTerraform(this._providerType),
            user_pool_id: cdktf.stringToTerraform(this._userPoolId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            attribute_mapping: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._attributeMapping),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            idp_identifiers: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._idpIdentifiers),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            provider_details: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._providerDetails),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            provider_name: {
                value: cdktf.stringToHclTerraform(this._providerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            provider_type: {
                value: cdktf.stringToHclTerraform(this._providerType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_pool_id: {
                value: cdktf.stringToHclTerraform(this._userPoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CognitoIdentityProvider = CognitoIdentityProvider;
_a = JSII_RTTI_SYMBOL_1;
CognitoIdentityProvider[_a] = { fqn: "@cdktf/provider-aws.cognitoIdentityProvider.CognitoIdentityProvider", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CognitoIdentityProvider.tfResourceType = "aws_cognito_identity_provider";
