// @winglang/sdk/lib/simulator/graph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Graph = void 0;
const tokens_1 = require("./tokens");
class Node {
    constructor(def) {
        this.def = def;
        this.dependencies = new Set();
        this.dependents = new Set();
    }
    get path() {
        return this.def.path;
    }
}
class Graph {
    constructor(resources = []) {
        this.byPath = {};
        for (const resource of resources) {
            this.byPath[resource.path] = new Node(resource);
        }
        // build the dependency graph
        for (const resource of resources) {
            const consumer = resource.path;
            // add explicit dependencies
            for (const dep of resource.deps ?? []) {
                this.recordDependency(consumer, dep);
            }
            // add implicit dependencies (e.g. from tokens in props)
            const implicitDeps = [];
            // collect all tokens from the props object (recursive) the "resolver" here is just a dummy
            // function that collects all tokens and returns a dummy value (we don't care about the
            // result).
            (0, tokens_1.resolveTokens)(resource.props ?? {}, (token) => {
                implicitDeps.push(token.path);
                return "[T]"; // <-- we don't really use the result, just need to return something
            });
            // do the same for resource.policy
            (0, tokens_1.resolveTokens)(resource.policy ?? {}, (token) => {
                implicitDeps.push(token.path);
                return "[T]";
            });
            // now add all implicit dependencies
            for (const dep of implicitDeps) {
                this.recordDependency(consumer, dep);
            }
        }
    }
    get nodes() {
        return Object.values(this.byPath);
    }
    tryFind(path) {
        const node = this.byPath[path];
        if (!node) {
            return undefined;
        }
        return node;
    }
    recordDependency(consumer, producer) {
        this.tryFind(consumer)?.dependencies.add(producer);
        this.tryFind(producer)?.dependents.add(consumer);
        // check for cyclic dependencies
        this.detectCycles(consumer);
        this.detectCycles(producer);
    }
    detectCycles(root) {
        const visited = new Set();
        const stack = new Set();
        const visit = (path) => {
            if (stack.has(path)) {
                throw new Error(`cyclic dependency detected: ${[...stack, path].join(" -> ")}`);
            }
            if (visited.has(path)) {
                return;
            }
            visited.add(path);
            stack.add(path);
            for (const dep of this.tryFind(path)?.dependencies ?? []) {
                visit(dep);
            }
            stack.delete(path);
        };
        visit(root);
    }
}
exports.Graph = Graph;
