// @cdktf/provider-aws/lib/sesv2-email-identity-feedback-attributes/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sesv2EmailIdentityFeedbackAttributes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_email_identity_feedback_attributes aws_sesv2_email_identity_feedback_attributes}
*/
class Sesv2EmailIdentityFeedbackAttributes extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Sesv2EmailIdentityFeedbackAttributes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Sesv2EmailIdentityFeedbackAttributes to import
    * @param importFromId The id of the existing Sesv2EmailIdentityFeedbackAttributes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_email_identity_feedback_attributes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Sesv2EmailIdentityFeedbackAttributes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sesv2_email_identity_feedback_attributes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sesv2_email_identity_feedback_attributes aws_sesv2_email_identity_feedback_attributes} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Sesv2EmailIdentityFeedbackAttributesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sesv2_email_identity_feedback_attributes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._emailForwardingEnabled = config.emailForwardingEnabled;
        this._emailIdentity = config.emailIdentity;
        this._id = config.id;
    }
    get emailForwardingEnabled() {
        return this.getBooleanAttribute('email_forwarding_enabled');
    }
    set emailForwardingEnabled(value) {
        this._emailForwardingEnabled = value;
    }
    resetEmailForwardingEnabled() {
        this._emailForwardingEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailForwardingEnabledInput() {
        return this._emailForwardingEnabled;
    }
    get emailIdentity() {
        return this.getStringAttribute('email_identity');
    }
    set emailIdentity(value) {
        this._emailIdentity = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailIdentityInput() {
        return this._emailIdentity;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            email_forwarding_enabled: cdktf.booleanToTerraform(this._emailForwardingEnabled),
            email_identity: cdktf.stringToTerraform(this._emailIdentity),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            email_forwarding_enabled: {
                value: cdktf.booleanToHclTerraform(this._emailForwardingEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            email_identity: {
                value: cdktf.stringToHclTerraform(this._emailIdentity),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Sesv2EmailIdentityFeedbackAttributes = Sesv2EmailIdentityFeedbackAttributes;
_a = JSII_RTTI_SYMBOL_1;
Sesv2EmailIdentityFeedbackAttributes[_a] = { fqn: "@cdktf/provider-aws.sesv2EmailIdentityFeedbackAttributes.Sesv2EmailIdentityFeedbackAttributes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Sesv2EmailIdentityFeedbackAttributes.tfResourceType = "aws_sesv2_email_identity_feedback_attributes";
