// @cdktf/provider-aws/lib/data-aws-instance/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsInstance = exports.DataAwsInstanceTimeoutsOutputReference = exports.dataAwsInstanceTimeoutsToHclTerraform = exports.dataAwsInstanceTimeoutsToTerraform = exports.DataAwsInstanceFilterList = exports.DataAwsInstanceFilterOutputReference = exports.dataAwsInstanceFilterToHclTerraform = exports.dataAwsInstanceFilterToTerraform = exports.DataAwsInstanceRootBlockDeviceList = exports.DataAwsInstanceRootBlockDeviceOutputReference = exports.dataAwsInstanceRootBlockDeviceToHclTerraform = exports.dataAwsInstanceRootBlockDeviceToTerraform = exports.DataAwsInstancePrivateDnsNameOptionsList = exports.DataAwsInstancePrivateDnsNameOptionsOutputReference = exports.dataAwsInstancePrivateDnsNameOptionsToHclTerraform = exports.dataAwsInstancePrivateDnsNameOptionsToTerraform = exports.DataAwsInstanceMetadataOptionsList = exports.DataAwsInstanceMetadataOptionsOutputReference = exports.dataAwsInstanceMetadataOptionsToHclTerraform = exports.dataAwsInstanceMetadataOptionsToTerraform = exports.DataAwsInstanceMaintenanceOptionsList = exports.DataAwsInstanceMaintenanceOptionsOutputReference = exports.dataAwsInstanceMaintenanceOptionsToHclTerraform = exports.dataAwsInstanceMaintenanceOptionsToTerraform = exports.DataAwsInstanceEphemeralBlockDeviceList = exports.DataAwsInstanceEphemeralBlockDeviceOutputReference = exports.dataAwsInstanceEphemeralBlockDeviceToHclTerraform = exports.dataAwsInstanceEphemeralBlockDeviceToTerraform = exports.DataAwsInstanceEnclaveOptionsList = exports.DataAwsInstanceEnclaveOptionsOutputReference = exports.dataAwsInstanceEnclaveOptionsToHclTerraform = exports.dataAwsInstanceEnclaveOptionsToTerraform = exports.DataAwsInstanceEbsBlockDeviceList = exports.DataAwsInstanceEbsBlockDeviceOutputReference = exports.dataAwsInstanceEbsBlockDeviceToHclTerraform = exports.dataAwsInstanceEbsBlockDeviceToTerraform = exports.DataAwsInstanceCreditSpecificationList = exports.DataAwsInstanceCreditSpecificationOutputReference = exports.dataAwsInstanceCreditSpecificationToHclTerraform = exports.dataAwsInstanceCreditSpecificationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsInstanceCreditSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceCreditSpecificationToTerraform = dataAwsInstanceCreditSpecificationToTerraform;
function dataAwsInstanceCreditSpecificationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceCreditSpecificationToHclTerraform = dataAwsInstanceCreditSpecificationToHclTerraform;
class DataAwsInstanceCreditSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cpu_credits - computed: true, optional: false, required: false
    get cpuCredits() {
        return this.getStringAttribute('cpu_credits');
    }
}
exports.DataAwsInstanceCreditSpecificationOutputReference = DataAwsInstanceCreditSpecificationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsInstanceCreditSpecificationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceCreditSpecificationOutputReference", version: "19.15.0" };
class DataAwsInstanceCreditSpecificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceCreditSpecificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceCreditSpecificationList = DataAwsInstanceCreditSpecificationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsInstanceCreditSpecificationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceCreditSpecificationList", version: "19.15.0" };
function dataAwsInstanceEbsBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceEbsBlockDeviceToTerraform = dataAwsInstanceEbsBlockDeviceToTerraform;
function dataAwsInstanceEbsBlockDeviceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceEbsBlockDeviceToHclTerraform = dataAwsInstanceEbsBlockDeviceToHclTerraform;
class DataAwsInstanceEbsBlockDeviceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_on_termination - computed: true, optional: false, required: false
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // snapshot_id - computed: true, optional: false, required: false
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    get tags() {
        return this._tags;
    }
    // throughput - computed: true, optional: false, required: false
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_id - computed: true, optional: false, required: false
    get volumeId() {
        return this.getStringAttribute('volume_id');
    }
    // volume_size - computed: true, optional: false, required: false
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsInstanceEbsBlockDeviceOutputReference = DataAwsInstanceEbsBlockDeviceOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEbsBlockDeviceOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEbsBlockDeviceOutputReference", version: "19.15.0" };
class DataAwsInstanceEbsBlockDeviceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceEbsBlockDeviceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceEbsBlockDeviceList = DataAwsInstanceEbsBlockDeviceList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEbsBlockDeviceList[_d] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEbsBlockDeviceList", version: "19.15.0" };
function dataAwsInstanceEnclaveOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceEnclaveOptionsToTerraform = dataAwsInstanceEnclaveOptionsToTerraform;
function dataAwsInstanceEnclaveOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceEnclaveOptionsToHclTerraform = dataAwsInstanceEnclaveOptionsToHclTerraform;
class DataAwsInstanceEnclaveOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsInstanceEnclaveOptionsOutputReference = DataAwsInstanceEnclaveOptionsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEnclaveOptionsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEnclaveOptionsOutputReference", version: "19.15.0" };
class DataAwsInstanceEnclaveOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceEnclaveOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceEnclaveOptionsList = DataAwsInstanceEnclaveOptionsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEnclaveOptionsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEnclaveOptionsList", version: "19.15.0" };
function dataAwsInstanceEphemeralBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceEphemeralBlockDeviceToTerraform = dataAwsInstanceEphemeralBlockDeviceToTerraform;
function dataAwsInstanceEphemeralBlockDeviceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceEphemeralBlockDeviceToHclTerraform = dataAwsInstanceEphemeralBlockDeviceToHclTerraform;
class DataAwsInstanceEphemeralBlockDeviceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // no_device - computed: true, optional: false, required: false
    get noDevice() {
        return this.getBooleanAttribute('no_device');
    }
    // virtual_name - computed: true, optional: false, required: false
    get virtualName() {
        return this.getStringAttribute('virtual_name');
    }
}
exports.DataAwsInstanceEphemeralBlockDeviceOutputReference = DataAwsInstanceEphemeralBlockDeviceOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEphemeralBlockDeviceOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEphemeralBlockDeviceOutputReference", version: "19.15.0" };
class DataAwsInstanceEphemeralBlockDeviceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceEphemeralBlockDeviceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceEphemeralBlockDeviceList = DataAwsInstanceEphemeralBlockDeviceList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEphemeralBlockDeviceList[_h] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceEphemeralBlockDeviceList", version: "19.15.0" };
function dataAwsInstanceMaintenanceOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceMaintenanceOptionsToTerraform = dataAwsInstanceMaintenanceOptionsToTerraform;
function dataAwsInstanceMaintenanceOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceMaintenanceOptionsToHclTerraform = dataAwsInstanceMaintenanceOptionsToHclTerraform;
class DataAwsInstanceMaintenanceOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_recovery - computed: true, optional: false, required: false
    get autoRecovery() {
        return this.getStringAttribute('auto_recovery');
    }
}
exports.DataAwsInstanceMaintenanceOptionsOutputReference = DataAwsInstanceMaintenanceOptionsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsInstanceMaintenanceOptionsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceMaintenanceOptionsOutputReference", version: "19.15.0" };
class DataAwsInstanceMaintenanceOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceMaintenanceOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceMaintenanceOptionsList = DataAwsInstanceMaintenanceOptionsList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsInstanceMaintenanceOptionsList[_k] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceMaintenanceOptionsList", version: "19.15.0" };
function dataAwsInstanceMetadataOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceMetadataOptionsToTerraform = dataAwsInstanceMetadataOptionsToTerraform;
function dataAwsInstanceMetadataOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceMetadataOptionsToHclTerraform = dataAwsInstanceMetadataOptionsToHclTerraform;
class DataAwsInstanceMetadataOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // http_endpoint - computed: true, optional: false, required: false
    get httpEndpoint() {
        return this.getStringAttribute('http_endpoint');
    }
    // http_protocol_ipv6 - computed: true, optional: false, required: false
    get httpProtocolIpv6() {
        return this.getStringAttribute('http_protocol_ipv6');
    }
    // http_put_response_hop_limit - computed: true, optional: false, required: false
    get httpPutResponseHopLimit() {
        return this.getNumberAttribute('http_put_response_hop_limit');
    }
    // http_tokens - computed: true, optional: false, required: false
    get httpTokens() {
        return this.getStringAttribute('http_tokens');
    }
    // instance_metadata_tags - computed: true, optional: false, required: false
    get instanceMetadataTags() {
        return this.getStringAttribute('instance_metadata_tags');
    }
}
exports.DataAwsInstanceMetadataOptionsOutputReference = DataAwsInstanceMetadataOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsInstanceMetadataOptionsOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceMetadataOptionsOutputReference", version: "19.15.0" };
class DataAwsInstanceMetadataOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceMetadataOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceMetadataOptionsList = DataAwsInstanceMetadataOptionsList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsInstanceMetadataOptionsList[_m] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceMetadataOptionsList", version: "19.15.0" };
function dataAwsInstancePrivateDnsNameOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstancePrivateDnsNameOptionsToTerraform = dataAwsInstancePrivateDnsNameOptionsToTerraform;
function dataAwsInstancePrivateDnsNameOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstancePrivateDnsNameOptionsToHclTerraform = dataAwsInstancePrivateDnsNameOptionsToHclTerraform;
class DataAwsInstancePrivateDnsNameOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_resource_name_dns_a_record - computed: true, optional: false, required: false
    get enableResourceNameDnsARecord() {
        return this.getBooleanAttribute('enable_resource_name_dns_a_record');
    }
    // enable_resource_name_dns_aaaa_record - computed: true, optional: false, required: false
    get enableResourceNameDnsAaaaRecord() {
        return this.getBooleanAttribute('enable_resource_name_dns_aaaa_record');
    }
    // hostname_type - computed: true, optional: false, required: false
    get hostnameType() {
        return this.getStringAttribute('hostname_type');
    }
}
exports.DataAwsInstancePrivateDnsNameOptionsOutputReference = DataAwsInstancePrivateDnsNameOptionsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsInstancePrivateDnsNameOptionsOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstancePrivateDnsNameOptionsOutputReference", version: "19.15.0" };
class DataAwsInstancePrivateDnsNameOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstancePrivateDnsNameOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstancePrivateDnsNameOptionsList = DataAwsInstancePrivateDnsNameOptionsList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsInstancePrivateDnsNameOptionsList[_p] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstancePrivateDnsNameOptionsList", version: "19.15.0" };
function dataAwsInstanceRootBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsInstanceRootBlockDeviceToTerraform = dataAwsInstanceRootBlockDeviceToTerraform;
function dataAwsInstanceRootBlockDeviceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsInstanceRootBlockDeviceToHclTerraform = dataAwsInstanceRootBlockDeviceToHclTerraform;
class DataAwsInstanceRootBlockDeviceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_on_termination - computed: true, optional: false, required: false
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    get tags() {
        return this._tags;
    }
    // throughput - computed: true, optional: false, required: false
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_id - computed: true, optional: false, required: false
    get volumeId() {
        return this.getStringAttribute('volume_id');
    }
    // volume_size - computed: true, optional: false, required: false
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsInstanceRootBlockDeviceOutputReference = DataAwsInstanceRootBlockDeviceOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsInstanceRootBlockDeviceOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceRootBlockDeviceOutputReference", version: "19.15.0" };
class DataAwsInstanceRootBlockDeviceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceRootBlockDeviceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceRootBlockDeviceList = DataAwsInstanceRootBlockDeviceList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsInstanceRootBlockDeviceList[_r] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceRootBlockDeviceList", version: "19.15.0" };
function dataAwsInstanceFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsInstanceFilterToTerraform = dataAwsInstanceFilterToTerraform;
function dataAwsInstanceFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsInstanceFilterToHclTerraform = dataAwsInstanceFilterToHclTerraform;
class DataAwsInstanceFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsInstanceFilterOutputReference = DataAwsInstanceFilterOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsInstanceFilterOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceFilterOutputReference", version: "19.15.0" };
class DataAwsInstanceFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsInstanceFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsInstanceFilterList = DataAwsInstanceFilterList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsInstanceFilterList[_t] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceFilterList", version: "19.15.0" };
function dataAwsInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.dataAwsInstanceTimeoutsToTerraform = dataAwsInstanceTimeoutsToTerraform;
function dataAwsInstanceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsInstanceTimeoutsToHclTerraform = dataAwsInstanceTimeoutsToHclTerraform;
class DataAwsInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._read = value.read;
        }
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.DataAwsInstanceTimeoutsOutputReference = DataAwsInstanceTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsInstanceTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstanceTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/instance aws_instance}
*/
class DataAwsInstance extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsInstance resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsInstance to import
    * @param importFromId The id of the existing DataAwsInstance that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/instance#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsInstance to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_instance", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/instance aws_instance} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsInstanceConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // credit_specification - computed: true, optional: false, required: false
        this._creditSpecification = new DataAwsInstanceCreditSpecificationList(this, "credit_specification", false);
        // ebs_block_device - computed: true, optional: false, required: false
        this._ebsBlockDevice = new DataAwsInstanceEbsBlockDeviceList(this, "ebs_block_device", true);
        // enclave_options - computed: true, optional: false, required: false
        this._enclaveOptions = new DataAwsInstanceEnclaveOptionsList(this, "enclave_options", false);
        // ephemeral_block_device - computed: true, optional: false, required: false
        this._ephemeralBlockDevice = new DataAwsInstanceEphemeralBlockDeviceList(this, "ephemeral_block_device", false);
        // maintenance_options - computed: true, optional: false, required: false
        this._maintenanceOptions = new DataAwsInstanceMaintenanceOptionsList(this, "maintenance_options", false);
        // metadata_options - computed: true, optional: false, required: false
        this._metadataOptions = new DataAwsInstanceMetadataOptionsList(this, "metadata_options", false);
        // private_dns_name_options - computed: true, optional: false, required: false
        this._privateDnsNameOptions = new DataAwsInstancePrivateDnsNameOptionsList(this, "private_dns_name_options", false);
        // root_block_device - computed: true, optional: false, required: false
        this._rootBlockDevice = new DataAwsInstanceRootBlockDeviceList(this, "root_block_device", true);
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsInstanceFilterList(this, "filter", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataAwsInstanceTimeoutsOutputReference(this, "timeouts");
        this._getPasswordData = config.fetchPasswordData;
        this._getUserData = config.fetchUserData;
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._instanceTags = config.instanceTags;
        this._tags = config.tags;
        this._filter.internalValue = config.filter;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ami - computed: true, optional: false, required: false
    get ami() {
        return this.getStringAttribute('ami');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // associate_public_ip_address - computed: true, optional: false, required: false
    get associatePublicIpAddress() {
        return this.getBooleanAttribute('associate_public_ip_address');
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    get creditSpecification() {
        return this._creditSpecification;
    }
    // disable_api_stop - computed: true, optional: false, required: false
    get disableApiStop() {
        return this.getBooleanAttribute('disable_api_stop');
    }
    // disable_api_termination - computed: true, optional: false, required: false
    get disableApiTermination() {
        return this.getBooleanAttribute('disable_api_termination');
    }
    get ebsBlockDevice() {
        return this._ebsBlockDevice;
    }
    // ebs_optimized - computed: true, optional: false, required: false
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    get enclaveOptions() {
        return this._enclaveOptions;
    }
    get ephemeralBlockDevice() {
        return this._ephemeralBlockDevice;
    }
    get fetchPasswordData() {
        return this.getBooleanAttribute('get_password_data');
    }
    set fetchPasswordData(value) {
        this._getPasswordData = value;
    }
    resetFetchPasswordData() {
        this._getPasswordData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fetchPasswordDataInput() {
        return this._getPasswordData;
    }
    get fetchUserData() {
        return this.getBooleanAttribute('get_user_data');
    }
    set fetchUserData(value) {
        this._getUserData = value;
    }
    resetFetchUserData() {
        this._getUserData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fetchUserDataInput() {
        return this._getUserData;
    }
    // host_id - computed: true, optional: false, required: false
    get hostId() {
        return this.getStringAttribute('host_id');
    }
    // host_resource_group_arn - computed: true, optional: false, required: false
    get hostResourceGroupArn() {
        return this.getStringAttribute('host_resource_group_arn');
    }
    // iam_instance_profile - computed: true, optional: false, required: false
    get iamInstanceProfile() {
        return this.getStringAttribute('iam_instance_profile');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    resetInstanceId() {
        this._instanceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    // instance_state - computed: true, optional: false, required: false
    get instanceState() {
        return this.getStringAttribute('instance_state');
    }
    get instanceTags() {
        return this.getStringMapAttribute('instance_tags');
    }
    set instanceTags(value) {
        this._instanceTags = value;
    }
    resetInstanceTags() {
        this._instanceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTagsInput() {
        return this._instanceTags;
    }
    // instance_type - computed: true, optional: false, required: false
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // ipv6_addresses - computed: true, optional: false, required: false
    get ipv6Addresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ipv6_addresses'));
    }
    // key_name - computed: true, optional: false, required: false
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    get maintenanceOptions() {
        return this._maintenanceOptions;
    }
    get metadataOptions() {
        return this._metadataOptions;
    }
    // monitoring - computed: true, optional: false, required: false
    get monitoring() {
        return this.getBooleanAttribute('monitoring');
    }
    // network_interface_id - computed: true, optional: false, required: false
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    // outpost_arn - computed: true, optional: false, required: false
    get outpostArn() {
        return this.getStringAttribute('outpost_arn');
    }
    // password_data - computed: true, optional: false, required: false
    get passwordData() {
        return this.getStringAttribute('password_data');
    }
    // placement_group - computed: true, optional: false, required: false
    get placementGroup() {
        return this.getStringAttribute('placement_group');
    }
    // placement_partition_number - computed: true, optional: false, required: false
    get placementPartitionNumber() {
        return this.getNumberAttribute('placement_partition_number');
    }
    // private_dns - computed: true, optional: false, required: false
    get privateDns() {
        return this.getStringAttribute('private_dns');
    }
    get privateDnsNameOptions() {
        return this._privateDnsNameOptions;
    }
    // private_ip - computed: true, optional: false, required: false
    get privateIp() {
        return this.getStringAttribute('private_ip');
    }
    // public_dns - computed: true, optional: false, required: false
    get publicDns() {
        return this.getStringAttribute('public_dns');
    }
    // public_ip - computed: true, optional: false, required: false
    get publicIp() {
        return this.getStringAttribute('public_ip');
    }
    get rootBlockDevice() {
        return this._rootBlockDevice;
    }
    // secondary_private_ips - computed: true, optional: false, required: false
    get secondaryPrivateIps() {
        return cdktf.Fn.tolist(this.getListAttribute('secondary_private_ips'));
    }
    // security_groups - computed: true, optional: false, required: false
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    // source_dest_check - computed: true, optional: false, required: false
    get sourceDestCheck() {
        return this.getBooleanAttribute('source_dest_check');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // tenancy - computed: true, optional: false, required: false
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
    // user_data - computed: true, optional: false, required: false
    get userData() {
        return this.getStringAttribute('user_data');
    }
    // user_data_base64 - computed: true, optional: false, required: false
    get userDataBase64() {
        return this.getStringAttribute('user_data_base64');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    get vpcSecurityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_security_group_ids'));
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            get_password_data: cdktf.booleanToTerraform(this._getPasswordData),
            get_user_data: cdktf.booleanToTerraform(this._getUserData),
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            instance_tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._instanceTags),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsInstanceFilterToTerraform, true)(this._filter.internalValue),
            timeouts: dataAwsInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            get_password_data: {
                value: cdktf.booleanToHclTerraform(this._getPasswordData),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            get_user_data: {
                value: cdktf.booleanToHclTerraform(this._getUserData),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._instanceTags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            filter: {
                value: cdktf.listMapperHcl(dataAwsInstanceFilterToHclTerraform, true)(this._filter.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DataAwsInstanceFilterList",
            },
            timeouts: {
                value: dataAwsInstanceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataAwsInstanceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsInstance = DataAwsInstance;
_v = JSII_RTTI_SYMBOL_1;
DataAwsInstance[_v] = { fqn: "@cdktf/provider-aws.dataAwsInstance.DataAwsInstance", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsInstance.tfResourceType = "aws_instance";
