// @cdktf/provider-aws/lib/finspace-kx-dataview/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FinspaceKxDataview = exports.FinspaceKxDataviewTimeoutsOutputReference = exports.finspaceKxDataviewTimeoutsToHclTerraform = exports.finspaceKxDataviewTimeoutsToTerraform = exports.FinspaceKxDataviewSegmentConfigurationsList = exports.FinspaceKxDataviewSegmentConfigurationsOutputReference = exports.finspaceKxDataviewSegmentConfigurationsToHclTerraform = exports.finspaceKxDataviewSegmentConfigurationsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function finspaceKxDataviewSegmentConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        db_paths: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.dbPaths),
        on_demand: cdktf.booleanToTerraform(struct.onDemand),
        volume_name: cdktf.stringToTerraform(struct.volumeName),
    };
}
exports.finspaceKxDataviewSegmentConfigurationsToTerraform = finspaceKxDataviewSegmentConfigurationsToTerraform;
function finspaceKxDataviewSegmentConfigurationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        db_paths: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.dbPaths),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        on_demand: {
            value: cdktf.booleanToHclTerraform(struct.onDemand),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        volume_name: {
            value: cdktf.stringToHclTerraform(struct.volumeName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.finspaceKxDataviewSegmentConfigurationsToHclTerraform = finspaceKxDataviewSegmentConfigurationsToHclTerraform;
class FinspaceKxDataviewSegmentConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dbPaths !== undefined) {
            hasAnyValues = true;
            internalValueResult.dbPaths = this._dbPaths;
        }
        if (this._onDemand !== undefined) {
            hasAnyValues = true;
            internalValueResult.onDemand = this._onDemand;
        }
        if (this._volumeName !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeName = this._volumeName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dbPaths = undefined;
            this._onDemand = undefined;
            this._volumeName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dbPaths = value.dbPaths;
            this._onDemand = value.onDemand;
            this._volumeName = value.volumeName;
        }
    }
    get dbPaths() {
        return this.getListAttribute('db_paths');
    }
    set dbPaths(value) {
        this._dbPaths = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbPathsInput() {
        return this._dbPaths;
    }
    get onDemand() {
        return this.getBooleanAttribute('on_demand');
    }
    set onDemand(value) {
        this._onDemand = value;
    }
    resetOnDemand() {
        this._onDemand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onDemandInput() {
        return this._onDemand;
    }
    get volumeName() {
        return this.getStringAttribute('volume_name');
    }
    set volumeName(value) {
        this._volumeName = value;
    }
    // Temporarily expose input value. Use with caution.
    get volumeNameInput() {
        return this._volumeName;
    }
}
exports.FinspaceKxDataviewSegmentConfigurationsOutputReference = FinspaceKxDataviewSegmentConfigurationsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FinspaceKxDataviewSegmentConfigurationsOutputReference[_a] = { fqn: "@cdktf/provider-aws.finspaceKxDataview.FinspaceKxDataviewSegmentConfigurationsOutputReference", version: "19.15.0" };
class FinspaceKxDataviewSegmentConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FinspaceKxDataviewSegmentConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FinspaceKxDataviewSegmentConfigurationsList = FinspaceKxDataviewSegmentConfigurationsList;
_b = JSII_RTTI_SYMBOL_1;
FinspaceKxDataviewSegmentConfigurationsList[_b] = { fqn: "@cdktf/provider-aws.finspaceKxDataview.FinspaceKxDataviewSegmentConfigurationsList", version: "19.15.0" };
function finspaceKxDataviewTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.finspaceKxDataviewTimeoutsToTerraform = finspaceKxDataviewTimeoutsToTerraform;
function finspaceKxDataviewTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.finspaceKxDataviewTimeoutsToHclTerraform = finspaceKxDataviewTimeoutsToHclTerraform;
class FinspaceKxDataviewTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FinspaceKxDataviewTimeoutsOutputReference = FinspaceKxDataviewTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FinspaceKxDataviewTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.finspaceKxDataview.FinspaceKxDataviewTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/finspace_kx_dataview aws_finspace_kx_dataview}
*/
class FinspaceKxDataview extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FinspaceKxDataview resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FinspaceKxDataview to import
    * @param importFromId The id of the existing FinspaceKxDataview that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/finspace_kx_dataview#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FinspaceKxDataview to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_finspace_kx_dataview", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/finspace_kx_dataview aws_finspace_kx_dataview} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FinspaceKxDataviewConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_finspace_kx_dataview',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // segment_configurations - computed: false, optional: true, required: false
        this._segmentConfigurations = new FinspaceKxDataviewSegmentConfigurationsList(this, "segment_configurations", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FinspaceKxDataviewTimeoutsOutputReference(this, "timeouts");
        this._autoUpdate = config.autoUpdate;
        this._availabilityZoneId = config.availabilityZoneId;
        this._azMode = config.azMode;
        this._changesetId = config.changesetId;
        this._databaseName = config.databaseName;
        this._description = config.description;
        this._environmentId = config.environmentId;
        this._id = config.id;
        this._name = config.name;
        this._readWrite = config.readWrite;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._segmentConfigurations.internalValue = config.segmentConfigurations;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoUpdate() {
        return this.getBooleanAttribute('auto_update');
    }
    set autoUpdate(value) {
        this._autoUpdate = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoUpdateInput() {
        return this._autoUpdate;
    }
    get availabilityZoneId() {
        return this.getStringAttribute('availability_zone_id');
    }
    set availabilityZoneId(value) {
        this._availabilityZoneId = value;
    }
    resetAvailabilityZoneId() {
        this._availabilityZoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneIdInput() {
        return this._availabilityZoneId;
    }
    get azMode() {
        return this.getStringAttribute('az_mode');
    }
    set azMode(value) {
        this._azMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get azModeInput() {
        return this._azMode;
    }
    get changesetId() {
        return this.getStringAttribute('changeset_id');
    }
    set changesetId(value) {
        this._changesetId = value;
    }
    resetChangesetId() {
        this._changesetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get changesetIdInput() {
        return this._changesetId;
    }
    // created_timestamp - computed: true, optional: false, required: false
    get createdTimestamp() {
        return this.getStringAttribute('created_timestamp');
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environmentId() {
        return this.getStringAttribute('environment_id');
    }
    set environmentId(value) {
        this._environmentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentIdInput() {
        return this._environmentId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_modified_timestamp - computed: true, optional: false, required: false
    get lastModifiedTimestamp() {
        return this.getStringAttribute('last_modified_timestamp');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get readWrite() {
        return this.getBooleanAttribute('read_write');
    }
    set readWrite(value) {
        this._readWrite = value;
    }
    resetReadWrite() {
        this._readWrite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readWriteInput() {
        return this._readWrite;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get segmentConfigurations() {
        return this._segmentConfigurations;
    }
    putSegmentConfigurations(value) {
        this._segmentConfigurations.internalValue = value;
    }
    resetSegmentConfigurations() {
        this._segmentConfigurations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentConfigurationsInput() {
        return this._segmentConfigurations.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_update: cdktf.booleanToTerraform(this._autoUpdate),
            availability_zone_id: cdktf.stringToTerraform(this._availabilityZoneId),
            az_mode: cdktf.stringToTerraform(this._azMode),
            changeset_id: cdktf.stringToTerraform(this._changesetId),
            database_name: cdktf.stringToTerraform(this._databaseName),
            description: cdktf.stringToTerraform(this._description),
            environment_id: cdktf.stringToTerraform(this._environmentId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            read_write: cdktf.booleanToTerraform(this._readWrite),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            segment_configurations: cdktf.listMapper(finspaceKxDataviewSegmentConfigurationsToTerraform, true)(this._segmentConfigurations.internalValue),
            timeouts: finspaceKxDataviewTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            auto_update: {
                value: cdktf.booleanToHclTerraform(this._autoUpdate),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            availability_zone_id: {
                value: cdktf.stringToHclTerraform(this._availabilityZoneId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            az_mode: {
                value: cdktf.stringToHclTerraform(this._azMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            changeset_id: {
                value: cdktf.stringToHclTerraform(this._changesetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            database_name: {
                value: cdktf.stringToHclTerraform(this._databaseName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            environment_id: {
                value: cdktf.stringToHclTerraform(this._environmentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            read_write: {
                value: cdktf.booleanToHclTerraform(this._readWrite),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            segment_configurations: {
                value: cdktf.listMapperHcl(finspaceKxDataviewSegmentConfigurationsToHclTerraform, true)(this._segmentConfigurations.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FinspaceKxDataviewSegmentConfigurationsList",
            },
            timeouts: {
                value: finspaceKxDataviewTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FinspaceKxDataviewTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FinspaceKxDataview = FinspaceKxDataview;
_d = JSII_RTTI_SYMBOL_1;
FinspaceKxDataview[_d] = { fqn: "@cdktf/provider-aws.finspaceKxDataview.FinspaceKxDataview", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
FinspaceKxDataview.tfResourceType = "aws_finspace_kx_dataview";
