// @cdktf/provider-aws/lib/data-aws-lambda-function/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLambdaFunction = exports.DataAwsLambdaFunctionVpcConfigList = exports.DataAwsLambdaFunctionVpcConfigOutputReference = exports.dataAwsLambdaFunctionVpcConfigToHclTerraform = exports.dataAwsLambdaFunctionVpcConfigToTerraform = exports.DataAwsLambdaFunctionTracingConfigList = exports.DataAwsLambdaFunctionTracingConfigOutputReference = exports.dataAwsLambdaFunctionTracingConfigToHclTerraform = exports.dataAwsLambdaFunctionTracingConfigToTerraform = exports.DataAwsLambdaFunctionLoggingConfigList = exports.DataAwsLambdaFunctionLoggingConfigOutputReference = exports.dataAwsLambdaFunctionLoggingConfigToHclTerraform = exports.dataAwsLambdaFunctionLoggingConfigToTerraform = exports.DataAwsLambdaFunctionFileSystemConfigList = exports.DataAwsLambdaFunctionFileSystemConfigOutputReference = exports.dataAwsLambdaFunctionFileSystemConfigToHclTerraform = exports.dataAwsLambdaFunctionFileSystemConfigToTerraform = exports.DataAwsLambdaFunctionEphemeralStorageList = exports.DataAwsLambdaFunctionEphemeralStorageOutputReference = exports.dataAwsLambdaFunctionEphemeralStorageToHclTerraform = exports.dataAwsLambdaFunctionEphemeralStorageToTerraform = exports.DataAwsLambdaFunctionEnvironmentList = exports.DataAwsLambdaFunctionEnvironmentOutputReference = exports.dataAwsLambdaFunctionEnvironmentToHclTerraform = exports.dataAwsLambdaFunctionEnvironmentToTerraform = exports.DataAwsLambdaFunctionDeadLetterConfigList = exports.DataAwsLambdaFunctionDeadLetterConfigOutputReference = exports.dataAwsLambdaFunctionDeadLetterConfigToHclTerraform = exports.dataAwsLambdaFunctionDeadLetterConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLambdaFunctionDeadLetterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionDeadLetterConfigToTerraform = dataAwsLambdaFunctionDeadLetterConfigToTerraform;
function dataAwsLambdaFunctionDeadLetterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionDeadLetterConfigToHclTerraform = dataAwsLambdaFunctionDeadLetterConfigToHclTerraform;
class DataAwsLambdaFunctionDeadLetterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // target_arn - computed: true, optional: false, required: false
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
}
exports.DataAwsLambdaFunctionDeadLetterConfigOutputReference = DataAwsLambdaFunctionDeadLetterConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionDeadLetterConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionDeadLetterConfigOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionDeadLetterConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionDeadLetterConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionDeadLetterConfigList = DataAwsLambdaFunctionDeadLetterConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionDeadLetterConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionDeadLetterConfigList", version: "19.15.0" };
function dataAwsLambdaFunctionEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionEnvironmentToTerraform = dataAwsLambdaFunctionEnvironmentToTerraform;
function dataAwsLambdaFunctionEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionEnvironmentToHclTerraform = dataAwsLambdaFunctionEnvironmentToHclTerraform;
class DataAwsLambdaFunctionEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // variables - computed: true, optional: false, required: false
        this._variables = new cdktf.StringMap(this, "variables");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get variables() {
        return this._variables;
    }
}
exports.DataAwsLambdaFunctionEnvironmentOutputReference = DataAwsLambdaFunctionEnvironmentOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionEnvironmentOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionEnvironmentOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionEnvironmentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionEnvironmentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionEnvironmentList = DataAwsLambdaFunctionEnvironmentList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionEnvironmentList[_d] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionEnvironmentList", version: "19.15.0" };
function dataAwsLambdaFunctionEphemeralStorageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionEphemeralStorageToTerraform = dataAwsLambdaFunctionEphemeralStorageToTerraform;
function dataAwsLambdaFunctionEphemeralStorageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionEphemeralStorageToHclTerraform = dataAwsLambdaFunctionEphemeralStorageToHclTerraform;
class DataAwsLambdaFunctionEphemeralStorageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
}
exports.DataAwsLambdaFunctionEphemeralStorageOutputReference = DataAwsLambdaFunctionEphemeralStorageOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionEphemeralStorageOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionEphemeralStorageOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionEphemeralStorageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionEphemeralStorageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionEphemeralStorageList = DataAwsLambdaFunctionEphemeralStorageList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionEphemeralStorageList[_f] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionEphemeralStorageList", version: "19.15.0" };
function dataAwsLambdaFunctionFileSystemConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionFileSystemConfigToTerraform = dataAwsLambdaFunctionFileSystemConfigToTerraform;
function dataAwsLambdaFunctionFileSystemConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionFileSystemConfigToHclTerraform = dataAwsLambdaFunctionFileSystemConfigToHclTerraform;
class DataAwsLambdaFunctionFileSystemConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // local_mount_path - computed: true, optional: false, required: false
    get localMountPath() {
        return this.getStringAttribute('local_mount_path');
    }
}
exports.DataAwsLambdaFunctionFileSystemConfigOutputReference = DataAwsLambdaFunctionFileSystemConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionFileSystemConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionFileSystemConfigOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionFileSystemConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionFileSystemConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionFileSystemConfigList = DataAwsLambdaFunctionFileSystemConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionFileSystemConfigList[_h] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionFileSystemConfigList", version: "19.15.0" };
function dataAwsLambdaFunctionLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionLoggingConfigToTerraform = dataAwsLambdaFunctionLoggingConfigToTerraform;
function dataAwsLambdaFunctionLoggingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionLoggingConfigToHclTerraform = dataAwsLambdaFunctionLoggingConfigToHclTerraform;
class DataAwsLambdaFunctionLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // application_log_level - computed: true, optional: false, required: false
    get applicationLogLevel() {
        return this.getStringAttribute('application_log_level');
    }
    // log_format - computed: true, optional: false, required: false
    get logFormat() {
        return this.getStringAttribute('log_format');
    }
    // log_group - computed: true, optional: false, required: false
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    // system_log_level - computed: true, optional: false, required: false
    get systemLogLevel() {
        return this.getStringAttribute('system_log_level');
    }
}
exports.DataAwsLambdaFunctionLoggingConfigOutputReference = DataAwsLambdaFunctionLoggingConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionLoggingConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionLoggingConfigOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionLoggingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionLoggingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionLoggingConfigList = DataAwsLambdaFunctionLoggingConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionLoggingConfigList[_k] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionLoggingConfigList", version: "19.15.0" };
function dataAwsLambdaFunctionTracingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionTracingConfigToTerraform = dataAwsLambdaFunctionTracingConfigToTerraform;
function dataAwsLambdaFunctionTracingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionTracingConfigToHclTerraform = dataAwsLambdaFunctionTracingConfigToHclTerraform;
class DataAwsLambdaFunctionTracingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataAwsLambdaFunctionTracingConfigOutputReference = DataAwsLambdaFunctionTracingConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionTracingConfigOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionTracingConfigOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionTracingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionTracingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionTracingConfigList = DataAwsLambdaFunctionTracingConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionTracingConfigList[_m] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionTracingConfigList", version: "19.15.0" };
function dataAwsLambdaFunctionVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaFunctionVpcConfigToTerraform = dataAwsLambdaFunctionVpcConfigToTerraform;
function dataAwsLambdaFunctionVpcConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaFunctionVpcConfigToHclTerraform = dataAwsLambdaFunctionVpcConfigToHclTerraform;
class DataAwsLambdaFunctionVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ipv6_allowed_for_dual_stack - computed: true, optional: false, required: false
    get ipv6AllowedForDualStack() {
        return this.getBooleanAttribute('ipv6_allowed_for_dual_stack');
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsLambdaFunctionVpcConfigOutputReference = DataAwsLambdaFunctionVpcConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionVpcConfigOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionVpcConfigOutputReference", version: "19.15.0" };
class DataAwsLambdaFunctionVpcConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaFunctionVpcConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaFunctionVpcConfigList = DataAwsLambdaFunctionVpcConfigList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionVpcConfigList[_p] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunctionVpcConfigList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_function aws_lambda_function}
*/
class DataAwsLambdaFunction extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsLambdaFunction resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsLambdaFunction to import
    * @param importFromId The id of the existing DataAwsLambdaFunction that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_function#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsLambdaFunction to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lambda_function", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_function aws_lambda_function} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLambdaFunctionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_function',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dead_letter_config - computed: true, optional: false, required: false
        this._deadLetterConfig = new DataAwsLambdaFunctionDeadLetterConfigList(this, "dead_letter_config", false);
        // environment - computed: true, optional: false, required: false
        this._environment = new DataAwsLambdaFunctionEnvironmentList(this, "environment", false);
        // ephemeral_storage - computed: true, optional: false, required: false
        this._ephemeralStorage = new DataAwsLambdaFunctionEphemeralStorageList(this, "ephemeral_storage", false);
        // file_system_config - computed: true, optional: false, required: false
        this._fileSystemConfig = new DataAwsLambdaFunctionFileSystemConfigList(this, "file_system_config", false);
        // logging_config - computed: true, optional: false, required: false
        this._loggingConfig = new DataAwsLambdaFunctionLoggingConfigList(this, "logging_config", false);
        // tracing_config - computed: true, optional: false, required: false
        this._tracingConfig = new DataAwsLambdaFunctionTracingConfigList(this, "tracing_config", false);
        // vpc_config - computed: true, optional: false, required: false
        this._vpcConfig = new DataAwsLambdaFunctionVpcConfigList(this, "vpc_config", false);
        this._functionName = config.functionName;
        this._id = config.id;
        this._qualifier = config.qualifier;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // architectures - computed: true, optional: false, required: false
    get architectures() {
        return this.getListAttribute('architectures');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // code_signing_config_arn - computed: true, optional: false, required: false
    get codeSigningConfigArn() {
        return this.getStringAttribute('code_signing_config_arn');
    }
    get deadLetterConfig() {
        return this._deadLetterConfig;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get environment() {
        return this._environment;
    }
    get ephemeralStorage() {
        return this._ephemeralStorage;
    }
    get fileSystemConfig() {
        return this._fileSystemConfig;
    }
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get functionNameInput() {
        return this._functionName;
    }
    // handler - computed: true, optional: false, required: false
    get handler() {
        return this.getStringAttribute('handler');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // image_uri - computed: true, optional: false, required: false
    get imageUri() {
        return this.getStringAttribute('image_uri');
    }
    // invoke_arn - computed: true, optional: false, required: false
    get invokeArn() {
        return this.getStringAttribute('invoke_arn');
    }
    // kms_key_arn - computed: true, optional: false, required: false
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    // last_modified - computed: true, optional: false, required: false
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    // layers - computed: true, optional: false, required: false
    get layers() {
        return this.getListAttribute('layers');
    }
    get loggingConfig() {
        return this._loggingConfig;
    }
    // memory_size - computed: true, optional: false, required: false
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    // qualified_arn - computed: true, optional: false, required: false
    get qualifiedArn() {
        return this.getStringAttribute('qualified_arn');
    }
    // qualified_invoke_arn - computed: true, optional: false, required: false
    get qualifiedInvokeArn() {
        return this.getStringAttribute('qualified_invoke_arn');
    }
    get qualifier() {
        return this.getStringAttribute('qualifier');
    }
    set qualifier(value) {
        this._qualifier = value;
    }
    resetQualifier() {
        this._qualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get qualifierInput() {
        return this._qualifier;
    }
    // reserved_concurrent_executions - computed: true, optional: false, required: false
    get reservedConcurrentExecutions() {
        return this.getNumberAttribute('reserved_concurrent_executions');
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
    // runtime - computed: true, optional: false, required: false
    get runtime() {
        return this.getStringAttribute('runtime');
    }
    // signing_job_arn - computed: true, optional: false, required: false
    get signingJobArn() {
        return this.getStringAttribute('signing_job_arn');
    }
    // signing_profile_version_arn - computed: true, optional: false, required: false
    get signingProfileVersionArn() {
        return this.getStringAttribute('signing_profile_version_arn');
    }
    // source_code_hash - computed: true, optional: false, required: false
    get sourceCodeHash() {
        return this.getStringAttribute('source_code_hash');
    }
    // source_code_size - computed: true, optional: false, required: false
    get sourceCodeSize() {
        return this.getNumberAttribute('source_code_size');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // timeout - computed: true, optional: false, required: false
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    get tracingConfig() {
        return this._tracingConfig;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            function_name: cdktf.stringToTerraform(this._functionName),
            id: cdktf.stringToTerraform(this._id),
            qualifier: cdktf.stringToTerraform(this._qualifier),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            function_name: {
                value: cdktf.stringToHclTerraform(this._functionName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            qualifier: {
                value: cdktf.stringToHclTerraform(this._qualifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsLambdaFunction = DataAwsLambdaFunction;
_q = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunction[_q] = { fqn: "@cdktf/provider-aws.dataAwsLambdaFunction.DataAwsLambdaFunction", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLambdaFunction.tfResourceType = "aws_lambda_function";
