// @cdktf/provider-aws/lib/lakeformation-resource/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LakeformationResource = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_resource aws_lakeformation_resource}
*/
class LakeformationResource extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LakeformationResource resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LakeformationResource to import
    * @param importFromId The id of the existing LakeformationResource that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_resource#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LakeformationResource to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lakeformation_resource", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lakeformation_resource aws_lakeformation_resource} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LakeformationResourceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lakeformation_resource',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._arn = config.arn;
        this._hybridAccessEnabled = config.hybridAccessEnabled;
        this._id = config.id;
        this._roleArn = config.roleArn;
        this._useServiceLinkedRole = config.useServiceLinkedRole;
        this._withFederation = config.withFederation;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get hybridAccessEnabled() {
        return this.getBooleanAttribute('hybrid_access_enabled');
    }
    set hybridAccessEnabled(value) {
        this._hybridAccessEnabled = value;
    }
    resetHybridAccessEnabled() {
        this._hybridAccessEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hybridAccessEnabledInput() {
        return this._hybridAccessEnabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_modified - computed: true, optional: false, required: false
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get useServiceLinkedRole() {
        return this.getBooleanAttribute('use_service_linked_role');
    }
    set useServiceLinkedRole(value) {
        this._useServiceLinkedRole = value;
    }
    resetUseServiceLinkedRole() {
        this._useServiceLinkedRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useServiceLinkedRoleInput() {
        return this._useServiceLinkedRole;
    }
    get withFederation() {
        return this.getBooleanAttribute('with_federation');
    }
    set withFederation(value) {
        this._withFederation = value;
    }
    resetWithFederation() {
        this._withFederation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get withFederationInput() {
        return this._withFederation;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            hybrid_access_enabled: cdktf.booleanToTerraform(this._hybridAccessEnabled),
            id: cdktf.stringToTerraform(this._id),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            use_service_linked_role: cdktf.booleanToTerraform(this._useServiceLinkedRole),
            with_federation: cdktf.booleanToTerraform(this._withFederation),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            hybrid_access_enabled: {
                value: cdktf.booleanToHclTerraform(this._hybridAccessEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            use_service_linked_role: {
                value: cdktf.booleanToHclTerraform(this._useServiceLinkedRole),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            with_federation: {
                value: cdktf.booleanToHclTerraform(this._withFederation),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LakeformationResource = LakeformationResource;
_a = JSII_RTTI_SYMBOL_1;
LakeformationResource[_a] = { fqn: "@cdktf/provider-aws.lakeformationResource.LakeformationResource", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LakeformationResource.tfResourceType = "aws_lakeformation_resource";
