// @cdktf/provider-aws/lib/data-aws-connect-bot-association/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectBotAssociation = exports.DataAwsConnectBotAssociationLexBotOutputReference = exports.dataAwsConnectBotAssociationLexBotToHclTerraform = exports.dataAwsConnectBotAssociationLexBotToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectBotAssociationLexBotToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lex_region: cdktf.stringToTerraform(struct.lexRegion),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.dataAwsConnectBotAssociationLexBotToTerraform = dataAwsConnectBotAssociationLexBotToTerraform;
function dataAwsConnectBotAssociationLexBotToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        lex_region: {
            value: cdktf.stringToHclTerraform(struct.lexRegion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsConnectBotAssociationLexBotToHclTerraform = dataAwsConnectBotAssociationLexBotToHclTerraform;
class DataAwsConnectBotAssociationLexBotOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lexRegion !== undefined) {
            hasAnyValues = true;
            internalValueResult.lexRegion = this._lexRegion;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lexRegion = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lexRegion = value.lexRegion;
            this._name = value.name;
        }
    }
    get lexRegion() {
        return this.getStringAttribute('lex_region');
    }
    set lexRegion(value) {
        this._lexRegion = value;
    }
    resetLexRegion() {
        this._lexRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lexRegionInput() {
        return this._lexRegion;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.DataAwsConnectBotAssociationLexBotOutputReference = DataAwsConnectBotAssociationLexBotOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectBotAssociationLexBotOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectBotAssociation.DataAwsConnectBotAssociationLexBotOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_bot_association aws_connect_bot_association}
*/
class DataAwsConnectBotAssociation extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsConnectBotAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsConnectBotAssociation to import
    * @param importFromId The id of the existing DataAwsConnectBotAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_bot_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsConnectBotAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_connect_bot_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/connect_bot_association aws_connect_bot_association} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectBotAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_bot_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // lex_bot - computed: false, optional: false, required: true
        this._lexBot = new DataAwsConnectBotAssociationLexBotOutputReference(this, "lex_bot");
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._lexBot.internalValue = config.lexBot;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get lexBot() {
        return this._lexBot;
    }
    putLexBot(value) {
        this._lexBot.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get lexBotInput() {
        return this._lexBot.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            lex_bot: dataAwsConnectBotAssociationLexBotToTerraform(this._lexBot.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            lex_bot: {
                value: dataAwsConnectBotAssociationLexBotToHclTerraform(this._lexBot.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsConnectBotAssociationLexBotList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsConnectBotAssociation = DataAwsConnectBotAssociation;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectBotAssociation[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectBotAssociation.DataAwsConnectBotAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectBotAssociation.tfResourceType = "aws_connect_bot_association";
