// @winglang/sdk/lib/.gen/providers/aws/lambda-event-source-mapping/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/lambda_event_source_mapping
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaEventSourceMapping = exports.LambdaEventSourceMappingSourceAccessConfigurationList = exports.LambdaEventSourceMappingSourceAccessConfigurationOutputReference = exports.lambdaEventSourceMappingSourceAccessConfigurationToHclTerraform = exports.lambdaEventSourceMappingSourceAccessConfigurationToTerraform = exports.LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigOutputReference = exports.lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToHclTerraform = exports.lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToTerraform = exports.LambdaEventSourceMappingSelfManagedEventSourceOutputReference = exports.lambdaEventSourceMappingSelfManagedEventSourceToHclTerraform = exports.lambdaEventSourceMappingSelfManagedEventSourceToTerraform = exports.LambdaEventSourceMappingScalingConfigOutputReference = exports.lambdaEventSourceMappingScalingConfigToHclTerraform = exports.lambdaEventSourceMappingScalingConfigToTerraform = exports.LambdaEventSourceMappingFilterCriteriaOutputReference = exports.lambdaEventSourceMappingFilterCriteriaToHclTerraform = exports.lambdaEventSourceMappingFilterCriteriaToTerraform = exports.LambdaEventSourceMappingFilterCriteriaFilterList = exports.LambdaEventSourceMappingFilterCriteriaFilterOutputReference = exports.lambdaEventSourceMappingFilterCriteriaFilterToHclTerraform = exports.lambdaEventSourceMappingFilterCriteriaFilterToTerraform = exports.LambdaEventSourceMappingDocumentDbEventSourceConfigOutputReference = exports.lambdaEventSourceMappingDocumentDbEventSourceConfigToHclTerraform = exports.lambdaEventSourceMappingDocumentDbEventSourceConfigToTerraform = exports.LambdaEventSourceMappingDestinationConfigOutputReference = exports.lambdaEventSourceMappingDestinationConfigToHclTerraform = exports.lambdaEventSourceMappingDestinationConfigToTerraform = exports.LambdaEventSourceMappingDestinationConfigOnFailureOutputReference = exports.lambdaEventSourceMappingDestinationConfigOnFailureToHclTerraform = exports.lambdaEventSourceMappingDestinationConfigOnFailureToTerraform = exports.LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigOutputReference = exports.lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToHclTerraform = exports.lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToTerraform = void 0;
const cdktf = __importStar(require("cdktf"));
function lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        consumer_group_id: cdktf.stringToTerraform(struct.consumerGroupId),
    };
}
exports.lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToTerraform = lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToTerraform;
function lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        consumer_group_id: {
            value: cdktf.stringToHclTerraform(struct.consumerGroupId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToHclTerraform = lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToHclTerraform;
class LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._consumerGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.consumerGroupId = this._consumerGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._consumerGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._consumerGroupId = value.consumerGroupId;
        }
    }
    get consumerGroupId() {
        return this.getStringAttribute('consumer_group_id');
    }
    set consumerGroupId(value) {
        this._consumerGroupId = value;
    }
    resetConsumerGroupId() {
        this._consumerGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get consumerGroupIdInput() {
        return this._consumerGroupId;
    }
}
exports.LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigOutputReference = LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigOutputReference;
function lambdaEventSourceMappingDestinationConfigOnFailureToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_arn: cdktf.stringToTerraform(struct.destinationArn),
    };
}
exports.lambdaEventSourceMappingDestinationConfigOnFailureToTerraform = lambdaEventSourceMappingDestinationConfigOnFailureToTerraform;
function lambdaEventSourceMappingDestinationConfigOnFailureToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination_arn: {
            value: cdktf.stringToHclTerraform(struct.destinationArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingDestinationConfigOnFailureToHclTerraform = lambdaEventSourceMappingDestinationConfigOnFailureToHclTerraform;
class LambdaEventSourceMappingDestinationConfigOnFailureOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationArn = this._destinationArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationArn = value.destinationArn;
        }
    }
    get destinationArn() {
        return this.getStringAttribute('destination_arn');
    }
    set destinationArn(value) {
        this._destinationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationArnInput() {
        return this._destinationArn;
    }
}
exports.LambdaEventSourceMappingDestinationConfigOnFailureOutputReference = LambdaEventSourceMappingDestinationConfigOnFailureOutputReference;
function lambdaEventSourceMappingDestinationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_failure: lambdaEventSourceMappingDestinationConfigOnFailureToTerraform(struct.onFailure),
    };
}
exports.lambdaEventSourceMappingDestinationConfigToTerraform = lambdaEventSourceMappingDestinationConfigToTerraform;
function lambdaEventSourceMappingDestinationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        on_failure: {
            value: lambdaEventSourceMappingDestinationConfigOnFailureToHclTerraform(struct.onFailure),
            isBlock: true,
            type: "list",
            storageClassType: "LambdaEventSourceMappingDestinationConfigOnFailureList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingDestinationConfigToHclTerraform = lambdaEventSourceMappingDestinationConfigToHclTerraform;
class LambdaEventSourceMappingDestinationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // on_failure - computed: false, optional: true, required: false
        this._onFailure = new LambdaEventSourceMappingDestinationConfigOnFailureOutputReference(this, "on_failure");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onFailure?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onFailure = this._onFailure?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._onFailure.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._onFailure.internalValue = value.onFailure;
        }
    }
    get onFailure() {
        return this._onFailure;
    }
    putOnFailure(value) {
        this._onFailure.internalValue = value;
    }
    resetOnFailure() {
        this._onFailure.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onFailureInput() {
        return this._onFailure.internalValue;
    }
}
exports.LambdaEventSourceMappingDestinationConfigOutputReference = LambdaEventSourceMappingDestinationConfigOutputReference;
function lambdaEventSourceMappingDocumentDbEventSourceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        collection_name: cdktf.stringToTerraform(struct.collectionName),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        full_document: cdktf.stringToTerraform(struct.fullDocument),
    };
}
exports.lambdaEventSourceMappingDocumentDbEventSourceConfigToTerraform = lambdaEventSourceMappingDocumentDbEventSourceConfigToTerraform;
function lambdaEventSourceMappingDocumentDbEventSourceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        collection_name: {
            value: cdktf.stringToHclTerraform(struct.collectionName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        database_name: {
            value: cdktf.stringToHclTerraform(struct.databaseName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        full_document: {
            value: cdktf.stringToHclTerraform(struct.fullDocument),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingDocumentDbEventSourceConfigToHclTerraform = lambdaEventSourceMappingDocumentDbEventSourceConfigToHclTerraform;
class LambdaEventSourceMappingDocumentDbEventSourceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._collectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.collectionName = this._collectionName;
        }
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._fullDocument !== undefined) {
            hasAnyValues = true;
            internalValueResult.fullDocument = this._fullDocument;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._collectionName = undefined;
            this._databaseName = undefined;
            this._fullDocument = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._collectionName = value.collectionName;
            this._databaseName = value.databaseName;
            this._fullDocument = value.fullDocument;
        }
    }
    get collectionName() {
        return this.getStringAttribute('collection_name');
    }
    set collectionName(value) {
        this._collectionName = value;
    }
    resetCollectionName() {
        this._collectionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get collectionNameInput() {
        return this._collectionName;
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get fullDocument() {
        return this.getStringAttribute('full_document');
    }
    set fullDocument(value) {
        this._fullDocument = value;
    }
    resetFullDocument() {
        this._fullDocument = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fullDocumentInput() {
        return this._fullDocument;
    }
}
exports.LambdaEventSourceMappingDocumentDbEventSourceConfigOutputReference = LambdaEventSourceMappingDocumentDbEventSourceConfigOutputReference;
function lambdaEventSourceMappingFilterCriteriaFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pattern: cdktf.stringToTerraform(struct.pattern),
    };
}
exports.lambdaEventSourceMappingFilterCriteriaFilterToTerraform = lambdaEventSourceMappingFilterCriteriaFilterToTerraform;
function lambdaEventSourceMappingFilterCriteriaFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pattern: {
            value: cdktf.stringToHclTerraform(struct.pattern),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingFilterCriteriaFilterToHclTerraform = lambdaEventSourceMappingFilterCriteriaFilterToHclTerraform;
class LambdaEventSourceMappingFilterCriteriaFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.pattern = this._pattern;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._pattern = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._pattern = value.pattern;
        }
    }
    get pattern() {
        return this.getStringAttribute('pattern');
    }
    set pattern(value) {
        this._pattern = value;
    }
    resetPattern() {
        this._pattern = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get patternInput() {
        return this._pattern;
    }
}
exports.LambdaEventSourceMappingFilterCriteriaFilterOutputReference = LambdaEventSourceMappingFilterCriteriaFilterOutputReference;
class LambdaEventSourceMappingFilterCriteriaFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LambdaEventSourceMappingFilterCriteriaFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LambdaEventSourceMappingFilterCriteriaFilterList = LambdaEventSourceMappingFilterCriteriaFilterList;
function lambdaEventSourceMappingFilterCriteriaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.listMapper(lambdaEventSourceMappingFilterCriteriaFilterToTerraform, true)(struct.filter),
    };
}
exports.lambdaEventSourceMappingFilterCriteriaToTerraform = lambdaEventSourceMappingFilterCriteriaToTerraform;
function lambdaEventSourceMappingFilterCriteriaToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter: {
            value: cdktf.listMapperHcl(lambdaEventSourceMappingFilterCriteriaFilterToHclTerraform, true)(struct.filter),
            isBlock: true,
            type: "set",
            storageClassType: "LambdaEventSourceMappingFilterCriteriaFilterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingFilterCriteriaToHclTerraform = lambdaEventSourceMappingFilterCriteriaToHclTerraform;
class LambdaEventSourceMappingFilterCriteriaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // filter - computed: false, optional: true, required: false
        this._filter = new LambdaEventSourceMappingFilterCriteriaFilterList(this, "filter", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filter.internalValue = value.filter;
        }
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
}
exports.LambdaEventSourceMappingFilterCriteriaOutputReference = LambdaEventSourceMappingFilterCriteriaOutputReference;
function lambdaEventSourceMappingScalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_concurrency: cdktf.numberToTerraform(struct.maximumConcurrency),
    };
}
exports.lambdaEventSourceMappingScalingConfigToTerraform = lambdaEventSourceMappingScalingConfigToTerraform;
function lambdaEventSourceMappingScalingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_concurrency: {
            value: cdktf.numberToHclTerraform(struct.maximumConcurrency),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingScalingConfigToHclTerraform = lambdaEventSourceMappingScalingConfigToHclTerraform;
class LambdaEventSourceMappingScalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumConcurrency = this._maximumConcurrency;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumConcurrency = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumConcurrency = value.maximumConcurrency;
        }
    }
    get maximumConcurrency() {
        return this.getNumberAttribute('maximum_concurrency');
    }
    set maximumConcurrency(value) {
        this._maximumConcurrency = value;
    }
    resetMaximumConcurrency() {
        this._maximumConcurrency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumConcurrencyInput() {
        return this._maximumConcurrency;
    }
}
exports.LambdaEventSourceMappingScalingConfigOutputReference = LambdaEventSourceMappingScalingConfigOutputReference;
function lambdaEventSourceMappingSelfManagedEventSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoints: cdktf.hashMapper(cdktf.stringToTerraform)(struct.endpoints),
    };
}
exports.lambdaEventSourceMappingSelfManagedEventSourceToTerraform = lambdaEventSourceMappingSelfManagedEventSourceToTerraform;
function lambdaEventSourceMappingSelfManagedEventSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        endpoints: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.endpoints),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingSelfManagedEventSourceToHclTerraform = lambdaEventSourceMappingSelfManagedEventSourceToHclTerraform;
class LambdaEventSourceMappingSelfManagedEventSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpoints !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpoints = this._endpoints;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpoints = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpoints = value.endpoints;
        }
    }
    get endpoints() {
        return this.getStringMapAttribute('endpoints');
    }
    set endpoints(value) {
        this._endpoints = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointsInput() {
        return this._endpoints;
    }
}
exports.LambdaEventSourceMappingSelfManagedEventSourceOutputReference = LambdaEventSourceMappingSelfManagedEventSourceOutputReference;
function lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        consumer_group_id: cdktf.stringToTerraform(struct.consumerGroupId),
    };
}
exports.lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToTerraform = lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToTerraform;
function lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        consumer_group_id: {
            value: cdktf.stringToHclTerraform(struct.consumerGroupId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToHclTerraform = lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToHclTerraform;
class LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._consumerGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.consumerGroupId = this._consumerGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._consumerGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._consumerGroupId = value.consumerGroupId;
        }
    }
    get consumerGroupId() {
        return this.getStringAttribute('consumer_group_id');
    }
    set consumerGroupId(value) {
        this._consumerGroupId = value;
    }
    resetConsumerGroupId() {
        this._consumerGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get consumerGroupIdInput() {
        return this._consumerGroupId;
    }
}
exports.LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigOutputReference = LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigOutputReference;
function lambdaEventSourceMappingSourceAccessConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.lambdaEventSourceMappingSourceAccessConfigurationToTerraform = lambdaEventSourceMappingSourceAccessConfigurationToTerraform;
function lambdaEventSourceMappingSourceAccessConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaEventSourceMappingSourceAccessConfigurationToHclTerraform = lambdaEventSourceMappingSourceAccessConfigurationToHclTerraform;
class LambdaEventSourceMappingSourceAccessConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._uri = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._uri = value.uri;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.LambdaEventSourceMappingSourceAccessConfigurationOutputReference = LambdaEventSourceMappingSourceAccessConfigurationOutputReference;
class LambdaEventSourceMappingSourceAccessConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LambdaEventSourceMappingSourceAccessConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LambdaEventSourceMappingSourceAccessConfigurationList = LambdaEventSourceMappingSourceAccessConfigurationList;
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/lambda_event_source_mapping aws_lambda_event_source_mapping}
*/
class LambdaEventSourceMapping extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LambdaEventSourceMapping resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LambdaEventSourceMapping to import
    * @param importFromId The id of the existing LambdaEventSourceMapping that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/lambda_event_source_mapping#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LambdaEventSourceMapping to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lambda_event_source_mapping", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/lambda_event_source_mapping aws_lambda_event_source_mapping} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LambdaEventSourceMappingConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_event_source_mapping',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.31.0',
                providerVersionConstraint: '~>5.31.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // amazon_managed_kafka_event_source_config - computed: false, optional: true, required: false
        this._amazonManagedKafkaEventSourceConfig = new LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigOutputReference(this, "amazon_managed_kafka_event_source_config");
        // destination_config - computed: false, optional: true, required: false
        this._destinationConfig = new LambdaEventSourceMappingDestinationConfigOutputReference(this, "destination_config");
        // document_db_event_source_config - computed: false, optional: true, required: false
        this._documentDbEventSourceConfig = new LambdaEventSourceMappingDocumentDbEventSourceConfigOutputReference(this, "document_db_event_source_config");
        // filter_criteria - computed: false, optional: true, required: false
        this._filterCriteria = new LambdaEventSourceMappingFilterCriteriaOutputReference(this, "filter_criteria");
        // scaling_config - computed: false, optional: true, required: false
        this._scalingConfig = new LambdaEventSourceMappingScalingConfigOutputReference(this, "scaling_config");
        // self_managed_event_source - computed: false, optional: true, required: false
        this._selfManagedEventSource = new LambdaEventSourceMappingSelfManagedEventSourceOutputReference(this, "self_managed_event_source");
        // self_managed_kafka_event_source_config - computed: false, optional: true, required: false
        this._selfManagedKafkaEventSourceConfig = new LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigOutputReference(this, "self_managed_kafka_event_source_config");
        // source_access_configuration - computed: false, optional: true, required: false
        this._sourceAccessConfiguration = new LambdaEventSourceMappingSourceAccessConfigurationList(this, "source_access_configuration", true);
        this._batchSize = config.batchSize;
        this._bisectBatchOnFunctionError = config.bisectBatchOnFunctionError;
        this._enabled = config.enabled;
        this._eventSourceArn = config.eventSourceArn;
        this._functionName = config.functionName;
        this._functionResponseTypes = config.functionResponseTypes;
        this._id = config.id;
        this._maximumBatchingWindowInSeconds = config.maximumBatchingWindowInSeconds;
        this._maximumRecordAgeInSeconds = config.maximumRecordAgeInSeconds;
        this._maximumRetryAttempts = config.maximumRetryAttempts;
        this._parallelizationFactor = config.parallelizationFactor;
        this._queues = config.queues;
        this._startingPosition = config.startingPosition;
        this._startingPositionTimestamp = config.startingPositionTimestamp;
        this._topics = config.topics;
        this._tumblingWindowInSeconds = config.tumblingWindowInSeconds;
        this._amazonManagedKafkaEventSourceConfig.internalValue = config.amazonManagedKafkaEventSourceConfig;
        this._destinationConfig.internalValue = config.destinationConfig;
        this._documentDbEventSourceConfig.internalValue = config.documentDbEventSourceConfig;
        this._filterCriteria.internalValue = config.filterCriteria;
        this._scalingConfig.internalValue = config.scalingConfig;
        this._selfManagedEventSource.internalValue = config.selfManagedEventSource;
        this._selfManagedKafkaEventSourceConfig.internalValue = config.selfManagedKafkaEventSourceConfig;
        this._sourceAccessConfiguration.internalValue = config.sourceAccessConfiguration;
    }
    get batchSize() {
        return this.getNumberAttribute('batch_size');
    }
    set batchSize(value) {
        this._batchSize = value;
    }
    resetBatchSize() {
        this._batchSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchSizeInput() {
        return this._batchSize;
    }
    get bisectBatchOnFunctionError() {
        return this.getBooleanAttribute('bisect_batch_on_function_error');
    }
    set bisectBatchOnFunctionError(value) {
        this._bisectBatchOnFunctionError = value;
    }
    resetBisectBatchOnFunctionError() {
        this._bisectBatchOnFunctionError = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bisectBatchOnFunctionErrorInput() {
        return this._bisectBatchOnFunctionError;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get eventSourceArn() {
        return this.getStringAttribute('event_source_arn');
    }
    set eventSourceArn(value) {
        this._eventSourceArn = value;
    }
    resetEventSourceArn() {
        this._eventSourceArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventSourceArnInput() {
        return this._eventSourceArn;
    }
    // function_arn - computed: true, optional: false, required: false
    get functionArn() {
        return this.getStringAttribute('function_arn');
    }
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get functionNameInput() {
        return this._functionName;
    }
    get functionResponseTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('function_response_types'));
    }
    set functionResponseTypes(value) {
        this._functionResponseTypes = value;
    }
    resetFunctionResponseTypes() {
        this._functionResponseTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get functionResponseTypesInput() {
        return this._functionResponseTypes;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_modified - computed: true, optional: false, required: false
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    // last_processing_result - computed: true, optional: false, required: false
    get lastProcessingResult() {
        return this.getStringAttribute('last_processing_result');
    }
    get maximumBatchingWindowInSeconds() {
        return this.getNumberAttribute('maximum_batching_window_in_seconds');
    }
    set maximumBatchingWindowInSeconds(value) {
        this._maximumBatchingWindowInSeconds = value;
    }
    resetMaximumBatchingWindowInSeconds() {
        this._maximumBatchingWindowInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumBatchingWindowInSecondsInput() {
        return this._maximumBatchingWindowInSeconds;
    }
    get maximumRecordAgeInSeconds() {
        return this.getNumberAttribute('maximum_record_age_in_seconds');
    }
    set maximumRecordAgeInSeconds(value) {
        this._maximumRecordAgeInSeconds = value;
    }
    resetMaximumRecordAgeInSeconds() {
        this._maximumRecordAgeInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumRecordAgeInSecondsInput() {
        return this._maximumRecordAgeInSeconds;
    }
    get maximumRetryAttempts() {
        return this.getNumberAttribute('maximum_retry_attempts');
    }
    set maximumRetryAttempts(value) {
        this._maximumRetryAttempts = value;
    }
    resetMaximumRetryAttempts() {
        this._maximumRetryAttempts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumRetryAttemptsInput() {
        return this._maximumRetryAttempts;
    }
    get parallelizationFactor() {
        return this.getNumberAttribute('parallelization_factor');
    }
    set parallelizationFactor(value) {
        this._parallelizationFactor = value;
    }
    resetParallelizationFactor() {
        this._parallelizationFactor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parallelizationFactorInput() {
        return this._parallelizationFactor;
    }
    get queues() {
        return this.getListAttribute('queues');
    }
    set queues(value) {
        this._queues = value;
    }
    resetQueues() {
        this._queues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queuesInput() {
        return this._queues;
    }
    get startingPosition() {
        return this.getStringAttribute('starting_position');
    }
    set startingPosition(value) {
        this._startingPosition = value;
    }
    resetStartingPosition() {
        this._startingPosition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startingPositionInput() {
        return this._startingPosition;
    }
    get startingPositionTimestamp() {
        return this.getStringAttribute('starting_position_timestamp');
    }
    set startingPositionTimestamp(value) {
        this._startingPositionTimestamp = value;
    }
    resetStartingPositionTimestamp() {
        this._startingPositionTimestamp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startingPositionTimestampInput() {
        return this._startingPositionTimestamp;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // state_transition_reason - computed: true, optional: false, required: false
    get stateTransitionReason() {
        return this.getStringAttribute('state_transition_reason');
    }
    get topics() {
        return cdktf.Fn.tolist(this.getListAttribute('topics'));
    }
    set topics(value) {
        this._topics = value;
    }
    resetTopics() {
        this._topics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get topicsInput() {
        return this._topics;
    }
    get tumblingWindowInSeconds() {
        return this.getNumberAttribute('tumbling_window_in_seconds');
    }
    set tumblingWindowInSeconds(value) {
        this._tumblingWindowInSeconds = value;
    }
    resetTumblingWindowInSeconds() {
        this._tumblingWindowInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tumblingWindowInSecondsInput() {
        return this._tumblingWindowInSeconds;
    }
    // uuid - computed: true, optional: false, required: false
    get uuid() {
        return this.getStringAttribute('uuid');
    }
    get amazonManagedKafkaEventSourceConfig() {
        return this._amazonManagedKafkaEventSourceConfig;
    }
    putAmazonManagedKafkaEventSourceConfig(value) {
        this._amazonManagedKafkaEventSourceConfig.internalValue = value;
    }
    resetAmazonManagedKafkaEventSourceConfig() {
        this._amazonManagedKafkaEventSourceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get amazonManagedKafkaEventSourceConfigInput() {
        return this._amazonManagedKafkaEventSourceConfig.internalValue;
    }
    get destinationConfig() {
        return this._destinationConfig;
    }
    putDestinationConfig(value) {
        this._destinationConfig.internalValue = value;
    }
    resetDestinationConfig() {
        this._destinationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationConfigInput() {
        return this._destinationConfig.internalValue;
    }
    get documentDbEventSourceConfig() {
        return this._documentDbEventSourceConfig;
    }
    putDocumentDbEventSourceConfig(value) {
        this._documentDbEventSourceConfig.internalValue = value;
    }
    resetDocumentDbEventSourceConfig() {
        this._documentDbEventSourceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get documentDbEventSourceConfigInput() {
        return this._documentDbEventSourceConfig.internalValue;
    }
    get filterCriteria() {
        return this._filterCriteria;
    }
    putFilterCriteria(value) {
        this._filterCriteria.internalValue = value;
    }
    resetFilterCriteria() {
        this._filterCriteria.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterCriteriaInput() {
        return this._filterCriteria.internalValue;
    }
    get scalingConfig() {
        return this._scalingConfig;
    }
    putScalingConfig(value) {
        this._scalingConfig.internalValue = value;
    }
    resetScalingConfig() {
        this._scalingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scalingConfigInput() {
        return this._scalingConfig.internalValue;
    }
    get selfManagedEventSource() {
        return this._selfManagedEventSource;
    }
    putSelfManagedEventSource(value) {
        this._selfManagedEventSource.internalValue = value;
    }
    resetSelfManagedEventSource() {
        this._selfManagedEventSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfManagedEventSourceInput() {
        return this._selfManagedEventSource.internalValue;
    }
    get selfManagedKafkaEventSourceConfig() {
        return this._selfManagedKafkaEventSourceConfig;
    }
    putSelfManagedKafkaEventSourceConfig(value) {
        this._selfManagedKafkaEventSourceConfig.internalValue = value;
    }
    resetSelfManagedKafkaEventSourceConfig() {
        this._selfManagedKafkaEventSourceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfManagedKafkaEventSourceConfigInput() {
        return this._selfManagedKafkaEventSourceConfig.internalValue;
    }
    get sourceAccessConfiguration() {
        return this._sourceAccessConfiguration;
    }
    putSourceAccessConfiguration(value) {
        this._sourceAccessConfiguration.internalValue = value;
    }
    resetSourceAccessConfiguration() {
        this._sourceAccessConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceAccessConfigurationInput() {
        return this._sourceAccessConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            batch_size: cdktf.numberToTerraform(this._batchSize),
            bisect_batch_on_function_error: cdktf.booleanToTerraform(this._bisectBatchOnFunctionError),
            enabled: cdktf.booleanToTerraform(this._enabled),
            event_source_arn: cdktf.stringToTerraform(this._eventSourceArn),
            function_name: cdktf.stringToTerraform(this._functionName),
            function_response_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._functionResponseTypes),
            id: cdktf.stringToTerraform(this._id),
            maximum_batching_window_in_seconds: cdktf.numberToTerraform(this._maximumBatchingWindowInSeconds),
            maximum_record_age_in_seconds: cdktf.numberToTerraform(this._maximumRecordAgeInSeconds),
            maximum_retry_attempts: cdktf.numberToTerraform(this._maximumRetryAttempts),
            parallelization_factor: cdktf.numberToTerraform(this._parallelizationFactor),
            queues: cdktf.listMapper(cdktf.stringToTerraform, false)(this._queues),
            starting_position: cdktf.stringToTerraform(this._startingPosition),
            starting_position_timestamp: cdktf.stringToTerraform(this._startingPositionTimestamp),
            topics: cdktf.listMapper(cdktf.stringToTerraform, false)(this._topics),
            tumbling_window_in_seconds: cdktf.numberToTerraform(this._tumblingWindowInSeconds),
            amazon_managed_kafka_event_source_config: lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToTerraform(this._amazonManagedKafkaEventSourceConfig.internalValue),
            destination_config: lambdaEventSourceMappingDestinationConfigToTerraform(this._destinationConfig.internalValue),
            document_db_event_source_config: lambdaEventSourceMappingDocumentDbEventSourceConfigToTerraform(this._documentDbEventSourceConfig.internalValue),
            filter_criteria: lambdaEventSourceMappingFilterCriteriaToTerraform(this._filterCriteria.internalValue),
            scaling_config: lambdaEventSourceMappingScalingConfigToTerraform(this._scalingConfig.internalValue),
            self_managed_event_source: lambdaEventSourceMappingSelfManagedEventSourceToTerraform(this._selfManagedEventSource.internalValue),
            self_managed_kafka_event_source_config: lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToTerraform(this._selfManagedKafkaEventSourceConfig.internalValue),
            source_access_configuration: cdktf.listMapper(lambdaEventSourceMappingSourceAccessConfigurationToTerraform, true)(this._sourceAccessConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            batch_size: {
                value: cdktf.numberToHclTerraform(this._batchSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            bisect_batch_on_function_error: {
                value: cdktf.booleanToHclTerraform(this._bisectBatchOnFunctionError),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            event_source_arn: {
                value: cdktf.stringToHclTerraform(this._eventSourceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            function_name: {
                value: cdktf.stringToHclTerraform(this._functionName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            function_response_types: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._functionResponseTypes),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            maximum_batching_window_in_seconds: {
                value: cdktf.numberToHclTerraform(this._maximumBatchingWindowInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            maximum_record_age_in_seconds: {
                value: cdktf.numberToHclTerraform(this._maximumRecordAgeInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            maximum_retry_attempts: {
                value: cdktf.numberToHclTerraform(this._maximumRetryAttempts),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            parallelization_factor: {
                value: cdktf.numberToHclTerraform(this._parallelizationFactor),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            queues: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._queues),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            starting_position: {
                value: cdktf.stringToHclTerraform(this._startingPosition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            starting_position_timestamp: {
                value: cdktf.stringToHclTerraform(this._startingPositionTimestamp),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            topics: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._topics),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tumbling_window_in_seconds: {
                value: cdktf.numberToHclTerraform(this._tumblingWindowInSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            amazon_managed_kafka_event_source_config: {
                value: lambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigToHclTerraform(this._amazonManagedKafkaEventSourceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingAmazonManagedKafkaEventSourceConfigList",
            },
            destination_config: {
                value: lambdaEventSourceMappingDestinationConfigToHclTerraform(this._destinationConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingDestinationConfigList",
            },
            document_db_event_source_config: {
                value: lambdaEventSourceMappingDocumentDbEventSourceConfigToHclTerraform(this._documentDbEventSourceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingDocumentDbEventSourceConfigList",
            },
            filter_criteria: {
                value: lambdaEventSourceMappingFilterCriteriaToHclTerraform(this._filterCriteria.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingFilterCriteriaList",
            },
            scaling_config: {
                value: lambdaEventSourceMappingScalingConfigToHclTerraform(this._scalingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingScalingConfigList",
            },
            self_managed_event_source: {
                value: lambdaEventSourceMappingSelfManagedEventSourceToHclTerraform(this._selfManagedEventSource.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingSelfManagedEventSourceList",
            },
            self_managed_kafka_event_source_config: {
                value: lambdaEventSourceMappingSelfManagedKafkaEventSourceConfigToHclTerraform(this._selfManagedKafkaEventSourceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaEventSourceMappingSelfManagedKafkaEventSourceConfigList",
            },
            source_access_configuration: {
                value: cdktf.listMapperHcl(lambdaEventSourceMappingSourceAccessConfigurationToHclTerraform, true)(this._sourceAccessConfiguration.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "LambdaEventSourceMappingSourceAccessConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LambdaEventSourceMapping = LambdaEventSourceMapping;
// =================
// STATIC PROPERTIES
// =================
LambdaEventSourceMapping.tfResourceType = "aws_lambda_event_source_mapping";
