// @cdktf/provider-aws/lib/opensearch-domain/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference = exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowToHclTerraform = exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowToTerraform = exports.OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference = exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToHclTerraform = exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToTerraform = exports.OpensearchDomainNodeToNodeEncryptionOutputReference = exports.opensearchDomainNodeToNodeEncryptionToHclTerraform = exports.opensearchDomainNodeToNodeEncryptionToTerraform = exports.OpensearchDomainLogPublishingOptionsList = exports.OpensearchDomainLogPublishingOptionsOutputReference = exports.opensearchDomainLogPublishingOptionsToHclTerraform = exports.opensearchDomainLogPublishingOptionsToTerraform = exports.OpensearchDomainEncryptAtRestOutputReference = exports.opensearchDomainEncryptAtRestToHclTerraform = exports.opensearchDomainEncryptAtRestToTerraform = exports.OpensearchDomainEbsOptionsOutputReference = exports.opensearchDomainEbsOptionsToHclTerraform = exports.opensearchDomainEbsOptionsToTerraform = exports.OpensearchDomainDomainEndpointOptionsOutputReference = exports.opensearchDomainDomainEndpointOptionsToHclTerraform = exports.opensearchDomainDomainEndpointOptionsToTerraform = exports.OpensearchDomainCognitoOptionsOutputReference = exports.opensearchDomainCognitoOptionsToHclTerraform = exports.opensearchDomainCognitoOptionsToTerraform = exports.OpensearchDomainClusterConfigOutputReference = exports.opensearchDomainClusterConfigToHclTerraform = exports.opensearchDomainClusterConfigToTerraform = exports.OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference = exports.opensearchDomainClusterConfigZoneAwarenessConfigToHclTerraform = exports.opensearchDomainClusterConfigZoneAwarenessConfigToTerraform = exports.OpensearchDomainClusterConfigColdStorageOptionsOutputReference = exports.opensearchDomainClusterConfigColdStorageOptionsToHclTerraform = exports.opensearchDomainClusterConfigColdStorageOptionsToTerraform = exports.OpensearchDomainAutoTuneOptionsOutputReference = exports.opensearchDomainAutoTuneOptionsToHclTerraform = exports.opensearchDomainAutoTuneOptionsToTerraform = exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleList = exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform = exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform = exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = exports.OpensearchDomainAdvancedSecurityOptionsOutputReference = exports.opensearchDomainAdvancedSecurityOptionsToHclTerraform = exports.opensearchDomainAdvancedSecurityOptionsToTerraform = exports.OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference = exports.opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform = exports.opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform = void 0;
exports.OpensearchDomain = exports.OpensearchDomainVpcOptionsOutputReference = exports.opensearchDomainVpcOptionsToHclTerraform = exports.opensearchDomainVpcOptionsToTerraform = exports.OpensearchDomainTimeoutsOutputReference = exports.opensearchDomainTimeoutsToHclTerraform = exports.opensearchDomainTimeoutsToTerraform = exports.OpensearchDomainSoftwareUpdateOptionsOutputReference = exports.opensearchDomainSoftwareUpdateOptionsToHclTerraform = exports.opensearchDomainSoftwareUpdateOptionsToTerraform = exports.OpensearchDomainSnapshotOptionsOutputReference = exports.opensearchDomainSnapshotOptionsToHclTerraform = exports.opensearchDomainSnapshotOptionsToTerraform = exports.OpensearchDomainOffPeakWindowOptionsOutputReference = exports.opensearchDomainOffPeakWindowOptionsToHclTerraform = exports.opensearchDomainOffPeakWindowOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        master_user_arn: cdktf.stringToTerraform(struct.masterUserArn),
        master_user_name: cdktf.stringToTerraform(struct.masterUserName),
        master_user_password: cdktf.stringToTerraform(struct.masterUserPassword),
    };
}
exports.opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform = opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform;
function opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        master_user_arn: {
            value: cdktf.stringToHclTerraform(struct.masterUserArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        master_user_name: {
            value: cdktf.stringToHclTerraform(struct.masterUserName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        master_user_password: {
            value: cdktf.stringToHclTerraform(struct.masterUserPassword),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform = opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform;
class OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._masterUserArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserArn = this._masterUserArn;
        }
        if (this._masterUserName !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserName = this._masterUserName;
        }
        if (this._masterUserPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserPassword = this._masterUserPassword;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._masterUserArn = undefined;
            this._masterUserName = undefined;
            this._masterUserPassword = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._masterUserArn = value.masterUserArn;
            this._masterUserName = value.masterUserName;
            this._masterUserPassword = value.masterUserPassword;
        }
    }
    get masterUserArn() {
        return this.getStringAttribute('master_user_arn');
    }
    set masterUserArn(value) {
        this._masterUserArn = value;
    }
    resetMasterUserArn() {
        this._masterUserArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserArnInput() {
        return this._masterUserArn;
    }
    get masterUserName() {
        return this.getStringAttribute('master_user_name');
    }
    set masterUserName(value) {
        this._masterUserName = value;
    }
    resetMasterUserName() {
        this._masterUserName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserNameInput() {
        return this._masterUserName;
    }
    get masterUserPassword() {
        return this.getStringAttribute('master_user_password');
    }
    set masterUserPassword(value) {
        this._masterUserPassword = value;
    }
    resetMasterUserPassword() {
        this._masterUserPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserPasswordInput() {
        return this._masterUserPassword;
    }
}
exports.OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference = OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference", version: "19.15.0" };
function opensearchDomainAdvancedSecurityOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        anonymous_auth_enabled: cdktf.booleanToTerraform(struct.anonymousAuthEnabled),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        internal_user_database_enabled: cdktf.booleanToTerraform(struct.internalUserDatabaseEnabled),
        master_user_options: opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToTerraform(struct.masterUserOptions),
    };
}
exports.opensearchDomainAdvancedSecurityOptionsToTerraform = opensearchDomainAdvancedSecurityOptionsToTerraform;
function opensearchDomainAdvancedSecurityOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        anonymous_auth_enabled: {
            value: cdktf.booleanToHclTerraform(struct.anonymousAuthEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        internal_user_database_enabled: {
            value: cdktf.booleanToHclTerraform(struct.internalUserDatabaseEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        master_user_options: {
            value: opensearchDomainAdvancedSecurityOptionsMasterUserOptionsToHclTerraform(struct.masterUserOptions),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainAdvancedSecurityOptionsToHclTerraform = opensearchDomainAdvancedSecurityOptionsToHclTerraform;
class OpensearchDomainAdvancedSecurityOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // master_user_options - computed: false, optional: true, required: false
        this._masterUserOptions = new OpensearchDomainAdvancedSecurityOptionsMasterUserOptionsOutputReference(this, "master_user_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._anonymousAuthEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.anonymousAuthEnabled = this._anonymousAuthEnabled;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._internalUserDatabaseEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.internalUserDatabaseEnabled = this._internalUserDatabaseEnabled;
        }
        if (this._masterUserOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterUserOptions = this._masterUserOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._anonymousAuthEnabled = undefined;
            this._enabled = undefined;
            this._internalUserDatabaseEnabled = undefined;
            this._masterUserOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._anonymousAuthEnabled = value.anonymousAuthEnabled;
            this._enabled = value.enabled;
            this._internalUserDatabaseEnabled = value.internalUserDatabaseEnabled;
            this._masterUserOptions.internalValue = value.masterUserOptions;
        }
    }
    get anonymousAuthEnabled() {
        return this.getBooleanAttribute('anonymous_auth_enabled');
    }
    set anonymousAuthEnabled(value) {
        this._anonymousAuthEnabled = value;
    }
    resetAnonymousAuthEnabled() {
        this._anonymousAuthEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get anonymousAuthEnabledInput() {
        return this._anonymousAuthEnabled;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get internalUserDatabaseEnabled() {
        return this.getBooleanAttribute('internal_user_database_enabled');
    }
    set internalUserDatabaseEnabled(value) {
        this._internalUserDatabaseEnabled = value;
    }
    resetInternalUserDatabaseEnabled() {
        this._internalUserDatabaseEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get internalUserDatabaseEnabledInput() {
        return this._internalUserDatabaseEnabled;
    }
    get masterUserOptions() {
        return this._masterUserOptions;
    }
    putMasterUserOptions(value) {
        this._masterUserOptions.internalValue = value;
    }
    resetMasterUserOptions() {
        this._masterUserOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterUserOptionsInput() {
        return this._masterUserOptions.internalValue;
    }
}
exports.OpensearchDomainAdvancedSecurityOptionsOutputReference = OpensearchDomainAdvancedSecurityOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
OpensearchDomainAdvancedSecurityOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAdvancedSecurityOptionsOutputReference", version: "19.15.0" };
function opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform;
function opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        unit: {
            value: cdktf.stringToHclTerraform(struct.unit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.numberToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform = opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform;
class OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference", version: "19.15.0" };
function opensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cron_expression_for_recurrence: cdktf.stringToTerraform(struct.cronExpressionForRecurrence),
        start_at: cdktf.stringToTerraform(struct.startAt),
        duration: opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform(struct.duration),
    };
}
exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = opensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform;
function opensearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cron_expression_for_recurrence: {
            value: cdktf.stringToHclTerraform(struct.cronExpressionForRecurrence),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_at: {
            value: cdktf.stringToHclTerraform(struct.startAt),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        duration: {
            value: opensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToHclTerraform(struct.duration),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform = opensearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform;
class OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // duration - computed: false, optional: false, required: true
        this._duration = new OpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference(this, "duration");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cronExpressionForRecurrence !== undefined) {
            hasAnyValues = true;
            internalValueResult.cronExpressionForRecurrence = this._cronExpressionForRecurrence;
        }
        if (this._startAt !== undefined) {
            hasAnyValues = true;
            internalValueResult.startAt = this._startAt;
        }
        if (this._duration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.duration = this._duration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cronExpressionForRecurrence = undefined;
            this._startAt = undefined;
            this._duration.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cronExpressionForRecurrence = value.cronExpressionForRecurrence;
            this._startAt = value.startAt;
            this._duration.internalValue = value.duration;
        }
    }
    get cronExpressionForRecurrence() {
        return this.getStringAttribute('cron_expression_for_recurrence');
    }
    set cronExpressionForRecurrence(value) {
        this._cronExpressionForRecurrence = value;
    }
    // Temporarily expose input value. Use with caution.
    get cronExpressionForRecurrenceInput() {
        return this._cronExpressionForRecurrence;
    }
    get startAt() {
        return this.getStringAttribute('start_at');
    }
    set startAt(value) {
        this._startAt = value;
    }
    // Temporarily expose input value. Use with caution.
    get startAtInput() {
        return this._startAt;
    }
    get duration() {
        return this._duration;
    }
    putDuration(value) {
        this._duration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration.internalValue;
    }
}
exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference;
_d = JSII_RTTI_SYMBOL_1;
OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference[_d] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference", version: "19.15.0" };
class OpensearchDomainAutoTuneOptionsMaintenanceScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OpensearchDomainAutoTuneOptionsMaintenanceScheduleList = OpensearchDomainAutoTuneOptionsMaintenanceScheduleList;
_e = JSII_RTTI_SYMBOL_1;
OpensearchDomainAutoTuneOptionsMaintenanceScheduleList[_e] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAutoTuneOptionsMaintenanceScheduleList", version: "19.15.0" };
function opensearchDomainAutoTuneOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        desired_state: cdktf.stringToTerraform(struct.desiredState),
        rollback_on_disable: cdktf.stringToTerraform(struct.rollbackOnDisable),
        use_off_peak_window: cdktf.booleanToTerraform(struct.useOffPeakWindow),
        maintenance_schedule: cdktf.listMapper(opensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform, true)(struct.maintenanceSchedule),
    };
}
exports.opensearchDomainAutoTuneOptionsToTerraform = opensearchDomainAutoTuneOptionsToTerraform;
function opensearchDomainAutoTuneOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        desired_state: {
            value: cdktf.stringToHclTerraform(struct.desiredState),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rollback_on_disable: {
            value: cdktf.stringToHclTerraform(struct.rollbackOnDisable),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        use_off_peak_window: {
            value: cdktf.booleanToHclTerraform(struct.useOffPeakWindow),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        maintenance_schedule: {
            value: cdktf.listMapperHcl(opensearchDomainAutoTuneOptionsMaintenanceScheduleToHclTerraform, true)(struct.maintenanceSchedule),
            isBlock: true,
            type: "set",
            storageClassType: "OpensearchDomainAutoTuneOptionsMaintenanceScheduleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainAutoTuneOptionsToHclTerraform = opensearchDomainAutoTuneOptionsToHclTerraform;
class OpensearchDomainAutoTuneOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // maintenance_schedule - computed: false, optional: true, required: false
        this._maintenanceSchedule = new OpensearchDomainAutoTuneOptionsMaintenanceScheduleList(this, "maintenance_schedule", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._desiredState !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredState = this._desiredState;
        }
        if (this._rollbackOnDisable !== undefined) {
            hasAnyValues = true;
            internalValueResult.rollbackOnDisable = this._rollbackOnDisable;
        }
        if (this._useOffPeakWindow !== undefined) {
            hasAnyValues = true;
            internalValueResult.useOffPeakWindow = this._useOffPeakWindow;
        }
        if (this._maintenanceSchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maintenanceSchedule = this._maintenanceSchedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._desiredState = undefined;
            this._rollbackOnDisable = undefined;
            this._useOffPeakWindow = undefined;
            this._maintenanceSchedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._desiredState = value.desiredState;
            this._rollbackOnDisable = value.rollbackOnDisable;
            this._useOffPeakWindow = value.useOffPeakWindow;
            this._maintenanceSchedule.internalValue = value.maintenanceSchedule;
        }
    }
    get desiredState() {
        return this.getStringAttribute('desired_state');
    }
    set desiredState(value) {
        this._desiredState = value;
    }
    // Temporarily expose input value. Use with caution.
    get desiredStateInput() {
        return this._desiredState;
    }
    get rollbackOnDisable() {
        return this.getStringAttribute('rollback_on_disable');
    }
    set rollbackOnDisable(value) {
        this._rollbackOnDisable = value;
    }
    resetRollbackOnDisable() {
        this._rollbackOnDisable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rollbackOnDisableInput() {
        return this._rollbackOnDisable;
    }
    get useOffPeakWindow() {
        return this.getBooleanAttribute('use_off_peak_window');
    }
    set useOffPeakWindow(value) {
        this._useOffPeakWindow = value;
    }
    resetUseOffPeakWindow() {
        this._useOffPeakWindow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useOffPeakWindowInput() {
        return this._useOffPeakWindow;
    }
    get maintenanceSchedule() {
        return this._maintenanceSchedule;
    }
    putMaintenanceSchedule(value) {
        this._maintenanceSchedule.internalValue = value;
    }
    resetMaintenanceSchedule() {
        this._maintenanceSchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceScheduleInput() {
        return this._maintenanceSchedule.internalValue;
    }
}
exports.OpensearchDomainAutoTuneOptionsOutputReference = OpensearchDomainAutoTuneOptionsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
OpensearchDomainAutoTuneOptionsOutputReference[_f] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainAutoTuneOptionsOutputReference", version: "19.15.0" };
function opensearchDomainClusterConfigColdStorageOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.opensearchDomainClusterConfigColdStorageOptionsToTerraform = opensearchDomainClusterConfigColdStorageOptionsToTerraform;
function opensearchDomainClusterConfigColdStorageOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainClusterConfigColdStorageOptionsToHclTerraform = opensearchDomainClusterConfigColdStorageOptionsToHclTerraform;
class OpensearchDomainClusterConfigColdStorageOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.OpensearchDomainClusterConfigColdStorageOptionsOutputReference = OpensearchDomainClusterConfigColdStorageOptionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
OpensearchDomainClusterConfigColdStorageOptionsOutputReference[_g] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainClusterConfigColdStorageOptionsOutputReference", version: "19.15.0" };
function opensearchDomainClusterConfigZoneAwarenessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_zone_count: cdktf.numberToTerraform(struct.availabilityZoneCount),
    };
}
exports.opensearchDomainClusterConfigZoneAwarenessConfigToTerraform = opensearchDomainClusterConfigZoneAwarenessConfigToTerraform;
function opensearchDomainClusterConfigZoneAwarenessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        availability_zone_count: {
            value: cdktf.numberToHclTerraform(struct.availabilityZoneCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainClusterConfigZoneAwarenessConfigToHclTerraform = opensearchDomainClusterConfigZoneAwarenessConfigToHclTerraform;
class OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availabilityZoneCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityZoneCount = this._availabilityZoneCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availabilityZoneCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availabilityZoneCount = value.availabilityZoneCount;
        }
    }
    get availabilityZoneCount() {
        return this.getNumberAttribute('availability_zone_count');
    }
    set availabilityZoneCount(value) {
        this._availabilityZoneCount = value;
    }
    resetAvailabilityZoneCount() {
        this._availabilityZoneCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneCountInput() {
        return this._availabilityZoneCount;
    }
}
exports.OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference = OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference", version: "19.15.0" };
function opensearchDomainClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dedicated_master_count: cdktf.numberToTerraform(struct.dedicatedMasterCount),
        dedicated_master_enabled: cdktf.booleanToTerraform(struct.dedicatedMasterEnabled),
        dedicated_master_type: cdktf.stringToTerraform(struct.dedicatedMasterType),
        instance_count: cdktf.numberToTerraform(struct.instanceCount),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        multi_az_with_standby_enabled: cdktf.booleanToTerraform(struct.multiAzWithStandbyEnabled),
        warm_count: cdktf.numberToTerraform(struct.warmCount),
        warm_enabled: cdktf.booleanToTerraform(struct.warmEnabled),
        warm_type: cdktf.stringToTerraform(struct.warmType),
        zone_awareness_enabled: cdktf.booleanToTerraform(struct.zoneAwarenessEnabled),
        cold_storage_options: opensearchDomainClusterConfigColdStorageOptionsToTerraform(struct.coldStorageOptions),
        zone_awareness_config: opensearchDomainClusterConfigZoneAwarenessConfigToTerraform(struct.zoneAwarenessConfig),
    };
}
exports.opensearchDomainClusterConfigToTerraform = opensearchDomainClusterConfigToTerraform;
function opensearchDomainClusterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dedicated_master_count: {
            value: cdktf.numberToHclTerraform(struct.dedicatedMasterCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        dedicated_master_enabled: {
            value: cdktf.booleanToHclTerraform(struct.dedicatedMasterEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        dedicated_master_type: {
            value: cdktf.stringToHclTerraform(struct.dedicatedMasterType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        instance_count: {
            value: cdktf.numberToHclTerraform(struct.instanceCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        multi_az_with_standby_enabled: {
            value: cdktf.booleanToHclTerraform(struct.multiAzWithStandbyEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        warm_count: {
            value: cdktf.numberToHclTerraform(struct.warmCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        warm_enabled: {
            value: cdktf.booleanToHclTerraform(struct.warmEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        warm_type: {
            value: cdktf.stringToHclTerraform(struct.warmType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        zone_awareness_enabled: {
            value: cdktf.booleanToHclTerraform(struct.zoneAwarenessEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        cold_storage_options: {
            value: opensearchDomainClusterConfigColdStorageOptionsToHclTerraform(struct.coldStorageOptions),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainClusterConfigColdStorageOptionsList",
        },
        zone_awareness_config: {
            value: opensearchDomainClusterConfigZoneAwarenessConfigToHclTerraform(struct.zoneAwarenessConfig),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainClusterConfigZoneAwarenessConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainClusterConfigToHclTerraform = opensearchDomainClusterConfigToHclTerraform;
class OpensearchDomainClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cold_storage_options - computed: false, optional: true, required: false
        this._coldStorageOptions = new OpensearchDomainClusterConfigColdStorageOptionsOutputReference(this, "cold_storage_options");
        // zone_awareness_config - computed: false, optional: true, required: false
        this._zoneAwarenessConfig = new OpensearchDomainClusterConfigZoneAwarenessConfigOutputReference(this, "zone_awareness_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dedicatedMasterCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterCount = this._dedicatedMasterCount;
        }
        if (this._dedicatedMasterEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterEnabled = this._dedicatedMasterEnabled;
        }
        if (this._dedicatedMasterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.dedicatedMasterType = this._dedicatedMasterType;
        }
        if (this._instanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceCount = this._instanceCount;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._multiAzWithStandbyEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.multiAzWithStandbyEnabled = this._multiAzWithStandbyEnabled;
        }
        if (this._warmCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmCount = this._warmCount;
        }
        if (this._warmEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmEnabled = this._warmEnabled;
        }
        if (this._warmType !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmType = this._warmType;
        }
        if (this._zoneAwarenessEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneAwarenessEnabled = this._zoneAwarenessEnabled;
        }
        if (this._coldStorageOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.coldStorageOptions = this._coldStorageOptions?.internalValue;
        }
        if (this._zoneAwarenessConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneAwarenessConfig = this._zoneAwarenessConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dedicatedMasterCount = undefined;
            this._dedicatedMasterEnabled = undefined;
            this._dedicatedMasterType = undefined;
            this._instanceCount = undefined;
            this._instanceType = undefined;
            this._multiAzWithStandbyEnabled = undefined;
            this._warmCount = undefined;
            this._warmEnabled = undefined;
            this._warmType = undefined;
            this._zoneAwarenessEnabled = undefined;
            this._coldStorageOptions.internalValue = undefined;
            this._zoneAwarenessConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dedicatedMasterCount = value.dedicatedMasterCount;
            this._dedicatedMasterEnabled = value.dedicatedMasterEnabled;
            this._dedicatedMasterType = value.dedicatedMasterType;
            this._instanceCount = value.instanceCount;
            this._instanceType = value.instanceType;
            this._multiAzWithStandbyEnabled = value.multiAzWithStandbyEnabled;
            this._warmCount = value.warmCount;
            this._warmEnabled = value.warmEnabled;
            this._warmType = value.warmType;
            this._zoneAwarenessEnabled = value.zoneAwarenessEnabled;
            this._coldStorageOptions.internalValue = value.coldStorageOptions;
            this._zoneAwarenessConfig.internalValue = value.zoneAwarenessConfig;
        }
    }
    get dedicatedMasterCount() {
        return this.getNumberAttribute('dedicated_master_count');
    }
    set dedicatedMasterCount(value) {
        this._dedicatedMasterCount = value;
    }
    resetDedicatedMasterCount() {
        this._dedicatedMasterCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterCountInput() {
        return this._dedicatedMasterCount;
    }
    get dedicatedMasterEnabled() {
        return this.getBooleanAttribute('dedicated_master_enabled');
    }
    set dedicatedMasterEnabled(value) {
        this._dedicatedMasterEnabled = value;
    }
    resetDedicatedMasterEnabled() {
        this._dedicatedMasterEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterEnabledInput() {
        return this._dedicatedMasterEnabled;
    }
    get dedicatedMasterType() {
        return this.getStringAttribute('dedicated_master_type');
    }
    set dedicatedMasterType(value) {
        this._dedicatedMasterType = value;
    }
    resetDedicatedMasterType() {
        this._dedicatedMasterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedMasterTypeInput() {
        return this._dedicatedMasterType;
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get multiAzWithStandbyEnabled() {
        return this.getBooleanAttribute('multi_az_with_standby_enabled');
    }
    set multiAzWithStandbyEnabled(value) {
        this._multiAzWithStandbyEnabled = value;
    }
    resetMultiAzWithStandbyEnabled() {
        this._multiAzWithStandbyEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multiAzWithStandbyEnabledInput() {
        return this._multiAzWithStandbyEnabled;
    }
    get warmCount() {
        return this.getNumberAttribute('warm_count');
    }
    set warmCount(value) {
        this._warmCount = value;
    }
    resetWarmCount() {
        this._warmCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmCountInput() {
        return this._warmCount;
    }
    get warmEnabled() {
        return this.getBooleanAttribute('warm_enabled');
    }
    set warmEnabled(value) {
        this._warmEnabled = value;
    }
    resetWarmEnabled() {
        this._warmEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmEnabledInput() {
        return this._warmEnabled;
    }
    get warmType() {
        return this.getStringAttribute('warm_type');
    }
    set warmType(value) {
        this._warmType = value;
    }
    resetWarmType() {
        this._warmType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmTypeInput() {
        return this._warmType;
    }
    get zoneAwarenessEnabled() {
        return this.getBooleanAttribute('zone_awareness_enabled');
    }
    set zoneAwarenessEnabled(value) {
        this._zoneAwarenessEnabled = value;
    }
    resetZoneAwarenessEnabled() {
        this._zoneAwarenessEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneAwarenessEnabledInput() {
        return this._zoneAwarenessEnabled;
    }
    get coldStorageOptions() {
        return this._coldStorageOptions;
    }
    putColdStorageOptions(value) {
        this._coldStorageOptions.internalValue = value;
    }
    resetColdStorageOptions() {
        this._coldStorageOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coldStorageOptionsInput() {
        return this._coldStorageOptions.internalValue;
    }
    get zoneAwarenessConfig() {
        return this._zoneAwarenessConfig;
    }
    putZoneAwarenessConfig(value) {
        this._zoneAwarenessConfig.internalValue = value;
    }
    resetZoneAwarenessConfig() {
        this._zoneAwarenessConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneAwarenessConfigInput() {
        return this._zoneAwarenessConfig.internalValue;
    }
}
exports.OpensearchDomainClusterConfigOutputReference = OpensearchDomainClusterConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
OpensearchDomainClusterConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainClusterConfigOutputReference", version: "19.15.0" };
function opensearchDomainCognitoOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        identity_pool_id: cdktf.stringToTerraform(struct.identityPoolId),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        user_pool_id: cdktf.stringToTerraform(struct.userPoolId),
    };
}
exports.opensearchDomainCognitoOptionsToTerraform = opensearchDomainCognitoOptionsToTerraform;
function opensearchDomainCognitoOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        identity_pool_id: {
            value: cdktf.stringToHclTerraform(struct.identityPoolId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_pool_id: {
            value: cdktf.stringToHclTerraform(struct.userPoolId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainCognitoOptionsToHclTerraform = opensearchDomainCognitoOptionsToHclTerraform;
class OpensearchDomainCognitoOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._identityPoolId !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityPoolId = this._identityPoolId;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._userPoolId !== undefined) {
            hasAnyValues = true;
            internalValueResult.userPoolId = this._userPoolId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._identityPoolId = undefined;
            this._roleArn = undefined;
            this._userPoolId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._identityPoolId = value.identityPoolId;
            this._roleArn = value.roleArn;
            this._userPoolId = value.userPoolId;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get identityPoolId() {
        return this.getStringAttribute('identity_pool_id');
    }
    set identityPoolId(value) {
        this._identityPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityPoolIdInput() {
        return this._identityPoolId;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
    set userPoolId(value) {
        this._userPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userPoolIdInput() {
        return this._userPoolId;
    }
}
exports.OpensearchDomainCognitoOptionsOutputReference = OpensearchDomainCognitoOptionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
OpensearchDomainCognitoOptionsOutputReference[_k] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainCognitoOptionsOutputReference", version: "19.15.0" };
function opensearchDomainDomainEndpointOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_endpoint: cdktf.stringToTerraform(struct.customEndpoint),
        custom_endpoint_certificate_arn: cdktf.stringToTerraform(struct.customEndpointCertificateArn),
        custom_endpoint_enabled: cdktf.booleanToTerraform(struct.customEndpointEnabled),
        enforce_https: cdktf.booleanToTerraform(struct.enforceHttps),
        tls_security_policy: cdktf.stringToTerraform(struct.tlsSecurityPolicy),
    };
}
exports.opensearchDomainDomainEndpointOptionsToTerraform = opensearchDomainDomainEndpointOptionsToTerraform;
function opensearchDomainDomainEndpointOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        custom_endpoint: {
            value: cdktf.stringToHclTerraform(struct.customEndpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        custom_endpoint_certificate_arn: {
            value: cdktf.stringToHclTerraform(struct.customEndpointCertificateArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        custom_endpoint_enabled: {
            value: cdktf.booleanToHclTerraform(struct.customEndpointEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enforce_https: {
            value: cdktf.booleanToHclTerraform(struct.enforceHttps),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        tls_security_policy: {
            value: cdktf.stringToHclTerraform(struct.tlsSecurityPolicy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainDomainEndpointOptionsToHclTerraform = opensearchDomainDomainEndpointOptionsToHclTerraform;
class OpensearchDomainDomainEndpointOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpoint = this._customEndpoint;
        }
        if (this._customEndpointCertificateArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpointCertificateArn = this._customEndpointCertificateArn;
        }
        if (this._customEndpointEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.customEndpointEnabled = this._customEndpointEnabled;
        }
        if (this._enforceHttps !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceHttps = this._enforceHttps;
        }
        if (this._tlsSecurityPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.tlsSecurityPolicy = this._tlsSecurityPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customEndpoint = undefined;
            this._customEndpointCertificateArn = undefined;
            this._customEndpointEnabled = undefined;
            this._enforceHttps = undefined;
            this._tlsSecurityPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customEndpoint = value.customEndpoint;
            this._customEndpointCertificateArn = value.customEndpointCertificateArn;
            this._customEndpointEnabled = value.customEndpointEnabled;
            this._enforceHttps = value.enforceHttps;
            this._tlsSecurityPolicy = value.tlsSecurityPolicy;
        }
    }
    get customEndpoint() {
        return this.getStringAttribute('custom_endpoint');
    }
    set customEndpoint(value) {
        this._customEndpoint = value;
    }
    resetCustomEndpoint() {
        this._customEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointInput() {
        return this._customEndpoint;
    }
    get customEndpointCertificateArn() {
        return this.getStringAttribute('custom_endpoint_certificate_arn');
    }
    set customEndpointCertificateArn(value) {
        this._customEndpointCertificateArn = value;
    }
    resetCustomEndpointCertificateArn() {
        this._customEndpointCertificateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointCertificateArnInput() {
        return this._customEndpointCertificateArn;
    }
    get customEndpointEnabled() {
        return this.getBooleanAttribute('custom_endpoint_enabled');
    }
    set customEndpointEnabled(value) {
        this._customEndpointEnabled = value;
    }
    resetCustomEndpointEnabled() {
        this._customEndpointEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customEndpointEnabledInput() {
        return this._customEndpointEnabled;
    }
    get enforceHttps() {
        return this.getBooleanAttribute('enforce_https');
    }
    set enforceHttps(value) {
        this._enforceHttps = value;
    }
    resetEnforceHttps() {
        this._enforceHttps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceHttpsInput() {
        return this._enforceHttps;
    }
    get tlsSecurityPolicy() {
        return this.getStringAttribute('tls_security_policy');
    }
    set tlsSecurityPolicy(value) {
        this._tlsSecurityPolicy = value;
    }
    resetTlsSecurityPolicy() {
        this._tlsSecurityPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsSecurityPolicyInput() {
        return this._tlsSecurityPolicy;
    }
}
exports.OpensearchDomainDomainEndpointOptionsOutputReference = OpensearchDomainDomainEndpointOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
OpensearchDomainDomainEndpointOptionsOutputReference[_l] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainDomainEndpointOptionsOutputReference", version: "19.15.0" };
function opensearchDomainEbsOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ebs_enabled: cdktf.booleanToTerraform(struct.ebsEnabled),
        iops: cdktf.numberToTerraform(struct.iops),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.opensearchDomainEbsOptionsToTerraform = opensearchDomainEbsOptionsToTerraform;
function opensearchDomainEbsOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ebs_enabled: {
            value: cdktf.booleanToHclTerraform(struct.ebsEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        iops: {
            value: cdktf.numberToHclTerraform(struct.iops),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        throughput: {
            value: cdktf.numberToHclTerraform(struct.throughput),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        volume_size: {
            value: cdktf.numberToHclTerraform(struct.volumeSize),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        volume_type: {
            value: cdktf.stringToHclTerraform(struct.volumeType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainEbsOptionsToHclTerraform = opensearchDomainEbsOptionsToHclTerraform;
class OpensearchDomainEbsOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ebsEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsEnabled = this._ebsEnabled;
        }
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._throughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._volumeSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeSize = this._volumeSize;
        }
        if (this._volumeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeType = this._volumeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ebsEnabled = undefined;
            this._iops = undefined;
            this._throughput = undefined;
            this._volumeSize = undefined;
            this._volumeType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ebsEnabled = value.ebsEnabled;
            this._iops = value.iops;
            this._throughput = value.throughput;
            this._volumeSize = value.volumeSize;
            this._volumeType = value.volumeType;
        }
    }
    get ebsEnabled() {
        return this.getBooleanAttribute('ebs_enabled');
    }
    set ebsEnabled(value) {
        this._ebsEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get ebsEnabledInput() {
        return this._ebsEnabled;
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    set throughput(value) {
        this._throughput = value;
    }
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get throughputInput() {
        return this._throughput;
    }
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    set volumeSize(value) {
        this._volumeSize = value;
    }
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInput() {
        return this._volumeSize;
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
}
exports.OpensearchDomainEbsOptionsOutputReference = OpensearchDomainEbsOptionsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
OpensearchDomainEbsOptionsOutputReference[_m] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainEbsOptionsOutputReference", version: "19.15.0" };
function opensearchDomainEncryptAtRestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.opensearchDomainEncryptAtRestToTerraform = opensearchDomainEncryptAtRestToTerraform;
function opensearchDomainEncryptAtRestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainEncryptAtRestToHclTerraform = opensearchDomainEncryptAtRestToHclTerraform;
class OpensearchDomainEncryptAtRestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._kmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.OpensearchDomainEncryptAtRestOutputReference = OpensearchDomainEncryptAtRestOutputReference;
_o = JSII_RTTI_SYMBOL_1;
OpensearchDomainEncryptAtRestOutputReference[_o] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainEncryptAtRestOutputReference", version: "19.15.0" };
function opensearchDomainLogPublishingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_log_group_arn: cdktf.stringToTerraform(struct.cloudwatchLogGroupArn),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_type: cdktf.stringToTerraform(struct.logType),
    };
}
exports.opensearchDomainLogPublishingOptionsToTerraform = opensearchDomainLogPublishingOptionsToTerraform;
function opensearchDomainLogPublishingOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cloudwatch_log_group_arn: {
            value: cdktf.stringToHclTerraform(struct.cloudwatchLogGroupArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        log_type: {
            value: cdktf.stringToHclTerraform(struct.logType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainLogPublishingOptionsToHclTerraform = opensearchDomainLogPublishingOptionsToHclTerraform;
class OpensearchDomainLogPublishingOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudwatchLogGroupArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogGroupArn = this._cloudwatchLogGroupArn;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logType !== undefined) {
            hasAnyValues = true;
            internalValueResult.logType = this._logType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cloudwatchLogGroupArn = undefined;
            this._enabled = undefined;
            this._logType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cloudwatchLogGroupArn = value.cloudwatchLogGroupArn;
            this._enabled = value.enabled;
            this._logType = value.logType;
        }
    }
    get cloudwatchLogGroupArn() {
        return this.getStringAttribute('cloudwatch_log_group_arn');
    }
    set cloudwatchLogGroupArn(value) {
        this._cloudwatchLogGroupArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogGroupArnInput() {
        return this._cloudwatchLogGroupArn;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get logType() {
        return this.getStringAttribute('log_type');
    }
    set logType(value) {
        this._logType = value;
    }
    // Temporarily expose input value. Use with caution.
    get logTypeInput() {
        return this._logType;
    }
}
exports.OpensearchDomainLogPublishingOptionsOutputReference = OpensearchDomainLogPublishingOptionsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
OpensearchDomainLogPublishingOptionsOutputReference[_p] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainLogPublishingOptionsOutputReference", version: "19.15.0" };
class OpensearchDomainLogPublishingOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OpensearchDomainLogPublishingOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OpensearchDomainLogPublishingOptionsList = OpensearchDomainLogPublishingOptionsList;
_q = JSII_RTTI_SYMBOL_1;
OpensearchDomainLogPublishingOptionsList[_q] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainLogPublishingOptionsList", version: "19.15.0" };
function opensearchDomainNodeToNodeEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.opensearchDomainNodeToNodeEncryptionToTerraform = opensearchDomainNodeToNodeEncryptionToTerraform;
function opensearchDomainNodeToNodeEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainNodeToNodeEncryptionToHclTerraform = opensearchDomainNodeToNodeEncryptionToHclTerraform;
class OpensearchDomainNodeToNodeEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.OpensearchDomainNodeToNodeEncryptionOutputReference = OpensearchDomainNodeToNodeEncryptionOutputReference;
_r = JSII_RTTI_SYMBOL_1;
OpensearchDomainNodeToNodeEncryptionOutputReference[_r] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainNodeToNodeEncryptionOutputReference", version: "19.15.0" };
function opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToTerraform = opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToTerraform;
function opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hours: {
            value: cdktf.numberToHclTerraform(struct.hours),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minutes: {
            value: cdktf.numberToHclTerraform(struct.minutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToHclTerraform = opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToHclTerraform;
class OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference = OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference;
_s = JSII_RTTI_SYMBOL_1;
OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference[_s] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference", version: "19.15.0" };
function opensearchDomainOffPeakWindowOptionsOffPeakWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        window_start_time: opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToTerraform(struct.windowStartTime),
    };
}
exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowToTerraform = opensearchDomainOffPeakWindowOptionsOffPeakWindowToTerraform;
function opensearchDomainOffPeakWindowOptionsOffPeakWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        window_start_time: {
            value: opensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeToHclTerraform(struct.windowStartTime),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainOffPeakWindowOptionsOffPeakWindowToHclTerraform = opensearchDomainOffPeakWindowOptionsOffPeakWindowToHclTerraform;
class OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // window_start_time - computed: false, optional: true, required: false
        this._windowStartTime = new OpensearchDomainOffPeakWindowOptionsOffPeakWindowWindowStartTimeOutputReference(this, "window_start_time");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._windowStartTime?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.windowStartTime = this._windowStartTime?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._windowStartTime.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._windowStartTime.internalValue = value.windowStartTime;
        }
    }
    get windowStartTime() {
        return this._windowStartTime;
    }
    putWindowStartTime(value) {
        this._windowStartTime.internalValue = value;
    }
    resetWindowStartTime() {
        this._windowStartTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get windowStartTimeInput() {
        return this._windowStartTime.internalValue;
    }
}
exports.OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference = OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference;
_t = JSII_RTTI_SYMBOL_1;
OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference[_t] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference", version: "19.15.0" };
function opensearchDomainOffPeakWindowOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        off_peak_window: opensearchDomainOffPeakWindowOptionsOffPeakWindowToTerraform(struct.offPeakWindow),
    };
}
exports.opensearchDomainOffPeakWindowOptionsToTerraform = opensearchDomainOffPeakWindowOptionsToTerraform;
function opensearchDomainOffPeakWindowOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        off_peak_window: {
            value: opensearchDomainOffPeakWindowOptionsOffPeakWindowToHclTerraform(struct.offPeakWindow),
            isBlock: true,
            type: "list",
            storageClassType: "OpensearchDomainOffPeakWindowOptionsOffPeakWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainOffPeakWindowOptionsToHclTerraform = opensearchDomainOffPeakWindowOptionsToHclTerraform;
class OpensearchDomainOffPeakWindowOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // off_peak_window - computed: false, optional: true, required: false
        this._offPeakWindow = new OpensearchDomainOffPeakWindowOptionsOffPeakWindowOutputReference(this, "off_peak_window");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._offPeakWindow?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.offPeakWindow = this._offPeakWindow?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._offPeakWindow.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._offPeakWindow.internalValue = value.offPeakWindow;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get offPeakWindow() {
        return this._offPeakWindow;
    }
    putOffPeakWindow(value) {
        this._offPeakWindow.internalValue = value;
    }
    resetOffPeakWindow() {
        this._offPeakWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get offPeakWindowInput() {
        return this._offPeakWindow.internalValue;
    }
}
exports.OpensearchDomainOffPeakWindowOptionsOutputReference = OpensearchDomainOffPeakWindowOptionsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
OpensearchDomainOffPeakWindowOptionsOutputReference[_u] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainOffPeakWindowOptionsOutputReference", version: "19.15.0" };
function opensearchDomainSnapshotOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        automated_snapshot_start_hour: cdktf.numberToTerraform(struct.automatedSnapshotStartHour),
    };
}
exports.opensearchDomainSnapshotOptionsToTerraform = opensearchDomainSnapshotOptionsToTerraform;
function opensearchDomainSnapshotOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        automated_snapshot_start_hour: {
            value: cdktf.numberToHclTerraform(struct.automatedSnapshotStartHour),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainSnapshotOptionsToHclTerraform = opensearchDomainSnapshotOptionsToHclTerraform;
class OpensearchDomainSnapshotOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._automatedSnapshotStartHour !== undefined) {
            hasAnyValues = true;
            internalValueResult.automatedSnapshotStartHour = this._automatedSnapshotStartHour;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._automatedSnapshotStartHour = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._automatedSnapshotStartHour = value.automatedSnapshotStartHour;
        }
    }
    get automatedSnapshotStartHour() {
        return this.getNumberAttribute('automated_snapshot_start_hour');
    }
    set automatedSnapshotStartHour(value) {
        this._automatedSnapshotStartHour = value;
    }
    // Temporarily expose input value. Use with caution.
    get automatedSnapshotStartHourInput() {
        return this._automatedSnapshotStartHour;
    }
}
exports.OpensearchDomainSnapshotOptionsOutputReference = OpensearchDomainSnapshotOptionsOutputReference;
_v = JSII_RTTI_SYMBOL_1;
OpensearchDomainSnapshotOptionsOutputReference[_v] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainSnapshotOptionsOutputReference", version: "19.15.0" };
function opensearchDomainSoftwareUpdateOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_software_update_enabled: cdktf.booleanToTerraform(struct.autoSoftwareUpdateEnabled),
    };
}
exports.opensearchDomainSoftwareUpdateOptionsToTerraform = opensearchDomainSoftwareUpdateOptionsToTerraform;
function opensearchDomainSoftwareUpdateOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_software_update_enabled: {
            value: cdktf.booleanToHclTerraform(struct.autoSoftwareUpdateEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainSoftwareUpdateOptionsToHclTerraform = opensearchDomainSoftwareUpdateOptionsToHclTerraform;
class OpensearchDomainSoftwareUpdateOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoSoftwareUpdateEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoSoftwareUpdateEnabled = this._autoSoftwareUpdateEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoSoftwareUpdateEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoSoftwareUpdateEnabled = value.autoSoftwareUpdateEnabled;
        }
    }
    get autoSoftwareUpdateEnabled() {
        return this.getBooleanAttribute('auto_software_update_enabled');
    }
    set autoSoftwareUpdateEnabled(value) {
        this._autoSoftwareUpdateEnabled = value;
    }
    resetAutoSoftwareUpdateEnabled() {
        this._autoSoftwareUpdateEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoSoftwareUpdateEnabledInput() {
        return this._autoSoftwareUpdateEnabled;
    }
}
exports.OpensearchDomainSoftwareUpdateOptionsOutputReference = OpensearchDomainSoftwareUpdateOptionsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
OpensearchDomainSoftwareUpdateOptionsOutputReference[_w] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainSoftwareUpdateOptionsOutputReference", version: "19.15.0" };
function opensearchDomainTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.opensearchDomainTimeoutsToTerraform = opensearchDomainTimeoutsToTerraform;
function opensearchDomainTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainTimeoutsToHclTerraform = opensearchDomainTimeoutsToHclTerraform;
class OpensearchDomainTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.OpensearchDomainTimeoutsOutputReference = OpensearchDomainTimeoutsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
OpensearchDomainTimeoutsOutputReference[_x] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainTimeoutsOutputReference", version: "19.15.0" };
function opensearchDomainVpcOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.opensearchDomainVpcOptionsToTerraform = opensearchDomainVpcOptionsToTerraform;
function opensearchDomainVpcOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.opensearchDomainVpcOptionsToHclTerraform = opensearchDomainVpcOptionsToHclTerraform;
class OpensearchDomainVpcOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    // availability_zones - computed: true, optional: false, required: false
    get availabilityZones() {
        return cdktf.Fn.tolist(this.getListAttribute('availability_zones'));
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.OpensearchDomainVpcOptionsOutputReference = OpensearchDomainVpcOptionsOutputReference;
_y = JSII_RTTI_SYMBOL_1;
OpensearchDomainVpcOptionsOutputReference[_y] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomainVpcOptionsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearch_domain aws_opensearch_domain}
*/
class OpensearchDomain extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a OpensearchDomain resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the OpensearchDomain to import
    * @param importFromId The id of the existing OpensearchDomain that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearch_domain#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the OpensearchDomain to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_opensearch_domain", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opensearch_domain aws_opensearch_domain} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OpensearchDomainConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opensearch_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // advanced_security_options - computed: false, optional: true, required: false
        this._advancedSecurityOptions = new OpensearchDomainAdvancedSecurityOptionsOutputReference(this, "advanced_security_options");
        // auto_tune_options - computed: false, optional: true, required: false
        this._autoTuneOptions = new OpensearchDomainAutoTuneOptionsOutputReference(this, "auto_tune_options");
        // cluster_config - computed: false, optional: true, required: false
        this._clusterConfig = new OpensearchDomainClusterConfigOutputReference(this, "cluster_config");
        // cognito_options - computed: false, optional: true, required: false
        this._cognitoOptions = new OpensearchDomainCognitoOptionsOutputReference(this, "cognito_options");
        // domain_endpoint_options - computed: false, optional: true, required: false
        this._domainEndpointOptions = new OpensearchDomainDomainEndpointOptionsOutputReference(this, "domain_endpoint_options");
        // ebs_options - computed: false, optional: true, required: false
        this._ebsOptions = new OpensearchDomainEbsOptionsOutputReference(this, "ebs_options");
        // encrypt_at_rest - computed: false, optional: true, required: false
        this._encryptAtRest = new OpensearchDomainEncryptAtRestOutputReference(this, "encrypt_at_rest");
        // log_publishing_options - computed: false, optional: true, required: false
        this._logPublishingOptions = new OpensearchDomainLogPublishingOptionsList(this, "log_publishing_options", true);
        // node_to_node_encryption - computed: false, optional: true, required: false
        this._nodeToNodeEncryption = new OpensearchDomainNodeToNodeEncryptionOutputReference(this, "node_to_node_encryption");
        // off_peak_window_options - computed: false, optional: true, required: false
        this._offPeakWindowOptions = new OpensearchDomainOffPeakWindowOptionsOutputReference(this, "off_peak_window_options");
        // snapshot_options - computed: false, optional: true, required: false
        this._snapshotOptions = new OpensearchDomainSnapshotOptionsOutputReference(this, "snapshot_options");
        // software_update_options - computed: false, optional: true, required: false
        this._softwareUpdateOptions = new OpensearchDomainSoftwareUpdateOptionsOutputReference(this, "software_update_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new OpensearchDomainTimeoutsOutputReference(this, "timeouts");
        // vpc_options - computed: false, optional: true, required: false
        this._vpcOptions = new OpensearchDomainVpcOptionsOutputReference(this, "vpc_options");
        this._accessPolicies = config.accessPolicies;
        this._advancedOptions = config.advancedOptions;
        this._domainName = config.domainName;
        this._engineVersion = config.engineVersion;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._advancedSecurityOptions.internalValue = config.advancedSecurityOptions;
        this._autoTuneOptions.internalValue = config.autoTuneOptions;
        this._clusterConfig.internalValue = config.clusterConfig;
        this._cognitoOptions.internalValue = config.cognitoOptions;
        this._domainEndpointOptions.internalValue = config.domainEndpointOptions;
        this._ebsOptions.internalValue = config.ebsOptions;
        this._encryptAtRest.internalValue = config.encryptAtRest;
        this._logPublishingOptions.internalValue = config.logPublishingOptions;
        this._nodeToNodeEncryption.internalValue = config.nodeToNodeEncryption;
        this._offPeakWindowOptions.internalValue = config.offPeakWindowOptions;
        this._snapshotOptions.internalValue = config.snapshotOptions;
        this._softwareUpdateOptions.internalValue = config.softwareUpdateOptions;
        this._timeouts.internalValue = config.timeouts;
        this._vpcOptions.internalValue = config.vpcOptions;
    }
    get accessPolicies() {
        return this.getStringAttribute('access_policies');
    }
    set accessPolicies(value) {
        this._accessPolicies = value;
    }
    resetAccessPolicies() {
        this._accessPolicies = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessPoliciesInput() {
        return this._accessPolicies;
    }
    get advancedOptions() {
        return this.getStringMapAttribute('advanced_options');
    }
    set advancedOptions(value) {
        this._advancedOptions = value;
    }
    resetAdvancedOptions() {
        this._advancedOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedOptionsInput() {
        return this._advancedOptions;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // dashboard_endpoint - computed: true, optional: false, required: false
    get dashboardEndpoint() {
        return this.getStringAttribute('dashboard_endpoint');
    }
    // domain_id - computed: true, optional: false, required: false
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    set engineVersion(value) {
        this._engineVersion = value;
    }
    resetEngineVersion() {
        this._engineVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engineVersionInput() {
        return this._engineVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kibana_endpoint - computed: true, optional: false, required: false
    get kibanaEndpoint() {
        return this.getStringAttribute('kibana_endpoint');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get advancedSecurityOptions() {
        return this._advancedSecurityOptions;
    }
    putAdvancedSecurityOptions(value) {
        this._advancedSecurityOptions.internalValue = value;
    }
    resetAdvancedSecurityOptions() {
        this._advancedSecurityOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedSecurityOptionsInput() {
        return this._advancedSecurityOptions.internalValue;
    }
    get autoTuneOptions() {
        return this._autoTuneOptions;
    }
    putAutoTuneOptions(value) {
        this._autoTuneOptions.internalValue = value;
    }
    resetAutoTuneOptions() {
        this._autoTuneOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoTuneOptionsInput() {
        return this._autoTuneOptions.internalValue;
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    putClusterConfig(value) {
        this._clusterConfig.internalValue = value;
    }
    resetClusterConfig() {
        this._clusterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterConfigInput() {
        return this._clusterConfig.internalValue;
    }
    get cognitoOptions() {
        return this._cognitoOptions;
    }
    putCognitoOptions(value) {
        this._cognitoOptions.internalValue = value;
    }
    resetCognitoOptions() {
        this._cognitoOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cognitoOptionsInput() {
        return this._cognitoOptions.internalValue;
    }
    get domainEndpointOptions() {
        return this._domainEndpointOptions;
    }
    putDomainEndpointOptions(value) {
        this._domainEndpointOptions.internalValue = value;
    }
    resetDomainEndpointOptions() {
        this._domainEndpointOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainEndpointOptionsInput() {
        return this._domainEndpointOptions.internalValue;
    }
    get ebsOptions() {
        return this._ebsOptions;
    }
    putEbsOptions(value) {
        this._ebsOptions.internalValue = value;
    }
    resetEbsOptions() {
        this._ebsOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsOptionsInput() {
        return this._ebsOptions.internalValue;
    }
    get encryptAtRest() {
        return this._encryptAtRest;
    }
    putEncryptAtRest(value) {
        this._encryptAtRest.internalValue = value;
    }
    resetEncryptAtRest() {
        this._encryptAtRest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptAtRestInput() {
        return this._encryptAtRest.internalValue;
    }
    get logPublishingOptions() {
        return this._logPublishingOptions;
    }
    putLogPublishingOptions(value) {
        this._logPublishingOptions.internalValue = value;
    }
    resetLogPublishingOptions() {
        this._logPublishingOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logPublishingOptionsInput() {
        return this._logPublishingOptions.internalValue;
    }
    get nodeToNodeEncryption() {
        return this._nodeToNodeEncryption;
    }
    putNodeToNodeEncryption(value) {
        this._nodeToNodeEncryption.internalValue = value;
    }
    resetNodeToNodeEncryption() {
        this._nodeToNodeEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeToNodeEncryptionInput() {
        return this._nodeToNodeEncryption.internalValue;
    }
    get offPeakWindowOptions() {
        return this._offPeakWindowOptions;
    }
    putOffPeakWindowOptions(value) {
        this._offPeakWindowOptions.internalValue = value;
    }
    resetOffPeakWindowOptions() {
        this._offPeakWindowOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get offPeakWindowOptionsInput() {
        return this._offPeakWindowOptions.internalValue;
    }
    get snapshotOptions() {
        return this._snapshotOptions;
    }
    putSnapshotOptions(value) {
        this._snapshotOptions.internalValue = value;
    }
    resetSnapshotOptions() {
        this._snapshotOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotOptionsInput() {
        return this._snapshotOptions.internalValue;
    }
    get softwareUpdateOptions() {
        return this._softwareUpdateOptions;
    }
    putSoftwareUpdateOptions(value) {
        this._softwareUpdateOptions.internalValue = value;
    }
    resetSoftwareUpdateOptions() {
        this._softwareUpdateOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get softwareUpdateOptionsInput() {
        return this._softwareUpdateOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcOptions() {
        return this._vpcOptions;
    }
    putVpcOptions(value) {
        this._vpcOptions.internalValue = value;
    }
    resetVpcOptions() {
        this._vpcOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcOptionsInput() {
        return this._vpcOptions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_policies: cdktf.stringToTerraform(this._accessPolicies),
            advanced_options: cdktf.hashMapper(cdktf.stringToTerraform)(this._advancedOptions),
            domain_name: cdktf.stringToTerraform(this._domainName),
            engine_version: cdktf.stringToTerraform(this._engineVersion),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            advanced_security_options: opensearchDomainAdvancedSecurityOptionsToTerraform(this._advancedSecurityOptions.internalValue),
            auto_tune_options: opensearchDomainAutoTuneOptionsToTerraform(this._autoTuneOptions.internalValue),
            cluster_config: opensearchDomainClusterConfigToTerraform(this._clusterConfig.internalValue),
            cognito_options: opensearchDomainCognitoOptionsToTerraform(this._cognitoOptions.internalValue),
            domain_endpoint_options: opensearchDomainDomainEndpointOptionsToTerraform(this._domainEndpointOptions.internalValue),
            ebs_options: opensearchDomainEbsOptionsToTerraform(this._ebsOptions.internalValue),
            encrypt_at_rest: opensearchDomainEncryptAtRestToTerraform(this._encryptAtRest.internalValue),
            log_publishing_options: cdktf.listMapper(opensearchDomainLogPublishingOptionsToTerraform, true)(this._logPublishingOptions.internalValue),
            node_to_node_encryption: opensearchDomainNodeToNodeEncryptionToTerraform(this._nodeToNodeEncryption.internalValue),
            off_peak_window_options: opensearchDomainOffPeakWindowOptionsToTerraform(this._offPeakWindowOptions.internalValue),
            snapshot_options: opensearchDomainSnapshotOptionsToTerraform(this._snapshotOptions.internalValue),
            software_update_options: opensearchDomainSoftwareUpdateOptionsToTerraform(this._softwareUpdateOptions.internalValue),
            timeouts: opensearchDomainTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_options: opensearchDomainVpcOptionsToTerraform(this._vpcOptions.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_policies: {
                value: cdktf.stringToHclTerraform(this._accessPolicies),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            advanced_options: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._advancedOptions),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            engine_version: {
                value: cdktf.stringToHclTerraform(this._engineVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            advanced_security_options: {
                value: opensearchDomainAdvancedSecurityOptionsToHclTerraform(this._advancedSecurityOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainAdvancedSecurityOptionsList",
            },
            auto_tune_options: {
                value: opensearchDomainAutoTuneOptionsToHclTerraform(this._autoTuneOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainAutoTuneOptionsList",
            },
            cluster_config: {
                value: opensearchDomainClusterConfigToHclTerraform(this._clusterConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainClusterConfigList",
            },
            cognito_options: {
                value: opensearchDomainCognitoOptionsToHclTerraform(this._cognitoOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainCognitoOptionsList",
            },
            domain_endpoint_options: {
                value: opensearchDomainDomainEndpointOptionsToHclTerraform(this._domainEndpointOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainDomainEndpointOptionsList",
            },
            ebs_options: {
                value: opensearchDomainEbsOptionsToHclTerraform(this._ebsOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainEbsOptionsList",
            },
            encrypt_at_rest: {
                value: opensearchDomainEncryptAtRestToHclTerraform(this._encryptAtRest.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainEncryptAtRestList",
            },
            log_publishing_options: {
                value: cdktf.listMapperHcl(opensearchDomainLogPublishingOptionsToHclTerraform, true)(this._logPublishingOptions.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "OpensearchDomainLogPublishingOptionsList",
            },
            node_to_node_encryption: {
                value: opensearchDomainNodeToNodeEncryptionToHclTerraform(this._nodeToNodeEncryption.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainNodeToNodeEncryptionList",
            },
            off_peak_window_options: {
                value: opensearchDomainOffPeakWindowOptionsToHclTerraform(this._offPeakWindowOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainOffPeakWindowOptionsList",
            },
            snapshot_options: {
                value: opensearchDomainSnapshotOptionsToHclTerraform(this._snapshotOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainSnapshotOptionsList",
            },
            software_update_options: {
                value: opensearchDomainSoftwareUpdateOptionsToHclTerraform(this._softwareUpdateOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainSoftwareUpdateOptionsList",
            },
            timeouts: {
                value: opensearchDomainTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "OpensearchDomainTimeouts",
            },
            vpc_options: {
                value: opensearchDomainVpcOptionsToHclTerraform(this._vpcOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "OpensearchDomainVpcOptionsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.OpensearchDomain = OpensearchDomain;
_z = JSII_RTTI_SYMBOL_1;
OpensearchDomain[_z] = { fqn: "@cdktf/provider-aws.opensearchDomain.OpensearchDomain", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
OpensearchDomain.tfResourceType = "aws_opensearch_domain";
