// @cdktf/provider-aws/lib/sagemaker-pipeline/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerPipeline = exports.SagemakerPipelinePipelineDefinitionS3LocationOutputReference = exports.sagemakerPipelinePipelineDefinitionS3LocationToHclTerraform = exports.sagemakerPipelinePipelineDefinitionS3LocationToTerraform = exports.SagemakerPipelineParallelismConfigurationOutputReference = exports.sagemakerPipelineParallelismConfigurationToHclTerraform = exports.sagemakerPipelineParallelismConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerPipelineParallelismConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_parallel_execution_steps: cdktf.numberToTerraform(struct.maxParallelExecutionSteps),
    };
}
exports.sagemakerPipelineParallelismConfigurationToTerraform = sagemakerPipelineParallelismConfigurationToTerraform;
function sagemakerPipelineParallelismConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_parallel_execution_steps: {
            value: cdktf.numberToHclTerraform(struct.maxParallelExecutionSteps),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerPipelineParallelismConfigurationToHclTerraform = sagemakerPipelineParallelismConfigurationToHclTerraform;
class SagemakerPipelineParallelismConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxParallelExecutionSteps !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxParallelExecutionSteps = this._maxParallelExecutionSteps;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxParallelExecutionSteps = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxParallelExecutionSteps = value.maxParallelExecutionSteps;
        }
    }
    get maxParallelExecutionSteps() {
        return this.getNumberAttribute('max_parallel_execution_steps');
    }
    set maxParallelExecutionSteps(value) {
        this._maxParallelExecutionSteps = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxParallelExecutionStepsInput() {
        return this._maxParallelExecutionSteps;
    }
}
exports.SagemakerPipelineParallelismConfigurationOutputReference = SagemakerPipelineParallelismConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerPipelineParallelismConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerPipeline.SagemakerPipelineParallelismConfigurationOutputReference", version: "19.15.0" };
function sagemakerPipelinePipelineDefinitionS3LocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        object_key: cdktf.stringToTerraform(struct.objectKey),
        version_id: cdktf.stringToTerraform(struct.versionId),
    };
}
exports.sagemakerPipelinePipelineDefinitionS3LocationToTerraform = sagemakerPipelinePipelineDefinitionS3LocationToTerraform;
function sagemakerPipelinePipelineDefinitionS3LocationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket: {
            value: cdktf.stringToHclTerraform(struct.bucket),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        object_key: {
            value: cdktf.stringToHclTerraform(struct.objectKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        version_id: {
            value: cdktf.stringToHclTerraform(struct.versionId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerPipelinePipelineDefinitionS3LocationToHclTerraform = sagemakerPipelinePipelineDefinitionS3LocationToHclTerraform;
class SagemakerPipelinePipelineDefinitionS3LocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._objectKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.objectKey = this._objectKey;
        }
        if (this._versionId !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionId = this._versionId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._objectKey = undefined;
            this._versionId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._objectKey = value.objectKey;
            this._versionId = value.versionId;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get objectKey() {
        return this.getStringAttribute('object_key');
    }
    set objectKey(value) {
        this._objectKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get objectKeyInput() {
        return this._objectKey;
    }
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    set versionId(value) {
        this._versionId = value;
    }
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionIdInput() {
        return this._versionId;
    }
}
exports.SagemakerPipelinePipelineDefinitionS3LocationOutputReference = SagemakerPipelinePipelineDefinitionS3LocationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SagemakerPipelinePipelineDefinitionS3LocationOutputReference[_b] = { fqn: "@cdktf/provider-aws.sagemakerPipeline.SagemakerPipelinePipelineDefinitionS3LocationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_pipeline aws_sagemaker_pipeline}
*/
class SagemakerPipeline extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SagemakerPipeline resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SagemakerPipeline to import
    * @param importFromId The id of the existing SagemakerPipeline that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_pipeline#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SagemakerPipeline to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sagemaker_pipeline", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_pipeline aws_sagemaker_pipeline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerPipelineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // parallelism_configuration - computed: false, optional: true, required: false
        this._parallelismConfiguration = new SagemakerPipelineParallelismConfigurationOutputReference(this, "parallelism_configuration");
        // pipeline_definition_s3_location - computed: false, optional: true, required: false
        this._pipelineDefinitionS3Location = new SagemakerPipelinePipelineDefinitionS3LocationOutputReference(this, "pipeline_definition_s3_location");
        this._id = config.id;
        this._pipelineDefinition = config.pipelineDefinition;
        this._pipelineDescription = config.pipelineDescription;
        this._pipelineDisplayName = config.pipelineDisplayName;
        this._pipelineName = config.pipelineName;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._parallelismConfiguration.internalValue = config.parallelismConfiguration;
        this._pipelineDefinitionS3Location.internalValue = config.pipelineDefinitionS3Location;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get pipelineDefinition() {
        return this.getStringAttribute('pipeline_definition');
    }
    set pipelineDefinition(value) {
        this._pipelineDefinition = value;
    }
    resetPipelineDefinition() {
        this._pipelineDefinition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineDefinitionInput() {
        return this._pipelineDefinition;
    }
    get pipelineDescription() {
        return this.getStringAttribute('pipeline_description');
    }
    set pipelineDescription(value) {
        this._pipelineDescription = value;
    }
    resetPipelineDescription() {
        this._pipelineDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineDescriptionInput() {
        return this._pipelineDescription;
    }
    get pipelineDisplayName() {
        return this.getStringAttribute('pipeline_display_name');
    }
    set pipelineDisplayName(value) {
        this._pipelineDisplayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineDisplayNameInput() {
        return this._pipelineDisplayName;
    }
    get pipelineName() {
        return this.getStringAttribute('pipeline_name');
    }
    set pipelineName(value) {
        this._pipelineName = value;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineNameInput() {
        return this._pipelineName;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get parallelismConfiguration() {
        return this._parallelismConfiguration;
    }
    putParallelismConfiguration(value) {
        this._parallelismConfiguration.internalValue = value;
    }
    resetParallelismConfiguration() {
        this._parallelismConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parallelismConfigurationInput() {
        return this._parallelismConfiguration.internalValue;
    }
    get pipelineDefinitionS3Location() {
        return this._pipelineDefinitionS3Location;
    }
    putPipelineDefinitionS3Location(value) {
        this._pipelineDefinitionS3Location.internalValue = value;
    }
    resetPipelineDefinitionS3Location() {
        this._pipelineDefinitionS3Location.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineDefinitionS3LocationInput() {
        return this._pipelineDefinitionS3Location.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            pipeline_definition: cdktf.stringToTerraform(this._pipelineDefinition),
            pipeline_description: cdktf.stringToTerraform(this._pipelineDescription),
            pipeline_display_name: cdktf.stringToTerraform(this._pipelineDisplayName),
            pipeline_name: cdktf.stringToTerraform(this._pipelineName),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            parallelism_configuration: sagemakerPipelineParallelismConfigurationToTerraform(this._parallelismConfiguration.internalValue),
            pipeline_definition_s3_location: sagemakerPipelinePipelineDefinitionS3LocationToTerraform(this._pipelineDefinitionS3Location.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_definition: {
                value: cdktf.stringToHclTerraform(this._pipelineDefinition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_description: {
                value: cdktf.stringToHclTerraform(this._pipelineDescription),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_display_name: {
                value: cdktf.stringToHclTerraform(this._pipelineDisplayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_name: {
                value: cdktf.stringToHclTerraform(this._pipelineName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            parallelism_configuration: {
                value: sagemakerPipelineParallelismConfigurationToHclTerraform(this._parallelismConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerPipelineParallelismConfigurationList",
            },
            pipeline_definition_s3_location: {
                value: sagemakerPipelinePipelineDefinitionS3LocationToHclTerraform(this._pipelineDefinitionS3Location.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerPipelinePipelineDefinitionS3LocationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SagemakerPipeline = SagemakerPipeline;
_c = JSII_RTTI_SYMBOL_1;
SagemakerPipeline[_c] = { fqn: "@cdktf/provider-aws.sagemakerPipeline.SagemakerPipeline", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SagemakerPipeline.tfResourceType = "aws_sagemaker_pipeline";
