// @cdktf/provider-aws/lib/datasync-location-s3/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncLocationS3 = exports.DatasyncLocationS3S3ConfigOutputReference = exports.datasyncLocationS3S3ConfigToHclTerraform = exports.datasyncLocationS3S3ConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncLocationS3S3ConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_access_role_arn: cdktf.stringToTerraform(struct.bucketAccessRoleArn),
    };
}
exports.datasyncLocationS3S3ConfigToTerraform = datasyncLocationS3S3ConfigToTerraform;
function datasyncLocationS3S3ConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_access_role_arn: {
            value: cdktf.stringToHclTerraform(struct.bucketAccessRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncLocationS3S3ConfigToHclTerraform = datasyncLocationS3S3ConfigToHclTerraform;
class DatasyncLocationS3S3ConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketAccessRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketAccessRoleArn = this._bucketAccessRoleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketAccessRoleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketAccessRoleArn = value.bucketAccessRoleArn;
        }
    }
    get bucketAccessRoleArn() {
        return this.getStringAttribute('bucket_access_role_arn');
    }
    set bucketAccessRoleArn(value) {
        this._bucketAccessRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketAccessRoleArnInput() {
        return this._bucketAccessRoleArn;
    }
}
exports.DatasyncLocationS3S3ConfigOutputReference = DatasyncLocationS3S3ConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncLocationS3S3ConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasyncLocationS3.DatasyncLocationS3S3ConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_s3 aws_datasync_location_s3}
*/
class DatasyncLocationS3 extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DatasyncLocationS3 resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DatasyncLocationS3 to import
    * @param importFromId The id of the existing DatasyncLocationS3 that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_s3#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DatasyncLocationS3 to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_datasync_location_s3", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_location_s3 aws_datasync_location_s3} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DatasyncLocationS3Config
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_location_s3',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // s3_config - computed: false, optional: false, required: true
        this._s3Config = new DatasyncLocationS3S3ConfigOutputReference(this, "s3_config");
        this._agentArns = config.agentArns;
        this._id = config.id;
        this._s3BucketArn = config.s3BucketArn;
        this._s3StorageClass = config.s3StorageClass;
        this._subdirectory = config.subdirectory;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._s3Config.internalValue = config.s3Config;
    }
    get agentArns() {
        return cdktf.Fn.tolist(this.getListAttribute('agent_arns'));
    }
    set agentArns(value) {
        this._agentArns = value;
    }
    resetAgentArns() {
        this._agentArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get agentArnsInput() {
        return this._agentArns;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get s3BucketArn() {
        return this.getStringAttribute('s3_bucket_arn');
    }
    set s3BucketArn(value) {
        this._s3BucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketArnInput() {
        return this._s3BucketArn;
    }
    get s3StorageClass() {
        return this.getStringAttribute('s3_storage_class');
    }
    set s3StorageClass(value) {
        this._s3StorageClass = value;
    }
    resetS3StorageClass() {
        this._s3StorageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3StorageClassInput() {
        return this._s3StorageClass;
    }
    get subdirectory() {
        return this.getStringAttribute('subdirectory');
    }
    set subdirectory(value) {
        this._subdirectory = value;
    }
    // Temporarily expose input value. Use with caution.
    get subdirectoryInput() {
        return this._subdirectory;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
    get s3Config() {
        return this._s3Config;
    }
    putS3Config(value) {
        this._s3Config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3ConfigInput() {
        return this._s3Config.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            agent_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._agentArns),
            id: cdktf.stringToTerraform(this._id),
            s3_bucket_arn: cdktf.stringToTerraform(this._s3BucketArn),
            s3_storage_class: cdktf.stringToTerraform(this._s3StorageClass),
            subdirectory: cdktf.stringToTerraform(this._subdirectory),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            s3_config: datasyncLocationS3S3ConfigToTerraform(this._s3Config.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            agent_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._agentArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_bucket_arn: {
                value: cdktf.stringToHclTerraform(this._s3BucketArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            s3_storage_class: {
                value: cdktf.stringToHclTerraform(this._s3StorageClass),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subdirectory: {
                value: cdktf.stringToHclTerraform(this._subdirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            s3_config: {
                value: datasyncLocationS3S3ConfigToHclTerraform(this._s3Config.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncLocationS3S3ConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DatasyncLocationS3 = DatasyncLocationS3;
_b = JSII_RTTI_SYMBOL_1;
DatasyncLocationS3[_b] = { fqn: "@cdktf/provider-aws.datasyncLocationS3.DatasyncLocationS3", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DatasyncLocationS3.tfResourceType = "aws_datasync_location_s3";
