// @cdktf/provider-aws/lib/apprunner-observability-configuration/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApprunnerObservabilityConfiguration = exports.ApprunnerObservabilityConfigurationTraceConfigurationOutputReference = exports.apprunnerObservabilityConfigurationTraceConfigurationToHclTerraform = exports.apprunnerObservabilityConfigurationTraceConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apprunnerObservabilityConfigurationTraceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        vendor: cdktf.stringToTerraform(struct.vendor),
    };
}
exports.apprunnerObservabilityConfigurationTraceConfigurationToTerraform = apprunnerObservabilityConfigurationTraceConfigurationToTerraform;
function apprunnerObservabilityConfigurationTraceConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        vendor: {
            value: cdktf.stringToHclTerraform(struct.vendor),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apprunnerObservabilityConfigurationTraceConfigurationToHclTerraform = apprunnerObservabilityConfigurationTraceConfigurationToHclTerraform;
class ApprunnerObservabilityConfigurationTraceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._vendor !== undefined) {
            hasAnyValues = true;
            internalValueResult.vendor = this._vendor;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._vendor = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._vendor = value.vendor;
        }
    }
    get vendor() {
        return this.getStringAttribute('vendor');
    }
    set vendor(value) {
        this._vendor = value;
    }
    resetVendor() {
        this._vendor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vendorInput() {
        return this._vendor;
    }
}
exports.ApprunnerObservabilityConfigurationTraceConfigurationOutputReference = ApprunnerObservabilityConfigurationTraceConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApprunnerObservabilityConfigurationTraceConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apprunnerObservabilityConfiguration.ApprunnerObservabilityConfigurationTraceConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_observability_configuration aws_apprunner_observability_configuration}
*/
class ApprunnerObservabilityConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApprunnerObservabilityConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApprunnerObservabilityConfiguration to import
    * @param importFromId The id of the existing ApprunnerObservabilityConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_observability_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApprunnerObservabilityConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_apprunner_observability_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/apprunner_observability_configuration aws_apprunner_observability_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApprunnerObservabilityConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apprunner_observability_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // trace_configuration - computed: false, optional: true, required: false
        this._traceConfiguration = new ApprunnerObservabilityConfigurationTraceConfigurationOutputReference(this, "trace_configuration");
        this._id = config.id;
        this._observabilityConfigurationName = config.observabilityConfigurationName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._traceConfiguration.internalValue = config.traceConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // latest - computed: true, optional: false, required: false
    get latest() {
        return this.getBooleanAttribute('latest');
    }
    get observabilityConfigurationName() {
        return this.getStringAttribute('observability_configuration_name');
    }
    set observabilityConfigurationName(value) {
        this._observabilityConfigurationName = value;
    }
    // Temporarily expose input value. Use with caution.
    get observabilityConfigurationNameInput() {
        return this._observabilityConfigurationName;
    }
    // observability_configuration_revision - computed: true, optional: false, required: false
    get observabilityConfigurationRevision() {
        return this.getNumberAttribute('observability_configuration_revision');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get traceConfiguration() {
        return this._traceConfiguration;
    }
    putTraceConfiguration(value) {
        this._traceConfiguration.internalValue = value;
    }
    resetTraceConfiguration() {
        this._traceConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get traceConfigurationInput() {
        return this._traceConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            observability_configuration_name: cdktf.stringToTerraform(this._observabilityConfigurationName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            trace_configuration: apprunnerObservabilityConfigurationTraceConfigurationToTerraform(this._traceConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            observability_configuration_name: {
                value: cdktf.stringToHclTerraform(this._observabilityConfigurationName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            trace_configuration: {
                value: apprunnerObservabilityConfigurationTraceConfigurationToHclTerraform(this._traceConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApprunnerObservabilityConfigurationTraceConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApprunnerObservabilityConfiguration = ApprunnerObservabilityConfiguration;
_b = JSII_RTTI_SYMBOL_1;
ApprunnerObservabilityConfiguration[_b] = { fqn: "@cdktf/provider-aws.apprunnerObservabilityConfiguration.ApprunnerObservabilityConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
ApprunnerObservabilityConfiguration.tfResourceType = "aws_apprunner_observability_configuration";
