// @cdktf/provider-aws/lib/organizations-account/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationsAccount = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/organizations_account aws_organizations_account}
*/
class OrganizationsAccount extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a OrganizationsAccount resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the OrganizationsAccount to import
    * @param importFromId The id of the existing OrganizationsAccount that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/organizations_account#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the OrganizationsAccount to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_organizations_account", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/organizations_account aws_organizations_account} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OrganizationsAccountConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_account',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._closeOnDeletion = config.closeOnDeletion;
        this._createGovcloud = config.createGovcloud;
        this._email = config.email;
        this._iamUserAccessToBilling = config.iamUserAccessToBilling;
        this._id = config.id;
        this._name = config.name;
        this._parentId = config.parentId;
        this._roleName = config.roleName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get closeOnDeletion() {
        return this.getBooleanAttribute('close_on_deletion');
    }
    set closeOnDeletion(value) {
        this._closeOnDeletion = value;
    }
    resetCloseOnDeletion() {
        this._closeOnDeletion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get closeOnDeletionInput() {
        return this._closeOnDeletion;
    }
    get createGovcloud() {
        return this.getBooleanAttribute('create_govcloud');
    }
    set createGovcloud(value) {
        this._createGovcloud = value;
    }
    resetCreateGovcloud() {
        this._createGovcloud = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createGovcloudInput() {
        return this._createGovcloud;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    // govcloud_id - computed: true, optional: false, required: false
    get govcloudId() {
        return this.getStringAttribute('govcloud_id');
    }
    get iamUserAccessToBilling() {
        return this.getStringAttribute('iam_user_access_to_billing');
    }
    set iamUserAccessToBilling(value) {
        this._iamUserAccessToBilling = value;
    }
    resetIamUserAccessToBilling() {
        this._iamUserAccessToBilling = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamUserAccessToBillingInput() {
        return this._iamUserAccessToBilling;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // joined_method - computed: true, optional: false, required: false
    get joinedMethod() {
        return this.getStringAttribute('joined_method');
    }
    // joined_timestamp - computed: true, optional: false, required: false
    get joinedTimestamp() {
        return this.getStringAttribute('joined_timestamp');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parentId() {
        return this.getStringAttribute('parent_id');
    }
    set parentId(value) {
        this._parentId = value;
    }
    resetParentId() {
        this._parentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentIdInput() {
        return this._parentId;
    }
    get roleName() {
        return this.getStringAttribute('role_name');
    }
    set roleName(value) {
        this._roleName = value;
    }
    resetRoleName() {
        this._roleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleNameInput() {
        return this._roleName;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            close_on_deletion: cdktf.booleanToTerraform(this._closeOnDeletion),
            create_govcloud: cdktf.booleanToTerraform(this._createGovcloud),
            email: cdktf.stringToTerraform(this._email),
            iam_user_access_to_billing: cdktf.stringToTerraform(this._iamUserAccessToBilling),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parent_id: cdktf.stringToTerraform(this._parentId),
            role_name: cdktf.stringToTerraform(this._roleName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            close_on_deletion: {
                value: cdktf.booleanToHclTerraform(this._closeOnDeletion),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            create_govcloud: {
                value: cdktf.booleanToHclTerraform(this._createGovcloud),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            email: {
                value: cdktf.stringToHclTerraform(this._email),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_user_access_to_billing: {
                value: cdktf.stringToHclTerraform(this._iamUserAccessToBilling),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent_id: {
                value: cdktf.stringToHclTerraform(this._parentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_name: {
                value: cdktf.stringToHclTerraform(this._roleName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.OrganizationsAccount = OrganizationsAccount;
_a = JSII_RTTI_SYMBOL_1;
OrganizationsAccount[_a] = { fqn: "@cdktf/provider-aws.organizationsAccount.OrganizationsAccount", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
OrganizationsAccount.tfResourceType = "aws_organizations_account";
