// @cdktf/provider-aws/lib/detective-invitation-accepter/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DetectiveInvitationAccepter = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/detective_invitation_accepter aws_detective_invitation_accepter}
*/
class DetectiveInvitationAccepter extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DetectiveInvitationAccepter resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DetectiveInvitationAccepter to import
    * @param importFromId The id of the existing DetectiveInvitationAccepter that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/detective_invitation_accepter#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DetectiveInvitationAccepter to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_detective_invitation_accepter", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/detective_invitation_accepter aws_detective_invitation_accepter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DetectiveInvitationAccepterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_detective_invitation_accepter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._graphArn = config.graphArn;
        this._id = config.id;
    }
    get graphArn() {
        return this.getStringAttribute('graph_arn');
    }
    set graphArn(value) {
        this._graphArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get graphArnInput() {
        return this._graphArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            graph_arn: cdktf.stringToTerraform(this._graphArn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            graph_arn: {
                value: cdktf.stringToHclTerraform(this._graphArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DetectiveInvitationAccepter = DetectiveInvitationAccepter;
_a = JSII_RTTI_SYMBOL_1;
DetectiveInvitationAccepter[_a] = { fqn: "@cdktf/provider-aws.detectiveInvitationAccepter.DetectiveInvitationAccepter", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DetectiveInvitationAccepter.tfResourceType = "aws_detective_invitation_accepter";
