// @cdktf/provider-aws/lib/verifiedaccess-trust-provider/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedaccessTrustProvider = exports.VerifiedaccessTrustProviderTimeoutsOutputReference = exports.verifiedaccessTrustProviderTimeoutsToHclTerraform = exports.verifiedaccessTrustProviderTimeoutsToTerraform = exports.VerifiedaccessTrustProviderOidcOptionsOutputReference = exports.verifiedaccessTrustProviderOidcOptionsToHclTerraform = exports.verifiedaccessTrustProviderOidcOptionsToTerraform = exports.VerifiedaccessTrustProviderDeviceOptionsOutputReference = exports.verifiedaccessTrustProviderDeviceOptionsToHclTerraform = exports.verifiedaccessTrustProviderDeviceOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function verifiedaccessTrustProviderDeviceOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        tenant_id: cdktf.stringToTerraform(struct.tenantId),
    };
}
exports.verifiedaccessTrustProviderDeviceOptionsToTerraform = verifiedaccessTrustProviderDeviceOptionsToTerraform;
function verifiedaccessTrustProviderDeviceOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        tenant_id: {
            value: cdktf.stringToHclTerraform(struct.tenantId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessTrustProviderDeviceOptionsToHclTerraform = verifiedaccessTrustProviderDeviceOptionsToHclTerraform;
class VerifiedaccessTrustProviderDeviceOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tenantId !== undefined) {
            hasAnyValues = true;
            internalValueResult.tenantId = this._tenantId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tenantId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tenantId = value.tenantId;
        }
    }
    get tenantId() {
        return this.getStringAttribute('tenant_id');
    }
    set tenantId(value) {
        this._tenantId = value;
    }
    resetTenantId() {
        this._tenantId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tenantIdInput() {
        return this._tenantId;
    }
}
exports.VerifiedaccessTrustProviderDeviceOptionsOutputReference = VerifiedaccessTrustProviderDeviceOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VerifiedaccessTrustProviderDeviceOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.verifiedaccessTrustProvider.VerifiedaccessTrustProviderDeviceOptionsOutputReference", version: "19.15.0" };
function verifiedaccessTrustProviderOidcOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authorization_endpoint: cdktf.stringToTerraform(struct.authorizationEndpoint),
        client_id: cdktf.stringToTerraform(struct.clientId),
        client_secret: cdktf.stringToTerraform(struct.clientSecret),
        issuer: cdktf.stringToTerraform(struct.issuer),
        scope: cdktf.stringToTerraform(struct.scope),
        token_endpoint: cdktf.stringToTerraform(struct.tokenEndpoint),
        user_info_endpoint: cdktf.stringToTerraform(struct.userInfoEndpoint),
    };
}
exports.verifiedaccessTrustProviderOidcOptionsToTerraform = verifiedaccessTrustProviderOidcOptionsToTerraform;
function verifiedaccessTrustProviderOidcOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        authorization_endpoint: {
            value: cdktf.stringToHclTerraform(struct.authorizationEndpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        client_id: {
            value: cdktf.stringToHclTerraform(struct.clientId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        client_secret: {
            value: cdktf.stringToHclTerraform(struct.clientSecret),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        issuer: {
            value: cdktf.stringToHclTerraform(struct.issuer),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        scope: {
            value: cdktf.stringToHclTerraform(struct.scope),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        token_endpoint: {
            value: cdktf.stringToHclTerraform(struct.tokenEndpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_info_endpoint: {
            value: cdktf.stringToHclTerraform(struct.userInfoEndpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessTrustProviderOidcOptionsToHclTerraform = verifiedaccessTrustProviderOidcOptionsToHclTerraform;
class VerifiedaccessTrustProviderOidcOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authorizationEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.authorizationEndpoint = this._authorizationEndpoint;
        }
        if (this._clientId !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientId = this._clientId;
        }
        if (this._clientSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientSecret = this._clientSecret;
        }
        if (this._issuer !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuer = this._issuer;
        }
        if (this._scope !== undefined) {
            hasAnyValues = true;
            internalValueResult.scope = this._scope;
        }
        if (this._tokenEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.tokenEndpoint = this._tokenEndpoint;
        }
        if (this._userInfoEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.userInfoEndpoint = this._userInfoEndpoint;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authorizationEndpoint = undefined;
            this._clientId = undefined;
            this._clientSecret = undefined;
            this._issuer = undefined;
            this._scope = undefined;
            this._tokenEndpoint = undefined;
            this._userInfoEndpoint = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authorizationEndpoint = value.authorizationEndpoint;
            this._clientId = value.clientId;
            this._clientSecret = value.clientSecret;
            this._issuer = value.issuer;
            this._scope = value.scope;
            this._tokenEndpoint = value.tokenEndpoint;
            this._userInfoEndpoint = value.userInfoEndpoint;
        }
    }
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    set authorizationEndpoint(value) {
        this._authorizationEndpoint = value;
    }
    resetAuthorizationEndpoint() {
        this._authorizationEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationEndpointInput() {
        return this._authorizationEndpoint;
    }
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    set clientId(value) {
        this._clientId = value;
    }
    resetClientId() {
        this._clientId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientIdInput() {
        return this._clientId;
    }
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    set clientSecret(value) {
        this._clientSecret = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientSecretInput() {
        return this._clientSecret;
    }
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    set issuer(value) {
        this._issuer = value;
    }
    resetIssuer() {
        this._issuer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get issuerInput() {
        return this._issuer;
    }
    get scope() {
        return this.getStringAttribute('scope');
    }
    set scope(value) {
        this._scope = value;
    }
    resetScope() {
        this._scope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope;
    }
    get tokenEndpoint() {
        return this.getStringAttribute('token_endpoint');
    }
    set tokenEndpoint(value) {
        this._tokenEndpoint = value;
    }
    resetTokenEndpoint() {
        this._tokenEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tokenEndpointInput() {
        return this._tokenEndpoint;
    }
    get userInfoEndpoint() {
        return this.getStringAttribute('user_info_endpoint');
    }
    set userInfoEndpoint(value) {
        this._userInfoEndpoint = value;
    }
    resetUserInfoEndpoint() {
        this._userInfoEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userInfoEndpointInput() {
        return this._userInfoEndpoint;
    }
}
exports.VerifiedaccessTrustProviderOidcOptionsOutputReference = VerifiedaccessTrustProviderOidcOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VerifiedaccessTrustProviderOidcOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.verifiedaccessTrustProvider.VerifiedaccessTrustProviderOidcOptionsOutputReference", version: "19.15.0" };
function verifiedaccessTrustProviderTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.verifiedaccessTrustProviderTimeoutsToTerraform = verifiedaccessTrustProviderTimeoutsToTerraform;
function verifiedaccessTrustProviderTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessTrustProviderTimeoutsToHclTerraform = verifiedaccessTrustProviderTimeoutsToHclTerraform;
class VerifiedaccessTrustProviderTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VerifiedaccessTrustProviderTimeoutsOutputReference = VerifiedaccessTrustProviderTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VerifiedaccessTrustProviderTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.verifiedaccessTrustProvider.VerifiedaccessTrustProviderTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_trust_provider aws_verifiedaccess_trust_provider}
*/
class VerifiedaccessTrustProvider extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedaccessTrustProvider resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedaccessTrustProvider to import
    * @param importFromId The id of the existing VerifiedaccessTrustProvider that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_trust_provider#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedaccessTrustProvider to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedaccess_trust_provider", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_trust_provider aws_verifiedaccess_trust_provider} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedaccessTrustProviderConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedaccess_trust_provider',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // device_options - computed: false, optional: true, required: false
        this._deviceOptions = new VerifiedaccessTrustProviderDeviceOptionsOutputReference(this, "device_options");
        // oidc_options - computed: false, optional: true, required: false
        this._oidcOptions = new VerifiedaccessTrustProviderOidcOptionsOutputReference(this, "oidc_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VerifiedaccessTrustProviderTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._deviceTrustProviderType = config.deviceTrustProviderType;
        this._id = config.id;
        this._policyReferenceName = config.policyReferenceName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._trustProviderType = config.trustProviderType;
        this._userTrustProviderType = config.userTrustProviderType;
        this._deviceOptions.internalValue = config.deviceOptions;
        this._oidcOptions.internalValue = config.oidcOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get deviceTrustProviderType() {
        return this.getStringAttribute('device_trust_provider_type');
    }
    set deviceTrustProviderType(value) {
        this._deviceTrustProviderType = value;
    }
    resetDeviceTrustProviderType() {
        this._deviceTrustProviderType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceTrustProviderTypeInput() {
        return this._deviceTrustProviderType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get policyReferenceName() {
        return this.getStringAttribute('policy_reference_name');
    }
    set policyReferenceName(value) {
        this._policyReferenceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyReferenceNameInput() {
        return this._policyReferenceName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get trustProviderType() {
        return this.getStringAttribute('trust_provider_type');
    }
    set trustProviderType(value) {
        this._trustProviderType = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustProviderTypeInput() {
        return this._trustProviderType;
    }
    get userTrustProviderType() {
        return this.getStringAttribute('user_trust_provider_type');
    }
    set userTrustProviderType(value) {
        this._userTrustProviderType = value;
    }
    resetUserTrustProviderType() {
        this._userTrustProviderType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userTrustProviderTypeInput() {
        return this._userTrustProviderType;
    }
    get deviceOptions() {
        return this._deviceOptions;
    }
    putDeviceOptions(value) {
        this._deviceOptions.internalValue = value;
    }
    resetDeviceOptions() {
        this._deviceOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceOptionsInput() {
        return this._deviceOptions.internalValue;
    }
    get oidcOptions() {
        return this._oidcOptions;
    }
    putOidcOptions(value) {
        this._oidcOptions.internalValue = value;
    }
    resetOidcOptions() {
        this._oidcOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oidcOptionsInput() {
        return this._oidcOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            device_trust_provider_type: cdktf.stringToTerraform(this._deviceTrustProviderType),
            id: cdktf.stringToTerraform(this._id),
            policy_reference_name: cdktf.stringToTerraform(this._policyReferenceName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            trust_provider_type: cdktf.stringToTerraform(this._trustProviderType),
            user_trust_provider_type: cdktf.stringToTerraform(this._userTrustProviderType),
            device_options: verifiedaccessTrustProviderDeviceOptionsToTerraform(this._deviceOptions.internalValue),
            oidc_options: verifiedaccessTrustProviderOidcOptionsToTerraform(this._oidcOptions.internalValue),
            timeouts: verifiedaccessTrustProviderTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            device_trust_provider_type: {
                value: cdktf.stringToHclTerraform(this._deviceTrustProviderType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy_reference_name: {
                value: cdktf.stringToHclTerraform(this._policyReferenceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            trust_provider_type: {
                value: cdktf.stringToHclTerraform(this._trustProviderType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_trust_provider_type: {
                value: cdktf.stringToHclTerraform(this._userTrustProviderType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            device_options: {
                value: verifiedaccessTrustProviderDeviceOptionsToHclTerraform(this._deviceOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessTrustProviderDeviceOptionsList",
            },
            oidc_options: {
                value: verifiedaccessTrustProviderOidcOptionsToHclTerraform(this._oidcOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessTrustProviderOidcOptionsList",
            },
            timeouts: {
                value: verifiedaccessTrustProviderTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VerifiedaccessTrustProviderTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedaccessTrustProvider = VerifiedaccessTrustProvider;
_d = JSII_RTTI_SYMBOL_1;
VerifiedaccessTrustProvider[_d] = { fqn: "@cdktf/provider-aws.verifiedaccessTrustProvider.VerifiedaccessTrustProvider", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedaccessTrustProvider.tfResourceType = "aws_verifiedaccess_trust_provider";
