// @cdktf/provider-aws/lib/auditmanager-framework-share/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditmanagerFrameworkShare = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework_share aws_auditmanager_framework_share}
*/
class AuditmanagerFrameworkShare extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AuditmanagerFrameworkShare resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AuditmanagerFrameworkShare to import
    * @param importFromId The id of the existing AuditmanagerFrameworkShare that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework_share#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AuditmanagerFrameworkShare to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_auditmanager_framework_share", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_framework_share aws_auditmanager_framework_share} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AuditmanagerFrameworkShareConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_auditmanager_framework_share',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._comment = config.comment;
        this._destinationAccount = config.destinationAccount;
        this._destinationRegion = config.destinationRegion;
        this._frameworkId = config.frameworkId;
    }
    get comment() {
        return this.getStringAttribute('comment');
    }
    set comment(value) {
        this._comment = value;
    }
    resetComment() {
        this._comment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commentInput() {
        return this._comment;
    }
    get destinationAccount() {
        return this.getStringAttribute('destination_account');
    }
    set destinationAccount(value) {
        this._destinationAccount = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationAccountInput() {
        return this._destinationAccount;
    }
    get destinationRegion() {
        return this.getStringAttribute('destination_region');
    }
    set destinationRegion(value) {
        this._destinationRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationRegionInput() {
        return this._destinationRegion;
    }
    get frameworkId() {
        return this.getStringAttribute('framework_id');
    }
    set frameworkId(value) {
        this._frameworkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get frameworkIdInput() {
        return this._frameworkId;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            comment: cdktf.stringToTerraform(this._comment),
            destination_account: cdktf.stringToTerraform(this._destinationAccount),
            destination_region: cdktf.stringToTerraform(this._destinationRegion),
            framework_id: cdktf.stringToTerraform(this._frameworkId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            comment: {
                value: cdktf.stringToHclTerraform(this._comment),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destination_account: {
                value: cdktf.stringToHclTerraform(this._destinationAccount),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destination_region: {
                value: cdktf.stringToHclTerraform(this._destinationRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            framework_id: {
                value: cdktf.stringToHclTerraform(this._frameworkId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AuditmanagerFrameworkShare = AuditmanagerFrameworkShare;
_a = JSII_RTTI_SYMBOL_1;
AuditmanagerFrameworkShare[_a] = { fqn: "@cdktf/provider-aws.auditmanagerFrameworkShare.AuditmanagerFrameworkShare", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AuditmanagerFrameworkShare.tfResourceType = "aws_auditmanager_framework_share";
