// @cdktf/provider-aws/lib/networkmanager-connect-peer/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkmanagerConnectPeer = exports.NetworkmanagerConnectPeerTimeoutsOutputReference = exports.networkmanagerConnectPeerTimeoutsToHclTerraform = exports.networkmanagerConnectPeerTimeoutsToTerraform = exports.NetworkmanagerConnectPeerBgpOptionsOutputReference = exports.networkmanagerConnectPeerBgpOptionsToHclTerraform = exports.networkmanagerConnectPeerBgpOptionsToTerraform = exports.NetworkmanagerConnectPeerConfigurationList = exports.NetworkmanagerConnectPeerConfigurationOutputReference = exports.networkmanagerConnectPeerConfigurationToHclTerraform = exports.networkmanagerConnectPeerConfigurationToTerraform = exports.NetworkmanagerConnectPeerConfigurationBgpConfigurationsList = exports.NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference = exports.networkmanagerConnectPeerConfigurationBgpConfigurationsToHclTerraform = exports.networkmanagerConnectPeerConfigurationBgpConfigurationsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkmanagerConnectPeerConfigurationBgpConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkmanagerConnectPeerConfigurationBgpConfigurationsToTerraform = networkmanagerConnectPeerConfigurationBgpConfigurationsToTerraform;
function networkmanagerConnectPeerConfigurationBgpConfigurationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkmanagerConnectPeerConfigurationBgpConfigurationsToHclTerraform = networkmanagerConnectPeerConfigurationBgpConfigurationsToHclTerraform;
class NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // core_network_address - computed: true, optional: false, required: false
    get coreNetworkAddress() {
        return this.getStringAttribute('core_network_address');
    }
    // core_network_asn - computed: true, optional: false, required: false
    get coreNetworkAsn() {
        return this.getNumberAttribute('core_network_asn');
    }
    // peer_address - computed: true, optional: false, required: false
    get peerAddress() {
        return this.getStringAttribute('peer_address');
    }
    // peer_asn - computed: true, optional: false, required: false
    get peerAsn() {
        return this.getNumberAttribute('peer_asn');
    }
}
exports.NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference = NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference", version: "19.15.0" };
class NetworkmanagerConnectPeerConfigurationBgpConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkmanagerConnectPeerConfigurationBgpConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkmanagerConnectPeerConfigurationBgpConfigurationsList = NetworkmanagerConnectPeerConfigurationBgpConfigurationsList;
_b = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerConfigurationBgpConfigurationsList[_b] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerConfigurationBgpConfigurationsList", version: "19.15.0" };
function networkmanagerConnectPeerConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkmanagerConnectPeerConfigurationToTerraform = networkmanagerConnectPeerConfigurationToTerraform;
function networkmanagerConnectPeerConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkmanagerConnectPeerConfigurationToHclTerraform = networkmanagerConnectPeerConfigurationToHclTerraform;
class NetworkmanagerConnectPeerConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // bgp_configurations - computed: true, optional: false, required: false
        this._bgpConfigurations = new NetworkmanagerConnectPeerConfigurationBgpConfigurationsList(this, "bgp_configurations", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get bgpConfigurations() {
        return this._bgpConfigurations;
    }
    // core_network_address - computed: true, optional: false, required: false
    get coreNetworkAddress() {
        return this.getStringAttribute('core_network_address');
    }
    // inside_cidr_blocks - computed: true, optional: false, required: false
    get insideCidrBlocks() {
        return cdktf.Fn.tolist(this.getListAttribute('inside_cidr_blocks'));
    }
    // peer_address - computed: true, optional: false, required: false
    get peerAddress() {
        return this.getStringAttribute('peer_address');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
}
exports.NetworkmanagerConnectPeerConfigurationOutputReference = NetworkmanagerConnectPeerConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerConfigurationOutputReference", version: "19.15.0" };
class NetworkmanagerConnectPeerConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkmanagerConnectPeerConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkmanagerConnectPeerConfigurationList = NetworkmanagerConnectPeerConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerConfigurationList[_d] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerConfigurationList", version: "19.15.0" };
function networkmanagerConnectPeerBgpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        peer_asn: cdktf.numberToTerraform(struct.peerAsn),
    };
}
exports.networkmanagerConnectPeerBgpOptionsToTerraform = networkmanagerConnectPeerBgpOptionsToTerraform;
function networkmanagerConnectPeerBgpOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        peer_asn: {
            value: cdktf.numberToHclTerraform(struct.peerAsn),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerConnectPeerBgpOptionsToHclTerraform = networkmanagerConnectPeerBgpOptionsToHclTerraform;
class NetworkmanagerConnectPeerBgpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._peerAsn !== undefined) {
            hasAnyValues = true;
            internalValueResult.peerAsn = this._peerAsn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._peerAsn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._peerAsn = value.peerAsn;
        }
    }
    get peerAsn() {
        return this.getNumberAttribute('peer_asn');
    }
    set peerAsn(value) {
        this._peerAsn = value;
    }
    resetPeerAsn() {
        this._peerAsn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get peerAsnInput() {
        return this._peerAsn;
    }
}
exports.NetworkmanagerConnectPeerBgpOptionsOutputReference = NetworkmanagerConnectPeerBgpOptionsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerBgpOptionsOutputReference[_e] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerBgpOptionsOutputReference", version: "19.15.0" };
function networkmanagerConnectPeerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.networkmanagerConnectPeerTimeoutsToTerraform = networkmanagerConnectPeerTimeoutsToTerraform;
function networkmanagerConnectPeerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerConnectPeerTimeoutsToHclTerraform = networkmanagerConnectPeerTimeoutsToHclTerraform;
class NetworkmanagerConnectPeerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.NetworkmanagerConnectPeerTimeoutsOutputReference = NetworkmanagerConnectPeerTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeerTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeerTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_peer aws_networkmanager_connect_peer}
*/
class NetworkmanagerConnectPeer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkmanagerConnectPeer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkmanagerConnectPeer to import
    * @param importFromId The id of the existing NetworkmanagerConnectPeer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_peer#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkmanagerConnectPeer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_networkmanager_connect_peer", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_connect_peer aws_networkmanager_connect_peer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkmanagerConnectPeerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkmanager_connect_peer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // configuration - computed: true, optional: false, required: false
        this._configuration = new NetworkmanagerConnectPeerConfigurationList(this, "configuration", false);
        // bgp_options - computed: false, optional: true, required: false
        this._bgpOptions = new NetworkmanagerConnectPeerBgpOptionsOutputReference(this, "bgp_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkmanagerConnectPeerTimeoutsOutputReference(this, "timeouts");
        this._connectAttachmentId = config.connectAttachmentId;
        this._coreNetworkAddress = config.coreNetworkAddress;
        this._id = config.id;
        this._insideCidrBlocks = config.insideCidrBlocks;
        this._peerAddress = config.peerAddress;
        this._subnetArn = config.subnetArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._bgpOptions.internalValue = config.bgpOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get configuration() {
        return this._configuration;
    }
    get connectAttachmentId() {
        return this.getStringAttribute('connect_attachment_id');
    }
    set connectAttachmentId(value) {
        this._connectAttachmentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectAttachmentIdInput() {
        return this._connectAttachmentId;
    }
    // connect_peer_id - computed: true, optional: false, required: false
    get connectPeerId() {
        return this.getStringAttribute('connect_peer_id');
    }
    get coreNetworkAddress() {
        return this.getStringAttribute('core_network_address');
    }
    set coreNetworkAddress(value) {
        this._coreNetworkAddress = value;
    }
    resetCoreNetworkAddress() {
        this._coreNetworkAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreNetworkAddressInput() {
        return this._coreNetworkAddress;
    }
    // core_network_id - computed: true, optional: false, required: false
    get coreNetworkId() {
        return this.getStringAttribute('core_network_id');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // edge_location - computed: true, optional: false, required: false
    get edgeLocation() {
        return this.getStringAttribute('edge_location');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get insideCidrBlocks() {
        return this.getListAttribute('inside_cidr_blocks');
    }
    set insideCidrBlocks(value) {
        this._insideCidrBlocks = value;
    }
    resetInsideCidrBlocks() {
        this._insideCidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get insideCidrBlocksInput() {
        return this._insideCidrBlocks;
    }
    get peerAddress() {
        return this.getStringAttribute('peer_address');
    }
    set peerAddress(value) {
        this._peerAddress = value;
    }
    // Temporarily expose input value. Use with caution.
    get peerAddressInput() {
        return this._peerAddress;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get subnetArn() {
        return this.getStringAttribute('subnet_arn');
    }
    set subnetArn(value) {
        this._subnetArn = value;
    }
    resetSubnetArn() {
        this._subnetArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetArnInput() {
        return this._subnetArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get bgpOptions() {
        return this._bgpOptions;
    }
    putBgpOptions(value) {
        this._bgpOptions.internalValue = value;
    }
    resetBgpOptions() {
        this._bgpOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bgpOptionsInput() {
        return this._bgpOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connect_attachment_id: cdktf.stringToTerraform(this._connectAttachmentId),
            core_network_address: cdktf.stringToTerraform(this._coreNetworkAddress),
            id: cdktf.stringToTerraform(this._id),
            inside_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(this._insideCidrBlocks),
            peer_address: cdktf.stringToTerraform(this._peerAddress),
            subnet_arn: cdktf.stringToTerraform(this._subnetArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            bgp_options: networkmanagerConnectPeerBgpOptionsToTerraform(this._bgpOptions.internalValue),
            timeouts: networkmanagerConnectPeerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            connect_attachment_id: {
                value: cdktf.stringToHclTerraform(this._connectAttachmentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            core_network_address: {
                value: cdktf.stringToHclTerraform(this._coreNetworkAddress),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            inside_cidr_blocks: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._insideCidrBlocks),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            peer_address: {
                value: cdktf.stringToHclTerraform(this._peerAddress),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_arn: {
                value: cdktf.stringToHclTerraform(this._subnetArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            bgp_options: {
                value: networkmanagerConnectPeerBgpOptionsToHclTerraform(this._bgpOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkmanagerConnectPeerBgpOptionsList",
            },
            timeouts: {
                value: networkmanagerConnectPeerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkmanagerConnectPeerTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkmanagerConnectPeer = NetworkmanagerConnectPeer;
_g = JSII_RTTI_SYMBOL_1;
NetworkmanagerConnectPeer[_g] = { fqn: "@cdktf/provider-aws.networkmanagerConnectPeer.NetworkmanagerConnectPeer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
NetworkmanagerConnectPeer.tfResourceType = "aws_networkmanager_connect_peer";
