// @cdktf/provider-aws/lib/data-aws-outposts-assets/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOutpostsAssets = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/outposts_assets aws_outposts_assets}
*/
class DataAwsOutpostsAssets extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsOutpostsAssets resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsOutpostsAssets to import
    * @param importFromId The id of the existing DataAwsOutpostsAssets that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/outposts_assets#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsOutpostsAssets to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_outposts_assets", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/outposts_assets aws_outposts_assets} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsOutpostsAssetsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_outposts_assets',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._arn = config.arn;
        this._hostIdFilter = config.hostIdFilter;
        this._id = config.id;
        this._statusIdFilter = config.statusIdFilter;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // asset_ids - computed: true, optional: false, required: false
    get assetIds() {
        return this.getListAttribute('asset_ids');
    }
    get hostIdFilter() {
        return cdktf.Fn.tolist(this.getListAttribute('host_id_filter'));
    }
    set hostIdFilter(value) {
        this._hostIdFilter = value;
    }
    resetHostIdFilter() {
        this._hostIdFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostIdFilterInput() {
        return this._hostIdFilter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get statusIdFilter() {
        return cdktf.Fn.tolist(this.getListAttribute('status_id_filter'));
    }
    set statusIdFilter(value) {
        this._statusIdFilter = value;
    }
    resetStatusIdFilter() {
        this._statusIdFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusIdFilterInput() {
        return this._statusIdFilter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            host_id_filter: cdktf.listMapper(cdktf.stringToTerraform, false)(this._hostIdFilter),
            id: cdktf.stringToTerraform(this._id),
            status_id_filter: cdktf.listMapper(cdktf.stringToTerraform, false)(this._statusIdFilter),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            host_id_filter: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._hostIdFilter),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status_id_filter: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._statusIdFilter),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsOutpostsAssets = DataAwsOutpostsAssets;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOutpostsAssets[_a] = { fqn: "@cdktf/provider-aws.dataAwsOutpostsAssets.DataAwsOutpostsAssets", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsOutpostsAssets.tfResourceType = "aws_outposts_assets";
