// @cdktf/provider-aws/lib/glue-resource-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueResourcePolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_resource_policy aws_glue_resource_policy}
*/
class GlueResourcePolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GlueResourcePolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GlueResourcePolicy to import
    * @param importFromId The id of the existing GlueResourcePolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_resource_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GlueResourcePolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_glue_resource_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/glue_resource_policy aws_glue_resource_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueResourcePolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_resource_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._enableHybrid = config.enableHybrid;
        this._id = config.id;
        this._policy = config.policy;
    }
    get enableHybrid() {
        return this.getStringAttribute('enable_hybrid');
    }
    set enableHybrid(value) {
        this._enableHybrid = value;
    }
    resetEnableHybrid() {
        this._enableHybrid = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableHybridInput() {
        return this._enableHybrid;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enable_hybrid: cdktf.stringToTerraform(this._enableHybrid),
            id: cdktf.stringToTerraform(this._id),
            policy: cdktf.stringToTerraform(this._policy),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            enable_hybrid: {
                value: cdktf.stringToHclTerraform(this._enableHybrid),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy: {
                value: cdktf.stringToHclTerraform(this._policy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GlueResourcePolicy = GlueResourcePolicy;
_a = JSII_RTTI_SYMBOL_1;
GlueResourcePolicy[_a] = { fqn: "@cdktf/provider-aws.glueResourcePolicy.GlueResourcePolicy", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GlueResourcePolicy.tfResourceType = "aws_glue_resource_policy";
