// @cdktf/provider-aws/lib/macie2-custom-data-identifier/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Macie2CustomDataIdentifier = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/macie2_custom_data_identifier aws_macie2_custom_data_identifier}
*/
class Macie2CustomDataIdentifier extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Macie2CustomDataIdentifier resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Macie2CustomDataIdentifier to import
    * @param importFromId The id of the existing Macie2CustomDataIdentifier that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/macie2_custom_data_identifier#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Macie2CustomDataIdentifier to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_macie2_custom_data_identifier", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/macie2_custom_data_identifier aws_macie2_custom_data_identifier} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Macie2CustomDataIdentifierConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_macie2_custom_data_identifier',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._ignoreWords = config.ignoreWords;
        this._keywords = config.keywords;
        this._maximumMatchDistance = config.maximumMatchDistance;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._regex = config.regex;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoreWords() {
        return cdktf.Fn.tolist(this.getListAttribute('ignore_words'));
    }
    set ignoreWords(value) {
        this._ignoreWords = value;
    }
    resetIgnoreWords() {
        this._ignoreWords = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreWordsInput() {
        return this._ignoreWords;
    }
    get keywords() {
        return cdktf.Fn.tolist(this.getListAttribute('keywords'));
    }
    set keywords(value) {
        this._keywords = value;
    }
    resetKeywords() {
        this._keywords = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keywordsInput() {
        return this._keywords;
    }
    get maximumMatchDistance() {
        return this.getNumberAttribute('maximum_match_distance');
    }
    set maximumMatchDistance(value) {
        this._maximumMatchDistance = value;
    }
    resetMaximumMatchDistance() {
        this._maximumMatchDistance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumMatchDistanceInput() {
        return this._maximumMatchDistance;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get regex() {
        return this.getStringAttribute('regex');
    }
    set regex(value) {
        this._regex = value;
    }
    resetRegex() {
        this._regex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            ignore_words: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ignoreWords),
            keywords: cdktf.listMapper(cdktf.stringToTerraform, false)(this._keywords),
            maximum_match_distance: cdktf.numberToTerraform(this._maximumMatchDistance),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            regex: cdktf.stringToTerraform(this._regex),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ignore_words: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._ignoreWords),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            keywords: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._keywords),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            maximum_match_distance: {
                value: cdktf.numberToHclTerraform(this._maximumMatchDistance),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            regex: {
                value: cdktf.stringToHclTerraform(this._regex),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Macie2CustomDataIdentifier = Macie2CustomDataIdentifier;
_a = JSII_RTTI_SYMBOL_1;
Macie2CustomDataIdentifier[_a] = { fqn: "@cdktf/provider-aws.macie2CustomDataIdentifier.Macie2CustomDataIdentifier", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Macie2CustomDataIdentifier.tfResourceType = "aws_macie2_custom_data_identifier";
