// @winglang/sdk/lib/core/connections.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connections = exports.CONNECTIONS_FILE_PATH = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const CONNECTIONS_SYMBOL = Symbol.for("@winglang/sdk.core.Connections");
exports.CONNECTIONS_FILE_PATH = "connections.json";
/**
 * Connections are a way to track relationships between constructs.
 */
class Connections {
    /**
     * Return the matching Connections of the given construct tree.
     */
    static of(construct) {
        let connections = construct.node.root[CONNECTIONS_SYMBOL];
        if (!connections) {
            connections = new Connections();
            construct.node.root[CONNECTIONS_SYMBOL] = connections;
        }
        return connections;
    }
    constructor() {
        this._connections = [];
    }
    /**
     * Adds a connection between two constructs. A connection is a piece of
     * metadata describing how one construct is related to another construct.
     */
    add(props) {
        const connection = props;
        // avoid duplicate connections
        if (this._connections.some((c) => c.source === connection.source &&
            c.sourceOp === connection.sourceOp &&
            c.target === connection.target &&
            c.targetOp === connection.targetOp &&
            c.name === connection.name)) {
            return;
        }
        this._connections.push(connection);
    }
    /**
     * Synthesize `connections.json` to the given directory.
     */
    synth(outdir) {
        const connections = this._connections.map((c) => ({
            source: c.source.node.path,
            sourceOp: c.sourceOp,
            target: c.target.node.path,
            targetOp: c.targetOp,
            name: c.name,
        }));
        const tree = {
            version: "connections-0.1",
            connections,
        };
        fs.writeFileSync(path.join(outdir, exports.CONNECTIONS_FILE_PATH), JSON.stringify(tree, undefined, 2), { encoding: "utf8" });
    }
}
exports.Connections = Connections;
_a = JSII_RTTI_SYMBOL_1;
Connections[_a] = { fqn: "@winglang/sdk.core.Connections", version: "0.0.0" };
