// @cdktf/provider-aws/lib/codecommit-approval-rule-template-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodecommitApprovalRuleTemplateAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codecommit_approval_rule_template_association aws_codecommit_approval_rule_template_association}
*/
class CodecommitApprovalRuleTemplateAssociation extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CodecommitApprovalRuleTemplateAssociation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CodecommitApprovalRuleTemplateAssociation to import
    * @param importFromId The id of the existing CodecommitApprovalRuleTemplateAssociation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codecommit_approval_rule_template_association#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CodecommitApprovalRuleTemplateAssociation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_codecommit_approval_rule_template_association", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/codecommit_approval_rule_template_association aws_codecommit_approval_rule_template_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodecommitApprovalRuleTemplateAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codecommit_approval_rule_template_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._approvalRuleTemplateName = config.approvalRuleTemplateName;
        this._id = config.id;
        this._repositoryName = config.repositoryName;
    }
    get approvalRuleTemplateName() {
        return this.getStringAttribute('approval_rule_template_name');
    }
    set approvalRuleTemplateName(value) {
        this._approvalRuleTemplateName = value;
    }
    // Temporarily expose input value. Use with caution.
    get approvalRuleTemplateNameInput() {
        return this._approvalRuleTemplateName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get repositoryName() {
        return this.getStringAttribute('repository_name');
    }
    set repositoryName(value) {
        this._repositoryName = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryNameInput() {
        return this._repositoryName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            approval_rule_template_name: cdktf.stringToTerraform(this._approvalRuleTemplateName),
            id: cdktf.stringToTerraform(this._id),
            repository_name: cdktf.stringToTerraform(this._repositoryName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            approval_rule_template_name: {
                value: cdktf.stringToHclTerraform(this._approvalRuleTemplateName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            repository_name: {
                value: cdktf.stringToHclTerraform(this._repositoryName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CodecommitApprovalRuleTemplateAssociation = CodecommitApprovalRuleTemplateAssociation;
_a = JSII_RTTI_SYMBOL_1;
CodecommitApprovalRuleTemplateAssociation[_a] = { fqn: "@cdktf/provider-aws.codecommitApprovalRuleTemplateAssociation.CodecommitApprovalRuleTemplateAssociation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CodecommitApprovalRuleTemplateAssociation.tfResourceType = "aws_codecommit_approval_rule_template_association";
