// @cdktf/provider-aws/lib/sagemaker-endpoint-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerEndpointConfigurationShadowProductionVariantsList = exports.SagemakerEndpointConfigurationShadowProductionVariantsOutputReference = exports.sagemakerEndpointConfigurationShadowProductionVariantsToHclTerraform = exports.sagemakerEndpointConfigurationShadowProductionVariantsToTerraform = exports.SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference = exports.sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToHclTerraform = exports.sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToTerraform = exports.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList = exports.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference = exports.sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToHclTerraform = exports.sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToTerraform = exports.SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference = exports.sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToHclTerraform = exports.sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToTerraform = exports.SagemakerEndpointConfigurationProductionVariantsList = exports.SagemakerEndpointConfigurationProductionVariantsOutputReference = exports.sagemakerEndpointConfigurationProductionVariantsToHclTerraform = exports.sagemakerEndpointConfigurationProductionVariantsToTerraform = exports.SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference = exports.sagemakerEndpointConfigurationProductionVariantsServerlessConfigToHclTerraform = exports.sagemakerEndpointConfigurationProductionVariantsServerlessConfigToTerraform = exports.SagemakerEndpointConfigurationProductionVariantsRoutingConfigList = exports.SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference = exports.sagemakerEndpointConfigurationProductionVariantsRoutingConfigToHclTerraform = exports.sagemakerEndpointConfigurationProductionVariantsRoutingConfigToTerraform = exports.SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference = exports.sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToHclTerraform = exports.sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToTerraform = exports.SagemakerEndpointConfigurationDataCaptureConfigOutputReference = exports.sagemakerEndpointConfigurationDataCaptureConfigToHclTerraform = exports.sagemakerEndpointConfigurationDataCaptureConfigToTerraform = exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList = exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference = exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToHclTerraform = exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToTerraform = exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference = exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToHclTerraform = exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToTerraform = exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference = exports.sagemakerEndpointConfigurationAsyncInferenceConfigToHclTerraform = exports.sagemakerEndpointConfigurationAsyncInferenceConfigToTerraform = exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference = exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToHclTerraform = exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToTerraform = exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference = exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToHclTerraform = exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToTerraform = exports.SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference = exports.sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToHclTerraform = exports.sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToTerraform = void 0;
exports.SagemakerEndpointConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_concurrent_invocations_per_instance: cdktf.numberToTerraform(struct.maxConcurrentInvocationsPerInstance),
    };
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToTerraform;
function sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_concurrent_invocations_per_instance: {
            value: cdktf.numberToHclTerraform(struct.maxConcurrentInvocationsPerInstance),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToHclTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToHclTerraform;
class SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConcurrentInvocationsPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentInvocationsPerInstance = this._maxConcurrentInvocationsPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConcurrentInvocationsPerInstance = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConcurrentInvocationsPerInstance = value.maxConcurrentInvocationsPerInstance;
        }
    }
    get maxConcurrentInvocationsPerInstance() {
        return this.getNumberAttribute('max_concurrent_invocations_per_instance');
    }
    set maxConcurrentInvocationsPerInstance(value) {
        this._maxConcurrentInvocationsPerInstance = value;
    }
    resetMaxConcurrentInvocationsPerInstance() {
        this._maxConcurrentInvocationsPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentInvocationsPerInstanceInput() {
        return this._maxConcurrentInvocationsPerInstance;
    }
}
exports.SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference = SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        error_topic: cdktf.stringToTerraform(struct.errorTopic),
        include_inference_response_in: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeInferenceResponseIn),
        success_topic: cdktf.stringToTerraform(struct.successTopic),
    };
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToTerraform;
function sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        error_topic: {
            value: cdktf.stringToHclTerraform(struct.errorTopic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        include_inference_response_in: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeInferenceResponseIn),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        success_topic: {
            value: cdktf.stringToHclTerraform(struct.successTopic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToHclTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToHclTerraform;
class SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._errorTopic !== undefined) {
            hasAnyValues = true;
            internalValueResult.errorTopic = this._errorTopic;
        }
        if (this._includeInferenceResponseIn !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeInferenceResponseIn = this._includeInferenceResponseIn;
        }
        if (this._successTopic !== undefined) {
            hasAnyValues = true;
            internalValueResult.successTopic = this._successTopic;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._errorTopic = undefined;
            this._includeInferenceResponseIn = undefined;
            this._successTopic = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._errorTopic = value.errorTopic;
            this._includeInferenceResponseIn = value.includeInferenceResponseIn;
            this._successTopic = value.successTopic;
        }
    }
    get errorTopic() {
        return this.getStringAttribute('error_topic');
    }
    set errorTopic(value) {
        this._errorTopic = value;
    }
    resetErrorTopic() {
        this._errorTopic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get errorTopicInput() {
        return this._errorTopic;
    }
    get includeInferenceResponseIn() {
        return cdktf.Fn.tolist(this.getListAttribute('include_inference_response_in'));
    }
    set includeInferenceResponseIn(value) {
        this._includeInferenceResponseIn = value;
    }
    resetIncludeInferenceResponseIn() {
        this._includeInferenceResponseIn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeInferenceResponseInInput() {
        return this._includeInferenceResponseIn;
    }
    get successTopic() {
        return this.getStringAttribute('success_topic');
    }
    set successTopic(value) {
        this._successTopic = value;
    }
    resetSuccessTopic() {
        this._successTopic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get successTopicInput() {
        return this._successTopic;
    }
}
exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference = SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        s3_failure_path: cdktf.stringToTerraform(struct.s3FailurePath),
        s3_output_path: cdktf.stringToTerraform(struct.s3OutputPath),
        notification_config: sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToTerraform(struct.notificationConfig),
    };
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToTerraform;
function sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_failure_path: {
            value: cdktf.stringToHclTerraform(struct.s3FailurePath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_output_path: {
            value: cdktf.stringToHclTerraform(struct.s3OutputPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        notification_config: {
            value: sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigToHclTerraform(struct.notificationConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToHclTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToHclTerraform;
class SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // notification_config - computed: false, optional: true, required: false
        this._notificationConfig = new SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigNotificationConfigOutputReference(this, "notification_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._s3FailurePath !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3FailurePath = this._s3FailurePath;
        }
        if (this._s3OutputPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3OutputPath = this._s3OutputPath;
        }
        if (this._notificationConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.notificationConfig = this._notificationConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._s3FailurePath = undefined;
            this._s3OutputPath = undefined;
            this._notificationConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._s3FailurePath = value.s3FailurePath;
            this._s3OutputPath = value.s3OutputPath;
            this._notificationConfig.internalValue = value.notificationConfig;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get s3FailurePath() {
        return this.getStringAttribute('s3_failure_path');
    }
    set s3FailurePath(value) {
        this._s3FailurePath = value;
    }
    resetS3FailurePath() {
        this._s3FailurePath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3FailurePathInput() {
        return this._s3FailurePath;
    }
    get s3OutputPath() {
        return this.getStringAttribute('s3_output_path');
    }
    set s3OutputPath(value) {
        this._s3OutputPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3OutputPathInput() {
        return this._s3OutputPath;
    }
    get notificationConfig() {
        return this._notificationConfig;
    }
    putNotificationConfig(value) {
        this._notificationConfig.internalValue = value;
    }
    resetNotificationConfig() {
        this._notificationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationConfigInput() {
        return this._notificationConfig.internalValue;
    }
}
exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference = SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationAsyncInferenceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_config: sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToTerraform(struct.clientConfig),
        output_config: sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToTerraform(struct.outputConfig),
    };
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigToTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigToTerraform;
function sagemakerEndpointConfigurationAsyncInferenceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        client_config: {
            value: sagemakerEndpointConfigurationAsyncInferenceConfigClientConfigToHclTerraform(struct.clientConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigList",
        },
        output_config: {
            value: sagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigToHclTerraform(struct.outputConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationAsyncInferenceConfigToHclTerraform = sagemakerEndpointConfigurationAsyncInferenceConfigToHclTerraform;
class SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // client_config - computed: false, optional: true, required: false
        this._clientConfig = new SagemakerEndpointConfigurationAsyncInferenceConfigClientConfigOutputReference(this, "client_config");
        // output_config - computed: false, optional: false, required: true
        this._outputConfig = new SagemakerEndpointConfigurationAsyncInferenceConfigOutputConfigOutputReference(this, "output_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientConfig = this._clientConfig?.internalValue;
        }
        if (this._outputConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputConfig = this._outputConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientConfig.internalValue = undefined;
            this._outputConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientConfig.internalValue = value.clientConfig;
            this._outputConfig.internalValue = value.outputConfig;
        }
    }
    get clientConfig() {
        return this._clientConfig;
    }
    putClientConfig(value) {
        this._clientConfig.internalValue = value;
    }
    resetClientConfig() {
        this._clientConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientConfigInput() {
        return this._clientConfig.internalValue;
    }
    get outputConfig() {
        return this._outputConfig;
    }
    putOutputConfig(value) {
        this._outputConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get outputConfigInput() {
        return this._outputConfig.internalValue;
    }
}
exports.SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference = SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        csv_content_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.csvContentTypes),
        json_content_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.jsonContentTypes),
    };
}
exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToTerraform = sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToTerraform;
function sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        csv_content_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.csvContentTypes),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        json_content_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.jsonContentTypes),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToHclTerraform = sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToHclTerraform;
class SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._csvContentTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.csvContentTypes = this._csvContentTypes;
        }
        if (this._jsonContentTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.jsonContentTypes = this._jsonContentTypes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._csvContentTypes = undefined;
            this._jsonContentTypes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._csvContentTypes = value.csvContentTypes;
            this._jsonContentTypes = value.jsonContentTypes;
        }
    }
    get csvContentTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('csv_content_types'));
    }
    set csvContentTypes(value) {
        this._csvContentTypes = value;
    }
    resetCsvContentTypes() {
        this._csvContentTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get csvContentTypesInput() {
        return this._csvContentTypes;
    }
    get jsonContentTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('json_content_types'));
    }
    set jsonContentTypes(value) {
        this._jsonContentTypes = value;
    }
    resetJsonContentTypes() {
        this._jsonContentTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonContentTypesInput() {
        return this._jsonContentTypes;
    }
}
exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference = SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference[_e] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        capture_mode: cdktf.stringToTerraform(struct.captureMode),
    };
}
exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToTerraform = sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToTerraform;
function sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        capture_mode: {
            value: cdktf.stringToHclTerraform(struct.captureMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToHclTerraform = sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToHclTerraform;
class SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._captureMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.captureMode = this._captureMode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._captureMode = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._captureMode = value.captureMode;
        }
    }
    get captureMode() {
        return this.getStringAttribute('capture_mode');
    }
    set captureMode(value) {
        this._captureMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get captureModeInput() {
        return this._captureMode;
    }
}
exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference = SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference[_f] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference", version: "19.15.0" };
class SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList = SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList;
_g = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList[_g] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList", version: "19.15.0" };
function sagemakerEndpointConfigurationDataCaptureConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_s3_uri: cdktf.stringToTerraform(struct.destinationS3Uri),
        enable_capture: cdktf.booleanToTerraform(struct.enableCapture),
        initial_sampling_percentage: cdktf.numberToTerraform(struct.initialSamplingPercentage),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        capture_content_type_header: sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToTerraform(struct.captureContentTypeHeader),
        capture_options: cdktf.listMapper(sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToTerraform, true)(struct.captureOptions),
    };
}
exports.sagemakerEndpointConfigurationDataCaptureConfigToTerraform = sagemakerEndpointConfigurationDataCaptureConfigToTerraform;
function sagemakerEndpointConfigurationDataCaptureConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.destinationS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enable_capture: {
            value: cdktf.booleanToHclTerraform(struct.enableCapture),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        initial_sampling_percentage: {
            value: cdktf.numberToHclTerraform(struct.initialSamplingPercentage),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        capture_content_type_header: {
            value: sagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderToHclTerraform(struct.captureContentTypeHeader),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderList",
        },
        capture_options: {
            value: cdktf.listMapperHcl(sagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsToHclTerraform, true)(struct.captureOptions),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationDataCaptureConfigToHclTerraform = sagemakerEndpointConfigurationDataCaptureConfigToHclTerraform;
class SagemakerEndpointConfigurationDataCaptureConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // capture_content_type_header - computed: false, optional: true, required: false
        this._captureContentTypeHeader = new SagemakerEndpointConfigurationDataCaptureConfigCaptureContentTypeHeaderOutputReference(this, "capture_content_type_header");
        // capture_options - computed: false, optional: false, required: true
        this._captureOptions = new SagemakerEndpointConfigurationDataCaptureConfigCaptureOptionsList(this, "capture_options", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationS3Uri = this._destinationS3Uri;
        }
        if (this._enableCapture !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableCapture = this._enableCapture;
        }
        if (this._initialSamplingPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialSamplingPercentage = this._initialSamplingPercentage;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._captureContentTypeHeader?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.captureContentTypeHeader = this._captureContentTypeHeader?.internalValue;
        }
        if (this._captureOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.captureOptions = this._captureOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationS3Uri = undefined;
            this._enableCapture = undefined;
            this._initialSamplingPercentage = undefined;
            this._kmsKeyId = undefined;
            this._captureContentTypeHeader.internalValue = undefined;
            this._captureOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationS3Uri = value.destinationS3Uri;
            this._enableCapture = value.enableCapture;
            this._initialSamplingPercentage = value.initialSamplingPercentage;
            this._kmsKeyId = value.kmsKeyId;
            this._captureContentTypeHeader.internalValue = value.captureContentTypeHeader;
            this._captureOptions.internalValue = value.captureOptions;
        }
    }
    get destinationS3Uri() {
        return this.getStringAttribute('destination_s3_uri');
    }
    set destinationS3Uri(value) {
        this._destinationS3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationS3UriInput() {
        return this._destinationS3Uri;
    }
    get enableCapture() {
        return this.getBooleanAttribute('enable_capture');
    }
    set enableCapture(value) {
        this._enableCapture = value;
    }
    resetEnableCapture() {
        this._enableCapture = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableCaptureInput() {
        return this._enableCapture;
    }
    get initialSamplingPercentage() {
        return this.getNumberAttribute('initial_sampling_percentage');
    }
    set initialSamplingPercentage(value) {
        this._initialSamplingPercentage = value;
    }
    // Temporarily expose input value. Use with caution.
    get initialSamplingPercentageInput() {
        return this._initialSamplingPercentage;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get captureContentTypeHeader() {
        return this._captureContentTypeHeader;
    }
    putCaptureContentTypeHeader(value) {
        this._captureContentTypeHeader.internalValue = value;
    }
    resetCaptureContentTypeHeader() {
        this._captureContentTypeHeader.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get captureContentTypeHeaderInput() {
        return this._captureContentTypeHeader.internalValue;
    }
    get captureOptions() {
        return this._captureOptions;
    }
    putCaptureOptions(value) {
        this._captureOptions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get captureOptionsInput() {
        return this._captureOptions.internalValue;
    }
}
exports.SagemakerEndpointConfigurationDataCaptureConfigOutputReference = SagemakerEndpointConfigurationDataCaptureConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationDataCaptureConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationDataCaptureConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_s3_uri: cdktf.stringToTerraform(struct.destinationS3Uri),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToTerraform = sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToTerraform;
function sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.destinationS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToHclTerraform = sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToHclTerraform;
class SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationS3Uri = this._destinationS3Uri;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationS3Uri = undefined;
            this._kmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationS3Uri = value.destinationS3Uri;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get destinationS3Uri() {
        return this.getStringAttribute('destination_s3_uri');
    }
    set destinationS3Uri(value) {
        this._destinationS3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationS3UriInput() {
        return this._destinationS3Uri;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference = SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationProductionVariantsRoutingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        routing_strategy: cdktf.stringToTerraform(struct.routingStrategy),
    };
}
exports.sagemakerEndpointConfigurationProductionVariantsRoutingConfigToTerraform = sagemakerEndpointConfigurationProductionVariantsRoutingConfigToTerraform;
function sagemakerEndpointConfigurationProductionVariantsRoutingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        routing_strategy: {
            value: cdktf.stringToHclTerraform(struct.routingStrategy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationProductionVariantsRoutingConfigToHclTerraform = sagemakerEndpointConfigurationProductionVariantsRoutingConfigToHclTerraform;
class SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._routingStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.routingStrategy = this._routingStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._routingStrategy = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._routingStrategy = value.routingStrategy;
        }
    }
    get routingStrategy() {
        return this.getStringAttribute('routing_strategy');
    }
    set routingStrategy(value) {
        this._routingStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get routingStrategyInput() {
        return this._routingStrategy;
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference = SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference", version: "19.15.0" };
class SagemakerEndpointConfigurationProductionVariantsRoutingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerEndpointConfigurationProductionVariantsRoutingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsRoutingConfigList = SagemakerEndpointConfigurationProductionVariantsRoutingConfigList;
_l = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsRoutingConfigList[_l] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsRoutingConfigList", version: "19.15.0" };
function sagemakerEndpointConfigurationProductionVariantsServerlessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_concurrency: cdktf.numberToTerraform(struct.maxConcurrency),
        memory_size_in_mb: cdktf.numberToTerraform(struct.memorySizeInMb),
        provisioned_concurrency: cdktf.numberToTerraform(struct.provisionedConcurrency),
    };
}
exports.sagemakerEndpointConfigurationProductionVariantsServerlessConfigToTerraform = sagemakerEndpointConfigurationProductionVariantsServerlessConfigToTerraform;
function sagemakerEndpointConfigurationProductionVariantsServerlessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_concurrency: {
            value: cdktf.numberToHclTerraform(struct.maxConcurrency),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        memory_size_in_mb: {
            value: cdktf.numberToHclTerraform(struct.memorySizeInMb),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        provisioned_concurrency: {
            value: cdktf.numberToHclTerraform(struct.provisionedConcurrency),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationProductionVariantsServerlessConfigToHclTerraform = sagemakerEndpointConfigurationProductionVariantsServerlessConfigToHclTerraform;
class SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrency = this._maxConcurrency;
        }
        if (this._memorySizeInMb !== undefined) {
            hasAnyValues = true;
            internalValueResult.memorySizeInMb = this._memorySizeInMb;
        }
        if (this._provisionedConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.provisionedConcurrency = this._provisionedConcurrency;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConcurrency = undefined;
            this._memorySizeInMb = undefined;
            this._provisionedConcurrency = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConcurrency = value.maxConcurrency;
            this._memorySizeInMb = value.memorySizeInMb;
            this._provisionedConcurrency = value.provisionedConcurrency;
        }
    }
    get maxConcurrency() {
        return this.getNumberAttribute('max_concurrency');
    }
    set maxConcurrency(value) {
        this._maxConcurrency = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrencyInput() {
        return this._maxConcurrency;
    }
    get memorySizeInMb() {
        return this.getNumberAttribute('memory_size_in_mb');
    }
    set memorySizeInMb(value) {
        this._memorySizeInMb = value;
    }
    // Temporarily expose input value. Use with caution.
    get memorySizeInMbInput() {
        return this._memorySizeInMb;
    }
    get provisionedConcurrency() {
        return this.getNumberAttribute('provisioned_concurrency');
    }
    set provisionedConcurrency(value) {
        this._provisionedConcurrency = value;
    }
    resetProvisionedConcurrency() {
        this._provisionedConcurrency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get provisionedConcurrencyInput() {
        return this._provisionedConcurrency;
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference = SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference[_m] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationProductionVariantsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accelerator_type: cdktf.stringToTerraform(struct.acceleratorType),
        container_startup_health_check_timeout_in_seconds: cdktf.numberToTerraform(struct.containerStartupHealthCheckTimeoutInSeconds),
        enable_ssm_access: cdktf.booleanToTerraform(struct.enableSsmAccess),
        initial_instance_count: cdktf.numberToTerraform(struct.initialInstanceCount),
        initial_variant_weight: cdktf.numberToTerraform(struct.initialVariantWeight),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        model_data_download_timeout_in_seconds: cdktf.numberToTerraform(struct.modelDataDownloadTimeoutInSeconds),
        model_name: cdktf.stringToTerraform(struct.modelName),
        variant_name: cdktf.stringToTerraform(struct.variantName),
        volume_size_in_gb: cdktf.numberToTerraform(struct.volumeSizeInGb),
        core_dump_config: sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToTerraform(struct.coreDumpConfig),
        routing_config: cdktf.listMapper(sagemakerEndpointConfigurationProductionVariantsRoutingConfigToTerraform, true)(struct.routingConfig),
        serverless_config: sagemakerEndpointConfigurationProductionVariantsServerlessConfigToTerraform(struct.serverlessConfig),
    };
}
exports.sagemakerEndpointConfigurationProductionVariantsToTerraform = sagemakerEndpointConfigurationProductionVariantsToTerraform;
function sagemakerEndpointConfigurationProductionVariantsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        accelerator_type: {
            value: cdktf.stringToHclTerraform(struct.acceleratorType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_startup_health_check_timeout_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.containerStartupHealthCheckTimeoutInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        enable_ssm_access: {
            value: cdktf.booleanToHclTerraform(struct.enableSsmAccess),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        initial_instance_count: {
            value: cdktf.numberToHclTerraform(struct.initialInstanceCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        initial_variant_weight: {
            value: cdktf.numberToHclTerraform(struct.initialVariantWeight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_data_download_timeout_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.modelDataDownloadTimeoutInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        model_name: {
            value: cdktf.stringToHclTerraform(struct.modelName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        variant_name: {
            value: cdktf.stringToHclTerraform(struct.variantName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        volume_size_in_gb: {
            value: cdktf.numberToHclTerraform(struct.volumeSizeInGb),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        core_dump_config: {
            value: sagemakerEndpointConfigurationProductionVariantsCoreDumpConfigToHclTerraform(struct.coreDumpConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigList",
        },
        routing_config: {
            value: cdktf.listMapperHcl(sagemakerEndpointConfigurationProductionVariantsRoutingConfigToHclTerraform, true)(struct.routingConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationProductionVariantsRoutingConfigList",
        },
        serverless_config: {
            value: sagemakerEndpointConfigurationProductionVariantsServerlessConfigToHclTerraform(struct.serverlessConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationProductionVariantsServerlessConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationProductionVariantsToHclTerraform = sagemakerEndpointConfigurationProductionVariantsToHclTerraform;
class SagemakerEndpointConfigurationProductionVariantsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // core_dump_config - computed: false, optional: true, required: false
        this._coreDumpConfig = new SagemakerEndpointConfigurationProductionVariantsCoreDumpConfigOutputReference(this, "core_dump_config");
        // routing_config - computed: false, optional: true, required: false
        this._routingConfig = new SagemakerEndpointConfigurationProductionVariantsRoutingConfigList(this, "routing_config", false);
        // serverless_config - computed: false, optional: true, required: false
        this._serverlessConfig = new SagemakerEndpointConfigurationProductionVariantsServerlessConfigOutputReference(this, "serverless_config");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._acceleratorType !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorType = this._acceleratorType;
        }
        if (this._containerStartupHealthCheckTimeoutInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerStartupHealthCheckTimeoutInSeconds = this._containerStartupHealthCheckTimeoutInSeconds;
        }
        if (this._enableSsmAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSsmAccess = this._enableSsmAccess;
        }
        if (this._initialInstanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialInstanceCount = this._initialInstanceCount;
        }
        if (this._initialVariantWeight !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialVariantWeight = this._initialVariantWeight;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._modelDataDownloadTimeoutInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataDownloadTimeoutInSeconds = this._modelDataDownloadTimeoutInSeconds;
        }
        if (this._modelName !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelName = this._modelName;
        }
        if (this._variantName !== undefined) {
            hasAnyValues = true;
            internalValueResult.variantName = this._variantName;
        }
        if (this._volumeSizeInGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeSizeInGb = this._volumeSizeInGb;
        }
        if (this._coreDumpConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.coreDumpConfig = this._coreDumpConfig?.internalValue;
        }
        if (this._routingConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.routingConfig = this._routingConfig?.internalValue;
        }
        if (this._serverlessConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverlessConfig = this._serverlessConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._acceleratorType = undefined;
            this._containerStartupHealthCheckTimeoutInSeconds = undefined;
            this._enableSsmAccess = undefined;
            this._initialInstanceCount = undefined;
            this._initialVariantWeight = undefined;
            this._instanceType = undefined;
            this._modelDataDownloadTimeoutInSeconds = undefined;
            this._modelName = undefined;
            this._variantName = undefined;
            this._volumeSizeInGb = undefined;
            this._coreDumpConfig.internalValue = undefined;
            this._routingConfig.internalValue = undefined;
            this._serverlessConfig.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._acceleratorType = value.acceleratorType;
            this._containerStartupHealthCheckTimeoutInSeconds = value.containerStartupHealthCheckTimeoutInSeconds;
            this._enableSsmAccess = value.enableSsmAccess;
            this._initialInstanceCount = value.initialInstanceCount;
            this._initialVariantWeight = value.initialVariantWeight;
            this._instanceType = value.instanceType;
            this._modelDataDownloadTimeoutInSeconds = value.modelDataDownloadTimeoutInSeconds;
            this._modelName = value.modelName;
            this._variantName = value.variantName;
            this._volumeSizeInGb = value.volumeSizeInGb;
            this._coreDumpConfig.internalValue = value.coreDumpConfig;
            this._routingConfig.internalValue = value.routingConfig;
            this._serverlessConfig.internalValue = value.serverlessConfig;
        }
    }
    get acceleratorType() {
        return this.getStringAttribute('accelerator_type');
    }
    set acceleratorType(value) {
        this._acceleratorType = value;
    }
    resetAcceleratorType() {
        this._acceleratorType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorTypeInput() {
        return this._acceleratorType;
    }
    get containerStartupHealthCheckTimeoutInSeconds() {
        return this.getNumberAttribute('container_startup_health_check_timeout_in_seconds');
    }
    set containerStartupHealthCheckTimeoutInSeconds(value) {
        this._containerStartupHealthCheckTimeoutInSeconds = value;
    }
    resetContainerStartupHealthCheckTimeoutInSeconds() {
        this._containerStartupHealthCheckTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerStartupHealthCheckTimeoutInSecondsInput() {
        return this._containerStartupHealthCheckTimeoutInSeconds;
    }
    get enableSsmAccess() {
        return this.getBooleanAttribute('enable_ssm_access');
    }
    set enableSsmAccess(value) {
        this._enableSsmAccess = value;
    }
    resetEnableSsmAccess() {
        this._enableSsmAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSsmAccessInput() {
        return this._enableSsmAccess;
    }
    get initialInstanceCount() {
        return this.getNumberAttribute('initial_instance_count');
    }
    set initialInstanceCount(value) {
        this._initialInstanceCount = value;
    }
    resetInitialInstanceCount() {
        this._initialInstanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialInstanceCountInput() {
        return this._initialInstanceCount;
    }
    get initialVariantWeight() {
        return this.getNumberAttribute('initial_variant_weight');
    }
    set initialVariantWeight(value) {
        this._initialVariantWeight = value;
    }
    resetInitialVariantWeight() {
        this._initialVariantWeight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialVariantWeightInput() {
        return this._initialVariantWeight;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get modelDataDownloadTimeoutInSeconds() {
        return this.getNumberAttribute('model_data_download_timeout_in_seconds');
    }
    set modelDataDownloadTimeoutInSeconds(value) {
        this._modelDataDownloadTimeoutInSeconds = value;
    }
    resetModelDataDownloadTimeoutInSeconds() {
        this._modelDataDownloadTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataDownloadTimeoutInSecondsInput() {
        return this._modelDataDownloadTimeoutInSeconds;
    }
    get modelName() {
        return this.getStringAttribute('model_name');
    }
    set modelName(value) {
        this._modelName = value;
    }
    // Temporarily expose input value. Use with caution.
    get modelNameInput() {
        return this._modelName;
    }
    get variantName() {
        return this.getStringAttribute('variant_name');
    }
    set variantName(value) {
        this._variantName = value;
    }
    resetVariantName() {
        this._variantName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get variantNameInput() {
        return this._variantName;
    }
    get volumeSizeInGb() {
        return this.getNumberAttribute('volume_size_in_gb');
    }
    set volumeSizeInGb(value) {
        this._volumeSizeInGb = value;
    }
    resetVolumeSizeInGb() {
        this._volumeSizeInGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInGbInput() {
        return this._volumeSizeInGb;
    }
    get coreDumpConfig() {
        return this._coreDumpConfig;
    }
    putCoreDumpConfig(value) {
        this._coreDumpConfig.internalValue = value;
    }
    resetCoreDumpConfig() {
        this._coreDumpConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreDumpConfigInput() {
        return this._coreDumpConfig.internalValue;
    }
    get routingConfig() {
        return this._routingConfig;
    }
    putRoutingConfig(value) {
        this._routingConfig.internalValue = value;
    }
    resetRoutingConfig() {
        this._routingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routingConfigInput() {
        return this._routingConfig.internalValue;
    }
    get serverlessConfig() {
        return this._serverlessConfig;
    }
    putServerlessConfig(value) {
        this._serverlessConfig.internalValue = value;
    }
    resetServerlessConfig() {
        this._serverlessConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverlessConfigInput() {
        return this._serverlessConfig.internalValue;
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsOutputReference = SagemakerEndpointConfigurationProductionVariantsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsOutputReference[_o] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsOutputReference", version: "19.15.0" };
class SagemakerEndpointConfigurationProductionVariantsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerEndpointConfigurationProductionVariantsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerEndpointConfigurationProductionVariantsList = SagemakerEndpointConfigurationProductionVariantsList;
_p = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationProductionVariantsList[_p] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationProductionVariantsList", version: "19.15.0" };
function sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_s3_uri: cdktf.stringToTerraform(struct.destinationS3Uri),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToTerraform = sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToTerraform;
function sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination_s3_uri: {
            value: cdktf.stringToHclTerraform(struct.destinationS3Uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToHclTerraform = sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToHclTerraform;
class SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationS3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationS3Uri = this._destinationS3Uri;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationS3Uri = undefined;
            this._kmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationS3Uri = value.destinationS3Uri;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get destinationS3Uri() {
        return this.getStringAttribute('destination_s3_uri');
    }
    set destinationS3Uri(value) {
        this._destinationS3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationS3UriInput() {
        return this._destinationS3Uri;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference = SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference[_q] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        routing_strategy: cdktf.stringToTerraform(struct.routingStrategy),
    };
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToTerraform = sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToTerraform;
function sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        routing_strategy: {
            value: cdktf.stringToHclTerraform(struct.routingStrategy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToHclTerraform = sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToHclTerraform;
class SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._routingStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.routingStrategy = this._routingStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._routingStrategy = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._routingStrategy = value.routingStrategy;
        }
    }
    get routingStrategy() {
        return this.getStringAttribute('routing_strategy');
    }
    set routingStrategy(value) {
        this._routingStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get routingStrategyInput() {
        return this._routingStrategy;
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference = SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference;
_r = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference[_r] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference", version: "19.15.0" };
class SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList = SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList;
_s = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList[_s] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList", version: "19.15.0" };
function sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_concurrency: cdktf.numberToTerraform(struct.maxConcurrency),
        memory_size_in_mb: cdktf.numberToTerraform(struct.memorySizeInMb),
        provisioned_concurrency: cdktf.numberToTerraform(struct.provisionedConcurrency),
    };
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToTerraform = sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToTerraform;
function sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_concurrency: {
            value: cdktf.numberToHclTerraform(struct.maxConcurrency),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        memory_size_in_mb: {
            value: cdktf.numberToHclTerraform(struct.memorySizeInMb),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        provisioned_concurrency: {
            value: cdktf.numberToHclTerraform(struct.provisionedConcurrency),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToHclTerraform = sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToHclTerraform;
class SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrency = this._maxConcurrency;
        }
        if (this._memorySizeInMb !== undefined) {
            hasAnyValues = true;
            internalValueResult.memorySizeInMb = this._memorySizeInMb;
        }
        if (this._provisionedConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.provisionedConcurrency = this._provisionedConcurrency;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConcurrency = undefined;
            this._memorySizeInMb = undefined;
            this._provisionedConcurrency = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConcurrency = value.maxConcurrency;
            this._memorySizeInMb = value.memorySizeInMb;
            this._provisionedConcurrency = value.provisionedConcurrency;
        }
    }
    get maxConcurrency() {
        return this.getNumberAttribute('max_concurrency');
    }
    set maxConcurrency(value) {
        this._maxConcurrency = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrencyInput() {
        return this._maxConcurrency;
    }
    get memorySizeInMb() {
        return this.getNumberAttribute('memory_size_in_mb');
    }
    set memorySizeInMb(value) {
        this._memorySizeInMb = value;
    }
    // Temporarily expose input value. Use with caution.
    get memorySizeInMbInput() {
        return this._memorySizeInMb;
    }
    get provisionedConcurrency() {
        return this.getNumberAttribute('provisioned_concurrency');
    }
    set provisionedConcurrency(value) {
        this._provisionedConcurrency = value;
    }
    resetProvisionedConcurrency() {
        this._provisionedConcurrency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get provisionedConcurrencyInput() {
        return this._provisionedConcurrency;
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference = SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference;
_t = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference[_t] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference", version: "19.15.0" };
function sagemakerEndpointConfigurationShadowProductionVariantsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accelerator_type: cdktf.stringToTerraform(struct.acceleratorType),
        container_startup_health_check_timeout_in_seconds: cdktf.numberToTerraform(struct.containerStartupHealthCheckTimeoutInSeconds),
        enable_ssm_access: cdktf.booleanToTerraform(struct.enableSsmAccess),
        initial_instance_count: cdktf.numberToTerraform(struct.initialInstanceCount),
        initial_variant_weight: cdktf.numberToTerraform(struct.initialVariantWeight),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        model_data_download_timeout_in_seconds: cdktf.numberToTerraform(struct.modelDataDownloadTimeoutInSeconds),
        model_name: cdktf.stringToTerraform(struct.modelName),
        variant_name: cdktf.stringToTerraform(struct.variantName),
        volume_size_in_gb: cdktf.numberToTerraform(struct.volumeSizeInGb),
        core_dump_config: sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToTerraform(struct.coreDumpConfig),
        routing_config: cdktf.listMapper(sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToTerraform, true)(struct.routingConfig),
        serverless_config: sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToTerraform(struct.serverlessConfig),
    };
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsToTerraform = sagemakerEndpointConfigurationShadowProductionVariantsToTerraform;
function sagemakerEndpointConfigurationShadowProductionVariantsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        accelerator_type: {
            value: cdktf.stringToHclTerraform(struct.acceleratorType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_startup_health_check_timeout_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.containerStartupHealthCheckTimeoutInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        enable_ssm_access: {
            value: cdktf.booleanToHclTerraform(struct.enableSsmAccess),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        initial_instance_count: {
            value: cdktf.numberToHclTerraform(struct.initialInstanceCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        initial_variant_weight: {
            value: cdktf.numberToHclTerraform(struct.initialVariantWeight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_data_download_timeout_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.modelDataDownloadTimeoutInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        model_name: {
            value: cdktf.stringToHclTerraform(struct.modelName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        variant_name: {
            value: cdktf.stringToHclTerraform(struct.variantName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        volume_size_in_gb: {
            value: cdktf.numberToHclTerraform(struct.volumeSizeInGb),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        core_dump_config: {
            value: sagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigToHclTerraform(struct.coreDumpConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigList",
        },
        routing_config: {
            value: cdktf.listMapperHcl(sagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigToHclTerraform, true)(struct.routingConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList",
        },
        serverless_config: {
            value: sagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigToHclTerraform(struct.serverlessConfig),
            isBlock: true,
            type: "list",
            storageClassType: "SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sagemakerEndpointConfigurationShadowProductionVariantsToHclTerraform = sagemakerEndpointConfigurationShadowProductionVariantsToHclTerraform;
class SagemakerEndpointConfigurationShadowProductionVariantsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // core_dump_config - computed: false, optional: true, required: false
        this._coreDumpConfig = new SagemakerEndpointConfigurationShadowProductionVariantsCoreDumpConfigOutputReference(this, "core_dump_config");
        // routing_config - computed: false, optional: true, required: false
        this._routingConfig = new SagemakerEndpointConfigurationShadowProductionVariantsRoutingConfigList(this, "routing_config", false);
        // serverless_config - computed: false, optional: true, required: false
        this._serverlessConfig = new SagemakerEndpointConfigurationShadowProductionVariantsServerlessConfigOutputReference(this, "serverless_config");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._acceleratorType !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorType = this._acceleratorType;
        }
        if (this._containerStartupHealthCheckTimeoutInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerStartupHealthCheckTimeoutInSeconds = this._containerStartupHealthCheckTimeoutInSeconds;
        }
        if (this._enableSsmAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSsmAccess = this._enableSsmAccess;
        }
        if (this._initialInstanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialInstanceCount = this._initialInstanceCount;
        }
        if (this._initialVariantWeight !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialVariantWeight = this._initialVariantWeight;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._modelDataDownloadTimeoutInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataDownloadTimeoutInSeconds = this._modelDataDownloadTimeoutInSeconds;
        }
        if (this._modelName !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelName = this._modelName;
        }
        if (this._variantName !== undefined) {
            hasAnyValues = true;
            internalValueResult.variantName = this._variantName;
        }
        if (this._volumeSizeInGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeSizeInGb = this._volumeSizeInGb;
        }
        if (this._coreDumpConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.coreDumpConfig = this._coreDumpConfig?.internalValue;
        }
        if (this._routingConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.routingConfig = this._routingConfig?.internalValue;
        }
        if (this._serverlessConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverlessConfig = this._serverlessConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._acceleratorType = undefined;
            this._containerStartupHealthCheckTimeoutInSeconds = undefined;
            this._enableSsmAccess = undefined;
            this._initialInstanceCount = undefined;
            this._initialVariantWeight = undefined;
            this._instanceType = undefined;
            this._modelDataDownloadTimeoutInSeconds = undefined;
            this._modelName = undefined;
            this._variantName = undefined;
            this._volumeSizeInGb = undefined;
            this._coreDumpConfig.internalValue = undefined;
            this._routingConfig.internalValue = undefined;
            this._serverlessConfig.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._acceleratorType = value.acceleratorType;
            this._containerStartupHealthCheckTimeoutInSeconds = value.containerStartupHealthCheckTimeoutInSeconds;
            this._enableSsmAccess = value.enableSsmAccess;
            this._initialInstanceCount = value.initialInstanceCount;
            this._initialVariantWeight = value.initialVariantWeight;
            this._instanceType = value.instanceType;
            this._modelDataDownloadTimeoutInSeconds = value.modelDataDownloadTimeoutInSeconds;
            this._modelName = value.modelName;
            this._variantName = value.variantName;
            this._volumeSizeInGb = value.volumeSizeInGb;
            this._coreDumpConfig.internalValue = value.coreDumpConfig;
            this._routingConfig.internalValue = value.routingConfig;
            this._serverlessConfig.internalValue = value.serverlessConfig;
        }
    }
    get acceleratorType() {
        return this.getStringAttribute('accelerator_type');
    }
    set acceleratorType(value) {
        this._acceleratorType = value;
    }
    resetAcceleratorType() {
        this._acceleratorType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorTypeInput() {
        return this._acceleratorType;
    }
    get containerStartupHealthCheckTimeoutInSeconds() {
        return this.getNumberAttribute('container_startup_health_check_timeout_in_seconds');
    }
    set containerStartupHealthCheckTimeoutInSeconds(value) {
        this._containerStartupHealthCheckTimeoutInSeconds = value;
    }
    resetContainerStartupHealthCheckTimeoutInSeconds() {
        this._containerStartupHealthCheckTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerStartupHealthCheckTimeoutInSecondsInput() {
        return this._containerStartupHealthCheckTimeoutInSeconds;
    }
    get enableSsmAccess() {
        return this.getBooleanAttribute('enable_ssm_access');
    }
    set enableSsmAccess(value) {
        this._enableSsmAccess = value;
    }
    resetEnableSsmAccess() {
        this._enableSsmAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSsmAccessInput() {
        return this._enableSsmAccess;
    }
    get initialInstanceCount() {
        return this.getNumberAttribute('initial_instance_count');
    }
    set initialInstanceCount(value) {
        this._initialInstanceCount = value;
    }
    resetInitialInstanceCount() {
        this._initialInstanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialInstanceCountInput() {
        return this._initialInstanceCount;
    }
    get initialVariantWeight() {
        return this.getNumberAttribute('initial_variant_weight');
    }
    set initialVariantWeight(value) {
        this._initialVariantWeight = value;
    }
    resetInitialVariantWeight() {
        this._initialVariantWeight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialVariantWeightInput() {
        return this._initialVariantWeight;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get modelDataDownloadTimeoutInSeconds() {
        return this.getNumberAttribute('model_data_download_timeout_in_seconds');
    }
    set modelDataDownloadTimeoutInSeconds(value) {
        this._modelDataDownloadTimeoutInSeconds = value;
    }
    resetModelDataDownloadTimeoutInSeconds() {
        this._modelDataDownloadTimeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataDownloadTimeoutInSecondsInput() {
        return this._modelDataDownloadTimeoutInSeconds;
    }
    get modelName() {
        return this.getStringAttribute('model_name');
    }
    set modelName(value) {
        this._modelName = value;
    }
    // Temporarily expose input value. Use with caution.
    get modelNameInput() {
        return this._modelName;
    }
    get variantName() {
        return this.getStringAttribute('variant_name');
    }
    set variantName(value) {
        this._variantName = value;
    }
    resetVariantName() {
        this._variantName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get variantNameInput() {
        return this._variantName;
    }
    get volumeSizeInGb() {
        return this.getNumberAttribute('volume_size_in_gb');
    }
    set volumeSizeInGb(value) {
        this._volumeSizeInGb = value;
    }
    resetVolumeSizeInGb() {
        this._volumeSizeInGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInGbInput() {
        return this._volumeSizeInGb;
    }
    get coreDumpConfig() {
        return this._coreDumpConfig;
    }
    putCoreDumpConfig(value) {
        this._coreDumpConfig.internalValue = value;
    }
    resetCoreDumpConfig() {
        this._coreDumpConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreDumpConfigInput() {
        return this._coreDumpConfig.internalValue;
    }
    get routingConfig() {
        return this._routingConfig;
    }
    putRoutingConfig(value) {
        this._routingConfig.internalValue = value;
    }
    resetRoutingConfig() {
        this._routingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routingConfigInput() {
        return this._routingConfig.internalValue;
    }
    get serverlessConfig() {
        return this._serverlessConfig;
    }
    putServerlessConfig(value) {
        this._serverlessConfig.internalValue = value;
    }
    resetServerlessConfig() {
        this._serverlessConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverlessConfigInput() {
        return this._serverlessConfig.internalValue;
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsOutputReference = SagemakerEndpointConfigurationShadowProductionVariantsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsOutputReference[_u] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsOutputReference", version: "19.15.0" };
class SagemakerEndpointConfigurationShadowProductionVariantsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerEndpointConfigurationShadowProductionVariantsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerEndpointConfigurationShadowProductionVariantsList = SagemakerEndpointConfigurationShadowProductionVariantsList;
_v = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfigurationShadowProductionVariantsList[_v] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfigurationShadowProductionVariantsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_endpoint_configuration aws_sagemaker_endpoint_configuration}
*/
class SagemakerEndpointConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SagemakerEndpointConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SagemakerEndpointConfiguration to import
    * @param importFromId The id of the existing SagemakerEndpointConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_endpoint_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SagemakerEndpointConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sagemaker_endpoint_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_endpoint_configuration aws_sagemaker_endpoint_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerEndpointConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_endpoint_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // async_inference_config - computed: false, optional: true, required: false
        this._asyncInferenceConfig = new SagemakerEndpointConfigurationAsyncInferenceConfigOutputReference(this, "async_inference_config");
        // data_capture_config - computed: false, optional: true, required: false
        this._dataCaptureConfig = new SagemakerEndpointConfigurationDataCaptureConfigOutputReference(this, "data_capture_config");
        // production_variants - computed: false, optional: false, required: true
        this._productionVariants = new SagemakerEndpointConfigurationProductionVariantsList(this, "production_variants", false);
        // shadow_production_variants - computed: false, optional: true, required: false
        this._shadowProductionVariants = new SagemakerEndpointConfigurationShadowProductionVariantsList(this, "shadow_production_variants", false);
        this._id = config.id;
        this._kmsKeyArn = config.kmsKeyArn;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._asyncInferenceConfig.internalValue = config.asyncInferenceConfig;
        this._dataCaptureConfig.internalValue = config.dataCaptureConfig;
        this._productionVariants.internalValue = config.productionVariants;
        this._shadowProductionVariants.internalValue = config.shadowProductionVariants;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get asyncInferenceConfig() {
        return this._asyncInferenceConfig;
    }
    putAsyncInferenceConfig(value) {
        this._asyncInferenceConfig.internalValue = value;
    }
    resetAsyncInferenceConfig() {
        this._asyncInferenceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get asyncInferenceConfigInput() {
        return this._asyncInferenceConfig.internalValue;
    }
    get dataCaptureConfig() {
        return this._dataCaptureConfig;
    }
    putDataCaptureConfig(value) {
        this._dataCaptureConfig.internalValue = value;
    }
    resetDataCaptureConfig() {
        this._dataCaptureConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataCaptureConfigInput() {
        return this._dataCaptureConfig.internalValue;
    }
    get productionVariants() {
        return this._productionVariants;
    }
    putProductionVariants(value) {
        this._productionVariants.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get productionVariantsInput() {
        return this._productionVariants.internalValue;
    }
    get shadowProductionVariants() {
        return this._shadowProductionVariants;
    }
    putShadowProductionVariants(value) {
        this._shadowProductionVariants.internalValue = value;
    }
    resetShadowProductionVariants() {
        this._shadowProductionVariants.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shadowProductionVariantsInput() {
        return this._shadowProductionVariants.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            kms_key_arn: cdktf.stringToTerraform(this._kmsKeyArn),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            async_inference_config: sagemakerEndpointConfigurationAsyncInferenceConfigToTerraform(this._asyncInferenceConfig.internalValue),
            data_capture_config: sagemakerEndpointConfigurationDataCaptureConfigToTerraform(this._dataCaptureConfig.internalValue),
            production_variants: cdktf.listMapper(sagemakerEndpointConfigurationProductionVariantsToTerraform, true)(this._productionVariants.internalValue),
            shadow_production_variants: cdktf.listMapper(sagemakerEndpointConfigurationShadowProductionVariantsToTerraform, true)(this._shadowProductionVariants.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_arn: {
                value: cdktf.stringToHclTerraform(this._kmsKeyArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            async_inference_config: {
                value: sagemakerEndpointConfigurationAsyncInferenceConfigToHclTerraform(this._asyncInferenceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerEndpointConfigurationAsyncInferenceConfigList",
            },
            data_capture_config: {
                value: sagemakerEndpointConfigurationDataCaptureConfigToHclTerraform(this._dataCaptureConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerEndpointConfigurationDataCaptureConfigList",
            },
            production_variants: {
                value: cdktf.listMapperHcl(sagemakerEndpointConfigurationProductionVariantsToHclTerraform, true)(this._productionVariants.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerEndpointConfigurationProductionVariantsList",
            },
            shadow_production_variants: {
                value: cdktf.listMapperHcl(sagemakerEndpointConfigurationShadowProductionVariantsToHclTerraform, true)(this._shadowProductionVariants.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SagemakerEndpointConfigurationShadowProductionVariantsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SagemakerEndpointConfiguration = SagemakerEndpointConfiguration;
_w = JSII_RTTI_SYMBOL_1;
SagemakerEndpointConfiguration[_w] = { fqn: "@cdktf/provider-aws.sagemakerEndpointConfiguration.SagemakerEndpointConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SagemakerEndpointConfiguration.tfResourceType = "aws_sagemaker_endpoint_configuration";
