// @cdktf/provider-aws/lib/data-aws-licensemanager-received-license/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLicensemanagerReceivedLicense = exports.DataAwsLicensemanagerReceivedLicenseValidityList = exports.DataAwsLicensemanagerReceivedLicenseValidityOutputReference = exports.dataAwsLicensemanagerReceivedLicenseValidityToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseValidityToTerraform = exports.DataAwsLicensemanagerReceivedLicenseReceivedMetadataList = exports.DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference = exports.dataAwsLicensemanagerReceivedLicenseReceivedMetadataToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseReceivedMetadataToTerraform = exports.DataAwsLicensemanagerReceivedLicenseLicenseMetadataList = exports.DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference = exports.dataAwsLicensemanagerReceivedLicenseLicenseMetadataToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseLicenseMetadataToTerraform = exports.DataAwsLicensemanagerReceivedLicenseIssuerList = exports.DataAwsLicensemanagerReceivedLicenseIssuerOutputReference = exports.dataAwsLicensemanagerReceivedLicenseIssuerToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseIssuerToTerraform = exports.DataAwsLicensemanagerReceivedLicenseEntitlementsList = exports.DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference = exports.dataAwsLicensemanagerReceivedLicenseEntitlementsToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseEntitlementsToTerraform = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToTerraform = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToTerraform = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList = exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToHclTerraform = exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToTerraform;
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToHclTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allow_early_check_in - computed: true, optional: false, required: false
    get allowEarlyCheckIn() {
        return this.getBooleanAttribute('allow_early_check_in');
    }
    // max_time_to_live_in_minutes - computed: true, optional: false, required: false
    get maxTimeToLiveInMinutes() {
        return this.getNumberAttribute('max_time_to_live_in_minutes');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToTerraform;
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToHclTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_time_to_live_in_minutes - computed: true, optional: false, required: false
    get maxTimeToLiveInMinutes() {
        return this.getNumberAttribute('max_time_to_live_in_minutes');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList[_d] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToTerraform;
function dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToHclTerraform = dataAwsLicensemanagerReceivedLicenseConsumptionConfigurationToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // borrow_configuration - computed: true, optional: false, required: false
        this._borrowConfiguration = new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationBorrowConfigurationList(this, "borrow_configuration", false);
        // provisional_configuration - computed: true, optional: false, required: false
        this._provisionalConfiguration = new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationProvisionalConfigurationList(this, "provisional_configuration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get borrowConfiguration() {
        return this._borrowConfiguration;
    }
    get provisionalConfiguration() {
        return this._provisionalConfiguration;
    }
    // renew_type - computed: true, optional: false, required: false
    get renewType() {
        return this.getStringAttribute('renew_type');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList = DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList[_f] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseEntitlementsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseEntitlementsToTerraform = dataAwsLicensemanagerReceivedLicenseEntitlementsToTerraform;
function dataAwsLicensemanagerReceivedLicenseEntitlementsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseEntitlementsToHclTerraform = dataAwsLicensemanagerReceivedLicenseEntitlementsToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allow_check_in - computed: true, optional: false, required: false
    get allowCheckIn() {
        return this.getBooleanAttribute('allow_check_in');
    }
    // max_count - computed: true, optional: false, required: false
    get maxCount() {
        return this.getNumberAttribute('max_count');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // unit - computed: true, optional: false, required: false
    get unit() {
        return this.getStringAttribute('unit');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference = DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseEntitlementsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseEntitlementsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseEntitlementsList = DataAwsLicensemanagerReceivedLicenseEntitlementsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseEntitlementsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseEntitlementsList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseIssuerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseIssuerToTerraform = dataAwsLicensemanagerReceivedLicenseIssuerToTerraform;
function dataAwsLicensemanagerReceivedLicenseIssuerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseIssuerToHclTerraform = dataAwsLicensemanagerReceivedLicenseIssuerToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseIssuerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key_fingerprint - computed: true, optional: false, required: false
    get keyFingerprint() {
        return this.getStringAttribute('key_fingerprint');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // sign_key - computed: true, optional: false, required: false
    get signKey() {
        return this.getStringAttribute('sign_key');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseIssuerOutputReference = DataAwsLicensemanagerReceivedLicenseIssuerOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseIssuerOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseIssuerOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseIssuerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseIssuerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseIssuerList = DataAwsLicensemanagerReceivedLicenseIssuerList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseIssuerList[_k] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseIssuerList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseLicenseMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseLicenseMetadataToTerraform = dataAwsLicensemanagerReceivedLicenseLicenseMetadataToTerraform;
function dataAwsLicensemanagerReceivedLicenseLicenseMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseLicenseMetadataToHclTerraform = dataAwsLicensemanagerReceivedLicenseLicenseMetadataToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference = DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseLicenseMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseLicenseMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseLicenseMetadataList = DataAwsLicensemanagerReceivedLicenseLicenseMetadataList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseLicenseMetadataList[_m] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseLicenseMetadataList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseReceivedMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseReceivedMetadataToTerraform = dataAwsLicensemanagerReceivedLicenseReceivedMetadataToTerraform;
function dataAwsLicensemanagerReceivedLicenseReceivedMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseReceivedMetadataToHclTerraform = dataAwsLicensemanagerReceivedLicenseReceivedMetadataToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allowed_operations - computed: true, optional: false, required: false
    get allowedOperations() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_operations'));
    }
    // received_status - computed: true, optional: false, required: false
    get receivedStatus() {
        return this.getStringAttribute('received_status');
    }
    // received_status_reason - computed: true, optional: false, required: false
    get receivedStatusReason() {
        return this.getStringAttribute('received_status_reason');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference = DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseReceivedMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseReceivedMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseReceivedMetadataList = DataAwsLicensemanagerReceivedLicenseReceivedMetadataList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseReceivedMetadataList[_p] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseReceivedMetadataList", version: "19.15.0" };
function dataAwsLicensemanagerReceivedLicenseValidityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLicensemanagerReceivedLicenseValidityToTerraform = dataAwsLicensemanagerReceivedLicenseValidityToTerraform;
function dataAwsLicensemanagerReceivedLicenseValidityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLicensemanagerReceivedLicenseValidityToHclTerraform = dataAwsLicensemanagerReceivedLicenseValidityToHclTerraform;
class DataAwsLicensemanagerReceivedLicenseValidityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // begin - computed: true, optional: false, required: false
    get begin() {
        return this.getStringAttribute('begin');
    }
    // end - computed: true, optional: false, required: false
    get end() {
        return this.getStringAttribute('end');
    }
}
exports.DataAwsLicensemanagerReceivedLicenseValidityOutputReference = DataAwsLicensemanagerReceivedLicenseValidityOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseValidityOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseValidityOutputReference", version: "19.15.0" };
class DataAwsLicensemanagerReceivedLicenseValidityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLicensemanagerReceivedLicenseValidityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLicensemanagerReceivedLicenseValidityList = DataAwsLicensemanagerReceivedLicenseValidityList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicenseValidityList[_r] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicenseValidityList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/licensemanager_received_license aws_licensemanager_received_license}
*/
class DataAwsLicensemanagerReceivedLicense extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsLicensemanagerReceivedLicense resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsLicensemanagerReceivedLicense to import
    * @param importFromId The id of the existing DataAwsLicensemanagerReceivedLicense that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/licensemanager_received_license#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsLicensemanagerReceivedLicense to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_licensemanager_received_license", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/licensemanager_received_license aws_licensemanager_received_license} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLicensemanagerReceivedLicenseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_licensemanager_received_license',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // consumption_configuration - computed: true, optional: false, required: false
        this._consumptionConfiguration = new DataAwsLicensemanagerReceivedLicenseConsumptionConfigurationList(this, "consumption_configuration", false);
        // entitlements - computed: true, optional: false, required: false
        this._entitlements = new DataAwsLicensemanagerReceivedLicenseEntitlementsList(this, "entitlements", true);
        // issuer - computed: true, optional: false, required: false
        this._issuer = new DataAwsLicensemanagerReceivedLicenseIssuerList(this, "issuer", false);
        // license_metadata - computed: true, optional: false, required: false
        this._licenseMetadata = new DataAwsLicensemanagerReceivedLicenseLicenseMetadataList(this, "license_metadata", true);
        // received_metadata - computed: true, optional: false, required: false
        this._receivedMetadata = new DataAwsLicensemanagerReceivedLicenseReceivedMetadataList(this, "received_metadata", false);
        // validity - computed: true, optional: false, required: false
        this._validity = new DataAwsLicensemanagerReceivedLicenseValidityList(this, "validity", false);
        this._id = config.id;
        this._licenseArn = config.licenseArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // beneficiary - computed: true, optional: false, required: false
    get beneficiary() {
        return this.getStringAttribute('beneficiary');
    }
    get consumptionConfiguration() {
        return this._consumptionConfiguration;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get entitlements() {
        return this._entitlements;
    }
    // home_region - computed: true, optional: false, required: false
    get homeRegion() {
        return this.getStringAttribute('home_region');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get issuer() {
        return this._issuer;
    }
    get licenseArn() {
        return this.getStringAttribute('license_arn');
    }
    set licenseArn(value) {
        this._licenseArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get licenseArnInput() {
        return this._licenseArn;
    }
    get licenseMetadata() {
        return this._licenseMetadata;
    }
    // license_name - computed: true, optional: false, required: false
    get licenseName() {
        return this.getStringAttribute('license_name');
    }
    // product_name - computed: true, optional: false, required: false
    get productName() {
        return this.getStringAttribute('product_name');
    }
    // product_sku - computed: true, optional: false, required: false
    get productSku() {
        return this.getStringAttribute('product_sku');
    }
    get receivedMetadata() {
        return this._receivedMetadata;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get validity() {
        return this._validity;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            license_arn: cdktf.stringToTerraform(this._licenseArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            license_arn: {
                value: cdktf.stringToHclTerraform(this._licenseArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsLicensemanagerReceivedLicense = DataAwsLicensemanagerReceivedLicense;
_s = JSII_RTTI_SYMBOL_1;
DataAwsLicensemanagerReceivedLicense[_s] = { fqn: "@cdktf/provider-aws.dataAwsLicensemanagerReceivedLicense.DataAwsLicensemanagerReceivedLicense", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLicensemanagerReceivedLicense.tfResourceType = "aws_licensemanager_received_license";
