// @cdktf/provider-aws/lib/data-aws-ssoadmin-application-assignments/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsoadminApplicationAssignments = exports.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList = exports.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference = exports.dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToHclTerraform = exports.dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToTerraform = dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToTerraform;
function dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToHclTerraform = dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToHclTerraform;
class DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // application_arn - computed: true, optional: false, required: false
    get applicationArn() {
        return this.getStringAttribute('application_arn');
    }
    // principal_id - computed: true, optional: false, required: false
    get principalId() {
        return this.getStringAttribute('principal_id');
    }
    // principal_type - computed: true, optional: false, required: false
    get principalType() {
        return this.getStringAttribute('principal_type');
    }
}
exports.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference = DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationAssignments.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference", version: "19.15.0" };
class DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList = DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationAssignments.DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_assignments aws_ssoadmin_application_assignments}
*/
class DataAwsSsoadminApplicationAssignments extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSsoadminApplicationAssignments resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSsoadminApplicationAssignments to import
    * @param importFromId The id of the existing DataAwsSsoadminApplicationAssignments that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_assignments#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSsoadminApplicationAssignments to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_application_assignments", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_assignments aws_ssoadmin_application_assignments} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSsoadminApplicationAssignmentsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_application_assignments',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // application_assignments - computed: false, optional: true, required: false
        this._applicationAssignments = new DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList(this, "application_assignments", false);
        this._applicationArn = config.applicationArn;
        this._applicationAssignments.internalValue = config.applicationAssignments;
    }
    get applicationArn() {
        return this.getStringAttribute('application_arn');
    }
    set applicationArn(value) {
        this._applicationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationArnInput() {
        return this._applicationArn;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get applicationAssignments() {
        return this._applicationAssignments;
    }
    putApplicationAssignments(value) {
        this._applicationAssignments.internalValue = value;
    }
    resetApplicationAssignments() {
        this._applicationAssignments.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationAssignmentsInput() {
        return this._applicationAssignments.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_arn: cdktf.stringToTerraform(this._applicationArn),
            application_assignments: cdktf.listMapper(dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToTerraform, true)(this._applicationAssignments.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_arn: {
                value: cdktf.stringToHclTerraform(this._applicationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            application_assignments: {
                value: cdktf.listMapperHcl(dataAwsSsoadminApplicationAssignmentsApplicationAssignmentsToHclTerraform, true)(this._applicationAssignments.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsSsoadminApplicationAssignmentsApplicationAssignmentsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSsoadminApplicationAssignments = DataAwsSsoadminApplicationAssignments;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationAssignments[_c] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationAssignments.DataAwsSsoadminApplicationAssignments", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSsoadminApplicationAssignments.tfResourceType = "aws_ssoadmin_application_assignments";
