// @cdktf/provider-aws/lib/quicksight-account-subscription/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuicksightAccountSubscription = exports.QuicksightAccountSubscriptionTimeoutsOutputReference = exports.quicksightAccountSubscriptionTimeoutsToHclTerraform = exports.quicksightAccountSubscriptionTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function quicksightAccountSubscriptionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
    };
}
exports.quicksightAccountSubscriptionTimeoutsToTerraform = quicksightAccountSubscriptionTimeoutsToTerraform;
function quicksightAccountSubscriptionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        read: {
            value: cdktf.stringToHclTerraform(struct.read),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.quicksightAccountSubscriptionTimeoutsToHclTerraform = quicksightAccountSubscriptionTimeoutsToHclTerraform;
class QuicksightAccountSubscriptionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
}
exports.QuicksightAccountSubscriptionTimeoutsOutputReference = QuicksightAccountSubscriptionTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
QuicksightAccountSubscriptionTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.quicksightAccountSubscription.QuicksightAccountSubscriptionTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_account_subscription aws_quicksight_account_subscription}
*/
class QuicksightAccountSubscription extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a QuicksightAccountSubscription resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the QuicksightAccountSubscription to import
    * @param importFromId The id of the existing QuicksightAccountSubscription that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_account_subscription#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the QuicksightAccountSubscription to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_quicksight_account_subscription", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/quicksight_account_subscription aws_quicksight_account_subscription} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options QuicksightAccountSubscriptionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_quicksight_account_subscription',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new QuicksightAccountSubscriptionTimeoutsOutputReference(this, "timeouts");
        this._accountName = config.accountName;
        this._activeDirectoryName = config.activeDirectoryName;
        this._adminGroup = config.adminGroup;
        this._authenticationMethod = config.authenticationMethod;
        this._authorGroup = config.authorGroup;
        this._awsAccountId = config.awsAccountId;
        this._contactNumber = config.contactNumber;
        this._directoryId = config.directoryId;
        this._edition = config.edition;
        this._emailAddress = config.emailAddress;
        this._firstName = config.firstName;
        this._id = config.id;
        this._lastName = config.lastName;
        this._notificationEmail = config.notificationEmail;
        this._readerGroup = config.readerGroup;
        this._realm = config.realm;
        this._timeouts.internalValue = config.timeouts;
    }
    get accountName() {
        return this.getStringAttribute('account_name');
    }
    set accountName(value) {
        this._accountName = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountNameInput() {
        return this._accountName;
    }
    // account_subscription_status - computed: true, optional: false, required: false
    get accountSubscriptionStatus() {
        return this.getStringAttribute('account_subscription_status');
    }
    get activeDirectoryName() {
        return this.getStringAttribute('active_directory_name');
    }
    set activeDirectoryName(value) {
        this._activeDirectoryName = value;
    }
    resetActiveDirectoryName() {
        this._activeDirectoryName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get activeDirectoryNameInput() {
        return this._activeDirectoryName;
    }
    get adminGroup() {
        return this.getListAttribute('admin_group');
    }
    set adminGroup(value) {
        this._adminGroup = value;
    }
    resetAdminGroup() {
        this._adminGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminGroupInput() {
        return this._adminGroup;
    }
    get authenticationMethod() {
        return this.getStringAttribute('authentication_method');
    }
    set authenticationMethod(value) {
        this._authenticationMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationMethodInput() {
        return this._authenticationMethod;
    }
    get authorGroup() {
        return this.getListAttribute('author_group');
    }
    set authorGroup(value) {
        this._authorGroup = value;
    }
    resetAuthorGroup() {
        this._authorGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorGroupInput() {
        return this._authorGroup;
    }
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    set awsAccountId(value) {
        this._awsAccountId = value;
    }
    resetAwsAccountId() {
        this._awsAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccountIdInput() {
        return this._awsAccountId;
    }
    get contactNumber() {
        return this.getStringAttribute('contact_number');
    }
    set contactNumber(value) {
        this._contactNumber = value;
    }
    resetContactNumber() {
        this._contactNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactNumberInput() {
        return this._contactNumber;
    }
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    set directoryId(value) {
        this._directoryId = value;
    }
    resetDirectoryId() {
        this._directoryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directoryIdInput() {
        return this._directoryId;
    }
    get edition() {
        return this.getStringAttribute('edition');
    }
    set edition(value) {
        this._edition = value;
    }
    // Temporarily expose input value. Use with caution.
    get editionInput() {
        return this._edition;
    }
    get emailAddress() {
        return this.getStringAttribute('email_address');
    }
    set emailAddress(value) {
        this._emailAddress = value;
    }
    resetEmailAddress() {
        this._emailAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailAddressInput() {
        return this._emailAddress;
    }
    get firstName() {
        return this.getStringAttribute('first_name');
    }
    set firstName(value) {
        this._firstName = value;
    }
    resetFirstName() {
        this._firstName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firstNameInput() {
        return this._firstName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lastName() {
        return this.getStringAttribute('last_name');
    }
    set lastName(value) {
        this._lastName = value;
    }
    resetLastName() {
        this._lastName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastNameInput() {
        return this._lastName;
    }
    get notificationEmail() {
        return this.getStringAttribute('notification_email');
    }
    set notificationEmail(value) {
        this._notificationEmail = value;
    }
    // Temporarily expose input value. Use with caution.
    get notificationEmailInput() {
        return this._notificationEmail;
    }
    get readerGroup() {
        return this.getListAttribute('reader_group');
    }
    set readerGroup(value) {
        this._readerGroup = value;
    }
    resetReaderGroup() {
        this._readerGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readerGroupInput() {
        return this._readerGroup;
    }
    get realm() {
        return this.getStringAttribute('realm');
    }
    set realm(value) {
        this._realm = value;
    }
    resetRealm() {
        this._realm = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get realmInput() {
        return this._realm;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_name: cdktf.stringToTerraform(this._accountName),
            active_directory_name: cdktf.stringToTerraform(this._activeDirectoryName),
            admin_group: cdktf.listMapper(cdktf.stringToTerraform, false)(this._adminGroup),
            authentication_method: cdktf.stringToTerraform(this._authenticationMethod),
            author_group: cdktf.listMapper(cdktf.stringToTerraform, false)(this._authorGroup),
            aws_account_id: cdktf.stringToTerraform(this._awsAccountId),
            contact_number: cdktf.stringToTerraform(this._contactNumber),
            directory_id: cdktf.stringToTerraform(this._directoryId),
            edition: cdktf.stringToTerraform(this._edition),
            email_address: cdktf.stringToTerraform(this._emailAddress),
            first_name: cdktf.stringToTerraform(this._firstName),
            id: cdktf.stringToTerraform(this._id),
            last_name: cdktf.stringToTerraform(this._lastName),
            notification_email: cdktf.stringToTerraform(this._notificationEmail),
            reader_group: cdktf.listMapper(cdktf.stringToTerraform, false)(this._readerGroup),
            realm: cdktf.stringToTerraform(this._realm),
            timeouts: quicksightAccountSubscriptionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            account_name: {
                value: cdktf.stringToHclTerraform(this._accountName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            active_directory_name: {
                value: cdktf.stringToHclTerraform(this._activeDirectoryName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            admin_group: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._adminGroup),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            authentication_method: {
                value: cdktf.stringToHclTerraform(this._authenticationMethod),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            author_group: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._authorGroup),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            aws_account_id: {
                value: cdktf.stringToHclTerraform(this._awsAccountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            contact_number: {
                value: cdktf.stringToHclTerraform(this._contactNumber),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            directory_id: {
                value: cdktf.stringToHclTerraform(this._directoryId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            edition: {
                value: cdktf.stringToHclTerraform(this._edition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            email_address: {
                value: cdktf.stringToHclTerraform(this._emailAddress),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            first_name: {
                value: cdktf.stringToHclTerraform(this._firstName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            last_name: {
                value: cdktf.stringToHclTerraform(this._lastName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            notification_email: {
                value: cdktf.stringToHclTerraform(this._notificationEmail),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            reader_group: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._readerGroup),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            realm: {
                value: cdktf.stringToHclTerraform(this._realm),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: quicksightAccountSubscriptionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "QuicksightAccountSubscriptionTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.QuicksightAccountSubscription = QuicksightAccountSubscription;
_b = JSII_RTTI_SYMBOL_1;
QuicksightAccountSubscription[_b] = { fqn: "@cdktf/provider-aws.quicksightAccountSubscription.QuicksightAccountSubscription", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
QuicksightAccountSubscription.tfResourceType = "aws_quicksight_account_subscription";
