// @cdktf/provider-aws/lib/data-aws-msk-broker-nodes/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsMskBrokerNodes = exports.DataAwsMskBrokerNodesNodeInfoListStructList = exports.DataAwsMskBrokerNodesNodeInfoListStructOutputReference = exports.dataAwsMskBrokerNodesNodeInfoListStructToHclTerraform = exports.dataAwsMskBrokerNodesNodeInfoListStructToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsMskBrokerNodesNodeInfoListStructToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsMskBrokerNodesNodeInfoListStructToTerraform = dataAwsMskBrokerNodesNodeInfoListStructToTerraform;
function dataAwsMskBrokerNodesNodeInfoListStructToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsMskBrokerNodesNodeInfoListStructToHclTerraform = dataAwsMskBrokerNodesNodeInfoListStructToHclTerraform;
class DataAwsMskBrokerNodesNodeInfoListStructOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // attached_eni_id - computed: true, optional: false, required: false
    get attachedEniId() {
        return this.getStringAttribute('attached_eni_id');
    }
    // broker_id - computed: true, optional: false, required: false
    get brokerId() {
        return this.getNumberAttribute('broker_id');
    }
    // client_subnet - computed: true, optional: false, required: false
    get clientSubnet() {
        return this.getStringAttribute('client_subnet');
    }
    // client_vpc_ip_address - computed: true, optional: false, required: false
    get clientVpcIpAddress() {
        return this.getStringAttribute('client_vpc_ip_address');
    }
    // endpoints - computed: true, optional: false, required: false
    get endpoints() {
        return cdktf.Fn.tolist(this.getListAttribute('endpoints'));
    }
    // node_arn - computed: true, optional: false, required: false
    get nodeArn() {
        return this.getStringAttribute('node_arn');
    }
}
exports.DataAwsMskBrokerNodesNodeInfoListStructOutputReference = DataAwsMskBrokerNodesNodeInfoListStructOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsMskBrokerNodesNodeInfoListStructOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsMskBrokerNodes.DataAwsMskBrokerNodesNodeInfoListStructOutputReference", version: "19.15.0" };
class DataAwsMskBrokerNodesNodeInfoListStructList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsMskBrokerNodesNodeInfoListStructOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsMskBrokerNodesNodeInfoListStructList = DataAwsMskBrokerNodesNodeInfoListStructList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsMskBrokerNodesNodeInfoListStructList[_b] = { fqn: "@cdktf/provider-aws.dataAwsMskBrokerNodes.DataAwsMskBrokerNodesNodeInfoListStructList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/msk_broker_nodes aws_msk_broker_nodes}
*/
class DataAwsMskBrokerNodes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsMskBrokerNodes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsMskBrokerNodes to import
    * @param importFromId The id of the existing DataAwsMskBrokerNodes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/msk_broker_nodes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsMskBrokerNodes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_msk_broker_nodes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/msk_broker_nodes aws_msk_broker_nodes} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsMskBrokerNodesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_broker_nodes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // node_info_list - computed: true, optional: false, required: false
        this._nodeInfoList = new DataAwsMskBrokerNodesNodeInfoListStructList(this, "node_info_list", false);
        this._clusterArn = config.clusterArn;
        this._id = config.id;
    }
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterArnInput() {
        return this._clusterArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get nodeInfoList() {
        return this._nodeInfoList;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_arn: {
                value: cdktf.stringToHclTerraform(this._clusterArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsMskBrokerNodes = DataAwsMskBrokerNodes;
_c = JSII_RTTI_SYMBOL_1;
DataAwsMskBrokerNodes[_c] = { fqn: "@cdktf/provider-aws.dataAwsMskBrokerNodes.DataAwsMskBrokerNodes", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsMskBrokerNodes.tfResourceType = "aws_msk_broker_nodes";
