// @winglang/sdk/lib/ui/http-client.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = exports.API_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const base_1 = require("./base");
const cloud_1 = require("../cloud");
const constants_1 = require("../constants");
const core_1 = require("../core");
/**
 * Global identifier for `HttpClient`.
 */
exports.API_FQN = (0, constants_1.fqnForType)("ui.HttpClient");
/**
 * An HttpClient can be used to make HTTP requests.
 * @noinflight
 */
class HttpClient extends base_1.VisualComponent {
    /**
     * Creates a new ui.HttpClient instance through the app.
     * @internal
     */
    static _newHttpClient(scope, id, label, getUrlHandler, getApiSpecHandler) {
        return core_1.App.of(scope).newAbstract(exports.API_FQN, scope, id, label, getUrlHandler, getApiSpecHandler);
    }
    constructor(scope, id, label, getUrlHandler, getApiSpecHandler) {
        super(scope, id);
        this.label = label;
        this.getUrlfn = new cloud_1.Function(this, "GetUrlHandler", getUrlHandler);
        this.getApiSpecfn = new cloud_1.Function(this, "GetApiSpecHandler", getApiSpecHandler);
    }
    /** @internal */
    _toUIComponent() {
        return {
            kind: "http-client",
            label: this.label,
            getUrlHandler: this.getUrlfn.node.path,
            getApiSpecHandler: this.getApiSpecfn.node.path,
        };
    }
    /** @internal */
    _toInflight() {
        throw new Error("Method not implemented.");
    }
}
exports.HttpClient = HttpClient;
_a = JSII_RTTI_SYMBOL_1;
HttpClient[_a] = { fqn: "@winglang/sdk.ui.HttpClient", version: "0.0.0" };
