// @winglang/sdk/lib/ui/section.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Section = exports.SECTION_FQN = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const base_1 = require("./base");
const button_1 = require("./button");
const field_1 = require("./field");
const constants_1 = require("../constants");
const core_1 = require("../core");
/**
 * Global identifier for `Section`.
 */
exports.SECTION_FQN = (0, constants_1.fqnForType)("ui.Section");
/**
 * A section can be used to group other visual components.
 * @noinflight
 */
class Section extends base_1.VisualComponent {
    /**
     * Creates a new ui.Section instance through the app.
     * @internal
     */
    static _newSection(scope, id, props = {}) {
        return core_1.App.of(scope).newAbstract(exports.SECTION_FQN, scope, id, props);
    }
    constructor(scope, id, props = {}) {
        super(scope, id);
        this.subcomponents = [];
        this.label = props.label;
        this.buttonIdx = 1;
        this.fieldIdx = 1;
    }
    /**
     * Adds a visual component to the section. The components
     * will be rendered in the order they were added.
     */
    add(...components) {
        for (const component of components) {
            if (component._newParent !== undefined) {
                throw new Error(`Cannot add ${component.node.path} to ${this.node.path} because it already belongs to ${component._newParent}.`);
            }
            component._newParent = this.node.path;
            this.subcomponents.push(component);
        }
    }
    /**
     * Adds a button to the section. Shorthand for `add(new ui.Button(...))`.
     */
    addButton(label, handler) {
        const id = `Button-${this.buttonIdx++}`;
        this.add(new button_1.Button(this, id, label, handler));
    }
    /**
     * Adds a field to the section. Shorthand for `add(new ui.Field(...))`.
     */
    addField(label, handler, props = {}) {
        const id = `Field-${this.fieldIdx++}`;
        this.add(new field_1.Field(this, id, label, handler, props));
    }
    /** @internal */
    _toUIComponent() {
        return {
            kind: "section",
            label: this.label,
            children: this.subcomponents.map((c) => c._toUIComponent()),
        };
    }
    /** @internal */
    _toInflight() {
        throw new Error("Method not implemented.");
    }
}
exports.Section = Section;
_a = JSII_RTTI_SYMBOL_1;
Section[_a] = { fqn: "@winglang/sdk.ui.Section", version: "0.0.0" };
