// @cdktf/provider-aws/lib/ce-anomaly-monitor/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CeAnomalyMonitor = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_anomaly_monitor aws_ce_anomaly_monitor}
*/
class CeAnomalyMonitor extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CeAnomalyMonitor resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CeAnomalyMonitor to import
    * @param importFromId The id of the existing CeAnomalyMonitor that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_anomaly_monitor#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CeAnomalyMonitor to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ce_anomaly_monitor", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ce_anomaly_monitor aws_ce_anomaly_monitor} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CeAnomalyMonitorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ce_anomaly_monitor',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._monitorDimension = config.monitorDimension;
        this._monitorSpecification = config.monitorSpecification;
        this._monitorType = config.monitorType;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get monitorDimension() {
        return this.getStringAttribute('monitor_dimension');
    }
    set monitorDimension(value) {
        this._monitorDimension = value;
    }
    resetMonitorDimension() {
        this._monitorDimension = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monitorDimensionInput() {
        return this._monitorDimension;
    }
    get monitorSpecification() {
        return this.getStringAttribute('monitor_specification');
    }
    set monitorSpecification(value) {
        this._monitorSpecification = value;
    }
    resetMonitorSpecification() {
        this._monitorSpecification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monitorSpecificationInput() {
        return this._monitorSpecification;
    }
    get monitorType() {
        return this.getStringAttribute('monitor_type');
    }
    set monitorType(value) {
        this._monitorType = value;
    }
    // Temporarily expose input value. Use with caution.
    get monitorTypeInput() {
        return this._monitorType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            monitor_dimension: cdktf.stringToTerraform(this._monitorDimension),
            monitor_specification: cdktf.stringToTerraform(this._monitorSpecification),
            monitor_type: cdktf.stringToTerraform(this._monitorType),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            monitor_dimension: {
                value: cdktf.stringToHclTerraform(this._monitorDimension),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            monitor_specification: {
                value: cdktf.stringToHclTerraform(this._monitorSpecification),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            monitor_type: {
                value: cdktf.stringToHclTerraform(this._monitorType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CeAnomalyMonitor = CeAnomalyMonitor;
_a = JSII_RTTI_SYMBOL_1;
CeAnomalyMonitor[_a] = { fqn: "@cdktf/provider-aws.ceAnomalyMonitor.CeAnomalyMonitor", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CeAnomalyMonitor.tfResourceType = "aws_ce_anomaly_monitor";
