// @cdktf/provider-aws/lib/data-aws-signer-signing-job/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSignerSigningJob = exports.DataAwsSignerSigningJobSourceList = exports.DataAwsSignerSigningJobSourceOutputReference = exports.dataAwsSignerSigningJobSourceToHclTerraform = exports.dataAwsSignerSigningJobSourceToTerraform = exports.DataAwsSignerSigningJobSourceS3List = exports.DataAwsSignerSigningJobSourceS3OutputReference = exports.dataAwsSignerSigningJobSourceS3ToHclTerraform = exports.dataAwsSignerSigningJobSourceS3ToTerraform = exports.DataAwsSignerSigningJobSignedObjectList = exports.DataAwsSignerSigningJobSignedObjectOutputReference = exports.dataAwsSignerSigningJobSignedObjectToHclTerraform = exports.dataAwsSignerSigningJobSignedObjectToTerraform = exports.DataAwsSignerSigningJobSignedObjectS3List = exports.DataAwsSignerSigningJobSignedObjectS3OutputReference = exports.dataAwsSignerSigningJobSignedObjectS3ToHclTerraform = exports.dataAwsSignerSigningJobSignedObjectS3ToTerraform = exports.DataAwsSignerSigningJobRevocationRecordList = exports.DataAwsSignerSigningJobRevocationRecordOutputReference = exports.dataAwsSignerSigningJobRevocationRecordToHclTerraform = exports.dataAwsSignerSigningJobRevocationRecordToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSignerSigningJobRevocationRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningJobRevocationRecordToTerraform = dataAwsSignerSigningJobRevocationRecordToTerraform;
function dataAwsSignerSigningJobRevocationRecordToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSignerSigningJobRevocationRecordToHclTerraform = dataAwsSignerSigningJobRevocationRecordToHclTerraform;
class DataAwsSignerSigningJobRevocationRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revoked_at - computed: true, optional: false, required: false
    get revokedAt() {
        return this.getStringAttribute('revoked_at');
    }
    // revoked_by - computed: true, optional: false, required: false
    get revokedBy() {
        return this.getStringAttribute('revoked_by');
    }
}
exports.DataAwsSignerSigningJobRevocationRecordOutputReference = DataAwsSignerSigningJobRevocationRecordOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobRevocationRecordOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobRevocationRecordOutputReference", version: "19.15.0" };
class DataAwsSignerSigningJobRevocationRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningJobRevocationRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningJobRevocationRecordList = DataAwsSignerSigningJobRevocationRecordList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobRevocationRecordList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobRevocationRecordList", version: "19.15.0" };
function dataAwsSignerSigningJobSignedObjectS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningJobSignedObjectS3ToTerraform = dataAwsSignerSigningJobSignedObjectS3ToTerraform;
function dataAwsSignerSigningJobSignedObjectS3ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSignerSigningJobSignedObjectS3ToHclTerraform = dataAwsSignerSigningJobSignedObjectS3ToHclTerraform;
class DataAwsSignerSigningJobSignedObjectS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
}
exports.DataAwsSignerSigningJobSignedObjectS3OutputReference = DataAwsSignerSigningJobSignedObjectS3OutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSignedObjectS3OutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSignedObjectS3OutputReference", version: "19.15.0" };
class DataAwsSignerSigningJobSignedObjectS3List extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningJobSignedObjectS3OutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningJobSignedObjectS3List = DataAwsSignerSigningJobSignedObjectS3List;
_d = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSignedObjectS3List[_d] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSignedObjectS3List", version: "19.15.0" };
function dataAwsSignerSigningJobSignedObjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningJobSignedObjectToTerraform = dataAwsSignerSigningJobSignedObjectToTerraform;
function dataAwsSignerSigningJobSignedObjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSignerSigningJobSignedObjectToHclTerraform = dataAwsSignerSigningJobSignedObjectToHclTerraform;
class DataAwsSignerSigningJobSignedObjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // s3 - computed: true, optional: false, required: false
        this._s3 = new DataAwsSignerSigningJobSignedObjectS3List(this, "s3", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get s3() {
        return this._s3;
    }
}
exports.DataAwsSignerSigningJobSignedObjectOutputReference = DataAwsSignerSigningJobSignedObjectOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSignedObjectOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSignedObjectOutputReference", version: "19.15.0" };
class DataAwsSignerSigningJobSignedObjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningJobSignedObjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningJobSignedObjectList = DataAwsSignerSigningJobSignedObjectList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSignedObjectList[_f] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSignedObjectList", version: "19.15.0" };
function dataAwsSignerSigningJobSourceS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningJobSourceS3ToTerraform = dataAwsSignerSigningJobSourceS3ToTerraform;
function dataAwsSignerSigningJobSourceS3ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSignerSigningJobSourceS3ToHclTerraform = dataAwsSignerSigningJobSourceS3ToHclTerraform;
class DataAwsSignerSigningJobSourceS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataAwsSignerSigningJobSourceS3OutputReference = DataAwsSignerSigningJobSourceS3OutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSourceS3OutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSourceS3OutputReference", version: "19.15.0" };
class DataAwsSignerSigningJobSourceS3List extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningJobSourceS3OutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningJobSourceS3List = DataAwsSignerSigningJobSourceS3List;
_h = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSourceS3List[_h] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSourceS3List", version: "19.15.0" };
function dataAwsSignerSigningJobSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningJobSourceToTerraform = dataAwsSignerSigningJobSourceToTerraform;
function dataAwsSignerSigningJobSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSignerSigningJobSourceToHclTerraform = dataAwsSignerSigningJobSourceToHclTerraform;
class DataAwsSignerSigningJobSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // s3 - computed: true, optional: false, required: false
        this._s3 = new DataAwsSignerSigningJobSourceS3List(this, "s3", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get s3() {
        return this._s3;
    }
}
exports.DataAwsSignerSigningJobSourceOutputReference = DataAwsSignerSigningJobSourceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSourceOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSourceOutputReference", version: "19.15.0" };
class DataAwsSignerSigningJobSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningJobSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningJobSourceList = DataAwsSignerSigningJobSourceList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJobSourceList[_k] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJobSourceList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/signer_signing_job aws_signer_signing_job}
*/
class DataAwsSignerSigningJob extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSignerSigningJob resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSignerSigningJob to import
    * @param importFromId The id of the existing DataAwsSignerSigningJob that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/signer_signing_job#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSignerSigningJob to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_signer_signing_job", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/signer_signing_job aws_signer_signing_job} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSignerSigningJobConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_signer_signing_job',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // revocation_record - computed: true, optional: false, required: false
        this._revocationRecord = new DataAwsSignerSigningJobRevocationRecordList(this, "revocation_record", false);
        // signed_object - computed: true, optional: false, required: false
        this._signedObject = new DataAwsSignerSigningJobSignedObjectList(this, "signed_object", false);
        // source - computed: true, optional: false, required: false
        this._source = new DataAwsSignerSigningJobSourceList(this, "source", false);
        this._id = config.id;
        this._jobId = config.jobId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // completed_at - computed: true, optional: false, required: false
    get completedAt() {
        return this.getStringAttribute('completed_at');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get jobId() {
        return this.getStringAttribute('job_id');
    }
    set jobId(value) {
        this._jobId = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobIdInput() {
        return this._jobId;
    }
    // job_invoker - computed: true, optional: false, required: false
    get jobInvoker() {
        return this.getStringAttribute('job_invoker');
    }
    // job_owner - computed: true, optional: false, required: false
    get jobOwner() {
        return this.getStringAttribute('job_owner');
    }
    // platform_display_name - computed: true, optional: false, required: false
    get platformDisplayName() {
        return this.getStringAttribute('platform_display_name');
    }
    // platform_id - computed: true, optional: false, required: false
    get platformId() {
        return this.getStringAttribute('platform_id');
    }
    // profile_name - computed: true, optional: false, required: false
    get profileName() {
        return this.getStringAttribute('profile_name');
    }
    // profile_version - computed: true, optional: false, required: false
    get profileVersion() {
        return this.getStringAttribute('profile_version');
    }
    // requested_by - computed: true, optional: false, required: false
    get requestedBy() {
        return this.getStringAttribute('requested_by');
    }
    get revocationRecord() {
        return this._revocationRecord;
    }
    // signature_expires_at - computed: true, optional: false, required: false
    get signatureExpiresAt() {
        return this.getStringAttribute('signature_expires_at');
    }
    get signedObject() {
        return this._signedObject;
    }
    get source() {
        return this._source;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_reason - computed: true, optional: false, required: false
    get statusReason() {
        return this.getStringAttribute('status_reason');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            job_id: cdktf.stringToTerraform(this._jobId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            job_id: {
                value: cdktf.stringToHclTerraform(this._jobId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSignerSigningJob = DataAwsSignerSigningJob;
_l = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningJob[_l] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningJob.DataAwsSignerSigningJob", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSignerSigningJob.tfResourceType = "aws_signer_signing_job";
