// @cdktf/provider-aws/lib/mq-broker/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MqBroker = exports.MqBrokerUserList = exports.MqBrokerUserOutputReference = exports.mqBrokerUserToHclTerraform = exports.mqBrokerUserToTerraform = exports.MqBrokerTimeoutsOutputReference = exports.mqBrokerTimeoutsToHclTerraform = exports.mqBrokerTimeoutsToTerraform = exports.MqBrokerMaintenanceWindowStartTimeOutputReference = exports.mqBrokerMaintenanceWindowStartTimeToHclTerraform = exports.mqBrokerMaintenanceWindowStartTimeToTerraform = exports.MqBrokerLogsOutputReference = exports.mqBrokerLogsToHclTerraform = exports.mqBrokerLogsToTerraform = exports.MqBrokerLdapServerMetadataOutputReference = exports.mqBrokerLdapServerMetadataToHclTerraform = exports.mqBrokerLdapServerMetadataToTerraform = exports.MqBrokerEncryptionOptionsOutputReference = exports.mqBrokerEncryptionOptionsToHclTerraform = exports.mqBrokerEncryptionOptionsToTerraform = exports.MqBrokerConfigurationOutputReference = exports.mqBrokerConfigurationToHclTerraform = exports.mqBrokerConfigurationToTerraform = exports.MqBrokerInstancesList = exports.MqBrokerInstancesOutputReference = exports.mqBrokerInstancesToHclTerraform = exports.mqBrokerInstancesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mqBrokerInstancesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.mqBrokerInstancesToTerraform = mqBrokerInstancesToTerraform;
function mqBrokerInstancesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.mqBrokerInstancesToHclTerraform = mqBrokerInstancesToHclTerraform;
class MqBrokerInstancesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // console_url - computed: true, optional: false, required: false
    get consoleUrl() {
        return this.getStringAttribute('console_url');
    }
    // endpoints - computed: true, optional: false, required: false
    get endpoints() {
        return this.getListAttribute('endpoints');
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
}
exports.MqBrokerInstancesOutputReference = MqBrokerInstancesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MqBrokerInstancesOutputReference[_a] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerInstancesOutputReference", version: "19.15.0" };
class MqBrokerInstancesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MqBrokerInstancesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MqBrokerInstancesList = MqBrokerInstancesList;
_b = JSII_RTTI_SYMBOL_1;
MqBrokerInstancesList[_b] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerInstancesList", version: "19.15.0" };
function mqBrokerConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mqBrokerConfigurationToTerraform = mqBrokerConfigurationToTerraform;
function mqBrokerConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        revision: {
            value: cdktf.numberToHclTerraform(struct.revision),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerConfigurationToHclTerraform = mqBrokerConfigurationToHclTerraform;
class MqBrokerConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._id = undefined;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._id = value.id;
            this._revision = value.revision;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get revision() {
        return this.getNumberAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    resetRevision() {
        this._revision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
}
exports.MqBrokerConfigurationOutputReference = MqBrokerConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MqBrokerConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerConfigurationOutputReference", version: "19.15.0" };
function mqBrokerEncryptionOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        use_aws_owned_key: cdktf.booleanToTerraform(struct.useAwsOwnedKey),
    };
}
exports.mqBrokerEncryptionOptionsToTerraform = mqBrokerEncryptionOptionsToTerraform;
function mqBrokerEncryptionOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_id: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        use_aws_owned_key: {
            value: cdktf.booleanToHclTerraform(struct.useAwsOwnedKey),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerEncryptionOptionsToHclTerraform = mqBrokerEncryptionOptionsToHclTerraform;
class MqBrokerEncryptionOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._useAwsOwnedKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.useAwsOwnedKey = this._useAwsOwnedKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._useAwsOwnedKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._useAwsOwnedKey = value.useAwsOwnedKey;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get useAwsOwnedKey() {
        return this.getBooleanAttribute('use_aws_owned_key');
    }
    set useAwsOwnedKey(value) {
        this._useAwsOwnedKey = value;
    }
    resetUseAwsOwnedKey() {
        this._useAwsOwnedKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useAwsOwnedKeyInput() {
        return this._useAwsOwnedKey;
    }
}
exports.MqBrokerEncryptionOptionsOutputReference = MqBrokerEncryptionOptionsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MqBrokerEncryptionOptionsOutputReference[_d] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerEncryptionOptionsOutputReference", version: "19.15.0" };
function mqBrokerLdapServerMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hosts: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.hosts),
        role_base: cdktf.stringToTerraform(struct.roleBase),
        role_name: cdktf.stringToTerraform(struct.roleName),
        role_search_matching: cdktf.stringToTerraform(struct.roleSearchMatching),
        role_search_subtree: cdktf.booleanToTerraform(struct.roleSearchSubtree),
        service_account_password: cdktf.stringToTerraform(struct.serviceAccountPassword),
        service_account_username: cdktf.stringToTerraform(struct.serviceAccountUsername),
        user_base: cdktf.stringToTerraform(struct.userBase),
        user_role_name: cdktf.stringToTerraform(struct.userRoleName),
        user_search_matching: cdktf.stringToTerraform(struct.userSearchMatching),
        user_search_subtree: cdktf.booleanToTerraform(struct.userSearchSubtree),
    };
}
exports.mqBrokerLdapServerMetadataToTerraform = mqBrokerLdapServerMetadataToTerraform;
function mqBrokerLdapServerMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hosts: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.hosts),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        role_base: {
            value: cdktf.stringToHclTerraform(struct.roleBase),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_name: {
            value: cdktf.stringToHclTerraform(struct.roleName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_search_matching: {
            value: cdktf.stringToHclTerraform(struct.roleSearchMatching),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_search_subtree: {
            value: cdktf.booleanToHclTerraform(struct.roleSearchSubtree),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        service_account_password: {
            value: cdktf.stringToHclTerraform(struct.serviceAccountPassword),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        service_account_username: {
            value: cdktf.stringToHclTerraform(struct.serviceAccountUsername),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_base: {
            value: cdktf.stringToHclTerraform(struct.userBase),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_role_name: {
            value: cdktf.stringToHclTerraform(struct.userRoleName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_search_matching: {
            value: cdktf.stringToHclTerraform(struct.userSearchMatching),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_search_subtree: {
            value: cdktf.booleanToHclTerraform(struct.userSearchSubtree),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerLdapServerMetadataToHclTerraform = mqBrokerLdapServerMetadataToHclTerraform;
class MqBrokerLdapServerMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hosts !== undefined) {
            hasAnyValues = true;
            internalValueResult.hosts = this._hosts;
        }
        if (this._roleBase !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleBase = this._roleBase;
        }
        if (this._roleName !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleName = this._roleName;
        }
        if (this._roleSearchMatching !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleSearchMatching = this._roleSearchMatching;
        }
        if (this._roleSearchSubtree !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleSearchSubtree = this._roleSearchSubtree;
        }
        if (this._serviceAccountPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccountPassword = this._serviceAccountPassword;
        }
        if (this._serviceAccountUsername !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccountUsername = this._serviceAccountUsername;
        }
        if (this._userBase !== undefined) {
            hasAnyValues = true;
            internalValueResult.userBase = this._userBase;
        }
        if (this._userRoleName !== undefined) {
            hasAnyValues = true;
            internalValueResult.userRoleName = this._userRoleName;
        }
        if (this._userSearchMatching !== undefined) {
            hasAnyValues = true;
            internalValueResult.userSearchMatching = this._userSearchMatching;
        }
        if (this._userSearchSubtree !== undefined) {
            hasAnyValues = true;
            internalValueResult.userSearchSubtree = this._userSearchSubtree;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hosts = undefined;
            this._roleBase = undefined;
            this._roleName = undefined;
            this._roleSearchMatching = undefined;
            this._roleSearchSubtree = undefined;
            this._serviceAccountPassword = undefined;
            this._serviceAccountUsername = undefined;
            this._userBase = undefined;
            this._userRoleName = undefined;
            this._userSearchMatching = undefined;
            this._userSearchSubtree = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hosts = value.hosts;
            this._roleBase = value.roleBase;
            this._roleName = value.roleName;
            this._roleSearchMatching = value.roleSearchMatching;
            this._roleSearchSubtree = value.roleSearchSubtree;
            this._serviceAccountPassword = value.serviceAccountPassword;
            this._serviceAccountUsername = value.serviceAccountUsername;
            this._userBase = value.userBase;
            this._userRoleName = value.userRoleName;
            this._userSearchMatching = value.userSearchMatching;
            this._userSearchSubtree = value.userSearchSubtree;
        }
    }
    get hosts() {
        return this.getListAttribute('hosts');
    }
    set hosts(value) {
        this._hosts = value;
    }
    resetHosts() {
        this._hosts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostsInput() {
        return this._hosts;
    }
    get roleBase() {
        return this.getStringAttribute('role_base');
    }
    set roleBase(value) {
        this._roleBase = value;
    }
    resetRoleBase() {
        this._roleBase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleBaseInput() {
        return this._roleBase;
    }
    get roleName() {
        return this.getStringAttribute('role_name');
    }
    set roleName(value) {
        this._roleName = value;
    }
    resetRoleName() {
        this._roleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleNameInput() {
        return this._roleName;
    }
    get roleSearchMatching() {
        return this.getStringAttribute('role_search_matching');
    }
    set roleSearchMatching(value) {
        this._roleSearchMatching = value;
    }
    resetRoleSearchMatching() {
        this._roleSearchMatching = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleSearchMatchingInput() {
        return this._roleSearchMatching;
    }
    get roleSearchSubtree() {
        return this.getBooleanAttribute('role_search_subtree');
    }
    set roleSearchSubtree(value) {
        this._roleSearchSubtree = value;
    }
    resetRoleSearchSubtree() {
        this._roleSearchSubtree = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleSearchSubtreeInput() {
        return this._roleSearchSubtree;
    }
    get serviceAccountPassword() {
        return this.getStringAttribute('service_account_password');
    }
    set serviceAccountPassword(value) {
        this._serviceAccountPassword = value;
    }
    resetServiceAccountPassword() {
        this._serviceAccountPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountPasswordInput() {
        return this._serviceAccountPassword;
    }
    get serviceAccountUsername() {
        return this.getStringAttribute('service_account_username');
    }
    set serviceAccountUsername(value) {
        this._serviceAccountUsername = value;
    }
    resetServiceAccountUsername() {
        this._serviceAccountUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountUsernameInput() {
        return this._serviceAccountUsername;
    }
    get userBase() {
        return this.getStringAttribute('user_base');
    }
    set userBase(value) {
        this._userBase = value;
    }
    resetUserBase() {
        this._userBase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userBaseInput() {
        return this._userBase;
    }
    get userRoleName() {
        return this.getStringAttribute('user_role_name');
    }
    set userRoleName(value) {
        this._userRoleName = value;
    }
    resetUserRoleName() {
        this._userRoleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userRoleNameInput() {
        return this._userRoleName;
    }
    get userSearchMatching() {
        return this.getStringAttribute('user_search_matching');
    }
    set userSearchMatching(value) {
        this._userSearchMatching = value;
    }
    resetUserSearchMatching() {
        this._userSearchMatching = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userSearchMatchingInput() {
        return this._userSearchMatching;
    }
    get userSearchSubtree() {
        return this.getBooleanAttribute('user_search_subtree');
    }
    set userSearchSubtree(value) {
        this._userSearchSubtree = value;
    }
    resetUserSearchSubtree() {
        this._userSearchSubtree = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userSearchSubtreeInput() {
        return this._userSearchSubtree;
    }
}
exports.MqBrokerLdapServerMetadataOutputReference = MqBrokerLdapServerMetadataOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MqBrokerLdapServerMetadataOutputReference[_e] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerLdapServerMetadataOutputReference", version: "19.15.0" };
function mqBrokerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit: cdktf.stringToTerraform(struct.audit),
        general: cdktf.booleanToTerraform(struct.general),
    };
}
exports.mqBrokerLogsToTerraform = mqBrokerLogsToTerraform;
function mqBrokerLogsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audit: {
            value: cdktf.stringToHclTerraform(struct.audit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        general: {
            value: cdktf.booleanToHclTerraform(struct.general),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerLogsToHclTerraform = mqBrokerLogsToHclTerraform;
class MqBrokerLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audit !== undefined) {
            hasAnyValues = true;
            internalValueResult.audit = this._audit;
        }
        if (this._general !== undefined) {
            hasAnyValues = true;
            internalValueResult.general = this._general;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audit = undefined;
            this._general = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audit = value.audit;
            this._general = value.general;
        }
    }
    get audit() {
        return this.getStringAttribute('audit');
    }
    set audit(value) {
        this._audit = value;
    }
    resetAudit() {
        this._audit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditInput() {
        return this._audit;
    }
    get general() {
        return this.getBooleanAttribute('general');
    }
    set general(value) {
        this._general = value;
    }
    resetGeneral() {
        this._general = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get generalInput() {
        return this._general;
    }
}
exports.MqBrokerLogsOutputReference = MqBrokerLogsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MqBrokerLogsOutputReference[_f] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerLogsOutputReference", version: "19.15.0" };
function mqBrokerMaintenanceWindowStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day_of_week: cdktf.stringToTerraform(struct.dayOfWeek),
        time_of_day: cdktf.stringToTerraform(struct.timeOfDay),
        time_zone: cdktf.stringToTerraform(struct.timeZone),
    };
}
exports.mqBrokerMaintenanceWindowStartTimeToTerraform = mqBrokerMaintenanceWindowStartTimeToTerraform;
function mqBrokerMaintenanceWindowStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day_of_week: {
            value: cdktf.stringToHclTerraform(struct.dayOfWeek),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        time_of_day: {
            value: cdktf.stringToHclTerraform(struct.timeOfDay),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        time_zone: {
            value: cdktf.stringToHclTerraform(struct.timeZone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerMaintenanceWindowStartTimeToHclTerraform = mqBrokerMaintenanceWindowStartTimeToHclTerraform;
class MqBrokerMaintenanceWindowStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dayOfWeek !== undefined) {
            hasAnyValues = true;
            internalValueResult.dayOfWeek = this._dayOfWeek;
        }
        if (this._timeOfDay !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeOfDay = this._timeOfDay;
        }
        if (this._timeZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeZone = this._timeZone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dayOfWeek = undefined;
            this._timeOfDay = undefined;
            this._timeZone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dayOfWeek = value.dayOfWeek;
            this._timeOfDay = value.timeOfDay;
            this._timeZone = value.timeZone;
        }
    }
    get dayOfWeek() {
        return this.getStringAttribute('day_of_week');
    }
    set dayOfWeek(value) {
        this._dayOfWeek = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayOfWeekInput() {
        return this._dayOfWeek;
    }
    get timeOfDay() {
        return this.getStringAttribute('time_of_day');
    }
    set timeOfDay(value) {
        this._timeOfDay = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeOfDayInput() {
        return this._timeOfDay;
    }
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    set timeZone(value) {
        this._timeZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeZoneInput() {
        return this._timeZone;
    }
}
exports.MqBrokerMaintenanceWindowStartTimeOutputReference = MqBrokerMaintenanceWindowStartTimeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MqBrokerMaintenanceWindowStartTimeOutputReference[_g] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerMaintenanceWindowStartTimeOutputReference", version: "19.15.0" };
function mqBrokerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mqBrokerTimeoutsToTerraform = mqBrokerTimeoutsToTerraform;
function mqBrokerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerTimeoutsToHclTerraform = mqBrokerTimeoutsToHclTerraform;
class MqBrokerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MqBrokerTimeoutsOutputReference = MqBrokerTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MqBrokerTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerTimeoutsOutputReference", version: "19.15.0" };
function mqBrokerUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        console_access: cdktf.booleanToTerraform(struct.consoleAccess),
        groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.groups),
        password: cdktf.stringToTerraform(struct.password),
        replication_user: cdktf.booleanToTerraform(struct.replicationUser),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.mqBrokerUserToTerraform = mqBrokerUserToTerraform;
function mqBrokerUserToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        console_access: {
            value: cdktf.booleanToHclTerraform(struct.consoleAccess),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        groups: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.groups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        password: {
            value: cdktf.stringToHclTerraform(struct.password),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        replication_user: {
            value: cdktf.booleanToHclTerraform(struct.replicationUser),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        username: {
            value: cdktf.stringToHclTerraform(struct.username),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.mqBrokerUserToHclTerraform = mqBrokerUserToHclTerraform;
class MqBrokerUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._consoleAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.consoleAccess = this._consoleAccess;
        }
        if (this._groups !== undefined) {
            hasAnyValues = true;
            internalValueResult.groups = this._groups;
        }
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._replicationUser !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicationUser = this._replicationUser;
        }
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._consoleAccess = undefined;
            this._groups = undefined;
            this._password = undefined;
            this._replicationUser = undefined;
            this._username = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._consoleAccess = value.consoleAccess;
            this._groups = value.groups;
            this._password = value.password;
            this._replicationUser = value.replicationUser;
            this._username = value.username;
        }
    }
    get consoleAccess() {
        return this.getBooleanAttribute('console_access');
    }
    set consoleAccess(value) {
        this._consoleAccess = value;
    }
    resetConsoleAccess() {
        this._consoleAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get consoleAccessInput() {
        return this._consoleAccess;
    }
    get groups() {
        return cdktf.Fn.tolist(this.getListAttribute('groups'));
    }
    set groups(value) {
        this._groups = value;
    }
    resetGroups() {
        this._groups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupsInput() {
        return this._groups;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get replicationUser() {
        return this.getBooleanAttribute('replication_user');
    }
    set replicationUser(value) {
        this._replicationUser = value;
    }
    resetReplicationUser() {
        this._replicationUser = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationUserInput() {
        return this._replicationUser;
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.MqBrokerUserOutputReference = MqBrokerUserOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MqBrokerUserOutputReference[_j] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerUserOutputReference", version: "19.15.0" };
class MqBrokerUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MqBrokerUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MqBrokerUserList = MqBrokerUserList;
_k = JSII_RTTI_SYMBOL_1;
MqBrokerUserList[_k] = { fqn: "@cdktf/provider-aws.mqBroker.MqBrokerUserList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/mq_broker aws_mq_broker}
*/
class MqBroker extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a MqBroker resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the MqBroker to import
    * @param importFromId The id of the existing MqBroker that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/mq_broker#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the MqBroker to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_mq_broker", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/mq_broker aws_mq_broker} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MqBrokerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_mq_broker',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // instances - computed: true, optional: false, required: false
        this._instances = new MqBrokerInstancesList(this, "instances", false);
        // configuration - computed: false, optional: true, required: false
        this._configuration = new MqBrokerConfigurationOutputReference(this, "configuration");
        // encryption_options - computed: false, optional: true, required: false
        this._encryptionOptions = new MqBrokerEncryptionOptionsOutputReference(this, "encryption_options");
        // ldap_server_metadata - computed: false, optional: true, required: false
        this._ldapServerMetadata = new MqBrokerLdapServerMetadataOutputReference(this, "ldap_server_metadata");
        // logs - computed: false, optional: true, required: false
        this._logs = new MqBrokerLogsOutputReference(this, "logs");
        // maintenance_window_start_time - computed: false, optional: true, required: false
        this._maintenanceWindowStartTime = new MqBrokerMaintenanceWindowStartTimeOutputReference(this, "maintenance_window_start_time");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MqBrokerTimeoutsOutputReference(this, "timeouts");
        // user - computed: false, optional: false, required: true
        this._user = new MqBrokerUserList(this, "user", true);
        this._applyImmediately = config.applyImmediately;
        this._authenticationStrategy = config.authenticationStrategy;
        this._autoMinorVersionUpgrade = config.autoMinorVersionUpgrade;
        this._brokerName = config.brokerName;
        this._dataReplicationMode = config.dataReplicationMode;
        this._dataReplicationPrimaryBrokerArn = config.dataReplicationPrimaryBrokerArn;
        this._deploymentMode = config.deploymentMode;
        this._engineType = config.engineType;
        this._engineVersion = config.engineVersion;
        this._hostInstanceType = config.hostInstanceType;
        this._id = config.id;
        this._publiclyAccessible = config.publiclyAccessible;
        this._securityGroups = config.securityGroups;
        this._storageType = config.storageType;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._configuration.internalValue = config.configuration;
        this._encryptionOptions.internalValue = config.encryptionOptions;
        this._ldapServerMetadata.internalValue = config.ldapServerMetadata;
        this._logs.internalValue = config.logs;
        this._maintenanceWindowStartTime.internalValue = config.maintenanceWindowStartTime;
        this._timeouts.internalValue = config.timeouts;
        this._user.internalValue = config.user;
    }
    get applyImmediately() {
        return this.getBooleanAttribute('apply_immediately');
    }
    set applyImmediately(value) {
        this._applyImmediately = value;
    }
    resetApplyImmediately() {
        this._applyImmediately = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applyImmediatelyInput() {
        return this._applyImmediately;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authenticationStrategy() {
        return this.getStringAttribute('authentication_strategy');
    }
    set authenticationStrategy(value) {
        this._authenticationStrategy = value;
    }
    resetAuthenticationStrategy() {
        this._authenticationStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationStrategyInput() {
        return this._authenticationStrategy;
    }
    get autoMinorVersionUpgrade() {
        return this.getBooleanAttribute('auto_minor_version_upgrade');
    }
    set autoMinorVersionUpgrade(value) {
        this._autoMinorVersionUpgrade = value;
    }
    resetAutoMinorVersionUpgrade() {
        this._autoMinorVersionUpgrade = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoMinorVersionUpgradeInput() {
        return this._autoMinorVersionUpgrade;
    }
    get brokerName() {
        return this.getStringAttribute('broker_name');
    }
    set brokerName(value) {
        this._brokerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get brokerNameInput() {
        return this._brokerName;
    }
    get dataReplicationMode() {
        return this.getStringAttribute('data_replication_mode');
    }
    set dataReplicationMode(value) {
        this._dataReplicationMode = value;
    }
    resetDataReplicationMode() {
        this._dataReplicationMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataReplicationModeInput() {
        return this._dataReplicationMode;
    }
    get dataReplicationPrimaryBrokerArn() {
        return this.getStringAttribute('data_replication_primary_broker_arn');
    }
    set dataReplicationPrimaryBrokerArn(value) {
        this._dataReplicationPrimaryBrokerArn = value;
    }
    resetDataReplicationPrimaryBrokerArn() {
        this._dataReplicationPrimaryBrokerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataReplicationPrimaryBrokerArnInput() {
        return this._dataReplicationPrimaryBrokerArn;
    }
    get deploymentMode() {
        return this.getStringAttribute('deployment_mode');
    }
    set deploymentMode(value) {
        this._deploymentMode = value;
    }
    resetDeploymentMode() {
        this._deploymentMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentModeInput() {
        return this._deploymentMode;
    }
    get engineType() {
        return this.getStringAttribute('engine_type');
    }
    set engineType(value) {
        this._engineType = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineTypeInput() {
        return this._engineType;
    }
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    set engineVersion(value) {
        this._engineVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineVersionInput() {
        return this._engineVersion;
    }
    get hostInstanceType() {
        return this.getStringAttribute('host_instance_type');
    }
    set hostInstanceType(value) {
        this._hostInstanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostInstanceTypeInput() {
        return this._hostInstanceType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instances() {
        return this._instances;
    }
    // pending_data_replication_mode - computed: true, optional: false, required: false
    get pendingDataReplicationMode() {
        return this.getStringAttribute('pending_data_replication_mode');
    }
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    set publiclyAccessible(value) {
        this._publiclyAccessible = value;
    }
    resetPubliclyAccessible() {
        this._publiclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publiclyAccessibleInput() {
        return this._publiclyAccessible;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    set storageType(value) {
        this._storageType = value;
    }
    resetStorageType() {
        this._storageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageTypeInput() {
        return this._storageType;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get configuration() {
        return this._configuration;
    }
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration.internalValue;
    }
    get encryptionOptions() {
        return this._encryptionOptions;
    }
    putEncryptionOptions(value) {
        this._encryptionOptions.internalValue = value;
    }
    resetEncryptionOptions() {
        this._encryptionOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionOptionsInput() {
        return this._encryptionOptions.internalValue;
    }
    get ldapServerMetadata() {
        return this._ldapServerMetadata;
    }
    putLdapServerMetadata(value) {
        this._ldapServerMetadata.internalValue = value;
    }
    resetLdapServerMetadata() {
        this._ldapServerMetadata.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ldapServerMetadataInput() {
        return this._ldapServerMetadata.internalValue;
    }
    get logs() {
        return this._logs;
    }
    putLogs(value) {
        this._logs.internalValue = value;
    }
    resetLogs() {
        this._logs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logsInput() {
        return this._logs.internalValue;
    }
    get maintenanceWindowStartTime() {
        return this._maintenanceWindowStartTime;
    }
    putMaintenanceWindowStartTime(value) {
        this._maintenanceWindowStartTime.internalValue = value;
    }
    resetMaintenanceWindowStartTime() {
        this._maintenanceWindowStartTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowStartTimeInput() {
        return this._maintenanceWindowStartTime.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get user() {
        return this._user;
    }
    putUser(value) {
        this._user.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get userInput() {
        return this._user.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            apply_immediately: cdktf.booleanToTerraform(this._applyImmediately),
            authentication_strategy: cdktf.stringToTerraform(this._authenticationStrategy),
            auto_minor_version_upgrade: cdktf.booleanToTerraform(this._autoMinorVersionUpgrade),
            broker_name: cdktf.stringToTerraform(this._brokerName),
            data_replication_mode: cdktf.stringToTerraform(this._dataReplicationMode),
            data_replication_primary_broker_arn: cdktf.stringToTerraform(this._dataReplicationPrimaryBrokerArn),
            deployment_mode: cdktf.stringToTerraform(this._deploymentMode),
            engine_type: cdktf.stringToTerraform(this._engineType),
            engine_version: cdktf.stringToTerraform(this._engineVersion),
            host_instance_type: cdktf.stringToTerraform(this._hostInstanceType),
            id: cdktf.stringToTerraform(this._id),
            publicly_accessible: cdktf.booleanToTerraform(this._publiclyAccessible),
            security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroups),
            storage_type: cdktf.stringToTerraform(this._storageType),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            configuration: mqBrokerConfigurationToTerraform(this._configuration.internalValue),
            encryption_options: mqBrokerEncryptionOptionsToTerraform(this._encryptionOptions.internalValue),
            ldap_server_metadata: mqBrokerLdapServerMetadataToTerraform(this._ldapServerMetadata.internalValue),
            logs: mqBrokerLogsToTerraform(this._logs.internalValue),
            maintenance_window_start_time: mqBrokerMaintenanceWindowStartTimeToTerraform(this._maintenanceWindowStartTime.internalValue),
            timeouts: mqBrokerTimeoutsToTerraform(this._timeouts.internalValue),
            user: cdktf.listMapper(mqBrokerUserToTerraform, true)(this._user.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            apply_immediately: {
                value: cdktf.booleanToHclTerraform(this._applyImmediately),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            authentication_strategy: {
                value: cdktf.stringToHclTerraform(this._authenticationStrategy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            auto_minor_version_upgrade: {
                value: cdktf.booleanToHclTerraform(this._autoMinorVersionUpgrade),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            broker_name: {
                value: cdktf.stringToHclTerraform(this._brokerName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_replication_mode: {
                value: cdktf.stringToHclTerraform(this._dataReplicationMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_replication_primary_broker_arn: {
                value: cdktf.stringToHclTerraform(this._dataReplicationPrimaryBrokerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deployment_mode: {
                value: cdktf.stringToHclTerraform(this._deploymentMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            engine_type: {
                value: cdktf.stringToHclTerraform(this._engineType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            engine_version: {
                value: cdktf.stringToHclTerraform(this._engineVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            host_instance_type: {
                value: cdktf.stringToHclTerraform(this._hostInstanceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            publicly_accessible: {
                value: cdktf.booleanToHclTerraform(this._publiclyAccessible),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            security_groups: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._securityGroups),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            storage_type: {
                value: cdktf.stringToHclTerraform(this._storageType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._subnetIds),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            configuration: {
                value: mqBrokerConfigurationToHclTerraform(this._configuration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MqBrokerConfigurationList",
            },
            encryption_options: {
                value: mqBrokerEncryptionOptionsToHclTerraform(this._encryptionOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MqBrokerEncryptionOptionsList",
            },
            ldap_server_metadata: {
                value: mqBrokerLdapServerMetadataToHclTerraform(this._ldapServerMetadata.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MqBrokerLdapServerMetadataList",
            },
            logs: {
                value: mqBrokerLogsToHclTerraform(this._logs.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MqBrokerLogsList",
            },
            maintenance_window_start_time: {
                value: mqBrokerMaintenanceWindowStartTimeToHclTerraform(this._maintenanceWindowStartTime.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "MqBrokerMaintenanceWindowStartTimeList",
            },
            timeouts: {
                value: mqBrokerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "MqBrokerTimeouts",
            },
            user: {
                value: cdktf.listMapperHcl(mqBrokerUserToHclTerraform, true)(this._user.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "MqBrokerUserList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.MqBroker = MqBroker;
_l = JSII_RTTI_SYMBOL_1;
MqBroker[_l] = { fqn: "@cdktf/provider-aws.mqBroker.MqBroker", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
MqBroker.tfResourceType = "aws_mq_broker";
