// @cdktf/provider-aws/lib/data-aws-dx-router-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDxRouterConfiguration = exports.DataAwsDxRouterConfigurationRouterList = exports.DataAwsDxRouterConfigurationRouterOutputReference = exports.dataAwsDxRouterConfigurationRouterToHclTerraform = exports.dataAwsDxRouterConfigurationRouterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsDxRouterConfigurationRouterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDxRouterConfigurationRouterToTerraform = dataAwsDxRouterConfigurationRouterToTerraform;
function dataAwsDxRouterConfigurationRouterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsDxRouterConfigurationRouterToHclTerraform = dataAwsDxRouterConfigurationRouterToHclTerraform;
class DataAwsDxRouterConfigurationRouterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // platform - computed: true, optional: false, required: false
    get platform() {
        return this.getStringAttribute('platform');
    }
    // router_type_identifier - computed: true, optional: false, required: false
    get routerTypeIdentifier() {
        return this.getStringAttribute('router_type_identifier');
    }
    // software - computed: true, optional: false, required: false
    get softwareAttribute() {
        return this.getStringAttribute('software');
    }
    // vendor - computed: true, optional: false, required: false
    get vendor() {
        return this.getStringAttribute('vendor');
    }
    // xslt_template_name - computed: true, optional: false, required: false
    get xsltTemplateName() {
        return this.getStringAttribute('xslt_template_name');
    }
    // xslt_template_name_for_mac_sec - computed: true, optional: false, required: false
    get xsltTemplateNameForMacSec() {
        return this.getStringAttribute('xslt_template_name_for_mac_sec');
    }
}
exports.DataAwsDxRouterConfigurationRouterOutputReference = DataAwsDxRouterConfigurationRouterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDxRouterConfigurationRouterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsDxRouterConfiguration.DataAwsDxRouterConfigurationRouterOutputReference", version: "19.15.0" };
class DataAwsDxRouterConfigurationRouterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDxRouterConfigurationRouterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDxRouterConfigurationRouterList = DataAwsDxRouterConfigurationRouterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDxRouterConfigurationRouterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsDxRouterConfiguration.DataAwsDxRouterConfigurationRouterList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_router_configuration aws_dx_router_configuration}
*/
class DataAwsDxRouterConfiguration extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDxRouterConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDxRouterConfiguration to import
    * @param importFromId The id of the existing DataAwsDxRouterConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_router_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDxRouterConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_dx_router_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_router_configuration aws_dx_router_configuration} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDxRouterConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_router_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // router - computed: true, optional: false, required: false
        this._router = new DataAwsDxRouterConfigurationRouterList(this, "router", false);
        this._id = config.id;
        this._routerTypeIdentifier = config.routerTypeIdentifier;
        this._virtualInterfaceId = config.virtualInterfaceId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // customer_router_config - computed: true, optional: false, required: false
    get customerRouterConfig() {
        return this.getStringAttribute('customer_router_config');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get router() {
        return this._router;
    }
    get routerTypeIdentifier() {
        return this.getStringAttribute('router_type_identifier');
    }
    set routerTypeIdentifier(value) {
        this._routerTypeIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get routerTypeIdentifierInput() {
        return this._routerTypeIdentifier;
    }
    get virtualInterfaceId() {
        return this.getStringAttribute('virtual_interface_id');
    }
    set virtualInterfaceId(value) {
        this._virtualInterfaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualInterfaceIdInput() {
        return this._virtualInterfaceId;
    }
    // virtual_interface_name - computed: true, optional: false, required: false
    get virtualInterfaceName() {
        return this.getStringAttribute('virtual_interface_name');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            router_type_identifier: cdktf.stringToTerraform(this._routerTypeIdentifier),
            virtual_interface_id: cdktf.stringToTerraform(this._virtualInterfaceId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            router_type_identifier: {
                value: cdktf.stringToHclTerraform(this._routerTypeIdentifier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            virtual_interface_id: {
                value: cdktf.stringToHclTerraform(this._virtualInterfaceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDxRouterConfiguration = DataAwsDxRouterConfiguration;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDxRouterConfiguration[_c] = { fqn: "@cdktf/provider-aws.dataAwsDxRouterConfiguration.DataAwsDxRouterConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDxRouterConfiguration.tfResourceType = "aws_dx_router_configuration";
