// @cdktf/provider-aws/lib/backup-global-settings/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupGlobalSettings = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_global_settings aws_backup_global_settings}
*/
class BackupGlobalSettings extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BackupGlobalSettings resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BackupGlobalSettings to import
    * @param importFromId The id of the existing BackupGlobalSettings that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_global_settings#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BackupGlobalSettings to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_backup_global_settings", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_global_settings aws_backup_global_settings} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BackupGlobalSettingsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_global_settings',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._globalSettings = config.globalSettings;
        this._id = config.id;
    }
    get globalSettings() {
        return this.getStringMapAttribute('global_settings');
    }
    set globalSettings(value) {
        this._globalSettings = value;
    }
    // Temporarily expose input value. Use with caution.
    get globalSettingsInput() {
        return this._globalSettings;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            global_settings: cdktf.hashMapper(cdktf.stringToTerraform)(this._globalSettings),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            global_settings: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._globalSettings),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BackupGlobalSettings = BackupGlobalSettings;
_a = JSII_RTTI_SYMBOL_1;
BackupGlobalSettings[_a] = { fqn: "@cdktf/provider-aws.backupGlobalSettings.BackupGlobalSettings", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
BackupGlobalSettings.tfResourceType = "aws_backup_global_settings";
