// @cdktf/provider-aws/lib/cleanrooms-collaboration/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CleanroomsCollaboration = exports.CleanroomsCollaborationTimeoutsOutputReference = exports.cleanroomsCollaborationTimeoutsToHclTerraform = exports.cleanroomsCollaborationTimeoutsToTerraform = exports.CleanroomsCollaborationMemberList = exports.CleanroomsCollaborationMemberOutputReference = exports.cleanroomsCollaborationMemberToHclTerraform = exports.cleanroomsCollaborationMemberToTerraform = exports.CleanroomsCollaborationDataEncryptionMetadataOutputReference = exports.cleanroomsCollaborationDataEncryptionMetadataToHclTerraform = exports.cleanroomsCollaborationDataEncryptionMetadataToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cleanroomsCollaborationDataEncryptionMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allow_clear_text: cdktf.booleanToTerraform(struct.allowClearText),
        allow_duplicates: cdktf.booleanToTerraform(struct.allowDuplicates),
        allow_joins_on_columns_with_different_names: cdktf.booleanToTerraform(struct.allowJoinsOnColumnsWithDifferentNames),
        preserve_nulls: cdktf.booleanToTerraform(struct.preserveNulls),
    };
}
exports.cleanroomsCollaborationDataEncryptionMetadataToTerraform = cleanroomsCollaborationDataEncryptionMetadataToTerraform;
function cleanroomsCollaborationDataEncryptionMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        allow_clear_text: {
            value: cdktf.booleanToHclTerraform(struct.allowClearText),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        allow_duplicates: {
            value: cdktf.booleanToHclTerraform(struct.allowDuplicates),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        allow_joins_on_columns_with_different_names: {
            value: cdktf.booleanToHclTerraform(struct.allowJoinsOnColumnsWithDifferentNames),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        preserve_nulls: {
            value: cdktf.booleanToHclTerraform(struct.preserveNulls),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cleanroomsCollaborationDataEncryptionMetadataToHclTerraform = cleanroomsCollaborationDataEncryptionMetadataToHclTerraform;
class CleanroomsCollaborationDataEncryptionMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowClearText !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowClearText = this._allowClearText;
        }
        if (this._allowDuplicates !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowDuplicates = this._allowDuplicates;
        }
        if (this._allowJoinsOnColumnsWithDifferentNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowJoinsOnColumnsWithDifferentNames = this._allowJoinsOnColumnsWithDifferentNames;
        }
        if (this._preserveNulls !== undefined) {
            hasAnyValues = true;
            internalValueResult.preserveNulls = this._preserveNulls;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowClearText = undefined;
            this._allowDuplicates = undefined;
            this._allowJoinsOnColumnsWithDifferentNames = undefined;
            this._preserveNulls = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowClearText = value.allowClearText;
            this._allowDuplicates = value.allowDuplicates;
            this._allowJoinsOnColumnsWithDifferentNames = value.allowJoinsOnColumnsWithDifferentNames;
            this._preserveNulls = value.preserveNulls;
        }
    }
    get allowClearText() {
        return this.getBooleanAttribute('allow_clear_text');
    }
    set allowClearText(value) {
        this._allowClearText = value;
    }
    // Temporarily expose input value. Use with caution.
    get allowClearTextInput() {
        return this._allowClearText;
    }
    get allowDuplicates() {
        return this.getBooleanAttribute('allow_duplicates');
    }
    set allowDuplicates(value) {
        this._allowDuplicates = value;
    }
    // Temporarily expose input value. Use with caution.
    get allowDuplicatesInput() {
        return this._allowDuplicates;
    }
    get allowJoinsOnColumnsWithDifferentNames() {
        return this.getBooleanAttribute('allow_joins_on_columns_with_different_names');
    }
    set allowJoinsOnColumnsWithDifferentNames(value) {
        this._allowJoinsOnColumnsWithDifferentNames = value;
    }
    // Temporarily expose input value. Use with caution.
    get allowJoinsOnColumnsWithDifferentNamesInput() {
        return this._allowJoinsOnColumnsWithDifferentNames;
    }
    get preserveNulls() {
        return this.getBooleanAttribute('preserve_nulls');
    }
    set preserveNulls(value) {
        this._preserveNulls = value;
    }
    // Temporarily expose input value. Use with caution.
    get preserveNullsInput() {
        return this._preserveNulls;
    }
}
exports.CleanroomsCollaborationDataEncryptionMetadataOutputReference = CleanroomsCollaborationDataEncryptionMetadataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CleanroomsCollaborationDataEncryptionMetadataOutputReference[_a] = { fqn: "@cdktf/provider-aws.cleanroomsCollaboration.CleanroomsCollaborationDataEncryptionMetadataOutputReference", version: "19.15.0" };
function cleanroomsCollaborationMemberToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account_id: cdktf.stringToTerraform(struct.accountId),
        display_name: cdktf.stringToTerraform(struct.displayName),
        member_abilities: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.memberAbilities),
    };
}
exports.cleanroomsCollaborationMemberToTerraform = cleanroomsCollaborationMemberToTerraform;
function cleanroomsCollaborationMemberToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        account_id: {
            value: cdktf.stringToHclTerraform(struct.accountId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        display_name: {
            value: cdktf.stringToHclTerraform(struct.displayName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        member_abilities: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.memberAbilities),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cleanroomsCollaborationMemberToHclTerraform = cleanroomsCollaborationMemberToHclTerraform;
class CleanroomsCollaborationMemberOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accountId !== undefined) {
            hasAnyValues = true;
            internalValueResult.accountId = this._accountId;
        }
        if (this._displayName !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayName = this._displayName;
        }
        if (this._memberAbilities !== undefined) {
            hasAnyValues = true;
            internalValueResult.memberAbilities = this._memberAbilities;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._accountId = undefined;
            this._displayName = undefined;
            this._memberAbilities = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._accountId = value.accountId;
            this._displayName = value.displayName;
            this._memberAbilities = value.memberAbilities;
        }
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get memberAbilities() {
        return this.getListAttribute('member_abilities');
    }
    set memberAbilities(value) {
        this._memberAbilities = value;
    }
    // Temporarily expose input value. Use with caution.
    get memberAbilitiesInput() {
        return this._memberAbilities;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.CleanroomsCollaborationMemberOutputReference = CleanroomsCollaborationMemberOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CleanroomsCollaborationMemberOutputReference[_b] = { fqn: "@cdktf/provider-aws.cleanroomsCollaboration.CleanroomsCollaborationMemberOutputReference", version: "19.15.0" };
class CleanroomsCollaborationMemberList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CleanroomsCollaborationMemberOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CleanroomsCollaborationMemberList = CleanroomsCollaborationMemberList;
_c = JSII_RTTI_SYMBOL_1;
CleanroomsCollaborationMemberList[_c] = { fqn: "@cdktf/provider-aws.cleanroomsCollaboration.CleanroomsCollaborationMemberList", version: "19.15.0" };
function cleanroomsCollaborationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cleanroomsCollaborationTimeoutsToTerraform = cleanroomsCollaborationTimeoutsToTerraform;
function cleanroomsCollaborationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cleanroomsCollaborationTimeoutsToHclTerraform = cleanroomsCollaborationTimeoutsToHclTerraform;
class CleanroomsCollaborationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CleanroomsCollaborationTimeoutsOutputReference = CleanroomsCollaborationTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CleanroomsCollaborationTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.cleanroomsCollaboration.CleanroomsCollaborationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cleanrooms_collaboration aws_cleanrooms_collaboration}
*/
class CleanroomsCollaboration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CleanroomsCollaboration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CleanroomsCollaboration to import
    * @param importFromId The id of the existing CleanroomsCollaboration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cleanrooms_collaboration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CleanroomsCollaboration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cleanrooms_collaboration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cleanrooms_collaboration aws_cleanrooms_collaboration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CleanroomsCollaborationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cleanrooms_collaboration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // data_encryption_metadata - computed: false, optional: true, required: false
        this._dataEncryptionMetadata = new CleanroomsCollaborationDataEncryptionMetadataOutputReference(this, "data_encryption_metadata");
        // member - computed: false, optional: true, required: false
        this._member = new CleanroomsCollaborationMemberList(this, "member", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CleanroomsCollaborationTimeoutsOutputReference(this, "timeouts");
        this._creatorDisplayName = config.creatorDisplayName;
        this._creatorMemberAbilities = config.creatorMemberAbilities;
        this._description = config.description;
        this._name = config.name;
        this._queryLogStatus = config.queryLogStatus;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._dataEncryptionMetadata.internalValue = config.dataEncryptionMetadata;
        this._member.internalValue = config.member;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get creatorDisplayName() {
        return this.getStringAttribute('creator_display_name');
    }
    set creatorDisplayName(value) {
        this._creatorDisplayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get creatorDisplayNameInput() {
        return this._creatorDisplayName;
    }
    get creatorMemberAbilities() {
        return this.getListAttribute('creator_member_abilities');
    }
    set creatorMemberAbilities(value) {
        this._creatorMemberAbilities = value;
    }
    // Temporarily expose input value. Use with caution.
    get creatorMemberAbilitiesInput() {
        return this._creatorMemberAbilities;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get queryLogStatus() {
        return this.getStringAttribute('query_log_status');
    }
    set queryLogStatus(value) {
        this._queryLogStatus = value;
    }
    // Temporarily expose input value. Use with caution.
    get queryLogStatusInput() {
        return this._queryLogStatus;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get dataEncryptionMetadata() {
        return this._dataEncryptionMetadata;
    }
    putDataEncryptionMetadata(value) {
        this._dataEncryptionMetadata.internalValue = value;
    }
    resetDataEncryptionMetadata() {
        this._dataEncryptionMetadata.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataEncryptionMetadataInput() {
        return this._dataEncryptionMetadata.internalValue;
    }
    get member() {
        return this._member;
    }
    putMember(value) {
        this._member.internalValue = value;
    }
    resetMember() {
        this._member.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get memberInput() {
        return this._member.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            creator_display_name: cdktf.stringToTerraform(this._creatorDisplayName),
            creator_member_abilities: cdktf.listMapper(cdktf.stringToTerraform, false)(this._creatorMemberAbilities),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            query_log_status: cdktf.stringToTerraform(this._queryLogStatus),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            data_encryption_metadata: cleanroomsCollaborationDataEncryptionMetadataToTerraform(this._dataEncryptionMetadata.internalValue),
            member: cdktf.listMapper(cleanroomsCollaborationMemberToTerraform, true)(this._member.internalValue),
            timeouts: cleanroomsCollaborationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            creator_display_name: {
                value: cdktf.stringToHclTerraform(this._creatorDisplayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            creator_member_abilities: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._creatorMemberAbilities),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            query_log_status: {
                value: cdktf.stringToHclTerraform(this._queryLogStatus),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            data_encryption_metadata: {
                value: cleanroomsCollaborationDataEncryptionMetadataToHclTerraform(this._dataEncryptionMetadata.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CleanroomsCollaborationDataEncryptionMetadataList",
            },
            member: {
                value: cdktf.listMapperHcl(cleanroomsCollaborationMemberToHclTerraform, true)(this._member.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CleanroomsCollaborationMemberList",
            },
            timeouts: {
                value: cleanroomsCollaborationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CleanroomsCollaborationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CleanroomsCollaboration = CleanroomsCollaboration;
_e = JSII_RTTI_SYMBOL_1;
CleanroomsCollaboration[_e] = { fqn: "@cdktf/provider-aws.cleanroomsCollaboration.CleanroomsCollaboration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CleanroomsCollaboration.tfResourceType = "aws_cleanrooms_collaboration";
