// @cdktf/provider-aws/lib/backup-vault-lock-configuration/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupVaultLockConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_vault_lock_configuration aws_backup_vault_lock_configuration}
*/
class BackupVaultLockConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BackupVaultLockConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BackupVaultLockConfiguration to import
    * @param importFromId The id of the existing BackupVaultLockConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_vault_lock_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BackupVaultLockConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_backup_vault_lock_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/backup_vault_lock_configuration aws_backup_vault_lock_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BackupVaultLockConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_vault_lock_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._backupVaultName = config.backupVaultName;
        this._changeableForDays = config.changeableForDays;
        this._id = config.id;
        this._maxRetentionDays = config.maxRetentionDays;
        this._minRetentionDays = config.minRetentionDays;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // backup_vault_arn - computed: true, optional: false, required: false
    get backupVaultArn() {
        return this.getStringAttribute('backup_vault_arn');
    }
    get backupVaultName() {
        return this.getStringAttribute('backup_vault_name');
    }
    set backupVaultName(value) {
        this._backupVaultName = value;
    }
    // Temporarily expose input value. Use with caution.
    get backupVaultNameInput() {
        return this._backupVaultName;
    }
    get changeableForDays() {
        return this.getNumberAttribute('changeable_for_days');
    }
    set changeableForDays(value) {
        this._changeableForDays = value;
    }
    resetChangeableForDays() {
        this._changeableForDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get changeableForDaysInput() {
        return this._changeableForDays;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get maxRetentionDays() {
        return this.getNumberAttribute('max_retention_days');
    }
    set maxRetentionDays(value) {
        this._maxRetentionDays = value;
    }
    resetMaxRetentionDays() {
        this._maxRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetentionDaysInput() {
        return this._maxRetentionDays;
    }
    get minRetentionDays() {
        return this.getNumberAttribute('min_retention_days');
    }
    set minRetentionDays(value) {
        this._minRetentionDays = value;
    }
    resetMinRetentionDays() {
        this._minRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minRetentionDaysInput() {
        return this._minRetentionDays;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            backup_vault_name: cdktf.stringToTerraform(this._backupVaultName),
            changeable_for_days: cdktf.numberToTerraform(this._changeableForDays),
            id: cdktf.stringToTerraform(this._id),
            max_retention_days: cdktf.numberToTerraform(this._maxRetentionDays),
            min_retention_days: cdktf.numberToTerraform(this._minRetentionDays),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            backup_vault_name: {
                value: cdktf.stringToHclTerraform(this._backupVaultName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            changeable_for_days: {
                value: cdktf.numberToHclTerraform(this._changeableForDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            max_retention_days: {
                value: cdktf.numberToHclTerraform(this._maxRetentionDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            min_retention_days: {
                value: cdktf.numberToHclTerraform(this._minRetentionDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BackupVaultLockConfiguration = BackupVaultLockConfiguration;
_a = JSII_RTTI_SYMBOL_1;
BackupVaultLockConfiguration[_a] = { fqn: "@cdktf/provider-aws.backupVaultLockConfiguration.BackupVaultLockConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
BackupVaultLockConfiguration.tfResourceType = "aws_backup_vault_lock_configuration";
