// @cdktf/provider-aws/lib/redshiftserverless-namespace/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessNamespace = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_namespace aws_redshiftserverless_namespace}
*/
class RedshiftserverlessNamespace extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftserverlessNamespace resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftserverlessNamespace to import
    * @param importFromId The id of the existing RedshiftserverlessNamespace that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_namespace#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftserverlessNamespace to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshiftserverless_namespace", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_namespace aws_redshiftserverless_namespace} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessNamespaceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_namespace',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._adminPasswordSecretKmsKeyId = config.adminPasswordSecretKmsKeyId;
        this._adminUserPassword = config.adminUserPassword;
        this._adminUsername = config.adminUsername;
        this._dbName = config.dbName;
        this._defaultIamRoleArn = config.defaultIamRoleArn;
        this._iamRoles = config.iamRoles;
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._logExports = config.logExports;
        this._manageAdminPassword = config.manageAdminPassword;
        this._namespaceName = config.namespaceName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // admin_password_secret_arn - computed: true, optional: false, required: false
    get adminPasswordSecretArn() {
        return this.getStringAttribute('admin_password_secret_arn');
    }
    get adminPasswordSecretKmsKeyId() {
        return this.getStringAttribute('admin_password_secret_kms_key_id');
    }
    set adminPasswordSecretKmsKeyId(value) {
        this._adminPasswordSecretKmsKeyId = value;
    }
    resetAdminPasswordSecretKmsKeyId() {
        this._adminPasswordSecretKmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminPasswordSecretKmsKeyIdInput() {
        return this._adminPasswordSecretKmsKeyId;
    }
    get adminUserPassword() {
        return this.getStringAttribute('admin_user_password');
    }
    set adminUserPassword(value) {
        this._adminUserPassword = value;
    }
    resetAdminUserPassword() {
        this._adminUserPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUserPasswordInput() {
        return this._adminUserPassword;
    }
    get adminUsername() {
        return this.getStringAttribute('admin_username');
    }
    set adminUsername(value) {
        this._adminUsername = value;
    }
    resetAdminUsername() {
        this._adminUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUsernameInput() {
        return this._adminUsername;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get dbName() {
        return this.getStringAttribute('db_name');
    }
    set dbName(value) {
        this._dbName = value;
    }
    resetDbName() {
        this._dbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbNameInput() {
        return this._dbName;
    }
    get defaultIamRoleArn() {
        return this.getStringAttribute('default_iam_role_arn');
    }
    set defaultIamRoleArn(value) {
        this._defaultIamRoleArn = value;
    }
    resetDefaultIamRoleArn() {
        this._defaultIamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultIamRoleArnInput() {
        return this._defaultIamRoleArn;
    }
    get iamRoles() {
        return cdktf.Fn.tolist(this.getListAttribute('iam_roles'));
    }
    set iamRoles(value) {
        this._iamRoles = value;
    }
    resetIamRoles() {
        this._iamRoles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRolesInput() {
        return this._iamRoles;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get logExports() {
        return cdktf.Fn.tolist(this.getListAttribute('log_exports'));
    }
    set logExports(value) {
        this._logExports = value;
    }
    resetLogExports() {
        this._logExports = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logExportsInput() {
        return this._logExports;
    }
    get manageAdminPassword() {
        return this.getBooleanAttribute('manage_admin_password');
    }
    set manageAdminPassword(value) {
        this._manageAdminPassword = value;
    }
    resetManageAdminPassword() {
        this._manageAdminPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get manageAdminPasswordInput() {
        return this._manageAdminPassword;
    }
    // namespace_id - computed: true, optional: false, required: false
    get namespaceId() {
        return this.getStringAttribute('namespace_id');
    }
    get namespaceName() {
        return this.getStringAttribute('namespace_name');
    }
    set namespaceName(value) {
        this._namespaceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceNameInput() {
        return this._namespaceName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            admin_password_secret_kms_key_id: cdktf.stringToTerraform(this._adminPasswordSecretKmsKeyId),
            admin_user_password: cdktf.stringToTerraform(this._adminUserPassword),
            admin_username: cdktf.stringToTerraform(this._adminUsername),
            db_name: cdktf.stringToTerraform(this._dbName),
            default_iam_role_arn: cdktf.stringToTerraform(this._defaultIamRoleArn),
            iam_roles: cdktf.listMapper(cdktf.stringToTerraform, false)(this._iamRoles),
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            log_exports: cdktf.listMapper(cdktf.stringToTerraform, false)(this._logExports),
            manage_admin_password: cdktf.booleanToTerraform(this._manageAdminPassword),
            namespace_name: cdktf.stringToTerraform(this._namespaceName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            admin_password_secret_kms_key_id: {
                value: cdktf.stringToHclTerraform(this._adminPasswordSecretKmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            admin_user_password: {
                value: cdktf.stringToHclTerraform(this._adminUserPassword),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            admin_username: {
                value: cdktf.stringToHclTerraform(this._adminUsername),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            db_name: {
                value: cdktf.stringToHclTerraform(this._dbName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            default_iam_role_arn: {
                value: cdktf.stringToHclTerraform(this._defaultIamRoleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            iam_roles: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._iamRoles),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_id: {
                value: cdktf.stringToHclTerraform(this._kmsKeyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            log_exports: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._logExports),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            manage_admin_password: {
                value: cdktf.booleanToHclTerraform(this._manageAdminPassword),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            namespace_name: {
                value: cdktf.stringToHclTerraform(this._namespaceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftserverlessNamespace = RedshiftserverlessNamespace;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessNamespace[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessNamespace.RedshiftserverlessNamespace", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessNamespace.tfResourceType = "aws_redshiftserverless_namespace";
