// @cdktf/provider-aws/lib/ecr-replication-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcrReplicationConfiguration = exports.EcrReplicationConfigurationReplicationConfigurationOutputReference = exports.ecrReplicationConfigurationReplicationConfigurationToHclTerraform = exports.ecrReplicationConfigurationReplicationConfigurationToTerraform = exports.EcrReplicationConfigurationReplicationConfigurationRuleList = exports.EcrReplicationConfigurationReplicationConfigurationRuleOutputReference = exports.ecrReplicationConfigurationReplicationConfigurationRuleToHclTerraform = exports.ecrReplicationConfigurationReplicationConfigurationRuleToTerraform = exports.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList = exports.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference = exports.ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToHclTerraform = exports.ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToTerraform = exports.EcrReplicationConfigurationReplicationConfigurationRuleDestinationList = exports.EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference = exports.ecrReplicationConfigurationReplicationConfigurationRuleDestinationToHclTerraform = exports.ecrReplicationConfigurationReplicationConfigurationRuleDestinationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecrReplicationConfigurationReplicationConfigurationRuleDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        region: cdktf.stringToTerraform(struct.region),
        registry_id: cdktf.stringToTerraform(struct.registryId),
    };
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleDestinationToTerraform = ecrReplicationConfigurationReplicationConfigurationRuleDestinationToTerraform;
function ecrReplicationConfigurationReplicationConfigurationRuleDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        registry_id: {
            value: cdktf.stringToHclTerraform(struct.registryId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleDestinationToHclTerraform = ecrReplicationConfigurationReplicationConfigurationRuleDestinationToHclTerraform;
class EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._registryId !== undefined) {
            hasAnyValues = true;
            internalValueResult.registryId = this._registryId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._region = undefined;
            this._registryId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._region = value.region;
            this._registryId = value.registryId;
        }
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    set registryId(value) {
        this._registryId = value;
    }
    // Temporarily expose input value. Use with caution.
    get registryIdInput() {
        return this._registryId;
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference = EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference", version: "19.15.0" };
class EcrReplicationConfigurationReplicationConfigurationRuleDestinationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcrReplicationConfigurationReplicationConfigurationRuleDestinationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleDestinationList = EcrReplicationConfigurationReplicationConfigurationRuleDestinationList;
_b = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleDestinationList[_b] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleDestinationList", version: "19.15.0" };
function ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.stringToTerraform(struct.filter),
        filter_type: cdktf.stringToTerraform(struct.filterType),
    };
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToTerraform = ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToTerraform;
function ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter: {
            value: cdktf.stringToHclTerraform(struct.filter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        filter_type: {
            value: cdktf.stringToHclTerraform(struct.filterType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToHclTerraform = ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToHclTerraform;
class EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter;
        }
        if (this._filterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterType = this._filterType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._filter = undefined;
            this._filterType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._filter = value.filter;
            this._filterType = value.filterType;
        }
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get filterType() {
        return this.getStringAttribute('filter_type');
    }
    set filterType(value) {
        this._filterType = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterTypeInput() {
        return this._filterType;
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference = EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference[_c] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference", version: "19.15.0" };
class EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList = EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList;
_d = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList[_d] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList", version: "19.15.0" };
function ecrReplicationConfigurationReplicationConfigurationRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination: cdktf.listMapper(ecrReplicationConfigurationReplicationConfigurationRuleDestinationToTerraform, true)(struct.destination),
        repository_filter: cdktf.listMapper(ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToTerraform, true)(struct.repositoryFilter),
    };
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleToTerraform = ecrReplicationConfigurationReplicationConfigurationRuleToTerraform;
function ecrReplicationConfigurationReplicationConfigurationRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination: {
            value: cdktf.listMapperHcl(ecrReplicationConfigurationReplicationConfigurationRuleDestinationToHclTerraform, true)(struct.destination),
            isBlock: true,
            type: "list",
            storageClassType: "EcrReplicationConfigurationReplicationConfigurationRuleDestinationList",
        },
        repository_filter: {
            value: cdktf.listMapperHcl(ecrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterToHclTerraform, true)(struct.repositoryFilter),
            isBlock: true,
            type: "list",
            storageClassType: "EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecrReplicationConfigurationReplicationConfigurationRuleToHclTerraform = ecrReplicationConfigurationReplicationConfigurationRuleToHclTerraform;
class EcrReplicationConfigurationReplicationConfigurationRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // destination - computed: false, optional: false, required: true
        this._destination = new EcrReplicationConfigurationReplicationConfigurationRuleDestinationList(this, "destination", false);
        // repository_filter - computed: false, optional: true, required: false
        this._repositoryFilter = new EcrReplicationConfigurationReplicationConfigurationRuleRepositoryFilterList(this, "repository_filter", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.destination = this._destination?.internalValue;
        }
        if (this._repositoryFilter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryFilter = this._repositoryFilter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._destination.internalValue = undefined;
            this._repositoryFilter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._destination.internalValue = value.destination;
            this._repositoryFilter.internalValue = value.repositoryFilter;
        }
    }
    get destination() {
        return this._destination;
    }
    putDestination(value) {
        this._destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination.internalValue;
    }
    get repositoryFilter() {
        return this._repositoryFilter;
    }
    putRepositoryFilter(value) {
        this._repositoryFilter.internalValue = value;
    }
    resetRepositoryFilter() {
        this._repositoryFilter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryFilterInput() {
        return this._repositoryFilter.internalValue;
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleOutputReference = EcrReplicationConfigurationReplicationConfigurationRuleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleOutputReference[_e] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleOutputReference", version: "19.15.0" };
class EcrReplicationConfigurationReplicationConfigurationRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcrReplicationConfigurationReplicationConfigurationRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationRuleList = EcrReplicationConfigurationReplicationConfigurationRuleList;
_f = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationRuleList[_f] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationRuleList", version: "19.15.0" };
function ecrReplicationConfigurationReplicationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        rule: cdktf.listMapper(ecrReplicationConfigurationReplicationConfigurationRuleToTerraform, true)(struct.rule),
    };
}
exports.ecrReplicationConfigurationReplicationConfigurationToTerraform = ecrReplicationConfigurationReplicationConfigurationToTerraform;
function ecrReplicationConfigurationReplicationConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        rule: {
            value: cdktf.listMapperHcl(ecrReplicationConfigurationReplicationConfigurationRuleToHclTerraform, true)(struct.rule),
            isBlock: true,
            type: "list",
            storageClassType: "EcrReplicationConfigurationReplicationConfigurationRuleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.ecrReplicationConfigurationReplicationConfigurationToHclTerraform = ecrReplicationConfigurationReplicationConfigurationToHclTerraform;
class EcrReplicationConfigurationReplicationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rule - computed: false, optional: false, required: true
        this._rule = new EcrReplicationConfigurationReplicationConfigurationRuleList(this, "rule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rule = this._rule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rule.internalValue = value.rule;
        }
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
}
exports.EcrReplicationConfigurationReplicationConfigurationOutputReference = EcrReplicationConfigurationReplicationConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EcrReplicationConfigurationReplicationConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfigurationReplicationConfigurationOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_replication_configuration aws_ecr_replication_configuration}
*/
class EcrReplicationConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EcrReplicationConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EcrReplicationConfiguration to import
    * @param importFromId The id of the existing EcrReplicationConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_replication_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EcrReplicationConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecr_replication_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/ecr_replication_configuration aws_ecr_replication_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EcrReplicationConfigurationConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_replication_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // replication_configuration - computed: false, optional: true, required: false
        this._replicationConfiguration = new EcrReplicationConfigurationReplicationConfigurationOutputReference(this, "replication_configuration");
        this._id = config.id;
        this._replicationConfiguration.internalValue = config.replicationConfiguration;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // registry_id - computed: true, optional: false, required: false
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    get replicationConfiguration() {
        return this._replicationConfiguration;
    }
    putReplicationConfiguration(value) {
        this._replicationConfiguration.internalValue = value;
    }
    resetReplicationConfiguration() {
        this._replicationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationConfigurationInput() {
        return this._replicationConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            replication_configuration: ecrReplicationConfigurationReplicationConfigurationToTerraform(this._replicationConfiguration.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            replication_configuration: {
                value: ecrReplicationConfigurationReplicationConfigurationToHclTerraform(this._replicationConfiguration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EcrReplicationConfigurationReplicationConfigurationList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EcrReplicationConfiguration = EcrReplicationConfiguration;
_h = JSII_RTTI_SYMBOL_1;
EcrReplicationConfiguration[_h] = { fqn: "@cdktf/provider-aws.ecrReplicationConfiguration.EcrReplicationConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EcrReplicationConfiguration.tfResourceType = "aws_ecr_replication_configuration";
