// @cdktf/provider-aws/lib/redshiftserverless-usage-limit/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessUsageLimit = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_usage_limit aws_redshiftserverless_usage_limit}
*/
class RedshiftserverlessUsageLimit extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedshiftserverlessUsageLimit resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedshiftserverlessUsageLimit to import
    * @param importFromId The id of the existing RedshiftserverlessUsageLimit that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_usage_limit#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedshiftserverlessUsageLimit to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshiftserverless_usage_limit", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/redshiftserverless_usage_limit aws_redshiftserverless_usage_limit} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessUsageLimitConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_usage_limit',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._amount = config.amount;
        this._breachAction = config.breachAction;
        this._id = config.id;
        this._period = config.period;
        this._resourceArn = config.resourceArn;
        this._usageType = config.usageType;
    }
    get amount() {
        return this.getNumberAttribute('amount');
    }
    set amount(value) {
        this._amount = value;
    }
    // Temporarily expose input value. Use with caution.
    get amountInput() {
        return this._amount;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get breachAction() {
        return this.getStringAttribute('breach_action');
    }
    set breachAction(value) {
        this._breachAction = value;
    }
    resetBreachAction() {
        this._breachAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breachActionInput() {
        return this._breachAction;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get period() {
        return this.getStringAttribute('period');
    }
    set period(value) {
        this._period = value;
    }
    resetPeriod() {
        this._period = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodInput() {
        return this._period;
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get usageType() {
        return this.getStringAttribute('usage_type');
    }
    set usageType(value) {
        this._usageType = value;
    }
    // Temporarily expose input value. Use with caution.
    get usageTypeInput() {
        return this._usageType;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            amount: cdktf.numberToTerraform(this._amount),
            breach_action: cdktf.stringToTerraform(this._breachAction),
            id: cdktf.stringToTerraform(this._id),
            period: cdktf.stringToTerraform(this._period),
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
            usage_type: cdktf.stringToTerraform(this._usageType),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            amount: {
                value: cdktf.numberToHclTerraform(this._amount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            breach_action: {
                value: cdktf.stringToHclTerraform(this._breachAction),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            period: {
                value: cdktf.stringToHclTerraform(this._period),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_arn: {
                value: cdktf.stringToHclTerraform(this._resourceArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            usage_type: {
                value: cdktf.stringToHclTerraform(this._usageType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedshiftserverlessUsageLimit = RedshiftserverlessUsageLimit;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessUsageLimit[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessUsageLimit.RedshiftserverlessUsageLimit", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessUsageLimit.tfResourceType = "aws_redshiftserverless_usage_limit";
