// @winglang/sdk/lib/.gen/providers/aws/dynamodb-table-item/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/dynamodb_table_item
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamodbTableItem = void 0;
const cdktf = __importStar(require("cdktf"));
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/dynamodb_table_item aws_dynamodb_table_item}
*/
class DynamodbTableItem extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DynamodbTableItem resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DynamodbTableItem to import
    * @param importFromId The id of the existing DynamodbTableItem that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/dynamodb_table_item#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DynamodbTableItem to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_dynamodb_table_item", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.31.0/docs/resources/dynamodb_table_item aws_dynamodb_table_item} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DynamodbTableItemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table_item',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.31.0',
                providerVersionConstraint: '~>5.31.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._hashKey = config.hashKey;
        this._id = config.id;
        this._item = config.item;
        this._rangeKey = config.rangeKey;
        this._tableName = config.tableName;
    }
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    set hashKey(value) {
        this._hashKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get hashKeyInput() {
        return this._hashKey;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get item() {
        return this.getStringAttribute('item');
    }
    set item(value) {
        this._item = value;
    }
    // Temporarily expose input value. Use with caution.
    get itemInput() {
        return this._item;
    }
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    set rangeKey(value) {
        this._rangeKey = value;
    }
    resetRangeKey() {
        this._rangeKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeKeyInput() {
        return this._rangeKey;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hash_key: cdktf.stringToTerraform(this._hashKey),
            id: cdktf.stringToTerraform(this._id),
            item: cdktf.stringToTerraform(this._item),
            range_key: cdktf.stringToTerraform(this._rangeKey),
            table_name: cdktf.stringToTerraform(this._tableName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            hash_key: {
                value: cdktf.stringToHclTerraform(this._hashKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            item: {
                value: cdktf.stringToHclTerraform(this._item),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            range_key: {
                value: cdktf.stringToHclTerraform(this._rangeKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            table_name: {
                value: cdktf.stringToHclTerraform(this._tableName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DynamodbTableItem = DynamodbTableItem;
// =================
// STATIC PROPERTIES
// =================
DynamodbTableItem.tfResourceType = "aws_dynamodb_table_item";
