// @cdktf/provider-aws/lib/data-aws-kendra-index/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKendraIndex = exports.DataAwsKendraIndexUserTokenConfigurationsList = exports.DataAwsKendraIndexUserTokenConfigurationsOutputReference = exports.dataAwsKendraIndexUserTokenConfigurationsToHclTerraform = exports.dataAwsKendraIndexUserTokenConfigurationsToTerraform = exports.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList = exports.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference = exports.dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToHclTerraform = exports.dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToTerraform = exports.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList = exports.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference = exports.dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToHclTerraform = exports.dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToTerraform = exports.DataAwsKendraIndexUserGroupResolutionConfigurationList = exports.DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference = exports.dataAwsKendraIndexUserGroupResolutionConfigurationToHclTerraform = exports.dataAwsKendraIndexUserGroupResolutionConfigurationToTerraform = exports.DataAwsKendraIndexServerSideEncryptionConfigurationList = exports.DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference = exports.dataAwsKendraIndexServerSideEncryptionConfigurationToHclTerraform = exports.dataAwsKendraIndexServerSideEncryptionConfigurationToTerraform = exports.DataAwsKendraIndexIndexStatisticsList = exports.DataAwsKendraIndexIndexStatisticsOutputReference = exports.dataAwsKendraIndexIndexStatisticsToHclTerraform = exports.dataAwsKendraIndexIndexStatisticsToTerraform = exports.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList = exports.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference = exports.dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToHclTerraform = exports.dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToTerraform = exports.DataAwsKendraIndexIndexStatisticsFaqStatisticsList = exports.DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference = exports.dataAwsKendraIndexIndexStatisticsFaqStatisticsToHclTerraform = exports.dataAwsKendraIndexIndexStatisticsFaqStatisticsToTerraform = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToHclTerraform = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToTerraform = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToHclTerraform = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToTerraform = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList = exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToHclTerraform = exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToTerraform = exports.DataAwsKendraIndexCapacityUnitsList = exports.DataAwsKendraIndexCapacityUnitsOutputReference = exports.dataAwsKendraIndexCapacityUnitsToHclTerraform = exports.dataAwsKendraIndexCapacityUnitsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsKendraIndexCapacityUnitsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexCapacityUnitsToTerraform = dataAwsKendraIndexCapacityUnitsToTerraform;
function dataAwsKendraIndexCapacityUnitsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexCapacityUnitsToHclTerraform = dataAwsKendraIndexCapacityUnitsToHclTerraform;
class DataAwsKendraIndexCapacityUnitsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // query_capacity_units - computed: true, optional: false, required: false
    get queryCapacityUnits() {
        return this.getNumberAttribute('query_capacity_units');
    }
    // storage_capacity_units - computed: true, optional: false, required: false
    get storageCapacityUnits() {
        return this.getNumberAttribute('storage_capacity_units');
    }
}
exports.DataAwsKendraIndexCapacityUnitsOutputReference = DataAwsKendraIndexCapacityUnitsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexCapacityUnitsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexCapacityUnitsOutputReference", version: "19.15.0" };
class DataAwsKendraIndexCapacityUnitsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexCapacityUnitsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexCapacityUnitsList = DataAwsKendraIndexCapacityUnitsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexCapacityUnitsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexCapacityUnitsList", version: "19.15.0" };
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToTerraform;
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToHclTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceToHclTerraform;
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // values_importance_map - computed: true, optional: false, required: false
        this._valuesImportanceMap = new cdktf.NumberMap(this, "values_importance_map");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // duration - computed: true, optional: false, required: false
    get duration() {
        return this.getStringAttribute('duration');
    }
    // freshness - computed: true, optional: false, required: false
    get freshness() {
        return this.getBooleanAttribute('freshness');
    }
    // importance - computed: true, optional: false, required: false
    get importance() {
        return this.getNumberAttribute('importance');
    }
    // rank_order - computed: true, optional: false, required: false
    get rankOrder() {
        return this.getStringAttribute('rank_order');
    }
    get valuesImportanceMap() {
        return this._valuesImportanceMap;
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference", version: "19.15.0" };
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList[_d] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList", version: "19.15.0" };
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToTerraform;
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToHclTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchToHclTerraform;
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // displayable - computed: true, optional: false, required: false
    get displayable() {
        return this.getBooleanAttribute('displayable');
    }
    // facetable - computed: true, optional: false, required: false
    get facetable() {
        return this.getBooleanAttribute('facetable');
    }
    // searchable - computed: true, optional: false, required: false
    get searchable() {
        return this.getBooleanAttribute('searchable');
    }
    // sortable - computed: true, optional: false, required: false
    get sortable() {
        return this.getBooleanAttribute('sortable');
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference", version: "19.15.0" };
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList[_f] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList", version: "19.15.0" };
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToTerraform;
function dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToHclTerraform = dataAwsKendraIndexDocumentMetadataConfigurationUpdatesToHclTerraform;
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // relevance - computed: true, optional: false, required: false
        this._relevance = new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesRelevanceList(this, "relevance", false);
        // search - computed: true, optional: false, required: false
        this._search = new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesSearchList(this, "search", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get relevance() {
        return this._relevance;
    }
    get search() {
        return this._search;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference", version: "19.15.0" };
class DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList = DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList[_h] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList", version: "19.15.0" };
function dataAwsKendraIndexIndexStatisticsFaqStatisticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexIndexStatisticsFaqStatisticsToTerraform = dataAwsKendraIndexIndexStatisticsFaqStatisticsToTerraform;
function dataAwsKendraIndexIndexStatisticsFaqStatisticsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexIndexStatisticsFaqStatisticsToHclTerraform = dataAwsKendraIndexIndexStatisticsFaqStatisticsToHclTerraform;
class DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // indexed_question_answers_count - computed: true, optional: false, required: false
    get indexedQuestionAnswersCount() {
        return this.getNumberAttribute('indexed_question_answers_count');
    }
}
exports.DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference = DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference", version: "19.15.0" };
class DataAwsKendraIndexIndexStatisticsFaqStatisticsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexIndexStatisticsFaqStatisticsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexIndexStatisticsFaqStatisticsList = DataAwsKendraIndexIndexStatisticsFaqStatisticsList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsFaqStatisticsList[_k] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsFaqStatisticsList", version: "19.15.0" };
function dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToTerraform = dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToTerraform;
function dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToHclTerraform = dataAwsKendraIndexIndexStatisticsTextDocumentStatisticsToHclTerraform;
class DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // indexed_text_bytes - computed: true, optional: false, required: false
    get indexedTextBytes() {
        return this.getNumberAttribute('indexed_text_bytes');
    }
    // indexed_text_documents_count - computed: true, optional: false, required: false
    get indexedTextDocumentsCount() {
        return this.getNumberAttribute('indexed_text_documents_count');
    }
}
exports.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference = DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference", version: "19.15.0" };
class DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList = DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList[_m] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList", version: "19.15.0" };
function dataAwsKendraIndexIndexStatisticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexIndexStatisticsToTerraform = dataAwsKendraIndexIndexStatisticsToTerraform;
function dataAwsKendraIndexIndexStatisticsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexIndexStatisticsToHclTerraform = dataAwsKendraIndexIndexStatisticsToHclTerraform;
class DataAwsKendraIndexIndexStatisticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // faq_statistics - computed: true, optional: false, required: false
        this._faqStatistics = new DataAwsKendraIndexIndexStatisticsFaqStatisticsList(this, "faq_statistics", false);
        // text_document_statistics - computed: true, optional: false, required: false
        this._textDocumentStatistics = new DataAwsKendraIndexIndexStatisticsTextDocumentStatisticsList(this, "text_document_statistics", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get faqStatistics() {
        return this._faqStatistics;
    }
    get textDocumentStatistics() {
        return this._textDocumentStatistics;
    }
}
exports.DataAwsKendraIndexIndexStatisticsOutputReference = DataAwsKendraIndexIndexStatisticsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsOutputReference", version: "19.15.0" };
class DataAwsKendraIndexIndexStatisticsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexIndexStatisticsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexIndexStatisticsList = DataAwsKendraIndexIndexStatisticsList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexIndexStatisticsList[_p] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexIndexStatisticsList", version: "19.15.0" };
function dataAwsKendraIndexServerSideEncryptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexServerSideEncryptionConfigurationToTerraform = dataAwsKendraIndexServerSideEncryptionConfigurationToTerraform;
function dataAwsKendraIndexServerSideEncryptionConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexServerSideEncryptionConfigurationToHclTerraform = dataAwsKendraIndexServerSideEncryptionConfigurationToHclTerraform;
class DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
}
exports.DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference = DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference", version: "19.15.0" };
class DataAwsKendraIndexServerSideEncryptionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexServerSideEncryptionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexServerSideEncryptionConfigurationList = DataAwsKendraIndexServerSideEncryptionConfigurationList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexServerSideEncryptionConfigurationList[_r] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexServerSideEncryptionConfigurationList", version: "19.15.0" };
function dataAwsKendraIndexUserGroupResolutionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexUserGroupResolutionConfigurationToTerraform = dataAwsKendraIndexUserGroupResolutionConfigurationToTerraform;
function dataAwsKendraIndexUserGroupResolutionConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexUserGroupResolutionConfigurationToHclTerraform = dataAwsKendraIndexUserGroupResolutionConfigurationToHclTerraform;
class DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // user_group_resolution_mode - computed: true, optional: false, required: false
    get userGroupResolutionMode() {
        return this.getStringAttribute('user_group_resolution_mode');
    }
}
exports.DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference = DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference", version: "19.15.0" };
class DataAwsKendraIndexUserGroupResolutionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexUserGroupResolutionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexUserGroupResolutionConfigurationList = DataAwsKendraIndexUserGroupResolutionConfigurationList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserGroupResolutionConfigurationList[_t] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserGroupResolutionConfigurationList", version: "19.15.0" };
function dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToTerraform = dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToTerraform;
function dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToHclTerraform = dataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationToHclTerraform;
class DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // group_attribute_field - computed: true, optional: false, required: false
    get groupAttributeField() {
        return this.getStringAttribute('group_attribute_field');
    }
    // user_name_attribute_field - computed: true, optional: false, required: false
    get userNameAttributeField() {
        return this.getStringAttribute('user_name_attribute_field');
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference = DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference", version: "19.15.0" };
class DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList = DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList;
_v = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList[_v] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList", version: "19.15.0" };
function dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToTerraform = dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToTerraform;
function dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToHclTerraform = dataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationToHclTerraform;
class DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // claim_regex - computed: true, optional: false, required: false
    get claimRegex() {
        return this.getStringAttribute('claim_regex');
    }
    // group_attribute_field - computed: true, optional: false, required: false
    get groupAttributeField() {
        return this.getStringAttribute('group_attribute_field');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // key_location - computed: true, optional: false, required: false
    get keyLocation() {
        return this.getStringAttribute('key_location');
    }
    // secrets_manager_arn - computed: true, optional: false, required: false
    get secretsManagerArn() {
        return this.getStringAttribute('secrets_manager_arn');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    // user_name_attribute_field - computed: true, optional: false, required: false
    get userNameAttributeField() {
        return this.getStringAttribute('user_name_attribute_field');
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference = DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference[_w] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference", version: "19.15.0" };
class DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList = DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList;
_x = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList[_x] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList", version: "19.15.0" };
function dataAwsKendraIndexUserTokenConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKendraIndexUserTokenConfigurationsToTerraform = dataAwsKendraIndexUserTokenConfigurationsToTerraform;
function dataAwsKendraIndexUserTokenConfigurationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsKendraIndexUserTokenConfigurationsToHclTerraform = dataAwsKendraIndexUserTokenConfigurationsToHclTerraform;
class DataAwsKendraIndexUserTokenConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // json_token_type_configuration - computed: true, optional: false, required: false
        this._jsonTokenTypeConfiguration = new DataAwsKendraIndexUserTokenConfigurationsJsonTokenTypeConfigurationList(this, "json_token_type_configuration", false);
        // jwt_token_type_configuration - computed: true, optional: false, required: false
        this._jwtTokenTypeConfiguration = new DataAwsKendraIndexUserTokenConfigurationsJwtTokenTypeConfigurationList(this, "jwt_token_type_configuration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get jsonTokenTypeConfiguration() {
        return this._jsonTokenTypeConfiguration;
    }
    get jwtTokenTypeConfiguration() {
        return this._jwtTokenTypeConfiguration;
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsOutputReference = DataAwsKendraIndexUserTokenConfigurationsOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsOutputReference[_y] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsOutputReference", version: "19.15.0" };
class DataAwsKendraIndexUserTokenConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKendraIndexUserTokenConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKendraIndexUserTokenConfigurationsList = DataAwsKendraIndexUserTokenConfigurationsList;
_z = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndexUserTokenConfigurationsList[_z] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndexUserTokenConfigurationsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_index aws_kendra_index}
*/
class DataAwsKendraIndex extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsKendraIndex resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsKendraIndex to import
    * @param importFromId The id of the existing DataAwsKendraIndex that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_index#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsKendraIndex to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_kendra_index", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/kendra_index aws_kendra_index} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsKendraIndexConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kendra_index',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // capacity_units - computed: true, optional: false, required: false
        this._capacityUnits = new DataAwsKendraIndexCapacityUnitsList(this, "capacity_units", false);
        // document_metadata_configuration_updates - computed: true, optional: false, required: false
        this._documentMetadataConfigurationUpdates = new DataAwsKendraIndexDocumentMetadataConfigurationUpdatesList(this, "document_metadata_configuration_updates", true);
        // index_statistics - computed: true, optional: false, required: false
        this._indexStatistics = new DataAwsKendraIndexIndexStatisticsList(this, "index_statistics", false);
        // server_side_encryption_configuration - computed: true, optional: false, required: false
        this._serverSideEncryptionConfiguration = new DataAwsKendraIndexServerSideEncryptionConfigurationList(this, "server_side_encryption_configuration", false);
        // user_group_resolution_configuration - computed: true, optional: false, required: false
        this._userGroupResolutionConfiguration = new DataAwsKendraIndexUserGroupResolutionConfigurationList(this, "user_group_resolution_configuration", false);
        // user_token_configurations - computed: true, optional: false, required: false
        this._userTokenConfigurations = new DataAwsKendraIndexUserTokenConfigurationsList(this, "user_token_configurations", false);
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get capacityUnits() {
        return this._capacityUnits;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get documentMetadataConfigurationUpdates() {
        return this._documentMetadataConfigurationUpdates;
    }
    // edition - computed: true, optional: false, required: false
    get edition() {
        return this.getStringAttribute('edition');
    }
    // error_message - computed: true, optional: false, required: false
    get errorMessage() {
        return this.getStringAttribute('error_message');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get indexStatistics() {
        return this._indexStatistics;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    get serverSideEncryptionConfiguration() {
        return this._serverSideEncryptionConfiguration;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // updated_at - computed: true, optional: false, required: false
    get updatedAt() {
        return this.getStringAttribute('updated_at');
    }
    // user_context_policy - computed: true, optional: false, required: false
    get userContextPolicy() {
        return this.getStringAttribute('user_context_policy');
    }
    get userGroupResolutionConfiguration() {
        return this._userGroupResolutionConfiguration;
    }
    get userTokenConfigurations() {
        return this._userTokenConfigurations;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsKendraIndex = DataAwsKendraIndex;
_0 = JSII_RTTI_SYMBOL_1;
DataAwsKendraIndex[_0] = { fqn: "@cdktf/provider-aws.dataAwsKendraIndex.DataAwsKendraIndex", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsKendraIndex.tfResourceType = "aws_kendra_index";
