// @cdktf/provider-aws/lib/vpc-ipam-preview-next-cidr/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcIpamPreviewNextCidr = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_preview_next_cidr aws_vpc_ipam_preview_next_cidr}
*/
class VpcIpamPreviewNextCidr extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VpcIpamPreviewNextCidr resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VpcIpamPreviewNextCidr to import
    * @param importFromId The id of the existing VpcIpamPreviewNextCidr that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_preview_next_cidr#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VpcIpamPreviewNextCidr to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpc_ipam_preview_next_cidr", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpc_ipam_preview_next_cidr aws_vpc_ipam_preview_next_cidr} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcIpamPreviewNextCidrConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam_preview_next_cidr',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._disallowedCidrs = config.disallowedCidrs;
        this._id = config.id;
        this._ipamPoolId = config.ipamPoolId;
        this._netmaskLength = config.netmaskLength;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // cidr - computed: true, optional: false, required: false
    get cidr() {
        return this.getStringAttribute('cidr');
    }
    get disallowedCidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('disallowed_cidrs'));
    }
    set disallowedCidrs(value) {
        this._disallowedCidrs = value;
    }
    resetDisallowedCidrs() {
        this._disallowedCidrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disallowedCidrsInput() {
        return this._disallowedCidrs;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipamPoolId() {
        return this.getStringAttribute('ipam_pool_id');
    }
    set ipamPoolId(value) {
        this._ipamPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get ipamPoolIdInput() {
        return this._ipamPoolId;
    }
    get netmaskLength() {
        return this.getNumberAttribute('netmask_length');
    }
    set netmaskLength(value) {
        this._netmaskLength = value;
    }
    resetNetmaskLength() {
        this._netmaskLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get netmaskLengthInput() {
        return this._netmaskLength;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            disallowed_cidrs: cdktf.listMapper(cdktf.stringToTerraform, false)(this._disallowedCidrs),
            id: cdktf.stringToTerraform(this._id),
            ipam_pool_id: cdktf.stringToTerraform(this._ipamPoolId),
            netmask_length: cdktf.numberToTerraform(this._netmaskLength),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            disallowed_cidrs: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._disallowedCidrs),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ipam_pool_id: {
                value: cdktf.stringToHclTerraform(this._ipamPoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            netmask_length: {
                value: cdktf.numberToHclTerraform(this._netmaskLength),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VpcIpamPreviewNextCidr = VpcIpamPreviewNextCidr;
_a = JSII_RTTI_SYMBOL_1;
VpcIpamPreviewNextCidr[_a] = { fqn: "@cdktf/provider-aws.vpcIpamPreviewNextCidr.VpcIpamPreviewNextCidr", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VpcIpamPreviewNextCidr.tfResourceType = "aws_vpc_ipam_preview_next_cidr";
