// @cdktf/provider-aws/lib/data-aws-ssoadmin-application-providers/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsoadminApplicationProviders = exports.DataAwsSsoadminApplicationProvidersApplicationProvidersList = exports.DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference = exports.dataAwsSsoadminApplicationProvidersApplicationProvidersToHclTerraform = exports.dataAwsSsoadminApplicationProvidersApplicationProvidersToTerraform = exports.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList = exports.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference = exports.dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToHclTerraform = exports.dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToTerraform = dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToTerraform;
function dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToHclTerraform = dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToHclTerraform;
class DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // icon_url - computed: true, optional: false, required: false
    get iconUrl() {
        return this.getStringAttribute('icon_url');
    }
}
exports.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference = DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationProviders.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference", version: "19.15.0" };
class DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList = DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationProviders.DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList", version: "19.15.0" };
function dataAwsSsoadminApplicationProvidersApplicationProvidersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        display_data: cdktf.listMapper(dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToTerraform, true)(struct.displayData),
    };
}
exports.dataAwsSsoadminApplicationProvidersApplicationProvidersToTerraform = dataAwsSsoadminApplicationProvidersApplicationProvidersToTerraform;
function dataAwsSsoadminApplicationProvidersApplicationProvidersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        display_data: {
            value: cdktf.listMapperHcl(dataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataToHclTerraform, true)(struct.displayData),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsSsoadminApplicationProvidersApplicationProvidersToHclTerraform = dataAwsSsoadminApplicationProvidersApplicationProvidersToHclTerraform;
class DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // display_data - computed: false, optional: true, required: false
        this._displayData = new DataAwsSsoadminApplicationProvidersApplicationProvidersDisplayDataList(this, "display_data", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._displayData?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayData = this._displayData?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._displayData.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._displayData.internalValue = value.displayData;
        }
    }
    // application_provider_arn - computed: true, optional: false, required: false
    get applicationProviderArn() {
        return this.getStringAttribute('application_provider_arn');
    }
    // federation_protocol - computed: true, optional: false, required: false
    get federationProtocol() {
        return this.getStringAttribute('federation_protocol');
    }
    get displayData() {
        return this._displayData;
    }
    putDisplayData(value) {
        this._displayData.internalValue = value;
    }
    resetDisplayData() {
        this._displayData.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayDataInput() {
        return this._displayData.internalValue;
    }
}
exports.DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference = DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationProviders.DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference", version: "19.15.0" };
class DataAwsSsoadminApplicationProvidersApplicationProvidersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSsoadminApplicationProvidersApplicationProvidersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSsoadminApplicationProvidersApplicationProvidersList = DataAwsSsoadminApplicationProvidersApplicationProvidersList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationProvidersApplicationProvidersList[_d] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationProviders.DataAwsSsoadminApplicationProvidersApplicationProvidersList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_providers aws_ssoadmin_application_providers}
*/
class DataAwsSsoadminApplicationProviders extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsSsoadminApplicationProviders resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsSsoadminApplicationProviders to import
    * @param importFromId The id of the existing DataAwsSsoadminApplicationProviders that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_providers#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsSsoadminApplicationProviders to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ssoadmin_application_providers", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ssoadmin_application_providers aws_ssoadmin_application_providers} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSsoadminApplicationProvidersConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_application_providers',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // application_providers - computed: false, optional: true, required: false
        this._applicationProviders = new DataAwsSsoadminApplicationProvidersApplicationProvidersList(this, "application_providers", false);
        this._applicationProviders.internalValue = config.applicationProviders;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get applicationProviders() {
        return this._applicationProviders;
    }
    putApplicationProviders(value) {
        this._applicationProviders.internalValue = value;
    }
    resetApplicationProviders() {
        this._applicationProviders.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationProvidersInput() {
        return this._applicationProviders.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_providers: cdktf.listMapper(dataAwsSsoadminApplicationProvidersApplicationProvidersToTerraform, true)(this._applicationProviders.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_providers: {
                value: cdktf.listMapperHcl(dataAwsSsoadminApplicationProvidersApplicationProvidersToHclTerraform, true)(this._applicationProviders.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsSsoadminApplicationProvidersApplicationProvidersList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsSsoadminApplicationProviders = DataAwsSsoadminApplicationProviders;
_e = JSII_RTTI_SYMBOL_1;
DataAwsSsoadminApplicationProviders[_e] = { fqn: "@cdktf/provider-aws.dataAwsSsoadminApplicationProviders.DataAwsSsoadminApplicationProviders", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSsoadminApplicationProviders.tfResourceType = "aws_ssoadmin_application_providers";
