// @cdktf/provider-aws/lib/cloudwatch-event-target/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventTargetSagemakerPipelineTargetOutputReference = exports.cloudwatchEventTargetSagemakerPipelineTargetToHclTerraform = exports.cloudwatchEventTargetSagemakerPipelineTargetToTerraform = exports.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList = exports.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference = exports.cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToHclTerraform = exports.cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToTerraform = exports.CloudwatchEventTargetRunCommandTargetsList = exports.CloudwatchEventTargetRunCommandTargetsOutputReference = exports.cloudwatchEventTargetRunCommandTargetsToHclTerraform = exports.cloudwatchEventTargetRunCommandTargetsToTerraform = exports.CloudwatchEventTargetRetryPolicyOutputReference = exports.cloudwatchEventTargetRetryPolicyToHclTerraform = exports.cloudwatchEventTargetRetryPolicyToTerraform = exports.CloudwatchEventTargetRedshiftTargetOutputReference = exports.cloudwatchEventTargetRedshiftTargetToHclTerraform = exports.cloudwatchEventTargetRedshiftTargetToTerraform = exports.CloudwatchEventTargetKinesisTargetOutputReference = exports.cloudwatchEventTargetKinesisTargetToHclTerraform = exports.cloudwatchEventTargetKinesisTargetToTerraform = exports.CloudwatchEventTargetInputTransformerOutputReference = exports.cloudwatchEventTargetInputTransformerToHclTerraform = exports.cloudwatchEventTargetInputTransformerToTerraform = exports.CloudwatchEventTargetHttpTargetOutputReference = exports.cloudwatchEventTargetHttpTargetToHclTerraform = exports.cloudwatchEventTargetHttpTargetToTerraform = exports.CloudwatchEventTargetEcsTargetOutputReference = exports.cloudwatchEventTargetEcsTargetToHclTerraform = exports.cloudwatchEventTargetEcsTargetToTerraform = exports.CloudwatchEventTargetEcsTargetPlacementConstraintList = exports.CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference = exports.cloudwatchEventTargetEcsTargetPlacementConstraintToHclTerraform = exports.cloudwatchEventTargetEcsTargetPlacementConstraintToTerraform = exports.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList = exports.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference = exports.cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToHclTerraform = exports.cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToTerraform = exports.CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference = exports.cloudwatchEventTargetEcsTargetNetworkConfigurationToHclTerraform = exports.cloudwatchEventTargetEcsTargetNetworkConfigurationToTerraform = exports.CloudwatchEventTargetEcsTargetCapacityProviderStrategyList = exports.CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference = exports.cloudwatchEventTargetEcsTargetCapacityProviderStrategyToHclTerraform = exports.cloudwatchEventTargetEcsTargetCapacityProviderStrategyToTerraform = exports.CloudwatchEventTargetDeadLetterConfigOutputReference = exports.cloudwatchEventTargetDeadLetterConfigToHclTerraform = exports.cloudwatchEventTargetDeadLetterConfigToTerraform = exports.CloudwatchEventTargetBatchTargetOutputReference = exports.cloudwatchEventTargetBatchTargetToHclTerraform = exports.cloudwatchEventTargetBatchTargetToTerraform = void 0;
exports.CloudwatchEventTarget = exports.CloudwatchEventTargetSqsTargetOutputReference = exports.cloudwatchEventTargetSqsTargetToHclTerraform = exports.cloudwatchEventTargetSqsTargetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudwatchEventTargetBatchTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        array_size: cdktf.numberToTerraform(struct.arraySize),
        job_attempts: cdktf.numberToTerraform(struct.jobAttempts),
        job_definition: cdktf.stringToTerraform(struct.jobDefinition),
        job_name: cdktf.stringToTerraform(struct.jobName),
    };
}
exports.cloudwatchEventTargetBatchTargetToTerraform = cloudwatchEventTargetBatchTargetToTerraform;
function cloudwatchEventTargetBatchTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        array_size: {
            value: cdktf.numberToHclTerraform(struct.arraySize),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        job_attempts: {
            value: cdktf.numberToHclTerraform(struct.jobAttempts),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        job_definition: {
            value: cdktf.stringToHclTerraform(struct.jobDefinition),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        job_name: {
            value: cdktf.stringToHclTerraform(struct.jobName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetBatchTargetToHclTerraform = cloudwatchEventTargetBatchTargetToHclTerraform;
class CloudwatchEventTargetBatchTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arraySize !== undefined) {
            hasAnyValues = true;
            internalValueResult.arraySize = this._arraySize;
        }
        if (this._jobAttempts !== undefined) {
            hasAnyValues = true;
            internalValueResult.jobAttempts = this._jobAttempts;
        }
        if (this._jobDefinition !== undefined) {
            hasAnyValues = true;
            internalValueResult.jobDefinition = this._jobDefinition;
        }
        if (this._jobName !== undefined) {
            hasAnyValues = true;
            internalValueResult.jobName = this._jobName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arraySize = undefined;
            this._jobAttempts = undefined;
            this._jobDefinition = undefined;
            this._jobName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arraySize = value.arraySize;
            this._jobAttempts = value.jobAttempts;
            this._jobDefinition = value.jobDefinition;
            this._jobName = value.jobName;
        }
    }
    get arraySize() {
        return this.getNumberAttribute('array_size');
    }
    set arraySize(value) {
        this._arraySize = value;
    }
    resetArraySize() {
        this._arraySize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arraySizeInput() {
        return this._arraySize;
    }
    get jobAttempts() {
        return this.getNumberAttribute('job_attempts');
    }
    set jobAttempts(value) {
        this._jobAttempts = value;
    }
    resetJobAttempts() {
        this._jobAttempts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jobAttemptsInput() {
        return this._jobAttempts;
    }
    get jobDefinition() {
        return this.getStringAttribute('job_definition');
    }
    set jobDefinition(value) {
        this._jobDefinition = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobDefinitionInput() {
        return this._jobDefinition;
    }
    get jobName() {
        return this.getStringAttribute('job_name');
    }
    set jobName(value) {
        this._jobName = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobNameInput() {
        return this._jobName;
    }
}
exports.CloudwatchEventTargetBatchTargetOutputReference = CloudwatchEventTargetBatchTargetOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetBatchTargetOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetBatchTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetDeadLetterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
    };
}
exports.cloudwatchEventTargetDeadLetterConfigToTerraform = cloudwatchEventTargetDeadLetterConfigToTerraform;
function cloudwatchEventTargetDeadLetterConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        arn: {
            value: cdktf.stringToHclTerraform(struct.arn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetDeadLetterConfigToHclTerraform = cloudwatchEventTargetDeadLetterConfigToHclTerraform;
class CloudwatchEventTargetDeadLetterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
}
exports.CloudwatchEventTargetDeadLetterConfigOutputReference = CloudwatchEventTargetDeadLetterConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetDeadLetterConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetDeadLetterConfigOutputReference", version: "19.15.0" };
function cloudwatchEventTargetEcsTargetCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.cloudwatchEventTargetEcsTargetCapacityProviderStrategyToTerraform = cloudwatchEventTargetEcsTargetCapacityProviderStrategyToTerraform;
function cloudwatchEventTargetEcsTargetCapacityProviderStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        base: {
            value: cdktf.numberToHclTerraform(struct.base),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        capacity_provider: {
            value: cdktf.stringToHclTerraform(struct.capacityProvider),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weight: {
            value: cdktf.numberToHclTerraform(struct.weight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetEcsTargetCapacityProviderStrategyToHclTerraform = cloudwatchEventTargetEcsTargetCapacityProviderStrategyToHclTerraform;
class CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._base !== undefined) {
            hasAnyValues = true;
            internalValueResult.base = this._base;
        }
        if (this._capacityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProvider = this._capacityProvider;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._base = undefined;
            this._capacityProvider = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._base = value.base;
            this._capacityProvider = value.capacityProvider;
            this._weight = value.weight;
        }
    }
    get base() {
        return this.getNumberAttribute('base');
    }
    set base(value) {
        this._base = value;
    }
    resetBase() {
        this._base = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseInput() {
        return this._base;
    }
    get capacityProvider() {
        return this.getStringAttribute('capacity_provider');
    }
    set capacityProvider(value) {
        this._capacityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderInput() {
        return this._capacityProvider;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference = CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference", version: "19.15.0" };
class CloudwatchEventTargetEcsTargetCapacityProviderStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventTargetEcsTargetCapacityProviderStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventTargetEcsTargetCapacityProviderStrategyList = CloudwatchEventTargetEcsTargetCapacityProviderStrategyList;
_d = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetCapacityProviderStrategyList[_d] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetCapacityProviderStrategyList", version: "19.15.0" };
function cloudwatchEventTargetEcsTargetNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.cloudwatchEventTargetEcsTargetNetworkConfigurationToTerraform = cloudwatchEventTargetEcsTargetNetworkConfigurationToTerraform;
function cloudwatchEventTargetEcsTargetNetworkConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        assign_public_ip: {
            value: cdktf.booleanToHclTerraform(struct.assignPublicIp),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        security_groups: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnets: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnets),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetEcsTargetNetworkConfigurationToHclTerraform = cloudwatchEventTargetEcsTargetNetworkConfigurationToHclTerraform;
class CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference = CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference", version: "19.15.0" };
function cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToTerraform = cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToTerraform;
function cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToHclTerraform = cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToHclTerraform;
class CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._field = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._field = value.field;
            this._type = value.type;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference = CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference[_f] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference", version: "19.15.0" };
class CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventTargetEcsTargetOrderedPlacementStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList = CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList;
_g = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList[_g] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList", version: "19.15.0" };
function cloudwatchEventTargetEcsTargetPlacementConstraintToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.cloudwatchEventTargetEcsTargetPlacementConstraintToTerraform = cloudwatchEventTargetEcsTargetPlacementConstraintToTerraform;
function cloudwatchEventTargetEcsTargetPlacementConstraintToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetEcsTargetPlacementConstraintToHclTerraform = cloudwatchEventTargetEcsTargetPlacementConstraintToHclTerraform;
class CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expression = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expression = value.expression;
            this._type = value.type;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    resetExpression() {
        this._expression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference = CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference[_h] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference", version: "19.15.0" };
class CloudwatchEventTargetEcsTargetPlacementConstraintList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventTargetEcsTargetPlacementConstraintOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventTargetEcsTargetPlacementConstraintList = CloudwatchEventTargetEcsTargetPlacementConstraintList;
_j = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetPlacementConstraintList[_j] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetPlacementConstraintList", version: "19.15.0" };
function cloudwatchEventTargetEcsTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_ecs_managed_tags: cdktf.booleanToTerraform(struct.enableEcsManagedTags),
        enable_execute_command: cdktf.booleanToTerraform(struct.enableExecuteCommand),
        group: cdktf.stringToTerraform(struct.group),
        launch_type: cdktf.stringToTerraform(struct.launchType),
        platform_version: cdktf.stringToTerraform(struct.platformVersion),
        propagate_tags: cdktf.stringToTerraform(struct.propagateTags),
        tags: cdktf.hashMapper(cdktf.stringToTerraform)(struct.tags),
        task_count: cdktf.numberToTerraform(struct.taskCount),
        task_definition_arn: cdktf.stringToTerraform(struct.taskDefinitionArn),
        capacity_provider_strategy: cdktf.listMapper(cloudwatchEventTargetEcsTargetCapacityProviderStrategyToTerraform, true)(struct.capacityProviderStrategy),
        network_configuration: cloudwatchEventTargetEcsTargetNetworkConfigurationToTerraform(struct.networkConfiguration),
        ordered_placement_strategy: cdktf.listMapper(cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToTerraform, true)(struct.orderedPlacementStrategy),
        placement_constraint: cdktf.listMapper(cloudwatchEventTargetEcsTargetPlacementConstraintToTerraform, true)(struct.placementConstraint),
    };
}
exports.cloudwatchEventTargetEcsTargetToTerraform = cloudwatchEventTargetEcsTargetToTerraform;
function cloudwatchEventTargetEcsTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_ecs_managed_tags: {
            value: cdktf.booleanToHclTerraform(struct.enableEcsManagedTags),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_execute_command: {
            value: cdktf.booleanToHclTerraform(struct.enableExecuteCommand),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        group: {
            value: cdktf.stringToHclTerraform(struct.group),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        launch_type: {
            value: cdktf.stringToHclTerraform(struct.launchType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        platform_version: {
            value: cdktf.stringToHclTerraform(struct.platformVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        propagate_tags: {
            value: cdktf.stringToHclTerraform(struct.propagateTags),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tags: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.tags),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        task_count: {
            value: cdktf.numberToHclTerraform(struct.taskCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        task_definition_arn: {
            value: cdktf.stringToHclTerraform(struct.taskDefinitionArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        capacity_provider_strategy: {
            value: cdktf.listMapperHcl(cloudwatchEventTargetEcsTargetCapacityProviderStrategyToHclTerraform, true)(struct.capacityProviderStrategy),
            isBlock: true,
            type: "set",
            storageClassType: "CloudwatchEventTargetEcsTargetCapacityProviderStrategyList",
        },
        network_configuration: {
            value: cloudwatchEventTargetEcsTargetNetworkConfigurationToHclTerraform(struct.networkConfiguration),
            isBlock: true,
            type: "list",
            storageClassType: "CloudwatchEventTargetEcsTargetNetworkConfigurationList",
        },
        ordered_placement_strategy: {
            value: cdktf.listMapperHcl(cloudwatchEventTargetEcsTargetOrderedPlacementStrategyToHclTerraform, true)(struct.orderedPlacementStrategy),
            isBlock: true,
            type: "list",
            storageClassType: "CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList",
        },
        placement_constraint: {
            value: cdktf.listMapperHcl(cloudwatchEventTargetEcsTargetPlacementConstraintToHclTerraform, true)(struct.placementConstraint),
            isBlock: true,
            type: "set",
            storageClassType: "CloudwatchEventTargetEcsTargetPlacementConstraintList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetEcsTargetToHclTerraform = cloudwatchEventTargetEcsTargetToHclTerraform;
class CloudwatchEventTargetEcsTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // capacity_provider_strategy - computed: false, optional: true, required: false
        this._capacityProviderStrategy = new CloudwatchEventTargetEcsTargetCapacityProviderStrategyList(this, "capacity_provider_strategy", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new CloudwatchEventTargetEcsTargetNetworkConfigurationOutputReference(this, "network_configuration");
        // ordered_placement_strategy - computed: false, optional: true, required: false
        this._orderedPlacementStrategy = new CloudwatchEventTargetEcsTargetOrderedPlacementStrategyList(this, "ordered_placement_strategy", false);
        // placement_constraint - computed: false, optional: true, required: false
        this._placementConstraint = new CloudwatchEventTargetEcsTargetPlacementConstraintList(this, "placement_constraint", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableEcsManagedTags !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableEcsManagedTags = this._enableEcsManagedTags;
        }
        if (this._enableExecuteCommand !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableExecuteCommand = this._enableExecuteCommand;
        }
        if (this._group !== undefined) {
            hasAnyValues = true;
            internalValueResult.group = this._group;
        }
        if (this._launchType !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchType = this._launchType;
        }
        if (this._platformVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.platformVersion = this._platformVersion;
        }
        if (this._propagateTags !== undefined) {
            hasAnyValues = true;
            internalValueResult.propagateTags = this._propagateTags;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._taskCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskCount = this._taskCount;
        }
        if (this._taskDefinitionArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskDefinitionArn = this._taskDefinitionArn;
        }
        if (this._capacityProviderStrategy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProviderStrategy = this._capacityProviderStrategy?.internalValue;
        }
        if (this._networkConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkConfiguration = this._networkConfiguration?.internalValue;
        }
        if (this._orderedPlacementStrategy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.orderedPlacementStrategy = this._orderedPlacementStrategy?.internalValue;
        }
        if (this._placementConstraint?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.placementConstraint = this._placementConstraint?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableEcsManagedTags = undefined;
            this._enableExecuteCommand = undefined;
            this._group = undefined;
            this._launchType = undefined;
            this._platformVersion = undefined;
            this._propagateTags = undefined;
            this._tags = undefined;
            this._taskCount = undefined;
            this._taskDefinitionArn = undefined;
            this._capacityProviderStrategy.internalValue = undefined;
            this._networkConfiguration.internalValue = undefined;
            this._orderedPlacementStrategy.internalValue = undefined;
            this._placementConstraint.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableEcsManagedTags = value.enableEcsManagedTags;
            this._enableExecuteCommand = value.enableExecuteCommand;
            this._group = value.group;
            this._launchType = value.launchType;
            this._platformVersion = value.platformVersion;
            this._propagateTags = value.propagateTags;
            this._tags = value.tags;
            this._taskCount = value.taskCount;
            this._taskDefinitionArn = value.taskDefinitionArn;
            this._capacityProviderStrategy.internalValue = value.capacityProviderStrategy;
            this._networkConfiguration.internalValue = value.networkConfiguration;
            this._orderedPlacementStrategy.internalValue = value.orderedPlacementStrategy;
            this._placementConstraint.internalValue = value.placementConstraint;
        }
    }
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    get group() {
        return this.getStringAttribute('group');
    }
    set group(value) {
        this._group = value;
    }
    resetGroup() {
        this._group = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupInput() {
        return this._group;
    }
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    set launchType(value) {
        this._launchType = value;
    }
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchTypeInput() {
        return this._launchType;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    set propagateTags(value) {
        this._propagateTags = value;
    }
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propagateTagsInput() {
        return this._propagateTags;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get taskCount() {
        return this.getNumberAttribute('task_count');
    }
    set taskCount(value) {
        this._taskCount = value;
    }
    resetTaskCount() {
        this._taskCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskCountInput() {
        return this._taskCount;
    }
    get taskDefinitionArn() {
        return this.getStringAttribute('task_definition_arn');
    }
    set taskDefinitionArn(value) {
        this._taskDefinitionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get taskDefinitionArnInput() {
        return this._taskDefinitionArn;
    }
    get capacityProviderStrategy() {
        return this._capacityProviderStrategy;
    }
    putCapacityProviderStrategy(value) {
        this._capacityProviderStrategy.internalValue = value;
    }
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get orderedPlacementStrategy() {
        return this._orderedPlacementStrategy;
    }
    putOrderedPlacementStrategy(value) {
        this._orderedPlacementStrategy.internalValue = value;
    }
    resetOrderedPlacementStrategy() {
        this._orderedPlacementStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orderedPlacementStrategyInput() {
        return this._orderedPlacementStrategy.internalValue;
    }
    get placementConstraint() {
        return this._placementConstraint;
    }
    putPlacementConstraint(value) {
        this._placementConstraint.internalValue = value;
    }
    resetPlacementConstraint() {
        this._placementConstraint.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementConstraintInput() {
        return this._placementConstraint.internalValue;
    }
}
exports.CloudwatchEventTargetEcsTargetOutputReference = CloudwatchEventTargetEcsTargetOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetEcsTargetOutputReference[_k] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetEcsTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetHttpTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        header_parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.headerParameters),
        path_parameter_values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.pathParameterValues),
        query_string_parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.queryStringParameters),
    };
}
exports.cloudwatchEventTargetHttpTargetToTerraform = cloudwatchEventTargetHttpTargetToTerraform;
function cloudwatchEventTargetHttpTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        header_parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.headerParameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        path_parameter_values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.pathParameterValues),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        query_string_parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.queryStringParameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetHttpTargetToHclTerraform = cloudwatchEventTargetHttpTargetToHclTerraform;
class CloudwatchEventTargetHttpTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headerParameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerParameters = this._headerParameters;
        }
        if (this._pathParameterValues !== undefined) {
            hasAnyValues = true;
            internalValueResult.pathParameterValues = this._pathParameterValues;
        }
        if (this._queryStringParameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.queryStringParameters = this._queryStringParameters;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._headerParameters = undefined;
            this._pathParameterValues = undefined;
            this._queryStringParameters = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._headerParameters = value.headerParameters;
            this._pathParameterValues = value.pathParameterValues;
            this._queryStringParameters = value.queryStringParameters;
        }
    }
    get headerParameters() {
        return this.getStringMapAttribute('header_parameters');
    }
    set headerParameters(value) {
        this._headerParameters = value;
    }
    resetHeaderParameters() {
        this._headerParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerParametersInput() {
        return this._headerParameters;
    }
    get pathParameterValues() {
        return this.getListAttribute('path_parameter_values');
    }
    set pathParameterValues(value) {
        this._pathParameterValues = value;
    }
    resetPathParameterValues() {
        this._pathParameterValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathParameterValuesInput() {
        return this._pathParameterValues;
    }
    get queryStringParameters() {
        return this.getStringMapAttribute('query_string_parameters');
    }
    set queryStringParameters(value) {
        this._queryStringParameters = value;
    }
    resetQueryStringParameters() {
        this._queryStringParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryStringParametersInput() {
        return this._queryStringParameters;
    }
}
exports.CloudwatchEventTargetHttpTargetOutputReference = CloudwatchEventTargetHttpTargetOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetHttpTargetOutputReference[_l] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetHttpTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetInputTransformerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_paths: cdktf.hashMapper(cdktf.stringToTerraform)(struct.inputPaths),
        input_template: cdktf.stringToTerraform(struct.inputTemplate),
    };
}
exports.cloudwatchEventTargetInputTransformerToTerraform = cloudwatchEventTargetInputTransformerToTerraform;
function cloudwatchEventTargetInputTransformerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        input_paths: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.inputPaths),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        input_template: {
            value: cdktf.stringToHclTerraform(struct.inputTemplate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetInputTransformerToHclTerraform = cloudwatchEventTargetInputTransformerToHclTerraform;
class CloudwatchEventTargetInputTransformerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputPaths !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputPaths = this._inputPaths;
        }
        if (this._inputTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputTemplate = this._inputTemplate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._inputPaths = undefined;
            this._inputTemplate = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._inputPaths = value.inputPaths;
            this._inputTemplate = value.inputTemplate;
        }
    }
    get inputPaths() {
        return this.getStringMapAttribute('input_paths');
    }
    set inputPaths(value) {
        this._inputPaths = value;
    }
    resetInputPaths() {
        this._inputPaths = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputPathsInput() {
        return this._inputPaths;
    }
    get inputTemplate() {
        return this.getStringAttribute('input_template');
    }
    set inputTemplate(value) {
        this._inputTemplate = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputTemplateInput() {
        return this._inputTemplate;
    }
}
exports.CloudwatchEventTargetInputTransformerOutputReference = CloudwatchEventTargetInputTransformerOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetInputTransformerOutputReference[_m] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetInputTransformerOutputReference", version: "19.15.0" };
function cloudwatchEventTargetKinesisTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partition_key_path: cdktf.stringToTerraform(struct.partitionKeyPath),
    };
}
exports.cloudwatchEventTargetKinesisTargetToTerraform = cloudwatchEventTargetKinesisTargetToTerraform;
function cloudwatchEventTargetKinesisTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        partition_key_path: {
            value: cdktf.stringToHclTerraform(struct.partitionKeyPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetKinesisTargetToHclTerraform = cloudwatchEventTargetKinesisTargetToHclTerraform;
class CloudwatchEventTargetKinesisTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._partitionKeyPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.partitionKeyPath = this._partitionKeyPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._partitionKeyPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._partitionKeyPath = value.partitionKeyPath;
        }
    }
    get partitionKeyPath() {
        return this.getStringAttribute('partition_key_path');
    }
    set partitionKeyPath(value) {
        this._partitionKeyPath = value;
    }
    resetPartitionKeyPath() {
        this._partitionKeyPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get partitionKeyPathInput() {
        return this._partitionKeyPath;
    }
}
exports.CloudwatchEventTargetKinesisTargetOutputReference = CloudwatchEventTargetKinesisTargetOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetKinesisTargetOutputReference[_o] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetKinesisTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetRedshiftTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database: cdktf.stringToTerraform(struct.database),
        db_user: cdktf.stringToTerraform(struct.dbUser),
        secrets_manager_arn: cdktf.stringToTerraform(struct.secretsManagerArn),
        sql: cdktf.stringToTerraform(struct.sql),
        statement_name: cdktf.stringToTerraform(struct.statementName),
        with_event: cdktf.booleanToTerraform(struct.withEvent),
    };
}
exports.cloudwatchEventTargetRedshiftTargetToTerraform = cloudwatchEventTargetRedshiftTargetToTerraform;
function cloudwatchEventTargetRedshiftTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        database: {
            value: cdktf.stringToHclTerraform(struct.database),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        db_user: {
            value: cdktf.stringToHclTerraform(struct.dbUser),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        secrets_manager_arn: {
            value: cdktf.stringToHclTerraform(struct.secretsManagerArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sql: {
            value: cdktf.stringToHclTerraform(struct.sql),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        statement_name: {
            value: cdktf.stringToHclTerraform(struct.statementName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        with_event: {
            value: cdktf.booleanToHclTerraform(struct.withEvent),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetRedshiftTargetToHclTerraform = cloudwatchEventTargetRedshiftTargetToHclTerraform;
class CloudwatchEventTargetRedshiftTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._database !== undefined) {
            hasAnyValues = true;
            internalValueResult.database = this._database;
        }
        if (this._dbUser !== undefined) {
            hasAnyValues = true;
            internalValueResult.dbUser = this._dbUser;
        }
        if (this._secretsManagerArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretsManagerArn = this._secretsManagerArn;
        }
        if (this._sql !== undefined) {
            hasAnyValues = true;
            internalValueResult.sql = this._sql;
        }
        if (this._statementName !== undefined) {
            hasAnyValues = true;
            internalValueResult.statementName = this._statementName;
        }
        if (this._withEvent !== undefined) {
            hasAnyValues = true;
            internalValueResult.withEvent = this._withEvent;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._database = undefined;
            this._dbUser = undefined;
            this._secretsManagerArn = undefined;
            this._sql = undefined;
            this._statementName = undefined;
            this._withEvent = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._database = value.database;
            this._dbUser = value.dbUser;
            this._secretsManagerArn = value.secretsManagerArn;
            this._sql = value.sql;
            this._statementName = value.statementName;
            this._withEvent = value.withEvent;
        }
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    set dbUser(value) {
        this._dbUser = value;
    }
    resetDbUser() {
        this._dbUser = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbUserInput() {
        return this._dbUser;
    }
    get secretsManagerArn() {
        return this.getStringAttribute('secrets_manager_arn');
    }
    set secretsManagerArn(value) {
        this._secretsManagerArn = value;
    }
    resetSecretsManagerArn() {
        this._secretsManagerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretsManagerArnInput() {
        return this._secretsManagerArn;
    }
    get sql() {
        return this.getStringAttribute('sql');
    }
    set sql(value) {
        this._sql = value;
    }
    resetSql() {
        this._sql = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlInput() {
        return this._sql;
    }
    get statementName() {
        return this.getStringAttribute('statement_name');
    }
    set statementName(value) {
        this._statementName = value;
    }
    resetStatementName() {
        this._statementName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statementNameInput() {
        return this._statementName;
    }
    get withEvent() {
        return this.getBooleanAttribute('with_event');
    }
    set withEvent(value) {
        this._withEvent = value;
    }
    resetWithEvent() {
        this._withEvent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get withEventInput() {
        return this._withEvent;
    }
}
exports.CloudwatchEventTargetRedshiftTargetOutputReference = CloudwatchEventTargetRedshiftTargetOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetRedshiftTargetOutputReference[_p] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetRedshiftTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maximum_event_age_in_seconds: cdktf.numberToTerraform(struct.maximumEventAgeInSeconds),
        maximum_retry_attempts: cdktf.numberToTerraform(struct.maximumRetryAttempts),
    };
}
exports.cloudwatchEventTargetRetryPolicyToTerraform = cloudwatchEventTargetRetryPolicyToTerraform;
function cloudwatchEventTargetRetryPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maximum_event_age_in_seconds: {
            value: cdktf.numberToHclTerraform(struct.maximumEventAgeInSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        maximum_retry_attempts: {
            value: cdktf.numberToHclTerraform(struct.maximumRetryAttempts),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetRetryPolicyToHclTerraform = cloudwatchEventTargetRetryPolicyToHclTerraform;
class CloudwatchEventTargetRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maximumEventAgeInSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumEventAgeInSeconds = this._maximumEventAgeInSeconds;
        }
        if (this._maximumRetryAttempts !== undefined) {
            hasAnyValues = true;
            internalValueResult.maximumRetryAttempts = this._maximumRetryAttempts;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maximumEventAgeInSeconds = undefined;
            this._maximumRetryAttempts = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maximumEventAgeInSeconds = value.maximumEventAgeInSeconds;
            this._maximumRetryAttempts = value.maximumRetryAttempts;
        }
    }
    get maximumEventAgeInSeconds() {
        return this.getNumberAttribute('maximum_event_age_in_seconds');
    }
    set maximumEventAgeInSeconds(value) {
        this._maximumEventAgeInSeconds = value;
    }
    resetMaximumEventAgeInSeconds() {
        this._maximumEventAgeInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumEventAgeInSecondsInput() {
        return this._maximumEventAgeInSeconds;
    }
    get maximumRetryAttempts() {
        return this.getNumberAttribute('maximum_retry_attempts');
    }
    set maximumRetryAttempts(value) {
        this._maximumRetryAttempts = value;
    }
    resetMaximumRetryAttempts() {
        this._maximumRetryAttempts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumRetryAttemptsInput() {
        return this._maximumRetryAttempts;
    }
}
exports.CloudwatchEventTargetRetryPolicyOutputReference = CloudwatchEventTargetRetryPolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetRetryPolicyOutputReference[_q] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetRetryPolicyOutputReference", version: "19.15.0" };
function cloudwatchEventTargetRunCommandTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.cloudwatchEventTargetRunCommandTargetsToTerraform = cloudwatchEventTargetRunCommandTargetsToTerraform;
function cloudwatchEventTargetRunCommandTargetsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetRunCommandTargetsToHclTerraform = cloudwatchEventTargetRunCommandTargetsToHclTerraform;
class CloudwatchEventTargetRunCommandTargetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.CloudwatchEventTargetRunCommandTargetsOutputReference = CloudwatchEventTargetRunCommandTargetsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetRunCommandTargetsOutputReference[_r] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetRunCommandTargetsOutputReference", version: "19.15.0" };
class CloudwatchEventTargetRunCommandTargetsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventTargetRunCommandTargetsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventTargetRunCommandTargetsList = CloudwatchEventTargetRunCommandTargetsList;
_s = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetRunCommandTargetsList[_s] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetRunCommandTargetsList", version: "19.15.0" };
function cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToTerraform = cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToTerraform;
function cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToHclTerraform = cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToHclTerraform;
class CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference = CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference;
_t = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference[_t] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference", version: "19.15.0" };
class CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList = CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList;
_u = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList[_u] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList", version: "19.15.0" };
function cloudwatchEventTargetSagemakerPipelineTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pipeline_parameter_list: cdktf.listMapper(cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToTerraform, true)(struct.pipelineParameterList),
    };
}
exports.cloudwatchEventTargetSagemakerPipelineTargetToTerraform = cloudwatchEventTargetSagemakerPipelineTargetToTerraform;
function cloudwatchEventTargetSagemakerPipelineTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pipeline_parameter_list: {
            value: cdktf.listMapperHcl(cloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructToHclTerraform, true)(struct.pipelineParameterList),
            isBlock: true,
            type: "set",
            storageClassType: "CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetSagemakerPipelineTargetToHclTerraform = cloudwatchEventTargetSagemakerPipelineTargetToHclTerraform;
class CloudwatchEventTargetSagemakerPipelineTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // pipeline_parameter_list - computed: false, optional: true, required: false
        this._pipelineParameterList = new CloudwatchEventTargetSagemakerPipelineTargetPipelineParameterListStructList(this, "pipeline_parameter_list", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pipelineParameterList?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.pipelineParameterList = this._pipelineParameterList?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._pipelineParameterList.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._pipelineParameterList.internalValue = value.pipelineParameterList;
        }
    }
    get pipelineParameterList() {
        return this._pipelineParameterList;
    }
    putPipelineParameterList(value) {
        this._pipelineParameterList.internalValue = value;
    }
    resetPipelineParameterList() {
        this._pipelineParameterList.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineParameterListInput() {
        return this._pipelineParameterList.internalValue;
    }
}
exports.CloudwatchEventTargetSagemakerPipelineTargetOutputReference = CloudwatchEventTargetSagemakerPipelineTargetOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetSagemakerPipelineTargetOutputReference[_v] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetSagemakerPipelineTargetOutputReference", version: "19.15.0" };
function cloudwatchEventTargetSqsTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_group_id: cdktf.stringToTerraform(struct.messageGroupId),
    };
}
exports.cloudwatchEventTargetSqsTargetToTerraform = cloudwatchEventTargetSqsTargetToTerraform;
function cloudwatchEventTargetSqsTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        message_group_id: {
            value: cdktf.stringToHclTerraform(struct.messageGroupId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventTargetSqsTargetToHclTerraform = cloudwatchEventTargetSqsTargetToHclTerraform;
class CloudwatchEventTargetSqsTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.messageGroupId = this._messageGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._messageGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._messageGroupId = value.messageGroupId;
        }
    }
    get messageGroupId() {
        return this.getStringAttribute('message_group_id');
    }
    set messageGroupId(value) {
        this._messageGroupId = value;
    }
    resetMessageGroupId() {
        this._messageGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messageGroupIdInput() {
        return this._messageGroupId;
    }
}
exports.CloudwatchEventTargetSqsTargetOutputReference = CloudwatchEventTargetSqsTargetOutputReference;
_w = JSII_RTTI_SYMBOL_1;
CloudwatchEventTargetSqsTargetOutputReference[_w] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTargetSqsTargetOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_target aws_cloudwatch_event_target}
*/
class CloudwatchEventTarget extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudwatchEventTarget resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudwatchEventTarget to import
    * @param importFromId The id of the existing CloudwatchEventTarget that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_target#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudwatchEventTarget to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudwatch_event_target", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_target aws_cloudwatch_event_target} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudwatchEventTargetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_target',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // batch_target - computed: false, optional: true, required: false
        this._batchTarget = new CloudwatchEventTargetBatchTargetOutputReference(this, "batch_target");
        // dead_letter_config - computed: false, optional: true, required: false
        this._deadLetterConfig = new CloudwatchEventTargetDeadLetterConfigOutputReference(this, "dead_letter_config");
        // ecs_target - computed: false, optional: true, required: false
        this._ecsTarget = new CloudwatchEventTargetEcsTargetOutputReference(this, "ecs_target");
        // http_target - computed: false, optional: true, required: false
        this._httpTarget = new CloudwatchEventTargetHttpTargetOutputReference(this, "http_target");
        // input_transformer - computed: false, optional: true, required: false
        this._inputTransformer = new CloudwatchEventTargetInputTransformerOutputReference(this, "input_transformer");
        // kinesis_target - computed: false, optional: true, required: false
        this._kinesisTarget = new CloudwatchEventTargetKinesisTargetOutputReference(this, "kinesis_target");
        // redshift_target - computed: false, optional: true, required: false
        this._redshiftTarget = new CloudwatchEventTargetRedshiftTargetOutputReference(this, "redshift_target");
        // retry_policy - computed: false, optional: true, required: false
        this._retryPolicy = new CloudwatchEventTargetRetryPolicyOutputReference(this, "retry_policy");
        // run_command_targets - computed: false, optional: true, required: false
        this._runCommandTargets = new CloudwatchEventTargetRunCommandTargetsList(this, "run_command_targets", false);
        // sagemaker_pipeline_target - computed: false, optional: true, required: false
        this._sagemakerPipelineTarget = new CloudwatchEventTargetSagemakerPipelineTargetOutputReference(this, "sagemaker_pipeline_target");
        // sqs_target - computed: false, optional: true, required: false
        this._sqsTarget = new CloudwatchEventTargetSqsTargetOutputReference(this, "sqs_target");
        this._arn = config.arn;
        this._eventBusName = config.eventBusName;
        this._id = config.id;
        this._input = config.input;
        this._inputPath = config.inputPath;
        this._roleArn = config.roleArn;
        this._rule = config.rule;
        this._targetId = config.targetId;
        this._batchTarget.internalValue = config.batchTarget;
        this._deadLetterConfig.internalValue = config.deadLetterConfig;
        this._ecsTarget.internalValue = config.ecsTarget;
        this._httpTarget.internalValue = config.httpTarget;
        this._inputTransformer.internalValue = config.inputTransformer;
        this._kinesisTarget.internalValue = config.kinesisTarget;
        this._redshiftTarget.internalValue = config.redshiftTarget;
        this._retryPolicy.internalValue = config.retryPolicy;
        this._runCommandTargets.internalValue = config.runCommandTargets;
        this._sagemakerPipelineTarget.internalValue = config.sagemakerPipelineTarget;
        this._sqsTarget.internalValue = config.sqsTarget;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get eventBusName() {
        return this.getStringAttribute('event_bus_name');
    }
    set eventBusName(value) {
        this._eventBusName = value;
    }
    resetEventBusName() {
        this._eventBusName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventBusNameInput() {
        return this._eventBusName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get input() {
        return this.getStringAttribute('input');
    }
    set input(value) {
        this._input = value;
    }
    resetInput() {
        this._input = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputInput() {
        return this._input;
    }
    get inputPath() {
        return this.getStringAttribute('input_path');
    }
    set inputPath(value) {
        this._inputPath = value;
    }
    resetInputPath() {
        this._inputPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputPathInput() {
        return this._inputPath;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get rule() {
        return this.getStringAttribute('rule');
    }
    set rule(value) {
        this._rule = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule;
    }
    get targetId() {
        return this.getStringAttribute('target_id');
    }
    set targetId(value) {
        this._targetId = value;
    }
    resetTargetId() {
        this._targetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetIdInput() {
        return this._targetId;
    }
    get batchTarget() {
        return this._batchTarget;
    }
    putBatchTarget(value) {
        this._batchTarget.internalValue = value;
    }
    resetBatchTarget() {
        this._batchTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchTargetInput() {
        return this._batchTarget.internalValue;
    }
    get deadLetterConfig() {
        return this._deadLetterConfig;
    }
    putDeadLetterConfig(value) {
        this._deadLetterConfig.internalValue = value;
    }
    resetDeadLetterConfig() {
        this._deadLetterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deadLetterConfigInput() {
        return this._deadLetterConfig.internalValue;
    }
    get ecsTarget() {
        return this._ecsTarget;
    }
    putEcsTarget(value) {
        this._ecsTarget.internalValue = value;
    }
    resetEcsTarget() {
        this._ecsTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ecsTargetInput() {
        return this._ecsTarget.internalValue;
    }
    get httpTarget() {
        return this._httpTarget;
    }
    putHttpTarget(value) {
        this._httpTarget.internalValue = value;
    }
    resetHttpTarget() {
        this._httpTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpTargetInput() {
        return this._httpTarget.internalValue;
    }
    get inputTransformer() {
        return this._inputTransformer;
    }
    putInputTransformer(value) {
        this._inputTransformer.internalValue = value;
    }
    resetInputTransformer() {
        this._inputTransformer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputTransformerInput() {
        return this._inputTransformer.internalValue;
    }
    get kinesisTarget() {
        return this._kinesisTarget;
    }
    putKinesisTarget(value) {
        this._kinesisTarget.internalValue = value;
    }
    resetKinesisTarget() {
        this._kinesisTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisTargetInput() {
        return this._kinesisTarget.internalValue;
    }
    get redshiftTarget() {
        return this._redshiftTarget;
    }
    putRedshiftTarget(value) {
        this._redshiftTarget.internalValue = value;
    }
    resetRedshiftTarget() {
        this._redshiftTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redshiftTargetInput() {
        return this._redshiftTarget.internalValue;
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    putRetryPolicy(value) {
        this._retryPolicy.internalValue = value;
    }
    resetRetryPolicy() {
        this._retryPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retryPolicyInput() {
        return this._retryPolicy.internalValue;
    }
    get runCommandTargets() {
        return this._runCommandTargets;
    }
    putRunCommandTargets(value) {
        this._runCommandTargets.internalValue = value;
    }
    resetRunCommandTargets() {
        this._runCommandTargets.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runCommandTargetsInput() {
        return this._runCommandTargets.internalValue;
    }
    get sagemakerPipelineTarget() {
        return this._sagemakerPipelineTarget;
    }
    putSagemakerPipelineTarget(value) {
        this._sagemakerPipelineTarget.internalValue = value;
    }
    resetSagemakerPipelineTarget() {
        this._sagemakerPipelineTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sagemakerPipelineTargetInput() {
        return this._sagemakerPipelineTarget.internalValue;
    }
    get sqsTarget() {
        return this._sqsTarget;
    }
    putSqsTarget(value) {
        this._sqsTarget.internalValue = value;
    }
    resetSqsTarget() {
        this._sqsTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqsTargetInput() {
        return this._sqsTarget.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            event_bus_name: cdktf.stringToTerraform(this._eventBusName),
            id: cdktf.stringToTerraform(this._id),
            input: cdktf.stringToTerraform(this._input),
            input_path: cdktf.stringToTerraform(this._inputPath),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            rule: cdktf.stringToTerraform(this._rule),
            target_id: cdktf.stringToTerraform(this._targetId),
            batch_target: cloudwatchEventTargetBatchTargetToTerraform(this._batchTarget.internalValue),
            dead_letter_config: cloudwatchEventTargetDeadLetterConfigToTerraform(this._deadLetterConfig.internalValue),
            ecs_target: cloudwatchEventTargetEcsTargetToTerraform(this._ecsTarget.internalValue),
            http_target: cloudwatchEventTargetHttpTargetToTerraform(this._httpTarget.internalValue),
            input_transformer: cloudwatchEventTargetInputTransformerToTerraform(this._inputTransformer.internalValue),
            kinesis_target: cloudwatchEventTargetKinesisTargetToTerraform(this._kinesisTarget.internalValue),
            redshift_target: cloudwatchEventTargetRedshiftTargetToTerraform(this._redshiftTarget.internalValue),
            retry_policy: cloudwatchEventTargetRetryPolicyToTerraform(this._retryPolicy.internalValue),
            run_command_targets: cdktf.listMapper(cloudwatchEventTargetRunCommandTargetsToTerraform, true)(this._runCommandTargets.internalValue),
            sagemaker_pipeline_target: cloudwatchEventTargetSagemakerPipelineTargetToTerraform(this._sagemakerPipelineTarget.internalValue),
            sqs_target: cloudwatchEventTargetSqsTargetToTerraform(this._sqsTarget.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            event_bus_name: {
                value: cdktf.stringToHclTerraform(this._eventBusName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            input: {
                value: cdktf.stringToHclTerraform(this._input),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            input_path: {
                value: cdktf.stringToHclTerraform(this._inputPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rule: {
                value: cdktf.stringToHclTerraform(this._rule),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_id: {
                value: cdktf.stringToHclTerraform(this._targetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            batch_target: {
                value: cloudwatchEventTargetBatchTargetToHclTerraform(this._batchTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetBatchTargetList",
            },
            dead_letter_config: {
                value: cloudwatchEventTargetDeadLetterConfigToHclTerraform(this._deadLetterConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetDeadLetterConfigList",
            },
            ecs_target: {
                value: cloudwatchEventTargetEcsTargetToHclTerraform(this._ecsTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetEcsTargetList",
            },
            http_target: {
                value: cloudwatchEventTargetHttpTargetToHclTerraform(this._httpTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetHttpTargetList",
            },
            input_transformer: {
                value: cloudwatchEventTargetInputTransformerToHclTerraform(this._inputTransformer.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetInputTransformerList",
            },
            kinesis_target: {
                value: cloudwatchEventTargetKinesisTargetToHclTerraform(this._kinesisTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetKinesisTargetList",
            },
            redshift_target: {
                value: cloudwatchEventTargetRedshiftTargetToHclTerraform(this._redshiftTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetRedshiftTargetList",
            },
            retry_policy: {
                value: cloudwatchEventTargetRetryPolicyToHclTerraform(this._retryPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetRetryPolicyList",
            },
            run_command_targets: {
                value: cdktf.listMapperHcl(cloudwatchEventTargetRunCommandTargetsToHclTerraform, true)(this._runCommandTargets.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetRunCommandTargetsList",
            },
            sagemaker_pipeline_target: {
                value: cloudwatchEventTargetSagemakerPipelineTargetToHclTerraform(this._sagemakerPipelineTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetSagemakerPipelineTargetList",
            },
            sqs_target: {
                value: cloudwatchEventTargetSqsTargetToHclTerraform(this._sqsTarget.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventTargetSqsTargetList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudwatchEventTarget = CloudwatchEventTarget;
_x = JSII_RTTI_SYMBOL_1;
CloudwatchEventTarget[_x] = { fqn: "@cdktf/provider-aws.cloudwatchEventTarget.CloudwatchEventTarget", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudwatchEventTarget.tfResourceType = "aws_cloudwatch_event_target";
