// @cdktf/provider-aws/lib/db-proxy-default-target-group/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbProxyDefaultTargetGroup = exports.DbProxyDefaultTargetGroupTimeoutsOutputReference = exports.dbProxyDefaultTargetGroupTimeoutsToHclTerraform = exports.dbProxyDefaultTargetGroupTimeoutsToTerraform = exports.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference = exports.dbProxyDefaultTargetGroupConnectionPoolConfigToHclTerraform = exports.dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_borrow_timeout: cdktf.numberToTerraform(struct.connectionBorrowTimeout),
        init_query: cdktf.stringToTerraform(struct.initQuery),
        max_connections_percent: cdktf.numberToTerraform(struct.maxConnectionsPercent),
        max_idle_connections_percent: cdktf.numberToTerraform(struct.maxIdleConnectionsPercent),
        session_pinning_filters: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.sessionPinningFilters),
    };
}
exports.dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform = dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform;
function dbProxyDefaultTargetGroupConnectionPoolConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        connection_borrow_timeout: {
            value: cdktf.numberToHclTerraform(struct.connectionBorrowTimeout),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        init_query: {
            value: cdktf.stringToHclTerraform(struct.initQuery),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_connections_percent: {
            value: cdktf.numberToHclTerraform(struct.maxConnectionsPercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_idle_connections_percent: {
            value: cdktf.numberToHclTerraform(struct.maxIdleConnectionsPercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        session_pinning_filters: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.sessionPinningFilters),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dbProxyDefaultTargetGroupConnectionPoolConfigToHclTerraform = dbProxyDefaultTargetGroupConnectionPoolConfigToHclTerraform;
class DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionBorrowTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionBorrowTimeout = this._connectionBorrowTimeout;
        }
        if (this._initQuery !== undefined) {
            hasAnyValues = true;
            internalValueResult.initQuery = this._initQuery;
        }
        if (this._maxConnectionsPercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConnectionsPercent = this._maxConnectionsPercent;
        }
        if (this._maxIdleConnectionsPercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxIdleConnectionsPercent = this._maxIdleConnectionsPercent;
        }
        if (this._sessionPinningFilters !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionPinningFilters = this._sessionPinningFilters;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connectionBorrowTimeout = undefined;
            this._initQuery = undefined;
            this._maxConnectionsPercent = undefined;
            this._maxIdleConnectionsPercent = undefined;
            this._sessionPinningFilters = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connectionBorrowTimeout = value.connectionBorrowTimeout;
            this._initQuery = value.initQuery;
            this._maxConnectionsPercent = value.maxConnectionsPercent;
            this._maxIdleConnectionsPercent = value.maxIdleConnectionsPercent;
            this._sessionPinningFilters = value.sessionPinningFilters;
        }
    }
    get connectionBorrowTimeout() {
        return this.getNumberAttribute('connection_borrow_timeout');
    }
    set connectionBorrowTimeout(value) {
        this._connectionBorrowTimeout = value;
    }
    resetConnectionBorrowTimeout() {
        this._connectionBorrowTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionBorrowTimeoutInput() {
        return this._connectionBorrowTimeout;
    }
    get initQuery() {
        return this.getStringAttribute('init_query');
    }
    set initQuery(value) {
        this._initQuery = value;
    }
    resetInitQuery() {
        this._initQuery = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initQueryInput() {
        return this._initQuery;
    }
    get maxConnectionsPercent() {
        return this.getNumberAttribute('max_connections_percent');
    }
    set maxConnectionsPercent(value) {
        this._maxConnectionsPercent = value;
    }
    resetMaxConnectionsPercent() {
        this._maxConnectionsPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConnectionsPercentInput() {
        return this._maxConnectionsPercent;
    }
    get maxIdleConnectionsPercent() {
        return this.getNumberAttribute('max_idle_connections_percent');
    }
    set maxIdleConnectionsPercent(value) {
        this._maxIdleConnectionsPercent = value;
    }
    resetMaxIdleConnectionsPercent() {
        this._maxIdleConnectionsPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxIdleConnectionsPercentInput() {
        return this._maxIdleConnectionsPercent;
    }
    get sessionPinningFilters() {
        return cdktf.Fn.tolist(this.getListAttribute('session_pinning_filters'));
    }
    set sessionPinningFilters(value) {
        this._sessionPinningFilters = value;
    }
    resetSessionPinningFilters() {
        this._sessionPinningFilters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionPinningFiltersInput() {
        return this._sessionPinningFilters;
    }
}
exports.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference = DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dbProxyDefaultTargetGroup.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference", version: "19.15.0" };
function dbProxyDefaultTargetGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dbProxyDefaultTargetGroupTimeoutsToTerraform = dbProxyDefaultTargetGroupTimeoutsToTerraform;
function dbProxyDefaultTargetGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dbProxyDefaultTargetGroupTimeoutsToHclTerraform = dbProxyDefaultTargetGroupTimeoutsToHclTerraform;
class DbProxyDefaultTargetGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DbProxyDefaultTargetGroupTimeoutsOutputReference = DbProxyDefaultTargetGroupTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroupTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.dbProxyDefaultTargetGroup.DbProxyDefaultTargetGroupTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_proxy_default_target_group aws_db_proxy_default_target_group}
*/
class DbProxyDefaultTargetGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DbProxyDefaultTargetGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DbProxyDefaultTargetGroup to import
    * @param importFromId The id of the existing DbProxyDefaultTargetGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_proxy_default_target_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DbProxyDefaultTargetGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_db_proxy_default_target_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/db_proxy_default_target_group aws_db_proxy_default_target_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DbProxyDefaultTargetGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_proxy_default_target_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // connection_pool_config - computed: false, optional: true, required: false
        this._connectionPoolConfig = new DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference(this, "connection_pool_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbProxyDefaultTargetGroupTimeoutsOutputReference(this, "timeouts");
        this._dbProxyName = config.dbProxyName;
        this._id = config.id;
        this._connectionPoolConfig.internalValue = config.connectionPoolConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get dbProxyName() {
        return this.getStringAttribute('db_proxy_name');
    }
    set dbProxyName(value) {
        this._dbProxyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbProxyNameInput() {
        return this._dbProxyName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get connectionPoolConfig() {
        return this._connectionPoolConfig;
    }
    putConnectionPoolConfig(value) {
        this._connectionPoolConfig.internalValue = value;
    }
    resetConnectionPoolConfig() {
        this._connectionPoolConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionPoolConfigInput() {
        return this._connectionPoolConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_proxy_name: cdktf.stringToTerraform(this._dbProxyName),
            id: cdktf.stringToTerraform(this._id),
            connection_pool_config: dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform(this._connectionPoolConfig.internalValue),
            timeouts: dbProxyDefaultTargetGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            db_proxy_name: {
                value: cdktf.stringToHclTerraform(this._dbProxyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            connection_pool_config: {
                value: dbProxyDefaultTargetGroupConnectionPoolConfigToHclTerraform(this._connectionPoolConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DbProxyDefaultTargetGroupConnectionPoolConfigList",
            },
            timeouts: {
                value: dbProxyDefaultTargetGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DbProxyDefaultTargetGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DbProxyDefaultTargetGroup = DbProxyDefaultTargetGroup;
_c = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroup[_c] = { fqn: "@cdktf/provider-aws.dbProxyDefaultTargetGroup.DbProxyDefaultTargetGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DbProxyDefaultTargetGroup.tfResourceType = "aws_db_proxy_default_target_group";
