// @cdktf/provider-aws/lib/transfer-server/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferServer = exports.TransferServerWorkflowDetailsOutputReference = exports.transferServerWorkflowDetailsToHclTerraform = exports.transferServerWorkflowDetailsToTerraform = exports.TransferServerWorkflowDetailsOnUploadOutputReference = exports.transferServerWorkflowDetailsOnUploadToHclTerraform = exports.transferServerWorkflowDetailsOnUploadToTerraform = exports.TransferServerWorkflowDetailsOnPartialUploadOutputReference = exports.transferServerWorkflowDetailsOnPartialUploadToHclTerraform = exports.transferServerWorkflowDetailsOnPartialUploadToTerraform = exports.TransferServerS3StorageOptionsOutputReference = exports.transferServerS3StorageOptionsToHclTerraform = exports.transferServerS3StorageOptionsToTerraform = exports.TransferServerProtocolDetailsOutputReference = exports.transferServerProtocolDetailsToHclTerraform = exports.transferServerProtocolDetailsToTerraform = exports.TransferServerEndpointDetailsOutputReference = exports.transferServerEndpointDetailsToHclTerraform = exports.transferServerEndpointDetailsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function transferServerEndpointDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address_allocation_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.addressAllocationIds),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
        vpc_endpoint_id: cdktf.stringToTerraform(struct.vpcEndpointId),
        vpc_id: cdktf.stringToTerraform(struct.vpcId),
    };
}
exports.transferServerEndpointDetailsToTerraform = transferServerEndpointDetailsToTerraform;
function transferServerEndpointDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        address_allocation_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.addressAllocationIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        vpc_endpoint_id: {
            value: cdktf.stringToHclTerraform(struct.vpcEndpointId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vpc_id: {
            value: cdktf.stringToHclTerraform(struct.vpcId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerEndpointDetailsToHclTerraform = transferServerEndpointDetailsToHclTerraform;
class TransferServerEndpointDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._addressAllocationIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressAllocationIds = this._addressAllocationIds;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        if (this._vpcEndpointId !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcEndpointId = this._vpcEndpointId;
        }
        if (this._vpcId !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcId = this._vpcId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._addressAllocationIds = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
            this._vpcEndpointId = undefined;
            this._vpcId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._addressAllocationIds = value.addressAllocationIds;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
            this._vpcEndpointId = value.vpcEndpointId;
            this._vpcId = value.vpcId;
        }
    }
    get addressAllocationIds() {
        return cdktf.Fn.tolist(this.getListAttribute('address_allocation_ids'));
    }
    set addressAllocationIds(value) {
        this._addressAllocationIds = value;
    }
    resetAddressAllocationIds() {
        this._addressAllocationIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressAllocationIdsInput() {
        return this._addressAllocationIds;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get vpcEndpointId() {
        return this.getStringAttribute('vpc_endpoint_id');
    }
    set vpcEndpointId(value) {
        this._vpcEndpointId = value;
    }
    resetVpcEndpointId() {
        this._vpcEndpointId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointIdInput() {
        return this._vpcEndpointId;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
}
exports.TransferServerEndpointDetailsOutputReference = TransferServerEndpointDetailsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
TransferServerEndpointDetailsOutputReference[_a] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerEndpointDetailsOutputReference", version: "19.15.0" };
function transferServerProtocolDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        as2_transports: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.as2Transports),
        passive_ip: cdktf.stringToTerraform(struct.passiveIp),
        set_stat_option: cdktf.stringToTerraform(struct.setStatOption),
        tls_session_resumption_mode: cdktf.stringToTerraform(struct.tlsSessionResumptionMode),
    };
}
exports.transferServerProtocolDetailsToTerraform = transferServerProtocolDetailsToTerraform;
function transferServerProtocolDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        as2_transports: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.as2Transports),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        passive_ip: {
            value: cdktf.stringToHclTerraform(struct.passiveIp),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        set_stat_option: {
            value: cdktf.stringToHclTerraform(struct.setStatOption),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tls_session_resumption_mode: {
            value: cdktf.stringToHclTerraform(struct.tlsSessionResumptionMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerProtocolDetailsToHclTerraform = transferServerProtocolDetailsToHclTerraform;
class TransferServerProtocolDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._as2Transports !== undefined) {
            hasAnyValues = true;
            internalValueResult.as2Transports = this._as2Transports;
        }
        if (this._passiveIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.passiveIp = this._passiveIp;
        }
        if (this._setStatOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.setStatOption = this._setStatOption;
        }
        if (this._tlsSessionResumptionMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.tlsSessionResumptionMode = this._tlsSessionResumptionMode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._as2Transports = undefined;
            this._passiveIp = undefined;
            this._setStatOption = undefined;
            this._tlsSessionResumptionMode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._as2Transports = value.as2Transports;
            this._passiveIp = value.passiveIp;
            this._setStatOption = value.setStatOption;
            this._tlsSessionResumptionMode = value.tlsSessionResumptionMode;
        }
    }
    get as2Transports() {
        return cdktf.Fn.tolist(this.getListAttribute('as2_transports'));
    }
    set as2Transports(value) {
        this._as2Transports = value;
    }
    resetAs2Transports() {
        this._as2Transports = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get as2TransportsInput() {
        return this._as2Transports;
    }
    get passiveIp() {
        return this.getStringAttribute('passive_ip');
    }
    set passiveIp(value) {
        this._passiveIp = value;
    }
    resetPassiveIp() {
        this._passiveIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passiveIpInput() {
        return this._passiveIp;
    }
    get setStatOption() {
        return this.getStringAttribute('set_stat_option');
    }
    set setStatOption(value) {
        this._setStatOption = value;
    }
    resetSetStatOption() {
        this._setStatOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setStatOptionInput() {
        return this._setStatOption;
    }
    get tlsSessionResumptionMode() {
        return this.getStringAttribute('tls_session_resumption_mode');
    }
    set tlsSessionResumptionMode(value) {
        this._tlsSessionResumptionMode = value;
    }
    resetTlsSessionResumptionMode() {
        this._tlsSessionResumptionMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsSessionResumptionModeInput() {
        return this._tlsSessionResumptionMode;
    }
}
exports.TransferServerProtocolDetailsOutputReference = TransferServerProtocolDetailsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
TransferServerProtocolDetailsOutputReference[_b] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerProtocolDetailsOutputReference", version: "19.15.0" };
function transferServerS3StorageOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        directory_listing_optimization: cdktf.stringToTerraform(struct.directoryListingOptimization),
    };
}
exports.transferServerS3StorageOptionsToTerraform = transferServerS3StorageOptionsToTerraform;
function transferServerS3StorageOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        directory_listing_optimization: {
            value: cdktf.stringToHclTerraform(struct.directoryListingOptimization),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerS3StorageOptionsToHclTerraform = transferServerS3StorageOptionsToHclTerraform;
class TransferServerS3StorageOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._directoryListingOptimization !== undefined) {
            hasAnyValues = true;
            internalValueResult.directoryListingOptimization = this._directoryListingOptimization;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._directoryListingOptimization = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._directoryListingOptimization = value.directoryListingOptimization;
        }
    }
    get directoryListingOptimization() {
        return this.getStringAttribute('directory_listing_optimization');
    }
    set directoryListingOptimization(value) {
        this._directoryListingOptimization = value;
    }
    resetDirectoryListingOptimization() {
        this._directoryListingOptimization = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directoryListingOptimizationInput() {
        return this._directoryListingOptimization;
    }
}
exports.TransferServerS3StorageOptionsOutputReference = TransferServerS3StorageOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
TransferServerS3StorageOptionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerS3StorageOptionsOutputReference", version: "19.15.0" };
function transferServerWorkflowDetailsOnPartialUploadToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execution_role: cdktf.stringToTerraform(struct.executionRole),
        workflow_id: cdktf.stringToTerraform(struct.workflowId),
    };
}
exports.transferServerWorkflowDetailsOnPartialUploadToTerraform = transferServerWorkflowDetailsOnPartialUploadToTerraform;
function transferServerWorkflowDetailsOnPartialUploadToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        execution_role: {
            value: cdktf.stringToHclTerraform(struct.executionRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        workflow_id: {
            value: cdktf.stringToHclTerraform(struct.workflowId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerWorkflowDetailsOnPartialUploadToHclTerraform = transferServerWorkflowDetailsOnPartialUploadToHclTerraform;
class TransferServerWorkflowDetailsOnPartialUploadOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._executionRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionRole = this._executionRole;
        }
        if (this._workflowId !== undefined) {
            hasAnyValues = true;
            internalValueResult.workflowId = this._workflowId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._executionRole = undefined;
            this._workflowId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._executionRole = value.executionRole;
            this._workflowId = value.workflowId;
        }
    }
    get executionRole() {
        return this.getStringAttribute('execution_role');
    }
    set executionRole(value) {
        this._executionRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleInput() {
        return this._executionRole;
    }
    get workflowId() {
        return this.getStringAttribute('workflow_id');
    }
    set workflowId(value) {
        this._workflowId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workflowIdInput() {
        return this._workflowId;
    }
}
exports.TransferServerWorkflowDetailsOnPartialUploadOutputReference = TransferServerWorkflowDetailsOnPartialUploadOutputReference;
_d = JSII_RTTI_SYMBOL_1;
TransferServerWorkflowDetailsOnPartialUploadOutputReference[_d] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerWorkflowDetailsOnPartialUploadOutputReference", version: "19.15.0" };
function transferServerWorkflowDetailsOnUploadToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execution_role: cdktf.stringToTerraform(struct.executionRole),
        workflow_id: cdktf.stringToTerraform(struct.workflowId),
    };
}
exports.transferServerWorkflowDetailsOnUploadToTerraform = transferServerWorkflowDetailsOnUploadToTerraform;
function transferServerWorkflowDetailsOnUploadToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        execution_role: {
            value: cdktf.stringToHclTerraform(struct.executionRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        workflow_id: {
            value: cdktf.stringToHclTerraform(struct.workflowId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerWorkflowDetailsOnUploadToHclTerraform = transferServerWorkflowDetailsOnUploadToHclTerraform;
class TransferServerWorkflowDetailsOnUploadOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._executionRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionRole = this._executionRole;
        }
        if (this._workflowId !== undefined) {
            hasAnyValues = true;
            internalValueResult.workflowId = this._workflowId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._executionRole = undefined;
            this._workflowId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._executionRole = value.executionRole;
            this._workflowId = value.workflowId;
        }
    }
    get executionRole() {
        return this.getStringAttribute('execution_role');
    }
    set executionRole(value) {
        this._executionRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleInput() {
        return this._executionRole;
    }
    get workflowId() {
        return this.getStringAttribute('workflow_id');
    }
    set workflowId(value) {
        this._workflowId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workflowIdInput() {
        return this._workflowId;
    }
}
exports.TransferServerWorkflowDetailsOnUploadOutputReference = TransferServerWorkflowDetailsOnUploadOutputReference;
_e = JSII_RTTI_SYMBOL_1;
TransferServerWorkflowDetailsOnUploadOutputReference[_e] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerWorkflowDetailsOnUploadOutputReference", version: "19.15.0" };
function transferServerWorkflowDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_partial_upload: transferServerWorkflowDetailsOnPartialUploadToTerraform(struct.onPartialUpload),
        on_upload: transferServerWorkflowDetailsOnUploadToTerraform(struct.onUpload),
    };
}
exports.transferServerWorkflowDetailsToTerraform = transferServerWorkflowDetailsToTerraform;
function transferServerWorkflowDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        on_partial_upload: {
            value: transferServerWorkflowDetailsOnPartialUploadToHclTerraform(struct.onPartialUpload),
            isBlock: true,
            type: "list",
            storageClassType: "TransferServerWorkflowDetailsOnPartialUploadList",
        },
        on_upload: {
            value: transferServerWorkflowDetailsOnUploadToHclTerraform(struct.onUpload),
            isBlock: true,
            type: "list",
            storageClassType: "TransferServerWorkflowDetailsOnUploadList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transferServerWorkflowDetailsToHclTerraform = transferServerWorkflowDetailsToHclTerraform;
class TransferServerWorkflowDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // on_partial_upload - computed: false, optional: true, required: false
        this._onPartialUpload = new TransferServerWorkflowDetailsOnPartialUploadOutputReference(this, "on_partial_upload");
        // on_upload - computed: false, optional: true, required: false
        this._onUpload = new TransferServerWorkflowDetailsOnUploadOutputReference(this, "on_upload");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onPartialUpload?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onPartialUpload = this._onPartialUpload?.internalValue;
        }
        if (this._onUpload?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onUpload = this._onUpload?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._onPartialUpload.internalValue = undefined;
            this._onUpload.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._onPartialUpload.internalValue = value.onPartialUpload;
            this._onUpload.internalValue = value.onUpload;
        }
    }
    get onPartialUpload() {
        return this._onPartialUpload;
    }
    putOnPartialUpload(value) {
        this._onPartialUpload.internalValue = value;
    }
    resetOnPartialUpload() {
        this._onPartialUpload.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onPartialUploadInput() {
        return this._onPartialUpload.internalValue;
    }
    get onUpload() {
        return this._onUpload;
    }
    putOnUpload(value) {
        this._onUpload.internalValue = value;
    }
    resetOnUpload() {
        this._onUpload.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onUploadInput() {
        return this._onUpload.internalValue;
    }
}
exports.TransferServerWorkflowDetailsOutputReference = TransferServerWorkflowDetailsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
TransferServerWorkflowDetailsOutputReference[_f] = { fqn: "@cdktf/provider-aws.transferServer.TransferServerWorkflowDetailsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_server aws_transfer_server}
*/
class TransferServer extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a TransferServer resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the TransferServer to import
    * @param importFromId The id of the existing TransferServer that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_server#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the TransferServer to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_transfer_server", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/transfer_server aws_transfer_server} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options TransferServerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_transfer_server',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint_details - computed: false, optional: true, required: false
        this._endpointDetails = new TransferServerEndpointDetailsOutputReference(this, "endpoint_details");
        // protocol_details - computed: false, optional: true, required: false
        this._protocolDetails = new TransferServerProtocolDetailsOutputReference(this, "protocol_details");
        // s3_storage_options - computed: false, optional: true, required: false
        this._s3StorageOptions = new TransferServerS3StorageOptionsOutputReference(this, "s3_storage_options");
        // workflow_details - computed: false, optional: true, required: false
        this._workflowDetails = new TransferServerWorkflowDetailsOutputReference(this, "workflow_details");
        this._certificate = config.certificate;
        this._directoryId = config.directoryId;
        this._domain = config.domain;
        this._endpointType = config.endpointType;
        this._forceDestroy = config.forceDestroy;
        this._function = config.function;
        this._hostKey = config.hostKey;
        this._id = config.id;
        this._identityProviderType = config.identityProviderType;
        this._invocationRole = config.invocationRole;
        this._loggingRole = config.loggingRole;
        this._postAuthenticationLoginBanner = config.postAuthenticationLoginBanner;
        this._preAuthenticationLoginBanner = config.preAuthenticationLoginBanner;
        this._protocols = config.protocols;
        this._securityPolicyName = config.securityPolicyName;
        this._structuredLogDestinations = config.structuredLogDestinations;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._url = config.url;
        this._endpointDetails.internalValue = config.endpointDetails;
        this._protocolDetails.internalValue = config.protocolDetails;
        this._s3StorageOptions.internalValue = config.s3StorageOptions;
        this._workflowDetails.internalValue = config.workflowDetails;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    set certificate(value) {
        this._certificate = value;
    }
    resetCertificate() {
        this._certificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificateInput() {
        return this._certificate;
    }
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    set directoryId(value) {
        this._directoryId = value;
    }
    resetDirectoryId() {
        this._directoryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directoryIdInput() {
        return this._directoryId;
    }
    get domain() {
        return this.getStringAttribute('domain');
    }
    set domain(value) {
        this._domain = value;
    }
    resetDomain() {
        this._domain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainInput() {
        return this._domain;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get endpointType() {
        return this.getStringAttribute('endpoint_type');
    }
    set endpointType(value) {
        this._endpointType = value;
    }
    resetEndpointType() {
        this._endpointType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointTypeInput() {
        return this._endpointType;
    }
    get forceDestroy() {
        return this.getBooleanAttribute('force_destroy');
    }
    set forceDestroy(value) {
        this._forceDestroy = value;
    }
    resetForceDestroy() {
        this._forceDestroy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceDestroyInput() {
        return this._forceDestroy;
    }
    get function() {
        return this.getStringAttribute('function');
    }
    set function(value) {
        this._function = value;
    }
    resetFunction() {
        this._function = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get functionInput() {
        return this._function;
    }
    get hostKey() {
        return this.getStringAttribute('host_key');
    }
    set hostKey(value) {
        this._hostKey = value;
    }
    resetHostKey() {
        this._hostKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostKeyInput() {
        return this._hostKey;
    }
    // host_key_fingerprint - computed: true, optional: false, required: false
    get hostKeyFingerprint() {
        return this.getStringAttribute('host_key_fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityProviderType() {
        return this.getStringAttribute('identity_provider_type');
    }
    set identityProviderType(value) {
        this._identityProviderType = value;
    }
    resetIdentityProviderType() {
        this._identityProviderType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identityProviderTypeInput() {
        return this._identityProviderType;
    }
    get invocationRole() {
        return this.getStringAttribute('invocation_role');
    }
    set invocationRole(value) {
        this._invocationRole = value;
    }
    resetInvocationRole() {
        this._invocationRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invocationRoleInput() {
        return this._invocationRole;
    }
    get loggingRole() {
        return this.getStringAttribute('logging_role');
    }
    set loggingRole(value) {
        this._loggingRole = value;
    }
    resetLoggingRole() {
        this._loggingRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingRoleInput() {
        return this._loggingRole;
    }
    get postAuthenticationLoginBanner() {
        return this.getStringAttribute('post_authentication_login_banner');
    }
    set postAuthenticationLoginBanner(value) {
        this._postAuthenticationLoginBanner = value;
    }
    resetPostAuthenticationLoginBanner() {
        this._postAuthenticationLoginBanner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postAuthenticationLoginBannerInput() {
        return this._postAuthenticationLoginBanner;
    }
    get preAuthenticationLoginBanner() {
        return this.getStringAttribute('pre_authentication_login_banner');
    }
    set preAuthenticationLoginBanner(value) {
        this._preAuthenticationLoginBanner = value;
    }
    resetPreAuthenticationLoginBanner() {
        this._preAuthenticationLoginBanner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preAuthenticationLoginBannerInput() {
        return this._preAuthenticationLoginBanner;
    }
    get protocols() {
        return cdktf.Fn.tolist(this.getListAttribute('protocols'));
    }
    set protocols(value) {
        this._protocols = value;
    }
    resetProtocols() {
        this._protocols = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolsInput() {
        return this._protocols;
    }
    get securityPolicyName() {
        return this.getStringAttribute('security_policy_name');
    }
    set securityPolicyName(value) {
        this._securityPolicyName = value;
    }
    resetSecurityPolicyName() {
        this._securityPolicyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityPolicyNameInput() {
        return this._securityPolicyName;
    }
    get structuredLogDestinations() {
        return cdktf.Fn.tolist(this.getListAttribute('structured_log_destinations'));
    }
    set structuredLogDestinations(value) {
        this._structuredLogDestinations = value;
    }
    resetStructuredLogDestinations() {
        this._structuredLogDestinations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get structuredLogDestinationsInput() {
        return this._structuredLogDestinations;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get url() {
        return this.getStringAttribute('url');
    }
    set url(value) {
        this._url = value;
    }
    resetUrl() {
        this._url = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get urlInput() {
        return this._url;
    }
    get endpointDetails() {
        return this._endpointDetails;
    }
    putEndpointDetails(value) {
        this._endpointDetails.internalValue = value;
    }
    resetEndpointDetails() {
        this._endpointDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointDetailsInput() {
        return this._endpointDetails.internalValue;
    }
    get protocolDetails() {
        return this._protocolDetails;
    }
    putProtocolDetails(value) {
        this._protocolDetails.internalValue = value;
    }
    resetProtocolDetails() {
        this._protocolDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolDetailsInput() {
        return this._protocolDetails.internalValue;
    }
    get s3StorageOptions() {
        return this._s3StorageOptions;
    }
    putS3StorageOptions(value) {
        this._s3StorageOptions.internalValue = value;
    }
    resetS3StorageOptions() {
        this._s3StorageOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3StorageOptionsInput() {
        return this._s3StorageOptions.internalValue;
    }
    get workflowDetails() {
        return this._workflowDetails;
    }
    putWorkflowDetails(value) {
        this._workflowDetails.internalValue = value;
    }
    resetWorkflowDetails() {
        this._workflowDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workflowDetailsInput() {
        return this._workflowDetails.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            certificate: cdktf.stringToTerraform(this._certificate),
            directory_id: cdktf.stringToTerraform(this._directoryId),
            domain: cdktf.stringToTerraform(this._domain),
            endpoint_type: cdktf.stringToTerraform(this._endpointType),
            force_destroy: cdktf.booleanToTerraform(this._forceDestroy),
            function: cdktf.stringToTerraform(this._function),
            host_key: cdktf.stringToTerraform(this._hostKey),
            id: cdktf.stringToTerraform(this._id),
            identity_provider_type: cdktf.stringToTerraform(this._identityProviderType),
            invocation_role: cdktf.stringToTerraform(this._invocationRole),
            logging_role: cdktf.stringToTerraform(this._loggingRole),
            post_authentication_login_banner: cdktf.stringToTerraform(this._postAuthenticationLoginBanner),
            pre_authentication_login_banner: cdktf.stringToTerraform(this._preAuthenticationLoginBanner),
            protocols: cdktf.listMapper(cdktf.stringToTerraform, false)(this._protocols),
            security_policy_name: cdktf.stringToTerraform(this._securityPolicyName),
            structured_log_destinations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._structuredLogDestinations),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            url: cdktf.stringToTerraform(this._url),
            endpoint_details: transferServerEndpointDetailsToTerraform(this._endpointDetails.internalValue),
            protocol_details: transferServerProtocolDetailsToTerraform(this._protocolDetails.internalValue),
            s3_storage_options: transferServerS3StorageOptionsToTerraform(this._s3StorageOptions.internalValue),
            workflow_details: transferServerWorkflowDetailsToTerraform(this._workflowDetails.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            certificate: {
                value: cdktf.stringToHclTerraform(this._certificate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            directory_id: {
                value: cdktf.stringToHclTerraform(this._directoryId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            domain: {
                value: cdktf.stringToHclTerraform(this._domain),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_type: {
                value: cdktf.stringToHclTerraform(this._endpointType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            force_destroy: {
                value: cdktf.booleanToHclTerraform(this._forceDestroy),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            function: {
                value: cdktf.stringToHclTerraform(this._function),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            host_key: {
                value: cdktf.stringToHclTerraform(this._hostKey),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identity_provider_type: {
                value: cdktf.stringToHclTerraform(this._identityProviderType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            invocation_role: {
                value: cdktf.stringToHclTerraform(this._invocationRole),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            logging_role: {
                value: cdktf.stringToHclTerraform(this._loggingRole),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            post_authentication_login_banner: {
                value: cdktf.stringToHclTerraform(this._postAuthenticationLoginBanner),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pre_authentication_login_banner: {
                value: cdktf.stringToHclTerraform(this._preAuthenticationLoginBanner),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            protocols: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._protocols),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            security_policy_name: {
                value: cdktf.stringToHclTerraform(this._securityPolicyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            structured_log_destinations: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._structuredLogDestinations),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            url: {
                value: cdktf.stringToHclTerraform(this._url),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_details: {
                value: transferServerEndpointDetailsToHclTerraform(this._endpointDetails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "TransferServerEndpointDetailsList",
            },
            protocol_details: {
                value: transferServerProtocolDetailsToHclTerraform(this._protocolDetails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "TransferServerProtocolDetailsList",
            },
            s3_storage_options: {
                value: transferServerS3StorageOptionsToHclTerraform(this._s3StorageOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "TransferServerS3StorageOptionsList",
            },
            workflow_details: {
                value: transferServerWorkflowDetailsToHclTerraform(this._workflowDetails.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "TransferServerWorkflowDetailsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.TransferServer = TransferServer;
_g = JSII_RTTI_SYMBOL_1;
TransferServer[_g] = { fqn: "@cdktf/provider-aws.transferServer.TransferServer", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
TransferServer.tfResourceType = "aws_transfer_server";
