// @cdktf/provider-aws/lib/waf-regex-match-set/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafRegexMatchSet = exports.WafRegexMatchSetRegexMatchTupleList = exports.WafRegexMatchSetRegexMatchTupleOutputReference = exports.wafRegexMatchSetRegexMatchTupleToHclTerraform = exports.wafRegexMatchSetRegexMatchTupleToTerraform = exports.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference = exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToHclTerraform = exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data: cdktf.stringToTerraform(struct.data),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform = wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform;
function wafRegexMatchSetRegexMatchTupleFieldToMatchToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        data: {
            value: cdktf.stringToHclTerraform(struct.data),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToHclTerraform = wafRegexMatchSetRegexMatchTupleFieldToMatchToHclTerraform;
class WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._data !== undefined) {
            hasAnyValues = true;
            internalValueResult.data = this._data;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._data = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._data = value.data;
            this._type = value.type;
        }
    }
    get data() {
        return this.getStringAttribute('data');
    }
    set data(value) {
        this._data = value;
    }
    resetData() {
        this._data = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataInput() {
        return this._data;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference = WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference[_a] = { fqn: "@cdktf/provider-aws.wafRegexMatchSet.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference", version: "19.15.0" };
function wafRegexMatchSetRegexMatchTupleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        regex_pattern_set_id: cdktf.stringToTerraform(struct.regexPatternSetId),
        text_transformation: cdktf.stringToTerraform(struct.textTransformation),
        field_to_match: wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform(struct.fieldToMatch),
    };
}
exports.wafRegexMatchSetRegexMatchTupleToTerraform = wafRegexMatchSetRegexMatchTupleToTerraform;
function wafRegexMatchSetRegexMatchTupleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        regex_pattern_set_id: {
            value: cdktf.stringToHclTerraform(struct.regexPatternSetId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        text_transformation: {
            value: cdktf.stringToHclTerraform(struct.textTransformation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        field_to_match: {
            value: wafRegexMatchSetRegexMatchTupleFieldToMatchToHclTerraform(struct.fieldToMatch),
            isBlock: true,
            type: "list",
            storageClassType: "WafRegexMatchSetRegexMatchTupleFieldToMatchList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.wafRegexMatchSetRegexMatchTupleToHclTerraform = wafRegexMatchSetRegexMatchTupleToHclTerraform;
class WafRegexMatchSetRegexMatchTupleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // field_to_match - computed: false, optional: false, required: true
        this._fieldToMatch = new WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference(this, "field_to_match");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._regexPatternSetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.regexPatternSetId = this._regexPatternSetId;
        }
        if (this._textTransformation !== undefined) {
            hasAnyValues = true;
            internalValueResult.textTransformation = this._textTransformation;
        }
        if (this._fieldToMatch?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.fieldToMatch = this._fieldToMatch?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._regexPatternSetId = undefined;
            this._textTransformation = undefined;
            this._fieldToMatch.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._regexPatternSetId = value.regexPatternSetId;
            this._textTransformation = value.textTransformation;
            this._fieldToMatch.internalValue = value.fieldToMatch;
        }
    }
    get regexPatternSetId() {
        return this.getStringAttribute('regex_pattern_set_id');
    }
    set regexPatternSetId(value) {
        this._regexPatternSetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get regexPatternSetIdInput() {
        return this._regexPatternSetId;
    }
    get textTransformation() {
        return this.getStringAttribute('text_transformation');
    }
    set textTransformation(value) {
        this._textTransformation = value;
    }
    // Temporarily expose input value. Use with caution.
    get textTransformationInput() {
        return this._textTransformation;
    }
    get fieldToMatch() {
        return this._fieldToMatch;
    }
    putFieldToMatch(value) {
        this._fieldToMatch.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldToMatchInput() {
        return this._fieldToMatch.internalValue;
    }
}
exports.WafRegexMatchSetRegexMatchTupleOutputReference = WafRegexMatchSetRegexMatchTupleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
WafRegexMatchSetRegexMatchTupleOutputReference[_b] = { fqn: "@cdktf/provider-aws.wafRegexMatchSet.WafRegexMatchSetRegexMatchTupleOutputReference", version: "19.15.0" };
class WafRegexMatchSetRegexMatchTupleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WafRegexMatchSetRegexMatchTupleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WafRegexMatchSetRegexMatchTupleList = WafRegexMatchSetRegexMatchTupleList;
_c = JSII_RTTI_SYMBOL_1;
WafRegexMatchSetRegexMatchTupleList[_c] = { fqn: "@cdktf/provider-aws.wafRegexMatchSet.WafRegexMatchSetRegexMatchTupleList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/waf_regex_match_set aws_waf_regex_match_set}
*/
class WafRegexMatchSet extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a WafRegexMatchSet resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the WafRegexMatchSet to import
    * @param importFromId The id of the existing WafRegexMatchSet that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/waf_regex_match_set#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the WafRegexMatchSet to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_waf_regex_match_set", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/waf_regex_match_set aws_waf_regex_match_set} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options WafRegexMatchSetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_waf_regex_match_set',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // regex_match_tuple - computed: false, optional: true, required: false
        this._regexMatchTuple = new WafRegexMatchSetRegexMatchTupleList(this, "regex_match_tuple", true);
        this._id = config.id;
        this._name = config.name;
        this._regexMatchTuple.internalValue = config.regexMatchTuple;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get regexMatchTuple() {
        return this._regexMatchTuple;
    }
    putRegexMatchTuple(value) {
        this._regexMatchTuple.internalValue = value;
    }
    resetRegexMatchTuple() {
        this._regexMatchTuple.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexMatchTupleInput() {
        return this._regexMatchTuple.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            regex_match_tuple: cdktf.listMapper(wafRegexMatchSetRegexMatchTupleToTerraform, true)(this._regexMatchTuple.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            regex_match_tuple: {
                value: cdktf.listMapperHcl(wafRegexMatchSetRegexMatchTupleToHclTerraform, true)(this._regexMatchTuple.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "WafRegexMatchSetRegexMatchTupleList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.WafRegexMatchSet = WafRegexMatchSet;
_d = JSII_RTTI_SYMBOL_1;
WafRegexMatchSet[_d] = { fqn: "@cdktf/provider-aws.wafRegexMatchSet.WafRegexMatchSet", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
WafRegexMatchSet.tfResourceType = "aws_waf_regex_match_set";
