// @cdktf/provider-aws/lib/licensemanager-license-configuration/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicensemanagerLicenseConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_license_configuration aws_licensemanager_license_configuration}
*/
class LicensemanagerLicenseConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LicensemanagerLicenseConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LicensemanagerLicenseConfiguration to import
    * @param importFromId The id of the existing LicensemanagerLicenseConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_license_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LicensemanagerLicenseConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_licensemanager_license_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/licensemanager_license_configuration aws_licensemanager_license_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LicensemanagerLicenseConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_licensemanager_license_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._licenseCount = config.licenseCount;
        this._licenseCountHardLimit = config.licenseCountHardLimit;
        this._licenseCountingType = config.licenseCountingType;
        this._licenseRules = config.licenseRules;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get licenseCount() {
        return this.getNumberAttribute('license_count');
    }
    set licenseCount(value) {
        this._licenseCount = value;
    }
    resetLicenseCount() {
        this._licenseCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountInput() {
        return this._licenseCount;
    }
    get licenseCountHardLimit() {
        return this.getBooleanAttribute('license_count_hard_limit');
    }
    set licenseCountHardLimit(value) {
        this._licenseCountHardLimit = value;
    }
    resetLicenseCountHardLimit() {
        this._licenseCountHardLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountHardLimitInput() {
        return this._licenseCountHardLimit;
    }
    get licenseCountingType() {
        return this.getStringAttribute('license_counting_type');
    }
    set licenseCountingType(value) {
        this._licenseCountingType = value;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountingTypeInput() {
        return this._licenseCountingType;
    }
    get licenseRules() {
        return this.getListAttribute('license_rules');
    }
    set licenseRules(value) {
        this._licenseRules = value;
    }
    resetLicenseRules() {
        this._licenseRules = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseRulesInput() {
        return this._licenseRules;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // owner_account_id - computed: true, optional: false, required: false
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            license_count: cdktf.numberToTerraform(this._licenseCount),
            license_count_hard_limit: cdktf.booleanToTerraform(this._licenseCountHardLimit),
            license_counting_type: cdktf.stringToTerraform(this._licenseCountingType),
            license_rules: cdktf.listMapper(cdktf.stringToTerraform, false)(this._licenseRules),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            license_count: {
                value: cdktf.numberToHclTerraform(this._licenseCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            license_count_hard_limit: {
                value: cdktf.booleanToHclTerraform(this._licenseCountHardLimit),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            license_counting_type: {
                value: cdktf.stringToHclTerraform(this._licenseCountingType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            license_rules: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._licenseRules),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LicensemanagerLicenseConfiguration = LicensemanagerLicenseConfiguration;
_a = JSII_RTTI_SYMBOL_1;
LicensemanagerLicenseConfiguration[_a] = { fqn: "@cdktf/provider-aws.licensemanagerLicenseConfiguration.LicensemanagerLicenseConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LicensemanagerLicenseConfiguration.tfResourceType = "aws_licensemanager_license_configuration";
