// @cdktf/provider-aws/lib/rolesanywhere-profile/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesanywhereProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_profile aws_rolesanywhere_profile}
*/
class RolesanywhereProfile extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RolesanywhereProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RolesanywhereProfile to import
    * @param importFromId The id of the existing RolesanywhereProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RolesanywhereProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_rolesanywhere_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/rolesanywhere_profile aws_rolesanywhere_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RolesanywhereProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rolesanywhere_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._durationSeconds = config.durationSeconds;
        this._enabled = config.enabled;
        this._id = config.id;
        this._managedPolicyArns = config.managedPolicyArns;
        this._name = config.name;
        this._requireInstanceProperties = config.requireInstanceProperties;
        this._roleArns = config.roleArns;
        this._sessionPolicy = config.sessionPolicy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get managedPolicyArns() {
        return cdktf.Fn.tolist(this.getListAttribute('managed_policy_arns'));
    }
    set managedPolicyArns(value) {
        this._managedPolicyArns = value;
    }
    resetManagedPolicyArns() {
        this._managedPolicyArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedPolicyArnsInput() {
        return this._managedPolicyArns;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get requireInstanceProperties() {
        return this.getBooleanAttribute('require_instance_properties');
    }
    set requireInstanceProperties(value) {
        this._requireInstanceProperties = value;
    }
    resetRequireInstanceProperties() {
        this._requireInstanceProperties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requireInstancePropertiesInput() {
        return this._requireInstanceProperties;
    }
    get roleArns() {
        return cdktf.Fn.tolist(this.getListAttribute('role_arns'));
    }
    set roleArns(value) {
        this._roleArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnsInput() {
        return this._roleArns;
    }
    get sessionPolicy() {
        return this.getStringAttribute('session_policy');
    }
    set sessionPolicy(value) {
        this._sessionPolicy = value;
    }
    resetSessionPolicy() {
        this._sessionPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionPolicyInput() {
        return this._sessionPolicy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            managed_policy_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._managedPolicyArns),
            name: cdktf.stringToTerraform(this._name),
            require_instance_properties: cdktf.booleanToTerraform(this._requireInstanceProperties),
            role_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._roleArns),
            session_policy: cdktf.stringToTerraform(this._sessionPolicy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            duration_seconds: {
                value: cdktf.numberToHclTerraform(this._durationSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            managed_policy_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._managedPolicyArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            require_instance_properties: {
                value: cdktf.booleanToHclTerraform(this._requireInstanceProperties),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            role_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._roleArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            session_policy: {
                value: cdktf.stringToHclTerraform(this._sessionPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RolesanywhereProfile = RolesanywhereProfile;
_a = JSII_RTTI_SYMBOL_1;
RolesanywhereProfile[_a] = { fqn: "@cdktf/provider-aws.rolesanywhereProfile.RolesanywhereProfile", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
RolesanywhereProfile.tfResourceType = "aws_rolesanywhere_profile";
