// @cdktf/provider-aws/lib/sagemaker-model-package-group-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerModelPackageGroupPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model_package_group_policy aws_sagemaker_model_package_group_policy}
*/
class SagemakerModelPackageGroupPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SagemakerModelPackageGroupPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SagemakerModelPackageGroupPolicy to import
    * @param importFromId The id of the existing SagemakerModelPackageGroupPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model_package_group_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SagemakerModelPackageGroupPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_sagemaker_model_package_group_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/sagemaker_model_package_group_policy aws_sagemaker_model_package_group_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerModelPackageGroupPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_model_package_group_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._modelPackageGroupName = config.modelPackageGroupName;
        this._resourcePolicy = config.resourcePolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get modelPackageGroupName() {
        return this.getStringAttribute('model_package_group_name');
    }
    set modelPackageGroupName(value) {
        this._modelPackageGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get modelPackageGroupNameInput() {
        return this._modelPackageGroupName;
    }
    get resourcePolicy() {
        return this.getStringAttribute('resource_policy');
    }
    set resourcePolicy(value) {
        this._resourcePolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePolicyInput() {
        return this._resourcePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            model_package_group_name: cdktf.stringToTerraform(this._modelPackageGroupName),
            resource_policy: cdktf.stringToTerraform(this._resourcePolicy),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            model_package_group_name: {
                value: cdktf.stringToHclTerraform(this._modelPackageGroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resource_policy: {
                value: cdktf.stringToHclTerraform(this._resourcePolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SagemakerModelPackageGroupPolicy = SagemakerModelPackageGroupPolicy;
_a = JSII_RTTI_SYMBOL_1;
SagemakerModelPackageGroupPolicy[_a] = { fqn: "@cdktf/provider-aws.sagemakerModelPackageGroupPolicy.SagemakerModelPackageGroupPolicy", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SagemakerModelPackageGroupPolicy.tfResourceType = "aws_sagemaker_model_package_group_policy";
