// @cdktf/provider-aws/lib/prometheus-scraper/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrometheusScraper = exports.PrometheusScraperTimeoutsOutputReference = exports.prometheusScraperTimeoutsToHclTerraform = exports.prometheusScraperTimeoutsToTerraform = exports.PrometheusScraperSourceList = exports.PrometheusScraperSourceOutputReference = exports.prometheusScraperSourceToHclTerraform = exports.prometheusScraperSourceToTerraform = exports.PrometheusScraperSourceEksList = exports.PrometheusScraperSourceEksOutputReference = exports.prometheusScraperSourceEksToHclTerraform = exports.prometheusScraperSourceEksToTerraform = exports.PrometheusScraperDestinationList = exports.PrometheusScraperDestinationOutputReference = exports.prometheusScraperDestinationToHclTerraform = exports.prometheusScraperDestinationToTerraform = exports.PrometheusScraperDestinationAmpList = exports.PrometheusScraperDestinationAmpOutputReference = exports.prometheusScraperDestinationAmpToHclTerraform = exports.prometheusScraperDestinationAmpToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function prometheusScraperDestinationAmpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        workspace_arn: cdktf.stringToTerraform(struct.workspaceArn),
    };
}
exports.prometheusScraperDestinationAmpToTerraform = prometheusScraperDestinationAmpToTerraform;
function prometheusScraperDestinationAmpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        workspace_arn: {
            value: cdktf.stringToHclTerraform(struct.workspaceArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.prometheusScraperDestinationAmpToHclTerraform = prometheusScraperDestinationAmpToHclTerraform;
class PrometheusScraperDestinationAmpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._workspaceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.workspaceArn = this._workspaceArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._workspaceArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._workspaceArn = value.workspaceArn;
        }
    }
    get workspaceArn() {
        return this.getStringAttribute('workspace_arn');
    }
    set workspaceArn(value) {
        this._workspaceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get workspaceArnInput() {
        return this._workspaceArn;
    }
}
exports.PrometheusScraperDestinationAmpOutputReference = PrometheusScraperDestinationAmpOutputReference;
_a = JSII_RTTI_SYMBOL_1;
PrometheusScraperDestinationAmpOutputReference[_a] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperDestinationAmpOutputReference", version: "19.15.0" };
class PrometheusScraperDestinationAmpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new PrometheusScraperDestinationAmpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.PrometheusScraperDestinationAmpList = PrometheusScraperDestinationAmpList;
_b = JSII_RTTI_SYMBOL_1;
PrometheusScraperDestinationAmpList[_b] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperDestinationAmpList", version: "19.15.0" };
function prometheusScraperDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        amp: cdktf.listMapper(prometheusScraperDestinationAmpToTerraform, true)(struct.amp),
    };
}
exports.prometheusScraperDestinationToTerraform = prometheusScraperDestinationToTerraform;
function prometheusScraperDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        amp: {
            value: cdktf.listMapperHcl(prometheusScraperDestinationAmpToHclTerraform, true)(struct.amp),
            isBlock: true,
            type: "list",
            storageClassType: "PrometheusScraperDestinationAmpList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.prometheusScraperDestinationToHclTerraform = prometheusScraperDestinationToHclTerraform;
class PrometheusScraperDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // amp - computed: false, optional: true, required: false
        this._amp = new PrometheusScraperDestinationAmpList(this, "amp", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._amp?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.amp = this._amp?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._amp.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._amp.internalValue = value.amp;
        }
    }
    get amp() {
        return this._amp;
    }
    putAmp(value) {
        this._amp.internalValue = value;
    }
    resetAmp() {
        this._amp.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ampInput() {
        return this._amp.internalValue;
    }
}
exports.PrometheusScraperDestinationOutputReference = PrometheusScraperDestinationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
PrometheusScraperDestinationOutputReference[_c] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperDestinationOutputReference", version: "19.15.0" };
class PrometheusScraperDestinationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new PrometheusScraperDestinationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.PrometheusScraperDestinationList = PrometheusScraperDestinationList;
_d = JSII_RTTI_SYMBOL_1;
PrometheusScraperDestinationList[_d] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperDestinationList", version: "19.15.0" };
function prometheusScraperSourceEksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cluster_arn: cdktf.stringToTerraform(struct.clusterArn),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.prometheusScraperSourceEksToTerraform = prometheusScraperSourceEksToTerraform;
function prometheusScraperSourceEksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cluster_arn: {
            value: cdktf.stringToHclTerraform(struct.clusterArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        subnet_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.prometheusScraperSourceEksToHclTerraform = prometheusScraperSourceEksToHclTerraform;
class PrometheusScraperSourceEksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clusterArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.clusterArn = this._clusterArn;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._clusterArn = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._clusterArn = value.clusterArn;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterArnInput() {
        return this._clusterArn;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.PrometheusScraperSourceEksOutputReference = PrometheusScraperSourceEksOutputReference;
_e = JSII_RTTI_SYMBOL_1;
PrometheusScraperSourceEksOutputReference[_e] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperSourceEksOutputReference", version: "19.15.0" };
class PrometheusScraperSourceEksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new PrometheusScraperSourceEksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.PrometheusScraperSourceEksList = PrometheusScraperSourceEksList;
_f = JSII_RTTI_SYMBOL_1;
PrometheusScraperSourceEksList[_f] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperSourceEksList", version: "19.15.0" };
function prometheusScraperSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        eks: cdktf.listMapper(prometheusScraperSourceEksToTerraform, true)(struct.eks),
    };
}
exports.prometheusScraperSourceToTerraform = prometheusScraperSourceToTerraform;
function prometheusScraperSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        eks: {
            value: cdktf.listMapperHcl(prometheusScraperSourceEksToHclTerraform, true)(struct.eks),
            isBlock: true,
            type: "list",
            storageClassType: "PrometheusScraperSourceEksList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.prometheusScraperSourceToHclTerraform = prometheusScraperSourceToHclTerraform;
class PrometheusScraperSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // eks - computed: false, optional: true, required: false
        this._eks = new PrometheusScraperSourceEksList(this, "eks", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._eks?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.eks = this._eks?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._eks.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._eks.internalValue = value.eks;
        }
    }
    get eks() {
        return this._eks;
    }
    putEks(value) {
        this._eks.internalValue = value;
    }
    resetEks() {
        this._eks.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eksInput() {
        return this._eks.internalValue;
    }
}
exports.PrometheusScraperSourceOutputReference = PrometheusScraperSourceOutputReference;
_g = JSII_RTTI_SYMBOL_1;
PrometheusScraperSourceOutputReference[_g] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperSourceOutputReference", version: "19.15.0" };
class PrometheusScraperSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new PrometheusScraperSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.PrometheusScraperSourceList = PrometheusScraperSourceList;
_h = JSII_RTTI_SYMBOL_1;
PrometheusScraperSourceList[_h] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperSourceList", version: "19.15.0" };
function prometheusScraperTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.prometheusScraperTimeoutsToTerraform = prometheusScraperTimeoutsToTerraform;
function prometheusScraperTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.prometheusScraperTimeoutsToHclTerraform = prometheusScraperTimeoutsToHclTerraform;
class PrometheusScraperTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.PrometheusScraperTimeoutsOutputReference = PrometheusScraperTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
PrometheusScraperTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraperTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/prometheus_scraper aws_prometheus_scraper}
*/
class PrometheusScraper extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a PrometheusScraper resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the PrometheusScraper to import
    * @param importFromId The id of the existing PrometheusScraper that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/prometheus_scraper#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the PrometheusScraper to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_prometheus_scraper", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/prometheus_scraper aws_prometheus_scraper} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options PrometheusScraperConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_prometheus_scraper',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // destination - computed: false, optional: true, required: false
        this._destination = new PrometheusScraperDestinationList(this, "destination", false);
        // source - computed: false, optional: true, required: false
        this._source = new PrometheusScraperSourceList(this, "source", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new PrometheusScraperTimeoutsOutputReference(this, "timeouts");
        this._alias = config.alias;
        this._scrapeConfiguration = config.scrapeConfiguration;
        this._tags = config.tags;
        this._destination.internalValue = config.destination;
        this._source.internalValue = config.source;
        this._timeouts.internalValue = config.timeouts;
    }
    get alias() {
        return this.getStringAttribute('alias');
    }
    set alias(value) {
        this._alias = value;
    }
    resetAlias() {
        this._alias = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    get scrapeConfiguration() {
        return this.getStringAttribute('scrape_configuration');
    }
    set scrapeConfiguration(value) {
        this._scrapeConfiguration = value;
    }
    // Temporarily expose input value. Use with caution.
    get scrapeConfigurationInput() {
        return this._scrapeConfiguration;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get destination() {
        return this._destination;
    }
    putDestination(value) {
        this._destination.internalValue = value;
    }
    resetDestination() {
        this._destination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination.internalValue;
    }
    get source() {
        return this._source;
    }
    putSource(value) {
        this._source.internalValue = value;
    }
    resetSource() {
        this._source.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            scrape_configuration: cdktf.stringToTerraform(this._scrapeConfiguration),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            destination: cdktf.listMapper(prometheusScraperDestinationToTerraform, true)(this._destination.internalValue),
            source: cdktf.listMapper(prometheusScraperSourceToTerraform, true)(this._source.internalValue),
            timeouts: prometheusScraperTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            alias: {
                value: cdktf.stringToHclTerraform(this._alias),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scrape_configuration: {
                value: cdktf.stringToHclTerraform(this._scrapeConfiguration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            destination: {
                value: cdktf.listMapperHcl(prometheusScraperDestinationToHclTerraform, true)(this._destination.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "PrometheusScraperDestinationList",
            },
            source: {
                value: cdktf.listMapperHcl(prometheusScraperSourceToHclTerraform, true)(this._source.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "PrometheusScraperSourceList",
            },
            timeouts: {
                value: prometheusScraperTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "PrometheusScraperTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.PrometheusScraper = PrometheusScraper;
_k = JSII_RTTI_SYMBOL_1;
PrometheusScraper[_k] = { fqn: "@cdktf/provider-aws.prometheusScraper.PrometheusScraper", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
PrometheusScraper.tfResourceType = "aws_prometheus_scraper";
