// @cdktf/provider-aws/lib/verifiedaccess-instance-trust-provider-attachment/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedaccessInstanceTrustProviderAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_trust_provider_attachment aws_verifiedaccess_instance_trust_provider_attachment}
*/
class VerifiedaccessInstanceTrustProviderAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedaccessInstanceTrustProviderAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedaccessInstanceTrustProviderAttachment to import
    * @param importFromId The id of the existing VerifiedaccessInstanceTrustProviderAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_trust_provider_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedaccessInstanceTrustProviderAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedaccess_instance_trust_provider_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_trust_provider_attachment aws_verifiedaccess_instance_trust_provider_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedaccessInstanceTrustProviderAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedaccess_instance_trust_provider_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._verifiedaccessInstanceId = config.verifiedaccessInstanceId;
        this._verifiedaccessTrustProviderId = config.verifiedaccessTrustProviderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get verifiedaccessInstanceId() {
        return this.getStringAttribute('verifiedaccess_instance_id');
    }
    set verifiedaccessInstanceId(value) {
        this._verifiedaccessInstanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get verifiedaccessInstanceIdInput() {
        return this._verifiedaccessInstanceId;
    }
    get verifiedaccessTrustProviderId() {
        return this.getStringAttribute('verifiedaccess_trust_provider_id');
    }
    set verifiedaccessTrustProviderId(value) {
        this._verifiedaccessTrustProviderId = value;
    }
    // Temporarily expose input value. Use with caution.
    get verifiedaccessTrustProviderIdInput() {
        return this._verifiedaccessTrustProviderId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            verifiedaccess_instance_id: cdktf.stringToTerraform(this._verifiedaccessInstanceId),
            verifiedaccess_trust_provider_id: cdktf.stringToTerraform(this._verifiedaccessTrustProviderId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            verifiedaccess_instance_id: {
                value: cdktf.stringToHclTerraform(this._verifiedaccessInstanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            verifiedaccess_trust_provider_id: {
                value: cdktf.stringToHclTerraform(this._verifiedaccessTrustProviderId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedaccessInstanceTrustProviderAttachment = VerifiedaccessInstanceTrustProviderAttachment;
_a = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceTrustProviderAttachment[_a] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceTrustProviderAttachment.VerifiedaccessInstanceTrustProviderAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedaccessInstanceTrustProviderAttachment.tfResourceType = "aws_verifiedaccess_instance_trust_provider_attachment";
