// @cdktf/provider-aws/lib/cloudfront-realtime-log-config/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudfrontRealtimeLogConfig = exports.CloudfrontRealtimeLogConfigEndpointOutputReference = exports.cloudfrontRealtimeLogConfigEndpointToHclTerraform = exports.cloudfrontRealtimeLogConfigEndpointToTerraform = exports.CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference = exports.cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToHclTerraform = exports.cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        stream_arn: cdktf.stringToTerraform(struct.streamArn),
    };
}
exports.cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform = cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform;
function cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        stream_arn: {
            value: cdktf.stringToHclTerraform(struct.streamArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToHclTerraform = cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToHclTerraform;
class CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._streamArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.streamArn = this._streamArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._roleArn = undefined;
            this._streamArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._roleArn = value.roleArn;
            this._streamArn = value.streamArn;
        }
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    set streamArn(value) {
        this._streamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamArnInput() {
        return this._streamArn;
    }
}
exports.CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference = CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudfrontRealtimeLogConfig.CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference", version: "19.15.0" };
function cloudfrontRealtimeLogConfigEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        stream_type: cdktf.stringToTerraform(struct.streamType),
        kinesis_stream_config: cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform(struct.kinesisStreamConfig),
    };
}
exports.cloudfrontRealtimeLogConfigEndpointToTerraform = cloudfrontRealtimeLogConfigEndpointToTerraform;
function cloudfrontRealtimeLogConfigEndpointToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        stream_type: {
            value: cdktf.stringToHclTerraform(struct.streamType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kinesis_stream_config: {
            value: cloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToHclTerraform(struct.kinesisStreamConfig),
            isBlock: true,
            type: "list",
            storageClassType: "CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudfrontRealtimeLogConfigEndpointToHclTerraform = cloudfrontRealtimeLogConfigEndpointToHclTerraform;
class CloudfrontRealtimeLogConfigEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kinesis_stream_config - computed: false, optional: false, required: true
        this._kinesisStreamConfig = new CloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference(this, "kinesis_stream_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._streamType !== undefined) {
            hasAnyValues = true;
            internalValueResult.streamType = this._streamType;
        }
        if (this._kinesisStreamConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisStreamConfig = this._kinesisStreamConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._streamType = undefined;
            this._kinesisStreamConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._streamType = value.streamType;
            this._kinesisStreamConfig.internalValue = value.kinesisStreamConfig;
        }
    }
    get streamType() {
        return this.getStringAttribute('stream_type');
    }
    set streamType(value) {
        this._streamType = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamTypeInput() {
        return this._streamType;
    }
    get kinesisStreamConfig() {
        return this._kinesisStreamConfig;
    }
    putKinesisStreamConfig(value) {
        this._kinesisStreamConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisStreamConfigInput() {
        return this._kinesisStreamConfig.internalValue;
    }
}
exports.CloudfrontRealtimeLogConfigEndpointOutputReference = CloudfrontRealtimeLogConfigEndpointOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudfrontRealtimeLogConfigEndpointOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudfrontRealtimeLogConfig.CloudfrontRealtimeLogConfigEndpointOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudfront_realtime_log_config aws_cloudfront_realtime_log_config}
*/
class CloudfrontRealtimeLogConfig extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudfrontRealtimeLogConfig resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudfrontRealtimeLogConfig to import
    * @param importFromId The id of the existing CloudfrontRealtimeLogConfig that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudfront_realtime_log_config#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudfrontRealtimeLogConfig to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudfront_realtime_log_config", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudfront_realtime_log_config aws_cloudfront_realtime_log_config} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudfrontRealtimeLogConfigConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_realtime_log_config',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint - computed: false, optional: false, required: true
        this._endpoint = new CloudfrontRealtimeLogConfigEndpointOutputReference(this, "endpoint");
        this._fields = config.fields;
        this._id = config.id;
        this._name = config.name;
        this._samplingRate = config.samplingRate;
        this._endpoint.internalValue = config.endpoint;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get fields() {
        return cdktf.Fn.tolist(this.getListAttribute('fields'));
    }
    set fields(value) {
        this._fields = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldsInput() {
        return this._fields;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get samplingRate() {
        return this.getNumberAttribute('sampling_rate');
    }
    set samplingRate(value) {
        this._samplingRate = value;
    }
    // Temporarily expose input value. Use with caution.
    get samplingRateInput() {
        return this._samplingRate;
    }
    get endpoint() {
        return this._endpoint;
    }
    putEndpoint(value) {
        this._endpoint.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointInput() {
        return this._endpoint.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            fields: cdktf.listMapper(cdktf.stringToTerraform, false)(this._fields),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            sampling_rate: cdktf.numberToTerraform(this._samplingRate),
            endpoint: cloudfrontRealtimeLogConfigEndpointToTerraform(this._endpoint.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            fields: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._fields),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            sampling_rate: {
                value: cdktf.numberToHclTerraform(this._samplingRate),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            endpoint: {
                value: cloudfrontRealtimeLogConfigEndpointToHclTerraform(this._endpoint.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudfrontRealtimeLogConfigEndpointList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudfrontRealtimeLogConfig = CloudfrontRealtimeLogConfig;
_c = JSII_RTTI_SYMBOL_1;
CloudfrontRealtimeLogConfig[_c] = { fqn: "@cdktf/provider-aws.cloudfrontRealtimeLogConfig.CloudfrontRealtimeLogConfig", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudfrontRealtimeLogConfig.tfResourceType = "aws_cloudfront_realtime_log_config";
