// @cdktf/provider-aws/lib/data-aws-codecatalyst-dev-environment/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCodecatalystDevEnvironment = exports.DataAwsCodecatalystDevEnvironmentRepositoriesList = exports.DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference = exports.dataAwsCodecatalystDevEnvironmentRepositoriesToHclTerraform = exports.dataAwsCodecatalystDevEnvironmentRepositoriesToTerraform = exports.DataAwsCodecatalystDevEnvironmentPersistentStorageList = exports.DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference = exports.dataAwsCodecatalystDevEnvironmentPersistentStorageToHclTerraform = exports.dataAwsCodecatalystDevEnvironmentPersistentStorageToTerraform = exports.DataAwsCodecatalystDevEnvironmentIdesList = exports.DataAwsCodecatalystDevEnvironmentIdesOutputReference = exports.dataAwsCodecatalystDevEnvironmentIdesToHclTerraform = exports.dataAwsCodecatalystDevEnvironmentIdesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsCodecatalystDevEnvironmentIdesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodecatalystDevEnvironmentIdesToTerraform = dataAwsCodecatalystDevEnvironmentIdesToTerraform;
function dataAwsCodecatalystDevEnvironmentIdesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodecatalystDevEnvironmentIdesToHclTerraform = dataAwsCodecatalystDevEnvironmentIdesToHclTerraform;
class DataAwsCodecatalystDevEnvironmentIdesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // runtime - computed: true, optional: false, required: false
    get runtime() {
        return this.getStringAttribute('runtime');
    }
}
exports.DataAwsCodecatalystDevEnvironmentIdesOutputReference = DataAwsCodecatalystDevEnvironmentIdesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentIdesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentIdesOutputReference", version: "19.15.0" };
class DataAwsCodecatalystDevEnvironmentIdesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodecatalystDevEnvironmentIdesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodecatalystDevEnvironmentIdesList = DataAwsCodecatalystDevEnvironmentIdesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentIdesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentIdesList", version: "19.15.0" };
function dataAwsCodecatalystDevEnvironmentPersistentStorageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodecatalystDevEnvironmentPersistentStorageToTerraform = dataAwsCodecatalystDevEnvironmentPersistentStorageToTerraform;
function dataAwsCodecatalystDevEnvironmentPersistentStorageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodecatalystDevEnvironmentPersistentStorageToHclTerraform = dataAwsCodecatalystDevEnvironmentPersistentStorageToHclTerraform;
class DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
}
exports.DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference = DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference", version: "19.15.0" };
class DataAwsCodecatalystDevEnvironmentPersistentStorageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodecatalystDevEnvironmentPersistentStorageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodecatalystDevEnvironmentPersistentStorageList = DataAwsCodecatalystDevEnvironmentPersistentStorageList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentPersistentStorageList[_d] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentPersistentStorageList", version: "19.15.0" };
function dataAwsCodecatalystDevEnvironmentRepositoriesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodecatalystDevEnvironmentRepositoriesToTerraform = dataAwsCodecatalystDevEnvironmentRepositoriesToTerraform;
function dataAwsCodecatalystDevEnvironmentRepositoriesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodecatalystDevEnvironmentRepositoriesToHclTerraform = dataAwsCodecatalystDevEnvironmentRepositoriesToHclTerraform;
class DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
        }
    }
    // branch_name - computed: true, optional: false, required: false
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    // repository_name - computed: true, optional: false, required: false
    get repositoryName() {
        return this.getStringAttribute('repository_name');
    }
}
exports.DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference = DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference", version: "19.15.0" };
class DataAwsCodecatalystDevEnvironmentRepositoriesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodecatalystDevEnvironmentRepositoriesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodecatalystDevEnvironmentRepositoriesList = DataAwsCodecatalystDevEnvironmentRepositoriesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironmentRepositoriesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironmentRepositoriesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codecatalyst_dev_environment aws_codecatalyst_dev_environment}
*/
class DataAwsCodecatalystDevEnvironment extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsCodecatalystDevEnvironment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsCodecatalystDevEnvironment to import
    * @param importFromId The id of the existing DataAwsCodecatalystDevEnvironment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codecatalyst_dev_environment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsCodecatalystDevEnvironment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_codecatalyst_dev_environment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codecatalyst_dev_environment aws_codecatalyst_dev_environment} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCodecatalystDevEnvironmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codecatalyst_dev_environment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ides - computed: true, optional: false, required: false
        this._ides = new DataAwsCodecatalystDevEnvironmentIdesList(this, "ides", false);
        // persistent_storage - computed: true, optional: false, required: false
        this._persistentStorage = new DataAwsCodecatalystDevEnvironmentPersistentStorageList(this, "persistent_storage", false);
        // repositories - computed: false, optional: true, required: false
        this._repositories = new DataAwsCodecatalystDevEnvironmentRepositoriesList(this, "repositories", false);
        this._alias = config.alias;
        this._creatorId = config.creatorId;
        this._envId = config.envId;
        this._id = config.id;
        this._projectName = config.projectName;
        this._spaceName = config.spaceName;
        this._tags = config.tags;
        this._repositories.internalValue = config.repositories;
    }
    get alias() {
        return this.getStringAttribute('alias');
    }
    set alias(value) {
        this._alias = value;
    }
    resetAlias() {
        this._alias = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    get creatorId() {
        return this.getStringAttribute('creator_id');
    }
    set creatorId(value) {
        this._creatorId = value;
    }
    resetCreatorId() {
        this._creatorId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get creatorIdInput() {
        return this._creatorId;
    }
    get envId() {
        return this.getStringAttribute('env_id');
    }
    set envId(value) {
        this._envId = value;
    }
    // Temporarily expose input value. Use with caution.
    get envIdInput() {
        return this._envId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ides() {
        return this._ides;
    }
    // inactivity_timeout_minutes - computed: true, optional: false, required: false
    get inactivityTimeoutMinutes() {
        return this.getNumberAttribute('inactivity_timeout_minutes');
    }
    // instance_type - computed: true, optional: false, required: false
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // last_updated_time - computed: true, optional: false, required: false
    get lastUpdatedTime() {
        return this.getStringAttribute('last_updated_time');
    }
    get persistentStorage() {
        return this._persistentStorage;
    }
    get projectName() {
        return this.getStringAttribute('project_name');
    }
    set projectName(value) {
        this._projectName = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectNameInput() {
        return this._projectName;
    }
    get spaceName() {
        return this.getStringAttribute('space_name');
    }
    set spaceName(value) {
        this._spaceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get spaceNameInput() {
        return this._spaceName;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_reason - computed: true, optional: false, required: false
    get statusReason() {
        return this.getStringAttribute('status_reason');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get repositories() {
        return this._repositories;
    }
    putRepositories(value) {
        this._repositories.internalValue = value;
    }
    resetRepositories() {
        this._repositories.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoriesInput() {
        return this._repositories.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            creator_id: cdktf.stringToTerraform(this._creatorId),
            env_id: cdktf.stringToTerraform(this._envId),
            id: cdktf.stringToTerraform(this._id),
            project_name: cdktf.stringToTerraform(this._projectName),
            space_name: cdktf.stringToTerraform(this._spaceName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            repositories: cdktf.listMapper(dataAwsCodecatalystDevEnvironmentRepositoriesToTerraform, true)(this._repositories.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            alias: {
                value: cdktf.stringToHclTerraform(this._alias),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            creator_id: {
                value: cdktf.stringToHclTerraform(this._creatorId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            env_id: {
                value: cdktf.stringToHclTerraform(this._envId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project_name: {
                value: cdktf.stringToHclTerraform(this._projectName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            space_name: {
                value: cdktf.stringToHclTerraform(this._spaceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            repositories: {
                value: cdktf.listMapperHcl(dataAwsCodecatalystDevEnvironmentRepositoriesToHclTerraform, true)(this._repositories.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsCodecatalystDevEnvironmentRepositoriesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsCodecatalystDevEnvironment = DataAwsCodecatalystDevEnvironment;
_g = JSII_RTTI_SYMBOL_1;
DataAwsCodecatalystDevEnvironment[_g] = { fqn: "@cdktf/provider-aws.dataAwsCodecatalystDevEnvironment.DataAwsCodecatalystDevEnvironment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCodecatalystDevEnvironment.tfResourceType = "aws_codecatalyst_dev_environment";
