// @cdktf/provider-aws/lib/data-aws-lambda-code-signing-config/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLambdaCodeSigningConfig = exports.DataAwsLambdaCodeSigningConfigPoliciesList = exports.DataAwsLambdaCodeSigningConfigPoliciesOutputReference = exports.dataAwsLambdaCodeSigningConfigPoliciesToHclTerraform = exports.dataAwsLambdaCodeSigningConfigPoliciesToTerraform = exports.DataAwsLambdaCodeSigningConfigAllowedPublishersList = exports.DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference = exports.dataAwsLambdaCodeSigningConfigAllowedPublishersToHclTerraform = exports.dataAwsLambdaCodeSigningConfigAllowedPublishersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLambdaCodeSigningConfigAllowedPublishersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaCodeSigningConfigAllowedPublishersToTerraform = dataAwsLambdaCodeSigningConfigAllowedPublishersToTerraform;
function dataAwsLambdaCodeSigningConfigAllowedPublishersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaCodeSigningConfigAllowedPublishersToHclTerraform = dataAwsLambdaCodeSigningConfigAllowedPublishersToHclTerraform;
class DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // signing_profile_version_arns - computed: true, optional: false, required: false
    get signingProfileVersionArns() {
        return cdktf.Fn.tolist(this.getListAttribute('signing_profile_version_arns'));
    }
}
exports.DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference = DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLambdaCodeSigningConfig.DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference", version: "19.15.0" };
class DataAwsLambdaCodeSigningConfigAllowedPublishersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaCodeSigningConfigAllowedPublishersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaCodeSigningConfigAllowedPublishersList = DataAwsLambdaCodeSigningConfigAllowedPublishersList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigAllowedPublishersList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLambdaCodeSigningConfig.DataAwsLambdaCodeSigningConfigAllowedPublishersList", version: "19.15.0" };
function dataAwsLambdaCodeSigningConfigPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLambdaCodeSigningConfigPoliciesToTerraform = dataAwsLambdaCodeSigningConfigPoliciesToTerraform;
function dataAwsLambdaCodeSigningConfigPoliciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsLambdaCodeSigningConfigPoliciesToHclTerraform = dataAwsLambdaCodeSigningConfigPoliciesToHclTerraform;
class DataAwsLambdaCodeSigningConfigPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // untrusted_artifact_on_deployment - computed: true, optional: false, required: false
    get untrustedArtifactOnDeployment() {
        return this.getStringAttribute('untrusted_artifact_on_deployment');
    }
}
exports.DataAwsLambdaCodeSigningConfigPoliciesOutputReference = DataAwsLambdaCodeSigningConfigPoliciesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigPoliciesOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsLambdaCodeSigningConfig.DataAwsLambdaCodeSigningConfigPoliciesOutputReference", version: "19.15.0" };
class DataAwsLambdaCodeSigningConfigPoliciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLambdaCodeSigningConfigPoliciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLambdaCodeSigningConfigPoliciesList = DataAwsLambdaCodeSigningConfigPoliciesList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigPoliciesList[_d] = { fqn: "@cdktf/provider-aws.dataAwsLambdaCodeSigningConfig.DataAwsLambdaCodeSigningConfigPoliciesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_code_signing_config aws_lambda_code_signing_config}
*/
class DataAwsLambdaCodeSigningConfig extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsLambdaCodeSigningConfig resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsLambdaCodeSigningConfig to import
    * @param importFromId The id of the existing DataAwsLambdaCodeSigningConfig that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_code_signing_config#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsLambdaCodeSigningConfig to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lambda_code_signing_config", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/lambda_code_signing_config aws_lambda_code_signing_config} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLambdaCodeSigningConfigConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_code_signing_config',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // allowed_publishers - computed: true, optional: false, required: false
        this._allowedPublishers = new DataAwsLambdaCodeSigningConfigAllowedPublishersList(this, "allowed_publishers", false);
        // policies - computed: true, optional: false, required: false
        this._policies = new DataAwsLambdaCodeSigningConfigPoliciesList(this, "policies", false);
        this._arn = config.arn;
        this._id = config.id;
    }
    get allowedPublishers() {
        return this._allowedPublishers;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // config_id - computed: true, optional: false, required: false
    get configId() {
        return this.getStringAttribute('config_id');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_modified - computed: true, optional: false, required: false
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    get policies() {
        return this._policies;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            arn: {
                value: cdktf.stringToHclTerraform(this._arn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsLambdaCodeSigningConfig = DataAwsLambdaCodeSigningConfig;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfig[_e] = { fqn: "@cdktf/provider-aws.dataAwsLambdaCodeSigningConfig.DataAwsLambdaCodeSigningConfig", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLambdaCodeSigningConfig.tfResourceType = "aws_lambda_code_signing_config";
