// @cdktf/provider-aws/lib/secretsmanager-secret-version/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsmanagerSecretVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/secretsmanager_secret_version aws_secretsmanager_secret_version}
*/
class SecretsmanagerSecretVersion extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecretsmanagerSecretVersion resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecretsmanagerSecretVersion to import
    * @param importFromId The id of the existing SecretsmanagerSecretVersion that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/secretsmanager_secret_version#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecretsmanagerSecretVersion to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_secretsmanager_secret_version", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/secretsmanager_secret_version aws_secretsmanager_secret_version} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecretsmanagerSecretVersionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_version',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._secretBinary = config.secretBinary;
        this._secretId = config.secretId;
        this._secretString = config.secretString;
        this._versionStages = config.versionStages;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get secretBinary() {
        return this.getStringAttribute('secret_binary');
    }
    set secretBinary(value) {
        this._secretBinary = value;
    }
    resetSecretBinary() {
        this._secretBinary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretBinaryInput() {
        return this._secretBinary;
    }
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretIdInput() {
        return this._secretId;
    }
    get secretString() {
        return this.getStringAttribute('secret_string');
    }
    set secretString(value) {
        this._secretString = value;
    }
    resetSecretString() {
        this._secretString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretStringInput() {
        return this._secretString;
    }
    // version_id - computed: true, optional: false, required: false
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    get versionStages() {
        return cdktf.Fn.tolist(this.getListAttribute('version_stages'));
    }
    set versionStages(value) {
        this._versionStages = value;
    }
    resetVersionStages() {
        this._versionStages = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionStagesInput() {
        return this._versionStages;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            secret_binary: cdktf.stringToTerraform(this._secretBinary),
            secret_id: cdktf.stringToTerraform(this._secretId),
            secret_string: cdktf.stringToTerraform(this._secretString),
            version_stages: cdktf.listMapper(cdktf.stringToTerraform, false)(this._versionStages),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secret_binary: {
                value: cdktf.stringToHclTerraform(this._secretBinary),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secret_id: {
                value: cdktf.stringToHclTerraform(this._secretId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secret_string: {
                value: cdktf.stringToHclTerraform(this._secretString),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version_stages: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._versionStages),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecretsmanagerSecretVersion = SecretsmanagerSecretVersion;
_a = JSII_RTTI_SYMBOL_1;
SecretsmanagerSecretVersion[_a] = { fqn: "@cdktf/provider-aws.secretsmanagerSecretVersion.SecretsmanagerSecretVersion", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecretsmanagerSecretVersion.tfResourceType = "aws_secretsmanager_secret_version";
