// @cdktf/provider-aws/lib/globalaccelerator-endpoint-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalacceleratorEndpointGroup = exports.GlobalacceleratorEndpointGroupTimeoutsOutputReference = exports.globalacceleratorEndpointGroupTimeoutsToHclTerraform = exports.globalacceleratorEndpointGroupTimeoutsToTerraform = exports.GlobalacceleratorEndpointGroupPortOverrideList = exports.GlobalacceleratorEndpointGroupPortOverrideOutputReference = exports.globalacceleratorEndpointGroupPortOverrideToHclTerraform = exports.globalacceleratorEndpointGroupPortOverrideToTerraform = exports.GlobalacceleratorEndpointGroupEndpointConfigurationList = exports.GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference = exports.globalacceleratorEndpointGroupEndpointConfigurationToHclTerraform = exports.globalacceleratorEndpointGroupEndpointConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function globalacceleratorEndpointGroupEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_ip_preservation_enabled: cdktf.booleanToTerraform(struct.clientIpPreservationEnabled),
        endpoint_id: cdktf.stringToTerraform(struct.endpointId),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.globalacceleratorEndpointGroupEndpointConfigurationToTerraform = globalacceleratorEndpointGroupEndpointConfigurationToTerraform;
function globalacceleratorEndpointGroupEndpointConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        client_ip_preservation_enabled: {
            value: cdktf.booleanToHclTerraform(struct.clientIpPreservationEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        endpoint_id: {
            value: cdktf.stringToHclTerraform(struct.endpointId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        weight: {
            value: cdktf.numberToHclTerraform(struct.weight),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.globalacceleratorEndpointGroupEndpointConfigurationToHclTerraform = globalacceleratorEndpointGroupEndpointConfigurationToHclTerraform;
class GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientIpPreservationEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientIpPreservationEnabled = this._clientIpPreservationEnabled;
        }
        if (this._endpointId !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointId = this._endpointId;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._clientIpPreservationEnabled = undefined;
            this._endpointId = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._clientIpPreservationEnabled = value.clientIpPreservationEnabled;
            this._endpointId = value.endpointId;
            this._weight = value.weight;
        }
    }
    get clientIpPreservationEnabled() {
        return this.getBooleanAttribute('client_ip_preservation_enabled');
    }
    set clientIpPreservationEnabled(value) {
        this._clientIpPreservationEnabled = value;
    }
    resetClientIpPreservationEnabled() {
        this._clientIpPreservationEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientIpPreservationEnabledInput() {
        return this._clientIpPreservationEnabled;
    }
    get endpointId() {
        return this.getStringAttribute('endpoint_id');
    }
    set endpointId(value) {
        this._endpointId = value;
    }
    resetEndpointId() {
        this._endpointId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointIdInput() {
        return this._endpointId;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference = GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference", version: "19.15.0" };
class GlobalacceleratorEndpointGroupEndpointConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlobalacceleratorEndpointGroupEndpointConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlobalacceleratorEndpointGroupEndpointConfigurationList = GlobalacceleratorEndpointGroupEndpointConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupEndpointConfigurationList[_b] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroupEndpointConfigurationList", version: "19.15.0" };
function globalacceleratorEndpointGroupPortOverrideToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_port: cdktf.numberToTerraform(struct.endpointPort),
        listener_port: cdktf.numberToTerraform(struct.listenerPort),
    };
}
exports.globalacceleratorEndpointGroupPortOverrideToTerraform = globalacceleratorEndpointGroupPortOverrideToTerraform;
function globalacceleratorEndpointGroupPortOverrideToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        endpoint_port: {
            value: cdktf.numberToHclTerraform(struct.endpointPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        listener_port: {
            value: cdktf.numberToHclTerraform(struct.listenerPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.globalacceleratorEndpointGroupPortOverrideToHclTerraform = globalacceleratorEndpointGroupPortOverrideToHclTerraform;
class GlobalacceleratorEndpointGroupPortOverrideOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointPort = this._endpointPort;
        }
        if (this._listenerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.listenerPort = this._listenerPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._endpointPort = undefined;
            this._listenerPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._endpointPort = value.endpointPort;
            this._listenerPort = value.listenerPort;
        }
    }
    get endpointPort() {
        return this.getNumberAttribute('endpoint_port');
    }
    set endpointPort(value) {
        this._endpointPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointPortInput() {
        return this._endpointPort;
    }
    get listenerPort() {
        return this.getNumberAttribute('listener_port');
    }
    set listenerPort(value) {
        this._listenerPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get listenerPortInput() {
        return this._listenerPort;
    }
}
exports.GlobalacceleratorEndpointGroupPortOverrideOutputReference = GlobalacceleratorEndpointGroupPortOverrideOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupPortOverrideOutputReference[_c] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroupPortOverrideOutputReference", version: "19.15.0" };
class GlobalacceleratorEndpointGroupPortOverrideList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlobalacceleratorEndpointGroupPortOverrideOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlobalacceleratorEndpointGroupPortOverrideList = GlobalacceleratorEndpointGroupPortOverrideList;
_d = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupPortOverrideList[_d] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroupPortOverrideList", version: "19.15.0" };
function globalacceleratorEndpointGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.globalacceleratorEndpointGroupTimeoutsToTerraform = globalacceleratorEndpointGroupTimeoutsToTerraform;
function globalacceleratorEndpointGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.globalacceleratorEndpointGroupTimeoutsToHclTerraform = globalacceleratorEndpointGroupTimeoutsToHclTerraform;
class GlobalacceleratorEndpointGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GlobalacceleratorEndpointGroupTimeoutsOutputReference = GlobalacceleratorEndpointGroupTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroupTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/globalaccelerator_endpoint_group aws_globalaccelerator_endpoint_group}
*/
class GlobalacceleratorEndpointGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GlobalacceleratorEndpointGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GlobalacceleratorEndpointGroup to import
    * @param importFromId The id of the existing GlobalacceleratorEndpointGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/globalaccelerator_endpoint_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GlobalacceleratorEndpointGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_globalaccelerator_endpoint_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/globalaccelerator_endpoint_group aws_globalaccelerator_endpoint_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlobalacceleratorEndpointGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_endpoint_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint_configuration - computed: false, optional: true, required: false
        this._endpointConfiguration = new GlobalacceleratorEndpointGroupEndpointConfigurationList(this, "endpoint_configuration", true);
        // port_override - computed: false, optional: true, required: false
        this._portOverride = new GlobalacceleratorEndpointGroupPortOverrideList(this, "port_override", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GlobalacceleratorEndpointGroupTimeoutsOutputReference(this, "timeouts");
        this._endpointGroupRegion = config.endpointGroupRegion;
        this._healthCheckIntervalSeconds = config.healthCheckIntervalSeconds;
        this._healthCheckPath = config.healthCheckPath;
        this._healthCheckPort = config.healthCheckPort;
        this._healthCheckProtocol = config.healthCheckProtocol;
        this._id = config.id;
        this._listenerArn = config.listenerArn;
        this._thresholdCount = config.thresholdCount;
        this._trafficDialPercentage = config.trafficDialPercentage;
        this._endpointConfiguration.internalValue = config.endpointConfiguration;
        this._portOverride.internalValue = config.portOverride;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get endpointGroupRegion() {
        return this.getStringAttribute('endpoint_group_region');
    }
    set endpointGroupRegion(value) {
        this._endpointGroupRegion = value;
    }
    resetEndpointGroupRegion() {
        this._endpointGroupRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointGroupRegionInput() {
        return this._endpointGroupRegion;
    }
    get healthCheckIntervalSeconds() {
        return this.getNumberAttribute('health_check_interval_seconds');
    }
    set healthCheckIntervalSeconds(value) {
        this._healthCheckIntervalSeconds = value;
    }
    resetHealthCheckIntervalSeconds() {
        this._healthCheckIntervalSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckIntervalSecondsInput() {
        return this._healthCheckIntervalSeconds;
    }
    get healthCheckPath() {
        return this.getStringAttribute('health_check_path');
    }
    set healthCheckPath(value) {
        this._healthCheckPath = value;
    }
    resetHealthCheckPath() {
        this._healthCheckPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckPathInput() {
        return this._healthCheckPath;
    }
    get healthCheckPort() {
        return this.getNumberAttribute('health_check_port');
    }
    set healthCheckPort(value) {
        this._healthCheckPort = value;
    }
    resetHealthCheckPort() {
        this._healthCheckPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckPortInput() {
        return this._healthCheckPort;
    }
    get healthCheckProtocol() {
        return this.getStringAttribute('health_check_protocol');
    }
    set healthCheckProtocol(value) {
        this._healthCheckProtocol = value;
    }
    resetHealthCheckProtocol() {
        this._healthCheckProtocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckProtocolInput() {
        return this._healthCheckProtocol;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get listenerArn() {
        return this.getStringAttribute('listener_arn');
    }
    set listenerArn(value) {
        this._listenerArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get listenerArnInput() {
        return this._listenerArn;
    }
    get thresholdCount() {
        return this.getNumberAttribute('threshold_count');
    }
    set thresholdCount(value) {
        this._thresholdCount = value;
    }
    resetThresholdCount() {
        this._thresholdCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thresholdCountInput() {
        return this._thresholdCount;
    }
    get trafficDialPercentage() {
        return this.getNumberAttribute('traffic_dial_percentage');
    }
    set trafficDialPercentage(value) {
        this._trafficDialPercentage = value;
    }
    resetTrafficDialPercentage() {
        this._trafficDialPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trafficDialPercentageInput() {
        return this._trafficDialPercentage;
    }
    get endpointConfiguration() {
        return this._endpointConfiguration;
    }
    putEndpointConfiguration(value) {
        this._endpointConfiguration.internalValue = value;
    }
    resetEndpointConfiguration() {
        this._endpointConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointConfigurationInput() {
        return this._endpointConfiguration.internalValue;
    }
    get portOverride() {
        return this._portOverride;
    }
    putPortOverride(value) {
        this._portOverride.internalValue = value;
    }
    resetPortOverride() {
        this._portOverride.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portOverrideInput() {
        return this._portOverride.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            endpoint_group_region: cdktf.stringToTerraform(this._endpointGroupRegion),
            health_check_interval_seconds: cdktf.numberToTerraform(this._healthCheckIntervalSeconds),
            health_check_path: cdktf.stringToTerraform(this._healthCheckPath),
            health_check_port: cdktf.numberToTerraform(this._healthCheckPort),
            health_check_protocol: cdktf.stringToTerraform(this._healthCheckProtocol),
            id: cdktf.stringToTerraform(this._id),
            listener_arn: cdktf.stringToTerraform(this._listenerArn),
            threshold_count: cdktf.numberToTerraform(this._thresholdCount),
            traffic_dial_percentage: cdktf.numberToTerraform(this._trafficDialPercentage),
            endpoint_configuration: cdktf.listMapper(globalacceleratorEndpointGroupEndpointConfigurationToTerraform, true)(this._endpointConfiguration.internalValue),
            port_override: cdktf.listMapper(globalacceleratorEndpointGroupPortOverrideToTerraform, true)(this._portOverride.internalValue),
            timeouts: globalacceleratorEndpointGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            endpoint_group_region: {
                value: cdktf.stringToHclTerraform(this._endpointGroupRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            health_check_interval_seconds: {
                value: cdktf.numberToHclTerraform(this._healthCheckIntervalSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            health_check_path: {
                value: cdktf.stringToHclTerraform(this._healthCheckPath),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            health_check_port: {
                value: cdktf.numberToHclTerraform(this._healthCheckPort),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            health_check_protocol: {
                value: cdktf.stringToHclTerraform(this._healthCheckProtocol),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            listener_arn: {
                value: cdktf.stringToHclTerraform(this._listenerArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            threshold_count: {
                value: cdktf.numberToHclTerraform(this._thresholdCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            traffic_dial_percentage: {
                value: cdktf.numberToHclTerraform(this._trafficDialPercentage),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            endpoint_configuration: {
                value: cdktf.listMapperHcl(globalacceleratorEndpointGroupEndpointConfigurationToHclTerraform, true)(this._endpointConfiguration.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "GlobalacceleratorEndpointGroupEndpointConfigurationList",
            },
            port_override: {
                value: cdktf.listMapperHcl(globalacceleratorEndpointGroupPortOverrideToHclTerraform, true)(this._portOverride.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "GlobalacceleratorEndpointGroupPortOverrideList",
            },
            timeouts: {
                value: globalacceleratorEndpointGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "GlobalacceleratorEndpointGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GlobalacceleratorEndpointGroup = GlobalacceleratorEndpointGroup;
_f = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroup[_f] = { fqn: "@cdktf/provider-aws.globalacceleratorEndpointGroup.GlobalacceleratorEndpointGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
GlobalacceleratorEndpointGroup.tfResourceType = "aws_globalaccelerator_endpoint_group";
