// @cdktf/provider-aws/lib/default-security-group/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultSecurityGroup = exports.DefaultSecurityGroupIngressList = exports.DefaultSecurityGroupIngressOutputReference = exports.defaultSecurityGroupIngressToHclTerraform = exports.defaultSecurityGroupIngressToTerraform = exports.DefaultSecurityGroupEgressList = exports.DefaultSecurityGroupEgressOutputReference = exports.defaultSecurityGroupEgressToHclTerraform = exports.defaultSecurityGroupEgressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function defaultSecurityGroupEgressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cidr_blocks: struct.cidrBlocks === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.cidrBlocks),
        description: struct.description === undefined ? null : cdktf.stringToTerraform(struct.description),
        from_port: struct.fromPort === undefined ? null : cdktf.numberToTerraform(struct.fromPort),
        ipv6_cidr_blocks: struct.ipv6CidrBlocks === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.ipv6CidrBlocks),
        prefix_list_ids: struct.prefixListIds === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.prefixListIds),
        protocol: struct.protocol === undefined ? null : cdktf.stringToTerraform(struct.protocol),
        security_groups: struct.securityGroups === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        self: struct.selfAttribute === undefined ? null : cdktf.booleanToTerraform(struct.selfAttribute),
        to_port: struct.toPort === undefined ? null : cdktf.numberToTerraform(struct.toPort),
    };
}
exports.defaultSecurityGroupEgressToTerraform = defaultSecurityGroupEgressToTerraform;
function defaultSecurityGroupEgressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cidr_blocks: {
            value: struct.cidrBlocks === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.cidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        description: {
            value: struct.description === undefined ? null : cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        from_port: {
            value: struct.fromPort === undefined ? null : cdktf.numberToHclTerraform(struct.fromPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        ipv6_cidr_blocks: {
            value: struct.ipv6CidrBlocks === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.ipv6CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        prefix_list_ids: {
            value: struct.prefixListIds === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.prefixListIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        protocol: {
            value: struct.protocol === undefined ? null : cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        security_groups: {
            value: struct.securityGroups === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        self: {
            value: struct.selfAttribute === undefined ? null : cdktf.booleanToHclTerraform(struct.selfAttribute),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        to_port: {
            value: struct.toPort === undefined ? null : cdktf.numberToHclTerraform(struct.toPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.defaultSecurityGroupEgressToHclTerraform = defaultSecurityGroupEgressToHclTerraform;
class DefaultSecurityGroupEgressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.cidrBlocks = this._cidrBlocks;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._ipv6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6CidrBlocks = this._ipv6CidrBlocks;
        }
        if (this._prefixListIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefixListIds = this._prefixListIds;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._self !== undefined) {
            hasAnyValues = true;
            internalValueResult.selfAttribute = this._self;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cidrBlocks = undefined;
            this._description = undefined;
            this._fromPort = undefined;
            this._ipv6CidrBlocks = undefined;
            this._prefixListIds = undefined;
            this._protocol = undefined;
            this._securityGroups = undefined;
            this._self = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cidrBlocks = value.cidrBlocks;
            this._description = value.description;
            this._fromPort = value.fromPort;
            this._ipv6CidrBlocks = value.ipv6CidrBlocks;
            this._prefixListIds = value.prefixListIds;
            this._protocol = value.protocol;
            this._securityGroups = value.securityGroups;
            this._self = value.selfAttribute;
            this._toPort = value.toPort;
        }
    }
    get cidrBlocks() {
        return this.getListAttribute('cidr_blocks');
    }
    set cidrBlocks(value) {
        this._cidrBlocks = value;
    }
    resetCidrBlocks() {
        this._cidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlocksInput() {
        return this._cidrBlocks;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get ipv6CidrBlocks() {
        return this.getListAttribute('ipv6_cidr_blocks');
    }
    set ipv6CidrBlocks(value) {
        this._ipv6CidrBlocks = value;
    }
    resetIpv6CidrBlocks() {
        this._ipv6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6CidrBlocksInput() {
        return this._ipv6CidrBlocks;
    }
    get prefixListIds() {
        return this.getListAttribute('prefix_list_ids');
    }
    set prefixListIds(value) {
        this._prefixListIds = value;
    }
    resetPrefixListIds() {
        this._prefixListIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixListIdsInput() {
        return this._prefixListIds;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get selfAttribute() {
        return this.getBooleanAttribute('self');
    }
    set selfAttribute(value) {
        this._self = value;
    }
    resetSelfAttribute() {
        this._self = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfAttributeInput() {
        return this._self;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.DefaultSecurityGroupEgressOutputReference = DefaultSecurityGroupEgressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DefaultSecurityGroupEgressOutputReference[_a] = { fqn: "@cdktf/provider-aws.defaultSecurityGroup.DefaultSecurityGroupEgressOutputReference", version: "19.15.0" };
class DefaultSecurityGroupEgressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DefaultSecurityGroupEgressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DefaultSecurityGroupEgressList = DefaultSecurityGroupEgressList;
_b = JSII_RTTI_SYMBOL_1;
DefaultSecurityGroupEgressList[_b] = { fqn: "@cdktf/provider-aws.defaultSecurityGroup.DefaultSecurityGroupEgressList", version: "19.15.0" };
function defaultSecurityGroupIngressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cidr_blocks: struct.cidrBlocks === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.cidrBlocks),
        description: struct.description === undefined ? null : cdktf.stringToTerraform(struct.description),
        from_port: struct.fromPort === undefined ? null : cdktf.numberToTerraform(struct.fromPort),
        ipv6_cidr_blocks: struct.ipv6CidrBlocks === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.ipv6CidrBlocks),
        prefix_list_ids: struct.prefixListIds === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.prefixListIds),
        protocol: struct.protocol === undefined ? null : cdktf.stringToTerraform(struct.protocol),
        security_groups: struct.securityGroups === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        self: struct.selfAttribute === undefined ? null : cdktf.booleanToTerraform(struct.selfAttribute),
        to_port: struct.toPort === undefined ? null : cdktf.numberToTerraform(struct.toPort),
    };
}
exports.defaultSecurityGroupIngressToTerraform = defaultSecurityGroupIngressToTerraform;
function defaultSecurityGroupIngressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cidr_blocks: {
            value: struct.cidrBlocks === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.cidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        description: {
            value: struct.description === undefined ? null : cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        from_port: {
            value: struct.fromPort === undefined ? null : cdktf.numberToHclTerraform(struct.fromPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        ipv6_cidr_blocks: {
            value: struct.ipv6CidrBlocks === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.ipv6CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        prefix_list_ids: {
            value: struct.prefixListIds === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.prefixListIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        protocol: {
            value: struct.protocol === undefined ? null : cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        security_groups: {
            value: struct.securityGroups === undefined ? null : cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroups),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        self: {
            value: struct.selfAttribute === undefined ? null : cdktf.booleanToHclTerraform(struct.selfAttribute),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        to_port: {
            value: struct.toPort === undefined ? null : cdktf.numberToHclTerraform(struct.toPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.defaultSecurityGroupIngressToHclTerraform = defaultSecurityGroupIngressToHclTerraform;
class DefaultSecurityGroupIngressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.cidrBlocks = this._cidrBlocks;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._ipv6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6CidrBlocks = this._ipv6CidrBlocks;
        }
        if (this._prefixListIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefixListIds = this._prefixListIds;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._self !== undefined) {
            hasAnyValues = true;
            internalValueResult.selfAttribute = this._self;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cidrBlocks = undefined;
            this._description = undefined;
            this._fromPort = undefined;
            this._ipv6CidrBlocks = undefined;
            this._prefixListIds = undefined;
            this._protocol = undefined;
            this._securityGroups = undefined;
            this._self = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cidrBlocks = value.cidrBlocks;
            this._description = value.description;
            this._fromPort = value.fromPort;
            this._ipv6CidrBlocks = value.ipv6CidrBlocks;
            this._prefixListIds = value.prefixListIds;
            this._protocol = value.protocol;
            this._securityGroups = value.securityGroups;
            this._self = value.selfAttribute;
            this._toPort = value.toPort;
        }
    }
    get cidrBlocks() {
        return this.getListAttribute('cidr_blocks');
    }
    set cidrBlocks(value) {
        this._cidrBlocks = value;
    }
    resetCidrBlocks() {
        this._cidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrBlocksInput() {
        return this._cidrBlocks;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get ipv6CidrBlocks() {
        return this.getListAttribute('ipv6_cidr_blocks');
    }
    set ipv6CidrBlocks(value) {
        this._ipv6CidrBlocks = value;
    }
    resetIpv6CidrBlocks() {
        this._ipv6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6CidrBlocksInput() {
        return this._ipv6CidrBlocks;
    }
    get prefixListIds() {
        return this.getListAttribute('prefix_list_ids');
    }
    set prefixListIds(value) {
        this._prefixListIds = value;
    }
    resetPrefixListIds() {
        this._prefixListIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixListIdsInput() {
        return this._prefixListIds;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get selfAttribute() {
        return this.getBooleanAttribute('self');
    }
    set selfAttribute(value) {
        this._self = value;
    }
    resetSelfAttribute() {
        this._self = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfAttributeInput() {
        return this._self;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.DefaultSecurityGroupIngressOutputReference = DefaultSecurityGroupIngressOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DefaultSecurityGroupIngressOutputReference[_c] = { fqn: "@cdktf/provider-aws.defaultSecurityGroup.DefaultSecurityGroupIngressOutputReference", version: "19.15.0" };
class DefaultSecurityGroupIngressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DefaultSecurityGroupIngressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DefaultSecurityGroupIngressList = DefaultSecurityGroupIngressList;
_d = JSII_RTTI_SYMBOL_1;
DefaultSecurityGroupIngressList[_d] = { fqn: "@cdktf/provider-aws.defaultSecurityGroup.DefaultSecurityGroupIngressList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_security_group aws_default_security_group}
*/
class DefaultSecurityGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DefaultSecurityGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DefaultSecurityGroup to import
    * @param importFromId The id of the existing DefaultSecurityGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_security_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DefaultSecurityGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_default_security_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/default_security_group aws_default_security_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DefaultSecurityGroupConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_default_security_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // egress - computed: true, optional: true, required: false
        this._egress = new DefaultSecurityGroupEgressList(this, "egress", true);
        // ingress - computed: true, optional: true, required: false
        this._ingress = new DefaultSecurityGroupIngressList(this, "ingress", true);
        this._egress.internalValue = config.egress;
        this._id = config.id;
        this._ingress.internalValue = config.ingress;
        this._revokeRulesOnDelete = config.revokeRulesOnDelete;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcId = config.vpcId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get egress() {
        return this._egress;
    }
    putEgress(value) {
        this._egress.internalValue = value;
    }
    resetEgress() {
        this._egress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress.internalValue;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ingress() {
        return this._ingress;
    }
    putIngress(value) {
        this._ingress.internalValue = value;
    }
    resetIngress() {
        this._ingress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressInput() {
        return this._ingress.internalValue;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // name_prefix - computed: true, optional: false, required: false
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get revokeRulesOnDelete() {
        return this.getBooleanAttribute('revoke_rules_on_delete');
    }
    set revokeRulesOnDelete(value) {
        this._revokeRulesOnDelete = value;
    }
    resetRevokeRulesOnDelete() {
        this._revokeRulesOnDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revokeRulesOnDeleteInput() {
        return this._revokeRulesOnDelete;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            egress: cdktf.listMapper(defaultSecurityGroupEgressToTerraform, false)(this._egress.internalValue),
            id: cdktf.stringToTerraform(this._id),
            ingress: cdktf.listMapper(defaultSecurityGroupIngressToTerraform, false)(this._ingress.internalValue),
            revoke_rules_on_delete: cdktf.booleanToTerraform(this._revokeRulesOnDelete),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            egress: {
                value: cdktf.listMapperHcl(defaultSecurityGroupEgressToHclTerraform, false)(this._egress.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DefaultSecurityGroupEgressList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ingress: {
                value: cdktf.listMapperHcl(defaultSecurityGroupIngressToHclTerraform, false)(this._ingress.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "DefaultSecurityGroupIngressList",
            },
            revoke_rules_on_delete: {
                value: cdktf.booleanToHclTerraform(this._revokeRulesOnDelete),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            vpc_id: {
                value: cdktf.stringToHclTerraform(this._vpcId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DefaultSecurityGroup = DefaultSecurityGroup;
_e = JSII_RTTI_SYMBOL_1;
DefaultSecurityGroup[_e] = { fqn: "@cdktf/provider-aws.defaultSecurityGroup.DefaultSecurityGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DefaultSecurityGroup.tfResourceType = "aws_default_security_group";
