// @cdktf/provider-aws/lib/securityhub-finding-aggregator/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubFindingAggregator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_finding_aggregator aws_securityhub_finding_aggregator}
*/
class SecurityhubFindingAggregator extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SecurityhubFindingAggregator resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SecurityhubFindingAggregator to import
    * @param importFromId The id of the existing SecurityhubFindingAggregator that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_finding_aggregator#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SecurityhubFindingAggregator to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_securityhub_finding_aggregator", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/securityhub_finding_aggregator aws_securityhub_finding_aggregator} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubFindingAggregatorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_finding_aggregator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._linkingMode = config.linkingMode;
        this._specifiedRegions = config.specifiedRegions;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get linkingMode() {
        return this.getStringAttribute('linking_mode');
    }
    set linkingMode(value) {
        this._linkingMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get linkingModeInput() {
        return this._linkingMode;
    }
    get specifiedRegions() {
        return cdktf.Fn.tolist(this.getListAttribute('specified_regions'));
    }
    set specifiedRegions(value) {
        this._specifiedRegions = value;
    }
    resetSpecifiedRegions() {
        this._specifiedRegions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get specifiedRegionsInput() {
        return this._specifiedRegions;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            linking_mode: cdktf.stringToTerraform(this._linkingMode),
            specified_regions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._specifiedRegions),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            linking_mode: {
                value: cdktf.stringToHclTerraform(this._linkingMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            specified_regions: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._specifiedRegions),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SecurityhubFindingAggregator = SecurityhubFindingAggregator;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubFindingAggregator[_a] = { fqn: "@cdktf/provider-aws.securityhubFindingAggregator.SecurityhubFindingAggregator", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubFindingAggregator.tfResourceType = "aws_securityhub_finding_aggregator";
