// @cdktf/provider-aws/lib/emr-instance-group/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmrInstanceGroup = exports.EmrInstanceGroupEbsConfigList = exports.EmrInstanceGroupEbsConfigOutputReference = exports.emrInstanceGroupEbsConfigToHclTerraform = exports.emrInstanceGroupEbsConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function emrInstanceGroupEbsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        size: cdktf.numberToTerraform(struct.size),
        type: cdktf.stringToTerraform(struct.type),
        volumes_per_instance: cdktf.numberToTerraform(struct.volumesPerInstance),
    };
}
exports.emrInstanceGroupEbsConfigToTerraform = emrInstanceGroupEbsConfigToTerraform;
function emrInstanceGroupEbsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iops: {
            value: cdktf.numberToHclTerraform(struct.iops),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        size: {
            value: cdktf.numberToHclTerraform(struct.size),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        volumes_per_instance: {
            value: cdktf.numberToHclTerraform(struct.volumesPerInstance),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.emrInstanceGroupEbsConfigToHclTerraform = emrInstanceGroupEbsConfigToHclTerraform;
class EmrInstanceGroupEbsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._volumesPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumesPerInstance = this._volumesPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iops = undefined;
            this._size = undefined;
            this._type = undefined;
            this._volumesPerInstance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iops = value.iops;
            this._size = value.size;
            this._type = value.type;
            this._volumesPerInstance = value.volumesPerInstance;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get volumesPerInstance() {
        return this.getNumberAttribute('volumes_per_instance');
    }
    set volumesPerInstance(value) {
        this._volumesPerInstance = value;
    }
    resetVolumesPerInstance() {
        this._volumesPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesPerInstanceInput() {
        return this._volumesPerInstance;
    }
}
exports.EmrInstanceGroupEbsConfigOutputReference = EmrInstanceGroupEbsConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EmrInstanceGroupEbsConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.emrInstanceGroup.EmrInstanceGroupEbsConfigOutputReference", version: "19.15.0" };
class EmrInstanceGroupEbsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrInstanceGroupEbsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrInstanceGroupEbsConfigList = EmrInstanceGroupEbsConfigList;
_b = JSII_RTTI_SYMBOL_1;
EmrInstanceGroupEbsConfigList[_b] = { fqn: "@cdktf/provider-aws.emrInstanceGroup.EmrInstanceGroupEbsConfigList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_instance_group aws_emr_instance_group}
*/
class EmrInstanceGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EmrInstanceGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EmrInstanceGroup to import
    * @param importFromId The id of the existing EmrInstanceGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_instance_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EmrInstanceGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_emr_instance_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/emr_instance_group aws_emr_instance_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EmrInstanceGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emr_instance_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ebs_config - computed: false, optional: true, required: false
        this._ebsConfig = new EmrInstanceGroupEbsConfigList(this, "ebs_config", true);
        this._autoscalingPolicy = config.autoscalingPolicy;
        this._bidPrice = config.bidPrice;
        this._clusterId = config.clusterId;
        this._configurationsJson = config.configurationsJson;
        this._ebsOptimized = config.ebsOptimized;
        this._id = config.id;
        this._instanceCount = config.instanceCount;
        this._instanceType = config.instanceType;
        this._name = config.name;
        this._ebsConfig.internalValue = config.ebsConfig;
    }
    get autoscalingPolicy() {
        return this.getStringAttribute('autoscaling_policy');
    }
    set autoscalingPolicy(value) {
        this._autoscalingPolicy = value;
    }
    resetAutoscalingPolicy() {
        this._autoscalingPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingPolicyInput() {
        return this._autoscalingPolicy;
    }
    get bidPrice() {
        return this.getStringAttribute('bid_price');
    }
    set bidPrice(value) {
        this._bidPrice = value;
    }
    resetBidPrice() {
        this._bidPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceInput() {
        return this._bidPrice;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get configurationsJson() {
        return this.getStringAttribute('configurations_json');
    }
    set configurationsJson(value) {
        this._configurationsJson = value;
    }
    resetConfigurationsJson() {
        this._configurationsJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationsJsonInput() {
        return this._configurationsJson;
    }
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    set ebsOptimized(value) {
        this._ebsOptimized = value;
    }
    resetEbsOptimized() {
        this._ebsOptimized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsOptimizedInput() {
        return this._ebsOptimized;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // running_instance_count - computed: true, optional: false, required: false
    get runningInstanceCount() {
        return this.getNumberAttribute('running_instance_count');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get ebsConfig() {
        return this._ebsConfig;
    }
    putEbsConfig(value) {
        this._ebsConfig.internalValue = value;
    }
    resetEbsConfig() {
        this._ebsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsConfigInput() {
        return this._ebsConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            autoscaling_policy: cdktf.stringToTerraform(this._autoscalingPolicy),
            bid_price: cdktf.stringToTerraform(this._bidPrice),
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            configurations_json: cdktf.stringToTerraform(this._configurationsJson),
            ebs_optimized: cdktf.booleanToTerraform(this._ebsOptimized),
            id: cdktf.stringToTerraform(this._id),
            instance_count: cdktf.numberToTerraform(this._instanceCount),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            name: cdktf.stringToTerraform(this._name),
            ebs_config: cdktf.listMapper(emrInstanceGroupEbsConfigToTerraform, true)(this._ebsConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            autoscaling_policy: {
                value: cdktf.stringToHclTerraform(this._autoscalingPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bid_price: {
                value: cdktf.stringToHclTerraform(this._bidPrice),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cluster_id: {
                value: cdktf.stringToHclTerraform(this._clusterId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            configurations_json: {
                value: cdktf.stringToHclTerraform(this._configurationsJson),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ebs_optimized: {
                value: cdktf.booleanToHclTerraform(this._ebsOptimized),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_count: {
                value: cdktf.numberToHclTerraform(this._instanceCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            instance_type: {
                value: cdktf.stringToHclTerraform(this._instanceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ebs_config: {
                value: cdktf.listMapperHcl(emrInstanceGroupEbsConfigToHclTerraform, true)(this._ebsConfig.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "EmrInstanceGroupEbsConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EmrInstanceGroup = EmrInstanceGroup;
_c = JSII_RTTI_SYMBOL_1;
EmrInstanceGroup[_c] = { fqn: "@cdktf/provider-aws.emrInstanceGroup.EmrInstanceGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
EmrInstanceGroup.tfResourceType = "aws_emr_instance_group";
