// @cdktf/provider-aws/lib/vpclattice-target-group/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpclatticeTargetGroup = exports.VpclatticeTargetGroupTimeoutsOutputReference = exports.vpclatticeTargetGroupTimeoutsToHclTerraform = exports.vpclatticeTargetGroupTimeoutsToTerraform = exports.VpclatticeTargetGroupConfigAOutputReference = exports.vpclatticeTargetGroupConfigAToHclTerraform = exports.vpclatticeTargetGroupConfigAToTerraform = exports.VpclatticeTargetGroupConfigHealthCheckOutputReference = exports.vpclatticeTargetGroupConfigHealthCheckToHclTerraform = exports.vpclatticeTargetGroupConfigHealthCheckToTerraform = exports.VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference = exports.vpclatticeTargetGroupConfigHealthCheckMatcherToHclTerraform = exports.vpclatticeTargetGroupConfigHealthCheckMatcherToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpclatticeTargetGroupConfigHealthCheckMatcherToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.vpclatticeTargetGroupConfigHealthCheckMatcherToTerraform = vpclatticeTargetGroupConfigHealthCheckMatcherToTerraform;
function vpclatticeTargetGroupConfigHealthCheckMatcherToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vpclatticeTargetGroupConfigHealthCheckMatcherToHclTerraform = vpclatticeTargetGroupConfigHealthCheckMatcherToHclTerraform;
class VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._value = value.value;
        }
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference = VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference[_a] = { fqn: "@cdktf/provider-aws.vpclatticeTargetGroup.VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference", version: "19.15.0" };
function vpclatticeTargetGroupConfigHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        health_check_interval_seconds: cdktf.numberToTerraform(struct.healthCheckIntervalSeconds),
        health_check_timeout_seconds: cdktf.numberToTerraform(struct.healthCheckTimeoutSeconds),
        healthy_threshold_count: cdktf.numberToTerraform(struct.healthyThresholdCount),
        path: cdktf.stringToTerraform(struct.path),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        protocol_version: cdktf.stringToTerraform(struct.protocolVersion),
        unhealthy_threshold_count: cdktf.numberToTerraform(struct.unhealthyThresholdCount),
        matcher: vpclatticeTargetGroupConfigHealthCheckMatcherToTerraform(struct.matcher),
    };
}
exports.vpclatticeTargetGroupConfigHealthCheckToTerraform = vpclatticeTargetGroupConfigHealthCheckToTerraform;
function vpclatticeTargetGroupConfigHealthCheckToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        health_check_interval_seconds: {
            value: cdktf.numberToHclTerraform(struct.healthCheckIntervalSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        health_check_timeout_seconds: {
            value: cdktf.numberToHclTerraform(struct.healthCheckTimeoutSeconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        healthy_threshold_count: {
            value: cdktf.numberToHclTerraform(struct.healthyThresholdCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol_version: {
            value: cdktf.stringToHclTerraform(struct.protocolVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        unhealthy_threshold_count: {
            value: cdktf.numberToHclTerraform(struct.unhealthyThresholdCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        matcher: {
            value: vpclatticeTargetGroupConfigHealthCheckMatcherToHclTerraform(struct.matcher),
            isBlock: true,
            type: "list",
            storageClassType: "VpclatticeTargetGroupConfigHealthCheckMatcherList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vpclatticeTargetGroupConfigHealthCheckToHclTerraform = vpclatticeTargetGroupConfigHealthCheckToHclTerraform;
class VpclatticeTargetGroupConfigHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // matcher - computed: false, optional: true, required: false
        this._matcher = new VpclatticeTargetGroupConfigHealthCheckMatcherOutputReference(this, "matcher");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._healthCheckIntervalSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheckIntervalSeconds = this._healthCheckIntervalSeconds;
        }
        if (this._healthCheckTimeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheckTimeoutSeconds = this._healthCheckTimeoutSeconds;
        }
        if (this._healthyThresholdCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthyThresholdCount = this._healthyThresholdCount;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._protocolVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocolVersion = this._protocolVersion;
        }
        if (this._unhealthyThresholdCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.unhealthyThresholdCount = this._unhealthyThresholdCount;
        }
        if (this._matcher?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.matcher = this._matcher?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._healthCheckIntervalSeconds = undefined;
            this._healthCheckTimeoutSeconds = undefined;
            this._healthyThresholdCount = undefined;
            this._path = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._protocolVersion = undefined;
            this._unhealthyThresholdCount = undefined;
            this._matcher.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._healthCheckIntervalSeconds = value.healthCheckIntervalSeconds;
            this._healthCheckTimeoutSeconds = value.healthCheckTimeoutSeconds;
            this._healthyThresholdCount = value.healthyThresholdCount;
            this._path = value.path;
            this._port = value.port;
            this._protocol = value.protocol;
            this._protocolVersion = value.protocolVersion;
            this._unhealthyThresholdCount = value.unhealthyThresholdCount;
            this._matcher.internalValue = value.matcher;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get healthCheckIntervalSeconds() {
        return this.getNumberAttribute('health_check_interval_seconds');
    }
    set healthCheckIntervalSeconds(value) {
        this._healthCheckIntervalSeconds = value;
    }
    resetHealthCheckIntervalSeconds() {
        this._healthCheckIntervalSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckIntervalSecondsInput() {
        return this._healthCheckIntervalSeconds;
    }
    get healthCheckTimeoutSeconds() {
        return this.getNumberAttribute('health_check_timeout_seconds');
    }
    set healthCheckTimeoutSeconds(value) {
        this._healthCheckTimeoutSeconds = value;
    }
    resetHealthCheckTimeoutSeconds() {
        this._healthCheckTimeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckTimeoutSecondsInput() {
        return this._healthCheckTimeoutSeconds;
    }
    get healthyThresholdCount() {
        return this.getNumberAttribute('healthy_threshold_count');
    }
    set healthyThresholdCount(value) {
        this._healthyThresholdCount = value;
    }
    resetHealthyThresholdCount() {
        this._healthyThresholdCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthyThresholdCountInput() {
        return this._healthyThresholdCount;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get protocolVersion() {
        return this.getStringAttribute('protocol_version');
    }
    set protocolVersion(value) {
        this._protocolVersion = value;
    }
    resetProtocolVersion() {
        this._protocolVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolVersionInput() {
        return this._protocolVersion;
    }
    get unhealthyThresholdCount() {
        return this.getNumberAttribute('unhealthy_threshold_count');
    }
    set unhealthyThresholdCount(value) {
        this._unhealthyThresholdCount = value;
    }
    resetUnhealthyThresholdCount() {
        this._unhealthyThresholdCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unhealthyThresholdCountInput() {
        return this._unhealthyThresholdCount;
    }
    get matcher() {
        return this._matcher;
    }
    putMatcher(value) {
        this._matcher.internalValue = value;
    }
    resetMatcher() {
        this._matcher.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matcherInput() {
        return this._matcher.internalValue;
    }
}
exports.VpclatticeTargetGroupConfigHealthCheckOutputReference = VpclatticeTargetGroupConfigHealthCheckOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VpclatticeTargetGroupConfigHealthCheckOutputReference[_b] = { fqn: "@cdktf/provider-aws.vpclatticeTargetGroup.VpclatticeTargetGroupConfigHealthCheckOutputReference", version: "19.15.0" };
function vpclatticeTargetGroupConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ip_address_type: cdktf.stringToTerraform(struct.ipAddressType),
        lambda_event_structure_version: cdktf.stringToTerraform(struct.lambdaEventStructureVersion),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        protocol_version: cdktf.stringToTerraform(struct.protocolVersion),
        vpc_identifier: cdktf.stringToTerraform(struct.vpcIdentifier),
        health_check: vpclatticeTargetGroupConfigHealthCheckToTerraform(struct.healthCheck),
    };
}
exports.vpclatticeTargetGroupConfigAToTerraform = vpclatticeTargetGroupConfigAToTerraform;
function vpclatticeTargetGroupConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ip_address_type: {
            value: cdktf.stringToHclTerraform(struct.ipAddressType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        lambda_event_structure_version: {
            value: cdktf.stringToHclTerraform(struct.lambdaEventStructureVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        protocol: {
            value: cdktf.stringToHclTerraform(struct.protocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol_version: {
            value: cdktf.stringToHclTerraform(struct.protocolVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vpc_identifier: {
            value: cdktf.stringToHclTerraform(struct.vpcIdentifier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        health_check: {
            value: vpclatticeTargetGroupConfigHealthCheckToHclTerraform(struct.healthCheck),
            isBlock: true,
            type: "list",
            storageClassType: "VpclatticeTargetGroupConfigHealthCheckList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vpclatticeTargetGroupConfigAToHclTerraform = vpclatticeTargetGroupConfigAToHclTerraform;
class VpclatticeTargetGroupConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // health_check - computed: false, optional: true, required: false
        this._healthCheck = new VpclatticeTargetGroupConfigHealthCheckOutputReference(this, "health_check");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipAddressType !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipAddressType = this._ipAddressType;
        }
        if (this._lambdaEventStructureVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambdaEventStructureVersion = this._lambdaEventStructureVersion;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._protocolVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocolVersion = this._protocolVersion;
        }
        if (this._vpcIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcIdentifier = this._vpcIdentifier;
        }
        if (this._healthCheck?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ipAddressType = undefined;
            this._lambdaEventStructureVersion = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._protocolVersion = undefined;
            this._vpcIdentifier = undefined;
            this._healthCheck.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ipAddressType = value.ipAddressType;
            this._lambdaEventStructureVersion = value.lambdaEventStructureVersion;
            this._port = value.port;
            this._protocol = value.protocol;
            this._protocolVersion = value.protocolVersion;
            this._vpcIdentifier = value.vpcIdentifier;
            this._healthCheck.internalValue = value.healthCheck;
        }
    }
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    set ipAddressType(value) {
        this._ipAddressType = value;
    }
    resetIpAddressType() {
        this._ipAddressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressTypeInput() {
        return this._ipAddressType;
    }
    get lambdaEventStructureVersion() {
        return this.getStringAttribute('lambda_event_structure_version');
    }
    set lambdaEventStructureVersion(value) {
        this._lambdaEventStructureVersion = value;
    }
    resetLambdaEventStructureVersion() {
        this._lambdaEventStructureVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaEventStructureVersionInput() {
        return this._lambdaEventStructureVersion;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get protocolVersion() {
        return this.getStringAttribute('protocol_version');
    }
    set protocolVersion(value) {
        this._protocolVersion = value;
    }
    resetProtocolVersion() {
        this._protocolVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolVersionInput() {
        return this._protocolVersion;
    }
    get vpcIdentifier() {
        return this.getStringAttribute('vpc_identifier');
    }
    set vpcIdentifier(value) {
        this._vpcIdentifier = value;
    }
    resetVpcIdentifier() {
        this._vpcIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdentifierInput() {
        return this._vpcIdentifier;
    }
    get healthCheck() {
        return this._healthCheck;
    }
    putHealthCheck(value) {
        this._healthCheck.internalValue = value;
    }
    resetHealthCheck() {
        this._healthCheck.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck.internalValue;
    }
}
exports.VpclatticeTargetGroupConfigAOutputReference = VpclatticeTargetGroupConfigAOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VpclatticeTargetGroupConfigAOutputReference[_c] = { fqn: "@cdktf/provider-aws.vpclatticeTargetGroup.VpclatticeTargetGroupConfigAOutputReference", version: "19.15.0" };
function vpclatticeTargetGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpclatticeTargetGroupTimeoutsToTerraform = vpclatticeTargetGroupTimeoutsToTerraform;
function vpclatticeTargetGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vpclatticeTargetGroupTimeoutsToHclTerraform = vpclatticeTargetGroupTimeoutsToHclTerraform;
class VpclatticeTargetGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpclatticeTargetGroupTimeoutsOutputReference = VpclatticeTargetGroupTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
VpclatticeTargetGroupTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.vpclatticeTargetGroup.VpclatticeTargetGroupTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpclattice_target_group aws_vpclattice_target_group}
*/
class VpclatticeTargetGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VpclatticeTargetGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VpclatticeTargetGroup to import
    * @param importFromId The id of the existing VpclatticeTargetGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpclattice_target_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VpclatticeTargetGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_vpclattice_target_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/vpclattice_target_group aws_vpclattice_target_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpclatticeTargetGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpclattice_target_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: false, optional: true, required: false
        this._config = new VpclatticeTargetGroupConfigAOutputReference(this, "config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpclatticeTargetGroupTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._config.internalValue = config.config;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    resetConfig() {
        this._config.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            config: vpclatticeTargetGroupConfigAToTerraform(this._config.internalValue),
            timeouts: vpclatticeTargetGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            config: {
                value: vpclatticeTargetGroupConfigAToHclTerraform(this._config.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VpclatticeTargetGroupConfigAList",
            },
            timeouts: {
                value: vpclatticeTargetGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VpclatticeTargetGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VpclatticeTargetGroup = VpclatticeTargetGroup;
_e = JSII_RTTI_SYMBOL_1;
VpclatticeTargetGroup[_e] = { fqn: "@cdktf/provider-aws.vpclatticeTargetGroup.VpclatticeTargetGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VpclatticeTargetGroup.tfResourceType = "aws_vpclattice_target_group";
