// @cdktf/provider-aws/lib/data-aws-budgets-budget/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBudgetsBudget = exports.DataAwsBudgetsBudgetPlannedLimitList = exports.DataAwsBudgetsBudgetPlannedLimitOutputReference = exports.dataAwsBudgetsBudgetPlannedLimitToHclTerraform = exports.dataAwsBudgetsBudgetPlannedLimitToTerraform = exports.DataAwsBudgetsBudgetNotificationList = exports.DataAwsBudgetsBudgetNotificationOutputReference = exports.dataAwsBudgetsBudgetNotificationToHclTerraform = exports.dataAwsBudgetsBudgetNotificationToTerraform = exports.DataAwsBudgetsBudgetCostTypesList = exports.DataAwsBudgetsBudgetCostTypesOutputReference = exports.dataAwsBudgetsBudgetCostTypesToHclTerraform = exports.dataAwsBudgetsBudgetCostTypesToTerraform = exports.DataAwsBudgetsBudgetCostFilterList = exports.DataAwsBudgetsBudgetCostFilterOutputReference = exports.dataAwsBudgetsBudgetCostFilterToHclTerraform = exports.dataAwsBudgetsBudgetCostFilterToTerraform = exports.DataAwsBudgetsBudgetCalculatedSpendList = exports.DataAwsBudgetsBudgetCalculatedSpendOutputReference = exports.dataAwsBudgetsBudgetCalculatedSpendToHclTerraform = exports.dataAwsBudgetsBudgetCalculatedSpendToTerraform = exports.DataAwsBudgetsBudgetCalculatedSpendActualSpendList = exports.DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference = exports.dataAwsBudgetsBudgetCalculatedSpendActualSpendToHclTerraform = exports.dataAwsBudgetsBudgetCalculatedSpendActualSpendToTerraform = exports.DataAwsBudgetsBudgetBudgetLimitList = exports.DataAwsBudgetsBudgetBudgetLimitOutputReference = exports.dataAwsBudgetsBudgetBudgetLimitToHclTerraform = exports.dataAwsBudgetsBudgetBudgetLimitToTerraform = exports.DataAwsBudgetsBudgetAutoAdjustDataList = exports.DataAwsBudgetsBudgetAutoAdjustDataOutputReference = exports.dataAwsBudgetsBudgetAutoAdjustDataToHclTerraform = exports.dataAwsBudgetsBudgetAutoAdjustDataToTerraform = exports.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList = exports.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference = exports.dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToHclTerraform = exports.dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToTerraform = dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToTerraform;
function dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToHclTerraform = dataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsToHclTerraform;
class DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // budget_adjustment_period - computed: true, optional: false, required: false
    get budgetAdjustmentPeriod() {
        return this.getNumberAttribute('budget_adjustment_period');
    }
    // lookback_available_periods - computed: true, optional: false, required: false
    get lookbackAvailablePeriods() {
        return this.getNumberAttribute('lookback_available_periods');
    }
}
exports.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference = DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList = DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList", version: "19.15.0" };
function dataAwsBudgetsBudgetAutoAdjustDataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetAutoAdjustDataToTerraform = dataAwsBudgetsBudgetAutoAdjustDataToTerraform;
function dataAwsBudgetsBudgetAutoAdjustDataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetAutoAdjustDataToHclTerraform = dataAwsBudgetsBudgetAutoAdjustDataToHclTerraform;
class DataAwsBudgetsBudgetAutoAdjustDataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // historical_options - computed: true, optional: false, required: false
        this._historicalOptions = new DataAwsBudgetsBudgetAutoAdjustDataHistoricalOptionsList(this, "historical_options", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_adjust_type - computed: true, optional: false, required: false
    get autoAdjustType() {
        return this.getStringAttribute('auto_adjust_type');
    }
    get historicalOptions() {
        return this._historicalOptions;
    }
    // last_auto_adjust_time - computed: true, optional: false, required: false
    get lastAutoAdjustTime() {
        return this.getStringAttribute('last_auto_adjust_time');
    }
}
exports.DataAwsBudgetsBudgetAutoAdjustDataOutputReference = DataAwsBudgetsBudgetAutoAdjustDataOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetAutoAdjustDataOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetAutoAdjustDataOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetAutoAdjustDataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetAutoAdjustDataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetAutoAdjustDataList = DataAwsBudgetsBudgetAutoAdjustDataList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetAutoAdjustDataList[_d] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetAutoAdjustDataList", version: "19.15.0" };
function dataAwsBudgetsBudgetBudgetLimitToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetBudgetLimitToTerraform = dataAwsBudgetsBudgetBudgetLimitToTerraform;
function dataAwsBudgetsBudgetBudgetLimitToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetBudgetLimitToHclTerraform = dataAwsBudgetsBudgetBudgetLimitToHclTerraform;
class DataAwsBudgetsBudgetBudgetLimitOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // amount - computed: true, optional: false, required: false
    get amount() {
        return this.getStringAttribute('amount');
    }
    // unit - computed: true, optional: false, required: false
    get unit() {
        return this.getStringAttribute('unit');
    }
}
exports.DataAwsBudgetsBudgetBudgetLimitOutputReference = DataAwsBudgetsBudgetBudgetLimitOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetBudgetLimitOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetBudgetLimitOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetBudgetLimitList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetBudgetLimitOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetBudgetLimitList = DataAwsBudgetsBudgetBudgetLimitList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetBudgetLimitList[_f] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetBudgetLimitList", version: "19.15.0" };
function dataAwsBudgetsBudgetCalculatedSpendActualSpendToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetCalculatedSpendActualSpendToTerraform = dataAwsBudgetsBudgetCalculatedSpendActualSpendToTerraform;
function dataAwsBudgetsBudgetCalculatedSpendActualSpendToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetCalculatedSpendActualSpendToHclTerraform = dataAwsBudgetsBudgetCalculatedSpendActualSpendToHclTerraform;
class DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // amount - computed: true, optional: false, required: false
    get amount() {
        return this.getStringAttribute('amount');
    }
    // unit - computed: true, optional: false, required: false
    get unit() {
        return this.getStringAttribute('unit');
    }
}
exports.DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference = DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetCalculatedSpendActualSpendList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetCalculatedSpendActualSpendOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetCalculatedSpendActualSpendList = DataAwsBudgetsBudgetCalculatedSpendActualSpendList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCalculatedSpendActualSpendList[_h] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCalculatedSpendActualSpendList", version: "19.15.0" };
function dataAwsBudgetsBudgetCalculatedSpendToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetCalculatedSpendToTerraform = dataAwsBudgetsBudgetCalculatedSpendToTerraform;
function dataAwsBudgetsBudgetCalculatedSpendToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetCalculatedSpendToHclTerraform = dataAwsBudgetsBudgetCalculatedSpendToHclTerraform;
class DataAwsBudgetsBudgetCalculatedSpendOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // actual_spend - computed: true, optional: false, required: false
        this._actualSpend = new DataAwsBudgetsBudgetCalculatedSpendActualSpendList(this, "actual_spend", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get actualSpend() {
        return this._actualSpend;
    }
}
exports.DataAwsBudgetsBudgetCalculatedSpendOutputReference = DataAwsBudgetsBudgetCalculatedSpendOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCalculatedSpendOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCalculatedSpendOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetCalculatedSpendList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetCalculatedSpendOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetCalculatedSpendList = DataAwsBudgetsBudgetCalculatedSpendList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCalculatedSpendList[_k] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCalculatedSpendList", version: "19.15.0" };
function dataAwsBudgetsBudgetCostFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetCostFilterToTerraform = dataAwsBudgetsBudgetCostFilterToTerraform;
function dataAwsBudgetsBudgetCostFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetCostFilterToHclTerraform = dataAwsBudgetsBudgetCostFilterToHclTerraform;
class DataAwsBudgetsBudgetCostFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataAwsBudgetsBudgetCostFilterOutputReference = DataAwsBudgetsBudgetCostFilterOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCostFilterOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCostFilterOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetCostFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetCostFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetCostFilterList = DataAwsBudgetsBudgetCostFilterList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCostFilterList[_m] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCostFilterList", version: "19.15.0" };
function dataAwsBudgetsBudgetCostTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetCostTypesToTerraform = dataAwsBudgetsBudgetCostTypesToTerraform;
function dataAwsBudgetsBudgetCostTypesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetCostTypesToHclTerraform = dataAwsBudgetsBudgetCostTypesToHclTerraform;
class DataAwsBudgetsBudgetCostTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // include_credit - computed: true, optional: false, required: false
    get includeCredit() {
        return this.getBooleanAttribute('include_credit');
    }
    // include_discount - computed: true, optional: false, required: false
    get includeDiscount() {
        return this.getBooleanAttribute('include_discount');
    }
    // include_other_subscription - computed: true, optional: false, required: false
    get includeOtherSubscription() {
        return this.getBooleanAttribute('include_other_subscription');
    }
    // include_recurring - computed: true, optional: false, required: false
    get includeRecurring() {
        return this.getBooleanAttribute('include_recurring');
    }
    // include_refund - computed: true, optional: false, required: false
    get includeRefund() {
        return this.getBooleanAttribute('include_refund');
    }
    // include_subscription - computed: true, optional: false, required: false
    get includeSubscription() {
        return this.getBooleanAttribute('include_subscription');
    }
    // include_support - computed: true, optional: false, required: false
    get includeSupport() {
        return this.getBooleanAttribute('include_support');
    }
    // include_tax - computed: true, optional: false, required: false
    get includeTax() {
        return this.getBooleanAttribute('include_tax');
    }
    // include_upfront - computed: true, optional: false, required: false
    get includeUpfront() {
        return this.getBooleanAttribute('include_upfront');
    }
    // use_amortized - computed: true, optional: false, required: false
    get useAmortized() {
        return this.getBooleanAttribute('use_amortized');
    }
    // use_blended - computed: true, optional: false, required: false
    get useBlended() {
        return this.getBooleanAttribute('use_blended');
    }
}
exports.DataAwsBudgetsBudgetCostTypesOutputReference = DataAwsBudgetsBudgetCostTypesOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCostTypesOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCostTypesOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetCostTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetCostTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetCostTypesList = DataAwsBudgetsBudgetCostTypesList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetCostTypesList[_p] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetCostTypesList", version: "19.15.0" };
function dataAwsBudgetsBudgetNotificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetNotificationToTerraform = dataAwsBudgetsBudgetNotificationToTerraform;
function dataAwsBudgetsBudgetNotificationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetNotificationToHclTerraform = dataAwsBudgetsBudgetNotificationToHclTerraform;
class DataAwsBudgetsBudgetNotificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // comparison_operator - computed: true, optional: false, required: false
    get comparisonOperator() {
        return this.getStringAttribute('comparison_operator');
    }
    // notification_type - computed: true, optional: false, required: false
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    // subscriber_email_addresses - computed: true, optional: false, required: false
    get subscriberEmailAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('subscriber_email_addresses'));
    }
    // subscriber_sns_topic_arns - computed: true, optional: false, required: false
    get subscriberSnsTopicArns() {
        return cdktf.Fn.tolist(this.getListAttribute('subscriber_sns_topic_arns'));
    }
    // threshold - computed: true, optional: false, required: false
    get threshold() {
        return this.getNumberAttribute('threshold');
    }
    // threshold_type - computed: true, optional: false, required: false
    get thresholdType() {
        return this.getStringAttribute('threshold_type');
    }
}
exports.DataAwsBudgetsBudgetNotificationOutputReference = DataAwsBudgetsBudgetNotificationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetNotificationOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetNotificationOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetNotificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetNotificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetNotificationList = DataAwsBudgetsBudgetNotificationList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetNotificationList[_r] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetNotificationList", version: "19.15.0" };
function dataAwsBudgetsBudgetPlannedLimitToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBudgetsBudgetPlannedLimitToTerraform = dataAwsBudgetsBudgetPlannedLimitToTerraform;
function dataAwsBudgetsBudgetPlannedLimitToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBudgetsBudgetPlannedLimitToHclTerraform = dataAwsBudgetsBudgetPlannedLimitToHclTerraform;
class DataAwsBudgetsBudgetPlannedLimitOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // amount - computed: true, optional: false, required: false
    get amount() {
        return this.getStringAttribute('amount');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // unit - computed: true, optional: false, required: false
    get unit() {
        return this.getStringAttribute('unit');
    }
}
exports.DataAwsBudgetsBudgetPlannedLimitOutputReference = DataAwsBudgetsBudgetPlannedLimitOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetPlannedLimitOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetPlannedLimitOutputReference", version: "19.15.0" };
class DataAwsBudgetsBudgetPlannedLimitList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBudgetsBudgetPlannedLimitOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBudgetsBudgetPlannedLimitList = DataAwsBudgetsBudgetPlannedLimitList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudgetPlannedLimitList[_t] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudgetPlannedLimitList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/budgets_budget aws_budgets_budget}
*/
class DataAwsBudgetsBudget extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsBudgetsBudget resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsBudgetsBudget to import
    * @param importFromId The id of the existing DataAwsBudgetsBudget that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/budgets_budget#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsBudgetsBudget to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_budgets_budget", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/budgets_budget aws_budgets_budget} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsBudgetsBudgetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_budgets_budget',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auto_adjust_data - computed: true, optional: false, required: false
        this._autoAdjustData = new DataAwsBudgetsBudgetAutoAdjustDataList(this, "auto_adjust_data", false);
        // budget_limit - computed: true, optional: false, required: false
        this._budgetLimit = new DataAwsBudgetsBudgetBudgetLimitList(this, "budget_limit", false);
        // calculated_spend - computed: true, optional: false, required: false
        this._calculatedSpend = new DataAwsBudgetsBudgetCalculatedSpendList(this, "calculated_spend", false);
        // cost_filter - computed: true, optional: false, required: false
        this._costFilter = new DataAwsBudgetsBudgetCostFilterList(this, "cost_filter", true);
        // cost_types - computed: true, optional: false, required: false
        this._costTypes = new DataAwsBudgetsBudgetCostTypesList(this, "cost_types", false);
        // notification - computed: true, optional: false, required: false
        this._notification = new DataAwsBudgetsBudgetNotificationList(this, "notification", true);
        // planned_limit - computed: true, optional: false, required: false
        this._plannedLimit = new DataAwsBudgetsBudgetPlannedLimitList(this, "planned_limit", true);
        this._accountId = config.accountId;
        this._id = config.id;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoAdjustData() {
        return this._autoAdjustData;
    }
    // budget_exceeded - computed: true, optional: false, required: false
    get budgetExceeded() {
        return this.getBooleanAttribute('budget_exceeded');
    }
    get budgetLimit() {
        return this._budgetLimit;
    }
    // budget_type - computed: true, optional: false, required: false
    get budgetType() {
        return this.getStringAttribute('budget_type');
    }
    get calculatedSpend() {
        return this._calculatedSpend;
    }
    get costFilter() {
        return this._costFilter;
    }
    get costTypes() {
        return this._costTypes;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get notification() {
        return this._notification;
    }
    get plannedLimit() {
        return this._plannedLimit;
    }
    // time_period_end - computed: true, optional: false, required: false
    get timePeriodEnd() {
        return this.getStringAttribute('time_period_end');
    }
    // time_period_start - computed: true, optional: false, required: false
    get timePeriodStart() {
        return this.getStringAttribute('time_period_start');
    }
    // time_unit - computed: true, optional: false, required: false
    get timeUnit() {
        return this.getStringAttribute('time_unit');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            account_id: {
                value: cdktf.stringToHclTerraform(this._accountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name_prefix: {
                value: cdktf.stringToHclTerraform(this._namePrefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsBudgetsBudget = DataAwsBudgetsBudget;
_u = JSII_RTTI_SYMBOL_1;
DataAwsBudgetsBudget[_u] = { fqn: "@cdktf/provider-aws.dataAwsBudgetsBudget.DataAwsBudgetsBudget", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsBudgetsBudget.tfResourceType = "aws_budgets_budget";
