// @cdktf/provider-aws/lib/cloudsearch-domain/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudsearchDomain = exports.CloudsearchDomainTimeoutsOutputReference = exports.cloudsearchDomainTimeoutsToHclTerraform = exports.cloudsearchDomainTimeoutsToTerraform = exports.CloudsearchDomainScalingParametersOutputReference = exports.cloudsearchDomainScalingParametersToHclTerraform = exports.cloudsearchDomainScalingParametersToTerraform = exports.CloudsearchDomainIndexFieldList = exports.CloudsearchDomainIndexFieldOutputReference = exports.cloudsearchDomainIndexFieldToHclTerraform = exports.cloudsearchDomainIndexFieldToTerraform = exports.CloudsearchDomainEndpointOptionsOutputReference = exports.cloudsearchDomainEndpointOptionsToHclTerraform = exports.cloudsearchDomainEndpointOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudsearchDomainEndpointOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enforce_https: cdktf.booleanToTerraform(struct.enforceHttps),
        tls_security_policy: cdktf.stringToTerraform(struct.tlsSecurityPolicy),
    };
}
exports.cloudsearchDomainEndpointOptionsToTerraform = cloudsearchDomainEndpointOptionsToTerraform;
function cloudsearchDomainEndpointOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enforce_https: {
            value: cdktf.booleanToHclTerraform(struct.enforceHttps),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        tls_security_policy: {
            value: cdktf.stringToHclTerraform(struct.tlsSecurityPolicy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudsearchDomainEndpointOptionsToHclTerraform = cloudsearchDomainEndpointOptionsToHclTerraform;
class CloudsearchDomainEndpointOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enforceHttps !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceHttps = this._enforceHttps;
        }
        if (this._tlsSecurityPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.tlsSecurityPolicy = this._tlsSecurityPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enforceHttps = undefined;
            this._tlsSecurityPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enforceHttps = value.enforceHttps;
            this._tlsSecurityPolicy = value.tlsSecurityPolicy;
        }
    }
    get enforceHttps() {
        return this.getBooleanAttribute('enforce_https');
    }
    set enforceHttps(value) {
        this._enforceHttps = value;
    }
    resetEnforceHttps() {
        this._enforceHttps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceHttpsInput() {
        return this._enforceHttps;
    }
    get tlsSecurityPolicy() {
        return this.getStringAttribute('tls_security_policy');
    }
    set tlsSecurityPolicy(value) {
        this._tlsSecurityPolicy = value;
    }
    resetTlsSecurityPolicy() {
        this._tlsSecurityPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsSecurityPolicyInput() {
        return this._tlsSecurityPolicy;
    }
}
exports.CloudsearchDomainEndpointOptionsOutputReference = CloudsearchDomainEndpointOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudsearchDomainEndpointOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomainEndpointOptionsOutputReference", version: "19.15.0" };
function cloudsearchDomainIndexFieldToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        analysis_scheme: cdktf.stringToTerraform(struct.analysisScheme),
        default_value: cdktf.stringToTerraform(struct.defaultValue),
        facet: cdktf.booleanToTerraform(struct.facet),
        highlight: cdktf.booleanToTerraform(struct.highlight),
        name: cdktf.stringToTerraform(struct.name),
        return: cdktf.booleanToTerraform(struct.return),
        search: cdktf.booleanToTerraform(struct.search),
        sort: cdktf.booleanToTerraform(struct.sort),
        source_fields: cdktf.stringToTerraform(struct.sourceFields),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.cloudsearchDomainIndexFieldToTerraform = cloudsearchDomainIndexFieldToTerraform;
function cloudsearchDomainIndexFieldToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        analysis_scheme: {
            value: cdktf.stringToHclTerraform(struct.analysisScheme),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        default_value: {
            value: cdktf.stringToHclTerraform(struct.defaultValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        facet: {
            value: cdktf.booleanToHclTerraform(struct.facet),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        highlight: {
            value: cdktf.booleanToHclTerraform(struct.highlight),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        return: {
            value: cdktf.booleanToHclTerraform(struct.return),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        search: {
            value: cdktf.booleanToHclTerraform(struct.search),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        sort: {
            value: cdktf.booleanToHclTerraform(struct.sort),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        source_fields: {
            value: cdktf.stringToHclTerraform(struct.sourceFields),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudsearchDomainIndexFieldToHclTerraform = cloudsearchDomainIndexFieldToHclTerraform;
class CloudsearchDomainIndexFieldOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._analysisScheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.analysisScheme = this._analysisScheme;
        }
        if (this._defaultValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultValue = this._defaultValue;
        }
        if (this._facet !== undefined) {
            hasAnyValues = true;
            internalValueResult.facet = this._facet;
        }
        if (this._highlight !== undefined) {
            hasAnyValues = true;
            internalValueResult.highlight = this._highlight;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._return !== undefined) {
            hasAnyValues = true;
            internalValueResult.return = this._return;
        }
        if (this._search !== undefined) {
            hasAnyValues = true;
            internalValueResult.search = this._search;
        }
        if (this._sort !== undefined) {
            hasAnyValues = true;
            internalValueResult.sort = this._sort;
        }
        if (this._sourceFields !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFields = this._sourceFields;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._analysisScheme = undefined;
            this._defaultValue = undefined;
            this._facet = undefined;
            this._highlight = undefined;
            this._name = undefined;
            this._return = undefined;
            this._search = undefined;
            this._sort = undefined;
            this._sourceFields = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._analysisScheme = value.analysisScheme;
            this._defaultValue = value.defaultValue;
            this._facet = value.facet;
            this._highlight = value.highlight;
            this._name = value.name;
            this._return = value.return;
            this._search = value.search;
            this._sort = value.sort;
            this._sourceFields = value.sourceFields;
            this._type = value.type;
        }
    }
    get analysisScheme() {
        return this.getStringAttribute('analysis_scheme');
    }
    set analysisScheme(value) {
        this._analysisScheme = value;
    }
    resetAnalysisScheme() {
        this._analysisScheme = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get analysisSchemeInput() {
        return this._analysisScheme;
    }
    get defaultValue() {
        return this.getStringAttribute('default_value');
    }
    set defaultValue(value) {
        this._defaultValue = value;
    }
    resetDefaultValue() {
        this._defaultValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultValueInput() {
        return this._defaultValue;
    }
    get facet() {
        return this.getBooleanAttribute('facet');
    }
    set facet(value) {
        this._facet = value;
    }
    resetFacet() {
        this._facet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get facetInput() {
        return this._facet;
    }
    get highlight() {
        return this.getBooleanAttribute('highlight');
    }
    set highlight(value) {
        this._highlight = value;
    }
    resetHighlight() {
        this._highlight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get highlightInput() {
        return this._highlight;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get return() {
        return this.getBooleanAttribute('return');
    }
    set return(value) {
        this._return = value;
    }
    resetReturn() {
        this._return = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get returnInput() {
        return this._return;
    }
    get search() {
        return this.getBooleanAttribute('search');
    }
    set search(value) {
        this._search = value;
    }
    resetSearch() {
        this._search = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get searchInput() {
        return this._search;
    }
    get sort() {
        return this.getBooleanAttribute('sort');
    }
    set sort(value) {
        this._sort = value;
    }
    resetSort() {
        this._sort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sortInput() {
        return this._sort;
    }
    get sourceFields() {
        return this.getStringAttribute('source_fields');
    }
    set sourceFields(value) {
        this._sourceFields = value;
    }
    resetSourceFields() {
        this._sourceFields = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFieldsInput() {
        return this._sourceFields;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CloudsearchDomainIndexFieldOutputReference = CloudsearchDomainIndexFieldOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudsearchDomainIndexFieldOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomainIndexFieldOutputReference", version: "19.15.0" };
class CloudsearchDomainIndexFieldList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudsearchDomainIndexFieldOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudsearchDomainIndexFieldList = CloudsearchDomainIndexFieldList;
_c = JSII_RTTI_SYMBOL_1;
CloudsearchDomainIndexFieldList[_c] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomainIndexFieldList", version: "19.15.0" };
function cloudsearchDomainScalingParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        desired_instance_type: cdktf.stringToTerraform(struct.desiredInstanceType),
        desired_partition_count: cdktf.numberToTerraform(struct.desiredPartitionCount),
        desired_replication_count: cdktf.numberToTerraform(struct.desiredReplicationCount),
    };
}
exports.cloudsearchDomainScalingParametersToTerraform = cloudsearchDomainScalingParametersToTerraform;
function cloudsearchDomainScalingParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        desired_instance_type: {
            value: cdktf.stringToHclTerraform(struct.desiredInstanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        desired_partition_count: {
            value: cdktf.numberToHclTerraform(struct.desiredPartitionCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        desired_replication_count: {
            value: cdktf.numberToHclTerraform(struct.desiredReplicationCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudsearchDomainScalingParametersToHclTerraform = cloudsearchDomainScalingParametersToHclTerraform;
class CloudsearchDomainScalingParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._desiredInstanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredInstanceType = this._desiredInstanceType;
        }
        if (this._desiredPartitionCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredPartitionCount = this._desiredPartitionCount;
        }
        if (this._desiredReplicationCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.desiredReplicationCount = this._desiredReplicationCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._desiredInstanceType = undefined;
            this._desiredPartitionCount = undefined;
            this._desiredReplicationCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._desiredInstanceType = value.desiredInstanceType;
            this._desiredPartitionCount = value.desiredPartitionCount;
            this._desiredReplicationCount = value.desiredReplicationCount;
        }
    }
    get desiredInstanceType() {
        return this.getStringAttribute('desired_instance_type');
    }
    set desiredInstanceType(value) {
        this._desiredInstanceType = value;
    }
    resetDesiredInstanceType() {
        this._desiredInstanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredInstanceTypeInput() {
        return this._desiredInstanceType;
    }
    get desiredPartitionCount() {
        return this.getNumberAttribute('desired_partition_count');
    }
    set desiredPartitionCount(value) {
        this._desiredPartitionCount = value;
    }
    resetDesiredPartitionCount() {
        this._desiredPartitionCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredPartitionCountInput() {
        return this._desiredPartitionCount;
    }
    get desiredReplicationCount() {
        return this.getNumberAttribute('desired_replication_count');
    }
    set desiredReplicationCount(value) {
        this._desiredReplicationCount = value;
    }
    resetDesiredReplicationCount() {
        this._desiredReplicationCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredReplicationCountInput() {
        return this._desiredReplicationCount;
    }
}
exports.CloudsearchDomainScalingParametersOutputReference = CloudsearchDomainScalingParametersOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudsearchDomainScalingParametersOutputReference[_d] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomainScalingParametersOutputReference", version: "19.15.0" };
function cloudsearchDomainTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudsearchDomainTimeoutsToTerraform = cloudsearchDomainTimeoutsToTerraform;
function cloudsearchDomainTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudsearchDomainTimeoutsToHclTerraform = cloudsearchDomainTimeoutsToHclTerraform;
class CloudsearchDomainTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudsearchDomainTimeoutsOutputReference = CloudsearchDomainTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudsearchDomainTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomainTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudsearch_domain aws_cloudsearch_domain}
*/
class CloudsearchDomain extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudsearchDomain resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudsearchDomain to import
    * @param importFromId The id of the existing CloudsearchDomain that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudsearch_domain#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudsearchDomain to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudsearch_domain", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudsearch_domain aws_cloudsearch_domain} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudsearchDomainConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudsearch_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint_options - computed: false, optional: true, required: false
        this._endpointOptions = new CloudsearchDomainEndpointOptionsOutputReference(this, "endpoint_options");
        // index_field - computed: false, optional: true, required: false
        this._indexField = new CloudsearchDomainIndexFieldList(this, "index_field", true);
        // scaling_parameters - computed: false, optional: true, required: false
        this._scalingParameters = new CloudsearchDomainScalingParametersOutputReference(this, "scaling_parameters");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudsearchDomainTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._multiAz = config.multiAz;
        this._name = config.name;
        this._endpointOptions.internalValue = config.endpointOptions;
        this._indexField.internalValue = config.indexField;
        this._scalingParameters.internalValue = config.scalingParameters;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // document_service_endpoint - computed: true, optional: false, required: false
    get documentServiceEndpoint() {
        return this.getStringAttribute('document_service_endpoint');
    }
    // domain_id - computed: true, optional: false, required: false
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get multiAz() {
        return this.getBooleanAttribute('multi_az');
    }
    set multiAz(value) {
        this._multiAz = value;
    }
    resetMultiAz() {
        this._multiAz = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multiAzInput() {
        return this._multiAz;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // search_service_endpoint - computed: true, optional: false, required: false
    get searchServiceEndpoint() {
        return this.getStringAttribute('search_service_endpoint');
    }
    get endpointOptions() {
        return this._endpointOptions;
    }
    putEndpointOptions(value) {
        this._endpointOptions.internalValue = value;
    }
    resetEndpointOptions() {
        this._endpointOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointOptionsInput() {
        return this._endpointOptions.internalValue;
    }
    get indexField() {
        return this._indexField;
    }
    putIndexField(value) {
        this._indexField.internalValue = value;
    }
    resetIndexField() {
        this._indexField.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get indexFieldInput() {
        return this._indexField.internalValue;
    }
    get scalingParameters() {
        return this._scalingParameters;
    }
    putScalingParameters(value) {
        this._scalingParameters.internalValue = value;
    }
    resetScalingParameters() {
        this._scalingParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scalingParametersInput() {
        return this._scalingParameters.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            multi_az: cdktf.booleanToTerraform(this._multiAz),
            name: cdktf.stringToTerraform(this._name),
            endpoint_options: cloudsearchDomainEndpointOptionsToTerraform(this._endpointOptions.internalValue),
            index_field: cdktf.listMapper(cloudsearchDomainIndexFieldToTerraform, true)(this._indexField.internalValue),
            scaling_parameters: cloudsearchDomainScalingParametersToTerraform(this._scalingParameters.internalValue),
            timeouts: cloudsearchDomainTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            multi_az: {
                value: cdktf.booleanToHclTerraform(this._multiAz),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            endpoint_options: {
                value: cloudsearchDomainEndpointOptionsToHclTerraform(this._endpointOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudsearchDomainEndpointOptionsList",
            },
            index_field: {
                value: cdktf.listMapperHcl(cloudsearchDomainIndexFieldToHclTerraform, true)(this._indexField.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "CloudsearchDomainIndexFieldList",
            },
            scaling_parameters: {
                value: cloudsearchDomainScalingParametersToHclTerraform(this._scalingParameters.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudsearchDomainScalingParametersList",
            },
            timeouts: {
                value: cloudsearchDomainTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CloudsearchDomainTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudsearchDomain = CloudsearchDomain;
_f = JSII_RTTI_SYMBOL_1;
CloudsearchDomain[_f] = { fqn: "@cdktf/provider-aws.cloudsearchDomain.CloudsearchDomain", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudsearchDomain.tfResourceType = "aws_cloudsearch_domain";
