// @cdktf/provider-aws/lib/auditmanager-assessment/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditmanagerAssessment = exports.AuditmanagerAssessmentScopeList = exports.AuditmanagerAssessmentScopeOutputReference = exports.auditmanagerAssessmentScopeToHclTerraform = exports.auditmanagerAssessmentScopeToTerraform = exports.AuditmanagerAssessmentScopeAwsServicesList = exports.AuditmanagerAssessmentScopeAwsServicesOutputReference = exports.auditmanagerAssessmentScopeAwsServicesToHclTerraform = exports.auditmanagerAssessmentScopeAwsServicesToTerraform = exports.AuditmanagerAssessmentScopeAwsAccountsList = exports.AuditmanagerAssessmentScopeAwsAccountsOutputReference = exports.auditmanagerAssessmentScopeAwsAccountsToHclTerraform = exports.auditmanagerAssessmentScopeAwsAccountsToTerraform = exports.AuditmanagerAssessmentAssessmentReportsDestinationList = exports.AuditmanagerAssessmentAssessmentReportsDestinationOutputReference = exports.auditmanagerAssessmentAssessmentReportsDestinationToHclTerraform = exports.auditmanagerAssessmentAssessmentReportsDestinationToTerraform = exports.AuditmanagerAssessmentRolesAllList = exports.AuditmanagerAssessmentRolesAllOutputReference = exports.auditmanagerAssessmentRolesAllToHclTerraform = exports.auditmanagerAssessmentRolesAllToTerraform = exports.AuditmanagerAssessmentRolesList = exports.AuditmanagerAssessmentRolesOutputReference = exports.auditmanagerAssessmentRolesToHclTerraform = exports.auditmanagerAssessmentRolesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function auditmanagerAssessmentRolesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        role_type: cdktf.stringToTerraform(struct.roleType),
    };
}
exports.auditmanagerAssessmentRolesToTerraform = auditmanagerAssessmentRolesToTerraform;
function auditmanagerAssessmentRolesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_type: {
            value: cdktf.stringToHclTerraform(struct.roleType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerAssessmentRolesToHclTerraform = auditmanagerAssessmentRolesToHclTerraform;
class AuditmanagerAssessmentRolesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._roleType !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleType = this._roleType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._roleArn = undefined;
            this._roleType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._roleArn = value.roleArn;
            this._roleType = value.roleType;
        }
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get roleType() {
        return this.getStringAttribute('role_type');
    }
    set roleType(value) {
        this._roleType = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleTypeInput() {
        return this._roleType;
    }
}
exports.AuditmanagerAssessmentRolesOutputReference = AuditmanagerAssessmentRolesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentRolesOutputReference[_a] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentRolesOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentRolesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentRolesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentRolesList = AuditmanagerAssessmentRolesList;
_b = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentRolesList[_b] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentRolesList", version: "19.15.0" };
function auditmanagerAssessmentRolesAllToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.auditmanagerAssessmentRolesAllToTerraform = auditmanagerAssessmentRolesAllToTerraform;
function auditmanagerAssessmentRolesAllToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.auditmanagerAssessmentRolesAllToHclTerraform = auditmanagerAssessmentRolesAllToHclTerraform;
class AuditmanagerAssessmentRolesAllOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // role_type - computed: true, optional: false, required: false
    get roleType() {
        return this.getStringAttribute('role_type');
    }
}
exports.AuditmanagerAssessmentRolesAllOutputReference = AuditmanagerAssessmentRolesAllOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentRolesAllOutputReference[_c] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentRolesAllOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentRolesAllList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentRolesAllOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentRolesAllList = AuditmanagerAssessmentRolesAllList;
_d = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentRolesAllList[_d] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentRolesAllList", version: "19.15.0" };
function auditmanagerAssessmentAssessmentReportsDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination: cdktf.stringToTerraform(struct.destination),
        destination_type: cdktf.stringToTerraform(struct.destinationType),
    };
}
exports.auditmanagerAssessmentAssessmentReportsDestinationToTerraform = auditmanagerAssessmentAssessmentReportsDestinationToTerraform;
function auditmanagerAssessmentAssessmentReportsDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        destination: {
            value: cdktf.stringToHclTerraform(struct.destination),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        destination_type: {
            value: cdktf.stringToHclTerraform(struct.destinationType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerAssessmentAssessmentReportsDestinationToHclTerraform = auditmanagerAssessmentAssessmentReportsDestinationToHclTerraform;
class AuditmanagerAssessmentAssessmentReportsDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destination !== undefined) {
            hasAnyValues = true;
            internalValueResult.destination = this._destination;
        }
        if (this._destinationType !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationType = this._destinationType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._destination = undefined;
            this._destinationType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._destination = value.destination;
            this._destinationType = value.destinationType;
        }
    }
    get destination() {
        return this.getStringAttribute('destination');
    }
    set destination(value) {
        this._destination = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination;
    }
    get destinationType() {
        return this.getStringAttribute('destination_type');
    }
    set destinationType(value) {
        this._destinationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationTypeInput() {
        return this._destinationType;
    }
}
exports.AuditmanagerAssessmentAssessmentReportsDestinationOutputReference = AuditmanagerAssessmentAssessmentReportsDestinationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentAssessmentReportsDestinationOutputReference[_e] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentAssessmentReportsDestinationOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentAssessmentReportsDestinationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentAssessmentReportsDestinationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentAssessmentReportsDestinationList = AuditmanagerAssessmentAssessmentReportsDestinationList;
_f = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentAssessmentReportsDestinationList[_f] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentAssessmentReportsDestinationList", version: "19.15.0" };
function auditmanagerAssessmentScopeAwsAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
    };
}
exports.auditmanagerAssessmentScopeAwsAccountsToTerraform = auditmanagerAssessmentScopeAwsAccountsToTerraform;
function auditmanagerAssessmentScopeAwsAccountsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerAssessmentScopeAwsAccountsToHclTerraform = auditmanagerAssessmentScopeAwsAccountsToHclTerraform;
class AuditmanagerAssessmentScopeAwsAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
}
exports.AuditmanagerAssessmentScopeAwsAccountsOutputReference = AuditmanagerAssessmentScopeAwsAccountsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeAwsAccountsOutputReference[_g] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeAwsAccountsOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentScopeAwsAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentScopeAwsAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentScopeAwsAccountsList = AuditmanagerAssessmentScopeAwsAccountsList;
_h = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeAwsAccountsList[_h] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeAwsAccountsList", version: "19.15.0" };
function auditmanagerAssessmentScopeAwsServicesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        service_name: cdktf.stringToTerraform(struct.serviceName),
    };
}
exports.auditmanagerAssessmentScopeAwsServicesToTerraform = auditmanagerAssessmentScopeAwsServicesToTerraform;
function auditmanagerAssessmentScopeAwsServicesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        service_name: {
            value: cdktf.stringToHclTerraform(struct.serviceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerAssessmentScopeAwsServicesToHclTerraform = auditmanagerAssessmentScopeAwsServicesToHclTerraform;
class AuditmanagerAssessmentScopeAwsServicesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceName = this._serviceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._serviceName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._serviceName = value.serviceName;
        }
    }
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    set serviceName(value) {
        this._serviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNameInput() {
        return this._serviceName;
    }
}
exports.AuditmanagerAssessmentScopeAwsServicesOutputReference = AuditmanagerAssessmentScopeAwsServicesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeAwsServicesOutputReference[_j] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeAwsServicesOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentScopeAwsServicesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentScopeAwsServicesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentScopeAwsServicesList = AuditmanagerAssessmentScopeAwsServicesList;
_k = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeAwsServicesList[_k] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeAwsServicesList", version: "19.15.0" };
function auditmanagerAssessmentScopeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        aws_accounts: cdktf.listMapper(auditmanagerAssessmentScopeAwsAccountsToTerraform, true)(struct.awsAccounts),
        aws_services: cdktf.listMapper(auditmanagerAssessmentScopeAwsServicesToTerraform, true)(struct.awsServices),
    };
}
exports.auditmanagerAssessmentScopeToTerraform = auditmanagerAssessmentScopeToTerraform;
function auditmanagerAssessmentScopeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        aws_accounts: {
            value: cdktf.listMapperHcl(auditmanagerAssessmentScopeAwsAccountsToHclTerraform, true)(struct.awsAccounts),
            isBlock: true,
            type: "set",
            storageClassType: "AuditmanagerAssessmentScopeAwsAccountsList",
        },
        aws_services: {
            value: cdktf.listMapperHcl(auditmanagerAssessmentScopeAwsServicesToHclTerraform, true)(struct.awsServices),
            isBlock: true,
            type: "set",
            storageClassType: "AuditmanagerAssessmentScopeAwsServicesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.auditmanagerAssessmentScopeToHclTerraform = auditmanagerAssessmentScopeToHclTerraform;
class AuditmanagerAssessmentScopeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // aws_accounts - computed: false, optional: true, required: false
        this._awsAccounts = new AuditmanagerAssessmentScopeAwsAccountsList(this, "aws_accounts", true);
        // aws_services - computed: false, optional: true, required: false
        this._awsServices = new AuditmanagerAssessmentScopeAwsServicesList(this, "aws_services", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._awsAccounts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsAccounts = this._awsAccounts?.internalValue;
        }
        if (this._awsServices?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsServices = this._awsServices?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._awsAccounts.internalValue = undefined;
            this._awsServices.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._awsAccounts.internalValue = value.awsAccounts;
            this._awsServices.internalValue = value.awsServices;
        }
    }
    get awsAccounts() {
        return this._awsAccounts;
    }
    putAwsAccounts(value) {
        this._awsAccounts.internalValue = value;
    }
    resetAwsAccounts() {
        this._awsAccounts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccountsInput() {
        return this._awsAccounts.internalValue;
    }
    get awsServices() {
        return this._awsServices;
    }
    putAwsServices(value) {
        this._awsServices.internalValue = value;
    }
    resetAwsServices() {
        this._awsServices.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsServicesInput() {
        return this._awsServices.internalValue;
    }
}
exports.AuditmanagerAssessmentScopeOutputReference = AuditmanagerAssessmentScopeOutputReference;
_l = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeOutputReference[_l] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeOutputReference", version: "19.15.0" };
class AuditmanagerAssessmentScopeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AuditmanagerAssessmentScopeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AuditmanagerAssessmentScopeList = AuditmanagerAssessmentScopeList;
_m = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessmentScopeList[_m] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessmentScopeList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_assessment aws_auditmanager_assessment}
*/
class AuditmanagerAssessment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AuditmanagerAssessment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AuditmanagerAssessment to import
    * @param importFromId The id of the existing AuditmanagerAssessment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_assessment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AuditmanagerAssessment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_auditmanager_assessment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/auditmanager_assessment aws_auditmanager_assessment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AuditmanagerAssessmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_auditmanager_assessment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // roles - computed: false, optional: false, required: true
        this._roles = new AuditmanagerAssessmentRolesList(this, "roles", true);
        // roles_all - computed: true, optional: false, required: false
        this._rolesAll = new AuditmanagerAssessmentRolesAllList(this, "roles_all", true);
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // assessment_reports_destination - computed: false, optional: true, required: false
        this._assessmentReportsDestination = new AuditmanagerAssessmentAssessmentReportsDestinationList(this, "assessment_reports_destination", false);
        // scope - computed: false, optional: true, required: false
        this._scope = new AuditmanagerAssessmentScopeList(this, "scope", false);
        this._description = config.description;
        this._frameworkId = config.frameworkId;
        this._name = config.name;
        this._roles.internalValue = config.roles;
        this._tags = config.tags;
        this._assessmentReportsDestination.internalValue = config.assessmentReportsDestination;
        this._scope.internalValue = config.scope;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get frameworkId() {
        return this.getStringAttribute('framework_id');
    }
    set frameworkId(value) {
        this._frameworkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get frameworkIdInput() {
        return this._frameworkId;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get roles() {
        return this._roles;
    }
    putRoles(value) {
        this._roles.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles.internalValue;
    }
    get rolesAll() {
        return this._rolesAll;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get assessmentReportsDestination() {
        return this._assessmentReportsDestination;
    }
    putAssessmentReportsDestination(value) {
        this._assessmentReportsDestination.internalValue = value;
    }
    resetAssessmentReportsDestination() {
        this._assessmentReportsDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assessmentReportsDestinationInput() {
        return this._assessmentReportsDestination.internalValue;
    }
    get scope() {
        return this._scope;
    }
    putScope(value) {
        this._scope.internalValue = value;
    }
    resetScope() {
        this._scope.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            framework_id: cdktf.stringToTerraform(this._frameworkId),
            name: cdktf.stringToTerraform(this._name),
            roles: cdktf.listMapper(auditmanagerAssessmentRolesToTerraform, false)(this._roles.internalValue),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            assessment_reports_destination: cdktf.listMapper(auditmanagerAssessmentAssessmentReportsDestinationToTerraform, true)(this._assessmentReportsDestination.internalValue),
            scope: cdktf.listMapper(auditmanagerAssessmentScopeToTerraform, true)(this._scope.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            framework_id: {
                value: cdktf.stringToHclTerraform(this._frameworkId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            roles: {
                value: cdktf.listMapperHcl(auditmanagerAssessmentRolesToHclTerraform, false)(this._roles.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "AuditmanagerAssessmentRolesList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            assessment_reports_destination: {
                value: cdktf.listMapperHcl(auditmanagerAssessmentAssessmentReportsDestinationToHclTerraform, true)(this._assessmentReportsDestination.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AuditmanagerAssessmentAssessmentReportsDestinationList",
            },
            scope: {
                value: cdktf.listMapperHcl(auditmanagerAssessmentScopeToHclTerraform, true)(this._scope.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AuditmanagerAssessmentScopeList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AuditmanagerAssessment = AuditmanagerAssessment;
_o = JSII_RTTI_SYMBOL_1;
AuditmanagerAssessment[_o] = { fqn: "@cdktf/provider-aws.auditmanagerAssessment.AuditmanagerAssessment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AuditmanagerAssessment.tfResourceType = "aws_auditmanager_assessment";
