// @cdktf/provider-aws/lib/data-aws-ecr-lifecycle-policy-document/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcrLifecyclePolicyDocument = exports.DataAwsEcrLifecyclePolicyDocumentRuleList = exports.DataAwsEcrLifecyclePolicyDocumentRuleOutputReference = exports.dataAwsEcrLifecyclePolicyDocumentRuleToHclTerraform = exports.dataAwsEcrLifecyclePolicyDocumentRuleToTerraform = exports.DataAwsEcrLifecyclePolicyDocumentRuleSelectionList = exports.DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference = exports.dataAwsEcrLifecyclePolicyDocumentRuleSelectionToHclTerraform = exports.dataAwsEcrLifecyclePolicyDocumentRuleSelectionToTerraform = exports.DataAwsEcrLifecyclePolicyDocumentRuleActionList = exports.DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference = exports.dataAwsEcrLifecyclePolicyDocumentRuleActionToHclTerraform = exports.dataAwsEcrLifecyclePolicyDocumentRuleActionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEcrLifecyclePolicyDocumentRuleActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleActionToTerraform = dataAwsEcrLifecyclePolicyDocumentRuleActionToTerraform;
function dataAwsEcrLifecyclePolicyDocumentRuleActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleActionToHclTerraform = dataAwsEcrLifecyclePolicyDocumentRuleActionToHclTerraform;
class DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference = DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference", version: "19.15.0" };
class DataAwsEcrLifecyclePolicyDocumentRuleActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcrLifecyclePolicyDocumentRuleActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleActionList = DataAwsEcrLifecyclePolicyDocumentRuleActionList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleActionList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleActionList", version: "19.15.0" };
function dataAwsEcrLifecyclePolicyDocumentRuleSelectionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count_number: cdktf.numberToTerraform(struct.countNumber),
        count_type: cdktf.stringToTerraform(struct.countType),
        count_unit: cdktf.stringToTerraform(struct.countUnit),
        tag_pattern_list: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tagPatternList),
        tag_prefix_list: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tagPrefixList),
        tag_status: cdktf.stringToTerraform(struct.tagStatus),
    };
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleSelectionToTerraform = dataAwsEcrLifecyclePolicyDocumentRuleSelectionToTerraform;
function dataAwsEcrLifecyclePolicyDocumentRuleSelectionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        count_number: {
            value: cdktf.numberToHclTerraform(struct.countNumber),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        count_type: {
            value: cdktf.stringToHclTerraform(struct.countType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        count_unit: {
            value: cdktf.stringToHclTerraform(struct.countUnit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tag_pattern_list: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tagPatternList),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        tag_prefix_list: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tagPrefixList),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        tag_status: {
            value: cdktf.stringToHclTerraform(struct.tagStatus),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleSelectionToHclTerraform = dataAwsEcrLifecyclePolicyDocumentRuleSelectionToHclTerraform;
class DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._countNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.countNumber = this._countNumber;
        }
        if (this._countType !== undefined) {
            hasAnyValues = true;
            internalValueResult.countType = this._countType;
        }
        if (this._countUnit !== undefined) {
            hasAnyValues = true;
            internalValueResult.countUnit = this._countUnit;
        }
        if (this._tagPatternList !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagPatternList = this._tagPatternList;
        }
        if (this._tagPrefixList !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagPrefixList = this._tagPrefixList;
        }
        if (this._tagStatus !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagStatus = this._tagStatus;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._countNumber = undefined;
            this._countType = undefined;
            this._countUnit = undefined;
            this._tagPatternList = undefined;
            this._tagPrefixList = undefined;
            this._tagStatus = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._countNumber = value.countNumber;
            this._countType = value.countType;
            this._countUnit = value.countUnit;
            this._tagPatternList = value.tagPatternList;
            this._tagPrefixList = value.tagPrefixList;
            this._tagStatus = value.tagStatus;
        }
    }
    get countNumber() {
        return this.getNumberAttribute('count_number');
    }
    set countNumber(value) {
        this._countNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get countNumberInput() {
        return this._countNumber;
    }
    get countType() {
        return this.getStringAttribute('count_type');
    }
    set countType(value) {
        this._countType = value;
    }
    // Temporarily expose input value. Use with caution.
    get countTypeInput() {
        return this._countType;
    }
    get countUnit() {
        return this.getStringAttribute('count_unit');
    }
    set countUnit(value) {
        this._countUnit = value;
    }
    resetCountUnit() {
        this._countUnit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countUnitInput() {
        return this._countUnit;
    }
    get tagPatternList() {
        return this.getListAttribute('tag_pattern_list');
    }
    set tagPatternList(value) {
        this._tagPatternList = value;
    }
    resetTagPatternList() {
        this._tagPatternList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagPatternListInput() {
        return this._tagPatternList;
    }
    get tagPrefixList() {
        return this.getListAttribute('tag_prefix_list');
    }
    set tagPrefixList(value) {
        this._tagPrefixList = value;
    }
    resetTagPrefixList() {
        this._tagPrefixList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagPrefixListInput() {
        return this._tagPrefixList;
    }
    get tagStatus() {
        return this.getStringAttribute('tag_status');
    }
    set tagStatus(value) {
        this._tagStatus = value;
    }
    // Temporarily expose input value. Use with caution.
    get tagStatusInput() {
        return this._tagStatus;
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference = DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference", version: "19.15.0" };
class DataAwsEcrLifecyclePolicyDocumentRuleSelectionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcrLifecyclePolicyDocumentRuleSelectionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleSelectionList = DataAwsEcrLifecyclePolicyDocumentRuleSelectionList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleSelectionList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleSelectionList", version: "19.15.0" };
function dataAwsEcrLifecyclePolicyDocumentRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        priority: cdktf.numberToTerraform(struct.priority),
        action: cdktf.listMapper(dataAwsEcrLifecyclePolicyDocumentRuleActionToTerraform, true)(struct.action),
        selection: cdktf.listMapper(dataAwsEcrLifecyclePolicyDocumentRuleSelectionToTerraform, true)(struct.selection),
    };
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleToTerraform = dataAwsEcrLifecyclePolicyDocumentRuleToTerraform;
function dataAwsEcrLifecyclePolicyDocumentRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        priority: {
            value: cdktf.numberToHclTerraform(struct.priority),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        action: {
            value: cdktf.listMapperHcl(dataAwsEcrLifecyclePolicyDocumentRuleActionToHclTerraform, true)(struct.action),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsEcrLifecyclePolicyDocumentRuleActionList",
        },
        selection: {
            value: cdktf.listMapperHcl(dataAwsEcrLifecyclePolicyDocumentRuleSelectionToHclTerraform, true)(struct.selection),
            isBlock: true,
            type: "list",
            storageClassType: "DataAwsEcrLifecyclePolicyDocumentRuleSelectionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataAwsEcrLifecyclePolicyDocumentRuleToHclTerraform = dataAwsEcrLifecyclePolicyDocumentRuleToHclTerraform;
class DataAwsEcrLifecyclePolicyDocumentRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: false, optional: true, required: false
        this._action = new DataAwsEcrLifecyclePolicyDocumentRuleActionList(this, "action", false);
        // selection - computed: false, optional: true, required: false
        this._selection = new DataAwsEcrLifecyclePolicyDocumentRuleSelectionList(this, "selection", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._selection?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.selection = this._selection?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._priority = undefined;
            this._action.internalValue = undefined;
            this._selection.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._priority = value.priority;
            this._action.internalValue = value.action;
            this._selection.internalValue = value.selection;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    resetAction() {
        this._action.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get selection() {
        return this._selection;
    }
    putSelection(value) {
        this._selection.internalValue = value;
    }
    resetSelection() {
        this._selection.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selectionInput() {
        return this._selection.internalValue;
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleOutputReference = DataAwsEcrLifecyclePolicyDocumentRuleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleOutputReference", version: "19.15.0" };
class DataAwsEcrLifecyclePolicyDocumentRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcrLifecyclePolicyDocumentRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcrLifecyclePolicyDocumentRuleList = DataAwsEcrLifecyclePolicyDocumentRuleList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocumentRuleList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocumentRuleList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_lifecycle_policy_document aws_ecr_lifecycle_policy_document}
*/
class DataAwsEcrLifecyclePolicyDocument extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsEcrLifecyclePolicyDocument resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsEcrLifecyclePolicyDocument to import
    * @param importFromId The id of the existing DataAwsEcrLifecyclePolicyDocument that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_lifecycle_policy_document#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsEcrLifecyclePolicyDocument to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_ecr_lifecycle_policy_document", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/ecr_lifecycle_policy_document aws_ecr_lifecycle_policy_document} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEcrLifecyclePolicyDocumentConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_lifecycle_policy_document',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: false, optional: true, required: false
        this._rule = new DataAwsEcrLifecyclePolicyDocumentRuleList(this, "rule", false);
        this._rule.internalValue = config.rule;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // json - computed: true, optional: false, required: false
    get json() {
        return this.getStringAttribute('json');
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            rule: cdktf.listMapper(dataAwsEcrLifecyclePolicyDocumentRuleToTerraform, true)(this._rule.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            rule: {
                value: cdktf.listMapperHcl(dataAwsEcrLifecyclePolicyDocumentRuleToHclTerraform, true)(this._rule.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataAwsEcrLifecyclePolicyDocumentRuleList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsEcrLifecyclePolicyDocument = DataAwsEcrLifecyclePolicyDocument;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEcrLifecyclePolicyDocument[_g] = { fqn: "@cdktf/provider-aws.dataAwsEcrLifecyclePolicyDocument.DataAwsEcrLifecyclePolicyDocument", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEcrLifecyclePolicyDocument.tfResourceType = "aws_ecr_lifecycle_policy_document";
