// @cdktf/provider-aws/lib/opsworks-permission/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksPermission = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opsworks_permission aws_opsworks_permission}
*/
class OpsworksPermission extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a OpsworksPermission resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the OpsworksPermission to import
    * @param importFromId The id of the existing OpsworksPermission that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opsworks_permission#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the OpsworksPermission to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_opsworks_permission", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/opsworks_permission aws_opsworks_permission} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OpsworksPermissionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_permission',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._allowSsh = config.allowSsh;
        this._allowSudo = config.allowSudo;
        this._id = config.id;
        this._level = config.level;
        this._stackId = config.stackId;
        this._userArn = config.userArn;
    }
    get allowSsh() {
        return this.getBooleanAttribute('allow_ssh');
    }
    set allowSsh(value) {
        this._allowSsh = value;
    }
    resetAllowSsh() {
        this._allowSsh = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowSshInput() {
        return this._allowSsh;
    }
    get allowSudo() {
        return this.getBooleanAttribute('allow_sudo');
    }
    set allowSudo(value) {
        this._allowSudo = value;
    }
    resetAllowSudo() {
        this._allowSudo = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowSudoInput() {
        return this._allowSudo;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get level() {
        return this.getStringAttribute('level');
    }
    set level(value) {
        this._level = value;
    }
    resetLevel() {
        this._level = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelInput() {
        return this._level;
    }
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    get stackIdInput() {
        return this._stackId;
    }
    get userArn() {
        return this.getStringAttribute('user_arn');
    }
    set userArn(value) {
        this._userArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get userArnInput() {
        return this._userArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allow_ssh: cdktf.booleanToTerraform(this._allowSsh),
            allow_sudo: cdktf.booleanToTerraform(this._allowSudo),
            id: cdktf.stringToTerraform(this._id),
            level: cdktf.stringToTerraform(this._level),
            stack_id: cdktf.stringToTerraform(this._stackId),
            user_arn: cdktf.stringToTerraform(this._userArn),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            allow_ssh: {
                value: cdktf.booleanToHclTerraform(this._allowSsh),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            allow_sudo: {
                value: cdktf.booleanToHclTerraform(this._allowSudo),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            level: {
                value: cdktf.stringToHclTerraform(this._level),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            stack_id: {
                value: cdktf.stringToHclTerraform(this._stackId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            user_arn: {
                value: cdktf.stringToHclTerraform(this._userArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.OpsworksPermission = OpsworksPermission;
_a = JSII_RTTI_SYMBOL_1;
OpsworksPermission[_a] = { fqn: "@cdktf/provider-aws.opsworksPermission.OpsworksPermission", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
OpsworksPermission.tfResourceType = "aws_opsworks_permission";
