// @cdktf/provider-aws/lib/lambda-alias/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaAlias = exports.LambdaAliasRoutingConfigOutputReference = exports.lambdaAliasRoutingConfigToHclTerraform = exports.lambdaAliasRoutingConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lambdaAliasRoutingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        additional_version_weights: cdktf.hashMapper(cdktf.numberToTerraform)(struct.additionalVersionWeights),
    };
}
exports.lambdaAliasRoutingConfigToTerraform = lambdaAliasRoutingConfigToTerraform;
function lambdaAliasRoutingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        additional_version_weights: {
            value: cdktf.hashMapperHcl(cdktf.numberToHclTerraform)(struct.additionalVersionWeights),
            isBlock: false,
            type: "map",
            storageClassType: "numberMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lambdaAliasRoutingConfigToHclTerraform = lambdaAliasRoutingConfigToHclTerraform;
class LambdaAliasRoutingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._additionalVersionWeights !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalVersionWeights = this._additionalVersionWeights;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._additionalVersionWeights = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._additionalVersionWeights = value.additionalVersionWeights;
        }
    }
    get additionalVersionWeights() {
        return this.getNumberMapAttribute('additional_version_weights');
    }
    set additionalVersionWeights(value) {
        this._additionalVersionWeights = value;
    }
    resetAdditionalVersionWeights() {
        this._additionalVersionWeights = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalVersionWeightsInput() {
        return this._additionalVersionWeights;
    }
}
exports.LambdaAliasRoutingConfigOutputReference = LambdaAliasRoutingConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LambdaAliasRoutingConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.lambdaAlias.LambdaAliasRoutingConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_alias aws_lambda_alias}
*/
class LambdaAlias extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a LambdaAlias resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the LambdaAlias to import
    * @param importFromId The id of the existing LambdaAlias that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_alias#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the LambdaAlias to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lambda_alias", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lambda_alias aws_lambda_alias} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LambdaAliasConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // routing_config - computed: false, optional: true, required: false
        this._routingConfig = new LambdaAliasRoutingConfigOutputReference(this, "routing_config");
        this._description = config.description;
        this._functionName = config.functionName;
        this._functionVersion = config.functionVersion;
        this._id = config.id;
        this._name = config.name;
        this._routingConfig.internalValue = config.routingConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get functionNameInput() {
        return this._functionName;
    }
    get functionVersion() {
        return this.getStringAttribute('function_version');
    }
    set functionVersion(value) {
        this._functionVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get functionVersionInput() {
        return this._functionVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // invoke_arn - computed: true, optional: false, required: false
    get invokeArn() {
        return this.getStringAttribute('invoke_arn');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get routingConfig() {
        return this._routingConfig;
    }
    putRoutingConfig(value) {
        this._routingConfig.internalValue = value;
    }
    resetRoutingConfig() {
        this._routingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routingConfigInput() {
        return this._routingConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            function_name: cdktf.stringToTerraform(this._functionName),
            function_version: cdktf.stringToTerraform(this._functionVersion),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            routing_config: lambdaAliasRoutingConfigToTerraform(this._routingConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            function_name: {
                value: cdktf.stringToHclTerraform(this._functionName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            function_version: {
                value: cdktf.stringToHclTerraform(this._functionVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            routing_config: {
                value: lambdaAliasRoutingConfigToHclTerraform(this._routingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "LambdaAliasRoutingConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.LambdaAlias = LambdaAlias;
_b = JSII_RTTI_SYMBOL_1;
LambdaAlias[_b] = { fqn: "@cdktf/provider-aws.lambdaAlias.LambdaAlias", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
LambdaAlias.tfResourceType = "aws_lambda_alias";
