// @cdktf/provider-aws/lib/data-aws-bedrock-foundation-models/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBedrockFoundationModels = exports.DataAwsBedrockFoundationModelsModelSummariesList = exports.DataAwsBedrockFoundationModelsModelSummariesOutputReference = exports.dataAwsBedrockFoundationModelsModelSummariesToHclTerraform = exports.dataAwsBedrockFoundationModelsModelSummariesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsBedrockFoundationModelsModelSummariesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsBedrockFoundationModelsModelSummariesToTerraform = dataAwsBedrockFoundationModelsModelSummariesToTerraform;
function dataAwsBedrockFoundationModelsModelSummariesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsBedrockFoundationModelsModelSummariesToHclTerraform = dataAwsBedrockFoundationModelsModelSummariesToHclTerraform;
class DataAwsBedrockFoundationModelsModelSummariesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // customizations_supported - computed: true, optional: false, required: false
    get customizationsSupported() {
        return cdktf.Fn.tolist(this.getListAttribute('customizations_supported'));
    }
    // inference_types_supported - computed: true, optional: false, required: false
    get inferenceTypesSupported() {
        return cdktf.Fn.tolist(this.getListAttribute('inference_types_supported'));
    }
    // input_modalities - computed: true, optional: false, required: false
    get inputModalities() {
        return cdktf.Fn.tolist(this.getListAttribute('input_modalities'));
    }
    // model_arn - computed: true, optional: false, required: false
    get modelArn() {
        return this.getStringAttribute('model_arn');
    }
    // model_id - computed: true, optional: false, required: false
    get modelId() {
        return this.getStringAttribute('model_id');
    }
    // model_name - computed: true, optional: false, required: false
    get modelName() {
        return this.getStringAttribute('model_name');
    }
    // output_modalities - computed: true, optional: false, required: false
    get outputModalities() {
        return cdktf.Fn.tolist(this.getListAttribute('output_modalities'));
    }
    // provider_name - computed: true, optional: false, required: false
    get providerName() {
        return this.getStringAttribute('provider_name');
    }
    // response_streaming_supported - computed: true, optional: false, required: false
    get responseStreamingSupported() {
        return this.getBooleanAttribute('response_streaming_supported');
    }
}
exports.DataAwsBedrockFoundationModelsModelSummariesOutputReference = DataAwsBedrockFoundationModelsModelSummariesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBedrockFoundationModelsModelSummariesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsBedrockFoundationModels.DataAwsBedrockFoundationModelsModelSummariesOutputReference", version: "19.15.0" };
class DataAwsBedrockFoundationModelsModelSummariesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsBedrockFoundationModelsModelSummariesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsBedrockFoundationModelsModelSummariesList = DataAwsBedrockFoundationModelsModelSummariesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsBedrockFoundationModelsModelSummariesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsBedrockFoundationModels.DataAwsBedrockFoundationModelsModelSummariesList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/bedrock_foundation_models aws_bedrock_foundation_models}
*/
class DataAwsBedrockFoundationModels extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsBedrockFoundationModels resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsBedrockFoundationModels to import
    * @param importFromId The id of the existing DataAwsBedrockFoundationModels that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/bedrock_foundation_models#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsBedrockFoundationModels to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_bedrock_foundation_models", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/bedrock_foundation_models aws_bedrock_foundation_models} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsBedrockFoundationModelsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_bedrock_foundation_models',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // model_summaries - computed: true, optional: false, required: false
        this._modelSummaries = new DataAwsBedrockFoundationModelsModelSummariesList(this, "model_summaries", false);
        this._byCustomizationType = config.byCustomizationType;
        this._byInferenceType = config.byInferenceType;
        this._byOutputModality = config.byOutputModality;
        this._byProvider = config.byProvider;
    }
    get byCustomizationType() {
        return this.getStringAttribute('by_customization_type');
    }
    set byCustomizationType(value) {
        this._byCustomizationType = value;
    }
    resetByCustomizationType() {
        this._byCustomizationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get byCustomizationTypeInput() {
        return this._byCustomizationType;
    }
    get byInferenceType() {
        return this.getStringAttribute('by_inference_type');
    }
    set byInferenceType(value) {
        this._byInferenceType = value;
    }
    resetByInferenceType() {
        this._byInferenceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get byInferenceTypeInput() {
        return this._byInferenceType;
    }
    get byOutputModality() {
        return this.getStringAttribute('by_output_modality');
    }
    set byOutputModality(value) {
        this._byOutputModality = value;
    }
    resetByOutputModality() {
        this._byOutputModality = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get byOutputModalityInput() {
        return this._byOutputModality;
    }
    get byProvider() {
        return this.getStringAttribute('by_provider');
    }
    set byProvider(value) {
        this._byProvider = value;
    }
    resetByProvider() {
        this._byProvider = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get byProviderInput() {
        return this._byProvider;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get modelSummaries() {
        return this._modelSummaries;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            by_customization_type: cdktf.stringToTerraform(this._byCustomizationType),
            by_inference_type: cdktf.stringToTerraform(this._byInferenceType),
            by_output_modality: cdktf.stringToTerraform(this._byOutputModality),
            by_provider: cdktf.stringToTerraform(this._byProvider),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            by_customization_type: {
                value: cdktf.stringToHclTerraform(this._byCustomizationType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            by_inference_type: {
                value: cdktf.stringToHclTerraform(this._byInferenceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            by_output_modality: {
                value: cdktf.stringToHclTerraform(this._byOutputModality),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            by_provider: {
                value: cdktf.stringToHclTerraform(this._byProvider),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsBedrockFoundationModels = DataAwsBedrockFoundationModels;
_c = JSII_RTTI_SYMBOL_1;
DataAwsBedrockFoundationModels[_c] = { fqn: "@cdktf/provider-aws.dataAwsBedrockFoundationModels.DataAwsBedrockFoundationModels", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsBedrockFoundationModels.tfResourceType = "aws_bedrock_foundation_models";
