// @cdktf/provider-aws/lib/s3-bucket-ownership-controls/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketOwnershipControls = exports.S3BucketOwnershipControlsRuleOutputReference = exports.s3BucketOwnershipControlsRuleToHclTerraform = exports.s3BucketOwnershipControlsRuleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketOwnershipControlsRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        object_ownership: cdktf.stringToTerraform(struct.objectOwnership),
    };
}
exports.s3BucketOwnershipControlsRuleToTerraform = s3BucketOwnershipControlsRuleToTerraform;
function s3BucketOwnershipControlsRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        object_ownership: {
            value: cdktf.stringToHclTerraform(struct.objectOwnership),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.s3BucketOwnershipControlsRuleToHclTerraform = s3BucketOwnershipControlsRuleToHclTerraform;
class S3BucketOwnershipControlsRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._objectOwnership !== undefined) {
            hasAnyValues = true;
            internalValueResult.objectOwnership = this._objectOwnership;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._objectOwnership = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._objectOwnership = value.objectOwnership;
        }
    }
    get objectOwnership() {
        return this.getStringAttribute('object_ownership');
    }
    set objectOwnership(value) {
        this._objectOwnership = value;
    }
    // Temporarily expose input value. Use with caution.
    get objectOwnershipInput() {
        return this._objectOwnership;
    }
}
exports.S3BucketOwnershipControlsRuleOutputReference = S3BucketOwnershipControlsRuleOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketOwnershipControlsRuleOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketOwnershipControls.S3BucketOwnershipControlsRuleOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_ownership_controls aws_s3_bucket_ownership_controls}
*/
class S3BucketOwnershipControls extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a S3BucketOwnershipControls resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the S3BucketOwnershipControls to import
    * @param importFromId The id of the existing S3BucketOwnershipControls that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_ownership_controls#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the S3BucketOwnershipControls to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_s3_bucket_ownership_controls", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/s3_bucket_ownership_controls aws_s3_bucket_ownership_controls} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketOwnershipControlsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_ownership_controls',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: false, optional: false, required: true
        this._rule = new S3BucketOwnershipControlsRuleOutputReference(this, "rule");
        this._bucket = config.bucket;
        this._id = config.id;
        this._rule.internalValue = config.rule;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            id: cdktf.stringToTerraform(this._id),
            rule: s3BucketOwnershipControlsRuleToTerraform(this._rule.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bucket: {
                value: cdktf.stringToHclTerraform(this._bucket),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rule: {
                value: s3BucketOwnershipControlsRuleToHclTerraform(this._rule.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "S3BucketOwnershipControlsRuleList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.S3BucketOwnershipControls = S3BucketOwnershipControls;
_b = JSII_RTTI_SYMBOL_1;
S3BucketOwnershipControls[_b] = { fqn: "@cdktf/provider-aws.s3BucketOwnershipControls.S3BucketOwnershipControls", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
S3BucketOwnershipControls.tfResourceType = "aws_s3_bucket_ownership_controls";
