// @cdktf/provider-aws/lib/lexv2models-slot-type/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lexv2ModelsSlotType = exports.Lexv2ModelsSlotTypeValueSelectionSettingList = exports.Lexv2ModelsSlotTypeValueSelectionSettingOutputReference = exports.lexv2ModelsSlotTypeValueSelectionSettingToHclTerraform = exports.lexv2ModelsSlotTypeValueSelectionSettingToTerraform = exports.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList = exports.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference = exports.lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToHclTerraform = exports.lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToTerraform = exports.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList = exports.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference = exports.lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToHclTerraform = exports.lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToTerraform = exports.Lexv2ModelsSlotTypeTimeoutsOutputReference = exports.lexv2ModelsSlotTypeTimeoutsToHclTerraform = exports.lexv2ModelsSlotTypeTimeoutsToTerraform = exports.Lexv2ModelsSlotTypeSlotTypeValuesList = exports.Lexv2ModelsSlotTypeSlotTypeValuesOutputReference = exports.lexv2ModelsSlotTypeSlotTypeValuesToHclTerraform = exports.lexv2ModelsSlotTypeSlotTypeValuesToTerraform = exports.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList = exports.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference = exports.lexv2ModelsSlotTypeSlotTypeValuesSynonymsToHclTerraform = exports.lexv2ModelsSlotTypeSlotTypeValuesSynonymsToTerraform = exports.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList = exports.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference = exports.lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToHclTerraform = exports.lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToTerraform = exports.Lexv2ModelsSlotTypeExternalSourceSettingList = exports.Lexv2ModelsSlotTypeExternalSourceSettingOutputReference = exports.lexv2ModelsSlotTypeExternalSourceSettingToHclTerraform = exports.lexv2ModelsSlotTypeExternalSourceSettingToTerraform = exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList = exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference = exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToHclTerraform = exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToTerraform = exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList = exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference = exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToHclTerraform = exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToTerraform = exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingList = exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference = exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingToHclTerraform = exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingToTerraform = exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList = exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference = exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToHclTerraform = exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        slot_type_id: cdktf.stringToTerraform(struct.slotTypeId),
    };
}
exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToTerraform = lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToTerraform;
function lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        slot_type_id: {
            value: cdktf.stringToHclTerraform(struct.slotTypeId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToHclTerraform = lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToHclTerraform;
class Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._slotTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.slotTypeId = this._slotTypeId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._slotTypeId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._slotTypeId = value.slotTypeId;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get slotTypeId() {
        return this.getStringAttribute('slot_type_id');
    }
    set slotTypeId(value) {
        this._slotTypeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get slotTypeIdInput() {
        return this._slotTypeId;
    }
}
exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference = Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference[_a] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList = Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList;
_b = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList[_b] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList", version: "19.15.0" };
function lexv2ModelsSlotTypeCompositeSlotTypeSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subslots: cdktf.listMapper(lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToTerraform, true)(struct.subslots),
    };
}
exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingToTerraform = lexv2ModelsSlotTypeCompositeSlotTypeSettingToTerraform;
function lexv2ModelsSlotTypeCompositeSlotTypeSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        subslots: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsToHclTerraform, true)(struct.subslots),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeCompositeSlotTypeSettingToHclTerraform = lexv2ModelsSlotTypeCompositeSlotTypeSettingToHclTerraform;
class Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // subslots - computed: false, optional: true, required: false
        this._subslots = new Lexv2ModelsSlotTypeCompositeSlotTypeSettingSubslotsList(this, "subslots", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subslots?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.subslots = this._subslots?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._subslots.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._subslots.internalValue = value.subslots;
        }
    }
    get subslots() {
        return this._subslots;
    }
    putSubslots(value) {
        this._subslots.internalValue = value;
    }
    resetSubslots() {
        this._subslots.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subslotsInput() {
        return this._subslots.internalValue;
    }
}
exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference = Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference[_c] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeCompositeSlotTypeSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeCompositeSlotTypeSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeCompositeSlotTypeSettingList = Lexv2ModelsSlotTypeCompositeSlotTypeSettingList;
_d = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeCompositeSlotTypeSettingList[_d] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeCompositeSlotTypeSettingList", version: "19.15.0" };
function lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_object_key: cdktf.stringToTerraform(struct.s3ObjectKey),
    };
}
exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToTerraform = lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToTerraform;
function lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_arn: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_bucket_name: {
            value: cdktf.stringToHclTerraform(struct.s3BucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_object_key: {
            value: cdktf.stringToHclTerraform(struct.s3ObjectKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToHclTerraform = lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToHclTerraform;
class Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyArn = this._kmsKeyArn;
        }
        if (this._s3BucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3ObjectKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3ObjectKey = this._s3ObjectKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._kmsKeyArn = undefined;
            this._s3BucketName = undefined;
            this._s3ObjectKey = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._kmsKeyArn = value.kmsKeyArn;
            this._s3BucketName = value.s3BucketName;
            this._s3ObjectKey = value.s3ObjectKey;
        }
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    get s3ObjectKey() {
        return this.getStringAttribute('s3_object_key');
    }
    set s3ObjectKey(value) {
        this._s3ObjectKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3ObjectKeyInput() {
        return this._s3ObjectKey;
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference = Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference;
_e = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference[_e] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList = Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList;
_f = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList[_f] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList", version: "19.15.0" };
function lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source: cdktf.listMapper(lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToTerraform, true)(struct.source),
    };
}
exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToTerraform = lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToTerraform;
function lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceToHclTerraform, true)(struct.source),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToHclTerraform = lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToHclTerraform;
class Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // source - computed: false, optional: true, required: false
        this._source = new Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingSourceList(this, "source", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._source?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._source.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._source.internalValue = value.source;
        }
    }
    get source() {
        return this._source;
    }
    putSource(value) {
        this._source.internalValue = value;
    }
    resetSource() {
        this._source.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source.internalValue;
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference = Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference;
_g = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference[_g] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList = Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList;
_h = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList[_h] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList", version: "19.15.0" };
function lexv2ModelsSlotTypeExternalSourceSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grammar_slot_type_setting: cdktf.listMapper(lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToTerraform, true)(struct.grammarSlotTypeSetting),
    };
}
exports.lexv2ModelsSlotTypeExternalSourceSettingToTerraform = lexv2ModelsSlotTypeExternalSourceSettingToTerraform;
function lexv2ModelsSlotTypeExternalSourceSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        grammar_slot_type_setting: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingToHclTerraform, true)(struct.grammarSlotTypeSetting),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeExternalSourceSettingToHclTerraform = lexv2ModelsSlotTypeExternalSourceSettingToHclTerraform;
class Lexv2ModelsSlotTypeExternalSourceSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // grammar_slot_type_setting - computed: false, optional: true, required: false
        this._grammarSlotTypeSetting = new Lexv2ModelsSlotTypeExternalSourceSettingGrammarSlotTypeSettingList(this, "grammar_slot_type_setting", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._grammarSlotTypeSetting?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grammarSlotTypeSetting = this._grammarSlotTypeSetting?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._grammarSlotTypeSetting.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._grammarSlotTypeSetting.internalValue = value.grammarSlotTypeSetting;
        }
    }
    get grammarSlotTypeSetting() {
        return this._grammarSlotTypeSetting;
    }
    putGrammarSlotTypeSetting(value) {
        this._grammarSlotTypeSetting.internalValue = value;
    }
    resetGrammarSlotTypeSetting() {
        this._grammarSlotTypeSetting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grammarSlotTypeSettingInput() {
        return this._grammarSlotTypeSetting.internalValue;
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingOutputReference = Lexv2ModelsSlotTypeExternalSourceSettingOutputReference;
_j = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingOutputReference[_j] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeExternalSourceSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeExternalSourceSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeExternalSourceSettingList = Lexv2ModelsSlotTypeExternalSourceSettingList;
_k = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeExternalSourceSettingList[_k] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeExternalSourceSettingList", version: "19.15.0" };
function lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToTerraform = lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToTerraform;
function lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToHclTerraform = lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToHclTerraform;
class Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._value = value.value;
        }
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference = Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference;
_l = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference[_l] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList = Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList;
_m = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList[_m] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList", version: "19.15.0" };
function lexv2ModelsSlotTypeSlotTypeValuesSynonymsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.lexv2ModelsSlotTypeSlotTypeValuesSynonymsToTerraform = lexv2ModelsSlotTypeSlotTypeValuesSynonymsToTerraform;
function lexv2ModelsSlotTypeSlotTypeValuesSynonymsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeSlotTypeValuesSynonymsToHclTerraform = lexv2ModelsSlotTypeSlotTypeValuesSynonymsToHclTerraform;
class Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._value = value.value;
        }
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference = Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference[_o] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeSlotTypeValuesSynonymsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList = Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList;
_p = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList[_p] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList", version: "19.15.0" };
function lexv2ModelsSlotTypeSlotTypeValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        slot_type_value: cdktf.listMapper(lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToTerraform, true)(struct.slotTypeValue),
        synonyms: cdktf.listMapper(lexv2ModelsSlotTypeSlotTypeValuesSynonymsToTerraform, true)(struct.synonyms),
    };
}
exports.lexv2ModelsSlotTypeSlotTypeValuesToTerraform = lexv2ModelsSlotTypeSlotTypeValuesToTerraform;
function lexv2ModelsSlotTypeSlotTypeValuesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        slot_type_value: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueToHclTerraform, true)(struct.slotTypeValue),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList",
        },
        synonyms: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeSlotTypeValuesSynonymsToHclTerraform, true)(struct.synonyms),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeSlotTypeValuesToHclTerraform = lexv2ModelsSlotTypeSlotTypeValuesToHclTerraform;
class Lexv2ModelsSlotTypeSlotTypeValuesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // slot_type_value - computed: false, optional: true, required: false
        this._slotTypeValue = new Lexv2ModelsSlotTypeSlotTypeValuesSlotTypeValueList(this, "slot_type_value", false);
        // synonyms - computed: false, optional: true, required: false
        this._synonyms = new Lexv2ModelsSlotTypeSlotTypeValuesSynonymsList(this, "synonyms", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._slotTypeValue?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.slotTypeValue = this._slotTypeValue?.internalValue;
        }
        if (this._synonyms?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.synonyms = this._synonyms?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._slotTypeValue.internalValue = undefined;
            this._synonyms.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._slotTypeValue.internalValue = value.slotTypeValue;
            this._synonyms.internalValue = value.synonyms;
        }
    }
    get slotTypeValue() {
        return this._slotTypeValue;
    }
    putSlotTypeValue(value) {
        this._slotTypeValue.internalValue = value;
    }
    resetSlotTypeValue() {
        this._slotTypeValue.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get slotTypeValueInput() {
        return this._slotTypeValue.internalValue;
    }
    get synonyms() {
        return this._synonyms;
    }
    putSynonyms(value) {
        this._synonyms.internalValue = value;
    }
    resetSynonyms() {
        this._synonyms.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get synonymsInput() {
        return this._synonyms.internalValue;
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesOutputReference = Lexv2ModelsSlotTypeSlotTypeValuesOutputReference;
_q = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesOutputReference[_q] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeSlotTypeValuesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeSlotTypeValuesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeSlotTypeValuesList = Lexv2ModelsSlotTypeSlotTypeValuesList;
_r = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeSlotTypeValuesList[_r] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeSlotTypeValuesList", version: "19.15.0" };
function lexv2ModelsSlotTypeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lexv2ModelsSlotTypeTimeoutsToTerraform = lexv2ModelsSlotTypeTimeoutsToTerraform;
function lexv2ModelsSlotTypeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeTimeoutsToHclTerraform = lexv2ModelsSlotTypeTimeoutsToHclTerraform;
class Lexv2ModelsSlotTypeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Lexv2ModelsSlotTypeTimeoutsOutputReference = Lexv2ModelsSlotTypeTimeoutsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeTimeoutsOutputReference[_s] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeTimeoutsOutputReference", version: "19.15.0" };
function lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audio_recognition_setting: cdktf.stringToTerraform(struct.audioRecognitionSetting),
    };
}
exports.lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToTerraform = lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToTerraform;
function lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audio_recognition_setting: {
            value: cdktf.stringToHclTerraform(struct.audioRecognitionSetting),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToHclTerraform = lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToHclTerraform;
class Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audioRecognitionSetting !== undefined) {
            hasAnyValues = true;
            internalValueResult.audioRecognitionSetting = this._audioRecognitionSetting;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._audioRecognitionSetting = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._audioRecognitionSetting = value.audioRecognitionSetting;
        }
    }
    get audioRecognitionSetting() {
        return this.getStringAttribute('audio_recognition_setting');
    }
    set audioRecognitionSetting(value) {
        this._audioRecognitionSetting = value;
    }
    resetAudioRecognitionSetting() {
        this._audioRecognitionSetting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get audioRecognitionSettingInput() {
        return this._audioRecognitionSetting;
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference = Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference;
_t = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference[_t] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList = Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList;
_u = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList[_u] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList", version: "19.15.0" };
function lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pattern: cdktf.stringToTerraform(struct.pattern),
    };
}
exports.lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToTerraform = lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToTerraform;
function lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pattern: {
            value: cdktf.stringToHclTerraform(struct.pattern),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToHclTerraform = lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToHclTerraform;
class Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.pattern = this._pattern;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._pattern = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._pattern = value.pattern;
        }
    }
    get pattern() {
        return this.getStringAttribute('pattern');
    }
    set pattern(value) {
        this._pattern = value;
    }
    // Temporarily expose input value. Use with caution.
    get patternInput() {
        return this._pattern;
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference = Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference;
_v = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference[_v] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList = Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList;
_w = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList[_w] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList", version: "19.15.0" };
function lexv2ModelsSlotTypeValueSelectionSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resolution_strategy: cdktf.stringToTerraform(struct.resolutionStrategy),
        advanced_recognition_setting: cdktf.listMapper(lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToTerraform, true)(struct.advancedRecognitionSetting),
        regex_filter: cdktf.listMapper(lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToTerraform, true)(struct.regexFilter),
    };
}
exports.lexv2ModelsSlotTypeValueSelectionSettingToTerraform = lexv2ModelsSlotTypeValueSelectionSettingToTerraform;
function lexv2ModelsSlotTypeValueSelectionSettingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resolution_strategy: {
            value: cdktf.stringToHclTerraform(struct.resolutionStrategy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        advanced_recognition_setting: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingToHclTerraform, true)(struct.advancedRecognitionSetting),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList",
        },
        regex_filter: {
            value: cdktf.listMapperHcl(lexv2ModelsSlotTypeValueSelectionSettingRegexFilterToHclTerraform, true)(struct.regexFilter),
            isBlock: true,
            type: "list",
            storageClassType: "Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsSlotTypeValueSelectionSettingToHclTerraform = lexv2ModelsSlotTypeValueSelectionSettingToHclTerraform;
class Lexv2ModelsSlotTypeValueSelectionSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // advanced_recognition_setting - computed: false, optional: true, required: false
        this._advancedRecognitionSetting = new Lexv2ModelsSlotTypeValueSelectionSettingAdvancedRecognitionSettingList(this, "advanced_recognition_setting", false);
        // regex_filter - computed: false, optional: true, required: false
        this._regexFilter = new Lexv2ModelsSlotTypeValueSelectionSettingRegexFilterList(this, "regex_filter", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resolutionStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.resolutionStrategy = this._resolutionStrategy;
        }
        if (this._advancedRecognitionSetting?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.advancedRecognitionSetting = this._advancedRecognitionSetting?.internalValue;
        }
        if (this._regexFilter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.regexFilter = this._regexFilter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._resolutionStrategy = undefined;
            this._advancedRecognitionSetting.internalValue = undefined;
            this._regexFilter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._resolutionStrategy = value.resolutionStrategy;
            this._advancedRecognitionSetting.internalValue = value.advancedRecognitionSetting;
            this._regexFilter.internalValue = value.regexFilter;
        }
    }
    get resolutionStrategy() {
        return this.getStringAttribute('resolution_strategy');
    }
    set resolutionStrategy(value) {
        this._resolutionStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get resolutionStrategyInput() {
        return this._resolutionStrategy;
    }
    get advancedRecognitionSetting() {
        return this._advancedRecognitionSetting;
    }
    putAdvancedRecognitionSetting(value) {
        this._advancedRecognitionSetting.internalValue = value;
    }
    resetAdvancedRecognitionSetting() {
        this._advancedRecognitionSetting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedRecognitionSettingInput() {
        return this._advancedRecognitionSetting.internalValue;
    }
    get regexFilter() {
        return this._regexFilter;
    }
    putRegexFilter(value) {
        this._regexFilter.internalValue = value;
    }
    resetRegexFilter() {
        this._regexFilter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexFilterInput() {
        return this._regexFilter.internalValue;
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingOutputReference = Lexv2ModelsSlotTypeValueSelectionSettingOutputReference;
_x = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingOutputReference[_x] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingOutputReference", version: "19.15.0" };
class Lexv2ModelsSlotTypeValueSelectionSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsSlotTypeValueSelectionSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsSlotTypeValueSelectionSettingList = Lexv2ModelsSlotTypeValueSelectionSettingList;
_y = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotTypeValueSelectionSettingList[_y] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotTypeValueSelectionSettingList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_slot_type aws_lexv2models_slot_type}
*/
class Lexv2ModelsSlotType extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Lexv2ModelsSlotType resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Lexv2ModelsSlotType to import
    * @param importFromId The id of the existing Lexv2ModelsSlotType that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_slot_type#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Lexv2ModelsSlotType to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lexv2models_slot_type", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_slot_type aws_lexv2models_slot_type} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Lexv2ModelsSlotTypeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lexv2models_slot_type',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // composite_slot_type_setting - computed: false, optional: true, required: false
        this._compositeSlotTypeSetting = new Lexv2ModelsSlotTypeCompositeSlotTypeSettingList(this, "composite_slot_type_setting", false);
        // external_source_setting - computed: false, optional: true, required: false
        this._externalSourceSetting = new Lexv2ModelsSlotTypeExternalSourceSettingList(this, "external_source_setting", false);
        // slot_type_values - computed: false, optional: true, required: false
        this._slotTypeValues = new Lexv2ModelsSlotTypeSlotTypeValuesList(this, "slot_type_values", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Lexv2ModelsSlotTypeTimeoutsOutputReference(this, "timeouts");
        // value_selection_setting - computed: false, optional: true, required: false
        this._valueSelectionSetting = new Lexv2ModelsSlotTypeValueSelectionSettingList(this, "value_selection_setting", false);
        this._botId = config.botId;
        this._botVersion = config.botVersion;
        this._description = config.description;
        this._localeId = config.localeId;
        this._name = config.name;
        this._parentSlotTypeSignature = config.parentSlotTypeSignature;
        this._compositeSlotTypeSetting.internalValue = config.compositeSlotTypeSetting;
        this._externalSourceSetting.internalValue = config.externalSourceSetting;
        this._slotTypeValues.internalValue = config.slotTypeValues;
        this._timeouts.internalValue = config.timeouts;
        this._valueSelectionSetting.internalValue = config.valueSelectionSetting;
    }
    get botId() {
        return this.getStringAttribute('bot_id');
    }
    set botId(value) {
        this._botId = value;
    }
    // Temporarily expose input value. Use with caution.
    get botIdInput() {
        return this._botId;
    }
    get botVersion() {
        return this.getStringAttribute('bot_version');
    }
    set botVersion(value) {
        this._botVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get botVersionInput() {
        return this._botVersion;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get localeId() {
        return this.getStringAttribute('locale_id');
    }
    set localeId(value) {
        this._localeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get localeIdInput() {
        return this._localeId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parentSlotTypeSignature() {
        return this.getStringAttribute('parent_slot_type_signature');
    }
    set parentSlotTypeSignature(value) {
        this._parentSlotTypeSignature = value;
    }
    resetParentSlotTypeSignature() {
        this._parentSlotTypeSignature = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentSlotTypeSignatureInput() {
        return this._parentSlotTypeSignature;
    }
    // slot_type_id - computed: true, optional: false, required: false
    get slotTypeId() {
        return this.getStringAttribute('slot_type_id');
    }
    get compositeSlotTypeSetting() {
        return this._compositeSlotTypeSetting;
    }
    putCompositeSlotTypeSetting(value) {
        this._compositeSlotTypeSetting.internalValue = value;
    }
    resetCompositeSlotTypeSetting() {
        this._compositeSlotTypeSetting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get compositeSlotTypeSettingInput() {
        return this._compositeSlotTypeSetting.internalValue;
    }
    get externalSourceSetting() {
        return this._externalSourceSetting;
    }
    putExternalSourceSetting(value) {
        this._externalSourceSetting.internalValue = value;
    }
    resetExternalSourceSetting() {
        this._externalSourceSetting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get externalSourceSettingInput() {
        return this._externalSourceSetting.internalValue;
    }
    get slotTypeValues() {
        return this._slotTypeValues;
    }
    putSlotTypeValues(value) {
        this._slotTypeValues.internalValue = value;
    }
    resetSlotTypeValues() {
        this._slotTypeValues.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get slotTypeValuesInput() {
        return this._slotTypeValues.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get valueSelectionSetting() {
        return this._valueSelectionSetting;
    }
    putValueSelectionSetting(value) {
        this._valueSelectionSetting.internalValue = value;
    }
    resetValueSelectionSetting() {
        this._valueSelectionSetting.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueSelectionSettingInput() {
        return this._valueSelectionSetting.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bot_id: cdktf.stringToTerraform(this._botId),
            bot_version: cdktf.stringToTerraform(this._botVersion),
            description: cdktf.stringToTerraform(this._description),
            locale_id: cdktf.stringToTerraform(this._localeId),
            name: cdktf.stringToTerraform(this._name),
            parent_slot_type_signature: cdktf.stringToTerraform(this._parentSlotTypeSignature),
            composite_slot_type_setting: cdktf.listMapper(lexv2ModelsSlotTypeCompositeSlotTypeSettingToTerraform, true)(this._compositeSlotTypeSetting.internalValue),
            external_source_setting: cdktf.listMapper(lexv2ModelsSlotTypeExternalSourceSettingToTerraform, true)(this._externalSourceSetting.internalValue),
            slot_type_values: cdktf.listMapper(lexv2ModelsSlotTypeSlotTypeValuesToTerraform, true)(this._slotTypeValues.internalValue),
            timeouts: lexv2ModelsSlotTypeTimeoutsToTerraform(this._timeouts.internalValue),
            value_selection_setting: cdktf.listMapper(lexv2ModelsSlotTypeValueSelectionSettingToTerraform, true)(this._valueSelectionSetting.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bot_id: {
                value: cdktf.stringToHclTerraform(this._botId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bot_version: {
                value: cdktf.stringToHclTerraform(this._botVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            locale_id: {
                value: cdktf.stringToHclTerraform(this._localeId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent_slot_type_signature: {
                value: cdktf.stringToHclTerraform(this._parentSlotTypeSignature),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            composite_slot_type_setting: {
                value: cdktf.listMapperHcl(lexv2ModelsSlotTypeCompositeSlotTypeSettingToHclTerraform, true)(this._compositeSlotTypeSetting.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Lexv2ModelsSlotTypeCompositeSlotTypeSettingList",
            },
            external_source_setting: {
                value: cdktf.listMapperHcl(lexv2ModelsSlotTypeExternalSourceSettingToHclTerraform, true)(this._externalSourceSetting.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Lexv2ModelsSlotTypeExternalSourceSettingList",
            },
            slot_type_values: {
                value: cdktf.listMapperHcl(lexv2ModelsSlotTypeSlotTypeValuesToHclTerraform, true)(this._slotTypeValues.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Lexv2ModelsSlotTypeSlotTypeValuesList",
            },
            timeouts: {
                value: lexv2ModelsSlotTypeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "Lexv2ModelsSlotTypeTimeouts",
            },
            value_selection_setting: {
                value: cdktf.listMapperHcl(lexv2ModelsSlotTypeValueSelectionSettingToHclTerraform, true)(this._valueSelectionSetting.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Lexv2ModelsSlotTypeValueSelectionSettingList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Lexv2ModelsSlotType = Lexv2ModelsSlotType;
_z = JSII_RTTI_SYMBOL_1;
Lexv2ModelsSlotType[_z] = { fqn: "@cdktf/provider-aws.lexv2ModelsSlotType.Lexv2ModelsSlotType", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Lexv2ModelsSlotType.tfResourceType = "aws_lexv2models_slot_type";
