// @cdktf/provider-aws/lib/networkmanager-vpc-attachment/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkmanagerVpcAttachment = exports.NetworkmanagerVpcAttachmentTimeoutsOutputReference = exports.networkmanagerVpcAttachmentTimeoutsToHclTerraform = exports.networkmanagerVpcAttachmentTimeoutsToTerraform = exports.NetworkmanagerVpcAttachmentOptionsOutputReference = exports.networkmanagerVpcAttachmentOptionsToHclTerraform = exports.networkmanagerVpcAttachmentOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkmanagerVpcAttachmentOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        appliance_mode_support: cdktf.booleanToTerraform(struct.applianceModeSupport),
        ipv6_support: cdktf.booleanToTerraform(struct.ipv6Support),
    };
}
exports.networkmanagerVpcAttachmentOptionsToTerraform = networkmanagerVpcAttachmentOptionsToTerraform;
function networkmanagerVpcAttachmentOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        appliance_mode_support: {
            value: cdktf.booleanToHclTerraform(struct.applianceModeSupport),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        ipv6_support: {
            value: cdktf.booleanToHclTerraform(struct.ipv6Support),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerVpcAttachmentOptionsToHclTerraform = networkmanagerVpcAttachmentOptionsToHclTerraform;
class NetworkmanagerVpcAttachmentOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._applianceModeSupport !== undefined) {
            hasAnyValues = true;
            internalValueResult.applianceModeSupport = this._applianceModeSupport;
        }
        if (this._ipv6Support !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6Support = this._ipv6Support;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._applianceModeSupport = undefined;
            this._ipv6Support = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._applianceModeSupport = value.applianceModeSupport;
            this._ipv6Support = value.ipv6Support;
        }
    }
    get applianceModeSupport() {
        return this.getBooleanAttribute('appliance_mode_support');
    }
    set applianceModeSupport(value) {
        this._applianceModeSupport = value;
    }
    resetApplianceModeSupport() {
        this._applianceModeSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applianceModeSupportInput() {
        return this._applianceModeSupport;
    }
    get ipv6Support() {
        return this.getBooleanAttribute('ipv6_support');
    }
    set ipv6Support(value) {
        this._ipv6Support = value;
    }
    resetIpv6Support() {
        this._ipv6Support = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6SupportInput() {
        return this._ipv6Support;
    }
}
exports.NetworkmanagerVpcAttachmentOptionsOutputReference = NetworkmanagerVpcAttachmentOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkmanagerVpcAttachmentOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkmanagerVpcAttachment.NetworkmanagerVpcAttachmentOptionsOutputReference", version: "19.15.0" };
function networkmanagerVpcAttachmentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkmanagerVpcAttachmentTimeoutsToTerraform = networkmanagerVpcAttachmentTimeoutsToTerraform;
function networkmanagerVpcAttachmentTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkmanagerVpcAttachmentTimeoutsToHclTerraform = networkmanagerVpcAttachmentTimeoutsToHclTerraform;
class NetworkmanagerVpcAttachmentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkmanagerVpcAttachmentTimeoutsOutputReference = NetworkmanagerVpcAttachmentTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
NetworkmanagerVpcAttachmentTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.networkmanagerVpcAttachment.NetworkmanagerVpcAttachmentTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_vpc_attachment aws_networkmanager_vpc_attachment}
*/
class NetworkmanagerVpcAttachment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkmanagerVpcAttachment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkmanagerVpcAttachment to import
    * @param importFromId The id of the existing NetworkmanagerVpcAttachment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_vpc_attachment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkmanagerVpcAttachment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_networkmanager_vpc_attachment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/networkmanager_vpc_attachment aws_networkmanager_vpc_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkmanagerVpcAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkmanager_vpc_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // options - computed: false, optional: true, required: false
        this._options = new NetworkmanagerVpcAttachmentOptionsOutputReference(this, "options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkmanagerVpcAttachmentTimeoutsOutputReference(this, "timeouts");
        this._coreNetworkId = config.coreNetworkId;
        this._id = config.id;
        this._subnetArns = config.subnetArns;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcArn = config.vpcArn;
        this._options.internalValue = config.options;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attachment_policy_rule_number - computed: true, optional: false, required: false
    get attachmentPolicyRuleNumber() {
        return this.getNumberAttribute('attachment_policy_rule_number');
    }
    // attachment_type - computed: true, optional: false, required: false
    get attachmentType() {
        return this.getStringAttribute('attachment_type');
    }
    // core_network_arn - computed: true, optional: false, required: false
    get coreNetworkArn() {
        return this.getStringAttribute('core_network_arn');
    }
    get coreNetworkId() {
        return this.getStringAttribute('core_network_id');
    }
    set coreNetworkId(value) {
        this._coreNetworkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get coreNetworkIdInput() {
        return this._coreNetworkId;
    }
    // edge_location - computed: true, optional: false, required: false
    get edgeLocation() {
        return this.getStringAttribute('edge_location');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_account_id - computed: true, optional: false, required: false
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    // resource_arn - computed: true, optional: false, required: false
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    // segment_name - computed: true, optional: false, required: false
    get segmentName() {
        return this.getStringAttribute('segment_name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get subnetArns() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_arns'));
    }
    set subnetArns(value) {
        this._subnetArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetArnsInput() {
        return this._subnetArns;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get vpcArn() {
        return this.getStringAttribute('vpc_arn');
    }
    set vpcArn(value) {
        this._vpcArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcArnInput() {
        return this._vpcArn;
    }
    get options() {
        return this._options;
    }
    putOptions(value) {
        this._options.internalValue = value;
    }
    resetOptions() {
        this._options.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            core_network_id: cdktf.stringToTerraform(this._coreNetworkId),
            id: cdktf.stringToTerraform(this._id),
            subnet_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetArns),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vpc_arn: cdktf.stringToTerraform(this._vpcArn),
            options: networkmanagerVpcAttachmentOptionsToTerraform(this._options.internalValue),
            timeouts: networkmanagerVpcAttachmentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            core_network_id: {
                value: cdktf.stringToHclTerraform(this._coreNetworkId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_arns: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._subnetArns),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            vpc_arn: {
                value: cdktf.stringToHclTerraform(this._vpcArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            options: {
                value: networkmanagerVpcAttachmentOptionsToHclTerraform(this._options.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkmanagerVpcAttachmentOptionsList",
            },
            timeouts: {
                value: networkmanagerVpcAttachmentTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkmanagerVpcAttachmentTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkmanagerVpcAttachment = NetworkmanagerVpcAttachment;
_c = JSII_RTTI_SYMBOL_1;
NetworkmanagerVpcAttachment[_c] = { fqn: "@cdktf/provider-aws.networkmanagerVpcAttachment.NetworkmanagerVpcAttachment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
NetworkmanagerVpcAttachment.tfResourceType = "aws_networkmanager_vpc_attachment";
