// @cdktf/provider-aws/lib/route53domains-delegation-signer-record/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53DomainsDelegationSignerRecord = exports.Route53DomainsDelegationSignerRecordTimeoutsOutputReference = exports.route53DomainsDelegationSignerRecordTimeoutsToHclTerraform = exports.route53DomainsDelegationSignerRecordTimeoutsToTerraform = exports.Route53DomainsDelegationSignerRecordSigningAttributesList = exports.Route53DomainsDelegationSignerRecordSigningAttributesOutputReference = exports.route53DomainsDelegationSignerRecordSigningAttributesToHclTerraform = exports.route53DomainsDelegationSignerRecordSigningAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function route53DomainsDelegationSignerRecordSigningAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        algorithm: cdktf.numberToTerraform(struct.algorithm),
        flags: cdktf.numberToTerraform(struct.flags),
        public_key: cdktf.stringToTerraform(struct.publicKey),
    };
}
exports.route53DomainsDelegationSignerRecordSigningAttributesToTerraform = route53DomainsDelegationSignerRecordSigningAttributesToTerraform;
function route53DomainsDelegationSignerRecordSigningAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        algorithm: {
            value: cdktf.numberToHclTerraform(struct.algorithm),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        flags: {
            value: cdktf.numberToHclTerraform(struct.flags),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        public_key: {
            value: cdktf.stringToHclTerraform(struct.publicKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.route53DomainsDelegationSignerRecordSigningAttributesToHclTerraform = route53DomainsDelegationSignerRecordSigningAttributesToHclTerraform;
class Route53DomainsDelegationSignerRecordSigningAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._algorithm !== undefined) {
            hasAnyValues = true;
            internalValueResult.algorithm = this._algorithm;
        }
        if (this._flags !== undefined) {
            hasAnyValues = true;
            internalValueResult.flags = this._flags;
        }
        if (this._publicKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.publicKey = this._publicKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._algorithm = undefined;
            this._flags = undefined;
            this._publicKey = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._algorithm = value.algorithm;
            this._flags = value.flags;
            this._publicKey = value.publicKey;
        }
    }
    get algorithm() {
        return this.getNumberAttribute('algorithm');
    }
    set algorithm(value) {
        this._algorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    get algorithmInput() {
        return this._algorithm;
    }
    get flags() {
        return this.getNumberAttribute('flags');
    }
    set flags(value) {
        this._flags = value;
    }
    // Temporarily expose input value. Use with caution.
    get flagsInput() {
        return this._flags;
    }
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    set publicKey(value) {
        this._publicKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get publicKeyInput() {
        return this._publicKey;
    }
}
exports.Route53DomainsDelegationSignerRecordSigningAttributesOutputReference = Route53DomainsDelegationSignerRecordSigningAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Route53DomainsDelegationSignerRecordSigningAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.route53DomainsDelegationSignerRecord.Route53DomainsDelegationSignerRecordSigningAttributesOutputReference", version: "19.15.0" };
class Route53DomainsDelegationSignerRecordSigningAttributesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Route53DomainsDelegationSignerRecordSigningAttributesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Route53DomainsDelegationSignerRecordSigningAttributesList = Route53DomainsDelegationSignerRecordSigningAttributesList;
_b = JSII_RTTI_SYMBOL_1;
Route53DomainsDelegationSignerRecordSigningAttributesList[_b] = { fqn: "@cdktf/provider-aws.route53DomainsDelegationSignerRecord.Route53DomainsDelegationSignerRecordSigningAttributesList", version: "19.15.0" };
function route53DomainsDelegationSignerRecordTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.route53DomainsDelegationSignerRecordTimeoutsToTerraform = route53DomainsDelegationSignerRecordTimeoutsToTerraform;
function route53DomainsDelegationSignerRecordTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.route53DomainsDelegationSignerRecordTimeoutsToHclTerraform = route53DomainsDelegationSignerRecordTimeoutsToHclTerraform;
class Route53DomainsDelegationSignerRecordTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.Route53DomainsDelegationSignerRecordTimeoutsOutputReference = Route53DomainsDelegationSignerRecordTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Route53DomainsDelegationSignerRecordTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.route53DomainsDelegationSignerRecord.Route53DomainsDelegationSignerRecordTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53domains_delegation_signer_record aws_route53domains_delegation_signer_record}
*/
class Route53DomainsDelegationSignerRecord extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Route53DomainsDelegationSignerRecord resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Route53DomainsDelegationSignerRecord to import
    * @param importFromId The id of the existing Route53DomainsDelegationSignerRecord that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53domains_delegation_signer_record#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Route53DomainsDelegationSignerRecord to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_route53domains_delegation_signer_record", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/route53domains_delegation_signer_record aws_route53domains_delegation_signer_record} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53DomainsDelegationSignerRecordConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53domains_delegation_signer_record',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // signing_attributes - computed: false, optional: true, required: false
        this._signingAttributes = new Route53DomainsDelegationSignerRecordSigningAttributesList(this, "signing_attributes", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Route53DomainsDelegationSignerRecordTimeoutsOutputReference(this, "timeouts");
        this._domainName = config.domainName;
        this._signingAttributes.internalValue = config.signingAttributes;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // dnssec_key_id - computed: true, optional: false, required: false
    get dnssecKeyId() {
        return this.getStringAttribute('dnssec_key_id');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get signingAttributes() {
        return this._signingAttributes;
    }
    putSigningAttributes(value) {
        this._signingAttributes.internalValue = value;
    }
    resetSigningAttributes() {
        this._signingAttributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get signingAttributesInput() {
        return this._signingAttributes.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            signing_attributes: cdktf.listMapper(route53DomainsDelegationSignerRecordSigningAttributesToTerraform, true)(this._signingAttributes.internalValue),
            timeouts: route53DomainsDelegationSignerRecordTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            domain_name: {
                value: cdktf.stringToHclTerraform(this._domainName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            signing_attributes: {
                value: cdktf.listMapperHcl(route53DomainsDelegationSignerRecordSigningAttributesToHclTerraform, true)(this._signingAttributes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Route53DomainsDelegationSignerRecordSigningAttributesList",
            },
            timeouts: {
                value: route53DomainsDelegationSignerRecordTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "Route53DomainsDelegationSignerRecordTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Route53DomainsDelegationSignerRecord = Route53DomainsDelegationSignerRecord;
_d = JSII_RTTI_SYMBOL_1;
Route53DomainsDelegationSignerRecord[_d] = { fqn: "@cdktf/provider-aws.route53DomainsDelegationSignerRecord.Route53DomainsDelegationSignerRecord", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Route53DomainsDelegationSignerRecord.tfResourceType = "aws_route53domains_delegation_signer_record";
