// @winglang/sdk/lib/expect/expect.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Util = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const strict_1 = __importStar(require("node:assert/strict"));
const core_1 = require("../core");
const std_1 = require("../std");
/**
 * The Util class provides methods for making assertions in tests,
 * such as comparing two strings for equality with rich error messages.
 * @inflight
 */
class Util {
    /**
     * @internal
     */
    static _toInflightType() {
        return core_1.InflightClient.forType(__filename, this.name);
    }
    /**
     * Compares two values for equality.
     * @param expected The expected value.
     * @param actual The value to test.
     * @throws Will throw an error if the actual value does not match the expected value.
     * @returns void
     */
    static equal(actual, expected) {
        strict_1.default.deepStrictEqual(actual, expected);
    }
    /**
     * Compares two values for inequality.
     * @param expected The expected value.
     * @param actual The value to test.
     * @throws Will throw an error if the actual value matches the expected value.
     * @returns void
     */
    static notEqual(actual, expected) {
        strict_1.default.notDeepStrictEqual(actual, expected);
    }
    /**
     * Checks if a value is nil.
     * @param actual The value to test.
     * @throws Will throw an error if the actual value is not nil.
     * @returns void
     */
    static nil(actual) {
        strict_1.default.ok(actual === null || actual === undefined, `Expected "${actual}" to be nil`);
    }
    /**
     * Checks if a value is not nil.
     * @param actual The value to test.
     * @throws Will throw an error if the actual value is nil.
     * @returns void
     */
    static notNil(actual) {
        strict_1.default.ok(actual !== null && actual !== undefined, `Expected "${actual}" to be not nil`);
    }
    /**
     * Checks if a string matches a regular expression pattern.
     * @param actual The string to test.
     * @param expected The regular expression pattern to match against.
     * @throws Will throw an error if the actual value does not match the expected regular expression pattern.
     * @returns void
     */
    static match(actual, expected) {
        const regex = std_1.Regex.compile(expected);
        const matches = regex.test(actual);
        if (!matches) {
            throw new strict_1.AssertionError({
                message: `The input did not match the regular expression ${expected}`,
            });
        }
    }
    /**
     * Checks if a string does not match a regular expression pattern.
     * @param actual The string to test.
     * @param expected The regular expression pattern to check against.
     * @throws Will throw an error if the actual value matches the expected regular expression pattern.
     * @returns void
     */
    static doesNotMatch(actual, expected) {
        const regex = std_1.Regex.compile(expected);
        const matches = regex.test(actual);
        if (matches) {
            throw new strict_1.AssertionError({
                message: `The input should not match the regular expression ${expected}`,
            });
        }
    }
    /**
     * Marks a test as failed.
     * @param message An optional message to include with the failure.
     * @throws Always throws an error with the provided message.
     * @returns void
     */
    static fail(message) {
        strict_1.default.fail(message);
    }
    /**
     * Asserts that a condition is truthy.
     * @param condition The condition to test.
     * @param message An optional message to include if the condition is falsy.
     * @throws Will throw an error if the condition is falsy.
     * @returns void
     */
    static ok(condition, message) {
        strict_1.default.ok(condition, message);
    }
    constructor() { }
}
exports.Util = Util;
_a = JSII_RTTI_SYMBOL_1;
Util[_a] = { fqn: "@winglang/sdk.expect.Util", version: "0.0.0" };
