// @cdktf/provider-aws/lib/appconfig-environment/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppconfigEnvironment = exports.AppconfigEnvironmentMonitorList = exports.AppconfigEnvironmentMonitorOutputReference = exports.appconfigEnvironmentMonitorToHclTerraform = exports.appconfigEnvironmentMonitorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appconfigEnvironmentMonitorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        alarm_arn: cdktf.stringToTerraform(struct.alarmArn),
        alarm_role_arn: cdktf.stringToTerraform(struct.alarmRoleArn),
    };
}
exports.appconfigEnvironmentMonitorToTerraform = appconfigEnvironmentMonitorToTerraform;
function appconfigEnvironmentMonitorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        alarm_arn: {
            value: cdktf.stringToHclTerraform(struct.alarmArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        alarm_role_arn: {
            value: cdktf.stringToHclTerraform(struct.alarmRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.appconfigEnvironmentMonitorToHclTerraform = appconfigEnvironmentMonitorToHclTerraform;
class AppconfigEnvironmentMonitorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._alarmArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.alarmArn = this._alarmArn;
        }
        if (this._alarmRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.alarmRoleArn = this._alarmRoleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._alarmArn = undefined;
            this._alarmRoleArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._alarmArn = value.alarmArn;
            this._alarmRoleArn = value.alarmRoleArn;
        }
    }
    get alarmArn() {
        return this.getStringAttribute('alarm_arn');
    }
    set alarmArn(value) {
        this._alarmArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get alarmArnInput() {
        return this._alarmArn;
    }
    get alarmRoleArn() {
        return this.getStringAttribute('alarm_role_arn');
    }
    set alarmRoleArn(value) {
        this._alarmRoleArn = value;
    }
    resetAlarmRoleArn() {
        this._alarmRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get alarmRoleArnInput() {
        return this._alarmRoleArn;
    }
}
exports.AppconfigEnvironmentMonitorOutputReference = AppconfigEnvironmentMonitorOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppconfigEnvironmentMonitorOutputReference[_a] = { fqn: "@cdktf/provider-aws.appconfigEnvironment.AppconfigEnvironmentMonitorOutputReference", version: "19.15.0" };
class AppconfigEnvironmentMonitorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppconfigEnvironmentMonitorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppconfigEnvironmentMonitorList = AppconfigEnvironmentMonitorList;
_b = JSII_RTTI_SYMBOL_1;
AppconfigEnvironmentMonitorList[_b] = { fqn: "@cdktf/provider-aws.appconfigEnvironment.AppconfigEnvironmentMonitorList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_environment aws_appconfig_environment}
*/
class AppconfigEnvironment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AppconfigEnvironment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AppconfigEnvironment to import
    * @param importFromId The id of the existing AppconfigEnvironment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_environment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AppconfigEnvironment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_appconfig_environment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/appconfig_environment aws_appconfig_environment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppconfigEnvironmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appconfig_environment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tags_all - computed: true, optional: false, required: false
        this._tagsAll = new cdktf.StringMap(this, "tags_all");
        // monitor - computed: false, optional: true, required: false
        this._monitor = new AppconfigEnvironmentMonitorList(this, "monitor", true);
        this._applicationId = config.applicationId;
        this._description = config.description;
        this._name = config.name;
        this._tags = config.tags;
        this._monitor.internalValue = config.monitor;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // environment_id - computed: true, optional: false, required: false
    get environmentId() {
        return this.getStringAttribute('environment_id');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this._tagsAll;
    }
    get monitor() {
        return this._monitor;
    }
    putMonitor(value) {
        this._monitor.internalValue = value;
    }
    resetMonitor() {
        this._monitor.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monitorInput() {
        return this._monitor.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            monitor: cdktf.listMapper(appconfigEnvironmentMonitorToTerraform, true)(this._monitor.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_id: {
                value: cdktf.stringToHclTerraform(this._applicationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            monitor: {
                value: cdktf.listMapperHcl(appconfigEnvironmentMonitorToHclTerraform, true)(this._monitor.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "AppconfigEnvironmentMonitorList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AppconfigEnvironment = AppconfigEnvironment;
_c = JSII_RTTI_SYMBOL_1;
AppconfigEnvironment[_c] = { fqn: "@cdktf/provider-aws.appconfigEnvironment.AppconfigEnvironment", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AppconfigEnvironment.tfResourceType = "aws_appconfig_environment";
