// @cdktf/provider-aws/lib/account-primary-contact/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountPrimaryContact = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/account_primary_contact aws_account_primary_contact}
*/
class AccountPrimaryContact extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AccountPrimaryContact resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AccountPrimaryContact to import
    * @param importFromId The id of the existing AccountPrimaryContact that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/account_primary_contact#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AccountPrimaryContact to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_account_primary_contact", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/account_primary_contact aws_account_primary_contact} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AccountPrimaryContactConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_account_primary_contact',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accountId = config.accountId;
        this._addressLine1 = config.addressLine1;
        this._addressLine2 = config.addressLine2;
        this._addressLine3 = config.addressLine3;
        this._city = config.city;
        this._companyName = config.companyName;
        this._countryCode = config.countryCode;
        this._districtOrCounty = config.districtOrCounty;
        this._fullName = config.fullName;
        this._id = config.id;
        this._phoneNumber = config.phoneNumber;
        this._postalCode = config.postalCode;
        this._stateOrRegion = config.stateOrRegion;
        this._websiteUrl = config.websiteUrl;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get addressLine1() {
        return this.getStringAttribute('address_line_1');
    }
    set addressLine1(value) {
        this._addressLine1 = value;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine1Input() {
        return this._addressLine1;
    }
    get addressLine2() {
        return this.getStringAttribute('address_line_2');
    }
    set addressLine2(value) {
        this._addressLine2 = value;
    }
    resetAddressLine2() {
        this._addressLine2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine2Input() {
        return this._addressLine2;
    }
    get addressLine3() {
        return this.getStringAttribute('address_line_3');
    }
    set addressLine3(value) {
        this._addressLine3 = value;
    }
    resetAddressLine3() {
        this._addressLine3 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine3Input() {
        return this._addressLine3;
    }
    get city() {
        return this.getStringAttribute('city');
    }
    set city(value) {
        this._city = value;
    }
    // Temporarily expose input value. Use with caution.
    get cityInput() {
        return this._city;
    }
    get companyName() {
        return this.getStringAttribute('company_name');
    }
    set companyName(value) {
        this._companyName = value;
    }
    resetCompanyName() {
        this._companyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get companyNameInput() {
        return this._companyName;
    }
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    set countryCode(value) {
        this._countryCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get countryCodeInput() {
        return this._countryCode;
    }
    get districtOrCounty() {
        return this.getStringAttribute('district_or_county');
    }
    set districtOrCounty(value) {
        this._districtOrCounty = value;
    }
    resetDistrictOrCounty() {
        this._districtOrCounty = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get districtOrCountyInput() {
        return this._districtOrCounty;
    }
    get fullName() {
        return this.getStringAttribute('full_name');
    }
    set fullName(value) {
        this._fullName = value;
    }
    // Temporarily expose input value. Use with caution.
    get fullNameInput() {
        return this._fullName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
    get postalCode() {
        return this.getStringAttribute('postal_code');
    }
    set postalCode(value) {
        this._postalCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get postalCodeInput() {
        return this._postalCode;
    }
    get stateOrRegion() {
        return this.getStringAttribute('state_or_region');
    }
    set stateOrRegion(value) {
        this._stateOrRegion = value;
    }
    resetStateOrRegion() {
        this._stateOrRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateOrRegionInput() {
        return this._stateOrRegion;
    }
    get websiteUrl() {
        return this.getStringAttribute('website_url');
    }
    set websiteUrl(value) {
        this._websiteUrl = value;
    }
    resetWebsiteUrl() {
        this._websiteUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get websiteUrlInput() {
        return this._websiteUrl;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            address_line_1: cdktf.stringToTerraform(this._addressLine1),
            address_line_2: cdktf.stringToTerraform(this._addressLine2),
            address_line_3: cdktf.stringToTerraform(this._addressLine3),
            city: cdktf.stringToTerraform(this._city),
            company_name: cdktf.stringToTerraform(this._companyName),
            country_code: cdktf.stringToTerraform(this._countryCode),
            district_or_county: cdktf.stringToTerraform(this._districtOrCounty),
            full_name: cdktf.stringToTerraform(this._fullName),
            id: cdktf.stringToTerraform(this._id),
            phone_number: cdktf.stringToTerraform(this._phoneNumber),
            postal_code: cdktf.stringToTerraform(this._postalCode),
            state_or_region: cdktf.stringToTerraform(this._stateOrRegion),
            website_url: cdktf.stringToTerraform(this._websiteUrl),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            account_id: {
                value: cdktf.stringToHclTerraform(this._accountId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            address_line_1: {
                value: cdktf.stringToHclTerraform(this._addressLine1),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            address_line_2: {
                value: cdktf.stringToHclTerraform(this._addressLine2),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            address_line_3: {
                value: cdktf.stringToHclTerraform(this._addressLine3),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            city: {
                value: cdktf.stringToHclTerraform(this._city),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            company_name: {
                value: cdktf.stringToHclTerraform(this._companyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            country_code: {
                value: cdktf.stringToHclTerraform(this._countryCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            district_or_county: {
                value: cdktf.stringToHclTerraform(this._districtOrCounty),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            full_name: {
                value: cdktf.stringToHclTerraform(this._fullName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            phone_number: {
                value: cdktf.stringToHclTerraform(this._phoneNumber),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            postal_code: {
                value: cdktf.stringToHclTerraform(this._postalCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            state_or_region: {
                value: cdktf.stringToHclTerraform(this._stateOrRegion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            website_url: {
                value: cdktf.stringToHclTerraform(this._websiteUrl),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AccountPrimaryContact = AccountPrimaryContact;
_a = JSII_RTTI_SYMBOL_1;
AccountPrimaryContact[_a] = { fqn: "@cdktf/provider-aws.accountPrimaryContact.AccountPrimaryContact", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AccountPrimaryContact.tfResourceType = "aws_account_primary_contact";
