// @cdktf/provider-aws/lib/lexv2models-bot-locale/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lexv2ModelsBotLocale = exports.Lexv2ModelsBotLocaleVoiceSettingsList = exports.Lexv2ModelsBotLocaleVoiceSettingsOutputReference = exports.lexv2ModelsBotLocaleVoiceSettingsToHclTerraform = exports.lexv2ModelsBotLocaleVoiceSettingsToTerraform = exports.Lexv2ModelsBotLocaleTimeoutsOutputReference = exports.lexv2ModelsBotLocaleTimeoutsToHclTerraform = exports.lexv2ModelsBotLocaleTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lexv2ModelsBotLocaleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lexv2ModelsBotLocaleTimeoutsToTerraform = lexv2ModelsBotLocaleTimeoutsToTerraform;
function lexv2ModelsBotLocaleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsBotLocaleTimeoutsToHclTerraform = lexv2ModelsBotLocaleTimeoutsToHclTerraform;
class Lexv2ModelsBotLocaleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Lexv2ModelsBotLocaleTimeoutsOutputReference = Lexv2ModelsBotLocaleTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotLocaleTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotLocale.Lexv2ModelsBotLocaleTimeoutsOutputReference", version: "19.15.0" };
function lexv2ModelsBotLocaleVoiceSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        engine: cdktf.stringToTerraform(struct.engine),
        voice_id: cdktf.stringToTerraform(struct.voiceId),
    };
}
exports.lexv2ModelsBotLocaleVoiceSettingsToTerraform = lexv2ModelsBotLocaleVoiceSettingsToTerraform;
function lexv2ModelsBotLocaleVoiceSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        engine: {
            value: cdktf.stringToHclTerraform(struct.engine),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        voice_id: {
            value: cdktf.stringToHclTerraform(struct.voiceId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.lexv2ModelsBotLocaleVoiceSettingsToHclTerraform = lexv2ModelsBotLocaleVoiceSettingsToHclTerraform;
class Lexv2ModelsBotLocaleVoiceSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._engine !== undefined) {
            hasAnyValues = true;
            internalValueResult.engine = this._engine;
        }
        if (this._voiceId !== undefined) {
            hasAnyValues = true;
            internalValueResult.voiceId = this._voiceId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._engine = undefined;
            this._voiceId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._engine = value.engine;
            this._voiceId = value.voiceId;
        }
    }
    get engine() {
        return this.getStringAttribute('engine');
    }
    set engine(value) {
        this._engine = value;
    }
    resetEngine() {
        this._engine = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get engineInput() {
        return this._engine;
    }
    get voiceId() {
        return this.getStringAttribute('voice_id');
    }
    set voiceId(value) {
        this._voiceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get voiceIdInput() {
        return this._voiceId;
    }
}
exports.Lexv2ModelsBotLocaleVoiceSettingsOutputReference = Lexv2ModelsBotLocaleVoiceSettingsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotLocaleVoiceSettingsOutputReference[_b] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotLocale.Lexv2ModelsBotLocaleVoiceSettingsOutputReference", version: "19.15.0" };
class Lexv2ModelsBotLocaleVoiceSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Lexv2ModelsBotLocaleVoiceSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Lexv2ModelsBotLocaleVoiceSettingsList = Lexv2ModelsBotLocaleVoiceSettingsList;
_c = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotLocaleVoiceSettingsList[_c] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotLocale.Lexv2ModelsBotLocaleVoiceSettingsList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_locale aws_lexv2models_bot_locale}
*/
class Lexv2ModelsBotLocale extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Lexv2ModelsBotLocale resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Lexv2ModelsBotLocale to import
    * @param importFromId The id of the existing Lexv2ModelsBotLocale that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_locale#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Lexv2ModelsBotLocale to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_lexv2models_bot_locale", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/lexv2models_bot_locale aws_lexv2models_bot_locale} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Lexv2ModelsBotLocaleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lexv2models_bot_locale',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Lexv2ModelsBotLocaleTimeoutsOutputReference(this, "timeouts");
        // voice_settings - computed: false, optional: true, required: false
        this._voiceSettings = new Lexv2ModelsBotLocaleVoiceSettingsList(this, "voice_settings", false);
        this._botId = config.botId;
        this._botVersion = config.botVersion;
        this._description = config.description;
        this._localeId = config.localeId;
        this._nLuIntentConfidenceThreshold = config.nLuIntentConfidenceThreshold;
        this._name = config.name;
        this._timeouts.internalValue = config.timeouts;
        this._voiceSettings.internalValue = config.voiceSettings;
    }
    get botId() {
        return this.getStringAttribute('bot_id');
    }
    set botId(value) {
        this._botId = value;
    }
    // Temporarily expose input value. Use with caution.
    get botIdInput() {
        return this._botId;
    }
    get botVersion() {
        return this.getStringAttribute('bot_version');
    }
    set botVersion(value) {
        this._botVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get botVersionInput() {
        return this._botVersion;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get localeId() {
        return this.getStringAttribute('locale_id');
    }
    set localeId(value) {
        this._localeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get localeIdInput() {
        return this._localeId;
    }
    get nLuIntentConfidenceThreshold() {
        return this.getNumberAttribute('n_lu_intent_confidence_threshold');
    }
    set nLuIntentConfidenceThreshold(value) {
        this._nLuIntentConfidenceThreshold = value;
    }
    // Temporarily expose input value. Use with caution.
    get nLuIntentConfidenceThresholdInput() {
        return this._nLuIntentConfidenceThreshold;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get voiceSettings() {
        return this._voiceSettings;
    }
    putVoiceSettings(value) {
        this._voiceSettings.internalValue = value;
    }
    resetVoiceSettings() {
        this._voiceSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get voiceSettingsInput() {
        return this._voiceSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bot_id: cdktf.stringToTerraform(this._botId),
            bot_version: cdktf.stringToTerraform(this._botVersion),
            description: cdktf.stringToTerraform(this._description),
            locale_id: cdktf.stringToTerraform(this._localeId),
            n_lu_intent_confidence_threshold: cdktf.numberToTerraform(this._nLuIntentConfidenceThreshold),
            name: cdktf.stringToTerraform(this._name),
            timeouts: lexv2ModelsBotLocaleTimeoutsToTerraform(this._timeouts.internalValue),
            voice_settings: cdktf.listMapper(lexv2ModelsBotLocaleVoiceSettingsToTerraform, true)(this._voiceSettings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            bot_id: {
                value: cdktf.stringToHclTerraform(this._botId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            bot_version: {
                value: cdktf.stringToHclTerraform(this._botVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            locale_id: {
                value: cdktf.stringToHclTerraform(this._localeId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            n_lu_intent_confidence_threshold: {
                value: cdktf.numberToHclTerraform(this._nLuIntentConfidenceThreshold),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: lexv2ModelsBotLocaleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "Lexv2ModelsBotLocaleTimeouts",
            },
            voice_settings: {
                value: cdktf.listMapperHcl(lexv2ModelsBotLocaleVoiceSettingsToHclTerraform, true)(this._voiceSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Lexv2ModelsBotLocaleVoiceSettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Lexv2ModelsBotLocale = Lexv2ModelsBotLocale;
_d = JSII_RTTI_SYMBOL_1;
Lexv2ModelsBotLocale[_d] = { fqn: "@cdktf/provider-aws.lexv2ModelsBotLocale.Lexv2ModelsBotLocale", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Lexv2ModelsBotLocale.tfResourceType = "aws_lexv2models_bot_locale";
