// @cdktf/provider-aws/lib/acmpca-certificate/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcmpcaCertificate = exports.AcmpcaCertificateValidityOutputReference = exports.acmpcaCertificateValidityToHclTerraform = exports.acmpcaCertificateValidityToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function acmpcaCertificateValidityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.acmpcaCertificateValidityToTerraform = acmpcaCertificateValidityToTerraform;
function acmpcaCertificateValidityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.acmpcaCertificateValidityToHclTerraform = acmpcaCertificateValidityToHclTerraform;
class AcmpcaCertificateValidityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AcmpcaCertificateValidityOutputReference = AcmpcaCertificateValidityOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateValidityOutputReference[_a] = { fqn: "@cdktf/provider-aws.acmpcaCertificate.AcmpcaCertificateValidityOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate aws_acmpca_certificate}
*/
class AcmpcaCertificate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AcmpcaCertificate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AcmpcaCertificate to import
    * @param importFromId The id of the existing AcmpcaCertificate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AcmpcaCertificate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_acmpca_certificate", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/acmpca_certificate aws_acmpca_certificate} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AcmpcaCertificateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_acmpca_certificate',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // validity - computed: false, optional: false, required: true
        this._validity = new AcmpcaCertificateValidityOutputReference(this, "validity");
        this._apiPassthrough = config.apiPassthrough;
        this._certificateAuthorityArn = config.certificateAuthorityArn;
        this._certificateSigningRequest = config.certificateSigningRequest;
        this._id = config.id;
        this._signingAlgorithm = config.signingAlgorithm;
        this._templateArn = config.templateArn;
        this._validity.internalValue = config.validity;
    }
    get apiPassthrough() {
        return this.getStringAttribute('api_passthrough');
    }
    set apiPassthrough(value) {
        this._apiPassthrough = value;
    }
    resetApiPassthrough() {
        this._apiPassthrough = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiPassthroughInput() {
        return this._apiPassthrough;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate - computed: true, optional: false, required: false
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    get certificateAuthorityArn() {
        return this.getStringAttribute('certificate_authority_arn');
    }
    set certificateAuthorityArn(value) {
        this._certificateAuthorityArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get certificateAuthorityArnInput() {
        return this._certificateAuthorityArn;
    }
    // certificate_chain - computed: true, optional: false, required: false
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    get certificateSigningRequest() {
        return this.getStringAttribute('certificate_signing_request');
    }
    set certificateSigningRequest(value) {
        this._certificateSigningRequest = value;
    }
    // Temporarily expose input value. Use with caution.
    get certificateSigningRequestInput() {
        return this._certificateSigningRequest;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get signingAlgorithm() {
        return this.getStringAttribute('signing_algorithm');
    }
    set signingAlgorithm(value) {
        this._signingAlgorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    get signingAlgorithmInput() {
        return this._signingAlgorithm;
    }
    get templateArn() {
        return this.getStringAttribute('template_arn');
    }
    set templateArn(value) {
        this._templateArn = value;
    }
    resetTemplateArn() {
        this._templateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get templateArnInput() {
        return this._templateArn;
    }
    get validity() {
        return this._validity;
    }
    putValidity(value) {
        this._validity.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get validityInput() {
        return this._validity.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_passthrough: cdktf.stringToTerraform(this._apiPassthrough),
            certificate_authority_arn: cdktf.stringToTerraform(this._certificateAuthorityArn),
            certificate_signing_request: cdktf.stringToTerraform(this._certificateSigningRequest),
            id: cdktf.stringToTerraform(this._id),
            signing_algorithm: cdktf.stringToTerraform(this._signingAlgorithm),
            template_arn: cdktf.stringToTerraform(this._templateArn),
            validity: acmpcaCertificateValidityToTerraform(this._validity.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            api_passthrough: {
                value: cdktf.stringToHclTerraform(this._apiPassthrough),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            certificate_authority_arn: {
                value: cdktf.stringToHclTerraform(this._certificateAuthorityArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            certificate_signing_request: {
                value: cdktf.stringToHclTerraform(this._certificateSigningRequest),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            signing_algorithm: {
                value: cdktf.stringToHclTerraform(this._signingAlgorithm),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            template_arn: {
                value: cdktf.stringToHclTerraform(this._templateArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            validity: {
                value: acmpcaCertificateValidityToHclTerraform(this._validity.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AcmpcaCertificateValidityList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AcmpcaCertificate = AcmpcaCertificate;
_b = JSII_RTTI_SYMBOL_1;
AcmpcaCertificate[_b] = { fqn: "@cdktf/provider-aws.acmpcaCertificate.AcmpcaCertificate", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
AcmpcaCertificate.tfResourceType = "aws_acmpca_certificate";
