// @cdktf/provider-aws/lib/datasync-task/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncTask = exports.DatasyncTaskTimeoutsOutputReference = exports.datasyncTaskTimeoutsToHclTerraform = exports.datasyncTaskTimeoutsToTerraform = exports.DatasyncTaskTaskReportConfigOutputReference = exports.datasyncTaskTaskReportConfigToHclTerraform = exports.datasyncTaskTaskReportConfigToTerraform = exports.DatasyncTaskTaskReportConfigS3DestinationOutputReference = exports.datasyncTaskTaskReportConfigS3DestinationToHclTerraform = exports.datasyncTaskTaskReportConfigS3DestinationToTerraform = exports.DatasyncTaskTaskReportConfigReportOverridesOutputReference = exports.datasyncTaskTaskReportConfigReportOverridesToHclTerraform = exports.datasyncTaskTaskReportConfigReportOverridesToTerraform = exports.DatasyncTaskScheduleOutputReference = exports.datasyncTaskScheduleToHclTerraform = exports.datasyncTaskScheduleToTerraform = exports.DatasyncTaskOptionsOutputReference = exports.datasyncTaskOptionsToHclTerraform = exports.datasyncTaskOptionsToTerraform = exports.DatasyncTaskIncludesOutputReference = exports.datasyncTaskIncludesToHclTerraform = exports.datasyncTaskIncludesToTerraform = exports.DatasyncTaskExcludesOutputReference = exports.datasyncTaskExcludesToHclTerraform = exports.datasyncTaskExcludesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncTaskExcludesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter_type: cdktf.stringToTerraform(struct.filterType),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.datasyncTaskExcludesToTerraform = datasyncTaskExcludesToTerraform;
function datasyncTaskExcludesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter_type: {
            value: cdktf.stringToHclTerraform(struct.filterType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskExcludesToHclTerraform = datasyncTaskExcludesToHclTerraform;
class DatasyncTaskExcludesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterType = this._filterType;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filterType = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filterType = value.filterType;
            this._value = value.value;
        }
    }
    get filterType() {
        return this.getStringAttribute('filter_type');
    }
    set filterType(value) {
        this._filterType = value;
    }
    resetFilterType() {
        this._filterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterTypeInput() {
        return this._filterType;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DatasyncTaskExcludesOutputReference = DatasyncTaskExcludesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncTaskExcludesOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskExcludesOutputReference", version: "19.15.0" };
function datasyncTaskIncludesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter_type: cdktf.stringToTerraform(struct.filterType),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.datasyncTaskIncludesToTerraform = datasyncTaskIncludesToTerraform;
function datasyncTaskIncludesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        filter_type: {
            value: cdktf.stringToHclTerraform(struct.filterType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskIncludesToHclTerraform = datasyncTaskIncludesToHclTerraform;
class DatasyncTaskIncludesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterType = this._filterType;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filterType = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filterType = value.filterType;
            this._value = value.value;
        }
    }
    get filterType() {
        return this.getStringAttribute('filter_type');
    }
    set filterType(value) {
        this._filterType = value;
    }
    resetFilterType() {
        this._filterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterTypeInput() {
        return this._filterType;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DatasyncTaskIncludesOutputReference = DatasyncTaskIncludesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DatasyncTaskIncludesOutputReference[_b] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskIncludesOutputReference", version: "19.15.0" };
function datasyncTaskOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        atime: cdktf.stringToTerraform(struct.atime),
        bytes_per_second: cdktf.numberToTerraform(struct.bytesPerSecond),
        gid: cdktf.stringToTerraform(struct.gid),
        log_level: cdktf.stringToTerraform(struct.logLevel),
        mtime: cdktf.stringToTerraform(struct.mtime),
        object_tags: cdktf.stringToTerraform(struct.objectTags),
        overwrite_mode: cdktf.stringToTerraform(struct.overwriteMode),
        posix_permissions: cdktf.stringToTerraform(struct.posixPermissions),
        preserve_deleted_files: cdktf.stringToTerraform(struct.preserveDeletedFiles),
        preserve_devices: cdktf.stringToTerraform(struct.preserveDevices),
        security_descriptor_copy_flags: cdktf.stringToTerraform(struct.securityDescriptorCopyFlags),
        task_queueing: cdktf.stringToTerraform(struct.taskQueueing),
        transfer_mode: cdktf.stringToTerraform(struct.transferMode),
        uid: cdktf.stringToTerraform(struct.uid),
        verify_mode: cdktf.stringToTerraform(struct.verifyMode),
    };
}
exports.datasyncTaskOptionsToTerraform = datasyncTaskOptionsToTerraform;
function datasyncTaskOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        atime: {
            value: cdktf.stringToHclTerraform(struct.atime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        bytes_per_second: {
            value: cdktf.numberToHclTerraform(struct.bytesPerSecond),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        gid: {
            value: cdktf.stringToHclTerraform(struct.gid),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        log_level: {
            value: cdktf.stringToHclTerraform(struct.logLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mtime: {
            value: cdktf.stringToHclTerraform(struct.mtime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        object_tags: {
            value: cdktf.stringToHclTerraform(struct.objectTags),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        overwrite_mode: {
            value: cdktf.stringToHclTerraform(struct.overwriteMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        posix_permissions: {
            value: cdktf.stringToHclTerraform(struct.posixPermissions),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        preserve_deleted_files: {
            value: cdktf.stringToHclTerraform(struct.preserveDeletedFiles),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        preserve_devices: {
            value: cdktf.stringToHclTerraform(struct.preserveDevices),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        security_descriptor_copy_flags: {
            value: cdktf.stringToHclTerraform(struct.securityDescriptorCopyFlags),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        task_queueing: {
            value: cdktf.stringToHclTerraform(struct.taskQueueing),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        transfer_mode: {
            value: cdktf.stringToHclTerraform(struct.transferMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uid: {
            value: cdktf.stringToHclTerraform(struct.uid),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        verify_mode: {
            value: cdktf.stringToHclTerraform(struct.verifyMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskOptionsToHclTerraform = datasyncTaskOptionsToHclTerraform;
class DatasyncTaskOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._atime !== undefined) {
            hasAnyValues = true;
            internalValueResult.atime = this._atime;
        }
        if (this._bytesPerSecond !== undefined) {
            hasAnyValues = true;
            internalValueResult.bytesPerSecond = this._bytesPerSecond;
        }
        if (this._gid !== undefined) {
            hasAnyValues = true;
            internalValueResult.gid = this._gid;
        }
        if (this._logLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        if (this._mtime !== undefined) {
            hasAnyValues = true;
            internalValueResult.mtime = this._mtime;
        }
        if (this._objectTags !== undefined) {
            hasAnyValues = true;
            internalValueResult.objectTags = this._objectTags;
        }
        if (this._overwriteMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.overwriteMode = this._overwriteMode;
        }
        if (this._posixPermissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.posixPermissions = this._posixPermissions;
        }
        if (this._preserveDeletedFiles !== undefined) {
            hasAnyValues = true;
            internalValueResult.preserveDeletedFiles = this._preserveDeletedFiles;
        }
        if (this._preserveDevices !== undefined) {
            hasAnyValues = true;
            internalValueResult.preserveDevices = this._preserveDevices;
        }
        if (this._securityDescriptorCopyFlags !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityDescriptorCopyFlags = this._securityDescriptorCopyFlags;
        }
        if (this._taskQueueing !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskQueueing = this._taskQueueing;
        }
        if (this._transferMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.transferMode = this._transferMode;
        }
        if (this._uid !== undefined) {
            hasAnyValues = true;
            internalValueResult.uid = this._uid;
        }
        if (this._verifyMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.verifyMode = this._verifyMode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._atime = undefined;
            this._bytesPerSecond = undefined;
            this._gid = undefined;
            this._logLevel = undefined;
            this._mtime = undefined;
            this._objectTags = undefined;
            this._overwriteMode = undefined;
            this._posixPermissions = undefined;
            this._preserveDeletedFiles = undefined;
            this._preserveDevices = undefined;
            this._securityDescriptorCopyFlags = undefined;
            this._taskQueueing = undefined;
            this._transferMode = undefined;
            this._uid = undefined;
            this._verifyMode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._atime = value.atime;
            this._bytesPerSecond = value.bytesPerSecond;
            this._gid = value.gid;
            this._logLevel = value.logLevel;
            this._mtime = value.mtime;
            this._objectTags = value.objectTags;
            this._overwriteMode = value.overwriteMode;
            this._posixPermissions = value.posixPermissions;
            this._preserveDeletedFiles = value.preserveDeletedFiles;
            this._preserveDevices = value.preserveDevices;
            this._securityDescriptorCopyFlags = value.securityDescriptorCopyFlags;
            this._taskQueueing = value.taskQueueing;
            this._transferMode = value.transferMode;
            this._uid = value.uid;
            this._verifyMode = value.verifyMode;
        }
    }
    get atime() {
        return this.getStringAttribute('atime');
    }
    set atime(value) {
        this._atime = value;
    }
    resetAtime() {
        this._atime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get atimeInput() {
        return this._atime;
    }
    get bytesPerSecond() {
        return this.getNumberAttribute('bytes_per_second');
    }
    set bytesPerSecond(value) {
        this._bytesPerSecond = value;
    }
    resetBytesPerSecond() {
        this._bytesPerSecond = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bytesPerSecondInput() {
        return this._bytesPerSecond;
    }
    get gid() {
        return this.getStringAttribute('gid');
    }
    set gid(value) {
        this._gid = value;
    }
    resetGid() {
        this._gid = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gidInput() {
        return this._gid;
    }
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    set logLevel(value) {
        this._logLevel = value;
    }
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logLevelInput() {
        return this._logLevel;
    }
    get mtime() {
        return this.getStringAttribute('mtime');
    }
    set mtime(value) {
        this._mtime = value;
    }
    resetMtime() {
        this._mtime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mtimeInput() {
        return this._mtime;
    }
    get objectTags() {
        return this.getStringAttribute('object_tags');
    }
    set objectTags(value) {
        this._objectTags = value;
    }
    resetObjectTags() {
        this._objectTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get objectTagsInput() {
        return this._objectTags;
    }
    get overwriteMode() {
        return this.getStringAttribute('overwrite_mode');
    }
    set overwriteMode(value) {
        this._overwriteMode = value;
    }
    resetOverwriteMode() {
        this._overwriteMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteModeInput() {
        return this._overwriteMode;
    }
    get posixPermissions() {
        return this.getStringAttribute('posix_permissions');
    }
    set posixPermissions(value) {
        this._posixPermissions = value;
    }
    resetPosixPermissions() {
        this._posixPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get posixPermissionsInput() {
        return this._posixPermissions;
    }
    get preserveDeletedFiles() {
        return this.getStringAttribute('preserve_deleted_files');
    }
    set preserveDeletedFiles(value) {
        this._preserveDeletedFiles = value;
    }
    resetPreserveDeletedFiles() {
        this._preserveDeletedFiles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preserveDeletedFilesInput() {
        return this._preserveDeletedFiles;
    }
    get preserveDevices() {
        return this.getStringAttribute('preserve_devices');
    }
    set preserveDevices(value) {
        this._preserveDevices = value;
    }
    resetPreserveDevices() {
        this._preserveDevices = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preserveDevicesInput() {
        return this._preserveDevices;
    }
    get securityDescriptorCopyFlags() {
        return this.getStringAttribute('security_descriptor_copy_flags');
    }
    set securityDescriptorCopyFlags(value) {
        this._securityDescriptorCopyFlags = value;
    }
    resetSecurityDescriptorCopyFlags() {
        this._securityDescriptorCopyFlags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityDescriptorCopyFlagsInput() {
        return this._securityDescriptorCopyFlags;
    }
    get taskQueueing() {
        return this.getStringAttribute('task_queueing');
    }
    set taskQueueing(value) {
        this._taskQueueing = value;
    }
    resetTaskQueueing() {
        this._taskQueueing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskQueueingInput() {
        return this._taskQueueing;
    }
    get transferMode() {
        return this.getStringAttribute('transfer_mode');
    }
    set transferMode(value) {
        this._transferMode = value;
    }
    resetTransferMode() {
        this._transferMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transferModeInput() {
        return this._transferMode;
    }
    get uid() {
        return this.getStringAttribute('uid');
    }
    set uid(value) {
        this._uid = value;
    }
    resetUid() {
        this._uid = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uidInput() {
        return this._uid;
    }
    get verifyMode() {
        return this.getStringAttribute('verify_mode');
    }
    set verifyMode(value) {
        this._verifyMode = value;
    }
    resetVerifyMode() {
        this._verifyMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get verifyModeInput() {
        return this._verifyMode;
    }
}
exports.DatasyncTaskOptionsOutputReference = DatasyncTaskOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DatasyncTaskOptionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskOptionsOutputReference", version: "19.15.0" };
function datasyncTaskScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        schedule_expression: cdktf.stringToTerraform(struct.scheduleExpression),
    };
}
exports.datasyncTaskScheduleToTerraform = datasyncTaskScheduleToTerraform;
function datasyncTaskScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        schedule_expression: {
            value: cdktf.stringToHclTerraform(struct.scheduleExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskScheduleToHclTerraform = datasyncTaskScheduleToHclTerraform;
class DatasyncTaskScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._scheduleExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheduleExpression = this._scheduleExpression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._scheduleExpression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._scheduleExpression = value.scheduleExpression;
        }
    }
    get scheduleExpression() {
        return this.getStringAttribute('schedule_expression');
    }
    set scheduleExpression(value) {
        this._scheduleExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleExpressionInput() {
        return this._scheduleExpression;
    }
}
exports.DatasyncTaskScheduleOutputReference = DatasyncTaskScheduleOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DatasyncTaskScheduleOutputReference[_d] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskScheduleOutputReference", version: "19.15.0" };
function datasyncTaskTaskReportConfigReportOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        deleted_override: cdktf.stringToTerraform(struct.deletedOverride),
        skipped_override: cdktf.stringToTerraform(struct.skippedOverride),
        transferred_override: cdktf.stringToTerraform(struct.transferredOverride),
        verified_override: cdktf.stringToTerraform(struct.verifiedOverride),
    };
}
exports.datasyncTaskTaskReportConfigReportOverridesToTerraform = datasyncTaskTaskReportConfigReportOverridesToTerraform;
function datasyncTaskTaskReportConfigReportOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        deleted_override: {
            value: cdktf.stringToHclTerraform(struct.deletedOverride),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        skipped_override: {
            value: cdktf.stringToHclTerraform(struct.skippedOverride),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        transferred_override: {
            value: cdktf.stringToHclTerraform(struct.transferredOverride),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        verified_override: {
            value: cdktf.stringToHclTerraform(struct.verifiedOverride),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskTaskReportConfigReportOverridesToHclTerraform = datasyncTaskTaskReportConfigReportOverridesToHclTerraform;
class DatasyncTaskTaskReportConfigReportOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deletedOverride !== undefined) {
            hasAnyValues = true;
            internalValueResult.deletedOverride = this._deletedOverride;
        }
        if (this._skippedOverride !== undefined) {
            hasAnyValues = true;
            internalValueResult.skippedOverride = this._skippedOverride;
        }
        if (this._transferredOverride !== undefined) {
            hasAnyValues = true;
            internalValueResult.transferredOverride = this._transferredOverride;
        }
        if (this._verifiedOverride !== undefined) {
            hasAnyValues = true;
            internalValueResult.verifiedOverride = this._verifiedOverride;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deletedOverride = undefined;
            this._skippedOverride = undefined;
            this._transferredOverride = undefined;
            this._verifiedOverride = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deletedOverride = value.deletedOverride;
            this._skippedOverride = value.skippedOverride;
            this._transferredOverride = value.transferredOverride;
            this._verifiedOverride = value.verifiedOverride;
        }
    }
    get deletedOverride() {
        return this.getStringAttribute('deleted_override');
    }
    set deletedOverride(value) {
        this._deletedOverride = value;
    }
    resetDeletedOverride() {
        this._deletedOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletedOverrideInput() {
        return this._deletedOverride;
    }
    get skippedOverride() {
        return this.getStringAttribute('skipped_override');
    }
    set skippedOverride(value) {
        this._skippedOverride = value;
    }
    resetSkippedOverride() {
        this._skippedOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skippedOverrideInput() {
        return this._skippedOverride;
    }
    get transferredOverride() {
        return this.getStringAttribute('transferred_override');
    }
    set transferredOverride(value) {
        this._transferredOverride = value;
    }
    resetTransferredOverride() {
        this._transferredOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transferredOverrideInput() {
        return this._transferredOverride;
    }
    get verifiedOverride() {
        return this.getStringAttribute('verified_override');
    }
    set verifiedOverride(value) {
        this._verifiedOverride = value;
    }
    resetVerifiedOverride() {
        this._verifiedOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get verifiedOverrideInput() {
        return this._verifiedOverride;
    }
}
exports.DatasyncTaskTaskReportConfigReportOverridesOutputReference = DatasyncTaskTaskReportConfigReportOverridesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DatasyncTaskTaskReportConfigReportOverridesOutputReference[_e] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskTaskReportConfigReportOverridesOutputReference", version: "19.15.0" };
function datasyncTaskTaskReportConfigS3DestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_access_role_arn: cdktf.stringToTerraform(struct.bucketAccessRoleArn),
        s3_bucket_arn: cdktf.stringToTerraform(struct.s3BucketArn),
        subdirectory: cdktf.stringToTerraform(struct.subdirectory),
    };
}
exports.datasyncTaskTaskReportConfigS3DestinationToTerraform = datasyncTaskTaskReportConfigS3DestinationToTerraform;
function datasyncTaskTaskReportConfigS3DestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_access_role_arn: {
            value: cdktf.stringToHclTerraform(struct.bucketAccessRoleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_bucket_arn: {
            value: cdktf.stringToHclTerraform(struct.s3BucketArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subdirectory: {
            value: cdktf.stringToHclTerraform(struct.subdirectory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskTaskReportConfigS3DestinationToHclTerraform = datasyncTaskTaskReportConfigS3DestinationToHclTerraform;
class DatasyncTaskTaskReportConfigS3DestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketAccessRoleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketAccessRoleArn = this._bucketAccessRoleArn;
        }
        if (this._s3BucketArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3BucketArn = this._s3BucketArn;
        }
        if (this._subdirectory !== undefined) {
            hasAnyValues = true;
            internalValueResult.subdirectory = this._subdirectory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketAccessRoleArn = undefined;
            this._s3BucketArn = undefined;
            this._subdirectory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketAccessRoleArn = value.bucketAccessRoleArn;
            this._s3BucketArn = value.s3BucketArn;
            this._subdirectory = value.subdirectory;
        }
    }
    get bucketAccessRoleArn() {
        return this.getStringAttribute('bucket_access_role_arn');
    }
    set bucketAccessRoleArn(value) {
        this._bucketAccessRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketAccessRoleArnInput() {
        return this._bucketAccessRoleArn;
    }
    get s3BucketArn() {
        return this.getStringAttribute('s3_bucket_arn');
    }
    set s3BucketArn(value) {
        this._s3BucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketArnInput() {
        return this._s3BucketArn;
    }
    get subdirectory() {
        return this.getStringAttribute('subdirectory');
    }
    set subdirectory(value) {
        this._subdirectory = value;
    }
    resetSubdirectory() {
        this._subdirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subdirectoryInput() {
        return this._subdirectory;
    }
}
exports.DatasyncTaskTaskReportConfigS3DestinationOutputReference = DatasyncTaskTaskReportConfigS3DestinationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DatasyncTaskTaskReportConfigS3DestinationOutputReference[_f] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskTaskReportConfigS3DestinationOutputReference", version: "19.15.0" };
function datasyncTaskTaskReportConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        output_type: cdktf.stringToTerraform(struct.outputType),
        report_level: cdktf.stringToTerraform(struct.reportLevel),
        s3_object_versioning: cdktf.stringToTerraform(struct.s3ObjectVersioning),
        report_overrides: datasyncTaskTaskReportConfigReportOverridesToTerraform(struct.reportOverrides),
        s3_destination: datasyncTaskTaskReportConfigS3DestinationToTerraform(struct.s3Destination),
    };
}
exports.datasyncTaskTaskReportConfigToTerraform = datasyncTaskTaskReportConfigToTerraform;
function datasyncTaskTaskReportConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        output_type: {
            value: cdktf.stringToHclTerraform(struct.outputType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        report_level: {
            value: cdktf.stringToHclTerraform(struct.reportLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        s3_object_versioning: {
            value: cdktf.stringToHclTerraform(struct.s3ObjectVersioning),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        report_overrides: {
            value: datasyncTaskTaskReportConfigReportOverridesToHclTerraform(struct.reportOverrides),
            isBlock: true,
            type: "list",
            storageClassType: "DatasyncTaskTaskReportConfigReportOverridesList",
        },
        s3_destination: {
            value: datasyncTaskTaskReportConfigS3DestinationToHclTerraform(struct.s3Destination),
            isBlock: true,
            type: "list",
            storageClassType: "DatasyncTaskTaskReportConfigS3DestinationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskTaskReportConfigToHclTerraform = datasyncTaskTaskReportConfigToHclTerraform;
class DatasyncTaskTaskReportConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // report_overrides - computed: false, optional: true, required: false
        this._reportOverrides = new DatasyncTaskTaskReportConfigReportOverridesOutputReference(this, "report_overrides");
        // s3_destination - computed: false, optional: false, required: true
        this._s3Destination = new DatasyncTaskTaskReportConfigS3DestinationOutputReference(this, "s3_destination");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._outputType !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputType = this._outputType;
        }
        if (this._reportLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.reportLevel = this._reportLevel;
        }
        if (this._s3ObjectVersioning !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3ObjectVersioning = this._s3ObjectVersioning;
        }
        if (this._reportOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.reportOverrides = this._reportOverrides?.internalValue;
        }
        if (this._s3Destination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Destination = this._s3Destination?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._outputType = undefined;
            this._reportLevel = undefined;
            this._s3ObjectVersioning = undefined;
            this._reportOverrides.internalValue = undefined;
            this._s3Destination.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._outputType = value.outputType;
            this._reportLevel = value.reportLevel;
            this._s3ObjectVersioning = value.s3ObjectVersioning;
            this._reportOverrides.internalValue = value.reportOverrides;
            this._s3Destination.internalValue = value.s3Destination;
        }
    }
    get outputType() {
        return this.getStringAttribute('output_type');
    }
    set outputType(value) {
        this._outputType = value;
    }
    resetOutputType() {
        this._outputType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputTypeInput() {
        return this._outputType;
    }
    get reportLevel() {
        return this.getStringAttribute('report_level');
    }
    set reportLevel(value) {
        this._reportLevel = value;
    }
    resetReportLevel() {
        this._reportLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reportLevelInput() {
        return this._reportLevel;
    }
    get s3ObjectVersioning() {
        return this.getStringAttribute('s3_object_versioning');
    }
    set s3ObjectVersioning(value) {
        this._s3ObjectVersioning = value;
    }
    resetS3ObjectVersioning() {
        this._s3ObjectVersioning = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3ObjectVersioningInput() {
        return this._s3ObjectVersioning;
    }
    get reportOverrides() {
        return this._reportOverrides;
    }
    putReportOverrides(value) {
        this._reportOverrides.internalValue = value;
    }
    resetReportOverrides() {
        this._reportOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reportOverridesInput() {
        return this._reportOverrides.internalValue;
    }
    get s3Destination() {
        return this._s3Destination;
    }
    putS3Destination(value) {
        this._s3Destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3DestinationInput() {
        return this._s3Destination.internalValue;
    }
}
exports.DatasyncTaskTaskReportConfigOutputReference = DatasyncTaskTaskReportConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DatasyncTaskTaskReportConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskTaskReportConfigOutputReference", version: "19.15.0" };
function datasyncTaskTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.datasyncTaskTimeoutsToTerraform = datasyncTaskTimeoutsToTerraform;
function datasyncTaskTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.datasyncTaskTimeoutsToHclTerraform = datasyncTaskTimeoutsToHclTerraform;
class DatasyncTaskTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.DatasyncTaskTimeoutsOutputReference = DatasyncTaskTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DatasyncTaskTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTaskTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_task aws_datasync_task}
*/
class DatasyncTask extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DatasyncTask resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DatasyncTask to import
    * @param importFromId The id of the existing DatasyncTask that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_task#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DatasyncTask to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_datasync_task", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/datasync_task aws_datasync_task} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DatasyncTaskConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_task',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // excludes - computed: false, optional: true, required: false
        this._excludes = new DatasyncTaskExcludesOutputReference(this, "excludes");
        // includes - computed: false, optional: true, required: false
        this._includes = new DatasyncTaskIncludesOutputReference(this, "includes");
        // options - computed: false, optional: true, required: false
        this._options = new DatasyncTaskOptionsOutputReference(this, "options");
        // schedule - computed: false, optional: true, required: false
        this._schedule = new DatasyncTaskScheduleOutputReference(this, "schedule");
        // task_report_config - computed: false, optional: true, required: false
        this._taskReportConfig = new DatasyncTaskTaskReportConfigOutputReference(this, "task_report_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DatasyncTaskTimeoutsOutputReference(this, "timeouts");
        this._cloudwatchLogGroupArn = config.cloudwatchLogGroupArn;
        this._destinationLocationArn = config.destinationLocationArn;
        this._id = config.id;
        this._name = config.name;
        this._sourceLocationArn = config.sourceLocationArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._excludes.internalValue = config.excludes;
        this._includes.internalValue = config.includes;
        this._options.internalValue = config.options;
        this._schedule.internalValue = config.schedule;
        this._taskReportConfig.internalValue = config.taskReportConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get cloudwatchLogGroupArn() {
        return this.getStringAttribute('cloudwatch_log_group_arn');
    }
    set cloudwatchLogGroupArn(value) {
        this._cloudwatchLogGroupArn = value;
    }
    resetCloudwatchLogGroupArn() {
        this._cloudwatchLogGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogGroupArnInput() {
        return this._cloudwatchLogGroupArn;
    }
    get destinationLocationArn() {
        return this.getStringAttribute('destination_location_arn');
    }
    set destinationLocationArn(value) {
        this._destinationLocationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationLocationArnInput() {
        return this._destinationLocationArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceLocationArn() {
        return this.getStringAttribute('source_location_arn');
    }
    set sourceLocationArn(value) {
        this._sourceLocationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceLocationArnInput() {
        return this._sourceLocationArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get excludes() {
        return this._excludes;
    }
    putExcludes(value) {
        this._excludes.internalValue = value;
    }
    resetExcludes() {
        this._excludes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludesInput() {
        return this._excludes.internalValue;
    }
    get includes() {
        return this._includes;
    }
    putIncludes(value) {
        this._includes.internalValue = value;
    }
    resetIncludes() {
        this._includes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includesInput() {
        return this._includes.internalValue;
    }
    get options() {
        return this._options;
    }
    putOptions(value) {
        this._options.internalValue = value;
    }
    resetOptions() {
        this._options.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options.internalValue;
    }
    get schedule() {
        return this._schedule;
    }
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    resetSchedule() {
        this._schedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    get taskReportConfig() {
        return this._taskReportConfig;
    }
    putTaskReportConfig(value) {
        this._taskReportConfig.internalValue = value;
    }
    resetTaskReportConfig() {
        this._taskReportConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskReportConfigInput() {
        return this._taskReportConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cloudwatch_log_group_arn: cdktf.stringToTerraform(this._cloudwatchLogGroupArn),
            destination_location_arn: cdktf.stringToTerraform(this._destinationLocationArn),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            source_location_arn: cdktf.stringToTerraform(this._sourceLocationArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            excludes: datasyncTaskExcludesToTerraform(this._excludes.internalValue),
            includes: datasyncTaskIncludesToTerraform(this._includes.internalValue),
            options: datasyncTaskOptionsToTerraform(this._options.internalValue),
            schedule: datasyncTaskScheduleToTerraform(this._schedule.internalValue),
            task_report_config: datasyncTaskTaskReportConfigToTerraform(this._taskReportConfig.internalValue),
            timeouts: datasyncTaskTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cloudwatch_log_group_arn: {
                value: cdktf.stringToHclTerraform(this._cloudwatchLogGroupArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            destination_location_arn: {
                value: cdktf.stringToHclTerraform(this._destinationLocationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            source_location_arn: {
                value: cdktf.stringToHclTerraform(this._sourceLocationArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            tags_all: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tagsAll),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            excludes: {
                value: datasyncTaskExcludesToHclTerraform(this._excludes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncTaskExcludesList",
            },
            includes: {
                value: datasyncTaskIncludesToHclTerraform(this._includes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncTaskIncludesList",
            },
            options: {
                value: datasyncTaskOptionsToHclTerraform(this._options.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncTaskOptionsList",
            },
            schedule: {
                value: datasyncTaskScheduleToHclTerraform(this._schedule.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncTaskScheduleList",
            },
            task_report_config: {
                value: datasyncTaskTaskReportConfigToHclTerraform(this._taskReportConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DatasyncTaskTaskReportConfigList",
            },
            timeouts: {
                value: datasyncTaskTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DatasyncTaskTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DatasyncTask = DatasyncTask;
_j = JSII_RTTI_SYMBOL_1;
DatasyncTask[_j] = { fqn: "@cdktf/provider-aws.datasyncTask.DatasyncTask", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DatasyncTask.tfResourceType = "aws_datasync_task";
