// @cdktf/provider-aws/lib/cloudwatch-event-endpoint/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventEndpoint = exports.CloudwatchEventEndpointRoutingConfigOutputReference = exports.cloudwatchEventEndpointRoutingConfigToHclTerraform = exports.cloudwatchEventEndpointRoutingConfigToTerraform = exports.CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigToHclTerraform = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigToTerraform = exports.CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToHclTerraform = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToTerraform = exports.CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToHclTerraform = exports.cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToTerraform = exports.CloudwatchEventEndpointReplicationConfigOutputReference = exports.cloudwatchEventEndpointReplicationConfigToHclTerraform = exports.cloudwatchEventEndpointReplicationConfigToTerraform = exports.CloudwatchEventEndpointEventBusList = exports.CloudwatchEventEndpointEventBusOutputReference = exports.cloudwatchEventEndpointEventBusToHclTerraform = exports.cloudwatchEventEndpointEventBusToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudwatchEventEndpointEventBusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_bus_arn: cdktf.stringToTerraform(struct.eventBusArn),
    };
}
exports.cloudwatchEventEndpointEventBusToTerraform = cloudwatchEventEndpointEventBusToTerraform;
function cloudwatchEventEndpointEventBusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        event_bus_arn: {
            value: cdktf.stringToHclTerraform(struct.eventBusArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointEventBusToHclTerraform = cloudwatchEventEndpointEventBusToHclTerraform;
class CloudwatchEventEndpointEventBusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._eventBusArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventBusArn = this._eventBusArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._eventBusArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._eventBusArn = value.eventBusArn;
        }
    }
    get eventBusArn() {
        return this.getStringAttribute('event_bus_arn');
    }
    set eventBusArn(value) {
        this._eventBusArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventBusArnInput() {
        return this._eventBusArn;
    }
}
exports.CloudwatchEventEndpointEventBusOutputReference = CloudwatchEventEndpointEventBusOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointEventBusOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointEventBusOutputReference", version: "19.15.0" };
class CloudwatchEventEndpointEventBusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventEndpointEventBusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventEndpointEventBusList = CloudwatchEventEndpointEventBusList;
_b = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointEventBusList[_b] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointEventBusList", version: "19.15.0" };
function cloudwatchEventEndpointReplicationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        state: cdktf.stringToTerraform(struct.state),
    };
}
exports.cloudwatchEventEndpointReplicationConfigToTerraform = cloudwatchEventEndpointReplicationConfigToTerraform;
function cloudwatchEventEndpointReplicationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        state: {
            value: cdktf.stringToHclTerraform(struct.state),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointReplicationConfigToHclTerraform = cloudwatchEventEndpointReplicationConfigToHclTerraform;
class CloudwatchEventEndpointReplicationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._state = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._state = value.state;
        }
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
}
exports.CloudwatchEventEndpointReplicationConfigOutputReference = CloudwatchEventEndpointReplicationConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointReplicationConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointReplicationConfigOutputReference", version: "19.15.0" };
function cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        health_check: cdktf.stringToTerraform(struct.healthCheck),
    };
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToTerraform;
function cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        health_check: {
            value: cdktf.stringToHclTerraform(struct.healthCheck),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToHclTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToHclTerraform;
class CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._healthCheck = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._healthCheck = value.healthCheck;
        }
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
}
exports.CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference = CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference[_d] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference", version: "19.15.0" };
function cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        route: cdktf.stringToTerraform(struct.route),
    };
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToTerraform;
function cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        route: {
            value: cdktf.stringToHclTerraform(struct.route),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToHclTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToHclTerraform;
class CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._route !== undefined) {
            hasAnyValues = true;
            internalValueResult.route = this._route;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._route = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._route = value.route;
        }
    }
    get route() {
        return this.getStringAttribute('route');
    }
    set route(value) {
        this._route = value;
    }
    resetRoute() {
        this._route = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routeInput() {
        return this._route;
    }
}
exports.CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference = CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference", version: "19.15.0" };
function cloudwatchEventEndpointRoutingConfigFailoverConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        primary: cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToTerraform(struct.primary),
        secondary: cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToTerraform(struct.secondary),
    };
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigToTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigToTerraform;
function cloudwatchEventEndpointRoutingConfigFailoverConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        primary: {
            value: cloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryToHclTerraform(struct.primary),
            isBlock: true,
            type: "list",
            storageClassType: "CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryList",
        },
        secondary: {
            value: cloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryToHclTerraform(struct.secondary),
            isBlock: true,
            type: "list",
            storageClassType: "CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointRoutingConfigFailoverConfigToHclTerraform = cloudwatchEventEndpointRoutingConfigFailoverConfigToHclTerraform;
class CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // primary - computed: false, optional: false, required: true
        this._primary = new CloudwatchEventEndpointRoutingConfigFailoverConfigPrimaryOutputReference(this, "primary");
        // secondary - computed: false, optional: false, required: true
        this._secondary = new CloudwatchEventEndpointRoutingConfigFailoverConfigSecondaryOutputReference(this, "secondary");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._primary?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.primary = this._primary?.internalValue;
        }
        if (this._secondary?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secondary = this._secondary?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._primary.internalValue = undefined;
            this._secondary.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._primary.internalValue = value.primary;
            this._secondary.internalValue = value.secondary;
        }
    }
    get primary() {
        return this._primary;
    }
    putPrimary(value) {
        this._primary.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get primaryInput() {
        return this._primary.internalValue;
    }
    get secondary() {
        return this._secondary;
    }
    putSecondary(value) {
        this._secondary.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryInput() {
        return this._secondary.internalValue;
    }
}
exports.CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference = CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference", version: "19.15.0" };
function cloudwatchEventEndpointRoutingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failover_config: cloudwatchEventEndpointRoutingConfigFailoverConfigToTerraform(struct.failoverConfig),
    };
}
exports.cloudwatchEventEndpointRoutingConfigToTerraform = cloudwatchEventEndpointRoutingConfigToTerraform;
function cloudwatchEventEndpointRoutingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        failover_config: {
            value: cloudwatchEventEndpointRoutingConfigFailoverConfigToHclTerraform(struct.failoverConfig),
            isBlock: true,
            type: "list",
            storageClassType: "CloudwatchEventEndpointRoutingConfigFailoverConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudwatchEventEndpointRoutingConfigToHclTerraform = cloudwatchEventEndpointRoutingConfigToHclTerraform;
class CloudwatchEventEndpointRoutingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // failover_config - computed: false, optional: false, required: true
        this._failoverConfig = new CloudwatchEventEndpointRoutingConfigFailoverConfigOutputReference(this, "failover_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failoverConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.failoverConfig = this._failoverConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failoverConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failoverConfig.internalValue = value.failoverConfig;
        }
    }
    get failoverConfig() {
        return this._failoverConfig;
    }
    putFailoverConfig(value) {
        this._failoverConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get failoverConfigInput() {
        return this._failoverConfig.internalValue;
    }
}
exports.CloudwatchEventEndpointRoutingConfigOutputReference = CloudwatchEventEndpointRoutingConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpointRoutingConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpointRoutingConfigOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_endpoint aws_cloudwatch_event_endpoint}
*/
class CloudwatchEventEndpoint extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudwatchEventEndpoint resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudwatchEventEndpoint to import
    * @param importFromId The id of the existing CloudwatchEventEndpoint that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_endpoint#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudwatchEventEndpoint to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cloudwatch_event_endpoint", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/cloudwatch_event_endpoint aws_cloudwatch_event_endpoint} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudwatchEventEndpointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // event_bus - computed: false, optional: false, required: true
        this._eventBus = new CloudwatchEventEndpointEventBusList(this, "event_bus", false);
        // replication_config - computed: false, optional: true, required: false
        this._replicationConfig = new CloudwatchEventEndpointReplicationConfigOutputReference(this, "replication_config");
        // routing_config - computed: false, optional: false, required: true
        this._routingConfig = new CloudwatchEventEndpointRoutingConfigOutputReference(this, "routing_config");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._roleArn = config.roleArn;
        this._eventBus.internalValue = config.eventBus;
        this._replicationConfig.internalValue = config.replicationConfig;
        this._routingConfig.internalValue = config.routingConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // endpoint_url - computed: true, optional: false, required: false
    get endpointUrl() {
        return this.getStringAttribute('endpoint_url');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get eventBus() {
        return this._eventBus;
    }
    putEventBus(value) {
        this._eventBus.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventBusInput() {
        return this._eventBus.internalValue;
    }
    get replicationConfig() {
        return this._replicationConfig;
    }
    putReplicationConfig(value) {
        this._replicationConfig.internalValue = value;
    }
    resetReplicationConfig() {
        this._replicationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationConfigInput() {
        return this._replicationConfig.internalValue;
    }
    get routingConfig() {
        return this._routingConfig;
    }
    putRoutingConfig(value) {
        this._routingConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get routingConfigInput() {
        return this._routingConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            event_bus: cdktf.listMapper(cloudwatchEventEndpointEventBusToTerraform, true)(this._eventBus.internalValue),
            replication_config: cloudwatchEventEndpointReplicationConfigToTerraform(this._replicationConfig.internalValue),
            routing_config: cloudwatchEventEndpointRoutingConfigToTerraform(this._routingConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            role_arn: {
                value: cdktf.stringToHclTerraform(this._roleArn),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            event_bus: {
                value: cdktf.listMapperHcl(cloudwatchEventEndpointEventBusToHclTerraform, true)(this._eventBus.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventEndpointEventBusList",
            },
            replication_config: {
                value: cloudwatchEventEndpointReplicationConfigToHclTerraform(this._replicationConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventEndpointReplicationConfigList",
            },
            routing_config: {
                value: cloudwatchEventEndpointRoutingConfigToHclTerraform(this._routingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudwatchEventEndpointRoutingConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudwatchEventEndpoint = CloudwatchEventEndpoint;
_h = JSII_RTTI_SYMBOL_1;
CloudwatchEventEndpoint[_h] = { fqn: "@cdktf/provider-aws.cloudwatchEventEndpoint.CloudwatchEventEndpoint", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
CloudwatchEventEndpoint.tfResourceType = "aws_cloudwatch_event_endpoint";
