// @cdktf/provider-aws/lib/data-aws-codeguruprofiler-profiling-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCodeguruprofilerProfilingGroup = exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusList = exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference = exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusToHclTerraform = exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusToTerraform = exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList = exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference = exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToHclTerraform = exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToTerraform = exports.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList = exports.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference = exports.dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToHclTerraform = exports.dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToTerraform = dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToTerraform;
function dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToHclTerraform = dataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigToHclTerraform;
class DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // profiling_enabled - computed: true, optional: false, required: false
    get profilingEnabled() {
        return this.getBooleanAttribute('profiling_enabled');
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference = DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference", version: "19.15.0" };
class DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList = DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList", version: "19.15.0" };
function dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToTerraform = dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToTerraform;
function dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToHclTerraform = dataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileToHclTerraform;
class DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // period - computed: true, optional: false, required: false
    get period() {
        return this.getStringAttribute('period');
    }
    // start - computed: true, optional: false, required: false
    get start() {
        return this.getStringAttribute('start');
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference = DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference", version: "19.15.0" };
class DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList = DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList[_d] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList", version: "19.15.0" };
function dataAwsCodeguruprofilerProfilingGroupProfilingStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusToTerraform = dataAwsCodeguruprofilerProfilingGroupProfilingStatusToTerraform;
function dataAwsCodeguruprofilerProfilingGroupProfilingStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCodeguruprofilerProfilingGroupProfilingStatusToHclTerraform = dataAwsCodeguruprofilerProfilingGroupProfilingStatusToHclTerraform;
class DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // latest_aggregated_profile - computed: true, optional: false, required: false
        this._latestAggregatedProfile = new DataAwsCodeguruprofilerProfilingGroupProfilingStatusLatestAggregatedProfileList(this, "latest_aggregated_profile", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // latest_agent_orchestrated_at - computed: true, optional: false, required: false
    get latestAgentOrchestratedAt() {
        return this.getStringAttribute('latest_agent_orchestrated_at');
    }
    // latest_agent_profile_reported_at - computed: true, optional: false, required: false
    get latestAgentProfileReportedAt() {
        return this.getStringAttribute('latest_agent_profile_reported_at');
    }
    get latestAggregatedProfile() {
        return this._latestAggregatedProfile;
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference = DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference", version: "19.15.0" };
class DataAwsCodeguruprofilerProfilingGroupProfilingStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCodeguruprofilerProfilingGroupProfilingStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCodeguruprofilerProfilingGroupProfilingStatusList = DataAwsCodeguruprofilerProfilingGroupProfilingStatusList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroupProfilingStatusList[_f] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroupProfilingStatusList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codeguruprofiler_profiling_group aws_codeguruprofiler_profiling_group}
*/
class DataAwsCodeguruprofilerProfilingGroup extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsCodeguruprofilerProfilingGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsCodeguruprofilerProfilingGroup to import
    * @param importFromId The id of the existing DataAwsCodeguruprofilerProfilingGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codeguruprofiler_profiling_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsCodeguruprofilerProfilingGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_codeguruprofiler_profiling_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/codeguruprofiler_profiling_group aws_codeguruprofiler_profiling_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCodeguruprofilerProfilingGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codeguruprofiler_profiling_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // agent_orchestration_config - computed: true, optional: false, required: false
        this._agentOrchestrationConfig = new DataAwsCodeguruprofilerProfilingGroupAgentOrchestrationConfigList(this, "agent_orchestration_config", false);
        // profiling_status - computed: true, optional: false, required: false
        this._profilingStatus = new DataAwsCodeguruprofilerProfilingGroupProfilingStatusList(this, "profiling_status", false);
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
        this._name = config.name;
    }
    get agentOrchestrationConfig() {
        return this._agentOrchestrationConfig;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // compute_platform - computed: true, optional: false, required: false
    get computePlatform() {
        return this.getStringAttribute('compute_platform');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get profilingStatus() {
        return this._profilingStatus;
    }
    get tags() {
        return this._tags;
    }
    // updated_at - computed: true, optional: false, required: false
    get updatedAt() {
        return this.getStringAttribute('updated_at');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsCodeguruprofilerProfilingGroup = DataAwsCodeguruprofilerProfilingGroup;
_g = JSII_RTTI_SYMBOL_1;
DataAwsCodeguruprofilerProfilingGroup[_g] = { fqn: "@cdktf/provider-aws.dataAwsCodeguruprofilerProfilingGroup.DataAwsCodeguruprofilerProfilingGroup", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCodeguruprofilerProfilingGroup.tfResourceType = "aws_codeguruprofiler_profiling_group";
