// @cdktf/provider-aws/lib/kms-custom-key-store/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsCustomKeyStore = exports.KmsCustomKeyStoreTimeoutsOutputReference = exports.kmsCustomKeyStoreTimeoutsToHclTerraform = exports.kmsCustomKeyStoreTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kmsCustomKeyStoreTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.kmsCustomKeyStoreTimeoutsToTerraform = kmsCustomKeyStoreTimeoutsToTerraform;
function kmsCustomKeyStoreTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.kmsCustomKeyStoreTimeoutsToHclTerraform = kmsCustomKeyStoreTimeoutsToHclTerraform;
class KmsCustomKeyStoreTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.KmsCustomKeyStoreTimeoutsOutputReference = KmsCustomKeyStoreTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KmsCustomKeyStoreTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kmsCustomKeyStore.KmsCustomKeyStoreTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kms_custom_key_store aws_kms_custom_key_store}
*/
class KmsCustomKeyStore extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a KmsCustomKeyStore resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the KmsCustomKeyStore to import
    * @param importFromId The id of the existing KmsCustomKeyStore that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kms_custom_key_store#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the KmsCustomKeyStore to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_kms_custom_key_store", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/kms_custom_key_store aws_kms_custom_key_store} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KmsCustomKeyStoreConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_custom_key_store',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new KmsCustomKeyStoreTimeoutsOutputReference(this, "timeouts");
        this._cloudHsmClusterId = config.cloudHsmClusterId;
        this._customKeyStoreName = config.customKeyStoreName;
        this._id = config.id;
        this._keyStorePassword = config.keyStorePassword;
        this._trustAnchorCertificate = config.trustAnchorCertificate;
        this._timeouts.internalValue = config.timeouts;
    }
    get cloudHsmClusterId() {
        return this.getStringAttribute('cloud_hsm_cluster_id');
    }
    set cloudHsmClusterId(value) {
        this._cloudHsmClusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudHsmClusterIdInput() {
        return this._cloudHsmClusterId;
    }
    get customKeyStoreName() {
        return this.getStringAttribute('custom_key_store_name');
    }
    set customKeyStoreName(value) {
        this._customKeyStoreName = value;
    }
    // Temporarily expose input value. Use with caution.
    get customKeyStoreNameInput() {
        return this._customKeyStoreName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyStorePassword() {
        return this.getStringAttribute('key_store_password');
    }
    set keyStorePassword(value) {
        this._keyStorePassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyStorePasswordInput() {
        return this._keyStorePassword;
    }
    get trustAnchorCertificate() {
        return this.getStringAttribute('trust_anchor_certificate');
    }
    set trustAnchorCertificate(value) {
        this._trustAnchorCertificate = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustAnchorCertificateInput() {
        return this._trustAnchorCertificate;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cloud_hsm_cluster_id: cdktf.stringToTerraform(this._cloudHsmClusterId),
            custom_key_store_name: cdktf.stringToTerraform(this._customKeyStoreName),
            id: cdktf.stringToTerraform(this._id),
            key_store_password: cdktf.stringToTerraform(this._keyStorePassword),
            trust_anchor_certificate: cdktf.stringToTerraform(this._trustAnchorCertificate),
            timeouts: kmsCustomKeyStoreTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cloud_hsm_cluster_id: {
                value: cdktf.stringToHclTerraform(this._cloudHsmClusterId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            custom_key_store_name: {
                value: cdktf.stringToHclTerraform(this._customKeyStoreName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key_store_password: {
                value: cdktf.stringToHclTerraform(this._keyStorePassword),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_anchor_certificate: {
                value: cdktf.stringToHclTerraform(this._trustAnchorCertificate),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: kmsCustomKeyStoreTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "KmsCustomKeyStoreTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.KmsCustomKeyStore = KmsCustomKeyStore;
_b = JSII_RTTI_SYMBOL_1;
KmsCustomKeyStore[_b] = { fqn: "@cdktf/provider-aws.kmsCustomKeyStore.KmsCustomKeyStore", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
KmsCustomKeyStore.tfResourceType = "aws_kms_custom_key_store";
