// @winglang/sdk/lib/.gen/providers/random/id/index.js
"use strict";
// https://registry.terraform.io/providers/hashicorp/random/3.5.1/docs/resources/id
// generated from terraform resource schema
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Id = void 0;
const cdktf = __importStar(require("cdktf"));
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/random/3.5.1/docs/resources/id random_id}
*/
class Id extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Id resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Id to import
    * @param importFromId The id of the existing Id that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/random/3.5.1/docs/resources/id#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Id to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "random_id", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/random/3.5.1/docs/resources/id random_id} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IdConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'random_id',
            terraformGeneratorMetadata: {
                providerName: 'random',
                providerVersion: '3.5.1',
                providerVersionConstraint: '~>3.5.1'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._byteLength = config.byteLength;
        this._keepers = config.keepers;
        this._prefix = config.prefix;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // b64_std - computed: true, optional: false, required: false
    get b64Std() {
        return this.getStringAttribute('b64_std');
    }
    // b64_url - computed: true, optional: false, required: false
    get b64Url() {
        return this.getStringAttribute('b64_url');
    }
    get byteLength() {
        return this.getNumberAttribute('byte_length');
    }
    set byteLength(value) {
        this._byteLength = value;
    }
    // Temporarily expose input value. Use with caution.
    get byteLengthInput() {
        return this._byteLength;
    }
    // dec - computed: true, optional: false, required: false
    get dec() {
        return this.getStringAttribute('dec');
    }
    // hex - computed: true, optional: false, required: false
    get hex() {
        return this.getStringAttribute('hex');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get keepers() {
        return this.getStringMapAttribute('keepers');
    }
    set keepers(value) {
        this._keepers = value;
    }
    resetKeepers() {
        this._keepers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keepersInput() {
        return this._keepers;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            byte_length: cdktf.numberToTerraform(this._byteLength),
            keepers: cdktf.hashMapper(cdktf.stringToTerraform)(this._keepers),
            prefix: cdktf.stringToTerraform(this._prefix),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            byte_length: {
                value: cdktf.numberToHclTerraform(this._byteLength),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            keepers: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._keepers),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            prefix: {
                value: cdktf.stringToHclTerraform(this._prefix),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Id = Id;
// =================
// STATIC PROPERTIES
// =================
Id.tfResourceType = "random_id";
