// @cdktf/provider-aws/lib/inspector2-organization-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inspector2OrganizationConfiguration = exports.Inspector2OrganizationConfigurationTimeoutsOutputReference = exports.inspector2OrganizationConfigurationTimeoutsToHclTerraform = exports.inspector2OrganizationConfigurationTimeoutsToTerraform = exports.Inspector2OrganizationConfigurationAutoEnableOutputReference = exports.inspector2OrganizationConfigurationAutoEnableToHclTerraform = exports.inspector2OrganizationConfigurationAutoEnableToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function inspector2OrganizationConfigurationAutoEnableToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ec2: cdktf.booleanToTerraform(struct.ec2),
        ecr: cdktf.booleanToTerraform(struct.ecr),
        lambda: cdktf.booleanToTerraform(struct.lambda),
        lambda_code: cdktf.booleanToTerraform(struct.lambdaCode),
    };
}
exports.inspector2OrganizationConfigurationAutoEnableToTerraform = inspector2OrganizationConfigurationAutoEnableToTerraform;
function inspector2OrganizationConfigurationAutoEnableToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ec2: {
            value: cdktf.booleanToHclTerraform(struct.ec2),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        ecr: {
            value: cdktf.booleanToHclTerraform(struct.ecr),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        lambda: {
            value: cdktf.booleanToHclTerraform(struct.lambda),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        lambda_code: {
            value: cdktf.booleanToHclTerraform(struct.lambdaCode),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.inspector2OrganizationConfigurationAutoEnableToHclTerraform = inspector2OrganizationConfigurationAutoEnableToHclTerraform;
class Inspector2OrganizationConfigurationAutoEnableOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ec2 !== undefined) {
            hasAnyValues = true;
            internalValueResult.ec2 = this._ec2;
        }
        if (this._ecr !== undefined) {
            hasAnyValues = true;
            internalValueResult.ecr = this._ecr;
        }
        if (this._lambda !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambda = this._lambda;
        }
        if (this._lambdaCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambdaCode = this._lambdaCode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ec2 = undefined;
            this._ecr = undefined;
            this._lambda = undefined;
            this._lambdaCode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ec2 = value.ec2;
            this._ecr = value.ecr;
            this._lambda = value.lambda;
            this._lambdaCode = value.lambdaCode;
        }
    }
    get ec2() {
        return this.getBooleanAttribute('ec2');
    }
    set ec2(value) {
        this._ec2 = value;
    }
    // Temporarily expose input value. Use with caution.
    get ec2Input() {
        return this._ec2;
    }
    get ecr() {
        return this.getBooleanAttribute('ecr');
    }
    set ecr(value) {
        this._ecr = value;
    }
    // Temporarily expose input value. Use with caution.
    get ecrInput() {
        return this._ecr;
    }
    get lambda() {
        return this.getBooleanAttribute('lambda');
    }
    set lambda(value) {
        this._lambda = value;
    }
    resetLambda() {
        this._lambda = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaInput() {
        return this._lambda;
    }
    get lambdaCode() {
        return this.getBooleanAttribute('lambda_code');
    }
    set lambdaCode(value) {
        this._lambdaCode = value;
    }
    resetLambdaCode() {
        this._lambdaCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaCodeInput() {
        return this._lambdaCode;
    }
}
exports.Inspector2OrganizationConfigurationAutoEnableOutputReference = Inspector2OrganizationConfigurationAutoEnableOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfigurationAutoEnableOutputReference[_a] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfigurationAutoEnableOutputReference", version: "19.15.0" };
function inspector2OrganizationConfigurationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.inspector2OrganizationConfigurationTimeoutsToTerraform = inspector2OrganizationConfigurationTimeoutsToTerraform;
function inspector2OrganizationConfigurationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.inspector2OrganizationConfigurationTimeoutsToHclTerraform = inspector2OrganizationConfigurationTimeoutsToHclTerraform;
class Inspector2OrganizationConfigurationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Inspector2OrganizationConfigurationTimeoutsOutputReference = Inspector2OrganizationConfigurationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfigurationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfigurationTimeoutsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector2_organization_configuration aws_inspector2_organization_configuration}
*/
class Inspector2OrganizationConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a Inspector2OrganizationConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the Inspector2OrganizationConfiguration to import
    * @param importFromId The id of the existing Inspector2OrganizationConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector2_organization_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the Inspector2OrganizationConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_inspector2_organization_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/inspector2_organization_configuration aws_inspector2_organization_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Inspector2OrganizationConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_inspector2_organization_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auto_enable - computed: false, optional: false, required: true
        this._autoEnable = new Inspector2OrganizationConfigurationAutoEnableOutputReference(this, "auto_enable");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Inspector2OrganizationConfigurationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._autoEnable.internalValue = config.autoEnable;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // max_account_limit_reached - computed: true, optional: false, required: false
    get maxAccountLimitReached() {
        return this.getBooleanAttribute('max_account_limit_reached');
    }
    get autoEnable() {
        return this._autoEnable;
    }
    putAutoEnable(value) {
        this._autoEnable.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableInput() {
        return this._autoEnable.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            auto_enable: inspector2OrganizationConfigurationAutoEnableToTerraform(this._autoEnable.internalValue),
            timeouts: inspector2OrganizationConfigurationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            auto_enable: {
                value: inspector2OrganizationConfigurationAutoEnableToHclTerraform(this._autoEnable.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "Inspector2OrganizationConfigurationAutoEnableList",
            },
            timeouts: {
                value: inspector2OrganizationConfigurationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "Inspector2OrganizationConfigurationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.Inspector2OrganizationConfiguration = Inspector2OrganizationConfiguration;
_c = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfiguration[_c] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
Inspector2OrganizationConfiguration.tfResourceType = "aws_inspector2_organization_configuration";
