// @cdktf/provider-aws/lib/data-aws-dx-location/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDxLocation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_location aws_dx_location}
*/
class DataAwsDxLocation extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsDxLocation resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsDxLocation to import
    * @param importFromId The id of the existing DataAwsDxLocation that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_location#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsDxLocation to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_dx_location", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/dx_location aws_dx_location} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDxLocationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_location',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._locationCode = config.locationCode;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // available_macsec_port_speeds - computed: true, optional: false, required: false
    get availableMacsecPortSpeeds() {
        return this.getListAttribute('available_macsec_port_speeds');
    }
    // available_port_speeds - computed: true, optional: false, required: false
    get availablePortSpeeds() {
        return this.getListAttribute('available_port_speeds');
    }
    // available_providers - computed: true, optional: false, required: false
    get availableProviders() {
        return this.getListAttribute('available_providers');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get locationCode() {
        return this.getStringAttribute('location_code');
    }
    set locationCode(value) {
        this._locationCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationCodeInput() {
        return this._locationCode;
    }
    // location_name - computed: true, optional: false, required: false
    get locationName() {
        return this.getStringAttribute('location_name');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location_code: cdktf.stringToTerraform(this._locationCode),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location_code: {
                value: cdktf.stringToHclTerraform(this._locationCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsDxLocation = DataAwsDxLocation;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDxLocation[_a] = { fqn: "@cdktf/provider-aws.dataAwsDxLocation.DataAwsDxLocation", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDxLocation.tfResourceType = "aws_dx_location";
