// @cdktf/provider-aws/lib/data-aws-cognito-identity-pool/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCognitoIdentityPool = exports.DataAwsCognitoIdentityPoolCognitoIdentityProvidersList = exports.DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference = exports.dataAwsCognitoIdentityPoolCognitoIdentityProvidersToHclTerraform = exports.dataAwsCognitoIdentityPoolCognitoIdentityProvidersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsCognitoIdentityPoolCognitoIdentityProvidersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCognitoIdentityPoolCognitoIdentityProvidersToTerraform = dataAwsCognitoIdentityPoolCognitoIdentityProvidersToTerraform;
function dataAwsCognitoIdentityPoolCognitoIdentityProvidersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataAwsCognitoIdentityPoolCognitoIdentityProvidersToHclTerraform = dataAwsCognitoIdentityPoolCognitoIdentityProvidersToHclTerraform;
class DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // client_id - computed: true, optional: false, required: false
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    // provider_name - computed: true, optional: false, required: false
    get providerName() {
        return this.getStringAttribute('provider_name');
    }
    // server_side_token_check - computed: true, optional: false, required: false
    get serverSideTokenCheck() {
        return this.getBooleanAttribute('server_side_token_check');
    }
}
exports.DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference = DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsCognitoIdentityPool.DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference", version: "19.15.0" };
class DataAwsCognitoIdentityPoolCognitoIdentityProvidersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCognitoIdentityPoolCognitoIdentityProvidersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCognitoIdentityPoolCognitoIdentityProvidersList = DataAwsCognitoIdentityPoolCognitoIdentityProvidersList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCognitoIdentityPoolCognitoIdentityProvidersList[_b] = { fqn: "@cdktf/provider-aws.dataAwsCognitoIdentityPool.DataAwsCognitoIdentityPoolCognitoIdentityProvidersList", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/cognito_identity_pool aws_cognito_identity_pool}
*/
class DataAwsCognitoIdentityPool extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsCognitoIdentityPool resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsCognitoIdentityPool to import
    * @param importFromId The id of the existing DataAwsCognitoIdentityPool that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/cognito_identity_pool#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsCognitoIdentityPool to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_cognito_identity_pool", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/cognito_identity_pool aws_cognito_identity_pool} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCognitoIdentityPoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cognito_identity_pool',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cognito_identity_providers - computed: true, optional: false, required: false
        this._cognitoIdentityProviders = new DataAwsCognitoIdentityPoolCognitoIdentityProvidersList(this, "cognito_identity_providers", true);
        // supported_login_providers - computed: true, optional: false, required: false
        this._supportedLoginProviders = new cdktf.StringMap(this, "supported_login_providers");
        this._id = config.id;
        this._identityPoolName = config.identityPoolName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // allow_classic_flow - computed: true, optional: false, required: false
    get allowClassicFlow() {
        return this.getBooleanAttribute('allow_classic_flow');
    }
    // allow_unauthenticated_identities - computed: true, optional: false, required: false
    get allowUnauthenticatedIdentities() {
        return this.getBooleanAttribute('allow_unauthenticated_identities');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get cognitoIdentityProviders() {
        return this._cognitoIdentityProviders;
    }
    // developer_provider_name - computed: true, optional: false, required: false
    get developerProviderName() {
        return this.getStringAttribute('developer_provider_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityPoolName() {
        return this.getStringAttribute('identity_pool_name');
    }
    set identityPoolName(value) {
        this._identityPoolName = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityPoolNameInput() {
        return this._identityPoolName;
    }
    // openid_connect_provider_arns - computed: true, optional: false, required: false
    get openidConnectProviderArns() {
        return cdktf.Fn.tolist(this.getListAttribute('openid_connect_provider_arns'));
    }
    // saml_provider_arns - computed: true, optional: false, required: false
    get samlProviderArns() {
        return this.getListAttribute('saml_provider_arns');
    }
    get supportedLoginProviders() {
        return this._supportedLoginProviders;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            identity_pool_name: cdktf.stringToTerraform(this._identityPoolName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            identity_pool_name: {
                value: cdktf.stringToHclTerraform(this._identityPoolName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tags: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._tags),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsCognitoIdentityPool = DataAwsCognitoIdentityPool;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCognitoIdentityPool[_c] = { fqn: "@cdktf/provider-aws.dataAwsCognitoIdentityPool.DataAwsCognitoIdentityPool", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCognitoIdentityPool.tfResourceType = "aws_cognito_identity_pool";
