// @cdktf/provider-aws/lib/data-aws-redshiftserverless-credentials/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftserverlessCredentials = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshiftserverless_credentials aws_redshiftserverless_credentials}
*/
class DataAwsRedshiftserverlessCredentials extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataAwsRedshiftserverlessCredentials resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataAwsRedshiftserverlessCredentials to import
    * @param importFromId The id of the existing DataAwsRedshiftserverlessCredentials that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshiftserverless_credentials#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataAwsRedshiftserverlessCredentials to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_redshiftserverless_credentials", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/data-sources/redshiftserverless_credentials aws_redshiftserverless_credentials} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRedshiftserverlessCredentialsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_credentials',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._dbName = config.dbName;
        this._durationSeconds = config.durationSeconds;
        this._id = config.id;
        this._workgroupName = config.workgroupName;
    }
    get dbName() {
        return this.getStringAttribute('db_name');
    }
    set dbName(value) {
        this._dbName = value;
    }
    resetDbName() {
        this._dbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbNameInput() {
        return this._dbName;
    }
    // db_password - computed: true, optional: false, required: false
    get dbPassword() {
        return this.getStringAttribute('db_password');
    }
    // db_user - computed: true, optional: false, required: false
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    // expiration - computed: true, optional: false, required: false
    get expiration() {
        return this.getStringAttribute('expiration');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get workgroupName() {
        return this.getStringAttribute('workgroup_name');
    }
    set workgroupName(value) {
        this._workgroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get workgroupNameInput() {
        return this._workgroupName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_name: cdktf.stringToTerraform(this._dbName),
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
            id: cdktf.stringToTerraform(this._id),
            workgroup_name: cdktf.stringToTerraform(this._workgroupName),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            db_name: {
                value: cdktf.stringToHclTerraform(this._dbName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            duration_seconds: {
                value: cdktf.numberToHclTerraform(this._durationSeconds),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workgroup_name: {
                value: cdktf.stringToHclTerraform(this._workgroupName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataAwsRedshiftserverlessCredentials = DataAwsRedshiftserverlessCredentials;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftserverlessCredentials[_a] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftserverlessCredentials.DataAwsRedshiftserverlessCredentials", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRedshiftserverlessCredentials.tfResourceType = "aws_redshiftserverless_credentials";
