// @cdktf/provider-aws/lib/verifiedaccess-instance-logging-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifiedaccessInstanceLoggingConfiguration = exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsToHclTerraform = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsToTerraform = exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToHclTerraform = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToTerraform = exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToHclTerraform = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToTerraform = exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToHclTerraform = exports.verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_group: cdktf.stringToTerraform(struct.logGroup),
    };
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToTerraform;
function verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        log_group: {
            value: cdktf.stringToHclTerraform(struct.logGroup),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToHclTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToHclTerraform;
class VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.logGroup = this._logGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logGroup = value.logGroup;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    set logGroup(value) {
        this._logGroup = value;
    }
    resetLogGroup() {
        this._logGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupInput() {
        return this._logGroup;
    }
}
exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference = VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference[_a] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceLoggingConfiguration.VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference", version: "19.15.0" };
function verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream: cdktf.stringToTerraform(struct.deliveryStream),
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToTerraform;
function verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delivery_stream: {
            value: cdktf.stringToHclTerraform(struct.deliveryStream),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToHclTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToHclTerraform;
class VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deliveryStream !== undefined) {
            hasAnyValues = true;
            internalValueResult.deliveryStream = this._deliveryStream;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deliveryStream = undefined;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deliveryStream = value.deliveryStream;
            this._enabled = value.enabled;
        }
    }
    get deliveryStream() {
        return this.getStringAttribute('delivery_stream');
    }
    set deliveryStream(value) {
        this._deliveryStream = value;
    }
    resetDeliveryStream() {
        this._deliveryStream = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deliveryStreamInput() {
        return this._deliveryStream;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference = VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference[_b] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceLoggingConfiguration.VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference", version: "19.15.0" };
function verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        bucket_owner: cdktf.stringToTerraform(struct.bucketOwner),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToTerraform;
function verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_name: {
            value: cdktf.stringToHclTerraform(struct.bucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        bucket_owner: {
            value: cdktf.stringToHclTerraform(struct.bucketOwner),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToHclTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToHclTerraform;
class VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._bucketOwner !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketOwner = this._bucketOwner;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._bucketOwner = undefined;
            this._enabled = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._bucketOwner = value.bucketOwner;
            this._enabled = value.enabled;
            this._prefix = value.prefix;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    resetBucketName() {
        this._bucketName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
    get bucketOwner() {
        return this.getStringAttribute('bucket_owner');
    }
    set bucketOwner(value) {
        this._bucketOwner = value;
    }
    resetBucketOwner() {
        this._bucketOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketOwnerInput() {
        return this._bucketOwner;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
}
exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference = VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference;
_c = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference[_c] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceLoggingConfiguration.VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference", version: "19.15.0" };
function verifiedaccessInstanceLoggingConfigurationAccessLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        include_trust_context: cdktf.booleanToTerraform(struct.includeTrustContext),
        log_version: cdktf.stringToTerraform(struct.logVersion),
        cloudwatch_logs: verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToTerraform(struct.cloudwatchLogs),
        kinesis_data_firehose: verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToTerraform(struct.kinesisDataFirehose),
        s3: verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToTerraform(struct.s3),
    };
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsToTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsToTerraform;
function verifiedaccessInstanceLoggingConfigurationAccessLogsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        include_trust_context: {
            value: cdktf.booleanToHclTerraform(struct.includeTrustContext),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        log_version: {
            value: cdktf.stringToHclTerraform(struct.logVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        cloudwatch_logs: {
            value: verifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsToHclTerraform(struct.cloudwatchLogs),
            isBlock: true,
            type: "list",
            storageClassType: "VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsList",
        },
        kinesis_data_firehose: {
            value: verifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseToHclTerraform(struct.kinesisDataFirehose),
            isBlock: true,
            type: "list",
            storageClassType: "VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseList",
        },
        s3: {
            value: verifiedaccessInstanceLoggingConfigurationAccessLogsS3ToHclTerraform(struct.s3),
            isBlock: true,
            type: "list",
            storageClassType: "VerifiedaccessInstanceLoggingConfigurationAccessLogsS3List",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.verifiedaccessInstanceLoggingConfigurationAccessLogsToHclTerraform = verifiedaccessInstanceLoggingConfigurationAccessLogsToHclTerraform;
class VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cloudwatch_logs - computed: false, optional: true, required: false
        this._cloudwatchLogs = new VerifiedaccessInstanceLoggingConfigurationAccessLogsCloudwatchLogsOutputReference(this, "cloudwatch_logs");
        // kinesis_data_firehose - computed: false, optional: true, required: false
        this._kinesisDataFirehose = new VerifiedaccessInstanceLoggingConfigurationAccessLogsKinesisDataFirehoseOutputReference(this, "kinesis_data_firehose");
        // s3 - computed: false, optional: true, required: false
        this._s3 = new VerifiedaccessInstanceLoggingConfigurationAccessLogsS3OutputReference(this, "s3");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._includeTrustContext !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeTrustContext = this._includeTrustContext;
        }
        if (this._logVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.logVersion = this._logVersion;
        }
        if (this._cloudwatchLogs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogs = this._cloudwatchLogs?.internalValue;
        }
        if (this._kinesisDataFirehose?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisDataFirehose = this._kinesisDataFirehose?.internalValue;
        }
        if (this._s3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3 = this._s3?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._includeTrustContext = undefined;
            this._logVersion = undefined;
            this._cloudwatchLogs.internalValue = undefined;
            this._kinesisDataFirehose.internalValue = undefined;
            this._s3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._includeTrustContext = value.includeTrustContext;
            this._logVersion = value.logVersion;
            this._cloudwatchLogs.internalValue = value.cloudwatchLogs;
            this._kinesisDataFirehose.internalValue = value.kinesisDataFirehose;
            this._s3.internalValue = value.s3;
        }
    }
    get includeTrustContext() {
        return this.getBooleanAttribute('include_trust_context');
    }
    set includeTrustContext(value) {
        this._includeTrustContext = value;
    }
    resetIncludeTrustContext() {
        this._includeTrustContext = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeTrustContextInput() {
        return this._includeTrustContext;
    }
    get logVersion() {
        return this.getStringAttribute('log_version');
    }
    set logVersion(value) {
        this._logVersion = value;
    }
    resetLogVersion() {
        this._logVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logVersionInput() {
        return this._logVersion;
    }
    get cloudwatchLogs() {
        return this._cloudwatchLogs;
    }
    putCloudwatchLogs(value) {
        this._cloudwatchLogs.internalValue = value;
    }
    resetCloudwatchLogs() {
        this._cloudwatchLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogsInput() {
        return this._cloudwatchLogs.internalValue;
    }
    get kinesisDataFirehose() {
        return this._kinesisDataFirehose;
    }
    putKinesisDataFirehose(value) {
        this._kinesisDataFirehose.internalValue = value;
    }
    resetKinesisDataFirehose() {
        this._kinesisDataFirehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisDataFirehoseInput() {
        return this._kinesisDataFirehose.internalValue;
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
}
exports.VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference = VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference[_d] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceLoggingConfiguration.VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference", version: "19.15.0" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_logging_configuration aws_verifiedaccess_instance_logging_configuration}
*/
class VerifiedaccessInstanceLoggingConfiguration extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VerifiedaccessInstanceLoggingConfiguration resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VerifiedaccessInstanceLoggingConfiguration to import
    * @param importFromId The id of the existing VerifiedaccessInstanceLoggingConfiguration that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_logging_configuration#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VerifiedaccessInstanceLoggingConfiguration to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "aws_verifiedaccess_instance_logging_configuration", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/aws/5.46.0/docs/resources/verifiedaccess_instance_logging_configuration aws_verifiedaccess_instance_logging_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VerifiedaccessInstanceLoggingConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_verifiedaccess_instance_logging_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '5.46.0',
                providerVersionConstraint: '~> 5.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_logs - computed: false, optional: false, required: true
        this._accessLogs = new VerifiedaccessInstanceLoggingConfigurationAccessLogsOutputReference(this, "access_logs");
        this._id = config.id;
        this._verifiedaccessInstanceId = config.verifiedaccessInstanceId;
        this._accessLogs.internalValue = config.accessLogs;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get verifiedaccessInstanceId() {
        return this.getStringAttribute('verifiedaccess_instance_id');
    }
    set verifiedaccessInstanceId(value) {
        this._verifiedaccessInstanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get verifiedaccessInstanceIdInput() {
        return this._verifiedaccessInstanceId;
    }
    get accessLogs() {
        return this._accessLogs;
    }
    putAccessLogs(value) {
        this._accessLogs.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessLogsInput() {
        return this._accessLogs.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            verifiedaccess_instance_id: cdktf.stringToTerraform(this._verifiedaccessInstanceId),
            access_logs: verifiedaccessInstanceLoggingConfigurationAccessLogsToTerraform(this._accessLogs.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            verifiedaccess_instance_id: {
                value: cdktf.stringToHclTerraform(this._verifiedaccessInstanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_logs: {
                value: verifiedaccessInstanceLoggingConfigurationAccessLogsToHclTerraform(this._accessLogs.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VerifiedaccessInstanceLoggingConfigurationAccessLogsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VerifiedaccessInstanceLoggingConfiguration = VerifiedaccessInstanceLoggingConfiguration;
_e = JSII_RTTI_SYMBOL_1;
VerifiedaccessInstanceLoggingConfiguration[_e] = { fqn: "@cdktf/provider-aws.verifiedaccessInstanceLoggingConfiguration.VerifiedaccessInstanceLoggingConfiguration", version: "19.15.0" };
// =================
// STATIC PROPERTIES
// =================
VerifiedaccessInstanceLoggingConfiguration.tfResourceType = "aws_verifiedaccess_instance_logging_configuration";
